::: {#pst-skip-link .skip-link .d-print-none}
[Skip to main content](#main-content)
:::

::: {#pst-scroll-pixel-helper}
:::

Back to top

::: search-button__wrapper
::: search-button__overlay
:::

::: search-button__search-container
[[Ctrl]{.kbd
.kbd-shortcut__modifier}+[K]{.kbd}]{.search-button__kbd-shortcut}
:::
:::

::: {.pst-async-banner-revealer .d-none}
:::

::: {.bd-header .navbar .navbar-expand-lg .bd-navbar .d-print-none}
:::

::: bd-container
::: {.bd-container__inner .bd-page-width}
::: {.bd-sidebar-primary .bd-sidebar}
::: {.sidebar-header-items .sidebar-primary__section}
:::

::: {.sidebar-primary-items__start .sidebar-primary__section}
::: sidebar-primary-item
[![None - Home](_static/logo.png){.logo__image
.only-light}](#){.navbar-brand .logo}
:::

::: sidebar-primary-item
:::

::: sidebar-primary-item
::: {.bd-toc-item .navbar-nav .active}
[User documentation]{.caption-text}

[NeuroML events]{.caption-text}

[The NeuroML Initiative]{.caption-text}

[Developer documentation]{.caption-text}

[Reference]{.caption-text}
:::
:::
:::

::: {.sidebar-primary-items__end .sidebar-primary__section}
:::

::: {#rtd-footer-container}
:::
:::

::: {#main-content .bd-main role="main"}
::: sbt-scroll-pixel-helper
:::

::: bd-content
::: bd-article-container
::: {.bd-header-article .d-print-none}
::: {.header-article-items .header-article__inner}
::: header-article-items__start
::: header-article-item
[]{.fa-solid .fa-bars}
:::
:::

::: header-article-items__end
::: header-article-item
::: article-header-buttons
::: {.dropdown .dropdown-source-buttons}

-   [[ ]{.btn__icon-container}
    [Repository]{.btn__text-container}](https://github.com/NeuroML/Documentation "Source repository"){.btn
    .btn-sm .btn-source-repository-button .dropdown-item target="_blank"
    bs-placement="left" bs-toggle="tooltip"}
-   [[ ]{.btn__icon-container} [Open
    issue]{.btn__text-container}](https://github.com/NeuroML/Documentation/issues/new?title=Issue%20on%20page%20%2FLanding.html&body=Your%20issue%20content%20here. "Open an issue"){.btn
    .btn-sm .btn-source-issues-button .dropdown-item target="_blank"
    bs-placement="left" bs-toggle="tooltip"}
:::

[ ]{.btn__icon-container}

[]{.fa-solid .fa-list}
:::
:::
:::
:::
:::

::: {#jb-print-docs-body .onlyprint}
# NeuroML

::: {#print-main-content}
::: {#jb-print-toc}
<div>

## Contents

</div>

[User documentation]{.caption-text}

-   [Mission and Aims](#document-Userdocs/MissionAndAims){.reference
    .internal .nav-link}
-   [How to use this documentation](#document-Userdocs/Usage){.reference
    .internal .nav-link}
-   [Get NeuroML](#document-Userdocs/GetNeuroML){.reference .internal
    .nav-link}
-   [Getting started with
    NeuroML](#document-Userdocs/GettingStarted){.reference .internal
    .nav-link}
    -   [Simulating a regular spiking Izhikevich
        neuron](#document-Userdocs/SingleNeuronExample){.reference
        .internal .nav-link}
    -   [Interactive single Izhikevich neuron NeuroML
        example](#document-Userdocs/NML2_examples/SingleNeuron){.reference
        .internal .nav-link}
    -   [A two population network of regular spiking Izhikevich
        neurons](#document-Userdocs/IzhikevichNetworkExample){.reference
        .internal .nav-link}
    -   [Interactive two population network
        example](#document-Userdocs/NML2_examples/IzhikevichNetwork){.reference
        .internal .nav-link}
    -   [Simulating a single compartment Hodgkin-Huxley
        neuron](#document-Userdocs/SingleCompartmentHHExample){.reference
        .internal .nav-link}
    -   [Interactive single compartment HH
        example](#document-Userdocs/NML2_examples/HH_single_compartment){.reference
        .internal .nav-link}
    -   [Simulating a multi compartment OLM
        neuron](#document-Userdocs/MultiCompartmentOLMexample){.reference
        .internal .nav-link}
    -   [Interactive multi-compartment OLM cell
        example](#document-Userdocs/NML2_examples/OLM){.reference
        .internal .nav-link}
    -   [Create novel NeuroML models from components on
        NeuroML-DB](#document-Userdocs/NML2_examples/NeuroML-DB){.reference
        .internal .nav-link}
-   [Finding and sharing NeuroML
    models](#document-Userdocs/FindingNeuroMLModels){.reference
    .internal .nav-link}
-   [Creating NeuroML
    models](#document-Userdocs/CreatingNeuroMLModels){.reference
    .internal .nav-link}
    -   [Converting models to NeuroML and sharing them on Open Source
        Brain](#document-Userdocs/ConvertingModels){.reference .internal
        .nav-link}
    -   [Handling Morphology
        Files](#document-Userdocs/ImportingMorphologyFiles){.reference
        .internal .nav-link}
    -   [HDF5 support](#document-Userdocs/HDF5){.reference .internal
        .nav-link}
    -   [Maintaining provenance in NeuroML
        models](#document-Userdocs/Provenance){.reference .internal
        .nav-link}
-   [Validating NeuroML
    Models](#document-Userdocs/ValidatingNeuroMLModels){.reference
    .internal .nav-link}
-   [Visualising NeuroML
    Models](#document-Userdocs/VisualisingNeuroMLModels){.reference
    .internal .nav-link}
    -   [Visualising and analysing ion channel
        models](#document-Userdocs/VisualisingChannels){.reference
        .internal .nav-link}
    -   [Visualising and analysing cell
        models](#document-Userdocs/VisualisingCells){.reference
        .internal .nav-link}
-   [Simulating NeuroML
    Models](#document-Userdocs/SimulatingNeuroMLModels){.reference
    .internal .nav-link}
-   [Optimising/fitting NeuroML
    Models](#document-Userdocs/OptimisingNeuroMLModels){.reference
    .internal .nav-link}
-   [Testing/validating NeuroML
    Models](#document-Userdocs/TestingNeuroMLModels){.reference
    .internal .nav-link}
-   [LEMS: Low Entropy Model
    Specification](#document-Userdocs/LEMS){.reference .internal
    .nav-link}
    -   [Model structure
        overview](#document-Userdocs/LEMSOverview){.reference .internal
        .nav-link}
    -   [Example 1: Dimensions, Units, ComponentTypes and
        Components](#document-Userdocs/LEMSExample1){.reference
        .internal .nav-link}
    -   [Example 2: tidying up example
        1](#document-Userdocs/LEMSExample2){.reference .internal
        .nav-link}
    -   [Main model](Landing.html#main-model){.reference .internal
        .nav-link}
    -   [Included files](Landing.html#included-files){.reference
        .internal .nav-link}
    -   [Example 3: Connection dependent synaptic
        components](#document-Userdocs/LEMSExample3){.reference
        .internal .nav-link}
    -   [Example 4: Kinetic
        schemes](#document-Userdocs/LEMSExample4){.reference .internal
        .nav-link}
    -   [Example 5: References and
        paths](#document-Userdocs/LEMSExample5){.reference .internal
        .nav-link}
    -   [Example 6: User defined types for simulation and
        display](#document-Userdocs/LEMSExample6){.reference .internal
        .nav-link}
    -   [Example 7: User defined types for networks and
        populations](#document-Userdocs/LEMSExample7){.reference
        .internal .nav-link}
    -   [Example 8: Regimes in Dynamics
        definitions](#document-Userdocs/LEMSExample8){.reference
        .internal .nav-link}
-   [Schema/Specification](#document-Userdocs/Specification){.reference
    .internal .nav-link}
    -   [NeuroML v2](#document-Userdocs/NeuroMLv2){.reference .internal
        .nav-link}
        -   [NeuroMLDocument](#document-Userdocs/Schemas/NeuroMLDocument){.reference
            .internal .nav-link}
        -   [NeuroMLCoreDimensions](#document-Userdocs/Schemas/NeuroMLCoreDimensions){.reference
            .internal .nav-link}
        -   [NeuroMLCoreCompTypes](#document-Userdocs/Schemas/NeuroMLCoreCompTypes){.reference
            .internal .nav-link}
        -   [Cells](#document-Userdocs/Schemas/Cells){.reference
            .internal .nav-link}
        -   [Channels](#document-Userdocs/Schemas/Channels){.reference
            .internal .nav-link}
        -   [Synapses](#document-Userdocs/Schemas/Synapses){.reference
            .internal .nav-link}
        -   [Inputs](#document-Userdocs/Schemas/Inputs){.reference
            .internal .nav-link}
        -   [Networks](#document-Userdocs/Schemas/Networks){.reference
            .internal .nav-link}
        -   [PyNN](#document-Userdocs/Schemas/PyNN){.reference .internal
            .nav-link}
        -   [Simulation](#document-Userdocs/Schemas/Simulation){.reference
            .internal .nav-link}
        -   [Index](#document-Userdocs/Schemas/Index){.reference
            .internal .nav-link}
    -   [NeuroML v1](#document-Userdocs/NeuroMLv1){.reference .internal
        .nav-link}
    -   [LEMS](#document-Userdocs/LEMSSchema){.reference .internal
        .nav-link}
        -   [Model
            structure](#document-Userdocs/LEMS_elements/Modelstructure){.reference
            .internal .nav-link}
        -   [Units and
            dimensions](#document-Userdocs/LEMS_elements/Unitsanddimensions){.reference
            .internal .nav-link}
        -   [Defining component
            types](#document-Userdocs/LEMS_elements/Definingcomponenttypes){.reference
            .internal .nav-link}
        -   [Dynamics](#document-Userdocs/LEMS_elements/Dynamics){.reference
            .internal .nav-link}
        -   [Structure](#document-Userdocs/LEMS_elements/Structure){.reference
            .internal .nav-link}
        -   [Simulation](#document-Userdocs/LEMS_elements/Simulation){.reference
            .internal .nav-link}
        -   [Procedure](#document-Userdocs/LEMS_elements/Procedure){.reference
            .internal .nav-link}
        -   [Defining
            Components](#document-Userdocs/LEMS_elements/DefiningComponents){.reference
            .internal .nav-link}
        -   [Geometry](#document-Userdocs/LEMS_elements/Geometry){.reference
            .internal .nav-link}
-   [NeuroML 2 and LEMS](#document-Userdocs/NeuroMLv2AndLEMS){.reference
    .internal .nav-link}
    -   [Conventions](#document-Userdocs/Conventions){.reference
        .internal .nav-link}
    -   [Units and
        dimensions](#document-Userdocs/UnitsAndDimensions){.reference
        .internal .nav-link}
    -   [Paths](#document-Userdocs/Paths){.reference .internal
        .nav-link}
    -   [Quantities and
        recording](#document-Userdocs/QuantitiesAndRecording){.reference
        .internal .nav-link}
    -   [LEMS Simulation
        files](#document-Userdocs/LEMSSimulation){.reference .internal
        .nav-link}
-   [Extending
    NeuroML](#document-Userdocs/ExtendingNeuroMLv2){.reference .internal
    .nav-link}
-   [Software and
    Tools](#document-Userdocs/Software/Software){.reference .internal
    .nav-link}
    -   [pyNeuroML](#document-Userdocs/Software/pyNeuroML){.reference
        .internal .nav-link}
    -   [libNeuroML](#document-Userdocs/Software/libNeuroML){.reference
        .internal .nav-link}
    -   [pyLEMS](#document-Userdocs/Software/pyLEMS){.reference
        .internal .nav-link}
    -   [NeuroMLlite](#document-Userdocs/Software/NeuroMLlite){.reference
        .internal .nav-link}
    -   [jNeuroML](#document-Userdocs/Software/jNeuroML){.reference
        .internal .nav-link}
    -   [jLEMS](#document-Userdocs/Software/jLEMS){.reference .internal
        .nav-link}
    -   [NeuroML C++
        API](#document-Userdocs/Software/NeuroML_API){.reference
        .internal .nav-link}
    -   [MatLab NeuroML
        Toolbox](#document-Userdocs/Software/MatLab){.reference
        .internal .nav-link}
    -   [Tools and resources with NeuroML
        support](#document-Userdocs/Software/SupportingTools){.reference
        .internal .nav-link}
        -   [Approaches to adding NeuroML
            support](#document-Userdocs/Software/Tools/Approaches){.reference
            .internal .nav-link}
        -   [NEURON and
            NeuroML](#document-Userdocs/Software/Tools/NEURON){.reference
            .internal .nav-link}
        -   [NetPyNE and
            NeuroML](#document-Userdocs/Software/Tools/NetPyNE){.reference
            .internal .nav-link}
        -   [PyNN and
            NeuroML](#document-Userdocs/Software/Tools/PyNN){.reference
            .internal .nav-link}
        -   [Brian and
            NeuroML](#document-Userdocs/Software/Tools/Brian){.reference
            .internal .nav-link}
        -   [MOOSE and
            NeuroML](#document-Userdocs/Software/Tools/MOOSE){.reference
            .internal .nav-link}
        -   [EDEN and
            NeuroML](#document-Userdocs/Software/Tools/EDEN){.reference
            .internal .nav-link}
        -   [Arbor and
            NeuroML](#document-Userdocs/Software/Tools/Arbor){.reference
            .internal .nav-link}
        -   [N2A and
            NeuroML](#document-Userdocs/Software/Tools/N2A){.reference
            .internal .nav-link}
        -   [NEST and
            NeuroML](#document-Userdocs/Software/Tools/NEST){.reference
            .internal .nav-link}
        -   [SWC and
            NeuroML](#document-Userdocs/Software/Tools/SWC){.reference
            .internal .nav-link}
-   [Citing NeuroML and related
    publications](#document-Userdocs/Publications){.reference .internal
    .nav-link}
-   [Frequently asked questions
    (FAQ)](#document-Userdocs/FAQ){.reference .internal .nav-link}
-   [Walk
    throughs](#document-Userdocs/Walkthroughs/Walkthroughs){.reference
    .internal .nav-link}
    -   [Converting Ray et al 2020 to
        NeuroML](#document-Userdocs/Walkthroughs/RayEtAl2020/RayEtAl2020){.reference
        .internal .nav-link}
        -   [Setting
            up](#document-Userdocs/Walkthroughs/RayEtAl2020/Setup){.reference
            .internal .nav-link}
        -   [Converting to
            NeuroML](#document-Userdocs/Walkthroughs/RayEtAl2020/Conversion){.reference
            .internal .nav-link}
        -   [Adding OMV
            tests](#document-Userdocs/Walkthroughs/RayEtAl2020/OMV){.reference
            .internal .nav-link}

[NeuroML events]{.caption-text}

-   [NeuroML outreach and events](#document-Events/Events){.reference
    .internal .nav-link}
-   [July 2024: NeuroML tutorial at CNS
    2024](#document-Events/202407-CNS2024){.reference .internal
    .nav-link}
-   [April 2024: NeuroML hackathon at HARMONY
    2024](#document-Events/202404-Harmony){.reference .internal
    .nav-link}
-   [June 2022: NeuroML tutorial at CNS\*2022 satellite
    tutorials](#document-Events/20220630-CNS2022){.reference .internal
    .nav-link}
-   [April 2022: NeuroML development workshop at HARMONY
    2022](#document-Events/202204-Harmony){.reference .internal
    .nav-link}
-   [October 2021: NeuroML development workshop at COMBINE
    meeting](#document-Events/202109-COMBINE){.reference .internal
    .nav-link}
-   [August 2021: NeuroML tutorial at INCF Training
    Weeks](#document-Events/202108-INCF-Training-Week){.reference
    .internal .nav-link}
-   [July 2021: NeuroML tutorial at
    CNS\*2021](#document-Events/202107-CNS2021){.reference .internal
    .nav-link}
-   [March 2021: NeuroML hackathon at HARMONY
    2021](#document-Events/202103-Harmony){.reference .internal
    .nav-link}
-   [March 2012: Fourth NeuroML Development
    Workshop](#document-Events/2012-Edinburgh){.reference .internal
    .nav-link}
-   [Past NeuroML Events](#document-Events/PastEvents){.reference
    .internal .nav-link}

[The NeuroML Initiative]{.caption-text}

-   [Getting in
    touch](#document-NeuroMLOrg/CommunicationChannels){.reference
    .internal .nav-link}
-   [Overview of standards in
    neuroscience](#document-NeuroMLOrg/Standards){.reference .internal
    .nav-link}
-   [A brief history of
    NeuroML](#document-NeuroMLOrg/History){.reference .internal
    .nav-link}
-   [NeuroML Editorial Board](#document-NeuroMLOrg/Board){.reference
    .internal .nav-link}
    -   [History of the NeuroML Editorial
        Board](#document-NeuroMLOrg/BoardHistory){.reference .internal
        .nav-link}
    -   [Workshop and Meeting
        reports](#document-NeuroMLOrg/BoardMeetingReports){.reference
        .internal .nav-link}
-   [NeuroML Scientific
    Committee](#document-NeuroMLOrg/ScientificCommittee){.reference
    .internal .nav-link}
-   [Funding and
    Acknowledgements](#document-NeuroMLOrg/Funding){.reference .internal
    .nav-link}
-   [Outreach and
    training](#document-NeuroMLOrg/OutreachTraining){.reference
    .internal .nav-link}
-   [NeuroML contributors](#document-NeuroMLOrg/Contributors){.reference
    .internal .nav-link}
-   [NeuroML repositories](#document-NeuroMLOrg/Repositories){.reference
    .internal .nav-link}
-   [Code of Conduct](#document-NeuroMLOrg/CoC){.reference .internal
    .nav-link}

[Developer documentation]{.caption-text}

-   [Overview](#document-Devdocs/Devdocs){.reference .internal
    .nav-link}
    -   [Contribution guidelines](#document-Devdocs/DevSOP){.reference
        .internal .nav-link}
    -   [Release Process](#document-Devdocs/ReleaseProcess){.reference
        .internal .nav-link}
    -   [Making changes to the NeuroML
        standard](#document-Devdocs/UpdatingStandard){.reference
        .internal .nav-link}
-   [Interaction with other languages and
    standards](#document-Devdocs/InteractionOtherBits){.reference
    .internal .nav-link}

[Reference]{.caption-text}

-   [Glossary](#document-Reference/Glossary){.reference .internal
    .nav-link}
-   [Bibliography](#document-Reference/zBibliography){.reference
    .internal .nav-link}
:::
:::
:::

::: {#searchbox}
:::

::: {#neuroml .section .tex2jax_ignore .mathjax_ignore}
[]{#landing}

# NeuroML[\#](#neuroml "Link to this heading"){.headerlink}

*A model description language for computational neuroscience.*

::: {.attention .admonition}
Read the latest NeuroML paper in eLife:

[The NeuroML ecosystem for standardized multi-scale modeling in
neuroscience](https://doi.org/10.7554/eLife.95135.3){.reference
.external}
([\[[SGM+25](Landing.html#id22 "Ankur Sinha, Padraig Gleeson, Bóris Marin, Salvador Dura-Bernal, Sotirios Panagiotou, Sharon Crook, Matteo Cantarelli, Robert C. Cannon, Andrew P. Davison, Harsha Gurnani, and Robin Angus Silver. The neuroml ecosystem for standardized multi-scale modeling in neuroscience. eLife, 13:RP95135, 2025. URL: https://doi.org/10.7554/eLife.95135, doi:10.7554/eLife.95135."){.reference
.internal}\]]{#id1})
:::

::: {#carousel-neuroml .carousel .slide ride="carousel"}
1.  
2.  
3.  

::: carousel-inner
::: {.carousel-item .active}
[![NeuroML is endorsed by the INCF.](_images/endorsed.png){.d-block
.w-100}](https://www.incf.org/neuroml){.reference .external
.image-reference}
:::

::: carousel-item
[![NeuroML is a COMBINE standard.](_images/combine.png){.d-block
.w-100}](http://co.mbine.org/standards/neuroml){.reference .external
.image-reference}
:::

::: carousel-item
[![Explore models using NeuroML on the Open Source Brain
platform.](_images/osbnivo_mod2.png){.d-block
.w-100}](http://www.opensourcebrain.org/projects){.reference .external
.image-reference}
:::

::: carousel-item
[![Featured tool - MOOSE](_images/moose_mod.png){.d-block
.w-100}](http://moose.sourceforge.net/){.reference .external
.image-reference}
:::

::: carousel-item
[![Explore the Open Worm project.](_images/openworm2-mod.png){.d-block
.w-100}](http://www.openworm.org/){.reference .external
.image-reference}
:::
:::

[[]{.carousel-control-prev-icon aria-hidden="true"}
[Previous]{.sr-only}](#carousel-neuroml){.carousel-control-prev
role="button" slide="prev"} [[]{.carousel-control-next-icon
aria-hidden="true"}
[Next]{.sr-only}](#carousel-neuroml){.carousel-control-next
role="button" slide="next"}
:::

\

NeuroML is an [[international, collaborative initiative]{.std
.std-ref}](Landing.html#missions-and-aims){.reference .internal} to
develop a language for describing detailed models of neural systems,
which will serve as a standard data format for defining and exchanging
descriptions of neuronal cell and network models. NeuroML is:

-   modular

-   standardised

-   structured

and this allows you to:

-   easily [[build]{.std
    .std-ref}](Landing.html#userdocs-creating-models){.reference
    .internal} and [[optimise]{.std
    .std-ref}](Landing.html#userdocs-optimising){.reference .internal}
    detailed models of neural systems

-   easily [[validate]{.std
    .std-ref}](Landing.html#userdocs-validating-models){.reference
    .internal} your models

-   easily [[visualise]{.std
    .std-ref}](Landing.html#userdocs-visualising-models){.reference
    .internal} your models

-   easily [[simulate]{.std
    .std-ref}](Landing.html#userdocs-simulators){.reference .internal}
    your models using a variety of simulators

-   easily [[analyse]{.std
    .std-ref}](Landing.html#userdocs-software){.reference .internal}
    your simulations

all using a [[well supported set of tools]{.std
.std-ref}](Landing.html#userdocs-software){.reference .internal} in the
powerful [Python](https://www.python.org){.reference .external}
programming language.

In this documentation, you will find information on [[using
NeuroML]{.std
.std-ref}](Landing.html#userdocs-getting-started-neuroml){.reference
.internal}, [[developing with NeuroML]{.std
.std-ref}](Landing.html#devdocs-overview){.reference .internal}, its
[[specification]{.std
.std-ref}](Landing.html#userdocs-specification){.reference .internal},
and the [[community]{.std
.std-ref}](Landing.html#neuromlorg-board){.reference .internal} that
maintains it.

For any queries, please contact the NeuroML community using any of our
[[communication channels]{.std
.std-ref}](Landing.html#contact){.reference .internal}.

::: {.toctree-wrapper .compound}
[]{#document-Userdocs/MissionAndAims}

::: {#mission-and-aims .section .tex2jax_ignore .mathjax_ignore}
[]{#missions-and-aims}

## Mission and Aims[\#](#mission-and-aims "Link to this heading"){.headerlink}

Computational models, based on detailed neuroanatomical and
electrophysiological data, are heavily used as an aid for understanding
the nervous system. NeuroML is an international, collaborative
initiative to develop a language for describing detailed models of
neural systems, which will serve as a standard data format for defining
and exchanging descriptions of neuronal cell and network models.

NeuroML specifications are developed by the [[NeuroML Editorial
Board]{.std .std-ref}](Landing.html#neuromlorg-board){.reference
.internal} and overseen by its [[Scientific Committee]{.std
.std-ref}](Landing.html#neuromlorg-scientificcommittee){.reference
.internal}. NeuroML is [endorsed by the
INCF](https://www.incf.org/sbp/neuroml){.reference .external}, and is
also an official [COMBINE
standard](http://co.mbine.org/standards/neuroml){.reference .external}.

The NeuroML project community develops an [XML (eXtensible Markup
Language)](https://wikipedia.org/XML){.reference .external} based
description language where [XML
Schemas](https://www.w3schools.com/xml/schema_intro.asp){.reference
.external} are used to define model specifications. The community also
develops and maintains a number of libraries (in [[Python, Java and
other languages]{.std
.std-ref}](Landing.html#userdocs-software){.reference .internal}) to
facilitate use of these specifications.

The **aims of the NeuroML initiative** are:

-   To create specifications for an XML-based language that describes
    the biophysics, anatomy and network architecture of neuronal systems
    at multiple scales

-   To facilitate the exchange of complex neuronal models between
    researchers, allowing for greater transparency and accessibility of
    models

-   To promote software tools which support NeuroML and support the
    development of new software and databases for neural modeling

-   To encourage researchers with models within the scope of NeuroML to
    exchange and publish their models in this format
:::

[]{#document-Userdocs/Usage}

::: {#how-to-use-this-documentation .section .tex2jax_ignore .mathjax_ignore}
[]{#userdocs-usage}

## How to use this documentation[\#](#how-to-use-this-documentation "Link to this heading"){.headerlink}

This documentation is generated using [Jupyter
books](https://jupyterbook.org/intro.html){.reference .external}. You
can learn more about the project on their website.

::: {#structure-and-navigation .section}
[]{#userdocs-usage-structure}

### Structure and navigation[\#](#structure-and-navigation "Link to this heading"){.headerlink}

::: {.tip .admonition}
Close the left hand side bar using the burger menu on the left side of
the top panel.

You can close the left hand side bar by clicking the burger menu on the
left side of the top panel. This increases the width of the middle
section of the documentation and can be helpful on smaller screens.
Clicking the hamburger menu again will re-open it.
:::

The documentation is divided into a few parts that can be seen in the
*left hand side navigation bar*:

-   **User documentation**: this includes documentation for anyone
    looking to use NeuroML

-   **NeuroML events**: any events related to NeuroML will be listed
    here

-   **The NeuroML Initiative**: this includes documentation on the
    NeuroML community

-   **Developer documentation**: this includes information for
    individuals looking to contribute to NeuroML (either the standard or
    the software)

-   **Reference**: this includes the glossary of terms and the
    bibliography.

Each part contains different chapters, which can each contain different
sections. Each page in the documentation also has its own navigation in
the *right hand side bar*.
:::

::: {#using-jupyter-notebooks-included-in-the-documentation .section}
[]{#userdocs-usage-jupyterbooks}

### Using Jupyter notebooks included in the documentation[\#](#using-jupyter-notebooks-included-in-the-documentation "Link to this heading"){.headerlink}

::: {.tip .admonition}
Familiar with Jupyter Notebooks? Skip ahead to the next section.

If you are familiar with Jupyter Notebooks, you can skip ahead to the
[[Getting started with NeuroML]{.std
.std-ref}](Landing.html#userdocs-getting-started-neuroml){.reference
.internal} section.
:::

The most important feature of Jupyter books is that it allows you to
include [Jupyter
notebooks](https://jupyter-notebook.readthedocs.io/en/stable/notebook.html){.reference
.external} in the documentation. This allows us to write documentation
which includes code examples that can be modified and executed by users
interactively in their browsers *without having to install anything on
their local machines*. For example, these are used in the [[Getting
Started]{.std
.std-ref}](Landing.html#userdocs-getting-started-neuroml){.reference
.internal} section.

Each Jupyter notebook in the documentation includes a rocket icon 🚀 in
the top bar:

<figure id="id1" class="align-center">
<img src="_images/izhikevich-rocket.png"
alt="Click the rocket icon in top panel of executable pages to execute them in Binder or Google Collaboratory." />
<figcaption><p><span class="caption-number">Fig. 1 </span><span
class="caption-text">Click the rocket icon in top panel of executable
pages to execute them in Binder or Google Collaboratory.</span><a
href="#id1" class="headerlink"
title="Link to this image">#</a></p></figcaption>
</figure>

Clicking this icon will allow you to run the Jupyter Notebook:

<figure id="id2" class="align-center">
<img src="_images/izhikevich-rocket-options.png"
alt="You can run the Jupyter Notebook on Binder or Google Colaboratory." />
<figcaption><p><span class="caption-number">Fig. 2 </span><span
class="caption-text">You can run the Jupyter Notebook on Binder or
Google Colaboratory.</span><a href="#id2" class="headerlink"
title="Link to this image">#</a></p></figcaption>
</figure>

You can choose from freely available services such as
[Binder](https://mybinder.org/){.reference .external} and [Google
Colaboratory](https://colab.research.google.com/){.reference .external}.
Both Binder and Google Colaboratory will take you to these services and
load the Jupyter Notebook for you to use. The Live code option uses
Binder but allows you to run the code in the current tab itself.
However, please note that this option does not include the full Jupyter
Notebook features that Binder and Google Colaboratory provide.

::: {.tip .admonition}
Run Binder and Google Colaboratory in a new tab.

It is suggested to right click and select "Open in new tab" so that the
tab with the NeuroML documentation remains open. In most browsers, you
can also use [`Ctrl`{.docutils .literal
.notranslate}]{.pre}` `{.docutils .literal .notranslate}[`+`{.docutils
.literal .notranslate}]{.pre}` `{.docutils .literal
.notranslate}[`click`{.docutils .literal .notranslate}]{.pre} to open
links in a new tab:
:::

<figure id="id3" class="align-center">
<a href="_images/izhikevich-binder.png"
class="reference internal image-reference"><img
src="_images/izhikevich-binder.png"
style="width: 576.0px; height: 289.2px;"
alt="Izhikevich example running in Binder." /></a>
<figcaption><p><span class="caption-number">Fig. 3 </span><span
class="caption-text">Izhikevich example running in Binder</span><a
href="#id3" class="headerlink"
title="Link to this image">#</a></p></figcaption>
</figure>

<figure id="id4" class="align-center">
<a href="_images/izhikevich-google.png"
class="reference internal image-reference"><img
src="_images/izhikevich-google.png"
style="width: 576.0px; height: 289.2px;"
alt="Izhikevich example running in Google Colaboratory." /></a>
<figcaption><p><span class="caption-number">Fig. 4 </span><span
class="caption-text">Izhikevich example running in Google
Colaboratory.</span><a href="#id4" class="headerlink"
title="Link to this image">#</a></p></figcaption>
</figure>

<figure id="id5" class="align-center">
<a href="_images/izhikevich-livecode.png"
class="reference internal image-reference"><img
src="_images/izhikevich-livecode.png"
style="width: 576.0px; height: 289.2px;"
alt="Izhikevich example running on Binder but using the Live Code option." /></a>
<figcaption><p><span class="caption-number">Fig. 5 </span><span
class="caption-text">Izhikevich example running in Binder but using the
Live Code option.</span><a href="#id5" class="headerlink"
title="Link to this image">#</a></p></figcaption>
</figure>

When running the Jupyter notebooks using these services, you can make
changes to the code and re-run it as required. On Binder and Google
Colaboratory, which provide the full range of Jupyter Notebook features,
you can also run all the code cells at once in sequence. Please see the
documentation pages to learn more about using Binder and Google
Colaboratory
[here](https://mybinder.readthedocs.io/en/latest/){.reference .external}
and
[here](https://colab.research.google.com/notebooks/basic_features_overview.ipynb){.reference
.external} respectively. General information on using Jupyter Notebooks
and the interface can be found in the documentation
[here](https://jupyter-notebook.readthedocs.io/en/stable/notebook.html#starting-the-notebook-server){.reference
.external}.

::: {#downloading-jupyter-notebooks-to-run-locally-on-your-machine .section}
[]{#userdocs-usage-jupyterbooks-locally}

#### Downloading Jupyter Notebooks to run locally on your machine[\#](#downloading-jupyter-notebooks-to-run-locally-on-your-machine "Link to this heading"){.headerlink}

Jupyter Notebooks can also be downloaded and run locally on your
machine. To download the notebooks, use the Download link in the top
panel:

<figure id="id6" class="align-center">
<img src="_images/jupyter-download.png"
alt="Jupyter notebooks can be downloaded using the Download link in the top panel." />
<figcaption><p><span class="caption-number">Fig. 6 </span><span
class="caption-text">Jupyter notebooks can be downloaded using the
Download link in the top panel.</span><a href="#id6" class="headerlink"
title="Link to this image">#</a></p></figcaption>
</figure>

You will need to install the Python Jupyter Notebook packages to do so.
Please refer to the Jupyter Notebook
[documentation](https://jupyter.readthedocs.io/en/latest/install/notebook-classic.html#alternative-for-experienced-python-users-installing-jupyter-with-pip){.reference
.external} to see how you can install Jupyter Notebooks. Additionally,
you will also need to install the [[NeuroML software]{.std
.std-ref}](Landing.html#userdocs-software){.reference .internal} to run
these notebooks. Information on using Jupyter Notebooks and the
interface can be found in the documentation
[here](https://jupyter-notebook.readthedocs.io/en/stable/notebook.html#starting-the-notebook-server){.reference
.external}.
:::
:::

::: {#downloading-this-documentation-as-pdf .section}
[]{#userdocs-usage-pdf}

### Downloading this documentation as PDF[\#](#downloading-this-documentation-as-pdf "Link to this heading"){.headerlink}

You can download this documentation as PDF pages for offline use.

To download individual pages, use the download icon in the top bar. This
will generate a PDF page of the current page for you, using your
browser's "print to file" functionality.

You can also download the complete book as a PDF
[here](https://docs.neuroml.org/_static/files/neuroml-documentation.pdf){.reference
.external}.
:::

::: {#reporting-bugs-and-issues .section}
[]{#userdocs-usage-bugs}

### Reporting bugs and issues[\#](#reporting-bugs-and-issues "Link to this heading"){.headerlink}

Please report any issues that you may find in the documentation so that
it can be improved. To report an issue on a particular page, you can use
the "open issue" link under the GitHub icon in the top panel.
Additionally, you can also suggest edits by editing the page in a fork
and opening a pull request using the "suggest and edit" link.

<figure id="id7" class="align-center">
<img src="_images/jupyterbook-issue.png"
alt="You can report issues and suggest edits to the documentation to help us improve it using the options in the GitHub icon in the top panel." />
<figcaption><p><span class="caption-number">Fig. 7 </span><span
class="caption-text">You can report issues and suggest edits to the
documentation to help us improve it using the options in the GitHub icon
in the top panel.</span><a href="#id7" class="headerlink"
title="Link to this image">#</a></p></figcaption>
</figure>

You can also always contact the NeuroML community using our
[[communication channels]{.std
.std-ref}](Landing.html#contact){.reference .internal} if required.
:::
:::

[]{#document-Userdocs/GetNeuroML}

::: {#get-neuroml .section .tex2jax_ignore .mathjax_ignore}
[]{#userdocs-get-neuroml}

## Get NeuroML[\#](#get-neuroml "Link to this heading"){.headerlink}

While one can use Jupyter Notebooks on different platforms
([Binder](https://binder.org){.reference .external}/[Open Source Brain
v2](https://v2.opensourcebrain.org){.reference .external}/[Google
Colab](https://colab.research.google.com){.reference .external}) to work
with NeuroML models (for example, the tutorials in this documentation
can mostly be run on Jupyter Notebooks), for certain use cases, it may
be preferable/necessary to install the stack on our own computers. One
such use case, for example, is when one needs to run large scale
simulations that require supercomputers/clusters to simulate.

::: {#what-you-need .section}
### What you need[\#](#what-you-need "Link to this heading"){.headerlink}

The NeuroML stack is written primarily in Python and Java, and so
requires:

-   a [supported](https://devguide.python.org/versions/){.reference
    .external}, working
    [Python](https://www.python.org/downloads/){.reference .external}
    installation

-   a working Java Runtime Environment (JRE)

The software stack is currently [tested
on](https://github.com/NeuroML/jNeuroML/blob/master/.github/workflows/ci.yml#L19C15-L19C44){.reference
.external}:

-   Python versions: 3.9--3.13 (3.12 is preferred)

-   Java versions 8, 11, 16, 17, 19 on these [operating systems
    (OS)](https://github.com/actions/runner-images){.reference
    .external}: Ubuntu 22.04 ("ubuntu-latest"), MacOS 14 Arm 64
    ("macos-latest"), Windows 2019 ("windows-2019")

Once you have these programming languages installed, all you need to do
is install [[pyNeuroML]{.std
.std-ref}](Landing.html#pyneuroml){.reference .internal}, and that will
install the other parts of the NeuroML software stack for you too.
Please see the [[pyNeuroML]{.std
.std-ref}](Landing.html#pyneuroml){.reference .internal} page for more
details.
:::
:::

[]{#document-Userdocs/GettingStarted}

::: {#getting-started-with-neuroml .section .tex2jax_ignore .mathjax_ignore}
[]{#userdocs-getting-started-neuroml}

## Getting started with NeuroML[\#](#getting-started-with-neuroml "Link to this heading"){.headerlink}

The best way to understand NeuroML is to work through NeuroML examples
to see how they are constructed and what they can do. We present below a
set of step-by-step guides to illustrate how models are written and
simulated using NeuroML.

  ------------------------------------------------------------------------------------------------------------------------------------------------------------ ----------------------------------------------------------------------------------------------------------------------------------------------------------- -------------------------------------------------------------------------------------------------------------------------------------------------------- --------------------------------------------------------------------------------------------------------------------------------
  [![](../_images/example-single-izhikevich2007cell-sim-v.png "Guide 1"){height="150"}](https://docs.neuroml.org/Userdocs/NML2_examples/SingleNeuron.html)     [![](../_images/example_izhikevich2007network_sim-spikes.png "Guide 2"){height="150"}](https://docs.neuroml.org/Userdocs/IzhikevichNetworkExample.html)     [![](../_images/HH_single_compartment_example_sim-v.png "Guide 3"){height="150"}](https://docs.neuroml.org/Userdocs/SingleCompartmentHHExample.html)     [![](../_images/olm.cell.xy.png "Guide 4"){height="150"}](https://docs.neuroml.org/Userdocs/MultiCompartmentOLMexample.html)  
  ------------------------------------------------------------------------------------------------------------------------------------------------------------ ----------------------------------------------------------------------------------------------------------------------------------------------------------- -------------------------------------------------------------------------------------------------------------------------------------------------------- --------------------------------------------------------------------------------------------------------------------------------

::: pst-scrollable-table-container
  Link to guide                                                                                                        Description                                                                                                                                                                                                                                                                                                                 Model life cycle stages
  -------------------------------------------------------------------------------------------------------------------- --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- ----------------------------------------------------------
                                                                                                                       **Introductory guides**                                                                                                                                                                                                                                                                                                     
  [[Guide 1]{.std .std-ref}](Landing.html#userdocs-getting-started-single-example){.reference .internal}               Create and simulate a simple regular spiking Izhikevich neuron in NeuroML                                                                                                                                                                                                                                                   Create, Validate, Simulate
  [[Guide 2]{.std .std-ref}](Landing.html#userdocs-gettingstarted-izhikevichnetwork){.reference .internal}             Create a network of two synaptically connected populations of Izhikevich neurons                                                                                                                                                                                                                                            Create, Validate, Visualise, Simulate
  [[Guide 3]{.std .std-ref}](Landing.html#userdocs-getting-started-single-compartment-example){.reference .internal}   Build and simulate a single compartment Hodgkin-Huxley neuron                                                                                                                                                                                                                                                               Create, Validate, Visualise, Simulate
  [[Guide 4]{.std .std-ref}](Landing.html#userdocs-getting-started-multi-compartment-example){.reference .internal}    Create and simulate a multi compartment hippocampal OLM neuron                                                                                                                                                                                                                                                              Create, Validate, Visualise, Simulate
                                                                                                                       **Advanced guides**                                                                                                                                                                                                                                                                                                         
  [Guide 5](https://docs.neuroml.org/Userdocs/NML2_examples/NeuroML-DB.html){.reference .external}                     Create novel NeuroML models from components on NeuroML-DB                                                                                                                                                                                                                                                                   Reuse, Create, Validate, Simulate
  [[Guide 6]{.std .std-ref}](Landing.html#userdocs-optimising){.reference .internal}                                   Optimise/fit NeuroML models to experimental data                                                                                                                                                                                                                                                                            Create, Validate, Simulate, Fit
  [[Guide 7]{.std .std-ref}](Landing.html#userdocs-extending){.reference .internal}                                    Extend NeuroML by creating a novel model type in LEMS                                                                                                                                                                                                                                                                       Create, Simulate
                                                                                                                       **Step by step walkthroughs**                                                                                                                                                                                                                                                                                               
  [[Guide 8]{.std .std-ref}](Landing.html#userdocs-creating-models-converting-conductance){.reference .internal}       Guide to converting cell models to NeuroML and sharing them on Open Source Brain                                                                                                                                                                                                                                            Create, Validate, Simulate, Share
  [[Guide 9]{.std .std-ref}](Landing.html#userdocs-walkthroughs-rayetal2020){.reference .internal}                     Conversion of Ray et al 2020 [\[[RAS20](Landing.html#id21 "Subhasis Ray, Zane N. Aldworth, and Mark A. Stopfer. Feedback inhibition and its control in an insect olfactory circuit. eLife, 9:e53281, 2020. URL: https://doi.org/10.7554/eLife.53281, doi:10.7554/eLife.53281."){.reference .internal}\]]{#id1} to NeuroML   Create, Validate, Visualise, Simulate, Extend using LEMS
:::

You do not need to install any software on your computers to run many of
the examples above. These examples are followed by a [Jupyter
notebook](https://jupyter.org/index.html){.reference .external} for you
to experiment with inside your browser ([[more info]{.std
.std-ref}](Landing.html#userdocs-usage-jupyterbooks){.reference
.internal}).

::: {.toctree-wrapper .compound}
[]{#document-Userdocs/SingleNeuronExample}

::: {#simulating-a-regular-spiking-izhikevich-neuron .section .tex2jax_ignore .mathjax_ignore}
[]{#userdocs-getting-started-single-example}

### Simulating a regular spiking Izhikevich neuron[\#](#simulating-a-regular-spiking-izhikevich-neuron "Link to this heading"){.headerlink}

::: {.tip .admonition}
See also the interactive version.

Note: this is a more detailed description of the first example which is
available as an [[interactive Juptyer
notebook]{.doc}](#document-Userdocs/NML2_examples/SingleNeuron){.reference
.internal} on the next page.
:::

In this section, we wish to simulate a single regular spiking Izhikevich
neuron
([\[[Izh07](Landing.html#id2 "Eugene M. Izhikevich. Dynamical systems in neuroscience. MIT Press, 2007."){.reference
.internal}\]]{#id1}) and record/visualise its membrane potential (as
shown in the figure below):

<figure id="id2" class="align-center">
<img src="_images/example-single-izhikevich2007cell-sim-v.png"
alt="Membrane potential for neuron recorded from the simulation" />
<figcaption><p><span class="caption-number">Fig. 8 </span><span
class="caption-text">Membrane potential of the simulated regular spiking
Izhikevich neuron.</span><a href="#id2" class="headerlink"
title="Link to this image">#</a></p></figcaption>
</figure>

This plot, saved as
[`example-single-izhikevich2007cell-sim-v.png`{.docutils .literal
.notranslate}]{.pre}, is generated using the following Python NeuroML
script:

::: {.highlight-python .notranslate}
::: highlight
    #!/usr/bin/env python3
    """
    Simulating a regular spiking Izhikevich neuron with NeuroML.

    File: izhikevich-single-neuron.py
    """

    from neuroml import NeuroMLDocument
    import neuroml.writers as writers
    from neuroml.utils import component_factory
    from neuroml.utils import validate_neuroml2
    from pyneuroml import pynml
    from pyneuroml.lems import LEMSSimulation
    import numpy as np


    # Create a new NeuroML model document
    # component_factory: form one: provide name of NeuroML class as string
    # advantage of this form: do not need to import all the ComponentType classes
    # before using them
    nml_doc = component_factory("NeuroMLDocument", id="IzhSingleNeuron")
    # component_factory: form two: provide class as argument
    # nml_doc = component_factory(NeuroMLDocument, id="IzhSingleNeuron")

    # Inspect it:
    nml_doc.info()

    # Also see contents:
    nml_doc.info(show_contents=True)

    # Define the Izhikevich cell and add it to the model in the document
    # the `add` will create and validate the new component, and add it to the
    # parent (nml_doc)
    izh0 = nml_doc.add(
        "Izhikevich2007Cell",
        id="izh2007RS0", v0="-60mV", C="100pF", k="0.7nS_per_mV", vr="-60mV",
        vt="-40mV", vpeak="35mV", a="0.03per_ms", b="-2nS", c="-50.0mV", d="100pA")

    # Exercise 1: give wrong units of a parameter/parameters
    # Exercise 2: skip out a few parameters

    # Inspect the component
    izh0.info()

    # Inspect the component, also show all members:
    izh0.info(True)

    # inspect the document
    nml_doc.info(show_contents=True)

    # Create a network and add it to the model
    # net = component_factory("Network", id="IzNet")
    # Throws an error: why?
    # Because a Population is necessary in a Network, but we have not provided one.
    # Two workarounds:
    # - create population first, and pass that to component_factory here
    # - disable validation
    net = nml_doc.add("Network", id="IzNet", validate=False)

    # Create a population of defined cells and add it to the model
    size0 = 1
    pop0 = net.add("Population", id="IzhPop0", component=izh0.id, size=size0)

    # Define an external stimulus and add it to the model
    pg = nml_doc.add(
        "PulseGenerator",
        id="pulseGen_%i" % 0, delay="0ms", duration="1000ms",
        amplitude="0.07 nA"
    )
    exp_input = net.add("ExplicitInput", target="%s[%i]" % (pop0.id, 0), input=pg.id)

    # Write the NeuroML model to a file
    nml_file = 'izhikevich2007_single_cell_network.nml'
    writers.NeuroMLWriter.write(nml_doc, nml_file)
    print("Written network file to: " + nml_file)

    # Validate the NeuroML model against the NeuroML schema
    validate_neuroml2(nml_file)

    ################################################################################
    # The NeuroML file has now been created and validated. The rest of the code
    # involves writing a LEMS simulation file to run an instance of the model

    # Create a simulation instance of the model
    simulation_id = "example-single-izhikevich2007cell-sim"
    simulation = LEMSSimulation(sim_id=simulation_id,
                                duration=1000, dt=0.1, simulation_seed=123)
    simulation.assign_simulation_target(net.id)
    simulation.include_neuroml2_file(nml_file)

    # Define the output file to store simulation outputs
    # we record the neuron's membrane potential
    simulation.create_output_file(
        "output0", "%s.v.dat" % simulation_id
    )
    simulation.add_column_to_output_file("output0", 'IzhPop0[0]', 'IzhPop0[0]/v')

    # Save the simulation to a file
    lems_simulation_file = simulation.save_to_file()

    # Run the simulation using the jNeuroML simulator
    pynml.run_lems_with_jneuroml(
        lems_simulation_file, max_memory="2G", nogui=True, plot=False
    )

    # Load the data from the file and plot the graph for the membrane potential
    # using the pynml generate_plot utility function.
    data_array = np.loadtxt("%s.v.dat" % simulation_id)
    pynml.generate_plot(
        [data_array[:, 0]], [data_array[:, 1]],
        "Membrane potential", show_plot_already=False,
        save_figure_to="%s-v.png" % simulation_id,
        xaxis="time (s)", yaxis="membrane potential (V)"
    )
:::
:::

::: {#declaring-the-model-in-neuroml .section}
[]{#userdocs-getting-started-single-example-declaring}

#### Declaring the model in NeuroML[\#](#declaring-the-model-in-neuroml "Link to this heading"){.headerlink}

::: {.admonition-python-is-the-suggested-programming-language-to-use-for-working-with-neuroml .admonition}
Python is the suggested programming language to use for working with
NeuroML.

The Python NeuroML tools and libraries provide a convenient, easy to use
interface to use NeuroML.
:::

Let us step through the different sections of the Python script. To
start writing a model in NeuroML, we first create a
[`NeuroMLDocument`{.docutils .literal .notranslate}]{.pre}. This
"document" represents the complete model and is the top level container
for everything that the model should contain.

::: {.highlight-python .notranslate}
::: highlight
    nml_doc = component_factory("NeuroMLDocument", id="IzhSingleNeuron")
:::
:::

Let us define an Izhikevich cell that we will use to simulate a neuron.
The Izhikevich neuron model can take sets of parameters to exhibit
different types of spiking behaviour. Here, we define a component
(object) of the general Izhikevich cell using parameters to show regular
spiking.

::: {.highlight-python .notranslate}
::: highlight
    izh0 = nml_doc.add(
        "Izhikevich2007Cell",
        id="izh2007RS0", v0="-60mV", C="100pF", k="0.7nS_per_mV", vr="-60mV",
        vt="-40mV", vpeak="35mV", a="0.03per_ms", b="-2nS", c="-50.0mV", d="100pA")
:::
:::

Now that the neuron has been defined and added to the document, we
declare a [[network]{.std
.std-ref}](Landing.html#schema-network){.reference .internal} with a
[[population]{.std .std-ref}](Landing.html#schema-population){.reference
.internal} of these neurons to create a network in a similar way. Here,
our model includes one network which includes only one population, which
in turn only consists of a single neuron. Once the network, its
populations, and their neurons have been declared, we again add them to
our model:

::: {.highlight-python .notranslate}
::: highlight
    net = nml_doc.add("Network", id="IzNet", validate=False)

    # Create a population of defined cells and add it to the model
    size0 = 1
    pop0 = net.add("Population", id="IzhPop0", component=izh0.id, size=size0)
:::
:::

Question: why did we disable validation when we created the new network
component?

::: {.highlight-python .notranslate}
::: highlight
    net = nml_doc.add("Network", id="IzNet", validate=False)
:::
:::

Let us try creating a network without disabling validation:

::: {.highlight-python .notranslate}
::: highlight
    net = nml_doc.add("Network", id="IzNet")
:::
:::

It will throw a validation error:

::: {.highlight-default .notranslate}
::: highlight
    ValueError: Validation failed:
    - Number of values for populations is below the minimum allowed, expected at least 1, found 0
:::
:::

This is because a network must have at least one population for it to be
valid. To fix this, we can either create the population before the
network, or we can disable validation. Here we chose to disable
validation because we knew we were immediately creating our population
and adding it to our network.

Moving on, since we are providing a single input to the single cell in
our network, we can add an [`ExplicitInput`{.docutils .literal
.notranslate}]{.pre} to our network. See the supplementary section on
the [`info`{.docutils .literal .notranslate}]{.pre} function below to
learn how you can find out that [`ExplicitInput`{.docutils .literal
.notranslate}]{.pre} could be used here.

The list of inputs included in the NeuroML specification can be found on
the [[inputs]{.std .std-ref}](Landing.html#schema-inputs){.reference
.internal} page. We use a [[pulse generator]{.std
.std-ref}](Landing.html#schema-pulsegenerator){.reference .internal}
here, creating a new component and adding it to our NeuroML document. To
connect it to our neuron, we specify the neuron as the
[`target`{.docutils .literal .notranslate}]{.pre} using an [[explicit
input]{.std .std-ref}](Landing.html#schema-explicitinput){.reference
.internal}.

::: {.highlight-python .notranslate}
::: highlight
    # Define an external stimulus and add it to the model
    pg = nml_doc.add(
        "PulseGenerator",
        id="pulseGen_%i" % 0, delay="0ms", duration="1000ms",
        amplitude="0.07 nA"
    )
    exp_input = net.add("ExplicitInput", target="%s[%i]" % (pop0.id, 0), input=pg.id)
:::
:::

This completes our model. It includes a single network, with one
population of one neuron that is driven by one pulse generator. At this
point, we can save our model to a file and validate it again to check if
it conforms to the NeuroML schema (more on this [[later]{.std
.std-ref}](Landing.html#userdocs-validating-models){.reference
.internal}).

::: {.highlight-python .notranslate}
::: highlight
    writers.NeuroMLWriter.write(nml_doc, nml_file)
    print("Written network file to: " + nml_file)

    # Validate the NeuroML model against the NeuroML schema
    validate_neuroml2(nml_file)
:::
:::

Note that the validation here will re-run the tests our component
factory and other methods use, but it also runs a series of additional
tests that can only be run on the complete model. So, it is necessary to
validate the model after it has been fully constructed.
:::

::: {#simulating-the-model .section}
[]{#userdocs-getting-started-single-example-simulating}

#### Simulating the model[\#](#simulating-the-model "Link to this heading"){.headerlink}

Until now, we have just declared the model in NeuroML. We have not,
however, included any information related to the simulation of this
model, e.g. how long to run it for, what to save from the simulation
etc.

With NeuroML v2, the information required to simulate the model is
provided using a [[LEMS Simulation file]{.std
.std-ref}](Landing.html#userdocs-lemssimulation){.reference .internal}.
We will not go into the details of LEMS just yet. We will limit
ourselves to the bits necessary to simulate our Izhikevich neuron only.

The following lines of code instantiate a new simulation with certain
simulation parameters: [`duration`{.docutils .literal
.notranslate}]{.pre}, [`dt`{.docutils .literal .notranslate}]{.pre},
[`simulation_seed`{.docutils .literal .notranslate}]{.pre}.
Additionally, they also define what information is being recorded from
the simulation. In this case, we create an output file, and then add a
new column to record the membrane potential [`v`{.docutils .literal
.notranslate}]{.pre} from our one neuron in the one population in it.
You can read more about recording from NeuroML simulations [[here]{.std
.std-ref}](Landing.html#userdocs-quantitiesandrecording){.reference
.internal}.

Finally, like we had saved our NeuroML model to a file, we also save our
LEMS document to a file.

::: {.highlight-python .notranslate}
::: highlight
    simulation_id = "example-single-izhikevich2007cell-sim"
    simulation = LEMSSimulation(sim_id=simulation_id,
                                duration=1000, dt=0.1, simulation_seed=123)
    simulation.assign_simulation_target(net.id)
    simulation.include_neuroml2_file(nml_file)

    # Define the output file to store simulation outputs
    # we record the neuron's membrane potential
    simulation.create_output_file(
        "output0", "%s.v.dat" % simulation_id
    )
    simulation.add_column_to_output_file("output0", 'IzhPop0[0]', 'IzhPop0[0]/v')

    # Save the simulation to a file
    lems_simulation_file = simulation.save_to_file()
:::
:::

Finally, [[pyNeuroML]{.std .std-ref}](Landing.html#pyneuroml){.reference
.internal} also includes functions that allow you to run the simulation
from the Python script itself:

::: {.highlight-python .notranslate}
::: highlight
    pynml.run_lems_with_jneuroml(
        lems_simulation_file, max_memory="2G", nogui=True, plot=False
    )
:::
:::

Here, we are running our simulation using the [[jNeuroML]{.std
.std-ref}](Landing.html#jneuroml){.reference .internal} simulator, which
is bundled with [[pyNeuroML]{.std
.std-ref}](Landing.html#pyneuroml){.reference .internal}. Since NeuroML
is a well defined standard, models defined in NeuroML can also be run
using other [[supported simulators]{.std
.std-ref}](Landing.html#userdocs-simulators){.reference .internal}.
:::

::: {#plotting-the-recorded-membrane-potential .section}
[]{#userdocs-getting-started-single-example-plotting}

#### Plotting the recorded membrane potential[\#](#plotting-the-recorded-membrane-potential "Link to this heading"){.headerlink}

Once we have simulated our model and the data has been collected in the
specified file, we can analyse the data. pyNeuroML also includes some
helpful functions to quickly plot various recorded variables. The last
few lines of code shows how the membrane potential plot at the top of
the page is generated.

::: {.highlight-python .notranslate}
::: highlight
    data_array = np.loadtxt("%s.v.dat" % simulation_id)
    pynml.generate_plot(
        [data_array[:, 0]], [data_array[:, 1]],
        "Membrane potential", show_plot_already=False,
        save_figure_to="%s-v.png" % simulation_id,
        xaxis="time (s)", yaxis="membrane potential (V)"
    )
:::
:::

On the next page, you will find an interactive Jupyter notebook where
you can play with this example. Click the "launch" button in the top
right hand corner to run the notebook in a configured service. *You do
not need to install any software on your computer to run these
notebooks.*
:::

::: {#supplementary-information .section}
[]{#userdocs-getting-started-single-example-supplementary}

#### Supplementary information[\#](#supplementary-information "Link to this heading"){.headerlink}

The sections here explain concepts that have been used above. These will
help give you a deeper understanding of NeuroML, so we do suggest you go
through them also.

::: {#the-generated-neuroml-model-xml .section}
[]{#userdocs-getting-started-single-example-supplementary-model}

##### The generated NeuroML model XML[\#](#the-generated-neuroml-model-xml "Link to this heading"){.headerlink}

Let us investigate the generated NeuroML XML file:

::: {.highlight-xml .notranslate}
::: highlight
    <neuroml xmlns="http://www.neuroml.org/schema/neuroml2"  xmlns:xs="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.neuroml.org/schema/neuroml2 https://raw.github.com/NeuroML/NeuroML2/development/Schemas/NeuroML2/NeuroML_v2.3.xsd" id="IzhSingleNeuron">
        <izhikevich2007Cell id="izh2007RS0" C="100pF" v0="-60mV" k="0.7nS_per_mV" vr="-60mV" vt="-40mV" vpeak="35mV" a="0.03per_ms" b="-2nS" c="-50.0mV" d="100pA"/>
        <pulseGenerator id="pulseGen_0" delay="0ms" duration="1000ms" amplitude="0.07 nA"/>
        <network id="IzNet">
            <population id="IzhPop0" component="izh2007RS0" size="1"/>
            <explicitInput target="IzhPop0[0]" input="pulseGen_0"/>
        </network>
    </neuroml>
:::
:::

NeuroML files are written in XML. So, they consist of tags and
attributes and can be processed by general purpose XML tools. Each
entity between chevrons is a *tag*: [`<..>`{.docutils .literal
.notranslate}]{.pre}, and each tag may have multiple *attributes* that
are defined using the [`name=value`{.docutils .literal
.notranslate}]{.pre} format. For example [`<neuroml`{.docutils .literal
.notranslate}]{.pre}` `{.docutils .literal .notranslate}[`..>`{.docutils
.literal .notranslate}]{.pre} is a tag, that contains the
[`id`{.docutils .literal .notranslate}]{.pre} attribute with value
[`NML2_SimpleIonChannel`{.docutils .literal .notranslate}]{.pre}.

::: {.admonition-xml-tutorial .admonition}
XML Tutorial

For details on XML, have a look through [this
tutorial](https://www.w3schools.com/xml/){.reference .external}.
:::

::: {.tip .admonition}
Is this XML well-formed?

A NeuroML file needs to be both 1) well-formed, as in complies with the
general rules of the XML language syntax, and 2) valid, i.e. contains
the expected NeuroML specific tags/attributes.

Is the XML shown above well-formed? See for yourself. Copy the NeuroML
file listed above and check it using an [online XML syntax
checker](https://www.w3schools.com/xml/xml_validator.asp){.reference
.external}.
:::

Let us step through this file to understand the different constructs
used in it. The first segment introduces the [`neuroml`{.docutils
.literal .notranslate}]{.pre} tag that includes information on the
specification that this NeuroML file adheres to.

::: {.highlight-xml .notranslate}
::: highlight
    <neuroml xmlns="http://www.neuroml.org/schema/neuroml2"  xmlns:xs="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.neuroml.org/schema/neuroml2 https://raw.github.com/NeuroML/NeuroML2/development/Schemas/NeuroML2/NeuroML_v2.3.xsd" id="IzhSingleNeuron">
:::
:::

The first attribute, [`xmlns`{.docutils .literal .notranslate}]{.pre}
defines the XML *namespace*. All the tags that are defined for use in
NeuroML are defined for use in the NeuroML namespace. This prevents
conflicts with other XML schemas that may use the same tags. Read more
on XML namespaces
[here](https://en.wikipedia.org/wiki/XML_namespace){.reference
.external}.

The remaining lines in this snippet refer to the *XML Schema* that is
defined for NeuroML. XML itself does not define any tags, so any tags
can be used in a general XML document. Here is an example of a valid XML
document, a simple HTML snippet:

::: {.highlight-xml .notranslate}
::: highlight
    <html>
    <head>
    <title>A title</title>
    </head>
    </html>
:::
:::

NeuroML, however, does not use these tags. It defines its own set of
standard tags using an [XML
Schema](http://www.w3.org/2001/XMLSchema-instance){.reference
.external}. In other words, the NeuroML XML schema defines the structure
and contents of a valid NeuroML document. Various tools can then compare
NeuroML documents to the NeuroML Schema to validate them.

::: {.admonition-purpose-of-the-neuroml-schema .admonition}
Purpose of the NeuroML schema

The NeuroML Schema defines the structure and contents of a valid NeuroML
document.
:::

The [`xmlns:xi`{.docutils .literal .notranslate}]{.pre} attribute
documents that NeuroML has a defined XML Schema. The next attribute,
[`xsi:schemaLocation`{.docutils .literal .notranslate}]{.pre} tells us
the locations of the NeuroML Schema. Here, two locations are provided:

-   the Web URL:
    [http://www.neuroml.org/schema/neuroml2](http://www.neuroml.org/schema/neuroml2){.reference
    .external},

-   and the location of the Schema Definition file (an [`xsd`{.docutils
    .literal .notranslate}]{.pre} file) relative to this example file in
    the GitHub repository.

We will look at the NeuroML schema in detail in later sections. All
NeuroML files must include the [`neuroml`{.docutils .literal
.notranslate}]{.pre} tag, and the attributes related to the NeuroML
Schema. The last attribute, [`id`{.docutils .literal
.notranslate}]{.pre} is the identification (or the name) of this
particular NeuroML document.

The remaining part of the file is the *declaration* of the model and its
dynamics:

::: {.highlight-xml .notranslate}
::: highlight
        <izhikevich2007Cell id="izh2007RS0" C="100pF" v0="-60mV" k="0.7nS_per_mV" vr="-60mV" vt="-40mV" vpeak="35mV" a="0.03per_ms" b="-2nS" c="-50.0mV" d="100pA"/>
        <pulseGenerator id="pulseGen_0" delay="0ms" duration="1000ms" amplitude="0.07 nA"/>
        <network id="IzNet">
            <population id="IzhPop0" component="izh2007RS0" size="1"/>
            <explicitInput target="IzhPop0[0]" input="pulseGen_0"/>
        </network>
:::
:::

The cell, is defined in the [`izhikevich2007Cell`{.docutils .literal
.notranslate}]{.pre} tag, which has a number of attributes as we saw
before (see [[here]{.std
.std-ref}](Landing.html#schema-izhikevich2007cell){.reference .internal}
for the schema definition):

-   [`id`{.docutils .literal .notranslate}]{.pre}: the name that we want
    to give to this cell. To refer to it later, for example,

-   [`v0`{.docutils .literal .notranslate}]{.pre}: the initial membrane
    potential for the cell,

-   [`C`{.docutils .literal .notranslate}]{.pre}: the leak conductance,

-   [`k`{.docutils .literal .notranslate}]{.pre}: conductance per
    voltage,

-   [`vr`{.docutils .literal .notranslate}]{.pre}: the membrane
    potential after a spike,

-   [`vt`{.docutils .literal .notranslate}]{.pre}: the threshold
    membrane potential, to detect a spike,

-   [`vpeak`{.docutils .literal .notranslate}]{.pre}: the peak membrane
    potential,

-   [`a`{.docutils .literal .notranslate}]{.pre}, [`b`{.docutils
    .literal .notranslate}]{.pre}, [`c`{.docutils .literal
    .notranslate}]{.pre}, and [`d`{.docutils .literal
    .notranslate}]{.pre}: are parameters of the Izhikevich neuron model.

Similarly, the [`pulseGenerator`{.docutils .literal .notranslate}]{.pre}
is also defined, and the [`network`{.docutils .literal
.notranslate}]{.pre} tag includes the [`population`{.docutils .literal
.notranslate}]{.pre} and [`explicitInput`{.docutils .literal
.notranslate}]{.pre}. We observe that even though we have declared the
entities, and the values for parameters that govern them, we do not
state what and how these parameters are used. This is because NeuroML is
a [declarative
language](https://en.wikipedia.org/wiki/Declarative_programming){.reference
.external} that defines the structure of models. We do not need to
define how the dynamics of the different parts of the model are
implemented. As we will see further below, these are already defined in
NeuroML.

::: {.admonition-neuroml-is-a-declarative-language .admonition}
NeuroML is a declarative language.

Users describe the various components of the model but do not need to
worry about how they are implemented.
:::

We have seen how an Izhikevich cell can be declared in NeuroML, with all
its parameters.

As is evident, XML files are excellent for storing structured data, but
may not be easy to write by hand. However, NeuroML users *are not
expected* to write in XML. They should use the Python tools as
demonstrated here.
:::

::: {#the-schema .section}
[]{#userdocs-getting-started-single-example-supplementary-schema}

##### The schema[\#](#the-schema "Link to this heading"){.headerlink}

Given that NeuroML develops a standard and defines what tags and
attributes can be used, let us see how these are defined for the
Izhikevich cell. The Izhikevich cell is defined in version 2 of the
NeuroML schema
[here](https://github.com/NeuroML/NeuroML2/blob/master/Schemas/NeuroML2/NeuroML_v2.0.xsd#L1422){.reference
.external}:

::: {.highlight-xml .notranslate}
::: highlight
        <xs:complexType name="Izhikevich2007Cell">
            <xs:complexContent>
                <xs:extension base="BaseCellMembPotCap">
                    <xs:attribute name="v0" type="Nml2Quantity_voltage" use="required"/>
                    <xs:attribute name="k" type="Nml2Quantity_conductancePerVoltage" use="required"/>
                    <xs:attribute name="vr" type="Nml2Quantity_voltage" use="required"/>
                    <xs:attribute name="vt" type="Nml2Quantity_voltage" use="required"/>
                    <xs:attribute name="vpeak" type="Nml2Quantity_voltage" use="required"/>
                    <xs:attribute name="a" type="Nml2Quantity_pertime" use="required"/>
                    <xs:attribute name="b" type="Nml2Quantity_conductance" use="required"/>
                    <xs:attribute name="c" type="Nml2Quantity_voltage" use="required"/>
                    <xs:attribute name="d" type="Nml2Quantity_current" use="required"/>
                </xs:extension>
            </xs:complexContent>
        </xs:complexType>
:::
:::

The [`xs:`{.docutils .literal .notranslate}]{.pre} prefix indicates that
these are all part of an XML Schema. The Izhikevich cell and all its
parameters are defined in the schema. As we saw before, parameters of
the model are defined as attributes in NeuroML files. So, here in the
schema, they are also defined as [`attributes`{.docutils .literal
.notranslate}]{.pre} of the [`complexType`{.docutils .literal
.notranslate}]{.pre} that the schema describes. The schema also
specifies which of the parameters are necessary, and what their
dimensions (units) are using the [`use`{.docutils .literal
.notranslate}]{.pre} and [`type`{.docutils .literal .notranslate}]{.pre}
properties.

This schema gives us all the information we need to describe an
Izhikevich cell in NeuroML. Using the specification in the Schema, any
number of Izhikevich cells can be defined in a NeuroML file with the
necessary parameter sets to create networks of Izhikevich cells.
:::

::: {#the-generated-lems-xml .section}
[]{#userdocs-getting-started-single-example-supplementary-lems}

##### The generated LEMS XML[\#](#the-generated-lems-xml "Link to this heading"){.headerlink}

The generated LEMS simulation file is shown below:

::: {.highlight-xml .notranslate}
::: highlight
    <Lems>

        <!-- 

            This LEMS file has been automatically generated using PyNeuroML v1.1.13 (libNeuroML v0.5.8)

         -->

        <!-- Specify which component to run -->
        <Target component="example-single-izhikevich2007cell-sim"/>

        <!-- Include core NeuroML2 ComponentType definitions -->
        <Include file="Cells.xml"/>
        <Include file="Networks.xml"/>
        <Include file="Simulation.xml"/>

        <Include file="izhikevich2007_single_cell_network.nml"/>

        <Simulation id="example-single-izhikevich2007cell-sim" length="1000ms" step="0.1ms" target="IzNet" seed="123">  <!-- Note seed: ensures same random numbers used every run -->
            <OutputFile id="output0" fileName="example-single-izhikevich2007cell-sim.v.dat">
                <OutputColumn id="IzhPop0[0]" quantity="IzhPop0[0]/v"/>
            </OutputFile>

        </Simulation>

    </Lems>
:::
:::

Similar to NeuroML, a [[LEMS Simulation file]{.std
.std-ref}](Landing.html#userdocs-lemssimulation){.reference .internal}
also has a well defined structure, i.e., a set of valid tags which
define the contents of the LEMS file. We observe that whereas the
NeuroML tags were related to the modelling parameters, the LEMS tags are
related to simulation. We also note that our NeuroML model has been
"included" in the LEMS file, so that all entities defined there are now
known to the LEMS simulation also. Like NeuroML, *users are not expected
to write the LEMS XML component by hand*. They should continue to use
the NeuroML Python tools.
:::

::: {#the-component-factory-function .section}
[]{#userdocs-getting-started-single-example-supplementary-component-factory}

##### The component_factory() function[\#](#the-component-factory-function "Link to this heading"){.headerlink}

In the code above, we've used the [`component_factory`{.code .docutils
.literal .notranslate}]{.pre} utility function that is included in the
[`neuroml.utils`{.code .docutils .literal .notranslate}]{.pre} module.
This is, as the name notes, a "factory function". When we provide the
name of a NeuroML component type (the Python class) to it as the first
argument along with any parameters, it will create a new component
(Python object) and return it to us to use, after running a few checks
under the hood:

-   is the created component valid?

-   are all the necessary parameters set?

-   are any extra parameters given?

We will see some of these checks in action later as we create more
components for our model.

The [`component_factory`{.docutils .literal .notranslate}]{.pre} can
accept two forms. We can either pass the component type (class) to the
function, or we can pass its name as a string. The difference is that we
do not need to [`import`{.docutils .literal .notranslate}]{.pre} the
class in our script before using it if we specify its name as a string.
The component factory function will import the class for us for us
internally. Either form works, so you can choose which you prefer. It is
important to only remain consistent and use one form to aid readability.
:::

::: {#the-add-function .section}
[]{#userdocs-getting-started-single-example-supplementary-add}

##### The add() function[\#](#the-add-function "Link to this heading"){.headerlink}

We've used another utility method in the code above: [`add`{.docutils
.literal .notranslate}]{.pre}. The [`add`{.docutils .literal
.notranslate}]{.pre} method calls the [`component_factory`{.docutils
.literal .notranslate}]{.pre} for us internally to create a new object
of the required component.

We could also use the [`component_factory`{.docutils .literal
.notranslate}]{.pre}, followed by [`add`{.docutils .literal
.notranslate}]{.pre}, which would result in the same thing:

::: {.highlight-python .notranslate}
::: highlight
    izh0 = component_factory(
        "Izhikevich2007Cell",
        id="izh2007RS0", v0="-60mV", C="100pF", k="0.7nS_per_mV", vr="-60mV",
        vt="-40mV", vpeak="35mV", a="0.03per_ms", b="-2nS", c="-50.0mV", d="100pA")
    nml_doc.add(izh0)
:::
:::

In fact, we could do it all without using either method:

::: {.highlight-python .notranslate}
::: highlight
    # from neuroml import Izhikevich2007Cell
    izh0 = neuroml.Izhikevich2007Cell(
        id="izh2007RS0", v0="-60mV", C="100pF", k="0.7nS_per_mV", vr="-60mV",
        vt="-40mV", vpeak="35mV", a="0.03per_ms", b="-2nS", c="-50.0mV", d="100pA")
    nml_doc.izhikevich2007_cells.append(izh0)
:::
:::

This last form is not suggested because here, the extra checks that the
[`component_factory`{.docutils .literal .notranslate}]{.pre} and
[`add`{.docutils .literal .notranslate}]{.pre} methods run are not
carried out. You also need to know the name of the variable in the
[`nml_doc`{.docutils .literal .notranslate}]{.pre} object to be able to
append to it. The output of the [`info`{.docutils .literal
.notranslate}]{.pre} method will list all the member names, but the
[`add`{.docutils .literal .notranslate}]{.pre} method inspects the
parent component and places the child in the right place for us.

An exercise here would be to try providing invalid arguments to the
[`add`{.docutils .literal .notranslate}]{.pre} or
[`component_factory`{.docutils .literal .notranslate}]{.pre} methods.
For example:

-   try giving the wrong units for a parameter

-   try leaving out a parameter

What happens?

For example, I have used the wrong units for the [`d`{.docutils .literal
.notranslate}]{.pre} parameter here, [`ms`{.docutils .literal
.notranslate}]{.pre} instead of [`pA`{.docutils .literal
.notranslate}]{.pre}:

::: {.highlight-default .notranslate}
::: highlight
    # or
    # izh0 = component_factory(
    izh0 = nml_doc.add(
        "Izhikevich2007Cell",
        id="izh2007RS0", v0="-60mV", C="100pF", k="0.7nS_per_mV", vr="-60mV",
        vt="-40mV", vpeak="35mV", a="0.03per_ms", b="-2nS", c="-50.0mV", d="100ms")
:::
:::

and it will throw a [`ValueError`{.docutils .literal
.notranslate}]{.pre} telling us that this does not match the expected
string for [`d`{.docutils .literal .notranslate}]{.pre}:

::: {.highlight-default .notranslate}
::: highlight
    ValueError: Validation failed:
    - Value "100ms" does not match xsd pattern restrictions: [['^(-?([0-9]*(\\.[0-9]+)?)([eE]-?[0-9]+)?[\\s]*(A|uA|nA|pA))$']]
:::
:::

The specific error here includes the "pattern restrictions" ([regular
expression](https://docs.python.org/3/howto/regex.html#regex-howto){.reference
.external}) for valid values of the [`d`{.docutils .literal
.notranslate}]{.pre} parameter. There are a number of tutorials on
regular expressions on the internet that you can use to learn more about
the meaning of the provided pattern restriction. The one restriction
that we are interested in here is that the value of [`d`{.docutils
.literal .notranslate}]{.pre} must end in one of [`A`{.docutils .literal
.notranslate}]{.pre}, [`uA`{.docutils .literal .notranslate}]{.pre},
[`nA`{.docutils .literal .notranslate}]{.pre}, or [`pA`{.docutils
.literal .notranslate}]{.pre}. Anything else will result in an invalid
value, and the factory will throw a [`ValueError`{.docutils .literal
.notranslate}]{.pre}.

The NeuroML specification declares valid units for all its components.
This allows us to validate models and components while building the
model---even before we have a complete model that we want to simulate.
In fact, NeuroML also defines a list of units and dimensions that can be
used.

::: {.admonition-units-in-neuroml .admonition}
Units in NeuroML

NeuroML defines a [[standard set of units]{.std
.std-ref}](Landing.html#schema-neuromlcoredimensions){.reference
.internal} that can be used in models. Learn more about units and
dimensions in NeuroML and LEMS [[here]{.std
.std-ref}](Landing.html#userdocs-unitsanddimensions){.reference
.internal}.
:::
:::

::: {#the-info-function .section}
[]{#userdocs-getting-started-single-example-supplementary-info}

##### The info() function[\#](#the-info-function "Link to this heading"){.headerlink}

Now that we have a document, what if we want to inspect it to see what
components it can hold, and what its current contents are? Each NeuroML
component type includes the [`info`{.code .docutils .literal
.notranslate}]{.pre} function that gives us a quick summary of
information about the component:

::: {.highlight-python .notranslate}
::: highlight
    nml_doc.info()

    # Also see contents:
    nml_doc.info(show_contents=True)
:::
:::

The output will be of this form:

::: {.highlight-default .notranslate}
::: highlight
    Please see the NeuroML standard schema documentation at https://docs.neuroml.org/Userdocs/NeuroMLv2.html for more information.

    Valid members for NeuroMLDocument are:
    * poisson_firing_synapses (class: PoissonFiringSynapse, Optional)
    * fixed_factor_concentration_models (class: FixedFactorConcentrationModel, Optional)
    * transient_poisson_firing_synapses (class: TransientPoissonFiringSynapse, Optional)
    * alpha_current_synapses (class: AlphaCurrentSynapse, Optional)
    * IF_curr_alpha (class: IF_curr_alpha, Optional)
    * alpha_synapses (class: AlphaSynapse, Optional)
    ...
:::
:::

This shows all the valid NeuroML components that the top level
[`NeuroMLDocument`{.docutils .literal .notranslate}]{.pre} component can
directly contain. It also tells us the component type (class)
corresponding to the component (object). It also tells us whether this
component is optional or required.

In the second form, where we also pass [`show_contents=True`{.docutils
.literal .notranslate}]{.pre}, it will also show the contents of each
member if any. We can use this to inspect our created Izhikevich cell
component:

::: {.highlight-python .notranslate}
::: highlight
    izh0.info(True)
:::
:::

The output will be:

::: {.highlight-default .notranslate}
::: highlight
    Izhikevich2007Cell -- Cell based on the modified Izhikevich model in Izhikevich 2007, Dynamical systems in neuroscience, MIT Press

    Please see the NeuroML standard schema documentation at https://docs.neuroml.org/Userdocs/NeuroMLv2.html for more information.

    Valid members for Izhikevich2007Cell are:
    * annotation (class: Annotation, Optional)
    * b (class: Nml2Quantity_conductance, Required)
            * Contents ('ids'/<objects>): -2nS

    * c (class: Nml2Quantity_voltage, Required)
            * Contents ('ids'/<objects>): -50.0mV

    * d (class: Nml2Quantity_current, Required)
            * Contents ('ids'/<objects>): 100pA

    * C (class: Nml2Quantity_capacitance, Required)
            * Contents ('ids'/<objects>): 100pF

    * v0 (class: Nml2Quantity_voltage, Required)
            * Contents ('ids'/<objects>): -60mV

    * k (class: Nml2Quantity_conductancePerVoltage, Required)
            * Contents ('ids'/<objects>): 0.7nS_per_mV

    * vr (class: Nml2Quantity_voltage, Required)
            * Contents ('ids'/<objects>): -60mV

    * neuro_lex_id (class: NeuroLexId, Optional)
    * metaid (class: MetaId, Optional)
    * vt (class: Nml2Quantity_voltage, Required)
            * Contents ('ids'/<objects>): -40mV

    * id (class: NmlId, Required)
            * Contents ('ids'/<objects>): izh2007RS0

    * notes (class: xs:string, Optional)
    * vpeak (class: Nml2Quantity_voltage, Required)
            * Contents ('ids'/<objects>): 35mV

    * properties (class: Property, Optional)
    * a (class: Nml2Quantity_pertime, Required)
            * Contents ('ids'/<objects>): 0.03per_ms
:::
:::

We can see that all the required parameters are correctly set for this
component.

We can also inspect the full document:

::: {.highlight-python .notranslate}
::: highlight
    nml_doc.info(show_contents=True)
:::
:::

Try running this at the beginning of the script right after creating the
document, and at the end when the model has been completed. You should
notice a major change, that our cell has been correctly added to the
document.

::: {.highlight-default .notranslate}
::: highlight
    ...
    * izhikevich2007_cells (class: Izhikevich2007Cell, Optional)
    *         * Contents ('ids'/<objects>): ['izh2007RS0']
    *
    ...
:::
:::

The [`info()`{.docutils .literal .notranslate}]{.pre} function is very
useful to see what components can belong to another. For example, to see
what components can be added to our [`net`{.docutils .literal
.notranslate}]{.pre} network, we can run this:

::: {.highlight-default .notranslate}
::: highlight
    net.info()

    Network -- Network containing:  **population** s ( potentially of type  **populationList** , and so specifying a list of cell  **location** s );  **projection** s ( with lists of  **connection** s ) and/or  **explicitConnection** s; and  **inputList** s ( with lists of  **input** s ) and/or  **explicitInput** s. Note: often in NeuroML this will be of type  **networkWithTemperature**  if there are temperature dependent elements ( e. g. ion channels ).

    Please see the NeuroML standard schema documentation at https://docs.neuroml.org/Userdocs/NeuroMLv2.html for more information.

    Valid members for Network are:
    * metaid (class: MetaId, Optional)
    * notes (class: xs:string, Optional)
    * properties (class: Property, Optional)
    * annotation (class: Annotation, Optional)
    * type (class: networkTypes, Optional)
    * temperature (class: Nml2Quantity_temperature, Optional)
    * neuro_lex_id (class: NeuroLexId, Optional)
    * spaces (class: Space, Optional)
    * regions (class: Region, Optional)
    * extracellular_properties (class: ExtracellularPropertiesLocal, Optional)
    * populations (class: Population, Required)
    * cell_sets (class: CellSet, Optional)
    * id (class: NmlId, Required)
    * synaptic_connections (class: SynapticConnection, Optional)
    * projections (class: Projection, Optional)
    * electrical_projections (class: ElectricalProjection, Optional)
    * continuous_projections (class: ContinuousProjection, Optional)
    * explicit_inputs (class: ExplicitInput, Optional)
    * input_lists (class: InputList, Optional)
:::
:::

This tells us what [`net`{.docutils .literal .notranslate}]{.pre} can
contain. For setting the input, for example, it would seem that we
should use one of either [`ExplicitInput`{.docutils .literal
.notranslate}]{.pre} or [`InputList`{.docutils .literal
.notranslate}]{.pre} here. The [`ctinfo`{.docutils .literal
.notranslate}]{.pre} function can be used to get more information about
these (next).
:::

::: {#the-ctinfo-function .section}
[]{#userdocs-getting-started-single-example-supplementary-ctinfo}

##### The ctinfo() function[\#](#the-ctinfo-function "Link to this heading"){.headerlink}

There are multiple ways of getting information on a component type. The
first, of course, is to look at the [[schema]{.std
.std-ref}](Landing.html#userdocs-neuromlv2){.reference .internal}
documentation online. The documentation for ExplicitInput is
[[here]{.std .std-ref}](Landing.html#schema-explicitinput){.reference
.internal}, and for InputList is [[here]{.std
.std-ref}](Landing.html#schema-inputlist){.reference .internal}. The
schema documentation will also include examples of usage for most
component types under the "Usage:Python" tab.

[`neuroml`{.docutils .literal .notranslate}]{.pre} includes the
[`ctinfo()`{.docutils .literal .notranslate}]{.pre} utility function,
that like the [`info()`{.docutils .literal .notranslate}]{.pre} method,
provides information about component types ([`ct`{.docutils .literal
.notranslate}]{.pre} in [`ctinfo`{.docutils .literal
.notranslate}]{.pre} stands for [`component`{.docutils .literal
.notranslate}]{.pre}` `{.docutils .literal
.notranslate}[`type`{.docutils .literal .notranslate}]{.pre}). Note that
component types are classes and the [`info()`{.docutils .literal
.notranslate}]{.pre} method cannot be used on them. It can only be used
once objects have been created from the component type classes.

So, we could do (create a new dummy object of the class and call
[`info()`{.docutils .literal .notranslate}]{.pre} on it):

::: {.highlight-python .notranslate}
::: highlight
    neuroml.ExplicitInput().info()
:::
:::

but [`ctinfo`{.docutils .literal .notranslate}]{.pre} will do this for
us:

::: {.highlight-python .notranslate}
::: highlight
    from neuroml.utils import ctinfo
    ctinfo("ExplicitInput")
    # or the second form:
    # ctinfo(neuroml.ExplicitInput)
    ExplicitInput -- An explicit input ( anything which extends  **basePointCurrent**  ) to a target cell in a population

    Please see the NeuroML standard schema documentation at https://docs.neuroml.org/Userdocs/NeuroMLv2.html for more information.

    Valid members for ExplicitInput are:
    * destination (class: xs:string, Optional)
    * target (class: xs:string, Required)
    * input (class: xs:string, Required)


    ctinfo("InputList")
    InputList -- An explicit list of  **input** s to a **population.**

    Please see the NeuroML standard schema documentation at https://docs.neuroml.org/Userdocs/NeuroMLv2.html for more information.

    Valid members for InputList are:
    * populations (class: NmlId, Required)
    * component (class: NmlId, Required)
    * input (class: Input, Optional)
    * input_ws (class: InputW, Optional)
    * id (class: NmlId, Required)
:::
:::

Finally, for completeness, we can also get information from the API
documentation for libNeuroML
[here](https://libneuroml.readthedocs.io/en/latest/){.reference
.external}. Since this is documentation that is "embedded" in the Python
classes, we can also use the Python [in-built help
function](https://docs.python.org/3/library/functions.html#help){.reference
.external} to see it:

::: {.highlight-default .notranslate}
::: highlight
    help(neuroml.ExplicitInput)
    Help on class ExplicitInput in module neuroml.nml.nml:

    class ExplicitInput(BaseWithoutId)
     |  ExplicitInput(target: 'one str (required)' = None, input: 'one str (required)' = None, destination: 'one str (optional)' = None, gds_collector_=None, **kwargs_)
     |  
     |  ExplicitInput -- An explicit input ( anything which extends  **basePointCurrent**  ) to a target cell in a population
     |  
     ...



    help(neuroml.InputList)
    Help on class InputList in module neuroml.nml.nml:

    class InputList(Base)
     |  InputList(id: 'one NonNegativeInteger (required)' = None, populations: 'one NmlId (required)' = None, component: 'one NmlId (required)' = None, input: 'list of Input(s) (optional)' = None, input_ws: 'list of InputW(s) (optional)' = None, gds_collector_=None, **kwargs_)
     |  
     |  InputList -- An explicit list of  **input** s to a **population.**
     ...
:::
:::

The information provided by the different sources will be similar, but
[`ctinfo()`{.docutils .literal .notranslate}]{.pre} is perhaps the most
NeuroML specific (whereas the Python [`help()`{.docutils .literal
.notranslate}]{.pre} function provides Python language related
information also.)

::: {.tip .admonition}
Use an integrated development environment (IDE):

[IDEs](https://en.wikipedia.org/wiki/Comparison_of_integrated_development_environments#Python){.reference
.external} make programming easier. For example, a good IDE will show
you the documentation that the [`help`{.docutils .literal
.notranslate}]{.pre} Python function shows.
:::

Another useful function is the [`ctparentinfo()`{.docutils .literal
.notranslate}]{.pre} function. Like [`info()`{.docutils .literal
.notranslate}]{.pre} it provides some information about the
component/object:

::: {.highlight-default .notranslate}
::: highlight
    ctparentinfo("InputList")
    InputList -- An explicit list of  **input** s to a **population.**

    Please see the NeuroML standard schema documentation at https://docs.neuroml.org/Userdocs/NeuroMLv2.html for more information.

    Valid parents for InputList are:
    * Network
            * input_lists (class: InputList, Optional)
:::
:::

This tells us that components of type [`InputList`{.docutils .literal
.notranslate}]{.pre} can be added to components of the
[`Network`{.docutils .literal .notranslate}]{.pre} type, in the
[`input_list`{.docutils .literal .notranslate}]{.pre} member. Of course,
we will use the [`add`{.docutils .literal .notranslate}]{.pre} function
in our network object [`net`{.docutils .literal .notranslate}]{.pre},
and that will add the component to the correct member.
:::

::: {#the-validate-function .section}
[]{#userdocs-getting-started-single-example-supplementary-validate}

##### The validate() function[\#](#the-validate-function "Link to this heading"){.headerlink}

We can check whether each component is valid using the
[`validate`{.docutils .literal .notranslate}]{.pre} function that each
component has. For example:

::: {.highlight-default .notranslate}
::: highlight
    net.validate()
:::
:::

This function does not return anything if the component is valid.
(Technically, if a function does not return anything in Python, it
returns [`None`{.docutils .literal .notranslate}]{.pre} by default, so
this returns [`None`{.docutils .literal .notranslate}]{.pre} if the
component is valid.) However, if it is not valid, it will throw a
[`ValueError`{.docutils .literal .notranslate}]{.pre}.
:::
:::
:::

[]{#document-Userdocs/NML2_examples/SingleNeuron}

::: {#interactive-single-izhikevich-neuron-neuroml-example .section .tex2jax_ignore .mathjax_ignore}
### Interactive single Izhikevich neuron NeuroML example[\#](#interactive-single-izhikevich-neuron-neuroml-example "Link to this heading"){.headerlink}

To run this interactive Jupyter Notebook when viewing the online NeuroML
documentation (e.g. via Binder or Google Colab), please click on the
rocket icon 🚀 in the top panel. For more information, please see [[how
to use this documentation]{.std
.std-ref}](Landing.html#userdocs-usage-jupyterbooks){.reference
.internal}.

This notebook creates a simple model in [NeuroML version
2](https://docs.neuroml.org/Userdocs/NeuroMLv2.html){.reference
.external}. It adds a [simple spiking neuron
model](https://docs.neuroml.org/Userdocs/Schemas/Cells.html#schema-izhikevich2007cell){.reference
.external} to a
[population](https://docs.neuroml.org/Userdocs/Schemas/Networks.html#schema-population){.reference
.external} and then the population to a
[network](https://docs.neuroml.org/Userdocs/Schemas/Networks.html#network){.reference
.external}. Then we create a [LEMS simulation
file](https://docs.neuroml.org/Userdocs/LEMSSimulation.html){.reference
.external} to specify how to to a simulate the model, and finally we
execute it using
[jNeuroML](https://docs.neuroml.org/Userdocs/Software/jNeuroML.html#jneuroml){.reference
.external}. The results of that simulation are plotted below.

![NeuroML Simulation
Plot](https://raw.githubusercontent.com/NeuroML/Documentation/development/source/Userdocs/NML2_examples/SingleNeuron.png)

See also a [[more detailed introduction to NeuroML and LEMS using this
example]{.std
.std-ref}](Landing.html#userdocs-getting-started-single-example){.reference
.internal}.

::: {#initial-setup-and-library-installs .section}
#### 1) Initial setup and library installs[\#](#initial-setup-and-library-installs "Link to this heading"){.headerlink}

Please uncomment the line below if you use the Google Colab (it does not
include these packages by default).

::: {.cell .docutils .container}
::: {.cell_input .docutils .container}
::: {.highlight-ipython3 .notranslate}
::: highlight
    #%pip install pyneuroml neuromllite NEURON
:::
:::
:::
:::

::: {.cell .docutils .container}
::: {.cell_input .docutils .container}
::: {.highlight-ipython3 .notranslate}
::: highlight
    from neuroml import NeuroMLDocument
    import neuroml.writers as writers
    from neuroml.utils import component_factory
    from neuroml.utils import validate_neuroml2
    from pyneuroml import pynml
    from pyneuroml.lems import LEMSSimulation
    import numpy as np
:::
:::
:::

::: {.cell_output .docutils .container}
::: {.output .traceback .highlight-ipythontb .notranslate}
::: highlight
    ---------------------------------------------------------------------------
    ModuleNotFoundError                       Traceback (most recent call last)
    Cell In[2], line 1
    ----> 1 from neuroml import NeuroMLDocument
          2 import neuroml.writers as writers
          3 from neuroml.utils import component_factory

    ModuleNotFoundError: No module named 'neuroml'
:::
:::
:::
:::
:::

::: {#declaring-the-neuroml-model .section}
#### 2) Declaring the NeuroML model[\#](#declaring-the-neuroml-model "Link to this heading"){.headerlink}

::: {#create-a-neuroml-document .section}
##### Create a NeuroML document[\#](#create-a-neuroml-document "Link to this heading"){.headerlink}

This is the container document to which the cells and the network will
be added.

::: {.cell .docutils .container}
::: {.cell_input .docutils .container}
::: {.highlight-ipython3 .notranslate}
::: highlight
    nml_doc = component_factory(NeuroMLDocument, id="IzhSingleNeuron")
:::
:::
:::
:::
:::

::: {#define-the-izhikevich-cell-and-add-it-to-the-model .section}
##### Define the Izhikevich cell and add it to the model[\#](#define-the-izhikevich-cell-and-add-it-to-the-model "Link to this heading"){.headerlink}

The [Izhikevich
model](https://www.izhikevich.org/publications/spikes.htm){.reference
.external} is a simple, 2 variable neuron model exhibiting a range of
neurophysiologically realistic spiking behaviours depending on the
parameters given. We use the
[izhikevich2007cell](https://docs.neuroml.org/Userdocs/Schemas/Cells.html#izhikevich2007cell){.reference
.external} version here.

::: {.cell .docutils .container}
::: {.cell_input .docutils .container}
::: {.highlight-ipython3 .notranslate}
::: highlight
    izh0 = nml_doc.add(
        "Izhikevich2007Cell",
        id="izh2007RS0", v0="-60mV", C="100pF", k="0.7nS_per_mV", vr="-60mV",
        vt="-40mV", vpeak="35mV", a="0.03per_ms", b="-2nS", c="-50.0mV", d="100pA")
:::
:::
:::
:::
:::

::: {#create-a-network-and-add-it-to-the-model .section}
##### Create a network and add it to the model[\#](#create-a-network-and-add-it-to-the-model "Link to this heading"){.headerlink}

We add a
[network](https://docs.neuroml.org/Userdocs/Schemas/Networks.html#network){.reference
.external} to the document created above.

::: {.cell .docutils .container}
::: {.cell_input .docutils .container}
::: {.highlight-ipython3 .notranslate}
::: highlight
    net = nml_doc.add("Network", id="IzNet", validate=False)
:::
:::
:::
:::
:::

::: {#create-a-population-of-defined-cells-and-add-it-to-the-model .section}
##### Create a population of defined cells and add it to the model[\#](#create-a-population-of-defined-cells-and-add-it-to-the-model "Link to this heading"){.headerlink}

A
[population](https://docs.neuroml.org/Userdocs/Schemas/Networks.html#schema-population){.reference
.external} of size 1 of these cells is created and then added to the
network.

::: {.cell .docutils .container}
::: {.cell_input .docutils .container}
::: {.highlight-ipython3 .notranslate}
::: highlight
    size0 = 1
    pop0 = net.add("Population", id="IzhPop0", component=izh0.id, size=size0)
:::
:::
:::
:::
:::

::: {#define-an-external-stimulus-and-add-it-to-the-model .section}
##### Define an external stimulus and add it to the model[\#](#define-an-external-stimulus-and-add-it-to-the-model "Link to this heading"){.headerlink}

On its own the cell will not spike, so we add a small current to it in
the form of a [pulse
generator](https://docs.neuroml.org/Userdocs/Schemas/Inputs.html#pulsegenerator){.reference
.external} which will apply a square pulse of current.

::: {.cell .docutils .container}
::: {.cell_input .docutils .container}
::: {.highlight-ipython3 .notranslate}
::: highlight
    pg = nml_doc.add(
        "PulseGenerator",
        id="pulseGen_%i" % 0, delay="0ms", duration="1000ms",
        amplitude="0.07 nA"
    )
    exp_input = net.add("ExplicitInput", target="%s[%i]" % (pop0.id, 0), input=pg.id)
:::
:::
:::
:::
:::
:::

::: {#write-print-and-validate-the-generated-file .section}
#### 3) Write, print and validate the generated file[\#](#write-print-and-validate-the-generated-file "Link to this heading"){.headerlink}

::: {#write-the-neuroml-model-to-a-file .section}
##### Write the NeuroML model to a file[\#](#write-the-neuroml-model-to-a-file "Link to this heading"){.headerlink}

::: {.cell .docutils .container}
::: {.cell_input .docutils .container}
::: {.highlight-ipython3 .notranslate}
::: highlight
    nml_file = 'izhikevich2007_single_cell_network.nml'
    writers.NeuroMLWriter.write(nml_doc, nml_file)
    print("Written network file to: " + nml_file)
:::
:::
:::

::: {.cell_output .docutils .container}
::: {.output .stream .highlight-myst-ansi .notranslate}
::: highlight
    Written network file to: izhikevich2007_single_cell_network.nml
:::
:::
:::
:::
:::

::: {#print-out-the-file .section}
##### Print out the file[\#](#print-out-the-file "Link to this heading"){.headerlink}

Here we print the XML for the saved NeuroML file.

::: {.cell .docutils .container}
::: {.cell_input .docutils .container}
::: {.highlight-ipython3 .notranslate}
::: highlight
    with open(nml_file) as f: 
       print(f.read())
:::
:::
:::

::: {.cell_output .docutils .container}
::: {.output .stream .highlight-myst-ansi .notranslate}
::: highlight
    <neuroml xmlns="http://www.neuroml.org/schema/neuroml2"  xmlns:xs="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.neuroml.org/schema/neuroml2 https://raw.github.com/NeuroML/NeuroML2/development/Schemas/NeuroML2/NeuroML_v2.3.xsd" id="IzhSingleNeuron">
        <izhikevich2007Cell id="izh2007RS0" C="100pF" v0="-60mV" k="0.7nS_per_mV" vr="-60mV" vt="-40mV" vpeak="35mV" a="0.03per_ms" b="-2nS" c="-50.0mV" d="100pA"/>
        <pulseGenerator id="pulseGen_0" delay="0ms" duration="1000ms" amplitude="0.07 nA"/>
        <network id="IzNet">
            <population id="IzhPop0" component="izh2007RS0" size="1"/>
            <explicitInput target="IzhPop0[0]" input="pulseGen_0"/>
        </network>
    </neuroml>
:::
:::
:::
:::
:::

::: {#validate-the-neuroml-model .section}
##### Validate the NeuroML model[\#](#validate-the-neuroml-model "Link to this heading"){.headerlink}

::: {.cell .docutils .container}
::: {.cell_input .docutils .container}
::: {.highlight-ipython3 .notranslate}
::: highlight
    validate_neuroml2(nml_file)
:::
:::
:::

::: {.cell_output .docutils .container}
::: {.output .stream .highlight-myst-ansi .notranslate}
::: highlight
    It's valid!
:::
:::
:::
:::
:::
:::

::: {#simulating-the-model .section}
#### 4) Simulating the model[\#](#simulating-the-model "Link to this heading"){.headerlink}

::: {#create-a-simulation-instance-of-the-model .section}
##### Create a simulation instance of the model[\#](#create-a-simulation-instance-of-the-model "Link to this heading"){.headerlink}

The NeuroML file does not contain any information on how long to
simulate the model for or what to save etc. For this we will need a
[LEMS simulation
file](https://docs.neuroml.org/Userdocs/LEMSSimulation.html){.reference
.external}.

::: {.cell .docutils .container}
::: {.cell_input .docutils .container}
::: {.highlight-ipython3 .notranslate}
::: highlight
    simulation_id = "example-single-izhikevich2007cell-sim"
    simulation = LEMSSimulation(sim_id=simulation_id,
                                duration=1000, dt=0.1, simulation_seed=123)
    simulation.assign_simulation_target(net.id)
    simulation.include_neuroml2_file(nml_file)
:::
:::
:::

::: {.cell_output .docutils .container}
::: {.output .stderr .highlight-myst-ansi .notranslate}
::: highlight
    pyNeuroML >>> INFO - Loading NeuroML2 file: /Users/padraig/git/Documentation/source/Userdocs/NML2_examples/izhikevich2007_single_cell_network.nml
:::
:::
:::
:::
:::

::: {#define-the-output-file-to-store-simulation-outputs .section}
##### Define the output file to store simulation outputs[\#](#define-the-output-file-to-store-simulation-outputs "Link to this heading"){.headerlink}

Here, we record the neuron's membrane potential to the specified data
file.

::: {.cell .docutils .container}
::: {.cell_input .docutils .container}
::: {.highlight-ipython3 .notranslate}
::: highlight
    simulation.create_output_file(
        "output0", "%s.v.dat" % simulation_id
    )
    simulation.add_column_to_output_file("output0", 'IzhPop0[0]', 'IzhPop0[0]/v')
:::
:::
:::
:::
:::

::: {#save-the-simulation-to-a-file .section}
##### Save the simulation to a file[\#](#save-the-simulation-to-a-file "Link to this heading"){.headerlink}

::: {.cell .docutils .container}
::: {.cell_input .docutils .container}
::: {.highlight-ipython3 .notranslate}
::: highlight
    lems_simulation_file = simulation.save_to_file()
    with open(lems_simulation_file) as f: 
       print(f.read())
:::
:::
:::

::: {.cell_output .docutils .container}
::: {.output .stream .highlight-myst-ansi .notranslate}
::: highlight
    <Lems>

        <!-- 

            This LEMS file has been automatically generated using PyNeuroML v1.1.10 (libNeuroML v0.5.8)

         -->

        <!-- Specify which component to run -->
        <Target component="example-single-izhikevich2007cell-sim"/>

        <!-- Include core NeuroML2 ComponentType definitions -->
        <Include file="Cells.xml"/>
        <Include file="Networks.xml"/>
        <Include file="Simulation.xml"/>

        <Include file="izhikevich2007_single_cell_network.nml"/>

        <Simulation id="example-single-izhikevich2007cell-sim" length="1000ms" step="0.1ms" target="IzNet" seed="123">  <!-- Note seed: ensures same random numbers used every run -->
            <OutputFile id="output0" fileName="example-single-izhikevich2007cell-sim.v.dat">
                <OutputColumn id="IzhPop0[0]" quantity="IzhPop0[0]/v"/>
            </OutputFile>

        </Simulation>

    </Lems>
:::
:::
:::
:::
:::
:::

::: {#run-the-simulation-using-the-jneuroml-simulator .section}
#### Run the simulation using the jNeuroML simulator[\#](#run-the-simulation-using-the-jneuroml-simulator "Link to this heading"){.headerlink}

::: {.cell .docutils .container}
::: {.cell_input .docutils .container}
::: {.highlight-ipython3 .notranslate}
::: highlight
    pynml.run_lems_with_jneuroml(
        lems_simulation_file, max_memory="2G", nogui=True, plot=False
    )
:::
:::
:::

::: {.cell_output .docutils .container}
::: {.output .stderr .highlight-myst-ansi .notranslate}
::: highlight
    pyNeuroML >>> INFO - Loading LEMS file: LEMS_example-single-izhikevich2007cell-sim.xml and running with jNeuroML
    pyNeuroML >>> INFO - Executing: (java -Xmx2G  -Djava.awt.headless=true -jar  "/opt/homebrew/anaconda3/envs/py39n/lib/python3.9/site-packages/pyneuroml/lib/jNeuroML-0.13.0-jar-with-dependencies.jar"  LEMS_example-single-izhikevich2007cell-sim.xml  -nogui -I '') in directory: .
    pyNeuroML >>> INFO - Command completed successfully!
:::
:::

::: {.output .text_plain .highlight-myst-ansi .notranslate}
::: highlight
    True
:::
:::
:::
:::
:::

::: {#plot-the-recorded-data .section}
#### Plot the recorded data[\#](#plot-the-recorded-data "Link to this heading"){.headerlink}

::: {.cell .docutils .container}
::: {.cell_input .docutils .container}
::: {.highlight-ipython3 .notranslate}
::: highlight
    # Load the data from the file and plot the graph for the membrane potential
    # using the pynml generate_plot utility function.
    data_array = np.loadtxt("%s.v.dat" % simulation_id)
    pynml.generate_plot(
        [data_array[:, 0]], [data_array[:, 1]],
        "Membrane potential", show_plot_already=True,
        xaxis="time (s)", yaxis="membrane potential (V)",
        save_figure_to="SingleNeuron.png"
        
    )
:::
:::
:::

::: {.cell_output .docutils .container}
::: {.output .stderr .highlight-myst-ansi .notranslate}
::: highlight
    pyNeuroML >>> INFO - Generating plot: Membrane potential
    pyNeuroML >>> INFO - Saving image to /Users/padraig/git/Documentation/source/Userdocs/NML2_examples/SingleNeuron.png of plot: Membrane potential
    pyNeuroML >>> INFO - Saved image to SingleNeuron.png of plot: Membrane potential
:::
:::

![](_images/b6da95dd80b21f297a7f10e0e0d2a1138c26f6aaa4a683d91e4e0827969083ac.png)

::: {.output .text_plain .highlight-myst-ansi .notranslate}
::: highlight
    <Axes: xlabel='time (s)', ylabel='membrane potential (V)'>
:::
:::
:::
:::
:::
:::

[]{#document-Userdocs/IzhikevichNetworkExample}

::: {#a-two-population-network-of-regular-spiking-izhikevich-neurons .section .tex2jax_ignore .mathjax_ignore}
[]{#userdocs-gettingstarted-izhikevichnetwork}

### A two population network of regular spiking Izhikevich neurons[\#](#a-two-population-network-of-regular-spiking-izhikevich-neurons "Link to this heading"){.headerlink}

Now that we have defined a cell, let us see how a network of these cells
may be declared and simulated. We will create a small network of cells,
simulate this network, and generate a plot of the spike times of the
cells (a raster plot):

<figure id="id1" class="align-center">
<img src="_images/example_izhikevich2007network_sim-spikes.png"
alt="Spike times of neurons recorded from the simulation" />
<figcaption><p><span class="caption-number">Fig. 9 </span><span
class="caption-text">Spike times of neurons in 2 populations recorded
from the simulation.</span><a href="#id1" class="headerlink"
title="Link to this image">#</a></p></figcaption>
</figure>

The Python script used to create the model, simulate it, and generate
this plot is below. Please note that this example uses the
[[NEURON]{.std .std-ref}](Landing.html#userdocs-neuron){.reference
.internal} simulator to simulate the model. Please ensure that the
[`NEURON_HOME`{.docutils .literal .notranslate}]{.pre} environment
variable is correctly set as noted [[here]{.std
.std-ref}](Landing.html#userdocs-neuron-envvar){.reference .internal}.

::: {.highlight-python .notranslate}
::: highlight
    #!/usr/bin/env python3
    """
    Create a simple network with two populations.
    """

    import random
    import numpy as np

    from neuroml.utils import component_factory
    from pyneuroml import pynml
    from pyneuroml.lems import LEMSSimulation
    import neuroml.writers as writers


    nml_doc = component_factory("NeuroMLDocument", id="IzNet")
    iz0 = nml_doc.add(
        "Izhikevich2007Cell",
        id="iz2007RS0",
        v0="-60mV",
        C="100pF",
        k="0.7nS_per_mV",
        vr="-60mV",
        vt="-40mV",
        vpeak="35mV",
        a="0.03per_ms",
        b="-2nS",
        c="-50.0mV",
        d="100pA",
    )

    # Inspect the component, also show all members:
    iz0.info(True)

    # Create a component of type ExpOneSynapse, and add it to the document
    syn0 = nml_doc.add(
        "ExpOneSynapse", id="syn0", gbase="65nS", erev="0mV", tau_decay="3ms"
    )
    # Check what we have so far:
    nml_doc.info(True)
    # Also try:
    print(nml_doc.summary())

    # create the network: turned of validation because we will add populations next
    net = nml_doc.add("Network", id="IzNet", validate=False)

    # create the first population
    size0 = 5
    pop0 = component_factory("Population", id="IzPop0", component=iz0.id, size=size0)
    # Set optional color property. Note: used later when generating plots
    pop0.add("Property", tag="color", value="0 0 .8")
    net.add(pop0)

    # create the second population
    size1 = 5
    pop1 = component_factory("Population", id="IzPop1", component=iz0.id, size=size1)
    pop1.add("Property", tag="color", value=".8 0 0")
    net.add(pop1)

    # network should be valid now that it contains populations
    net.validate()

    # create a projection from one population to another
    proj = net.add(
        "Projection",
        id="proj",
        presynaptic_population=pop0.id,
        postsynaptic_population=pop1.id,
        synapse=syn0.id,
    )

    # We do two things in the loop:
    # - add pulse generator inputs to population 1 to make neurons spike
    # - create synapses between the two populations with a particular probability
    random.seed(123)
    prob_connection = 0.8
    count = 0
    for pre in range(0, size0):
        # pulse generator as explicit stimulus
        pg = nml_doc.add(
            "PulseGenerator",
            id="pg_%i" % pre,
            delay="0ms",
            duration="10000ms",
            amplitude="%f nA" % (0.1 + 0.1 * random.random()),
        )

        exp_input = net.add(
            "ExplicitInput", target="%s[%i]" % (pop0.id, pre), input=pg.id
        )

        # synapses between populations
        for post in range(0, size1):
            if random.random() <= prob_connection:
                syn = proj.add(
                    "Connection",
                    id=count,
                    pre_cell_id="../%s[%i]" % (pop0.id, pre),
                    post_cell_id="../%s[%i]" % (pop1.id, post),
                )
                count += 1

    nml_doc.info(True)
    print(nml_doc.summary())

    # write model to file and validate
    nml_file = "izhikevich2007_network.nml"
    writers.NeuroMLWriter.write(nml_doc, nml_file)

    print("Written network file to: " + nml_file)
    pynml.validate_neuroml2(nml_file)

    # Create simulation, and record data
    simulation_id = "example_izhikevich2007network_sim"
    simulation = LEMSSimulation(
        sim_id=simulation_id, duration=1000, dt=0.1, simulation_seed=123
    )
    simulation.assign_simulation_target(net.id)
    simulation.include_neuroml2_file(nml_file)

    simulation.create_event_output_file(
        "pop0", "%s.0.spikes.dat" % simulation_id, format="ID_TIME"
    )
    for pre in range(0, size0):
        simulation.add_selection_to_event_output_file(
            "pop0", pre, "IzPop0[{}]".format(pre), "spike"
        )

    simulation.create_event_output_file(
        "pop1", "%s.1.spikes.dat" % simulation_id, format="ID_TIME"
    )
    for pre in range(0, size1):
        simulation.add_selection_to_event_output_file(
            "pop1", pre, "IzPop1[{}]".format(pre), "spike"
        )

    lems_simulation_file = simulation.save_to_file()

    # Run the simulation
    pynml.run_lems_with_jneuroml_neuron(
        lems_simulation_file, max_memory="2G", nogui=True, plot=False
    )

    # Load the data from the file and plot the spike times
    # using the pynml generate_plot utility function.
    data_array_0 = np.loadtxt("%s.0.spikes.dat" % simulation_id)
    data_array_1 = np.loadtxt("%s.1.spikes.dat" % simulation_id)
    times_0 = data_array_0[:, 1]
    times_1 = data_array_1[:, 1]
    ids_0 = data_array_0[:, 0]
    ids_1 = [id + size0 for id in data_array_1[:, 0]]
    pynml.generate_plot(
        [times_0, times_1],
        [ids_0, ids_1],
        "Spike times",
        show_plot_already=False,
        save_figure_to="%s-spikes.png" % simulation_id,
        xaxis="time (s)",
        yaxis="cell ID",
        colors=["b", "r"],
        linewidths=["0", "0"],
        markers=[".", "."],
    )
:::
:::

As with the previous example, we will step through this script to see
how the various components of the network are declared in NeuroML before
running the simulation and generating the plot. We will use the same
helper functions to inspect the model as we build it:
[`component_factory`{.docutils .literal .notranslate}]{.pre},
[`add`{.docutils .literal .notranslate}]{.pre}, [`info`{.docutils
.literal .notranslate}]{.pre}, [`summary`{.docutils .literal
.notranslate}]{.pre}.

::: {#declaring-the-model-in-neuroml .section}
[]{#userdocs-gettingstarted-izhikevichnetwork-declaring}

#### Declaring the model in NeuroML[\#](#declaring-the-model-in-neuroml "Link to this heading"){.headerlink}

To declare the complete network model, we must again first declare its
core entities:

::: {.highlight-python .notranslate}
::: highlight
    nml_doc = component_factory("NeuroMLDocument", id="IzNet")
    iz0 = nml_doc.add(
        "Izhikevich2007Cell",
        id="iz2007RS0",
        v0="-60mV",
        C="100pF",
        k="0.7nS_per_mV",
        vr="-60mV",
        vt="-40mV",
        vpeak="35mV",
        a="0.03per_ms",
        b="-2nS",
        c="-50.0mV",
        d="100pA",
    )

    # Inspect the component, also show all members:
    iz0.info(True)

    # Create a component of type ExpOneSynapse, and add it to the document
    syn0 = nml_doc.add(
        "ExpOneSynapse", id="syn0", gbase="65nS", erev="0mV", tau_decay="3ms"
    )
:::
:::

Here, we create a new document, declare the [[Izhikevich neuron]{.std
.std-ref}](Landing.html#schema-izhikevich2007cell){.reference
.internal}, and also declare the synapse that we are going to use to
connect one population of neurons to the other. We use the [[ExpOne
Synapse]{.std .std-ref}](Landing.html#schema-exponesynapse){.reference
.internal} here, where the conductance of the synapse increases
instantaneously by a constant value [`gbase`{.docutils .literal
.notranslate}]{.pre} on receiving a spike, and then decays exponentially
with a decay constant [`tauDecay`{.docutils .literal
.notranslate}]{.pre}.

Let's inspect our model document so far:

::: {.highlight-default .notranslate}
::: highlight
    nml_doc.info(True)
    Please see the NeuroML standard schema documentation at https://docs.neuroml.org/Userdocs/NeuroMLv2.html for more information.

    Valid members for NeuroMLDocument are:
    ...
    * izhikevich2007_cells (class: Izhikevich2007Cell, Optional)
            * Contents ('ids'/<objects>): ['iz2007RS0']

    * ad_ex_ia_f_cells (class: AdExIaFCell, Optional)
    * fitz_hugh_nagumo_cells (class: FitzHughNagumoCell, Optional)
    * fitz_hugh_nagumo1969_cells (class: FitzHughNagumo1969Cell, Optional)
    * pinsky_rinzel_ca3_cells (class: PinskyRinzelCA3Cell, Optional)
    * pulse_generators (class: PulseGenerator, Optional)
    * pulse_generator_dls (class: PulseGeneratorDL, Optional)
    * id (class: NmlId, Required)
            * Contents ('ids'/<objects>): IzNet

    * sine_generators (class: SineGenerator, Optional)
    ...
    * IF_curr_exp (class: IF_curr_exp, Optional)
    * exp_one_synapses (class: ExpOneSynapse, Optional)
            * Contents ('ids'/<objects>): ['syn0']

    * IF_cond_alpha (class: IF_cond_alpha, Optional)
    * exp_two_synapses (class: ExpTwoSynapse, Optional)
    ...
:::
:::

Let's also get a summary:

::: {.highlight-default .notranslate}
::: highlight
    print(nml_doc.summary())
    *******************************************************
    * NeuroMLDocument: IzNet
    *
    *  ExpOneSynapse: ['syn0']
    *  Izhikevich2007Cell: ['iz2007RS0']
    *
    *******************************************************
:::
:::

We can now declare our [[network]{.std
.std-ref}](Landing.html#schema-network){.reference .internal} with 2
[[populations]{.std
.std-ref}](Landing.html#schema-population){.reference .internal} of
these cells. Note: setting a color as a [[property]{.std
.std-ref}](Landing.html#schema-property){.reference .internal} is
optional, but is used in when we generate our plots below.

::: {.highlight-python .notranslate}
::: highlight
    # create the network: turned of validation because we will add populations next
    net = nml_doc.add("Network", id="IzNet", validate=False)

    # create the first population
    size0 = 5
    pop0 = component_factory("Population", id="IzPop0", component=iz0.id, size=size0)
    # Set optional color property. Note: used later when generating plots
    pop0.add("Property", tag="color", value="0 0 .8")
    net.add(pop0)

    # create the second population
    size1 = 5
    pop1 = component_factory("Population", id="IzPop1", component=iz0.id, size=size1)
    pop1.add("Property", tag="color", value=".8 0 0")
    net.add(pop1)
:::
:::

We can test to see if the network is now valid, since we have added the
required populations to it:

::: {.highlight-default .notranslate}
::: highlight
    net.validate()
:::
:::

This function does not return anything if the component is valid. If it
is invalid, however, it will throw a [`ValueError`{.docutils .literal
.notranslate}]{.pre}.

We can now create [[projections]{.std
.std-ref}](Landing.html#schema-projection){.reference .internal} between
the two populations based on some probability of connection. To do this,
we iterate over each post-synaptic neuron for each pre-synaptic neuron
and draw a random number between 0 and 1. If the drawn number is less
than the required probability of connection, the connection is created.

While we are iterating over all our pre-synaptic cells here, we also add
external inputs to them using [[ExplicitInputs]{.std
.std-ref}](Landing.html#schema-explicitinput){.reference .internal}
(this could have been done in a different loop, but it is convenient to
also do this here).

::: {.highlight-python .notranslate}
::: highlight
    # create a projection from one population to another
    proj = net.add(
        "Projection",
        id="proj",
        presynaptic_population=pop0.id,
        postsynaptic_population=pop1.id,
        synapse=syn0.id,
    )

    # We do two things in the loop:
    # - add pulse generator inputs to population 1 to make neurons spike
    # - create synapses between the two populations with a particular probability
    random.seed(123)
    prob_connection = 0.8
    count = 0
    for pre in range(0, size0):
        # pulse generator as explicit stimulus
        pg = nml_doc.add(
            "PulseGenerator",
            id="pg_%i" % pre,
            delay="0ms",
            duration="10000ms",
            amplitude="%f nA" % (0.1 + 0.1 * random.random()),
        )

        exp_input = net.add(
            "ExplicitInput", target="%s[%i]" % (pop0.id, pre), input=pg.id
        )

        # synapses between populations
        for post in range(0, size1):
            if random.random() <= prob_connection:
                syn = proj.add(
                    "Connection",
                    id=count,
                    pre_cell_id="../%s[%i]" % (pop0.id, pre),
                    post_cell_id="../%s[%i]" % (pop1.id, post),
                )
                count += 1
:::
:::

Let us inspect our model again to confirm that we have it set up
correctly.

::: {.highlight-default .notranslate}
::: highlight
    nml_doc.info(True)
    Please see the NeuroML standard schema documentation at https://docs.neuroml.org/Userdocs/NeuroMLv2.html for more information.

    Valid members for NeuroMLDocument are:
    * biophysical_properties (class: BiophysicalProperties, Optional)
    * SpikeSourcePoisson (class: SpikeSourcePoisson, Optional)
    * cells (class: Cell, Optional)
    * networks (class: Network, Optional)
            * Contents ('ids'/<objects>): ['IzNet']

    * cell2_ca_poolses (class: Cell2CaPools, Optional)
    ...
    * izhikevich2007_cells (class: Izhikevich2007Cell, Optional)
            * Contents ('ids'/<objects>): ['iz2007RS0']

    * ad_ex_ia_f_cells (class: AdExIaFCell, Optional)
    * fitz_hugh_nagumo_cells (class: FitzHughNagumoCell, Optional)
    * fitz_hugh_nagumo1969_cells (class: FitzHughNagumo1969Cell, Optional)
    * pinsky_rinzel_ca3_cells (class: PinskyRinzelCA3Cell, Optional)
    * pulse_generators (class: PulseGenerator, Optional)
            * Contents ('ids'/<objects>): ['pg_0', 'pg_1', 'pg_2', 'pg_3', 'pg_4']

    * pulse_generator_dls (class: PulseGeneratorDL, Optional)
    * id (class: NmlId, Required)
            * Contents ('ids'/<objects>): IzNet

    ...
    * exp_one_synapses (class: ExpOneSynapse, Optional)
            * Contents ('ids'/<objects>): ['syn0']

    * IF_cond_alpha (class: IF_cond_alpha, Optional)
    ..


    print(nml_doc.summary())
    *******************************************************
    * NeuroMLDocument: IzNet
    *
    *  ExpOneSynapse: ['syn0']
    *  Izhikevich2007Cell: ['iz2007RS0']
    *  PulseGenerator: ['pg_0', 'pg_1', 'pg_2', 'pg_3', 'pg_4']
    *
    *  Network: IzNet
    *
    *   10 cells in 2 populations
    *     Population: IzPop0 with 5 components of type iz2007RS0
    *       Properties: color=0 0 .8;
    *     Population: IzPop1 with 5 components of type iz2007RS0
    *       Properties: color=.8 0 0;
    *
    *   20 connections in 1 projections
    *     Projection: proj from IzPop0 to IzPop1, synapse: syn0
    *       20 connections: [(Connection 0: 0 -> 0), ...]
    *
    *   0 inputs in 0 input lists
    *
    *   5 explicit inputs (outside of input lists)
    *     Explicit Input of type pg_0 to IzPop0(cell 0), destination: unspecified
    *     Explicit Input of type pg_1 to IzPop0(cell 1), destination: unspecified
    *     Explicit Input of type pg_2 to IzPop0(cell 2), destination: unspecified
    *     Explicit Input of type pg_3 to IzPop0(cell 3), destination: unspecified
    *     Explicit Input of type pg_4 to IzPop0(cell 4), destination: unspecified
    *
    *******************************************************
:::
:::

We can now save and validate our model.

::: {.highlight-python .notranslate}
::: highlight
    # write model to file and validate
    nml_file = "izhikevich2007_network.nml"
    writers.NeuroMLWriter.write(nml_doc, nml_file)

    print("Written network file to: " + nml_file)
    pynml.validate_neuroml2(nml_file)
:::
:::

::: {#the-generated-neuroml-model .section}
##### The generated NeuroML model[\#](#the-generated-neuroml-model "Link to this heading"){.headerlink}

Let us take a look at the generated NeuroML model

::: {.highlight-xml .notranslate}
::: highlight
    <neuroml xmlns="http://www.neuroml.org/schema/neuroml2"  xmlns:xs="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.neuroml.org/schema/neuroml2 https://raw.github.com/NeuroML/NeuroML2/development/Schemas/NeuroML2/NeuroML_v2.3.xsd" id="IzNet">
        <expOneSynapse id="syn0" gbase="65nS" erev="0mV" tauDecay="3ms"/>
        <izhikevich2007Cell id="iz2007RS0" C="100pF" v0="-60mV" k="0.7nS_per_mV" vr="-60mV" vt="-40mV" vpeak="35mV" a="0.03per_ms" b="-2nS" c="-50.0mV" d="100pA"/>
        <pulseGenerator id="pg_0" delay="0ms" duration="10000ms" amplitude="0.105236 nA"/>
        <pulseGenerator id="pg_1" delay="0ms" duration="10000ms" amplitude="0.153620 nA"/>
        <pulseGenerator id="pg_2" delay="0ms" duration="10000ms" amplitude="0.124516 nA"/>
        <pulseGenerator id="pg_3" delay="0ms" duration="10000ms" amplitude="0.131546 nA"/>
        <pulseGenerator id="pg_4" delay="0ms" duration="10000ms" amplitude="0.102124 nA"/>
        <network id="IzNet">
            <population id="IzPop0" component="iz2007RS0" size="5" type="population">
                <property tag="color" value="0 0 .8"/>
            </population>
            <population id="IzPop1" component="iz2007RS0" size="5" type="population">
                <property tag="color" value=".8 0 0"/>
            </population>
            <projection id="proj" presynapticPopulation="IzPop0" postsynapticPopulation="IzPop1" synapse="syn0">
                <connection id="0" preCellId="../IzPop0[0]" postCellId="../IzPop1[0]"/>
                <connection id="1" preCellId="../IzPop0[0]" postCellId="../IzPop1[1]"/>
                <connection id="2" preCellId="../IzPop0[0]" postCellId="../IzPop1[2]"/>
                <connection id="3" preCellId="../IzPop0[0]" postCellId="../IzPop1[4]"/>
                <connection id="4" preCellId="../IzPop0[1]" postCellId="../IzPop1[0]"/>
                <connection id="5" preCellId="../IzPop0[1]" postCellId="../IzPop1[2]"/>
                <connection id="6" preCellId="../IzPop0[1]" postCellId="../IzPop1[3]"/>
                <connection id="7" preCellId="../IzPop0[1]" postCellId="../IzPop1[4]"/>
                <connection id="8" preCellId="../IzPop0[2]" postCellId="../IzPop1[0]"/>
                <connection id="9" preCellId="../IzPop0[2]" postCellId="../IzPop1[1]"/>
                <connection id="10" preCellId="../IzPop0[2]" postCellId="../IzPop1[2]"/>
                <connection id="11" preCellId="../IzPop0[2]" postCellId="../IzPop1[3]"/>
                <connection id="12" preCellId="../IzPop0[2]" postCellId="../IzPop1[4]"/>
                <connection id="13" preCellId="../IzPop0[3]" postCellId="../IzPop1[0]"/>
                <connection id="14" preCellId="../IzPop0[3]" postCellId="../IzPop1[2]"/>
                <connection id="15" preCellId="../IzPop0[3]" postCellId="../IzPop1[3]"/>
                <connection id="16" preCellId="../IzPop0[3]" postCellId="../IzPop1[4]"/>
                <connection id="17" preCellId="../IzPop0[4]" postCellId="../IzPop1[1]"/>
                <connection id="18" preCellId="../IzPop0[4]" postCellId="../IzPop1[2]"/>
                <connection id="19" preCellId="../IzPop0[4]" postCellId="../IzPop1[4]"/>
            </projection>
            <explicitInput target="IzPop0[0]" input="pg_0"/>
            <explicitInput target="IzPop0[1]" input="pg_1"/>
            <explicitInput target="IzPop0[2]" input="pg_2"/>
            <explicitInput target="IzPop0[3]" input="pg_3"/>
            <explicitInput target="IzPop0[4]" input="pg_4"/>
        </network>
    </neuroml>
:::
:::

It should now be easy to see how the model is clearly declared in the
NeuroML file. Observe how entities are referenced in NeuroML depending
on their location in the document architecture. Here, [[population]{.std
.std-ref}](Landing.html#schema-population){.reference .internal} and
[[projection]{.std .std-ref}](Landing.html#schema-projection){.reference
.internal} are at the same level. The synaptic connections using the
[[connection]{.std .std-ref}](Landing.html#schema-connection){.reference
.internal} tag are at the next level. So, in the [[connection]{.std
.std-ref}](Landing.html#schema-connection){.reference .internal} tags,
populations are to be referred to as [`../`{.docutils .literal
.notranslate}]{.pre} which indicates the previous level. The
[[explicitinput]{.std
.std-ref}](Landing.html#schema-explicitinput){.reference .internal} tag
is at the same level as the [[population]{.std
.std-ref}](Landing.html#schema-population){.reference .internal} and
[[projection]{.std .std-ref}](Landing.html#schema-projection){.reference
.internal} tags, so we do *not* need to use [`../`{.docutils .literal
.notranslate}]{.pre} here to reference them.

Another point worth noting here is that because we have defined a
population of the same components by specifying a size rather than by
individually adding components to it, we can refer to the entities of
the population using the common [`[..]`{.docutils .literal
.notranslate}]{.pre} index operator.

The advantage of such a declarative format is that we can also easily
get information on our model from the NeuroML file. Similar to the
[`summary()`{.docutils .literal .notranslate}]{.pre} function that we
have used so far, [[pyNeuroML]{.std
.std-ref}](Landing.html#pyneuroml){.reference .internal} also includes
the helper [`pynml-summary`{.docutils .literal .notranslate}]{.pre}
script that can be used to get summaries of NeuroML models from their
NeuroML files:

::: {.highlight-console .notranslate}
::: highlight
    $ pynml-summary izhikevich2007_network.nml
    *******************************************************
    * NeuroMLDocument: IzNet
    *
    *   ExpOneSynapse: ['syn0']
    *   Izhikevich2007Cell: ['iz2007RS0']
    *   PulseGenerator: ['pulseGen_0', 'pulseGen_1', 'pulseGen_2', 'pulseGen_3', 'pulseGen_4']
    *
    *  Network: IzNet
    *
    *   10 cells in 2 populations
    *     Population: IzPop0 with 5 components of type iz2007RS0
    *     Population: IzPop1 with 5 components of type iz2007RS0
    *
    *   20 connections in 1 projections
    *     Projection: proj from IzPop0 to IzPop1, synapse: syn0
    *       20 connections: [(Connection 0: 0 -> 0), ...]
    *
    *   0 inputs in 0 input lists
    *
    *   5 explicit inputs (outside of input lists)
    *     Explicit Input of type pulseGen_0 to IzPop0(cell 0), destination: unspecified
    *     Explicit Input of type pulseGen_1 to IzPop0(cell 1), destination: unspecified
    *     Explicit Input of type pulseGen_2 to IzPop0(cell 2), destination: unspecified
    *     Explicit Input of type pulseGen_3 to IzPop0(cell 3), destination: unspecified
    *     Explicit Input of type pulseGen_4 to IzPop0(cell 4), destination: unspecified
    *
    *******************************************************
:::
:::

We can also generate a graphical summary of our model using
[`pynml`{.docutils .literal .notranslate}]{.pre} from [[pyNeuroML]{.std
.std-ref}](Landing.html#pyneuroml){.reference .internal}:

::: {.highlight-console .notranslate}
::: highlight
    $ pynml izhikevich2007_network.nml -graph 3
:::
:::

This generates the following model summary diagram:

<figure id="id2" class="align-center">
<a href="_images/IzNet.gv.png"
class="reference internal image-reference"><img
src="_images/IzNet.gv.png" style="width: 756.5px; height: 155.0px;"
alt="Model summary graph generated using pynml and the dot tool." /></a>
<figcaption><p><span class="caption-number">Fig. 10 </span><span
class="caption-text">A summary graph of the model generated using pynml
and the dot tool.</span><a href="#id2" class="headerlink"
title="Link to this image">#</a></p></figcaption>
</figure>

Other options for [`pynml`{.docutils .literal .notranslate}]{.pre}
produce other views, e.g individual connections:

::: {.highlight-console .notranslate}
::: highlight
    $ pynml izhikevich2007_network.nml -graph -1
:::
:::

<figure id="id3" class="align-center">
<a href="_images/IzNet-1.gv.png"
class="reference internal image-reference"><img
src="_images/IzNet-1.gv.png"
style="width: 485.79999999999995px; height: 108.5px;"
alt="Model summary graph showing individual connections between cells in the populations." /></a>
<figcaption><p><span class="caption-number">Fig. 11 </span><span
class="caption-text">Model summary graph showing individual connections
between cells in the populations.</span><a href="#id3"
class="headerlink" title="Link to this image">#</a></p></figcaption>
</figure>

In our very simple network here, neurons do not have morphologies and
are not distributed in space. In later examples, however, we will also
see how summary figures of the network that show the morphologies,
locations of different layers and neurons, and so on can also be
generated using the NeuroML tools.
:::
:::

::: {#simulating-the-model .section}
[]{#userdocs-gettingstarted-izhikevichnetwork-simulating}

#### Simulating the model[\#](#simulating-the-model "Link to this heading"){.headerlink}

Now that we have our model set up, we can proceed to simulating it. We
create our simulation, and setup the information we want to record from
it.

::: {.highlight-python .notranslate}
::: highlight
    # Create simulation, and record data
    simulation_id = "example_izhikevich2007network_sim"
    simulation = LEMSSimulation(
        sim_id=simulation_id, duration=1000, dt=0.1, simulation_seed=123
    )
    simulation.assign_simulation_target(net.id)
    simulation.include_neuroml2_file(nml_file)

    simulation.create_event_output_file(
        "pop0", "%s.0.spikes.dat" % simulation_id, format="ID_TIME"
    )
    for pre in range(0, size0):
        simulation.add_selection_to_event_output_file(
            "pop0", pre, "IzPop0[{}]".format(pre), "spike"
        )

    simulation.create_event_output_file(
        "pop1", "%s.1.spikes.dat" % simulation_id, format="ID_TIME"
    )
    for pre in range(0, size1):
        simulation.add_selection_to_event_output_file(
            "pop1", pre, "IzPop1[{}]".format(pre), "spike"
        )

    lems_simulation_file = simulation.save_to_file()
:::
:::

The generated LEMS file is here:

::: {.highlight-xml .notranslate}
::: highlight
    <Lems>

        <!-- 

            This LEMS file has been automatically generated using PyNeuroML v1.1.13 (libNeuroML v0.5.8)

         -->

        <!-- Specify which component to run -->
        <Target component="example_izhikevich2007network_sim"/>

        <!-- Include core NeuroML2 ComponentType definitions -->
        <Include file="Cells.xml"/>
        <Include file="Networks.xml"/>
        <Include file="Simulation.xml"/>

        <Include file="izhikevich2007_network.nml"/>

        <Simulation id="example_izhikevich2007network_sim" length="1000ms" step="0.1ms" target="IzNet" seed="123">  <!-- Note seed: ensures same random numbers used every run -->
            <EventOutputFile id="pop0" fileName="example_izhikevich2007network_sim.0.spikes.dat" format="ID_TIME">
                <EventSelection id="0" select="IzPop0[0]" eventPort="spike"/>
                <EventSelection id="1" select="IzPop0[1]" eventPort="spike"/>
                <EventSelection id="2" select="IzPop0[2]" eventPort="spike"/>
                <EventSelection id="3" select="IzPop0[3]" eventPort="spike"/>
                <EventSelection id="4" select="IzPop0[4]" eventPort="spike"/>
            </EventOutputFile>

            <EventOutputFile id="pop1" fileName="example_izhikevich2007network_sim.1.spikes.dat" format="ID_TIME">
                <EventSelection id="0" select="IzPop1[0]" eventPort="spike"/>
                <EventSelection id="1" select="IzPop1[1]" eventPort="spike"/>
                <EventSelection id="2" select="IzPop1[2]" eventPort="spike"/>
                <EventSelection id="3" select="IzPop1[3]" eventPort="spike"/>
                <EventSelection id="4" select="IzPop1[4]" eventPort="spike"/>
            </EventOutputFile>

        </Simulation>

    </Lems>
:::
:::

Where we had generated a graphical summary of the model before, we can
now also generate graphical summaries of the simulation using
[`pynml`{.docutils .literal .notranslate}]{.pre} and the
[`-lems-graph`{.docutils .literal .notranslate}]{.pre} option. This
dives deeper into the LEMS definition of the cells, showing more of the
underlying dynamics of the components:

::: {.highlight-console .notranslate}
::: highlight
    $ pynml LEMS_example_izhikevich2007network_sim.xml -lems-graph
:::
:::

Here is the generated summary graph:

<figure id="id4" class="align-center">
<a href="_images/LEMS_example_izhikevich2007network_sim.png"
class="reference internal image-reference"><img
src="_images/LEMS_example_izhikevich2007network_sim.png"
style="width: 365.5px; height: 760.0px;"
alt="Model summary graph generated using pynml." /></a>
<figcaption><p><span class="caption-number">Fig. 12 </span><span
class="caption-text">A summary graph of the model generated using pynml
-lems-graph.</span><a href="#id4" class="headerlink"
title="Link to this image">#</a></p></figcaption>
</figure>

It shows a top-down breakdown of the simulation: from the network, to
the populations, to the cell types, leading up to the components that
these cells are made of (more on Components later). Let us add the
necessary code to run our simulation, this time using the well known
NEURON simulator:

::: {.highlight-python .notranslate}
::: highlight
    # Run the simulation
    pynml.run_lems_with_jneuroml_neuron(
        lems_simulation_file, max_memory="2G", nogui=True, plot=False
    )
:::
:::
:::

::: {#plotting-recorded-spike-times .section}
[]{#userdocs-gettingstarted-izhikevichnetwork-plotting}

#### Plotting recorded spike times[\#](#plotting-recorded-spike-times "Link to this heading"){.headerlink}

To analyse the outputs from the simulation, we can again plot the
information we recorded. In the previous example, we had recorded and
plotted the membrane potentials from our cell. Here, we have recorded
the spike times. So let us plot them to generate our figure:

::: {.highlight-python .notranslate}
::: highlight
    # Load the data from the file and plot the spike times
    # using the pynml generate_plot utility function.
    data_array_0 = np.loadtxt("%s.0.spikes.dat" % simulation_id)
    data_array_1 = np.loadtxt("%s.1.spikes.dat" % simulation_id)
    times_0 = data_array_0[:, 1]
    times_1 = data_array_1[:, 1]
    ids_0 = data_array_0[:, 0]
    ids_1 = [id + size0 for id in data_array_1[:, 0]]
    pynml.generate_plot(
        [times_0, times_1],
        [ids_0, ids_1],
        "Spike times",
        show_plot_already=False,
        save_figure_to="%s-spikes.png" % simulation_id,
        xaxis="time (s)",
        yaxis="cell ID",
        colors=["b", "r"],
        linewidths=["0", "0"],
        markers=[".", "."],
    )
:::
:::

Observe how we are using the same [`generate_plot`{.docutils .literal
.notranslate}]{.pre} utility function as before: it is general enough to
plot different recorded quantities. Under the hood, it passes this
information to Python's Matplotlib library. This produces the raster
plot shown at the top of the page.

This concludes our second example. Here, we have seen how to create,
simulate, and record from a simple two population network of single
compartment point neurons. The next section is an interactive notebook
that you can use to play with this example. After that we will move on
to the next example: a neuron model using Hodgkin Huxley style ion
channels.
:::
:::

[]{#document-Userdocs/NML2_examples/IzhikevichNetwork}

::: {#interactive-two-population-network-example .section .tex2jax_ignore .mathjax_ignore}
### Interactive two population network example[\#](#interactive-two-population-network-example "Link to this heading"){.headerlink}

To run this interactive Jupyter Notebook, please click on the rocket
icon 🚀 in the top panel. For more information, please see [[how to use
this documentation]{.std
.std-ref}](Landing.html#userdocs-usage-jupyterbooks){.reference
.internal}. Please uncomment the line below if you use the Google Colab.
(It does not include these packages by default).

::: {.cell .docutils .container}
::: {.cell_input .docutils .container}
::: {.highlight-ipython3 .notranslate}
::: highlight
    #%pip install pyneuroml neuromllite NEURON
:::
:::
:::
:::

::: {.cell .docutils .container}
::: {.cell_input .docutils .container}
::: {.highlight-ipython3 .notranslate}
::: highlight
    #!/usr/bin/env python3
    """
    Create a simple network with two populations.
    """

    import random
    import numpy as np

    from neuroml.utils import component_factory
    from pyneuroml import pynml
    from pyneuroml.lems import LEMSSimulation
    import neuroml.writers as writers
:::
:::
:::

::: {.cell_output .docutils .container}
::: {.output .traceback .highlight-ipythontb .notranslate}
::: highlight
    ---------------------------------------------------------------------------
    ModuleNotFoundError                       Traceback (most recent call last)
    Cell In[2], line 9
          6 import random
          7 import numpy as np
    ----> 9 from neuroml.utils import component_factory
         10 from pyneuroml import pynml
         11 from pyneuroml.lems import LEMSSimulation

    ModuleNotFoundError: No module named 'neuroml'
:::
:::
:::
:::

::: {#declaring-the-neuroml-model .section}
#### Declaring the NeuroML model[\#](#declaring-the-neuroml-model "Link to this heading"){.headerlink}

::: {#create-a-neuroml-document .section}
##### Create a NeuroML document[\#](#create-a-neuroml-document "Link to this heading"){.headerlink}

::: {.cell .docutils .container}
::: {.cell_input .docutils .container}
::: {.highlight-ipython3 .notranslate}
::: highlight
    nml_doc = component_factory("NeuroMLDocument", id="IzNet")
:::
:::
:::
:::
:::

::: {#declare-the-izhikevich-cell-and-add-it-to-the-model-document .section}
##### Declare the Izhikevich cell and add it to the model document[\#](#declare-the-izhikevich-cell-and-add-it-to-the-model-document "Link to this heading"){.headerlink}

::: {.cell .docutils .container}
::: {.cell_input .docutils .container}
::: {.highlight-ipython3 .notranslate}
::: highlight
    iz0 = nml_doc.add(
        "Izhikevich2007Cell",
        id="iz2007RS0",
        v0="-60mV",
        C="100pF",
        k="0.7nS_per_mV",
        vr="-60mV",
        vt="-40mV",
        vpeak="35mV",
        a="0.03per_ms",
        b="-2nS",
        c="-50.0mV",
        d="100pA",
    )
    # Inspect the component, also show all members:
    iz0.info(True)
:::
:::
:::

::: {.cell_output .docutils .container}
::: {.output .stream .highlight-myst-ansi .notranslate}
::: highlight
    Izhikevich2007Cell -- Cell based on the modified Izhikevich model in Izhikevich 2007, Dynamical systems in neuroscience, MIT Press

    Please see the NeuroML standard schema documentation at https://docs.neuroml.org/Userdocs/NeuroMLv2.html for more information.

    Valid members for Izhikevich2007Cell are:
    * a (class: Nml2Quantity_pertime, Required)
        * Contents ('ids'/<objects>): 0.03per_ms

    * C (class: Nml2Quantity_capacitance, Required)
        * Contents ('ids'/<objects>): 100pF

    * annotation (class: Annotation, Optional)
    * b (class: Nml2Quantity_conductance, Required)
        * Contents ('ids'/<objects>): -2nS

    * metaid (class: MetaId, Optional)
    * c (class: Nml2Quantity_voltage, Required)
        * Contents ('ids'/<objects>): -50.0mV

    * d (class: Nml2Quantity_current, Required)
        * Contents ('ids'/<objects>): 100pA

    * neuro_lex_id (class: NeuroLexId, Optional)
    * v0 (class: Nml2Quantity_voltage, Required)
        * Contents ('ids'/<objects>): -60mV

    * properties (class: Property, Optional)
    * k (class: Nml2Quantity_conductancePerVoltage, Required)
        * Contents ('ids'/<objects>): 0.7nS_per_mV

    * notes (class: xs:string, Optional)
    * vr (class: Nml2Quantity_voltage, Required)
        * Contents ('ids'/<objects>): -60mV

    * vt (class: Nml2Quantity_voltage, Required)
        * Contents ('ids'/<objects>): -40mV

    * vpeak (class: Nml2Quantity_voltage, Required)
        * Contents ('ids'/<objects>): 35mV

    * id (class: NmlId, Required)
        * Contents ('ids'/<objects>): iz2007RS0
:::
:::

::: {.output .text_plain .highlight-myst-ansi .notranslate}
::: highlight
    "Izhikevich2007Cell -- Cell based on the modified Izhikevich model in Izhikevich 2007, Dynamical systems in neuroscience, MIT Press\n\nPlease see the NeuroML standard schema documentation at https://docs.neuroml.org/Userdocs/NeuroMLv2.html for more information.\n\nValid members for Izhikevich2007Cell are:\n* a (class: Nml2Quantity_pertime, Required)\n\t* Contents ('ids'/<objects>): 0.03per_ms\n\n* C (class: Nml2Quantity_capacitance, Required)\n\t* Contents ('ids'/<objects>): 100pF\n\n* annotation (class: Annotation, Optional)\n* b (class: Nml2Quantity_conductance, Required)\n\t* Contents ('ids'/<objects>): -2nS\n\n* metaid (class: MetaId, Optional)\n* c (class: Nml2Quantity_voltage, Required)\n\t* Contents ('ids'/<objects>): -50.0mV\n\n* d (class: Nml2Quantity_current, Required)\n\t* Contents ('ids'/<objects>): 100pA\n\n* neuro_lex_id (class: NeuroLexId, Optional)\n* v0 (class: Nml2Quantity_voltage, Required)\n\t* Contents ('ids'/<objects>): -60mV\n\n* properties (class: Property, Optional)\n* k (class: Nml2Quantity_conductancePerVoltage, Required)\n\t* Contents ('ids'/<objects>): 0.7nS_per_mV\n\n* notes (class: xs:string, Optional)\n* vr (class: Nml2Quantity_voltage, Required)\n\t* Contents ('ids'/<objects>): -60mV\n\n* vt (class: Nml2Quantity_voltage, Required)\n\t* Contents ('ids'/<objects>): -40mV\n\n* vpeak (class: Nml2Quantity_voltage, Required)\n\t* Contents ('ids'/<objects>): 35mV\n\n* id (class: NmlId, Required)\n\t* Contents ('ids'/<objects>): iz2007RS0\n\n"
:::
:::
:::
:::
:::

::: {#declare-the-synapse-and-add-it-to-the-model-document .section}
##### Declare the Synapse and add it to the model document[\#](#declare-the-synapse-and-add-it-to-the-model-document "Link to this heading"){.headerlink}

::: {.cell .docutils .container}
::: {.cell_input .docutils .container}
::: {.highlight-ipython3 .notranslate}
::: highlight
    syn0 = nml_doc.add(
        "ExpOneSynapse", id="syn0", gbase="65nS", erev="0mV", tau_decay="3ms"
    )
:::
:::
:::
:::

::: {.cell .docutils .container}
::: {.cell_input .docutils .container}
::: {.highlight-ipython3 .notranslate}
::: highlight
    nml_doc.info(True)
:::
:::
:::

::: {.cell_output .docutils .container}
::: {.output .stream .highlight-myst-ansi .notranslate}
::: highlight
    Please see the NeuroML standard schema documentation at https://docs.neuroml.org/Userdocs/NeuroMLv2.html for more information.

    Valid members for NeuroMLDocument are:
    * compound_inputs (class: CompoundInput, Optional)
    * compound_input_dls (class: CompoundInputDL, Optional)
    * includes (class: IncludeType, Optional)
    * voltage_clamps (class: VoltageClamp, Optional)
    * extracellular_properties (class: ExtracellularProperties, Optional)
    * voltage_clamp_triples (class: VoltageClampTriple, Optional)
    * intracellular_properties (class: IntracellularProperties, Optional)
    * spike_arrays (class: SpikeArray, Optional)
    * morphology (class: Morphology, Optional)
    * timed_synaptic_inputs (class: TimedSynapticInput, Optional)
    * ion_channel (class: IonChannel, Optional)
    * spike_generators (class: SpikeGenerator, Optional)
    * ion_channel_hhs (class: IonChannelHH, Optional)
    * spike_generator_randoms (class: SpikeGeneratorRandom, Optional)
    * ion_channel_v_shifts (class: IonChannelVShift, Optional)
    * spike_generator_poissons (class: SpikeGeneratorPoisson, Optional)
    * ion_channel_kses (class: IonChannelKS, Optional)
    * spike_generator_ref_poissons (class: SpikeGeneratorRefPoisson, Optional)
    * decaying_pool_concentration_models (class: DecayingPoolConcentrationModel, Optional)
    * poisson_firing_synapses (class: PoissonFiringSynapse, Optional)
    * fixed_factor_concentration_models (class: FixedFactorConcentrationModel, Optional)
    * transient_poisson_firing_synapses (class: TransientPoissonFiringSynapse, Optional)
    * alpha_current_synapses (class: AlphaCurrentSynapse, Optional)
    * IF_curr_alpha (class: IF_curr_alpha, Optional)
    * alpha_synapses (class: AlphaSynapse, Optional)
    * IF_curr_exp (class: IF_curr_exp, Optional)
    * exp_one_synapses (class: ExpOneSynapse, Optional)
        * Contents ('ids'/<objects>): ['syn0']

    * IF_cond_alpha (class: IF_cond_alpha, Optional)
    * exp_two_synapses (class: ExpTwoSynapse, Optional)
    * IF_cond_exp (class: IF_cond_exp, Optional)
    * exp_three_synapses (class: ExpThreeSynapse, Optional)
    * EIF_cond_exp_isfa_ista (class: EIF_cond_exp_isfa_ista, Optional)
    * blocking_plastic_synapses (class: BlockingPlasticSynapse, Optional)
    * EIF_cond_alpha_isfa_ista (class: EIF_cond_alpha_isfa_ista, Optional)
    * double_synapses (class: DoubleSynapse, Optional)
    * HH_cond_exp (class: HH_cond_exp, Optional)
    * gap_junctions (class: GapJunction, Optional)
    * exp_cond_synapses (class: ExpCondSynapse, Optional)
    * silent_synapses (class: SilentSynapse, Optional)
    * alpha_cond_synapses (class: AlphaCondSynapse, Optional)
    * linear_graded_synapses (class: LinearGradedSynapse, Optional)
    * exp_curr_synapses (class: ExpCurrSynapse, Optional)
    * graded_synapses (class: GradedSynapse, Optional)
    * alpha_curr_synapses (class: AlphaCurrSynapse, Optional)
    * annotation (class: Annotation, Optional)
    * biophysical_properties (class: BiophysicalProperties, Optional)
    * SpikeSourcePoisson (class: SpikeSourcePoisson, Optional)
    * cells (class: Cell, Optional)
    * networks (class: Network, Optional)
    * cell2_ca_poolses (class: Cell2CaPools, Optional)
    * ComponentType (class: ComponentType, Optional)
    * base_cells (class: BaseCell, Optional)
    * iaf_tau_cells (class: IafTauCell, Optional)
    * properties (class: Property, Optional)
    * iaf_tau_ref_cells (class: IafTauRefCell, Optional)
    * notes (class: xs:string, Optional)
    * iaf_cells (class: IafCell, Optional)
    * metaid (class: MetaId, Optional)
    * iaf_ref_cells (class: IafRefCell, Optional)
    * izhikevich_cells (class: IzhikevichCell, Optional)
    * izhikevich2007_cells (class: Izhikevich2007Cell, Optional)
        * Contents ('ids'/<objects>): ['iz2007RS0']

    * ad_ex_ia_f_cells (class: AdExIaFCell, Optional)
    * fitz_hugh_nagumo_cells (class: FitzHughNagumoCell, Optional)
    * fitz_hugh_nagumo1969_cells (class: FitzHughNagumo1969Cell, Optional)
    * pinsky_rinzel_ca3_cells (class: PinskyRinzelCA3Cell, Optional)
    * pulse_generators (class: PulseGenerator, Optional)
    * pulse_generator_dls (class: PulseGeneratorDL, Optional)
    * sine_generators (class: SineGenerator, Optional)
    * sine_generator_dls (class: SineGeneratorDL, Optional)
    * ramp_generators (class: RampGenerator, Optional)
    * ramp_generator_dls (class: RampGeneratorDL, Optional)
    * id (class: NmlId, Required)
        * Contents ('ids'/<objects>): IzNet
:::
:::

::: {.output .text_plain .highlight-myst-ansi .notranslate}
::: highlight
    "Please see the NeuroML standard schema documentation at https://docs.neuroml.org/Userdocs/NeuroMLv2.html for more information.\n\nValid members for NeuroMLDocument are:\n* compound_inputs (class: CompoundInput, Optional)\n* compound_input_dls (class: CompoundInputDL, Optional)\n* includes (class: IncludeType, Optional)\n* voltage_clamps (class: VoltageClamp, Optional)\n* extracellular_properties (class: ExtracellularProperties, Optional)\n* voltage_clamp_triples (class: VoltageClampTriple, Optional)\n* intracellular_properties (class: IntracellularProperties, Optional)\n* spike_arrays (class: SpikeArray, Optional)\n* morphology (class: Morphology, Optional)\n* timed_synaptic_inputs (class: TimedSynapticInput, Optional)\n* ion_channel (class: IonChannel, Optional)\n* spike_generators (class: SpikeGenerator, Optional)\n* ion_channel_hhs (class: IonChannelHH, Optional)\n* spike_generator_randoms (class: SpikeGeneratorRandom, Optional)\n* ion_channel_v_shifts (class: IonChannelVShift, Optional)\n* spike_generator_poissons (class: SpikeGeneratorPoisson, Optional)\n* ion_channel_kses (class: IonChannelKS, Optional)\n* spike_generator_ref_poissons (class: SpikeGeneratorRefPoisson, Optional)\n* decaying_pool_concentration_models (class: DecayingPoolConcentrationModel, Optional)\n* poisson_firing_synapses (class: PoissonFiringSynapse, Optional)\n* fixed_factor_concentration_models (class: FixedFactorConcentrationModel, Optional)\n* transient_poisson_firing_synapses (class: TransientPoissonFiringSynapse, Optional)\n* alpha_current_synapses (class: AlphaCurrentSynapse, Optional)\n* IF_curr_alpha (class: IF_curr_alpha, Optional)\n* alpha_synapses (class: AlphaSynapse, Optional)\n* IF_curr_exp (class: IF_curr_exp, Optional)\n* exp_one_synapses (class: ExpOneSynapse, Optional)\n\t* Contents ('ids'/<objects>): ['syn0']\n\n* IF_cond_alpha (class: IF_cond_alpha, Optional)\n* exp_two_synapses (class: ExpTwoSynapse, Optional)\n* IF_cond_exp (class: IF_cond_exp, Optional)\n* exp_three_synapses (class: ExpThreeSynapse, Optional)\n* EIF_cond_exp_isfa_ista (class: EIF_cond_exp_isfa_ista, Optional)\n* blocking_plastic_synapses (class: BlockingPlasticSynapse, Optional)\n* EIF_cond_alpha_isfa_ista (class: EIF_cond_alpha_isfa_ista, Optional)\n* double_synapses (class: DoubleSynapse, Optional)\n* HH_cond_exp (class: HH_cond_exp, Optional)\n* gap_junctions (class: GapJunction, Optional)\n* exp_cond_synapses (class: ExpCondSynapse, Optional)\n* silent_synapses (class: SilentSynapse, Optional)\n* alpha_cond_synapses (class: AlphaCondSynapse, Optional)\n* linear_graded_synapses (class: LinearGradedSynapse, Optional)\n* exp_curr_synapses (class: ExpCurrSynapse, Optional)\n* graded_synapses (class: GradedSynapse, Optional)\n* alpha_curr_synapses (class: AlphaCurrSynapse, Optional)\n* annotation (class: Annotation, Optional)\n* biophysical_properties (class: BiophysicalProperties, Optional)\n* SpikeSourcePoisson (class: SpikeSourcePoisson, Optional)\n* cells (class: Cell, Optional)\n* networks (class: Network, Optional)\n* cell2_ca_poolses (class: Cell2CaPools, Optional)\n* ComponentType (class: ComponentType, Optional)\n* base_cells (class: BaseCell, Optional)\n* iaf_tau_cells (class: IafTauCell, Optional)\n* properties (class: Property, Optional)\n* iaf_tau_ref_cells (class: IafTauRefCell, Optional)\n* notes (class: xs:string, Optional)\n* iaf_cells (class: IafCell, Optional)\n* metaid (class: MetaId, Optional)\n* iaf_ref_cells (class: IafRefCell, Optional)\n* izhikevich_cells (class: IzhikevichCell, Optional)\n* izhikevich2007_cells (class: Izhikevich2007Cell, Optional)\n\t* Contents ('ids'/<objects>): ['iz2007RS0']\n\n* ad_ex_ia_f_cells (class: AdExIaFCell, Optional)\n* fitz_hugh_nagumo_cells (class: FitzHughNagumoCell, Optional)\n* fitz_hugh_nagumo1969_cells (class: FitzHughNagumo1969Cell, Optional)\n* pinsky_rinzel_ca3_cells (class: PinskyRinzelCA3Cell, Optional)\n* pulse_generators (class: PulseGenerator, Optional)\n* pulse_generator_dls (class: PulseGeneratorDL, Optional)\n* sine_generators (class: SineGenerator, Optional)\n* sine_generator_dls (class: SineGeneratorDL, Optional)\n* ramp_generators (class: RampGenerator, Optional)\n* ramp_generator_dls (class: RampGeneratorDL, Optional)\n* id (class: NmlId, Required)\n\t* Contents ('ids'/<objects>): IzNet\n\n"
:::
:::
:::
:::

::: {.cell .docutils .container}
::: {.cell_input .docutils .container}
::: {.highlight-ipython3 .notranslate}
::: highlight
    print(nml_doc.summary())
:::
:::
:::

::: {.cell_output .docutils .container}
::: {.output .stream .highlight-myst-ansi .notranslate}
::: highlight
    *******************************************************
    * NeuroMLDocument: IzNet
    *
    *  ExpOneSynapse: ['syn0']
    *  Izhikevich2007Cell: ['iz2007RS0']
    *
    *******************************************************
:::
:::
:::
:::
:::

::: {#declare-a-network-and-add-it-to-the-model-document .section}
##### Declare a Network and add it to the model document[\#](#declare-a-network-and-add-it-to-the-model-document "Link to this heading"){.headerlink}

::: {.cell .docutils .container}
::: {.cell_input .docutils .container}
::: {.highlight-ipython3 .notranslate}
::: highlight
    net = nml_doc.add("Network", id="IzNet", validate=False)
:::
:::
:::
:::
:::

::: {#create-two-populations .section}
##### Create two populations[\#](#create-two-populations "Link to this heading"){.headerlink}

::: {.cell .docutils .container}
::: {.cell_input .docutils .container}
::: {.highlight-ipython3 .notranslate}
::: highlight
    # create the first population
    size0 = 5
    pop0 = component_factory("Population", id="IzPop0", component=iz0.id, size=size0, type="population")
    # Set optional color property. Note: used later when generating plots
    pop0.add("Property", tag="color", value="0 0 .8")
    net.add(pop0)

    # create the second population
    size1 = 5
    pop1 = component_factory("Population", id="IzPop1", component=iz0.id, size=size1, type="population")
    pop1.add("Property", tag="color", value=".8 0 0")
    net.add(pop1)
:::
:::
:::

::: {.cell_output .docutils .container}
::: {.output .text_plain .highlight-myst-ansi .notranslate}
::: highlight
    <neuroml.nml.nml.Population at 0x7f53ace425c0>
:::
:::
:::
:::

::: {.cell .docutils .container}
::: {.cell_input .docutils .container}
::: {.highlight-ipython3 .notranslate}
::: highlight
    net.validate()
:::
:::
:::
:::
:::

::: {#declare-projections .section}
##### Declare projections[\#](#declare-projections "Link to this heading"){.headerlink}

::: {.cell .docutils .container}
::: {.cell_input .docutils .container}
::: {.highlight-ipython3 .notranslate}
::: highlight
    # create a projection from one population to another
    proj = net.add(
        "Projection",
        id="proj",
        presynaptic_population=pop0.id,
        postsynaptic_population=pop1.id,
        synapse=syn0.id,
    )
:::
:::
:::
:::
:::

::: {#add-the-projections-between-populations-and-the-external-inputs .section}
##### Add the projections between populations and the external inputs[\#](#add-the-projections-between-populations-and-the-external-inputs "Link to this heading"){.headerlink}

::: {.cell .docutils .container}
::: {.cell_input .docutils .container}
::: {.highlight-ipython3 .notranslate}
::: highlight
    random.seed(123)
    prob_connection = 0.8
    count = 0
    for pre in range(0, size0):
        # pulse generator as explicit stimulus
        pg = nml_doc.add(
            "PulseGenerator",
            id="pg_%i" % pre,
            delay="0ms",
            duration="10000ms",
            amplitude="%f nA" % (0.1 + 0.1 * random.random()),
        )

        exp_input = net.add(
            "ExplicitInput", target="%s[%i]" % (pop0.id, pre), input=pg.id
        )

        # synapses between populations
        for post in range(0, size1):
            if random.random() <= prob_connection:
                syn = proj.add(
                    "Connection",
                    id=count,
                    pre_cell_id="../%s[%i]" % (pop0.id, pre),
                    post_cell_id="../%s[%i]" % (pop1.id, post),
                )
                count += 1
:::
:::
:::
:::

::: {.cell .docutils .container}
::: {.cell_input .docutils .container}
::: {.highlight-ipython3 .notranslate}
::: highlight
    nml_doc.info(True)
:::
:::
:::

::: {.cell_output .docutils .container}
::: {.output .stream .highlight-myst-ansi .notranslate}
::: highlight
    Please see the NeuroML standard schema documentation at https://docs.neuroml.org/Userdocs/NeuroMLv2.html for more information.

    Valid members for NeuroMLDocument are:
    * compound_inputs (class: CompoundInput, Optional)
    * compound_input_dls (class: CompoundInputDL, Optional)
    * includes (class: IncludeType, Optional)
    * voltage_clamps (class: VoltageClamp, Optional)
    * extracellular_properties (class: ExtracellularProperties, Optional)
    * voltage_clamp_triples (class: VoltageClampTriple, Optional)
    * intracellular_properties (class: IntracellularProperties, Optional)
    * spike_arrays (class: SpikeArray, Optional)
    * morphology (class: Morphology, Optional)
    * timed_synaptic_inputs (class: TimedSynapticInput, Optional)
    * ion_channel (class: IonChannel, Optional)
    * spike_generators (class: SpikeGenerator, Optional)
    * ion_channel_hhs (class: IonChannelHH, Optional)
    * spike_generator_randoms (class: SpikeGeneratorRandom, Optional)
    * ion_channel_v_shifts (class: IonChannelVShift, Optional)
    * spike_generator_poissons (class: SpikeGeneratorPoisson, Optional)
    * ion_channel_kses (class: IonChannelKS, Optional)
    * spike_generator_ref_poissons (class: SpikeGeneratorRefPoisson, Optional)
    * decaying_pool_concentration_models (class: DecayingPoolConcentrationModel, Optional)
    * poisson_firing_synapses (class: PoissonFiringSynapse, Optional)
    * fixed_factor_concentration_models (class: FixedFactorConcentrationModel, Optional)
    * transient_poisson_firing_synapses (class: TransientPoissonFiringSynapse, Optional)
    * alpha_current_synapses (class: AlphaCurrentSynapse, Optional)
    * IF_curr_alpha (class: IF_curr_alpha, Optional)
    * alpha_synapses (class: AlphaSynapse, Optional)
    * IF_curr_exp (class: IF_curr_exp, Optional)
    * exp_one_synapses (class: ExpOneSynapse, Optional)
        * Contents ('ids'/<objects>): ['syn0']

    * IF_cond_alpha (class: IF_cond_alpha, Optional)
    * exp_two_synapses (class: ExpTwoSynapse, Optional)
    * IF_cond_exp (class: IF_cond_exp, Optional)
    * exp_three_synapses (class: ExpThreeSynapse, Optional)
    * EIF_cond_exp_isfa_ista (class: EIF_cond_exp_isfa_ista, Optional)
    * blocking_plastic_synapses (class: BlockingPlasticSynapse, Optional)
    * EIF_cond_alpha_isfa_ista (class: EIF_cond_alpha_isfa_ista, Optional)
    * double_synapses (class: DoubleSynapse, Optional)
    * HH_cond_exp (class: HH_cond_exp, Optional)
    * gap_junctions (class: GapJunction, Optional)
    * exp_cond_synapses (class: ExpCondSynapse, Optional)
    * silent_synapses (class: SilentSynapse, Optional)
    * alpha_cond_synapses (class: AlphaCondSynapse, Optional)
    * linear_graded_synapses (class: LinearGradedSynapse, Optional)
    * exp_curr_synapses (class: ExpCurrSynapse, Optional)
    * graded_synapses (class: GradedSynapse, Optional)
    * alpha_curr_synapses (class: AlphaCurrSynapse, Optional)
    * annotation (class: Annotation, Optional)
    * biophysical_properties (class: BiophysicalProperties, Optional)
    * SpikeSourcePoisson (class: SpikeSourcePoisson, Optional)
    * cells (class: Cell, Optional)
    * networks (class: Network, Optional)
        * Contents ('ids'/<objects>): ['IzNet']

    * cell2_ca_poolses (class: Cell2CaPools, Optional)
    * ComponentType (class: ComponentType, Optional)
    * base_cells (class: BaseCell, Optional)
    * iaf_tau_cells (class: IafTauCell, Optional)
    * properties (class: Property, Optional)
    * iaf_tau_ref_cells (class: IafTauRefCell, Optional)
    * notes (class: xs:string, Optional)
    * iaf_cells (class: IafCell, Optional)
    * metaid (class: MetaId, Optional)
    * iaf_ref_cells (class: IafRefCell, Optional)
    * izhikevich_cells (class: IzhikevichCell, Optional)
    * izhikevich2007_cells (class: Izhikevich2007Cell, Optional)
        * Contents ('ids'/<objects>): ['iz2007RS0']

    * ad_ex_ia_f_cells (class: AdExIaFCell, Optional)
    * fitz_hugh_nagumo_cells (class: FitzHughNagumoCell, Optional)
    * fitz_hugh_nagumo1969_cells (class: FitzHughNagumo1969Cell, Optional)
    * pinsky_rinzel_ca3_cells (class: PinskyRinzelCA3Cell, Optional)
    * pulse_generators (class: PulseGenerator, Optional)
        * Contents ('ids'/<objects>): ['pg_0', 'pg_1', 'pg_2', 'pg_3', 'pg_4']

    * pulse_generator_dls (class: PulseGeneratorDL, Optional)
    * sine_generators (class: SineGenerator, Optional)
    * sine_generator_dls (class: SineGeneratorDL, Optional)
    * ramp_generators (class: RampGenerator, Optional)
    * ramp_generator_dls (class: RampGeneratorDL, Optional)
    * id (class: NmlId, Required)
        * Contents ('ids'/<objects>): IzNet
:::
:::

::: {.output .text_plain .highlight-myst-ansi .notranslate}
::: highlight
    "Please see the NeuroML standard schema documentation at https://docs.neuroml.org/Userdocs/NeuroMLv2.html for more information.\n\nValid members for NeuroMLDocument are:\n* compound_inputs (class: CompoundInput, Optional)\n* compound_input_dls (class: CompoundInputDL, Optional)\n* includes (class: IncludeType, Optional)\n* voltage_clamps (class: VoltageClamp, Optional)\n* extracellular_properties (class: ExtracellularProperties, Optional)\n* voltage_clamp_triples (class: VoltageClampTriple, Optional)\n* intracellular_properties (class: IntracellularProperties, Optional)\n* spike_arrays (class: SpikeArray, Optional)\n* morphology (class: Morphology, Optional)\n* timed_synaptic_inputs (class: TimedSynapticInput, Optional)\n* ion_channel (class: IonChannel, Optional)\n* spike_generators (class: SpikeGenerator, Optional)\n* ion_channel_hhs (class: IonChannelHH, Optional)\n* spike_generator_randoms (class: SpikeGeneratorRandom, Optional)\n* ion_channel_v_shifts (class: IonChannelVShift, Optional)\n* spike_generator_poissons (class: SpikeGeneratorPoisson, Optional)\n* ion_channel_kses (class: IonChannelKS, Optional)\n* spike_generator_ref_poissons (class: SpikeGeneratorRefPoisson, Optional)\n* decaying_pool_concentration_models (class: DecayingPoolConcentrationModel, Optional)\n* poisson_firing_synapses (class: PoissonFiringSynapse, Optional)\n* fixed_factor_concentration_models (class: FixedFactorConcentrationModel, Optional)\n* transient_poisson_firing_synapses (class: TransientPoissonFiringSynapse, Optional)\n* alpha_current_synapses (class: AlphaCurrentSynapse, Optional)\n* IF_curr_alpha (class: IF_curr_alpha, Optional)\n* alpha_synapses (class: AlphaSynapse, Optional)\n* IF_curr_exp (class: IF_curr_exp, Optional)\n* exp_one_synapses (class: ExpOneSynapse, Optional)\n\t* Contents ('ids'/<objects>): ['syn0']\n\n* IF_cond_alpha (class: IF_cond_alpha, Optional)\n* exp_two_synapses (class: ExpTwoSynapse, Optional)\n* IF_cond_exp (class: IF_cond_exp, Optional)\n* exp_three_synapses (class: ExpThreeSynapse, Optional)\n* EIF_cond_exp_isfa_ista (class: EIF_cond_exp_isfa_ista, Optional)\n* blocking_plastic_synapses (class: BlockingPlasticSynapse, Optional)\n* EIF_cond_alpha_isfa_ista (class: EIF_cond_alpha_isfa_ista, Optional)\n* double_synapses (class: DoubleSynapse, Optional)\n* HH_cond_exp (class: HH_cond_exp, Optional)\n* gap_junctions (class: GapJunction, Optional)\n* exp_cond_synapses (class: ExpCondSynapse, Optional)\n* silent_synapses (class: SilentSynapse, Optional)\n* alpha_cond_synapses (class: AlphaCondSynapse, Optional)\n* linear_graded_synapses (class: LinearGradedSynapse, Optional)\n* exp_curr_synapses (class: ExpCurrSynapse, Optional)\n* graded_synapses (class: GradedSynapse, Optional)\n* alpha_curr_synapses (class: AlphaCurrSynapse, Optional)\n* annotation (class: Annotation, Optional)\n* biophysical_properties (class: BiophysicalProperties, Optional)\n* SpikeSourcePoisson (class: SpikeSourcePoisson, Optional)\n* cells (class: Cell, Optional)\n* networks (class: Network, Optional)\n\t* Contents ('ids'/<objects>): ['IzNet']\n\n* cell2_ca_poolses (class: Cell2CaPools, Optional)\n* ComponentType (class: ComponentType, Optional)\n* base_cells (class: BaseCell, Optional)\n* iaf_tau_cells (class: IafTauCell, Optional)\n* properties (class: Property, Optional)\n* iaf_tau_ref_cells (class: IafTauRefCell, Optional)\n* notes (class: xs:string, Optional)\n* iaf_cells (class: IafCell, Optional)\n* metaid (class: MetaId, Optional)\n* iaf_ref_cells (class: IafRefCell, Optional)\n* izhikevich_cells (class: IzhikevichCell, Optional)\n* izhikevich2007_cells (class: Izhikevich2007Cell, Optional)\n\t* Contents ('ids'/<objects>): ['iz2007RS0']\n\n* ad_ex_ia_f_cells (class: AdExIaFCell, Optional)\n* fitz_hugh_nagumo_cells (class: FitzHughNagumoCell, Optional)\n* fitz_hugh_nagumo1969_cells (class: FitzHughNagumo1969Cell, Optional)\n* pinsky_rinzel_ca3_cells (class: PinskyRinzelCA3Cell, Optional)\n* pulse_generators (class: PulseGenerator, Optional)\n\t* Contents ('ids'/<objects>): ['pg_0', 'pg_1', 'pg_2', 'pg_3', 'pg_4']\n\n* pulse_generator_dls (class: PulseGeneratorDL, Optional)\n* sine_generators (class: SineGenerator, Optional)\n* sine_generator_dls (class: SineGeneratorDL, Optional)\n* ramp_generators (class: RampGenerator, Optional)\n* ramp_generator_dls (class: RampGeneratorDL, Optional)\n* id (class: NmlId, Required)\n\t* Contents ('ids'/<objects>): IzNet\n\n"
:::
:::
:::
:::

::: {.cell .docutils .container}
::: {.cell_input .docutils .container}
::: {.highlight-ipython3 .notranslate}
::: highlight
    print(nml_doc.summary())
:::
:::
:::

::: {.cell_output .docutils .container}
::: {.output .stream .highlight-myst-ansi .notranslate}
::: highlight
    *******************************************************
    * NeuroMLDocument: IzNet
    *
    *  ExpOneSynapse: ['syn0']
    *  Izhikevich2007Cell: ['iz2007RS0']
    *  PulseGenerator: ['pg_0', 'pg_1', 'pg_2', 'pg_3', 'pg_4']
    *
    *  Network: IzNet
    *
    *   10 cells in 2 populations 
    *     Population: IzPop0 with 5 components of type iz2007RS0
    *       Properties: color=0 0 .8; 
    *     Population: IzPop1 with 5 components of type iz2007RS0
    *       Properties: color=.8 0 0; 
    *
    *   20 connections in 1 projections 
    *     Projection: proj from IzPop0 to IzPop1, synapse: syn0
    *       20 connections: [(Connection 0: 0 -> 0), ...]
    *
    *   0 inputs in 0 input lists 
    *
    *   5 explicit inputs (outside of input lists)
    *     Explicit Input of type pg_0 to IzPop0(cell 0), destination: unspecified
    *     Explicit Input of type pg_1 to IzPop0(cell 1), destination: unspecified
    *     Explicit Input of type pg_2 to IzPop0(cell 2), destination: unspecified
    *     Explicit Input of type pg_3 to IzPop0(cell 3), destination: unspecified
    *     Explicit Input of type pg_4 to IzPop0(cell 4), destination: unspecified
    *
    *******************************************************
:::
:::
:::
:::
:::

::: {#write-the-neuroml-model-to-a-neuroml-file-and-validate-it .section}
##### Write the NeuroML model to a NeuroML file and validate it[\#](#write-the-neuroml-model-to-a-neuroml-file-and-validate-it "Link to this heading"){.headerlink}

::: {.cell .docutils .container}
::: {.cell_input .docutils .container}
::: {.highlight-ipython3 .notranslate}
::: highlight
    nml_file = 'izhikevich2007_network.nml'
    writers.NeuroMLWriter.write(nml_doc, nml_file)

    print("Written network file to: " + nml_file)
    pynml.validate_neuroml2(nml_file)
:::
:::
:::

::: {.cell_output .docutils .container}
::: {.output .stderr .highlight-myst-ansi .notranslate}
::: highlight
    pyNeuroML >>> INFO - Executing: (java -Xmx400M  -jar  "/home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/.venv/lib/python3.10/site-packages/pyneuroml/lib/jNeuroML-0.12.0-jar-with-dependencies.jar" -validate "izhikevich2007_network.nml" ) in directory: .
:::
:::

::: {.output .stream .highlight-myst-ansi .notranslate}
::: highlight
    Written network file to: izhikevich2007_network.nml
:::
:::

::: {.output .stderr .highlight-myst-ansi .notranslate}
::: highlight
    pyNeuroML >>> INFO - Command completed. Output: 
      jNeuroML >>   jNeuroML v0.12.0
      jNeuroML >>  Validating: /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/izhikevich2007_network.nml
      jNeuroML >>  Valid against schema and all tests
      jNeuroML >>  No warnings
      jNeuroML >>  
      jNeuroML >>  Validated 1 files: All valid and no warnings
      jNeuroML >>  
      jNeuroML >>  
:::
:::

::: {.output .text_plain .highlight-myst-ansi .notranslate}
::: highlight
    True
:::
:::
:::
:::
:::
:::

::: {#simulating-the-model .section}
#### Simulating the model[\#](#simulating-the-model "Link to this heading"){.headerlink}

::: {#create-a-simulation-instance-of-the-model .section}
##### Create a simulation instance of the model[\#](#create-a-simulation-instance-of-the-model "Link to this heading"){.headerlink}

::: {.cell .docutils .container}
::: {.cell_input .docutils .container}
::: {.highlight-ipython3 .notranslate}
::: highlight
    simulation_id = "example_izhikevich2007network_sim"
    simulation = LEMSSimulation(sim_id=simulation_id,
                                duration=1000, dt=0.1, simulation_seed=123)
    simulation.assign_simulation_target(net.id)
    simulation.include_neuroml2_file(nml_file)
:::
:::
:::

::: {.cell_output .docutils .container}
::: {.output .stderr .highlight-myst-ansi .notranslate}
::: highlight
    pyNeuroML >>> INFO - Loading NeuroML2 file: /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/izhikevich2007_network.nml
:::
:::
:::
:::
:::

::: {#define-the-output-file-to-store-spikes .section}
##### Define the output file to store spikes[\#](#define-the-output-file-to-store-spikes "Link to this heading"){.headerlink}

::: {.cell .docutils .container}
::: {.cell_input .docutils .container}
::: {.highlight-ipython3 .notranslate}
::: highlight
    simulation.create_event_output_file(
        "pop0", "%s.0.spikes.dat" % simulation_id, format='ID_TIME'
    )
    for pre in range(0, size0):
        simulation.add_selection_to_event_output_file(
            "pop0", pre, 'IzPop0[{}]'.format(pre), 'spike')

    simulation.create_event_output_file(
        "pop1", "%s.1.spikes.dat" % simulation_id, format='ID_TIME'
    )
    for pre in range(0, size1):
        simulation.add_selection_to_event_output_file(
            "pop1", pre, 'IzPop1[{}]'.format(pre), 'spike')
:::
:::
:::
:::
:::
:::

::: {#save-the-simulation-to-a-file .section}
#### Save the simulation to a file[\#](#save-the-simulation-to-a-file "Link to this heading"){.headerlink}

::: {.cell .docutils .container}
::: {.cell_input .docutils .container}
::: {.highlight-ipython3 .notranslate}
::: highlight
    lems_simulation_file = simulation.save_to_file()
:::
:::
:::
:::
:::

::: {#run-the-simulation-using-jneuroml .section}
#### Run the simulation using jNeuroML[\#](#run-the-simulation-using-jneuroml "Link to this heading"){.headerlink}

::: {.cell .docutils .container}
::: {.cell_input .docutils .container}
::: {.highlight-ipython3 .notranslate}
::: highlight
    pynml.run_lems_with_jneuroml_neuron(
        lems_simulation_file, max_memory="2G", nogui=True, plot=False
    )
:::
:::
:::

::: {.cell_output .docutils .container}
::: {.output .stderr .highlight-myst-ansi .notranslate}
::: highlight
    pyNeuroML >>> INFO - Loading LEMS file: LEMS_example_izhikevich2007network_sim.xml and running with jNeuroML_NEURON
    pyNeuroML >>> INFO - Executing: (java -Xmx2G  -Djava.awt.headless=true -jar  "/home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/.venv/lib/python3.10/site-packages/pyneuroml/lib/jNeuroML-0.12.0-jar-with-dependencies.jar"  "LEMS_example_izhikevich2007network_sim.xml"  -neuron -run -compile -nogui -I '') in directory: .
    pyNeuroML >>> INFO - Command completed. Output: 
      jNeuroML >>   jNeuroML v0.12.0
      jNeuroML >>  (INFO) Reading from: /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/LEMS_example_izhikevich2007network_sim.xml
      jNeuroML >>  (INFO) Creating NeuronWriter to output files to /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples
      jNeuroML >>  (INFO) Adding simulation Component(id=example_izhikevich2007network_sim type=Simulation) of network/component: IzNet (Type: network)
      jNeuroML >>  (INFO) Adding population: IzPop0
      jNeuroML >>  (INFO) -- Writing to mod: /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/iz2007RS0.mod
      jNeuroML >>  (INFO) File /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/iz2007RS0.mod exists and is identical
      jNeuroML >>  (INFO) Adding population: IzPop1
      jNeuroML >>  (INFO) -- Mod file for: iz2007RS0 has already been created
      jNeuroML >>  (INFO) Adding projections/connections...
      jNeuroML >>  (INFO) -- Writing to mod: /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/syn0.mod
      jNeuroML >>  (INFO) File /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/syn0.mod exists and is identical
      jNeuroML >>  (INFO) -- Writing to mod: /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/pg_0.mod
      jNeuroML >>  (INFO) File /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/pg_0.mod exists and is identical
      jNeuroML >>  (INFO) -- Writing to mod: /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/pg_1.mod
      jNeuroML >>  (INFO) File /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/pg_1.mod exists and is identical
      jNeuroML >>  (INFO) -- Writing to mod: /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/pg_2.mod
      jNeuroML >>  (INFO) File /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/pg_2.mod exists and is identical
      jNeuroML >>  (INFO) -- Writing to mod: /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/pg_3.mod
      jNeuroML >>  (INFO) File /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/pg_3.mod exists and is identical
      jNeuroML >>  (INFO) -- Writing to mod: /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/pg_4.mod
      jNeuroML >>  (INFO) File /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/pg_4.mod exists and is identical
      jNeuroML >>  (INFO) Trying to compile mods in: /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples
      jNeuroML >>  (INFO) Going to compile the mod files in: /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples, forcing recompile: false
      jNeuroML >>  (INFO) Parent dir: /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples
      jNeuroML >>  (INFO) Assuming *nix environment...
      jNeuroML >>  (INFO) Name of file to be created: /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/x86_64/libnrnmech.la
      jNeuroML >>  (INFO) Name of file to be created: /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/x86_64/libnrnmech.so
      jNeuroML >>  (INFO) Name of file to be created: /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/x86_64/.libs/libnrnmech.so
      jNeuroML >>  (INFO) commandToExecute: /usr/bin/nrnivmodl
      jNeuroML >>  (INFO) Found previously compiled file: /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/x86_64/libnrnmech.so
      jNeuroML >>  (INFO) Going to check if mods in /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples are newer than /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/x86_64/libnrnmech.so
      jNeuroML >>  (INFO) Going to check if mods in /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples are newer than /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/x86_64/.libs/libnrnmech.so
      jNeuroML >>  (INFO) Not being asked to recompile, and no mod files exist in /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples which are newer than /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/x86_64/.libs/libnrnmech.so
      jNeuroML >>  (INFO) Success in compiling mods: true
      jNeuroML >>  (INFO) Have successfully executed command: python /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/LEMS_example_izhikevich2007network_sim_nrn.py
      jNeuroML >>  (INFO) NRN Output >>> 
      jNeuroML >>  (INFO) NRN Output >>>     Starting simulation in NEURON of 1000ms generated from NeuroML2 model...
      jNeuroML >>  (INFO) NRN Output >>> 
      jNeuroML >>  (INFO) NRN Output >>> Population IzPop0 contains 5 instance(s) of component: iz2007RS0 of type: izhikevich2007Cell
      jNeuroML >>  (INFO) NRN Output >>> Population IzPop1 contains 5 instance(s) of component: iz2007RS0 of type: izhikevich2007Cell
      jNeuroML >>  (INFO) NRN Output >>> Adding projection: proj, from IzPop0 to IzPop1 with synapse syn0, 20 connection(s)
      jNeuroML >>  (INFO) NRN Output >>> Setting up the network to simulate took 0.000990 seconds
      jNeuroML >>  (INFO) NRN Output >>> Running a simulation of 1000.0ms (dt = 0.1ms; seed=123)
      jNeuroML >>  (INFO) NRN Output >>> Finished NEURON simulation in 0.022036 seconds (0.000367 mins)...
      jNeuroML >>  (INFO) NRN Output >>> Saving results at t=999.9999999996382...
      jNeuroML >>  (INFO) NRN Output >>> Saved data to: time.dat
      jNeuroML >>  (INFO) NRN Output >>> Saved data to: example_izhikevich2007network_sim.1.spikes.dat
      jNeuroML >>  (INFO) NRN Output >>> Saved data to: example_izhikevich2007network_sim.0.spikes.dat
      jNeuroML >>  (INFO) NRN Output >>> Finished saving results in 0.002883 seconds
      jNeuroML >>  (INFO) NRN Output >>> Done
      jNeuroML >>  (INFO) Exit value for running NEURON: 0
      jNeuroML >>  
:::
:::

::: {.output .text_plain .highlight-myst-ansi .notranslate}
::: highlight
    True
:::
:::
:::
:::
:::

::: {#plot-the-recorded-data .section}
#### Plot the recorded data[\#](#plot-the-recorded-data "Link to this heading"){.headerlink}

::: {.cell .docutils .container}
::: {.cell_input .docutils .container}
::: {.highlight-ipython3 .notranslate}
::: highlight
    # Load the data from the file and plot the spike times
    # using the pynml generate_plot utility function.
    data_array_0 = np.loadtxt("%s.0.spikes.dat" % simulation_id)
    data_array_1 = np.loadtxt("%s.1.spikes.dat" % simulation_id)
    times_0 = data_array_0[:,1]
    times_1 = data_array_1[:,1]
    ids_0 = data_array_0[:,0]
    ids_1 = [id+size0 for id in data_array_1[:,0]]
    pynml.generate_plot(
        [times_0,times_1], [ids_0,ids_1],
        "Spike times", show_plot_already=False,
        save_figure_to="%s-spikes.png" % simulation_id,
        xaxis="time (s)", yaxis="cell ID",
        colors=['b','r'],
        linewidths=['0','0'], markers=['.','.'],
    )
:::
:::
:::

::: {.cell_output .docutils .container}
::: {.output .stderr .highlight-myst-ansi .notranslate}
::: highlight
    pyNeuroML >>> INFO - Generating plot: Spike times
    /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/.venv/lib64/python3.10/site-packages/pyneuroml/plot/Plot.py:174: UserWarning: marker is redundantly defined by the 'marker' keyword argument and the fmt string "o" (-> marker='o'). The keyword argument will take precedence.
      plt.plot(
    pyNeuroML >>> INFO - Saving image to /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/example_izhikevich2007network_sim-spikes.png of plot: Spike times
    pyNeuroML >>> INFO - Saved image to example_izhikevich2007network_sim-spikes.png of plot: Spike times
:::
:::

::: {.output .text_plain .highlight-myst-ansi .notranslate}
::: highlight
    <AxesSubplot: xlabel='time (s)', ylabel='cell ID'>
:::
:::

![](_images/35387de40bb3779c6e78fd8b3b87dcd64e3e7aafd3790055adfb904c42de8715.png)
:::
:::
:::
:::

[]{#document-Userdocs/SingleCompartmentHHExample}

::: {#simulating-a-single-compartment-hodgkin-huxley-neuron .section .tex2jax_ignore .mathjax_ignore}
[]{#userdocs-getting-started-single-compartment-example}

### Simulating a single compartment Hodgkin-Huxley neuron[\#](#simulating-a-single-compartment-hodgkin-huxley-neuron "Link to this heading"){.headerlink}

In this section we will model and simulate a Hodgkin-Huxley (HH) neuron
([\[[HH52](Landing.html#id14 "Alan L. Hodgkin and Andrew F. Huxley. A quantitative description of membrane current and its application to conduction and excitation in nerve. The Journal of physiology, 117(4):500, 1952."){.reference
.internal}\]]{#id1}). A Hodgkin-Huxley neuron includes Sodium (Na),
Potassium (K), and leak ion channels. For further information on this
neuron model, please see
[here](https://hodgkin-huxley-tutorial.readthedocs.io/en/latest/index.html){.reference
.external}.

<figure id="id2" class="align-center">
<img src="_images/HH_single_compartment_example_sim-v.png"
alt="Membrane potential for neuron recorded from the simulation" />
<figcaption><p><span class="caption-number">Fig. 13 </span><span
class="caption-text">Membrane potential of the simulated Hodgkin-Huxley
neuron.</span><a href="#id2" class="headerlink"
title="Link to this image">#</a></p></figcaption>
</figure>

This plot, saved as [`HH_single_compartment_example_sim-v.png`{.docutils
.literal .notranslate}]{.pre} is generated using the following Python
NeuroML script:

::: {.highlight-python .notranslate}
::: highlight
    #!/usr/bin/env python3
    """
    Create a network with a single HH cell, and simulate it.

    File: hh-single-compartment.py

    Copyright 2023 NeuroML contributors
    Author: Ankur Sinha <sanjay DOT ankur AT gmail DOT com>
    """

    import math
    import neuroml
    from neuroml import NeuroMLDocument
    from neuroml import Network, Population
    from neuroml import PulseGenerator, ExplicitInput
    import numpy as np
    from pyneuroml import pynml
    from pyneuroml.lems import LEMSSimulation
    from neuroml.utils import component_factory


    def main():
        """Main function

        Include the NeuroML model into a LEMS simulation file, run it, plot some
        data.
        """
        # Simulation bits
        sim_id = "HH_single_compartment_example_sim"
        simulation = LEMSSimulation(
            sim_id=sim_id, duration=300, dt=0.01, simulation_seed=123
        )
        # Include the NeuroML model file
        simulation.include_neuroml2_file(create_network())
        # Assign target for the simulation
        simulation.assign_simulation_target("single_hh_cell_network")

        # Recording information from the simulation
        simulation.create_output_file(id="output0", file_name=sim_id + ".dat")
        simulation.add_column_to_output_file(
            "output0", column_id="pop0[0]/v", quantity="pop0[0]/v"
        )
        simulation.add_column_to_output_file(
            "output0", column_id="pop0[0]/iChannels", quantity="pop0[0]/iChannels"
        )
        simulation.add_column_to_output_file(
            "output0",
            column_id="pop0[0]/na/iDensity",
            quantity="pop0[0]/biophys/membraneProperties/na_channels/iDensity/",
        )
        simulation.add_column_to_output_file(
            "output0",
            column_id="pop0[0]/k/iDensity",
            quantity="pop0[0]/biophys/membraneProperties/k_channels/iDensity/",
        )

        # Save LEMS simulation to file
        sim_file = simulation.save_to_file()

        # Run the simulation using the default jNeuroML simulator
        pynml.run_lems_with_jneuroml(sim_file, max_memory="2G", nogui=True, plot=False)
        # Plot the data
        plot_data(sim_id)


    def plot_data(sim_id):
        """Plot the sim data.

        Load the data from the file and plot the graph for the membrane potential
        using the pynml generate_plot utility function.

        :sim_id: ID of simulaton

        """
        data_array = np.loadtxt(sim_id + ".dat")
        pynml.generate_plot(
            [data_array[:, 0]],
            [data_array[:, 1]],
            "Membrane potential",
            show_plot_already=False,
            save_figure_to=sim_id + "-v.png",
            xaxis="time (s)",
            yaxis="membrane potential (V)",
        )
        pynml.generate_plot(
            [data_array[:, 0]],
            [data_array[:, 2]],
            "channel current",
            show_plot_already=False,
            save_figure_to=sim_id + "-i.png",
            xaxis="time (s)",
            yaxis="channel current (A)",
        )
        pynml.generate_plot(
            [data_array[:, 0], data_array[:, 0]],
            [data_array[:, 3], data_array[:, 4]],
            "current density",
            labels=["Na", "K"],
            show_plot_already=False,
            save_figure_to=sim_id + "-iden.png",
            xaxis="time (s)",
            yaxis="current density (A_per_m2)",
        )


    def create_na_channel():
        """Create the Na channel.

        This will create the Na channel and save it to a file.
        It will also validate this file.

        returns: name of the created file
        """
        na_channel = component_factory(
            "IonChannelHH",
            id="na_channel",
            notes="Sodium channel for HH cell",
            conductance="10pS",
            species="na",
            validate=False,
        )
        gate_m = component_factory(
            "GateHHRates",
            id="m",
            instances="3",
            notes="m gate for na channel",
            validate=False,
        )
        m_forward_rate = component_factory(
            "HHRate", type="HHExpLinearRate", rate="1per_ms", midpoint="-40mV", scale="10mV"
        )
        m_reverse_rate = component_factory(
            "HHRate", type="HHExpRate", rate="4per_ms", midpoint="-65mV", scale="-18mV"
        )

        gate_m.add(m_forward_rate, hint="forward_rate", validate=False)
        gate_m.add(m_reverse_rate, hint="reverse_rate")
        na_channel.add(gate_m)

        gate_h = component_factory(
            "GateHHRates",
            id="h",
            instances="1",
            notes="h gate for na channel",
            validate=False,
        )
        h_forward_rate = component_factory(
            "HHRate", type="HHExpRate", rate="0.07per_ms", midpoint="-65mV", scale="-20mV"
        )
        h_reverse_rate = component_factory(
            "HHRate", type="HHSigmoidRate", rate="1per_ms", midpoint="-35mV", scale="10mV"
        )
        gate_h.add(h_forward_rate, hint="forward_rate", validate=False)
        gate_h.add(h_reverse_rate, hint="reverse_rate")
        na_channel.add(gate_h)

        na_channel_doc = component_factory(
            "NeuroMLDocument", id="na_channel", notes="Na channel for HH neuron"
        )
        na_channel_fn = "HH_example_na_channel.nml"
        na_channel_doc.add(na_channel)
        na_channel_doc.validate(recursive=True)

        pynml.write_neuroml2_file(
            nml2_doc=na_channel_doc, nml2_file_name=na_channel_fn, validate=True
        )

        return na_channel_fn


    def create_k_channel():
        """Create the K channel

        This will create the K channel and save it to a file.
        It will also validate this file.

        :returns: name of the K channel file
        """
        k_channel = component_factory(
            "IonChannelHH",
            id="k_channel",
            notes="Potassium channel for HH cell",
            conductance="10pS",
            species="k",
            validate=False,
        )
        gate_n = component_factory(
            "GateHHRates",
            id="n",
            instances="4",
            notes="n gate for k channel",
            validate=False,
        )
        n_forward_rate = component_factory(
            "HHRate",
            type="HHExpLinearRate",
            rate="0.1per_ms",
            midpoint="-55mV",
            scale="10mV",
        )
        n_reverse_rate = component_factory(
            "HHRate", type="HHExpRate", rate="0.125per_ms", midpoint="-65mV", scale="-80mV"
        )
        gate_n.add(n_forward_rate, hint="forward_rate", validate=False)
        gate_n.add(n_reverse_rate, hint="reverse_rate")
        k_channel.add(gate_n)

        k_channel_doc = component_factory(
            "NeuroMLDocument", id="k_channel", notes="k channel for HH neuron"
        )
        k_channel_fn = "HH_example_k_channel.nml"
        k_channel_doc.add(k_channel)
        k_channel_doc.validate(recursive=True)

        pynml.write_neuroml2_file(
            nml2_doc=k_channel_doc, nml2_file_name=k_channel_fn, validate=True
        )

        return k_channel_fn


    def create_leak_channel():
        """Create a leak channel

        This will create the leak channel and save it to a file.
        It will also validate this file.

        :returns: name of leak channel nml file
        """
        leak_channel = component_factory(
            "IonChannelHH", id="leak_channel", conductance="10pS", notes="Leak conductance"
        )
        leak_channel_doc = component_factory(
            "NeuroMLDocument", id="leak_channel", notes="leak channel for HH neuron"
        )
        leak_channel_fn = "HH_example_leak_channel.nml"
        leak_channel_doc.add(leak_channel)
        leak_channel_doc.validate(recursive=True)

        pynml.write_neuroml2_file(
            nml2_doc=leak_channel_doc, nml2_file_name=leak_channel_fn, validate=True
        )

        return leak_channel_fn


    def create_cell():
        """Create the cell.

        :returns: name of the cell nml file
        """
        # Create the nml file and add the ion channels
        hh_cell_doc = NeuroMLDocument(id="cell", notes="HH cell")
        hh_cell_fn = "HH_example_cell.nml"

        # Define a cell
        hh_cell = hh_cell_doc.add(
            "Cell", id="hh_cell", notes="A single compartment HH cell"
        )  # type: neuroml.Cell
        hh_cell.info(show_contents=True)

        # Channel density for Na channel
        hh_cell.add_channel_density(
            hh_cell_doc,
            cd_id="na_channels",
            cond_density="120.0 mS_per_cm2",
            erev="50.0 mV",
            ion="na",
            ion_channel="na_channel",
            ion_chan_def_file=create_na_channel(),
        )

        # Channel density for k channel
        hh_cell.add_channel_density(
            hh_cell_doc,
            cd_id="k_channels",
            cond_density="360 S_per_m2",
            erev="-77mV",
            ion="k",
            ion_channel="k_channel",
            ion_chan_def_file=create_k_channel(),
        )
        # Leak channel
        hh_cell.add_channel_density(
            hh_cell_doc,
            cd_id="leak_channels",
            cond_density="3.0 S_per_m2",
            erev="-54.3mV",
            ion="non_specific",
            ion_channel="leak_channel",
            ion_chan_def_file=create_leak_channel(),
        )

        # Other membrane properties
        hh_cell.add_membrane_property("SpikeThresh", value="-20mV")
        hh_cell.set_specific_capacitance("1.0 uF_per_cm2")
        hh_cell.set_init_memb_potential("-65mV")

        hh_cell.set_resistivity("0.03 kohm_cm")

        # We want a diameter such that area is 1000 micro meter^2
        # surface area of a sphere is 4pi r^2 = 4pi diam^2
        diam = math.sqrt(1000 / math.pi)
        hh_cell.add_segment(
            prox=[0, 0, 0, diam],
            dist=[0, 0, 0, diam],
            name="soma",
            parent=None,
            fraction_along=1.0,
            group="soma_0",
        )

        hh_cell_doc.validate(recursive=True)
        pynml.write_neuroml2_file(
            nml2_doc=hh_cell_doc, nml2_file_name=hh_cell_fn, validate=True
        )
        return hh_cell_fn


    def create_network():
        """Create the network

        :returns: name of network nml file
        """
        net_doc = component_factory(
            "NeuroMLDocument", id="network", notes="HH cell network"
        )
        net_doc_fn = "HH_example_net.nml"
        net_doc.add("IncludeType", href=create_cell())
        net = net_doc.add("Network", id="single_hh_cell_network", validate=False)

        # Create a population: convenient to create many cells of the same type
        pop = net.add(
            "Population",
            id="pop0",
            notes="A population for our cell",
            component="hh_cell",
            size=1,
        )

        # Input
        pulsegen = net_doc.add(
            "PulseGenerator",
            id="pg",
            notes="Simple pulse generator",
            delay="100ms",
            duration="100ms",
            amplitude="0.08nA",
        )

        exp_input = net.add("ExplicitInput", target="pop0[0]", input="pg")

        net_doc.validate(recursive=True)

        pynml.write_neuroml2_file(
            nml2_doc=net_doc, nml2_file_name=net_doc_fn, validate=True
        )
        return net_doc_fn


    if __name__ == "__main__":
        main()
:::
:::

::: {#declaring-the-model-in-neuroml .section}
[]{#userdocs-getting-started-single-compartment-example-model}

#### Declaring the model in NeuroML[\#](#declaring-the-model-in-neuroml "Link to this heading"){.headerlink}

Similar to previous examples, we will first declare the model, visualise
it, and then simulate it. The HH neuron model is more complex than the
[[Izhikevich neuron model]{.std
.std-ref}](Landing.html#userdocs-getting-started-single-example){.reference
.internal} we have seen so far. For example, it includes voltage-gated
ion channels. We will first implement these ion channels in NeuroML,
then add them to a cell. We will then create a network of one cell which
will will stimulate with external input to record the membrane
potential.

As you can also see in the script, since this is a slightly more complex
model, we have modularised our code into different functions that carry
out different tasks. Let us now step through the script in a bottom-up
fashion. We start with the ion channels and build the network
simulation.

::: {#declaring-ion-channels .section}
[]{#userdocs-getting-started-single-compartment-example-model-channels}

##### Declaring ion channels[\#](#declaring-ion-channels "Link to this heading"){.headerlink}

::: {.admonition-note-you-might-not-need-to-define-your-ion-channels-in-python-every-time .admonition}
Note: you might not need to define your ion channels in Python every
time...

In this example, all parts of the model, including the ion channels, are
defined from scratch in Python and then NeuroML files in XML are
generated and saved. For many modelling projects however, ion channel
XML files will be reused from other models, and can just be included in
the cells that use them with: [`<include`{.docutils .literal
.notranslate}]{.pre}` `{.docutils .literal
.notranslate}[`href="my_channel.nml"/>`{.docutils .literal
.notranslate}]{.pre}. See [[ here]{.std
.std-ref}](Landing.html#userdocs-finding-models){.reference .internal}
for tips on where to find ion channel models in NeuroML.
:::

Let us look at the definition of the Sodium (Na) channel in NeuroML:

::: {.highlight-python .notranslate}
::: highlight
    def create_na_channel():
        """Create the Na channel.

        This will create the Na channel and save it to a file.
        It will also validate this file.

        returns: name of the created file
        """
        na_channel = component_factory(
            "IonChannelHH",
            id="na_channel",
            notes="Sodium channel for HH cell",
            conductance="10pS",
            species="na",
            validate=False,
        )
        gate_m = component_factory(
            "GateHHRates",
            id="m",
            instances="3",
            notes="m gate for na channel",
            validate=False,
        )
        m_forward_rate = component_factory(
            "HHRate", type="HHExpLinearRate", rate="1per_ms", midpoint="-40mV", scale="10mV"
        )
        m_reverse_rate = component_factory(
            "HHRate", type="HHExpRate", rate="4per_ms", midpoint="-65mV", scale="-18mV"
        )

        gate_m.add(m_forward_rate, hint="forward_rate", validate=False)
        gate_m.add(m_reverse_rate, hint="reverse_rate")
        na_channel.add(gate_m)

        gate_h = component_factory(
            "GateHHRates",
            id="h",
            instances="1",
            notes="h gate for na channel",
            validate=False,
        )
        h_forward_rate = component_factory(
            "HHRate", type="HHExpRate", rate="0.07per_ms", midpoint="-65mV", scale="-20mV"
        )
        h_reverse_rate = component_factory(
            "HHRate", type="HHSigmoidRate", rate="1per_ms", midpoint="-35mV", scale="10mV"
        )
        gate_h.add(h_forward_rate, hint="forward_rate", validate=False)
        gate_h.add(h_reverse_rate, hint="reverse_rate")
        na_channel.add(gate_h)

        na_channel_doc = component_factory(
            "NeuroMLDocument", id="na_channel", notes="Na channel for HH neuron"
        )
        na_channel_fn = "HH_example_na_channel.nml"
        na_channel_doc.add(na_channel)
        na_channel_doc.validate(recursive=True)

        pynml.write_neuroml2_file(
            nml2_doc=na_channel_doc, nml2_file_name=na_channel_fn, validate=True
        )

        return na_channel_fn
:::
:::

Here, we define the two gates, [`m`{.docutils .literal
.notranslate}]{.pre} and [`h`{.docutils .literal .notranslate}]{.pre},
with their forward and reverse rates and add them to the channel. Next,
we create a NeuroML document and save this channel (only this channel
that we've just defined) to a NeuroML file and validate it. So we now
have our Na channel defined in a separate NeuroML file that can be used
in multiple models and shared:

::: {.highlight-xml .notranslate}
::: highlight
    <neuroml xmlns="http://www.neuroml.org/schema/neuroml2"  xmlns:xs="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.neuroml.org/schema/neuroml2 https://raw.github.com/NeuroML/NeuroML2/development/Schemas/NeuroML2/NeuroML_v2.3.xsd" id="na_channel">
        <notes>Na channel for HH neuron</notes>
        <ionChannelHH id="na_channel" species="na" conductance="10pS">
            <notes>Sodium channel for HH cell</notes>
            <gateHHrates id="m" instances="3">
                <notes>m gate for na channel</notes>
                <forwardRate type="HHExpLinearRate" rate="1per_ms" midpoint="-40mV" scale="10mV"/>
                <reverseRate type="HHExpRate" rate="4per_ms" midpoint="-65mV" scale="-18mV"/>
            </gateHHrates>
            <gateHHrates id="h" instances="1">
                <notes>h gate for na channel</notes>
                <forwardRate type="HHExpRate" rate="0.07per_ms" midpoint="-65mV" scale="-20mV"/>
                <reverseRate type="HHSigmoidRate" rate="1per_ms" midpoint="-35mV" scale="10mV"/>
            </gateHHrates>
        </ionChannelHH>
    </neuroml>
:::
:::

The various rate equations ([[HHExpLinearRate]{.std
.std-ref}](Landing.html#schema-hhexplinearrate){.reference .internal},
[[HHExpRate]{.std .std-ref}](Landing.html#schema-hhexprate){.reference
.internal}, [[HHSigmoidRate]{.std
.std-ref}](Landing.html#schema-hhsigmoidrate){.reference .internal} that
can be used in the gate (here [[gateHHrates]{.std
.std-ref}](Landing.html#schema-gatehhrates){.reference .internal}, but
other forms such as [[gateHHtauInf]{.std
.std-ref}](Landing.html#schema-gatehhtauinf){.reference .internal} and
[[gateHHInstantaneous]{.std
.std-ref}](Landing.html#schema-gatehhinstantaneous){.reference
.internal} can be used) are defined in the NeuroML [[schema]{.std
.std-ref}](Landing.html#schema-channels){.reference .internal}.

Also note that since we'll want to *include* this file in other NeuroML
files, we make the function return the name of the file. This is an
implementation detail, and there are other ways of doing this too. We
could have hard-coded this in all our functions or defined it as a
global variable in the script for example. If we were using
object-oriented programming, we could have created a class and stored
this information as a class or object variable.

The K and leak channels are defined in a similar way:

::: {.highlight-python .notranslate}
::: highlight
    def create_k_channel():
        """Create the K channel

        This will create the K channel and save it to a file.
        It will also validate this file.

        :returns: name of the K channel file
        """
        k_channel = component_factory(
            "IonChannelHH",
            id="k_channel",
            notes="Potassium channel for HH cell",
            conductance="10pS",
            species="k",
            validate=False,
        )
        gate_n = component_factory(
            "GateHHRates",
            id="n",
            instances="4",
            notes="n gate for k channel",
            validate=False,
        )
        n_forward_rate = component_factory(
            "HHRate",
            type="HHExpLinearRate",
            rate="0.1per_ms",
            midpoint="-55mV",
            scale="10mV",
        )
        n_reverse_rate = component_factory(
            "HHRate", type="HHExpRate", rate="0.125per_ms", midpoint="-65mV", scale="-80mV"
        )
        gate_n.add(n_forward_rate, hint="forward_rate", validate=False)
        gate_n.add(n_reverse_rate, hint="reverse_rate")
        k_channel.add(gate_n)

        k_channel_doc = component_factory(
            "NeuroMLDocument", id="k_channel", notes="k channel for HH neuron"
        )
        k_channel_fn = "HH_example_k_channel.nml"
        k_channel_doc.add(k_channel)
        k_channel_doc.validate(recursive=True)

        pynml.write_neuroml2_file(
            nml2_doc=k_channel_doc, nml2_file_name=k_channel_fn, validate=True
        )

        return k_channel_fn


    def create_leak_channel():
        """Create a leak channel

        This will create the leak channel and save it to a file.
        It will also validate this file.

        :returns: name of leak channel nml file
        """
        leak_channel = component_factory(
            "IonChannelHH", id="leak_channel", conductance="10pS", notes="Leak conductance"
        )
        leak_channel_doc = component_factory(
            "NeuroMLDocument", id="leak_channel", notes="leak channel for HH neuron"
        )
        leak_channel_fn = "HH_example_leak_channel.nml"
        leak_channel_doc.add(leak_channel)
        leak_channel_doc.validate(recursive=True)

        pynml.write_neuroml2_file(
            nml2_doc=leak_channel_doc, nml2_file_name=leak_channel_fn, validate=True
        )

        return leak_channel_fn
:::
:::

They are also saved in their own NeuroML files, which have also been
validated. The file for the K channel:

::: {.highlight-xml .notranslate}
::: highlight
    <neuroml xmlns="http://www.neuroml.org/schema/neuroml2"  xmlns:xs="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.neuroml.org/schema/neuroml2 https://raw.github.com/NeuroML/NeuroML2/development/Schemas/NeuroML2/NeuroML_v2.3.xsd" id="k_channel">
        <notes>k channel for HH neuron</notes>
        <ionChannelHH id="k_channel" species="k" conductance="10pS">
            <notes>Potassium channel for HH cell</notes>
            <gateHHrates id="n" instances="4">
                <notes>n gate for k channel</notes>
                <forwardRate type="HHExpLinearRate" rate="0.1per_ms" midpoint="-55mV" scale="10mV"/>
                <reverseRate type="HHExpRate" rate="0.125per_ms" midpoint="-65mV" scale="-80mV"/>
            </gateHHrates>
        </ionChannelHH>
    </neuroml>
:::
:::

For the leak channel:

::: {.highlight-xml .notranslate}
::: highlight
    <neuroml xmlns="http://www.neuroml.org/schema/neuroml2"  xmlns:xs="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.neuroml.org/schema/neuroml2 https://raw.github.com/NeuroML/NeuroML2/development/Schemas/NeuroML2/NeuroML_v2.3.xsd" id="leak_channel">
        <notes>leak channel for HH neuron</notes>
        <ionChannelHH id="leak_channel" conductance="10pS">
            <notes>Leak conductance</notes>
        </ionChannelHH>
    </neuroml>
:::
:::
:::

::: {#declaring-the-cell .section}
[]{#userdocs-getting-started-single-compartment-example-model-cell}

##### Declaring the cell[\#](#declaring-the-cell "Link to this heading"){.headerlink}

Now that we have declared our ion channels, we can start constructing
our [[cell]{.std .std-ref}](Landing.html#schema-cell){.reference
.internal} in a different function.

::: {.highlight-python .notranslate}
::: highlight
    def create_cell():
        """Create the cell.

        :returns: name of the cell nml file
        """
        # Create the nml file and add the ion channels
        hh_cell_doc = NeuroMLDocument(id="cell", notes="HH cell")
        hh_cell_fn = "HH_example_cell.nml"

        # Define a cell
        hh_cell = hh_cell_doc.add(
            "Cell", id="hh_cell", notes="A single compartment HH cell"
        )  # type: neuroml.Cell
        hh_cell.info(show_contents=True)

        # Channel density for Na channel
        hh_cell.add_channel_density(
            hh_cell_doc,
            cd_id="na_channels",
            cond_density="120.0 mS_per_cm2",
            erev="50.0 mV",
            ion="na",
            ion_channel="na_channel",
            ion_chan_def_file=create_na_channel(),
        )

        # Channel density for k channel
        hh_cell.add_channel_density(
            hh_cell_doc,
            cd_id="k_channels",
            cond_density="360 S_per_m2",
            erev="-77mV",
            ion="k",
            ion_channel="k_channel",
            ion_chan_def_file=create_k_channel(),
        )
        # Leak channel
        hh_cell.add_channel_density(
            hh_cell_doc,
            cd_id="leak_channels",
            cond_density="3.0 S_per_m2",
            erev="-54.3mV",
            ion="non_specific",
            ion_channel="leak_channel",
            ion_chan_def_file=create_leak_channel(),
        )

        # Other membrane properties
        hh_cell.add_membrane_property("SpikeThresh", value="-20mV")
        hh_cell.set_specific_capacitance("1.0 uF_per_cm2")
        hh_cell.set_init_memb_potential("-65mV")

        hh_cell.set_resistivity("0.03 kohm_cm")

        # We want a diameter such that area is 1000 micro meter^2
        # surface area of a sphere is 4pi r^2 = 4pi diam^2
        diam = math.sqrt(1000 / math.pi)
        hh_cell.add_segment(
            prox=[0, 0, 0, diam],
            dist=[0, 0, 0, diam],
            name="soma",
            parent=None,
            fraction_along=1.0,
            group="soma_0",
        )

        hh_cell_doc.validate(recursive=True)
        pynml.write_neuroml2_file(
            nml2_doc=hh_cell_doc, nml2_file_name=hh_cell_fn, validate=True
        )
        return hh_cell_fn
:::
:::

Let us walk through this function:

::: {.highlight-python .notranslate}
::: highlight
        na_channel_doc.add(na_channel)
        na_channel_doc.validate(recursive=True)

        pynml.write_neuroml2_file(
            nml2_doc=na_channel_doc, nml2_file_name=na_channel_fn, validate=True
        )

        return na_channel_fn
:::
:::

We start by creating a new NeuroML document that we will use to save
this cell, and adding the cell to it.

A [[Cell]{.std .std-ref}](Landing.html#schema-cell){.reference
.internal} component has a number of child/children components that we
need to now populate:

::: {.highlight-pycon .notranslate}
::: highlight
    Cell -- Cell with  **segment** s specified in a  **morphology**  element along with details on its  **biophysicalProperties** . NOTE: this can only be correctly simulated using jLEMS when there is a single segment in the cell, and **v**  of this cell represents the membrane potential in that isopotential segment.

    Please see the NeuroML standard schema documentation at https://docs.neuroml.org/Userdocs/NeuroMLv2.html for more information.

    Valid members for Cell are:
    * morphology_attr (class: NmlId, Optional)
    * biophysical_properties_attr (class: NmlId, Optional)
    * morphology (class: Morphology, Optional)
            * Contents ('ids'/<objects>): 'morphology'

    * neuro_lex_id (class: NeuroLexId, Optional)
    * metaid (class: MetaId, Optional)
    * biophysical_properties (class: BiophysicalProperties, Optional)
            * Contents ('ids'/<objects>): 'biophys'

    * id (class: NmlId, Required)
            * Contents ('ids'/<objects>): hh_cell

    * notes (class: xs:string, Optional)
            * Contents ('ids'/<objects>): A single compartment HH cell

    * properties (class: Property, Optional)
    * annotation (class: Annotation, Optional)
:::
:::

We can see that the [[morphology]{.std
.std-ref}](Landing.html#schema-morphology){.reference .internal} and
[[biophysical properties]{.std
.std-ref}](Landing.html#schema-biophysicalproperties){.reference
.internal} components have already been initialised for us. We now need
to add the required components to them.

We begin with the biophysical properties. Biophysical properties are
themselves split into two:

-   the [[membrane properties]{.std
    .std-ref}](Landing.html#schema-membraneproperties){.reference
    .internal}

-   the [[intracellular properties]{.std
    .std-ref}](Landing.html#schema-intracellularproperties){.reference
    .internal}

Let us look at membrane properties first. The [[schema]{.std
.std-ref}](Landing.html#schema-membraneproperties){.reference .internal}
shows that membrane properties has two *child* elements:

-   [[initMembPotential]{.std
    .std-ref}](Landing.html#schema-initmembpotential){.reference
    .internal}

-   [[spikeThresh]{.std
    .std-ref}](Landing.html#schema-spikethresh){.reference .internal}

and three *children* elements:

-   [[specificCapacitances]{.std
    .std-ref}](Landing.html#schema-specificcapacitance){.reference
    .internal}

-   [[populations]{.std
    .std-ref}](Landing.html#schema-basechannelpopulation){.reference
    .internal}

-   [[channelDensities]{.std
    .std-ref}](Landing.html#schema-basechanneldensity){.reference
    .internal}

::: {.tip .admonition}
Child elements vs Children elements

When an element specifies a **Child** subelement, it will only have one
of these present (it could have zero). **Children** explicitly says that
there can be zero, one or many subelements.
:::

So, we start with the ion-channels which are distributed along the
membrane with some density. A number of helpful functions are available
to us: [`add_channel_density`{.docutils .literal .notranslate}]{.pre},
[`add_membrane_property`{.docutils .literal .notranslate}]{.pre},
[`set_specific_capacitance`{.docutils .literal .notranslate}]{.pre},
[`set_init_memb_potential`{.docutils .literal .notranslate}]{.pre}: For
example, for the Na channels:

::: {.highlight-python .notranslate}
::: highlight
        # Channel density for Na channel
        hh_cell.add_channel_density(
            hh_cell_doc,
            cd_id="na_channels",
            cond_density="120.0 mS_per_cm2",
            erev="50.0 mV",
            ion="na",
            ion_channel="na_channel",
            ion_chan_def_file=create_na_channel(),
        )
:::
:::

and similarly for the K and leak channels. Now, since the ion-channels
were created in other files, we need to make this document aware of
their declarations. To do this, reference the other files in the
[`ion_chan_def_file`{.docutils .literal .notranslate}]{.pre} argument of
the [`add_channel_density`{.docutils .literal .notranslate}]{.pre}
method. Under the hood, this will [`include`{.docutils .literal
.notranslate}]{.pre} the ion channel definition file we have created in
this cell document using an [`IncludeType`{.docutils .literal
.notranslate}]{.pre} component. Each document we want to include gets
appended to the list of [`includes`{.docutils .literal
.notranslate}]{.pre} for the document.

Next, we add the other child and children elements: the [[Specific
Capacitance]{.std
.std-ref}](Landing.html#schema-specificcapacitance){.reference
.internal}, the [[Spike Threshold]{.std
.std-ref}](Landing.html#schema-spikethresh){.reference .internal}, the
[[InitMembPotential]{.std
.std-ref}](Landing.html#schema-initmembpotential){.reference .internal}.
This completes the membrane properties. We then add the intracellular
properties next: [[Resistivity]{.std
.std-ref}](Landing.html#schema-resistivity){.reference .internal}.

::: {.highlight-python .notranslate}
::: highlight
        # Other membrane properties
        hh_cell.add_membrane_property("SpikeThresh", value="-20mV")
        hh_cell.set_specific_capacitance("1.0 uF_per_cm2")
        hh_cell.set_init_memb_potential("-65mV")

        hh_cell.set_resistivity("0.03 kohm_cm")
:::
:::

Next, we add the [[Morphology]{.std
.std-ref}](Landing.html#schema-morphology){.reference .internal} related
information for our cell. Here, we are only creating a single
compartment cell with only one segment. We will look into
multi-compartment cells with more segments in later examples:

::: {.highlight-python .notranslate}
::: highlight
        diam = math.sqrt(1000 / math.pi)
        hh_cell.add_segment(
            prox=[0, 0, 0, diam],
            dist=[0, 0, 0, diam],
            name="soma",
            parent=None,
            fraction_along=1.0,
            group="soma_0",
        )
:::
:::

A [[segment]{.std .std-ref}](Landing.html#schema-segment){.reference
.internal} has [`proximal`{.docutils .literal .notranslate}]{.pre} and
[`distal`{.docutils .literal .notranslate}]{.pre} child elements which
describe the extent of the segment. These are described using a
[[Point3DWithDiam]{.std
.std-ref}](Landing.html#schema-point3dwithdiam){.reference .internal}
object, which the [`add_segment`{.docutils .literal .notranslate}]{.pre}
function creates for us.

This completes our cell. We add it to our NeuroML document, and save
(and validate) it. The resulting NeuroML file is:

::: {.highlight-xml .notranslate}
::: highlight
    <neuroml xmlns="http://www.neuroml.org/schema/neuroml2"  xmlns:xs="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.neuroml.org/schema/neuroml2 https://raw.github.com/NeuroML/NeuroML2/development/Schemas/NeuroML2/NeuroML_v2.3.xsd" id="cell">
        <notes>HH cell</notes>
        <include href="HH_example_na_channel.nml"/>
        <include href="HH_example_k_channel.nml"/>
        <include href="HH_example_leak_channel.nml"/>
        <cell id="hh_cell">
            <notes>A single compartment HH cell</notes>
            <morphology id="hh_cell_morph">
                <segment id="0" name="soma">
                    <proximal x="0.0" y="0.0" z="0.0" diameter="17.841241161527712"/>
                    <distal x="0.0" y="0.0" z="0.0" diameter="17.841241161527712"/>
                </segment>
            </morphology>
            <biophysicalProperties id="hh_b_prop">
                <membraneProperties>
                    <channelDensity id="na_channels" ionChannel="na_channel" condDensity="120.0 mS_per_cm2" erev="50.0 mV" ion="na"/>
                    <channelDensity id="k_channels" ionChannel="k_channel" condDensity="360 S_per_m2" erev="-77mV" ion="k"/>
                    <channelDensity id="leak_channels" ionChannel="leak_channel" condDensity="3.0 S_per_m2" erev="-54.3mV" ion="non_specific"/>
                    <spikeThresh value="-20mV"/>
                    <specificCapacitance value="1.0 uF_per_cm2"/>
                    <initMembPotential value="-65mV"/>
                </membraneProperties>
                <intracellularProperties>
                    <resistivity value="0.03 kohm_cm"/>
                </intracellularProperties>
            </biophysicalProperties>
        </cell>
    </neuroml>
:::
:::

We now have our cell defined in a separate NeuroML file, that can be
re-used and shared.
:::

::: {#declaring-the-network .section}
[]{#userdocs-getting-started-single-compartment-example-model-network}

##### Declaring the network[\#](#declaring-the-network "Link to this heading"){.headerlink}

We now use our cell in a network. A [[network in NeuroML]{.std
.std-ref}](Landing.html#schema-network){.reference .internal} has
multiple children elements: [[populations]{.std
.std-ref}](Landing.html#schema-population){.reference .internal},
[[projections]{.std
.std-ref}](Landing.html#schema-projection){.reference .internal},
[[inputLists]{.std .std-ref}](Landing.html#schema-inputlist){.reference
.internal} and so on. Here we are going to only create a network with
one cell, and an [[explicit input]{.std
.std-ref}](Landing.html#schema-explicitinput){.reference .internal} to
the cell:

::: {.highlight-python .notranslate}
::: highlight
    def create_network():
        """Create the network

        :returns: name of network nml file
        """
        net_doc = component_factory(
            "NeuroMLDocument", id="network", notes="HH cell network"
        )
        net_doc_fn = "HH_example_net.nml"
        net_doc.add("IncludeType", href=create_cell())
        net = net_doc.add("Network", id="single_hh_cell_network", validate=False)

        # Create a population: convenient to create many cells of the same type
        pop = net.add(
            "Population",
            id="pop0",
            notes="A population for our cell",
            component="hh_cell",
            size=1,
        )

        # Input
        pulsegen = net_doc.add(
            "PulseGenerator",
            id="pg",
            notes="Simple pulse generator",
            delay="100ms",
            duration="100ms",
            amplitude="0.08nA",
        )

        exp_input = net.add("ExplicitInput", target="pop0[0]", input="pg")

        net_doc.validate(recursive=True)

        pynml.write_neuroml2_file(
            nml2_doc=net_doc, nml2_file_name=net_doc_fn, validate=True
        )
        return net_doc_fn
:::
:::

We start in the same way, by creating a new NeuroML document and
including our cell file into it. We then create a [[population]{.std
.std-ref}](Landing.html#schema-population){.reference .internal}
comprising of a single cell. We create a [[pulse generator]{.std
.std-ref}](Landing.html#schema-pulsegenerator){.reference .internal} as
an [[explicit input]{.std
.std-ref}](Landing.html#schema-explicitinput){.reference .internal},
which targets our population. Note that as the schema documentation for
[`ExplicitInput`{.docutils .literal .notranslate}]{.pre} notes, any
current source (any component that *extends* [[basePointCurrent]{.std
.std-ref}](Landing.html#schema-basepointcurrent){.reference .internal})
can be used as an [`ExplicitInput`{.docutils .literal
.notranslate}]{.pre}.

We add all of these to the [[network]{.std
.std-ref}](Landing.html#schema-network){.reference .internal} and save
(and validate) our network file. The NeuroML file generated is below:

::: {.highlight-xml .notranslate}
::: highlight
    <neuroml xmlns="http://www.neuroml.org/schema/neuroml2"  xmlns:xs="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.neuroml.org/schema/neuroml2 https://raw.github.com/NeuroML/NeuroML2/development/Schemas/NeuroML2/NeuroML_v2.3.xsd" id="network">
        <notes>HH cell network</notes>
        <include href="HH_example_cell.nml"/>
        <pulseGenerator id="pg" delay="100ms" duration="100ms" amplitude="0.08nA">
            <notes>Simple pulse generator</notes>
        </pulseGenerator>
        <network id="single_hh_cell_network">
            <population id="pop0" component="hh_cell" size="1">
                <notes>A population for our cell</notes>
            </population>
            <explicitInput target="pop0[0]" input="pg"/>
        </network>
    </neuroml>
:::
:::
:::

::: {#the-generated-neuroml-model .section}
[]{#userdocs-getting-started-single-compartment-example-model-generatedmodel}

##### The generated NeuroML model[\#](#the-generated-neuroml-model "Link to this heading"){.headerlink}

Before we look at simulating the model, we can inspect our model to
check for correctness. All our NeuroML files were validated when they
were created already, so we do not need to run this step again. However,
if required, this can be easily done:

::: {.highlight-console .notranslate}
::: highlight
    pynml -validate HH_*nml
:::
:::

Next, we can visualise our model using the information noted in the
[[visualising NeuroML models]{.std
.std-ref}](Landing.html#userdocs-visualising-models){.reference
.internal} page (including the [`-v`{.docutils .literal
.notranslate}]{.pre} verbose option for more information on the cell):

::: {.highlight-console .notranslate}
::: highlight
    pynml-summary HH_example_net.nml -v
    *******************************************************
    * NeuroMLDocument: network
    *
    *  IonChannelHH: ['k_channel', 'leak_channel', 'na_channel']
    *  PulseGenerator: ['pg']
    *
    *  Cell: hh_cell
    *    <Segment|0|soma>
    *      Parent segment: None (root segment)
    *      (0.0, 0.0, 0.0), diam 17.841241161527712um -> (0.0, 0.0, 0.0), diam 17.841241161527712um; seg length: 0.0 um
    *      Surface area: 1000.0 um2, volume: 2973.5401935879518 um3
    *    Total length of 1 segment: 0.0 um; total area: 1000.0 um2
    *
    *    Channel density: na_channels on all;       conductance of 120.0 mS_per_cm2 through ion chan na_channel with ion na, erev: 50.0 mV
    *      Channel is on <Segment|0|soma>,  total conductance: 1200.0 S_per_m2 x 1e-09 m2 = 1.2000000000000002e-06 S (1200000.0000000002 pS)
    *    Channel density: k_channels on all;        conductance of 360 S_per_m2 through ion chan k_channel with ion k, erev: -77mV
    *      Channel is on <Segment|0|soma>,  total conductance: 360.0 S_per_m2 x 1e-09 m2 = 3.6000000000000005e-07 S (360000.00000000006 pS)
    *    Channel density: leak_channels on all;     conductance of 3.0 S_per_m2 through ion chan leak_channel with ion non_specific, erev: -54.3mV
    *      Channel is on <Segment|0|soma>,  total conductance: 3.0 S_per_m2 x 1e-09 m2 = 3.0000000000000004e-09 S (3000.0000000000005 pS)
    *
    *    Specific capacitance on all: 1.0 uF_per_cm2
    *      Capacitance of <Segment|0|soma>, total capacitance: 0.01 F_per_m2 x 1e-09 m2 = 1.0000000000000001e-11 F (10.000000000000002 pF)
    *
    *  Network: single_hh_cell_network
    *
    *   1 cells in 1 populations
    *     Population: pop0 with 1 components of type hh_cell
    *
    *   0 connections in 0 projections
    *
    *   0 inputs in 0 input lists
    *
    *   1 explicit inputs (outside of input lists)
    *     Explicit Input of type pg to pop0(cell 0), destination: unspecified
    *
    *******************************************************
:::
:::

Since our model is a single compartment model with only one cell, it
doesn't have any 3D structure to visualise. We can check the
connectivity graph of the model:

::: {.highlight-console .notranslate}
::: highlight
    pynml -graph 10 HH_example_net.nml
:::
:::

which will give us this figure:

<figure id="id3" class="align-center">
<a href="_images/single_hh_cell_network.gv.png"
class="reference internal image-reference"><img
src="_images/single_hh_cell_network.gv.png"
style="width: 368.4px; height: 184.2px;"
alt="Level 10 network graph generated by pynml" /></a>
<figcaption><p><span class="caption-number">Fig. 14 </span><span
class="caption-text">Level 10 network graph generated by pynml</span><a
href="#id3" class="headerlink"
title="Link to this image">#</a></p></figcaption>
</figure>

::: {#analysing-channels .section}
[]{#userdocs-getting-started-single-compartment-example-model-generatedmodel-analysingchannels}

###### Analysing channels[\#](#analysing-channels "Link to this heading"){.headerlink}

Finally, we can analyse the ion channels that we've declared using the
[`pynml-channelanalysis`{.docutils .literal .notranslate}]{.pre}
utility:

::: {.highlight-console .notranslate}
::: highlight
    pynml-channelanalysis HH_example_k_channel.nml
:::
:::

This generates graphs to show the behaviour of the channel:

::: container-fluid
::: {.row .my-2 .py-2}
::: {.col-sm-6 .px-2}
<figure id="id4" class="align-center">
<a href="_images/HH_example_k_channel_2.png"
class="reference internal image-reference"><img
src="_images/HH_example_k_channel_2.png"
style="width: 384.0px; height: 288.0px;"
alt="Steady state behaviour of the K ion channel" /></a>
<figcaption><p><span class="caption-number">Fig. 15 </span><span
class="caption-text">Steady state behaviour of the K ion
channel.</span><a href="#id4" class="headerlink"
title="Link to this image">#</a></p></figcaption>
</figure>
:::

::: {.col-sm-6 .px-2}
<figure id="id5" class="align-center">
<a href="_images/HH_example_k_channel_1.png"
class="reference internal image-reference"><img
src="_images/HH_example_k_channel_1.png"
style="width: 384.0px; height: 288.0px;"
alt="Time course of the K ion channel" /></a>
<figcaption><p><span class="caption-number">Fig. 16 </span><span
class="caption-text">Time course of the K ion channel.</span><a
href="#id5" class="headerlink"
title="Link to this image">#</a></p></figcaption>
</figure>
:::
:::
:::

Similarly, we can get these for the Na channel also:

::: {.highlight-console .notranslate}
::: highlight
    pynml-channelanalysis HH_example_na_channel.nml
:::
:::

::: container-fluid
::: {.row .my-2 .py-2}
::: {.col-sm-6 .px-2}
<figure id="id6" class="align-center">
<a href="_images/HH_example_na_channel_2.png"
class="reference internal image-reference"><img
src="_images/HH_example_na_channel_2.png"
style="width: 384.0px; height: 288.0px;"
alt="Steady state behaviour of the Na ion channel" /></a>
<figcaption><p><span class="caption-number">Fig. 17 </span><span
class="caption-text">Steady state behaviour of the Na ion
channel.</span><a href="#id6" class="headerlink"
title="Link to this image">#</a></p></figcaption>
</figure>
:::

::: {.col-sm-6 .px-2}
<figure id="id7" class="align-center">
<a href="_images/HH_example_na_channel_1.png"
class="reference internal image-reference"><img
src="_images/HH_example_na_channel_1.png"
style="width: 384.0px; height: 288.0px;"
alt="Time course of the Na ion channel" /></a>
<figcaption><p><span class="caption-number">Fig. 18 </span><span
class="caption-text">Time course of the Na ion channel.</span><a
href="#id7" class="headerlink"
title="Link to this image">#</a></p></figcaption>
</figure>
:::
:::
:::
:::
:::
:::

::: {#simulating-the-model .section}
[]{#userdocs-getting-started-single-compartment-example-simulating}

#### Simulating the model[\#](#simulating-the-model "Link to this heading"){.headerlink}

Now that we have declared and inspected our network model and all its
components, we can proceed to simulate it. We do this in the
[`main`{.docutils .literal .notranslate}]{.pre} function:

::: {.highlight-python .notranslate}
::: highlight
    def main():
        """Main function

        Include the NeuroML model into a LEMS simulation file, run it, plot some
        data.
        """
        # Simulation bits
        sim_id = "HH_single_compartment_example_sim"
        simulation = LEMSSimulation(
            sim_id=sim_id, duration=300, dt=0.01, simulation_seed=123
        )
        # Include the NeuroML model file
        simulation.include_neuroml2_file(create_network())
        # Assign target for the simulation
        simulation.assign_simulation_target("single_hh_cell_network")

        # Recording information from the simulation
        simulation.create_output_file(id="output0", file_name=sim_id + ".dat")
        simulation.add_column_to_output_file(
            "output0", column_id="pop0[0]/v", quantity="pop0[0]/v"
        )
        simulation.add_column_to_output_file(
            "output0", column_id="pop0[0]/iChannels", quantity="pop0[0]/iChannels"
        )
        simulation.add_column_to_output_file(
            "output0",
            column_id="pop0[0]/na/iDensity",
            quantity="pop0[0]/biophys/membraneProperties/na_channels/iDensity/",
        )
        simulation.add_column_to_output_file(
            "output0",
            column_id="pop0[0]/k/iDensity",
            quantity="pop0[0]/biophys/membraneProperties/k_channels/iDensity/",
        )

        # Save LEMS simulation to file
        sim_file = simulation.save_to_file()

        # Run the simulation using the default jNeuroML simulator
        pynml.run_lems_with_jneuroml(sim_file, max_memory="2G", nogui=True, plot=False)
        # Plot the data
        plot_data(sim_id)
:::
:::

Here we first create a [`LEMSSimulation`{.docutils .literal
.notranslate}]{.pre} instance and include our network NeuroML file in
it. We must inform LEMS what the target of the simulation is. In our
case, it's the id of our network, [`single_hh_cell_network`{.docutils
.literal .notranslate}]{.pre}:

::: {.highlight-python .notranslate}
::: highlight
        sim_id = "HH_single_compartment_example_sim"
        simulation = LEMSSimulation(
            sim_id=sim_id, duration=300, dt=0.01, simulation_seed=123
        )
        # Include the NeuroML model file
        simulation.include_neuroml2_file(create_network())
        # Assign target for the simulation
        simulation.assign_simulation_target("single_hh_cell_network")
:::
:::

We also want to record some information, so we create an output file
first with an [`id`{.docutils .literal .notranslate}]{.pre} of
[`output0`{.docutils .literal .notranslate}]{.pre}:

::: {.highlight-python .notranslate}
::: highlight
        simulation.create_output_file(id="output0", file_name=sim_id + ".dat")
:::
:::

Now, we can record any quantity that is exposed by NeuroML (any
[`exposure`{.docutils .literal .notranslate}]{.pre}). For example, we
add a column for the membrane potential [`v`{.docutils .literal
.notranslate}]{.pre} of the [[cell]{.std
.std-ref}](Landing.html#schema-cell){.reference .internal} which would
be the *0th* (and only) cell in our population [`pop0`{.docutils
.literal .notranslate}]{.pre}: [`pop0[0]/v`{.docutils .literal
.notranslate}]{.pre}. We can also record the current in the channels:
[`pop[0]/iChannels`{.docutils .literal .notranslate}]{.pre} We can also
record the [[current density]{.std
.std-ref}](Landing.html#schema-channeldensity){.reference .internal}
[`iDensity`{.docutils .literal .notranslate}]{.pre} for the channels, so
we also record these.

::: {.highlight-python .notranslate}
::: highlight
        simulation.add_column_to_output_file(
            "output0", column_id="pop0[0]/v", quantity="pop0[0]/v"
        )
        simulation.add_column_to_output_file(
            "output0", column_id="pop0[0]/iChannels", quantity="pop0[0]/iChannels"
        )
        simulation.add_column_to_output_file(
            "output0",
            column_id="pop0[0]/na/iDensity",
            quantity="pop0[0]/biophys/membraneProperties/na_channels/iDensity/",
        )
        simulation.add_column_to_output_file(
            "output0",
            column_id="pop0[0]/k/iDensity",
            quantity="pop0[0]/biophys/membraneProperties/k_channels/iDensity/",
        )
:::
:::

We then save the LEMS simulation file, run our simulation with the
default [[jNeuroML]{.std .std-ref}](Landing.html#jneuroml){.reference
.internal} simulator.
:::

::: {#plotting-the-recorded-variables .section}
[]{#userdocs-getting-started-single-compartment-example-plotting}

#### Plotting the recorded variables[\#](#plotting-the-recorded-variables "Link to this heading"){.headerlink}

To plot the variables that we recorded, we read the data and use the
[`generate_plot`{.docutils .literal .notranslate}]{.pre} utility
function:

::: {.highlight-python .notranslate}
::: highlight
    def plot_data(sim_id):
        """Plot the sim data.

        Load the data from the file and plot the graph for the membrane potential
        using the pynml generate_plot utility function.

        :sim_id: ID of simulaton

        """
        data_array = np.loadtxt(sim_id + ".dat")
        pynml.generate_plot(
            [data_array[:, 0]],
            [data_array[:, 1]],
            "Membrane potential",
            show_plot_already=False,
            save_figure_to=sim_id + "-v.png",
            xaxis="time (s)",
            yaxis="membrane potential (V)",
        )
        pynml.generate_plot(
            [data_array[:, 0]],
            [data_array[:, 2]],
            "channel current",
            show_plot_already=False,
            save_figure_to=sim_id + "-i.png",
            xaxis="time (s)",
            yaxis="channel current (A)",
        )
        pynml.generate_plot(
            [data_array[:, 0], data_array[:, 0]],
            [data_array[:, 3], data_array[:, 4]],
            "current density",
            labels=["Na", "K"],
            show_plot_already=False,
            save_figure_to=sim_id + "-iden.png",
            xaxis="time (s)",
            yaxis="current density (A_per_m2)",
        )
:::
:::

This generates the following graphs:

::: container-fluid
::: {.row .my-2 .py-2}
::: {.col-sm-6 .px-2}
<figure id="id8" class="align-center">
<a href="_images/HH_single_compartment_example_sim-v.png"
class="reference internal image-reference"><img
src="_images/HH_single_compartment_example_sim-v.png"
style="width: 383.4px; height: 277.2px;" alt="Membrane potential" /></a>
<figcaption><p><span class="caption-number">Fig. 19 </span><span
class="caption-text">Membrane potential</span><a href="#id8"
class="headerlink" title="Link to this image">#</a></p></figcaption>
</figure>
:::

::: {.col-sm-6 .px-2}
<figure id="id9" class="align-center">
<a href="_images/HH_single_compartment_example_sim-i.png"
class="reference internal image-reference"><img
src="_images/HH_single_compartment_example_sim-i.png"
style="width: 375.0px; height: 289.2px;" alt="Channel current" /></a>
<figcaption><p><span class="caption-number">Fig. 20 </span><span
class="caption-text">Channel current.</span><a href="#id9"
class="headerlink" title="Link to this image">#</a></p></figcaption>
</figure>
:::

::: {.col-sm-6 .px-2}
<figure id="id10" class="align-center">
<a href="_images/HH_single_compartment_example_sim-iden.png"
class="reference internal image-reference"><img
src="_images/HH_single_compartment_example_sim-iden.png"
style="width: 375.59999999999997px; height: 277.2px;"
alt="Channel current densities" /></a>
<figcaption><p><span class="caption-number">Fig. 21 </span><span
class="caption-text">Channel current densities</span><a href="#id10"
class="headerlink" title="Link to this image">#</a></p></figcaption>
</figure>
:::
:::
:::

This concludes out third example. Here we have seen how to create,
simulate, record, and visualise a single compartment Hodgkin-Huxley
neuron. In the next section, you will find an interactive notebook where
you can play with this example.
:::
:::

[]{#document-Userdocs/NML2_examples/HH_single_compartment}

::: {#interactive-single-compartment-hh-example .section .tex2jax_ignore .mathjax_ignore}
### Interactive single compartment HH example[\#](#interactive-single-compartment-hh-example "Link to this heading"){.headerlink}

To run this interactive Jupyter Notebook, please click on the rocket
icon 🚀 in the top panel. For more information, please see [[how to use
this documentation]{.std
.std-ref}](Landing.html#userdocs-usage-jupyterbooks){.reference
.internal}. Please uncomment the line below if you use the Google Colab.
(It does not include these packages by default).

::: {.cell .docutils .container}
::: {.cell_input .docutils .container}
::: {.highlight-ipython3 .notranslate}
::: highlight
    #%pip install pyneuroml neuromllite NEURON
:::
:::
:::
:::

::: {.cell .docutils .container}
::: {.cell_input .docutils .container}
::: {.highlight-ipython3 .notranslate}
::: highlight
    import math
    from neuroml import NeuroMLDocument
    from neuroml import Cell
    from neuroml import IonChannelHH
    from neuroml import GateHHRates
    from neuroml import BiophysicalProperties
    from neuroml import MembraneProperties
    from neuroml import ChannelDensity
    from neuroml import HHRate
    from neuroml import SpikeThresh
    from neuroml import SpecificCapacitance
    from neuroml import InitMembPotential
    from neuroml import IntracellularProperties
    from neuroml import IncludeType
    from neuroml import Resistivity
    from neuroml import Morphology, Segment, Point3DWithDiam
    from neuroml import Network, Population
    from neuroml import PulseGenerator, ExplicitInput
    import numpy as np
    from pyneuroml import pynml
    from pyneuroml.lems import LEMSSimulation
:::
:::
:::

::: {.cell_output .docutils .container}
::: {.output .traceback .highlight-ipythontb .notranslate}
::: highlight
    ---------------------------------------------------------------------------
    ModuleNotFoundError                       Traceback (most recent call last)
    Cell In[2], line 2
          1 import math
    ----> 2 from neuroml import NeuroMLDocument
          3 from neuroml import Cell
          4 from neuroml import IonChannelHH

    ModuleNotFoundError: No module named 'neuroml'
:::
:::
:::
:::

::: {#declare-the-model .section}
#### Declare the model[\#](#declare-the-model "Link to this heading"){.headerlink}

::: {#create-ion-channels .section}
##### Create ion channels[\#](#create-ion-channels "Link to this heading"){.headerlink}

::: {.cell .docutils .container}
::: {.cell_input .docutils .container}
::: {.highlight-ipython3 .notranslate}
::: highlight
    def create_na_channel():
        """Create the Na channel.

        This will create the Na channel and save it to a file.
        It will also validate this file.

        returns: name of the created file
        """
        na_channel = IonChannelHH(id="na_channel", notes="Sodium channel for HH cell", conductance="10pS", species="na")
        gate_m = GateHHRates(id="m", instances="3", notes="m gate for na channel")

        m_forward_rate = HHRate(type="HHExpLinearRate", rate="1per_ms", midpoint="-40mV", scale="10mV")
        m_reverse_rate = HHRate(type="HHExpRate", rate="4per_ms", midpoint="-65mV", scale="-18mV")
        gate_m.forward_rate = m_forward_rate
        gate_m.reverse_rate = m_reverse_rate
        na_channel.gate_hh_rates.append(gate_m)

        gate_h = GateHHRates(id="h", instances="1", notes="h gate for na channel")
        h_forward_rate = HHRate(type="HHExpRate", rate="0.07per_ms", midpoint="-65mV", scale="-20mV")
        h_reverse_rate = HHRate(type="HHSigmoidRate", rate="1per_ms", midpoint="-35mV", scale="10mV")
        gate_h.forward_rate = h_forward_rate
        gate_h.reverse_rate = h_reverse_rate
        na_channel.gate_hh_rates.append(gate_h)

        na_channel_doc = NeuroMLDocument(id="na_channel", notes="Na channel for HH neuron")
        na_channel_fn = "HH_example_na_channel.nml"
        na_channel_doc.ion_channel_hhs.append(na_channel)

        pynml.write_neuroml2_file(nml2_doc=na_channel_doc, nml2_file_name=na_channel_fn, validate=True)

        return na_channel_fn
:::
:::
:::
:::

::: {.cell .docutils .container}
::: {.cell_input .docutils .container}
::: {.highlight-ipython3 .notranslate}
::: highlight
    def create_k_channel():
        """Create the K channel

        This will create the K channel and save it to a file.
        It will also validate this file.

        :returns: name of the K channel file
        """
        k_channel = IonChannelHH(id="k_channel", notes="Potassium channel for HH cell", conductance="10pS", species="k")
        gate_n = GateHHRates(id="n", instances="4", notes="n gate for k channel")
        n_forward_rate = HHRate(type="HHExpLinearRate", rate="0.1per_ms", midpoint="-55mV", scale="10mV")
        n_reverse_rate = HHRate(type="HHExpRate", rate="0.125per_ms", midpoint="-65mV", scale="-80mV")
        gate_n.forward_rate = n_forward_rate
        gate_n.reverse_rate = n_reverse_rate
        k_channel.gate_hh_rates.append(gate_n)

        k_channel_doc = NeuroMLDocument(id="k_channel", notes="k channel for HH neuron")
        k_channel_fn = "HH_example_k_channel.nml"
        k_channel_doc.ion_channel_hhs.append(k_channel)

        pynml.write_neuroml2_file(nml2_doc=k_channel_doc, nml2_file_name=k_channel_fn, validate=True)

        return k_channel_fn
:::
:::
:::
:::

::: {.cell .docutils .container}
::: {.cell_input .docutils .container}
::: {.highlight-ipython3 .notranslate}
::: highlight
    def create_leak_channel():
        """Create a leak channel

        This will create the leak channel and save it to a file.
        It will also validate this file.

        :returns: name of leak channel nml file
        """
        leak_channel = IonChannelHH(id="leak_channel", conductance="10pS", notes="Leak conductance")
        leak_channel_doc = NeuroMLDocument(id="leak_channel", notes="leak channel for HH neuron")
        leak_channel_fn = "HH_example_leak_channel.nml"
        leak_channel_doc.ion_channel_hhs.append(leak_channel)

        pynml.write_neuroml2_file(nml2_doc=leak_channel_doc, nml2_file_name=leak_channel_fn, validate=True)

        return leak_channel_fn
:::
:::
:::
:::
:::

::: {#create-cell .section}
##### Create cell[\#](#create-cell "Link to this heading"){.headerlink}

::: {.cell .docutils .container}
::: {.cell_input .docutils .container}
::: {.highlight-ipython3 .notranslate}
::: highlight
    def create_cell():
        """Create the cell.

        :returns: name of the cell nml file
        """
        # Create the nml file and add the ion channels
        hh_cell_doc = NeuroMLDocument(id="cell", notes="HH cell")
        hh_cell_fn = "HH_example_cell.nml"
        hh_cell_doc.includes.append(IncludeType(href=create_na_channel()))
        hh_cell_doc.includes.append(IncludeType(href=create_k_channel()))
        hh_cell_doc.includes.append(IncludeType(href=create_leak_channel()))

        # Define a cell
        hh_cell = Cell(id="hh_cell", notes="A single compartment HH cell")

        # Define its biophysical properties
        bio_prop = BiophysicalProperties(id="hh_b_prop")
        #  notes="Biophysical properties for HH cell")

        # Membrane properties are a type of biophysical properties
        mem_prop = MembraneProperties()
        # Add membrane properties to the biophysical properties
        bio_prop.membrane_properties = mem_prop

        # Append to cell
        hh_cell.biophysical_properties = bio_prop

        # Channel density for Na channel
        na_channel_density = ChannelDensity(id="na_channels", cond_density="120.0 mS_per_cm2", erev="50.0 mV", ion="na", ion_channel="na_channel")
        mem_prop.channel_densities.append(na_channel_density)

        # Channel density for k channel
        k_channel_density = ChannelDensity(id="k_channels", cond_density="360 S_per_m2", erev="-77mV", ion="k", ion_channel="k_channel")
        mem_prop.channel_densities.append(k_channel_density)

        # Leak channel
        leak_channel_density = ChannelDensity(id="leak_channels", cond_density="3.0 S_per_m2", erev="-54.3mV", ion="non_specific", ion_channel="leak_channel")
        mem_prop.channel_densities.append(leak_channel_density)

        # Other membrane properties
        mem_prop.spike_threshes.append(SpikeThresh(value="-20mV"))
        mem_prop.specific_capacitances.append(SpecificCapacitance(value="1.0 uF_per_cm2"))
        mem_prop.init_memb_potentials.append(InitMembPotential(value="-65mV"))

        intra_prop = IntracellularProperties()
        intra_prop.resistivities.append(Resistivity(value="0.03 kohm_cm"))

        # Add to biological properties
        bio_prop.intracellular_properties = intra_prop

        # Morphology
        morph = Morphology(id="hh_cell_morph")
        #  notes="Simple morphology for the HH cell")
        seg = Segment(id="0", name="soma", notes="Soma segment")
        # We want a diameter such that area is 1000 micro meter^2
        # surface area of a sphere is 4pi r^2 = 4pi diam^2
        diam = math.sqrt(1000 / math.pi)
        proximal = distal = Point3DWithDiam(x="0", y="0", z="0", diameter=str(diam))
        seg.proximal = proximal
        seg.distal = distal
        morph.segments.append(seg)
        hh_cell.morphology = morph

        hh_cell_doc.cells.append(hh_cell)
        pynml.write_neuroml2_file(nml2_doc=hh_cell_doc, nml2_file_name=hh_cell_fn, validate=True)
        return hh_cell_fn
:::
:::
:::
:::
:::

::: {#create-a-network .section}
##### Create a network[\#](#create-a-network "Link to this heading"){.headerlink}

::: {.cell .docutils .container}
::: {.cell_input .docutils .container}
::: {.highlight-ipython3 .notranslate}
::: highlight
    def create_network():
        """Create the network

        :returns: name of network nml file
        """
        net_doc = NeuroMLDocument(id="network",
                                  notes="HH cell network")
        net_doc_fn = "HH_example_net.nml"
        net_doc.includes.append(IncludeType(href=create_cell()))
        # Create a population: convenient to create many cells of the same type
        pop = Population(id="pop0", notes="A population for our cell", component="hh_cell", size=1)
        # Input
        pulsegen = PulseGenerator(id="pg", notes="Simple pulse generator", delay="100ms", duration="100ms", amplitude="0.08nA")

        exp_input = ExplicitInput(target="pop0[0]", input="pg")

        net = Network(id="single_hh_cell_network", note="A network with a single population")
        net_doc.pulse_generators.append(pulsegen)
        net.explicit_inputs.append(exp_input)
        net.populations.append(pop)
        net_doc.networks.append(net)

        pynml.write_neuroml2_file(nml2_doc=net_doc, nml2_file_name=net_doc_fn, validate=True)
        return net_doc_fn
:::
:::
:::
:::
:::
:::

::: {#plot-the-data-we-record .section}
#### Plot the data we record[\#](#plot-the-data-we-record "Link to this heading"){.headerlink}

::: {.cell .docutils .container}
::: {.cell_input .docutils .container}
::: {.highlight-ipython3 .notranslate}
::: highlight
    def plot_data(sim_id):
        """Plot the sim data.

        Load the data from the file and plot the graph for the membrane potential
        using the pynml generate_plot utility function.

        :sim_id: ID of simulaton

        """
        data_array = np.loadtxt(sim_id + ".dat")
        pynml.generate_plot([data_array[:, 0]], [data_array[:, 1]], "Membrane potential", show_plot_already=False, save_figure_to=sim_id + "-v.png", xaxis="time (s)", yaxis="membrane potential (V)")
        pynml.generate_plot([data_array[:, 0]], [data_array[:, 2]], "channel current", show_plot_already=False, save_figure_to=sim_id + "-i.png", xaxis="time (s)", yaxis="channel current (A)")
        pynml.generate_plot([data_array[:, 0], data_array[:, 0]], [data_array[:, 3], data_array[:, 4]], "current density", labels=["Na", "K"], show_plot_already=False, save_figure_to=sim_id + "-iden.png", xaxis="time (s)", yaxis="current density (A_per_m2)")
:::
:::
:::
:::
:::

::: {#create-and-run-the-simulation .section}
#### Create and run the simulation[\#](#create-and-run-the-simulation "Link to this heading"){.headerlink}

Create the simulation, run it, record data, and plot the recorded
information.

::: {.cell .docutils .container}
::: {.cell_input .docutils .container}
::: {.highlight-ipython3 .notranslate}
::: highlight
    def main():
        """Main function

        Include the NeuroML model into a LEMS simulation file, run it, plot some
        data.
        """
        # Simulation bits
        sim_id = "HH_single_compartment_example_sim"
        simulation = LEMSSimulation(sim_id=sim_id, duration=300, dt=0.01, simulation_seed=123)
        # Include the NeuroML model file
        simulation.include_neuroml2_file(create_network())
        # Assign target for the simulation
        simulation.assign_simulation_target("single_hh_cell_network")

        # Recording information from the simulation
        simulation.create_output_file(id="output0", file_name=sim_id + ".dat")
        simulation.add_column_to_output_file("output0", column_id="pop0[0]/v", quantity="pop0[0]/v")
        simulation.add_column_to_output_file("output0", column_id="pop0[0]/iChannels", quantity="pop0[0]/iChannels")
        simulation.add_column_to_output_file("output0", column_id="pop0[0]/na/iDensity", quantity="pop0[0]/hh_b_prop/membraneProperties/na_channels/iDensity/")
        simulation.add_column_to_output_file("output0", column_id="pop0[0]/k/iDensity", quantity="pop0[0]/hh_b_prop/membraneProperties/k_channels/iDensity/")

        # Save LEMS simulation to file
        sim_file = simulation.save_to_file()

        # Run the simulation using the default jNeuroML simulator
        pynml.run_lems_with_jneuroml(sim_file, max_memory="2G", nogui=True, plot=False)
        # Plot the data
        plot_data(sim_id)
:::
:::
:::
:::

::: {.cell .docutils .container}
::: {.cell_input .docutils .container}
::: {.highlight-ipython3 .notranslate}
::: highlight
    if __name__ == "__main__":
        main()
:::
:::
:::

::: {.cell_output .docutils .container}
::: {.output .stream .highlight-myst-ansi .notranslate}
::: highlight
    pyNeuroML >>> Written LEMS Simulation HH_single_compartment_example_sim to file: LEMS_HH_single_compartment_example_sim.xml
    pyNeuroML >>> Generating plot: Membrane potential
:::
:::

::: {.output .stderr .highlight-myst-ansi .notranslate}
::: highlight
    /usr/lib/python3.9/site-packages/pyneuroml/pynml.py:1688: MatplotlibDeprecationWarning: 
    The set_window_title function was deprecated in Matplotlib 3.4 and will be removed two minor releases later. Use manager.set_window_title or GUI-specific methods instead.
      fig.canvas.set_window_title(title)
    /usr/lib/python3.9/site-packages/pyneuroml/pynml.py:1727: UserWarning: marker is redundantly defined by the 'marker' keyword argument and the fmt string "o" (-> marker='o'). The keyword argument will take precedence.
      plt.plot(xvalues[i], yvalues[i], 'o', marker=marker, markersize=markersize, linestyle=linestyle, linewidth=linewidth, label=label)
    /usr/lib/python3.9/site-packages/pyneuroml/pynml.py:1688: MatplotlibDeprecationWarning: 
    The set_window_title function was deprecated in Matplotlib 3.4 and will be removed two minor releases later. Use manager.set_window_title or GUI-specific methods instead.
      fig.canvas.set_window_title(title)
    /usr/lib/python3.9/site-packages/pyneuroml/pynml.py:1727: UserWarning: marker is redundantly defined by the 'marker' keyword argument and the fmt string "o" (-> marker='o'). The keyword argument will take precedence.
      plt.plot(xvalues[i], yvalues[i], 'o', marker=marker, markersize=markersize, linestyle=linestyle, linewidth=linewidth, label=label)
:::
:::

::: {.output .stream .highlight-myst-ansi .notranslate}
::: highlight
    pyNeuroML >>> Saved image to HH_single_compartment_example_sim-v.png of plot: Membrane potential
    pyNeuroML >>> Generating plot: channel current
    pyNeuroML >>> Saved image to HH_single_compartment_example_sim-i.png of plot: channel current
    pyNeuroML >>> Generating plot: current density
    pyNeuroML >>> Saved image to HH_single_compartment_example_sim-iden.png of plot: current density
:::
:::

::: {.output .stderr .highlight-myst-ansi .notranslate}
::: highlight
    /usr/lib/python3.9/site-packages/pyneuroml/pynml.py:1688: MatplotlibDeprecationWarning: 
    The set_window_title function was deprecated in Matplotlib 3.4 and will be removed two minor releases later. Use manager.set_window_title or GUI-specific methods instead.
      fig.canvas.set_window_title(title)
    /usr/lib/python3.9/site-packages/pyneuroml/pynml.py:1727: UserWarning: marker is redundantly defined by the 'marker' keyword argument and the fmt string "o" (-> marker='o'). The keyword argument will take precedence.
      plt.plot(xvalues[i], yvalues[i], 'o', marker=marker, markersize=markersize, linestyle=linestyle, linewidth=linewidth, label=label)
    /usr/lib/python3.9/site-packages/pyneuroml/pynml.py:1727: UserWarning: marker is redundantly defined by the 'marker' keyword argument and the fmt string "o" (-> marker='o'). The keyword argument will take precedence.
      plt.plot(xvalues[i], yvalues[i], 'o', marker=marker, markersize=markersize, linestyle=linestyle, linewidth=linewidth, label=label)
:::
:::

![](_images/fce843b119e9ef931e2fb37076f9f8b9a2f35cda36396498580c7374c2b8a456.png)
![](_images/7d2474292dcb2b6a573030e67d9af7d7e579e2c3d8e56c1e32a56d343a0becb7.png)
![](_images/523586e534cca4ab457b29b08c7b91242a2585c1d118638969d1474917a40993.png)
:::
:::
:::
:::

[]{#document-Userdocs/MultiCompartmentOLMexample}

::: {#simulating-a-multi-compartment-olm-neuron .section .tex2jax_ignore .mathjax_ignore}
[]{#userdocs-getting-started-multi-compartment-example}

### Simulating a multi compartment OLM neuron[\#](#simulating-a-multi-compartment-olm-neuron "Link to this heading"){.headerlink}

In this section we will model and simulate a multi-compartment
Oriens-lacunosum moleculare (OLM) interneuron cell from the rodent
hippocampal CA1 network model developed by Bezaire et al.
([\[[BRB+16](Landing.html#id15 "Marianne J Bezaire, Ivan Raikov, Kelly Burk, Dhrumil Vyas, and Ivan Soltesz. Interneuronal mechanisms of hippocampal theta oscillations in a full-scale model of the rodent ca1 circuit. eLife, 5:e18566, dec 2016. URL: https://doi.org/10.7554/eLife.18566, doi:10.7554/eLife.18566."){.reference
.internal}\]]{#id1}). The complete network model can be seen [here on
GitHub](https://github.com/mbezaire/ca1){.reference .external}, and
[here on Open Source
Brain](https://www.opensourcebrain.org/projects/nc_ca1){.reference
.external}.

<figure id="id3" class="align-center">
<a href="_images/olm.cell.xy.png"
class="reference internal image-reference"><img
src="_images/olm.cell.xy.png" style="width: 50%;"
alt="Morphology of constructed OLM cell in xy plane" /></a>
<figcaption><p><span class="caption-number">Fig. 22 </span><span
class="caption-text">Morphology of OLM cell in xy plane</span><a
href="#id3" class="headerlink"
title="Link to this image">#</a></p></figcaption>
</figure>

<figure id="id4" class="align-center">
<a href="_images/olm_example_sim_seg0_soma0-v.png"
class="reference internal image-reference"><img
src="_images/olm_example_sim_seg0_soma0-v.png" style="width: 50%;"
alt="Membrane potential for neuron recorded from the simulation at the soma" /></a>
<figcaption><p><span class="caption-number">Fig. 23 </span><span
class="caption-text">Membrane potential of the simulated OLM cell at the
soma.</span><a href="#id4" class="headerlink"
title="Link to this image">#</a></p></figcaption>
</figure>

This plot, saved as [`olm_example_sim_seg0_soma0-v.png`{.docutils
.literal .notranslate}]{.pre} is generated using the following Python
NeuroML script:

::: {.highlight-python .notranslate}
::: highlight
    #!/usr/bin/env python3
    """
    Multi-compartmental OLM cell example

    File: olm-example.py

    Copyright 2023 NeuroML contributors
    Authors: Padraig Gleeson, Ankur Sinha
    """

    import neuroml
    from neuroml import NeuroMLDocument
    from neuroml.utils import component_factory
    from pyneuroml import pynml
    from pyneuroml.lems import LEMSSimulation
    from pyneuroml.plot.PlotMorphology import plot_2D
    import numpy as np


    def main():
        """Main function

        Include the NeuroML model into a LEMS simulation file, run it, plot some
        data.
        """
        # Simulation bits
        sim_id = "olm_example_sim"
        simulation = LEMSSimulation(sim_id=sim_id, duration=600, dt=0.01, simulation_seed=123)
        # Include the NeuroML model file
        simulation.include_neuroml2_file(create_olm_network())
        # Assign target for the simulation
        simulation.assign_simulation_target("single_olm_cell_network")

        # Recording information from the simulation
        simulation.create_output_file(id="output0", file_name=sim_id + ".dat")
        simulation.add_column_to_output_file("output0", column_id="pop0_0_v", quantity="pop0[0]/v")
        simulation.add_column_to_output_file("output0",
                                             column_id="pop0_0_v_Seg0_soma_0",
                                             quantity="pop0/0/olm/0/v")
        simulation.add_column_to_output_file("output0",
                                             column_id="pop0_0_v_Seg1_soma_0",
                                             quantity="pop0/0/olm/1/v")
        simulation.add_column_to_output_file("output0",
                                             column_id="pop0_0_v_Seg0_axon_0",
                                             quantity="pop0/0/olm/2/v")
        simulation.add_column_to_output_file("output0",
                                             column_id="pop0_0_v_Seg1_axon_0",
                                             quantity="pop0/0/olm/3/v")
        simulation.add_column_to_output_file("output0",
                                             column_id="pop0_0_v_Seg0_dend_0",
                                             quantity="pop0/0/olm/4/v")
        simulation.add_column_to_output_file("output0",
                                             column_id="pop0_0_v_Seg1_dend_0",
                                             quantity="pop0/0/olm/6/v")
        simulation.add_column_to_output_file("output0",
                                             column_id="pop0_0_v_Seg0_dend_1",
                                             quantity="pop0/0/olm/5/v")
        simulation.add_column_to_output_file("output0",
                                             column_id="pop0_0_v_Seg1_dend_1",
                                             quantity="pop0/0/olm/7/v")
        # Save LEMS simulation to file
        sim_file = simulation.save_to_file()

        # Run the simulation using the NEURON simulator
        pynml.run_lems_with_jneuroml_neuron(sim_file, max_memory="2G", nogui=True,
                                            plot=False, skip_run=False)
        # Plot the data
        plot_data(sim_id)


    def plot_data(sim_id):
        """Plot the sim data.

        Load the data from the file and plot the graph for the membrane potential
        using the pynml generate_plot utility function.

        :sim_id: ID of simulaton

        """
        data_array = np.loadtxt(sim_id + ".dat")
        pynml.generate_plot([data_array[:, 0]], [data_array[:, 1]], "Membrane potential (soma seg 0)", show_plot_already=False, save_figure_to=sim_id + "_seg0_soma0-v.png", xaxis="time (s)", yaxis="membrane potential (V)")
        pynml.generate_plot([data_array[:, 0]], [data_array[:, 2]], "Membrane potential (soma seg 1)", show_plot_already=False, save_figure_to=sim_id + "_seg1_soma0-v.png", xaxis="time (s)", yaxis="membrane potential (V)")
        pynml.generate_plot([data_array[:, 0]], [data_array[:, 3]], "Membrane potential (axon seg 0)", show_plot_already=False, save_figure_to=sim_id + "_seg0_axon0-v.png", xaxis="time (s)", yaxis="membrane potential (V)")
        pynml.generate_plot([data_array[:, 0]], [data_array[:, 4]], "Membrane potential (axon seg 1)", show_plot_already=False, save_figure_to=sim_id + "_seg1_axon0-v.png", xaxis="time (s)", yaxis="membrane potential (V)")


    def create_olm_network():
        """Create the network

        :returns: name of network nml file
        """
        net_doc = NeuroMLDocument(id="network",
                                  notes="OLM cell network")
        net_doc_fn = "olm_example_net.nml"
        net_doc.add("IncludeType", href=create_olm_cell())
        net = net_doc.add("Network", id="single_olm_cell_network", validate=False)
        # Create a population: convenient to create many cells of the same type
        pop = net.add("Population", id="pop0", notes="A population for our cell",
                      component="olm", size=1, type="populationList",
                      validate=False)
        pop.add("Instance", id=0, location=component_factory("Location", x=0., y=0., z=0.))
        # Input
        net_doc.add("PulseGenerator", id="pg_olm", notes="Simple pulse generator", delay="100ms", duration="100ms", amplitude="0.08nA")

        net.add("ExplicitInput", target="pop0[0]", input="pg_olm")

        pynml.write_neuroml2_file(nml2_doc=net_doc, nml2_file_name=net_doc_fn, validate=True)
        return net_doc_fn


    def create_olm_cell():
        """Create the complete cell.

        :returns: cell object
        """
        nml_cell_doc = component_factory("NeuroMLDocument", id="oml_cell")
        cell = nml_cell_doc.add("Cell", id="olm", neuro_lex_id="NLXCELL:091206")  # type neuroml.Cell
        nml_cell_file = cell.id + ".cell.nml"

        cell.summary()
        cell.info(show_contents=True)
        cell.morphology.info(show_contents=True)

        # Add two soma segments to an unbranched segment group
        cell.add_unbranched_segment_group("soma_0")
        diam = 10.0
        soma_0 = cell.add_segment(
            prox=[0.0, 0.0, 0.0, diam],
            dist=[0.0, 10., 0.0, diam],
            name="Seg0_soma_0",
            group_id="soma_0",
            seg_type="soma"
        )

        soma_1 = cell.add_segment(
            prox=None,
            dist=[0.0, 10. + 10., 0.0, diam],
            name="Seg1_soma_0",
            parent=soma_0,
            group_id="soma_0",
            seg_type="soma"
        )

        # Add axon segments
        diam = 1.5
        cell.add_unbranched_segments(
            [
                [0.0, 0.0, 0.0, diam],
                [0.0, -75, 0.0, diam],
                [0.0, -150, 0.0, diam],
            ],
            parent=soma_0,
            fraction_along=0.0,
            group_id="axon_0",
            seg_type="axon"
        )

        # Add 2 dendrite segments, using the branching utility function

        diam = 3.0
        cell.add_unbranched_segments(
            [
                [0.0, 20.0, 0.0, diam],
                [100, 120, 0.0, diam],
                [177, 197, 0.0, diam],
            ],
            parent=soma_1,
            fraction_along=1.0,
            group_id="dend_0",
            seg_type="dendrite"
        )

        cell.add_unbranched_segments(
            [
                [0.0, 20.0, 0.0, diam],
                [-100, 120, 0.0, diam],
                [-177, 197, 0.0, diam],
            ],
            parent=soma_1,
            fraction_along=1.0,
            group_id="dend_1",
            seg_type="dendrite"
        )

        # color groups for morphology plots
        den_seg_group = cell.get_segment_group("dendrite_group")
        den_seg_group.add("Property", tag="color", value="0.8 0 0")

        ax_seg_group = cell.get_segment_group("axon_group")
        ax_seg_group.add("Property", tag="color", value="0 0.8 0")

        soma_seg_group = cell.get_segment_group("soma_group")
        soma_seg_group.add("Property", tag="color", value="0 0 0.8")

        # Other cell properties
        cell.set_init_memb_potential("-67mV")
        cell.set_resistivity("0.15 kohm_cm")
        cell.set_specific_capacitance("1.3 uF_per_cm2")
        cell.set_spike_thresh("-20 mV")

        # channels
        # leak
        cell.add_channel_density(nml_cell_doc,
                                 cd_id="leak_all",
                                 cond_density="0.01 mS_per_cm2",
                                 ion_channel="leak_chan",
                                 ion_chan_def_file="olm-example/leak_chan.channel.nml",
                                 erev="-67mV",
                                 ion="non_specific")
        # HCNolm_soma
        cell.add_channel_density(nml_cell_doc,
                                 cd_id="HCNolm_soma",
                                 cond_density="0.5 mS_per_cm2",
                                 ion_channel="HCNolm",
                                 ion_chan_def_file="olm-example/HCNolm.channel.nml",
                                 erev="-32.9mV",
                                 ion="h",
                                 group_id="soma_group")
        # Kdrfast_soma
        cell.add_channel_density(nml_cell_doc,
                                 cd_id="Kdrfast_soma",
                                 cond_density="73.37 mS_per_cm2",
                                 ion_channel="Kdrfast",
                                 ion_chan_def_file="olm-example/Kdrfast.channel.nml",
                                 erev="-77mV",
                                 ion="k",
                                 group_id="soma_group")
        # Kdrfast_dendrite
        cell.add_channel_density(nml_cell_doc,
                                 cd_id="Kdrfast_dendrite",
                                 cond_density="105.8 mS_per_cm2",
                                 ion_channel="Kdrfast",
                                 ion_chan_def_file="olm-example/Kdrfast.channel.nml",
                                 erev="-77mV",
                                 ion="k",
                                 group_id="dendrite_group")
        # Kdrfast_axon
        cell.add_channel_density(nml_cell_doc,
                                 cd_id="Kdrfast_axon",
                                 cond_density="117.392 mS_per_cm2",
                                 ion_channel="Kdrfast",
                                 ion_chan_def_file="olm-example/Kdrfast.channel.nml",
                                 erev="-77mV",
                                 ion="k",
                                 group_id="axon_group")
        # KvAolm_soma
        cell.add_channel_density(nml_cell_doc,
                                 cd_id="KvAolm_soma",
                                 cond_density="4.95 mS_per_cm2",
                                 ion_channel="KvAolm",
                                 ion_chan_def_file="olm-example/KvAolm.channel.nml",
                                 erev="-77mV",
                                 ion="k",
                                 group_id="soma_group")
        # KvAolm_dendrite
        cell.add_channel_density(nml_cell_doc,
                                 cd_id="KvAolm_dendrite",
                                 cond_density="2.8 mS_per_cm2",
                                 ion_channel="KvAolm",
                                 ion_chan_def_file="olm-example/KvAolm.channel.nml",
                                 erev="-77mV",
                                 ion="k",
                                 group_id="dendrite_group")
        # Nav_soma
        cell.add_channel_density(nml_cell_doc,
                                 cd_id="Nav_soma",
                                 cond_density="10.7 mS_per_cm2",
                                 ion_channel="Nav",
                                 ion_chan_def_file="olm-example/Nav.channel.nml",
                                 erev="50mV",
                                 ion="na",
                                 group_id="soma_group")
        # Nav_dendrite
        cell.add_channel_density(nml_cell_doc,
                                 cd_id="Nav_dendrite",
                                 cond_density="23.4 mS_per_cm2",
                                 ion_channel="Nav",
                                 ion_chan_def_file="olm-example/Nav.channel.nml",
                                 erev="50mV",
                                 ion="na",
                                 group_id="dendrite_group")
        # Nav_axon
        cell.add_channel_density(nml_cell_doc,
                                 cd_id="Nav_axon",
                                 cond_density="17.12 mS_per_cm2",
                                 ion_channel="Nav",
                                 ion_chan_def_file="olm-example/Nav.channel.nml",
                                 erev="50mV",
                                 ion="na",
                                 group_id="axon_group")

        cell.optimise_segment_groups()
        cell.validate(recursive=True)
        pynml.write_neuroml2_file(nml_cell_doc, nml_cell_file, True, True)
        plot_2D(nml_cell_file, plane2d="xy", nogui=True,
                save_to_file="olm.cell.xy.png")
        return nml_cell_file


    if __name__ == "__main__":
        main()
:::
:::

::: {#declaring-the-model-in-neuroml .section}
#### Declaring the model in NeuroML[\#](#declaring-the-model-in-neuroml "Link to this heading"){.headerlink}

Similar to previous examples, we will first declare the model, visualise
it, and then simulate it. The OLM model is slightly more complex than
the HH neuron model we had worked with in the [[previous tutorial]{.std
.std-ref}](Landing.html#userdocs-getting-started-single-compartment-example){.reference
.internal} since it includes multiple compartments. However, where we
had declared the ion-channels ourselves in the previous example, here
will will not do so. We will *include* channels that have been
pre-defined in NeuroML to demonstrate how components defined in NeuroML
can be easily re-used in models.

We will follow the same method as before. We will first define the cell,
create a network with one instance of the cell, and then simulate it to
record and plot the membrane potential from different segments.

::: {#declaring-the-cell .section}
##### Declaring the cell[\#](#declaring-the-cell "Link to this heading"){.headerlink}

To keep our Python script modularised, we start constructing our
[[cell]{.std .std-ref}](Landing.html#schema-cell){.reference .internal}
in a separate function.

::: {.highlight-python .notranslate}
::: highlight
    def create_olm_cell():
        """Create the complete cell.

        :returns: cell object
        """
        nml_cell_doc = component_factory("NeuroMLDocument", id="oml_cell")
        cell = nml_cell_doc.add("Cell", id="olm", neuro_lex_id="NLXCELL:091206")  # type neuroml.Cell
        nml_cell_file = cell.id + ".cell.nml"

        cell.summary()
        cell.info(show_contents=True)
        cell.morphology.info(show_contents=True)

        # Add two soma segments to an unbranched segment group
        cell.add_unbranched_segment_group("soma_0")
        diam = 10.0
        soma_0 = cell.add_segment(
            prox=[0.0, 0.0, 0.0, diam],
            dist=[0.0, 10., 0.0, diam],
            name="Seg0_soma_0",
            group_id="soma_0",
            seg_type="soma"
        )

        soma_1 = cell.add_segment(
            prox=None,
            dist=[0.0, 10. + 10., 0.0, diam],
            name="Seg1_soma_0",
            parent=soma_0,
            group_id="soma_0",
            seg_type="soma"
        )

        # Add axon segments
        diam = 1.5
        cell.add_unbranched_segments(
            [
                [0.0, 0.0, 0.0, diam],
                [0.0, -75, 0.0, diam],
                [0.0, -150, 0.0, diam],
            ],
            parent=soma_0,
            fraction_along=0.0,
            group_id="axon_0",
            seg_type="axon"
        )

        # Add 2 dendrite segments, using the branching utility function

        diam = 3.0
        cell.add_unbranched_segments(
            [
                [0.0, 20.0, 0.0, diam],
                [100, 120, 0.0, diam],
                [177, 197, 0.0, diam],
            ],
            parent=soma_1,
            fraction_along=1.0,
            group_id="dend_0",
            seg_type="dendrite"
        )

        cell.add_unbranched_segments(
            [
                [0.0, 20.0, 0.0, diam],
                [-100, 120, 0.0, diam],
                [-177, 197, 0.0, diam],
            ],
            parent=soma_1,
            fraction_along=1.0,
            group_id="dend_1",
            seg_type="dendrite"
        )

        # color groups for morphology plots
        den_seg_group = cell.get_segment_group("dendrite_group")
        den_seg_group.add("Property", tag="color", value="0.8 0 0")

        ax_seg_group = cell.get_segment_group("axon_group")
        ax_seg_group.add("Property", tag="color", value="0 0.8 0")

        soma_seg_group = cell.get_segment_group("soma_group")
        soma_seg_group.add("Property", tag="color", value="0 0 0.8")

        # Other cell properties
        cell.set_init_memb_potential("-67mV")
        cell.set_resistivity("0.15 kohm_cm")
        cell.set_specific_capacitance("1.3 uF_per_cm2")
        cell.set_spike_thresh("-20 mV")

        # channels
        # leak
        cell.add_channel_density(nml_cell_doc,
                                 cd_id="leak_all",
                                 cond_density="0.01 mS_per_cm2",
                                 ion_channel="leak_chan",
                                 ion_chan_def_file="olm-example/leak_chan.channel.nml",
                                 erev="-67mV",
                                 ion="non_specific")
        # HCNolm_soma
        cell.add_channel_density(nml_cell_doc,
                                 cd_id="HCNolm_soma",
                                 cond_density="0.5 mS_per_cm2",
                                 ion_channel="HCNolm",
                                 ion_chan_def_file="olm-example/HCNolm.channel.nml",
                                 erev="-32.9mV",
                                 ion="h",
                                 group_id="soma_group")
        # Kdrfast_soma
        cell.add_channel_density(nml_cell_doc,
                                 cd_id="Kdrfast_soma",
                                 cond_density="73.37 mS_per_cm2",
                                 ion_channel="Kdrfast",
                                 ion_chan_def_file="olm-example/Kdrfast.channel.nml",
                                 erev="-77mV",
                                 ion="k",
                                 group_id="soma_group")
        # Kdrfast_dendrite
        cell.add_channel_density(nml_cell_doc,
                                 cd_id="Kdrfast_dendrite",
                                 cond_density="105.8 mS_per_cm2",
                                 ion_channel="Kdrfast",
                                 ion_chan_def_file="olm-example/Kdrfast.channel.nml",
                                 erev="-77mV",
                                 ion="k",
                                 group_id="dendrite_group")
        # Kdrfast_axon
        cell.add_channel_density(nml_cell_doc,
                                 cd_id="Kdrfast_axon",
                                 cond_density="117.392 mS_per_cm2",
                                 ion_channel="Kdrfast",
                                 ion_chan_def_file="olm-example/Kdrfast.channel.nml",
                                 erev="-77mV",
                                 ion="k",
                                 group_id="axon_group")
        # KvAolm_soma
        cell.add_channel_density(nml_cell_doc,
                                 cd_id="KvAolm_soma",
                                 cond_density="4.95 mS_per_cm2",
                                 ion_channel="KvAolm",
                                 ion_chan_def_file="olm-example/KvAolm.channel.nml",
                                 erev="-77mV",
                                 ion="k",
                                 group_id="soma_group")
        # KvAolm_dendrite
        cell.add_channel_density(nml_cell_doc,
                                 cd_id="KvAolm_dendrite",
                                 cond_density="2.8 mS_per_cm2",
                                 ion_channel="KvAolm",
                                 ion_chan_def_file="olm-example/KvAolm.channel.nml",
                                 erev="-77mV",
                                 ion="k",
                                 group_id="dendrite_group")
        # Nav_soma
        cell.add_channel_density(nml_cell_doc,
                                 cd_id="Nav_soma",
                                 cond_density="10.7 mS_per_cm2",
                                 ion_channel="Nav",
                                 ion_chan_def_file="olm-example/Nav.channel.nml",
                                 erev="50mV",
                                 ion="na",
                                 group_id="soma_group")
        # Nav_dendrite
        cell.add_channel_density(nml_cell_doc,
                                 cd_id="Nav_dendrite",
                                 cond_density="23.4 mS_per_cm2",
                                 ion_channel="Nav",
                                 ion_chan_def_file="olm-example/Nav.channel.nml",
                                 erev="50mV",
                                 ion="na",
                                 group_id="dendrite_group")
        # Nav_axon
        cell.add_channel_density(nml_cell_doc,
                                 cd_id="Nav_axon",
                                 cond_density="17.12 mS_per_cm2",
                                 ion_channel="Nav",
                                 ion_chan_def_file="olm-example/Nav.channel.nml",
                                 erev="50mV",
                                 ion="na",
                                 group_id="axon_group")
:::
:::

Let us walk through this function:

::: {.highlight-python .notranslate}
::: highlight
        nml_cell_doc = component_factory("NeuroMLDocument", id="oml_cell")
        cell = nml_cell_doc.add("Cell", id="olm", neuro_lex_id="NLXCELL:091206")  # type neuroml.Cell
        nml_cell_file = cell.id + ".cell.nml"

        cell.summary()
        cell.info(show_contents=True)
        cell.morphology.info(show_contents=True)
:::
:::

We create a new model document that will hold the cell model. Then, we
create and add a new [[Cell]{.std
.std-ref}](Landing.html#schema-cell){.reference .internal} using the
[`add`{.docutils .literal .notranslate}]{.pre} method to the document.
We also provide a [`neuro_lex_id`{.docutils .literal
.notranslate}]{.pre} here, which is the [NeuroLex ontology
identifier](https://scicrunch.org/scicrunch/interlex/view/ilx_0105030?searchTerm=oriens-lacunosum%20moleculare){.reference
.external}. This allows us to better connect models to biological
concepts.

As we have seen in the [[single Izhikevich neuron example]{.std
.std-ref}](Landing.html#userdocs-getting-started-single-example-supplementary-add){.reference
.internal}, the [`add`{.docutils .literal .notranslate}]{.pre} method
calls the [`component_factory`{.docutils .literal .notranslate}]{.pre}
to create the component object for us. For the [`Cell`{.docutils
.literal .notranslate}]{.pre} component type, it does a number of extra
things for us to set up, or initialise, the cell.

We have a number of ways of inspecting the cell. The
[`summary`{.docutils .literal .notranslate}]{.pre} function provides a
very short summary of the cell. This is useful to quickly get a high
level overview of it:

::: {.highlight-pycon .notranslate}
::: highlight
    >>> cell.summary()
    *******************************************************
    * Cell: olm
    * Notes: None
    * Segments: 0
    * SegmentGroups: 4
    *******************************************************
:::
:::

We can also use the general [[info function]{.std
.std-ref}](Landing.html#userdocs-getting-started-single-example-supplementary-info){.reference
.internal} to inspect the cell:

::: {.highlight-pycon .notranslate}
::: highlight
    >>> cell.info(show_contents=True)
    Cell -- Cell with  **segment** s specified in a  **morphology**  element along with details on its  **biophysicalProperties** . NOTE: this can only be correctly simulated using jLEMS when there is a single segment in the cell, and **v**  of this cell represents the membrane potential in that isopotential segment.

    Please see the NeuroML standard schema documentation at https://docs.neuroml.org/Userdocs/NeuroMLv2.html for more information.

    Valid members for Cell are:
    * biophysical_properties_attr (class: NmlId, Optional)
    * morphology (class: Morphology, Optional)
            * Contents ('ids'/<objects>): 'morphology'

    * neuro_lex_id (class: NeuroLexId, Optional)
            * Contents ('ids'/<objects>): NLXCELL:091206

    * metaid (class: MetaId, Optional)
    * biophysical_properties (class: BiophysicalProperties, Optional)
            * Contents ('ids'/<objects>): 'biophys'

    * id (class: NmlId, Required)
            * Contents ('ids'/<objects>): olm

    * notes (class: xs:string, Optional)
    * properties (class: Property, Optional)
    * annotation (class: Annotation, Optional)
    * morphology_attr (class: NmlId, Optional)
:::
:::

We see the cell already contains [`biophysical_properties`{.docutils
.literal .notranslate}]{.pre} or [`morphology`{.docutils .literal
.notranslate}]{.pre}. Because these are components of the cell that are
expected to be used, these were added automatically for us when the new
component was created.

Let us take a look at the morphology of the cell:

::: {.highlight-pycon .notranslate}
::: highlight
    >>> cell.morphology.info(show_contents=True)
    Morphology -- The collection of  **segment** s which specify the 3D structure of the cell, along with a number of  **segmentGroup** s

    Please see the NeuroML standard schema documentation at https://docs.neuroml.org/Userdocs/NeuroMLv2.html for more information.

    Valid members for Morphology are:
    * segments (class: Segment, Required)
    * metaid (class: MetaId, Optional)
    * segment_groups (class: SegmentGroup, Optional)
            * Contents ('ids'/<objects>): ['all', 'soma_group', 'axon_group', 'dendrite_group']

    * id (class: NmlId, Required)
            * Contents ('ids'/<objects>): morphology

    * notes (class: xs:string, Optional)
    * properties (class: Property, Optional)
    * annotation (class: Annotation, Optional)
:::
:::

We see that there are no segments in the cell because we have not added
any. However, there are already a number of "default" segment groups
that were automatically added for us: [`all`{.docutils .literal
.notranslate}]{.pre}, [`soma_group`{.docutils .literal
.notranslate}]{.pre}, [`axon_group`{.docutils .literal
.notranslate}]{.pre}, [`dendrite_group`{.docutils .literal
.notranslate}]{.pre}. These groups allow us to keep track of all the
segments, and of the segments forming the soma, the axon, and the
dendrites of the cell respectively. Take a look at the [[conventions
page]{.std
.std-ref}](Landing.html#userdocs-conventions-segments){.reference
.internal} for more information on these.

We now have an empty cell. Since we are building a multi-compartmental
cell, we now proceed to define the detailed morphology of the cell. We
do this by adding [[segments]{.std
.std-ref}](Landing.html#schema-segment){.reference .internal} and
grouping them in to [[segment groups]{.std
.std-ref}](Landing.html#schema-segmentgroup){.reference .internal}. We
can add segments using the [`add_segment`{.docutils .literal
.notranslate}]{.pre} utility function, as we do for the segments forming
the soma. Here, our soma has two segments.

::: {.highlight-python .notranslate}
::: highlight
        # Add two soma segments to an unbranched segment group
        cell.add_unbranched_segment_group("soma_0")
        diam = 10.0
        soma_0 = cell.add_segment(
            prox=[0.0, 0.0, 0.0, diam],
            dist=[0.0, 10., 0.0, diam],
            name="Seg0_soma_0",
            group_id="soma_0",
            seg_type="soma"
        )

        soma_1 = cell.add_segment(
            prox=None,
            dist=[0.0, 10. + 10., 0.0, diam],
            name="Seg1_soma_0",
            parent=soma_0,
            group_id="soma_0",
            seg_type="soma"
        )
:::
:::

The utility function takes the dimensions of the segment---it's
[[proximal]{.std .std-ref}](Landing.html#schema-proximal){.reference
.internal} and [[distal]{.std
.std-ref}](Landing.html#schema-distal){.reference .internal}
co-ordinates and the diameter to create a segment of the provided name.
Additionally, since segments need to be contiguous, it makes the first
segment the *parent* of the second, to build a chain. Finally, it places
the segment into the specified segment group and the default groups that
we also have and adds the segment to the cell's morphology.

Note that by default, the [`add_segment`{.docutils .literal
.notranslate}]{.pre} function does not know if the segments are
contiguous, i.e., that they form an unbranched branch of the cell. We
could have added segments here that do not line up in a chain, when
building different parts of a cell for example. In this case, we know
that the two soma segments must be contiguous, and that they are on the
same unbranched branch (i.e. a continuous section without any branching
points on it), so we create an unbranched segment group first using the
[`add_unbranched_segment_group`{.docutils .literal .notranslate}]{.pre}.

If we were only creating cell morphologies, this would not not matter
much. Even if the two segments were not included in a group of
unbranched segments, they would still be connected. However, for
simulation, simulators such as NEURON need to know which parts of the
cell form unbranched sections so that they can apply the [cable
equation](https://en.wikipedia.org/wiki/Cable_theory#Deriving_the_cable_equation){.reference
.external} and break them into smaller segments to simulate the electric
current through them. (See
[\[[CGH+07](Landing.html#id8 "Sharon Crook, Padraig Gleeson, Fred Howell, Joseph Svitak, and R. Angus Silver. MorphML: level 1 of the NeuroML standards for neuronal morphology data and model specification. Neuroinformatics, 5(2):96–104, 5 2007. doi:10.1007/s12021-007-0003-6."){.reference
.internal}\]]{#id2} for more information on how different simulators
simulate cells with detailed morphologies.)

Next, we can call the same functions multiple times to add soma,
dendritic, and axonal segments to our cell but this can get quite
lengthy. To easily add unbranched contiguous lists of segments to the
cell, we can directly use the [`add_unbranched_segments`{.docutils
.literal .notranslate}]{.pre} utility function. Here we use it to create
an axonal segment group, and two dendritic groups each with two
segments. The first point we provide is the proximal (starting) of the
dendrite. The next two points are the distal (ends) of each segment
forming the section.

::: {.highlight-python .notranslate}
::: highlight
        # Add axon segments
        diam = 1.5
        cell.add_unbranched_segments(
            [
                [0.0, 0.0, 0.0, diam],
                [0.0, -75, 0.0, diam],
                [0.0, -150, 0.0, diam],
            ],
            parent=soma_0,
            fraction_along=0.0,
            group_id="axon_0",
            seg_type="axon"
        )

        # Add 2 dendrite segments, using the branching utility function

        diam = 3.0
        cell.add_unbranched_segments(
            [
                [0.0, 20.0, 0.0, diam],
                [100, 120, 0.0, diam],
                [177, 197, 0.0, diam],
            ],
            parent=soma_1,
            fraction_along=1.0,
            group_id="dend_0",
            seg_type="dendrite"
        )

        cell.add_unbranched_segments(
            [
                [0.0, 20.0, 0.0, diam],
                [-100, 120, 0.0, diam],
                [-177, 197, 0.0, diam],
            ],
            parent=soma_1,
            fraction_along=1.0,
            group_id="dend_1",
            seg_type="dendrite"
        )
:::
:::

We repeat this process to create more dendritic and axonal sections of
contiguous segments.

Finally, we add an extra colour property to the three primary segment
groups that can be used when generating morphology graphs:

::: {.highlight-python .notranslate}
::: highlight
        # color groups for morphology plots
        den_seg_group = cell.get_segment_group("dendrite_group")
        den_seg_group.add("Property", tag="color", value="0.8 0 0")

        ax_seg_group = cell.get_segment_group("axon_group")
        ax_seg_group.add("Property", tag="color", value="0 0.8 0")

        soma_seg_group = cell.get_segment_group("soma_group")
        soma_seg_group.add("Property", tag="color", value="0 0 0.8")
:::
:::

We have now completed adding the morphological information to our cell.
Next, we proceed to our [[biophysical properties]{.std
.std-ref}](Landing.html#schema-biophysicalproperties){.reference
.internal}, e.g.:

-   the [[membrane properties]{.std
    .std-ref}](Landing.html#schema-membraneproperties){.reference
    .internal}

    -   [[spike threshold]{.std
        .std-ref}](Landing.html#schema-spikethresh){.reference
        .internal}

    -   [[initial membrane potential]{.std
        .std-ref}](Landing.html#schema-initmembpotential){.reference
        .internal}

    -   [[channel densities]{.std
        .std-ref}](Landing.html#schema-basechanneldensity){.reference
        .internal}

    -   [[specifc capacitances]{.std
        .std-ref}](Landing.html#schema-specificcapacitance){.reference
        .internal}

-   the [[intracellular properties]{.std
    .std-ref}](Landing.html#schema-intracellularproperties){.reference
    .internal}

    -   [[resistivity]{.std
        .std-ref}](Landing.html#schema-resistivity){.reference
        .internal}

We use more helpful utility functions to set these values

::: {.highlight-python .notranslate}
::: highlight
        # Other cell properties
        cell.set_init_memb_potential("-67mV")
        cell.set_resistivity("0.15 kohm_cm")
        cell.set_specific_capacitance("1.3 uF_per_cm2")
:::
:::

For setting channel densities, we have the
[`add_channel_density`{.docutils .literal .notranslate}]{.pre} function:

::: {.highlight-python .notranslate}
::: highlight
        # channels
        # leak
        cell.add_channel_density(nml_cell_doc,
                                 cd_id="leak_all",
                                 cond_density="0.01 mS_per_cm2",
                                 ion_channel="leak_chan",
                                 ion_chan_def_file="olm-example/leak_chan.channel.nml",
                                 erev="-67mV",
                                 ion="non_specific")
        # HCNolm_soma
        cell.add_channel_density(nml_cell_doc,
                                 cd_id="HCNolm_soma",
                                 cond_density="0.5 mS_per_cm2",
                                 ion_channel="HCNolm",
                                 ion_chan_def_file="olm-example/HCNolm.channel.nml",
                                 erev="-32.9mV",
                                 ion="h",
                                 group_id="soma_group")
        # Kdrfast_soma
        cell.add_channel_density(nml_cell_doc,
                                 cd_id="Kdrfast_soma",
                                 cond_density="73.37 mS_per_cm2",
                                 ion_channel="Kdrfast",
                                 ion_chan_def_file="olm-example/Kdrfast.channel.nml",
                                 erev="-77mV",
                                 ion="k",
                                 group_id="soma_group")
        # Kdrfast_dendrite
        cell.add_channel_density(nml_cell_doc,
                                 cd_id="Kdrfast_dendrite",
                                 cond_density="105.8 mS_per_cm2",
                                 ion_channel="Kdrfast",
                                 ion_chan_def_file="olm-example/Kdrfast.channel.nml",
                                 erev="-77mV",
                                 ion="k",
                                 group_id="dendrite_group")
        # Kdrfast_axon
        cell.add_channel_density(nml_cell_doc,
                                 cd_id="Kdrfast_axon",
                                 cond_density="117.392 mS_per_cm2",
                                 ion_channel="Kdrfast",
                                 ion_chan_def_file="olm-example/Kdrfast.channel.nml",
                                 erev="-77mV",
                                 ion="k",
                                 group_id="axon_group")
        # KvAolm_soma
        cell.add_channel_density(nml_cell_doc,
                                 cd_id="KvAolm_soma",
                                 cond_density="4.95 mS_per_cm2",
                                 ion_channel="KvAolm",
                                 ion_chan_def_file="olm-example/KvAolm.channel.nml",
                                 erev="-77mV",
                                 ion="k",
                                 group_id="soma_group")
        # KvAolm_dendrite
        cell.add_channel_density(nml_cell_doc,
                                 cd_id="KvAolm_dendrite",
                                 cond_density="2.8 mS_per_cm2",
                                 ion_channel="KvAolm",
                                 ion_chan_def_file="olm-example/KvAolm.channel.nml",
                                 erev="-77mV",
                                 ion="k",
                                 group_id="dendrite_group")
        # Nav_soma
        cell.add_channel_density(nml_cell_doc,
                                 cd_id="Nav_soma",
                                 cond_density="10.7 mS_per_cm2",
                                 ion_channel="Nav",
                                 ion_chan_def_file="olm-example/Nav.channel.nml",
                                 erev="50mV",
                                 ion="na",
                                 group_id="soma_group")
        # Nav_dendrite
        cell.add_channel_density(nml_cell_doc,
                                 cd_id="Nav_dendrite",
                                 cond_density="23.4 mS_per_cm2",
                                 ion_channel="Nav",
                                 ion_chan_def_file="olm-example/Nav.channel.nml",
                                 erev="50mV",
                                 ion="na",
                                 group_id="dendrite_group")
        # Nav_axon
        cell.add_channel_density(nml_cell_doc,
                                 cd_id="Nav_axon",
                                 cond_density="17.12 mS_per_cm2",
                                 ion_channel="Nav",
                                 ion_chan_def_file="olm-example/Nav.channel.nml",
                                 erev="50mV",
                                 ion="na",
:::
:::

Note that we are not writing our channel files from scratch here. We are
re-using already written NeuroML channel definitions by simply including
their NeuroML definition files.

This completes the definition of our cell. We now run the level one
validation, write it to a file while also running a complete (level one
and level two) validation using pyNeuroML. We also generate the
morphology plot shown on the top of this page.

::: {.highlight-python .notranslate}
::: highlight
        cell.optimise_segment_groups()
        cell.validate(recursive=True)
        pynml.write_neuroml2_file(nml_cell_doc, nml_cell_file, True, True)
:::
:::

The resulting NeuroML file is:

::: {.highlight-xml .notranslate}
::: highlight
    <neuroml xmlns="http://www.neuroml.org/schema/neuroml2"  xmlns:xs="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.neuroml.org/schema/neuroml2 https://raw.github.com/NeuroML/NeuroML2/development/Schemas/NeuroML2/NeuroML_v2.3.1.xsd" id="oml_cell">
        <include href="olm-example/leak_chan.channel.nml"/>
        <include href="olm-example/HCNolm.channel.nml"/>
        <include href="olm-example/Kdrfast.channel.nml"/>
        <include href="olm-example/KvAolm.channel.nml"/>
        <include href="olm-example/Nav.channel.nml"/>
        <cell id="olm" neuroLexId="NLXCELL:091206">
            <morphology id="morphology">
                <segment id="0" name="Seg0_soma_0">
                    <proximal x="0.0" y="0.0" z="0.0" diameter="10.0"/>
                    <distal x="0.0" y="10.0" z="0.0" diameter="10.0"/>
                </segment>
                <segment id="1" name="Seg1_soma_0">
                    <parent segment="0"/>
                    <distal x="0.0" y="20.0" z="0.0" diameter="10.0"/>
                </segment>
                <segment id="2" name="Seg0_axon_0">
                    <parent segment="0" fractionAlong="0.0"/>
                    <proximal x="0.0" y="0.0" z="0.0" diameter="1.5"/>
                    <distal x="0.0" y="-75.0" z="0.0" diameter="1.5"/>
                </segment>
                <segment id="3" name="Seg1_axon_0">
                    <parent segment="2"/>
                    <proximal x="0.0" y="-75.0" z="0.0" diameter="1.5"/>
                    <distal x="0.0" y="-150.0" z="0.0" diameter="1.5"/>
                </segment>
                <segment id="4" name="Seg0_dend_0">
                    <parent segment="1"/>
                    <proximal x="0.0" y="20.0" z="0.0" diameter="3.0"/>
                    <distal x="100.0" y="120.0" z="0.0" diameter="3.0"/>
                </segment>
                <segment id="5" name="Seg1_dend_0">
                    <parent segment="4"/>
                    <proximal x="100.0" y="120.0" z="0.0" diameter="3.0"/>
                    <distal x="177.0" y="197.0" z="0.0" diameter="3.0"/>
                </segment>
                <segment id="6" name="Seg0_dend_1">
                    <parent segment="1"/>
                    <proximal x="0.0" y="20.0" z="0.0" diameter="3.0"/>
                    <distal x="-100.0" y="120.0" z="0.0" diameter="3.0"/>
                </segment>
                <segment id="7" name="Seg1_dend_1">
                    <parent segment="6"/>
                    <proximal x="-100.0" y="120.0" z="0.0" diameter="3.0"/>
                    <distal x="-177.0" y="197.0" z="0.0" diameter="3.0"/>
                </segment>
                <segmentGroup id="soma_0" neuroLexId="sao864921383">
                    <member segment="0"/>
                    <member segment="1"/>
                </segmentGroup>
                <segmentGroup id="axon_0" neuroLexId="sao864921383">
                    <member segment="2"/>
                    <member segment="3"/>
                </segmentGroup>
                <segmentGroup id="dend_0" neuroLexId="sao864921383">
                    <member segment="4"/>
                    <member segment="5"/>
                </segmentGroup>
                <segmentGroup id="dend_1" neuroLexId="sao864921383">
                    <member segment="6"/>
                    <member segment="7"/>
                </segmentGroup>
                <segmentGroup id="soma_group" neuroLexId="GO:0043025">
                    <notes>Default soma segment group for the cell</notes>
                    <property tag="color" value="0 0 0.8"/>
                    <include segmentGroup="soma_0"/>
                </segmentGroup>
                <segmentGroup id="axon_group" neuroLexId="GO:0030424">
                    <notes>Default axon segment group for the cell</notes>
                    <property tag="color" value="0 0.8 0"/>
                    <include segmentGroup="axon_0"/>
                </segmentGroup>
                <segmentGroup id="dendrite_group" neuroLexId="GO:0030425">
                    <notes>Default dendrite segment group for the cell</notes>
                    <property tag="color" value="0.8 0 0"/>
                    <include segmentGroup="dend_0"/>
                    <include segmentGroup="dend_1"/>
                </segmentGroup>
                <segmentGroup id="all">
                    <notes>Default segment group for all segments in the cell</notes>
                    <include segmentGroup="axon_0"/>
                    <include segmentGroup="dend_0"/>
                    <include segmentGroup="dend_1"/>
                    <include segmentGroup="soma_0"/>
                </segmentGroup>
            </morphology>
            <biophysicalProperties id="biophys">
                <membraneProperties>
                    <channelDensity id="leak_all" ionChannel="leak_chan" condDensity="0.01 mS_per_cm2" erev="-67mV" ion="non_specific"/>
                    <channelDensity id="HCNolm_soma" ionChannel="HCNolm" condDensity="0.5 mS_per_cm2" erev="-32.9mV" segmentGroup="soma_group" ion="h"/>
                    <channelDensity id="Kdrfast_soma" ionChannel="Kdrfast" condDensity="73.37 mS_per_cm2" erev="-77mV" segmentGroup="soma_group" ion="k"/>
                    <channelDensity id="Kdrfast_dendrite" ionChannel="Kdrfast" condDensity="105.8 mS_per_cm2" erev="-77mV" segmentGroup="dendrite_group" ion="k"/>
                    <channelDensity id="Kdrfast_axon" ionChannel="Kdrfast" condDensity="117.392 mS_per_cm2" erev="-77mV" segmentGroup="axon_group" ion="k"/>
                    <channelDensity id="KvAolm_soma" ionChannel="KvAolm" condDensity="4.95 mS_per_cm2" erev="-77mV" segmentGroup="soma_group" ion="k"/>
                    <channelDensity id="KvAolm_dendrite" ionChannel="KvAolm" condDensity="2.8 mS_per_cm2" erev="-77mV" segmentGroup="dendrite_group" ion="k"/>
                    <channelDensity id="Nav_soma" ionChannel="Nav" condDensity="10.7 mS_per_cm2" erev="50mV" segmentGroup="soma_group" ion="na"/>
                    <channelDensity id="Nav_dendrite" ionChannel="Nav" condDensity="23.4 mS_per_cm2" erev="50mV" segmentGroup="dendrite_group" ion="na"/>
                    <channelDensity id="Nav_axon" ionChannel="Nav" condDensity="17.12 mS_per_cm2" erev="50mV" segmentGroup="axon_group" ion="na"/>
                    <spikeThresh value="-20 mV"/>
                    <specificCapacitance value="1.3 uF_per_cm2"/>
                    <initMembPotential value="-67mV"/>
                </membraneProperties>
                <intracellularProperties>
                    <resistivity value="0.15 kohm_cm"/>
                </intracellularProperties>
            </biophysicalProperties>
        </cell>
    </neuroml>
:::
:::

We can now already inspect our cell using the NeuroML tools. We have
already generated the morphology plot in our script, but we can also do
it using [`pynml`{.docutils .literal .notranslate}]{.pre}:

::: {.highlight-console .notranslate}
::: highlight
    pynml -png olm.cell.png
    ...
    pyNeuroML >>> Writing to: /home/asinha/Documents/02_Code/00_mine/2020-OSB/NeuroML-Documentation/source/Userdocs/NML2_examples/olm.cell.png
:::
:::

This gives us a figure of the morphology of our cell, similar to the one
we've already generated:

<figure id="id5" class="align-center">
<img src="_images/olm.cell.png"
alt="Figure showing the morphology of the OLM cell generated from the NeuroML definition." />
<figcaption><p><span class="caption-number">Fig. 24 </span><span
class="caption-text">Figure showing the morphology of the OLM cell
generated from the NeuroML definition.</span><a href="#id5"
class="headerlink" title="Link to this image">#</a></p></figcaption>
</figure>
:::

::: {#declaring-the-network .section}
##### Declaring the network[\#](#declaring-the-network "Link to this heading"){.headerlink}

We now use our cell in a network. Similar to our previous example, we
are going to only create a network with one cell, and an [[explicit
input]{.std .std-ref}](Landing.html#schema-explicitinput){.reference
.internal} to the cell:

::: {.highlight-python .notranslate}
::: highlight
    def create_olm_network():
        """Create the network

        :returns: name of network nml file
        """
        net_doc = NeuroMLDocument(id="network",
                                  notes="OLM cell network")
        net_doc_fn = "olm_example_net.nml"
        net_doc.add("IncludeType", href=create_olm_cell())
        net = net_doc.add("Network", id="single_olm_cell_network", validate=False)
        # Create a population: convenient to create many cells of the same type
        pop = net.add("Population", id="pop0", notes="A population for our cell",
                      component="olm", size=1, type="populationList",
                      validate=False)
        pop.add("Instance", id=0, location=component_factory("Location", x=0., y=0., z=0.))
        # Input
        net_doc.add("PulseGenerator", id="pg_olm", notes="Simple pulse generator", delay="100ms", duration="100ms", amplitude="0.08nA")

        net.add("ExplicitInput", target="pop0[0]", input="pg_olm")

        pynml.write_neuroml2_file(nml2_doc=net_doc, nml2_file_name=net_doc_fn, validate=True)
        return net_doc_fn
:::
:::

We start in the same way, by creating a new NeuroML document and
including our cell file into it. We then create a [[population]{.std
.std-ref}](Landing.html#schema-population){.reference .internal}
comprising of a single cell. We create a [[pulse generator]{.std
.std-ref}](Landing.html#schema-pulsegenerator){.reference .internal} as
an [[explicit input]{.std
.std-ref}](Landing.html#schema-explicitinput){.reference .internal},
which targets our population. Note that as the schema documentation for
[`ExplicitInput`{.docutils .literal .notranslate}]{.pre} notes, any
current source (any component that *extends* [[basePointCurrent]{.std
.std-ref}](Landing.html#schema-basepointcurrent){.reference .internal})
can be used as an [`ExplicitInput`{.docutils .literal
.notranslate}]{.pre}.

We add all of these to the [[network]{.std
.std-ref}](Landing.html#schema-network){.reference .internal} and save
(and validate) our network file. The NeuroML file generated is below:

::: {.highlight-xml .notranslate}
::: highlight
    <neuroml xmlns="http://www.neuroml.org/schema/neuroml2"  xmlns:xs="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.neuroml.org/schema/neuroml2 https://raw.github.com/NeuroML/NeuroML2/development/Schemas/NeuroML2/NeuroML_v2.3.1.xsd" id="network">
        <notes>OLM cell network</notes>
        <include href="olm.cell.nml"/>
        <pulseGenerator id="pg_olm" delay="100ms" duration="100ms" amplitude="0.08nA">
            <notes>Simple pulse generator</notes>
        </pulseGenerator>
        <network id="single_olm_cell_network">
            <population id="pop0" component="olm" size="1" type="populationList">
                <notes>A population for our cell</notes>
                <instance id="0">
                    <location x="0.0" y="0.0" z="0.0"/>
                </instance>
            </population>
            <explicitInput target="pop0[0]" input="pg_olm"/>
        </network>
    </neuroml>
:::
:::
:::

::: {#the-generated-neuroml-model .section}
##### The generated NeuroML model[\#](#the-generated-neuroml-model "Link to this heading"){.headerlink}

Before we look at simulating the model, we can inspect our model to
check for correctness. All our NeuroML files were validated when they
were created already, so we do not need to run this step again. However,
if required, this can be easily done:

::: {.highlight-console .notranslate}
::: highlight
    pynml -validate olm*nml
:::
:::

Next, we can visualise our model using the information noted in the
[[visualising NeuroML models]{.std
.std-ref}](Landing.html#userdocs-visualising-models){.reference
.internal} page (including the [`-v`{.docutils .literal
.notranslate}]{.pre} verbose option for more information on the cell):

::: {.highlight-console .notranslate}
::: highlight
    pynml-summary olm_example_net.nml -v
    *******************************************************
    * NeuroMLDocument: network
    *
    *  ComponentType: ['Bezaire_HCNolm_tau', 'Bezaire_Kdrfast_betaq', 'Bezaire_KvAolm_taub', 'Bezaire_Nav_alphah']
    *  IonChannel: ['HCNolm', 'Kdrfast', 'KvAolm', 'Nav', 'leak_chan']
    *  PulseGenerator: ['pg_olm']
    *
    *  Cell: olm
    *    <Segment|0|Seg0_soma_0>
    *      Parent segment: None (root segment)
    *      (0.0, 0.0, 0.0), diam 10.0um -> (0.0, 10.0, 0.0), diam 10.0um; seg length: 10.0 um
    *      Surface area: 314.1592653589793 um2, volume: 785.3981633974482 um3
    *    <Segment|1|Seg1_soma_0>
    *      Parent segment: 0
    *      None -> (0.0, 20.0, 0.0), diam 10.0um; seg length: 10.0 um
    *      Surface area: 314.1592653589793 um2, volume: 785.3981633974482 um3
    *    <Segment|2|Seg0_axon_0>
    *      Parent segment: 0
    *      (0.0, 0.0, 0.0), diam 1.5um -> (0.0, -75.0, 0.0), diam 1.5um; seg length: 75.0 um
    *      Surface area: 353.4291735288517 um2, volume: 132.53594007331938 um3
    *    <Segment|3|Seg1_axon_0>
    *      Parent segment: 2
    *      None -> (0.0, -150.0, 0.0), diam 1.5um; seg length: 75.0 um
    *      Surface area: 353.4291735288517 um2, volume: 132.53594007331938 um3
    *    <Segment|4|Seg0_dend_0>
    *      Parent segment: 1
    *      (0.0, 20.0, 0.0), diam 3.0um -> (100.0, 120.0, 0.0), diam 3.0um; seg length: 141.4213562373095 um
    *      Surface area: 1332.8648814475098 um2, volume: 999.6486610856323 um3
    *    <Segment|5|Seg1_dend_0>
    *      Parent segment: 4
    *      None -> (177.0, 197.0, 0.0), diam 3.0um; seg length: 108.89444430272832 um
    *      Surface area: 1026.3059587145826 um2, volume: 769.7294690359369 um3
    *    <Segment|6|Seg0_dend_1>
    *      Parent segment: 1
    *      (0.0, 20.0, 0.0), diam 3.0um -> (-100.0, 120.0, 0.0), diam 3.0um; seg length: 141.4213562373095 um
    *      Surface area: 1332.8648814475098 um2, volume: 999.6486610856323 um3
    *    <Segment|7|Seg1_dend_1>
    *      Parent segment: 6
    *      None -> (-177.0, 197.0, 0.0), diam 3.0um; seg length: 108.89444430272832 um
    *      Surface area: 1026.3059587145826 um2, volume: 769.7294690359369 um3
    *    Total length of 8 segments: 670.6316010800756 um; total area: 6053.518558099847 um2
    *
    *    SegmentGroup: all, 8 member(s),    0 included group(s);    contains 8 segments in total
    *    SegmentGroup: soma_group,  2 member(s),    1 included group(s);   contains 2 segments in total
    *    SegmentGroup: axon_group,  2 member(s),    1 included group(s);   contains 2 segments in total
    *    SegmentGroup: dendrite_group,      4 member(s),    2 included group(s);    contains 4 segments in total
    *    SegmentGroup: soma_0,      2 member(s),    0 included group(s);   contains 2 segments in total
    *    SegmentGroup: axon_0,      2 member(s),    0 included group(s);   contains 2 segments in total
    *    SegmentGroup: dend_0,      2 member(s),    0 included group(s);   contains 2 segments in total
    *    SegmentGroup: dend_1,      2 member(s),    0 included group(s);   contains 2 segments in total
    *
    *    Channel density: leak_all on all;  conductance of 0.01 mS_per_cm2 through ion chan leak_chan with ion non_specific, erev: -67mV
    *      Channel is on <Segment|0|Seg0_soma_0>,   total conductance: 0.1 S_per_m2 x 3.1415926535897934e-10 m2 = 3.1415926535897936e-11 S (31.41592653589794 pS)
    *      Channel is on <Segment|1|Seg1_soma_0>,   total conductance: 0.1 S_per_m2 x 3.1415926535897934e-10 m2 = 3.1415926535897936e-11 S (31.41592653589794 pS)
    *      Channel is on <Segment|2|Seg0_axon_0>,   total conductance: 0.1 S_per_m2 x 3.534291735288517e-10 m2 = 3.534291735288518e-11 S (35.34291735288518 pS)
    *      Channel is on <Segment|3|Seg1_axon_0>,   total conductance: 0.1 S_per_m2 x 3.534291735288517e-10 m2 = 3.534291735288518e-11 S (35.34291735288518 pS)
    *      Channel is on <Segment|4|Seg0_dend_0>,   total conductance: 0.1 S_per_m2 x 1.3328648814475097e-09 m2 = 1.3328648814475097e-10 S (133.28648814475096 pS)
    *      Channel is on <Segment|5|Seg1_dend_0>,   total conductance: 0.1 S_per_m2 x 1.0263059587145826e-09 m2 = 1.0263059587145826e-10 S (102.63059587145825 pS)
    *      Channel is on <Segment|6|Seg0_dend_1>,   total conductance: 0.1 S_per_m2 x 1.3328648814475097e-09 m2 = 1.3328648814475097e-10 S (133.28648814475096 pS)
    *      Channel is on <Segment|7|Seg1_dend_1>,   total conductance: 0.1 S_per_m2 x 1.0263059587145826e-09 m2 = 1.0263059587145826e-10 S (102.63059587145825 pS)
    *    Channel density: HCNolm_soma on soma_group;        conductance of 0.5 mS_per_cm2 through ion chan HCNolm with ion h, erev: -32.9mV
    *      Channel is on <Segment|0|Seg0_soma_0>,   total conductance: 5.0 S_per_m2 x 3.1415926535897934e-10 m2 = 1.5707963267948968e-09 S (1570.796326794897 pS)
    *      Channel is on <Segment|1|Seg1_soma_0>,   total conductance: 5.0 S_per_m2 x 3.1415926535897934e-10 m2 = 1.5707963267948968e-09 S (1570.796326794897 pS)
    *    Channel density: Kdrfast_soma on soma_group;       conductance of 73.37 mS_per_cm2 through ion chan Kdrfast with ion k, erev: -77mV
    *      Channel is on <Segment|0|Seg0_soma_0>,   total conductance: 733.7 S_per_m2 x 3.1415926535897934e-10 m2 = 2.3049865299388314e-07 S (230498.65299388315 pS)
    *      Channel is on <Segment|1|Seg1_soma_0>,   total conductance: 733.7 S_per_m2 x 3.1415926535897934e-10 m2 = 2.3049865299388314e-07 S (230498.65299388315 pS)
    *    Channel density: Kdrfast_dendrite on dendrite_group;       conductance of 105.8 mS_per_cm2 through ion chan Kdrfast with ion k, erev: -77mV
    *      Channel is on <Segment|4|Seg0_dend_0>,   total conductance: 1058.0 S_per_m2 x 1.3328648814475097e-09 m2 = 1.4101710445714652e-06 S (1410171.0445714653 pS)
    *      Channel is on <Segment|5|Seg1_dend_0>,   total conductance: 1058.0 S_per_m2 x 1.0263059587145826e-09 m2 = 1.0858317043200284e-06 S (1085831.7043200284 pS)
    *      Channel is on <Segment|6|Seg0_dend_1>,   total conductance: 1058.0 S_per_m2 x 1.3328648814475097e-09 m2 = 1.4101710445714652e-06 S (1410171.0445714653 pS)
    *      Channel is on <Segment|7|Seg1_dend_1>,   total conductance: 1058.0 S_per_m2 x 1.0263059587145826e-09 m2 = 1.0858317043200284e-06 S (1085831.7043200284 pS)
    *    Channel density: Kdrfast_axon on axon_group;       conductance of 117.392 mS_per_cm2 through ion chan Kdrfast with ion k, erev: -77mV
    *      Channel is on <Segment|2|Seg0_axon_0>,   total conductance: 1173.92 S_per_m2 x 3.534291735288517e-10 m2 = 4.1489757538898964e-07 S (414897.57538898964 pS)
    *      Channel is on <Segment|3|Seg1_axon_0>,   total conductance: 1173.92 S_per_m2 x 3.534291735288517e-10 m2 = 4.1489757538898964e-07 S (414897.57538898964 pS)
    *    Channel density: KvAolm_soma on soma_group;        conductance of 4.95 mS_per_cm2 through ion chan KvAolm with ion k, erev: -77mV
    *      Channel is on <Segment|0|Seg0_soma_0>,   total conductance: 49.5 S_per_m2 x 3.1415926535897934e-10 m2 = 1.5550883635269477e-08 S (15550.883635269476 pS)
    *      Channel is on <Segment|1|Seg1_soma_0>,   total conductance: 49.5 S_per_m2 x 3.1415926535897934e-10 m2 = 1.5550883635269477e-08 S (15550.883635269476 pS)
    *    Channel density: KvAolm_dendrite on dendrite_group;        conductance of 2.8 mS_per_cm2 through ion chan KvAolm with ion k, erev: -77mV
    *      Channel is on <Segment|4|Seg0_dend_0>,   total conductance: 28.0 S_per_m2 x 1.3328648814475097e-09 m2 = 3.7320216680530273e-08 S (37320.21668053028 pS)
    *      Channel is on <Segment|5|Seg1_dend_0>,   total conductance: 28.0 S_per_m2 x 1.0263059587145826e-09 m2 = 2.8736566844008313e-08 S (28736.566844008314 pS)
    *      Channel is on <Segment|6|Seg0_dend_1>,   total conductance: 28.0 S_per_m2 x 1.3328648814475097e-09 m2 = 3.7320216680530273e-08 S (37320.21668053028 pS)
    *      Channel is on <Segment|7|Seg1_dend_1>,   total conductance: 28.0 S_per_m2 x 1.0263059587145826e-09 m2 = 2.8736566844008313e-08 S (28736.566844008314 pS)
    *    Channel density: Nav_soma on soma_group;   conductance of 10.7 mS_per_cm2 through ion chan Nav with ion na, erev: 50mV
    *      Channel is on <Segment|0|Seg0_soma_0>,   total conductance: 107.0 S_per_m2 x 3.1415926535897934e-10 m2 = 3.361504139341079e-08 S (33615.04139341079 pS)
    *      Channel is on <Segment|1|Seg1_soma_0>,   total conductance: 107.0 S_per_m2 x 3.1415926535897934e-10 m2 = 3.361504139341079e-08 S (33615.04139341079 pS)
    *    Channel density: Nav_dendrite on dendrite_group;   conductance of 23.4 mS_per_cm2 through ion chan Nav with ion na, erev: 50mV
    *      Channel is on <Segment|4|Seg0_dend_0>,   total conductance: 234.0 S_per_m2 x 1.3328648814475097e-09 m2 = 3.118903822587173e-07 S (311890.3822587173 pS)
    *      Channel is on <Segment|5|Seg1_dend_0>,   total conductance: 234.0 S_per_m2 x 1.0263059587145826e-09 m2 = 2.401555943392123e-07 S (240155.59433921232 pS)
    *      Channel is on <Segment|6|Seg0_dend_1>,   total conductance: 234.0 S_per_m2 x 1.3328648814475097e-09 m2 = 3.118903822587173e-07 S (311890.3822587173 pS)
    *      Channel is on <Segment|7|Seg1_dend_1>,   total conductance: 234.0 S_per_m2 x 1.0263059587145826e-09 m2 = 2.401555943392123e-07 S (240155.59433921232 pS)
    *    Channel density: Nav_axon on axon_group;   conductance of 17.12 mS_per_cm2 through ion chan Nav with ion na, erev: 50mV
    *      Channel is on <Segment|2|Seg0_axon_0>,   total conductance: 171.20000000000002 S_per_m2 x 3.534291735288517e-10 m2 = 6.050707450813942e-08 S (60507.07450813943 pS)
    *      Channel is on <Segment|3|Seg1_axon_0>,   total conductance: 171.20000000000002 S_per_m2 x 3.534291735288517e-10 m2 = 6.050707450813942e-08 S (60507.07450813943 pS)
    *
    *    Specific capacitance on all: 1.3 uF_per_cm2
    *      Capacitance of <Segment|0|Seg0_soma_0>,  total capacitance: 0.013000000000000001 F_per_m2 x 3.1415926535897934e-10 m2 = 4.084070449666732e-12 F (4.084070449666732 pF)
    *      Capacitance of <Segment|1|Seg1_soma_0>,  total capacitance: 0.013000000000000001 F_per_m2 x 3.1415926535897934e-10 m2 = 4.084070449666732e-12 F (4.084070449666732 pF)
    *      Capacitance of <Segment|2|Seg0_axon_0>,  total capacitance: 0.013000000000000001 F_per_m2 x 3.534291735288517e-10 m2 = 4.594579255875073e-12 F (4.594579255875073 pF)
    *      Capacitance of <Segment|3|Seg1_axon_0>,  total capacitance: 0.013000000000000001 F_per_m2 x 3.534291735288517e-10 m2 = 4.594579255875073e-12 F (4.594579255875073 pF)
    *      Capacitance of <Segment|4|Seg0_dend_0>,  total capacitance: 0.013000000000000001 F_per_m2 x 1.3328648814475097e-09 m2 = 1.732724345881763e-11 F (17.32724345881763 pF)
    *      Capacitance of <Segment|5|Seg1_dend_0>,  total capacitance: 0.013000000000000001 F_per_m2 x 1.0263059587145826e-09 m2 = 1.3341977463289574e-11 F (13.341977463289574 pF)
    *      Capacitance of <Segment|6|Seg0_dend_1>,  total capacitance: 0.013000000000000001 F_per_m2 x 1.3328648814475097e-09 m2 = 1.732724345881763e-11 F (17.32724345881763 pF)
    *      Capacitance of <Segment|7|Seg1_dend_1>,  total capacitance: 0.013000000000000001 F_per_m2 x 1.0263059587145826e-09 m2 = 1.3341977463289574e-11 F (13.341977463289574 pF)
    *
    *  Network: single_olm_cell_network
    *
    *   1 cells in 1 populations
    *     Population: pop0 with 1 components of type olm
    *       Locations: [(0, 0, 0), ...]
    *
    *   0 connections in 0 projections
    *
    *   0 inputs in 0 input lists
    *
    *   1 explicit inputs (outside of input lists)
    *     Explicit Input of type pg_olm to pop0(cell 0), destination: unspecified
    *
    *******************************************************
:::
:::

We can check the connectivity graph of the model:

::: {.highlight-console .notranslate}
::: highlight
    pynml -graph 10 olm_example_net.nml
:::
:::

which will give us this figure:

<figure id="id6" class="align-center">
<a href="_images/single_olm_cell_network.gv.png"
class="reference internal image-reference"><img
src="_images/single_olm_cell_network.gv.png"
style="width: 368.4px; height: 184.2px;"
alt="Level 10 network graph generated by pynml" /></a>
<figcaption><p><span class="caption-number">Fig. 25 </span><span
class="caption-text">Level 10 network graph generated by pynml</span><a
href="#id6" class="headerlink"
title="Link to this image">#</a></p></figcaption>
</figure>
:::
:::

::: {#simulating-the-model .section}
#### Simulating the model[\#](#simulating-the-model "Link to this heading"){.headerlink}

Please note that this example uses the [[NEURON]{.std
.std-ref}](Landing.html#userdocs-neuron){.reference .internal} simulator
to simulate the model. Please ensure that the [`NEURON_HOME`{.docutils
.literal .notranslate}]{.pre} environment variable is correctly set as
noted [[here]{.std
.std-ref}](Landing.html#userdocs-neuron-envvar){.reference .internal}.

Now that we have declared and inspected our network model and all its
components, we can proceed to simulate it. We do this in the
[`main`{.docutils .literal .notranslate}]{.pre} function:

::: {.highlight-python .notranslate}
::: highlight
    def main():
        """Main function

        Include the NeuroML model into a LEMS simulation file, run it, plot some
        data.
        """
        # Simulation bits
        sim_id = "olm_example_sim"
        simulation = LEMSSimulation(sim_id=sim_id, duration=600, dt=0.01, simulation_seed=123)
        # Include the NeuroML model file
        simulation.include_neuroml2_file(create_olm_network())
        # Assign target for the simulation
        simulation.assign_simulation_target("single_olm_cell_network")

        # Recording information from the simulation
        simulation.create_output_file(id="output0", file_name=sim_id + ".dat")
        simulation.add_column_to_output_file("output0", column_id="pop0_0_v", quantity="pop0[0]/v")
        simulation.add_column_to_output_file("output0",
                                             column_id="pop0_0_v_Seg0_soma_0",
                                             quantity="pop0/0/olm/0/v")
        simulation.add_column_to_output_file("output0",
                                             column_id="pop0_0_v_Seg1_soma_0",
                                             quantity="pop0/0/olm/1/v")
        simulation.add_column_to_output_file("output0",
                                             column_id="pop0_0_v_Seg0_axon_0",
                                             quantity="pop0/0/olm/2/v")
        simulation.add_column_to_output_file("output0",
                                             column_id="pop0_0_v_Seg1_axon_0",
                                             quantity="pop0/0/olm/3/v")
        simulation.add_column_to_output_file("output0",
                                             column_id="pop0_0_v_Seg0_dend_0",
                                             quantity="pop0/0/olm/4/v")
        simulation.add_column_to_output_file("output0",
                                             column_id="pop0_0_v_Seg1_dend_0",
                                             quantity="pop0/0/olm/6/v")
        simulation.add_column_to_output_file("output0",
                                             column_id="pop0_0_v_Seg0_dend_1",
                                             quantity="pop0/0/olm/5/v")
        simulation.add_column_to_output_file("output0",
                                             column_id="pop0_0_v_Seg1_dend_1",
                                             quantity="pop0/0/olm/7/v")
        # Save LEMS simulation to file
        sim_file = simulation.save_to_file()

        # Run the simulation using the NEURON simulator
        pynml.run_lems_with_jneuroml_neuron(sim_file, max_memory="2G", nogui=True,
                                            plot=False, skip_run=False)
        # Plot the data
        plot_data(sim_id)
:::
:::

Here we first create a [`LEMSSimulation`{.docutils .literal
.notranslate}]{.pre} instance and include our network NeuroML file in
it. We must inform LEMS what the target of the simulation is. In our
case, it's the id of our network, [`single_olm_cell_network`{.docutils
.literal .notranslate}]{.pre}:

::: {.highlight-python .notranslate}
::: highlight
        simulation = LEMSSimulation(sim_id=sim_id, duration=600, dt=0.01, simulation_seed=123)
        # Include the NeuroML model file
        simulation.include_neuroml2_file(create_olm_network())
        # Assign target for the simulation
        simulation.assign_simulation_target("single_olm_cell_network")
:::
:::

We also want to record some information, so we create an output file
first with an [`id`{.docutils .literal .notranslate}]{.pre} of
[`output0`{.docutils .literal .notranslate}]{.pre}:

::: {.highlight-python .notranslate}
::: highlight
        simulation.add_column_to_output_file("output0", column_id="pop0_0_v", quantity="pop0[0]/v")
:::
:::

Now, we can record any quantity that is exposed by NeuroML (any
[`exposure`{.docutils .literal .notranslate}]{.pre}). Here, for example,
we add columns for the membrane potentials [`v`{.docutils .literal
.notranslate}]{.pre} of the different segments of the [[cell]{.std
.std-ref}](Landing.html#schema-cell){.reference .internal}.

::: {.highlight-python .notranslate}
::: highlight
        simulation.add_column_to_output_file("output0",
                                             column_id="pop0_0_v_Seg0_soma_0",
                                             quantity="pop0/0/olm/0/v")
        simulation.add_column_to_output_file("output0",
                                             column_id="pop0_0_v_Seg1_soma_0",
                                             quantity="pop0/0/olm/1/v")
        simulation.add_column_to_output_file("output0",
                                             column_id="pop0_0_v_Seg0_axon_0",
                                             quantity="pop0/0/olm/2/v")
        simulation.add_column_to_output_file("output0",
                                             column_id="pop0_0_v_Seg1_axon_0",
                                             quantity="pop0/0/olm/3/v")
        simulation.add_column_to_output_file("output0",
                                             column_id="pop0_0_v_Seg0_dend_0",
                                             quantity="pop0/0/olm/4/v")
        simulation.add_column_to_output_file("output0",
                                             column_id="pop0_0_v_Seg1_dend_0",
                                             quantity="pop0/0/olm/6/v")
        simulation.add_column_to_output_file("output0",
                                             column_id="pop0_0_v_Seg0_dend_1",
                                             quantity="pop0/0/olm/5/v")
        simulation.add_column_to_output_file("output0",
                                             column_id="pop0_0_v_Seg1_dend_1",
                                             quantity="pop0/0/olm/7/v")
:::
:::

The path required to point to the [`quantity`{.docutils .literal
.notranslate}]{.pre} (exposure) to be recorded needs to be correctly
provided. Here, where we use a [[population list]{.std
.std-ref}](Landing.html#schema-populationlist){.reference .internal}
that includes an [[instance]{.std
.std-ref}](Landing.html#schema-instance){.reference .internal} of the
cell, it is: [`population_id/instance_id/cell`{.docutils .literal
.notranslate}]{.pre}` `{.docutils .literal
.notranslate}[`component`{.docutils .literal
.notranslate}]{.pre}` `{.docutils .literal
.notranslate}[`type/segment`{.docutils .literal
.notranslate}]{.pre}` `{.docutils .literal
.notranslate}[`id/exposure`{.docutils .literal .notranslate}]{.pre}.
(See tickets
[15](https://github.com/NeuroML/Documentation/issues/15){.reference
.external} and
[16](https://github.com/NeuroML/Documentation/issues/16){.reference
.external})

We then save the LEMS simulation file, and run our simulation with the
[[NEURON]{.std
.std-ref}](Landing.html#userdocs-simulating-models-neuron){.reference
.internal} simulator (since the default [[jNeuroML]{.std
.std-ref}](Landing.html#jneuroml){.reference .internal} simulator can
only simulate single compartment cells).

::: {.highlight-python .notranslate}
::: highlight
        sim_file = simulation.save_to_file()

        # Run the simulation using the NEURON simulator
        pynml.run_lems_with_jneuroml_neuron(sim_file, max_memory="2G", nogui=True,
                                            plot=False, skip_run=False)
:::
:::
:::

::: {#plotting-the-recorded-variables .section}
#### Plotting the recorded variables[\#](#plotting-the-recorded-variables "Link to this heading"){.headerlink}

To plot the variables that we recorded, we write a simple function that
reads the data and uses the [`generate_plot`{.docutils .literal
.notranslate}]{.pre} utility function which generates the membrane
potential graphs for different segments.

::: {.highlight-python .notranslate}
::: highlight
    def plot_data(sim_id):
        """Plot the sim data.

        Load the data from the file and plot the graph for the membrane potential
        using the pynml generate_plot utility function.

        :sim_id: ID of simulaton

        """
        data_array = np.loadtxt(sim_id + ".dat")
        pynml.generate_plot([data_array[:, 0]], [data_array[:, 1]], "Membrane potential (soma seg 0)", show_plot_already=False, save_figure_to=sim_id + "_seg0_soma0-v.png", xaxis="time (s)", yaxis="membrane potential (V)")
        pynml.generate_plot([data_array[:, 0]], [data_array[:, 2]], "Membrane potential (soma seg 1)", show_plot_already=False, save_figure_to=sim_id + "_seg1_soma0-v.png", xaxis="time (s)", yaxis="membrane potential (V)")
        pynml.generate_plot([data_array[:, 0]], [data_array[:, 3]], "Membrane potential (axon seg 0)", show_plot_already=False, save_figure_to=sim_id + "_seg0_axon0-v.png", xaxis="time (s)", yaxis="membrane potential (V)")
        pynml.generate_plot([data_array[:, 0]], [data_array[:, 4]], "Membrane potential (axon seg 1)", show_plot_already=False, save_figure_to=sim_id + "_seg1_axon0-v.png", xaxis="time (s)", yaxis="membrane potential (V)")
:::
:::

This concludes this example. Here we have seen how to create, simulate,
record, and visualise a multi-compartment neuron. In the next section,
you will find an interactive notebook where you can play with this
example.
:::
:::

[]{#document-Userdocs/NML2_examples/OLM}

::: {#interactive-multi-compartment-olm-cell-example .section .tex2jax_ignore .mathjax_ignore}
### Interactive multi-compartment OLM cell example[\#](#interactive-multi-compartment-olm-cell-example "Link to this heading"){.headerlink}

To run this interactive Jupyter Notebook, please click on the rocket
icon 🚀 in the top panel. For more information, please see [[how to use
this documentation]{.std
.std-ref}](Landing.html#userdocs-usage-jupyterbooks){.reference
.internal}. Please uncomment the line below if you use the Google Colab.
(It does not include these packages by default).

::: {.cell .docutils .container}
::: {.cell_input .docutils .container}
::: {.highlight-ipython3 .notranslate}
::: highlight
    #%pip install pyneuroml neuromllite NEURON
:::
:::
:::
:::

::: {.cell .docutils .container}
::: {.cell_input .docutils .container}
::: {.highlight-ipython3 .notranslate}
::: highlight
    #!/usr/bin/env python3
    """
    Multi-compartmental OLM cell example

    File: olm-example.py

    Copyright 2023 NeuroML contributors
    Authors: Padraig Gleeson, Ankur Sinha
    """

    import neuroml
    from neuroml import NeuroMLDocument
    from neuroml.utils import component_factory
    from pyneuroml import pynml
    from pyneuroml.lems import LEMSSimulation
    from pyneuroml.plot.PlotMorphology import plot_2D
    import numpy as np
:::
:::
:::

::: {.cell_output .docutils .container}
::: {.output .traceback .highlight-ipythontb .notranslate}
::: highlight
    ---------------------------------------------------------------------------
    ModuleNotFoundError                       Traceback (most recent call last)
    Cell In[2], line 11
          1 #!/usr/bin/env python3
          2 """
          3 Multi-compartmental OLM cell example
          4 
       (...)      8 Authors: Padraig Gleeson, Ankur Sinha
          9 """
    ---> 11 import neuroml
         12 from neuroml import NeuroMLDocument
         13 from neuroml.utils import component_factory

    ModuleNotFoundError: No module named 'neuroml'
:::
:::
:::
:::

The CellBuilder module file can be found in the same folder as the
Python script. It is used to define the helper functions that we use in
our main file.

::: {#declaring-the-neuroml-model .section}
#### Declaring the NeuroML model[\#](#declaring-the-neuroml-model "Link to this heading"){.headerlink}

::: {#create-the-cell .section}
##### Create the cell[\#](#create-the-cell "Link to this heading"){.headerlink}

In this example, we do not create the ion channels. We include ion
channels that are already provided in NeuroML files.

::: {.cell .docutils .container}
::: {.cell_input .docutils .container}
::: {.highlight-ipython3 .notranslate}
::: highlight
    def create_olm_cell():
        """Create the complete cell.

        :returns: cell object
        """
        nml_cell_doc = component_factory("NeuroMLDocument", id="oml_cell")
        cell = nml_cell_doc.add("Cell", id="olm", neuro_lex_id="NLXCELL:091206")  # type neuroml.Cell
        nml_cell_file = cell.id + ".cell.nml"

        cell.summary()
        cell.info(show_contents=True)
        cell.morphology.info(show_contents=True)

        # Add two soma segments to an unbranched segment group
        cell.add_unbranched_segment_group("soma_0")
        diam = 10.0
        soma_0 = cell.add_segment(
            prox=[0.0, 0.0, 0.0, diam],
            dist=[0.0, 10., 0.0, diam],
            name="Seg0_soma_0",
            group_id="soma_0",
            seg_type="soma"
        )

        soma_1 = cell.add_segment(
            prox=None,
            dist=[0.0, 10. + 10., 0.0, diam],
            name="Seg1_soma_0",
            parent=soma_0,
            group_id="soma_0",
            seg_type="soma"
        )

        # Add axon segments
        diam = 1.5
        cell.add_unbranched_segments(
            [
                [0.0, 0.0, 0.0, diam],
                [0.0, -75, 0.0, diam],
                [0.0, -150, 0.0, diam],
            ],
            parent=soma_0,
            fraction_along=0.0,
            group_id="axon_0",
            seg_type="axon"
        )

        # Add 2 dendrite segments, using the branching utility function

        diam = 3.0
        cell.add_unbranched_segments(
            [
                [0.0, 20.0, 0.0, diam],
                [100, 120, 0.0, diam],
                [177, 197, 0.0, diam],
            ],
            parent=soma_1,
            fraction_along=1.0,
            group_id="dend_0",
            seg_type="dendrite"
        )

        cell.add_unbranched_segments(
            [
                [0.0, 20.0, 0.0, diam],
                [-100, 120, 0.0, diam],
                [-177, 197, 0.0, diam],
            ],
            parent=soma_1,
            fraction_along=1.0,
            group_id="dend_1",
            seg_type="dendrite"
        )

        # color groups for morphology plots
        den_seg_group = cell.get_segment_group("dendrite_group")
        den_seg_group.add("Property", tag="color", value="0.8 0 0")

        ax_seg_group = cell.get_segment_group("axon_group")
        ax_seg_group.add("Property", tag="color", value="0 0.8 0")

        soma_seg_group = cell.get_segment_group("soma_group")
        soma_seg_group.add("Property", tag="color", value="0 0 0.8")

        # Other cell properties
        cell.set_init_memb_potential("-67mV")
        cell.set_resistivity("0.15 kohm_cm")
        cell.set_specific_capacitance("1.3 uF_per_cm2")

        # channels
        # leak
        cell.add_channel_density(nml_cell_doc,
                                 cd_id="leak_all",
                                 cond_density="0.01 mS_per_cm2",
                                 ion_channel="leak_chan",
                                 ion_chan_def_file="olm-example/leak_chan.channel.nml",
                                 erev="-67mV",
                                 ion="non_specific")
        # HCNolm_soma
        cell.add_channel_density(nml_cell_doc,
                                 cd_id="HCNolm_soma",
                                 cond_density="0.5 mS_per_cm2",
                                 ion_channel="HCNolm",
                                 ion_chan_def_file="olm-example/HCNolm.channel.nml",
                                 erev="-32.9mV",
                                 ion="h",
                                 group_id="soma_group")
        # Kdrfast_soma
        cell.add_channel_density(nml_cell_doc,
                                 cd_id="Kdrfast_soma",
                                 cond_density="73.37 mS_per_cm2",
                                 ion_channel="Kdrfast",
                                 ion_chan_def_file="olm-example/Kdrfast.channel.nml",
                                 erev="-77mV",
                                 ion="k",
                                 group_id="soma_group")
        # Kdrfast_dendrite
        cell.add_channel_density(nml_cell_doc,
                                 cd_id="Kdrfast_dendrite",
                                 cond_density="105.8 mS_per_cm2",
                                 ion_channel="Kdrfast",
                                 ion_chan_def_file="olm-example/Kdrfast.channel.nml",
                                 erev="-77mV",
                                 ion="k",
                                 group_id="dendrite_group")
        # Kdrfast_axon
        cell.add_channel_density(nml_cell_doc,
                                 cd_id="Kdrfast_axon",
                                 cond_density="117.392 mS_per_cm2",
                                 ion_channel="Kdrfast",
                                 ion_chan_def_file="olm-example/Kdrfast.channel.nml",
                                 erev="-77mV",
                                 ion="k",
                                 group_id="axon_group")
        # KvAolm_soma
        cell.add_channel_density(nml_cell_doc,
                                 cd_id="KvAolm_soma",
                                 cond_density="4.95 mS_per_cm2",
                                 ion_channel="KvAolm",
                                 ion_chan_def_file="olm-example/KvAolm.channel.nml",
                                 erev="-77mV",
                                 ion="k",
                                 group_id="soma_group")
        # KvAolm_dendrite
        cell.add_channel_density(nml_cell_doc,
                                 cd_id="KvAolm_dendrite",
                                 cond_density="2.8 mS_per_cm2",
                                 ion_channel="KvAolm",
                                 ion_chan_def_file="olm-example/KvAolm.channel.nml",
                                 erev="-77mV",
                                 ion="k",
                                 group_id="dendrite_group")
        # Nav_soma
        cell.add_channel_density(nml_cell_doc,
                                 cd_id="Nav_soma",
                                 cond_density="10.7 mS_per_cm2",
                                 ion_channel="Nav",
                                 ion_chan_def_file="olm-example/Nav.channel.nml",
                                 erev="50mV",
                                 ion="na",
                                 group_id="soma_group")
        # Nav_dendrite
        cell.add_channel_density(nml_cell_doc,
                                 cd_id="Nav_dendrite",
                                 cond_density="23.4 mS_per_cm2",
                                 ion_channel="Nav",
                                 ion_chan_def_file="olm-example/Nav.channel.nml",
                                 erev="50mV",
                                 ion="na",
                                 group_id="dendrite_group")
        # Nav_axon
        cell.add_channel_density(nml_cell_doc,
                                 cd_id="Nav_axon",
                                 cond_density="17.12 mS_per_cm2",
                                 ion_channel="Nav",
                                 ion_chan_def_file="olm-example/Nav.channel.nml",
                                 erev="50mV",
                                 ion="na",
                                 group_id="axon_group")

        cell.optimise_segment_groups()
        cell.validate(recursive=True)
        pynml.write_neuroml2_file(nml_cell_doc, nml_cell_file, True, True)
        plot_2D(nml_cell_file, plane2d="xy", nogui=True,
                save_to_file="olm.cell.xy.png")
        return nml_cell_file
:::
:::
:::
:::
:::

::: {#create-the-network .section}
##### Create the network[\#](#create-the-network "Link to this heading"){.headerlink}

::: {.cell .docutils .container}
::: {.cell_input .docutils .container}
::: {.highlight-ipython3 .notranslate}
::: highlight
    def create_olm_network():
        """Create the network

        :returns: name of network nml file
        """
        net_doc = NeuroMLDocument(id="network",
                                  notes="OLM cell network")
        net_doc_fn = "olm_example_net.nml"
        net_doc.add("IncludeType", href=create_olm_cell())
        net = net_doc.add("Network", id="single_olm_cell_network", validate=False)
        # Create a population: convenient to create many cells of the same type
        pop = net.add("Population", id="pop0", notes="A population for our cell",
                      component="olm", size=1, type="populationList",
                      validate=False)
        pop.add("Instance", id=0, location=component_factory("Location", x=0., y=0., z=0.))
        # Input
        net_doc.add("PulseGenerator", id="pg_olm", notes="Simple pulse generator", delay="100ms", duration="100ms", amplitude="0.08nA")

        net.add("ExplicitInput", target="pop0[0]", input="pg_olm")

        pynml.write_neuroml2_file(nml2_doc=net_doc, nml2_file_name=net_doc_fn, validate=True)
        return net_doc_fn
:::
:::
:::
:::
:::
:::

::: {#plot-the-data-we-record .section}
#### Plot the data we record[\#](#plot-the-data-we-record "Link to this heading"){.headerlink}

::: {.cell .docutils .container}
::: {.cell_input .docutils .container}
::: {.highlight-ipython3 .notranslate}
::: highlight
    def plot_data(sim_id):
        """Plot the sim data.

        Load the data from the file and plot the graph for the membrane potential
        using the pynml generate_plot utility function.

        :sim_id: ID of simulaton

        """
        data_array = np.loadtxt(sim_id + ".dat")
        pynml.generate_plot([data_array[:, 0]], [data_array[:, 1]], "Membrane potential (soma seg 0)", show_plot_already=False, save_figure_to=sim_id + "_seg0_soma0-v.png", xaxis="time (s)", yaxis="membrane potential (V)")
        pynml.generate_plot([data_array[:, 0]], [data_array[:, 2]], "Membrane potential (soma seg 1)", show_plot_already=False, save_figure_to=sim_id + "_seg1_soma0-v.png", xaxis="time (s)", yaxis="membrane potential (V)")
        pynml.generate_plot([data_array[:, 0]], [data_array[:, 3]], "Membrane potential (axon seg 0)", show_plot_already=False, save_figure_to=sim_id + "_seg0_axon0-v.png", xaxis="time (s)", yaxis="membrane potential (V)")
        pynml.generate_plot([data_array[:, 0]], [data_array[:, 4]], "Membrane potential (axon seg 1)", show_plot_already=False, save_figure_to=sim_id + "_seg1_axon0-v.png", xaxis="time (s)", yaxis="membrane potential (V)")
:::
:::
:::
:::
:::

::: {#create-and-run-the-simulaton .section}
#### Create and run the simulaton[\#](#create-and-run-the-simulaton "Link to this heading"){.headerlink}

::: {.cell .docutils .container}
::: {.cell_input .docutils .container}
::: {.highlight-ipython3 .notranslate}
::: highlight
    def main():
        """Main function

        Include the NeuroML model into a LEMS simulation file, run it, plot some
        data.
        """
        # Simulation bits
        sim_id = "olm_example_sim"
        simulation = LEMSSimulation(sim_id=sim_id, duration=600, dt=0.01, simulation_seed=123)
        # Include the NeuroML model file
        simulation.include_neuroml2_file(create_olm_network())
        # Assign target for the simulation
        simulation.assign_simulation_target("single_olm_cell_network")

        # Recording information from the simulation
        simulation.create_output_file(id="output0", file_name=sim_id + ".dat")
        simulation.add_column_to_output_file("output0", column_id="pop0_0_v", quantity="pop0[0]/v")
        simulation.add_column_to_output_file("output0",
                                             column_id="pop0_0_v_Seg0_soma_0",
                                             quantity="pop0/0/olm/0/v")
        simulation.add_column_to_output_file("output0",
                                             column_id="pop0_0_v_Seg1_soma_0",
                                             quantity="pop0/0/olm/1/v")
        simulation.add_column_to_output_file("output0",
                                             column_id="pop0_0_v_Seg0_axon_0",
                                             quantity="pop0/0/olm/2/v")
        simulation.add_column_to_output_file("output0",
                                             column_id="pop0_0_v_Seg1_axon_0",
                                             quantity="pop0/0/olm/3/v")
        simulation.add_column_to_output_file("output0",
                                             column_id="pop0_0_v_Seg0_dend_0",
                                             quantity="pop0/0/olm/4/v")
        simulation.add_column_to_output_file("output0",
                                             column_id="pop0_0_v_Seg1_dend_0",
                                             quantity="pop0/0/olm/6/v")
        simulation.add_column_to_output_file("output0",
                                             column_id="pop0_0_v_Seg0_dend_1",
                                             quantity="pop0/0/olm/5/v")
        simulation.add_column_to_output_file("output0",
                                             column_id="pop0_0_v_Seg1_dend_1",
                                             quantity="pop0/0/olm/7/v")
        # Save LEMS simulation to file
        sim_file = simulation.save_to_file()

        # Run the simulation using the NEURON simulator
        pynml.run_lems_with_jneuroml_neuron(sim_file, max_memory="2G", nogui=True,
                                            plot=False, skip_run=False)
        # Plot the data
        plot_data(sim_id)
:::
:::
:::
:::

::: {.cell .docutils .container}
::: {.cell_input .docutils .container}
::: {.highlight-ipython3 .notranslate}
::: highlight
    if __name__ == "__main__":
        main()
:::
:::
:::

::: {.cell_output .docutils .container}
::: {.output .stderr .highlight-myst-ansi .notranslate}
::: highlight
    pyNeuroML >>> INFO - Executing: (java -Xmx400M  -jar  "/home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/.venv/lib/python3.10/site-packages/pyneuroml/lib/jNeuroML-0.12.0-jar-with-dependencies.jar" -validate "olm.cell.nml" ) in directory: .
:::
:::

::: {.output .stream .highlight-myst-ansi .notranslate}
::: highlight
    *******************************************************
    * Cell: olm
    * Notes: None
    * Segments: 0
    * SegmentGroups: 4
    *******************************************************
    Cell -- Cell with  **segment** s specified in a  **morphology**  element along with details on its  **biophysicalProperties** . NOTE: this can only be correctly simulated using jLEMS when there is a single segment in the cell, and **v**  of this cell represents the membrane potential in that isopotential segment.

    Please see the NeuroML standard schema documentation at https://docs.neuroml.org/Userdocs/NeuroMLv2.html for more information.

    Valid members for Cell are:
    * morphology_attr (class: NmlId, Optional)
    * biophysical_properties_attr (class: NmlId, Optional)
    * morphology (class: Morphology, Optional)
        * Contents ('ids'/<objects>): 'morphology'

    * neuro_lex_id (class: NeuroLexId, Optional)
        * Contents ('ids'/<objects>): NLXCELL:091206

    * biophysical_properties (class: BiophysicalProperties, Optional)
        * Contents ('ids'/<objects>): 'biophys'

    * annotation (class: Annotation, Optional)
    * properties (class: Property, Optional)
    * notes (class: xs:string, Optional)
    * id (class: NmlId, Required)
        * Contents ('ids'/<objects>): olm

    * metaid (class: MetaId, Optional)

    Morphology -- The collection of  **segment** s which specify the 3D structure of the cell, along with a number of  **segmentGroup** s

    Please see the NeuroML standard schema documentation at https://docs.neuroml.org/Userdocs/NeuroMLv2.html for more information.

    Valid members for Morphology are:
    * segments (class: Segment, Required)
    * segment_groups (class: SegmentGroup, Optional)
        * Contents ('ids'/<objects>): ['all', 'soma_group', 'axon_group', 'dendrite_group']

    * annotation (class: Annotation, Optional)
    * properties (class: Property, Optional)
    * notes (class: xs:string, Optional)
    * id (class: NmlId, Required)
        * Contents ('ids'/<objects>): morphology

    * metaid (class: MetaId, Optional)
:::
:::

::: {.output .stderr .highlight-myst-ansi .notranslate}
::: highlight
    pyNeuroML >>> INFO - Command completed. Output: 
      jNeuroML >>   jNeuroML v0.12.0
      jNeuroML >>  Validating: /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/olm.cell.nml
      jNeuroML >>  Valid against schema and all tests
      jNeuroML >>  No warnings
      jNeuroML >>  
      jNeuroML >>  Validated 1 files: All valid and no warnings
      jNeuroML >>  
      jNeuroML >>  
    pyNeuroML >>> INFO - Loading NeuroML2 file: olm.cell.nml
    pyNeuroML >>> INFO - Including included files (included already: [])
    pyNeuroML >>> INFO - Loading NeuroML2 file: /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/olm-example/leak_chan.channel.nml
    pyNeuroML >>> INFO - Including included files (included already: [])
    pyNeuroML >>> INFO - Loading NeuroML2 file: /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/olm-example/HCNolm.channel.nml
    pyNeuroML >>> INFO - Including included files (included already: ['/home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/olm-example/leak_chan.channel.nml'])
    pyNeuroML >>> INFO - Loading NeuroML2 file: /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/olm-example/Kdrfast.channel.nml
    pyNeuroML >>> INFO - Including included files (included already: ['/home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/olm-example/leak_chan.channel.nml', '/home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/olm-example/HCNolm.channel.nml'])
    pyNeuroML >>> INFO - Loading NeuroML2 file: /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/olm-example/KvAolm.channel.nml
    pyNeuroML >>> INFO - Including included files (included already: ['/home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/olm-example/leak_chan.channel.nml', '/home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/olm-example/HCNolm.channel.nml', '/home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/olm-example/Kdrfast.channel.nml'])
    pyNeuroML >>> INFO - Loading NeuroML2 file: /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/olm-example/Nav.channel.nml
    pyNeuroML >>> INFO - Including included files (included already: ['/home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/olm-example/leak_chan.channel.nml', '/home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/olm-example/HCNolm.channel.nml', '/home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/olm-example/Kdrfast.channel.nml', '/home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/olm-example/KvAolm.channel.nml'])
    pyNeuroML >>> INFO - Executing: (java -Xmx400M  -jar  "/home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/.venv/lib/python3.10/site-packages/pyneuroml/lib/jNeuroML-0.12.0-jar-with-dependencies.jar" -validate "olm_example_net.nml" ) in directory: .
:::
:::

::: {.output .stream .highlight-myst-ansi .notranslate}
::: highlight
    Saved image on plane xy to /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/olm.cell.xy.png of plot: 2D plot of olm from olm.cell.nml
:::
:::

::: {.output .stderr .highlight-myst-ansi .notranslate}
::: highlight
    pyNeuroML >>> INFO - Command completed. Output: 
      jNeuroML >>   jNeuroML v0.12.0
      jNeuroML >>  Validating: /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/olm_example_net.nml
      jNeuroML >>  Valid against schema and all tests
      jNeuroML >>  No warnings
      jNeuroML >>  
      jNeuroML >>  Validated 1 files: All valid and no warnings
      jNeuroML >>  
      jNeuroML >>  
    pyNeuroML >>> INFO - Loading NeuroML2 file: /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/olm_example_net.nml
    pyNeuroML >>> INFO - Loading NeuroML2 file: /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/olm.cell.nml
    pyNeuroML >>> INFO - Loading NeuroML2 file: /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/olm-example/leak_chan.channel.nml
    pyNeuroML >>> INFO - Loading NeuroML2 file: /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/olm-example/HCNolm.channel.nml
    pyNeuroML >>> INFO - Loading NeuroML2 file: /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/olm-example/Kdrfast.channel.nml
    pyNeuroML >>> INFO - Loading NeuroML2 file: /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/olm-example/KvAolm.channel.nml
    pyNeuroML >>> INFO - Loading NeuroML2 file: /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/olm-example/Nav.channel.nml
    pyNeuroML >>> INFO - Loading LEMS file: LEMS_olm_example_sim.xml and running with jNeuroML_NEURON
    pyNeuroML >>> INFO - Executing: (java -Xmx2G  -Djava.awt.headless=true -jar  "/home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/.venv/lib/python3.10/site-packages/pyneuroml/lib/jNeuroML-0.12.0-jar-with-dependencies.jar"  "LEMS_olm_example_sim.xml"  -neuron -run -compile -nogui -I '') in directory: .
    pyNeuroML >>> INFO - Command completed. Output: 
      jNeuroML >>   jNeuroML v0.12.0
      jNeuroML >>  (INFO) Reading from: /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/LEMS_olm_example_sim.xml
      jNeuroML >>  (INFO) Creating NeuronWriter to output files to /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples
      jNeuroML >>  (INFO) Adding simulation Component(id=olm_example_sim type=Simulation) of network/component: single_olm_cell_network (Type: network)
      jNeuroML >>  (INFO) Adding population: pop0
      jNeuroML >>  (INFO) -- Writing to hoc: /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/olm.hoc
      jNeuroML >>  (INFO) -- Writing to mod: /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/leak_chan.mod
      jNeuroML >>  (INFO) File /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/leak_chan.mod exists and is identical
      jNeuroML >>  (INFO) -- Writing to mod: /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/HCNolm.mod
      jNeuroML >>  (INFO) File /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/HCNolm.mod exists and is identical
      jNeuroML >>  (INFO) -- Writing to mod: /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/Kdrfast.mod
      jNeuroML >>  (INFO) File /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/Kdrfast.mod exists and is identical
      jNeuroML >>  (INFO) -- Writing to mod: /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/KvAolm.mod
      jNeuroML >>  (INFO) File /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/KvAolm.mod exists and is identical
      jNeuroML >>  (INFO) -- Writing to mod: /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/Nav.mod
      jNeuroML >>  (INFO) File /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/Nav.mod exists and is identical
      jNeuroML >>  (INFO) Adding projections/connections...
      jNeuroML >>  (INFO) -- Writing to mod: /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/pg_olm.mod
      jNeuroML >>  (INFO) File /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/pg_olm.mod exists and is identical
      jNeuroML >>  (INFO) Trying to compile mods in: /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples
      jNeuroML >>  (INFO) Going to compile the mod files in: /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples, forcing recompile: false
      jNeuroML >>  (INFO) Parent dir: /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples
      jNeuroML >>  (INFO) Assuming *nix environment...
      jNeuroML >>  (INFO) Name of file to be created: /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/x86_64/libnrnmech.la
      jNeuroML >>  (INFO) Name of file to be created: /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/x86_64/libnrnmech.so
      jNeuroML >>  (INFO) Name of file to be created: /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/x86_64/.libs/libnrnmech.so
      jNeuroML >>  (INFO) commandToExecute: /usr/bin/nrnivmodl
      jNeuroML >>  (INFO) Found previously compiled file: /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/x86_64/libnrnmech.so
      jNeuroML >>  (INFO) Going to check if mods in /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples are newer than /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/x86_64/libnrnmech.so
      jNeuroML >>  (INFO) Going to check if mods in /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples are newer than /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/x86_64/.libs/libnrnmech.so
      jNeuroML >>  (INFO) Not being asked to recompile, and no mod files exist in /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples which are newer than /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/x86_64/.libs/libnrnmech.so
      jNeuroML >>  (INFO) Success in compiling mods: true
      jNeuroML >>  (INFO) Have successfully executed command: python /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/LEMS_olm_example_sim_nrn.py
      jNeuroML >>  (INFO) NRN Output >>> 
      jNeuroML >>  (INFO) NRN Output >>>     Starting simulation in NEURON of 600ms generated from NeuroML2 model...
      jNeuroML >>  (INFO) NRN Output >>> 
      jNeuroML >>  (INFO) NRN Output >>> Population pop0 contains 1 instance(s) of component: olm of type: cell
      jNeuroML >>  (INFO) NRN Output >>>     1 
      jNeuroML >>  (INFO) NRN Output >>> Setting up the network to simulate took 0.006119 seconds
      jNeuroML >>  (INFO) NRN Output >>> Running a simulation of 600.0ms (dt = 0.01ms; seed=123)
      jNeuroML >>  (INFO) NRN Output >>> Finished NEURON simulation in 0.424608 seconds (0.007077 mins)...
      jNeuroML >>  (INFO) NRN Output >>> Saving results at t=599.9999999995268...
      jNeuroML >>  (INFO) NRN Output >>> Saved data to: time.dat
      jNeuroML >>  (INFO) NRN Output >>> Saved data to: olm_example_sim.dat
      jNeuroML >>  (INFO) NRN Output >>> Finished saving results in 0.455872 seconds
      jNeuroML >>  (INFO) NRN Output >>> Done
      jNeuroML >>  (INFO) Exit value for running NEURON: 0
      jNeuroML >>  
    pyNeuroML >>> INFO - Generating plot: Membrane potential (soma seg 0)
    /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/.venv/lib64/python3.10/site-packages/pyneuroml/plot/Plot.py:186: UserWarning: marker is redundantly defined by the 'marker' keyword argument and the fmt string "o" (-> marker='o'). The keyword argument will take precedence.
      plt.plot(
    pyNeuroML >>> INFO - Saving image to /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/olm_example_sim_seg0_soma0-v.png of plot: Membrane potential (soma seg 0)
:::
:::

::: {.output .stderr .highlight-myst-ansi .notranslate}
::: highlight
    pyNeuroML >>> INFO - Saved image to olm_example_sim_seg0_soma0-v.png of plot: Membrane potential (soma seg 0)
    pyNeuroML >>> INFO - Generating plot: Membrane potential (soma seg 1)
    pyNeuroML >>> INFO - Saving image to /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/olm_example_sim_seg1_soma0-v.png of plot: Membrane potential (soma seg 1)
    pyNeuroML >>> INFO - Saved image to olm_example_sim_seg1_soma0-v.png of plot: Membrane potential (soma seg 1)
    pyNeuroML >>> INFO - Generating plot: Membrane potential (axon seg 0)
    pyNeuroML >>> INFO - Saving image to /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/olm_example_sim_seg0_axon0-v.png of plot: Membrane potential (axon seg 0)
    pyNeuroML >>> INFO - Saved image to olm_example_sim_seg0_axon0-v.png of plot: Membrane potential (axon seg 0)
    pyNeuroML >>> INFO - Generating plot: Membrane potential (axon seg 1)
    pyNeuroML >>> INFO - Saving image to /home/asinha/Documents/02_Code/00_mine/NeuroML/documentation/source/Userdocs/NML2_examples/olm_example_sim_seg1_axon0-v.png of plot: Membrane potential (axon seg 1)
    pyNeuroML >>> INFO - Saved image to olm_example_sim_seg1_axon0-v.png of plot: Membrane potential (axon seg 1)
:::
:::

![](_images/2e1bea36fe18d51db608aa0f8ab5ea17531213bda9b78245318583898b1049c5.png)
![](_images/ec193126c160d9b3a694a430ec1c3ce01ff44331eb30ff5cecfa6f5b966145d2.png)
![](_images/ec193126c160d9b3a694a430ec1c3ce01ff44331eb30ff5cecfa6f5b966145d2.png)
![](_images/ec193126c160d9b3a694a430ec1c3ce01ff44331eb30ff5cecfa6f5b966145d2.png)
![](_images/21da5d20d226f9ba010af0c85727743b7d0f604a11d927ff7b85e04705273bf6.png)
:::
:::
:::
:::

[]{#document-Userdocs/NML2_examples/NeuroML-DB}

::: {#create-novel-neuroml-models-from-components-on-neuroml-db .section .tex2jax_ignore .mathjax_ignore}
### Create novel NeuroML models from components on NeuroML-DB[\#](#create-novel-neuroml-models-from-components-on-neuroml-db "Link to this heading"){.headerlink}

This notebook demonstrates how to access the [NeuroML-DB
database](https://neuroml-db.org){.reference .external}, extract
elements in NeuroML format and use them to create new models

::: {.cell .docutils .container}
::: {.cell_input .docutils .container}
::: {.highlight-ipython3 .notranslate}
::: highlight
    from pyneuroml import pynml
    import urllib.request, json 
    import requests
    import os
:::
:::
:::

::: {.cell_output .docutils .container}
::: {.output .traceback .highlight-ipythontb .notranslate}
::: highlight
    ---------------------------------------------------------------------------
    ModuleNotFoundError                       Traceback (most recent call last)
    Cell In[1], line 1
    ----> 1 from pyneuroml import pynml
          2 import urllib.request, json 
          3 import requests

    ModuleNotFoundError: No module named 'pyneuroml'
:::
:::
:::
:::

::: {#search-for-download-and-analyse-channels-accessed-via-the-neuroml-db-api .section}
#### 1) Search for, download and analyse channels accessed via the NeuroML-DB API[\#](#search-for-download-and-analyse-channels-accessed-via-the-neuroml-db-api "Link to this heading"){.headerlink}

::: {#search-in-the-database-for-a-particular-type-of-channel .section}
##### 1.1) Search in the database for a particular type of channel[\#](#search-in-the-database-for-a-particular-type-of-channel "Link to this heading"){.headerlink}

::: {.cell .docutils .container}
::: {.cell_input .docutils .container}
::: {.highlight-ipython3 .notranslate}
::: highlight
    types = {'cell':'NMLCL','channel':'NMLCH'}

    # Helper method for search
    def search_neuromldb(search_term, type=None):
        
        with urllib.request.urlopen('https://neuroml-db.org/api/search?q=%s' % search_term.replace(' ','%20')) as url:
            data = json.load(url)

        for l in data:
            if type!=None:
                for type_ in types:
                    if type==type_ and not types[type_] in l['Model_ID']:
                        data.remove(l)
            if l in data:
                print('%s: %s, %s %s %s'%(l['Model_ID'],l['Name'],l['First_Author'],l['Second_Author'],l['Publication_Year']))
            
        return data

    data = search_neuromldb("Fast Sodium", 'channel')
:::
:::
:::

::: {.cell_output .docutils .container}
::: {.output .stream .highlight-myst-ansi .notranslate}
::: highlight
    NMLCH000023: NaF Inactivating Fast Sodium, Maex De Schutter 1998
    NMLCH001490: NaTs Fast Inactivating Sodium, Gouwens Berg 2018
    NMLCH001398: NaTa Fast Inactivating Sodium, Hay Hill 2011
    NMLCH000171: NaF Fast Transient Inactivating Sodium, Traub Contreras 2005
    NMLCH000170: NaF Fast Transient Inactivating Sodium, Traub Contreras 2005
    NMLCH000169: NaF Fast Transient Inactivating Sodium, Traub Contreras 2005
    NMLCH000168: NaF Fast Transient Inactivating Sodium, Traub Contreras 2005
    NMLCH000167: NaF Fast Transient Inactivating Sodium, Traub Contreras 2005
    NMLCH000166: NaF Fast Transient Inactivating Sodium, Traub Contreras 2005
    NMLCH000165: NaF Fast Transient Inactivating Sodium, Traub Contreras 2005
    NMLCH000164: NaP Persistent Noninactivating Sodium, Traub Buhl 2003
    NMLCH000163: NaF Fast Transient Inactivating Sodium, Traub Contreras 2005
    NMLCH000162: NaF Fast Transient Inactivating Sodium, Traub Contreras 2005
    NMLCH000161: NaF Fast Transient Inactivating Sodium, Traub Contreras 2005
    NMLCH000160: NaF Fast Transient Inactivating Sodium, Traub Buhl 2003
    NMLCH000159: NaF Fast Transient Inactivating Sodium, Traub Contreras 2005
    NMLCH000158: NaF Fast Transient Inactivating Sodium, Traub Contreras 2005
    NMLCH000157: NaF Fast Transient Inactivating Sodium, Traub Buhl 2003
    NMLCH000131: NaF Fast Sodium, Pospischil Toledo-Rodriguez 2008
    NMLCH000111: NaTs Fast Inactivating Sodium, Colbert Pan 2002
    NMLCH000110: NaTa Fast Inactivating Sodium, Colbert Pan 2002
    NMLCH000008: NaF Inactivating Fast Sodium, Maex De Schutter 1998
:::
:::
:::
:::
:::

::: {#select-one-of-these-results-and-download-it-and-browse-the-contents .section}
##### 1.2) Select one of these results and download it, and browse the contents[\#](#select-one-of-these-results-and-download-it-and-browse-the-contents "Link to this heading"){.headerlink}

::: {.cell .docutils .container}
::: {.cell_input .docutils .container}
::: {.highlight-ipython3 .notranslate}
::: highlight
    # Helper method to retrieve a NeuroML file based on modelID
    def get_model_from_neuromldb(model_id, type):
      
        fname = '%s.%s.nml'%(model_id, type)

        url = 'https://neuroml-db.org/render_xml_file?modelID=%s'%model_id
        r = requests.get(url)
        with open(fname , 'wb') as f:
            f.write(r.content)

        return pynml.read_neuroml2_file(fname), fname


    # Choose one of the channels
    chan_model_id = 'NMLCH001398'

    na_chan_doc, na_chan_fname_orig = get_model_from_neuromldb(chan_model_id, 'channel')

    na_chan = na_chan_doc.ion_channel[0] # select the first/only ion channel in the nml doc

    na_chan_fname = '%s.channel.nml'%na_chan.id
    os.rename(na_chan_fname_orig, na_chan_fname) # Rename for clarity

    print('Channel %s (in file %s) has notes: %s'%(na_chan.id, na_chan_fname, na_chan.notes))
:::
:::
:::

::: {.cell_output .docutils .container}
::: {.output .stderr .highlight-myst-ansi .notranslate}
::: highlight
    pyNeuroML >>> INFO - Loading NeuroML2 file: NMLCH001398.channel.nml
:::
:::

::: {.output .stream .highlight-myst-ansi .notranslate}
::: highlight
    Channel NaTa_t (in file NaTa_t.channel.nml) has notes: Fast inactivating Na+ current
                
    Comment from original mod file: 
    :Reference :Colbert and Pan 2002
:::
:::
:::
:::
:::

::: {#plot-the-time-course-and-steady-state-of-this-channel .section}
##### 1.3) Plot the time course and steady state of this channel[\#](#plot-the-time-course-and-steady-state-of-this-channel "Link to this heading"){.headerlink}

::: {.cell .docutils .container}
::: {.cell_input .docutils .container}
::: {.highlight-ipython3 .notranslate}
::: highlight
    from pyneuroml.analysis.NML2ChannelAnalysis import run

    na_erev = 50 # mV
    run(channel_files=[na_chan_fname], ivCurve=True, erev=na_erev, clampDelay=5, clampDuration=10, duration=20)
:::
:::
:::

::: {.cell_output .docutils .container}
::: {.output .stderr .highlight-myst-ansi .notranslate}
::: highlight
    pyNeuroML >>> INFO - Loading NeuroML2 file: NaTa_t.channel.nml
    pyNeuroML >>> INFO - Loading NeuroML2 file: NaTa_t.channel.nml
    pyNeuroML >>> INFO - Loading NeuroML2 file: NaTa_t.channel.nml
    pyNeuroML >>> INFO - Loading LEMS file: /Users/padraig/git/Documentation/source/Userdocs/NML2_examples/LEMS_Test_NaTa_t.xml and running with jNeuroML
    pyNeuroML >>> INFO - Executing: (java -Xmx400M  -Djava.awt.headless=true -jar  "/opt/homebrew/anaconda3/envs/py39n/lib/python3.9/site-packages/pyneuroml/lib/jNeuroML-0.13.0-jar-with-dependencies.jar"  /Users/padraig/git/Documentation/source/Userdocs/NML2_examples/LEMS_Test_NaTa_t.xml  -nogui -I '') in directory: .
    pyNeuroML >>> INFO - Command completed successfully!
:::
:::

![](_images/4a56a474313c1712b353a7a3f085449661b7fd7bd3ad4890b429cf4953c34026.png)
![](_images/56d999547a37a2d351ead4521d2cd3a85d2739d8e00846d41c37a75edf9b698c.png)
![](_images/b2aa4981e8cb6b519d72cd0d1e76aa6759e3e3913d827eeee9b12defecf38c63.png)
![](_images/5532522cf7b99fb6f0dead129fa2c8e0c46d4a126add2a89309fec5a8405545b.png)
![](_images/d3d84013cc4343a3e29f989e171b2aa9d3a91931c647b322d965986c06ad3967.png)
:::
:::
:::

::: {#find-a-potassium-channel .section}
##### 1.4) Find a Potassium channel[\#](#find-a-potassium-channel "Link to this heading"){.headerlink}

::: {.cell .docutils .container}
::: {.cell_input .docutils .container}
::: {.highlight-ipython3 .notranslate}
::: highlight
    data = search_neuromldb("Fast Potassium", 'channel')
:::
:::
:::

::: {.cell_output .docutils .container}
::: {.output .stream .highlight-myst-ansi .notranslate}
::: highlight
    NMLCH000113: K Fast Noninactivating Potassium, Rettig Wunder 1992
    NMLCH001627: KDr Fast Delayed Rectifier Potassium , Bezaire Raikov 2016
    NMLCH001611: KDr Fast Delayed Rectifier Potassium, Bezaire Raikov 2016
    NMLCH001609: KDr Fast Delayed Rectifier Potassium , Bezaire Raikov 2016
    NMLCH001549: K Fast Potassium, Boyle Cohen 2008
    NMLCH001529: KT Fast Inactivating Potassium, Gouwens Berg 2018
    NMLCH001465: K Fast Noninactivating Potassium, Gouwens Berg 2018
    NMLCH001400: KTst Fast Inactivating Potassium, Hay Hill 2011
    NMLCH001394: K Fast Noninactivating Potassium, Hay Hill 2011
    NMLCH000156: IM M Type Potassium, Traub Buhl 2003
    NMLCH000155: KDr Delayed Rectifier Potassium, Traub Buhl 2003
    NMLCH000154: KDr Delayed Rectifier Potassium Channel for Fast Spiking (FS) Interneurons, Traub Contreras 2005
    NMLCH000153: KCa BK Type Fast Calcium Dependent Potassium, Traub Buhl 2003
    NMLCH000152: KCa BK Type Fast Calcium Dependent Potassium, Traub Buhl 2003
    NMLCH000151: KCa AHP Type Calcium Dependent Potassium, Traub Buhl 2003
    NMLCH000149: KCa Slow AHP Type Calcium Dependent Potassium, Traub Buhl 2003
    NMLCH000148: IA A Type Potassium, Traub Buhl 2003
    NMLCH000146: K2 Type Slowly Activating and Inactivating Potassium, Traub Buhl 2003
    NMLCH000123: K Fast Potassium, Korngreen Sakmann 2000
    NMLCH000108: KTst Fast Inactivating Potassium, Korngreen Sakmann 2000
:::
:::
:::
:::

::: {.cell .docutils .container}
::: {.cell_input .docutils .container}
::: {.highlight-ipython3 .notranslate}
::: highlight
    # Download one of these

    k_chan_doc, k_chan_fname_orig = get_model_from_neuromldb('NMLCH000113', 'channel')

    k_chan = k_chan_doc.ion_channel[0] # select the first ion channel in the nml doc

    k_chan_fname = '%s.channel.nml'%k_chan.id
    os.rename(k_chan_fname_orig, k_chan_fname) # Rename for clarity

    print('Channel %s (in file %s) has notes: %s'%(k_chan.id, k_chan_fname, k_chan.notes))

    k_erev = -77
    run(channel_files=[k_chan_fname], ivCurve=True, erev=k_erev)
:::
:::
:::

::: {.cell_output .docutils .container}
::: {.output .stderr .highlight-myst-ansi .notranslate}
::: highlight
    pyNeuroML >>> INFO - Loading NeuroML2 file: NMLCH000113.channel.nml
    pyNeuroML >>> INFO - Loading NeuroML2 file: SKv3_1.channel.nml
    pyNeuroML >>> INFO - Loading NeuroML2 file: SKv3_1.channel.nml
    pyNeuroML >>> INFO - Loading NeuroML2 file: SKv3_1.channel.nml
    pyNeuroML >>> INFO - Loading LEMS file: /Users/padraig/git/Documentation/source/Userdocs/NML2_examples/LEMS_Test_SKv3_1.xml and running with jNeuroML
    pyNeuroML >>> INFO - Executing: (java -Xmx400M  -Djava.awt.headless=true -jar  "/opt/homebrew/anaconda3/envs/py39n/lib/python3.9/site-packages/pyneuroml/lib/jNeuroML-0.13.0-jar-with-dependencies.jar"  /Users/padraig/git/Documentation/source/Userdocs/NML2_examples/LEMS_Test_SKv3_1.xml  -nogui -I '') in directory: .
:::
:::

::: {.output .stream .highlight-myst-ansi .notranslate}
::: highlight
    Channel SKv3_1 (in file SKv3_1.channel.nml) has notes: Fast, non inactivating K+ current
                
    Comment from original mod file: 
    :Reference : :      Characterization of a Shaw-related potassium channel family in rat brain, The EMBO Journal, vol.11, no.7,2473-2486 (1992)
:::
:::

::: {.output .stderr .highlight-myst-ansi .notranslate}
::: highlight
    pyNeuroML >>> INFO - Command completed successfully!
:::
:::

![](_images/39f32324242c88ea6fbae0ad8a168db12e69b5d4003d3aea56582b6a8dd6d1cd.png)
![](_images/e7d9d45258035bf5b93bf809355fac7ebbf1889be2e005ea86bf616c5ae186cb.png)
![](_images/b1d3fc1ed74d01cb82fe356e2a98ce3ecc57e951417ad2310ba7caa0587cc44c.png)
![](_images/09b056ba3bcee90926ca2d8092d94e42615a0c2169f2d563a24732ed19aaf577.png)
![](_images/fb5a80742e218f1e3932e62a5e81cb3c1b8b113c9dc0e6a3645ac2459f3c02d7.png)
:::
:::
:::

::: {#find-a-passive-leak-current .section}
##### 1.5) Find a passive (leak) current[\#](#find-a-passive-leak-current "Link to this heading"){.headerlink}

::: {.cell .docutils .container}
::: {.cell_input .docutils .container}
::: {.highlight-ipython3 .notranslate}
::: highlight
    data = search_neuromldb("Leak", 'channel')

    pas_chan_doc, pas_chan_fname = get_model_from_neuromldb(data[0]['Model_ID'], 'channel')
    pas_chan = pas_chan_doc.ion_channel[0] # select the first ion channel in the nml doc

    print('Channel %s has notes: %s'%(pas_chan.id, pas_chan.notes))
:::
:::
:::

::: {.cell_output .docutils .container}
::: {.output .stream .highlight-myst-ansi .notranslate}
::: highlight
    NMLCH000015: Passive Leak, Maex De Schutter 1998
    NMLCH001623: Passive Leak, Bezaire Raikov 2016
    NMLCH001601: Passive Leak, Smith Smith 2013
    NMLCH001590: Passive Leak, Prinz Bucher 2004
    NMLCH001545: Passive Leak, Boyle Cohen 2008
    NMLCH001471: Passive Leak, Gouwens Berg 2018
    NMLCH001427: Passive Leak, Hodgkin Huxley 1952
    NMLCH001403: Passive Leak, Hay Hill 2011
    NMLCH000172: Passive Leak, Traub Contreras 2005
    NMLCH000130: Passive Leak, Pospischil Toledo-Rodriguez 2008
    NMLCH000114: Passive Leak, Hodgkin Huxley 1952
    NMLCH000095: Passive Leak, Vervaeke Lorincz 2010
    NMLCH000027: Passive Leak, De Schutter Bower 1994
    NMLCH000024: Passive Leak, Maex De Schutter 1998
    NMLCH000016: Passive Leak, Maex De Schutter 1998
    NMLCH000007: Passive Leak, Migliore Ferrante 2005
    NMLCH000139: Passive Leak, Hodgkin Huxley 1952
:::
:::

::: {.output .stderr .highlight-myst-ansi .notranslate}
::: highlight
    pyNeuroML >>> INFO - Loading NeuroML2 file: NMLCH000015.channel.nml
:::
:::

::: {.output .stream .highlight-myst-ansi .notranslate}
::: highlight
    Channel GranPassiveCond has notes: Simple leak conductance for Granule cell
:::
:::
:::
:::
:::
:::

::: {#create-a-new-cell-model-using-these-channels .section}
#### 2) Create a new cell model using these channels[\#](#create-a-new-cell-model-using-these-channels "Link to this heading"){.headerlink}

::: {#create-the-cell-object-add-channels-and-save-to-file .section}
##### 2.1) Create the Cell object, add channels and save to file[\#](#create-the-cell-object-add-channels-and-save-to-file "Link to this heading"){.headerlink}

::: {.cell .docutils .container}
::: {.cell_input .docutils .container}
::: {.highlight-ipython3 .notranslate}
::: highlight
    from neuroml import *
    from neuroml.utils import component_factory
    import neuroml.writers as writers


    nml_doc = NeuroMLDocument(id="TestCell")

    cell = component_factory("Cell", id="novel_cell")
    nml_doc.add(cell)

    cell.add_segment(prox=[0,0,0,17.841242], 
                     dist=[0,0,0,17.841242], 
                     seg_type='soma')

    cell.set_resistivity('0.03 kohm_cm')
    cell.set_init_memb_potential('-65mV')
    cell.set_specific_capacitance('1.0 uF_per_cm2')
    cell.set_spike_thresh('0mV')


    cell.add_channel_density(nml_doc,
                            cd_id='%s_chans'%na_chan.id,
                            ion_channel=na_chan.id,
                            cond_density='150 mS_per_cm2',
                            ion_chan_def_file=na_chan_fname,
                            erev="%s mV"%na_erev)

    cell.add_channel_density(nml_doc,
                            cd_id='%s_chans'%k_chan.id,
                            ion_channel=k_chan.id,
                            cond_density='36 mS_per_cm2',
                            ion_chan_def_file=k_chan_fname,
                            erev="%s mV"%k_erev)

    cell.add_channel_density(nml_doc,
                            cd_id='%s_chans'%pas_chan.id,
                            ion_channel=pas_chan.id,
                            cond_density='0.3 mS_per_cm2',
                            ion_chan_def_file=pas_chan_fname,
                            erev="-65 mV")



    cell_file = "%s.cell.nml"%nml_doc.id
    writers.NeuroMLWriter.write(nml_doc, cell_file)


    print("Written cell file to: " + cell_file)

    from neuroml.utils import validate_neuroml2

    validate_neuroml2(cell_file)

    #!cat TestCell.cell.nml


    pynml.summary(nml_doc, verbose=True)
:::
:::
:::

::: {.cell_output .docutils .container}
::: {.output .stream .highlight-myst-ansi .notranslate}
::: highlight
    Warning: Segment group all already exists.
    Warning: Segment group soma_group already exists.
    Written cell file to: TestCell.cell.nml
    It's valid!
    *******************************************************
    * NeuroMLDocument: TestCell
    *
    *
    *  Cell: novel_cell
    *    <Segment|0|Seg0>
    *      Parent segment: None (root segment)
    *      (0.0, 0.0, 0.0), diam 17.841242um -> (0.0, 0.0, 0.0), diam 17.841242um; seg length: 0.0 um
    *      Surface area: 1000.0000939925986 um2, volume: 2973.540612824116 um3
    *    Total length of 1 segment: 0.0 um; total area: 1000.0000939925986 um2
    *  
    *    SegmentGroup: soma_group,  1 member(s),    0 included group(s);    contains 1 segment, id: 0
    *    SegmentGroup: all, 1 member(s),    0 included group(s);    contains 1 segment, id: 0
    *  
    *    Channel density: NaTa_t_chans on all;  conductance of 150 mS_per_cm2 through ion chan NaTa_t with ion non_specific, erev: 50 mV
    *      Channel is on <Segment|0|Seg0>,    total conductance: 1500.0 S_per_m2 x 1.0000000939925986e-09 m2 = 1.500000140988898e-06 S (1500000.140988898 pS)
    *    Channel density: SKv3_1_chans on all;  conductance of 36 mS_per_cm2 through ion chan SKv3_1 with ion non_specific, erev: -77 mV
    *      Channel is on <Segment|0|Seg0>,    total conductance: 360.0 S_per_m2 x 1.0000000939925986e-09 m2 = 3.600000338373355e-07 S (360000.0338373355 pS)
    *    Channel density: GranPassiveCond_chans on all; conductance of 0.3 mS_per_cm2 through ion chan GranPassiveCond with ion non_specific, erev: -65 mV
    *      Channel is on <Segment|0|Seg0>,    total conductance: 3.0 S_per_m2 x 1.0000000939925986e-09 m2 = 3.000000281977796e-09 S (3000.000281977796 pS)
    *  
    *    Specific capacitance on all: 1.0 uF_per_cm2
    *      Capacitance of <Segment|0|Seg0>,   total capacitance: 0.01 F_per_m2 x 1.0000000939925986e-09 m2 = 1.0000000939925987e-11 F (10.000000939925988 pF)
    *
    *******************************************************
:::
:::

::: {.output .stderr .highlight-myst-ansi .notranslate}
::: highlight
    /opt/homebrew/anaconda3/envs/py39n/lib/python3.9/site-packages/neuroml/nml/generatedssupersuper.py:176: UserWarning: morphology has already been assigned.  Use `force=True` to overwrite. Hint: you can make changes to the already added object as required without needing to re-add it because only references to the objects are added, not their values.
      warnings.warn(
    /opt/homebrew/anaconda3/envs/py39n/lib/python3.9/site-packages/neuroml/nml/generatedssupersuper.py:176: UserWarning: biophysical_properties has already been assigned.  Use `force=True` to overwrite. Hint: you can make changes to the already added object as required without needing to re-add it because only references to the objects are added, not their values.
      warnings.warn(
    /opt/homebrew/anaconda3/envs/py39n/lib/python3.9/site-packages/neuroml/nml/generatedssupersuper.py:176: UserWarning: intracellular_properties has already been assigned.  Use `force=True` to overwrite. Hint: you can make changes to the already added object as required without needing to re-add it because only references to the objects are added, not their values.
      warnings.warn(
    /opt/homebrew/anaconda3/envs/py39n/lib/python3.9/site-packages/neuroml/nml/generatedssupersuper.py:176: UserWarning: membrane_properties has already been assigned.  Use `force=True` to overwrite. Hint: you can make changes to the already added object as required without needing to re-add it because only references to the objects are added, not their values.
      warnings.warn(
:::
:::
:::
:::
:::

::: {#generate-a-plot-of-the-actvity-of-the-cell-under-current-clamp-input .section}
##### 2.2) Generate a plot of the actvity of the cell under current clamp input[\#](#generate-a-plot-of-the-actvity-of-the-cell-under-current-clamp-input "Link to this heading"){.headerlink}

::: {.cell .docutils .container}
::: {.cell_input .docutils .container}
::: {.highlight-ipython3 .notranslate}
::: highlight
    from pyneuroml.analysis import generate_current_vs_frequency_curve

    generate_current_vs_frequency_curve(cell_file,
        cell.id,
        start_amp_nA=-0.02,
        end_amp_nA=0.06,
        step_nA=0.01,
        pre_zero_pulse=20,
        post_zero_pulse=20,
        analysis_duration=100,
        temperature='34degC',
        plot_voltage_traces=True)
:::
:::
:::

::: {.cell_output .docutils .container}
::: {.output .stderr .highlight-myst-ansi .notranslate}
::: highlight
    pyNeuroML >>> INFO - Loading NeuroML2 file: /Users/padraig/git/Documentation/source/Userdocs/NML2_examples/TestCell.cell.nml
    pyNeuroML >>> INFO - Loading NeuroML2 file: /Users/padraig/git/Documentation/source/Userdocs/NML2_examples/NaTa_t.channel.nml
    pyNeuroML >>> INFO - Loading NeuroML2 file: /Users/padraig/git/Documentation/source/Userdocs/NML2_examples/SKv3_1.channel.nml
    pyNeuroML >>> INFO - Loading NeuroML2 file: /Users/padraig/git/Documentation/source/Userdocs/NML2_examples/NMLCH000015.channel.nml
    pyNeuroML >>> INFO - Executing: (java -Xmx400M  -jar  "/opt/homebrew/anaconda3/envs/py39n/lib/python3.9/site-packages/pyneuroml/lib/jNeuroML-0.13.0-jar-with-dependencies.jar" -validate iv_novel_cell.net.nml ) in directory: .
    pyNeuroML >>> INFO - Command completed successfully!
    pyNeuroML >>> INFO - Loading NeuroML2 file: /Users/padraig/git/Documentation/source/Userdocs/NML2_examples/iv_novel_cell.net.nml
    pyNeuroML >>> INFO - Loading NeuroML2 file: /Users/padraig/git/Documentation/source/Userdocs/NML2_examples/TestCell.cell.nml
    pyNeuroML >>> INFO - Loading NeuroML2 file: /Users/padraig/git/Documentation/source/Userdocs/NML2_examples/NaTa_t.channel.nml
    pyNeuroML >>> INFO - Loading NeuroML2 file: /Users/padraig/git/Documentation/source/Userdocs/NML2_examples/SKv3_1.channel.nml
    pyNeuroML >>> INFO - Loading NeuroML2 file: /Users/padraig/git/Documentation/source/Userdocs/NML2_examples/NMLCH000015.channel.nml
    pyNeuroML >>> INFO - Loading LEMS file: LEMS_iv_novel_cell.xml and running with jNeuroML
    pyNeuroML >>> INFO - Executing: (java -Xmx400M  -Djava.awt.headless=true -jar  "/opt/homebrew/anaconda3/envs/py39n/lib/python3.9/site-packages/pyneuroml/lib/jNeuroML-0.13.0-jar-with-dependencies.jar"  LEMS_iv_novel_cell.xml  -nogui -I '') in directory: .
    pyNeuroML >>> INFO - Command completed successfully!
    /opt/homebrew/anaconda3/envs/py39n/lib/python3.9/site-packages/numpy/core/fromnumeric.py:3432: RuntimeWarning: Mean of empty slice.
      return _methods._mean(a, axis=axis, dtype=dtype,
    /opt/homebrew/anaconda3/envs/py39n/lib/python3.9/site-packages/numpy/core/_methods.py:190: RuntimeWarning: invalid value encountered in double_scalars
      ret = ret.dtype.type(ret / rcount)
    pyNeuroML >>> INFO - Generating plot: Membrane potential traces for: TestCell.cell.nml
    pyNeuroML >>> INFO - Generating plot: Firing frequency versus injected current for: TestCell.cell.nml
:::
:::

![](_images/6fa63944a2646426dff6ade512d68a6ba6a53b9b7057c9bcd55eea8f8e1dba21.png)
![](_images/d3f762fb7ec27fd7339831d1e89aa8dd8020512703457350a7edf9b4b328f4ff.png)

::: {.output .text_plain .highlight-myst-ansi .notranslate}
::: highlight
    {-0.02: 0.0,
     -0.01: 0.0,
     0.0: 0.0,
     0.01: 0.0,
     0.02: 0.0,
     0.03: 50.0,
     0.04: 70.0,
     0.05: 90.0}
:::
:::
:::
:::
:::
:::

::: {#search-for-download-and-analyse-a-complete-neuroml-cell .section}
#### 3) Search for, download and analyse a complete NeuroML cell[\#](#search-for-download-and-analyse-a-complete-neuroml-cell "Link to this heading"){.headerlink}

::: {#search-for-and-download-the-cell .section}
##### 3.1) Search for and download the cell[\#](#search-for-and-download-the-cell "Link to this heading"){.headerlink}

::: {.cell .docutils .container}
::: {.cell_input .docutils .container}
::: {.highlight-ipython3 .notranslate}
::: highlight
    data = search_neuromldb("Burst Accommodating Martinotti", 'cell')
:::
:::
:::

::: {.cell_output .docutils .container}
::: {.output .stream .highlight-myst-ansi .notranslate}
::: highlight
    NMLCL000109: Layer 2/3 Burst Accommodating Martinotti Cell (3), Markram Muller 2015
    NMLCL000395: Layer 6 Burst Non-accommodating Martinotti Cell (4), Markram Muller 2015
    NMLCL000393: Layer 6 Burst Non-accommodating Martinotti Cell (2), Markram Muller 2015
    NMLCL000345: Layer 4 Burst Non-accommodating Martinotti Cell (4), Markram Muller 2015
    NMLCL000343: Layer 4 Burst Non-accommodating Martinotti Cell (2), Markram Muller 2015
    NMLCL000315: Layer 2/3 Burst Non-accommodating Martinotti Cell (4), Markram Muller 2015
    NMLCL000313: Layer 2/3 Burst Non-accommodating Martinotti Cell (2), Markram Muller 2015
    NMLCL000191: Layer 6 Burst Accommodating Martinotti Cell (5), Markram Muller 2015
    NMLCL000190: Layer 6 Burst Accommodating Martinotti Cell (4), Markram Muller 2015
    NMLCL000188: Layer 6 Burst Accommodating Martinotti Cell (2), Markram Muller 2015
    NMLCL000161: Layer 5 Burst Accommodating Martinotti Cell (5), Markram Muller 2015
    NMLCL000160: Layer 5 Burst Accommodating Martinotti Cell (4), Markram Muller 2015
    NMLCL000158: Layer 5 Burst Accommodating Martinotti Cell (2), Markram Muller 2015
    NMLCL000157: Layer 5 Burst Accommodating Martinotti Cell (1), Markram Muller 2015
    NMLCL000136: Layer 4 Burst Accommodating Martinotti Cell (5), Markram Muller 2015
    NMLCL000134: Layer 4 Burst Accommodating Martinotti Cell (3), Markram Muller 2015
    NMLCL000133: Layer 4 Burst Accommodating Martinotti Cell (2), Markram Muller 2015
    NMLCL000111: Layer 2/3 Burst Accommodating Martinotti Cell (5), Markram Muller 2015
    NMLCL000108: Layer 2/3 Burst Accommodating Martinotti Cell (2), Markram Muller 2015
    NMLCL000187: Layer 6 Burst Accommodating Martinotti Cell (1), Markram Muller 2015
:::
:::
:::
:::

::: {.cell .docutils .container}
::: {.cell_input .docutils .container}
::: {.highlight-ipython3 .notranslate}
::: highlight
    model_id = 'NMLCL000109'

    def get_model_details_from_neuromldb(model_id):

        url = 'https://neuroml-db.org/api/model?id=%s'%model_id
        
        with urllib.request.urlopen(url) as res:
            model_details = json.load(res)
            
        for k in model_details['model']: print('%s:\t%s'%(k,model_details['model'][k]))
            
        return model_details
        

    model_details = get_model_details_from_neuromldb(model_id)
:::
:::
:::

::: {.cell_output .docutils .container}
::: {.output .stream .highlight-myst-ansi .notranslate}
::: highlight
    Model_ID:   NMLCL000109
    Status: CURRENT
    Errors: None
    Status_Timestamp:   2018-12-24T16:29:21+00:00
    Type:   Cell
    Equations:  2061
    Runtime_Per_Step:   0.000510815108024267
    Max_Stable_DT:  0.0625
    Max_Stable_DT_Error:    0.969373115260111
    Max_Stable_DT_Benchmark_RunTime:    4.903825037032964
    Optimal_DT: 0.00781557069252977
    Optimal_DT_Error:   0.122802372280653
    Optimal_DT_Benchmark_RunTime:   39.21518682026722
    Optimal_DT_a:   0.000968992248062015
    Optimal_DT_b:   15.8634962901345
    Optimal_DT_c:   -0.0136485748710791
    CVODE_baseline_step_frequency:  13111.7719061286
    CVODE_steps_per_spike:  766.973690134512
    CVODE_Benchmark_RunTime:    10.615508666396945
    Name:   Layer 2/3 Burst Accommodating Martinotti Cell (3)
    Directory_Path: /var/www/NeuroMLmodels/NMLCL000109
    File_Name:  bAC217_L23_MC_40be3bf0e8_0_0.cell.nml
    File_Updated:   None
    File_MD5_Checksum:  bcf74ccd8c6840f6c8f7551084c989b7
    File:   /var/www/NeuroMLmodels/NMLCL000109/bAC217_L23_MC_40be3bf0e8_0_0.cell.nml
    Publication_ID: 6000246
    Upload_Time:    2016-12-14T14:29:41+00:00
    Notes:  None
    ID_Helper:  175
    Sections:   40
    Compartments:   204
    Stability_Range_Low:    None
    Stability_Range_High:   None
    Is_Passive: 0
    Is_Intrinsically_Spiking:   0
    Resting_Voltage:    -71.2726975181673
    Rheobase_Low:   0.0469207763671875
    Rheobase_High:  0.048065185546875
    Threshold_Current_Low:  0.234375
    Threshold_Current_High: 0.29296875
    Bias_Voltage:   -80.0
    Bias_Current:   -0.0221022438906573
    CVODE_Active:   None
    Threshold:  None
    Is_GLIF:    0
    V_Variable: None
    Steady_State_Delay: None
    AP12AmplitudeDrop:  1.2698800006924
    AP1SSAmplitudeChange:   2.3903765783198
    AP1Amplitude:   72.1822400011024
    AP1WidthHalfHeight: 0.53
    AP1WidthPeakToTrough:   5.03000000000009
    AP1RateOfChangePeakToTrough:    -17.8182926253306
    AP1AHPDepth:    17.4437719043123
    AP2Amplitude:   70.91236000041
    AP2WidthHalfHeight: 0.55
    AP2WidthPeakToTrough:   6.22000000000003
    AP2RateOfChangePeakToTrough:    -14.2206642462703
    AP2AHPDepth:    17.5401716113915
    AP12AmplitudeChangePercent: -1.75926931704115
    AP12HalfWidthChangePercent: 3.77358490566034
    AP12RateOfChangePeakToTroughPercentChange:  -20.1906459541805
    AP12AHPDepthPercentChange:  0.552631091532312
    InputResistance:    386.187465741356
    AP1DelayMean:   20.6200000000001
    AP1DelaySD: 0.0
    AP2DelayMean:   72.1300000000001
    AP2DelaySD: 0.0
    Burst1ISIMean:  60.255
    Burst1ISISD:    0.0
    InitialAccommodationMean:   -75.0
    SSAccommodationMean:    -50.0
    AccommodationRateToSS:  -0.143802128271498
    AccommodationAtSSMean:  -83.5991575049674
    AccommodationRateMeanAtSS:  166.006825685025
    ISICV:  2.95112315366106
    ISIMedian:  216.41
    ISIBurstMeanChange: 33.954571927781
    SpikeRateStrongStim:    10.0
    AP1DelayMeanStrongStim: 7.30999999999995
    AP1DelaySDStrongStim:   0.0
    AP2DelayMeanStrongStim: 33.8900000000001
    AP2DelaySDStrongStim:   0.0
    Burst1ISIMeanStrongStim:    28.135
    Burst1ISISDStrongStim:  0.0
    RampFirstSpike: 2046.34
    FrequencyFilterType:    Low-Pass
    FrequencyPassAbove: 29.0
    FrequencyPassBelow: 58.7316087498722
    Ephyz_Cluster_ID:   /3/1/4/
    Channel_Type:   None
    Time_Step:  None
    Max_Stable_DT_Benchmark_RunTime_HH: 911.1330720253954
    Optimal_DT_Benchmark_RunTime_HH:    466.46500114887556
    CVODE_Benchmark_RunTime_HH: 498.01998845545205
    Stability_Range_Low_Corr:   -1.875
    Stability_Range_High_Corr:  15.0
:::
:::
:::
:::

::: {.cell .docutils .container}
::: {.cell_input .docutils .container}
::: {.highlight-ipython3 .notranslate}
::: highlight
    def get_full_model_from_neuromldb(model_id):
        
        fname = '%s.nml.zip'%(model_id)

        url = 'https://neuroml-db.org/GetModelZip?modelID=%s&version=NeuroML'%model_id
        r = requests.get(url)
        with open(fname , 'wb') as f:
            f.write(r.content)
        

        import zipfile
        with zipfile.ZipFile(fname, 'r') as z:
            z.extractall('.')

        
        print('Saved as %s'%fname)
        
    get_full_model_from_neuromldb(model_id)

    detailed_cell_file = model_details['model']['File_Name']

    validate_neuroml2(detailed_cell_file)
:::
:::
:::

::: {.cell_output .docutils .container}
::: {.output .stream .highlight-myst-ansi .notranslate}
::: highlight
    Saved as NMLCL000109.nml.zip
    It's valid!
:::
:::
:::
:::
:::

::: {#generate-3d-views-of-the-cell .section}
##### 3.2) Generate 3D views of the cell[\#](#generate-3d-views-of-the-cell "Link to this heading"){.headerlink}

::: {.cell .docutils .container}
::: {.cell_input .docutils .container}
::: {.highlight-ipython3 .notranslate}
::: highlight
    from pyneuroml.plot.PlotMorphology import plot_2D

    planes = ['yz', 'xz', 'xy']
    for plane in planes:

        plot_2D(detailed_cell_file,
                plane2d  = plane,
                min_width = 0,
                verbose= False,
                nogui = True,
                square=False)

    detailed_cell_doc = pynml.read_neuroml2_file(detailed_cell_file)
    detailed_cell = detailed_cell_doc.cells[0]
:::
:::
:::

::: {.cell_output .docutils .container}
::: {.output .stderr .highlight-myst-ansi .notranslate}
::: highlight
    pyNeuroML >>> INFO - Loading NeuroML2 file: bAC217_L23_MC_40be3bf0e8_0_0.cell.nml
    pyNeuroML >>> INFO - Loading NeuroML2 file: bAC217_L23_MC_40be3bf0e8_0_0.cell.nml
    pyNeuroML >>> INFO - Loading NeuroML2 file: bAC217_L23_MC_40be3bf0e8_0_0.cell.nml
    pyNeuroML >>> INFO - Loading NeuroML2 file: bAC217_L23_MC_40be3bf0e8_0_0.cell.nml
:::
:::

![](_images/311f26e8521b553eb3294ccf6409e24912e380c92535323da32bd6aa0111080a.png)
![](_images/10f7b384128918f1a4783bb32cf043728fd73a477f08fda6a6c09e9f49a0dd1d.png)
![](_images/65f04ed77b9509a582a77992bacb723f0ebc056e88ab96c78af71d744f2150dd.png)
:::
:::
:::

::: {#analyse-cell-spiking-behaviour .section}
##### 3.3) Analyse cell spiking behaviour[\#](#analyse-cell-spiking-behaviour "Link to this heading"){.headerlink}

::: {.cell .docutils .container}
::: {.cell_input .docutils .container}
::: {.highlight-ipython3 .notranslate}
::: highlight
    martinotti_cell = detailed_cell_doc.cells[0]

    generate_current_vs_frequency_curve(detailed_cell_file,
        martinotti_cell.id,
        start_amp_nA=-0.02,
        end_amp_nA=0.1,
        step_nA=0.02,
        pre_zero_pulse=100,
        post_zero_pulse=100,
        analysis_duration=1000,
        plot_voltage_traces=True,
        simulator='jNeuroML_NEURON')
:::
:::
:::

::: {.cell_output .docutils .container}
::: {.output .stderr .highlight-myst-ansi .notranslate}
::: highlight
    pyNeuroML >>> INFO - Loading NeuroML2 file: /Users/padraig/git/Documentation/source/Userdocs/NML2_examples/bAC217_L23_MC_40be3bf0e8_0_0.cell.nml
    pyNeuroML >>> INFO - Loading NeuroML2 file: /Users/padraig/git/Documentation/source/Userdocs/NML2_examples/K_Tst.channel.nml
    pyNeuroML >>> INFO - Loading NeuroML2 file: /Users/padraig/git/Documentation/source/Userdocs/NML2_examples/SKv3_1.channel.nml
    pyNeuroML >>> INFO - Loading NeuroML2 file: /Users/padraig/git/Documentation/source/Userdocs/NML2_examples/Nap_Et2.channel.nml
    pyNeuroML >>> INFO - Loading NeuroML2 file: /Users/padraig/git/Documentation/source/Userdocs/NML2_examples/NaTs2_t.channel.nml
    pyNeuroML >>> INFO - Loading NeuroML2 file: /Users/padraig/git/Documentation/source/Userdocs/NML2_examples/Ih.channel.nml
    pyNeuroML >>> INFO - Loading NeuroML2 file: /Users/padraig/git/Documentation/source/Userdocs/NML2_examples/pas.channel.nml
    pyNeuroML >>> INFO - Loading NeuroML2 file: /Users/padraig/git/Documentation/source/Userdocs/NML2_examples/Im.channel.nml
    pyNeuroML >>> INFO - Loading NeuroML2 file: /Users/padraig/git/Documentation/source/Userdocs/NML2_examples/NaTa_t.channel.nml
    pyNeuroML >>> INFO - Loading NeuroML2 file: /Users/padraig/git/Documentation/source/Userdocs/NML2_examples/Ca_LVAst.channel.nml
    pyNeuroML >>> INFO - Loading NeuroML2 file: /Users/padraig/git/Documentation/source/Userdocs/NML2_examples/SK_E2.channel.nml
    pyNeuroML >>> INFO - Loading NeuroML2 file: /Users/padraig/git/Documentation/source/Userdocs/NML2_examples/K_Pst.channel.nml
    pyNeuroML >>> INFO - Loading NeuroML2 file: /Users/padraig/git/Documentation/source/Userdocs/NML2_examples/CaDynamics_E2_NML2.nml
    pyNeuroML >>> INFO - Loading NeuroML2 file: /Users/padraig/git/Documentation/source/Userdocs/NML2_examples/Ca.channel.nml
    pyNeuroML >>> INFO - Executing: (java -Xmx400M  -jar  "/opt/homebrew/anaconda3/envs/py39n/lib/python3.9/site-packages/pyneuroml/lib/jNeuroML-0.13.0-jar-with-dependencies.jar" -validate iv_bAC217_L23_MC_40be3bf0e8_0_0.net.nml ) in directory: .
    pyNeuroML >>> INFO - Command completed successfully!
    pyNeuroML >>> INFO - Loading NeuroML2 file: /Users/padraig/git/Documentation/source/Userdocs/NML2_examples/iv_bAC217_L23_MC_40be3bf0e8_0_0.net.nml
    pyNeuroML >>> INFO - Loading NeuroML2 file: /Users/padraig/git/Documentation/source/Userdocs/NML2_examples/bAC217_L23_MC_40be3bf0e8_0_0.cell.nml
    pyNeuroML >>> INFO - Loading NeuroML2 file: /Users/padraig/git/Documentation/source/Userdocs/NML2_examples/K_Tst.channel.nml
    pyNeuroML >>> INFO - Loading NeuroML2 file: /Users/padraig/git/Documentation/source/Userdocs/NML2_examples/SKv3_1.channel.nml
    pyNeuroML >>> INFO - Loading NeuroML2 file: /Users/padraig/git/Documentation/source/Userdocs/NML2_examples/Nap_Et2.channel.nml
    pyNeuroML >>> INFO - Loading NeuroML2 file: /Users/padraig/git/Documentation/source/Userdocs/NML2_examples/NaTs2_t.channel.nml
    pyNeuroML >>> INFO - Loading NeuroML2 file: /Users/padraig/git/Documentation/source/Userdocs/NML2_examples/Ih.channel.nml
    pyNeuroML >>> INFO - Loading NeuroML2 file: /Users/padraig/git/Documentation/source/Userdocs/NML2_examples/pas.channel.nml
    pyNeuroML >>> INFO - Loading NeuroML2 file: /Users/padraig/git/Documentation/source/Userdocs/NML2_examples/Im.channel.nml
    pyNeuroML >>> INFO - Loading NeuroML2 file: /Users/padraig/git/Documentation/source/Userdocs/NML2_examples/NaTa_t.channel.nml
    pyNeuroML >>> INFO - Loading NeuroML2 file: /Users/padraig/git/Documentation/source/Userdocs/NML2_examples/Ca_LVAst.channel.nml
    pyNeuroML >>> INFO - Loading NeuroML2 file: /Users/padraig/git/Documentation/source/Userdocs/NML2_examples/SK_E2.channel.nml
    pyNeuroML >>> INFO - Loading NeuroML2 file: /Users/padraig/git/Documentation/source/Userdocs/NML2_examples/K_Pst.channel.nml
    pyNeuroML >>> INFO - Loading NeuroML2 file: /Users/padraig/git/Documentation/source/Userdocs/NML2_examples/CaDynamics_E2_NML2.nml
    pyNeuroML >>> INFO - Loading NeuroML2 file: /Users/padraig/git/Documentation/source/Userdocs/NML2_examples/Ca.channel.nml
    pyNeuroML >>> INFO - Loading LEMS file: LEMS_iv_bAC217_L23_MC_40be3bf0e8_0_0.xml and running with jNeuroML_NEURON
    pyNeuroML >>> INFO - Executing: (java -Xmx400M  -Djava.awt.headless=true -jar  "/opt/homebrew/anaconda3/envs/py39n/lib/python3.9/site-packages/pyneuroml/lib/jNeuroML-0.13.0-jar-with-dependencies.jar"  LEMS_iv_bAC217_L23_MC_40be3bf0e8_0_0.xml  -neuron -run -compile -nogui -I '') in directory: .
    pyNeuroML >>> INFO - Command completed successfully!
    /opt/homebrew/anaconda3/envs/py39n/lib/python3.9/site-packages/numpy/core/fromnumeric.py:3432: RuntimeWarning: Mean of empty slice.
      return _methods._mean(a, axis=axis, dtype=dtype,
    /opt/homebrew/anaconda3/envs/py39n/lib/python3.9/site-packages/numpy/core/_methods.py:190: RuntimeWarning: invalid value encountered in double_scalars
      ret = ret.dtype.type(ret / rcount)
    pyNeuroML >>> INFO - Generating plot: Membrane potential traces for: bAC217_L23_MC_40be3bf0e8_0_0.cell.nml
    pyNeuroML >>> INFO - Generating plot: Firing frequency versus injected current for: bAC217_L23_MC_40be3bf0e8_0_0.cell.nml
:::
:::

![](_images/486734328bf2595816a031df84ce456e26a3f31c3bce9e2f25b38ae50d9bf955.png)
![](_images/1921130adb0e4a2f5e4ebd1edccbfa8c217b6321b08f07a04036b3eca21588d9.png)

::: {.output .text_plain .highlight-myst-ansi .notranslate}
::: highlight
    {-0.02: 0.0, 0.0: 0.0, 0.02: 0.0, 0.04: 0.0, 0.06: 5.0, 0.08: 8.0, 0.1: 10.0}
:::
:::
:::
:::
:::
:::

::: {#create-a-small-network-using-the-cells .section}
#### 4) Create a small network using the cells[\#](#create-a-small-network-using-the-cells "Link to this heading"){.headerlink}

::: {#generate-the-neuroml-network .section}
##### 4.1) Generate the NeuroML network[\#](#generate-the-neuroml-network "Link to this heading"){.headerlink}

::: {.cell .docutils .container}
::: {.cell_input .docutils .container}
::: {.highlight-ipython3 .notranslate}
::: highlight
    from neuroml.utils import component_factory
    from pyneuroml import pynml
    from pyneuroml.lems import LEMSSimulation
    import neuroml.writers as writers
    import random

    nml_doc = component_factory("NeuroMLDocument", id="NML_DB_Net")


    ### Create the network
    net = nml_doc.add("Network", id="NML_DB_Net", validate=False)
    net.type="networkWithTemperature" 
    net.temperature="34.0degC"

    ### Add a synapse
    syn0 = nml_doc.add(
        "ExpOneSynapse", id="syn0", gbase="65nS", erev="0mV", tau_decay="3ms"
    )


    ## Create the first population
    size_exc = 4
    nml_doc.add("IncludeType", href=cell_file)

    pop_exc = component_factory("Population", id="Exc", component=cell.id, size=size_exc, type="population")
    # Set optional color property. Note: used later when generating plots
    ##pop0.add("Property", tag="color", value="0 0 .8")
    net.add(pop_exc)

    ## Create the second population
    size_inh = 4
    nml_doc.add("IncludeType", href=detailed_cell_file)

    pop_inh = component_factory("Population", id="Inh", component=detailed_cell.id, size=size_inh, type="population")
    # Set optional color property. Note: used later when generating plots
    ##pop1.add("Property", tag="color", value="0 0 .8")
    net.add(pop_inh)

    ## Create connections and inputs
    random.seed(123)
    prob_connection = 0.8

    proj_count = 0

    projection = Projection(
                        id="Proj_exc_inh",
                        presynaptic_population=pop_exc.id,
                        postsynaptic_population=pop_inh.id,
                        synapse=syn0.id,
                    )

    net.projections.append(projection)

    for i in range(0, size_exc):
        for j in range(0, size_inh):
            if random.random() <= prob_connection: # probablistic connection...
                connection = ConnectionWD(
                    id=proj_count,
                    pre_cell_id="%s[%i]" % (pop_exc.id, i),
                    post_cell_id="%s[%i]" % (pop_inh.id, j),
                    weight=random.random(),
                    delay='0ms'
                )
                projection.add(connection)
                proj_count += 1

    for i in range(0, size_exc):
        # pulse generator as explicit stimulus
        pg = nml_doc.add(
            "PulseGenerator",
            id="pg_exc_%i" % i,
            delay="20ms",
            duration="260ms",
            amplitude="%f nA" % (0.03 + 0.01 * random.random()),
        )
        
        exp_input = net.add(
            "ExplicitInput", target="%s[%i]" % (pop_exc.id, i), input=pg.id
        )

    for i in range(0, size_inh):
        # pulse generator as explicit stimulus
        pg = nml_doc.add(
            "PulseGenerator",
            id="pg_inh_%i"%i,
            delay="20ms",
            duration="260ms",
            amplitude="%f nA" % (0.02 + 0.02 * random.random()),
        )
        
        exp_input = net.add(
            "ExplicitInput", target="%s[%i]" % (pop_inh.id, i), input=pg.id
        )


    print(nml_doc.summary())

    nml_net_file = 'NML_DB_network.net.nml'
    writers.NeuroMLWriter.write(nml_doc, nml_net_file)

    print("Written network file to: " + nml_net_file)
    pynml.validate_neuroml2(nml_net_file)
:::
:::
:::

::: {.cell_output .docutils .container}
::: {.output .stderr .highlight-myst-ansi .notranslate}
::: highlight
    pyNeuroML >>> INFO - Executing: (java -Xmx400M  -jar  "/opt/homebrew/anaconda3/envs/py39n/lib/python3.9/site-packages/pyneuroml/lib/jNeuroML-0.13.0-jar-with-dependencies.jar" -validate NML_DB_network.net.nml ) in directory: .
:::
:::

::: {.output .stream .highlight-myst-ansi .notranslate}
::: highlight
    *******************************************************
    * NeuroMLDocument: NML_DB_Net
    *
    *  ExpOneSynapse: ['syn0']
    *  IncludeType: ['TestCell.cell.nml', 'bAC217_L23_MC_40be3bf0e8_0_0.cell.nml']
    *  PulseGenerator: ['pg_exc_0', 'pg_exc_1', 'pg_exc_2', 'pg_exc_3', 'pg_inh_0', 'pg_inh_1', 'pg_inh_2', 'pg_inh_3']
    *
    *  Network: NML_DB_Net (temperature: 34.0degC)
    *
    *   8 cells in 2 populations 
    *     Population: Exc with 4 components of type novel_cell
    *     Population: Inh with 4 components of type bAC217_L23_MC_40be3bf0e8_0_0
    *
    *   13 connections in 1 projections 
    *     Projection: Proj_exc_inh from Exc to Inh, synapse: syn0
    *       13 connections (wd): [(Connection 0: 0 -> 0, weight: 0.087187, delay: 0.00000 ms), ...]
    *
    *   0 inputs in 0 input lists 
    *
    *   8 explicit inputs (outside of input lists)
    *     Explicit Input of type pg_exc_0 to Exc(cell 0), destination: unspecified
    *     Explicit Input of type pg_exc_1 to Exc(cell 1), destination: unspecified
    *     Explicit Input of type pg_exc_2 to Exc(cell 2), destination: unspecified
    *     Explicit Input of type pg_exc_3 to Exc(cell 3), destination: unspecified
    *     Explicit Input of type pg_inh_0 to Inh(cell 0), destination: unspecified
    *     Explicit Input of type pg_inh_1 to Inh(cell 1), destination: unspecified
    *     Explicit Input of type pg_inh_2 to Inh(cell 2), destination: unspecified
    *     Explicit Input of type pg_inh_3 to Inh(cell 3), destination: unspecified
    *
    *******************************************************
    Written network file to: NML_DB_network.net.nml
:::
:::

::: {.output .stderr .highlight-myst-ansi .notranslate}
::: highlight
    pyNeuroML >>> INFO - Command completed successfully!
    pyNeuroML >>> INFO - Output: 
      jNeuroML >>   jNeuroML v0.13.0
      jNeuroML >>  Validating: /Users/padraig/git/Documentation/source/Userdocs/NML2_examples/NML_DB_network.net.nml
      jNeuroML >>  Valid against schema and all tests
      jNeuroML >>  No warnings
      jNeuroML >>  
      jNeuroML >>  Validated 1 files: All valid and no warnings
      jNeuroML >>  
      jNeuroML >>  
:::
:::

::: {.output .text_plain .highlight-myst-ansi .notranslate}
::: highlight
    True
:::
:::
:::
:::
:::

::: {#generate-views-of-the-network .section}
##### 4.2) Generate views of the network[\#](#generate-views-of-the-network "Link to this heading"){.headerlink}

::: {.cell .docutils .container}
::: {.cell_input .docutils .container}
::: {.highlight-ipython3 .notranslate}
::: highlight
    from pyneuroml.pynml import generate_nmlgraph

    generate_nmlgraph(nml_net_file, level=-1, engine="dot")

    from IPython.display import Image
    Image(filename='%s.gv.png'%net.id,width=500)
:::
:::
:::

::: {.cell_output .docutils .container}
::: {.output .stderr .highlight-myst-ansi .notranslate}
::: highlight
    pyNeuroML >>> INFO - Converting NML_DB_network.net.nml to graphical form, level -1, engine dot
:::
:::

::: {.output .stream .highlight-myst-ansi .notranslate}
::: highlight
    neuromllite >>> Initiating GraphViz handler, level -1, engine: dot, seed: 1234
    Parsing: NML_DB_network.net.nml
    Loaded: NML_DB_network.net.nml as NeuroMLDocument
    neuromllite >>> Document: NML_DB_Net
    neuromllite >>> Network: NML_DB_Net
    neuromllite >>> Population: Exc, component: novel_cell (Cell), size: 4 cells, properties: {}
    neuromllite >>> Population: Inh, component: bAC217_L23_MC_40be3bf0e8_0_0 (Cell), size: 4 cells, properties: {}
    neuromllite >>> GRAPH PROJ: Proj_exc_inh (Exc (4) -> Inh (4), projection): w 1.0; wtot: 4.849021209484878; sign: 1; cond: 65.0 nS (65nS); all: {'projection': 58.85091239849243, 'inhibitory': -1e+100, 'excitatory': -1e+100, 'electricalProjection': -1e+100, 'continuousProjection': -1e+100} -> {'projection': 1.380338363104285, 'inhibitory': 1e+100, 'excitatory': 1e+100, 'electricalProjection': 1e+100, 'continuousProjection': 1e+100}
    neuromllite >>>  - conn Exc_0 -> Inh_0: 5.66713403897108 (5.66713403897108)
    neuromllite >>>  - conn Exc_0 -> Inh_1: 7.000515270998535 (7.000515270998535)
    neuromllite >>>  - conn Exc_0 -> Inh_3: 34.85313260220526 (34.85313260220526)
    neuromllite >>>  - conn Exc_1 -> Inh_0: 55.385630230408985 (55.385630230408985)
    neuromllite >>>  - conn Exc_1 -> Inh_1: 21.91908271210294 (21.91908271210294)
    neuromllite >>>  - conn Exc_1 -> Inh_2: 15.935617913644716 (15.935617913644716)
    neuromllite >>>  - conn Exc_1 -> Inh_3: 28.357926571989168 (28.357926571989168)
    neuromllite >>>  - conn Exc_2 -> Inh_0: 38.84396519171941 (38.84396519171941)
    neuromllite >>>  - conn Exc_2 -> Inh_1: 20.504833292831094 (20.504833292831094)
    neuromllite >>>  - conn Exc_2 -> Inh_2: 58.85091239849243 (58.85091239849243)
    neuromllite >>>  - conn Exc_2 -> Inh_3: 9.241379188047565 (9.241379188047565)
    neuromllite >>>  - conn Exc_3 -> Inh_0: 1.380338363104285 (1.380338363104285)
    neuromllite >>>  - conn Exc_3 -> Inh_2: 17.245910842001596 (17.245910842001596)
    neuromllite >>> Generating graph for: NML_DB_Net
    neuromllite >>> **************************************
    neuromllite >>> * Settings for GraphVizHandler: 
    neuromllite >>> *
    neuromllite >>> *    engine:                  dot
    neuromllite >>> *    level:                   -1
    neuromllite >>> *    is_cell_level:           True
    neuromllite >>> *    CUTOFF_INH_SYN_MV:       -50
    neuromllite >>> *    include_ext_inputs:      True
    neuromllite >>> *    include_input_pops:      True
    neuromllite >>> *    scale_by_post_pop_size:  True
    neuromllite >>> *    scale_by_post_pop_cond:  True
    neuromllite >>> *    min_weight_to_show:      0
    neuromllite >>> *    show_chem_conns:         True
    neuromllite >>> *    show_elect_conns:        True
    neuromllite >>> *    show_cont_conns:         True
    neuromllite >>> *    output_format:           png
    neuromllite >>> *
    neuromllite >>> * Used values: 
    neuromllite >>> *    syn_conds_used:          syn0:     65.0 nS (65nS)
    neuromllite >>> *
    neuromllite >>> **************************************
:::
:::

::: {.output .stderr .highlight-myst-ansi .notranslate}
::: highlight
    /opt/homebrew/anaconda3/envs/py39n/lib/python3.9/subprocess.py:1052: ResourceWarning: subprocess 25497 is still running
      _warn("subprocess %s is still running" % self.pid,
    ResourceWarning: Enable tracemalloc to get the object allocation traceback
    pyNeuroML >>> INFO - Done with GraphViz...
:::
:::

[![](_images/cbcc1f3b0b4c7e625953206eac46859f31f199aa62d7a8ea9eb208ec67dda9c9.png){style="width: 500px;"}](_images/cbcc1f3b0b4c7e625953206eac46859f31f199aa62d7a8ea9eb208ec67dda9c9.png){.reference
.internal .image-reference}
:::
:::
:::

::: {#generate-lems-file-to-simulate-the-network .section}
##### 4.3) Generate LEMS file to simulate the network[\#](#generate-lems-file-to-simulate-the-network "Link to this heading"){.headerlink}

See documentation about [LEMS Simulation
files](https://docs.neuroml.org/Userdocs/LEMSSimulation.html){.reference
.external}.

::: {.cell .docutils .container}
::: {.cell_input .docutils .container}
::: {.highlight-ipython3 .notranslate}
::: highlight
    simulation_id = "NML_DB_network_sim"
    simulation = LEMSSimulation(sim_id=simulation_id,
                                duration=300, dt=0.025, simulation_seed=123)
    simulation.assign_simulation_target(net.id)
    simulation.include_neuroml2_file(nml_net_file, include_included=False)


    pops = [pop_exc, pop_inh]

    for pop in pops:
        simulation.create_event_output_file(
            f"{pop.id}_spikes", f"{pop.id}.spikes", format='ID_TIME'
        )
        simulation.create_output_file(pop.id, "%s.v.dat" % pop.id)

        for pre in range(0, pop.size):
            next_cell = '{}[{}]'.format(pop.id,pre)
            simulation.add_selection_to_event_output_file(
                f"{pop.id}_spikes", pop.component, next_cell, 'spike')

            simulation.add_column_to_output_file(pop.id, f"v{pre}", f"{next_cell}/v")
        
        
    lems_simulation_file = simulation.save_to_file()

    print(f"Saved LEMS Simulation file to: {lems_simulation_file}")
:::
:::
:::

::: {.cell_output .docutils .container}
::: {.output .stream .highlight-myst-ansi .notranslate}
::: highlight
    Saved LEMS Simulation file to: LEMS_NML_DB_network_sim.xml
:::
:::
:::
:::
:::

::: {#run-simulation-of-the-network-using-neuron .section}
##### 4.4) Run simulation of the network using NEURON[\#](#run-simulation-of-the-network-using-neuron "Link to this heading"){.headerlink}

::: {.cell .docutils .container}
::: {.cell_input .docutils .container}
::: {.highlight-ipython3 .notranslate}
::: highlight
    traces = pynml.run_lems_with_jneuroml_neuron(
        lems_simulation_file, max_memory="2G", nogui=True, load_saved_data=True, plot=True
    )
:::
:::
:::

::: {.cell_output .docutils .container}
::: {.output .stderr .highlight-myst-ansi .notranslate}
::: highlight
    pyNeuroML >>> INFO - Loading LEMS file: LEMS_NML_DB_network_sim.xml and running with jNeuroML_NEURON
    pyNeuroML >>> INFO - Executing: (java -Xmx2G  -Djava.awt.headless=true -jar  "/opt/homebrew/anaconda3/envs/py39n/lib/python3.9/site-packages/pyneuroml/lib/jNeuroML-0.13.0-jar-with-dependencies.jar"  LEMS_NML_DB_network_sim.xml  -neuron -run -compile -nogui -I '') in directory: .
    pyNeuroML >>> INFO - Command completed successfully!
    pyNeuroML >>> WARNING - Reloading: Data loaded from ./Exc.v.dat (jNeuroML_NEURON)
    pyNeuroML >>> WARNING - Reloading: Data loaded from ./Inh.v.dat (jNeuroML_NEURON)
:::
:::

![](_images/b2f5f0a7b7b908b19a44be5ea8289dfd0fe0d44b6e3017b6dbd63b20f3e06d2f.png)
:::
:::
:::
:::
:::
:::
:::

[]{#document-Userdocs/FindingNeuroMLModels}

::: {#finding-and-sharing-neuroml-models .section .tex2jax_ignore .mathjax_ignore}
[]{#userdocs-finding-models}

## Finding and sharing NeuroML models[\#](#finding-and-sharing-neuroml-models "Link to this heading"){.headerlink}

There are an increasing number of repositories where you can find
NeuroML models, many of which will are accepting submissions from the
community who wish to share their work in this format.

::: {#neuroml-db-the-neuroml-database .section}
[]{#userdocs-finding-models-neuromldb}

### NeuroML-DB: The NeuroML Database[\#](#neuroml-db-the-neuroml-database "Link to this heading"){.headerlink}

::: {.tip .admonition}
Read the NeuroML-DB paper!

Read the paper: [NeuroML-DB: Sharing and characterizing data-driven
neuroscience models described in
NeuroML](https://doi.org/10.1371/journal.pcbi.1010941){.reference
.external}([\[[BHG+23](Landing.html#id23 "Justas Birgiolas, Vergil Haynes, Padraig Gleeson, Richard C. Gerkin, Suzanne W. Dietrich, and Sharon Crook. Neuroml-db: sharing and characterizing data-driven neuroscience models described in neuroml. PLOS Computational Biology, 19(3):1-29, 2023. URL: https://doi.org/10.1371/journal.pcbi.1010941, doi:10.1371/journal.pcbi.1010941."){.reference
.internal}\]]{#id1}) .
:::

<figure id="id2" class="align-center">
<img src="_images/NML-DB.png" alt="NeuroML Database" />
<figcaption><p><span class="caption-number">Fig. 26 </span><span
class="caption-text">The NeuroML Database contains NeuroML files for
many <a href="https://neuroml-db.org/model_info?model_id=NMLCL000938"
class="reference external">cells</a> (left above), <a
href="https://neuroml-db.org/model_info?model_id=NMLCH000103"
class="reference external">channels</a> (right) and synapses taken from
Open Source Brain, Blue Brain Project, Allen Institute and
more.</span><a href="#id2" class="headerlink"
title="Link to this image">#</a></p></figcaption>
</figure>

Go

\

The [NeuroML Database](https://neuroml-db.org/){.reference .external} is
a relational database that provides a means for sharing NeuroML model
descriptions and their components. One of its goals is to contribute to
an efficient tool chain for model development using NeuroML. This
emphasis allows the database design and subsequent searching to take
advantage of this specific format. In particular, the NeuroML database
allows for efficient searches over the components of models and metadata
that are associated with a hierarchical NeuroML model description.

The NeuroML Database is developed and maintained by the [ICON
Lab](https://iconlab.asu.edu/){.reference .external} at [Arizona State
University](https://asu.edu/){.reference .external}.

To submit your NeuroML model to NeuroML-DB, please see the information
on [this page](https://neuroml-db.org/about){.reference .external}.
:::

::: {#open-source-brain .section}
[]{#userdocs-finding-models-osbv1}

### Open Source Brain[\#](#open-source-brain "Link to this heading"){.headerlink}

<figure id="id3" class="align-center">
<img src="_images/OSBv1.png" alt="Open Source Brain" />
<figcaption><p><span class="caption-number">Fig. 27 </span><span
class="caption-text">Examples of NeuroML 2 models visualised on Open
Source Brain. A) <a
href="https://www.opensourcebrain.org/projects/hodgkin-huxley-tutorial?explorer=https%3A%2F%2Fraw.githubusercontent.com%2Fopensourcebrain%2Ftutorials%2Fdevelopment%2Fmodels%2FhodgkinHuxley%2FGEPPETTO.json"
class="reference external">Hodgkin Huxley model</a> interactive
tutorial. B) Integrate and fire network model of cortical column (<a
href="https://www.opensourcebrain.org/projects/potjansdiesmann2014"
class="reference external">Potjans and Diesmann 2014</a>), showing
network connectivity. C) Cortical model with multicompartmental cells
(<a href="https://www.opensourcebrain.org/projects/thalamocortical"
class="reference external">Traub et al. 2005</a>), showing network
properties and simulated membrane potential activity. D) Model of C.
elegans nervous system from <a
href="https://www.opensourcebrain.org/projects/c302/"
class="reference external">OpenWorm project</a>. All
visualisation/analysis/simulation enabled due to models being in
standardised NeuroML format.</span><a href="#id3" class="headerlink"
title="Link to this image">#</a></p></figcaption>
</figure>

Go

\

[Open Source Brain](https://www.opensourcebrain.org){.reference
.external} is a platform for sharing, viewing, analysing, and simulating
standardized models from different brain regions and species. An index
of various NeuroML models on Open Source Brain and their validation
status can be seen
[here](https://github.com/OpenSourceBrain/.github/blob/main/testsheet/README.md){.reference
.external}.

To add your NeuroML model to Open Source Brain, please see the
information on [[this page]{.std
.std-ref}](Landing.html#userdocs-creating-models-converting-conductance){.reference
.internal}.
:::

::: {#other-related-projects .section}
[]{#userdocs-finding-models-others}

### Other related projects[\#](#other-related-projects "Link to this heading"){.headerlink}

::: {.admonition .note}
Note

Needs introductory text.
:::
:::

::: {#neuromorpho-org .section}
[]{#userdocs-finding-models-neuromorpho}

### [NeuroMorpho.Org](http://NeuroMorpho.Org){.reference .external}[\#](#neuromorpho-org "Link to this heading"){.headerlink}

[NeuroMorpho.Org](https://neuromorpho.org){.reference .external} is a
database of digitally reconstructed neurons. This resource can be used
to retrieve reconstructed neuronal morphologies of multiple cell types
from a number of species. The database can be browsed by neuron type,
brain area, species, contributing lab, or cells can be searched for
according to various morphometric criteria or the associated metadata.

There is a utility present on the site to view the cells in 3D (based on
Robert Cannon's Cvapp), which can also save the morphologies in NeuroML
2 format.

A tutorial on getting data from
[NeuroMorpho.Org](http://NeuroMorpho.Org){.reference .external} in
NeuroML format can be found
[here](https://github.com/NeuralEnsemble/NeuroinformaticsTutorial/blob/master/Exercises/Exercise1_NeuroMorpho_to_OSB.md){.reference
.external}.
:::

::: {#openworm .section}
[]{#userdocs-finding-models-openworm}

### OpenWorm[\#](#openworm "Link to this heading"){.headerlink}

The [OpenWorm project](http://www.openworm.org){.reference .external}
aims to create a simulation platform to build digital *in-silico* living
systems, starting with a C. elegans virtual organism simulation. The
simulations and associated tools are being developed in a fully open
source manner.

NeuroML is being used for the description of the 302 neurons in the
[worm's nervous
system](https://www.opensourcebrain.org/projects/c302/){.reference
.external}, both for morphological description of the cells and their
electrical properties.
:::

::: {#allen-institute .section}
[]{#userdocs-finding-models-alleninstitute}

### Allen Institute[\#](#allen-institute "Link to this heading"){.headerlink}

Multiple cell models as produced by the [Allen
Institute](https://alleninstitute.org/){.reference .external} as part of
their large scale brain modelling efforts are available in NeuroML
format
[here](https://github.com/OpenSourceBrain/AllenInstituteNeuroML){.reference
.external}.
:::

::: {#blue-brain-project .section}
[]{#userdocs-finding-models-bluebrainproject}

### Blue Brain Project[\#](#blue-brain-project "Link to this heading"){.headerlink}

The detailed cortical cell models from the [Blue Brain
Project](http://bluebrain.epfl.ch){.reference .external} have been
converted to NeuroML format, along with the ion channels from the
[Channelpedia database](https://channelpedia.epfl.ch){.reference
.external}. See
[here](https://github.com/OpenSourceBrain/BlueBrainProjectShowcase){.reference
.external} for details.
:::
:::

[]{#document-Userdocs/CreatingNeuroMLModels}

::: {#creating-neuroml-models .section .tex2jax_ignore .mathjax_ignore}
[]{#userdocs-creating-models}

## Creating NeuroML models[\#](#creating-neuroml-models "Link to this heading"){.headerlink}

There are 3 main ways of developing a new model of a neuronal system in
NeuroML

**1) Reuse elements from previous NeuroML models**

There are an increasing number of resources where you can find and
analyse previously developed NeuroML models to use as the basis for a
new model. See [[here]{.std
.std-ref}](Landing.html#userdocs-finding-models){.reference .internal}
for details.

**2) Writing models from scratch using Python NeuroML tools**

The toolchain around NeuroML means that it is possible to create a model
in NeuroML format from the start. Please see the [[Getting Started with
NeuroML section]{.std
.std-ref}](Landing.html#userdocs-getting-started-neuroml){.reference
.internal} for quick examples on how you can use [[pyNeuroML]{.std
.std-ref}](Landing.html#pyneuroml){.reference .internal} to create
NeuroML models and run them.

**3) Convert a published model developed in a simulator specific format
to NeuroML**

Most computational models used in publications are released in the
particular format used by the authors during their research, often in a
general purpose simulator like [[NEURON]{.std
.std-ref}](Landing.html#userdocs-supporting-apps-neuron){.reference
.internal}. Many of these can be found on
[ModelDB](https://senselab.med.yale.edu/ModelDB/default){.reference
.external}. Converting one of these to NeuroML format will mean that all
further developments/modifications of the model will be standards
compliant, and will give access to all of the NeuroML compliant tools
for visualising/analysing/optimising/sharing the model, as well as
providing multiple options for executing the model across multiple
simulators.

The next page is a **step by step guide** to creating a new NeuroML
model based on an existing published model, verifying its behaviour, and
sharing it with the community on the Open Source Brain platform.

::: {.toctree-wrapper .compound}
[]{#document-Userdocs/ConvertingModels}

::: {#converting-models-to-neuroml-and-sharing-them-on-open-source-brain .section .tex2jax_ignore .mathjax_ignore}
[]{#userdocs-creating-models-converting-conductance}

### Converting models to NeuroML and sharing them on Open Source Brain[\#](#converting-models-to-neuroml-and-sharing-them-on-open-source-brain "Link to this heading"){.headerlink}

::: {.tip .admonition}
Walk throughs available

Look at the [[Walkthroughs chapter]{.std
.std-ref}](Landing.html#userdocs-walkthroughs){.reference .internal} for
worked examples.
:::

The figure below is taken from the supplementary information of the
[[Open Source Brain paper]{.std
.std-ref}](Landing.html#papers-osb){.reference .internal}, and gives a
quick overview of the steps required and tools available for converting
a model to NeuroML and sharing it on the OSB platform.

<figure id="id2" class="align-center">
<a href="_images/osb-conversion.png"
class="reference internal image-reference"><img
src="_images/osb-conversion.png" style="width: 60%;"
alt="Schamatic of process of converting models to NeuroML" /></a>
<figcaption><p><span class="caption-number">Fig. 28 </span><span
class="caption-text">Procedures and tools to convert models from native
formats to NeuroML and PyNN (Taken from Gleeson et al. 2019 <span
id="id1">[<a href="Landing.html#id12" class="reference internal"
title="Padraig Gleeson, Matteo Cantarelli, Boris Marin, Adrian Quintana, Matt Earnshaw, Sadra Sadeh, Eugenio Piasini, Justas Birgiolas, Robert C. Cannon, N. Alex Cayco-Gajic, Sharon Crook, Andrew P. Davison, Salvador Dura-Bernal, András Ecker, Michael L. Hines, Giovanni Idili, Frederic Lanore, Stephen D. Larson, William W. Lytton, Amitava Majumdar, Robert A. McDougal, Subhashini Sivagnanam, Sergio Solinas, Rokas Stanislovas, Sacha J. van Albada, Werner van Geit, and R. Angus Silver. Open source brain: a collaborative resource for visualizing, analyzing, simulating, and developing standardized models of neurons and circuits. Neuron, 103(3):395–411, 2019. doi:10.1016/j.neuron.2019.05.019.">GCM+19</a>]</span>)</span><a
href="#id2" class="headerlink"
title="Link to this image">#</a></p></figcaption>
</figure>

::: {#step-1-find-the-original-model-code .section}
#### Step 1) Find the original model code[\#](#step-1-find-the-original-model-code "Link to this heading"){.headerlink}

While it should in principle be possible to create the model based only
on the description in the accompanying publication, having the original
code is invaluable. The original code allows the identification of all
parameters related to the model, and it is required to verify the
dynamical behaviour of the NeuroML equivalent.

Scripts for an increasing number of published models are available on
[ModelDB](https://modeldb.science){.reference .external}. ModelDB models
are also published as [GitHub
repositories](https://github.com/ModelDBRepository){.reference
.external}. Forks of these are also managed in the [Open Source Brain
GitHub organization](https://github.com/opensourcebrain/){.reference
.external} and indexed on [Open Source Brain version
2](https://v2.opensourcebrain.org/){.reference .external}.

So, the first step is to obtain the original model code and verify that
this can be run to reproduce the published results.
:::

::: {#step-2-create-github-and-osb-accounts-for-sharing-the-code .section}
[]{#userdocs-creating-models-converting-conductance-github}

#### Step 2) Create GitHub and OSB accounts for sharing the code[\#](#step-2-create-github-and-osb-accounts-for-sharing-the-code "Link to this heading"){.headerlink}

::: {#a-sign-up-to-github-and-osb .section}
##### 2a) Sign up to GitHub and OSB[\#](#a-sign-up-to-github-and-osb "Link to this heading"){.headerlink}

Sign up to [GitHub](https://github.com/signup){.reference .external} to
be able to share the updated code publicly. Next, sign up to [Open
Source
Brain](https://www.opensourcebrain.org/account/register){.reference
.external}, and adding a reference to your GitHub user account will help
link between the two resources.
:::

::: {#b-create-github-repository .section}
##### 2b) Create GitHub repository[\#](#b-create-github-repository "Link to this heading"){.headerlink}

Create a new [GitHub
repository](https://docs.github.com/en/repositories){.reference
.external} for your new model. There are plenty of examples of
repositories containing NeuroML [on
OSB](https://github.com/orgs/OpenSourceBrain/repositories){.reference
.external}. It's fine to share the code under your own user account, but
if you would like to host it at
[OpenSourceBrain](https://github.com/OpenSourceBrain){.github .reference
.external}, please [get in contact with the OSB
team](https://docs.opensourcebrain.org/General/Contacts.html){.reference
.external}.

Now you can commit the scripts for original version of the model to your
GitHub repository. **Please check what the license/redistribution
conditions are for the code!** Authors who have shared their code on
[ModelDB](https://senselab.med.yale.edu/ModelDB/default){.reference
.external} are generally happy for the code to be reused, but it is good
to get in contact with them as a courtesy to let them know your plans
with the model. They will generally be very supportive as long as the
original publications are referenced, and will often have useful
information on any updated versions of the model. Adding or updating a
[README
file](https://docs.opensourcebrain.org/OSBv1/Write_Your_Project_Documentation.html#add-a-readme-file-in-your-github-repository-and-reuse-it-on-your-osb-projects){.reference
.external} will be valuable for anyone who comes across the model on
GitHub.
:::

::: {#c-create-osb-project .section}
##### 2c) Create OSB project[\#](#c-create-osb-project "Link to this heading"){.headerlink}

Now you can create a project on OSB which will point to the GitHub
repository and will be able to find any NeuroML models committed to it.
You can also add a link back to the original archived version on
ModelDB, and even reuse your README on GitHub as a description. For more
details on this see
[here](https://docs.opensourcebrain.org/OSBv1/Creating_Your_Own_Project.html){.reference
.external}.
:::
:::

::: {#step-3-improve-and-test-original-model-code .section}
#### Step 3) Improve and test original model code[\#](#step-3-improve-and-test-original-model-code "Link to this heading"){.headerlink}

With the original simulator code shared on GitHub, and a README updated
to describe it, new users will be able to clone the repository and start
using the code as shared by the authors. Some updates may be required
and any changes from the original version will be recorded under the Git
history visible on GitHub.

::: {#a-make-simpler-modularised-versions-of-original-model-scripts .section}
##### 3a) Make simpler/modularised versions of original model scripts[\#](#a-make-simpler-modularised-versions-of-original-model-scripts "Link to this heading"){.headerlink}

Many of the model scripts which get released on ModelDB aim to reproduce
one or two of the figures from the associated publication. However,
these scripts can be quite complex, and mix simulation with some
analysis of the results. They don't always provide a single, simple run
of the model with standard parameters, which would be the target for a
first version of the model in NeuroML.

Therefore it would be useful to create some additional scripts (reusing
cell/channel definition files as much as possible) illustrating the
baseline behaviour of the model, including:

-   A simple script with a single cell (or one for each if multiple
    cells present) - applying a simple current pulse into each (e.g.
    [example1](https://github.com/mbezaire/ca1/blob/development/NeuroML2/olm.hoc){.reference
    .external},
    [example2](https://github.com/OpenSourceBrain/MiglioreEtAl14_OlfactoryBulb3D/blob/master/NEURON/mitral.hoc){.reference
    .external})

-   A single compartment (soma only) example with all the ion channels
    (ideally one where channels can easily be added/commented out) -
    apply current pulse
    ([example](https://github.com/OpenSourceBrain/BlueBrainProjectShowcase/blob/master/NMC/NEURON/Test_Soma.hoc){.reference
    .external} in NEURON)

-   A passive version of multi-compartmental cell with multiple
    locations recorded

-   A multi-compartmental cell with multiple channels and calcium
    dynamics, with the channels specified in separate files

These will be much easier to compare to equivalents in NeuroML.
:::

::: {#b-add-omv-tests .section}
[]{#userdocs-creating-models-converting-conductance-add-omv-tests}

##### 3b) Add OMV tests[\#](#b-add-omv-tests "Link to this heading"){.headerlink}

::: {.dropdown .tip .admonition}
Optional, but recommended.

This step is optional, but highly recommended to create automated tests
on the behaviour of the model.
:::

Once you have some scripts which illustrate (in plots/saved data) the
baseline expected behaviour of your model (spiketimes, rate of firing
etc.), it would be good to put some checks in place which can be run to
ensure this behaviour stays consistent across changes/commits to your
repository, different versions of the underlying simulator, as well as
providing a target for what the NeuroML version of the model should
produce.

The **[Open Source Brain Model validation
framework](https://github.com/OpenSourceBrain/osb-model-validation){.reference
.external} (OMV)** is designed for exactly this, allowing small scripts
to be added to your repository stating what files to execute in what
simulation engine and what the expected properties of generated output
should be. These tests can be run on your local machine during
development, but can also be easily integrated with [GitHub
Actions](https://github.com/features/actions){.reference .external},
allowing tests across multiple simulators to be run every time there is
a commit to the repository
([example](https://github.com/OpenSourceBrain/IzhikevichModel/actions/runs/1520638984){.reference
.external}).

To start using this for your project, install OMV and test running it on
your local machine ([`omv`{.docutils .literal
.notranslate}]{.pre}` `{.docutils .literal .notranslate}[`all`{.docutils
.literal .notranslate}]{.pre}) on some standard examples (e.g. [Hay et
al.](https://github.com/OpenSourceBrain/L5bPyrCellHayEtAl2011){.reference
.external}).

Add OMV tests for your native simulator scripts
([example](https://github.com/OpenSourceBrain/GranCellSolinasEtAl10/blob/master/NEURON/.test.nrnpy.omt){.reference
.external}), e.g. test the spike times of cell when simple current pulse
applied. Commit this file to GitHub, along with a GitHub Actions
workflow
([example](https://github.com/OpenSourceBrain/IzhikevichModel/blob/master/.github/workflows/omv-ci.yml){.reference
.external}), and look for runs under the Actions tab of your project on
GitHub.

Later, you can add OMV tests too for the equivalent NeuroML versions,
reusing the Model Emergent Property ([`*.mep`{.docutils .literal
.notranslate}]{.pre}) file
([example](https://github.com/mbezaire/ca1/blob/development/NeuroML2/cells/tests/.test.sca.jnmlnrn.omt){.reference
.external}), thus testing that the behaviours of the 2 versions are the
same (within a certain tolerance).
:::
:::

::: {#create-a-version-of-the-model-in-neuroml-2 .section}
[]{#userdocs-creating-models-converting-conductance-initialnml2}

#### 4) Create a version of the model in NeuroML 2[\#](#create-a-version-of-the-model-in-neuroml-2 "Link to this heading"){.headerlink}

::: {#a-create-a-lems-simulation-file-to-run-the-model .section}
##### 4a) Create a LEMS Simulation file to run the model[\#](#a-create-a-lems-simulation-file-to-run-the-model "Link to this heading"){.headerlink}

A [[LEMS Simulation file]{.std
.std-ref}](Landing.html#userdocs-lemssimulation){.reference .internal}
is required to specify how to run a simulation of the NeuroML model, how
long to run, what to plot/save etc. Create a LEMS\*.xml
([example](https://github.com/OpenSourceBrain/BlueBrainProjectShowcase/blob/master/NMC/NeuroML2/LEMS_Soma_AllNML2.xml){.reference
.external}) with \*.net.nml
([example](https://github.com/OpenSourceBrain/BlueBrainProjectShowcase/blob/master/NMC/NeuroML2/Soma_AllNML2.net.nml){.reference
.external}) and \*.cell.nml
([example](https://github.com/OpenSourceBrain/BlueBrainProjectShowcase/blob/master/NMC/NeuroML2/Soma_AllNML2.cell.nml){.reference
.external}) for **a cell with only a soma** (don't try to match a full
multi-compartmental cell with all channels to the original version at
this early stage).

Start off with only passive parameters (capacitance, axial resistance
and 1 leak current) set; gradually add channels as in 4b) below; apply a
current pulse and save soma membrane potential to file.

Ensure all [`*nml`{.docutils .literal .notranslate}]{.pre} files are
[[valid]{.std
.std-ref}](Landing.html#userdocs-validating-models){.reference
.internal}. Ensure the [`LEMS*.xml`{.docutils .literal
.notranslate}]{.pre} runs with [`jnml`{.docutils .literal
.notranslate}]{.pre}; visually compare the behaviour with original
simple script from the previous section.

Ensure the [`LEMS*.xml`{.docutils .literal .notranslate}]{.pre} runs
with [`jnml`{.docutils .literal .notranslate}]{.pre}` `{.docutils
.literal .notranslate}[`-neuron`{.docutils .literal
.notranslate}]{.pre}, producing similar behaviour. If there is a good
correspondence, add OMV tests for the NeuroML version, using the Model
Emergent Property ([`*.mep`{.docutils .literal .notranslate}]{.pre})
file from the original script's test.

When ready, commit the LEMS/NeuroML code to GitHub.
:::

::: {#b-convert-channels-to-neuroml .section}
[]{#userdocs-creating-models-converting-conductance-convert-channels}

##### 4b) Convert channels to NeuroML[\#](#b-convert-channels-to-neuroml "Link to this heading"){.headerlink}

Restructure/annotate/comment channel files in the original model to be
as clear as possible and ideally have all use the same overall structure
(e.g. see mod files
[here](https://github.com/mbezaire/ca1/tree/development){.reference
.external}).

(Optional) Create a (Python) script/notebook which contains the core
activation variable expressions for the channels; this can be useful to
restructure/test/plot/alter units of the expressions before generating
the equivalent in NeuroML
([example](https://github.com/OpenSourceBrain/PINGnets/blob/master/NeuroML2/ConvertChannels.ipynb){.reference
.external}).

If you are using NEURON, use [`pynml-modchananalysis`{.docutils .literal
.notranslate}]{.pre} to generate plots of the activation variables for
the channels in the mod files
([example1](https://github.com/NeuroML/pyNeuroML/blob/master/examples/analyseNaMod.sh){.reference
.external},
[example2](https://github.com/OpenSourceBrain/BlueBrainProjectShowcase/blob/master/NMC/NEURON/analyse.sh){.reference
.external}).

Start from an existing similar example of an ion channel in NeuroML
([examples1](https://github.com/OpenSourceBrain/AllenInstituteNeuroML/tree/master/CellTypesDatabase/models/NeuroML2){.reference
.external},
[examples2](https://github.com/RokasSt/Thalamocortical/tree/master/NeuroML2/channels){.reference
.external},
[examples3](https://github.com/mbezaire/ca1/tree/development/NeuroML2/channels){.reference
.external}).

Use [`pynml-channelanalysis`{.docutils .literal .notranslate}]{.pre} to
generate similar plots for your NeuroML based channels as your mod
channels; these can easily be plotted for adding to your GitHub repo as
summary pages
([example1](https://github.com/mbezaire/ca1/blob/development/NeuroML2/channels/channel_summary/README.md){.reference
.external},
[example2](https://github.com/OpenSourceBrain/MiglioreEtAl14_OlfactoryBulb3D/blob/master/NeuroML2/Channels/channel_summary/README.md){.reference
.external}).

Create a script to load the output of mod analysis and nml analysis and
compare the outputs
([example](https://github.com/OpenSourceBrain/BlueBrainProjectShowcase/blob/master/NMC/NeuroML2/compare_nml2_mods.py){.reference
.external}).
:::

::: {#c-compare-single-compartment-cell-with-channels .section}
[]{#userdocs-creating-models-converting-conductance-compare-single-comp}

##### 4c) Compare single compartment cell with channels[\#](#c-compare-single-compartment-cell-with-channels "Link to this heading"){.headerlink}

Ensure you have a passive soma example in NeuroML which reproduces the
behaviour of an equivalent passibe version inthe original format (from
steps 3a and 4a above).

Gradually test the cell with passive conductance and *each channel
individually*. Plot v along with rate variables for each channel &
compare how they look during current pulse ([example in
NEURON](https://github.com/OpenSourceBrain/BlueBrainProjectShowcase/blob/master/NMC/NEURON/Test_Soma.hoc){.reference
.external} vs [example in
NeuroML](https://github.com/OpenSourceBrain/BlueBrainProjectShowcase/blob/master/NMC/NeuroML2/Soma_AllNML2.cell.nml){.reference
.external} and
[LEMS](https://github.com/OpenSourceBrain/BlueBrainProjectShowcase/blob/master/NMC/NeuroML2/LEMS_Soma_AllNML2.xml){.reference
.external})

Test these in [`jnml`{.docutils .literal .notranslate}]{.pre} first,
then in Neuron with [`jnml`{.docutils .literal
.notranslate}]{.pre}` `{.docutils .literal
.notranslate}[`-neuron`{.docutils .literal .notranslate}]{.pre}.

When you are happy with each of the channels, try the soma with all of
the channels in place, with the same channel density as present in the
soma of the original cell.
:::

::: {#d-compare-multi-compartmental-cell-incorporating-channels .section}
[]{#userdocs-creating-models-converting-conductance-compare-multi-comp}

##### 4d) Compare multi-compartmental cell incorporating channels[\#](#d-compare-multi-compartmental-cell-incorporating-channels "Link to this heading"){.headerlink}

If the model was created in NEURON, export the 3D morphology from the
original NEURON scripts using pyNeuroML
([example](https://github.com/OpenSourceBrain/SmithEtAl2013-L23DendriticSpikes/blob/master/NeuroML2/export_nml2.py){.reference
.external}); this will be easier if there is a hoc script with just a
single cell instance as in section 1). While there is the option to use
[`includeBiophysicalProperties=True`{.docutils .literal
.notranslate}]{.pre} and this will attempt to export the conductance
densities on different groups, it may be better to consolidate these and
add them afterwards using correctly named groups and the most efficient
representation of conductance density to group relationships
([example](https://github.com/OpenSourceBrain/MiglioreEtAl14_OlfactoryBulb3D/blob/master/Python/Export/export_mitral.py){.reference
.external}).

::: {.highlight-python .notranslate}
::: highlight
    from pyneuroml.neuron import export_to_neuroml2
    ..
    export_to_neuroml2("test.hoc", "test.morphonly.cell.nml", includeBiophysicalProperties=False)
:::
:::

Alternatively manually add the [`<channelDensity>`{.docutils .literal
.notranslate}]{.pre} elements to the cell file (as
[here](https://github.com/OpenSourceBrain/SmithEtAl2013-L23DendriticSpikes/blob/master/NeuroML2/L23_NoHotSpot.cell.nml#L16711){.reference
.external}).

You can use the tools for [[visualising NeuroML Models]{.std
.std-ref}](Landing.html#userdocs-visualising-models){.reference
.internal} to compare how these versions look agains the originals.

As with the single compartment example, it's best to **start off with
the passive case**, i.e no active channels on the soma or dendrites, and
compare that to the original code (for membrane potential at multiple
locations!), and gradually add channels.

Many projects on OSB were originally converted from the original format
(NEURON, GENESIS, etc.) to NeuroML v1 using [[neuroConstruct]{.std
.std-ref}](Landing.html#userdocs-supporting-apps-neuroconstruct){.reference
.internal} (see
[here](http://www.opensourcebrain.org/search_custom_field?f%5B%5D=43&amp;op%5B43%5D=~&amp;v%5B43%5D%5B%5D=neuroConstruct){.reference
.external} for a list of these). neuroConstruct has good support for
export to NeuroML v2, and this code could form the basis for your
conversion. More on using neuroConstruct
[here](http://www.opensourcebrain.org/docs#Using_neuroConstruct_Based_Projects){.reference
.external} and details on conversion of models to NeuroML v1
[here](http://www.neuroconstruct.org/docs/importneuron.html#Converting+mod+file%2FGENESIS+script+channels+into+ChannelML){.reference
.external}.

Note: you can also export other morphologies from
[NeuroMorpho.org](https://neuromorpho.org){.reference .external} in
NeuroML2 format
([example](https://github.com/NeuralEnsemble/NeuroinformaticsTutorial/blob/master/Exercises/Exercise1_NeuroMorpho_to_OSB.md){.reference
.external}) to try out different reconstructions of the same cell type
with your complement of channels.
:::
:::

::: {#e-re-optimising-cell-models .section}
[]{#userdocs-creating-models-converting-conductance-reoptimise}

#### 4e) (Re)optimising cell models[\#](#e-re-optimising-cell-models "Link to this heading"){.headerlink}

You can use
[Neurotune](https://github.com/NeuralEnsemble/neurotune/){.reference
.external} inside pyNeuroML to re-optimise your cell models. An example
is
[here](https://github.com/NeuroML/pyNeuroML/blob/master/examples/tuneHHCell.py){.reference
.external}, and a full sequence of optimising a NeuroML model against
data in NWB can be found [[here]{.std
.std-ref}](Landing.html#userdocs-optimising){.reference .internal}.
:::

::: {#f-create-an-equivalent-network-model-in-neuroml .section}
#### 4f) Create an equivalent network model in NeuroML[\#](#f-create-an-equivalent-network-model-in-neuroml "Link to this heading"){.headerlink}

Creating an equivalent of a complex network model originally built in
hoc for example in NeuroML is not trivial. The guide to network building
with libNeuroML [[here]{.std
.std-ref}](Landing.html#userdocs-gettingstarted-izhikevichnetwork){.reference
.internal} is a good place to start.

See also [[NeuroMLlite]{.std
.std-ref}](Landing.html#neuromllite){.reference .internal}.
:::

::: {#access-view-and-run-your-model-on-osb .section}
#### 5) Access, view and run your model on OSB[\#](#access-view-and-run-your-model-on-osb "Link to this heading"){.headerlink}

When you're happy that a version of the model is behaving correctly in
NeuroML, you can try visualising it on OSB.

See
[here](https://docs.opensourcebrain.org/OSBv1/Five_Minute_Introduction.html){.reference
.external} for more details about viewing and simulating projects on
OSB.
:::

::: {#share-and-collaborate .section}
#### 6) Share and collaborate[\#](#share-and-collaborate "Link to this heading"){.headerlink}

There is more information on how you can disseminate and promote your
model once it is on OSB in the main documentation for that platform:
[https://docs.opensourcebrain.org](https://docs.opensourcebrain.org){.reference
.external}.

Consider sharing parts of the model on [[other NeuroML supporting
resources]{.std
.std-ref}](Landing.html#userdocs-finding-models){.reference .internal}
(e.g. cell and channel files on NeuroML-DB).
:::
:::

[]{#document-Userdocs/ImportingMorphologyFiles}

::: {#handling-morphology-files .section .tex2jax_ignore .mathjax_ignore}
[]{#userdocs-importing-morphology-files}

### Handling Morphology Files[\#](#handling-morphology-files "Link to this heading"){.headerlink}

A number of formats are used in neuroscience to encode neuronal
morphologies obtained from experiments involving neuronal
reconstructions. This page provides general information on these
formats, and documents how they may be converted to NeuroML 2 for use in
computational models.

::: {#terminology .section}
[]{#userdocs-importing-morphology-files-terminology}

#### Terminology[\#](#terminology "Link to this heading"){.headerlink}

<figure id="id2" class="align-center">
<img src="_images/MorphologyNeuroML2.png"
alt="Morphologies in NeuroML 2." />
<figcaption><p><span class="caption-number">Fig. 29 </span><span
class="caption-text">Specification of morphologies in <strong>NeuroML
2</strong>. More details can be found for each element in the
specification, e.g. <a href="Landing.html#schema-cell"
class="reference internal"><span
class="std std-ref">&lt;cell&gt;</span></a>, <a
href="Landing.html#schema-morphology" class="reference internal"><span
class="std std-ref">&lt;morphology&gt;</span></a>, <a
href="Landing.html#schema-segment" class="reference internal"><span
class="std std-ref">&lt;segment&gt;</span></a>, <a
href="Landing.html#schema-segmentgroup" class="reference internal"><span
class="std std-ref">&lt;segmentGroup&gt;</span></a>, <a
href="Landing.html#schema-proximal" class="reference internal"><span
class="std std-ref">&lt;proximal&gt;</span></a>, <a
href="Landing.html#schema-distal" class="reference internal"><span
class="std std-ref">&lt;distal&gt;</span></a>.</span><a href="#id2"
class="headerlink" title="Link to this image">#</a></p></figcaption>
</figure>

<figure id="id3" class="align-center">
<a href="_images/crook2007-morphml-figure1.png"
class="reference internal image-reference"><img
src="_images/crook2007-morphml-figure1.png" style="width: 500px;"
alt="Figure 1 from {cite}`Crook2007` showing different representations of neuronal morphology." /></a>
<figcaption><p><span class="caption-number">Fig. 30 </span><span
class="caption-text">Figure 1 from <span id="id1">[<a
href="Landing.html#id8" class="reference internal"
title="Sharon Crook, Padraig Gleeson, Fred Howell, Joseph Svitak, and R. Angus Silver. MorphML: level 1 of the NeuroML standards for neuronal morphology data and model specification. Neuroinformatics, 5(2):96–104, 5 2007. doi:10.1007/s12021-007-0003-6.">CGH+07</a>]</span>,
a schematic comparing handling of morphological information for a simple
cell by different applications. a) Schematic of biological cell
structure to be represented. b) <strong>Neurolucida
reconstruction</strong> where the soma is represented by an outline and
three-dimensional points are specified along each branch. c)
<strong>NEURON simulator</strong> format where cell structure is
specified in <em>sections</em>. Only the center of the section is
simulated unless the nseg parameter is greater than one. d)
<strong>GENESIS simulator representation</strong> using compartments
that are cylinders except for the soma, which can be spherical. The
optimal length of each compartment is determined by the electrotonic
length. e) <strong>MorphML representation</strong> (i.e. NeuroML v1)
where any of the information shown in panels b through d can be
encoded.</span><a href="#id3" class="headerlink"
title="Link to this image">#</a></p></figcaption>
</figure>

All formats have their own terminology that is used to refer to
different parts of the cell.

In
[NEURON](https://neuronsimulator.github.io/nrn/python/modelspec/programmatic/topology/geometry.html){.reference
.external}:

-   a [`section`{.code .docutils .literal .notranslate}]{.pre} is an
    unbranched contiguous cell region

-   the morphology of a cell is defined by 3D points, [`pt3D`{.docutils
    .literal .notranslate}]{.pre}

-   for simulation, one can specify how many segments a section should
    be divided into, given by [`nseg`{.code .docutils .literal
    .notranslate}]{.pre}

In NeuroML:

-   segments are 3D points describing the cell morphology

-   continuous, unbranched segments groups, would form a section

-   the [`numberInternalDivisions`{.code .docutils .literal
    .notranslate}]{.pre} property can be used to set the number of
    divisions a segment or segment group should be divided into for
    simulation
:::

::: {#cell-validity .section}
[]{#userdocs-importing-morphology-files-validity}

#### Cell validity[\#](#cell-validity "Link to this heading"){.headerlink}

In general, it is usually necessary to examine NeuroML cells converted
from various formats, especially experimental reconstructions, before
they can be used in simulations. This is because reconstructions may not
always contain all the information necessary to simulate the cell.

Two potential problems that must be checked are:

-   Point of connection of dendritic branches to the soma: e.g., in
    Neurolucida, there is no explicit soma but usually only an outline.

-   Zero length sections: NEURON can work with zero segment lengths
    (consecutive [`pt3d`{.docutils .literal .notranslate}]{.pre} points
    being equal), but a standard mapping of this may not be supported in
    other simulators such as GENESIS.

An incomplete list of checks to make to ensure a valid cell is (taken
from [[neuroConstruct]{.std
.std-ref}](Landing.html#userdocs-supporting-apps-neuroconstruct){.reference
.internal}):

-   Only one segment should be without a parent (root)

-   All segments must have sections

-   All segments must have endpoints

-   All segments must have unique IDs

-   All segments must have unique names

-   All sections must have unique names

-   Segments after the first in a section must only be connected to 1
    parent

-   Only one segment may be spherical and must belong to the
    [`soma_group`{.code .docutils .literal .notranslate}]{.pre}
    SegmentGroup

-   The cell must have at least one segment

-   The cell must have at least one soma section, i.e., which is in the
    [`soma_group`{.code .docutils .literal .notranslate}]{.pre}

-   The cell must have a cell name

The NeuroML validation tools will check for some of these and report
errors where possible.
:::

::: {#formats .section}
[]{#userdocs-importing-morphology-files-formats}

#### Formats[\#](#formats "Link to this heading"){.headerlink}

::: {#neuroml2 .section}
[]{#userdocs-importing-morphology-files-formats-neuroml}

##### NeuroML2[\#](#neuroml2 "Link to this heading"){.headerlink}

In NeuroML, morphologies are encapsulated in the [[morphology]{.std
.std-ref}](Landing.html#schema-morphology){.reference .internal}
modelling element. A morphology includes [[segments]{.std
.std-ref}](Landing.html#schema-segment){.reference .internal} and
[[segments groups]{.std
.std-ref}](Landing.html#schema-segmentgroup){.reference .internal}, and
these can be used to refer to parts of the cell's morphology, for
example, when placing ionic conductances. A number of conventions for
use in morphologies are listed [[here]{.std
.std-ref}](Landing.html#userdocs-conventions-segments){.reference
.internal}.

::: {#morphologies-can-be-stored-in-external-files .section}
[]{#userdocs-importing-morphology-files-formats-neuroml-external}

###### Morphologies can be stored in external files[\#](#morphologies-can-be-stored-in-external-files "Link to this heading"){.headerlink}

::: {.note .admonition}
Requires jNeuroML v0.13.2, pyNeuroML v1.3.2

The functionality to store morphology information in external files was
implemented in jNeuroML v0.13.2, and pyNeuroML v1.3.2. Please ensure you
are using these or newer versions to use this feature.
:::

Usually, morphologies are embedded in NeuroML cell definition files,
[[for example]{.std
.std-ref}](Landing.html#userdocs-getting-started-multi-compartment-example){.reference
.internal}:

::: {.highlight-xml .notranslate}
::: highlight
        <cell id="pyr_soma_m_in_b_in">
        <!-- ... -->

            <morphology id="morph0">

                <segment >
                <!-- more segments and segment groups -->

            </morphology>

            <biophysicalProperties id="biophys1">
                <!-- biophysical properties contents -->
            </biophysicalProperties>

        </cell>
:::
:::

However, morphologies (and [[biophysical properties]{.std
.std-ref}](Landing.html#schema-biophysicalproperties){.reference
.internal}) can also be stored as "standalone" entities outside the cell
definition and referred to. Further, they can also be stored in external
files that may be "included" in the cell definition file (using the
[IncludeType](https://libneuroml.readthedocs.io/en/latest/userdocs/coreclasses.html#includetype){.reference
.external} model element). This allows the re-use of morphology and
biophysical properties in multiple cell models:

::: {.highlight-xml .notranslate}
::: highlight
        <cell id="pyr_soma_m_out_b_out" morphology="morph0" biophysicalProperties="biophys1">
            <!-- cell contents without morphology and biophysical properties -->
        </cell>

        <!-- Potentially in other files... -->

        <morphology id="morph0">

            <segment >
            <!-- more segments and segment groups -->

        </morphology>

        <biophysicalProperties id="biophys1">
            <!-- biophysical properties contents -->
        </biophysicalProperties>
:::
:::
:::
:::

::: {#neuron .section}
[]{#userdocs-importing-morphology-files-formats-neuron}

##### NEURON[\#](#neuron "Link to this heading"){.headerlink}

There is no fixed format in NEURON for specifying morphologies. However,
cells created in NEURON may be exported to NeuroML2 format using the
[[`export_to_neuroml2`{.docutils .literal
.notranslate}]{.pre}](https://pyneuroml.readthedocs.io/en/stable/pyneuroml.neuron.html#pyneuroml.neuron.export_to_neuroml2){.reference
.external} method included in [[pyNeuroML]{.std
.std-ref}](Landing.html#pyneuroml){.reference .internal}
([example](https://github.com/OpenSourceBrain/SmithEtAl2013-L23DendriticSpikes/blob/master/NeuroML2/export_nml2.py){.reference
.external}).
:::

::: {#genesis .section}
[]{#userdocs-importing-morphology-files-formats-genesis}

##### GENESIS[\#](#genesis "Link to this heading"){.headerlink}

The format for a GENESIS cell description is given
[here](http://www.genesis-sim.org/GENESIS/Hyperdoc/Manual-25.html#readcell){.reference
.external}.
:::

::: {#cvapp-swc-files .section}
[]{#userdocs-importing-morphology-files-formats-cvapp-swc}

##### CVApp/SWC files[\#](#cvapp-swc-files "Link to this heading"){.headerlink}

Please see this [[page]{.std
.std-ref}](Landing.html#userdocs-swc){.reference .internal}.
:::

::: {#neurolucida .section}
[]{#userdocs-importing-morphology-files-formats-neurolucida}

##### Neurolucida[\#](#neurolucida "Link to this heading"){.headerlink}

The [Neurolucida](http://www.mbfbioscience.com/neurolucida){.reference
.external} file format is used by MicroBrightField products to store
information on neuronal reconstructions. Both binary and ASCII format
files can be generated by these products. The format allows recording of
various anatomical features, not only neuronal processes such as
dendrites and cell bodies, but can record other micro-anatomical
features of potential interest to anatomists. Not all of these features
will be relevant when constructing a single cell computational model.
:::
:::

::: {#tools .section}
[]{#userdocs-importing-morphology-files-tools}

#### Tools[\#](#tools "Link to this heading"){.headerlink}

::: {#cvapp .section}
[]{#userdocs-importing-morphology-files-tools-cvapp}

##### CVApp[\#](#cvapp "Link to this heading"){.headerlink}

The standalone
[CVApp](https://github.com/NeuroML/Cvapp-NeuroMorpho.org){.reference
.external} tool provides an interface to visualize SWC files and export
them into NeuroML2. For more information, please see this [[page]{.std
.std-ref}](Landing.html#userdocs-swc){.reference .internal}.
:::

::: {#neuroconstruct .section}
[]{#userdocs-importing-morphology-files-tools-neuroconstruct}

##### neuroConstruct[\#](#neuroconstruct "Link to this heading"){.headerlink}

[[neuroConstruct]{.std
.std-ref}](Landing.html#userdocs-supporting-apps-neuroconstruct){.reference
.internal} includes functionality to interactively import GENESIS,
NEURON, CVapp (SWC), Neurolucida, and older MorphML formats to NeuroML2.
Please see the [neuroConstruct
documentation](http://www.neuroconstruct.org/docs/import.html){.reference
.external} for more information.
:::

::: {#pyneuroml .section}
[]{#userdocs-importing-morphology-files-tools-pyneuroml}

##### pyNeuroML[\#](#pyneuroml "Link to this heading"){.headerlink}

[[pyNeuroML]{.std .std-ref}](Landing.html#pyneuroml){.reference
.internal} includes functionality to convert NEURON files into NeuroML
using the [[`export_to_neuroml2`{.docutils .literal
.notranslate}]{.pre}](https://pyneuroml.readthedocs.io/en/stable/pyneuroml.neuron.html#pyneuroml.neuron.export_to_neuroml2){.reference
.external} method included in [[pyNeuroML]{.std
.std-ref}](Landing.html#pyneuroml){.reference .internal}
([example](https://github.com/OpenSourceBrain/SmithEtAl2013-L23DendriticSpikes/blob/master/NeuroML2/export_nml2.py){.reference
.external}).
:::
:::
:::

[]{#document-Userdocs/HDF5}

::: {#hdf5-support .section .tex2jax_ignore .mathjax_ignore}
[]{#userdocs-hdf5}

### HDF5 support[\#](#hdf5-support "Link to this heading"){.headerlink}

The XML serializations of large NeuroML models can be prohibitive to
store. For such cases, NeuroML also includes support for saving models
in the binary [HDF5](https://www.hdfgroup.org/solutions/hdf5){.reference
.external} format via the [NeuroMLHdf5Writer in
libNeuroML](https://libneuroml.readthedocs.io/en/stable/userdocs/writers.html#neuroml.writers.NeuroMLHdf5Writer){.reference
.external}. The same format can be exported also from the Java API
([example](https://github.com/NeuroML/org.neuroml.model/blob/master/src/test/java/org/neuroml/model/test/HDF5Test.java){.reference
.external}).

The format of the export is documented below:

-   [[Network]{.std .std-ref}](Landing.html#schema-network){.reference
    .internal} is exported as a [`network`{.docutils .literal
    .notranslate}]{.pre} HDF5 group with [`id`{.docutils .literal
    .notranslate}]{.pre}, [`notes`{.docutils .literal
    .notranslate}]{.pre}, and the [`temperature`{.docutils .literal
    .notranslate}]{.pre} (optional) stored as attributes.

-   [[Population]{.std
    .std-ref}](Landing.html#schema-population){.reference .internal} is
    exported as a group with id [`population_<id`{.docutils .literal
    .notranslate}]{.pre}` `{.docutils .literal
    .notranslate}[`of`{.docutils .literal
    .notranslate}]{.pre}` `{.docutils .literal
    .notranslate}[`the`{.docutils .literal
    .notranslate}]{.pre}` `{.docutils .literal
    .notranslate}[`population>`{.docutils .literal .notranslate}]{.pre}
    with [`id`{.docutils .literal .notranslate}]{.pre},
    [`component`{.docutils .literal .notranslate}]{.pre},
    [`size`{.docutils .literal .notranslate}]{.pre}, [`type`{.docutils
    .literal .notranslate}]{.pre}, and [`property`{.docutils .literal
    .notranslate}]{.pre} tags stored as attributes.

    -   If the population is a [[population list]{.std
        .std-ref}](Landing.html#schema-populationlist){.reference
        .internal} that includes [[instances]{.std
        .std-ref}](Landing.html#schema-instance){.reference .internal}
        of cells, the locations of cells (x, y, z), these are stored in
        a 3 column table ("chunked array") with a row per instance.

-   [[Projection]{.std
    .std-ref}](Landing.html#schema-projection){.reference .internal} is
    exported as a group with id [`project_<id`{.docutils .literal
    .notranslate}]{.pre}` `{.docutils .literal
    .notranslate}[`of`{.docutils .literal
    .notranslate}]{.pre}` `{.docutils .literal
    .notranslate}[`the`{.docutils .literal
    .notranslate}]{.pre}` `{.docutils .literal
    .notranslate}[`projection>`{.docutils .literal .notranslate}]{.pre}
    with [`id`{.docutils .literal .notranslate}]{.pre},
    [`type`{.docutils .literal .notranslate}]{.pre},
    [`presynapticPopulation`{.docutils .literal .notranslate}]{.pre},
    [`postSynapticPopulation`{.docutils .literal .notranslate}]{.pre},
    [`synapse`{.docutils .literal .notranslate}]{.pre} as attributes.

    -   [[Connection]{.std
        .std-ref}](Landing.html#schema-connection){.reference .internal}
        and [[ConnectionWD]{.std
        .std-ref}](Landing.html#schema-connectionwd){.reference
        .internal} elements in projections are stored as rows in a table
        with the first two columns as the [`pre_cell_id`{.docutils
        .literal .notranslate}]{.pre} and [`post_cell_id`{.docutils
        .literal .notranslate}]{.pre} respectively, and the successive
        columns for the necessary attributes.

-   [[ElectricalProjection]{.std
    .std-ref}](Landing.html#schema-electricalprojection){.reference
    .internal} is exported similar to Projection with the
    [[ElectricalConnection]{.std
    .std-ref}](Landing.html#schema-electricalconnection){.reference
    .internal}, [[ElectricalConnectionInstance]{.std
    .std-ref}](Landing.html#schema-electricalconnectioninstance){.reference
    .internal}, and [[ElectricalConnectionInstanceW]{.std
    .std-ref}](Landing.html#schema-electricalconnectioninstancew){.reference
    .internal} entries stored in tables.

-   [[ContinuousProjection]{.std
    .std-ref}](Landing.html#schema-continuousprojection){.reference
    .internal} is exported similar to Projection with the
    [[ContinuousConnection]{.std
    .std-ref}](Landing.html#schema-continuousconnection){.reference
    .internal}, [[ContinuousConnectionInstance]{.std
    .std-ref}](Landing.html#schema-continuousconnectioninstance){.reference
    .internal}, and [[ContinuousConnectionInstanceW]{.std
    .std-ref}](Landing.html#schema-continuousconnectioninstancew){.reference
    .internal} entries stored in tables.

-   [[InputList]{.std
    .std-ref}](Landing.html#schema-inputlist){.reference .internal} is
    exported similar to Projection with the [[Input]{.std
    .std-ref}](Landing.html#schema-input){.reference .internal}, and
    [[InputW]{.std .std-ref}](Landing.html#schema-inputw){.reference
    .internal} entries stored in tables.

For more details, the source code of these export functions can be seen
[here in the libNeuroML
repository](https://github.com/NeuralEnsemble/libNeuroML/blob/2d8112178d8d82b07a20f8395ec22a23a6323a6c/neuroml/nml/helper_methods.py#L2548){.reference
.external} and [here in
org.neuroml.model](https://github.com/NeuroML/org.neuroml.model/blob/master/src/main/java/org/neuroml/model/util/hdf5/NeuroMLHDF5Writer.java){.reference
.external}.

HDF5 NeuroML files can be read and processed by [`jnml`{.docutils
.literal .notranslate}]{.pre} and [`pynml`{.docutils .literal
.notranslate}]{.pre} in the same way as XML files (see
[here](https://github.com/OpenSourceBrain/OpenCortex/tree/master/examples/HDF5){.reference
.external} for LEMS Simulation file examples which reference HDF5
NeuroML models).
:::

[]{#document-Userdocs/Provenance}

::: {#maintaining-provenance-in-neuroml-models .section .tex2jax_ignore .mathjax_ignore}
[]{#userdocs-provenance}

### Maintaining provenance in NeuroML models[\#](#maintaining-provenance-in-neuroml-models "Link to this heading"){.headerlink}

It is important to include metadata related to the "provenance" of model
elements in NeuroML models. This ensures that the sources of different
components of different models can be easily obtained/verified. For
example, all NeuroML models should ideally include information about the
original implementations, original research papers, and the original
creators for all model elements.

NeuroML supports the inclusion of such information in the
[[annotation]{.std .std-ref}](Landing.html#schema-annotation){.reference
.internal} model element. Annotations in NeuroML can include metadata
using the [Resource Description Framework
(RDF)](https://en.wikipedia.org/wiki/Resource_Description_Framework){.reference
.external}. The advantage of using RDF is that it makes the metadata
machine readable which enables automated analysis and parsing.

::: {#annotation-specification .section}
[]{#userdocs-provenance-spec}

#### Annotation specification[\#](#annotation-specification "Link to this heading"){.headerlink}

NeuroML follows the [COMBINE specification for
annotations](https://github.com/combine-org/combine-specifications/blob/main/specifications/qualifiers-1.1.md){.reference
.external}. This specification provides a set of well defined
relationships that can be used to connect model elements to various
metadata. These are included in the NeuroML schema as [[core NeuroML
component types]{.std
.std-ref}](Landing.html#schema-neuromlcorecomptypes){.reference
.internal}.

The suggested format for inclusion of the RDF metadata in annotations is
[Minimum information required in the annotation of models
(MIRIAM)](https://en.wikipedia.org/wiki/Minimum_information_required_in_the_annotation_of_models){.reference
.external}.
:::

::: {#creating-annotations .section}
[]{#userdocs-provenance-creating}

#### Creating annotations[\#](#creating-annotations "Link to this heading"){.headerlink}

[[pyNeuroML]{.std .std-ref}](Landing.html#pyneuroml){.reference
.internal} includes the
[create_annotation](https://pyneuroml.readthedocs.io/en/development/pyneuroml.annotations.html#pyneuroml.annotations.Annotation.create_annotation){.reference
.external} function for the addition of annotations to model elements.
Please see the API documentation for explanation on its usage. Note that
this functionality is not included by default when installing pyNeuroML.
One must install the annotation extra:

::: {.highlight-bash .notranslate}
::: highlight
    pip install pyneuroml[annotations]
:::
:::

An example of an annotation for the NeuroML conversion of Ray et al 2020
[\[[RAS20](Landing.html#id21 "Subhasis Ray, Zane N. Aldworth, and Mark A. Stopfer. Feedback inhibition and its control in an insect olfactory circuit. eLife, 9:e53281, 2020. URL: https://doi.org/10.7554/eLife.53281, doi:10.7554/eLife.53281."){.reference
.internal}\]]{#id1} is shown below:

::: {.highlight-python .notranslate}
::: highlight
        annotation = create_annotation(
            subject=cell.id, title="Giant GABAergic Neuron model",
            description="Subhasis Ray, Zane N Aldworth, Mark A Stopfer (2020) Feedback inhibition and its control in an insect olfactory circuit eLife 9:e53281.",
            annotation_style="miriam",
            serialization_format="pretty-xml",
            xml_header=False,
            citations={"https://doi.org/10.7554/eLife.53281": "Subhasis Ray, Zane N Aldworth, Mark A Stopfer (2020) Feedback inhibition and its control in an insect olfactory circuit eLife 9:e53281."},
            sources={"https://modeldb.science/262670": "ModelDB",
                     "https://github.com/OpenSourceBrain/262670": "GitHub",
                     "https://v1.opensourcebrain.org/projects/locust-mushroom-body": "Open Source Brain"},
            authors={"Subhasis Ray":
                     {"https://orcid.org/0000-0003-2566-7146": "orcid"},
                     },
            contributors={
                "Ankur Sinha": {"https://orcid.org/0000-0001-7568-7167": "orcid"},
            },
            creation_date="2024-04-25"
        )
        cell.annotation = neuroml.Annotation([annotation])
:::
:::

This generates the following RDF annotation in the MIRIAM format to be
included in the NeuroML cell file:

::: {.highlight-xml .notranslate}
::: highlight
        <annotation>
            <rdf:RDF
              xmlns:dc="http://purl.org/dc/elements/1.1/"
              xmlns:dcterms="http://purl.org/dc/terms/"
              xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#"
              xmlns:bqmodel="http://biomodels.net/model-qualifiers/"
            >
              <rdf:Description rdf:about="GGN">
                <dc:title>Giant GABAergic Neuron model</dc:title>
                <dc:description>Subhasis Ray, Zane N Aldworth, Mark A Stopfer (2020) Feedback inhibition and its control in an insect olfactory circuit eLife 9:e53281.</dc:description>
                <dc:source>
                  <rdf:Bag>
                    <rdf:_1 rdf:resource="https://modeldb.science/262670"/>
                  </rdf:Bag>
                </dc:source>
                <dc:source>
                  <rdf:Bag>
                    <rdf:_1 rdf:resource="https://github.com/OpenSourceBrain/262670"/>
                  </rdf:Bag>
                </dc:source>
                <dc:source>
                  <rdf:Bag>
                    <rdf:_1 rdf:resource="https://v1.opensourcebrain.org/projects/locust-mushroom-body"/>
                  </rdf:Bag>
                </dc:source>
                <bqmodel:isDescribedBy>
                  <rdf:Bag>
                    <rdf:_1 rdf:resource="https://doi.org/10.7554/eLife.53281"/>
                  </rdf:Bag>
                </bqmodel:isDescribedBy>
                <dc:creator>
                  <rdf:Bag>
                    <rdf:_1>Subhasis Ray</rdf:_1>
                    <rdf:_2>https://orcid.org/0000-0003-2566-7146</rdf:_2>
                  </rdf:Bag>
                </dc:creator>
                <dc:contributor>
                  <rdf:Bag>
                    <rdf:_1>Ankur Sinha</rdf:_1>
                    <rdf:_2>https://orcid.org/0000-0001-7568-7167</rdf:_2>
                  </rdf:Bag>
                </dc:contributor>
                <dcterms:created>
                  <rdf:Description>
                    <dcterms:W3CDTF>2024-04-25</dcterms:W3CDTF>
                  </rdf:Description>
                </dcterms:created>
              </rdf:Description>
            </rdf:RDF>

        </annotation>
:::
:::

pyNeuroML also includes functions to extract data from annotations.
:::
:::
:::
:::

[]{#document-Userdocs/ValidatingNeuroMLModels}

::: {#validating-neuroml-models .section .tex2jax_ignore .mathjax_ignore}
[]{#userdocs-validating-models}

## Validating NeuroML Models[\#](#validating-neuroml-models "Link to this heading"){.headerlink}

::: {.tip .admonition}
Validate NeuroML 2 files before using them.

It is good practice to validate NeuroML 2 files to check them for
correctness before using them.
:::

Models described in NeuroML must adhere to the NeuroML
[[specification]{.std
.std-ref}](Landing.html#userdocs-specification){.reference .internal}.
This allows all NeuroML models to be checked for correctness:
**validation**. There are a number of ways of **validating** NeuroML
model files.

::: {#using-the-command-line-tools .section}
[]{#userdocs-validating-models-cli}

### Using the command line tools[\#](#using-the-command-line-tools "Link to this heading"){.headerlink}

Both [`pynml`{.docutils .literal .notranslate}]{.pre} (provided by
[[pyNeuroML]{.std .std-ref}](Landing.html#pyneuroml){.reference
.internal}) and [`jnml`{.docutils .literal .notranslate}]{.pre}
(provided by [[jNeuroML]{.std
.std-ref}](Landing.html#jneuroml){.reference .internal}) can validate
individual NeuroML files:

::: {.highlight-console .notranslate}
::: highlight
    Usage:

    # For NeuroML 2
    jnml -validate <NML file(s)>
    pynml <NML file(s)> -validate

    # For NeuroML 1 (deprecated)
    jnml -validatev1 <NML file>
    pynml <NML file(s)> -validatev1
:::
:::
:::

::: {#using-the-python-api .section}
[]{#userdocs-validating-models-python}

### Using the Python API[\#](#using-the-python-api "Link to this heading"){.headerlink}

The [[pyNeuroML]{.std .std-ref}](Landing.html#pyneuroml){.reference
.internal} Python API provides a number of methods to validate NeuroML 2
files. The first is the aptly named [`validate_neuroml2`{.docutils
.literal .notranslate}]{.pre} function:

::: {.highlight-python .notranslate}
::: highlight
    from pyneuroml.pynml import validate_neuroml2

    ...

    validate_neuroml2(nml_filename)
:::
:::

Similarly, the [`validate_neuroml1`{.docutils .literal
.notranslate}]{.pre} function can be used to validate NeuroML v1 files.

If you are loading NeuroML files into your Python script, the
[`read_neuroml2_file`{.docutils .literal .notranslate}]{.pre} function
also includes validation:

::: {.highlight-python .notranslate}
::: highlight
    from pyneuroml.pynml import read_neuroml2_file


    ....


    read_neuroml2_file(nml_filename, include_includes=True, check_validity_pre_include=True)
:::
:::

This will read (load) the provided NeuroML 2 file and all the files that
are recursively included by it, and validate them all while it loads
them.
:::

::: {#list-of-validation-tests .section}
[]{#userdocs-validating-models-tests}

### List of validation tests[\#](#list-of-validation-tests "Link to this heading"){.headerlink}

These tests are made against the Schema document.

::: pst-scrollable-table-container
  Test                   Description
  ---------------------- ---------------------------------------------------------------------------------------------
  Check names            Check that names of all elements, attributes, parameters match those provided in the schema
  Check types            Check that the types of all included elements
  Check values           Check that values follow given restrictions
  Check inclusion        Check that required elements are included
  Check cardinality      Check the number of elements
  Check hierarchy        Check that child/children elements are included in the correct parent elements
  Check sequence order   Check that child/children elements are included in the correct order
:::

These are additional validation tests that are run on models (defined
[here](https://github.com/NeuroML/org.neuroml.model/blob/development/src/main/java/org/neuroml/model/util/NeuroML2Validator.java#L57){.reference
.external}):

::: pst-scrollable-table-container
  Test                                                                                                                                                                      Description
  ------------------------------------------------------------------------------------------------------------------------------------------------------------------------- ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
  Check top level ids                                                                                                                                                       Check that top level (root) elements have unique ids
  Check [[Network]{.std .std-ref}](Landing.html#schema-network){.reference .internal} level ids                                                                             Check that child/children of the [[Network]{.std .std-ref}](Landing.html#schema-network){.reference .internal} element have unique ids
  Check [[Cell]{.std .std-ref}](Landing.html#schema-cell){.reference .internal} [[Segment]{.std .std-ref}](Landing.html#schema-segment){.reference .internal} ids           Check that all [[Segment]{.std .std-ref}](Landing.html#schema-segment){.reference .internal}s in a [[Cell]{.std .std-ref}](Landing.html#schema-cell){.reference .internal} have unique ids
  Check single [[Segment]{.std .std-ref}](Landing.html#schema-segment){.reference .internal} without parent                                                                 Check that only one [[Segment]{.std .std-ref}](Landing.html#schema-segment){.reference .internal} is without parents (the soma [[Segment]{.std .std-ref}](Landing.html#schema-segment){.reference .internal})
  Check [[SegmentGroup]{.std .std-ref}](Landing.html#schema-segmentgroup){.reference .internal} ids                                                                         Check that all [[SegmentGroup]{.std .std-ref}](Landing.html#schema-segmentgroup){.reference .internal}s in a [[Cell]{.std .std-ref}](Landing.html#schema-cell){.reference .internal} have unique ids
  Check [[Member]{.std .std-ref}](Landing.html#schema-member){.reference .internal} segment ids exist                                                                       Check that [[Segment]{.std .std-ref}](Landing.html#schema-segment){.reference .internal}s referred to in [[SegmentGroup]{.std .std-ref}](Landing.html#schema-segmentgroup){.reference .internal} [[Member]{.std .std-ref}](Landing.html#schema-member){.reference .internal}s exist
  Check [[SegmentGroup]{.std .std-ref}](Landing.html#schema-segmentgroup){.reference .internal} definition                                                                  Check that [[SegmentGroup]{.std .std-ref}](Landing.html#schema-segmentgroup){.reference .internal}s being referenced are defined
  Check [[SegmentGroup]{.std .std-ref}](Landing.html#schema-segmentgroup){.reference .internal} definition order                                                            Check that [[SegmentGroup]{.std .std-ref}](Landing.html#schema-segmentgroup){.reference .internal}s are defined before being referenced
  Check included [[SegmentGroup]{.std .std-ref}](Landing.html#schema-segmentgroup){.reference .internal}s                                                                   Check that [[SegmentGroup]{.std .std-ref}](Landing.html#schema-segmentgroup){.reference .internal}s referenced by [[Include]{.std .std-ref}](Landing.html#schema-include){.reference .internal} elements of other [[SegmentGroup]{.std .std-ref}](Landing.html#schema-segmentgroup){.reference .internal}s exist
  Check [`numberInternalDivisions`{.docutils .literal .notranslate}]{.pre}                                                                                                  Check that [[SegmentGroup]{.std .std-ref}](Landing.html#schema-segmentgroup){.reference .internal}s define [`numberInternalDivisions`{.docutils .literal .notranslate}]{.pre} (used by simulators to discretize un-branched branches into compartments for simulation)
  Check included model files                                                                                                                                                Check that model files included by other files exist
  Check [[Population]{.std .std-ref}](Landing.html#schema-population){.reference .internal} component                                                                       Check that a component id provided to a [[Population]{.std .std-ref}](Landing.html#schema-population){.reference .internal} exists
  Check ion channel exists                                                                                                                                                  Check that an ion channel used to define a [[ChannelDensity]{.std .std-ref}](Landing.html#schema-channeldensity){.reference .internal} element exists
  Check concentration model species                                                                                                                                         Check that the species used in [[ConcentrationModel]{.std .std-ref}](Landing.html#schema-concentrationmodel){.reference .internal} elements are defined
  Check [[Population]{.std .std-ref}](Landing.html#schema-population){.reference .internal} size                                                                            Check that the [`size`{.docutils .literal .notranslate}]{.pre} attribute of a [[PopulationList]{.std .std-ref}](Landing.html#schema-populationlist){.reference .internal} matches the number of defined [[Instance]{.std .std-ref}](Landing.html#schema-instance){.reference .internal}s
  Check [[Projection]{.std .std-ref}](Landing.html#schema-projection){.reference .internal} component                                                                       Check that [[Population]{.std .std-ref}](Landing.html#schema-population){.reference .internal}s used in the [[Projection]{.std .std-ref}](Landing.html#schema-projection){.reference .internal} elements exist
  Check [[Connection]{.std .std-ref}](Landing.html#schema-connection){.reference .internal} [[Segment]{.std .std-ref}](Landing.html#schema-segment){.reference .internal}   Check that the [[Segment]{.std .std-ref}](Landing.html#schema-segment){.reference .internal} used in [[Connection]{.std .std-ref}](Landing.html#schema-connection){.reference .internal} elements exist
  Check [[Connection]{.std .std-ref}](Landing.html#schema-connection){.reference .internal} pre/post cells                                                                  Check that the pre- and post-synaptic cells used in [[Connection]{.std .std-ref}](Landing.html#schema-connection){.reference .internal} elements exist and are correctly specified
  Check [[Synapse]{.std .std-ref}](Landing.html#schema-synapses){.reference .internal}                                                                                      Check that the [[Synapse]{.std .std-ref}](Landing.html#schema-synapses){.reference .internal} component used in a [[Projection]{.std .std-ref}](Landing.html#schema-projection){.reference .internal} element exists
  Check root id                                                                                                                                                             Check that the root [[Segment]{.std .std-ref}](Landing.html#schema-segment){.reference .internal} in a [[Cell]{.std .std-ref}](Landing.html#schema-cell){.reference .internal} morphology has id (0)
:::
:::
:::

[]{#document-Userdocs/VisualisingNeuroMLModels}

::: {#visualising-neuroml-models .section .tex2jax_ignore .mathjax_ignore}
[]{#userdocs-visualising-models}

## Visualising NeuroML Models[\#](#visualising-neuroml-models "Link to this heading"){.headerlink}

A number of the [[NeuroML software tools]{.std
.std-ref}](Landing.html#userdocs-software){.reference .internal} can be
used to easily visualise models described in NeuroML.

::: {#get-a-quick-summary-of-your-model .section}
[]{#userdoc-visualising-models-summary}

### Get a quick summary of your model[\#](#get-a-quick-summary-of-your-model "Link to this heading"){.headerlink}

::: {#using-command-line-tools .section}
[]{#userdoc-visualising-models-summary-cli}

#### Using command line tools[\#](#using-command-line-tools "Link to this heading"){.headerlink}

You can get a quick summary of your NeuroML model using the
[`pynml-summary`{.docutils .literal .notranslate}]{.pre} command line
tool that is provided by [[pyNeuroML]{.std
.std-ref}](Landing.html#pyneuroml){.reference .internal}:

::: {.highlight-console .notranslate}
::: highlight
    Usage:
    pynml-summary <NeuroML file>
:::
:::

For example, to get a quick summary of the [Primary Auditory Cortex
model by Dave
Beeman](https://github.com/OpenSourceBrain/ACnet2/blob/master/neuroConstruct/generatedNeuroML2/MediumNet.net.nml){.reference
.external} (see it [here on Open Source
Brain](https://www.opensourcebrain.org/projects/acnet2){.reference
.external}), one can run:

::: {.highlight-console .notranslate}
::: highlight
    pynml-summary MediumNet.net.nml

    *******************************************************
    * NeuroMLDocument: network_ACnet2
    *
    *  PulseGenerator: ['BackgroundRandomIClamps']
    *
    *  Network: network_ACnet2 (temperature: 6.3 degC)
    *
    *   60 cells in 2 populations
    *     Population: baskets_12 with 12 components of type bask
    *       Locations: [(372.5585, 75.3425, 459.2106), ...]
    *       Properties: color=0.0 0.19921875 0.59765625;
    *     Population: pyramidals_48 with 48 components of type pyr_4_sym
    *       Locations: [(64.2564, 0.6838, 94.8305), ...]
    *       Properties: color=0.796875 0.0 0.0;
    *
    *   984 connections in 4 projections
    *     Projection: SmallNet_bask_bask from baskets_12 to baskets_12, synapse: GABA_syn_inh
    *       60 connections: [(Connection 0: 3:0(0.41661) -> 0:0(0.68577)), ...]
    *     Projection: SmallNet_bask_pyr from baskets_12 to pyramidals_48, synapse: GABA_syn
    *       336 connections: [(Connection 0: 10:0(0.05824) -> 0:6(0.02628)), ...]
    *     Projection: SmallNet_pyr_bask from pyramidals_48 to baskets_12, synapse: AMPA_syn_inh
    *       252 connections: [(Connection 0: 1:0(0.89734) -> 0:1(0.09495)), ...]
    *     Projection: SmallNet_pyr_pyr from pyramidals_48 to pyramidals_48, synapse: AMPA_syn
    *       336 connections: [(Connection 0: 14:0(0.52814) -> 0:3(0.10797)), ...]
    *
    *   14 inputs in 1 input lists
    *     Input list: BackgroundRandomIClamps to pyramidals_48, component BackgroundRandomIClamps
    *       14 inputs: [(Input 0: 37:0(0.500000)), ...]
    *
    *******************************************************
:::
:::
:::

::: {#using-pyneuroml .section}
[]{#userdoc-visualising-models-summary-pyneuroml}

#### Using pyNeuroML[\#](#using-pyneuroml "Link to this heading"){.headerlink}

You can also get a summary of your model from within your
[[pyNeuroML]{.std .std-ref}](Landing.html#pyneuroml){.reference
.internal} script itself using the [`summary`{.docutils .literal
.notranslate}]{.pre} function:

::: {.highlight-python .notranslate}
::: highlight
    import pyneuroml.pynml

    ...


    pyneuroml.pynml.summary(nml2_doc)
:::
:::
:::
:::

::: {#view-the-3d-structure-of-your-model .section}
[]{#userdoc-visualising-models-png}

### View the 3D structure of your model[\#](#view-the-3d-structure-of-your-model "Link to this heading"){.headerlink}

::: {#userdoc-visualising-models-png-cli .section}
[]{#id1}

#### Using command line tools[\#](#userdoc-visualising-models-png-cli "Link to this heading"){.headerlink}

You can generate an image of the 3D structure of the NeuroML model using
the [`pynml`{.docutils .literal .notranslate}]{.pre} command provided by
[[pyNeuroML]{.std .std-ref}](Landing.html#pyneuroml){.reference
.internal}, or using the [`jnml`{.docutils .literal .notranslate}]{.pre}
command provided by [[jNeuroML]{.std
.std-ref}](Landing.html#jneuroml){.reference .internal}:

::: {.highlight-console .notranslate}
::: highlight
    Usage:
    pynml -png/-svg <NeuroML file>
    jnml -png/-svg <NeuroML file>
:::
:::

For example, to generate a PNG image of the Auditory Cortex model used
above, we can use (use [`-svg`{.docutils .literal .notranslate}]{.pre}
to generate a vectorised SVG image instead of a PNG):

::: {.highlight-console .notranslate}
::: highlight
    pynml -png MediumNet.net.nml
:::
:::

This generates the following image showing different views of the
network :

<figure id="id7" class="align-center">
<a href="_images/Acnet-medium.net.png"
class="reference internal image-reference"><img
src="_images/Acnet-medium.net.png"
style="width: 325.0px; height: 567.4px;"
alt="Graphical view of the Auditory Cortex model generated with pynml" /></a>
<figcaption><p><span class="caption-number">Fig. 31 </span><span
class="caption-text">Graphical view of the Auditory Cortex model
generated with pynml</span><a href="#id7" class="headerlink"
title="Link to this image">#</a></p></figcaption>
</figure>

An visualiser is also included in pyneuroml as
[`pynml-plotmorph`{.docutils .literal .notranslate}]{.pre} which
includes both 2D and 3D views:

::: {.highlight-console .notranslate}
::: highlight
    Usage:
    pynml-plotmorph <NeuroML file>
    pynml-plotmorph -i <NeuroML file>
:::
:::

<figure id="id8" class="align-center">
<a href="_images/20231122-ACNet.png"
class="reference internal image-reference"><img
src="_images/20231122-ACNet.png" style="width: 30%;"
alt="A network visualised with `pynml-plotmorph`" /></a>
<figcaption><p><span class="caption-number">Fig. 32 </span><span
class="caption-text">Matplotlib based 2D visualisation of a network with
<span class="pre"><code
class="docutils literal notranslate">pynml-plotmorph</code></span>.</span><a
href="#id8" class="headerlink"
title="Link to this image">#</a></p></figcaption>
</figure>

Your browser does not support the video tag

\
*Example network visualised interactively using \`pynml-plotmorph\`*\
\

You can also generate graphical representations that can be viewed with
the [Persistence of Vision Raytracer
(POV-Ray)](http://povray.org/){.reference .external} tool using the
[`pynml-povray`{.docutils .literal .notranslate}]{.pre} tool. For
example:

::: {.highlight-console .notranslate}
::: highlight
    pynml-povray MediumNet.net.nml -scalez 8
    povray Antialias=On Antialias_Depth=10 Antialias_Threshold=0.1 Output_to_File=y Output_File_Type=N Output_File_Name=Acnet-medium.povray +W1200 +H900 MediumNet.net.nml.pov
:::
:::

generates this image:

<figure id="id9" class="align-center">
<a href="_images/Acnet-medium.povray.png"
class="reference internal image-reference"><img
src="_images/Acnet-medium.povray.png"
style="width: 600.0px; height: 450.0px;"
alt="Graphical view of the Auditory Cortex model generated with pynml-povray and POV-Ray" /></a>
<figcaption><p><span class="caption-number">Fig. 33 </span><span
class="caption-text">Graphical view of the Auditory Cortex model
generated with pynml-povray and POV-Ray</span><a href="#id9"
class="headerlink" title="Link to this image">#</a></p></figcaption>
</figure>

You can also use POV-Ray interactively. Please refer to the [official
website](http://povray.org/download/){.reference .external} for more
information on installing and using POV-Ray. On Fedora Linux systems,
you can install it from the Fedora repositories using [`dnf`{.docutils
.literal .notranslate}]{.pre}:

::: {.highlight-console .notranslate}
::: highlight
    sudo dnf install povray
:::
:::
:::

::: {#userdoc-visualising-models-png-pyneuroml .section}
[]{#id2}

#### Using pyNeuroML[\#](#userdoc-visualising-models-png-pyneuroml "Link to this heading"){.headerlink}

These functions are also exposed as Python functions in
[[pyNeuroML]{.std .std-ref}](Landing.html#pyneuroml){.reference
.internal}, so that you can use them directly in Python scripts:

::: {.highlight-python .notranslate}
::: highlight
    import pyneuroml.pynml

    pyneuroml.pynml.nml2_to_png(nml2_doc)
    pyneuroml.pynml.nml2_to_svg(nml2_doc)


    from pyneuroml.plot.PlotMorphology import plot_2D
    from pyneuroml.plot.PlotMorphologyVispy import plot_interactive_3D

    plot_2D(nml2_doc)
    plot_interactive_3D(nml2_doc)
:::
:::

::: {.tip .admonition}
Open Source Brain uses NeuroML.

The [Open Source Brain
platform](https://www.opensourcebrain.org){.reference .external}
generates the interactive visualisations from NeuroML sources. See the
Auditory Cortex model on Open Source Brain
[here](https://www.opensourcebrain.org/projects/acnet2){.reference
.external}.
:::
:::
:::

::: {#view-the-connectivity-graph-of-your-model .section}
[]{#userdoc-visualising-models-graph}

### View the connectivity graph of your model[\#](#view-the-connectivity-graph-of-your-model "Link to this heading"){.headerlink}

::: {#userdoc-visualising-models-graph-cli .section}
[]{#id3}

#### Using command line tools[\#](#userdoc-visualising-models-graph-cli "Link to this heading"){.headerlink}

::: {.tip .dropdown .admonition}
Use levels to generate connectivity graphs with different levels of
detail.

Positive values for levels will generate figures at the population
level, while negative values will generate them at the level of cells.
:::

You can generate an image of the 3D structure of the NeuroML model using
[`pynml`{.docutils .literal .notranslate}]{.pre}:

::: {.highlight-console .notranslate}
::: highlight
    Usage:
    pynml <NeuroML file> -graph <level, engine>
:::
:::

For example, to generate a PNG image of the Auditory Cortex model used
above, we can use:

::: {.highlight-console .notranslate}
::: highlight
    pynml MediumNet.net.nml -graph 1d
:::
:::

This generates the following image showing different views of the
network :

<figure id="id10" class="align-center">
<img src="_images/Acnet-medium-graph-level1.png"
alt="Level 1 network graph generated by pynml" />
<figcaption><p><span class="caption-number">Fig. 34 </span><span
class="caption-text">Level 1 network graph generated by pynml</span><a
href="#id10" class="headerlink"
title="Link to this image">#</a></p></figcaption>
</figure>

You can modify the level of detail included in the graph by using
different values of levels. For example, this command generates a level
5 graph:

::: {.highlight-console .notranslate}
::: highlight
    pynml MediumNet.net.nml -graph 5d
:::
:::

<figure id="id11" class="align-center">
<a href="_images/Acnet-medium-graph-level5.png"
class="reference internal image-reference"><img
src="_images/Acnet-medium-graph-level5.png"
style="width: 345.0px; height: 372.59999999999997px;"
alt="Level 5 network graph generated by pynml" /></a>
<figcaption><p><span class="caption-number">Fig. 35 </span><span
class="caption-text">Level 5 network graph generated by pynml</span><a
href="#id11" class="headerlink"
title="Link to this image">#</a></p></figcaption>
</figure>
:::

::: {#userdoc-visualising-models-graph-pyneuroml .section}
[]{#id4}

#### Using pyNeuroML[\#](#userdoc-visualising-models-graph-pyneuroml "Link to this heading"){.headerlink}

You can also generated these figures from within your [[pyNeuroML]{.std
.std-ref}](Landing.html#pyneuroml){.reference .internal} script itself
using the [`generate_nmlgraph`{.docutils .literal .notranslate}]{.pre}
function:

::: {.highlight-python .notranslate}
::: highlight
    import pyneuroml.pynml

    ...


    pyneuroml.pynml.generate_nmlgraph(nml2_doc, level="1", engine="dot")
:::
:::
:::
:::

::: {#view-the-connectivity-matrices-of-the-model .section}
[]{#userdoc-visualising-models-matrix}

### View the connectivity matrices of the model[\#](#view-the-connectivity-matrices-of-the-model "Link to this heading"){.headerlink}

You can generate the connectivity matrices of projections between
neuronal populations of the NeuroML model using [`pynml`{.docutils
.literal .notranslate}]{.pre}:

::: {.highlight-console .notranslate}
::: highlight
    Usage:
    pynml <NeuroML file> -matrix <level>
:::
:::

For example, to generate a PNG image of the connectivity matrices in the
Auditory Cortex model used above, we can use:

::: {.highlight-console .notranslate}
::: highlight
    pynml MediumNet.net.nml -matrix 1
:::
:::

This generates the following images showing different views of the
connectivity matrices in the network :

::: container-fluid
::: {.row .my-2 .py-2}
::: {.col-sm-6 .px-2}
[![Connectivity matrix generated by pynml - sum of signed
weights](_images/Acnet-matrix-1.png){style="width: 320.0px; height: 240.0px;"}](_images/Acnet-matrix-1.png){.reference
.internal .image-reference}
:::

::: {.col-sm-6 .px-2}
[![Connectivity matrix generated by pynml - average magnitude of
conductance received by post-synaptic
neurons](_images/Acnet-matrix-2.png){style="width: 320.0px; height: 240.0px;"}](_images/Acnet-matrix-2.png){.reference
.internal .image-reference}
:::

::: {.col-sm-6 .px-2}
[![Connectivity matrix generated by pynml - number of
connections](_images/Acnet-matrix-3.png){style="width: 320.0px; height: 240.0px;"}](_images/Acnet-matrix-3.png){.reference
.internal .image-reference}
:::

::: {.col-sm-6 .px-2}
[![Connectivity matrix generated by pynml - average conductance received
by post-synaptic
neurons](_images/Acnet-matrix-4.png){style="width: 320.0px; height: 240.0px;"}](_images/Acnet-matrix-4.png){.reference
.internal .image-reference}
:::

::: {.col-sm-6 .px-2}
[![Connectivity matrix generated by pynml - average weight per
connection](_images/Acnet-matrix-5.png){style="width: 320.0px; height: 240.0px;"}](_images/Acnet-matrix-5.png){.reference
.internal .image-reference}
:::
:::
:::
:::

::: {#view-graph-of-the-simulation-instance-of-the-model .section}
[]{#userdoc-visualising-models-lemsgraph}

### View graph of the simulation instance of the model[\#](#view-graph-of-the-simulation-instance-of-the-model "Link to this heading"){.headerlink}

::: {#userdoc-visualising-models-lemsgraph-cli .section}
[]{#id5}

#### Using command line tools[\#](#userdoc-visualising-models-lemsgraph-cli "Link to this heading"){.headerlink}

When you have created a simulation instance of the NeuroML model using
LEMS, you can also visualise this using [`pynml`{.docutils .literal
.notranslate}]{.pre} or [`jnml`{.docutils .literal .notranslate}]{.pre}:

::: {.highlight-console .notranslate}
::: highlight
    Usage:
    pynml <LEMS simulation file> -lems-graph
    jnml <LEMS simulation file> -lems-graph
:::
:::

For example, to generate the LEMS graph for the [[Izhikevich neuron
network example]{.std
.std-ref}](Landing.html#userdocs-gettingstarted-izhikevichnetwork){.reference
.internal}, we will use:

::: {.highlight-console .notranslate}
::: highlight
    jnml LEMS_example_izhikevich2007network_sim.xml -lems-graph
:::
:::

will generate:

<figure id="id12" class="align-center">
<a href="_images/LEMS_example_izhikevich2007network_sim.png"
class="reference internal image-reference"><img
src="_images/LEMS_example_izhikevich2007network_sim.png"
style="width: 365.5px; height: 760.0px;"
alt="Model summary graph generated using jnml." /></a>
<figcaption><p><span class="caption-number">Fig. 36 </span><span
class="caption-text">A summary graph of the model generated using
jnml.</span><a href="#id12" class="headerlink"
title="Link to this image">#</a></p></figcaption>
</figure>

Note that the [`-lems-graph`{.docutils .literal .notranslate}]{.pre}
option does not take options for levels of detail. It shows all the
details of the simulation instance, and so is better suited for simpler
models rather than detailed conductance based network models. For
example, for the Auditory Cortex model, [[`this`{.xref .download
.docutils .literal
.notranslate}]{.pre}](_downloads/54d9372f3c36e040552bd9cd5af44e43/Acnet-LEMS.png){.reference
.download .internal download=""} very very detailed image is generated
(please click to open: it is too large to display in the page).
:::

::: {#userdoc-visualising-models-lemsgraph-pyneuroml .section}
[]{#id6}

#### Using pyNeuroML[\#](#userdoc-visualising-models-lemsgraph-pyneuroml "Link to this heading"){.headerlink}

You can also generated these figures from within your [[pyNeuroML]{.std
.std-ref}](Landing.html#pyneuroml){.reference .internal} script itself
using the [`generate_lemsgraph`{.docutils .literal .notranslate}]{.pre}
function:

::: {.highlight-python .notranslate}
::: highlight
    import pyneuroml.pynml

    ...

    pyneuroml.pynml.generate_lemsgraph(lems_file)
:::
:::
:::
:::

::: {#viewing-analysing-ion-channel-dynamics .section}
### Viewing/analysing ion channel dynamics[\#](#viewing-analysing-ion-channel-dynamics "Link to this heading"){.headerlink}

There is a dedicated section on [[visualising and analysing ion channel
models]{.std
.std-ref}](Landing.html#userdocs-visualising-channels){.reference
.internal}.
:::

::: {.toctree-wrapper .compound}
[]{#document-Userdocs/VisualisingChannels}

::: {#visualising-and-analysing-ion-channel-models .section .tex2jax_ignore .mathjax_ignore}
[]{#userdocs-visualising-channels}

### Visualising and analysing ion channel models[\#](#visualising-and-analysing-ion-channel-models "Link to this heading"){.headerlink}

A core part of NeuroML is the ability to specify voltage dependent (and
potentially concentration dependent) membrane conductances, which are
due to ion channels.

::: {#help-converting-examining-channels-in-neuroml .section}
#### Help converting/examining channels in NeuroML[\#](#help-converting-examining-channels-in-neuroml "Link to this heading"){.headerlink}

Converting your own ion channel models to NeuroML is facilitated by
examples (e.g. [a simple HH Na+
channel](https://github.com/NeuroML/NeuroML2/tree/master/examples/NML2_SimpleIonChannel.nml){.reference
.external}) and the specification documentation (e.g. for
[[\<ionChannelHH\>]{.std
.std-ref}](Landing.html#schema-ionchannelhh){.reference .internal},
[[\<gateHHrates\>]{.std
.std-ref}](Landing.html#schema-gatehhrates){.reference .internal},
[[\<HHExpLinearRate\>]{.std
.std-ref}](Landing.html#schema-hhexplinearrate){.reference .internal},
but there are also a number of software tools which can be used to view
the internal properties of the ion channels, as well as their behaviour.

::: {.tip .admonition}
Converting cell models to NeuroML

Note: there is a full guide to [[Converting cell models to NeuroML and
sharing them on Open Source Brain]{.std
.std-ref}](Landing.html#userdocs-creating-models-converting-conductance){.reference
.internal} which uses some of the tools and methods below.
:::

::: {#use-jnml-info-note-not-in-pynml-yet .section}
##### 1) Use jnml -info (note not in pynml yet...)[\#](#use-jnml-info-note-not-in-pynml-yet "Link to this heading"){.headerlink}

[[jNeuroML]{.std .std-ref}](Landing.html#jneuroml){.reference .internal}
can be used on channel files for a quick summary of the contents.

::: {.highlight-default .notranslate}
::: highlight
    > jnml NaConductance.channel.nml -info

     jNeuroML v0.12.0

    Information on contents of NeuroML 2 file
    Ion Channel NaConductance:
        ID: NaConductance
        Description: HH Na Channel
        Gates:
            gate m:
                instances: 3
                forward rate: 1e3 * (v - (-0.04))/0.01 / ( 1 - exp(-(v - (-0.04)) / 0.01))
                reverse rate: 4e3 * exp((v - (-0.065))/-0.018)
            gate h:
                instances: 1
                forward rate: 70 * exp((v - (-0.065))/-0.02)
                reverse rate: 1e3 /(1 + exp((v - (-0.035))/0.01))
:::
:::
:::

::: {#use-pynml-utilities-in-pyneuroml .section}
[]{#userdoc-pynml-channelanalysis}

##### 2) Use pynml utilities in pyNeuroML[\#](#use-pynml-utilities-in-pyneuroml "Link to this heading"){.headerlink}

[[pyNeuroML]{.std .std-ref}](Landing.html#pyneuroml){.reference
.internal} comes with 3 utilities which help enable examination of the
properties of ion channels, both based on NeuroML and NEURON mod files.

::: {.highlight-default .notranslate}
::: highlight
    pynml-channelanalysis NaConductance.channel.nml   # Analyse a NeuroML 2 channel
    pynml-plotchan cell.nml   # Plot distribution of peak channel conductances over cell morphology
    pynml-modchananalysis NaConductance               # Analyse a NEURON channel e.g. from NaConductance.mod
:::
:::

![pynml_channelanalyse](_images/pynml-channelanalysis.png)

[`pynml-channelanalysis`{.docutils .literal .notranslate}]{.pre}
includes a number of options for generating graphs of channel activity
under different conditions (see
[here](https://github.com/OpenSourceBrain/BlueBrainProjectShowcase/tree/master/NMC/NeuroML2#analyse-channel-properties){.reference
.external} for details).

::: {.highlight-default .notranslate}
::: highlight
    pynml-channelanalysis NaTa_t.channel.nml  -erev 55 -stepTargetVoltage 10 -clampDuration 5 -i -duration 7 -clampDelay 1
:::
:::

![pynml_analyse_Na](_images/NaTa.png)

[`pynml-plotchan`{.docutils .literal .notranslate}]{.pre} will plot the
distribution of peak conductances of all channels in a cell over its
morphology, and also show distribution as a function of distance from
the soma. For example, the figure below shows the distribution of the Ih
channel in the layer 5 pyramidal neuron model on the neuronal morphology
on the left, and the value at different distances from the soma on the
right.

![pynml_analyse_plotchan](_images/Ih-combined.png)

This functionality is also available as a [Python
function](https://pyneuroml.readthedocs.io/en/stable/pyneuroml.analysis.html#pyneuroml.analysis.ChannelDensityPlot.plot_channel_densities){.reference
.external} for use in scripts.
:::

::: {#load-cell-model-on-to-osbv1-analyse-the-channels .section}
##### 4) Load cell model on to OSBv1 & analyse the channels[\#](#load-cell-model-on-to-osbv1-analyse-the-channels "Link to this heading"){.headerlink}

Open Source Brain (version 1) includes channel analysis functionalities.

![osb_channel_analysis](_images/osb-channel-analysis.png)
:::

::: {#export-to-one-of-the-supported-simulators .section}
##### 5) Export to one of the supported simulators[\#](#export-to-one-of-the-supported-simulators "Link to this heading"){.headerlink}

Exporting to Neuron say ([`jnml`{.docutils .literal
.notranslate}]{.pre}` `{.docutils .literal
.notranslate}[`LEMS_NML2_Ex5_DetCell.xml`{.docutils .literal
.notranslate}]{.pre}` `{.docutils .literal
.notranslate}[`-neuron`{.docutils .literal .notranslate}]{.pre}) will
produce mod files with the "flattened" equations:

::: {.highlight-default .notranslate}
::: highlight
    ...
    DERIVATIVE states {
        rates()
        m_q' = rate_m_q
        h_q' = rate_h_q

    }

    PROCEDURE rates() {

        m_forwardRate_x = (v -  m_forwardRate_midpoint ) /  m_forwardRate_scale ? evaluable
        if (m_forwardRate_x  != 0)  {
            m_forwardRate_r = m_forwardRate_rate  *  m_forwardRate_x  / (1 - exp(0 -  m_forwardRate_x )) ? evaluable cdv
        } else if (m_forwardRate_x  == 0)  {
            m_forwardRate_r = m_forwardRate_rate ? evaluable cdv
        }
    ...
:::
:::

Exporting to Brian 2 ([`jnml`{.docutils .literal
.notranslate}]{.pre}` `{.docutils .literal
.notranslate}[`LEMS_NML2_Ex5_DetCell.xml`{.docutils .literal
.notranslate}]{.pre}` `{.docutils .literal
.notranslate}[`-brian2`{.docutils .literal .notranslate}]{.pre}) will
also produce a large file with the explicit expressions...

::: {.highlight-default .notranslate}
::: highlight
    ...
    hhcell_eqs=Equations('''
        dbioPhys1_membraneProperties_NaConductances_NaConductance_m_q/dt = ((bioPhys1_membraneProperties_NaConductances_NaConductance_m_inf - bioPhys1_membraneProperties_NaConductances_NaConductance_m_q) / bioPhys1_membraneProperties_NaConductances_NaConductance_m_tau) :  1
        dbioPhys1_membraneProperties_NaConductances_NaConductance_h_q/dt = ((bioPhys1_membraneProperties_NaConductances_NaConductance_h_inf - bioPhys1_membraneProperties_NaConductances_NaConductance_h_q) / bioPhys1_membraneProperties_NaConductances_NaConductance_h_tau) :  1
        dbioPhys1_membraneProperties_KConductances_KConductance_n_q/dt = ((bioPhys1_membraneProperties_KConductances_KConductance_n_inf - bioPhys1_membraneProperties_KConductances_KConductance_n_q) / bioPhys1_membraneProperties_KConductances_KConductance_n_tau) :  1
        dv/dt = ((iChannels + iSyn) / totCap) :  volt
        morph1_0_LEN = 1.0 * meter : meter
    ...
        bioPhys1_membraneProperties_KConductances_erev = -0.077 * volt : volt
        bioPhys1_membraneProperties_KConductances_condDensity = 360.0 * kilogram**-1 * meter**-4 * second**3 * amp**2 : kilogram**-1 * meter**-4 * second**3 * amp**2
        bioPhys1_membraneProperties_KConductances_KConductance_conductance = 1.0E-11 * siemens : siemens
        bioPhys1_membraneProperties_KConductances_KConductance_n_instances = 4.0: 1
        bioPhys1_membraneProperties_KConductances_KConductance_n_forwardRate_rate = 100.0 * second**-1 : second**-1
        bioPhys1_membraneProperties_KConductances_KConductance_n_forwardRate_midpoint = -0.055 * volt : volt
        bioPhys1_membraneProperties_KConductances_KConductance_n_forwardRate_scale = 0.01 * volt : volt
        bioPhys1_membraneProperties_KConductances_KConductance_n_reverseRate_rate = 125.0 * second**-1 : second**-1
        bioPhys1_membraneProperties_KConductances_KConductance_n_reverseRate_midpoint = -0.065 * volt : volt
        bioPhys1_membraneProperties_KConductances_KConductance_n_reverseRate_scale = -0.08 * volt : volt
:::
:::

Both very verbose, but it's possible to see at least what explicit
expressions are being used for the channels...
:::

::: {#use-neuroml-db .section}
##### 6) Use NeuroML-DB[\#](#use-neuroml-db "Link to this heading"){.headerlink}

[NeuroML-DB](https://neuroml-db.org){.reference .external} also provides
analysis features for Ion channels.

![neuromldb-channel-analysis](_images/neuromldb-channel-analysis.png)
:::
:::
:::

[]{#document-Userdocs/VisualisingCells}

::: {#visualising-and-analysing-cell-models .section .tex2jax_ignore .mathjax_ignore}
[]{#userdocs-visualising-cells}

### Visualising and analysing cell models[\#](#visualising-and-analysing-cell-models "Link to this heading"){.headerlink}

The NeuroML ecosystem include a number of utilities for analysis and
visualisation of cells. Cell morphologies can either be visualised
programmatically using the core tools, or using the many advanced
neuroinformatics tools in the ecosystem that support NeuroML. In
addition to the resources listed below, you can also use the
visualisation features of any other tools that read NeuroML. E.g.,
[[NetPyNE and NetPyNE-UI]{.std
.std-ref}](Landing.html#userdocs-supporting-apps-netpyne){.reference
.internal}, [[neuroConstruct]{.std
.std-ref}](Landing.html#userdocs-supporting-apps-neuroconstruct){.reference
.internal}, [[Arbor]{.std
.std-ref}](Landing.html#userdocs-supporting-apps-arbor){.reference
.internal} and others.

::: {#visualising-morphology-of-multi-compartmental-cell-models .section}
[]{#userdocs-visualising-cells-morph}

#### Visualising morphology of multi-compartmental cell models[\#](#visualising-morphology-of-multi-compartmental-cell-models "Link to this heading"){.headerlink}

Multi-compartmental cells can be visualised using the
[plot_2D](https://pyneuroml.readthedocs.io/en/development/pyneuroml.plot.html#pyneuroml.plot.PlotMorphology.plot_2D){.reference
.external} and
[plot_interactive_3D](https://pyneuroml.readthedocs.io/en/development/pyneuroml.plot.html#pyneuroml.plot.PlotMorphology.plot_interactive_3D){.reference
.external} methods included in [[pyNeuroML]{.std
.std-ref}](Landing.html#pyneuroml){.reference .internal}. This
functionality is also exposed via the [`pynml-plotmorph`{.docutils
.literal .notranslate}]{.pre} command line tool.

<figure id="id1" class="align-center">
<a href="_images/test_morphology_plot_2d_Cell_497232312_cell_nml_xy.png"
class="reference internal image-reference"><img
src="_images/test_morphology_plot_2d_Cell_497232312_cell_nml_xy.png"
style="width: 50%;"
alt="Morphology of example cell plotted with `plot_2D`" /></a>
<figcaption><p><span class="caption-number">Fig. 37 </span><span
class="caption-text">Morphology of example cell plotted with <span
class="pre"><code
class="docutils literal notranslate">plot_2D</code></span> in the X-Y
plane.</span><a href="#id1" class="headerlink"
title="Link to this image">#</a></p></figcaption>
</figure>

Your browser does not support the video tag

\
*Morphology of example cell visualised interactively using
\`plot_interactive_3D\`*

::: {#visualising-morphology-of-multi-compartmental-cell-models-in-neuroml-db .section}
[]{#userdocs-visualising-cells-morph-nmldb}

##### Visualising morphology of multi-compartmental cell models in NeuroML-db[\#](#visualising-morphology-of-multi-compartmental-cell-models-in-neuroml-db "Link to this heading"){.headerlink}

The [[NeuroML-DB]{.std
.std-ref}](Landing.html#userdocs-finding-models-neuromldb){.reference
.internal} platform shows detailed cell morphologies of all cells
included in its database.

<figure id="id2" class="align-center">
<a href="_images/nml-db-morphology.png"
class="reference internal image-reference"><img
src="_images/nml-db-morphology.png" style="width: 70%;"
alt="Morphology of cell shown in NeuroML-DB." /></a>
<figcaption><p><span class="caption-number">Fig. 38 </span><span
class="caption-text">Visualisation of morphology of an example cell on
NeuroML-DB.</span><a href="#id2" class="headerlink"
title="Link to this image">#</a></p></figcaption>
</figure>
:::

::: {#visualising-morphology-of-multi-compartmental-cell-models-in-open-source-brain .section}
[]{#userdocs-visualising-cells-morph-osb}

##### Visualising morphology of multi-compartmental cell models in Open Source Brain[\#](#visualising-morphology-of-multi-compartmental-cell-models-in-open-source-brain "Link to this heading"){.headerlink}

The [[Open Source Brain]{.std
.std-ref}](Landing.html#userdocs-finding-models-osbv1){.reference
.internal} platform also provides advanced visualisation capabilities
that can be used to visualise the morphologies of NeuroML cells.

<figure id="id3" class="align-center">
<a href="_images/osb-morphology.png"
class="reference internal image-reference"><img
src="_images/osb-morphology.png" style="width: 70%;"
alt="Morphology of cell shown on Open Source Brain." /></a>
<figcaption><p><span class="caption-number">Fig. 39 </span><span
class="caption-text">Interactive visualisation of morphology of an
example cell on Open Source Brain.</span><a href="#id3"
class="headerlink" title="Link to this image">#</a></p></figcaption>
</figure>
:::
:::

::: {#analysing-cell-electrophysiology .section}
[]{#userdocs-visualising-cells-ephys}

#### Analysing cell electrophysiology[\#](#analysing-cell-electrophysiology "Link to this heading"){.headerlink}

The core tools also include utilities to aid in the analysis of cell
electrophysiology. [[pyNeuroML]{.std
.std-ref}](Landing.html#pyneuroml){.reference .internal} includes the
[generate_current_vs_frequency_curve](https://pyneuroml.readthedocs.io/en/development/pyneuroml.analysis.html#pyneuroml.analysis.generate_current_vs_frequency_curve){.reference
.external} utility function that can be used to generate
current-frequency, current-sub-threshold voltage, and to plot voltage
traces generated at the soma for different current injections. For
example, we can analyse the [[OLM cell from our tutorial]{.std
.std-ref}](Landing.html#userdocs-getting-started-multi-compartment-example){.reference
.internal}:

::: {.highlight-default .notranslate}
::: highlight
    generate_current_vs_frequency_curve("source/Userdocs/NML2_examples/olm.cell.nml", "olm", simulator="jNeuroML_NEURON", plot_iv=True, plot_if=True, plot_voltage_traces=True)
:::
:::

This will generate these figures:

<figure id="id4" class="align-center">
<a href="_images/olm-cell-fi.png"
class="reference internal image-reference"><img
src="_images/olm-cell-fi.png" style="width: 70%;"
alt="F-I curve for OLM cell generated using `generate_current_vs_frequency_curve`." /></a>
<figcaption><p><span class="caption-number">Fig. 40 </span><span
class="caption-text">F-I curve for OLM cell generated using <span
class="pre"><code
class="docutils literal notranslate">generate_current_vs_frequency_curve</code></span>.</span><a
href="#id4" class="headerlink"
title="Link to this image">#</a></p></figcaption>
</figure>

<figure id="id5" class="align-center">
<a href="_images/olm-cell-subthresholdVi.png"
class="reference internal image-reference"><img
src="_images/olm-cell-subthresholdVi.png" style="width: 70%;"
alt="Current vs sub-threshold voltage curve for OLM cell generated using `generate_current_vs_frequency_curve`." /></a>
<figcaption><p><span class="caption-number">Fig. 41 </span><span
class="caption-text">Current vs sub-threshold voltage curve for OLM cell
generated using <span class="pre"><code
class="docutils literal notranslate">generate_current_vs_frequency_curve</code></span>.</span><a
href="#id5" class="headerlink"
title="Link to this image">#</a></p></figcaption>
</figure>

<figure id="id6" class="align-center">
<a href="_images/olm-cell-voltage-traces.png"
class="reference internal image-reference"><img
src="_images/olm-cell-voltage-traces.png" style="width: 70%;"
alt="Voltage traces for OLM cell with different injection currents generated using `generate_current_vs_frequency_curve`." /></a>
<figcaption><p><span class="caption-number">Fig. 42 </span><span
class="caption-text">Voltage traces for OLM cell with different
injection currents generated using <span class="pre"><code
class="docutils literal notranslate">generate_current_vs_frequency_curve</code></span>.</span><a
href="#id6" class="headerlink"
title="Link to this image">#</a></p></figcaption>
</figure>
:::
:::
:::
:::

[]{#document-Userdocs/SimulatingNeuroMLModels}

::: {#simulating-neuroml-models .section .tex2jax_ignore .mathjax_ignore}
[]{#userdocs-simulators}

## Simulating NeuroML Models[\#](#simulating-neuroml-models "Link to this heading"){.headerlink}

::: {.tip .admonition}
Validate NeuroML 2 files before using them.

It is good practice to [[validate NeuroML 2 files]{.std
.std-ref}](Landing.html#userdocs-validating-models){.reference
.internal} to check them for correctness before simulating them.
:::

::: {#using-open-source-brain .section}
[]{#userdocs-simulating-models-osb}

### Using Open Source Brain[\#](#using-open-source-brain "Link to this heading"){.headerlink}

Models that have already been converted to NeuroML and added to the
[Open Source Brain](https://www.opensourcebrain.org/){.reference
.external} platform can be simulated through your browser.

<figure id="id1" class="align-center">
<a href="_images/OSBv1.png"
class="reference internal image-reference"><img src="_images/OSBv1.png"
style="width: 801.0px; height: 461.5px;" alt="Open Source Brain" /></a>
<figcaption><p><span class="caption-number">Fig. 43 </span><span
class="caption-text">Examples of NeuroML 2 models visualised on Open
Source Brain. A) <a
href="https://www.opensourcebrain.org/projects/hodgkin-huxley-tutorial?explorer=https%3A%2F%2Fraw.githubusercontent.com%2Fopensourcebrain%2Ftutorials%2Fdevelopment%2Fmodels%2FhodgkinHuxley%2FGEPPETTO.json"
class="reference external">Hodgkin Huxley model</a> interactive
tutorial. B) Integrate and fire network model of cortical column (<a
href="https://www.opensourcebrain.org/projects/potjansdiesmann2014"
class="reference external">Potjans and Diesmann 2014</a>), showing
network connectivity. C) Cortical model with multicompartmental cells
(<a href="https://www.opensourcebrain.org/projects/thalamocortical"
class="reference external">Traub et al. 2005</a>), showing network
properties and simulated membrane potential activity. D) Model of C.
elegans nervous system from <a
href="https://www.opensourcebrain.org/projects/c302/"
class="reference external">OpenWorm project</a>. All
visualisation/analysis/simulation enabled due to models being in
standardised NeuroML format.</span><a href="#id1" class="headerlink"
title="Link to this image">#</a></p></figcaption>
</figure>

Most of the [OSB example
projects](https://www.opensourcebrain.org/projects){.reference
.external} feature prebuilt NeuroML models which can be simulated in
this way.

A discussion on the steps required for sharing your own models on OSB,
with a view to simulating them on the platform, can be found
[[here]{.std
.std-ref}](Landing.html#userdocs-creating-models-converting-conductance){.reference
.internal}.
:::

::: {#using-jneuroml-pyneuroml .section}
[]{#userdocs-simulating-models-jnml}[]{#userdocs-simulating-models-jlems}

### Using jNeuroML/pyNeuroML[\#](#using-jneuroml-pyneuroml "Link to this heading"){.headerlink}

[[jLEMS]{.std .std-ref}](Landing.html#jlems){.reference .internal} is
the reference implementation of the LEMS language in Java, and can be
used to simulate single compartment models written in NeuroML/LEMS. It
is included in both [[jNeuroML]{.std
.std-ref}](Landing.html#jneuroml){.reference .internal} and
[[pyNeuroML]{.std .std-ref}](Landing.html#pyneuroml){.reference
.internal}.

<figure id="id2" class="align-center">
<a href="_images/pynml_jnml.svg"
class="reference internal image-reference"><img
src="_images/pynml_jnml.svg" style="width: 500px;"
alt="jNeuroML and pyNeuroML" /></a>
<figcaption><p><span class="caption-number">Fig. 44 </span><span
class="caption-text">Relationship between <a href="Landing.html#jlems"
class="reference internal"><span class="std std-ref">jLEMS</span></a>,
<a href="Landing.html#jneuroml" class="reference internal"><span
class="std std-ref">jNeuroML</span></a> and <a
href="Landing.html#pyneuroml" class="reference internal"><span
class="std std-ref">pyNeuroML</span></a>.</span><a href="#id2"
class="headerlink" title="Link to this image">#</a></p></figcaption>
</figure>

[[jNeuroML]{.std .std-ref}](Landing.html#jneuroml){.reference .internal}
and [[pyNeuroML]{.std .std-ref}](Landing.html#pyneuroml){.reference
.internal} can be used at the command line as follows, when a [[LEMS
Simulation file]{.std
.std-ref}](Landing.html#userdocs-lemssimulation){.reference .internal}
has been created to describe what to simulate/plot/save:

::: {.highlight-console .notranslate}
::: highlight
    # Simulate the model using jNeuroML
    jnml <LEMS simulation file>

    # Simulate the model using pyNeuroML
    pynml <LEMS simulation file>
:::
:::

You can also run LEMS simulations using jNeuroML straight from a Python
script using the [[pyNeuroML]{.std
.std-ref}](Landing.html#pyneuroml){.reference .internal} API:

::: {.highlight-python .notranslate}
::: highlight
    from pyneuroml.pynml import run_lems_with_jneuroml

    ...

    run_lems_with_jneuroml(lems_file_name)
:::
:::
:::

::: {#using-neuron .section}
[]{#userdocs-simulating-models-neuron}

### Using NEURON[\#](#using-neuron "Link to this heading"){.headerlink}

For more complex models that can not be simulated using jLEMS (e.g.
incorporating multicompartmental cells), we can use the [[NEURON]{.std
.std-ref}](Landing.html#userdocs-neuron){.reference .internal}
simulator, also using [[jNeuroML]{.std
.std-ref}](Landing.html#jneuroml){.reference .internal} or
[[pyNeuroML]{.std .std-ref}](Landing.html#pyneuroml){.reference
.internal}, pointing at a [[LEMS Simulation file]{.std
.std-ref}](Landing.html#userdocs-lemssimulation){.reference .internal}
describing what to simulate, and using the [`-neuron`{.docutils .literal
.notranslate}]{.pre} option:

::: {.highlight-console .notranslate}
::: highlight
    # Simulate the model using NEURON with python/hoc/mod files generated by jNeuroML
    jnml <LEMS simulation file> -neuron -run

    # Simulate the model using NEURON with python/hoc/mod files generated by pyNeuroML
    pynml <LEMS simulation file> -neuron -run
:::
:::

You can also run LEMS simulations using the NEURON simulator using the
[[pyNeuroML]{.std .std-ref}](Landing.html#pyneuroml){.reference
.internal} API:

::: {.highlight-python .notranslate}
::: highlight
    from pyneuroml.pynml import run_lems_with_jneuroml_neuron

    ...

    run_lems_with_jneuroml_neuron(lems_file_name)
:::
:::

There is a **dedicated page on NEURON/NeuroML interactions**
[[here]{.std .std-ref}](Landing.html#userdocs-neuron){.reference
.internal}.
:::

::: {#using-netpyne .section}
[]{#userdocs-simulating-models-netpyne}

### Using NetPyNE[\#](#using-netpyne "Link to this heading"){.headerlink}

You can also generate and run [NetPyNE](https://netpyne.org){.reference
.external} code from NeuroML. To generate and run NetPyNE code, use
[[jNeuroML]{.std .std-ref}](Landing.html#jneuroml){.reference .internal}
or [[pyNeuroML]{.std .std-ref}](Landing.html#pyneuroml){.reference
.internal}:

::: {.highlight-console .notranslate}
::: highlight
    # Simulate the model using NetPyNE with python/hoc/mod files generated by jNeuroML
    jnml <LEMS simulation file> -netpyne -run

    # Simulate the model using NetPyNE with python/hoc/mod files generated by pyNeuroML
    pynml <LEMS simulation file> -netpyne -run
:::
:::

The main generated Python file name will end in [`_netpyne.py`{.docutils
.literal .notranslate}]{.pre}.

You can also run LEMS simulations using the NetPyNE simulator using the
[[pyNeuroML]{.std .std-ref}](Landing.html#pyneuroml){.reference
.internal} API:

::: {.highlight-python .notranslate}
::: highlight
    from pyneuroml.pynml import run_lems_with_jneuroml_netpyne

    ...

    run_lems_with_jneuroml_netpyne(lems_file_name)
:::
:::

There is a **dedicated page on NetPyNE/NeuroML interactions**
[[here]{.std .std-ref}](Landing.html#userdocs-netpyne){.reference
.internal}.
:::

::: {#using-brian2 .section}
[]{#userdocs-simulating-models-brian2}

### Using Brian2[\#](#using-brian2 "Link to this heading"){.headerlink}

You can export single component NeuroML models to Python scripts for
running them using the [Brian2](https://briansimulator.org){.reference
.external} simulator:

::: {.highlight-console .notranslate}
::: highlight
    # Using jnml
    jnml <LEMS simulation file> -brian2

    # Using pynml
    pynml <LEMS simulation file> -brian2
:::
:::

You can also run LEMS simulations using the Brian2 simulator using the
[[pyNeuroML]{.std .std-ref}](Landing.html#pyneuroml){.reference
.internal} API:

::: {.highlight-python .notranslate}
::: highlight
    from pyneuroml.pynml import run_lems_with_jneuroml_brian2

    ...

    run_lems_with_jneuroml_brian2(lems_file_name)
:::
:::

There is a **dedicated page on Brian/NeuroML interactions** [[here]{.std
.std-ref}](Landing.html#userdocs-brian){.reference .internal}.
:::

::: {#using-moose .section}
[]{#userdocs-simulating-models-moose}

### Using MOOSE[\#](#using-moose "Link to this heading"){.headerlink}

You can export NeuroML models to the MOOSE simulator format using
[[jNeuroML]{.std .std-ref}](Landing.html#jneuroml){.reference .internal}
or [[pyNeuroML]{.std .std-ref}](Landing.html#pyneuroml){.reference
.internal}, pointing at a [[LEMS Simulation file]{.std
.std-ref}](Landing.html#userdocs-lemssimulation){.reference .internal}
describing what to simulate, and using the [`-moose`{.docutils .literal
.notranslate}]{.pre} option:

::: {.highlight-console .notranslate}
::: highlight
    # Using jnml
    jnml <LEMS simulation file> -moose

    # Using pynml
    pynml <LEMS simulation file> -moose
:::
:::

There is a **dedicated page on MOOSE/NeuroML interactions** [[here]{.std
.std-ref}](Landing.html#userdocs-moose){.reference .internal}.
:::

::: {#using-eden .section}
[]{#userdocs-simulating-models-eden}

### Using EDEN[\#](#using-eden "Link to this heading"){.headerlink}

The EDEN simulator can load and simulate NeuroML v2 models.

There is a **dedicated page on EDEN/NeuroML interactions** [[here]{.std
.std-ref}](Landing.html#userdocs-eden){.reference .internal}.
:::

::: {#using-arbor .section}
[]{#userdocs-simulating-models-arbor}

### Using Arbor[\#](#using-arbor "Link to this heading"){.headerlink}

You can import NeuroML models to the Arbor simulator.

There is a **dedicated page on Arbor/NeuroML interactions** [[here]{.std
.std-ref}](Landing.html#userdocs-arbor){.reference .internal}.
:::
:::

[]{#document-Userdocs/OptimisingNeuroMLModels}

::: {#optimising-fitting-neuroml-models .section .tex2jax_ignore .mathjax_ignore}
[]{#userdocs-optimising}

## Optimising/fitting NeuroML Models[\#](#optimising-fitting-neuroml-models "Link to this heading"){.headerlink}

[[pyNeuroML]{.std .std-ref}](Landing.html#pyneuroml){.reference
.internal} includes the
[NeuroMLTuner](https://pyneuroml.readthedocs.io/en/development/pyneuroml.tune.html#module-pyneuroml.tune.NeuroMLTuner){.reference
.external} module for the tuning and optimisation of NeuroML models
against provided data. This uses the
[Neurotune](https://github.com/NeuralEnsemble/neurotune/){.reference
.external} Python package for the fitting of models using evolutionary
computation.

This page will walk through an example model optimisation.

::: container-fluid
::: {.row .my-2 .py-2}
::: {.col-sm-6 .px-2}
<figure id="id3" class="align-default">
<img src="_images/fitted_izhikevich_sim-exp-v.png"
alt="Membrane potential from example experimental data." />
<figcaption><p><span class="caption-number">Fig. 45 </span><span
class="caption-text">Membrane potential from the experimental
data.</span><a href="#id3" class="headerlink"
title="Link to this image">#</a></p></figcaption>
</figure>
:::

::: {.col-sm-6 .px-2}
<figure id="id4" class="align-default">
<img src="_images/fitted_izhikevich_output.png"
alt="Membrane potential obtained from example fitted model." />
<figcaption><p><span class="caption-number">Fig. 46 </span><span
class="caption-text">Membrane potential obtained from the model with
highest fitness.</span><a href="#id4" class="headerlink"
title="Link to this image">#</a></p></figcaption>
</figure>
:::
:::
:::

The Python script used to run the optimisation and generate the graphs
is given below. This can be adapted for other optimisations.

::: {.highlight-python .notranslate}
::: highlight
    #!/usr/bin/env python3
    """
    Example file showing the tuning of an Izhikevich neuron using pyNeuroML.

    File: source/Userdocs/NML2_examples/tune-izhikevich.py

    Copyright 2023 NeuroML contributors
    """


    from pyneuroml.tune.NeuroMLTuner import run_optimisation
    import pynwb  # type: ignore
    import numpy as np
    from pyelectro.utils import simple_network_analysis
    from typing import List, Dict, Tuple
    from pyneuroml.pynml import write_neuroml2_file
    from pyneuroml.pynml import generate_plot
    from pyneuroml.pynml import run_lems_with_jneuroml
    from neuroml import (
        NeuroMLDocument,
        Izhikevich2007Cell,
        PulseGenerator,
        Network,
        Population,
        ExplicitInput,
    )
    from hdmf.container import Container
    from pyneuroml.lems.LEMSSimulation import LEMSSimulation

    import sys


    def get_data_metrics(datafile: Container) -> Tuple[Dict, Dict, Dict]:
        """Analyse the data to get metrics to tune against.

        :returns: metrics from pyelectro analysis, currents, and the membrane potential values

        """
        analysis_results = {}
        currents = {}
        memb_vals = {}
        total_acquisitions = len(datafile.acquisition)

        for acq in range(1, total_acquisitions):
            print("Going over acquisition # {}".format(acq))

            # stimulus lasts about 1000ms, so we take about the first 1500 ms
            data_v = (
                datafile.acquisition["CurrentClampSeries_{:02d}".format(acq)].data[:15000] * 1000.0
            )
            # get sampling rate from the data
            sampling_rate = datafile.acquisition[
                "CurrentClampSeries_{:02d}".format(acq)
            ].rate
            # generate time steps from sampling rate
            data_t = np.arange(0, len(data_v) / sampling_rate, 1.0 / sampling_rate) * 1000.0
            # run the analysis
            analysis_results[acq] = simple_network_analysis({acq: data_v}, data_t)

            # extract current from description, but can be extracted from other
            # locations also, such as the CurrentClampStimulus series.
            data_i = (
                datafile.acquisition["CurrentClampSeries_{:02d}".format(acq)]
                .description.split("(")[1]
                .split("~")[1]
                .split(" ")[0]
            )
            currents[acq] = data_i
            memb_vals[acq] = (data_t, data_v)

        return (analysis_results, currents, memb_vals)


    def tune_izh_model(acq_list: List, metrics_from_data: Dict, currents: Dict) -> Dict:
        """Tune networks model against the data.

        Here we generate a network with the necessary number of Izhikevich cells,
        one for each current stimulus, and tune them against the experimental data.

        :param acq_list: list of indices of acquisitions/sweeps to tune against
        :type acq_list: list
        :param metrics_from_data: dictionary with the sweep number as index, and
            the dictionary containing metrics generated from the analysis
        :type metrics_from_data: dict
        :param currents: dictionary with sweep number as index and stimulus current
            value
        """

        # length of simulation of the cells---should match the length of the
        # experiment
        sim_time = 1500.0
        # Create a NeuroML template network simulation file that we will use for
        # the tuning
        template_doc = NeuroMLDocument(id="IzhTuneNet")
        # Add an Izhikevich cell with some parameters to the document
        template_doc.izhikevich2007_cells.append(
            Izhikevich2007Cell(
                id="Izh2007",
                C="100pF",
                v0="-60mV",
                k="0.7nS_per_mV",
                vr="-60mV",
                vt="-40mV",
                vpeak="35mV",
                a="0.03per_ms",
                b="-2nS",
                c="-50.0mV",
                d="100pA",
            )
        )
        template_doc.networks.append(Network(id="Network0"))
        # Add a cell for each acquisition list
        popsize = len(acq_list)
        template_doc.networks[0].populations.append(
            Population(id="Pop0", component="Izh2007", size=popsize)
        )

        # Add a current source for each cell, matching the currents that
        # were used in the experimental study.
        counter = 0
        for acq in acq_list:
            template_doc.pulse_generators.append(
                PulseGenerator(
                    id="Stim{}".format(counter),
                    delay="80ms",
                    duration="1000ms",
                    amplitude="{}pA".format(currents[acq]),
                )
            )
            template_doc.networks[0].explicit_inputs.append(
                ExplicitInput(
                    target="Pop0[{}]".format(counter), input="Stim{}".format(counter)
                )
            )
            counter = counter + 1

        # Print a summary
        print(template_doc.summary())

        # Write to a neuroml file and validate it.
        reference = "TuneIzhFergusonPyr3"
        template_filename = "{}.net.nml".format(reference)
        write_neuroml2_file(template_doc, template_filename, validate=True)

        # Now for the tuning bits

        # format is type:id/variable:id/units
        # supported types: cell/channel/izhikevich2007cell
        # supported variables:
        #  - channel: vShift
        #  - cell: channelDensity, vShift_channelDensity, channelDensityNernst,
        #  erev_id, erev_ion, specificCapacitance, resistivity
        #  - izhikevich2007Cell: all available attributes

        # we want to tune these parameters within these ranges
        # param: (min, max)
        parameters = {
            "izhikevich2007Cell:Izh2007/C/pF": (100, 300),
            "izhikevich2007Cell:Izh2007/k/nS_per_mV": (0.01, 2),
            "izhikevich2007Cell:Izh2007/vr/mV": (-70, -50),
            "izhikevich2007Cell:Izh2007/vt/mV": (-60, 0),
            "izhikevich2007Cell:Izh2007/vpeak/mV": (35, 70),
            "izhikevich2007Cell:Izh2007/a/per_ms": (0.001, 0.4),
            "izhikevich2007Cell:Izh2007/b/nS": (-10, 10),
            "izhikevich2007Cell:Izh2007/c/mV": (-65, -10),
            "izhikevich2007Cell:Izh2007/d/pA": (50, 500),
        }  # type: Dict[str, Tuple[float, float]]

        # Set up our target data and so on
        ctr = 0
        target_data = {}
        weights = {}
        for acq in acq_list:
            # data to fit to:
            # format: path/to/variable:metric
            # metric from pyelectro, for example:
            # https://pyelectro.readthedocs.io/en/latest/pyelectro.html?highlight=mean_spike_frequency#pyelectro.analysis.mean_spike_frequency
            mean_spike_frequency = "Pop0[{}]/v:mean_spike_frequency".format(ctr)
            average_last_1percent = "Pop0[{}]/v:average_last_1percent".format(ctr)
            first_spike_time = "Pop0[{}]/v:first_spike_time".format(ctr)

            # each metric can have an associated weight
            weights[mean_spike_frequency] = 1
            weights[average_last_1percent] = 1
            weights[first_spike_time] = 1

            # value of the target data from our data set
            target_data[mean_spike_frequency] = metrics_from_data[acq][
                "{}:mean_spike_frequency".format(acq)
            ]
            target_data[average_last_1percent] = metrics_from_data[acq][
                "{}:average_last_1percent".format(acq)
            ]
            target_data[first_spike_time] = metrics_from_data[acq][
                "{}:first_spike_time".format(acq)
            ]

            # only add these if the experimental data includes them
            # these are only generated for traces with spikes
            if "{}:average_maximum".format(acq) in metrics_from_data[acq]:
                average_maximum = "Pop0[{}]/v:average_maximum".format(ctr)
                weights[average_maximum] = 1
                target_data[average_maximum] = metrics_from_data[acq][
                    "{}:average_maximum".format(acq)
                ]
            if "{}:average_minimum".format(acq) in metrics_from_data[acq]:
                average_minimum = "Pop0[{}]/v:average_minimum".format(ctr)
                weights[average_minimum] = 1
                target_data[average_minimum] = metrics_from_data[acq][
                    "{}:average_minimum".format(acq)
                ]

            ctr = ctr + 1

        # simulator to use
        simulator = "jNeuroML"

        return run_optimisation(
            # Prefix for new files
            prefix="TuneIzh",
            # Name of the NeuroML template file
            neuroml_file=template_filename,
            # Name of the network
            target="Network0",
            # Parameters to be fitted
            parameters=list(parameters.keys()),
            # Our max and min constraints
            min_constraints=[v[0] for v in parameters.values()],
            max_constraints=[v[1] for v in parameters.values()],
            # Weights we set for parameters
            weights=weights,
            # The experimental metrics to fit to
            target_data=target_data,
            # Simulation time
            sim_time=sim_time,
            # EC parameters
            population_size=100,
            max_evaluations=500,
            num_selected=30,
            num_offspring=50,
            mutation_rate=0.9,
            num_elites=3,
            # Seed value
            seed=12345,
            # Simulator
            simulator=simulator,
            dt=0.025,
            show_plot_already='-nogui' not in sys.argv,
            save_to_file="fitted_izhikevich_fitness.png",
            save_to_file_scatter="fitted_izhikevich_scatter.png",
            save_to_file_hist="fitted_izhikevich_hist.png",
            save_to_file_output="fitted_izhikevich_output.png",
            num_parallel_evaluations=4,
        )


    def run_fitted_cell_simulation(
        sweeps_to_tune_against: List, tuning_report: Dict, simulation_id: str
    ) -> None:
        """Run a simulation with the values obtained from the fitting

        :param tuning_report: tuning report from the optimser
        :type tuning_report: Dict
        :param simulation_id: text id of simulation
        :type simulation_id: str

        """
        # get the fittest variables
        fittest_vars = tuning_report["fittest vars"]
        C = str(fittest_vars["izhikevich2007Cell:Izh2007/C/pF"]) + "pF"
        k = str(fittest_vars["izhikevich2007Cell:Izh2007/k/nS_per_mV"]) + "nS_per_mV"
        vr = str(fittest_vars["izhikevich2007Cell:Izh2007/vr/mV"]) + "mV"
        vt = str(fittest_vars["izhikevich2007Cell:Izh2007/vt/mV"]) + "mV"
        vpeak = str(fittest_vars["izhikevich2007Cell:Izh2007/vpeak/mV"]) + "mV"
        a = str(fittest_vars["izhikevich2007Cell:Izh2007/a/per_ms"]) + "per_ms"
        b = str(fittest_vars["izhikevich2007Cell:Izh2007/b/nS"]) + "nS"
        c = str(fittest_vars["izhikevich2007Cell:Izh2007/c/mV"]) + "mV"
        d = str(fittest_vars["izhikevich2007Cell:Izh2007/d/pA"]) + "pA"

        # Create a simulation using our obtained parameters.
        # Note that the tuner generates a graph with the fitted values already, but
        # we want to keep a copy of our fitted cell also, so we'll create a NeuroML
        # Document ourselves also.
        sim_time = 1500.0
        simulation_doc = NeuroMLDocument(id="FittedNet")
        # Add an Izhikevich cell with some parameters to the document
        simulation_doc.izhikevich2007_cells.append(
            Izhikevich2007Cell(
                id="Izh2007",
                C=C,
                v0="-60mV",
                k=k,
                vr=vr,
                vt=vt,
                vpeak=vpeak,
                a=a,
                b=b,
                c=c,
                d=d,
            )
        )
        simulation_doc.networks.append(Network(id="Network0"))
        # Add a cell for each acquisition list
        popsize = len(sweeps_to_tune_against)
        simulation_doc.networks[0].populations.append(
            Population(id="Pop0", component="Izh2007", size=popsize)
        )

        # Add a current source for each cell, matching the currents that
        # were used in the experimental study.
        counter = 0
        for acq in sweeps_to_tune_against:
            simulation_doc.pulse_generators.append(
                PulseGenerator(
                    id="Stim{}".format(counter),
                    delay="80ms",
                    duration="1000ms",
                    amplitude="{}pA".format(currents[acq]),
                )
            )
            simulation_doc.networks[0].explicit_inputs.append(
                ExplicitInput(
                    target="Pop0[{}]".format(counter), input="Stim{}".format(counter)
                )
            )
            counter = counter + 1

        # Print a summary
        print(simulation_doc.summary())

        # Write to a neuroml file and validate it.
        reference = "FittedIzhFergusonPyr3"
        simulation_filename = "{}.net.nml".format(reference)
        write_neuroml2_file(simulation_doc, simulation_filename, validate=True)

        simulation = LEMSSimulation(
            sim_id=simulation_id,
            duration=sim_time,
            dt=0.1,
            target="Network0",
            simulation_seed=54321,
        )
        simulation.include_neuroml2_file(simulation_filename)
        simulation.create_output_file("output0", "{}.v.dat".format(simulation_id))
        counter = 0
        for acq in sweeps_to_tune_against:
            simulation.add_column_to_output_file(
                "output0", "Pop0[{}]".format(counter), "Pop0[{}]/v".format(counter)
            )
            counter = counter + 1
        simulation_file = simulation.save_to_file()
        # simulate
        run_lems_with_jneuroml(simulation_file, max_memory="2G", nogui=True, plot=False)


    def plot_sim_data(
        sweeps_to_tune_against: List, simulation_id: str, memb_pots: Dict
    ) -> None:
        """Plot data from our fitted simulation

        :param simulation_id: string id of simulation
        :type simulation_id: str
        """
        # Plot
        data_array = np.loadtxt("%s.v.dat" % simulation_id)

        # construct data for plotting
        counter = 0
        time_vals_list = []
        sim_v_list = []
        data_v_list = []
        data_t_list = []
        stim_vals = []
        for acq in sweeps_to_tune_against:
            stim_vals.append("{}pA".format(currents[acq]))

            # remains the same for all columns
            time_vals_list.append(data_array[:, 0] * 1000.0)
            sim_v_list.append(data_array[:, counter + 1] * 1000.0)

            data_v_list.append(memb_pots[acq][1])
            data_t_list.append(memb_pots[acq][0])

            counter = counter + 1

        # Model membrane potential plot
        generate_plot(
            xvalues=time_vals_list,
            yvalues=sim_v_list,
            labels=stim_vals,
            title="Membrane potential (model)",
            show_plot_already=False,
            save_figure_to="%s-model-v.png" % simulation_id,
            xaxis="time (ms)",
            yaxis="membrane potential (mV)",
        )
        # data membrane potential plot
        generate_plot(
            xvalues=data_t_list,
            yvalues=data_v_list,
            labels=stim_vals,
            title="Membrane potential (exp)",
            show_plot_already=False,
            save_figure_to="%s-exp-v.png" % simulation_id,
            xaxis="time (ms)",
            yaxis="membrane potential (mV)",
        )


    if __name__ == "__main__":

        # set the default size for generated plots
        # https://matplotlib.org/stable/tutorials/introductory/customizing.html#a-sample-matplotlibrc-file
        import matplotlib as mpl
        mpl.rcParams["figure.figsize"] = [18, 12]

        io = pynwb.NWBHDF5IO("./FergusonEtAl2015_PYR3.nwb", "r")
        datafile = io.read()

        analysis_results, currents, memb_pots = get_data_metrics(datafile)

        # Choose what sweeps to tune against.
        # There are 33 sweeps: 1..33.
        # sweeps_to_tune_against = [1, 2, 15, 30, 31, 32, 33]
        sweeps_to_tune_against = [16,21]
        report = tune_izh_model(sweeps_to_tune_against, analysis_results, currents)

        simulation_id = "fitted_izhikevich_sim"
        run_fitted_cell_simulation(sweeps_to_tune_against, report, simulation_id)

        plot_sim_data(sweeps_to_tune_against, simulation_id, memb_pots)

        # close the data file
        io.close()
:::
:::

::: {#loading-data-and-calculating-metrics-to-use-for-optimisation .section}
[]{#userdocs-optimising-data}

### Loading data and calculating metrics to use for optimisation[\#](#loading-data-and-calculating-metrics-to-use-for-optimisation "Link to this heading"){.headerlink}

The first step in the optimisation of the model is to obtain the data
that the model is to be fitted against. In this example, we will use the
data set of CA1 pyramidal cell recordings using an intact whole
hippocampus preparation, including recordings of rebound firing
[\[[FHA+15](Landing.html#id17 "KA Ferguson, CYL Huh, B Amilhon, S Williams, and FK Skinner. Data set of CA1 pyramidal cell recordings using an intact whole hippocampus preparation, including recordings of rebound firing (V2). May 2015. URL: https://doi.org/10.5281/zenodo.17794, doi:10.5281/zenodo.17794."){.reference
.internal}\]]{#id1}. The data set is provided in the [Neurodata Without
Borders](https://nwb.org){.reference .external} (NWB) format. It can can
be downloaded [here on the Open Source Brain
repository](https://github.com/OpenSourceBrain/NWBShowcase/tree/master/FergusonEtAl2015){.reference
.external}, and can also be viewed on the [NWB
Explorer](https://nwbexplorer.opensourcebrain.org){.reference .external}
web application:

<figure id="id5" class="align-default">
<img src="_images/fitted_izhikevich_screenshot_nwbexplorer.png"
alt="Screenshot showing two recordings from FergusonEtAl2015_PYR3.nwb in NWB Explorer." />
<figcaption><p><span class="caption-number">Fig. 47 </span><span
class="caption-text">Screenshot showing two recordings from
FergusonEtAl2015_PYR3.nwb in NWB Explorer.</span><a href="#id5"
class="headerlink" title="Link to this image">#</a></p></figcaption>
</figure>

For this example, we will use the
[FergusonEtAl2015_PYR3.nwb](https://github.com/OpenSourceBrain/NWBShowcase/blob/master/FergusonEtAl2015/FergusonEtAl2015_PYR3.nwb){.reference
.external} data file. We use the
[PyNWB](https://pynwb.readthedocs.io/en/stable/){.reference .external}
package to read it, and then pass the loaded data to our
[`get_data_metrics`{.docutils .literal .notranslate}]{.pre} function to
extract the metrics we want to use for model fitting.

::: {.highlight-python .notranslate}
::: highlight
        io = pynwb.NWBHDF5IO("./FergusonEtAl2015_PYR3.nwb", "r")
        datafile = io.read()

        analysis_results, currents, memb_pots = get_data_metrics(datafile)
:::
:::

Similar to [[libNeuroML]{.std
.std-ref}](Landing.html#libneuroml){.reference .internal}, PyNWB
provides a Python object model to interact with NWB files. You can learn
more on using PyNWB in its
[documentation](https://pynwb.readthedocs.io/en/stable/){.reference
.external}.

Here, the data file includes recordings from multiple (33 in total)
current clamp experiments that are numbered from 1 through 33. We
iterate over each recording individually to extract the membrane
potential values and store them in [`data_v`{.docutils .literal
.notranslate}]{.pre}. For each, we also calculate the time stamps for
the recordings from the provided sampling rate. We pass this information
to the [`simple_network_analysis`{.docutils .literal
.notranslate}]{.pre} function provided by the
[PyElectro](https://github.com/NeuralEnsemble/pyelectro){.reference
.external} Python package to calculate features (metrics) that we will
use for fitting a neuron model.

::: {.highlight-python .notranslate}
::: highlight
    def get_data_metrics(datafile: Container) -> Tuple[Dict, Dict, Dict]:
        """Analyse the data to get metrics to tune against.

        :returns: metrics from pyelectro analysis, currents, and the membrane potential values

        """
        analysis_results = {}
        currents = {}
        memb_vals = {}
        total_acquisitions = len(datafile.acquisition)

        for acq in range(1, total_acquisitions):
            print("Going over acquisition # {}".format(acq))

            # stimulus lasts about 1000ms, so we take about the first 1500 ms
            data_v = (
                datafile.acquisition["CurrentClampSeries_{:02d}".format(acq)].data[:15000] * 1000.0
            )
            # get sampling rate from the data
            sampling_rate = datafile.acquisition[
                "CurrentClampSeries_{:02d}".format(acq)
            ].rate
            # generate time steps from sampling rate
            data_t = np.arange(0, len(data_v) / sampling_rate, 1.0 / sampling_rate) * 1000.0
            # run the analysis
            analysis_results[acq] = simple_network_analysis({acq: data_v}, data_t)

            # extract current from description, but can be extracted from other
            # locations also, such as the CurrentClampStimulus series.
            data_i = (
                datafile.acquisition["CurrentClampSeries_{:02d}".format(acq)]
                .description.split("(")[1]
                .split("~")[1]
                .split(" ")[0]
            )
            currents[acq] = data_i
            memb_vals[acq] = (data_t, data_v)

        return (analysis_results, currents, memb_vals)
:::
:::

The features calculated by PyElectro for each recording, which we store
in [`analysis_results`{.docutils .literal .notranslate}]{.pre}, can be
seen below:

::: {.highlight-console .notranslate}
::: highlight
    Going over acquisition # 1
    pyelectro >>> {   '1:average_last_1percent': -60.4182980855306,
    pyelectro >>>     '1:max_peak_no': 0,
    pyelectro >>>     '1:maximum': -57.922367,
    pyelectro >>>     '1:mean_spike_frequency': 0,
    pyelectro >>>     '1:min_peak_no': 0,
    pyelectro >>>     '1:minimum': -60.729984}
    Going over acquisition # 2
    pyelectro >>> {   '2:average_last_1percent': -60.2773068745931,
    pyelectro >>>     '2:max_peak_no': 0,
    pyelectro >>>     '2:maximum': -56.182865,
    pyelectro >>>     '2:mean_spike_frequency': 0,
    pyelectro >>>     '2:min_peak_no': 0,
    pyelectro >>>     '2:minimum': -60.882572}
    Going over acquisition # 3
    pyelectro >>> {   '3:average_last_1percent': -60.175174713134766,
    pyelectro >>>     '3:max_peak_no': 0,
    pyelectro >>>     '3:maximum': -54.22974,
    pyelectro >>>     '3:mean_spike_frequency': 0,
    pyelectro >>>     '3:min_peak_no': 0,
    pyelectro >>>     '3:minimum': -60.7605}
    Going over acquisition # 4
    pyelectro >>> {   '4:average_last_1percent': -60.11576716105143,
    pyelectro >>>     '4:max_peak_no': 0,
    pyelectro >>>     '4:maximum': -49.133305,
    pyelectro >>>     '4:mean_spike_frequency': 0,
    pyelectro >>>     '4:min_peak_no': 0,
    pyelectro >>>     '4:minimum': -60.607914}
    Going over acquisition # 5
    pyelectro >>> {   '5:average_last_1percent': -59.628299713134766,
    pyelectro >>>     '5:max_peak_no': 0,
    pyelectro >>>     '5:maximum': -48.645023,
    pyelectro >>>     '5:mean_spike_frequency': 0,
    pyelectro >>>     '5:min_peak_no': 0,
    pyelectro >>>     '5:minimum': -60.241703}
    Going over acquisition # 6
    pyelectro >>> {   '6:average_last_1percent': -60.04679743448893,
    pyelectro >>>     '6:max_peak_no': 0,
    pyelectro >>>     '6:maximum': -45.16602,
    pyelectro >>>     '6:mean_spike_frequency': 0,
    pyelectro >>>     '6:min_peak_no': 0,
    pyelectro >>>     '6:minimum': -60.699467}
    Going over acquisition # 7
    pyelectro >>> {   '7:average_last_1percent': -59.88566462198893,
    pyelectro >>>     '7:average_maximum': 66.3147,
    pyelectro >>>     '7:average_minimum': -47.9126,
    pyelectro >>>     '7:first_spike_time': 482.20000000000005,
    pyelectro >>>     '7:max_peak_no': 2,
    pyelectro >>>     '7:maximum': 67.38281,
    pyelectro >>>     '7:mean_spike_frequency': 0,
    pyelectro >>>     '7:min_peak_no': 1,
    pyelectro >>>     '7:minimum': -60.15015}
    Going over acquisition # 8
    pyelectro >>> {   '8:average_last_1percent': -60.5531857808431,
    pyelectro >>>     '8:average_maximum': 64.63623,
    pyelectro >>>     '8:average_minimum': -46.05103,
    pyelectro >>>     '8:first_spike_time': 280.8,
    pyelectro >>>     '8:max_peak_no': 2,
    pyelectro >>>     '8:maximum': 65.460205,
    pyelectro >>>     '8:mean_spike_frequency': 0,
    pyelectro >>>     '8:min_peak_no': 1,
    pyelectro >>>     '8:minimum': -61.096195}
    Going over acquisition # 9
    pyelectro >>> {   '9:average_last_1percent': -60.2797482808431,
    pyelectro >>>     '9:average_maximum': 62.187195,
    pyelectro >>>     '9:average_minimum': -45.867924,
    pyelectro >>>     '9:first_spike_time': 192.10000000000002,
    pyelectro >>>     '9:interspike_time_covar': 0.12604539832023948,
    pyelectro >>>     '9:max_interspike_time': 233.69999999999993,
    pyelectro >>>     '9:max_peak_no': 4,
    pyelectro >>>     '9:maximum': 64.42261,
    pyelectro >>>     '9:mean_spike_frequency': 4.984216647283602,
    pyelectro >>>     '9:min_interspike_time': 172.29999999999995,
    pyelectro >>>     '9:min_peak_no': 3,
    pyelectro >>>     '9:minimum': -60.91309,
    pyelectro >>>     '9:peak_decay_exponent': -0.008125577959852965,
    pyelectro >>>     '9:peak_linear_gradient': -0.007648055557429393,
    pyelectro >>>     '9:spike_broadening': 0.891046031985908,
    pyelectro >>>     '9:spike_frequency_adaptation': -0.05850411039850073,
    pyelectro >>>     '9:spike_width_adaptation': 0.02767948174212246,
    pyelectro >>>     '9:trough_decay_exponent': -0.00318728965696189,
    pyelectro >>>     '9:trough_phase_adaptation': -0.05016262394149966}
    Going over acquisition # 10
    pyelectro >>> {   '10:average_last_1percent': -60.4292844136556,
    pyelectro >>>     '10:average_maximum': 59.680183,
    pyelectro >>>     '10:average_minimum': -44.731144,
    pyelectro >>>     '10:first_spike_time': 156.9,
    pyelectro >>>     '10:interspike_time_covar': 0.5779639183148945,
    pyelectro >>>     '10:max_interspike_time': 438.5000000000001,
    pyelectro >>>     '10:max_peak_no': 5,
    pyelectro >>>     '10:maximum': 62.072758,
    pyelectro >>>     '10:mean_spike_frequency': 4.553215708594194,
    pyelectro >>>     '10:min_interspike_time': 132.60000000000005,
    pyelectro >>>     '10:min_peak_no': 4,
    pyelectro >>>     '10:minimum': -60.882572,
    pyelectro >>>     '10:peak_decay_exponent': -0.009585017031582235,
    pyelectro >>>     '10:peak_linear_gradient': -0.005260966229876463,
    pyelectro >>>     '10:spike_broadening': 0.8756680523402136,
    pyelectro >>>     '10:spike_frequency_adaptation': -0.04780471479504103,
    pyelectro >>>     '10:spike_width_adaptation': 0.014551159295128686,
    pyelectro >>>     '10:trough_decay_exponent': -0.006056437839814275,
    pyelectro >>>     '10:trough_phase_adaptation': -0.04269124477477909}
    Going over acquisition # 11
    pyelectro >>> {   '11:average_last_1percent': -60.84635798136393,
    pyelectro >>>     '11:average_maximum': 58.73414,
    pyelectro >>>     '11:average_minimum': -43.800358,
    pyelectro >>>     '11:first_spike_time': 138.1,
    pyelectro >>>     '11:interspike_time_covar': 0.18317620745649893,
    pyelectro >>>     '11:max_interspike_time': 179.9999999999999,
    pyelectro >>>     '11:max_peak_no': 5,
    pyelectro >>>     '11:maximum': 61.03516,
    pyelectro >>>     '11:mean_spike_frequency': 7.033585370142431,
    pyelectro >>>     '11:min_interspike_time': 106.50000000000003,
    pyelectro >>>     '11:min_peak_no': 4,
    pyelectro >>>     '11:minimum': -61.248783,
    pyelectro >>>     '11:peak_decay_exponent': -0.010372120562797708,
    pyelectro >>>     '11:peak_linear_gradient': -0.0074866848970347325,
    pyelectro >>>     '11:spike_broadening': 0.8498887121142943,
    pyelectro >>>     '11:spike_frequency_adaptation': -0.052381521145715274,
    pyelectro >>>     '11:spike_width_adaptation': 0.025414793671653328,
    pyelectro >>>     '11:trough_decay_exponent': -0.007552906636393599,
    pyelectro >>>     '11:trough_phase_adaptation': -0.04473631982885844}
    Going over acquisition # 12
    pyelectro >>> {   '12:average_last_1percent': -61.085208892822266,
    pyelectro >>>     '12:average_maximum': 58.481857,
    pyelectro >>>     '12:average_minimum': -42.974857,
    pyelectro >>>     '12:first_spike_time': 127.40000000000002,
    pyelectro >>>     '12:interspike_time_covar': 0.16275057704467177,
    pyelectro >>>     '12:max_interspike_time': 136.5,
    pyelectro >>>     '12:max_peak_no': 6,
    pyelectro >>>     '12:maximum': 61.737064,
    pyelectro >>>     '12:mean_spike_frequency': 8.791981712678037,
    pyelectro >>>     '12:min_interspike_time': 84.60000000000001,
    pyelectro >>>     '12:min_peak_no': 5,
    pyelectro >>>     '12:minimum': -61.462406,
    pyelectro >>>     '12:peak_decay_exponent': -0.015987075984851808,
    pyelectro >>>     '12:peak_linear_gradient': -0.009383652380440125,
    pyelectro >>>     '12:spike_broadening': 0.8205694396572242,
    pyelectro >>>     '12:spike_frequency_adaptation': -0.04259621402895674,
    pyelectro >>>     '12:spike_width_adaptation': 0.0228428573204052,
    pyelectro >>>     '12:trough_decay_exponent': -0.012180031782655684,
    pyelectro >>>     '12:trough_phase_adaptation': 0.0003391093549627843}
    Going over acquisition # 13
    pyelectro >>> {   '13:average_last_1percent': -60.6233762105306,
    pyelectro >>>     '13:average_maximum': 56.980137,
    pyelectro >>>     '13:average_minimum': -42.205814,
    pyelectro >>>     '13:first_spike_time': 122.9,
    pyelectro >>>     '13:interspike_time_covar': 0.1989630987796946,
    pyelectro >>>     '13:max_interspike_time': 138.9000000000001,
    pyelectro >>>     '13:max_peak_no': 8,
    pyelectro >>>     '13:maximum': 61.30982,
    pyelectro >>>     '13:mean_spike_frequency': 9.743875278396434,
    pyelectro >>>     '13:min_interspike_time': 75.4,
    pyelectro >>>     '13:min_peak_no': 7,
    pyelectro >>>     '13:minimum': -60.974125,
    pyelectro >>>     '13:peak_decay_exponent': -0.01856642711769415,
    pyelectro >>>     '13:peak_linear_gradient': -0.009561076077386068,
    pyelectro >>>     '13:spike_broadening': 0.803052014150605,
    pyelectro >>>     '13:spike_frequency_adaptation': -0.028079821139188187,
    pyelectro >>>     '13:spike_width_adaptation': 0.01504310702538977,
    pyelectro >>>     '13:trough_decay_exponent': -0.013208504624154408,
    pyelectro >>>     '13:trough_phase_adaptation': -0.025379665674913895}
    Going over acquisition # 14
    pyelectro >>> {   '14:average_last_1percent': -60.723270416259766,
    pyelectro >>>     '14:average_maximum': 55.986195,
    pyelectro >>>     '14:average_minimum': -41.54587,
    pyelectro >>>     '14:first_spike_time': 114.8,
    pyelectro >>>     '14:interspike_time_covar': 0.34335772265161896,
    pyelectro >>>     '14:max_interspike_time': 194.39999999999998,
    pyelectro >>>     '14:max_peak_no': 9,
    pyelectro >>>     '14:maximum': 60.882572,
    pyelectro >>>     '14:mean_spike_frequency': 8.785416209092906,
    pyelectro >>>     '14:min_interspike_time': 74.40000000000002,
    pyelectro >>>     '14:min_peak_no': 8,
    pyelectro >>>     '14:minimum': -61.126713,
    pyelectro >>>     '14:peak_decay_exponent': -0.022541304298167242,
    pyelectro >>>     '14:peak_linear_gradient': -0.008000988888256885,
    pyelectro >>>     '14:spike_broadening': 0.8009562042583951,
    pyelectro >>>     '14:spike_frequency_adaptation': -0.022434594832088855,
    pyelectro >>>     '14:spike_width_adaptation': 0.010424354074822617,
    pyelectro >>>     '14:trough_decay_exponent': -0.018754566142487872,
    pyelectro >>>     '14:trough_phase_adaptation': -0.019014319738344054}
    Going over acquisition # 15
    pyelectro >>> {   '15:average_last_1percent': -60.99833552042643,
    pyelectro >>>     '15:average_maximum': 55.89295,
    pyelectro >>>     '15:average_minimum': -40.78892,
    pyelectro >>>     '15:first_spike_time': 113.2,
    pyelectro >>>     '15:interspike_time_covar': 0.6436697297327385,
    pyelectro >>>     '15:max_interspike_time': 311.30000000000007,
    pyelectro >>>     '15:max_peak_no': 8,
    pyelectro >>>     '15:maximum': 60.91309,
    pyelectro >>>     '15:mean_spike_frequency': 8.201523140011716,
    pyelectro >>>     '15:min_interspike_time': 71.60000000000001,
    pyelectro >>>     '15:min_peak_no': 7,
    pyelectro >>>     '15:minimum': -61.370853,
    pyelectro >>>     '15:peak_decay_exponent': -0.025953113905923406,
    pyelectro >>>     '15:peak_linear_gradient': -0.008306657481016694,
    pyelectro >>>     '15:spike_broadening': 0.7782197474305453,
    pyelectro >>>     '15:spike_frequency_adaptation': -0.030010388928284993,
    pyelectro >>>     '15:spike_width_adaptation': 0.012108100430332605,
    pyelectro >>>     '15:trough_decay_exponent': -0.01262141611091244,
    pyelectro >>>     '15:trough_phase_adaptation': -0.025746376793896804}
    Going over acquisition # 16
    pyelectro >>> {   '16:average_last_1percent': -60.380863189697266,
    pyelectro >>>     '16:average_maximum': 54.52382,
    pyelectro >>>     '16:average_minimum': -39.78882,
    pyelectro >>>     '16:first_spike_time': 108.9,
    pyelectro >>>     '16:interspike_time_covar': 0.23652534644271225,
    pyelectro >>>     '16:max_interspike_time': 123.00000000000011,
    pyelectro >>>     '16:max_peak_no': 11,
    pyelectro >>>     '16:maximum': 60.7605,
    pyelectro >>>     '16:mean_spike_frequency': 10.8837614279495,
    pyelectro >>>     '16:min_interspike_time': 59.30000000000001,
    pyelectro >>>     '16:min_peak_no': 10,
    pyelectro >>>     '16:minimum': -60.79102,
    pyelectro >>>     '16:peak_decay_exponent': -0.03301104578192642,
    pyelectro >>>     '16:peak_linear_gradient': -0.007545664792227554,
    pyelectro >>>     '16:spike_broadening': 0.7546314677569799,
    pyelectro >>>     '16:spike_frequency_adaptation': -0.013692136410690963,
    pyelectro >>>     '16:spike_width_adaptation': 0.008664177864358623,
    pyelectro >>>     '16:trough_decay_exponent': -0.023836469122601508,
    pyelectro >>>     '16:trough_phase_adaptation': -0.010081239597430595}
    Going over acquisition # 17
    pyelectro >>> {   '17:average_last_1percent': -60.552982330322266,
    pyelectro >>>     '17:average_maximum': 54.44642,
    pyelectro >>>     '17:average_minimum': -39.008247,
    pyelectro >>>     '17:first_spike_time': 105.6,
    pyelectro >>>     '17:interspike_time_covar': 0.19651182311074483,
    pyelectro >>>     '17:max_interspike_time': 106.29999999999995,
    pyelectro >>>     '17:max_peak_no': 10,
    pyelectro >>>     '17:maximum': 60.63843,
    pyelectro >>>     '17:mean_spike_frequency': 11.506008693428791,
    pyelectro >>>     '17:min_interspike_time': 58.60000000000002,
    pyelectro >>>     '17:min_peak_no': 9,
    pyelectro >>>     '17:minimum': -61.03516,
    pyelectro >>>     '17:peak_decay_exponent': -0.03684157763477531,
    pyelectro >>>     '17:peak_linear_gradient': -0.009090863209461205,
    pyelectro >>>     '17:spike_broadening': 0.7534694949245309,
    pyelectro >>>     '17:spike_frequency_adaptation': -0.023373852901912264,
    pyelectro >>>     '17:spike_width_adaptation': 0.011268432654001511,
    pyelectro >>>     '17:trough_decay_exponent': -0.020590018720385343,
    pyelectro >>>     '17:trough_phase_adaptation': -0.00906121257722172}
    Going over acquisition # 18
    pyelectro >>> {   '18:average_last_1percent': -60.64799372355143,
    pyelectro >>>     '18:average_maximum': 53.783077,
    pyelectro >>>     '18:average_minimum': -38.424683,
    pyelectro >>>     '18:first_spike_time': 104.2,
    pyelectro >>>     '18:interspike_time_covar': 0.2502832189694502,
    pyelectro >>>     '18:max_interspike_time': 124.59999999999991,
    pyelectro >>>     '18:max_peak_no': 11,
    pyelectro >>>     '18:maximum': 60.63843,
    pyelectro >>>     '18:mean_spike_frequency': 11.420740063956146,
    pyelectro >>>     '18:min_interspike_time': 53.09999999999998,
    pyelectro >>>     '18:min_peak_no': 10,
    pyelectro >>>     '18:minimum': -61.03516,
    pyelectro >>>     '18:peak_decay_exponent': -0.04141134556729957,
    pyelectro >>>     '18:peak_linear_gradient': -0.008476351143979814,
    pyelectro >>>     '18:spike_broadening': 0.7403041191114148,
    pyelectro >>>     '18:spike_frequency_adaptation': -0.01809717600857398,
    pyelectro >>>     '18:spike_width_adaptation': 0.009174879706803092,
    pyelectro >>>     '18:trough_decay_exponent': -0.02760451378209885,
    pyelectro >>>     '18:trough_phase_adaptation': -0.005774543184281237}
    Going over acquisition # 19
    pyelectro >>> {   '19:average_last_1percent': -60.855106353759766,
    pyelectro >>>     '19:average_maximum': 53.430737,
    pyelectro >>>     '19:average_minimum': -37.713623,
    pyelectro >>>     '19:first_spike_time': 102.50000000000001,
    pyelectro >>>     '19:interspike_time_covar': 0.25333327224227414,
    pyelectro >>>     '19:max_interspike_time': 114.69999999999993,
    pyelectro >>>     '19:max_peak_no': 11,
    pyelectro >>>     '19:maximum': 60.607914,
    pyelectro >>>     '19:mean_spike_frequency': 12.47038284075321,
    pyelectro >>>     '19:min_interspike_time': 51.8,
    pyelectro >>>     '19:min_peak_no': 10,
    pyelectro >>>     '19:minimum': -61.248783,
    pyelectro >>>     '19:peak_decay_exponent': -0.04918301935790627,
    pyelectro >>>     '19:peak_linear_gradient': -0.008553290998046907,
    pyelectro >>>     '19:spike_broadening': 0.7301692622822238,
    pyelectro >>>     '19:spike_frequency_adaptation': -0.015561797159916213,
    pyelectro >>>     '19:spike_width_adaptation': 0.010054185105794627,
    pyelectro >>>     '19:trough_decay_exponent': -0.03413795061471875,
    pyelectro >>>     '19:trough_phase_adaptation': -0.011967671377256838}
    Going over acquisition # 20
    pyelectro >>> {   '20:average_last_1percent': -60.793460845947266,
    pyelectro >>>     '20:average_maximum': 53.11169,
    pyelectro >>>     '20:average_minimum': -37.045288,
    pyelectro >>>     '20:first_spike_time': 101.4,
    pyelectro >>>     '20:interspike_time_covar': 0.2865607615520669,
    pyelectro >>>     '20:max_interspike_time': 123.0,
    pyelectro >>>     '20:max_peak_no': 11,
    pyelectro >>>     '20:maximum': 60.51636,
    pyelectro >>>     '20:mean_spike_frequency': 12.624668602449184,
    pyelectro >>>     '20:min_interspike_time': 49.099999999999994,
    pyelectro >>>     '20:min_peak_no': 10,
    pyelectro >>>     '20:minimum': -61.30982,
    pyelectro >>>     '20:peak_decay_exponent': -0.053836942989781186,
    pyelectro >>>     '20:peak_linear_gradient': -0.009554089132365705,
    pyelectro >>>     '20:spike_broadening': 0.7067401817806985,
    pyelectro >>>     '20:spike_frequency_adaptation': -0.02046601020346991,
    pyelectro >>>     '20:spike_width_adaptation': 0.01032699280307952,
    pyelectro >>>     '20:trough_decay_exponent': -0.03229413870032492,
    pyelectro >>>     '20:trough_phase_adaptation': -0.009902402143729637}
    Going over acquisition # 21
    pyelectro >>> {   '21:average_last_1percent': -59.912113189697266,
    pyelectro >>>     '21:average_maximum': 51.912754,
    pyelectro >>>     '21:average_minimum': -35.964966,
    pyelectro >>>     '21:first_spike_time': 100.4,
    pyelectro >>>     '21:interspike_time_covar': 0.31784939578511834,
    pyelectro >>>     '21:max_interspike_time': 130.10000000000002,
    pyelectro >>>     '21:max_peak_no': 13,
    pyelectro >>>     '21:maximum': 61.15723,
    pyelectro >>>     '21:mean_spike_frequency': 12.369858777445623,
    pyelectro >>>     '21:min_interspike_time': 45.10000000000002,
    pyelectro >>>     '21:min_peak_no': 12,
    pyelectro >>>     '21:minimum': -61.614994,
    pyelectro >>>     '21:peak_decay_exponent': -0.06340663337165775,
    pyelectro >>>     '21:peak_linear_gradient': -0.009514554022982258,
    pyelectro >>>     '21:spike_broadening': 0.6805457955255507,
    pyelectro >>>     '21:spike_frequency_adaptation': -0.012321447021551269,
    pyelectro >>>     '21:spike_width_adaptation': 0.007303096579113352,
    pyelectro >>>     '21:trough_decay_exponent': -0.03236374133246423,
    pyelectro >>>     '21:trough_phase_adaptation': -0.009705621425080494}
    Going over acquisition # 22
    pyelectro >>> {   '22:average_last_1percent': -60.0850461324056,
    pyelectro >>>     '22:average_maximum': 51.325874,
    pyelectro >>>     '22:average_minimum': -35.22746,
    pyelectro >>>     '22:first_spike_time': 97.7,
    pyelectro >>>     '22:interspike_time_covar': 0.3280130255436152,
    pyelectro >>>     '22:max_interspike_time': 123.30000000000007,
    pyelectro >>>     '22:max_peak_no': 13,
    pyelectro >>>     '22:maximum': 60.91309,
    pyelectro >>>     '22:mean_spike_frequency': 12.784998934583422,
    pyelectro >>>     '22:min_interspike_time': 40.60000000000001,
    pyelectro >>>     '22:min_peak_no': 12,
    pyelectro >>>     '22:minimum': -60.51636,
    pyelectro >>>     '22:peak_decay_exponent': -0.07398649685748288,
    pyelectro >>>     '22:peak_linear_gradient': -0.008846573199048182,
    pyelectro >>>     '22:spike_broadening': 0.673572178798493,
    pyelectro >>>     '22:spike_frequency_adaptation': -0.01394751742502263,
    pyelectro >>>     '22:spike_width_adaptation': 0.00745978471908774,
    pyelectro >>>     '22:trough_decay_exponent': -0.03985576781966312,
    pyelectro >>>     '22:trough_phase_adaptation': -0.012949190338366346}
    Going over acquisition # 23
    pyelectro >>> {   '23:average_last_1percent': -60.42582575480143,
    pyelectro >>>     '23:average_maximum': 50.883705,
    pyelectro >>>     '23:average_minimum': -34.70553,
    pyelectro >>>     '23:first_spike_time': 97.0,
    pyelectro >>>     '23:interspike_time_covar': 0.3025008293643565,
    pyelectro >>>     '23:max_interspike_time': 113.00000000000011,
    pyelectro >>>     '23:max_peak_no': 14,
    pyelectro >>>     '23:maximum': 61.126713,
    pyelectro >>>     '23:mean_spike_frequency': 13.453378867846421,
    pyelectro >>>     '23:min_interspike_time': 37.599999999999994,
    pyelectro >>>     '23:min_peak_no': 13,
    pyelectro >>>     '23:minimum': -60.79102,
    pyelectro >>>     '23:peak_decay_exponent': -0.09454776279913378,
    pyelectro >>>     '23:peak_linear_gradient': -0.007913299554316041,
    pyelectro >>>     '23:spike_broadening': 0.6664213397577756,
    pyelectro >>>     '23:spike_frequency_adaptation': -0.014974747741974222,
    pyelectro >>>     '23:spike_width_adaptation': 0.0067844847504214744,
    pyelectro >>>     '23:trough_decay_exponent': -0.04357925225307838,
    pyelectro >>>     '23:trough_phase_adaptation': -0.006437508651280852}
    Going over acquisition # 24
    pyelectro >>> {   '24:average_last_1percent': -60.48380915323893,
    pyelectro >>>     '24:average_maximum': 50.66572,
    pyelectro >>>     '24:average_minimum': -34.043533,
    pyelectro >>>     '24:first_spike_time': 95.9,
    pyelectro >>>     '24:interspike_time_covar': 0.2757061784222999,
    pyelectro >>>     '24:max_interspike_time': 106.29999999999995,
    pyelectro >>>     '24:max_peak_no': 14,
    pyelectro >>>     '24:maximum': 61.2793,
    pyelectro >>>     '24:mean_spike_frequency': 13.685651121170649,
    pyelectro >>>     '24:min_interspike_time': 42.5,
    pyelectro >>>     '24:min_peak_no': 13,
    pyelectro >>>     '24:minimum': -61.03516,
    pyelectro >>>     '24:peak_decay_exponent': -0.10291055243646331,
    pyelectro >>>     '24:peak_linear_gradient': -0.008156801002617309,
    pyelectro >>>     '24:spike_broadening': 0.647962840377368,
    pyelectro >>>     '24:spike_frequency_adaptation': -0.01033640102347251,
    pyelectro >>>     '24:spike_width_adaptation': 0.0070544336550024695,
    pyelectro >>>     '24:trough_decay_exponent': -0.04136814132208841,
    pyelectro >>>     '24:trough_phase_adaptation': -0.007165702258804302}
    Going over acquisition # 25
    pyelectro >>> {   '25:average_last_1percent': -60.056766510009766,
    pyelectro >>>     '25:average_maximum': 50.34093,
    pyelectro >>>     '25:average_minimum': -33.228947,
    pyelectro >>>     '25:first_spike_time': 95.60000000000001,
    pyelectro >>>     '25:interspike_time_covar': 0.28833246313094774,
    pyelectro >>>     '25:max_interspike_time': 100.80000000000007,
    pyelectro >>>     '25:max_peak_no': 14,
    pyelectro >>>     '25:maximum': 61.40137,
    pyelectro >>>     '25:mean_spike_frequency': 14.023732470334416,
    pyelectro >>>     '25:min_interspike_time': 39.10000000000001,
    pyelectro >>>     '25:min_peak_no': 13,
    pyelectro >>>     '25:minimum': -60.607914,
    pyelectro >>>     '25:peak_decay_exponent': -0.1047695739806843,
    pyelectro >>>     '25:peak_linear_gradient': -0.00879119329941481,
    pyelectro >>>     '25:spike_broadening': 0.6394636967007732,
    pyelectro >>>     '25:spike_frequency_adaptation': -0.011357738090467376,
    pyelectro >>>     '25:spike_width_adaptation': 0.007198805900434394,
    pyelectro >>>     '25:trough_decay_exponent': -0.03898706127522965,
    pyelectro >>>     '25:trough_phase_adaptation': -0.005791081007349543}
    Going over acquisition # 26
    pyelectro >>> {   '26:average_last_1percent': -60.2272580464681,
    pyelectro >>>     '26:average_maximum': 49.776356,
    pyelectro >>>     '26:average_minimum': -32.534084,
    pyelectro >>>     '26:first_spike_time': 94.89999999999999,
    pyelectro >>>     '26:interspike_time_covar': 0.3174070553930572,
    pyelectro >>>     '26:max_interspike_time': 115.70000000000005,
    pyelectro >>>     '26:max_peak_no': 14,
    pyelectro >>>     '26:maximum': 61.15723,
    pyelectro >>>     '26:mean_spike_frequency': 14.697569248162802,
    pyelectro >>>     '26:min_interspike_time': 35.60000000000001,
    pyelectro >>>     '26:min_peak_no': 13,
    pyelectro >>>     '26:minimum': -60.729984,
    pyelectro >>>     '26:peak_decay_exponent': -0.11931629839276492,
    pyelectro >>>     '26:peak_linear_gradient': -0.009603020729143796,
    pyelectro >>>     '26:spike_broadening': 0.6253471365146865,
    pyelectro >>>     '26:spike_frequency_adaptation': -0.015585439927950483,
    pyelectro >>>     '26:spike_width_adaptation': 0.007759081127414656,
    pyelectro >>>     '26:trough_decay_exponent': -0.049101688628808246,
    pyelectro >>>     '26:trough_phase_adaptation': -0.012139424609633551}
    Going over acquisition # 27
    pyelectro >>> {   '27:average_last_1percent': -60.3578732808431,
    pyelectro >>>     '27:average_maximum': 49.30769,
    pyelectro >>>     '27:average_minimum': -32.003548,
    pyelectro >>>     '27:first_spike_time': 93.60000000000001,
    pyelectro >>>     '27:interspike_time_covar': 0.309017296765978,
    pyelectro >>>     '27:max_interspike_time': 109.0,
    pyelectro >>>     '27:max_peak_no': 14,
    pyelectro >>>     '27:maximum': 61.43189,
    pyelectro >>>     '27:mean_spike_frequency': 14.729209154769997,
    pyelectro >>>     '27:min_interspike_time': 32.19999999999999,
    pyelectro >>>     '27:min_peak_no': 13,
    pyelectro >>>     '27:minimum': -60.7605,
    pyelectro >>>     '27:peak_decay_exponent': -0.12538878925370048,
    pyelectro >>>     '27:peak_linear_gradient': -0.009067695791685005,
    pyelectro >>>     '27:spike_broadening': 0.6066765033439258,
    pyelectro >>>     '27:spike_frequency_adaptation': -0.015492078035720953,
    pyelectro >>>     '27:spike_width_adaptation': 0.007539073044770246,
    pyelectro >>>     '27:trough_decay_exponent': -0.05115040330367209,
    pyelectro >>>     '27:trough_phase_adaptation': -0.012545852013557039}
    Going over acquisition # 28
    pyelectro >>> {   '28:average_last_1percent': -60.3798459370931,
    pyelectro >>>     '28:average_maximum': 48.999027,
    pyelectro >>>     '28:average_minimum': -31.055996,
    pyelectro >>>     '28:first_spike_time': 93.4,
    pyelectro >>>     '28:interspike_time_covar': 0.30636145843159784,
    pyelectro >>>     '28:max_interspike_time': 104.70000000000005,
    pyelectro >>>     '28:max_peak_no': 15,
    pyelectro >>>     '28:maximum': 61.2793,
    pyelectro >>>     '28:mean_spike_frequency': 14.760147601476012,
    pyelectro >>>     '28:min_interspike_time': 34.20000000000002,
    pyelectro >>>     '28:min_peak_no': 14,
    pyelectro >>>     '28:minimum': -60.943607,
    pyelectro >>>     '28:peak_decay_exponent': -0.16496940386452533,
    pyelectro >>>     '28:peak_linear_gradient': -0.007631694348641044,
    pyelectro >>>     '28:spike_broadening': 0.5953810644123492,
    pyelectro >>>     '28:spike_frequency_adaptation': -0.014342884276047468,
    pyelectro >>>     '28:spike_width_adaptation': 0.006650416835633624,
    pyelectro >>>     '28:trough_decay_exponent': -0.04668774074305247,
    pyelectro >>>     '28:trough_phase_adaptation': 0.006738292518776989}
    Going over acquisition # 29
    pyelectro >>> {   '29:average_last_1percent': -60.662235260009766,
    pyelectro >>>     '29:average_maximum': 48.664642,
    pyelectro >>>     '29:average_minimum': -30.39551,
    pyelectro >>>     '29:first_spike_time': 93.10000000000001,
    pyelectro >>>     '29:interspike_time_covar': 0.5343365146969321,
    pyelectro >>>     '29:max_interspike_time': 183.60000000000002,
    pyelectro >>>     '29:max_peak_no': 14,
    pyelectro >>>     '29:maximum': 61.370853,
    pyelectro >>>     '29:mean_spike_frequency': 14.458903347792235,
    pyelectro >>>     '29:min_interspike_time': 25.599999999999994,
    pyelectro >>>     '29:min_peak_no': 13,
    pyelectro >>>     '29:minimum': -61.187748,
    pyelectro >>>     '29:peak_decay_exponent': -0.15153571047754788,
    pyelectro >>>     '29:peak_linear_gradient': -0.007862338819211844,
    pyelectro >>>     '29:spike_broadening': 0.5870269764023183,
    pyelectro >>>     '29:spike_frequency_adaptation': -0.016055090374903703,
    pyelectro >>>     '29:spike_width_adaptation': 0.00754720476623759,
    pyelectro >>>     '29:trough_decay_exponent': -0.05167436400749407,
    pyelectro >>>     '29:trough_phase_adaptation': 0.006928896725677521}
    Going over acquisition # 30
    pyelectro >>> {   '30:average_last_1percent': -60.68766657511393,
    pyelectro >>>     '30:average_maximum': 48.13843,
    pyelectro >>>     '30:average_minimum': -29.626032,
    pyelectro >>>     '30:first_spike_time': 92.80000000000001,
    pyelectro >>>     '30:interspike_time_covar': 0.34247923269889735,
    pyelectro >>>     '30:max_interspike_time': 103.89999999999998,
    pyelectro >>>     '30:max_peak_no': 15,
    pyelectro >>>     '30:maximum': 61.767582,
    pyelectro >>>     '30:mean_spike_frequency': 15.688032272523532,
    pyelectro >>>     '30:min_interspike_time': 24.599999999999994,
    pyelectro >>>     '30:min_peak_no': 14,
    pyelectro >>>     '30:minimum': -61.248783,
    pyelectro >>>     '30:peak_decay_exponent': -0.19194537406993217,
    pyelectro >>>     '30:peak_linear_gradient': -0.006994401512067068,
    pyelectro >>>     '30:spike_broadening': 0.5770492253613585,
    pyelectro >>>     '30:spike_frequency_adaptation': -0.01538364588730742,
    pyelectro >>>     '30:spike_width_adaptation': 0.0069293286774622966,
    pyelectro >>>     '30:trough_decay_exponent': -0.04326627973106321,
    pyelectro >>>     '30:trough_phase_adaptation': 0.0064812799362231775}
    Going over acquisition # 31
    pyelectro >>> {   '31:average_last_1percent': -60.63212458292643,
    pyelectro >>>     '31:average_maximum': 48.024498,
    pyelectro >>>     '31:average_minimum': -29.024399,
    pyelectro >>>     '31:first_spike_time': 92.4,
    pyelectro >>>     '31:interspike_time_covar': 0.406847478416553,
    pyelectro >>>     '31:max_interspike_time': 133.5999999999999,
    pyelectro >>>     '31:max_peak_no': 15,
    pyelectro >>>     '31:maximum': 61.889652,
    pyelectro >>>     '31:mean_spike_frequency': 14.704337779644996,
    pyelectro >>>     '31:min_interspike_time': 29.5,
    pyelectro >>>     '31:min_peak_no': 14,
    pyelectro >>>     '31:minimum': -61.30982,
    pyelectro >>>     '31:peak_decay_exponent': -0.1671016453568311,
    pyelectro >>>     '31:peak_linear_gradient': -0.0086990196695813,
    pyelectro >>>     '31:spike_broadening': 0.5569432887124698,
    pyelectro >>>     '31:spike_frequency_adaptation': -0.014767300558908368,
    pyelectro >>>     '31:spike_width_adaptation': 0.006743383637276833,
    pyelectro >>>     '31:trough_decay_exponent': -0.04051025499900451,
    pyelectro >>>     '31:trough_phase_adaptation': 0.006526251236739548}
    Going over acquisition # 32
    pyelectro >>> {   '32:average_last_1percent': -59.76054255167643,
    pyelectro >>>     '32:average_maximum': 47.896324,
    pyelectro >>>     '32:average_minimum': -27.88653,
    pyelectro >>>     '32:first_spike_time': 91.30000000000001,
    pyelectro >>>     '32:interspike_time_covar': 0.34354448310799324,
    pyelectro >>>     '32:max_interspike_time': 106.60000000000002,
    pyelectro >>>     '32:max_peak_no': 15,
    pyelectro >>>     '32:maximum': 62.469486,
    pyelectro >>>     '32:mean_spike_frequency': 15.222355115798628,
    pyelectro >>>     '32:min_interspike_time': 30.700000000000003,
    pyelectro >>>     '32:min_peak_no': 14,
    pyelectro >>>     '32:minimum': -60.42481,
    pyelectro >>>     '32:peak_decay_exponent': -0.2052962133940038,
    pyelectro >>>     '32:peak_linear_gradient': -0.00818069814788547,
    pyelectro >>>     '32:spike_broadening': 0.5505132639070699,
    pyelectro >>>     '32:spike_frequency_adaptation': -0.0127147931736426,
    pyelectro >>>     '32:spike_width_adaptation': 0.006794945084249822,
    pyelectro >>>     '32:trough_decay_exponent': -0.045165955567810896,
    pyelectro >>>     '32:trough_phase_adaptation': 0.00648520248429949}
    Going over acquisition # 33
    pyelectro >>> {   '33:average_last_1percent': -59.76237360636393,
    pyelectro >>>     '33:average_maximum': 47.544212,
    pyelectro >>>     '33:average_minimum': -27.22403,
    pyelectro >>>     '33:first_spike_time': 91.4,
    pyelectro >>>     '33:interspike_time_covar': 0.9530683477414146,
    pyelectro >>>     '33:max_interspike_time': 322.6,
    pyelectro >>>     '33:max_peak_no': 14,
    pyelectro >>>     '33:maximum': 62.28638,
    pyelectro >>>     '33:mean_spike_frequency': 13.151239251390995,
    pyelectro >>>     '33:min_interspike_time': 29.0,
    pyelectro >>>     '33:min_peak_no': 13,
    pyelectro >>>     '33:minimum': -60.42481,
    pyelectro >>>     '33:peak_decay_exponent': -0.21944646857580366,
    pyelectro >>>     '33:peak_linear_gradient': -0.00986060329457358,
    pyelectro >>>     '33:spike_broadening': 0.5524597059029492,
    pyelectro >>>     '33:spike_frequency_adaptation': -0.01666625273183203,
    pyelectro >>>     '33:spike_width_adaptation': 0.006743589954469429,
    pyelectro >>>     '33:trough_decay_exponent': -0.03685620725832345,
    pyelectro >>>     '33:trough_phase_adaptation': -0.012225503917922204}
:::
:::

We now have the following information:

-   [`analysis_results`{.docutils .literal .notranslate}]{.pre}: the
    results of the analysis by PyElectro; we need these to set the
    target values for our fitting

-   [`currents`{.docutils .literal .notranslate}]{.pre}: the value of
    stimulation current for each sweep we've chosen; we need this for
    our models

-   [`memb_vals`{.docutils .literal .notranslate}]{.pre}: the time
    series of the membrane potentials and recordings times; we'll use
    this to plot the membrane potentials later to compare our fitted
    model against
:::

::: {#running-the-optimisation .section}
[]{#userdocs-optimising-running}

### Running the optimisation[\#](#running-the-optimisation "Link to this heading"){.headerlink}

To run the optimisation, we want to choose which of the 33 time series
we want to fit our model against. Ideally, we would want to fit our
model to all of them. Here, however, for simplicity and to keep the
computation time in check, we only pick two of the 33 sweeps. (As an
exercise, you can change the list to see how that affects your fitting.)

::: {.highlight-python .notranslate}
::: highlight
        sweeps_to_tune_against = [16,21]
        report = tune_izh_model(sweeps_to_tune_against, analysis_results, currents)
:::
:::

The Neurotune optimiser uses the evolutionary computation method
provided by the
[Inspyred](https://github.com/aarongarrett/inspyred/){.reference
.external} package. In short:

-   the evolutionary algorithm starts with a population of models, each
    with a random value for a set of parameters constrained by a max/min
    value we have supplied

-   it then calculates a fitness value for each model by comparing the
    features generated by the model to the target features that we
    provide

-   in each generation, it finds the fittest models (parents)

-   it mutates these to generate the next generation of models
    (offspring)

-   it replaces the least fit models with fittest of the new individuals

The idea is that by calculating the fittest parents and offspring, it
will find the candidate models that fit the provided target data best.
You can read more about evolutionary computation online (e.g.
[Wikipedia](https://en.wikipedia.org/wiki/Evolutionary_algorithm){.reference
.external}). More information on model fitting in computational
neuroscience can also be found in the literature. For example, see this
review
[\[[PBM04](Landing.html#id19 "Astrid A. Prinz, Dirk Bucher, and Eve Marder. Similar network activity from disparate circuit parameters. Nature Neuroscience, 7(12):1345–1352, 2004. doi:10.1038/nn1352."){.reference
.internal},
[RGF+11](Landing.html#id18 "Cyrille Rossant, Dan F. Goodman, Bertrand Fontaine, Jonathan Platkiewicz, Anna Magnusson, and Romain Brette. Fitting neuron models to spike trains. Frontiers in Neuroscience, 5:9, 2011. URL: https://www.frontiersin.org/article/10.3389/fnins.2011.00009, doi:10.3389/fnins.2011.00009."){.reference
.internal}\]]{#id2}.

Here, we follow the following steps:

-   we set up a template NeuroML model that will be passed to the
    optimiser

-   we list the parameters we want to fit, and provide the extents of
    their state spaces

-   we list the target features that the optimiser will use to calculate
    fitness, and set their weights

-   finally, we use the [`run_optimisation`{.docutils .literal
    .notranslate}]{.pre} function to run the optimisation

The [`tune_izh_model`{.docutils .literal .notranslate}]{.pre} function
shown below is the main workhorse function that does our fitting:

::: {.highlight-python .notranslate}
::: highlight
    def tune_izh_model(acq_list: List, metrics_from_data: Dict, currents: Dict) -> Dict:
        """Tune networks model against the data.

        Here we generate a network with the necessary number of Izhikevich cells,
        one for each current stimulus, and tune them against the experimental data.

        :param acq_list: list of indices of acquisitions/sweeps to tune against
        :type acq_list: list
        :param metrics_from_data: dictionary with the sweep number as index, and
            the dictionary containing metrics generated from the analysis
        :type metrics_from_data: dict
        :param currents: dictionary with sweep number as index and stimulus current
            value
        """

        # length of simulation of the cells---should match the length of the
        # experiment
        sim_time = 1500.0
        # Create a NeuroML template network simulation file that we will use for
        # the tuning
        template_doc = NeuroMLDocument(id="IzhTuneNet")
        # Add an Izhikevich cell with some parameters to the document
        template_doc.izhikevich2007_cells.append(
            Izhikevich2007Cell(
                id="Izh2007",
                C="100pF",
                v0="-60mV",
                k="0.7nS_per_mV",
                vr="-60mV",
                vt="-40mV",
                vpeak="35mV",
                a="0.03per_ms",
                b="-2nS",
                c="-50.0mV",
                d="100pA",
            )
        )
        template_doc.networks.append(Network(id="Network0"))
        # Add a cell for each acquisition list
        popsize = len(acq_list)
        template_doc.networks[0].populations.append(
            Population(id="Pop0", component="Izh2007", size=popsize)
        )

        # Add a current source for each cell, matching the currents that
        # were used in the experimental study.
        counter = 0
        for acq in acq_list:
            template_doc.pulse_generators.append(
                PulseGenerator(
                    id="Stim{}".format(counter),
                    delay="80ms",
                    duration="1000ms",
                    amplitude="{}pA".format(currents[acq]),
                )
            )
            template_doc.networks[0].explicit_inputs.append(
                ExplicitInput(
                    target="Pop0[{}]".format(counter), input="Stim{}".format(counter)
                )
            )
            counter = counter + 1

        # Print a summary
        print(template_doc.summary())

        # Write to a neuroml file and validate it.
        reference = "TuneIzhFergusonPyr3"
        template_filename = "{}.net.nml".format(reference)
        write_neuroml2_file(template_doc, template_filename, validate=True)

        # Now for the tuning bits

        # format is type:id/variable:id/units
        # supported types: cell/channel/izhikevich2007cell
        # supported variables:
        #  - channel: vShift
        #  - cell: channelDensity, vShift_channelDensity, channelDensityNernst,
        #  erev_id, erev_ion, specificCapacitance, resistivity
        #  - izhikevich2007Cell: all available attributes

        # we want to tune these parameters within these ranges
        # param: (min, max)
        parameters = {
            "izhikevich2007Cell:Izh2007/C/pF": (100, 300),
            "izhikevich2007Cell:Izh2007/k/nS_per_mV": (0.01, 2),
            "izhikevich2007Cell:Izh2007/vr/mV": (-70, -50),
            "izhikevich2007Cell:Izh2007/vt/mV": (-60, 0),
            "izhikevich2007Cell:Izh2007/vpeak/mV": (35, 70),
            "izhikevich2007Cell:Izh2007/a/per_ms": (0.001, 0.4),
            "izhikevich2007Cell:Izh2007/b/nS": (-10, 10),
            "izhikevich2007Cell:Izh2007/c/mV": (-65, -10),
            "izhikevich2007Cell:Izh2007/d/pA": (50, 500),
        }  # type: Dict[str, Tuple[float, float]]

        # Set up our target data and so on
        ctr = 0
        target_data = {}
        weights = {}
        for acq in acq_list:
            # data to fit to:
            # format: path/to/variable:metric
            # metric from pyelectro, for example:
            # https://pyelectro.readthedocs.io/en/latest/pyelectro.html?highlight=mean_spike_frequency#pyelectro.analysis.mean_spike_frequency
            mean_spike_frequency = "Pop0[{}]/v:mean_spike_frequency".format(ctr)
            average_last_1percent = "Pop0[{}]/v:average_last_1percent".format(ctr)
            first_spike_time = "Pop0[{}]/v:first_spike_time".format(ctr)

            # each metric can have an associated weight
            weights[mean_spike_frequency] = 1
            weights[average_last_1percent] = 1
            weights[first_spike_time] = 1

            # value of the target data from our data set
            target_data[mean_spike_frequency] = metrics_from_data[acq][
                "{}:mean_spike_frequency".format(acq)
            ]
            target_data[average_last_1percent] = metrics_from_data[acq][
                "{}:average_last_1percent".format(acq)
            ]
            target_data[first_spike_time] = metrics_from_data[acq][
                "{}:first_spike_time".format(acq)
            ]

            # only add these if the experimental data includes them
            # these are only generated for traces with spikes
            if "{}:average_maximum".format(acq) in metrics_from_data[acq]:
                average_maximum = "Pop0[{}]/v:average_maximum".format(ctr)
                weights[average_maximum] = 1
                target_data[average_maximum] = metrics_from_data[acq][
                    "{}:average_maximum".format(acq)
                ]
            if "{}:average_minimum".format(acq) in metrics_from_data[acq]:
                average_minimum = "Pop0[{}]/v:average_minimum".format(ctr)
                weights[average_minimum] = 1
                target_data[average_minimum] = metrics_from_data[acq][
                    "{}:average_minimum".format(acq)
                ]

            ctr = ctr + 1

        # simulator to use
        simulator = "jNeuroML"

        return run_optimisation(
            # Prefix for new files
            prefix="TuneIzh",
            # Name of the NeuroML template file
            neuroml_file=template_filename,
            # Name of the network
            target="Network0",
            # Parameters to be fitted
            parameters=list(parameters.keys()),
            # Our max and min constraints
            min_constraints=[v[0] for v in parameters.values()],
            max_constraints=[v[1] for v in parameters.values()],
            # Weights we set for parameters
            weights=weights,
            # The experimental metrics to fit to
            target_data=target_data,
            # Simulation time
            sim_time=sim_time,
            # EC parameters
            population_size=100,
            max_evaluations=500,
            num_selected=30,
            num_offspring=50,
            mutation_rate=0.9,
            num_elites=3,
            # Seed value
            seed=12345,
            # Simulator
            simulator=simulator,
            dt=0.025,
            show_plot_already='-nogui' not in sys.argv,
            save_to_file="fitted_izhikevich_fitness.png",
            save_to_file_scatter="fitted_izhikevich_scatter.png",
            save_to_file_hist="fitted_izhikevich_hist.png",
            save_to_file_output="fitted_izhikevich_output.png",
            num_parallel_evaluations=4,
        )
:::
:::

Let us walk through the different sections of this function.

::: {#writing-a-template-model .section}
[]{#userdocs-optimising-running-template}

#### Writing a template model[\#](#writing-a-template-model "Link to this heading"){.headerlink}

In this example, we want to fit the parameters of an [[Izhikevich
cell]{.std .std-ref}](Landing.html#schema-izhikevich2007cell){.reference
.internal} to our data such that simulating the cell then gives us
membrane potentials similar to those observed in the experiment.
Following the [[Izhikevich network example]{.std
.std-ref}](Landing.html#userdocs-getting-started-single-example){.reference
.internal}, we set up a template network with one Izhikevich cell for
each experimental recording that we want to fit. For each of these
cells, we provide a current stimulus matching the current used in the
current clamp experiments that we obtained our recordings from:

::: {.highlight-python .notranslate}
::: highlight
        # length of simulation of the cells---should match the length of the
        # experiment
        sim_time = 1500.0
        # Create a NeuroML template network simulation file that we will use for
        # the tuning
        template_doc = NeuroMLDocument(id="IzhTuneNet")
        # Add an Izhikevich cell with some parameters to the document
        template_doc.izhikevich2007_cells.append(
            Izhikevich2007Cell(
                id="Izh2007",
                C="100pF",
                v0="-60mV",
                k="0.7nS_per_mV",
                vr="-60mV",
                vt="-40mV",
                vpeak="35mV",
                a="0.03per_ms",
                b="-2nS",
                c="-50.0mV",
                d="100pA",
            )
        )
        template_doc.networks.append(Network(id="Network0"))
        # Add a cell for each acquisition list
        popsize = len(acq_list)
        template_doc.networks[0].populations.append(
            Population(id="Pop0", component="Izh2007", size=popsize)
        )

        # Add a current source for each cell, matching the currents that
        # were used in the experimental study.
        counter = 0
        for acq in acq_list:
            template_doc.pulse_generators.append(
                PulseGenerator(
                    id="Stim{}".format(counter),
                    delay="80ms",
                    duration="1000ms",
                    amplitude="{}pA".format(currents[acq]),
                )
            )
            template_doc.networks[0].explicit_inputs.append(
                ExplicitInput(
                    target="Pop0[{}]".format(counter), input="Stim{}".format(counter)
                )
            )
            counter = counter + 1

        # Print a summary
        print(template_doc.summary())

        # Write to a neuroml file and validate it.
        reference = "TuneIzhFergusonPyr3"
        template_filename = "{}.net.nml".format(reference)
        write_neuroml2_file(template_doc, template_filename, validate=True)
:::
:::

The resultant network template model for our two chosen recordings is
shown below:

::: {.highlight-xml .notranslate}
::: highlight
    <neuroml xmlns="http://www.neuroml.org/schema/neuroml2"  xmlns:xs="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.neuroml.org/schema/neuroml2 https://raw.github.com/NeuroML/NeuroML2/development/Schemas/NeuroML2/NeuroML_v2.2.xsd" id="IzhTuneNet">
        <izhikevich2007Cell id="Izh2007" C="100pF" v0="-60mV" k="0.7nS_per_mV" vr="-60mV" vt="-40mV" vpeak="35mV" a="0.03per_ms" b="-2nS" c="-50.0mV" d="100pA"/>
        <pulseGenerator id="Stim0" delay="80ms" duration="1000ms" amplitude="152.0pA"/>
        <pulseGenerator id="Stim1" delay="80ms" duration="1000ms" amplitude="202.0pA"/>
        <network id="Network0">
            <population id="Pop0" component="Izh2007" size="2"/>
            <explicitInput target="Pop0[0]" input="Stim0"/>
            <explicitInput target="Pop0[1]" input="Stim1"/>
        </network>
    </neuroml>
:::
:::

Please note that the initial parameters of the Izhikevich Cell do not
matter here because the optimiser will modify these to run the candidate
simulations.
:::

::: {#setting-up-the-optimisation-parameters .section}
[]{#userdocs-optimising-running-params}

#### Setting up the optimisation parameters[\#](#setting-up-the-optimisation-parameters "Link to this heading"){.headerlink}

The next step is to set the features/metrics that we want to fit:

The [`parameters`{.docutils .literal .notranslate}]{.pre} dictionary
contains the specifications of the parameters that we wish to fit, along
with their minimum and maximum permitted values.

::: {.highlight-python .notranslate}
::: highlight
        # we want to tune these parameters within these ranges
        # param: (min, max)
        parameters = {
            "izhikevich2007Cell:Izh2007/C/pF": (100, 300),
            "izhikevich2007Cell:Izh2007/k/nS_per_mV": (0.01, 2),
            "izhikevich2007Cell:Izh2007/vr/mV": (-70, -50),
            "izhikevich2007Cell:Izh2007/vt/mV": (-60, 0),
            "izhikevich2007Cell:Izh2007/vpeak/mV": (35, 70),
            "izhikevich2007Cell:Izh2007/a/per_ms": (0.001, 0.4),
            "izhikevich2007Cell:Izh2007/b/nS": (-10, 10),
            "izhikevich2007Cell:Izh2007/c/mV": (-65, -10),
            "izhikevich2007Cell:Izh2007/d/pA": (50, 500),
        }  # type: Dict[str, Tuple[float, float]]
:::
:::

The format of the parameter specification is:
[`ComponentType:ComponentID/VariableName[:VariableID]/Units`{.docutils
.literal .notranslate}]{.pre}. So, for example, to fit the Capacitance
of the Izhikevich cell, our parameter specification string is:
[`izhikevich2007Cell:Izh2007/C/pF`{.docutils .literal
.notranslate}]{.pre}.

All NeuroML [[Cell]{.std
.std-ref}](Landing.html#schema-cells){.reference .internal} and
[[Channel]{.std .std-ref}](Landing.html#schema-channels){.reference
.internal} ComponentTypes can be fitted using the
[`NeuroMLTuner`{.docutils .literal .notranslate}]{.pre}.

Next, we specify the target data that we want to fit against.

::: {.highlight-python .notranslate}
::: highlight
        # Set up our target data and so on
        ctr = 0
        target_data = {}
        weights = {}
        for acq in acq_list:
            # data to fit to:
            # format: path/to/variable:metric
            # metric from pyelectro, for example:
            # https://pyelectro.readthedocs.io/en/latest/pyelectro.html?highlight=mean_spike_frequency#pyelectro.analysis.mean_spike_frequency
            mean_spike_frequency = "Pop0[{}]/v:mean_spike_frequency".format(ctr)
            average_last_1percent = "Pop0[{}]/v:average_last_1percent".format(ctr)
            first_spike_time = "Pop0[{}]/v:first_spike_time".format(ctr)

            # each metric can have an associated weight
            weights[mean_spike_frequency] = 1
            weights[average_last_1percent] = 1
            weights[first_spike_time] = 1

            # value of the target data from our data set
            target_data[mean_spike_frequency] = metrics_from_data[acq][
                "{}:mean_spike_frequency".format(acq)
            ]
            target_data[average_last_1percent] = metrics_from_data[acq][
                "{}:average_last_1percent".format(acq)
            ]
            target_data[first_spike_time] = metrics_from_data[acq][
                "{}:first_spike_time".format(acq)
            ]

            # only add these if the experimental data includes them
            # these are only generated for traces with spikes
            if "{}:average_maximum".format(acq) in metrics_from_data[acq]:
                average_maximum = "Pop0[{}]/v:average_maximum".format(ctr)
                weights[average_maximum] = 1
                target_data[average_maximum] = metrics_from_data[acq][
                    "{}:average_maximum".format(acq)
                ]
            if "{}:average_minimum".format(acq) in metrics_from_data[acq]:
                average_minimum = "Pop0[{}]/v:average_minimum".format(ctr)
                weights[average_minimum] = 1
                target_data[average_minimum] = metrics_from_data[acq][
                    "{}:average_minimum".format(acq)
                ]

            ctr = ctr + 1
:::
:::

As we have set up a cell for each recording that we want to fit to, we
must also set the target value for each cell. We pick four features from
a subset of features that PyElectro provided us with:

-   [`mean_spike_frequency`{.docutils .literal .notranslate}]{.pre}

-   [`average_last_1percent`{.docutils .literal .notranslate}]{.pre}

-   [`average_maximum`{.docutils .literal .notranslate}]{.pre}

-   [`average_minimum`{.docutils .literal .notranslate}]{.pre}

The last two can only be calculated for membrane potential data that
includes spikes. Since a few of the experimental recordings to not show
any spikes, these two metrics will not be calculated for them. So, we
only add them for the corresponding cell only if they are present in the
features for the chosen recording.

The format for the [`target_data`{.docutils .literal
.notranslate}]{.pre} is similar to that of the [`parameters`{.docutils
.literal .notranslate}]{.pre}. The keys of the [`target_data`{.docutils
.literal .notranslate}]{.pre} dictionary are the specifications for the
metrics. The format for these is:
[`path/to/variable:pyelectro`{.docutils .literal
.notranslate}]{.pre}` `{.docutils .literal
.notranslate}[`metric`{.docutils .literal .notranslate}]{.pre}. You can
learn more about constructing paths in NeuroML [[here]{.std
.std-ref}](Landing.html#userdocs-paths){.reference .internal}. The value
for the each key is the corresponding metric that was calculated for us
by PyElectro (in [`analysis_results`{.docutils .literal
.notranslate}]{.pre}). The for loop will set the
[`target_data`{.docutils .literal .notranslate}]{.pre} to this (printed
by pyNeuroML when we run the script):

::: {.highlight-python .notranslate}
::: highlight
    target_data =  {
      'Pop0[0]/v:mean_spike_frequency': 7.033585370142431,
      'Pop0[0]/v:average_last_1percent': -60.84635798136393,
      'Pop0[0]/v:average_maximum': 58.73414,
      'Pop0[0]/v:average_minimum': -43.800358,
      'Pop0[1]/v:mean_spike_frequency': 10.8837614279495,
      'Pop0[1]/v:average_last_1percent': -60.380863189697266,
      'Pop0[1]/v:average_maximum': 54.52382,
      'Pop0[1]/v:average_minimum': -39.78882
    }
:::
:::

Similarly, we also set up the weights for each target metric in the
[`weights`{.docutils .literal .notranslate}]{.pre} variable:

::: {.highlight-python .notranslate}
::: highlight
    weights = {
     'Pop0[0]/v:mean_spike_frequency': 1,
     'Pop0[0]/v:average_last_1percent': 1,
     'Pop0[0]/v: average_maximum': 1,
     'Pop0[0]/v:average_minimum': 1,
     'Pop0[1]/v:mean_spike_frequency': 1,
     'Pop0[1]/v:average_last_1percent': 1,
     'Pop0[1]/v:average_maximum': 1,
     'Pop0[1]/v:average_minimum': 1
     }
:::
:::

For simplicity, we set the weights for all as [`1`{.docutils .literal
.notranslate}]{.pre} here.
:::

::: {#calling-the-optimisation-function .section}
[]{#userdocs-optimising-running-call}

#### Calling the optimisation function[\#](#calling-the-optimisation-function "Link to this heading"){.headerlink}

The last step is to call our [`run_optimisation`{.docutils .literal
.notranslate}]{.pre} function with the various parameters that we have
set up. Here, for simplicity, we use the [`jNeuroML`{.docutils .literal
.notranslate}]{.pre} simulator. For multi-compartmental models, however,
we will need to use the [`jNeuroML_NEURON`{.docutils .literal
.notranslate}]{.pre} simulator (since [`jNeuroML`{.docutils .literal
.notranslate}]{.pre} only supports single compartment simulations). A
number of arguments to the function are specific to evolutionary
computation, and their discussion is beyond the scope of this tutorial.

::: {.highlight-python .notranslate}
::: highlight
        # simulator to use
        simulator = "jNeuroML"

        return run_optimisation(
            # Prefix for new files
            prefix="TuneIzh",
            # Name of the NeuroML template file
            neuroml_file=template_filename,
            # Name of the network
            target="Network0",
            # Parameters to be fitted
            parameters=list(parameters.keys()),
            # Our max and min constraints
            min_constraints=[v[0] for v in parameters.values()],
            max_constraints=[v[1] for v in parameters.values()],
            # Weights we set for parameters
            weights=weights,
            # The experimental metrics to fit to
            target_data=target_data,
            # Simulation time
            sim_time=sim_time,
            # EC parameters
            population_size=100,
            max_evaluations=500,
            num_selected=30,
            num_offspring=50,
            mutation_rate=0.9,
            num_elites=3,
            # Seed value
            seed=12345,
            # Simulator
            simulator=simulator,
            dt=0.025,
            show_plot_already='-nogui' not in sys.argv,
            save_to_file="fitted_izhikevich_fitness.png",
            save_to_file_scatter="fitted_izhikevich_scatter.png",
            save_to_file_hist="fitted_izhikevich_hist.png",
            save_to_file_output="fitted_izhikevich_output.png",
            num_parallel_evaluations=4,
        )
:::
:::

The [`run_optimisation`{.docutils .literal .notranslate}]{.pre} function
will print out the optimisation report, and also return it so that it
can be stored in a variable for further use. The terminal output is
shown below:

::: {.highlight-console .notranslate}
::: highlight
    Ran 500 evaluations (pop: 100) in 582.205449 seconds (9.703424 mins total; 1.164411s per eval)

    ---------- Best candidate ------------------------------------------
    {   'Pop0[0]/v:average_last_1percent': -59.276969863333285,
        'Pop0[0]/v:average_maximum': 47.35760225,
        'Pop0[0]/v:average_minimum': -53.95061271428572,
        'Pop0[0]/v:first_spike_time': 170.1,
        'Pop0[0]/v:interspike_time_covar': 0.1330373936860586,
        'Pop0[0]/v:max_interspike_time': 190.57499999999982,
        'Pop0[0]/v:max_peak_no': 8,
        'Pop0[0]/v:maximum': 47.427714,
        'Pop0[0]/v:mean_spike_frequency': 6.957040276293886,
        'Pop0[0]/v:min_interspike_time': 135.25000000000003,
        'Pop0[0]/v:min_peak_no': 7,
        'Pop0[0]/v:minimum': -68.13577,
        'Pop0[0]/v:peak_decay_exponent': 0.0003379360943630205,
        'Pop0[0]/v:peak_linear_gradient': -3.270149536895308e-05,
        'Pop0[0]/v:spike_broadening': 0.982357731987536,
        'Pop0[0]/v:spike_frequency_adaptation': -0.016935379943933133,
        'Pop0[0]/v:spike_width_adaptation': 0.011971808793771004,
        'Pop0[0]/v:trough_decay_exponent': -0.0008421760726029059,
        'Pop0[0]/v:trough_phase_adaptation': -0.014231837120099502,
        'Pop0[1]/v:average_last_1percent': -59.28251401166662,
        'Pop0[1]/v:average_maximum': 47.242452454545464,
        'Pop0[1]/v:average_minimum': -48.287914,
        'Pop0[1]/v:first_spike_time': 146.7,
        'Pop0[1]/v:interspike_time_covar': 0.01075626702836981,
        'Pop0[1]/v:max_interspike_time': 91.67499999999998,
        'Pop0[1]/v:max_peak_no': 11,
        'Pop0[1]/v:maximum': 47.423363,
        'Pop0[1]/v:mean_spike_frequency': 10.973033769511423,
        'Pop0[1]/v:min_interspike_time': 88.20000000000002,
        'Pop0[1]/v:min_peak_no': 10,
        'Pop0[1]/v:minimum': -62.58064000000001,
        'Pop0[1]/v:peak_decay_exponent': 0.0008036004162568405,
        'Pop0[1]/v:peak_linear_gradient': -0.00012436953066659044,
        'Pop0[1]/v:spike_broadening': 0.9877761288704633,
        'Pop0[1]/v:spike_frequency_adaptation': 0.0064956079899488595,
        'Pop0[1]/v:spike_width_adaptation': 0.008982392557695507,
        'Pop0[1]/v:trough_decay_exponent': -0.004658690933014975,
        'Pop0[1]/v:trough_phase_adaptation': 0.009514671770845617}

    TARGETS:
    {   'Pop0[0]/v:average_last_1percent': -60.84635798136393,
        'Pop0[0]/v:average_maximum': 58.73414,
        'Pop0[0]/v:average_minimum': -43.800358,
        'Pop0[0]/v:mean_spike_frequency': 7.033585370142431,
        'Pop0[1]/v:average_last_1percent': -60.380863189697266,
        'Pop0[1]/v:average_maximum': 54.52382,
        'Pop0[1]/v:average_minimum': -39.78882,
        'Pop0[1]/v:mean_spike_frequency': 10.8837614279495}

    TUNED VALUES:
    {   'Pop0[0]/v:average_last_1percent': -59.276969863333285,
        'Pop0[0]/v:average_maximum': 47.35760225,
        'Pop0[0]/v:average_minimum': -53.95061271428572,
        'Pop0[0]/v:mean_spike_frequency': 6.957040276293886,
        'Pop0[1]/v:average_last_1percent': -59.28251401166662,
        'Pop0[1]/v:average_maximum': 47.242452454545464,
        'Pop0[1]/v:average_minimum': -48.287914,
        'Pop0[1]/v:mean_spike_frequency': 10.973033769511423}

    FITNESS: 0.003633

    FITTEST: {   'izhikevich2007Cell:Izh2007/C/pF': 240.6982897890555,
        'izhikevich2007Cell:Izh2007/a/per_ms': 0.03863507615280202,
        'izhikevich2007Cell:Izh2007/b/nS': 2.0112449831346746,
        'izhikevich2007Cell:Izh2007/c/mV': -43.069939785498356,
        'izhikevich2007Cell:Izh2007/d/pA': 212.50982499591083,
        'izhikevich2007Cell:Izh2007/k/nS_per_mV': 0.24113869560362797,
        'izhikevich2007Cell:Izh2007/vpeak/mV': 47.44063356996336,
        'izhikevich2007Cell:Izh2007/vr/mV': -59.283747806929135,
        'izhikevich2007Cell:Izh2007/vt/mV': -48.9131459978619}
:::
:::

It will also generate a number of plots (shown below):

-   showing the evolution of the parameters being fitted, with
    indications of the fitness value: larger circles mean more fitness

-   the change in the overall fitness value as the population evolves

-   distributions of the values of the parameters being fitted, with
    indications of the fitness value: darker lines mean higher fitness

<figure id="id6" class="align-default">
<img src="_images/fitted_izhikevich_scatter.png"
alt="Evolution of parameters." />
<figcaption><p><span class="caption-number">Fig. 48 </span><span
class="caption-text">The figure shows the values of various parameters
throughout the evolution, with larger circles having higher values of
fitness.</span><a href="#id6" class="headerlink"
title="Link to this image">#</a></p></figcaption>
</figure>

<figure id="id7" class="align-default">
<img src="_images/fitted_izhikevich_fitness.png"
alt="Evolution of fitness." />
<figcaption><p><span class="caption-number">Fig. 49 </span><span
class="caption-text">The figure shows the trend of the fitness
throughout the evolution.</span><a href="#id7" class="headerlink"
title="Link to this image">#</a></p></figcaption>
</figure>

<figure id="id8" class="align-default">
<img src="_images/fitted_izhikevich_hist.png"
alt="Histograms of values of fitting parameters." />
<figcaption><p><span class="caption-number">Fig. 50 </span><span
class="caption-text">The figure shows the distribution of values that
for each parameter throughout the evolution. Darker lines have higher
fitness values.</span><a href="#id8" class="headerlink"
title="Link to this image">#</a></p></figcaption>
</figure>
:::
:::

::: {#viewing-results .section}
[]{#userdocs-optimising-results}

### Viewing results[\#](#viewing-results "Link to this heading"){.headerlink}

The tuner also generates a plot with the membrane potential of a cell
using the fitted parameter values (shown on the top of the page). Here,
to document how the fitted parameters are to be extracted from the
output of the [`run_optimisation`{.docutils .literal
.notranslate}]{.pre} function, we also construct a model to use the
fitted parameters ourselves and plot the membrane potential to compare
it against the experimental data.

::: {#extracting-results-and-running-a-fitted-model .section}
[]{#userdocs-optimising-results-run-model}

#### Extracting results and running a fitted model[\#](#extracting-results-and-running-a-fitted-model "Link to this heading"){.headerlink}

This is done in the [`run_fitted_cell_simulation`{.docutils .literal
.notranslate}]{.pre} function:

::: {.highlight-python .notranslate}
::: highlight
    def run_fitted_cell_simulation(
        sweeps_to_tune_against: List, tuning_report: Dict, simulation_id: str
    ) -> None:
        """Run a simulation with the values obtained from the fitting

        :param tuning_report: tuning report from the optimser
        :type tuning_report: Dict
        :param simulation_id: text id of simulation
        :type simulation_id: str

        """
        # get the fittest variables
        fittest_vars = tuning_report["fittest vars"]
        C = str(fittest_vars["izhikevich2007Cell:Izh2007/C/pF"]) + "pF"
        k = str(fittest_vars["izhikevich2007Cell:Izh2007/k/nS_per_mV"]) + "nS_per_mV"
        vr = str(fittest_vars["izhikevich2007Cell:Izh2007/vr/mV"]) + "mV"
        vt = str(fittest_vars["izhikevich2007Cell:Izh2007/vt/mV"]) + "mV"
        vpeak = str(fittest_vars["izhikevich2007Cell:Izh2007/vpeak/mV"]) + "mV"
        a = str(fittest_vars["izhikevich2007Cell:Izh2007/a/per_ms"]) + "per_ms"
        b = str(fittest_vars["izhikevich2007Cell:Izh2007/b/nS"]) + "nS"
        c = str(fittest_vars["izhikevich2007Cell:Izh2007/c/mV"]) + "mV"
        d = str(fittest_vars["izhikevich2007Cell:Izh2007/d/pA"]) + "pA"

        # Create a simulation using our obtained parameters.
        # Note that the tuner generates a graph with the fitted values already, but
        # we want to keep a copy of our fitted cell also, so we'll create a NeuroML
        # Document ourselves also.
        sim_time = 1500.0
        simulation_doc = NeuroMLDocument(id="FittedNet")
        # Add an Izhikevich cell with some parameters to the document
        simulation_doc.izhikevich2007_cells.append(
            Izhikevich2007Cell(
                id="Izh2007",
                C=C,
                v0="-60mV",
                k=k,
                vr=vr,
                vt=vt,
                vpeak=vpeak,
                a=a,
                b=b,
                c=c,
                d=d,
            )
        )
        simulation_doc.networks.append(Network(id="Network0"))
        # Add a cell for each acquisition list
        popsize = len(sweeps_to_tune_against)
        simulation_doc.networks[0].populations.append(
            Population(id="Pop0", component="Izh2007", size=popsize)
        )

        # Add a current source for each cell, matching the currents that
        # were used in the experimental study.
        counter = 0
        for acq in sweeps_to_tune_against:
            simulation_doc.pulse_generators.append(
                PulseGenerator(
                    id="Stim{}".format(counter),
                    delay="80ms",
                    duration="1000ms",
                    amplitude="{}pA".format(currents[acq]),
                )
            )
            simulation_doc.networks[0].explicit_inputs.append(
                ExplicitInput(
                    target="Pop0[{}]".format(counter), input="Stim{}".format(counter)
                )
            )
            counter = counter + 1

        # Print a summary
        print(simulation_doc.summary())

        # Write to a neuroml file and validate it.
        reference = "FittedIzhFergusonPyr3"
        simulation_filename = "{}.net.nml".format(reference)
        write_neuroml2_file(simulation_doc, simulation_filename, validate=True)

        simulation = LEMSSimulation(
            sim_id=simulation_id,
            duration=sim_time,
            dt=0.1,
            target="Network0",
            simulation_seed=54321,
        )
        simulation.include_neuroml2_file(simulation_filename)
        simulation.create_output_file("output0", "{}.v.dat".format(simulation_id))
        counter = 0
        for acq in sweeps_to_tune_against:
            simulation.add_column_to_output_file(
                "output0", "Pop0[{}]".format(counter), "Pop0[{}]/v".format(counter)
            )
            counter = counter + 1
        simulation_file = simulation.save_to_file()
        # simulate
        run_lems_with_jneuroml(simulation_file, max_memory="2G", nogui=True, plot=False)
:::
:::

First, we extract the fitted parameters from the dictionary returned by
the [`run_optimisation`{.docutils .literal .notranslate}]{.pre}
function. Then, we use these parameters to set up a simple NeuroML
network and run a test simulation, recording the values of membrane
potentials generated by the cells. Please note that the current stimulus
to the cells in this test model must also match the values that were
used in the experiment, and so also in the fitting.
:::

::: {#plotting-model-generated-and-experimentally-recorded-membrane-potentials .section}
[]{#userdocs-optimising-results-plotting}

#### Plotting model generated and experimentally recorded membrane potentials[\#](#plotting-model-generated-and-experimentally-recorded-membrane-potentials "Link to this heading"){.headerlink}

Finally, in the [`plot_sim_data`{.docutils .literal .notranslate}]{.pre}
function, we plot the membrane potentials from our fitted cells and the
experimental data to see visually inspect the results of our fitting:

::: {.highlight-python .notranslate}
::: highlight
    def plot_sim_data(
        sweeps_to_tune_against: List, simulation_id: str, memb_pots: Dict
    ) -> None:
        """Plot data from our fitted simulation

        :param simulation_id: string id of simulation
        :type simulation_id: str
        """
        # Plot
        data_array = np.loadtxt("%s.v.dat" % simulation_id)

        # construct data for plotting
        counter = 0
        time_vals_list = []
        sim_v_list = []
        data_v_list = []
        data_t_list = []
        stim_vals = []
        for acq in sweeps_to_tune_against:
            stim_vals.append("{}pA".format(currents[acq]))

            # remains the same for all columns
            time_vals_list.append(data_array[:, 0] * 1000.0)
            sim_v_list.append(data_array[:, counter + 1] * 1000.0)

            data_v_list.append(memb_pots[acq][1])
            data_t_list.append(memb_pots[acq][0])

            counter = counter + 1

        # Model membrane potential plot
        generate_plot(
            xvalues=time_vals_list,
            yvalues=sim_v_list,
            labels=stim_vals,
            title="Membrane potential (model)",
            show_plot_already=False,
            save_figure_to="%s-model-v.png" % simulation_id,
            xaxis="time (ms)",
            yaxis="membrane potential (mV)",
        )
        # data membrane potential plot
        generate_plot(
            xvalues=data_t_list,
            yvalues=data_v_list,
            labels=stim_vals,
            title="Membrane potential (exp)",
            show_plot_already=False,
            save_figure_to="%s-exp-v.png" % simulation_id,
            xaxis="time (ms)",
            yaxis="membrane potential (mV)",
        )
:::
:::

This generates the following figures:

::: container-fluid
::: {.row .my-2 .py-2}
::: {.col-sm-6 .px-2}
<figure id="id9" class="align-default">
<img src="_images/fitted_izhikevich_sim-exp-v.png"
alt="Membrane potential from example experimental data." />
<figcaption><p><span class="caption-number">Fig. 51 </span><span
class="caption-text">Membrane potential from the experimental
data.</span><a href="#id9" class="headerlink"
title="Link to this image">#</a></p></figcaption>
</figure>
:::

::: {.col-sm-6 .px-2}
<figure id="id10" class="align-default">
<img src="_images/fitted_izhikevich_output.png"
alt="Membrane potential obtained from example fitted model." />
<figcaption><p><span class="caption-number">Fig. 52 </span><span
class="caption-text">Membrane potential obtained from the model with
highest fitness.</span><a href="#id10" class="headerlink"
title="Link to this image">#</a></p></figcaption>
</figure>
:::
:::
:::

We can clearly see the similarity between our fitted model and the
experimental data. A number of tweaks can be made to improve the
fitting. For example, pyNeuroML also provides a two staged optimisation
function:
[run_2stage_optimisation](https://pyneuroml.readthedocs.io/en/development/pyneuroml.tune.html#pyneuroml.tune.NeuroMLTuner.run_2stage_optimization){.reference
.external} that allows users to optimise sets of parameters in two
different stages. The graphs also show ranges of parameters that provide
fits, so users can also hand-tune their models further as required.
:::
:::
:::

[]{#document-Userdocs/TestingNeuroMLModels}

::: {#testing-validating-neuroml-models .section .tex2jax_ignore .mathjax_ignore}
[]{#userdocs-testing-validating-models}

## Testing/validating NeuroML Models[\#](#testing-validating-neuroml-models "Link to this heading"){.headerlink}

Models described in NeuroML can be run across multiple simulators, and
it it essential that the activity (e.g. spike times) of the models are
as close as possible across all of these independently developed
platforms.

It is also important to validate that the behaviour of a given NeuroML
model reproduces some recorded aspect of the biological equivalent.

::: {#testing-behaviour-of-neuroml-models-across-simulators .section}
[]{#userdocs-testing-models}

### Testing behaviour of NeuroML models across simulators[\#](#testing-behaviour-of-neuroml-models-across-simulators "Link to this heading"){.headerlink}

This type of testing addresses the question: **Does a given NeuroML
model produce the same results when run across multiple simulators?**

::: {#omv-open-source-brain-model-validation-framework .section}
[]{#userdocs-testing-models-omv}

#### OMV - Open Source Brain Model Validation framework[\#](#omv-open-source-brain-model-validation-framework "Link to this heading"){.headerlink}

The OSB Model Validation framework was originally developed as an
automated model validation package for [Open Source
Brain](http://www.opensourcebrain.org){.reference .external} projects,
which can be used for testing model behaviour across many [simulation
engines](https://github.com/OpenSourceBrain/osb-model-validation/tree/master/omv/engines){.reference
.external} both:

-   on your local machine when developing models

-   on [GitHub Actions](https://github.com/features/actions){.reference
    .external}, to ensure tests pass on every commit.

This framework has been used to test the 30+ NeuroML and PyNN models
described in the [Open Source Brain paper (Gleeson et al.
2019)](https://www.cell.com/neuron/fulltext/S0896-6273(19)30444-1){.reference
.external}, and [many
more](https://github.com/OpenSourceBrain/.github/blob/main/testsheet/README.md){.reference
.external}.

See
[OpenSourceBrain/osb-model-validation](https://github.com/OpenSourceBrain/osb-model-validation){.github
.reference .external} for more details.
:::
:::

::: {#validating-that-neuroml-model-reproduce-biological-activity .section}
[]{#userdocs-validating-models-bio}

### Validating that NeuroML model reproduce biological activity[\#](#validating-that-neuroml-model-reproduce-biological-activity "Link to this heading"){.headerlink}

This type of testing addresses the question: **How well does a given
NeuroML model replicate the activity as seen in real
neurons/channels/networks?**

::: {#sciunit-neuronunit .section}
#### SciUnit/NeuronUnit[\#](#sciunit-neuronunit "Link to this heading"){.headerlink}

[SciUnit](https://scidash.org/sciunit.html){.reference .external} is a
Python framework for test-driven validation of scientific models, and
[NeuronUnit](https://scidash.org/neuronunit.html){.reference .external}
is a package based on this for data-driven validation of neuron and ion
channel models. See also [SciDash](https://scidash.org/){.reference
.external} for more information.

Interactive Jupyter notebooks for running NeuronUnit examples can be
found [this
repository](https://github.com/scidash/neuronunit/tree/master/docs){.reference
.external}.

TODO: Add details on using
[SciUnit](https://scidash.org/sciunit.html){.reference .external} and
[NeuronUnit](https://scidash.org/neuronunit.html){.reference .external}
with NeuroML models.
:::
:::
:::

[]{#document-Userdocs/LEMS}

::: {#lems-low-entropy-model-specification .section .tex2jax_ignore .mathjax_ignore}
[]{#userdocs-lems}

## LEMS: Low Entropy Model Specification[\#](#lems-low-entropy-model-specification "Link to this heading"){.headerlink}

A language for specifying hierarchical models based on fundamental
physical relationships

::: {.dropdown .tip .admonition}
LEMS

For an in-depth guide to LEMS, please see the research paper: [LEMS: a
language for expressing complex biological models in concise and
hierarchical form and its use in underpinning NeuroML
2](https://www.frontiersin.org/articles/10.3389/fninf.2014.00079/full){.reference
.external}. Documentation on the structure of the LEMS language can be
found [[here]{.std
.std-ref}](Landing.html#userdocs-lemsschema){.reference .internal}.
:::

LEMS is being developed to provide a compact, minimally redundant,
human-readable, human-writable, declarative way of expressing models of
biological systems.

It differs from other systems such as CellML or SBML in its requirement
to be human writable and the inclusion of basic physical concepts such
as dimensionality and physical nesting as part of the language. The main
goal is to enable model developers to write declarative models in LEMS
in much the same way as software developers write software applications
in computer languages such as in C, Java or Python. The examples shown
here use XML for expressing models as text, but LEMS is not primarily an
XML language. Rather it defines a set of structures for representing
models. The reference implementation also supports a more concise
indentation-based format for representing models.

There are two independent implementations of LEMS: jLEMS, written in
Java and pyLEMS written in Python. Both are hosted on the
[github.com/LEMS](http://github.com/LEMS){.reference .external}.

::: {#capabilities .section}
[]{#userdocs-lems-capabilities}

### Capabilities[\#](#capabilities "Link to this heading"){.headerlink}

You can define ComponentTypes (e.g. a "HH channel" or "a bi-exponential
synapse") which express the general properties of a particular type of
thing that goes in a model. This includes saying what parameters they
have, what child elements they are allowed, and how they behave (the
equations).

You can then define Components based on these types by supplying values
for the parameters and adding any child elements that are required, so,
for example, a bi-exponential synapse model with rise time 1ms and decay
5ms would be a component.

ComponentTypes can extend other ComponentTypes to add extra parameters,
fix certain values, and otherwise modify their behavior. Components can
extend other Components to reuse specified parameter values. There is
also a loose notion of abstract types, so a component can accept
children with a particular lineage without needing to know exactly what
type they are. This can be used, for example, to define cells that
accept synaptic connections provided they have a particular signature.

Each ComponentType can have a Dynamics element that specifies how it
behaves: what the state variables are, the equations that govern them,
and what happens when events are sent or received. The interpreter takes
a model consisting of type and component elements referenced from a
network, builds an instance from them and runs it.

For those familiar with object oriented languages, the
ComponentType/Component distinction is close to the normal
Class/Instance distinction. When the model is run, the same pattern
applies again, with the Components acting as class definitions, with
their "instances" actually containing the state variables in the running
mode.
:::

::: {#background .section}
[]{#userdocs-lems-background}

### Background[\#](#background "Link to this heading"){.headerlink}

The March 2010 NeuroML meeting
([minutes](https://docs.neuroml.org/_static/files/NeuroMLWorkshop2010.pdf){.reference
.external}) identified a need to extend the capability within NeuroML
for expressing a range of models of synapses. It was decided that the
hitherto adopted approach of defining parameterized building blocks to
construct models by combining blocks and setting parameters was unlikely
to be flexible enough to cope with the needs for synapse models. This is
not obvious a-priori, since, for example, the pre NeuroML 2.0 ion
channel building blocks are fully adequate to describe the dynamics of a
wide range existing channel models. But there appears to be no such
commonality in models used for synapses, where the mechanisms used range
from highly detailed biochemical models to much more abstract ones.

This work also has antecedents in Catacomb 3, which was essentially a
GUI for a component definition system and model builder using a type
system similar to that proposed here. Much of the XML processing code
used in the interpreter was taken from PSICS which itself currently uses
the "building block" approach to model specification. The need for
user-defined types has been considered with respect to future PSICS
development, and this proposal also reflects potential requirements for
PSICS.
:::

::: {#example .section}
[]{#userdocs-lems-example}

### Example[\#](#example "Link to this heading"){.headerlink}

Here is the XML for a simple integrate-and-fire cell definition:

::: {.highlight-xml .notranslate}
::: highlight
     <ComponentType name="refractiaf">
         <Parameter name="threshold" dimension="voltage"/>
         <Parameter name="refractoryPeriod" dimension="time"/>
         <Parameter name="capacitance" dimension="capacitance"/>
         <Parameter name="vleak" dimension="voltage"/>
         <Parameter name="gleak" dimension="conductance"/>

         <Parameter name="current" dimension="current"/>
         <Parameter name="vreset" dimension="voltage"/>
         <Parameter name="deltaV" dimension="voltage"/>
         <Parameter name="v0" dimension="voltage"/>

         <EventPort name="out" direction="out"/>
         <EventPort name="in" direction="in"/>

         <Exposure name="v" dimension="voltage"/>

         <Dynamics>
             <StateVariable name="v" exposure="v" dimension="voltage" />   
             <StateVariable name="tin" dimension="time"/>
             <OnStart>
                 <StateAssignment variable="v" value="v0"/>
             </OnStart>

             <Regime name="refr">
                 <OnEntry>
                     <StateAssignment variable="tin" value="t" />
                     <StateAssignment variable="v" value="vreset" />
                 </OnEntry>
                 <OnCondition test="t .gt. tin + refractoryPeriod">
                     <Transition regime="int" />
                 </OnCondition>
             </Regime>

             <Regime name="int" initial="true">
                 <TimeDerivative variable="v" value="(current + gleak * (vleak - v)) / capacitance" />
                 <OnCondition test="v .gt. threshold">
                     <EventOut port="out" />
                     <Transition regime="refr" />
                 </OnCondition>

             </Regime>
         </Dynamics>

     </ComponentType>
:::
:::

Once this definition is available, a particular model using this
structure can be specified with the following XML:

::: {.highlight-xml .notranslate}
::: highlight
    <refractiaf threshold="-40mV" refractoryPeriod="5ms" capacitance="1nF" vleak="-80mV" gleak="100pS" vreset="-70mV" v0="-70mV" deltaV="10mV" />
:::
:::

More complex models will have nested components and other types of
parameters, but the basic principle of separating out the equations and
parameters for reusable model components, such that the equations are
only stated once, remains the same.
:::

::: {.toctree-wrapper .compound}
[]{#document-Userdocs/LEMSOverview}

::: {#model-structure-overview .section .tex2jax_ignore .mathjax_ignore}
[]{#userdocs-lemsoverview}

### Model structure overview[\#](#model-structure-overview "Link to this heading"){.headerlink}

Models are based on user-defined types (the term ComponentType is used
in the XML) that contain parameter declarations, reference declarations
and specification of what children an instance of a type can have.
Typically they also contain a Dynamics specification which can contain
build-time and run-time declarations. Build-time declarations apply when
a simulation is set up, for example to connect cells. Run-time
declarations specify the state variables, equations and events that are
involved.

An instance of a ComponentType is a model Component It specifies a
particular set of parameters for a given ComponentType. It says nothing
about state variables: in a simulation, typically many run-time
instances will correspond to a single model component definition, and
several model component definitions will use the same type. A run-time
instance holds its own set of state variables as defined by the Type
definition and a reference to its component for the parameter values
specific to that particular model component. The update rules come from
the type definition. As such, neither the ComponentType nor the
Component is properly a "prototype" for the runtime instance.

::: {#defining-componenttypes .section}
[]{#userdocs-lemsoverview-defining-component-types}

#### Defining ComponentTypes[\#](#defining-componenttypes "Link to this heading"){.headerlink}

ComponentTypes are declared as, for example:

::: {.highlight-xml .notranslate}
::: highlight
    <ComponentType name="myCell">
       <Parameter name="threshold" dimension="voltage" />
    </ComponentType>
:::
:::

A Component based on such a type is expressed as:

::: {.highlight-xml .notranslate}
::: highlight
    <Component type="myCell" threshold="dimensional_quantity" />
:::
:::

The quoted value for 'threshold' here is a rich quantity with size and
dimensions, typically consisting of a numerical value and a unit symbol.
Assignments like this are the only place unit symbols can occur.
Equations and expressions relate rich types, independent of any
particular unit system.

An equivalent way of writing the above in shorthand notation (using an
example of a string with size and dimension for threshold) is:

::: {.highlight-xml .notranslate}
::: highlight
    <myCell threshold="-30 mV" />
:::
:::

A type can contain elements for specifying the following aspects of the
structure and parameters of a model component:

-   Parameter - dimensional quantities that remain fixed within a model

-   Child - a required single sub-component of a given type

-   Children - variable number of sub-components of the given type

-   ComponentRef - a reference to a top-level component definition.

-   Link - a reference to a component definition relative to the
    referrer

-   Attachments - for build-time connections

-   EventPort - for run-time discrete event communication

-   Exposure - quantities that can be accessed from other components

-   Requirement - quantities that must be accessible to the component
    for it to make sense

-   DerivedParameter - like parameters, but derived from some other
    quantity in the model

The "EventPort" and "Attachments" declarations don't have any
corresponding elements in their model component specification. They only
affect how the component can be used when a model is instantiated.
EventPorts specify that a model can send or receive events, and should
match up with declarations in its Dynamics specification. An
"Attachments" declaration specifies that a run-time instance can have
dynamically generated attachments as, for example, when a new synapse
run-time instance is added to a cell for each incoming connection.
:::

::: {#inheritance .section}
[]{#userdocs-lemsoverview-inheritance}

#### Inheritance[\#](#inheritance "Link to this heading"){.headerlink}

A type can extend another type, adding new parameters, or supplying
values (SetParam) for inherited parameters. As well as reducing
duplication, the key application of this is with the Child and Children
declarations, where a type can specify that it needs a child or children
of a particular supertype, but doesn't care about which particular
sub-type is used in a model. This applies, for example, where a cell
requires synapses that compute a quantity with dimensions current, but
doesn't need access to any other parts of the synapse Dynamics.
:::

::: {#run-time-dynamics .section}
[]{#userdocs-lemsoverview-runtime-dynamics}

#### Run-time Dynamics[\#](#run-time-dynamics "Link to this heading"){.headerlink}

Run time Dynamics are included within a Dynamics block in a type
specification. They include declaration of:

-   state variables

-   first order differential equations with respect to time of state
    variables

-   derived quantities - things computed in terms of other local
    quantities or computed from other run-time instances

Run time Dynamics can be grouped into Regimes, where only one regime is
operative at a given time for a particular run-time instance. Regimes
have access to all the variables in the parent instance and can define
their own local variables.

Dynamics can also contain event blocks:

-   OnStart blocks contain any initialization declarations needed when a
    run-time state is instantiated

-   OnEvent blocks specify what happens when an event is received on a
    specified port

-   OnEntry blocks (only within regimes) specify things that should
    happen each time the system enters that regime.

-   OnCondition blocks have a test condition and specify what should
    happen when it is met.

Blocks may contain state variable assignments, event sending directives
and transition directives to indicate that the system should change from
one regime to another.
:::

::: {#build-time-structure .section}
[]{#userdocs-lemsoverview-buildtime-structure}

#### Build-time Structure[\#](#build-time-structure "Link to this heading"){.headerlink}

Build-time Structure defines the structure of a multi-component model.
Currently there are:

-   MultiInstantiate - for declaring that a component yields multiple
    run-time instances corresponding to a particular model component.
    Eg, for defining populations of cells.

-   ForEach - for iterating over multiple instances in the run-time
    structure

-   EventConnection - for connecting ports between run-time instances
:::

::: {#other .section}
[]{#userdocs-lemsoverview-other}

#### Other[\#](#other "Link to this heading"){.headerlink}

There are also Run, Show and Record Dynamics for creating type
definitions that define simulations and what should be recorded or
displayed from such a simulation.
:::

::: {#observations .section}
[]{#userdocs-lemsoverview-observations}

#### Observations[\#](#observations "Link to this heading"){.headerlink}

The numerous references to "run-time instances" above is problematic,
since the structures do not dictate any particular way of building a
simulator or running a model. In particular, there is no requirement
that a component or Dynamics declaration should give rise to any
particular collection of state variables that could be interpreted as a
run-time instance in the state of a simulator.

So, it is convenient to think of eventual state instances, and that is
indeed how the reference interpreter works, but the model specification
structure should avoid anything that is specific to this picture.
:::

::: {#type-specification-examples .section}
[]{#userdocs-lemsoverview-typespecific-examples}

#### Type specification examples[\#](#type-specification-examples "Link to this heading"){.headerlink}

Examples of type definitions using the various types of child element:

::: {.highlight-xml .notranslate}
::: highlight
    <ComponentType name="synapse">
       <EventPort direction="in" />
    </ComponentType>
:::
:::

says that instances of components using this type can receive events.

::: {.highlight-xml .notranslate}
::: highlight
    <ComponentType name="HHChannel">
       <Children name="gates" type="HHGate" />
    </ComponentType>
:::
:::

says that a HHChannel can have gates.

::: {.highlight-xml .notranslate}
::: highlight
    <ComponentType name="HHGate">
       <Child name="Forward" type="HHRate" />
       <Child name="Reverse" type="HHRate" />
    </ComponentType>
:::
:::

says that a HHGate has two children called Forward and Reverse, each of
type HHRate.

::: {.highlight-xml .notranslate}
::: highlight
    <ComponentType name="synapseCell">
       <Attachments name="synapses" type="synapse" />
    </ComponentType>
:::
:::

says that instances of components based on the synapseCell type can have
instances of component based on the synapse type attached to them at
build time.

::: {.highlight-xml .notranslate}
::: highlight
    <ComponentType name="Population">
       <ComponentRef name="component" type="Component" />
    </ComponentType>
:::
:::

says that components based on the Population type need a reference to a
component of type Component (ie, anything) (which would then be used as
the thing to be repeated in the population, but it doesn't say that
here).

::: {.highlight-xml .notranslate}
::: highlight
    <ComponentType name="EventConnectivity">
       <Link name="source" type="Population" />
    </ComponentType>
:::
:::

says that EventConnectivity components need a relative path to a local
component of type Population which will be accessed via the name
"source". The model component declarations corresponding to the channel
and gate types would be:

::: {.highlight-xml .notranslate}
::: highlight
    <Component type="HHChanne">
       <Component type="HHGate">
          <Component type="some_hh_gate_type" role="Forward" />
          <Component type="some_hh_gate_type" role="Reverse" />
       </Component>
    </Component>
:::
:::

or, in the shorthand notation:

::: {.highlight-xml .notranslate}
::: highlight
    <HHChannel>
       <HHGate>
          <Forward type="some_hh_gate_type" />
          <Reverse type="some_hh_gate_type" />
       </HHGate>
    </HHChannel>
:::
:::

For the population type it would be:

::: {.highlight-xml .notranslate}
::: highlight
    <Component id="myPopulation" type="population" component="myCellModel" />
:::
:::

And for the connections:

::: {.highlight-xml .notranslate}
::: highlight
    <Component type="EventConnectivity" source="myPopulation" />
:::
:::
:::
:::

[]{#document-Userdocs/LEMSExample1}

::: {#example-1-dimensions-units-componenttypes-and-components .section .tex2jax_ignore .mathjax_ignore}
[]{#userdocs-lemsexample1}

### Example 1: Dimensions, Units, ComponentTypes and Components[\#](#example-1-dimensions-units-componenttypes-and-components "Link to this heading"){.headerlink}

This page is structured as a walk-through of a single example explaining
the various elements as they occur.

::: {.highlight-xml .notranslate}
::: highlight
    <Lems>

        <Target component="sim1" />

        <Dimension name="voltage" m="1" l="2" t="-3" i="-1" />
        <Dimension name="time" t="1" />
        <Dimension name="per_time" t="-1" />
        <Dimension name="conductance" m="-1" l="-2" t="3" i="2" />
        <Dimension name="capacitance" m="-1" l="-2" t="4" i="2" />
        <Dimension name="current" i="1" />


        <ComponentType name="iaf1">
            <Parameter name="threshold" dimension="voltage" />
            <Parameter name="refractoryPeriod" dimension="time" />
            <Parameter name="capacitance" dimension="capacitance" />
        </ComponentType>


        <Unit symbol="mV" dimension="voltage" power="-3" />
        <Unit symbol="ms" dimension="time" power="-3" />
        <Unit symbol="pS" dimension="conductance" power="-12" />
        <Unit symbol="nS" dimension="conductance" power="-9" />
        <Unit symbol="uF" dimension="capacitance" power="-6" />
        <Unit symbol="nF" dimension="capacitance" power="-9" />
        <Unit symbol="pF" dimension="capacitance" power="-12" />
        <Unit symbol="per_ms" dimension="per_time" power="3" />
        <Unit symbol="pA" dimension="current" power="-12" />

        <iaf1 id="celltype_a" threshold="-30 mV" refractoryPeriod="2 ms" capacitance="3uF" />
        <!-- or -->
        <Component id="ctb" type="iaf1" threshold="-30 mV" refractoryPeriod="2 ms" capacitance="1uF" />


        <ComponentType name="iaf2" extends="iaf1">
            <Fixed parameter="threshold" value="-45mV" />
        </ComponentType>

        <ComponentType name="iaf3" extends="iaf1">
            <Parameter name="leakConductance" dimension="conductance" />
            <Parameter name="leakReversal" dimension="voltage" />
            <Parameter name="deltaV" dimension="voltage" />

            <EventPort name="spikes-in" direction="in" />
            <Exposure name="v" dimension="voltage" />

            <Dynamics>
                <StateVariable name="v" exposure="v" dimension="voltage" />
                <TimeDerivative variable="v" value="leakConductance * (leakReversal - v) / capacitance" />

                <OnEvent port="spikes-in">
                    <StateAssignment variable="v" value="v + deltaV" />
                </OnEvent>
            </Dynamics>

        </ComponentType>


        <ComponentType name="spikeGenerator">
            <Parameter name="period" dimension="time" />
            <EventPort name="a" direction="out" />
            <Exposure name="tsince" dimension="time" />
            <Dynamics>
                <StateVariable name="tsince" exposure="tsince" dimension="time" />
                <TimeDerivative variable="tsince" value="1" />
                <OnCondition test="tsince .gt. period">
                    <StateAssignment variable="tsince" value="0" />
                    <EventOut port="a" />
                </OnCondition>
            </Dynamics>
        </ComponentType>


        <ComponentType name="spikeGenerator2" extends="spikeGenerator">
            <Dynamics>
                <StateVariable name="tlast" dimension="time" />
                <DerivedVariable name="tsince" dimension="time" exposure="tsince" value="t - tlast" />
                <OnCondition test="t - tlast .gt. period">
                    <StateAssignment variable="tlast" value="t" />
                    <EventOut port="a" />
                </OnCondition>
            </Dynamics>
        </ComponentType>


        <ComponentType name="HHRate">
            <Parameter name="rate" dimension="per_time" />
            <Parameter name="midpoint" dimension="voltage" />
            <Parameter name="scale" dimension="voltage" />
            <Requirement name="v" dimension="voltage" />
            <Exposure name="r" dimension="per_time" />
        </ComponentType>


        <ComponentType name="HHExpRate" extends="HHRate">
            <Dynamics>
                <DerivedVariable name="r" dimension="per_time" exposure="r" value="rate * exp((v - midpoint)/scale)" />
            </Dynamics>
        </ComponentType>


        <ComponentType name="HHSigmoidRate" extends="HHRate">
            <Dynamics>
                <DerivedVariable name="r" dimension="per_time" exposure="r" value="rate / (1 + exp( -(v - midpoint)/scale))" />
            </Dynamics>
        </ComponentType>


        <ComponentType name="HHExpLinearRate" extends="HHRate">
            <Dynamics>
                <DerivedVariable name="x" dimension="none" value="(v - midpoint) / scale" />
                <DerivedVariable name="r" dimension="per_time" exposure="r" value="rate * x / (1 - exp(-x))" />
            </Dynamics>
        </ComponentType>


        <ComponentType name="HHGate0">
            <Parameter name="power" dimension="none" />
            <Child name="Forward" type="HHRate" />
            <Child name="Reverse" type="HHRate" />
            <Requirement name="v" dimension="voltage" />
            <Exposure name="fcond" dimension="none" />
            <Dynamics>
                <StateVariable name="q" dimension="none" />
                <DerivedVariable dimension="per_time" name="rf" select="Forward/r" />
                <DerivedVariable dimension="per_time" name="rr" select="Reverse/r" />
                <TimeDerivative variable="q" value="rf * (1 - q) - rr * q" />
                <DerivedVariable name="fcond" dimension="none" exposure="fcond" value="q^power" />
            </Dynamics>
        </ComponentType>


        <ComponentType name="HHGate">
            <Parameter name="power" dimension="none" />
            <Child name="Forward" type="HHRate" />
            <Child name="Reverse" type="HHRate" />
            <Requirement name="v" dimension="voltage" />
            <Exposure name="fcond" dimension="none" />
            <Dynamics>
                <StateVariable name="x" dimension="none" />
                <DerivedVariable name="ex" dimension="none" value="exp(x)" />
                <DerivedVariable name="q" dimension="none" value="ex / (1 + ex)" />
                <DerivedVariable name="rf" dimension="per_time" select="Forward/r" />
                <DerivedVariable name="rr" dimension="per_time" select="Reverse/r" />
                <TimeDerivative variable="x" value="(1 + ex)^2 / ex * (rf * (1 - q) - rr * q)" />
                <DerivedVariable name="fcond" dimension="none" exposure="fcond" value="q^power" />
            </Dynamics>
        </ComponentType>


        <ComponentType name="HHChannel">
            <Parameter name="conductance" dimension="conductance" />
            <Children name="gates" type="HHGate" />
            <Exposure name="g" dimension="conductance" />
            <Dynamics>
                <DerivedVariable name="gatefeff" dimension="none"
                                 select="gates[*]/fcond" reduce="multiply" />
                <DerivedVariable name="g" exposure="g"
                                 dimension="conductance" value="conductance * gatefeff" />
            </Dynamics>
        </ComponentType>


        <HHChannel id="na" conductance="20pS">
            <HHGate id="m" power="3">
                <Forward type="HHExpLinearRate" rate="1.per_ms"
                         midpoint="-40mV" scale="10mV" />
                <Reverse type="HHExpRate" rate="4per_ms" midpoint="-65mV"
                         scale="-18mV" />
            </HHGate>
            <HHGate id="h" power="1">
                <Forward type="HHExpRate" rate="0.07per_ms"
                         midpoint="-65.mV" scale="-20.mV" />
                <Reverse type="HHSigmoidRate" rate="1per_ms"
                         midpoint="-35mV" scale="10mV" />
            </HHGate>
        </HHChannel>


        <HHChannel id="k" conductance="20pS">
            <HHGate id="n" power="4">
                <Forward type="HHExpLinearRate" rate="0.1per_ms"
                         midpoint="-55mV" scale="10mV" />
                <Reverse type="HHExpRate" rate="0.125per_ms"
                         midpoint="-65mV" scale="-80mV" />
            </HHGate>
        </HHChannel>


        <ComponentType name="ChannelPopulation">
            <ComponentReference name="channel" type="HHChannel" />
            <Parameter name="number" dimension="none" />
            <Parameter name="erev" dimension="voltage" />
            <Requirement name="v" dimension="voltage" />
            <Exposure name="current" dimension="current" />

            <Dynamics>
                <DerivedVariable name="channelg" dimension="conductance" select="channel/g" />
                <DerivedVariable name="geff" dimension="conductance" value="channelg * number" />
                <DerivedVariable name="current" dimension="current" exposure="current" value="geff * (erev - v)" />
            </Dynamics>

            <Structure>
                <ChildInstance component="channel" />
            </Structure>
        </ComponentType>



        <ComponentType name="HHCell">
            <Parameter name="capacitance" dimension="capacitance" />
            <Children name="populations" type="ChannelPopulation" />
            <Parameter name="injection" dimension="current" />
            <Parameter name="v0" dimension="voltage" />
            <Exposure name="v" dimension="voltage" />
            <Dynamics>
                <OnStart>
                    <StateAssignment variable="v" value="v0" />
                </OnStart>

                <DerivedVariable name="totcurrent"
                                 dimension="current" select="populations[*]/current"
                                 reduce="add" />
                <StateVariable name="v" exposure="v"
                               dimension="voltage" />
                <TimeDerivative variable="v"
                                value="(totcurrent + injection) / capacitance" />
            </Dynamics>
        </ComponentType>



        <HHCell id="hhcell_1" capacitance="1pF" injection="4pA" v0="-60mV">
            <ChannelPopulation channel="na" number="6000"  erev="50mV" />
            <ChannelPopulation channel="k" number="1800" erev="-77mV" />
        </HHCell>


        <Component id="celltype_c" type="iaf3" leakConductance="3 pS" refractoryPeriod="3 ms" threshold="45 mV" leakReversal="-50 mV" deltaV="5mV" capacitance="1uF" />

        <Component id="gen1" type="spikeGenerator" period="30ms" />

        <Component id="gen2" type="spikeGenerator2" period="32ms" />

        <Component id="iaf3cpt" type="iaf3" leakReversal="-50mV" deltaV="50mV" threshold="-30mV" leakConductance="50pS" refractoryPeriod="4ms" capacitance="1pF" />


        <Include file="SimpleNetwork.xml" />


        <Network id="net1">
            <Population id="p1" component="gen1" size="1" />
            <Population id="p2" component="gen2" size="1" />
            <Population id="p3" component="iaf3cpt" size="1" />

            <Population id="hhpop" component="hhcell_1" size="1" />

            <EventConnectivity id="p1-p3" source="p1" target="p3">
                <Connections type="AllAll" />
            </EventConnectivity>
        </Network>


        <Include file="SingleSimulation.xml" />
        
        <Simulation id="sim1" length="80ms" step="0.01ms" target="net1">
            <Display id="d0" title="Example 1: Dimensions, Units, ComponentTypes and Components" 
                     timeScale="1ms" xmin="-10" xmax="90" ymin="-90" ymax="60">
                <Line id="tsince" quantity="p1[0]/tsince" scale="1ms" timeScale="1ms" color="#00c000" />
                <Line id="p3v" quantity="p3[0]/v" scale="1mV" timeScale="1ms" color="#0000f0" />
                <Line id="p0v" quantity="hhpop[0]/v" scale="1mV" timeScale="1ms" color="#ff4040" />
            </Display>
        </Simulation>

    </Lems>
:::
:::

The whole model is wrapped in a block which, for now, is called "Lems"
(Low Entropy Model Specification). Then we define the dimensions that
will be used in this model. Typically these would be loaded from an
external file along with various other stuff, not repeated in each
model, but it is included here in the interests of having a single file
for everything.

::: {.highlight-xml .notranslate}
::: highlight
    <Dimension name="voltage" m="1" l="2" t="3" i="-1" />
    <Dimension name="time" t="1" />
    <Dimension name="per_time" t="-1" />
    <Dimension name="conductance" m="-1" l="-2" t="3" i="2" />
    <Dimension name="capacitance" m="-1" l="-2" t="4" i="2" />
    <Dimension name="current" i="1" />
:::
:::

Each dimension element just associates a dimension name with the
exponents for mass, length, time and current.

At this stage, one can begin defining component types. This is done with
the ComponentType element and child Parameter elements. A simple cell
model with three parameters could be defined as:

::: {.highlight-xml .notranslate}
::: highlight
    <ComponentType name="cell1">     
       <Parameter name="threshold" dimension="voltage" />     
       <Parameter name="refractoryPeriod" dimension="time" />     
       <Parameter name="capacitance" dimension="capacitance" />    
    </ComponentType>
:::
:::

Each of the Parameter elements defines a parameter that should be
supplied when a component is defined based on this type. Before we can
define a component though, we need some units to use in setting those
values.

Defining a unit involves supplying the symbol, dimension and the power
of ten by which it is scaled from the IS base unit. Note that units have
a symbol, not a name. This is because they occur as a component of an
assignment expression such as 'threshold="-45mV"' not as a reference
such as 'dimension="voltage"'. In general, where one component refers to
another, then the attribute value is the name of the thing being
referred to, and the attribute name is the lower case version of the
type of the thing being referred to. Thus when a dimension is declared
with \<Dimension name="voltage".../\> then it is referred to from a
Parameter as . This holds for all references to components of a
particular type.

Returning to the units, this model will use the following, which
normally would also be loaded from an external file of standard
settings.

::: {.highlight-xml .notranslate}
::: highlight
    <Unit symbol="mV" dimension="voltage" powTen="-3" />
    <Unit symbol="ms" dimension="time" powTen="-3" />
    <Unit symbol="pS" dimension="conductance" powTen="-12" />
    <Unit symbol="nS" dimension="conductance" powTen="-9" />
    <Unit symbol="uF" dimension="capacitance" powTen="-6" />
    <Unit symbol="nF" dimension="capacitance" powTen="-9" />
    <Unit symbol="pF" dimension="capacitance" powTen="-12" />
    <Unit symbol="per_ms" dimension="per_time" powTen="3" />
    <Unit symbol="pA" dimension="current" powTen="-12" />
:::
:::

Once the units are available it is possible to define a component. There
are two equivalent ways of doing this: either by using the Component
element and setting its type, or by using the type as a new XML element.
The latter may be a little more readable, but for a simple component
like this it doesn't make much difference. For more complicated
components with nested children though, the second form is definitely
clearer (eg see the HHChannel examples later).

::: {.highlight-xml .notranslate}
::: highlight
    <Component id="ctb" type="cell1" threshold="-30 mV" refractoryPeriod="2 ms" capacitance="1uF" />
    <cell1 id="celltype_a" threshold="-30 mV" refractoryPeriod="2 ms" capacitance="3uF" />
:::
:::

In specifying a component, a value must be supplied for each of the
parameters defined in the corresponding type. The value is composed of a
number and a unit. It can't include expressions with multiple units for
the values so, for example, to express an acceleration you couldn't
write "3 m s\^-2". Instead would need to define a unit element for the
compound unit (and a dimension element for acceleration) and use that.

Specifying all the parameters for each component can lead to
duplication. Suppose, for example, you want to build a range of cell
models all based on cell1, but you don't want to change the threshold.
You could define a new type without the threshold, but it is neater to
still use the same type but specify that you are restricting attention
to the set that all have a particular value for the threshold. This can
be done by creating a new type that extends the cell1 type and includes
a Fixed element to fix the threshold:

::: {.highlight-xml .notranslate}
::: highlight
    <ComponentType name="cell2" extends="cell1">     
       <Fixed parameter="threshold" value="-45mV" />
    </ComponentType>
:::
:::

The cell2 type can now be used by only setting the remaining two
parameters.

As well as restricting types when you extend them, you can also add new
parameters as shown in the next type. This also introduces an EventPort,
to indicate that instances of components built from this type can
receive events, and, finally, a Dynamics block. This is where the
Dynamics of instances of the component can be specified. The phrase
"instances of the component" is intentional. The type itself doesn't
"behave": it is just a definition. A component built from the type
doesn't "behave" either: it is just a set of parameter values linked
back to the type. The thing that "behaves" is an instance in a runnable
model that actually contains state variables. In general, many
components may be based on one type, and one component may give rise to
many instances in a running model.

Here is a basic capacitative cell with a leaking potential and a simple
event handler.

::: {.highlight-xml .notranslate}
::: highlight
    <ComponentType name="cell3" extends="cell1">     
       <Parameter name="leakConductance" dimension="conductance" />     
       <Parameter name="leakReversal" dimension="voltage" />     
       <Parameter name="deltaV" dimension="voltage" />     
       <EventPort name="spikes-in" direction="in" />     
       <Exposure name="v" dimension="voltage" />     
       <Dynamics>        
          <StateVariable name="v" exposure="v" dimension="voltage" />        
          <TimeDerivative variable="v" value="leakConductance * (leakReversal - v) / capacitance" />        
          <OnEvent port="spikes-in">             
             <StateAssignment variable="v" value="v + deltaV" />        
          </OnEvent>     
       </Dynamics>
    </ComponentType>
:::
:::

The Dynamics involves a single state variable, a voltage called "v", and
one equation, expressing how v drifts towards the leak reversal
potential. The event block specifies what happens when an instance
receives an event. In this case the state variable v is bumped up by
deltaV. The value attribute in the TimeDerivative element is an
expression involving the parameters and the state variables. It gives
the right hand side of a first order differential equation dv/dt =
(...). Expressions follow normal operator precedence rules with "\^" for
general powers and exp(x) for exponentials.

Below is another example of a Dynamics, this time with an output port
and a condition testing block that sends an event when the condition
becomes true. The test attribute in the OnCondition element is a boolean
valued expression. These use Fortran style operators (.gt. and .lt. for
\> and \<) to avoid confusion with xml angle brackets.

::: {.highlight-xml .notranslate}
::: highlight
    <ComponentType name="spikeGenerator">     
       <Parameter name="period" dimension="time" />     
       <EventPort name="a" direction="out" />     
       <Exposure name="tsince" dimension="time" />     
       <Dynamics>         
          <StateVariable name="tsince" exposure="tsince" dimension="time" />         
          <TimeDerivative variable="tsince" value="1" />         
          <OnCondition test="tsince .gt. period">             
             <StateAssignment variable="tsince" value="0" />             
             <EventOut port="a" />         
          </OnCondition>     
       </Dynamics>
    </ComponentType>
:::
:::

The above model is one way of writing a regular event generator. It has
a state variable that grows in sync with t until it reaches a threshold
when the event fires and it is reset. The model below achieves the same
effect without solving a differential equation. Instead, it asks for
access to the global time variable ("t" is the one global variable that
is always available) and uses that in the test condition. \[aside -
there's a slight problem here since t exists even if the model doesn't
define a dimension called time\].

::: {.highlight-xml .notranslate}
::: highlight
    <ComponentType name="spikeGenerator2" extends="spikeGenerator">     
       <Dynamics>         
          <GlobalVariable name="t" dimension="time" />         
          <StateVariable name="tlast" dimension="time" />         
          <DerivedVariable name="tsince" exposure="tsince" value="t - tlast" />         
          <OnCondition test="t - tlast .gt. period">             
             <StateAssignment variable="tlast" value="t" />             
             <EventOut port="a" />         
          </OnCondition>     
       </Dynamics>
    </ComponentType>
:::
:::

The examples so far have all been of very simple components which just
had a single set of parameters. Real models however require rather more
structure than this with components having children of various types and
possibly multiple children of certain types. To illustrate this, the
next example shows how the concept of an ion channel using
Hodgkin-Huxley Dynamics can be defined.

Starting from the bottom, we define the different types of rate
equations that occur. These will supply terms in the equations for the
derivatives of the gating particles. There are three different
expressions used in the HH equations, but they can all be expressed with
three parameters, rate, midpoint and scale. We first define a general
rate class, and then extend it for the three cases.

The HHRate Dynamics shows two new constructs. An Exposure declares that
the component makes a quantity available to other components. A
Requirement specifies that the component needs to know about a variable
that it doesn't define itself. When it is used in a model, the specified
variable must be available (and have the right dimension) in the parent
component or one of its more remote ancestors.

Note that the general HHRate class defines an Exposure without a
Dynamics block to actually set its value. This is analogous to an
abstract class in java: you can't actually make a component out of the
HHRate element directly (the interpreter will complain) but any
component using a HHRate will know it has an exposed variable called
"r". The types that extend HHRate have to supply a value for "r" before
they are fully defined and ready to be used.

So here is the basic HHRate and its three extensions:

::: {.highlight-xml .notranslate}
::: highlight
    <ComponentType name="HHRate">     
       <Parameter name="rate" dimension="per_time" />     
       <Parameter name="midpoint" dimension="voltage" />     
       <Parameter name="scale" dimension="voltage" />     
       <Exposure name="r" dimension="per_time" />    
    </ComponentType>
    <ComponentType name="HHExpRate" extends="HHRate">     
       <Dynamics>         
          <DerivedVariable name="r" exposure="r" value="rate * exp((v - midpoint)/scale)" />     
       </Dynamics>
    </ComponentType>
    <ComponentType name="HHSigmoidRate" extends="HHRate">     
       <Dynamics>         
          <DerivedVariable name="r" exposure="r" value="rate / (1 + exp(0 - (v - midpoint)/scale))" />     
       </Dynamics>
    </ComponentType>
    <ComponentType name="HHExpLinearRate" extends="HHRate">     
       <Dynamics>         
          <DerivedVariable name="x" value="(v - midpoint) / scale" />         
          <DerivedVariable name="r" exposure="r" value="rate * x / (1 - exp(0 - x))" />     
       </Dynamics>
    </ComponentType>
:::
:::

Now the rate elements are available, they can be used to define a
component for a gate in a HH model. This introduces the Child element
which says that components built using this type must include a
subcomponent of the specified type. A HH gate needs subcomponents for
the forward and reverse rates.

::: {.highlight-xml .notranslate}
::: highlight
    <ComponentType name="HHGate0">     
       <Parameter name="power" dimension="none" />      
       <Child name="Forward" type="HHRate" />     
       <Child name="Reverse" type="HHRate" />     
       <Exposure name="fcond" dimension="none" />     
       <Requirement name="v" dimension="voltage" />     
       <Dynamics>         
          <StateVariable name="q" dimension="none" />         
          <DerivedVariable name="rf" select="Forward/r" />         
          <DerivedVariable name="rr" select="Reverse/r" />          
          <TimeDerivative variable="q" value="rf * (1 - q) - rr * q" />         
          <DerivedVariable name="fcond" exposure="fcond" value="q^power" />     
       </Dynamics>
    </ComponentType>
:::
:::

The above is a perfectly reasonable way to define a HH gate but
unfortunately it needs smarter numerics than the simple forward Euler
rule used in the proof of concept interpreter. Running this model with
the Euler method leads to numerical instabilities. Happily, this problem
can be circumvented without improving the numerics by changing the state
variable. Instead of q which is defined on \[0, 1\] you can use x
defined on (-infinity, infinity) which works much better with a naive
integration scheme. This is what it looks like with x instead of q:

::: {.highlight-xml .notranslate}
::: highlight
    <ComponentType name="HHGate">     
       <Parameter name="power" dimension="none" />      
       <Child name="Forward" type="HHRate" />     
       <Child name="Reverse" type="HHRate" />     
       <Exposure name="fcond" dimension="none" />     
       <Requirement name="v" dimension="voltage" />     
       <Dynamics>         
          <StateVariable name="x" dimension="none" />         
          <DerivedVariable name="ex" dimension="none" value="exp(x)" />         
          <DerivedVariable name="q" dimension="none" value="ex / (1 + ex)" />         
          <DerivedVariable name="rf" select="Forward/r" />         
          <DerivedVariable name="rr" select="Reverse/r" />          
          <TimeDerivative variable="x" value="(1 + ex)^2 / ex * (rf * (1 - q) - rr * q)" />         
          <DerivedVariable name="fcond" exposure="fcond" value="q^power" />     
       </Dynamics>
    </ComponentType>
:::
:::

Now the gate type has been defined, it can be used to say what a HH
Channel actually is. In this picture, a channel just has a conductance
and one or more gates:

::: {.highlight-xml .notranslate}
::: highlight
    <ComponentType name="HHChannel">     
       <Parameter name="conductance" dimension="conductance" />     
       <Children name="gates" type="HHGate" min="0" max="4" />     
       <Requirement name="v" dimension="voltage" />     
       <Exposure name="g" dimension="conductance" />          
       <Dynamics>                  
          <DerivedVariable name="gatefeff" select="gates[*]/fcond" reduce="multiply" />         
          <DerivedVariable name="g" exposure="g" value="conductance * gatefeff" />     
       </Dynamics>
    </ComponentType>
:::
:::

This introduces one new construct, the Children element, that allows for
an indeterminate number of children of a given type. This means that the
same type can be used for potassium channels with only one gate, sodium
channels with two gates or indeed other channels with more gates. The
first derived variable in the Dynamics block uses a xpath-style
selection function to process the indeterminate number of children. In
this case it computes the produce of the fcond variables from the
different gates.

With these definitions in place, it is now possible to define some
channel models. The classic Hodgkin-Huxley sodium channel can be
represented as:

::: {.highlight-xml .notranslate}
::: highlight
    <HHChannel id="na" conductance="20pS">     
       <HHGate id="m" power="3">         
          <Forward type="HHExpLinearRate" rate="1.per_ms" midpoint="-40mV" scale="10mV" />         
          <Reverse type="HHExpRate" rate="4per_ms" midpoint="-65mV" scale="-18mV" />     
       </HHGate>          
       <HHGate id="h" power="1">         
          <Forward type="HHExpRate" rate="0.07per_ms" midpoint="-65.mV" scale="-20.mV" />         
          <Reverse type="HHSigmoidRate" rate="1per_ms" midpoint="-35mV" scale="10mV" />     
       </HHGate>
    </HHChannel>
:::
:::

The potassium channel uses exactly the same types, but has only one
gate:

::: {.highlight-xml .notranslate}
::: highlight
    <HHChannel id="k" conductance="20pS">     
       <HHGate id="n" power="4">         
          <Forward type="HHExpLinearRate" rate="0.1per_ms" midpoint="-55mV" scale="10mV" />         
          <Reverse type="HHExpRate" rate="0.125per_ms" midpoint="-65mV" scale="-80mV" />     
       </HHGate>
    </HHChannel>
:::
:::

These channel models are an example where the ability to use the type
name as the XML tag makes the model much clearer: the alternative just
with three levels of Component elements would look rather unhelpful.

Although the channel models have now been defined, they still need to be
used in a cell before anything can be run. For this we'll just define a
basic channel population type. There is one new construct here: the
ComponentRef element which in this case says that a channel population
needs a reference to a component of type HHChannel. This is much like a
Child element, but instead of the component being defined then and there
inside the channel population, there is just a reference to it.

The Dynamics block for a cannel population just computes the total
conductance and then the current, in this case using Ohm's law.

::: {.highlight-xml .notranslate}
::: highlight
    <ComponentType name="ChannelPopulation">     
       <ComponentRef name="channel" type="HHChannel" />     
       <Parameter name="number" dimension="none" />     
       <Parameter name="erev" dimension="voltage" />     
       <Requirement name="v" dimension="voltage" />     
       <Exposure name="current" dimension="current" />     
       <Dynamics>         
          <DerivedVariable name="channelg" select="channel/g" />         
          <DerivedVariable name="geff" value="channelg * number" />         
          <DerivedVariable name="current" exposure="current" value="geff * (erev - v)" />     
       </Dynamics>
    </ComponentType>
:::
:::

To use these populations, they need inserting in a cell. The following
type represents a simple cell with a number of populations and an option
to inject a current so it does something interesting.

::: {.highlight-xml .notranslate}
::: highlight
    <ComponentType name="HHCell">     
       <Parameter name="capacitance" dimension="capacitance" />     
       <Children name="populations" type="ChannelPopulation" />     
       <Parameter name="injection" dimension="current" />     
       <Parameter name="v0" dimension="voltage" />     
       <Exposure name="v" dimension="voltage" />     
       <Dynamics>         
          <OnStart>              
             <StateAssignment variable="v" value="v0" />         
          </OnStart>             
          <DerivedVariable name="totcurrent" select="populations[*]/current" reduce="add" />         
          <StateVariable name="v" dimension="voltage" />          
          <TimeDerivative variable="v" value="(totcurrent + injection) / capacitance" />     
       </Dynamics>
    </ComponentType>
:::
:::

This Dynamics block introduces the OnStart element which is much like
the OnEvent elements earlier, except the block applies only when the
simulation starts. In this case it just sets the voltage to a value
supplied as a parameter. The Dynamics block uses another selector
function "sum(..." to sum the currents delivered by the various
populations.

Now all the definitions are in place to define a cell model with a
couple of channel populations:

::: {.highlight-xml .notranslate}
::: highlight
    <HHCell id="hhcell_1" capacitance="1pF" injection="4pA" v0="-60mV">     
       <ChannelPopulation channel="na" number="6000" erev="50mV" />     
       <ChannelPopulation channel="k" number="1800" erev="-77mV" />
    </HHCell>
:::
:::

To go with this cell type, we can define some components using the other
types defined earlier. Note how celltype_d is based on an existing
component via the "extends" attribute and only replaces one parameter
value.

::: {.highlight-xml .notranslate}
::: highlight
    <Component id="celltype_c" type="iaf3" leakConductance="3 pS" refractoryPeriod="3 ms" threshold="45 mV" leakReversal="-50 mV" deltaV="5mV" capacitance="1uF" />
    <Component id="celltype_d" extends="celltype_c" leakConductance="5 pS" />
    <Component id="gen1" type="spikeGenerator" period="30ms" />
    <Component id="gen2" type="spikeGenerator2" period="32ms" />
    <Component id="cell3cpt" type="cell3" leakReversal="-50mV" deltaV="50mV" threshold="-30mV" leakConductance="50pS" refractoryPeriod="4ms" capacitance="1pF" />
:::
:::

Finally a simulation element says what component is to be run and for
how long. It also contains an embedded display element so the results of
the simulation can be visualized. These are also user-defined types:
their definitions will be presented in example 6.

::: {.highlight-xml .notranslate}
::: highlight
    <Simulation length="80ms" step="0.05ms" target="hhcell_1">     
       <Display unit="ms">         
          <Line quantity="v" unit="mV" color="#0000f0" />     
       </Display>
    </Simulation>
:::
:::

That's it. When this model is run it produces the figure shown below
(after rescaling a bit).

<figure id="id1" class="align-center">
<img src="_images/lems_example4.png"
alt="LEMS GUI showing simulation output graphs" />
<figcaption><p><span class="caption-number">Fig. 53 </span><span
class="caption-text">LEMS GUI showing simulation output graphs</span><a
href="#id1" class="headerlink"
title="Link to this image">#</a></p></figcaption>
</figure>
:::

[]{#document-Userdocs/LEMSExample2}

::: {#example-2-tidying-up-example-1 .section .tex2jax_ignore .mathjax_ignore}
[]{#userdocs-lemsexample2}

### Example 2: tidying up example 1[\#](#example-2-tidying-up-example-1 "Link to this heading"){.headerlink}

This models is the same as in example 1, except that the definitions
have been split out into several self-contained files.

The main file, included below, uses the Include element to include
definitions from other files. Each file is only read once, even if
several files include it. Because some of these files, such as the HH
channel definitions, are intended to be used on their own, they include
all the dimension definitions they need. These may also occur in other
files with the same dimension names. This is fine as long as the
dimensions being declared are the same. An error will be reported if a
new definition is supplied that changes any of the values. The same
applies for Unit definitions. For other element types names and ids must
be unique. An id or name can't appear twice, even if the content of the
elements is the same.
:::

::: {#main-model .section .tex2jax_ignore .mathjax_ignore}
### Main model[\#](#main-model "Link to this heading"){.headerlink}

This defines a few components, then a network that uses them and a
simulation to run it all. The HHCell component refers to channel types
coming from the included hhmodels.xml file which in turn depends on
hhcell.xml and hhchannel.xml.

::: {.highlight-xml .notranslate}
::: highlight
    <Lems>
     
        <Target component="sim1"/>
     
        <Include file="ex2dims.xml"/>
        <Include file="hhchannel.xml"/> 
     
        <Include file="hhcell.xml"/>
        <Include file="spikegenerators.xml"/>
        <Include file="hhmodels.xml"/>
        <Include file="misciaf.xml"/> 

        <Include file="SimpleNetwork.xml"/>

        <HHCell id="hhcell_1" capacitance="1pF" injection="4pA" v0="-60mV">
            <ChannelPopulation channel="na" number="6000" erev="50mV"/>
            <ChannelPopulation channel="k" number="1800" erev="-77mV"/>
        </HHCell>
     
        <Component id="gen1" type="spikeGenerator" period="30ms"/>

        <Component id="gen2" type="spikeGenerator2" period="32ms"/>

        <Component id="iaf3cpt" type="iaf3" leakReversal="-50mV" deltaV="50mV" threshold="-30mV" leakConductance="50pS"
               refractoryPeriod="4ms" capacitance="1pF"/>


        <Network id="net1">
            <Population id="p1" component="gen1" size="1"/>
            <Population id="p2" component="gen2" size="1"/>
            <Population id="p3" component="iaf3cpt" size="1"/>
       
            <Population id="hhpop" component="hhcell_1" size="1"/>
        
       
            <EventConnectivity id="p1-p3" source="p1" target="p3">
                <Connections type="AllAll"/>
            </EventConnectivity>
        </Network>
        
        <Include file="SingleSimulation.xml" />
        
        <Simulation id="sim1" length="80ms" step="0.01ms" target="net1">
            <Display id="d0" title="Example 2" timeScale="1ms" xmin="-10" xmax="90" ymin="-90" ymax="60">
                <Line id="tsince" quantity="p1[0]/tsince" scale="1ms" timeScale="1ms" color="#00c000" />
                <Line id="p3v" quantity="p3[0]/v" scale="1mV" timeScale="1ms" color="#0000f0" />
                <Line id="p0v" quantity="hhpop[0]/v" scale="1mV" timeScale="1ms" color="#ff4040" />
            </Display>
        </Simulation>

    </Lems>
:::
:::
:::

::: {#included-files .section .tex2jax_ignore .mathjax_ignore}
### Included files[\#](#included-files "Link to this heading"){.headerlink}

::: {.highlight-xml .notranslate}
::: highlight
    <Lems> 

        <Dimension name="voltage" m="1" l="2" t="-3" i="-1"/>
        <Dimension name="time" t="1"/>
        <Dimension name="per_time" t="-1"/>
        <Dimension name="conductance" m="-1" l="-2" t="3" i="2"/>
        <Dimension name="capacitance" m="-1" l="-2" t="4" i="2"/>
        <Dimension name="current" i="1"/>
        <Dimension name="temperature" k="1"/>

        <Unit symbol="mV" dimension="voltage" power="-3"/> 
        <Unit symbol="ms" dimension="time" power="-3"/> 
        <Unit symbol="pS" dimension="conductance" power="-12"/>
        <Unit symbol="nS" dimension="conductance" power="-9"/>
        <Unit symbol="uF" dimension="capacitance" power="-6"/>
        <Unit symbol="nF" dimension="capacitance" power="-9"/>
        <Unit symbol="pF" dimension="capacitance" power="-12"/>
        <Unit symbol="per_ms" dimension="per_time" power="3"/>
        <Unit symbol="pA" dimension="current" power="-12"/>
        <Unit symbol="nA" dimension="current" power="-9"/>
        <Unit symbol="degC" dimension="temperature" offset="273.15"/>

    </Lems>
:::
:::

The file hhchannel.xml contains complete definitions of a fairly general
HH-style channel model with any number of gates based on the three
standard types used in the original HH work.

::: {.highlight-xml .notranslate}
::: highlight
    <Lems>

        <Dimension name="voltage" m="1" l="2" t="-3" i="-1"/>
        <Dimension name="time" t="1"/>
        <Dimension name="per_time" t="-1"/>
        <Dimension name="conductance" m="-1" l="-2" t="3" i="2"/>
        <Dimension name="capacitance" m="-1" l="-2" t="4" i="2"/>
        <Dimension name="current" i="1"/>

        
        <ComponentType name="HHRate">
            <Parameter name="rate" dimension="per_time"/>
            <Parameter name="midpoint" dimension="voltage"/>
            <Parameter name="scale" dimension="voltage"/>
            <Requirement name="v" dimension="voltage"/>
            <Exposure name="r" dimension="per_time"/>
        </ComponentType>


        <ComponentType name="HHExpRate" extends="HHRate">
            <Dynamics>
                <DerivedVariable name="r" exposure="r"  dimension="per_time" value="rate * exp((v - midpoint)/scale)"/>
            </Dynamics>
        </ComponentType>


        <ComponentType name="HHSigmoidRate" extends="HHRate">
            <Dynamics>
                <DerivedVariable name="r" dimension="per_time" exposure="r" value="rate / (1 + exp( -(v - midpoint)/scale))"/>
            </Dynamics>
        </ComponentType>


        <ComponentType name="HHExpLinearRate" extends="HHRate">
            <Dynamics>
                <DerivedVariable name="x" dimension="none" value="(v - midpoint) / scale"/>
                <DerivedVariable name="r" dimension="per_time" exposure="r" value="rate * x / (1 - exp(-x))"/>
            </Dynamics>
        </ComponentType>
     

        <ComponentType name="HHGate0">
            <Parameter name="power" dimension="none"/> 
            <Child name="Forward" type="HHRate"/>
            <Child name="Reverse" type="HHRate"/>
            <Requirement name="v" dimension="voltage"/>
            <Exposure name="fcond" dimension="none"/>
        
            <Dynamics simultaneous="true">
                <StateVariable name="q" dimension="none"/>
                <DerivedVariable dimension="per_time" name="rf" select="Forward/r"/>
                <DerivedVariable dimension="per_time" name="rr" select="Reverse/r"/> 
                <TimeDerivative variable="q" value="rf * (1 - q) - rr * q"/>
                <DerivedVariable name="fcond" dimension="none" exposure="fcond" value="q^power"/>
            </Dynamics>    
        </ComponentType>


        <Include file="hhaltgate.xml"/>

        
        <ComponentType name="HHChannel">
            <Parameter name="conductance" dimension="conductance"/>
            <Children name="gates" type="HHGate"/>
            <Exposure name="g" dimension="conductance"/>
            <Dynamics simultaneous="false">
                <DerivedVariable name="gatefeff" dimension="none" select="gates[*]/fcond" reduce="multiply"/>
                <DerivedVariable name="g" exposure="g" dimension="conductance" value="conductance * gatefeff"/>
            </Dynamics>
        </ComponentType>


    </Lems>
:::
:::

As mentioned in example1, the numerics are too feeble to cope with this
gate definition though, so a change of variables is employed instead:

::: {.highlight-xml .notranslate}
::: highlight
    <Lems>
     
        <ComponentType name="HHGate">
            <Parameter name="power" dimension="none"/> 
            <Child name="Forward" type="HHRate"/>
            <Child name="Reverse" type="HHRate"/>
            <Requirement name="v" dimension="voltage"/>
            <Exposure name="fcond" dimension="none"/>
        
       
            <Dynamics simultaneous="false">
                <StateVariable name="x" dimension="none"/>
                <DerivedVariable name="ex" dimension="none" value="exp(x)"/>
                <DerivedVariable name="q" dimension="none" value="ex / (1 + ex)"/>
                <DerivedVariable name="rf" dimension="per_time" select="Forward/r"/>
                <DerivedVariable name="rr" dimension="per_time" select="Reverse/r"/> 
            
                <TimeDerivative variable="x" value="(1 + ex)^2 / ex * (rf * (1 - q) - rr * q)"/>
          
                <DerivedVariable name="fcond" dimension="none" exposure="fcond" value="q^power"/>
            </Dynamics>    
        </ComponentType>

    </Lems>
:::
:::

The file hhcell.xml defines a simple cell model with some populations of
HH channels.

::: {.highlight-xml .notranslate}
::: highlight
    <Lems>

        <Include file="hhchannel.xml"/>

        <Dimension name="voltage" m="1" l="2" t="-3" i="-1"/>
        <Dimension name="capacitance" m="-1" l="-2" t="4" i="2"/>
        <Dimension name="current" i="1"/>

        
        <ComponentType name="ChannelPopulation">
            <ComponentReference name="channel" type="HHChannel"/>
            <Parameter name="number" dimension="none"/>
            <Parameter name="erev" dimension="voltage"/>
            <Requirement name="v" dimension="voltage"/>
            <Exposure name="current" dimension="current"/>
            <Exposure name="geff" dimension="conductance"/>
      
            <Structure>
                <ChildInstance component="channel"/>
            </Structure>
      
            <Dynamics simultaneous="false">
                <DerivedVariable name="channelg" dimension="conductance" select="channel/g"/>
                <DerivedVariable name="geff" exposure="geff" value="channelg * number"/>
                <DerivedVariable name="current" exposure="current" value="geff * (erev - v)"/>
            </Dynamics>
        
        </ComponentType>


        <ComponentType name="HHCell">
            <Parameter name="capacitance" dimension="capacitance"/>
            <Children name="populations" type="ChannelPopulation"/>
            <Parameter name="injection" dimension="current"/>
            <Parameter name="v0" dimension="voltage"/>
            <Exposure name="v" dimension="voltage"/>
        
            <Dynamics simultaneous="true">
                <OnStart>
                    <StateAssignment variable="v" value="v0"/>
                </OnStart>
      
                <DerivedVariable name="totcurrent" dimension="current" select="populations[*]/current" reduce="add"/>
                <StateVariable name="v" exposure="v" dimension="voltage"/> 
                <TimeDerivative variable="v" value="(totcurrent + injection) / capacitance"/>
            </Dynamics>
        </ComponentType>
     
    </Lems>
:::
:::

A couple of spike generators.

::: {.highlight-xml .notranslate}
::: highlight
    <Lems>

        <Dimension name="time" t="1"/>
     

        <ComponentType name="spikeGenerator">
            <Parameter name="period" dimension="time"/>
            <EventPort name="a" direction="out"/>
            <Exposure name="tsince" dimension="time"/>
            <Dynamics>
                <StateVariable name="tsince" exposure="tsince" dimension="time"/>
                <TimeDerivative variable="tsince" value="1"/>
                <OnCondition test="tsince .gt. period">
                    <StateAssignment variable="tsince" value="0"/>
                    <EventOut port="a"/>
                </OnCondition>
            </Dynamics>
        </ComponentType>


        <ComponentType name="spikeGenerator2" extends="spikeGenerator">
            <Dynamics>
                <DerivedVariable name="tsince" exposure="tsince" value="t - tlast"/>
                <StateVariable name="tlast" dimension="time"/>
                <OnCondition test="t - tlast .gt. period">
                    <StateAssignment variable="tlast" value="t"/>
                    <EventOut port="a"/>
                </OnCondition>
            </Dynamics>
        </ComponentType>
     

    </Lems>
:::
:::

And now the components themselves. These are the standard HH sodium and
potassium channels (as used in Rallpack3).

::: {.highlight-xml .notranslate}
::: highlight
    <Lems>
        <Include file="hhchannel.xml"/>

        <Unit symbol="mV" dimension="voltage" power="-3"/> 
        <Unit symbol="per_ms" dimension="per_time" power="3"/>
        <Unit symbol="pS" dimension="conductance" power="-12"/>

        
        <HHChannel id="na" conductance="20pS">
            <HHGate id="m" power="3">
                <Forward type="HHExpLinearRate" rate="1.per_ms" midpoint="-40mV" scale="10mV"/>
                <Reverse type="HHExpRate" rate="4per_ms" midpoint="-65mV" scale="-18mV"/>
            </HHGate>
        
            <HHGate id="h" power="1">
                <Forward type="HHExpRate" rate="0.07per_ms" midpoint="-65.mV" scale="-20.mV"/>
                <Reverse type="HHSigmoidRate" rate="1per_ms" midpoint="-35mV" scale="10mV"/>
            </HHGate>
        </HHChannel>


        <HHChannel id="k" conductance="20pS">
            <HHGate id="n" power="4">
                <Forward type="HHExpLinearRate" rate="0.1per_ms" midpoint="-55mV" scale="10mV"/>
                <Reverse type="HHExpRate" rate="0.125per_ms" midpoint="-65mV" scale="-80mV"/>
            </HHGate>
        </HHChannel>

    </Lems>
:::
:::

Some miscellaneous iaf models.

::: {.highlight-xml .notranslate}
::: highlight
    <Lems>
        
        <Include file="elecdims.xml"/>
       

        <ComponentType name="iaf1">
            <Parameter name="threshold" dimension="voltage"/>
            <Parameter name="refractoryPeriod" dimension="time"/>
            <Parameter name="capacitance" dimension="capacitance"/>
        </ComponentType>


        <ComponentType name="iaf3" extends="iaf1">
            <Parameter name="leakConductance" dimension="conductance"/>
            <Parameter name="leakReversal" dimension="voltage"/>
            <Parameter name="deltaV" dimension="voltage"/>
            <EventPort name="spikes-in" direction="in"/>
            <Exposure name="v" dimension="voltage"/>

            <Dynamics>
                <StateVariable name="v" exposure="v" dimension="voltage"/>
                <TimeDerivative variable="v" value="leakConductance * (leakReversal - v) / capacitance"/>
                 
                <OnEvent port="spikes-in">
                    <StateAssignment variable="v" value="v + deltaV"/>
                </OnEvent>      
            </Dynamics>

        </ComponentType>
     

    </Lems>
:::
:::

Finally, a small collection of dimension definitions useful for things
like the miscellaneous iaf cell definitions.

::: {.highlight-xml .notranslate}
::: highlight
    <Lems>
        <Dimension name="voltage" m="1" l="2" t="-3" i="-1"/>
        <Dimension name="time" t="1"/>
        <Dimension name="conductance" m="-1" l="-2" t="3" i="2"/>
        <Dimension name="capacitance" m="-1" l="-2" t="4" i="2"/>
        <Dimension name="current" i="1"/>
        
    </Lems>
:::
:::
:::

[]{#document-Userdocs/LEMSExample3}

::: {#example-3-connection-dependent-synaptic-components .section .tex2jax_ignore .mathjax_ignore}
[]{#userdocs-lemsexample3}

### Example 3: Connection dependent synaptic components[\#](#example-3-connection-dependent-synaptic-components "Link to this heading"){.headerlink}

In many models, a synapse is only created where a connection exists.
This means that the model of the receiving cell should only declare that
particular types of synapse can be added to it, not the actual synapse
sub-components themselves.

Not much is needed beyond the elements described in example 1 except for
some extensions to the component that declares the connectivity and a
new child element in the component that the synapses are attached to.
The full example is shown below. The synapse type includes an EventPort
just like the previously defined cell type. The cell type however
includes a new child element: Attachments defined as:

::: {.highlight-xml .notranslate}
::: highlight
    <Attachments name="synapses" type="synapse" />
:::
:::

This operates rather like the Children element except that when a
component is defined using this type the sub-elements are not included
in the component definition. Instead it indicates that instances of
components of the particular type may be attached later when the model
is actually run.
:::

[]{#document-Userdocs/LEMSExample4}

::: {#example-4-kinetic-schemes .section .tex2jax_ignore .mathjax_ignore}
[]{#userdocs-lemsexample4}

### Example 4: Kinetic schemes[\#](#example-4-kinetic-schemes "Link to this heading"){.headerlink}

The existing components provide everything necessary to define types
that allow a model to specify a kinetic scheme (Markov model). The
missing ingredient is the Dynamics element to actually expresses how
instances of the components develop through time.

First then, the following definitions can be used to express ion channel
models where the channel state is represented by an occupancy vector
among a number of distinct states with rates for the transitions between
states.

::: {.highlight-xml .notranslate}
::: highlight
    <ComponentType name="KSGate">
       <Parameter name="power" dimension="none" />
       <Parameter name="deltaV" dimension="voltage" />
       <Children name="states" type="KSState" />
       <Children name="transitions" type="KSTransition" />
    </ComponentType>
    <ComponentType name="KSState">
       <Parameter name="relativeConductance" dimension="none" />
       <Dynamics>
          <StateVariable name="occupancy" dimension="none" />
          <DerivedVariable name="q" value="relativeConductance * occupancy" />
       </Dynamics>
    </ComponentType>
    <ComponentType name="KSClosedState" extends="KSState">
       <Fixed parameter="relativeConductance" value="0" />
    </ComponentType>
    <ComponentType name="KSOpenState" extends="KSState">
       <Fixed parameter="relativeConductance" value="1" />
    </ComponentType>
    <ComponentType name="KSTransition">
       <Link name="from" type="KSState" />
       <Link name="to" type="KSState" />
       <Requirement name="v" dimension="voltage" />
       <Exposure name="rf" dimension="per_time" />
       <Exposure name="rr" dimension="per_time" />
    </ComponentType>
    <ComponentType name="KSChannel">
       <Parameter name="conductance" dimension="conductance" />
       <Children name="gates" type="KSGate" />
       <Exposure name="g" dimension="conductance" />
       <Dynamics>
          <DerivedVariable name="fopen" dimension="none" select="gates[*]/fopen" reduce="multiply" />
          <DerivedVariable name="g" exposure="g" dimension="conductance" value="fopen * conductance" />
       </Dynamics>
    </ComponentType>
:::
:::

This says that a gate can contain any number of states and transitions.
A state has an occupancy variable, and a transition has links to two
states giving the source and target states for the transition.

The transition element here is an abstract element because it doesn't
provide a Dynamics block but just specifies what quantities transitions
should privide via the two exposures. One of the most useful forms of
transition is a damped Boltzman equation which can be parameterizd as
follows:

::: {.highlight-xml .notranslate}
::: highlight
    <ComponentType name="VHalfTransition" extends="KSTransition">
       <Parameter name="vHalf" dimension="voltage" />
       <Parameter name="z" dimension="none" />
       <Parameter name="gamma" dimension="none" />
       <Parameter name="tau" dimension="time" />
       <Parameter name="tauMin" dimension="time" />
       <Constant name="kte" dimension="voltage" value="25.3mV" />
       <Requirement name="v" dimension="voltage" />
       <Dynamics>
          <DerivedVariable name="rf0" dimension="per_time" value="exp(z * gamma * (v - vHalf) / kte) / tau" />
          <DerivedVariable name="rr0" dimension="per_time" value="exp(-z * (1 - gamma) * (v - vHalf) / kte) / tau" />
          <DerivedVariable name="rf" exposure="rf" dimension="per_time" value="1 / (1/rf0 + tauMin)" />
          <DerivedVariable name="rr" exposure="rr" dimension="per_time" value="1 / (1/rr0 + tauMin)" />
       </Dynamics>
    </ComponentType>
:::
:::

Given these definitions, we can express a couple of simple channel
models that use kinetic schemes. There is nothing special about these
models. They are just examples used in PSICS that produce spikes (albeit
rather unnatural looking ones) when used together.

::: {.highlight-xml .notranslate}
::: highlight
    <KSChannel id="na1" conductance="20pS">
       <KSGate power="1" deltaV="0.1mV">
          <KSClosedState id="c1" />
          <KSClosedState id="c2" />
          <KSOpenState id="o1" relativeConductance="1" />
          <KSClosedState id="c3" />
          <VHalfTransition from="c1" to="c2" vHalf="-35mV" z="2.5" gamma="0.8" tau="0.15ms" tauMin="0.001ms" />
          <VHalfTransition from="c2" to="o1" vHalf="-35mV" z="2.5" gamma="0.8" tau="0.15ms" tauMin="0.001ms" />
          <VHalfTransition from="o1" to="c3" vHalf="-70mV" z="1.1" gamma="0.90" tau="8.0ms" tauMin="0.01ms" />
       </KSGate>
    </KSChannel>
    <KSChannel id="k1" conductance="30pS">
       <KSGate power="1" deltaV="0.1mV">
          <KSClosedState id="c1" />
          <KSOpenState id="o1" />
          <VHalfTransition from="c1" to="o1" vHalf="0mV" z="1.5" gamma="0.75" tau="3.2ms" tauMin="0.3ms" />
       </KSGate>
    </KSChannel>
:::
:::

This has all been done with the existing components. They allow types to
be defined for expressing kinetic schemes, and models can be expressed
that use these types, but there is nothing so far that says that the
model actually is governed by a kinetic scheme. In particular, there is
an "occupancy" state variable in each state element for which there is
no governing equation and the rates generate "rf" and "rr" quantities
that are not unused anywhere.

What is needed is a new element in the Dynamics block to link these
together and say that the rates apply to the changes of occupancy among
the state elements. This is done by adding a KineticScheme element to
the Dynamics block for a gate as follows (this now shows the full
definition of the KSGate element):

::: {.highlight-xml .notranslate}
::: highlight
    <ComponentType name="KSGate">
       <Parameter name="power" dimension="none" />
       <Parameter name="deltaV" dimension="voltage" />
       <Children name="states" type="KSState" />
       <Children name="transitions" type="KSTransition" />
       <Dynamics>
          <KineticScheme name="ks">
             <Nodes children="states" variable="occupancy" />
             <Edges children="transitions" sourceNodeName="from" targetNodeName="to" forwardRate="rf" reverseRate="rr" />
             <Tabulable variable="v" increment="deltaV" />
          </KineticScheme>
          <DerivedVariable name="q" dimension="none" select="states[*]/q" reduce="add" />
          <DerivedVariable name="fopen" exposure="fopen" dimension="none" value="q^power" />
       </Dynamics>
    </ComponentType>
:::
:::

The new part here is the KineticScheme element and its children Nodes,
Edges and Tabulable. The Nodes element says which elements in the parent
container are goverened by the scheme, and which variable in those
elements represents the relative occupancy.

The Edges element is a little more complicated. It has to say not only
which elements define the transitions, but how the fields in the
transitions map to things the scheme knows about. For a transition in a
kinetic scheme, you need to know which state the transition comes from,
which it goes to, and how fast it goes. It is possible (as here) that a
single transition defines both directions, in which case it must also
say which variable in the target objects provides the reverse transition
rates. This is what the last four attributes of the Edges element do.

The Tabulable element is a temporary convenience for implementation
purposes. In this case it says that the rates depend only on v and that
the transition matrices can be cached an reused on a grid of spacing
deltaV rather than recomputed every time. This is not used in the 0.2.1
version of the interpreter.

Note that the KineticScheme element doesn't say anything about what the
outputs are. All it does is control the occupancy state variable in the
state elements. The interpretation of these quantities is specified in
the normal way with the two DerivedVaraible declarations. No special
elements are needed in the scheme itself.

To actually use these models we need cell and population elements to
link them all together. There is nothing new here - it all works just as
for HH channels. The rest of the example4.xml file is:

::: {.highlight-xml .notranslate}
::: highlight
    <ComponentType name="ChannelPopulation">
       <ComponentRef name="channel" type="KSChannel" />
       <Parameter name="number" dimension="none" />
       <Parameter name="erev" dimension="voltage" />
       <Requirement name="v" dimension="voltage" />
       <Dynamics>
          <DerivedVariable name="channelg" dimension="conductance" select="channel/g" />
          <DerivedVariable name="geff" value="channelg * number" />
          <DerivedVariable name="current" value="geff * (erev - v)" />
       </Dynamics>
    </ComponentType>
    <ComponentType name="KSCell">
       <Parameter name="capacitance" dimension="capacitance" />
       <Children name="populations" type="ChannelPopulation" />
       <Parameter name="injection" dimension="current" />
       <Parameter name="v0" dimension="voltage" />
       <Dynamics>
          <OnStart>
             <StateAssignment variable="v" value="v0" />
          </OnStart>
          <DerivedVariable name="totcurrent" dimension="current" select="sum(populations[*]/current)" />
          <StateVariable name="v" dimension="voltage" />
          <TimeDerivative variable="v" value="(totcurrent + injection) / capacitance" />
       </Dynamics>
    </ComponentType>
    <KSCell id="kscell_1" capacitance="1pF" injection="1pA" v0="-60mV">
       <ChannelPopulation channel="na1" number="600" erev="50mV" />
       <ChannelPopulation channel="k1" number="180" erev="-77mV" />
    </KSCell>

    <Network id="net1">
       <XPopulation id="kspop" component="kscell_1" size="1" />
    </Network>

    <Simulation length="80ms" step="0.07ms" target="net1">
       <Display timeScale="ms">
          <Line quantity="kspop[0]/v" scale="mV" color="#ff4040" />
       </Display>
    </Simulation>
:::
:::

When run, this produces:

<figure id="id1" class="align-center">
<img src="_images/lems_example4.png"
alt="LEMS GUI showing simulation output graphs" />
<figcaption><p><span class="caption-number">Fig. 54 </span><span
class="caption-text">LEMS GUI showing simulation output graphs</span><a
href="#id1" class="headerlink"
title="Link to this image">#</a></p></figcaption>
</figure>

There are clearly some initialization issues but the basic Dynamics is
the same as the PSICS version of this model.
:::

[]{#document-Userdocs/LEMSExample5}

::: {#example-5-references-and-paths .section .tex2jax_ignore .mathjax_ignore}
[]{#userdocs-lemsexample5}

### Example 5: References and paths[\#](#example-5-references-and-paths "Link to this heading"){.headerlink}

The ChannelPopulation type used earlier repeats a common model
specification error in that it makes the reversal potential of a
population of channels a parameter of the population (often it is made a
parameter of the channel specification, which is equally bad):

::: {.highlight-xml .notranslate}
::: highlight
    <ComponentType name="ChannelPopulation">
       <ComponentRef name="channel" type="KSChannel" />
       <Parameter name="number" dimension="none" />
       <Parameter name="erev" dimension="voltage" />
    </ComponentType>
:::
:::

In fact, of course, the reversal potential is not a property of a
channel population, or of a channel. It depends on the environment the
channel is put in and the ions it is permeable to. But, it is needed in
the Dynamics specification for the population so just putting it in as a
parameter solves the immediate problem. In the process, however, it
introduces the potential for easily creating contradictory models, by,
for example setting different reversals for populations of the same type
of channel.

A much better approach is to let the channel just say what it is
permeable to. Some other element in the model can define the membrane
reversal potentials for different channels, and the channel population
object should then look up the relevant value for the permeant ion of
its channel. This provides a cleaner expression of what is there,
removes redundancy and lowers the entropy of the model specification.

The following three types are sufficient to provide a simple framework
to centralize the definitions of species and reversal potentials on one
place:

::: {.highlight-xml .notranslate}
::: highlight
    <ComponentType name="Species">
       <Text name="name" />
       <Parameter name="charge" dimension="none" />
    </ComponentType>
    <ComponentType name="Environment">
       <Children name="membranePotentials" type="MembranePotential" />
    </ComponentType>
    <ComponentType name="MembranePotential">
       <ComponentRef name="species" type="Species" />
       <Parameter name="reversal" dimension="voltage" />
    </ComponentType>
:::
:::

Once these are available, they can be used to define some species, and
to create an environment component that sets their reversal potentials:

::: {.highlight-xml .notranslate}
::: highlight
    <Species id="Na" name="Sodium" charge="1" />
    <Species id="K" name="Potassium" charge="1" />
    <Species id="Ca" name="Calcium" charge="1" />
    <Environment id="env1">
       <MembranePotential species="Na" reversal="50mV" />
       <MembranePotential species="K" reversal="-80mV" />
    </Environment>
:::
:::

The next step is to add a species reference to the channel type, so that
channel definitions can say what species they are permeant to.

::: {.highlight-xml .notranslate}
::: highlight
    <ComponentType name="KSChannel">
       <Parameter name="conductance" dimension="conductance" />
       <ComponentRef name="species" type="Species" />
       <Children name="gates" type="KSGate" />
       <Dynamics>
          <DerivedVariable name="fopen" dimension="none" select="gates[*]/fopen" reduce="multiply" />
          <DerivedVariable name="g" dimension="conductance" value="fopen * conductance" />
       </Dynamics>
    </ComponentType>
:::
:::

Finally the channel population type needs modifying to add a derived
parameter that addresses the reversal potential from the membrane
properties:

::: {.highlight-xml .notranslate}
::: highlight
    <ComponentType name="ChannelPopulation">
       <ComponentRef name="channel" type="KSChannel" />
       <Parameter name="number" dimension="none" />
       <Requirement name="v" dimension="voltage" />
       <DerivedParameter name="erev" dimension="voltage" select="//MenbranePotential[species = channel/species]/reversal" />
       <Dynamics>
          <DerivedVariable name="channelg" dimension="conductance" select="channel/g" />
          <DerivedVariable name="geff" value="channelg * number" />
          <DerivedVariable name="current" value="geff * (erev - v)" />
       </Dynamics>
    </ComponentType>
:::
:::

This introduces a new construct, the DerivedParameter specification that
defines a local parameter "erev" to hold the quantity from the specified
path:

::: {.highlight-xml .notranslate}
::: highlight
    <DerivedParameter name="erev" dimension="voltage" select="//MenbranePotential[species = channel/species]/reversal" />
:::
:::

The path here uses XPath like syntax operating on the component tree in
the model. In this case, it finds all the elements of thpe
MembranePotential in the model. The predicate selects the one for which
the species is the same as the species referred to from the channel used
for this population. Finally, it takes the "reversal" parameter from the
membrane potential component. This is made locally available as the
parameter "erev".

The Dynamics of this model is exactly the same as example 4. The full
model including both the type definitions and the components is included
below.

::: {.highlight-xml .notranslate}
::: highlight
    <Lems>

        <Target component="sim1"/>

        <Include file="ex2dims.xml"/>

        <ComponentType name="Species">
            <Text name="name"/>
            <Parameter name="charge" dimension="none"/>
        </ComponentType>


        <ComponentType name="Environment">
            <Children name="membranePotentials" type="MembranePotential"/>
        </ComponentType>


        <ComponentType name="MembranePotential">
            <ComponentReference name="species" type="Species"/>
            <Parameter name="reversal" dimension="voltage"/>
        </ComponentType>

        <Species id="Na" name="Sodium" charge="1"/>
        <Species id="K" name="Potassium" charge="1"/>
        <Species id="Ca" name="Calcium" charge="1"/>


        <Environment id="env1">
            <MembranePotential species="Na" reversal="50mV"/>
            <MembranePotential species="K" reversal="-80mV"/>    
        </Environment>


        <ComponentType name="KSChannel">
            <Parameter name="conductance" dimension="conductance"/>
            <ComponentReference name="species" type="Species"/>
            <Children name="gates" type="KSGate"/>
            <Exposure name="g" dimension="conductance"/>
        
            <Dynamics>
                <DerivedVariable name="fopen" dimension="none" select="gates[*]/fopen" reduce="multiply"/>
                <DerivedVariable name="g" exposure="g" dimension="conductance" value="fopen * conductance"/>
            </Dynamics>
        </ComponentType>


        
        <ComponentType name="KSGate">
            <Parameter name="power" dimension="none"/>
            <Parameter name="deltaV" dimension="voltage"/>
            <Children name="states" type="KSState"/>
            <Children name="transitions" type="KSTransition"/>
            <Exposure name="fopen" dimension="none"/>
        
            <Dynamics>   
                <KineticScheme name="ks" nodes="states" stateVariable="occupancy"
                               edges="transitions" edgeSource="from" edgeTarget="to" 
                               forwardRate="rf" reverseRate="rr" dependency="v" step="deltaV"/>
        
                <DerivedVariable name="q" dimension="none" select="states[*]/q" reduce="add"/>
                <DerivedVariable name="fopen" exposure="fopen" dimension="none" value="q^power"/>
            </Dynamics>
        </ComponentType>

        
        <ComponentType name="KSState">
            <Parameter name="relativeConductance" dimension="none"/>
            <Exposure name="q" dimension="none"/>
            <Exposure name="occupancy" dimension="none"/>
       
            <Dynamics>
                <StateVariable name="occupancy" exposure="occupancy" dimension="none"/>
                <DerivedVariable name="q" exposure="q" value="relativeConductance * occupancy"/>
            </Dynamics>
        </ComponentType>

        <ComponentType name="KSClosedState" extends="KSState">
            <Fixed parameter="relativeConductance" value="0"/>
        </ComponentType>


        <ComponentType name="KSOpenState" extends="KSState">
            <Fixed parameter="relativeConductance" value="1"/>
        </ComponentType>


        <ComponentType name="KSTransition">
            <Link name="from" type="KSState"/>
            <Link name="to" type="KSState"/>
            <Exposure name="rf" dimension="per_time"/>
            <Exposure name="rr" dimension="per_time"/>
     
        </ComponentType>


        <ComponentType name="VHalfTransition" extends="KSTransition">
            <Parameter name="vHalf" dimension="voltage"/>
            <Parameter name="z" dimension="none"/>
            <Parameter name="gamma" dimension="none"/>
            <Parameter name="tau" dimension="time"/>
            <Parameter name="tauMin" dimension="time"/>
            <Constant name="kte" dimension="voltage" value="25.3mV"/>
            <Requirement name="v" dimension="voltage"/>  
        
            <Dynamics>
                <DerivedVariable name="rf0" dimension="per_time" value="exp(z * gamma * (v - vHalf) / kte) / tau"/>
                <DerivedVariable name="rr0" dimension="per_time" value="exp(-z * (1 - gamma) * (v - vHalf) / kte) / tau"/>
                <DerivedVariable name="rf" exposure="rf" dimension="per_time" value="1 / (1/rf0 + tauMin)"/>
                <DerivedVariable name="rr" exposure="rr" dimension="per_time" value="1 / (1/rr0 + tauMin)"/>
            </Dynamics>
        </ComponentType>






        <KSChannel id="na1" conductance="20pS" species="Na">
            <KSGate power="1" deltaV="0.1mV">
                <KSClosedState id="c1"/>
                <KSClosedState id="c2"/>
                <KSOpenState id="o1" relativeConductance="1"/>
                <KSClosedState id="c3"/>
                <VHalfTransition from="c1" to="c2" vHalf = "-35mV" z="2.5" gamma="0.8" tau="0.15ms" tauMin="0.001ms"/>
                <VHalfTransition from="c2" to="o1" vHalf = "-35mV" z="2.5" gamma="0.8" tau="0.15ms" tauMin="0.001ms"/>
                <VHalfTransition from="o1" to="c3" vHalf = "-70mV" z="1.1" gamma="0.90" tau="8.0ms" tauMin="0.01ms"/>         
            </KSGate>
        </KSChannel>


        <KSChannel id="k1" conductance="30pS" species="K">
            <KSGate power="1" deltaV="0.1mV">
                <KSClosedState id="c1"/>
                <KSOpenState id="o1"/>
                <VHalfTransition from="c1" to="o1" vHalf = "0mV" z="1.5" gamma="0.75" tau="3.2ms" tauMin="0.3ms"/>
            </KSGate>
        </KSChannel>




        <ComponentType name="ChannelPopulation">
            <ComponentReference name="channel" type="KSChannel"/>
            <Parameter name="number" dimension="none"/>
            <Requirement name="v" dimension="voltage"/>
            <Exposure name="current" dimension="current"/>
            <DerivedParameter name="erev" dimension="voltage" select="//MembranePotential[species=channel/species]/reversal"/>
            <Dynamics>
         
                <DerivedVariable name="channelg" dimension="conductance" select="channel/g"/>
                <DerivedVariable name="geff" value="channelg * number"/>
                <DerivedVariable name="current" exposure="current" value="geff * (erev - v)"/>
            </Dynamics>    
        
        
            <Structure>    
                <ChildInstance component="channel"/>    
            </Structure>
        </ComponentType>




        <ComponentType name="KSCell">
            <Parameter name="capacitance" dimension="capacitance"/>
            <ComponentReference name="environment" type="Environment"/>
            <Children name="populations" type="ChannelPopulation"/>
            <Parameter name="injection" dimension="current"/>
            <Parameter name="v0" dimension="voltage"/>
            <Exposure name="v" dimension="voltage"/>
            <Dynamics>
                <OnStart>
                    <StateAssignment variable="v" value="v0"/>
                </OnStart>
      
                <DerivedVariable name="totcurrent" dimension="current" select="populations[*]/current" reduce="add"/>
                <StateVariable name="v" exposure="v" dimension="voltage"/> 
                <TimeDerivative variable="v" value="(totcurrent + injection) / capacitance"/>
            </Dynamics>
        </ComponentType>



        <KSCell id="kscell_1" capacitance="0.4pF" injection="1pA" v0="-60mV" environment="env1">
            <ChannelPopulation channel="na1" number="400"/>
            <ChannelPopulation channel="k1" number="180"/>
        </KSCell>

        
        <Include file="SingleSimulation.xml" />
        
        <Simulation id="sim1" length="80ms" step="0.05ms" target="kscell_1">
            <Display id="d0" title="Example 5: References and paths" timeScale="1ms" xmin="-10" xmax="90" ymin="-90" ymax="60">
                <Line id="v" quantity="v" scale="1mV" timeScale="1ms" color="#0000f0"/>
            </Display>
        </Simulation>

    </Lems>
:::
:::
:::

[]{#document-Userdocs/LEMSExample6}

::: {#example-6-user-defined-types-for-simulation-and-display .section .tex2jax_ignore .mathjax_ignore}
[]{#userdocs-lemsexample6}

### Example 6: User defined types for simulation and display[\#](#example-6-user-defined-types-for-simulation-and-display "Link to this heading"){.headerlink}

Up until now, the examples have used a set of simple Simulation, Display
and Line constructs without explaining how they are defined. This shows
what is needed in the Dynamics block to let the user defined types to
specify that they actually define a runnable simulation or settings that
can be used to display results.

This means that the user can select their own names for the different
parameters required for a simulation, and, more importantly, simulation
and display attributes can be added to existing type definitions to make
multi-faceted type definitions that can both be run on their own or as
part of a larger simulation.

Example 6 shows two new elements that can be used in the Dynamics block,
Run and Show as illustrated in the following user-defined type that
defines a simulation:

::: {.highlight-xml .notranslate}
::: highlight
    <ComponentType name="Simulation">
       <Parameter name="length" dimension="time" />
       <Parameter name="step" dimension="time" />
       <ComponentRef name="target" type="HHCell" />
       <Children name="displays" type="Display" />
       <Dynamics>
          <StateVariable name="t" dimension="time" />
          <Run component="target" variable="t" increment="step" total="length" />
          <Show src="displays" />
       </Dynamics>
    </ComponentType>
:::
:::

The 'component' attribute of the Run element specifies which parameter
of the type contains the reference to the component that should actually
be run. The 'step' and 'increment' attributes specify the parameters
that hold the timestep and total runtime. The 'variable' attribute is
for future use - at present, the independent variable is always 't'.

A Run element can be added to the Dynamics block in any type definition
to make it independently runnable.

Running a simulation without any output is rarely much use, so there are
two futher elements that can be included in the Dynamics block: Show and
Record. The 'src' attribute of the Show element points to the components
that should be shown. These in turn can contain other Show elements but
eventually everything pointed to by a Show element should contain one or
more Record elements. These specify what will actually be sent as
output. They have the path to the variable as the 'quantity' attribute,
its scale as the 'scale' attribute and the line color for plotting.

The following two types show one way that these can be combined to allow
the user to express a display object containing one or more lines.

::: {.highlight-xml .notranslate}
::: highlight
    <ComponentType name="Display">
       <Parameter name="timeScale" dimension="time" />
       <Children name="lines" type="Line" />
       <Dynamics>
          <Show src="lines" scale="timeScale" />
       </Dynamics>
    </ComponentType>
    <ComponentType name="Line">
       <Parameter name="scale" dimension="*" />
       <Text name="color" />
       <Path name="quantity" />
       <Dynamics>
          <Record quantity="quantity" scale="scale" color="color" />
       </Dynamics>
    </ComponentType>
:::
:::

Once these have been defined, a component can be constructed that uses
them as follows:

::: {.highlight-xml .notranslate}
::: highlight
    <Simulation id="sim1" length="80ms" step="0.05ms" target="hhcell_1">
       <Display timeScale="1ms">
          <Line id="V" quantity="v" scale="1mV" color="#0000f0" />
          <Line id="Na_q" quantity="NaPop/geff" scale="1nS" color="#f00000" />
          <Line id="K_q" quantity="KPop/geff" scale="1nS" color="#00f000" />
       </Display>
    </Simulation>
:::
:::

When run, this produces the output shown below:

<figure id="id1" class="align-center">
<img src="_images/lems_example6.png"
alt="LEMS GUI showing simulation output graphs" />
<figcaption><p><span class="caption-number">Fig. 55 </span><span
class="caption-text">LEMS GUI showing simulation output graphs</span><a
href="#id1" class="headerlink"
title="Link to this image">#</a></p></figcaption>
</figure>

Note how the scale attributes are set to 1mV and 1nS for the different
lines so that they show up on the same axes.
:::

[]{#document-Userdocs/LEMSExample7}

::: {#example-7-user-defined-types-for-networks-and-populations .section .tex2jax_ignore .mathjax_ignore}
[]{#userdocs-lemsexample7}

### Example 7: User defined types for networks and populations[\#](#example-7-user-defined-types-for-networks-and-populations "Link to this heading"){.headerlink}

This example shows how the standard component type structures can be
used to declare components for simple networks. The following three
definitions allow networks to be constructed containing fixed size
populations of a particular component type.

::: {.highlight-xml .notranslate}
::: highlight
    <ComponentType name="Network">
       <Children name="populations" type="Population" />
       <Children name="connectivities" type="EventConnectivity" />
    </ComponentType>
    <ComponentType name="Population">
       <ComponentRef name="component" type="Component" />
       <Parameter name="size" dimension="none" />
    </ComponentType>
    <ComponentType name="EventConnectivity">
       <Link name="source" type="Population" />
       <Link name="target" type="Population" />
       <Child name="Connections" type="ConnectionPattern" />
    </ComponentType>
:::
:::

The harder part is to provide elements in the Dynamics blocks to express
what should be done with components based on these types. The Network
element doesn't pose any problems because the default behavior on
instantiation will do the right thing: it will instantiate each of the
child populations and EventConnectivity elements.

But the population element needs to say that its instantiation involves
making 'size' instances of the component referred to by the 'component'
reference, where 'size' is the value supplied for the size parameter in
a component specification. This can be done by including a Build element
inside the Dynamics block:

::: {.highlight-xml .notranslate}
::: highlight
    <ComponentType name="Population">
       <ComponentRef name="component" type="Component" />
       <Parameter name="size" dimension="none" />
       <Dynamics>
          <Build>
             <MultiInstantiate number="size" component="component" />
          </Build>
       </Dynamics>
    </ComponentType>
:::
:::

The MultiInstantiate specification says that there should be 'size'
instances of the component referred to in the 'component' parameter
created when the model is built. This overrides the default behavior.
\[TODO: what is the Build element content corresponding to the default
behavior?\].

This serves to create some rather simple populations. More complex
specifications, such as putting one instance at each point of a grid
satisfying a particular constraint could be handled via first declaring
elements to form the grid, and then using selectors that pick the points
in the population element to actually put the cells at \[its not clear
to me how much more would be required to make this work, other than
implementing proper xpath-like selectors\].

The following three types define a general connectivity structure with
an abstract ConnectionPattern type, and a specific instance for All-All
connectivity.

::: {.highlight-xml .notranslate}
::: highlight
    <ComponentType name="EventConnectivity">
       <Link name="source" type="Population" />
       <Link name="target" type="Population" />
       <Child name="Connections" type="ConnectionPattern" />
    </ComponentType>
    <ComponentType name="ConnectionPattern">
    </ComponentType>
    <ComponentType name="AllAll" extends="ConnectionPattern">
       <Dynamics>
          <Build>
             <ForEach instances="../source" as="a">
                <ForEach instances="../target" as="b">
                   <EventConnection from="a" to="b" />
                </ForEach>
             </ForEach>
          </Build>
       </Dynamics>
    </ComponentType>
:::
:::

The Build element in the AllAll pattern uses a new ForEach construct and
the EventConnectin element from before. The ForEach element operates
selects each instance matching its 'instances' attribute, and applies
the enclosing directives, much in the same way as for-each in XSL. The
proof of concept interpreter also has Choose, When and Otherwise
elements that operate much like their XSL equivalents, although these
are not used in this example.

With these definitions in place, a network simulation can be defined
with the following:

::: {.highlight-xml .notranslate}
::: highlight
    <Network id="net1">
       <Population id="p1" component="gen1" size="2" />
       <Population id="p3" component="iaf3cpt" size="3" />
       <EventConnectivity id="p1-p3" source="p1" target="p3">
          <Connections type="AllAll" />
       </EventConnectivity>
    </Network>

    <Simulation id="sim1" length="80ms" step="0.05ms" target="net1">
       <Display timeScale="1ms">
          <Line id="gen_v" quantity="p3[0]/v" scale="1mV" color="#0000f0" />
          <Line id="gen_tsince" quantity="p1[0]/tsince" scale="1ms" color="#00c000" />
       </Display>
    </Simulation>
:::
:::

The output when the model is run is shown below, followed by the full
listing.

<figure id="id1" class="align-center">
<img src="_images/lems_example7.png"
alt="LEMS GUI showing simulation output graphs" />
<figcaption><p><span class="caption-number">Fig. 56 </span><span
class="caption-text">LEMS GUI showing simulation output graphs</span><a
href="#id1" class="headerlink"
title="Link to this image">#</a></p></figcaption>
</figure>

::: {.highlight-xml .notranslate}
::: highlight
    <Lems>
     
        <Target component="sim1"/> 
     
     
        <Include file="ex2dims.xml"/>
        <Include file="spikegenerators.xml"/>
        <Include file="misciaf.xml"/>

     
        <Component id="gen1" type="spikeGenerator" period="30ms"/>

        <Component id="gen2" type="spikeGenerator2" period="32ms"/>

        <Component id="iaf3cpt" type="iaf3" leakReversal="-50mV" deltaV="50mV" threshold="-30mV" leakConductance="50pS"
               refractoryPeriod="4ms" capacitance="1pF"/>


        <ComponentType name="Network">
            <Children name="populations" type="Population"/>
            <Children name="connectivities" type="EventConnectivity"/>
        </ComponentType>


        <ComponentType name="Population">
            <ComponentReference name="component" type="Component"/>
            <Parameter name="size" dimension="none"/>
       
            <Structure>
                <MultiInstantiate number="size" component="component"/>
            </Structure>
      
        </ComponentType>


        <ComponentType name="EventConnectivity">
            <Link name="source" type="Population"/>
            <Link name="target" type="Population"/>
            <Child name="Connections" type="ConnectionPattern"/>
        </ComponentType>


        <ComponentType name="ConnectionPattern"/>


        <ComponentType name="AllAll" extends="ConnectionPattern">
            <Structure>
                <ForEach instances="../source" as="a">
                    <ForEach instances="../target" as="b">
                        <EventConnection from="a" to="b"/>
                    </ForEach>
                </ForEach>    
            </Structure>
        </ComponentType>


        <Network id="net1">
            <Population id="p1" component="gen1" size="2"/>
            <Population id="p3" component="iaf3cpt" size="3"/>
         
            <EventConnectivity id="p1-p3" source="p1" target="p3">
                <Connections type="AllAll"/>
            </EventConnectivity>
        </Network>


        <Include file="SingleSimulation.xml" />
        
        <Simulation id="sim1" length="80ms" step="0.05ms" target="net1">
            <Display id="d0" title="Example 7: User defined types for networks and populations" timeScale="1ms" xmin="-10" xmax="90" ymin="-50" ymax="90">
                <Line id="gen_v" quantity="p3[0]/v" scale="1mV" timeScale="1ms" color="#0000f0"/>
                <Line id="gen_tsince" quantity="p1[0]/tsince" scale="1ms" timeScale="1ms" color="#00c000"/>
            </Display>
        </Simulation>

    </Lems>
:::
:::
:::

[]{#document-Userdocs/LEMSExample8}

::: {#example-8-regimes-in-dynamics-definitions .section .tex2jax_ignore .mathjax_ignore}
[]{#userdocs-lemsexample8}

### Example 8: Regimes in Dynamics definitions[\#](#example-8-regimes-in-dynamics-definitions "Link to this heading"){.headerlink}

This example introduces the Regime, Transition and OnEntry elements
within a Dynamics block. Rather than having a single state instance, the
entity can be on one of the defined regimes at any given time. The
Transition element occurring inside a condition block serves to move it
from one regime to another. The OnEntry block inside a regime can
contain initialization directives that apply each time the entity enters
that regime.

::: {.highlight-xml .notranslate}
::: highlight
    <ComponentType name="refractiaf">
       <Parameter name="threshold" dimension="voltage" />
       <Parameter name="refractoryPeriod" dimension="time" />
       <Parameter name="capacitance" dimension="capacitance" />
       <Parameter name="vleak" dimension="voltage" />
       <Parameter name="gleak" dimension="conductance" />
       <Parameter name="current" dimension="current" />
       <Parameter name="vreset" dimension="voltage" />
       <Parameter name="deltaV" dimension="voltage" />
       <Parameter name="v0" dimension="voltage" />
       <EventPort name="out" direction="out" />
       <EventPort name="in" direction="in" />
       <Dynamics>
          <StateVariable name="v" dimension="voltage" />
          <OnStart>
             <StateAssignment variable="v" value="v0" />
          </OnStart>
          <Regime name="refr">
             <StateVariable name="tin" dimension="time" />
             <OnEntry>
                <StateAssignment variable="tin" value="t" />
                <StateAssignment variable="v" value="vreset" />
             </OnEntry>
             <OnCondition test="t .gt. tin + refractoryPeriod">
                <Transition regime="int" />
             </OnCondition>
          </Regime>
          <Regime name="int" initial="true">
             <TimeDerivative variable="v" value="(current + gleak * (vleak - v)) / capacitance" />
             <OnCondition test="v .gt. threshold">
                <EventOut port="out" />
                <Transition regime="refr" />
             </OnCondition>
             <OnEvent port="in">
                <StateAssignment variable="v" value="v + deltaV" />
             </OnEvent>
          </Regime>
       </Dynamics>
    </ComponentType>
:::
:::

Full listing:

::: {.highlight-xml .notranslate}
::: highlight
    <Lems>

        <Target component="sim1"/> 


        <Include file="ex2dims.xml"/>
        <Include file="spikegenerators.xml"/>
        <Include file="misciaf.xml"/>


        <ComponentType name="refractiaf">
            <Parameter name="threshold" dimension="voltage"/>
            <Parameter name="refractoryPeriod" dimension="time"/>
            <Parameter name="capacitance" dimension="capacitance"/>
            <Parameter name="vleak" dimension="voltage"/>
            <Parameter name="gleak" dimension="conductance"/>

            <Parameter name="current" dimension="current"/>
            <Parameter name="vreset" dimension="voltage"/>
            <Parameter name="deltaV" dimension="voltage"/>
            <Parameter name="v0" dimension="voltage"/>

            <EventPort name="out" direction="out"/>
            <EventPort name="in" direction="in"/>

            <Exposure name="v" dimension="voltage"/>

            <Dynamics>     
                <StateVariable name="v" exposure="v" dimension="voltage" /> 
                <StateVariable name="tin" dimension="time"/>       
                
                <OnStart>
                    <StateAssignment variable="v" value="v0"/>
                </OnStart>

                <Regime name="refr">         
                    <OnEntry>             
                        <StateAssignment variable="tin" value="t" />             
                        <StateAssignment variable="v" value="vreset" />          
                    </OnEntry>                   
                    <OnCondition test="t .gt. tin + refractoryPeriod">                 
                        <Transition regime="int" />             
                    </OnCondition>         
                </Regime>          

                <Regime name="int" initial="true">         
                    <TimeDerivative variable="v" value="(current + gleak * (vleak - v)) / capacitance" />         
                    <OnCondition test="v .gt. threshold">             
                        <EventOut port="out" />             
                        <Transition regime="refr" />         
                    </OnCondition>         
                    <OnEvent port="in">
                        <StateAssignment variable="v" value="v + deltaV"/>
                    </OnEvent>

                </Regime>
            </Dynamics>

        </ComponentType>


        <Component id="gen1" type="spikeGenerator" period="7ms"/>

        
        <Component id="multiregime" type="refractiaf" threshold="-50mV" v0="-80mV"
                   refractoryPeriod="20ms" capacitance="1pF" vreset="-80mV" vleak="-90mV" 
                   gleak="5pS" current="0.00001nA" deltaV="5mV"/>


        <ComponentType name="Network">
            <Children name="populations" type="Population"/>
            <Children name="connectivities" type="EventConnectivity"/>
        </ComponentType>


        <ComponentType name="Population">
            <ComponentReference name="component" type="Component"/>
            <Parameter name="size" dimension="none"/>
            <Structure>
                    <MultiInstantiate number="size" component="component"/>
            </Structure>
        </ComponentType>


        <ComponentType name="EventConnectivity">
            <Link name="source" type="Population"/>
            <Link name="target" type="Population"/>
            <Child name="Connections" type="ConnectionPattern"/>
        </ComponentType>


        <ComponentType name="ConnectionPattern"/>


        <ComponentType name="AllAll" extends="ConnectionPattern">
            <Structure>
                <ForEach instances="../source" as="a">
                    <ForEach instances="../target" as="b">
                        <EventConnection from="a" to="b"/>
                    </ForEach>
                </ForEach>    
            </Structure>
        </ComponentType>


        <Network id="net1">
            <Population id="p1" component="gen1" size="1"/>
            <Population id="p3" component="multiregime" size="2"/>

            <EventConnectivity id="p1-p3" source="p1" target="p3">
                <Connections type="AllAll"/>
            </EventConnectivity>
        </Network>


        <Include file="SingleSimulation.xml" />

        <Simulation id="sim1" length="80ms" step="0.05ms" target="net1">
            <Display id="d0" title="Example 8: Regimes in dynamics definitions" timeScale="1ms" xmin="-10" xmax="90" ymin="-90" ymax="20">
                <Line id="gen_vmr" quantity="p3[0]/v" scale="1mV" timeScale="1ms" color="#00c000"/>
                <Line id="gen_sv" quantity="p1[0]/tsince" scale="1ms" timeScale="1ms" color="#f00000"/>
            </Display>
        </Simulation>


    </Lems>
:::
:::
:::
:::
:::

[]{#document-Userdocs/Specification}

::: {#schema-specification .section .tex2jax_ignore .mathjax_ignore}
[]{#userdocs-specification}

## Schema/Specification[\#](#schema-specification "Link to this heading"){.headerlink}

::: {.admonition-neuroml-v2-3-is-the-current-stable-release-of-the-language-and-is-described-below .admonition}
NeuroML v2.3 is the current stable release of the language, and is
described below.

For an overview of the various releases of the language see: [[A brief
history of NeuroML]{.std .std-ref}](Landing.html#history){.reference
.internal}.
:::

We've briefly seen the XML representation of NeuroML models and
simulations in the [[Getting Started]{.std
.std-ref}](Landing.html#userdocs-getting-started-neuroml){.reference
.internal} tutorials. Here, we dive a little deeper into the underlying
details of NeuroML.

XML itself does not define a set of standard tags: any tags may be used
as long as the resultant document is
[well-formed](https://en.wikipedia.org/wiki/Well-formed_document){.reference
.external}. Therefore, NeuroML defines a standard set of XML elements
(the tags and attributes which specify the model and parameters, e.g.
[`<iafCell`{.docutils .literal .notranslate}]{.pre}` `{.docutils
.literal .notranslate}[`id="iaf"`{.docutils .literal
.notranslate}]{.pre}` `{.docutils .literal
.notranslate}[`leakReversal="-60mV"...>`{.docutils .literal
.notranslate}]{.pre}) that may be used in NeuroML documents: the NeuroML
[XML Schema
Definition](https://en.wikipedia.org/wiki/XML_Schema_(W3C)){.reference
.external}. This is referred to as the NeuroML *schema* or the NeuroML
*specification*.

As the wiki page says:

> <div>
>
> XSD (XML Schema Definition), a recommendation of the World Wide Web
> Consortium (W3C), specifies how to formally describe the elements in
> an Extensible Markup Language (XML) document. It can be used by
> programmers to verify each piece of item content in a document, to
> assure it adheres to the description of the element it is placed in.
>
> </div>

This gives us an idea of the advantages of using an XML based system.
All NeuroML models must use these pre-defined tags/components---this is
what we check for when we [[validate NeuroML models]{.std
.std-ref}](Landing.html#userdocs-validating-models){.reference
.internal}. A valid NeuroML model is said to adhere to the NeuroML
schema.

::: {.note .admonition}
Purpose of the NeuroML specification/schema.

The NeuroML schema/specification defines the structure of a valid
NeuroML document. The [[core NeuroML tools]{.std
.std-ref}](Landing.html#userdocs-software){.reference .internal} adhere
to this specification and can read/write/interpret the language
correctly.
:::

In the next section, we learn more about the NeuroML 2 schema, and see
how the dynamics of the NeuroML 2 entities are defined in LEMS.

::: {.toctree-wrapper .compound}
[]{#document-Userdocs/NeuroMLv2}

::: {#neuroml-v2 .section .tex2jax_ignore .mathjax_ignore}
[]{#userdocs-neuromlv2}

### NeuroML v2[\#](#neuroml-v2 "Link to this heading"){.headerlink}

The current stable version of NeuroML is v2.3, and the XSD Schema for
this can be found
[here](https://github.com/NeuroML/NeuroML2/blob/master/Schemas/NeuroML2/NeuroML_v2.3.xsd){.reference
.external}. The following figure, taken from Cannon et al. 2014
([\[[CGC+14](Landing.html#id10 "Robert C. Cannon, Padraig Gleeson, Sharon Crook, Gautham Ganapathy, Boris Marin, Eugenio Piasini, and R. Angus Silver. LEMS: a language for expressing complex biological models in concise and hierarchical form and its use in underpinning NeuroML 2. Frontiers in Neuroinformatics, 2014. doi:10.3389/fninf.2014.00079."){.reference
.internal}\]]{#id1}) shows some of the core elements defined in NeuroML
version 2 (note: these key elements haven't changed since that
publication).

<figure id="id2" class="align-center">
<img src="_images/Figure6a.png"
alt="Elements defined in the NeuroML schema, version 2." />
<figcaption><p><span class="caption-number">Fig. 57 </span><span
class="caption-text">Elements defined in the NeuroML schema, version
2.</span><a href="#id2" class="headerlink"
title="Link to this image">#</a></p></figcaption>
</figure>

You can see the complete definitions of NeuroML 2 entities in the
following pages. You can also search this documentation for specific
entities that you may be using in your NeuroML models.

Examples of files using the NeuroML 2 schema, and some of the elements
they use are:

::: pst-scrollable-table-container
  Example file                                                                                                                                                                                            NeuroML elements used
  ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- -----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
  [A simple cell with a morphology & segments arranged into groups](https://github.com/NeuroML/NeuroML2/tree/master/examples/NML2_SimpleMorphology.nml){.reference .external}                             [[\<cell\>]{.std .std-ref}](Landing.html#schema-cell){.reference .internal}, [[\<morphology\>]{.std .std-ref}](Landing.html#schema-morphology){.reference .internal}, [[\<segment\>]{.std .std-ref}](Landing.html#schema-segment){.reference .internal}, [[\<segmentGroup\>]{.std .std-ref}](Landing.html#schema-segmentgroup){.reference .internal}
  [A cell specifying biophysical properties (channel densities, passive electrical properties, etc.)](https://github.com/NeuroML/NeuroML2/tree/master/examples/NML2_FullCell.nml){.reference .external}   [[\<membraneProperties\>]{.std .std-ref}](Landing.html#schema-membraneproperties){.reference .internal}, [[\<intracellularProperties\>]{.std .std-ref}](Landing.html#schema-intracellularproperties){.reference .internal}, [[\<channelDensity\>]{.std .std-ref}](Landing.html#schema-channeldensity){.reference .internal}
  [A simple HH Na+ channel](https://github.com/NeuroML/NeuroML2/tree/master/examples/NML2_SimpleIonChannel.nml){.reference .external}                                                                     [[\<ionChannelHH\>]{.std .std-ref}](Landing.html#schema-ionchannelhh){.reference .internal}, [[\<gateHHrates\>]{.std .std-ref}](Landing.html#schema-gatehhrates){.reference .internal}, [[\<HHExpLinearRate\>]{.std .std-ref}](Landing.html#schema-hhexplinearrate){.reference .internal}
  [Some of the simplified spiking neuron models which are supported](https://github.com/NeuroML/NeuroML2/tree/master/examples/NML2_AbstractCells.nml){.reference .external}                               [[\<iafCell\>]{.std .std-ref}](Landing.html#schema-iafcell){.reference .internal}, [[\<izhikevich2007Cell\>]{.std .std-ref}](Landing.html#schema-izhikevich2007cell){.reference .internal}, [[\<adExIaFCell\>]{.std .std-ref}](Landing.html#schema-adexiafcell){.reference .internal}, [[\<fitzHughNagumoCell\>]{.std .std-ref}](Landing.html#schema-fitzhughnagumocell){.reference .internal}
  [Synapse models](https://github.com/NeuroML/NeuroML2/tree/master/examples/NML2_SynapseTypes.nml){.reference .external}                                                                                  [[\<alphaSynapse\>]{.std .std-ref}](Landing.html#schema-alphasynapse){.reference .internal}, [[\<expTwoSynapse\>]{.std .std-ref}](Landing.html#schema-exptwosynapse){.reference .internal}, [[\<blockingPlasticSynapse\>]{.std .std-ref}](Landing.html#schema-blockingplasticsynapse){.reference .internal}, [[\<doubleSynapse\>]{.std .std-ref}](Landing.html#schema-doublesynapse){.reference .internal}
  [A network of cells positioned in 3D and synaptically connected](https://github.com/NeuroML/NeuroML2/tree/master/examples/NML2_InstanceBasedNetwork.nml){.reference .external}                          [[\<network\>]{.std .std-ref}](Landing.html#schema-network){.reference .internal}, [[\<population\>]{.std .std-ref}](Landing.html#schema-population){.reference .internal}, [[\<projection\>]{.std .std-ref}](Landing.html#schema-projection){.reference .internal}, [[\<connection\>]{.std .std-ref}](Landing.html#schema-connection){.reference .internal}, [[\<inputList\>]{.std .std-ref}](Landing.html#schema-inputlist){.reference .internal}
:::

NeuroML files containing the XML representation of the model can be
[[validated]{.std
.std-ref}](Landing.html#userdocs-validating-models){.reference
.internal} to ensure all of the correct tags/attributes are present.

**But** how do we know how the model is actually meant to use the
specified attributes in an element? The schema only says that
[`leakReversal`{.docutils .literal .notranslate}]{.pre},
[`thresh`{.docutils .literal .notranslate}]{.pre}, etc. are allowed
attributes on [`iafCell`{.docutils .literal .notranslate}]{.pre}, but
how are these used to calculate the membrane potential? The answer lies
in another, lower-level language, called LEMS (Low Entropy Model
Specification).

While valid NeuroML entities are contained in the schema, their
underlying mathematical structure and composition rules must also be
defined. For this, NeuroML version 2 makes use of [[LEMS]{.std
.std-ref}](Landing.html#userdocs-lems){.reference .internal}.

::: {.toctree-wrapper .compound}
[]{#document-Userdocs/Schemas/NeuroMLDocument}

::: {#neuromldocument .section .tex2jax_ignore .mathjax_ignore}
[]{#schema-neuroml-main}

#### NeuroMLDocument[\#](#neuromldocument "Link to this heading"){.headerlink}

**The main NeuroML container class, and other associated types that do
not fit into the other categories**

------------------------------------------------------------------------

Schema against which NeuroML based on these should be valid:
[NeuroML_v2.3.xsd](https://github.com/NeuroML/NeuroML2/tree/master/Schemas/NeuroML2/NeuroML_v2.3.xsd){.reference
.external}. Please file any issues or questions at the [issue tracker
here](https://github.com/NeuroML/NeuroML2/issues){.reference .external}.

------------------------------------------------------------------------

::: {#schema-neuromldocument .section}
[]{#id1}

##### NeuroMLDocument[\#](#schema-neuromldocument "Link to this heading"){.headerlink}

::: {.sd-tab-set .docutils}
Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
      <xs:element name="neuroml" type="NeuroMLDocument">
        <xs:annotation>
          <xs:documentation>The root NeuroML element.</xs:documentation>
        </xs:annotation>
      </xs:element>

      <xs:complexType name="NeuroMLDocument">
        <xs:complexContent>
          <xs:extension base="Standalone">
            <xs:sequence>
              <xs:element name="include" type="IncludeType" minOccurs="0" maxOccurs="unbounded"/>
              <xs:element name="extracellularProperties" type="ExtracellularProperties" minOccurs="0" maxOccurs="unbounded"/>
              <xs:element name="intracellularProperties" type="IntracellularProperties" minOccurs="0" maxOccurs="unbounded"/>
              <xs:element name="morphology" type="Morphology" minOccurs="0" maxOccurs="unbounded"/>
              <xs:element name="ionChannel" type="IonChannel" minOccurs="0" maxOccurs="unbounded"/>
              <xs:element name="ionChannelHH" type="IonChannelHH" minOccurs="0" maxOccurs="unbounded"/>
              <xs:element name="ionChannelVShift" type="IonChannelVShift" minOccurs="0" maxOccurs="unbounded"/>
              <xs:element name="ionChannelKS" type="IonChannelKS" minOccurs="0" maxOccurs="unbounded"/>
              <xs:group ref="ConcentrationModelTypes"/>
              <xs:group ref="SynapseTypes"/>
              <xs:element name="biophysicalProperties" type="BiophysicalProperties" minOccurs="0" maxOccurs="unbounded"/>
              <xs:group ref="CellTypes"/>
              <xs:group ref="InputTypes"/>
              <xs:group ref="PyNNCellTypes"/>
              <xs:group ref="PyNNSynapseTypes"/>
              <xs:group ref="PyNNInputTypes"/>
              <xs:element name="network" type="Network" minOccurs="0" maxOccurs="unbounded"/>
              <xs:element name="ComponentType" type="ComponentType" minOccurs="0" maxOccurs="unbounded"/>
            </xs:sequence>
          </xs:extension>
        </xs:complexContent>
      </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=NeuroMLDocument){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import NeuroMLDocument
    from neuroml.utils import component_factory

    variable = component_factory(
        NeuroMLDocument,
        id: 'a NmlId (required)' = None
        metaid: 'a MetaId (optional)' = None
        notes: 'a string (optional)' = None
        properties: 'list of Property(s) (optional)' = None
        annotation: 'a Annotation (optional)' = None
        includes: 'list of IncludeType(s) (optional)' = None
        extracellular_properties: 'list of ExtracellularProperties(s) (optional)' = None
        intracellular_properties: 'list of IntracellularProperties(s) (optional)' = None
        morphology: 'list of Morphology(s) (optional)' = None
        ion_channel: 'list of IonChannel(s) (optional)' = None
        ion_channel_hhs: 'list of IonChannelHH(s) (optional)' = None
        ion_channel_v_shifts: 'list of IonChannelVShift(s) (optional)' = None
        ion_channel_kses: 'list of IonChannelKS(s) (optional)' = None
        decaying_pool_concentration_models: 'list of DecayingPoolConcentrationModel(s) (optional)' = None
        fixed_factor_concentration_models: 'list of FixedFactorConcentrationModel(s) (optional)' = None
        alpha_current_synapses: 'list of AlphaCurrentSynapse(s) (optional)' = None
        alpha_synapses: 'list of AlphaSynapse(s) (optional)' = None
        exp_one_synapses: 'list of ExpOneSynapse(s) (optional)' = None
        exp_two_synapses: 'list of ExpTwoSynapse(s) (optional)' = None
        exp_three_synapses: 'list of ExpThreeSynapse(s) (optional)' = None
        blocking_plastic_synapses: 'list of BlockingPlasticSynapse(s) (optional)' = None
        double_synapses: 'list of DoubleSynapse(s) (optional)' = None
        gap_junctions: 'list of GapJunction(s) (optional)' = None
        silent_synapses: 'list of SilentSynapse(s) (optional)' = None
        linear_graded_synapses: 'list of LinearGradedSynapse(s) (optional)' = None
        graded_synapses: 'list of GradedSynapse(s) (optional)' = None
        biophysical_properties: 'list of BiophysicalProperties(s) (optional)' = None
        cells: 'list of Cell(s) (optional)' = None
        cell2_ca_poolses: 'list of Cell2CaPools(s) (optional)' = None
        base_cells: 'list of BaseCell(s) (optional)' = None
        iaf_tau_cells: 'list of IafTauCell(s) (optional)' = None
        iaf_tau_ref_cells: 'list of IafTauRefCell(s) (optional)' = None
        iaf_cells: 'list of IafCell(s) (optional)' = None
        iaf_ref_cells: 'list of IafRefCell(s) (optional)' = None
        izhikevich_cells: 'list of IzhikevichCell(s) (optional)' = None
        izhikevich2007_cells: 'list of Izhikevich2007Cell(s) (optional)' = None
        ad_ex_ia_f_cells: 'list of AdExIaFCell(s) (optional)' = None
        fitz_hugh_nagumo_cells: 'list of FitzHughNagumoCell(s) (optional)' = None
        fitz_hugh_nagumo1969_cells: 'list of FitzHughNagumo1969Cell(s) (optional)' = None
        pinsky_rinzel_ca3_cells: 'list of PinskyRinzelCA3Cell(s) (optional)' = None
        hindmarshRose1984Cell: 'list of HindmarshRose1984Cell(s) (optional)' = None
        pulse_generators: 'list of PulseGenerator(s) (optional)' = None
        pulse_generator_dls: 'list of PulseGeneratorDL(s) (optional)' = None
        sine_generators: 'list of SineGenerator(s) (optional)' = None
        sine_generator_dls: 'list of SineGeneratorDL(s) (optional)' = None
        ramp_generators: 'list of RampGenerator(s) (optional)' = None
        ramp_generator_dls: 'list of RampGeneratorDL(s) (optional)' = None
        compound_inputs: 'list of CompoundInput(s) (optional)' = None
        compound_input_dls: 'list of CompoundInputDL(s) (optional)' = None
        voltage_clamps: 'list of VoltageClamp(s) (optional)' = None
        voltage_clamp_triples: 'list of VoltageClampTriple(s) (optional)' = None
        spike_arrays: 'list of SpikeArray(s) (optional)' = None
        timed_synaptic_inputs: 'list of TimedSynapticInput(s) (optional)' = None
        spike_generators: 'list of SpikeGenerator(s) (optional)' = None
        spike_generator_randoms: 'list of SpikeGeneratorRandom(s) (optional)' = None
        spike_generator_poissons: 'list of SpikeGeneratorPoisson(s) (optional)' = None
        spike_generator_ref_poissons: 'list of SpikeGeneratorRefPoisson(s) (optional)' = None
        poisson_firing_synapses: 'list of PoissonFiringSynapse(s) (optional)' = None
        transient_poisson_firing_synapses: 'list of TransientPoissonFiringSynapse(s) (optional)' = None
        IF_curr_alpha: 'list of IF_curr_alpha(s) (optional)' = None
        IF_curr_exp: 'list of IF_curr_exp(s) (optional)' = None
        IF_cond_alpha: 'list of IF_cond_alpha(s) (optional)' = None
        IF_cond_exp: 'list of IF_cond_exp(s) (optional)' = None
        EIF_cond_exp_isfa_ista: 'list of EIF_cond_exp_isfa_ista(s) (optional)' = None
        EIF_cond_alpha_isfa_ista: 'list of EIF_cond_alpha_isfa_ista(s) (optional)' = None
        HH_cond_exp: 'list of HH_cond_exp(s) (optional)' = None
        exp_cond_synapses: 'list of ExpCondSynapse(s) (optional)' = None
        alpha_cond_synapses: 'list of AlphaCondSynapse(s) (optional)' = None
        exp_curr_synapses: 'list of ExpCurrSynapse(s) (optional)' = None
        alpha_curr_synapses: 'list of AlphaCurrSynapse(s) (optional)' = None
        SpikeSourcePoisson: 'list of SpikeSourcePoisson(s) (optional)' = None
        networks: 'list of Network(s) (optional)' = None
        ComponentType: 'list of ComponentType(s) (optional)' = None
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <neuroml xmlns="http://www.neuroml.org/schema/neuroml2"  xmlns:xs="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.neuroml.org/schema/neuroml2 https://raw.github.com/NeuroML/NeuroML2/development/Schemas/NeuroML2/NeuroML_v2.3.xsd" id="HL23PYR">
        <include href="HL23PYR.cell.nml"/>
        <pulseGenerator id="pg_HL23PYR" delay="50ms" duration="200ms" amplitude="0.2nA">
            <notes>Simple pulse generator</notes>
        </pulseGenerator>
        <network id="HL23PYRNet" type="networkWithTemperature" temperature="34 degC">
            <population id="HL23PYR_pop" component="HL23PYR" type="populationList">
                <property tag="color" value="0.3220229197377431 0.19279726280626452 0.37635392246534727"/>
                <property tag="region" value="L23"/>
                <instance id="0">
                    <location x="0.0" y="0.0" z="0.0"/>
                </instance>
            </population>
            <inputList id="stim_iclamp_HL23PYR" population="HL23PYR_pop" component="pg_HL23PYR">
                <input id="0" target="../HL23PYR_pop/0" destination="synapses"/>
            </inputList>
        </network>
    </neuroml>
:::
:::
:::
:::
:::

::: {#includetype .section}
[]{#schema-includetype}

##### IncludeType[\#](#includetype "Link to this heading"){.headerlink}

*Used to include other documents into each other.*

::: {.sd-tab-set .docutils}
Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
      <xs:complexType name="IncludeType">
        <xs:attribute name="href" use="required" type="xs:anyURI"/>
      </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=IncludeType){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import IncludeType
    from neuroml.utils import component_factory

    variable = component_factory(
        IncludeType,
        href: 'a anyURI (required)' = None)
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <neuroml xmlns="http://www.neuroml.org/schema/neuroml2"  xmlns:xs="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.neuroml.org/schema/neuroml2 https://raw.github.com/NeuroML/NeuroML2/development/Schemas/NeuroML2/NeuroML_v2.3.xsd" id="HL23PYR">
        <include href="A.cell.nml"/>
        ..
    </neuroml>
:::
:::
:::
:::
:::
:::

[]{#document-Userdocs/Schemas/NeuroMLCoreDimensions}

::: {#neuromlcoredimensions .section .tex2jax_ignore .mathjax_ignore}
[]{#schema-neuromlcoredimensions}

#### NeuroMLCoreDimensions[\#](#neuromlcoredimensions "Link to this heading"){.headerlink}

Original ComponentType definitions:
[NeuroMLCoreDimensions.xml](https://github.com/NeuroML/NeuroML2/blob/master/NeuroML2CoreTypes//NeuroMLCoreDimensions.xml){.reference
.external}. Schema against which NeuroML based on these should be valid:
[NeuroML_v2.3.xsd](https://github.com/NeuroML/NeuroML2/tree/master/Schemas/NeuroML2/NeuroML_v2.3.xsd){.reference
.external}. Generated on 14/08/24 from
[this](https://github.com/NeuroML/NeuroML2/commit/352244cff605cb1ba24fa7c11757dc818fe90fd2){.reference
.external} commit. Please file any issues or questions at the [issue
tracker here](https://github.com/NeuroML/NeuroML2/issues){.reference
.external}.

------------------------------------------------------------------------

::: {#dimensions .section}
[]{#schema-dimensions}

##### Dimensions[\#](#dimensions "Link to this heading"){.headerlink}

::: {#area .section}
[]{#schema-dimensions-area}

###### area[\#](#area "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .sd-col-6 .sd-col-xs-6 .sd-col-sm-6 .sd-col-md-6 .sd-col-lg-6 .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Dimensions
:::

L^2^
:::
:::
:::

::: {.sd-col .sd-d-flex-row .sd-col-6 .sd-col-xs-6 .sd-col-sm-6 .sd-col-md-6 .sd-col-lg-6 .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Units
:::

-   Defined unit: [[cm2]{.std .std-ref}](#schema-units-cm2){.reference
    .internal}

-   Defined unit: [[m2]{.std .std-ref}](#schema-units-m2){.reference
    .internal}

-   Defined unit: [[um2]{.std .std-ref}](#schema-units-um2){.reference
    .internal}
:::
:::
:::
:::
:::
:::

::: {#capacitance .section}
[]{#schema-dimensions-capacitance}

###### capacitance[\#](#capacitance "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .sd-col-6 .sd-col-xs-6 .sd-col-sm-6 .sd-col-md-6 .sd-col-lg-6 .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Dimensions
:::

M^-1^ L^-2^ T^4^ I^2^
:::
:::
:::

::: {.sd-col .sd-d-flex-row .sd-col-6 .sd-col-xs-6 .sd-col-sm-6 .sd-col-md-6 .sd-col-lg-6 .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Units
:::

-   Defined unit: [[F]{.std .std-ref}](#schema-units-f){.reference
    .internal}

-   Defined unit: [[nF]{.std .std-ref}](#schema-units-nf){.reference
    .internal}

-   Defined unit: [[pF]{.std .std-ref}](#schema-units-pf){.reference
    .internal}

-   Defined unit: [[uF]{.std .std-ref}](#schema-units-uf){.reference
    .internal}
:::
:::
:::

::: {.sd-col .sd-d-flex-row .sd-col-12 .sd-col-xs-12 .sd-col-sm-12 .sd-col-md-12 .sd-col-lg-12 .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Schema
:::

::: {.highlight-xml .notranslate}
::: highlight
    <xs:simpleType name="Nml2Quantity_capacitance">
      <xs:restriction base="xs:string">
        <xs:pattern value="-?([0-9]*(\.[0-9]+)?)([eE]-?[0-9]+)?[\s]*(F|uF|nF|pF)"/>
      </xs:restriction>
    </xs:simpleType>
:::
:::
:::
:::
:::
:::
:::
:::

::: {#charge .section}
[]{#schema-dimensions-charge}

###### charge[\#](#charge "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .sd-col-6 .sd-col-xs-6 .sd-col-sm-6 .sd-col-md-6 .sd-col-lg-6 .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Dimensions
:::

T^1^ I^1^
:::
:::
:::

::: {.sd-col .sd-d-flex-row .sd-col-6 .sd-col-xs-6 .sd-col-sm-6 .sd-col-md-6 .sd-col-lg-6 .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Units
:::

-   Defined unit: [[C]{.std .std-ref}](#schema-units-c){.reference
    .internal}

-   Defined unit: [[e]{.std .std-ref}](#schema-units-e){.reference
    .internal}
:::
:::
:::
:::
:::
:::

::: {#charge-per-mole .section}
[]{#schema-dimensions-charge-per-mole}

###### charge_per_mole[\#](#charge-per-mole "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .sd-col-6 .sd-col-xs-6 .sd-col-sm-6 .sd-col-md-6 .sd-col-lg-6 .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Dimensions
:::

T^1^ I^1^ N^-1^
:::
:::
:::

::: {.sd-col .sd-d-flex-row .sd-col-6 .sd-col-xs-6 .sd-col-sm-6 .sd-col-md-6 .sd-col-lg-6 .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Units
:::

-   Defined unit: [[C_per_mol]{.std
    .std-ref}](#schema-units-c-per-mol){.reference .internal}

-   Defined unit: [[nA_ms_per_amol]{.std
    .std-ref}](#schema-units-na-ms-per-amol){.reference .internal}

-   Defined unit: [[pC_per_umol]{.std
    .std-ref}](#schema-units-pc-per-umol){.reference .internal}
:::
:::
:::
:::
:::
:::

::: {#concentration .section}
[]{#schema-dimensions-concentration}

###### concentration[\#](#concentration "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .sd-col-6 .sd-col-xs-6 .sd-col-sm-6 .sd-col-md-6 .sd-col-lg-6 .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Dimensions
:::

L^-3^ N^1^
:::
:::
:::

::: {.sd-col .sd-d-flex-row .sd-col-6 .sd-col-xs-6 .sd-col-sm-6 .sd-col-md-6 .sd-col-lg-6 .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Units
:::

-   Defined unit: [[M]{.std .std-ref}](#schema-units-m){.reference
    .internal}

-   Defined unit: [[mM]{.std .std-ref}](#schema-units-mm){.reference
    .internal}

-   Defined unit: [[mol_per_cm3]{.std
    .std-ref}](#schema-units-mol-per-cm3){.reference .internal}

-   Defined unit: [[mol_per_m3]{.std
    .std-ref}](#schema-units-mol-per-m3){.reference .internal}
:::
:::
:::

::: {.sd-col .sd-d-flex-row .sd-col-12 .sd-col-xs-12 .sd-col-sm-12 .sd-col-md-12 .sd-col-lg-12 .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Schema
:::

::: {.highlight-xml .notranslate}
::: highlight
    <xs:simpleType name="Nml2Quantity_concentration">
      <xs:restriction base="xs:string">
        <xs:pattern value="-?([0-9]*(\.[0-9]+)?)([eE]-?[0-9]+)?[\s]*(mol_per_m3|mol_per_cm3|M|mM)"/>
      </xs:restriction>
    </xs:simpleType>
:::
:::
:::
:::
:::
:::
:::
:::

::: {#conductance .section}
[]{#schema-dimensions-conductance}

###### conductance[\#](#conductance "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .sd-col-6 .sd-col-xs-6 .sd-col-sm-6 .sd-col-md-6 .sd-col-lg-6 .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Dimensions
:::

M^-1^ L^-2^ T^3^ I^2^
:::
:::
:::

::: {.sd-col .sd-d-flex-row .sd-col-6 .sd-col-xs-6 .sd-col-sm-6 .sd-col-md-6 .sd-col-lg-6 .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Units
:::

-   Defined unit: [[S]{.std .std-ref}](#schema-units-s){.reference
    .internal}

-   Defined unit: [[mS]{.std .std-ref}](#schema-units-ms){.reference
    .internal}

-   Defined unit: [[nS]{.std .std-ref}](#schema-units-ns){.reference
    .internal}

-   Defined unit: [[pS]{.std .std-ref}](#schema-units-ps){.reference
    .internal}

-   Defined unit: [[uS]{.std .std-ref}](#schema-units-us){.reference
    .internal}
:::
:::
:::

::: {.sd-col .sd-d-flex-row .sd-col-12 .sd-col-xs-12 .sd-col-sm-12 .sd-col-md-12 .sd-col-lg-12 .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Schema
:::

::: {.highlight-xml .notranslate}
::: highlight
    <xs:simpleType name="Nml2Quantity_conductance">
      <xs:restriction base="xs:string">
        <xs:pattern value="-?([0-9]*(\.[0-9]+)?)([eE]-?[0-9]+)?[\s]*(S|mS|uS|nS|pS)"/>
      </xs:restriction>
    </xs:simpleType>
:::
:::
:::
:::
:::
:::
:::
:::

::: {#conductancedensity .section}
[]{#schema-dimensions-conductancedensity}

###### conductanceDensity[\#](#conductancedensity "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .sd-col-6 .sd-col-xs-6 .sd-col-sm-6 .sd-col-md-6 .sd-col-lg-6 .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Dimensions
:::

M^-1^ L^-4^ T^3^ I^2^
:::
:::
:::

::: {.sd-col .sd-d-flex-row .sd-col-6 .sd-col-xs-6 .sd-col-sm-6 .sd-col-md-6 .sd-col-lg-6 .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Units
:::

-   Defined unit: [[S_per_cm2]{.std
    .std-ref}](#schema-units-s-per-cm2){.reference .internal}

-   Defined unit: [[S_per_m2]{.std
    .std-ref}](#schema-units-s-per-m2){.reference .internal}

-   Defined unit: [[mS_per_cm2]{.std
    .std-ref}](#schema-units-ms-per-cm2){.reference .internal}

-   Defined unit: [[uS_per_cm2]{.std
    .std-ref}](#schema-units-us-per-cm2){.reference .internal}
:::
:::
:::
:::
:::
:::

::: {#conductance-per-voltage .section}
[]{#schema-dimensions-conductance-per-voltage}

###### conductance_per_voltage[\#](#conductance-per-voltage "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .sd-col-6 .sd-col-xs-6 .sd-col-sm-6 .sd-col-md-6 .sd-col-lg-6 .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Dimensions
:::

M^-2^ L^-4^ T^6^ I^3^
:::
:::
:::

::: {.sd-col .sd-d-flex-row .sd-col-6 .sd-col-xs-6 .sd-col-sm-6 .sd-col-md-6 .sd-col-lg-6 .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Units
:::

-   Defined unit: [[S_per_V]{.std
    .std-ref}](#schema-units-s-per-v){.reference .internal}

-   Defined unit: [[nS_per_mV]{.std
    .std-ref}](#schema-units-ns-per-mv){.reference .internal}
:::
:::
:::

::: {.sd-col .sd-d-flex-row .sd-col-12 .sd-col-xs-12 .sd-col-sm-12 .sd-col-md-12 .sd-col-lg-12 .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Schema
:::

::: {.highlight-xml .notranslate}
::: highlight
    <xs:simpleType name="Nml2Quantity_conductancePerVoltage">
      <xs:restriction base="xs:string">
        <xs:pattern value="-?([0-9]*(\.[0-9]+)?)([eE]-?[0-9]+)?[\s]*(S_per_V|nS_per_mV)"/>
      </xs:restriction>
    </xs:simpleType>
:::
:::
:::
:::
:::
:::
:::
:::

::: {#current .section}
[]{#schema-dimensions-current}

###### current[\#](#current "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .sd-col-6 .sd-col-xs-6 .sd-col-sm-6 .sd-col-md-6 .sd-col-lg-6 .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Dimensions
:::

I^1^
:::
:::
:::

::: {.sd-col .sd-d-flex-row .sd-col-6 .sd-col-xs-6 .sd-col-sm-6 .sd-col-md-6 .sd-col-lg-6 .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Units
:::

-   Defined unit: [[A]{.std .std-ref}](#schema-units-a){.reference
    .internal}

-   Defined unit: [[nA]{.std .std-ref}](#schema-units-na){.reference
    .internal}

-   Defined unit: [[pA]{.std .std-ref}](#schema-units-pa){.reference
    .internal}

-   Defined unit: [[uA]{.std .std-ref}](#schema-units-ua){.reference
    .internal}
:::
:::
:::

::: {.sd-col .sd-d-flex-row .sd-col-12 .sd-col-xs-12 .sd-col-sm-12 .sd-col-md-12 .sd-col-lg-12 .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Schema
:::

::: {.highlight-xml .notranslate}
::: highlight
    <xs:simpleType name="Nml2Quantity_current">
      <xs:restriction base="xs:string">
        <xs:pattern value="-?([0-9]*(\.[0-9]+)?)([eE]-?[0-9]+)?[\s]*(A|uA|nA|pA)"/>
      </xs:restriction>
    </xs:simpleType>
:::
:::
:::
:::
:::
:::
:::
:::

::: {#currentdensity .section}
[]{#schema-dimensions-currentdensity}

###### currentDensity[\#](#currentdensity "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .sd-col-6 .sd-col-xs-6 .sd-col-sm-6 .sd-col-md-6 .sd-col-lg-6 .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Dimensions
:::

L^-2^ I^1^
:::
:::
:::

::: {.sd-col .sd-d-flex-row .sd-col-6 .sd-col-xs-6 .sd-col-sm-6 .sd-col-md-6 .sd-col-lg-6 .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Units
:::

-   Defined unit: [[A_per_m2]{.std
    .std-ref}](#schema-units-a-per-m2){.reference .internal}

-   Defined unit: [[mA_per_cm2]{.std
    .std-ref}](#schema-units-ma-per-cm2){.reference .internal}

-   Defined unit: [[uA_per_cm2]{.std
    .std-ref}](#schema-units-ua-per-cm2){.reference .internal}
:::
:::
:::
:::
:::
:::

::: {#schema-dimensions-idealgasconstantdims .section}
[]{#idealgasconstantdims}

###### idealGasConstantDims[\#](#schema-dimensions-idealgasconstantdims "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .sd-col-6 .sd-col-xs-6 .sd-col-sm-6 .sd-col-md-6 .sd-col-lg-6 .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Dimensions
:::

M^1^ L^2^ T^-2^ K^-1^ N^-1^
:::
:::
:::

::: {.sd-col .sd-d-flex-row .sd-col-6 .sd-col-xs-6 .sd-col-sm-6 .sd-col-md-6 .sd-col-lg-6 .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Units
:::

-   Defined unit: [[J_per_K_per_mol]{.std
    .std-ref}](#schema-units-j-per-k-per-mol){.reference .internal}

-   Defined unit: [[fJ_per_K_per_umol]{.std
    .std-ref}](#schema-units-fj-per-k-per-umol){.reference .internal}
:::
:::
:::
:::
:::
:::

::: {#length .section}
[]{#schema-dimensions-length}

###### length[\#](#length "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .sd-col-6 .sd-col-xs-6 .sd-col-sm-6 .sd-col-md-6 .sd-col-lg-6 .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Dimensions
:::

L^1^
:::
:::
:::

::: {.sd-col .sd-d-flex-row .sd-col-6 .sd-col-xs-6 .sd-col-sm-6 .sd-col-md-6 .sd-col-lg-6 .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Units
:::

-   Defined unit: [[cm]{.std .std-ref}](#schema-units-cm){.reference
    .internal}

-   Defined unit: [[m]{.std .std-ref}](#id1){.reference .internal}

-   Defined unit: [[um]{.std .std-ref}](#schema-units-um){.reference
    .internal}
:::
:::
:::

::: {.sd-col .sd-d-flex-row .sd-col-12 .sd-col-xs-12 .sd-col-sm-12 .sd-col-md-12 .sd-col-lg-12 .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Schema
:::

::: {.highlight-xml .notranslate}
::: highlight
    <xs:simpleType name="Nml2Quantity_length">
      <xs:restriction base="xs:string">
        <xs:pattern value="-?([0-9]*(\.[0-9]+)?)([eE]-?[0-9]+)?[\s]*(m|cm|um)"/>
      </xs:restriction>
    </xs:simpleType>
:::
:::
:::
:::
:::
:::
:::
:::

::: {#per-time .section}
[]{#schema-dimensions-per-time}

###### per_time[\#](#per-time "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .sd-col-6 .sd-col-xs-6 .sd-col-sm-6 .sd-col-md-6 .sd-col-lg-6 .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Dimensions
:::

T^-1^
:::
:::
:::

::: {.sd-col .sd-d-flex-row .sd-col-6 .sd-col-xs-6 .sd-col-sm-6 .sd-col-md-6 .sd-col-lg-6 .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Units
:::

-   Defined unit: [[Hz]{.std .std-ref}](#schema-units-hz){.reference
    .internal}

-   Defined unit: [[per_hour]{.std
    .std-ref}](#schema-units-per-hour){.reference .internal}

-   Defined unit: [[per_min]{.std
    .std-ref}](#schema-units-per-min){.reference .internal}

-   Defined unit: [[per_ms]{.std
    .std-ref}](#schema-units-per-ms){.reference .internal}

-   Defined unit: [[per_s]{.std
    .std-ref}](#schema-units-per-s){.reference .internal}
:::
:::
:::

::: {.sd-col .sd-d-flex-row .sd-col-12 .sd-col-xs-12 .sd-col-sm-12 .sd-col-md-12 .sd-col-lg-12 .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Schema
:::

::: {.highlight-xml .notranslate}
::: highlight
    <xs:simpleType name="Nml2Quantity_pertime">
      <xs:restriction base="xs:string">
        <xs:pattern value="-?([0-9]*(\.[0-9]+)?)([eE]-?[0-9]+)?[\s]*(per_s|per_ms|Hz)"/>
      </xs:restriction>
    </xs:simpleType>
:::
:::
:::
:::
:::
:::
:::
:::

::: {#per-voltage .section}
[]{#schema-dimensions-per-voltage}

###### per_voltage[\#](#per-voltage "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .sd-col-6 .sd-col-xs-6 .sd-col-sm-6 .sd-col-md-6 .sd-col-lg-6 .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Dimensions
:::

M^-1^ L^-2^ T^3^ I^1^
:::
:::
:::

::: {.sd-col .sd-d-flex-row .sd-col-6 .sd-col-xs-6 .sd-col-sm-6 .sd-col-md-6 .sd-col-lg-6 .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Units
:::

-   Defined unit: [[per_V]{.std
    .std-ref}](#schema-units-per-v){.reference .internal}

-   Defined unit: [[per_mV]{.std
    .std-ref}](#schema-units-per-mv){.reference .internal}
:::
:::
:::
:::
:::
:::

::: {#permeability .section}
[]{#schema-dimensions-permeability}

###### permeability[\#](#permeability "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .sd-col-6 .sd-col-xs-6 .sd-col-sm-6 .sd-col-md-6 .sd-col-lg-6 .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Dimensions
:::

L^1^ T^-1^
:::
:::
:::

::: {.sd-col .sd-d-flex-row .sd-col-6 .sd-col-xs-6 .sd-col-sm-6 .sd-col-md-6 .sd-col-lg-6 .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Units
:::

-   Defined unit: [[cm_per_ms]{.std
    .std-ref}](#schema-units-cm-per-ms){.reference .internal}

-   Defined unit: [[cm_per_s]{.std
    .std-ref}](#schema-units-cm-per-s){.reference .internal}

-   Defined unit: [[m_per_s]{.std
    .std-ref}](#schema-units-m-per-s){.reference .internal}

-   Defined unit: [[um_per_ms]{.std
    .std-ref}](#schema-units-um-per-ms){.reference .internal}
:::
:::
:::

::: {.sd-col .sd-d-flex-row .sd-col-12 .sd-col-xs-12 .sd-col-sm-12 .sd-col-md-12 .sd-col-lg-12 .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Schema
:::

::: {.highlight-xml .notranslate}
::: highlight
    <xs:simpleType name="Nml2Quantity_permeability">
      <xs:restriction base="xs:string">
        <xs:pattern value="-?([0-9]*(\.[0-9]+)?)([eE]-?[0-9]+)?[\s]*(m_per_s|um_per_ms|cm_per_s|cm_per_ms)"/>
      </xs:restriction>
    </xs:simpleType>
:::
:::
:::
:::
:::
:::
:::
:::

::: {#resistance .section}
[]{#schema-dimensions-resistance}

###### resistance[\#](#resistance "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .sd-col-6 .sd-col-xs-6 .sd-col-sm-6 .sd-col-md-6 .sd-col-lg-6 .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Dimensions
:::

M^1^ L^2^ T^-3^ I^-2^
:::
:::
:::

::: {.sd-col .sd-d-flex-row .sd-col-6 .sd-col-xs-6 .sd-col-sm-6 .sd-col-md-6 .sd-col-lg-6 .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Units
:::

-   Defined unit: [[Mohm]{.std .std-ref}](#schema-units-mohm){.reference
    .internal}

-   Defined unit: [[kohm]{.std .std-ref}](#schema-units-kohm){.reference
    .internal}

-   Defined unit: [[ohm]{.std .std-ref}](#schema-units-ohm){.reference
    .internal}
:::
:::
:::

::: {.sd-col .sd-d-flex-row .sd-col-12 .sd-col-xs-12 .sd-col-sm-12 .sd-col-md-12 .sd-col-lg-12 .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Schema
:::

::: {.highlight-xml .notranslate}
::: highlight
    <xs:simpleType name="Nml2Quantity_resistance">
      <xs:restriction base="xs:string">
        <xs:pattern value="-?([0-9]*(\.[0-9]+)?)([eE]-?[0-9]+)?[\s]*(ohm|kohm|Mohm)"/>
      </xs:restriction>
    </xs:simpleType>
:::
:::
:::
:::
:::
:::
:::
:::

::: {#resistivity .section}
[]{#schema-dimensions-resistivity}

###### resistivity[\#](#resistivity "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .sd-col-6 .sd-col-xs-6 .sd-col-sm-6 .sd-col-md-6 .sd-col-lg-6 .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Dimensions
:::

M^2^ L^2^ T^-3^ I^-2^
:::
:::
:::

::: {.sd-col .sd-d-flex-row .sd-col-6 .sd-col-xs-6 .sd-col-sm-6 .sd-col-md-6 .sd-col-lg-6 .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Units
:::

-   Defined unit: [[kohm_cm]{.std
    .std-ref}](#schema-units-kohm-cm){.reference .internal}

-   Defined unit: [[ohm_cm]{.std
    .std-ref}](#schema-units-ohm-cm){.reference .internal}

-   Defined unit: [[ohm_m]{.std
    .std-ref}](#schema-units-ohm-m){.reference .internal}
:::
:::
:::

::: {.sd-col .sd-d-flex-row .sd-col-12 .sd-col-xs-12 .sd-col-sm-12 .sd-col-md-12 .sd-col-lg-12 .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Schema
:::

::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="Resistivity">
      <xs:complexContent>
        <xs:extension base="BaseWithoutId">
          <xs:attribute name="value" type="Nml2Quantity_resistivity" use="required"/>
          <xs:attribute name="segmentGroup" type="NmlId" use="optional" default="all"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::
:::
:::
:::
:::
:::

::: {#rho-factor .section}
[]{#schema-dimensions-rho-factor}

###### rho_factor[\#](#rho-factor "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .sd-col-6 .sd-col-xs-6 .sd-col-sm-6 .sd-col-md-6 .sd-col-lg-6 .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Dimensions
:::

L^-1^ T^-1^ I^-1^ N^1^
:::
:::
:::

::: {.sd-col .sd-d-flex-row .sd-col-6 .sd-col-xs-6 .sd-col-sm-6 .sd-col-md-6 .sd-col-lg-6 .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Units
:::

-   Defined unit: [[mol_per_cm_per_uA_per_ms]{.std
    .std-ref}](#schema-units-mol-per-cm-per-ua-per-ms){.reference
    .internal}

-   Defined unit: [[mol_per_m_per_A_per_s]{.std
    .std-ref}](#schema-units-mol-per-m-per-a-per-s){.reference
    .internal}

-   Defined unit: [[umol_per_cm_per_nA_per_ms]{.std
    .std-ref}](#schema-units-umol-per-cm-per-na-per-ms){.reference
    .internal}
:::
:::
:::

::: {.sd-col .sd-d-flex-row .sd-col-12 .sd-col-xs-12 .sd-col-sm-12 .sd-col-md-12 .sd-col-lg-12 .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Schema
:::

::: {.highlight-xml .notranslate}
::: highlight
    <xs:simpleType name="Nml2Quantity_rhoFactor">
      <xs:restriction base="xs:string">
        <xs:pattern value="-?([0-9]*(\.[0-9]+)?)([eE]-?[0-9]+)?[\s]*(mol_per_m_per_A_per_s|mol_per_cm_per_uA_per_ms)"/>
      </xs:restriction>
    </xs:simpleType>
:::
:::
:::
:::
:::
:::
:::
:::

::: {#specificcapacitance .section}
[]{#schema-dimensions-specificcapacitance}

###### specificCapacitance[\#](#specificcapacitance "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .sd-col-6 .sd-col-xs-6 .sd-col-sm-6 .sd-col-md-6 .sd-col-lg-6 .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Dimensions
:::

M^-1^ L^-4^ T^4^ I^2^
:::
:::
:::

::: {.sd-col .sd-d-flex-row .sd-col-6 .sd-col-xs-6 .sd-col-sm-6 .sd-col-md-6 .sd-col-lg-6 .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Units
:::

-   Defined unit: [[F_per_m2]{.std
    .std-ref}](#schema-units-f-per-m2){.reference .internal}

-   Defined unit: [[uF_per_cm2]{.std
    .std-ref}](#schema-units-uf-per-cm2){.reference .internal}
:::
:::
:::
:::
:::
:::

::: {#substance .section}
[]{#schema-dimensions-substance}

###### substance[\#](#substance "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .sd-col-6 .sd-col-xs-6 .sd-col-sm-6 .sd-col-md-6 .sd-col-lg-6 .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Dimensions
:::

N^1^
:::
:::
:::

::: {.sd-col .sd-d-flex-row .sd-col-6 .sd-col-xs-6 .sd-col-sm-6 .sd-col-md-6 .sd-col-lg-6 .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Units
:::

-   Defined unit: [[mol]{.std .std-ref}](#schema-units-mol){.reference
    .internal}
:::
:::
:::
:::
:::
:::

::: {#temperature .section}
[]{#schema-dimensions-temperature}

###### temperature[\#](#temperature "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .sd-col-6 .sd-col-xs-6 .sd-col-sm-6 .sd-col-md-6 .sd-col-lg-6 .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Dimensions
:::

K^1^
:::
:::
:::

::: {.sd-col .sd-d-flex-row .sd-col-6 .sd-col-xs-6 .sd-col-sm-6 .sd-col-md-6 .sd-col-lg-6 .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Units
:::

-   Defined unit: [[K]{.std .std-ref}](#schema-units-k){.reference
    .internal}

-   Defined unit: [[degC]{.std .std-ref}](#schema-units-degc){.reference
    .internal}
:::
:::
:::

::: {.sd-col .sd-d-flex-row .sd-col-12 .sd-col-xs-12 .sd-col-sm-12 .sd-col-md-12 .sd-col-lg-12 .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Schema
:::

::: {.highlight-xml .notranslate}
::: highlight
    <xs:simpleType name="Nml2Quantity_temperature">
      <xs:restriction base="xs:string">
        <xs:pattern value="-?([0-9]*(\.[0-9]+)?)([eE]-?[0-9]+)?[\s]*(degC)"/>
      </xs:restriction>
    </xs:simpleType>
:::
:::
:::
:::
:::
:::
:::
:::

::: {#time .section}
[]{#schema-dimensions-time}

###### time[\#](#time "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .sd-col-6 .sd-col-xs-6 .sd-col-sm-6 .sd-col-md-6 .sd-col-lg-6 .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Dimensions
:::

T^1^
:::
:::
:::

::: {.sd-col .sd-d-flex-row .sd-col-6 .sd-col-xs-6 .sd-col-sm-6 .sd-col-md-6 .sd-col-lg-6 .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Units
:::

-   Defined unit: [[hour]{.std .std-ref}](#schema-units-hour){.reference
    .internal}

-   Defined unit: [[min]{.std .std-ref}](#schema-units-min){.reference
    .internal}

-   Defined unit: [[ms]{.std .std-ref}](#id3){.reference .internal}

-   Defined unit: [[s]{.std .std-ref}](#id5){.reference .internal}
:::
:::
:::

::: {.sd-col .sd-d-flex-row .sd-col-12 .sd-col-xs-12 .sd-col-sm-12 .sd-col-md-12 .sd-col-lg-12 .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Schema
:::

::: {.highlight-xml .notranslate}
::: highlight
    <xs:simpleType name="Nml2Quantity_time">
      <xs:restriction base="xs:string">
        <xs:pattern value="-?([0-9]*(\.[0-9]+)?)([eE]-?[0-9]+)?[\s]*(s|ms)"/>
      </xs:restriction>
    </xs:simpleType>
:::
:::
:::
:::
:::
:::
:::
:::

::: {#voltage .section}
[]{#schema-dimensions-voltage}

###### voltage[\#](#voltage "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .sd-col-6 .sd-col-xs-6 .sd-col-sm-6 .sd-col-md-6 .sd-col-lg-6 .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Dimensions
:::

M^1^ L^2^ T^-3^ I^-1^
:::
:::
:::

::: {.sd-col .sd-d-flex-row .sd-col-6 .sd-col-xs-6 .sd-col-sm-6 .sd-col-md-6 .sd-col-lg-6 .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Units
:::

-   Defined unit: [[V]{.std .std-ref}](#schema-units-v){.reference
    .internal}

-   Defined unit: [[mV]{.std .std-ref}](#schema-units-mv){.reference
    .internal}
:::
:::
:::

::: {.sd-col .sd-d-flex-row .sd-col-12 .sd-col-xs-12 .sd-col-sm-12 .sd-col-md-12 .sd-col-lg-12 .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Schema
:::

::: {.highlight-xml .notranslate}
::: highlight
    <xs:simpleType name="Nml2Quantity_voltage">
      <xs:restriction base="xs:string">
        <xs:pattern value="-?([0-9]*(\.[0-9]+)?)([eE]-?[0-9]+)?[\s]*(V|mV)"/>
      </xs:restriction>
    </xs:simpleType>
:::
:::
:::
:::
:::
:::
:::
:::

::: {#volume .section}
[]{#schema-dimensions-volume}

###### volume[\#](#volume "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .sd-col-6 .sd-col-xs-6 .sd-col-sm-6 .sd-col-md-6 .sd-col-lg-6 .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Dimensions
:::

L^3^
:::
:::
:::

::: {.sd-col .sd-d-flex-row .sd-col-6 .sd-col-xs-6 .sd-col-sm-6 .sd-col-md-6 .sd-col-lg-6 .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Units
:::

-   Defined unit: [[cm3]{.std .std-ref}](#schema-units-cm3){.reference
    .internal}

-   Defined unit: [[litre]{.std
    .std-ref}](#schema-units-litre){.reference .internal}

-   Defined unit: [[m3]{.std .std-ref}](#schema-units-m3){.reference
    .internal}

-   Defined unit: [[um3]{.std .std-ref}](#schema-units-um3){.reference
    .internal}
:::
:::
:::
:::
:::
:::
:::

::: {#units .section}
##### Units[\#](#units "Link to this heading"){.headerlink}

::: {#a .section}
[]{#schema-units-a}

###### A[\#](#a "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[current]{.std
    .std-ref}](#schema-dimensions-current){.reference .internal}

-   Power of 10: 0
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 A = 1.00e+09 [[nA]{.std .std-ref}](#schema-units-na){.reference
    .internal}

-   1 A = 1.00e+12 [[pA]{.std .std-ref}](#schema-units-pa){.reference
    .internal}

-   1 A = 1.00e+06 [[uA]{.std .std-ref}](#schema-units-ua){.reference
    .internal}
:::
:::
:::
:::
:::
:::

::: {#a-per-m2 .section}
[]{#schema-units-a-per-m2}

###### A_per_m2[\#](#a-per-m2 "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[currentDensity]{.std
    .std-ref}](#schema-dimensions-currentdensity){.reference .internal}

-   Power of 10: 0
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 A_per_m2 = 0.1 [[mA_per_cm2]{.std
    .std-ref}](#schema-units-ma-per-cm2){.reference .internal}

-   1 A_per_m2 = 100 [[uA_per_cm2]{.std
    .std-ref}](#schema-units-ua-per-cm2){.reference .internal}
:::
:::
:::
:::
:::
:::

::: {#c .section}
[]{#schema-units-c}

###### C[\#](#c "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[charge]{.std
    .std-ref}](#schema-dimensions-charge){.reference .internal}

-   Power of 10: 0
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 C = 6.24e+18 [[e]{.std .std-ref}](#schema-units-e){.reference
    .internal}
:::
:::
:::
:::
:::
:::

::: {#c-per-mol .section}
[]{#schema-units-c-per-mol}

###### C_per_mol[\#](#c-per-mol "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[charge_per_mole]{.std
    .std-ref}](#schema-dimensions-charge-per-mole){.reference .internal}

-   Power of 10: 0
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 C_per_mol = 1e-06 [[nA_ms_per_amol]{.std
    .std-ref}](#schema-units-na-ms-per-amol){.reference .internal}

-   1 C_per_mol = 1.00e+06 [[pC_per_umol]{.std
    .std-ref}](#schema-units-pc-per-umol){.reference .internal}
:::
:::
:::
:::
:::
:::

::: {#f .section}
[]{#schema-units-f}

###### F[\#](#f "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[capacitance]{.std
    .std-ref}](#schema-dimensions-capacitance){.reference .internal}

-   Power of 10: 0
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 F = 1.00e+09 [[nF]{.std .std-ref}](#schema-units-nf){.reference
    .internal}

-   1 F = 1.00e+12 [[pF]{.std .std-ref}](#schema-units-pf){.reference
    .internal}

-   1 F = 1.00e+06 [[uF]{.std .std-ref}](#schema-units-uf){.reference
    .internal}
:::
:::
:::
:::
:::
:::

::: {#f-per-m2 .section}
[]{#schema-units-f-per-m2}

###### F_per_m2[\#](#f-per-m2 "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[specificCapacitance]{.std
    .std-ref}](#schema-dimensions-specificcapacitance){.reference
    .internal}

-   Power of 10: 0
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 F_per_m2 = 100 [[uF_per_cm2]{.std
    .std-ref}](#schema-units-uf-per-cm2){.reference .internal}
:::
:::
:::
:::
:::
:::

::: {#hz .section}
[]{#schema-units-hz}

###### Hz[\#](#hz "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[per_time]{.std
    .std-ref}](#schema-dimensions-per-time){.reference .internal}

-   Power of 10: 0
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 Hz = 3600 [[per_hour]{.std
    .std-ref}](#schema-units-per-hour){.reference .internal}

-   1 Hz = 60 [[per_min]{.std
    .std-ref}](#schema-units-per-min){.reference .internal}

-   1 Hz = 0.001 [[per_ms]{.std
    .std-ref}](#schema-units-per-ms){.reference .internal}

-   1 Hz = 1 [[per_s]{.std .std-ref}](#schema-units-per-s){.reference
    .internal}
:::
:::
:::
:::
:::
:::

::: {#j-per-k-per-mol .section}
[]{#schema-units-j-per-k-per-mol}

###### J_per_K_per_mol[\#](#j-per-k-per-mol "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[idealGasConstantDims]{.std
    .std-ref}](#schema-dimensions-idealgasconstantdims){.reference
    .internal}

-   Power of 10: 0
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 J_per_K_per_mol = 1.00e+09 [[fJ_per_K_per_umol]{.std
    .std-ref}](#schema-units-fj-per-k-per-umol){.reference .internal}
:::
:::
:::
:::
:::
:::

::: {#k .section}
[]{#schema-units-k}

###### K[\#](#k "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[temperature]{.std
    .std-ref}](#schema-dimensions-temperature){.reference .internal}

-   Power of 10: 0
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 K = -272.15 [[degC]{.std .std-ref}](#schema-units-degc){.reference
    .internal}
:::
:::
:::
:::
:::
:::

::: {#m .section}
[]{#schema-units-m}

###### M[\#](#m "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[concentration]{.std
    .std-ref}](#schema-dimensions-concentration){.reference .internal}

-   Power of 10: 3
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 M = 1000 [[mM]{.std .std-ref}](#schema-units-mm){.reference
    .internal}

-   1 M = 0.001 [[mol_per_cm3]{.std
    .std-ref}](#schema-units-mol-per-cm3){.reference .internal}

-   1 M = 1000 [[mol_per_m3]{.std
    .std-ref}](#schema-units-mol-per-m3){.reference .internal}
:::
:::
:::
:::
:::
:::

::: {#mohm .section}
[]{#schema-units-mohm}

###### Mohm[\#](#mohm "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[resistance]{.std
    .std-ref}](#schema-dimensions-resistance){.reference .internal}

-   Power of 10: 6
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 Mohm = 1000 [[kohm]{.std .std-ref}](#schema-units-kohm){.reference
    .internal}

-   1 Mohm = 1.00e+06 [[ohm]{.std
    .std-ref}](#schema-units-ohm){.reference .internal}
:::
:::
:::
:::
:::
:::

::: {#s .section}
[]{#schema-units-s}

###### S[\#](#s "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[conductance]{.std
    .std-ref}](#schema-dimensions-conductance){.reference .internal}

-   Power of 10: 0
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 S = 1000 [[mS]{.std .std-ref}](#schema-units-ms){.reference
    .internal}

-   1 S = 1.00e+09 [[nS]{.std .std-ref}](#schema-units-ns){.reference
    .internal}

-   1 S = 1.00e+12 [[pS]{.std .std-ref}](#schema-units-ps){.reference
    .internal}

-   1 S = 1.00e+06 [[uS]{.std .std-ref}](#schema-units-us){.reference
    .internal}
:::
:::
:::
:::
:::
:::

::: {#s-per-v .section}
[]{#schema-units-s-per-v}

###### S_per_V[\#](#s-per-v "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[conductance_per_voltage]{.std
    .std-ref}](#schema-dimensions-conductance-per-voltage){.reference
    .internal}

-   Power of 10: 0
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 S_per_V = 1.00e+06 [[nS_per_mV]{.std
    .std-ref}](#schema-units-ns-per-mv){.reference .internal}
:::
:::
:::
:::
:::
:::

::: {#s-per-cm2 .section}
[]{#schema-units-s-per-cm2}

###### S_per_cm2[\#](#s-per-cm2 "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[conductanceDensity]{.std
    .std-ref}](#schema-dimensions-conductancedensity){.reference
    .internal}

-   Power of 10: 4
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 S_per_cm2 = 10000 [[S_per_m2]{.std
    .std-ref}](#schema-units-s-per-m2){.reference .internal}

-   1 S_per_cm2 = 1000 [[mS_per_cm2]{.std
    .std-ref}](#schema-units-ms-per-cm2){.reference .internal}

-   1 S_per_cm2 = 1.00e+06 [[uS_per_cm2]{.std
    .std-ref}](#schema-units-us-per-cm2){.reference .internal}
:::
:::
:::
:::
:::
:::

::: {#s-per-m2 .section}
[]{#schema-units-s-per-m2}

###### S_per_m2[\#](#s-per-m2 "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[conductanceDensity]{.std
    .std-ref}](#schema-dimensions-conductancedensity){.reference
    .internal}

-   Power of 10: 0
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 S_per_m2 = 0.0001 [[S_per_cm2]{.std
    .std-ref}](#schema-units-s-per-cm2){.reference .internal}

-   1 S_per_m2 = 0.1 [[mS_per_cm2]{.std
    .std-ref}](#schema-units-ms-per-cm2){.reference .internal}

-   1 S_per_m2 = 100 [[uS_per_cm2]{.std
    .std-ref}](#schema-units-us-per-cm2){.reference .internal}
:::
:::
:::
:::
:::
:::

::: {#v .section}
[]{#schema-units-v}

###### V[\#](#v "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[voltage]{.std
    .std-ref}](#schema-dimensions-voltage){.reference .internal}

-   Power of 10: 0
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 V = 1000 [[mV]{.std .std-ref}](#schema-units-mv){.reference
    .internal}
:::
:::
:::
:::
:::
:::

::: {#cm .section}
[]{#schema-units-cm}

###### cm[\#](#cm "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[length]{.std
    .std-ref}](#schema-dimensions-length){.reference .internal}

-   Power of 10: -2
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 cm = 0.01 [[m]{.std .std-ref}](#id1){.reference .internal}

-   1 cm = 10000 [[um]{.std .std-ref}](#schema-units-um){.reference
    .internal}
:::
:::
:::
:::
:::
:::

::: {#cm2 .section}
[]{#schema-units-cm2}

###### cm2[\#](#cm2 "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[area]{.std
    .std-ref}](#schema-dimensions-area){.reference .internal}

-   Power of 10: -4
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 cm2 = 0.0001 [[m2]{.std .std-ref}](#schema-units-m2){.reference
    .internal}

-   1 cm2 = 1.00e+08 [[um2]{.std
    .std-ref}](#schema-units-um2){.reference .internal}
:::
:::
:::
:::
:::
:::

::: {#cm3 .section}
[]{#schema-units-cm3}

###### cm3[\#](#cm3 "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[volume]{.std
    .std-ref}](#schema-dimensions-volume){.reference .internal}

-   Power of 10: -6
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 cm3 = 0.001 [[litre]{.std
    .std-ref}](#schema-units-litre){.reference .internal}

-   1 cm3 = 1e-06 [[m3]{.std .std-ref}](#schema-units-m3){.reference
    .internal}

-   1 cm3 = 1.00e+12 [[um3]{.std
    .std-ref}](#schema-units-um3){.reference .internal}
:::
:::
:::
:::
:::
:::

::: {#cm-per-ms .section}
[]{#schema-units-cm-per-ms}

###### cm_per_ms[\#](#cm-per-ms "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[permeability]{.std
    .std-ref}](#schema-dimensions-permeability){.reference .internal}

-   Power of 10: 1
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 cm_per_ms = 1000 [[cm_per_s]{.std
    .std-ref}](#schema-units-cm-per-s){.reference .internal}

-   1 cm_per_ms = 10 [[m_per_s]{.std
    .std-ref}](#schema-units-m-per-s){.reference .internal}

-   1 cm_per_ms = 10000 [[um_per_ms]{.std
    .std-ref}](#schema-units-um-per-ms){.reference .internal}
:::
:::
:::
:::
:::
:::

::: {#cm-per-s .section}
[]{#schema-units-cm-per-s}

###### cm_per_s[\#](#cm-per-s "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[permeability]{.std
    .std-ref}](#schema-dimensions-permeability){.reference .internal}

-   Power of 10: -2
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 cm_per_s = 0.001 [[cm_per_ms]{.std
    .std-ref}](#schema-units-cm-per-ms){.reference .internal}

-   1 cm_per_s = 0.01 [[m_per_s]{.std
    .std-ref}](#schema-units-m-per-s){.reference .internal}

-   1 cm_per_s = 10 [[um_per_ms]{.std
    .std-ref}](#schema-units-um-per-ms){.reference .internal}
:::
:::
:::
:::
:::
:::

::: {#degc .section}
[]{#schema-units-degc}

###### degC[\#](#degc "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[temperature]{.std
    .std-ref}](#schema-dimensions-temperature){.reference .internal}

-   Power of 10: 0

-   Offset: 273.15
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 degC = 274.15 [[K]{.std .std-ref}](#schema-units-k){.reference
    .internal}
:::
:::
:::
:::
:::
:::

::: {#e .section}
[]{#schema-units-e}

###### e[\#](#e "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[charge]{.std
    .std-ref}](#schema-dimensions-charge){.reference .internal}

-   Power of 10: 0

-   Scale: 1.602176634e-19
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 e = 1.6022e-19 [[C]{.std .std-ref}](#schema-units-c){.reference
    .internal}
:::
:::
:::
:::
:::
:::

::: {#fj-per-k-per-umol .section}
[]{#schema-units-fj-per-k-per-umol}

###### fJ_per_K_per_umol[\#](#fj-per-k-per-umol "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[idealGasConstantDims]{.std
    .std-ref}](#schema-dimensions-idealgasconstantdims){.reference
    .internal}

-   Power of 10: -9
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 fJ_per_K_per_umol = 1e-09 [[J_per_K_per_mol]{.std
    .std-ref}](#schema-units-j-per-k-per-mol){.reference .internal}
:::
:::
:::
:::
:::
:::

::: {#hour .section}
[]{#schema-units-hour}

###### hour[\#](#hour "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[time]{.std
    .std-ref}](#schema-dimensions-time){.reference .internal}

-   Power of 10: 0

-   Scale: 3600.0
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 hour = 60 [[min]{.std .std-ref}](#schema-units-min){.reference
    .internal}

-   1 hour = 3.60e+06 [[ms]{.std .std-ref}](#id3){.reference .internal}

-   1 hour = 3600 [[s]{.std .std-ref}](#id5){.reference .internal}
:::
:::
:::
:::
:::
:::

::: {#kohm .section}
[]{#schema-units-kohm}

###### kohm[\#](#kohm "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[resistance]{.std
    .std-ref}](#schema-dimensions-resistance){.reference .internal}

-   Power of 10: 3
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 kohm = 0.001 [[Mohm]{.std
    .std-ref}](#schema-units-mohm){.reference .internal}

-   1 kohm = 1000 [[ohm]{.std .std-ref}](#schema-units-ohm){.reference
    .internal}
:::
:::
:::
:::
:::
:::

::: {#kohm-cm .section}
[]{#schema-units-kohm-cm}

###### kohm_cm[\#](#kohm-cm "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[resistivity]{.std
    .std-ref}](#schema-dimensions-resistivity){.reference .internal}

-   Power of 10: 1
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 kohm_cm = 1000 [[ohm_cm]{.std
    .std-ref}](#schema-units-ohm-cm){.reference .internal}

-   1 kohm_cm = 10 [[ohm_m]{.std
    .std-ref}](#schema-units-ohm-m){.reference .internal}
:::
:::
:::
:::
:::
:::

::: {#litre .section}
[]{#schema-units-litre}

###### litre[\#](#litre "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[volume]{.std
    .std-ref}](#schema-dimensions-volume){.reference .internal}

-   Power of 10: -3
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 litre = 1000 [[cm3]{.std .std-ref}](#schema-units-cm3){.reference
    .internal}

-   1 litre = 0.001 [[m3]{.std .std-ref}](#schema-units-m3){.reference
    .internal}

-   1 litre = 1.00e+15 [[um3]{.std
    .std-ref}](#schema-units-um3){.reference .internal}
:::
:::
:::
:::
:::
:::

::: {#id1 .section}
[]{#id2}

###### m[\#](#id1 "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[length]{.std
    .std-ref}](#schema-dimensions-length){.reference .internal}

-   Power of 10: 0
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 m = 100 [[cm]{.std .std-ref}](#schema-units-cm){.reference
    .internal}

-   1 m = 1.00e+06 [[um]{.std .std-ref}](#schema-units-um){.reference
    .internal}
:::
:::
:::
:::
:::
:::

::: {#m2 .section}
[]{#schema-units-m2}

###### m2[\#](#m2 "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[area]{.std
    .std-ref}](#schema-dimensions-area){.reference .internal}

-   Power of 10: 0
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 m2 = 10000 [[cm2]{.std .std-ref}](#schema-units-cm2){.reference
    .internal}

-   1 m2 = 1.00e+12 [[um2]{.std .std-ref}](#schema-units-um2){.reference
    .internal}
:::
:::
:::
:::
:::
:::

::: {#m3 .section}
[]{#schema-units-m3}

###### m3[\#](#m3 "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[volume]{.std
    .std-ref}](#schema-dimensions-volume){.reference .internal}

-   Power of 10: 0
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 m3 = 1.00e+06 [[cm3]{.std .std-ref}](#schema-units-cm3){.reference
    .internal}

-   1 m3 = 1000 [[litre]{.std .std-ref}](#schema-units-litre){.reference
    .internal}

-   1 m3 = 1.00e+18 [[um3]{.std .std-ref}](#schema-units-um3){.reference
    .internal}
:::
:::
:::
:::
:::
:::

::: {#ma-per-cm2 .section}
[]{#schema-units-ma-per-cm2}

###### mA_per_cm2[\#](#ma-per-cm2 "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[currentDensity]{.std
    .std-ref}](#schema-dimensions-currentdensity){.reference .internal}

-   Power of 10: 1
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 mA_per_cm2 = 10 [[A_per_m2]{.std
    .std-ref}](#schema-units-a-per-m2){.reference .internal}

-   1 mA_per_cm2 = 1000 [[uA_per_cm2]{.std
    .std-ref}](#schema-units-ua-per-cm2){.reference .internal}
:::
:::
:::
:::
:::
:::

::: {#mm .section}
[]{#schema-units-mm}

###### mM[\#](#mm "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[concentration]{.std
    .std-ref}](#schema-dimensions-concentration){.reference .internal}

-   Power of 10: 0
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 mM = 0.001 [[M]{.std .std-ref}](#schema-units-m){.reference
    .internal}

-   1 mM = 1e-06 [[mol_per_cm3]{.std
    .std-ref}](#schema-units-mol-per-cm3){.reference .internal}

-   1 mM = 1 [[mol_per_m3]{.std
    .std-ref}](#schema-units-mol-per-m3){.reference .internal}
:::
:::
:::
:::
:::
:::

::: {#ms .section}
[]{#schema-units-ms}

###### mS[\#](#ms "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[conductance]{.std
    .std-ref}](#schema-dimensions-conductance){.reference .internal}

-   Power of 10: -3
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 mS = 0.001 [[S]{.std .std-ref}](#schema-units-s){.reference
    .internal}

-   1 mS = 1.00e+06 [[nS]{.std .std-ref}](#schema-units-ns){.reference
    .internal}

-   1 mS = 1.00e+09 [[pS]{.std .std-ref}](#schema-units-ps){.reference
    .internal}

-   1 mS = 1000 [[uS]{.std .std-ref}](#schema-units-us){.reference
    .internal}
:::
:::
:::
:::
:::
:::

::: {#ms-per-cm2 .section}
[]{#schema-units-ms-per-cm2}

###### mS_per_cm2[\#](#ms-per-cm2 "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[conductanceDensity]{.std
    .std-ref}](#schema-dimensions-conductancedensity){.reference
    .internal}

-   Power of 10: 1
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 mS_per_cm2 = 0.001 [[S_per_cm2]{.std
    .std-ref}](#schema-units-s-per-cm2){.reference .internal}

-   1 mS_per_cm2 = 10 [[S_per_m2]{.std
    .std-ref}](#schema-units-s-per-m2){.reference .internal}

-   1 mS_per_cm2 = 1000 [[uS_per_cm2]{.std
    .std-ref}](#schema-units-us-per-cm2){.reference .internal}
:::
:::
:::
:::
:::
:::

::: {#mv .section}
[]{#schema-units-mv}

###### mV[\#](#mv "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[voltage]{.std
    .std-ref}](#schema-dimensions-voltage){.reference .internal}

-   Power of 10: -3
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 mV = 0.001 [[V]{.std .std-ref}](#schema-units-v){.reference
    .internal}
:::
:::
:::
:::
:::
:::

::: {#m-per-s .section}
[]{#schema-units-m-per-s}

###### m_per_s[\#](#m-per-s "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[permeability]{.std
    .std-ref}](#schema-dimensions-permeability){.reference .internal}

-   Power of 10: 0
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 m_per_s = 0.1 [[cm_per_ms]{.std
    .std-ref}](#schema-units-cm-per-ms){.reference .internal}

-   1 m_per_s = 100 [[cm_per_s]{.std
    .std-ref}](#schema-units-cm-per-s){.reference .internal}

-   1 m_per_s = 1000 [[um_per_ms]{.std
    .std-ref}](#schema-units-um-per-ms){.reference .internal}
:::
:::
:::
:::
:::
:::

::: {#min .section}
[]{#schema-units-min}

###### min[\#](#min "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[time]{.std
    .std-ref}](#schema-dimensions-time){.reference .internal}

-   Power of 10: 0

-   Scale: 60.0
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 min = 0.016667 [[hour]{.std
    .std-ref}](#schema-units-hour){.reference .internal}

-   1 min = 6.00e+04 [[ms]{.std .std-ref}](#id3){.reference .internal}

-   1 min = 60 [[s]{.std .std-ref}](#id5){.reference .internal}
:::
:::
:::
:::
:::
:::

::: {#mol .section}
[]{#schema-units-mol}

###### mol[\#](#mol "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[substance]{.std
    .std-ref}](#schema-dimensions-substance){.reference .internal}

-   Power of 10: 0
:::
:::
:::
:::
:::
:::

::: {#mol-per-cm3 .section}
[]{#schema-units-mol-per-cm3}

###### mol_per_cm3[\#](#mol-per-cm3 "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[concentration]{.std
    .std-ref}](#schema-dimensions-concentration){.reference .internal}

-   Power of 10: 6
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 mol_per_cm3 = 1000 [[M]{.std
    .std-ref}](#schema-units-m){.reference .internal}

-   1 mol_per_cm3 = 1.00e+06 [[mM]{.std
    .std-ref}](#schema-units-mm){.reference .internal}

-   1 mol_per_cm3 = 1.00e+06 [[mol_per_m3]{.std
    .std-ref}](#schema-units-mol-per-m3){.reference .internal}
:::
:::
:::
:::
:::
:::

::: {#mol-per-cm-per-ua-per-ms .section}
[]{#schema-units-mol-per-cm-per-ua-per-ms}

###### mol_per_cm_per_uA_per_ms[\#](#mol-per-cm-per-ua-per-ms "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[rho_factor]{.std
    .std-ref}](#schema-dimensions-rho-factor){.reference .internal}

-   Power of 10: 11
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 mol_per_cm_per_uA_per_ms = 1.00e+11 [[mol_per_m_per_A_per_s]{.std
    .std-ref}](#schema-units-mol-per-m-per-a-per-s){.reference
    .internal}

-   1 mol_per_cm_per_uA_per_ms = 1000 [[umol_per_cm_per_nA_per_ms]{.std
    .std-ref}](#schema-units-umol-per-cm-per-na-per-ms){.reference
    .internal}
:::
:::
:::
:::
:::
:::

::: {#mol-per-m3 .section}
[]{#schema-units-mol-per-m3}

###### mol_per_m3[\#](#mol-per-m3 "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[concentration]{.std
    .std-ref}](#schema-dimensions-concentration){.reference .internal}

-   Power of 10: 0
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 mol_per_m3 = 0.001 [[M]{.std
    .std-ref}](#schema-units-m){.reference .internal}

-   1 mol_per_m3 = 1 [[mM]{.std .std-ref}](#schema-units-mm){.reference
    .internal}

-   1 mol_per_m3 = 1e-06 [[mol_per_cm3]{.std
    .std-ref}](#schema-units-mol-per-cm3){.reference .internal}
:::
:::
:::
:::
:::
:::

::: {#mol-per-m-per-a-per-s .section}
[]{#schema-units-mol-per-m-per-a-per-s}

###### mol_per_m_per_A_per_s[\#](#mol-per-m-per-a-per-s "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[rho_factor]{.std
    .std-ref}](#schema-dimensions-rho-factor){.reference .internal}

-   Power of 10: 0
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 mol_per_m_per_A_per_s = 1e-11 [[mol_per_cm_per_uA_per_ms]{.std
    .std-ref}](#schema-units-mol-per-cm-per-ua-per-ms){.reference
    .internal}

-   1 mol_per_m_per_A_per_s = 1e-08 [[umol_per_cm_per_nA_per_ms]{.std
    .std-ref}](#schema-units-umol-per-cm-per-na-per-ms){.reference
    .internal}
:::
:::
:::
:::
:::
:::

::: {#id3 .section}
[]{#id4}

###### ms[\#](#id3 "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[time]{.std
    .std-ref}](#schema-dimensions-time){.reference .internal}

-   Power of 10: -3
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 ms = 2.7778e-07 [[hour]{.std
    .std-ref}](#schema-units-hour){.reference .internal}

-   1 ms = 1.6667e-05 [[min]{.std
    .std-ref}](#schema-units-min){.reference .internal}

-   1 ms = 0.001 [[s]{.std .std-ref}](#id5){.reference .internal}
:::
:::
:::
:::
:::
:::

::: {#na .section}
[]{#schema-units-na}

###### nA[\#](#na "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[current]{.std
    .std-ref}](#schema-dimensions-current){.reference .internal}

-   Power of 10: -9
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 nA = 1e-09 [[A]{.std .std-ref}](#schema-units-a){.reference
    .internal}

-   1 nA = 1000 [[pA]{.std .std-ref}](#schema-units-pa){.reference
    .internal}

-   1 nA = 0.001 [[uA]{.std .std-ref}](#schema-units-ua){.reference
    .internal}
:::
:::
:::
:::
:::
:::

::: {#na-ms-per-amol .section}
[]{#schema-units-na-ms-per-amol}

###### nA_ms_per_amol[\#](#na-ms-per-amol "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[charge_per_mole]{.std
    .std-ref}](#schema-dimensions-charge-per-mole){.reference .internal}

-   Power of 10: 6
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 nA_ms_per_amol = 1.00e+06 [[C_per_mol]{.std
    .std-ref}](#schema-units-c-per-mol){.reference .internal}

-   1 nA_ms_per_amol = 1.00e+12 [[pC_per_umol]{.std
    .std-ref}](#schema-units-pc-per-umol){.reference .internal}
:::
:::
:::
:::
:::
:::

::: {#nf .section}
[]{#schema-units-nf}

###### nF[\#](#nf "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[capacitance]{.std
    .std-ref}](#schema-dimensions-capacitance){.reference .internal}

-   Power of 10: -9
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 nF = 1e-09 [[F]{.std .std-ref}](#schema-units-f){.reference
    .internal}

-   1 nF = 1000 [[pF]{.std .std-ref}](#schema-units-pf){.reference
    .internal}

-   1 nF = 0.001 [[uF]{.std .std-ref}](#schema-units-uf){.reference
    .internal}
:::
:::
:::
:::
:::
:::

::: {#ns .section}
[]{#schema-units-ns}

###### nS[\#](#ns "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[conductance]{.std
    .std-ref}](#schema-dimensions-conductance){.reference .internal}

-   Power of 10: -9
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 nS = 1e-09 [[S]{.std .std-ref}](#schema-units-s){.reference
    .internal}

-   1 nS = 1e-06 [[mS]{.std .std-ref}](#schema-units-ms){.reference
    .internal}

-   1 nS = 1000 [[pS]{.std .std-ref}](#schema-units-ps){.reference
    .internal}

-   1 nS = 0.001 [[uS]{.std .std-ref}](#schema-units-us){.reference
    .internal}
:::
:::
:::
:::
:::
:::

::: {#ns-per-mv .section}
[]{#schema-units-ns-per-mv}

###### nS_per_mV[\#](#ns-per-mv "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[conductance_per_voltage]{.std
    .std-ref}](#schema-dimensions-conductance-per-voltage){.reference
    .internal}

-   Power of 10: -6
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 nS_per_mV = 1e-06 [[S_per_V]{.std
    .std-ref}](#schema-units-s-per-v){.reference .internal}
:::
:::
:::
:::
:::
:::

::: {#ohm .section}
[]{#schema-units-ohm}

###### ohm[\#](#ohm "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[resistance]{.std
    .std-ref}](#schema-dimensions-resistance){.reference .internal}

-   Power of 10: 0
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 ohm = 1e-06 [[Mohm]{.std .std-ref}](#schema-units-mohm){.reference
    .internal}

-   1 ohm = 0.001 [[kohm]{.std .std-ref}](#schema-units-kohm){.reference
    .internal}
:::
:::
:::
:::
:::
:::

::: {#ohm-cm .section}
[]{#schema-units-ohm-cm}

###### ohm_cm[\#](#ohm-cm "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[resistivity]{.std
    .std-ref}](#schema-dimensions-resistivity){.reference .internal}

-   Power of 10: -2
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 ohm_cm = 0.001 [[kohm_cm]{.std
    .std-ref}](#schema-units-kohm-cm){.reference .internal}

-   1 ohm_cm = 0.01 [[ohm_m]{.std
    .std-ref}](#schema-units-ohm-m){.reference .internal}
:::
:::
:::
:::
:::
:::

::: {#ohm-m .section}
[]{#schema-units-ohm-m}

###### ohm_m[\#](#ohm-m "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[resistivity]{.std
    .std-ref}](#schema-dimensions-resistivity){.reference .internal}

-   Power of 10: 0
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 ohm_m = 0.1 [[kohm_cm]{.std
    .std-ref}](#schema-units-kohm-cm){.reference .internal}

-   1 ohm_m = 100 [[ohm_cm]{.std
    .std-ref}](#schema-units-ohm-cm){.reference .internal}
:::
:::
:::
:::
:::
:::

::: {#pa .section}
[]{#schema-units-pa}

###### pA[\#](#pa "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[current]{.std
    .std-ref}](#schema-dimensions-current){.reference .internal}

-   Power of 10: -12
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 pA = 1e-12 [[A]{.std .std-ref}](#schema-units-a){.reference
    .internal}

-   1 pA = 0.001 [[nA]{.std .std-ref}](#schema-units-na){.reference
    .internal}

-   1 pA = 1e-06 [[uA]{.std .std-ref}](#schema-units-ua){.reference
    .internal}
:::
:::
:::
:::
:::
:::

::: {#pc-per-umol .section}
[]{#schema-units-pc-per-umol}

###### pC_per_umol[\#](#pc-per-umol "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[charge_per_mole]{.std
    .std-ref}](#schema-dimensions-charge-per-mole){.reference .internal}

-   Power of 10: -6
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 pC_per_umol = 1e-06 [[C_per_mol]{.std
    .std-ref}](#schema-units-c-per-mol){.reference .internal}

-   1 pC_per_umol = 1e-12 [[nA_ms_per_amol]{.std
    .std-ref}](#schema-units-na-ms-per-amol){.reference .internal}
:::
:::
:::
:::
:::
:::

::: {#pf .section}
[]{#schema-units-pf}

###### pF[\#](#pf "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[capacitance]{.std
    .std-ref}](#schema-dimensions-capacitance){.reference .internal}

-   Power of 10: -12
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 pF = 1e-12 [[F]{.std .std-ref}](#schema-units-f){.reference
    .internal}

-   1 pF = 0.001 [[nF]{.std .std-ref}](#schema-units-nf){.reference
    .internal}

-   1 pF = 1e-06 [[uF]{.std .std-ref}](#schema-units-uf){.reference
    .internal}
:::
:::
:::
:::
:::
:::

::: {#ps .section}
[]{#schema-units-ps}

###### pS[\#](#ps "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[conductance]{.std
    .std-ref}](#schema-dimensions-conductance){.reference .internal}

-   Power of 10: -12
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 pS = 1e-12 [[S]{.std .std-ref}](#schema-units-s){.reference
    .internal}

-   1 pS = 1e-09 [[mS]{.std .std-ref}](#schema-units-ms){.reference
    .internal}

-   1 pS = 0.001 [[nS]{.std .std-ref}](#schema-units-ns){.reference
    .internal}

-   1 pS = 1e-06 [[uS]{.std .std-ref}](#schema-units-us){.reference
    .internal}
:::
:::
:::
:::
:::
:::

::: {#per-v .section}
[]{#schema-units-per-v}

###### per_V[\#](#per-v "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[per_voltage]{.std
    .std-ref}](#schema-dimensions-per-voltage){.reference .internal}

-   Power of 10: 0
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 per_V = 0.001 [[per_mV]{.std
    .std-ref}](#schema-units-per-mv){.reference .internal}
:::
:::
:::
:::
:::
:::

::: {#per-hour .section}
[]{#schema-units-per-hour}

###### per_hour[\#](#per-hour "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[per_time]{.std
    .std-ref}](#schema-dimensions-per-time){.reference .internal}

-   Power of 10: 0

-   Scale: 0.00027777777778
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 per_hour = 0.00027778 [[Hz]{.std
    .std-ref}](#schema-units-hz){.reference .internal}

-   1 per_hour = 0.016667 [[per_min]{.std
    .std-ref}](#schema-units-per-min){.reference .internal}

-   1 per_hour = 2.7778e-07 [[per_ms]{.std
    .std-ref}](#schema-units-per-ms){.reference .internal}

-   1 per_hour = 0.00027778 [[per_s]{.std
    .std-ref}](#schema-units-per-s){.reference .internal}
:::
:::
:::
:::
:::
:::

::: {#per-mv .section}
[]{#schema-units-per-mv}

###### per_mV[\#](#per-mv "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[per_voltage]{.std
    .std-ref}](#schema-dimensions-per-voltage){.reference .internal}

-   Power of 10: 3
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 per_mV = 1000 [[per_V]{.std
    .std-ref}](#schema-units-per-v){.reference .internal}
:::
:::
:::
:::
:::
:::

::: {#per-min .section}
[]{#schema-units-per-min}

###### per_min[\#](#per-min "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[per_time]{.std
    .std-ref}](#schema-dimensions-per-time){.reference .internal}

-   Power of 10: 0

-   Scale: 0.01666666667
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 per_min = 0.016667 [[Hz]{.std
    .std-ref}](#schema-units-hz){.reference .internal}

-   1 per_min = 60 [[per_hour]{.std
    .std-ref}](#schema-units-per-hour){.reference .internal}

-   1 per_min = 1.6667e-05 [[per_ms]{.std
    .std-ref}](#schema-units-per-ms){.reference .internal}

-   1 per_min = 0.016667 [[per_s]{.std
    .std-ref}](#schema-units-per-s){.reference .internal}
:::
:::
:::
:::
:::
:::

::: {#per-ms .section}
[]{#schema-units-per-ms}

###### per_ms[\#](#per-ms "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[per_time]{.std
    .std-ref}](#schema-dimensions-per-time){.reference .internal}

-   Power of 10: 3
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 per_ms = 1000 [[Hz]{.std .std-ref}](#schema-units-hz){.reference
    .internal}

-   1 per_ms = 3.60e+06 [[per_hour]{.std
    .std-ref}](#schema-units-per-hour){.reference .internal}

-   1 per_ms = 6.00e+04 [[per_min]{.std
    .std-ref}](#schema-units-per-min){.reference .internal}

-   1 per_ms = 1000 [[per_s]{.std
    .std-ref}](#schema-units-per-s){.reference .internal}
:::
:::
:::
:::
:::
:::

::: {#per-s .section}
[]{#schema-units-per-s}

###### per_s[\#](#per-s "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[per_time]{.std
    .std-ref}](#schema-dimensions-per-time){.reference .internal}

-   Power of 10: 0
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 per_s = 1 [[Hz]{.std .std-ref}](#schema-units-hz){.reference
    .internal}

-   1 per_s = 3600 [[per_hour]{.std
    .std-ref}](#schema-units-per-hour){.reference .internal}

-   1 per_s = 60 [[per_min]{.std
    .std-ref}](#schema-units-per-min){.reference .internal}

-   1 per_s = 0.001 [[per_ms]{.std
    .std-ref}](#schema-units-per-ms){.reference .internal}
:::
:::
:::
:::
:::
:::

::: {#id5 .section}
[]{#id6}

###### s[\#](#id5 "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[time]{.std
    .std-ref}](#schema-dimensions-time){.reference .internal}

-   Power of 10: 0
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 s = 0.00027778 [[hour]{.std
    .std-ref}](#schema-units-hour){.reference .internal}

-   1 s = 0.016667 [[min]{.std .std-ref}](#schema-units-min){.reference
    .internal}

-   1 s = 1000 [[ms]{.std .std-ref}](#id3){.reference .internal}
:::
:::
:::
:::
:::
:::

::: {#ua .section}
[]{#schema-units-ua}

###### uA[\#](#ua "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[current]{.std
    .std-ref}](#schema-dimensions-current){.reference .internal}

-   Power of 10: -6
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 uA = 1e-06 [[A]{.std .std-ref}](#schema-units-a){.reference
    .internal}

-   1 uA = 1000 [[nA]{.std .std-ref}](#schema-units-na){.reference
    .internal}

-   1 uA = 1.00e+06 [[pA]{.std .std-ref}](#schema-units-pa){.reference
    .internal}
:::
:::
:::
:::
:::
:::

::: {#ua-per-cm2 .section}
[]{#schema-units-ua-per-cm2}

###### uA_per_cm2[\#](#ua-per-cm2 "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[currentDensity]{.std
    .std-ref}](#schema-dimensions-currentdensity){.reference .internal}

-   Power of 10: -2
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 uA_per_cm2 = 0.01 [[A_per_m2]{.std
    .std-ref}](#schema-units-a-per-m2){.reference .internal}

-   1 uA_per_cm2 = 0.001 [[mA_per_cm2]{.std
    .std-ref}](#schema-units-ma-per-cm2){.reference .internal}
:::
:::
:::
:::
:::
:::

::: {#uf .section}
[]{#schema-units-uf}

###### uF[\#](#uf "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[capacitance]{.std
    .std-ref}](#schema-dimensions-capacitance){.reference .internal}

-   Power of 10: -6
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 uF = 1e-06 [[F]{.std .std-ref}](#schema-units-f){.reference
    .internal}

-   1 uF = 1000 [[nF]{.std .std-ref}](#schema-units-nf){.reference
    .internal}

-   1 uF = 1.00e+06 [[pF]{.std .std-ref}](#schema-units-pf){.reference
    .internal}
:::
:::
:::
:::
:::
:::

::: {#uf-per-cm2 .section}
[]{#schema-units-uf-per-cm2}

###### uF_per_cm2[\#](#uf-per-cm2 "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[specificCapacitance]{.std
    .std-ref}](#schema-dimensions-specificcapacitance){.reference
    .internal}

-   Power of 10: -2
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 uF_per_cm2 = 0.01 [[F_per_m2]{.std
    .std-ref}](#schema-units-f-per-m2){.reference .internal}
:::
:::
:::
:::
:::
:::

::: {#us .section}
[]{#schema-units-us}

###### uS[\#](#us "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[conductance]{.std
    .std-ref}](#schema-dimensions-conductance){.reference .internal}

-   Power of 10: -6
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 uS = 1e-06 [[S]{.std .std-ref}](#schema-units-s){.reference
    .internal}

-   1 uS = 0.001 [[mS]{.std .std-ref}](#schema-units-ms){.reference
    .internal}

-   1 uS = 1000 [[nS]{.std .std-ref}](#schema-units-ns){.reference
    .internal}

-   1 uS = 1.00e+06 [[pS]{.std .std-ref}](#schema-units-ps){.reference
    .internal}
:::
:::
:::
:::
:::
:::

::: {#us-per-cm2 .section}
[]{#schema-units-us-per-cm2}

###### uS_per_cm2[\#](#us-per-cm2 "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[conductanceDensity]{.std
    .std-ref}](#schema-dimensions-conductancedensity){.reference
    .internal}

-   Power of 10: -2
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 uS_per_cm2 = 1e-06 [[S_per_cm2]{.std
    .std-ref}](#schema-units-s-per-cm2){.reference .internal}

-   1 uS_per_cm2 = 0.01 [[S_per_m2]{.std
    .std-ref}](#schema-units-s-per-m2){.reference .internal}

-   1 uS_per_cm2 = 0.001 [[mS_per_cm2]{.std
    .std-ref}](#schema-units-ms-per-cm2){.reference .internal}
:::
:::
:::
:::
:::
:::

::: {#um .section}
[]{#schema-units-um}

###### um[\#](#um "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[length]{.std
    .std-ref}](#schema-dimensions-length){.reference .internal}

-   Power of 10: -6
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 um = 0.0001 [[cm]{.std .std-ref}](#schema-units-cm){.reference
    .internal}

-   1 um = 1e-06 [[m]{.std .std-ref}](#id1){.reference .internal}
:::
:::
:::
:::
:::
:::

::: {#um2 .section}
[]{#schema-units-um2}

###### um2[\#](#um2 "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[area]{.std
    .std-ref}](#schema-dimensions-area){.reference .internal}

-   Power of 10: -12
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 um2 = 1e-08 [[cm2]{.std .std-ref}](#schema-units-cm2){.reference
    .internal}

-   1 um2 = 1e-12 [[m2]{.std .std-ref}](#schema-units-m2){.reference
    .internal}
:::
:::
:::
:::
:::
:::

::: {#um3 .section}
[]{#schema-units-um3}

###### um3[\#](#um3 "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[volume]{.std
    .std-ref}](#schema-dimensions-volume){.reference .internal}

-   Power of 10: -18
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 um3 = 1e-12 [[cm3]{.std .std-ref}](#schema-units-cm3){.reference
    .internal}

-   1 um3 = 1e-15 [[litre]{.std
    .std-ref}](#schema-units-litre){.reference .internal}

-   1 um3 = 1e-18 [[m3]{.std .std-ref}](#schema-units-m3){.reference
    .internal}
:::
:::
:::
:::
:::
:::

::: {#um-per-ms .section}
[]{#schema-units-um-per-ms}

###### um_per_ms[\#](#um-per-ms "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[permeability]{.std
    .std-ref}](#schema-dimensions-permeability){.reference .internal}

-   Power of 10: -3
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 um_per_ms = 0.0001 [[cm_per_ms]{.std
    .std-ref}](#schema-units-cm-per-ms){.reference .internal}

-   1 um_per_ms = 0.1 [[cm_per_s]{.std
    .std-ref}](#schema-units-cm-per-s){.reference .internal}

-   1 um_per_ms = 0.001 [[m_per_s]{.std
    .std-ref}](#schema-units-m-per-s){.reference .internal}
:::
:::
:::
:::
:::
:::

::: {#umol-per-cm-per-na-per-ms .section}
[]{#schema-units-umol-per-cm-per-na-per-ms}

###### umol_per_cm_per_nA_per_ms[\#](#umol-per-cm-per-na-per-ms "Link to this heading"){.headerlink}

::: {.sd-container-fluid .sd-sphinx-override .sd-mb-4 .docutils}
::: {.sd-row .sd-g-2 .sd-g-xs-2 .sd-g-sm-2 .sd-g-md-2 .sd-g-lg-2 .docutils}
::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Summary
:::

-   Dimension: [[rho_factor]{.std
    .std-ref}](#schema-dimensions-rho-factor){.reference .internal}

-   Power of 10: 8
:::
:::
:::

::: {.sd-col .sd-d-flex-row .docutils}
::: {.sd-card .sd-sphinx-override .sd-w-100 .sd-shadow-sm .docutils}
::: {.sd-card-body .docutils}
::: {.sd-card-title .sd-font-weight-bold .docutils}
Conversions
:::

-   1 umol_per_cm_per_nA_per_ms = 0.001 [[mol_per_cm_per_uA_per_ms]{.std
    .std-ref}](#schema-units-mol-per-cm-per-ua-per-ms){.reference
    .internal}

-   1 umol_per_cm_per_nA_per_ms = 1.00e+08 [[mol_per_m_per_A_per_s]{.std
    .std-ref}](#schema-units-mol-per-m-per-a-per-s){.reference
    .internal}
:::
:::
:::
:::
:::
:::
:::
:::

[]{#document-Userdocs/Schemas/NeuroMLCoreCompTypes}

::: {#neuromlcorecomptypes .section .tex2jax_ignore .mathjax_ignore}
[]{#schema-neuromlcorecomptypes}

#### NeuroMLCoreCompTypes[\#](#neuromlcorecomptypes "Link to this heading"){.headerlink}

Original ComponentType definitions:
[NeuroMLCoreCompTypes.xml](https://github.com/NeuroML/NeuroML2/blob/master/NeuroML2CoreTypes//NeuroMLCoreCompTypes.xml){.reference
.external}. Schema against which NeuroML based on these should be valid:
[NeuroML_v2.3.xsd](https://github.com/NeuroML/NeuroML2/tree/master/Schemas/NeuroML2/NeuroML_v2.3.xsd){.reference
.external}. Generated on 14/08/24 from
[this](https://github.com/NeuroML/NeuroML2/commit/352244cff605cb1ba24fa7c11757dc818fe90fd2){.reference
.external} commit. Please file any issues or questions at the [issue
tracker here](https://github.com/NeuroML/NeuroML2/issues){.reference
.external}.

------------------------------------------------------------------------

::: {#notes .section}
[]{#schema-notes}

##### notes[\#](#notes "Link to this heading"){.headerlink}

*Human readable notes/description for a Component.*

::: {.sd-tab-set .docutils}
Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:simpleType name="Notes">
      <xs:annotation>
        <xs:documentation>Textual human readable notes related to the element in question. It's useful to put these into
             the NeuroML files instead of XML comments, as the notes can be extracted and repeated in the files to which the NeuroML is mapped.
                </xs:documentation>
      </xs:annotation>
      <xs:restriction base="xs:string"/>
    </xs:simpleType>
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <notes>A Simple Spiking cell for testing purposes</notes>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <notes>Multicompartmental cell</notes>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <notes>Leak conductance</notes>
:::
:::
:::
:::
:::

::: {#annotation .section}
[]{#schema-annotation}

##### annotation[\#](#annotation "Link to this heading"){.headerlink}

*A structured annotation containing metadata, specifically RDF or
[[property]{.std .std-ref}](#schema-property){.reference .internal}
elements.*

::: {.sd-tab-set .docutils}
Child list

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **r  |                                                 | [[rd        |
| df:R |                                                 | f_RDF]{.std |
| DF** |                                                 | .std-       |
|      |                                                 | ref}](#sche |
|      |                                                 | ma-rdf-rdf) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Children list

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **pr |                                                 | [[pro       |
| oper |                                                 | perty]{.std |
| ty** |                                                 | .std-r      |
|      |                                                 | ef}](#schem |
|      |                                                 | a-property) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="Annotation">
      <xs:complexContent>
        <xs:extension base="BaseWithoutId">
          <xs:sequence>
            <xs:any processContents="skip" minOccurs="0" maxOccurs="unbounded"/>
          </xs:sequence>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=Annotation){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import Annotation
    from neuroml.utils import component_factory

    variable = component_factory(
        Annotation,
        anytypeobjs_=None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <annotation>
        <rdf:RDF xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#" xmlns:bqbiol="http://biomodels.net/biology-qualifiers/">
            <rdf:Description rdf:about="HippoCA1Cell">
                <bqbiol:is>
                    <rdf:Bag>
                        <rdf:li rdf:resource="urn:miriam:neurondb:258"/>
                    </rdf:Bag>
                </bqbiol:is>
            </rdf:Description>
        </rdf:RDF>
    </annotation>
:::
:::
:::
:::
:::

::: {#property .section}
[]{#schema-property}

##### property[\#](#property "Link to this heading"){.headerlink}

*A property ( a **tag** and **value** pair ), which can be on any
[[baseStandalone]{.std .std-ref}](#schema-basestandalone){.reference
.internal} either as a direct child, or within an [[annotation]{.std
.std-ref}](#schema-annotation){.reference .internal}. Generally
something which helps the visual display or facilitates simulation of a
Component, but is not a core physiological property. Common examples
include: **numberInternalDivisions,** equivalent of nseg in NEURON;
**radius,** for a radius to use in graphical displays for abstract cells
( i.e. without defined morphologies ); **color,** the color to use for a
[[population]{.std .std-ref}](Landing.html#schema-population){.reference
.internal} or [[populationList]{.std
.std-ref}](Landing.html#schema-populationlist){.reference .internal} of
cells; **recommended_dt_ms,** the recommended timestep to use for
simulating a [[network]{.std
.std-ref}](Landing.html#schema-network){.reference .internal},
**recommended_duration_ms** the recommended duration to use when running
a [[network]{.std .std-ref}](Landing.html#schema-network){.reference
.internal}.*

::: {.sd-tab-set .docutils}
Text fields

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| *      | Name of the property                                         |
| *tag** |                                                              |
+--------+--------------------------------------------------------------+
| **v    | Value of the property                                        |
| alue** |                                                              |
+--------+--------------------------------------------------------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="Property">
      <xs:complexContent>
        <xs:extension base="BaseWithoutId">
          <xs:attribute name="tag" type="xs:string" use="required"/>
          <xs:attribute name="value" type="xs:string" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=Property){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import Property
    from neuroml.utils import component_factory

    variable = component_factory(
        Property,
        tag: 'a string (required)' = None,
        value: 'a string (required)' = None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <property tag="numberInternalDivisions" value="9"/>
:::
:::
:::
:::
:::

::: {#basestandalone .section}
[]{#schema-basestandalone}

##### *baseStandalone*[\#](#basestandalone "Link to this heading"){.headerlink}

*Base type of any Component which can have [[notes]{.std
.std-ref}](#schema-notes){.reference .internal}, [[annotation]{.std
.std-ref}](#schema-annotation){.reference .internal}, or a
[[property]{.std .std-ref}](#schema-property){.reference .internal}
list.*

::: {.sd-tab-set .docutils}
Child list

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    |                                                 | [[          |
| *not |                                                 | notes]{.std |
| es** |                                                 | .st         |
|      |                                                 | d-ref}](#sc |
|      |                                                 | hema-notes) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   |                                                 | [[annot     |
| anno |                                                 | ation]{.std |
| tati |                                                 | .std-ref    |
| on** |                                                 | }](#schema- |
|      |                                                 | annotation) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Children list

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **pr |                                                 | [[pro       |
| oper |                                                 | perty]{.std |
| ty** |                                                 | .std-r      |
|      |                                                 | ef}](#schem |
|      |                                                 | a-property) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="Standalone">
      <xs:complexContent>
        <xs:extension base="Base">
          <xs:sequence>
            <xs:element name="notes" type="Notes" minOccurs="0"/>
            <xs:element name="property" type="Property" minOccurs="0" maxOccurs="unbounded"/>
            <xs:element name="annotation" type="Annotation" minOccurs="0"/>
          </xs:sequence>
          <xs:attribute name="metaid" type="MetaId" use="optional"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::
:::
:::

::: {#rdf-rdf .section}
[]{#schema-rdf-rdf}

##### rdf_RDF[\#](#rdf-rdf "Link to this heading"){.headerlink}

*Structured block in an [[annotation]{.std
.std-ref}](#schema-annotation){.reference .internal} based on RDF. See
[OpenSourceBrain/OSB_API](https://github.com/OpenSourceBrain/OSB_API/blob/master/python/examples/grancelllayer.xml){.github
.reference .external}.*

::: {.sd-tab-set .docutils}
Text fields

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| *      |                                                              |
| *xmlns |                                                              |
| :rdf** |                                                              |
+--------+--------------------------------------------------------------+
:::
:::

Child list

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **r  |                                                 | [           |
| df:D |                                                 | [rdf_Descri |
| escr |                                                 | ption]{.std |
| ipti |                                                 | .s          |
| on** |                                                 | td-ref}](#s |
|      |                                                 | chema-rdf-d |
|      |                                                 | escription) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::
:::
:::

::: {#rdf-description .section}
[]{#schema-rdf-description}

##### rdf_Description[\#](#rdf-description "Link to this heading"){.headerlink}

*Structured block in an [[annotation]{.std
.std-ref}](#schema-annotation){.reference .internal} based on RDF.*

::: {.sd-tab-set .docutils}
Text fields

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| *      |                                                              |
| *rdf:a |                                                              |
| bout** |                                                              |
+--------+--------------------------------------------------------------+
:::
:::

Child list

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   |                                                 | [[bqbiol_en |
| bqbi |                                                 | codes]{.std |
| ol:e |                                                 | .           |
| ncod |                                                 | std-ref}](# |
| es** |                                                 | schema-bqbi |
|      |                                                 | ol-encodes) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   |                                                 | [[bqbiol_ha |
| bqbi |                                                 | sPart]{.std |
| ol:h |                                                 | .           |
| asPa |                                                 | std-ref}](# |
| rt** |                                                 | schema-bqbi |
|      |                                                 | ol-haspart) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   |                                                 | [[bq        |
| bqbi |                                                 | biol_hasPro |
| ol:h |                                                 | perty]{.std |
| asPr |                                                 | .std-       |
| oper |                                                 | ref}](#sche |
| ty** |                                                 | ma-bqbiol-h |
|      |                                                 | asproperty) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    |                                                 | [[b         |
| *bqb |                                                 | qbiol_hasVe |
| iol: |                                                 | rsion]{.std |
| hasV |                                                 | .std        |
| ersi |                                                 | -ref}](#sch |
| on** |                                                 | ema-bqbiol- |
|      |                                                 | hasversion) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    |                                                 | [[bqbi      |
| *bqb |                                                 | ol_is]{.std |
| iol: |                                                 | .std-re     |
| is** |                                                 | f}](#schema |
|      |                                                 | -bqbiol-is) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **bq |                                                 | [[bqbi      |
| biol |                                                 | ol_isDescri |
| :isD |                                                 | bedBy]{.std |
| escr |                                                 | .std-re     |
| ibed |                                                 | f}](#schema |
| By** |                                                 | -bqbiol-isd |
|      |                                                 | escribedby) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   |                                                 | [[bq        |
| bqbi |                                                 | biol_isEnco |
| ol:i |                                                 | dedBy]{.std |
| sEnc |                                                 | .std-       |
| oded |                                                 | ref}](#sche |
| By** |                                                 | ma-bqbiol-i |
|      |                                                 | sencodedby) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   |                                                 | [[bq        |
| bqbi |                                                 | biol_isHomo |
| ol:i |                                                 | logTo]{.std |
| sHom |                                                 | .std-       |
| olog |                                                 | ref}](#sche |
| To** |                                                 | ma-bqbiol-i |
|      |                                                 | shomologto) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **b  |                                                 | [           |
| qbio |                                                 | [bqbiol_isP |
| l:is |                                                 | artOf]{.std |
| Part |                                                 | .s          |
| Of** |                                                 | td-ref}](#s |
|      |                                                 | chema-bqbio |
|      |                                                 | l-ispartof) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **b  |                                                 | [[bqb       |
| qbio |                                                 | iol_isPrope |
| l:is |                                                 | rtyOf]{.std |
| Prop |                                                 | .std-r      |
| erty |                                                 | ef}](#schem |
| Of** |                                                 | a-bqbiol-is |
|      |                                                 | propertyof) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   |                                                 | [[bq        |
| bqbi |                                                 | biol_isVers |
| ol:i |                                                 | ionOf]{.std |
| sVer |                                                 | .std-       |
| sion |                                                 | ref}](#sche |
| Of** |                                                 | ma-bqbiol-i |
|      |                                                 | sversionof) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **b  |                                                 | [           |
| qbio |                                                 | [bqbiol_occ |
| l:oc |                                                 | ursIn]{.std |
| curs |                                                 | .s          |
| In** |                                                 | td-ref}](#s |
|      |                                                 | chema-bqbio |
|      |                                                 | l-occursin) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **b  |                                                 | [           |
| qbio |                                                 | [bqbiol_has |
| l:ha |                                                 | Taxon]{.std |
| sTax |                                                 | .s          |
| on** |                                                 | td-ref}](#s |
|      |                                                 | chema-bqbio |
|      |                                                 | l-hastaxon) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   |                                                 | [[bqmod     |
| bqmo |                                                 | el_is]{.std |
| del: |                                                 | .std-ref    |
| is** |                                                 | }](#schema- |
|      |                                                 | bqmodel-is) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    |                                                 | [[bqmod     |
| *bqm |                                                 | el_isDescri |
| odel |                                                 | bedBy]{.std |
| :isD |                                                 | .std-ref    |
| escr |                                                 | }](#schema- |
| ibed |                                                 | bqmodel-isd |
| By** |                                                 | escribedby) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    |                                                 | [[bqmod     |
| *bqm |                                                 | el_isDerive |
| odel |                                                 | dFrom]{.std |
| :isD |                                                 | .std-ref    |
| eriv |                                                 | }](#schema- |
| edFr |                                                 | bqmodel-isd |
| om** |                                                 | erivedfrom) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::
:::
:::

::: {#basebqbiol .section}
[]{#schema-basebqbiol}

##### *baseBqbiol*[\#](#basebqbiol "Link to this heading"){.headerlink}

*Structured block in an [[annotation]{.std
.std-ref}](#schema-annotation){.reference .internal} based on RDF.*

::: {.sd-tab-set .docutils}
Child list

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **r  |                                                 | [[rd        |
| df:B |                                                 | f_Bag]{.std |
| ag** |                                                 | .std-       |
|      |                                                 | ref}](#sche |
|      |                                                 | ma-rdf-bag) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::
:::
:::

::: {#bqbiol-encodes .section}
[]{#schema-bqbiol-encodes}

##### bqbiol_encodes[\#](#bqbiol-encodes "Link to this heading"){.headerlink}

extends *[[baseBqbiol]{.std .std-ref}](#schema-basebqbiol){.reference
.internal}*

*See
[http://co.mbine.org/standards/qualifiers](http://co.mbine.org/standards/qualifiers){.reference
.external}.*
:::

::: {#bqbiol-haspart .section}
[]{#schema-bqbiol-haspart}

##### bqbiol_hasPart[\#](#bqbiol-haspart "Link to this heading"){.headerlink}

extends *[[baseBqbiol]{.std .std-ref}](#schema-basebqbiol){.reference
.internal}*

*See
[http://co.mbine.org/standards/qualifiers](http://co.mbine.org/standards/qualifiers){.reference
.external}.*
:::

::: {#bqbiol-hasproperty .section}
[]{#schema-bqbiol-hasproperty}

##### bqbiol_hasProperty[\#](#bqbiol-hasproperty "Link to this heading"){.headerlink}

extends *[[baseBqbiol]{.std .std-ref}](#schema-basebqbiol){.reference
.internal}*

*See
[http://co.mbine.org/standards/qualifiers](http://co.mbine.org/standards/qualifiers){.reference
.external}.*
:::

::: {#bqbiol-hasversion .section}
[]{#schema-bqbiol-hasversion}

##### bqbiol_hasVersion[\#](#bqbiol-hasversion "Link to this heading"){.headerlink}

extends *[[baseBqbiol]{.std .std-ref}](#schema-basebqbiol){.reference
.internal}*

*See
[http://co.mbine.org/standards/qualifiers](http://co.mbine.org/standards/qualifiers){.reference
.external}.*
:::

::: {#bqbiol-is .section}
[]{#schema-bqbiol-is}

##### bqbiol_is[\#](#bqbiol-is "Link to this heading"){.headerlink}

extends *[[baseBqbiol]{.std .std-ref}](#schema-basebqbiol){.reference
.internal}*

*See
[http://co.mbine.org/standards/qualifiers](http://co.mbine.org/standards/qualifiers){.reference
.external}.*
:::

::: {#bqbiol-isdescribedby .section}
[]{#schema-bqbiol-isdescribedby}

##### bqbiol_isDescribedBy[\#](#bqbiol-isdescribedby "Link to this heading"){.headerlink}

extends *[[baseBqbiol]{.std .std-ref}](#schema-basebqbiol){.reference
.internal}*

*See
[http://co.mbine.org/standards/qualifiers](http://co.mbine.org/standards/qualifiers){.reference
.external}.*
:::

::: {#bqbiol-isencodedby .section}
[]{#schema-bqbiol-isencodedby}

##### bqbiol_isEncodedBy[\#](#bqbiol-isencodedby "Link to this heading"){.headerlink}

extends *[[baseBqbiol]{.std .std-ref}](#schema-basebqbiol){.reference
.internal}*

*See
[http://co.mbine.org/standards/qualifiers](http://co.mbine.org/standards/qualifiers){.reference
.external}.*
:::

::: {#bqbiol-ishomologto .section}
[]{#schema-bqbiol-ishomologto}

##### bqbiol_isHomologTo[\#](#bqbiol-ishomologto "Link to this heading"){.headerlink}

extends *[[baseBqbiol]{.std .std-ref}](#schema-basebqbiol){.reference
.internal}*

*See
[http://co.mbine.org/standards/qualifiers](http://co.mbine.org/standards/qualifiers){.reference
.external}.*
:::

::: {#bqbiol-ispartof .section}
[]{#schema-bqbiol-ispartof}

##### bqbiol_isPartOf[\#](#bqbiol-ispartof "Link to this heading"){.headerlink}

extends *[[baseBqbiol]{.std .std-ref}](#schema-basebqbiol){.reference
.internal}*

*See
[http://co.mbine.org/standards/qualifiers](http://co.mbine.org/standards/qualifiers){.reference
.external}.*
:::

::: {#bqbiol-ispropertyof .section}
[]{#schema-bqbiol-ispropertyof}

##### bqbiol_isPropertyOf[\#](#bqbiol-ispropertyof "Link to this heading"){.headerlink}

extends *[[baseBqbiol]{.std .std-ref}](#schema-basebqbiol){.reference
.internal}*

*See
[http://co.mbine.org/standards/qualifiers](http://co.mbine.org/standards/qualifiers){.reference
.external}.*
:::

::: {#bqbiol-isversionof .section}
[]{#schema-bqbiol-isversionof}

##### bqbiol_isVersionOf[\#](#bqbiol-isversionof "Link to this heading"){.headerlink}

extends *[[baseBqbiol]{.std .std-ref}](#schema-basebqbiol){.reference
.internal}*

*See
[http://co.mbine.org/standards/qualifiers](http://co.mbine.org/standards/qualifiers){.reference
.external}.*

::: {.sd-tab-set .docutils}
Text fields

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| **xm   |                                                              |
| lns:bq |                                                              |
| biol** |                                                              |
+--------+--------------------------------------------------------------+
:::
:::
:::
:::

::: {#bqbiol-occursin .section}
[]{#schema-bqbiol-occursin}

##### bqbiol_occursIn[\#](#bqbiol-occursin "Link to this heading"){.headerlink}

extends *[[baseBqbiol]{.std .std-ref}](#schema-basebqbiol){.reference
.internal}*

*See
[http://co.mbine.org/standards/qualifiers](http://co.mbine.org/standards/qualifiers){.reference
.external}.*
:::

::: {#bqbiol-hastaxon .section}
[]{#schema-bqbiol-hastaxon}

##### bqbiol_hasTaxon[\#](#bqbiol-hastaxon "Link to this heading"){.headerlink}

extends *[[baseBqbiol]{.std .std-ref}](#schema-basebqbiol){.reference
.internal}*

*See
[http://co.mbine.org/standards/qualifiers](http://co.mbine.org/standards/qualifiers){.reference
.external}.*
:::

::: {#bqmodel-is .section}
[]{#schema-bqmodel-is}

##### bqmodel_is[\#](#bqmodel-is "Link to this heading"){.headerlink}

extends *[[baseBqbiol]{.std .std-ref}](#schema-basebqbiol){.reference
.internal}*

*See
[http://co.mbine.org/standards/qualifiers](http://co.mbine.org/standards/qualifiers){.reference
.external}.*
:::

::: {#bqmodel-isdescribedby .section}
[]{#schema-bqmodel-isdescribedby}

##### bqmodel_isDescribedBy[\#](#bqmodel-isdescribedby "Link to this heading"){.headerlink}

extends *[[baseBqbiol]{.std .std-ref}](#schema-basebqbiol){.reference
.internal}*

*See
[http://co.mbine.org/standards/qualifiers](http://co.mbine.org/standards/qualifiers){.reference
.external}.*

::: {.sd-tab-set .docutils}
Text fields

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| **xml  |                                                              |
| ns:bqm |                                                              |
| odel** |                                                              |
+--------+--------------------------------------------------------------+
:::
:::
:::
:::

::: {#bqmodel-isderivedfrom .section}
[]{#schema-bqmodel-isderivedfrom}

##### bqmodel_isDerivedFrom[\#](#bqmodel-isderivedfrom "Link to this heading"){.headerlink}

extends *[[baseBqbiol]{.std .std-ref}](#schema-basebqbiol){.reference
.internal}*

*See
[http://co.mbine.org/standards/qualifiers](http://co.mbine.org/standards/qualifiers){.reference
.external}.*
:::

::: {#rdf-bag .section}
[]{#schema-rdf-bag}

##### rdf_Bag[\#](#rdf-bag "Link to this heading"){.headerlink}

*Structured block in an [[annotation]{.std
.std-ref}](#schema-annotation){.reference .internal} based on RDF.*

::: {.sd-tab-set .docutils}
Children list

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   |                                                 | [schema:rd  |
| rdf: |                                                 | f:li]{.xref |
| li** |                                                 | .std        |
|      |                                                 | .std-ref}   |
+------+-------------------------------------------------+-------------+
:::
:::
:::
:::

::: {#rdf-li .section}
[]{#schema-rdf-li}

##### rdf_li[\#](#rdf-li "Link to this heading"){.headerlink}

*Structured block in an [[annotation]{.std
.std-ref}](#schema-annotation){.reference .internal} based on RDF.*

::: {.sd-tab-set .docutils}
Text fields

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| **rd   |                                                              |
| f:reso |                                                              |
| urce** |                                                              |
+--------+--------------------------------------------------------------+
:::
:::
:::
:::

::: {#point3dwithdiam .section}
[]{#schema-point3dwithdiam}

##### point3DWithDiam[\#](#point3dwithdiam "Link to this heading"){.headerlink}

*Base type for ComponentTypes which specify an ( **x,** **y,** **z** )
coordinate along with a **diameter.** Note: no dimension used in the
attributes for these coordinates! These are assumed to have dimension
micrometer ( 10\^-6 m ). This is due to micrometers being the default
option for the majority of neuronal morphology formats, and dimensions
are omitted here to facilitate reading and writing of morphologies in
NeuroML.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **di | Diameter of the ppoint. Note: no dimension      | Di          |
| amet | used, see description of *point3DWithDiam* for  | mensionless |
| er** | details.                                        |             |
+------+-------------------------------------------------+-------------+
| *    | x coordinate of the point. Note: no dimension   | Di          |
| *x** | used, see description of *point3DWithDiam* for  | mensionless |
|      | details.                                        |             |
+------+-------------------------------------------------+-------------+
| *    | y coordinate of the ppoint. Note: no dimension  | Di          |
| *y** | used, see description of *point3DWithDiam* for  | mensionless |
|      | details.                                        |             |
+------+-------------------------------------------------+-------------+
| *    | z coordinate of the ppoint. Note: no dimension  | Di          |
| *z** | used, see description of *point3DWithDiam* for  | mensionless |
|      | details.                                        |             |
+------+-------------------------------------------------+-------------+
:::
:::

Constants

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------------------+-----------------------------------+-------------+
| **MICRON** = 1um   |                                   | [[l         |
|                    |                                   | ength]{.std |
|                    |                                   | .std-       |
|                    |                                   | ref}](Landi |
|                    |                                   | ng.html#sch |
|                    |                                   | ema-dimensi |
|                    |                                   | ons-length) |
|                    |                                   | {.reference |
|                    |                                   | .internal}  |
+--------------------+-----------------------------------+-------------+
:::
:::

Derived parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   | A dimensional quantity given by half the        | [[l         |
| radi | \_diameter.                                     | ength]{.std |
| us** |                                                 | .std-       |
|      |                                                 | ref}](Landi |
|      |                                                 | ng.html#sch |
|      |                                                 | ema-dimensi |
|      |                                                 | ons-length) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::

   **radius** = MICRON \* diameter / 2

::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **x  | A version of \_x with dimension length.         | [[l         |
| Leng |                                                 | ength]{.std |
| th** |                                                 | .std-       |
|      |                                                 | ref}](Landi |
|      |                                                 | ng.html#sch |
|      |                                                 | ema-dimensi |
|      |                                                 | ons-length) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::

   **xLength** = MICRON \* x

::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **y  | A version of \_y with dimension length.         | [[l         |
| Leng |                                                 | ength]{.std |
| th** |                                                 | .std-       |
|      |                                                 | ref}](Landi |
|      |                                                 | ng.html#sch |
|      |                                                 | ema-dimensi |
|      |                                                 | ons-length) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::

   **yLength** = MICRON \* y

::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **z  | A version of \_z with dimension length.         | [[l         |
| Leng |                                                 | ength]{.std |
| th** |                                                 | .std-       |
|      |                                                 | ref}](Landi |
|      |                                                 | ng.html#sch |
|      |                                                 | ema-dimensi |
|      |                                                 | ons-length) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::

   **zLength** = MICRON \* z
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="Point3DWithDiam">
      <xs:complexContent>
        <xs:extension base="BaseWithoutId">
          <xs:attribute name="x" type="xs:double" use="required"/>
          <xs:attribute name="y" type="xs:double" use="required"/>
          <xs:attribute name="z" type="xs:double" use="required"/>
          <xs:attribute name="diameter" type="DoubleGreaterThanZero" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=Point3DWithDiam){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import Point3DWithDiam
    from neuroml.utils import component_factory

    variable = component_factory(
        Point3DWithDiam,
        x: 'a double (required)' = None,
        y: 'a double (required)' = None,
        z: 'a double (required)' = None,
        diameter: 'a DoubleGreaterThanZero (required)' = None,
    )
:::
:::
:::
:::
:::
:::

[]{#document-Userdocs/Schemas/Cells}

::: {#cells .section .tex2jax_ignore .mathjax_ignore}
[]{#schema-cells}

#### Cells[\#](#cells "Link to this heading"){.headerlink}

**Defines both abstract cell models ( e.g. [[izhikevichCell]{.std
.std-ref}](#schema-izhikevichcell){.reference .internal}, adaptive
exponential integrate and fire cell, [[adExIaFCell]{.std
.std-ref}](#schema-adexiafcell){.reference .internal} ), point
conductance based cell models ( [[pointCellCondBased]{.std
.std-ref}](#schema-pointcellcondbased){.reference .internal},
[[pointCellCondBasedCa]{.std
.std-ref}](#schema-pointcellcondbasedca){.reference .internal} ) and
cells models ( [[cell]{.std .std-ref}](#schema-cell){.reference
.internal} ) which specify the [[morphology]{.std
.std-ref}](#schema-morphology){.reference .internal} ( containing
[[segment]{.std .std-ref}](#schema-segment){.reference .internal}s ) and
[[biophysicalProperties]{.std
.std-ref}](#schema-biophysicalproperties){.reference .internal}
separately.**

------------------------------------------------------------------------

Original ComponentType definitions:
[Cells.xml](https://github.com/NeuroML/NeuroML2/blob/master/NeuroML2CoreTypes//Cells.xml){.reference
.external}. Schema against which NeuroML based on these should be valid:
[NeuroML_v2.3.xsd](https://github.com/NeuroML/NeuroML2/tree/master/Schemas/NeuroML2/NeuroML_v2.3.xsd){.reference
.external}. Generated on 14/08/24 from
[this](https://github.com/NeuroML/NeuroML2/commit/352244cff605cb1ba24fa7c11757dc818fe90fd2){.reference
.external} commit. Please file any issues or questions at the [issue
tracker here](https://github.com/NeuroML/NeuroML2/issues){.reference
.external}.

------------------------------------------------------------------------

::: {#basecell .section}
[]{#schema-basecell}

##### *baseCell*[\#](#basecell "Link to this heading"){.headerlink}

extends *[[baseStandalone]{.std
.std-ref}](Landing.html#schema-basestandalone){.reference .internal}*

*Base type of any cell ( e.g. point neuron like
[[izhikevich2007Cell]{.std
.std-ref}](#schema-izhikevich2007cell){.reference .internal}, or a
morphologically detailed [[cell]{.std
.std-ref}](#schema-cell){.reference .internal} with [[segment]{.std
.std-ref}](#schema-segment){.reference .internal}s ) which can be used
in a [[population]{.std
.std-ref}](Landing.html#schema-population){.reference .internal}.*

::: {.sd-tab-set .docutils}
Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="BaseCell">
      <xs:complexContent>
        <xs:extension base="Standalone">
          <xs:attribute name="neuroLexId" type="NeuroLexId" use="optional"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=BaseCell){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import BaseCell
    from neuroml.utils import component_factory

    variable = component_factory(
        BaseCell,
        id: 'a NmlId (required)' = None,
        metaid: 'a MetaId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        neuro_lex_id: 'a NeuroLexId (optional)' = None,
        extensiontype_=None,
    )
:::
:::
:::
:::
:::

::: {#basespikingcell .section}
[]{#schema-basespikingcell}

##### *baseSpikingCell*[\#](#basespikingcell "Link to this heading"){.headerlink}

extends *[[baseCell]{.std .std-ref}](#schema-basecell){.reference
.internal}*

*Base type of any cell which can emit **spike** events.*

::: {.sd-tab-set .docutils}
Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Spike event                                     | Direction:  |
| *spi |                                                 | out         |
| ke** |                                                 |             |
+------+-------------------------------------------------+-------------+
:::
:::
:::
:::

::: {#basecellmembpot .section}
[]{#schema-basecellmembpot}

##### *baseCellMembPot*[\#](#basecellmembpot "Link to this heading"){.headerlink}

extends *[[baseSpikingCell]{.std
.std-ref}](#schema-basespikingcell){.reference .internal}*

*Any spiking cell which has a membrane potential **v** with units of
voltage ( as opposed to a dimensionless membrane potential used in
[[baseCellMembPotDL]{.std
.std-ref}](#schema-basecellmembpotdl){.reference .internal} ).*

::: {.sd-tab-set .docutils}
Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Membrane potential                              | [[vo        |
| *v** |                                                 | ltage]{.std |
|      |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Spike event *(from [[baseSpikingCell]{.std      | Direction:  |
| *spi | .std-ref}](#schema-basespikingcell){.reference  | out         |
| ke** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::
:::
:::

::: {#basecellmembpotdl .section}
[]{#schema-basecellmembpotdl}

##### *baseCellMembPotDL*[\#](#basecellmembpotdl "Link to this heading"){.headerlink}

extends *[[baseSpikingCell]{.std
.std-ref}](#schema-basespikingcell){.reference .internal}*

*Any spiking cell which has a dimensioness membrane potential, **V** (
as opposed to a membrane potential units of voltage,
[[baseCellMembPot]{.std .std-ref}](#schema-basecellmembpot){.reference
.internal} ).*

::: {.sd-tab-set .docutils}
Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Membrane potential                              | Di          |
| *V** |                                                 | mensionless |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Spike event *(from [[baseSpikingCell]{.std      | Direction:  |
| *spi | .std-ref}](#schema-basespikingcell){.reference  | out         |
| ke** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::
:::
:::

::: {#basechannelpopulation .section}
[]{#schema-basechannelpopulation}

##### *baseChannelPopulation*[\#](#basechannelpopulation "Link to this heading"){.headerlink}

extends *[[baseVoltageDepPointCurrent]{.std
.std-ref}](Landing.html#schema-basevoltagedeppointcurrent){.reference
.internal}*

*Base type for any current produced by a population of channels, all of
which are of type **ionChannel**.*

::: {.sd-tab-set .docutils}
Component References

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   |                                                 | [[baseIonCh |
| ionC |                                                 | annel]{.std |
| hann |                                                 | .s          |
| el** |                                                 | td-ref}](La |
|      |                                                 | nding.html# |
|      |                                                 | schema-base |
|      |                                                 | ionchannel) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | The total (usually time varying) current        | [[cu        |
| *i** | produced by this ComponentType *(from           | rrent]{.std |
|      | [[basePointCurrent]{.std                        | .std-r      |
|      | .std-ref}](L                                    | ef}](Landin |
|      | anding.html#schema-basepointcurrent){.reference | g.html#sche |
|      | .internal})*                                    | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | The current may vary with the voltage exposed   | [[vo        |
| *v** | by the ComponentType on which this is placed    | ltage]{.std |
|      | *(from [[baseVoltageDepPointCurrent]{.std       | .std-r      |
|      | .std-ref}](Landing.htm                          | ef}](Landin |
|      | l#schema-basevoltagedeppointcurrent){.reference | g.html#sche |
|      | .internal})*                                    | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::
:::
:::

::: {#channelpopulation .section}
[]{#schema-channelpopulation}

##### channelPopulation[\#](#channelpopulation "Link to this heading"){.headerlink}

extends *[[baseChannelPopulation]{.std
.std-ref}](#schema-basechannelpopulation){.reference .internal}*

*Population of a **number** of ohmic ion channels. These each produce a
conductance **channelg** across a reversal potential **erev,** giving a
total current **i.** Note that active membrane currents are more
frequently specified as a density over an area of the [[cell]{.std
.std-ref}](#schema-cell){.reference .internal} using
[[channelDensity]{.std .std-ref}](#schema-channeldensity){.reference
.internal}.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **er | The reversal potential of the current produced  | [[vo        |
| ev** |                                                 | ltage]{.std |
|      |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   | The number of channels present. This will be    | Di          |
| numb | multiplied by the time varying conductance of   | mensionless |
| er** | the individual ion channel (which extends       |             |
|      | *baseIonChannel*) to produce the total          |             |
|      | conductance                                     |             |
+------+-------------------------------------------------+-------------+
:::
:::

Constants

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------------------+-----------------------------------+-------------+
| **vShift** = 0mV   | Set to a constant 0mV here to     | [[vo        |
|                    | allow ion channels which use      | ltage]{.std |
|                    | \_vShift in their rate variable   | .std-r      |
|                    | expressions to be used with       | ef}](Landin |
|                    | *channelPopulation*, not just     | g.html#sche |
|                    | with *channelDensityVShift*       | ma-dimensio |
|                    | (where \_vShift would be          | ns-voltage) |
|                    | explicitly set)                   | {.reference |
|                    |                                   | .internal}  |
+--------------------+-----------------------------------+-------------+
:::
:::

Text fields

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| *      | Which ion flows through the channel. Note: ideally this      |
| *ion** | needs to be a property of ionChannel only, but it's here as  |
|        | it makes it easier to select channelPopulations transmitting |
|        | specific ions.                                               |
+--------+--------------------------------------------------------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | The total (usually time varying) current        | [[cu        |
| *i** | produced by this ComponentType *(from           | rrent]{.std |
|      | [[basePointCurrent]{.std                        | .std-r      |
|      | .std-ref}](L                                    | ef}](Landin |
|      | anding.html#schema-basepointcurrent){.reference | g.html#sche |
|      | .internal})*                                    | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | The current may vary with the voltage exposed   | [[vo        |
| *v** | by the ComponentType on which this is placed    | ltage]{.std |
|      | *(from [[baseVoltageDepPointCurrent]{.std       | .std-r      |
|      | .std-ref}](Landing.htm                          | ef}](Landin |
|      | l#schema-basevoltagedeppointcurrent){.reference | g.html#sche |
|      | .internal})*                                    | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***Structure***

:   CHILD INSTANCE: **ionChannel**

***Derived Variables***

:   **channelg** = ionChannel-\>g

:   **geff** = channelg \* number

:   **i** = geff \* (erev - v) (exposed as **i**)
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="ChannelPopulation">
      <xs:complexContent>
        <xs:extension base="Base">
          <xs:sequence>
            <xs:element name="variableParameter" type="VariableParameter" minOccurs="0" maxOccurs="unbounded"/>
          </xs:sequence>
          <xs:attribute name="ionChannel" type="NmlId" use="required"/>
          <xs:attribute name="number" type="NonNegativeInteger" use="required"/>
          <xs:attribute name="erev" type="Nml2Quantity_voltage" use="required"/>
          <xs:attribute name="segmentGroup" type="NmlId" use="optional" default="all"/>
          <xs:attribute name="segment" type="NonNegativeInteger" use="optional"/>
          <xs:attribute name="ion" type="NmlId" use="required">
            <xs:annotation>
                </xs:annotation>
          </xs:attribute>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=ChannelPopulation){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import ChannelPopulation
    from neuroml.utils import component_factory

    variable = component_factory(
        ChannelPopulation,
        id: 'a NmlId (required)' = None,
        ion_channel: 'a NmlId (required)' = None,
        number: 'a NonNegativeInteger (required)' = None,
        erev: 'a Nml2Quantity_voltage (required)' = None,
        segment_groups: 'a NmlId (optional)' = 'all',
        segments: 'a NonNegativeInteger (optional)' = None,
        ion: 'a NmlId (required)' = None,
        variable_parameters: 'list of VariableParameter(s) (optional)' = None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <channelPopulation id="naChansDend" ionChannel="NaConductance" segment="2" number="120000" erev="50mV" ion="na"/>
:::
:::
:::
:::
:::

::: {#channelpopulationnernst .section}
[]{#schema-channelpopulationnernst}

##### channelPopulationNernst[\#](#channelpopulationnernst "Link to this heading"){.headerlink}

extends *[[baseChannelPopulation]{.std
.std-ref}](#schema-basechannelpopulation){.reference .internal}*

*Population of a **number** of channels with a time varying reversal
potential **erev** determined by Nernst equation. Note: hard coded for
Ca only!*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   | The number of channels present. This will be    | Di          |
| numb | multiplied by the time varying conductance of   | mensionless |
| er** | the individual ion channel (which extends       |             |
|      | *baseIonChannel*) to produce the total          |             |
|      | conductance                                     |             |
+------+-------------------------------------------------+-------------+
:::
:::

Constants

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------------------+-----------------------------------+-------------+
| **R** = 8.3144621  |                                   | [[idea      |
| J_per_K_per_mol    |                                   | lGasConstan |
|                    |                                   | tDims]{.std |
|                    |                                   | .std-ref    |
|                    |                                   | }](Landing. |
|                    |                                   | html#schema |
|                    |                                   | -dimensions |
|                    |                                   | -idealgasco |
|                    |                                   | nstantdims) |
|                    |                                   | {.reference |
|                    |                                   | .internal}  |
+--------------------+-----------------------------------+-------------+
| **zCa** = 2        |                                   | Di          |
|                    |                                   | mensionless |
+--------------------+-----------------------------------+-------------+
| **F** = 96485.3    |                                   | [           |
| C_per_mol          |                                   | [charge_per |
|                    |                                   | _mole]{.std |
|                    |                                   | .st         |
|                    |                                   | d-ref}](Lan |
|                    |                                   | ding.html#s |
|                    |                                   | chema-dimen |
|                    |                                   | sions-charg |
|                    |                                   | e-per-mole) |
|                    |                                   | {.reference |
|                    |                                   | .internal}  |
+--------------------+-----------------------------------+-------------+
| **vShift** = 0mV   | Set to a constant 0mV here to     | [[vo        |
|                    | allow ion channels which use      | ltage]{.std |
|                    | \_vShift in their rate variable   | .std-r      |
|                    | expressions to be used with       | ef}](Landin |
|                    | *channelPopulation*, not just     | g.html#sche |
|                    | with *channelDensityVShift*       | ma-dimensio |
|                    | (where \_vShift would be          | ns-voltage) |
|                    | explicitly set)                   | {.reference |
|                    |                                   | .internal}  |
+--------------------+-----------------------------------+-------------+
:::
:::

Text fields

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| *      | Which ion flows through the channel. Note: ideally this      |
| *ion** | needs to be a property of ionChannel only, but it's here as  |
|        | it makes it easier to select channelPopulations transmitting |
|        | specific ions.                                               |
+--------+--------------------------------------------------------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **er | The reversal potential of the current produced, | [[vo        |
| ev** | calculated from \_caConcExt and \_caConc        | ltage]{.std |
|      |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | The total (usually time varying) current        | [[cu        |
| *i** | produced by this ComponentType *(from           | rrent]{.std |
|      | [[basePointCurrent]{.std                        | .std-r      |
|      | .std-ref}](L                                    | ef}](Landin |
|      | anding.html#schema-basepointcurrent){.reference | g.html#sche |
|      | .internal})*                                    | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   | The internal Ca2+ concentration, as             | [[concentr  |
| caCo | calculated/exposed by the parent Component      | ation]{.std |
| nc** |                                                 | .           |
|      |                                                 | std-ref}](L |
|      |                                                 | anding.html |
|      |                                                 | #schema-dim |
|      |                                                 | ensions-con |
|      |                                                 | centration) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | The external Ca2+ concentration, as             | [[concentr  |
| *caC | calculated/exposed by the parent Component      | ation]{.std |
| oncE |                                                 | .           |
| xt** |                                                 | std-ref}](L |
|      |                                                 | anding.html |
|      |                                                 | #schema-dim |
|      |                                                 | ensions-con |
|      |                                                 | centration) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **t  | The temperature to use in the calculation of    | [[temper    |
| empe | \_erev. Note this is generally exposed by a     | ature]{.std |
| ratu | *networkWithTemperature*.                       | .std-ref}]  |
| re** |                                                 | (Landing.ht |
|      |                                                 | ml#schema-d |
|      |                                                 | imensions-t |
|      |                                                 | emperature) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | The current may vary with the voltage exposed   | [[vo        |
| *v** | by the ComponentType on which this is placed    | ltage]{.std |
|      | *(from [[baseVoltageDepPointCurrent]{.std       | .std-r      |
|      | .std-ref}](Landing.htm                          | ef}](Landin |
|      | l#schema-basevoltagedeppointcurrent){.reference | g.html#sche |
|      | .internal})*                                    | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***Structure***

:   CHILD INSTANCE: **ionChannel**

***Derived Variables***

:   **singleChannelConductance** = ionChannel-\>g

:   **totalConductance** = singleChannelConductance \* number

:   **erev** = (R \* temperature / (zCa \* F)) \* log(caConcExt /
    caConc) (exposed as **erev**)

:   **i** = totalConductance \* (erev - v) (exposed as **i**)
:::
:::
:::

::: {#basechanneldensity .section}
[]{#schema-basechanneldensity}

##### *baseChannelDensity*[\#](#basechanneldensity "Link to this heading"){.headerlink}

*Base type for a current of density **iDensity** distributed on an area
of a [[cell]{.std .std-ref}](#schema-cell){.reference .internal},
flowing through the specified **ionChannel.** Instances of this (
normally [[channelDensity]{.std
.std-ref}](#schema-channeldensity){.reference .internal} ) are specified
in the [[membraneProperties]{.std
.std-ref}](#schema-membraneproperties){.reference .internal} of the
[[cell]{.std .std-ref}](#schema-cell){.reference .internal}.*

::: {.sd-tab-set .docutils}
Component References

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   |                                                 | [[baseIonCh |
| ionC |                                                 | annel]{.std |
| hann |                                                 | .s          |
| el** |                                                 | td-ref}](La |
|      |                                                 | nding.html# |
|      |                                                 | schema-base |
|      |                                                 | ionchannel) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **iD |                                                 | [[currentDe |
| ensi |                                                 | nsity]{.std |
| ty** |                                                 | .s          |
|      |                                                 | td-ref}](La |
|      |                                                 | nding.html# |
|      |                                                 | schema-dime |
|      |                                                 | nsions-curr |
|      |                                                 | entdensity) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    |                                                 | [[vo        |
| *v** |                                                 | ltage]{.std |
|      |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::
:::
:::

::: {#basechanneldensitycond .section}
[]{#schema-basechanneldensitycond}

##### *baseChannelDensityCond*[\#](#basechanneldensitycond "Link to this heading"){.headerlink}

extends *[[baseChannelDensity]{.std
.std-ref}](#schema-basechanneldensity){.reference .internal}*

*Base type for distributed conductances on an area of a cell producing a
( not necessarily ohmic ) current.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **c  |                                                 | [[co        |
| ondD |                                                 | nductanceDe |
| ensi |                                                 | nsity]{.std |
| ty** |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-conducta |
|      |                                                 | ncedensity) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **gD |                                                 | [[co        |
| ensi |                                                 | nductanceDe |
| ty** |                                                 | nsity]{.std |
|      |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-conducta |
|      |                                                 | ncedensity) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **iD | *(from [[baseChannelDensity]{.std               | [[currentDe |
| ensi | .s                                              | nsity]{.std |
| ty** | td-ref}](#schema-basechanneldensity){.reference | .s          |
|      | .internal})*                                    | td-ref}](La |
|      |                                                 | nding.html# |
|      |                                                 | schema-dime |
|      |                                                 | nsions-curr |
|      |                                                 | entdensity) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseChannelDensity]{.std               | [[vo        |
| *v** | .s                                              | ltage]{.std |
|      | td-ref}](#schema-basechanneldensity){.reference | .std-r      |
|      | .internal})*                                    | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::
:::
:::

::: {#variableparameter .section}
[]{#schema-variableparameter}

##### variableParameter[\#](#variableparameter "Link to this heading"){.headerlink}

*Specifies a **parameter** ( e.g. condDensity ) which can vary its value
across a **segmentGroup.** The value is calculated from **value**
attribute of the [[inhomogeneousValue]{.std
.std-ref}](#schema-inhomogeneousvalue){.reference .internal} subelement.
This element is normally a child of [[channelDensityNonUniform]{.std
.std-ref}](#schema-channeldensitynonuniform){.reference .internal},
[[channelDensityNonUniformNernst]{.std
.std-ref}](#schema-channeldensitynonuniformnernst){.reference .internal}
or [[channelDensityNonUniformGHK]{.std
.std-ref}](#schema-channeldensitynonuniformghk){.reference .internal}
and is used to calculate the value of the conductance, etc. which will
vary on different parts of the cell. The **segmentGroup** specified here
needs to define an [[inhomogeneousParameter]{.std
.std-ref}](#schema-inhomogeneousparameter){.reference .internal} (
referenced from **inhomogeneousParameter** in the
[[inhomogeneousValue]{.std
.std-ref}](#schema-inhomogeneousvalue){.reference .internal} ), which
calculates a **variable** ( e.g. p ) varying across the cell ( e.g.
based on the path length from soma ), which is then used in the
**value** attribute of the [[inhomogeneousValue]{.std
.std-ref}](#schema-inhomogeneousvalue){.reference .internal} ( so for
example condDensity = f( p ) ).*

::: {.sd-tab-set .docutils}
Text fields

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| *      |                                                              |
| *param |                                                              |
| eter** |                                                              |
+--------+--------------------------------------------------------------+
| **se   |                                                              |
| gmentG |                                                              |
| roup** |                                                              |
+--------+--------------------------------------------------------------+
:::
:::

Child list

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   |                                                 | [[in        |
| inho |                                                 | homogeneous |
| moge |                                                 | Value]{.std |
| neou |                                                 | .std-       |
| sVal |                                                 | ref}](#sche |
| ue** |                                                 | ma-inhomoge |
|      |                                                 | neousvalue) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="VariableParameter">
      <xs:sequence>
        <xs:element name="inhomogeneousValue" type="InhomogeneousValue" minOccurs="0"/>
      </xs:sequence>
      <xs:attribute name="parameter" type="xs:string" use="required"/>
      <xs:attribute name="segmentGroup" type="xs:string" use="required"/>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=VariableParameter){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import VariableParameter
    from neuroml.utils import component_factory

    variable = component_factory(
        VariableParameter,
        parameter: 'a string (required)' = None,
        segment_groups: 'a string (required)' = None,
        inhomogeneous_value: 'a InhomogeneousValue (optional)' = None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <variableParameter parameter="condDensity" segmentGroup="dendrite_group">
        <inhomogeneousValue inhomogeneousParameter="dendrite_group_x1" value="5e-7 * exp(-p/200)"/>
    </variableParameter>
:::
:::
:::
:::
:::

::: {#inhomogeneousvalue .section}
[]{#schema-inhomogeneousvalue}

##### inhomogeneousValue[\#](#inhomogeneousvalue "Link to this heading"){.headerlink}

*Specifies the **value** of an **inhomogeneousParameter.** For usage see
[[variableParameter]{.std
.std-ref}](#schema-variableparameter){.reference .internal}.*

::: {.sd-tab-set .docutils}
Text fields

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| **     |                                                              |
| inhomo |                                                              |
| geneou |                                                              |
| sParam |                                                              |
| eter** |                                                              |
+--------+--------------------------------------------------------------+
| **v    |                                                              |
| alue** |                                                              |
+--------+--------------------------------------------------------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="InhomogeneousValue">
      <xs:complexContent>
        <xs:extension base="BaseWithoutId">
          <xs:attribute name="inhomogeneousParameter" type="xs:string" use="required"/>
          <xs:attribute name="value" type="xs:string" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=InhomogeneousValue){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import InhomogeneousValue
    from neuroml.utils import component_factory

    variable = component_factory(
        InhomogeneousValue,
        inhomogeneous_parameters: 'a string (required)' = None,
        value: 'a string (required)' = None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <inhomogeneousValue inhomogeneousParameter="dendrite_group_x1" value="5e-7 * exp(-p/200)"/>
:::
:::
:::
:::
:::

::: {#channeldensitynonuniform .section}
[]{#schema-channeldensitynonuniform}

##### channelDensityNonUniform[\#](#channeldensitynonuniform "Link to this heading"){.headerlink}

extends *[[baseChannelDensity]{.std
.std-ref}](#schema-basechanneldensity){.reference .internal}*

*Specifies a time varying ohmic conductance density, which is
distributed on a region of the **cell.** The conductance density of the
channel is not uniform, but is set using the [[variableParameter]{.std
.std-ref}](#schema-variableparameter){.reference .internal}. Note, there
is no dynamical description of this in LEMS yet, as this type only makes
sense for multicompartmental cells. A ComponentType for this needs to be
present to enable export of NeuroML 2 multicompartmental cells via
LEMS/jNeuroML to NEURON.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **er | The reversal potential of the current produced  | [[vo        |
| ev** |                                                 | ltage]{.std |
|      |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Constants

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------------------+-----------------------------------+-------------+
| **ZERO_CURR_DENS** |                                   | [[currentDe |
| = 0 A_per_m2       |                                   | nsity]{.std |
|                    |                                   | .s          |
|                    |                                   | td-ref}](La |
|                    |                                   | nding.html# |
|                    |                                   | schema-dime |
|                    |                                   | nsions-curr |
|                    |                                   | entdensity) |
|                    |                                   | {.reference |
|                    |                                   | .internal}  |
+--------------------+-----------------------------------+-------------+
:::
:::

Text fields

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| **se   |                                                              |
| gmentG |                                                              |
| roup** |                                                              |
+--------+--------------------------------------------------------------+
| *      | Which ion flows through the channel. Note: ideally this      |
| *ion** | needs to be a property of ionChannel only, but it's here as  |
|        | it makes it easier to select channelPopulations transmitting |
|        | specific ions.                                               |
+--------+--------------------------------------------------------------+
:::
:::

Child list

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    |                                                 | [[v         |
| *var |                                                 | ariablePara |
| iabl |                                                 | meter]{.std |
| ePar |                                                 | .std        |
| amet |                                                 | -ref}](#sch |
| er** |                                                 | ema-variabl |
|      |                                                 | eparameter) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **iD | *(from [[baseChannelDensity]{.std               | [[currentDe |
| ensi | .s                                              | nsity]{.std |
| ty** | td-ref}](#schema-basechanneldensity){.reference | .s          |
|      | .internal})*                                    | td-ref}](La |
|      |                                                 | nding.html# |
|      |                                                 | schema-dime |
|      |                                                 | nsions-curr |
|      |                                                 | entdensity) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseChannelDensity]{.std               | [[vo        |
| *v** | .s                                              | ltage]{.std |
|      | td-ref}](#schema-basechanneldensity){.reference | .std-r      |
|      | .internal})*                                    | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***Structure***

:   CHILD INSTANCE: **ionChannel**

***Derived Variables***

:   **iDensity** = ZERO_CURR_DENS (exposed as **iDensity**)
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="ChannelDensityNonUniform">
      <xs:complexContent>
        <xs:extension base="Base">
          <xs:sequence>
            <xs:element name="variableParameter" type="VariableParameter" minOccurs="0" maxOccurs="unbounded"/>
          </xs:sequence>
          <xs:attribute name="ionChannel" type="NmlId" use="required"/>
          <xs:attribute name="erev" type="Nml2Quantity_voltage" use="required"/>
          <xs:attribute name="ion" type="NmlId" use="required">
            <xs:annotation>
                </xs:annotation>
          </xs:attribute>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=ChannelDensityNonUniform){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import ChannelDensityNonUniform
    from neuroml.utils import component_factory

    variable = component_factory(
        ChannelDensityNonUniform,
        id: 'a NmlId (required)' = None,
        ion_channel: 'a NmlId (required)' = None,
        erev: 'a Nml2Quantity_voltage (required)' = None,
        ion: 'a NmlId (required)' = None,
        variable_parameters: 'list of VariableParameter(s) (optional)' = None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <channelDensityNonUniform id="nonuniform_na_chans" ionChannel="NaConductance" erev="50mV" ion="na">
        <variableParameter parameter="condDensity" segmentGroup="dendrite_group">
            <inhomogeneousValue inhomogeneousParameter="dendrite_group_x1" value="5e-7 * exp(-p/200)"/>
        </variableParameter>
    </channelDensityNonUniform>
:::
:::
:::
:::
:::

::: {#channeldensitynonuniformnernst .section}
[]{#schema-channeldensitynonuniformnernst}

##### channelDensityNonUniformNernst[\#](#channeldensitynonuniformnernst "Link to this heading"){.headerlink}

extends *[[baseChannelDensity]{.std
.std-ref}](#schema-basechanneldensity){.reference .internal}*

*Specifies a time varying conductance density, which is distributed on a
region of the **cell,** and whose reversal potential is calculated from
the Nernst equation. Hard coded for Ca only!. The conductance density of
the channel is not uniform, but is set using the
[[variableParameter]{.std
.std-ref}](#schema-variableparameter){.reference .internal}. Note, there
is no dynamical description of this in LEMS yet, as this type only makes
sense for multicompartmental cells. A ComponentType for this needs to be
present to enable export of NeuroML 2 multicompartmental cells via
LEMS/jNeuroML to NEURON.*

::: {.sd-tab-set .docutils}
Constants

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------------------+-----------------------------------+-------------+
| **ZERO_CURR_DENS** |                                   | [[currentDe |
| = 0 A_per_m2       |                                   | nsity]{.std |
|                    |                                   | .s          |
|                    |                                   | td-ref}](La |
|                    |                                   | nding.html# |
|                    |                                   | schema-dime |
|                    |                                   | nsions-curr |
|                    |                                   | entdensity) |
|                    |                                   | {.reference |
|                    |                                   | .internal}  |
+--------------------+-----------------------------------+-------------+
:::
:::

Text fields

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| **se   |                                                              |
| gmentG |                                                              |
| roup** |                                                              |
+--------+--------------------------------------------------------------+
| *      | Which ion flows through the channel. Note: ideally this      |
| *ion** | needs to be a property of ionChannel only, but it's here as  |
|        | it makes it easier to select channelPopulations transmitting |
|        | specific ions.                                               |
+--------+--------------------------------------------------------------+
:::
:::

Child list

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    |                                                 | [[v         |
| *var |                                                 | ariablePara |
| iabl |                                                 | meter]{.std |
| ePar |                                                 | .std        |
| amet |                                                 | -ref}](#sch |
| er** |                                                 | ema-variabl |
|      |                                                 | eparameter) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **iD | *(from [[baseChannelDensity]{.std               | [[currentDe |
| ensi | .s                                              | nsity]{.std |
| ty** | td-ref}](#schema-basechanneldensity){.reference | .s          |
|      | .internal})*                                    | td-ref}](La |
|      |                                                 | nding.html# |
|      |                                                 | schema-dime |
|      |                                                 | nsions-curr |
|      |                                                 | entdensity) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseChannelDensity]{.std               | [[vo        |
| *v** | .s                                              | ltage]{.std |
|      | td-ref}](#schema-basechanneldensity){.reference | .std-r      |
|      | .internal})*                                    | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***Structure***

:   CHILD INSTANCE: **ionChannel**

***Derived Variables***

:   **iDensity** = ZERO_CURR_DENS (exposed as **iDensity**)
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="ChannelDensityNonUniformNernst">
      <xs:complexContent>
        <xs:extension base="Base">
          <xs:sequence>
            <xs:element name="variableParameter" type="VariableParameter" minOccurs="0" maxOccurs="unbounded"/>
          </xs:sequence>
          <xs:attribute name="ionChannel" type="NmlId" use="required"/>
          <xs:attribute name="ion" type="NmlId" use="required">
            <xs:annotation>
                </xs:annotation>
          </xs:attribute>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=ChannelDensityNonUniformNernst){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import ChannelDensityNonUniformNernst
    from neuroml.utils import component_factory

    variable = component_factory(
        ChannelDensityNonUniformNernst,
        id: 'a NmlId (required)' = None,
        ion_channel: 'a NmlId (required)' = None,
        ion: 'a NmlId (required)' = None,
        variable_parameters: 'list of VariableParameter(s) (optional)' = None,
    )
:::
:::
:::
:::
:::

::: {#channeldensitynonuniformghk .section}
[]{#schema-channeldensitynonuniformghk}

##### channelDensityNonUniformGHK[\#](#channeldensitynonuniformghk "Link to this heading"){.headerlink}

extends *[[baseChannelDensity]{.std
.std-ref}](#schema-basechanneldensity){.reference .internal}*

*Specifies a time varying conductance density, which is distributed on a
region of the **cell,** and whose current is calculated from the
Goldman-Hodgkin-Katz equation. Hard coded for Ca only!. The conductance
density of the channel is not uniform, but is set using the
[[variableParameter]{.std
.std-ref}](#schema-variableparameter){.reference .internal}. Note, there
is no dynamical description of this in LEMS yet, as this type only makes
sense for multicompartmental cells. A ComponentType for this needs to be
present to enable export of NeuroML 2 multicompartmental cells via
LEMS/jNeuroML to NEURON.*

::: {.sd-tab-set .docutils}
Constants

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------------------+-----------------------------------+-------------+
| **ZERO_CURR_DENS** |                                   | [[currentDe |
| = 0 A_per_m2       |                                   | nsity]{.std |
|                    |                                   | .s          |
|                    |                                   | td-ref}](La |
|                    |                                   | nding.html# |
|                    |                                   | schema-dime |
|                    |                                   | nsions-curr |
|                    |                                   | entdensity) |
|                    |                                   | {.reference |
|                    |                                   | .internal}  |
+--------------------+-----------------------------------+-------------+
:::
:::

Text fields

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| **se   |                                                              |
| gmentG |                                                              |
| roup** |                                                              |
+--------+--------------------------------------------------------------+
| *      | Which ion flows through the channel. Note: ideally this      |
| *ion** | needs to be a property of ionChannel only, but it's here as  |
|        | it makes it easier to select channelPopulations transmitting |
|        | specific ions.                                               |
+--------+--------------------------------------------------------------+
:::
:::

Child list

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    |                                                 | [[v         |
| *var |                                                 | ariablePara |
| iabl |                                                 | meter]{.std |
| ePar |                                                 | .std        |
| amet |                                                 | -ref}](#sch |
| er** |                                                 | ema-variabl |
|      |                                                 | eparameter) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **iD | *(from [[baseChannelDensity]{.std               | [[currentDe |
| ensi | .s                                              | nsity]{.std |
| ty** | td-ref}](#schema-basechanneldensity){.reference | .s          |
|      | .internal})*                                    | td-ref}](La |
|      |                                                 | nding.html# |
|      |                                                 | schema-dime |
|      |                                                 | nsions-curr |
|      |                                                 | entdensity) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseChannelDensity]{.std               | [[vo        |
| *v** | .s                                              | ltage]{.std |
|      | td-ref}](#schema-basechanneldensity){.reference | .std-r      |
|      | .internal})*                                    | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***Structure***

:   CHILD INSTANCE: **ionChannel**

***Derived Variables***

:   **iDensity** = ZERO_CURR_DENS (exposed as **iDensity**)
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="ChannelDensityNonUniformGHK">
      <xs:complexContent>
        <xs:extension base="Base">
          <xs:sequence>
            <xs:element name="variableParameter" type="VariableParameter" minOccurs="0" maxOccurs="unbounded"/>
          </xs:sequence>
          <xs:attribute name="ionChannel" type="NmlId" use="required"/>
          <xs:attribute name="ion" type="NmlId" use="required">
            <xs:annotation>
                </xs:annotation>
          </xs:attribute>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=ChannelDensityNonUniformGHK){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import ChannelDensityNonUniformGHK
    from neuroml.utils import component_factory

    variable = component_factory(
        ChannelDensityNonUniformGHK,
        id: 'a NmlId (required)' = None,
        ion_channel: 'a NmlId (required)' = None,
        ion: 'a NmlId (required)' = None,
        variable_parameters: 'list of VariableParameter(s) (optional)' = None,
    )
:::
:::
:::
:::
:::

::: {#channeldensity .section}
[]{#schema-channeldensity}

##### channelDensity[\#](#channeldensity "Link to this heading"){.headerlink}

extends *[[baseChannelDensityCond]{.std
.std-ref}](#schema-basechanneldensitycond){.reference .internal}*

*Specifies a time varying ohmic conductance density, **gDensity,** which
is distributed on an area of the **cell** ( specified in
[[membraneProperties]{.std
.std-ref}](#schema-membraneproperties){.reference .internal} ) with
fixed reversal potential **erev** producing a current density
**iDensity**.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **c  | *(from [[baseChannelDensityCond]{.std           | [[co        |
| ondD | .std-r                                          | nductanceDe |
| ensi | ef}](#schema-basechanneldensitycond){.reference | nsity]{.std |
| ty** | .internal})*                                    | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-conducta |
|      |                                                 | ncedensity) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **er | The reversal potential of the current produced  | [[vo        |
| ev** |                                                 | ltage]{.std |
|      |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Constants

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------------------+-----------------------------------+-------------+
| **vShift** = 0mV   | Set to a constant 0mV here to     | [[vo        |
|                    | allow ion channels which use      | ltage]{.std |
|                    | \_vShift in their rate variable   | .std-r      |
|                    | expressions to be used with       | ef}](Landin |
|                    | *channelDensity*, not just with   | g.html#sche |
|                    | *channelDensityVShift* (where     | ma-dimensio |
|                    | \_vShift would be explicitly set) | ns-voltage) |
|                    |                                   | {.reference |
|                    |                                   | .internal}  |
+--------------------+-----------------------------------+-------------+
:::
:::

Text fields

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| **se   | Which *segmentGroup* the channelDensity is placed on. If     |
| gmentG | this is missing, it implies it is placed on all \_segment_s  |
| roup** | of the *cell*                                                |
+--------+--------------------------------------------------------------+
| *      | Which ion flows through the channel. Note: ideally this      |
| *ion** | needs to be a property of ionChannel only, but it's here as  |
|        | it makes it easier to select channelPopulations transmitting |
|        | specific ions.                                               |
+--------+--------------------------------------------------------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **gD | *(from [[baseChannelDensityCond]{.std           | [[co        |
| ensi | .std-r                                          | nductanceDe |
| ty** | ef}](#schema-basechanneldensitycond){.reference | nsity]{.std |
|      | .internal})*                                    | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-conducta |
|      |                                                 | ncedensity) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **iD | *(from [[baseChannelDensity]{.std               | [[currentDe |
| ensi | .s                                              | nsity]{.std |
| ty** | td-ref}](#schema-basechanneldensity){.reference | .s          |
|      | .internal})*                                    | td-ref}](La |
|      |                                                 | nding.html# |
|      |                                                 | schema-dime |
|      |                                                 | nsions-curr |
|      |                                                 | entdensity) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseChannelDensity]{.std               | [[vo        |
| *v** | .s                                              | ltage]{.std |
|      | td-ref}](#schema-basechanneldensity){.reference | .std-r      |
|      | .internal})*                                    | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***Structure***

:   CHILD INSTANCE: **ionChannel**

***Derived Variables***

:   **channelf** = ionChannel-\>fopen

:   **gDensity** = condDensity \* channelf (exposed as **gDensity**)

:   **iDensity** = gDensity \* (erev - v) (exposed as **iDensity**)
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="ChannelDensity">
      <xs:complexContent>
        <xs:extension base="Base">
          <xs:sequence>
            <xs:element name="variableParameter" type="VariableParameter" minOccurs="0" maxOccurs="unbounded"/>
          </xs:sequence>
          <xs:attribute name="ionChannel" type="NmlId" use="required"/>
          <xs:attribute name="condDensity" type="Nml2Quantity_conductanceDensity" use="optional"/>
          <xs:attribute name="erev" type="Nml2Quantity_voltage" use="required"/>
          <xs:attribute name="segmentGroup" type="NmlId" use="optional" default="all"/>
          <xs:attribute name="segment" type="NonNegativeInteger" use="optional"/>
          <xs:attribute name="ion" type="NmlId" use="required">
            <xs:annotation>
                </xs:annotation>
          </xs:attribute>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=ChannelDensity){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import ChannelDensity
    from neuroml.utils import component_factory

    variable = component_factory(
        ChannelDensity,
        id: 'a NmlId (required)' = None,
        ion_channel: 'a NmlId (required)' = None,
        cond_density: 'a Nml2Quantity_conductanceDensity (optional)' = None,
        erev: 'a Nml2Quantity_voltage (required)' = None,
        segment_groups: 'a NmlId (optional)' = 'all',
        segments: 'a NonNegativeInteger (optional)' = None,
        ion: 'a NmlId (required)' = None,
        variable_parameters: 'list of VariableParameter(s) (optional)' = None,
        extensiontype_=None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <channelDensity id="pasChans" ionChannel="pas" condDensity="3.0 S_per_m2" erev="-70mV" ion="non_specific"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <channelDensity id="naChansSoma" ionChannel="NaConductance" segmentGroup="soma_group" condDensity="120.0 mS_per_cm2" erev="50mV" ion="na"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <channelDensity id="naChans" ionChannel="HH_Na" segmentGroup="soma_group" condDensity="120.0 mS_per_cm2" ion="na" erev="50mV"/>
:::
:::
:::
:::
:::

::: {#channeldensityvshift .section}
[]{#schema-channeldensityvshift}

##### channelDensityVShift[\#](#channeldensityvshift "Link to this heading"){.headerlink}

extends [[channelDensity]{.std
.std-ref}](#schema-channeldensity){.reference .internal}

*Same as [[channelDensity]{.std
.std-ref}](#schema-channeldensity){.reference .internal}, but with a
**vShift** parameter to change voltage activation of gates. The exact
usage of **vShift** in expressions for rates is determined by the
individual gates.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **c  | *(from [[baseChannelDensityCond]{.std           | [[co        |
| ondD | .std-r                                          | nductanceDe |
| ensi | ef}](#schema-basechanneldensitycond){.reference | nsity]{.std |
| ty** | .internal})*                                    | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-conducta |
|      |                                                 | ncedensity) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **er | The reversal potential of the current produced  | [[vo        |
| ev** | *(from [[channelDensity]{.std                   | ltage]{.std |
|      | .std-ref}](#schema-channeldensity){.reference   | .std-r      |
|      | .internal})*                                    | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   |                                                 | [[vo        |
| vShi |                                                 | ltage]{.std |
| ft** |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Text fields

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| **se   | Which *segmentGroup* the channelDensity is placed on. If     |
| gmentG | this is missing, it implies it is placed on all \_segment_s  |
| roup** | of the *cell*                                                |
+--------+--------------------------------------------------------------+
| *      | Which ion flows through the channel. Note: ideally this      |
| *ion** | needs to be a property of ionChannel only, but it's here as  |
|        | it makes it easier to select channelPopulations transmitting |
|        | specific ions.                                               |
+--------+--------------------------------------------------------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **gD | *(from [[baseChannelDensityCond]{.std           | [[co        |
| ensi | .std-r                                          | nductanceDe |
| ty** | ef}](#schema-basechanneldensitycond){.reference | nsity]{.std |
|      | .internal})*                                    | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-conducta |
|      |                                                 | ncedensity) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **iD | *(from [[baseChannelDensity]{.std               | [[currentDe |
| ensi | .s                                              | nsity]{.std |
| ty** | td-ref}](#schema-basechanneldensity){.reference | .s          |
|      | .internal})*                                    | td-ref}](La |
|      |                                                 | nding.html# |
|      |                                                 | schema-dime |
|      |                                                 | nsions-curr |
|      |                                                 | entdensity) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseChannelDensity]{.std               | [[vo        |
| *v** | .s                                              | ltage]{.std |
|      | td-ref}](#schema-basechanneldensity){.reference | .std-r      |
|      | .internal})*                                    | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="ChannelDensityVShift">
      <xs:complexContent>
        <xs:extension base="ChannelDensity">
          <xs:attribute name="vShift" type="Nml2Quantity_voltage" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=ChannelDensityVShift){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import ChannelDensityVShift
    from neuroml.utils import component_factory

    variable = component_factory(
        ChannelDensityVShift,
        id: 'a NmlId (required)' = None,
        ion_channel: 'a NmlId (required)' = None,
        cond_density: 'a Nml2Quantity_conductanceDensity (optional)' = None,
        erev: 'a Nml2Quantity_voltage (required)' = None,
        segment_groups: 'a NmlId (optional)' = 'all',
        segments: 'a NonNegativeInteger (optional)' = None,
        ion: 'a NmlId (required)' = None,
        variable_parameters: 'list of VariableParameter(s) (optional)' = None,
        v_shift: 'a Nml2Quantity_voltage (required)' = None,
    )
:::
:::
:::
:::
:::

::: {#channeldensitynernst .section}
[]{#schema-channeldensitynernst}

##### channelDensityNernst[\#](#channeldensitynernst "Link to this heading"){.headerlink}

extends *[[baseChannelDensityCond]{.std
.std-ref}](#schema-basechanneldensitycond){.reference .internal}*

*Specifies a time varying conductance density, **gDensity,** which is
distributed on an area of the **cell,** producing a current density
**iDensity** and whose reversal potential is calculated from the Nernst
equation. Hard coded for Ca only! See
[OpenSourceBrain/ghk-nernst](https://github.com/OpenSourceBrain/ghk-nernst){.github
.reference .external}.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **c  | *(from [[baseChannelDensityCond]{.std           | [[co        |
| ondD | .std-r                                          | nductanceDe |
| ensi | ef}](#schema-basechanneldensitycond){.reference | nsity]{.std |
| ty** | .internal})*                                    | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-conducta |
|      |                                                 | ncedensity) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Constants

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------------------+-----------------------------------+-------------+
| **R** = 8.3144621  |                                   | [[idea      |
| J_per_K_per_mol    |                                   | lGasConstan |
|                    |                                   | tDims]{.std |
|                    |                                   | .std-ref    |
|                    |                                   | }](Landing. |
|                    |                                   | html#schema |
|                    |                                   | -dimensions |
|                    |                                   | -idealgasco |
|                    |                                   | nstantdims) |
|                    |                                   | {.reference |
|                    |                                   | .internal}  |
+--------------------+-----------------------------------+-------------+
| **zCa** = 2        |                                   | Di          |
|                    |                                   | mensionless |
+--------------------+-----------------------------------+-------------+
| **F** = 96485.3    |                                   | [           |
| C_per_mol          |                                   | [charge_per |
|                    |                                   | _mole]{.std |
|                    |                                   | .st         |
|                    |                                   | d-ref}](Lan |
|                    |                                   | ding.html#s |
|                    |                                   | chema-dimen |
|                    |                                   | sions-charg |
|                    |                                   | e-per-mole) |
|                    |                                   | {.reference |
|                    |                                   | .internal}  |
+--------------------+-----------------------------------+-------------+
:::
:::

Text fields

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| **se   | Which *segmentGroup* the channelDensityNernst is placed on.  |
| gmentG | If this is missing, it implies it is placed on all           |
| roup** | \_segment_s of the *cell*                                    |
+--------+--------------------------------------------------------------+
| *      | Which ion flows through the channel. Note: ideally this      |
| *ion** | needs to be a property of ionChannel only, but it's here as  |
|        | it makes it easier to select channelPopulations transmitting |
|        | specific ions.                                               |
+--------+--------------------------------------------------------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **er | The reversal potential of the current produced, | [[vo        |
| ev** | calculated from caConcExt and caConc            | ltage]{.std |
|      |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **gD | *(from [[baseChannelDensityCond]{.std           | [[co        |
| ensi | .std-r                                          | nductanceDe |
| ty** | ef}](#schema-basechanneldensitycond){.reference | nsity]{.std |
|      | .internal})*                                    | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-conducta |
|      |                                                 | ncedensity) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **iD | *(from [[baseChannelDensity]{.std               | [[currentDe |
| ensi | .s                                              | nsity]{.std |
| ty** | td-ref}](#schema-basechanneldensity){.reference | .s          |
|      | .internal})*                                    | td-ref}](La |
|      |                                                 | nding.html# |
|      |                                                 | schema-dime |
|      |                                                 | nsions-curr |
|      |                                                 | entdensity) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   |                                                 | [[concentr  |
| caCo |                                                 | ation]{.std |
| nc** |                                                 | .           |
|      |                                                 | std-ref}](L |
|      |                                                 | anding.html |
|      |                                                 | #schema-dim |
|      |                                                 | ensions-con |
|      |                                                 | centration) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    |                                                 | [[concentr  |
| *caC |                                                 | ation]{.std |
| oncE |                                                 | .           |
| xt** |                                                 | std-ref}](L |
|      |                                                 | anding.html |
|      |                                                 | #schema-dim |
|      |                                                 | ensions-con |
|      |                                                 | centration) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **t  |                                                 | [[temper    |
| empe |                                                 | ature]{.std |
| ratu |                                                 | .std-ref}]  |
| re** |                                                 | (Landing.ht |
|      |                                                 | ml#schema-d |
|      |                                                 | imensions-t |
|      |                                                 | emperature) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseChannelDensity]{.std               | [[vo        |
| *v** | .s                                              | ltage]{.std |
|      | td-ref}](#schema-basechanneldensity){.reference | .std-r      |
|      | .internal})*                                    | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***Structure***

:   CHILD INSTANCE: **ionChannel**

***Derived Variables***

:   **channelf** = ionChannel-\>fopen

***Conditional Derived Variables***

:   IF caConcExt \> 0 THEN

:     **gDensity** = condDensity \* channelf  (exposed as **gDensity**)

:   IF caConcExt \<= 0 THEN

:     **gDensity** = 0  (exposed as **gDensity**)

:   IF caConcExt \> 0 THEN

:     **erev** = (R \* temperature / (zCa \* F)) \* log(caConcExt /
    caConc)  (exposed as **erev**)

:   IF caConcExt \<= 0 THEN

:     **erev** = 0  (exposed as **erev**)

:   IF caConcExt \> 0 THEN

:     **iDensity** = gDensity \* (erev - v)  (exposed as **iDensity**)

:   IF caConcExt \<= 0 THEN

:     **iDensity** = 0  (exposed as **iDensity**)
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="ChannelDensityNernst">
      <xs:complexContent>
        <xs:extension base="Base">
          <xs:sequence>
            <xs:element name="variableParameter" type="VariableParameter" minOccurs="0" maxOccurs="unbounded"/>
          </xs:sequence>
          <xs:attribute name="ionChannel" type="NmlId" use="required"/>
          <xs:attribute name="condDensity" type="Nml2Quantity_conductanceDensity" use="optional"/>
          <xs:attribute name="segmentGroup" type="NmlId" use="optional" default="all"/>
          <xs:attribute name="segment" type="NmlId" use="optional"/>
          <xs:attribute name="ion" type="NmlId" use="required">
            <xs:annotation>
                </xs:annotation>
          </xs:attribute>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=ChannelDensityNernst){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import ChannelDensityNernst
    from neuroml.utils import component_factory

    variable = component_factory(
        ChannelDensityNernst,
        id: 'a NmlId (required)' = None,
        ion_channel: 'a NmlId (required)' = None,
        cond_density: 'a Nml2Quantity_conductanceDensity (optional)' = None,
        segment_groups: 'a NmlId (optional)' = 'all',
        segments: 'a NonNegativeInteger (optional)' = None,
        ion: 'a NmlId (required)' = None,
        variable_parameters: 'list of VariableParameter(s) (optional)' = None,
        extensiontype_=None,
    )
:::
:::
:::
:::
:::

::: {#channeldensitynernstca2 .section}
[]{#schema-channeldensitynernstca2}

##### channelDensityNernstCa2[\#](#channeldensitynernstca2 "Link to this heading"){.headerlink}

extends *[[baseChannelDensityCond]{.std
.std-ref}](#schema-basechanneldensitycond){.reference .internal}*

*This component is similar to the original component type
[[channelDensityNernst]{.std
.std-ref}](#schema-channeldensitynernst){.reference .internal} but it is
changed in order to have a reversal potential that depends on a second
independent Ca++ pool ( ca2 ). See
[OpenSourceBrain/ghk-nernst](https://github.com/OpenSourceBrain/ghk-nernst){.github
.reference .external}.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **c  | *(from [[baseChannelDensityCond]{.std           | [[co        |
| ondD | .std-r                                          | nductanceDe |
| ensi | ef}](#schema-basechanneldensitycond){.reference | nsity]{.std |
| ty** | .internal})*                                    | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-conducta |
|      |                                                 | ncedensity) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Constants

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------------------+-----------------------------------+-------------+
| **R** = 8.3144621  |                                   | [[idea      |
| J_per_K_per_mol    |                                   | lGasConstan |
|                    |                                   | tDims]{.std |
|                    |                                   | .std-ref    |
|                    |                                   | }](Landing. |
|                    |                                   | html#schema |
|                    |                                   | -dimensions |
|                    |                                   | -idealgasco |
|                    |                                   | nstantdims) |
|                    |                                   | {.reference |
|                    |                                   | .internal}  |
+--------------------+-----------------------------------+-------------+
| **zCa** = 2        |                                   | Di          |
|                    |                                   | mensionless |
+--------------------+-----------------------------------+-------------+
| **F** = 96485.3    |                                   | [           |
| C_per_mol          |                                   | [charge_per |
|                    |                                   | _mole]{.std |
|                    |                                   | .st         |
|                    |                                   | d-ref}](Lan |
|                    |                                   | ding.html#s |
|                    |                                   | chema-dimen |
|                    |                                   | sions-charg |
|                    |                                   | e-per-mole) |
|                    |                                   | {.reference |
|                    |                                   | .internal}  |
+--------------------+-----------------------------------+-------------+
:::
:::

Text fields

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| **se   | Which *segmentGroup* the channelDensityNernstCa2 is placed   |
| gmentG | on. If this is missing, it implies it is placed on all       |
| roup** | \_segment_s of the *cell*                                    |
+--------+--------------------------------------------------------------+
| *      | Which ion flows through the channel. Note: ideally this      |
| *ion** | needs to be a property of ionChannel only, but it's here as  |
|        | it makes it easier to select channelPopulations transmitting |
|        | specific ions.                                               |
+--------+--------------------------------------------------------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **er | The reversal potential of the current produced  | [[vo        |
| ev** |                                                 | ltage]{.std |
|      |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **gD | *(from [[baseChannelDensityCond]{.std           | [[co        |
| ensi | .std-r                                          | nductanceDe |
| ty** | ef}](#schema-basechanneldensitycond){.reference | nsity]{.std |
|      | .internal})*                                    | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-conducta |
|      |                                                 | ncedensity) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **iD | *(from [[baseChannelDensity]{.std               | [[currentDe |
| ensi | .s                                              | nsity]{.std |
| ty** | td-ref}](#schema-basechanneldensity){.reference | .s          |
|      | .internal})*                                    | td-ref}](La |
|      |                                                 | nding.html# |
|      |                                                 | schema-dime |
|      |                                                 | nsions-curr |
|      |                                                 | entdensity) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **c  |                                                 | [[concentr  |
| aCon |                                                 | ation]{.std |
| c2** |                                                 | .           |
|      |                                                 | std-ref}](L |
|      |                                                 | anding.html |
|      |                                                 | #schema-dim |
|      |                                                 | ensions-con |
|      |                                                 | centration) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   |                                                 | [[concentr  |
| caCo |                                                 | ation]{.std |
| ncEx |                                                 | .           |
| t2** |                                                 | std-ref}](L |
|      |                                                 | anding.html |
|      |                                                 | #schema-dim |
|      |                                                 | ensions-con |
|      |                                                 | centration) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **t  |                                                 | [[temper    |
| empe |                                                 | ature]{.std |
| ratu |                                                 | .std-ref}]  |
| re** |                                                 | (Landing.ht |
|      |                                                 | ml#schema-d |
|      |                                                 | imensions-t |
|      |                                                 | emperature) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseChannelDensity]{.std               | [[vo        |
| *v** | .s                                              | ltage]{.std |
|      | td-ref}](#schema-basechanneldensity){.reference | .std-r      |
|      | .internal})*                                    | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***Structure***

:   CHILD INSTANCE: **ionChannel**

***Derived Variables***

:   **channelf** = ionChannel-\>fopen

***Conditional Derived Variables***

:   IF caConcExt2 \> 0 THEN

:     **gDensity** = condDensity \* channelf  (exposed as **gDensity**)

:   IF caConcExt2 \<= 0 THEN

:     **gDensity** = 0  (exposed as **gDensity**)

:   IF caConcExt2 \> 0 THEN

:     **erev** = (R \* temperature / (zCa \* F)) \* log(caConcExt2 /
    caConc2)  (exposed as **erev**)

:   IF caConcExt2 \<= 0 THEN

:     **erev** = 0  (exposed as **erev**)

:   IF caConcExt2 \> 0 THEN

:     **iDensity** = gDensity \* (erev - v)  (exposed as **iDensity**)

:   IF caConcExt2 \<= 0 THEN

:     **iDensity** = 0  (exposed as **iDensity**)
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="ChannelDensityNernstCa2">
      <xs:complexContent>
        <xs:extension base="ChannelDensityNernst">
          </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=ChannelDensityNernstCa2){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import ChannelDensityNernstCa2
    from neuroml.utils import component_factory

    variable = component_factory(
        ChannelDensityNernstCa2,
        id: 'a NmlId (required)' = None,
        ion_channel: 'a NmlId (required)' = None,
        cond_density: 'a Nml2Quantity_conductanceDensity (optional)' = None,
        segment_groups: 'a NmlId (optional)' = 'all',
        segments: 'a NonNegativeInteger (optional)' = None,
        ion: 'a NmlId (required)' = None,
        variable_parameters: 'list of VariableParameter(s) (optional)' = None,
    )
:::
:::
:::
:::
:::

::: {#channeldensityghk .section}
[]{#schema-channeldensityghk}

##### channelDensityGHK[\#](#channeldensityghk "Link to this heading"){.headerlink}

extends *[[baseChannelDensity]{.std
.std-ref}](#schema-basechanneldensity){.reference .internal}*

*Specifies a time varying conductance density, **gDensity,** which is
distributed on an area of the cell, producing a current density
**iDensity** and whose reversal potential is calculated from the Goldman
Hodgkin Katz equation. Hard coded for Ca only! See
[OpenSourceBrain/ghk-nernst](https://github.com/OpenSourceBrain/ghk-nernst){.github
.reference .external}.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **pe |                                                 | [[permeab   |
| rmea |                                                 | ility]{.std |
| bili |                                                 | .std-ref}]( |
| ty** |                                                 | Landing.htm |
|      |                                                 | l#schema-di |
|      |                                                 | mensions-pe |
|      |                                                 | rmeability) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Constants

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------------------+-----------------------------------+-------------+
| **R** = 8.3144621  |                                   | [[idea      |
| J_per_K_per_mol    |                                   | lGasConstan |
|                    |                                   | tDims]{.std |
|                    |                                   | .std-ref    |
|                    |                                   | }](Landing. |
|                    |                                   | html#schema |
|                    |                                   | -dimensions |
|                    |                                   | -idealgasco |
|                    |                                   | nstantdims) |
|                    |                                   | {.reference |
|                    |                                   | .internal}  |
+--------------------+-----------------------------------+-------------+
| **zCa** = 2        |                                   | Di          |
|                    |                                   | mensionless |
+--------------------+-----------------------------------+-------------+
| **F** = 96485.3    |                                   | [           |
| C_per_mol          |                                   | [charge_per |
|                    |                                   | _mole]{.std |
|                    |                                   | .st         |
|                    |                                   | d-ref}](Lan |
|                    |                                   | ding.html#s |
|                    |                                   | chema-dimen |
|                    |                                   | sions-charg |
|                    |                                   | e-per-mole) |
|                    |                                   | {.reference |
|                    |                                   | .internal}  |
+--------------------+-----------------------------------+-------------+
:::
:::

Text fields

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| **se   | Which *segmentGroup* the channelDensityGHK is placed on. If  |
| gmentG | this is missing, it implies it is placed on all \_segment_s  |
| roup** | of the *cell*                                                |
+--------+--------------------------------------------------------------+
| *      | Which ion flows through the channel. Note: ideally this      |
| *ion** | needs to be a property of ionChannel only, but it's here as  |
|        | it makes it easier to select channelPopulations transmitting |
|        | specific ions.                                               |
+--------+--------------------------------------------------------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **iD | *(from [[baseChannelDensity]{.std               | [[currentDe |
| ensi | .s                                              | nsity]{.std |
| ty** | td-ref}](#schema-basechanneldensity){.reference | .s          |
|      | .internal})*                                    | td-ref}](La |
|      |                                                 | nding.html# |
|      |                                                 | schema-dime |
|      |                                                 | nsions-curr |
|      |                                                 | entdensity) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   |                                                 | [[concentr  |
| caCo |                                                 | ation]{.std |
| nc** |                                                 | .           |
|      |                                                 | std-ref}](L |
|      |                                                 | anding.html |
|      |                                                 | #schema-dim |
|      |                                                 | ensions-con |
|      |                                                 | centration) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    |                                                 | [[concentr  |
| *caC |                                                 | ation]{.std |
| oncE |                                                 | .           |
| xt** |                                                 | std-ref}](L |
|      |                                                 | anding.html |
|      |                                                 | #schema-dim |
|      |                                                 | ensions-con |
|      |                                                 | centration) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **t  |                                                 | [[temper    |
| empe |                                                 | ature]{.std |
| ratu |                                                 | .std-ref}]  |
| re** |                                                 | (Landing.ht |
|      |                                                 | ml#schema-d |
|      |                                                 | imensions-t |
|      |                                                 | emperature) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseChannelDensity]{.std               | [[vo        |
| *v** | .s                                              | ltage]{.std |
|      | td-ref}](#schema-basechanneldensity){.reference | .std-r      |
|      | .internal})*                                    | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***Structure***

:   CHILD INSTANCE: **ionChannel**

***Derived Variables***

:   **K** = (zCa \* F) / (R \* temperature)

:   **expKv** = exp(-1 \* K \* v)

:   **channelf** = ionChannel-\>fopen

***Conditional Derived Variables***

:   IF caConcExt \> 0 THEN

:     **iDensity** = -1 \* channelf \* permeability \* zCa \* F \* K \*
    v \* ( caConc - (caConcExt \* expKv) ) / (1 - expKv)  (exposed as
    **iDensity**)

:   IF caConcExt \<= 0 THEN

:     **iDensity** = 0  (exposed as **iDensity**)
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="ChannelDensityGHK">
      <xs:complexContent>
        <xs:extension base="Base">
          <xs:attribute name="ionChannel" type="NmlId" use="required"/>
          <xs:attribute name="permeability" type="Nml2Quantity_permeability" use="required"/>
          <xs:attribute name="segmentGroup" type="NmlId" use="optional" default="all"/>
          <xs:attribute name="segment" type="NmlId" use="optional"/>
          <xs:attribute name="ion" type="NmlId" use="required">
            <xs:annotation>
                </xs:annotation>
          </xs:attribute>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=ChannelDensityGHK){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import ChannelDensityGHK
    from neuroml.utils import component_factory

    variable = component_factory(
        ChannelDensityGHK,
        id: 'a NmlId (required)' = None,
        ion_channel: 'a NmlId (required)' = None,
        permeability: 'a Nml2Quantity_permeability (required)' = None,
        segment_groups: 'a NmlId (optional)' = 'all',
        segments: 'a NonNegativeInteger (optional)' = None,
        ion: 'a NmlId (required)' = None,
    )
:::
:::
:::
:::
:::

::: {#channeldensityghk2 .section}
[]{#schema-channeldensityghk2}

##### channelDensityGHK2[\#](#channeldensityghk2 "Link to this heading"){.headerlink}

extends *[[baseChannelDensityCond]{.std
.std-ref}](#schema-basechanneldensitycond){.reference .internal}*

*Time varying conductance density, **gDensity,** which is distributed on
an area of the cell, producing a current density **iDensity.** Modified
version of Jaffe et al. 1994 ( used also in Lawrence et al. 2006 ). See
[OpenSourceBrain/ghk-nernst](https://github.com/OpenSourceBrain/ghk-nernst){.github
.reference .external}.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **c  | *(from [[baseChannelDensityCond]{.std           | [[co        |
| ondD | .std-r                                          | nductanceDe |
| ensi | ef}](#schema-basechanneldensitycond){.reference | nsity]{.std |
| ty** | .internal})*                                    | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-conducta |
|      |                                                 | ncedensity) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Constants

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------------------+-----------------------------------+-------------+
| **VOLT_SCALE** = 1 |                                   | [[vo        |
| mV                 |                                   | ltage]{.std |
|                    |                                   | .std-r      |
|                    |                                   | ef}](Landin |
|                    |                                   | g.html#sche |
|                    |                                   | ma-dimensio |
|                    |                                   | ns-voltage) |
|                    |                                   | {.reference |
|                    |                                   | .internal}  |
+--------------------+-----------------------------------+-------------+
| **CONC_SCALE** = 1 |                                   | [[concentr  |
| mM                 |                                   | ation]{.std |
|                    |                                   | .           |
|                    |                                   | std-ref}](L |
|                    |                                   | anding.html |
|                    |                                   | #schema-dim |
|                    |                                   | ensions-con |
|                    |                                   | centration) |
|                    |                                   | {.reference |
|                    |                                   | .internal}  |
+--------------------+-----------------------------------+-------------+
| **TEMP_SCALE** = 1 |                                   | [[temper    |
| K                  |                                   | ature]{.std |
|                    |                                   | .std-ref}]  |
|                    |                                   | (Landing.ht |
|                    |                                   | ml#schema-d |
|                    |                                   | imensions-t |
|                    |                                   | emperature) |
|                    |                                   | {.reference |
|                    |                                   | .internal}  |
+--------------------+-----------------------------------+-------------+
:::
:::

Text fields

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| **se   | Which *segmentGroup* the channelDensityGHK2 is placed on. If |
| gmentG | this is missing, it implies it is placed on all \_segment_s  |
| roup** | of the *cell*                                                |
+--------+--------------------------------------------------------------+
| *      | Which ion flows through the channel. Note: ideally this      |
| *ion** | needs to be a property of ionChannel only, but it's here as  |
|        | it makes it easier to select channelPopulations transmitting |
|        | specific ions.                                               |
+--------+--------------------------------------------------------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **gD | *(from [[baseChannelDensityCond]{.std           | [[co        |
| ensi | .std-r                                          | nductanceDe |
| ty** | ef}](#schema-basechanneldensitycond){.reference | nsity]{.std |
|      | .internal})*                                    | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-conducta |
|      |                                                 | ncedensity) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **iD | *(from [[baseChannelDensity]{.std               | [[currentDe |
| ensi | .s                                              | nsity]{.std |
| ty** | td-ref}](#schema-basechanneldensity){.reference | .s          |
|      | .internal})*                                    | td-ref}](La |
|      |                                                 | nding.html# |
|      |                                                 | schema-dime |
|      |                                                 | nsions-curr |
|      |                                                 | entdensity) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   |                                                 | [[concentr  |
| caCo |                                                 | ation]{.std |
| nc** |                                                 | .           |
|      |                                                 | std-ref}](L |
|      |                                                 | anding.html |
|      |                                                 | #schema-dim |
|      |                                                 | ensions-con |
|      |                                                 | centration) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    |                                                 | [[concentr  |
| *caC |                                                 | ation]{.std |
| oncE |                                                 | .           |
| xt** |                                                 | std-ref}](L |
|      |                                                 | anding.html |
|      |                                                 | #schema-dim |
|      |                                                 | ensions-con |
|      |                                                 | centration) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **t  |                                                 | [[temper    |
| empe |                                                 | ature]{.std |
| ratu |                                                 | .std-ref}]  |
| re** |                                                 | (Landing.ht |
|      |                                                 | ml#schema-d |
|      |                                                 | imensions-t |
|      |                                                 | emperature) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseChannelDensity]{.std               | [[vo        |
| *v** | .s                                              | ltage]{.std |
|      | td-ref}](#schema-basechanneldensity){.reference | .std-r      |
|      | .internal})*                                    | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***Structure***

:   CHILD INSTANCE: **ionChannel**

***Derived Variables***

:   **V** = v / VOLT_SCALE

:   **ca_conc_i** = caConc / CONC_SCALE

:   **ca_conc_ext** = caConcExt / CONC_SCALE

:   **T** = temperature / TEMP_SCALE

:   **channelf** = ionChannel-\>fopen

:   **gDensity** = condDensity \* channelf (exposed as **gDensity**)

:   **tmp** = (25 \* T) / (293.15 \* 2)

***Conditional Derived Variables***

:   IF V/tmp = 0. THEN

:     **pOpen** = tmp \* 1e-3 \* (1 - ((ca_conc_i/ca_conc_ext) \*
    exp(V/tmp))) \* (1 - (V/tmp)/2)

:   IF V/tmp != 0. THEN

:     **pOpen** = tmp \* 1e-3 \* (1 - ((ca_conc_i/ca_conc_ext) \*
    exp(V/tmp))) \* ((V/tmp) / (exp(V/tmp) - 1))

:   IF ca_conc_ext \> 0 THEN

:     **iDensity** = gDensity \* pOpen  (exposed as **iDensity**)

:   IF ca_conc_ext \<= 0 THEN

:     **iDensity** = 0  (exposed as **iDensity**)
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="ChannelDensityGHK2">
      <xs:complexContent>
        <xs:extension base="Base">
          <xs:attribute name="ionChannel" type="NmlId" use="required"/>
          <xs:attribute name="condDensity" type="Nml2Quantity_conductanceDensity" use="optional"/>
          <xs:attribute name="segmentGroup" type="NmlId" use="optional" default="all"/>
          <xs:attribute name="segment" type="NmlId" use="optional"/>
          <xs:attribute name="ion" type="NmlId" use="required">
            <xs:annotation>
                </xs:annotation>
          </xs:attribute>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=ChannelDensityGHK2){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import ChannelDensityGHK2
    from neuroml.utils import component_factory

    variable = component_factory(
        ChannelDensityGHK2,
        id: 'a NmlId (required)' = None,
        ion_channel: 'a NmlId (required)' = None,
        cond_density: 'a Nml2Quantity_conductanceDensity (optional)' = None,
        segment_groups: 'a NmlId (optional)' = 'all',
        segments: 'a NonNegativeInteger (optional)' = None,
        ion: 'a NmlId (required)' = None,
    )
:::
:::
:::
:::
:::

::: {#pointcellcondbased .section}
[]{#schema-pointcellcondbased}

##### pointCellCondBased[\#](#pointcellcondbased "Link to this heading"){.headerlink}

extends *[[baseCellMembPotCap]{.std
.std-ref}](#schema-basecellmembpotcap){.reference .internal}*

*Simple model of a conductance based cell, with no separate
[[morphology]{.std .std-ref}](#schema-morphology){.reference .internal}
element, just an absolute capacitance **C,** and a set of channel
**populations.** Note: use of [[cell]{.std
.std-ref}](#schema-cell){.reference .internal} is generally preferable (
and more widely supported ), even for a single compartment cell.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Total capacitance of the cell membrane *(from   | [[capaci    |
| *C** | [[baseCellMembPotCap]{.std                      | tance]{.std |
|      | .s                                              | .std-ref}]  |
|      | td-ref}](#schema-basecellmembpotcap){.reference | (Landing.ht |
|      | .internal})*                                    | ml#schema-d |
|      |                                                 | imensions-c |
|      |                                                 | apacitance) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   | The voltage threshold above which the cell is   | [[vo        |
| thre | considered to be \_spiking                      | ltage]{.std |
| sh** |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   | The initial membrane potential of the cell      | [[vo        |
| v0** |                                                 | ltage]{.std |
|      |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Children list

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **p  |                                                 | [[baseC     |
| opul |                                                 | hannelPopul |
| atio |                                                 | ation]{.std |
| ns** |                                                 | .std-ref    |
|      |                                                 | }](#schema- |
|      |                                                 | basechannel |
|      |                                                 | population) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Total current crossing the cell membrane *(from | [[cu        |
| *iMe | [[baseCellMembPotCap]{.std                      | rrent]{.std |
| mb** | .s                                              | .std-r      |
|      | td-ref}](#schema-basecellmembpotcap){.reference | ef}](Landin |
|      | .internal})*                                    | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **iS | Total current due to synaptic inputs *(from     | [[cu        |
| yn** | [[baseCellMembPotCap]{.std                      | rrent]{.std |
|      | .s                                              | .std-r      |
|      | td-ref}](#schema-basecellmembpotcap){.reference | ef}](Landin |
|      | .internal})*                                    | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | Membrane potential *(from                       | [[vo        |
| *v** | [[baseCellMembPot]{.std                         | ltage]{.std |
|      | .std-ref}](#schema-basecellmembpot){.reference  | .std-r      |
|      | .internal})*                                    | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Spike event *(from [[baseSpikingCell]{.std      | Direction:  |
| *spi | .std-ref}](#schema-basespikingcell){.reference  | out         |
| ke** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Attachments

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **sy |                                                 | [[          |
| naps |                                                 | basePointCu |
| es** |                                                 | rrent]{.std |
|      |                                                 | .std        |
|      |                                                 | -ref}](Land |
|      |                                                 | ing.html#sc |
|      |                                                 | hema-basepo |
|      |                                                 | intcurrent) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***State Variables***

:   **v**: [[voltage]{.std
    .std-ref}](Landing.html#schema-dimensions-voltage){.reference
    .internal}  (exposed as **v**)

:   **spiking**: Dimensionless

***On Start***

:   **v** = v0

:   **spiking** = 0

***On Conditions***

:   IF v \> thresh AND spiking \< 0.5 THEN

:      **spiking** = 1

:      EVENT OUT on port: **spike**

:   IF v \< thresh THEN

:      **spiking** = 0

***Derived Variables***

:   **iChannels** = populations\[\*\]-\>i(reduce method: add)

:   **iSyn** = synapses\[\*\]-\>i(reduce method: add) (exposed as
    **iSyn**)

:   **iMemb** = iChannels + iSyn (exposed as **iMemb**)

***Time Derivatives***

:   d **v** /dt = iMemb / C
:::
:::
:::

::: {#pointcellcondbasedca .section}
[]{#schema-pointcellcondbasedca}

##### pointCellCondBasedCa[\#](#pointcellcondbasedca "Link to this heading"){.headerlink}

extends *[[baseCellMembPotCap]{.std
.std-ref}](#schema-basecellmembpotcap){.reference .internal}*

*TEMPORARY: Point cell with conductances and Ca concentration info. Not
yet fully tested!!!*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Total capacitance of the cell membrane *(from   | [[capaci    |
| *C** | [[baseCellMembPotCap]{.std                      | tance]{.std |
|      | .s                                              | .std-ref}]  |
|      | td-ref}](#schema-basecellmembpotcap){.reference | (Landing.ht |
|      | .internal})*                                    | ml#schema-d |
|      |                                                 | imensions-c |
|      |                                                 | apacitance) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   | The voltage threshold above which the cell is   | [[vo        |
| thre | considered to be \_spiking                      | ltage]{.std |
| sh** |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   | The initial membrane potential of the cell      | [[vo        |
| v0** |                                                 | ltage]{.std |
|      |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Children list

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **p  |                                                 | [[baseC     |
| opul |                                                 | hannelPopul |
| atio |                                                 | ation]{.std |
| ns** |                                                 | .std-ref    |
|      |                                                 | }](#schema- |
|      |                                                 | basechannel |
|      |                                                 | population) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **c  |                                                 | [[co        |
| once |                                                 | ncentration |
| ntra |                                                 | Model]{.std |
| tion |                                                 | .std-       |
| Mode |                                                 | ref}](#sche |
| ls** |                                                 | ma-concentr |
|      |                                                 | ationmodel) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   |                                                 | [[concentr  |
| caCo |                                                 | ation]{.std |
| nc** |                                                 | .           |
|      |                                                 | std-ref}](L |
|      |                                                 | anding.html |
|      |                                                 | #schema-dim |
|      |                                                 | ensions-con |
|      |                                                 | centration) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **i  |                                                 | [[cu        |
| Ca** |                                                 | rrent]{.std |
|      |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | Total current crossing the cell membrane *(from | [[cu        |
| *iMe | [[baseCellMembPotCap]{.std                      | rrent]{.std |
| mb** | .s                                              | .std-r      |
|      | td-ref}](#schema-basecellmembpotcap){.reference | ef}](Landin |
|      | .internal})*                                    | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **iS | Total current due to synaptic inputs *(from     | [[cu        |
| yn** | [[baseCellMembPotCap]{.std                      | rrent]{.std |
|      | .s                                              | .std-r      |
|      | td-ref}](#schema-basecellmembpotcap){.reference | ef}](Landin |
|      | .internal})*                                    | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | Membrane potential *(from                       | [[vo        |
| *v** | [[baseCellMembPot]{.std                         | ltage]{.std |
|      | .std-ref}](#schema-basecellmembpot){.reference  | .std-r      |
|      | .internal})*                                    | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Spike event *(from [[baseSpikingCell]{.std      | Direction:  |
| *spi | .std-ref}](#schema-basespikingcell){.reference  | out         |
| ke** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Attachments

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **sy |                                                 | [[          |
| naps |                                                 | basePointCu |
| es** |                                                 | rrent]{.std |
|      |                                                 | .std        |
|      |                                                 | -ref}](Land |
|      |                                                 | ing.html#sc |
|      |                                                 | hema-basepo |
|      |                                                 | intcurrent) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***State Variables***

:   **v**: [[voltage]{.std
    .std-ref}](Landing.html#schema-dimensions-voltage){.reference
    .internal}  (exposed as **v**)

:   **spiking**: Dimensionless

***On Start***

:   **v** = v0

:   **spiking** = 0

***On Conditions***

:   IF v \> thresh AND spiking \< 0.5 THEN

:      **spiking** = 1

:      EVENT OUT on port: **spike**

:   IF v \< thresh THEN

:      **spiking** = 0

***Derived Variables***

:   **iChannels** = populations\[\*\]-\>i(reduce method: add)

:   **iCa** = populations\[ion='ca'\]-\>i(reduce method: add) (exposed
    as **iCa**)

:   **caConc**
    = concentrationModels\[species='ca'\]-\>concentration(reduce method:
    add) (exposed as **caConc**)

:   **iSyn** = synapses\[\*\]-\>i(reduce method: add) (exposed as
    **iSyn**)

:   **iMemb** = iChannels + iSyn (exposed as **iMemb**)

***Time Derivatives***

:   d **v** /dt = iMemb / C
:::
:::
:::

::: {#distal .section}
[]{#schema-distal}

##### distal[\#](#distal "Link to this heading"){.headerlink}

extends [[point3DWithDiam]{.std
.std-ref}](Landing.html#schema-point3dwithdiam){.reference .internal}

*Point on a [[segment]{.std .std-ref}](#schema-segment){.reference
.internal} furthest from the soma. Should always be present in the
description of a [[segment]{.std .std-ref}](#schema-segment){.reference
.internal}, unlike [[proximal]{.std
.std-ref}](#schema-proximal){.reference .internal}.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **di | Diameter of the ppoint. Note: no dimension      | Di          |
| amet | used, see description of *point3DWithDiam* for  | mensionless |
| er** | details. *(from [[point3DWithDiam]{.std         |             |
|      | .std-ref}](                                     |             |
|      | Landing.html#schema-point3dwithdiam){.reference |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| *    | x coordinate of the point. Note: no dimension   | Di          |
| *x** | used, see description of *point3DWithDiam* for  | mensionless |
|      | details. *(from [[point3DWithDiam]{.std         |             |
|      | .std-ref}](                                     |             |
|      | Landing.html#schema-point3dwithdiam){.reference |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| *    | y coordinate of the ppoint. Note: no dimension  | Di          |
| *y** | used, see description of *point3DWithDiam* for  | mensionless |
|      | details. *(from [[point3DWithDiam]{.std         |             |
|      | .std-ref}](                                     |             |
|      | Landing.html#schema-point3dwithdiam){.reference |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| *    | z coordinate of the ppoint. Note: no dimension  | Di          |
| *z** | used, see description of *point3DWithDiam* for  | mensionless |
|      | details. *(from [[point3DWithDiam]{.std         |             |
|      | .std-ref}](                                     |             |
|      | Landing.html#schema-point3dwithdiam){.reference |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Derived parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   | A dimensional quantity given by half the        | [[l         |
| radi | \_diameter. *(from [[point3DWithDiam]{.std      | ength]{.std |
| us** | .std-ref}](                                     | .std-       |
|      | Landing.html#schema-point3dwithdiam){.reference | ref}](Landi |
|      | .internal})*                                    | ng.html#sch |
|      |                                                 | ema-dimensi |
|      |                                                 | ons-length) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::

   **radius** = MICRON \* diameter / 2

::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **x  | A version of \_x with dimension length. *(from  | [[l         |
| Leng | [[point3DWithDiam]{.std                         | ength]{.std |
| th** | .std-ref}](                                     | .std-       |
|      | Landing.html#schema-point3dwithdiam){.reference | ref}](Landi |
|      | .internal})*                                    | ng.html#sch |
|      |                                                 | ema-dimensi |
|      |                                                 | ons-length) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::

   **xLength** = MICRON \* x

::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **y  | A version of \_y with dimension length. *(from  | [[l         |
| Leng | [[point3DWithDiam]{.std                         | ength]{.std |
| th** | .std-ref}](                                     | .std-       |
|      | Landing.html#schema-point3dwithdiam){.reference | ref}](Landi |
|      | .internal})*                                    | ng.html#sch |
|      |                                                 | ema-dimensi |
|      |                                                 | ons-length) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::

   **yLength** = MICRON \* y

::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **z  | A version of \_z with dimension length. *(from  | [[l         |
| Leng | [[point3DWithDiam]{.std                         | ength]{.std |
| th** | .std-ref}](                                     | .std-       |
|      | Landing.html#schema-point3dwithdiam){.reference | ref}](Landi |
|      | .internal})*                                    | ng.html#sch |
|      |                                                 | ema-dimensi |
|      |                                                 | ons-length) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::

   **zLength** = MICRON \* z
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <distal x="10" y="0" z="0" diameter="10"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <distal x="20" y="0" z="0" diameter="3"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <distal x="30" y="0" z="0" diameter="1"/>
:::
:::
:::
:::
:::

::: {#proximal .section}
[]{#schema-proximal}

##### proximal[\#](#proximal "Link to this heading"){.headerlink}

extends [[point3DWithDiam]{.std
.std-ref}](Landing.html#schema-point3dwithdiam){.reference .internal}

*Point on a [[segment]{.std .std-ref}](#schema-segment){.reference
.internal} closest to the soma. Note, the proximal point can be omitted,
and in this case is defined as being the point **fractionAlong** between
the proximal and [[distal]{.std .std-ref}](#schema-distal){.reference
.internal} point of the [[parent]{.std
.std-ref}](#schema-parent){.reference .internal}, i.e. if
**fractionAlong** = 1 ( as it is by default ) it will be the **distal**
on the parent, or if **fractionAlong** = 0, it will be the proximal
point. If between 0 and 1, it is the linear interpolation between the
two points.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **di | Diameter of the ppoint. Note: no dimension      | Di          |
| amet | used, see description of *point3DWithDiam* for  | mensionless |
| er** | details. *(from [[point3DWithDiam]{.std         |             |
|      | .std-ref}](                                     |             |
|      | Landing.html#schema-point3dwithdiam){.reference |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| *    | x coordinate of the point. Note: no dimension   | Di          |
| *x** | used, see description of *point3DWithDiam* for  | mensionless |
|      | details. *(from [[point3DWithDiam]{.std         |             |
|      | .std-ref}](                                     |             |
|      | Landing.html#schema-point3dwithdiam){.reference |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| *    | y coordinate of the ppoint. Note: no dimension  | Di          |
| *y** | used, see description of *point3DWithDiam* for  | mensionless |
|      | details. *(from [[point3DWithDiam]{.std         |             |
|      | .std-ref}](                                     |             |
|      | Landing.html#schema-point3dwithdiam){.reference |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| *    | z coordinate of the ppoint. Note: no dimension  | Di          |
| *z** | used, see description of *point3DWithDiam* for  | mensionless |
|      | details. *(from [[point3DWithDiam]{.std         |             |
|      | .std-ref}](                                     |             |
|      | Landing.html#schema-point3dwithdiam){.reference |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Derived parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   | A dimensional quantity given by half the        | [[l         |
| radi | \_diameter. *(from [[point3DWithDiam]{.std      | ength]{.std |
| us** | .std-ref}](                                     | .std-       |
|      | Landing.html#schema-point3dwithdiam){.reference | ref}](Landi |
|      | .internal})*                                    | ng.html#sch |
|      |                                                 | ema-dimensi |
|      |                                                 | ons-length) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::

   **radius** = MICRON \* diameter / 2

::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **x  | A version of \_x with dimension length. *(from  | [[l         |
| Leng | [[point3DWithDiam]{.std                         | ength]{.std |
| th** | .std-ref}](                                     | .std-       |
|      | Landing.html#schema-point3dwithdiam){.reference | ref}](Landi |
|      | .internal})*                                    | ng.html#sch |
|      |                                                 | ema-dimensi |
|      |                                                 | ons-length) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::

   **xLength** = MICRON \* x

::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **y  | A version of \_y with dimension length. *(from  | [[l         |
| Leng | [[point3DWithDiam]{.std                         | ength]{.std |
| th** | .std-ref}](                                     | .std-       |
|      | Landing.html#schema-point3dwithdiam){.reference | ref}](Landi |
|      | .internal})*                                    | ng.html#sch |
|      |                                                 | ema-dimensi |
|      |                                                 | ons-length) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::

   **yLength** = MICRON \* y

::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **z  | A version of \_z with dimension length. *(from  | [[l         |
| Leng | [[point3DWithDiam]{.std                         | ength]{.std |
| th** | .std-ref}](                                     | .std-       |
|      | Landing.html#schema-point3dwithdiam){.reference | ref}](Landi |
|      | .internal})*                                    | ng.html#sch |
|      |                                                 | ema-dimensi |
|      |                                                 | ons-length) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::

   **zLength** = MICRON \* z
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <proximal x="0" y="0" z="0" diameter="10"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <proximal x="25" y="0" z="0" diameter="0.1"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <proximal x="0" y="0" z="0" diameter="10"/>
:::
:::
:::
:::
:::

::: {#parent .section}
[]{#schema-parent}

##### parent[\#](#parent "Link to this heading"){.headerlink}

*Specifies the [[segment]{.std .std-ref}](#schema-segment){.reference
.internal} which is this segment's parent. The **fractionAlong**
specifies where it is connected, usually 1 ( the default value ),
meaning the [[distal]{.std .std-ref}](#schema-distal){.reference
.internal} point of the parent, or 0, meaning the [[proximal]{.std
.std-ref}](#schema-proximal){.reference .internal} point. If it is
between these, a linear interpolation between the 2 points should be
used.*

::: {.sd-tab-set .docutils}
Text fields

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| **seg  | The id of the parent segment                                 |
| ment** |                                                              |
+--------+--------------------------------------------------------------+
| **fra  | The fraction along the the parent segment at which this      |
| ctionA | segment is attached. For usage see *proximal*                |
| long** |                                                              |
+--------+--------------------------------------------------------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="SegmentParent">
      <xs:complexContent>
        <xs:extension base="BaseWithoutId">
          <xs:attribute name="segment" type="NonNegativeInteger" use="required"/>
          <xs:attribute name="fractionAlong" type="ZeroToOne" use="optional" default="1"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <parent segment="0"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <parent segment="1"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <parent segment="2" fractionAlong="0.5"/>
:::
:::
:::
:::
:::

::: {#segment .section}
[]{#schema-segment}

##### segment[\#](#segment "Link to this heading"){.headerlink}

*A segment defines the smallest unit within a possibly branching
structure ( [[morphology]{.std .std-ref}](#schema-morphology){.reference
.internal} ), such as a dendrite or axon. Its **id** should be a
nonnegative integer ( usually soma/root = 0 ). Its end points are given
by the [[proximal]{.std .std-ref}](#schema-proximal){.reference
.internal} and [[distal]{.std .std-ref}](#schema-distal){.reference
.internal} points. The [[proximal]{.std
.std-ref}](#schema-proximal){.reference .internal} point can be omitted,
usually because it is the same as a point on the [[parent]{.std
.std-ref}](#schema-parent){.reference .internal} segment, see
[[proximal]{.std .std-ref}](#schema-proximal){.reference .internal} for
details. [[parent]{.std .std-ref}](#schema-parent){.reference .internal}
specifies the parent segment. The first segment of a [[cell]{.std
.std-ref}](#schema-cell){.reference .internal} ( with no [[parent]{.std
.std-ref}](#schema-parent){.reference .internal} ) usually represents
the soma. The shape is normally a cylinder ( radii of the
[[proximal]{.std .std-ref}](#schema-proximal){.reference .internal} and
[[distal]{.std .std-ref}](#schema-distal){.reference .internal} equal,
but positions different ) or a conical frustum ( radii and positions
different ). If the x, y, x positions of the [[proximal]{.std
.std-ref}](#schema-proximal){.reference .internal} and [[distal]{.std
.std-ref}](#schema-distal){.reference .internal} are equal, the segment
can be interpreted as a sphere, and in this case the radii of these
points must be equal. NOTE: LEMS does not yet support multicompartmental
modelling, so the Dynamics here is only appropriate for single
compartment modelling.*

::: {.sd-tab-set .docutils}
Constants

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------------------+-----------------------------------+-------------+
| **LEN** = 1m       |                                   | [[l         |
|                    |                                   | ength]{.std |
|                    |                                   | .std-       |
|                    |                                   | ref}](Landi |
|                    |                                   | ng.html#sch |
|                    |                                   | ema-dimensi |
|                    |                                   | ons-length) |
|                    |                                   | {.reference |
|                    |                                   | .internal}  |
+--------------------+-----------------------------------+-------------+
:::
:::

Text fields

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| **     | An optional name for the segment. Convenient for providing a |
| name** | suitable variable name for generated code, e.g. soma, dend0  |
+--------+--------------------------------------------------------------+
:::
:::

Child list

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   |                                                 | [[p         |
| pare |                                                 | arent]{.std |
| nt** |                                                 | .std        |
|      |                                                 | -ref}](#sch |
|      |                                                 | ema-parent) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   |                                                 | [[d         |
| dist |                                                 | istal]{.std |
| al** |                                                 | .std        |
|      |                                                 | -ref}](#sch |
|      |                                                 | ema-distal) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **pr |                                                 | [[pro       |
| oxim |                                                 | ximal]{.std |
| al** |                                                 | .std-r      |
|      |                                                 | ef}](#schem |
|      |                                                 | a-proximal) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   |                                                 | [[l         |
| leng |                                                 | ength]{.std |
| th** |                                                 | .std-       |
|      |                                                 | ref}](Landi |
|      |                                                 | ng.html#sch |
|      |                                                 | ema-dimensi |
|      |                                                 | ons-length) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **r  |                                                 | [[l         |
| adDi |                                                 | ength]{.std |
| st** |                                                 | .std-       |
|      |                                                 | ref}](Landi |
|      |                                                 | ng.html#sch |
|      |                                                 | ema-dimensi |
|      |                                                 | ons-length) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **s  |                                                 | [           |
| urfa |                                                 | [area]{.std |
| ceAr |                                                 | .st         |
| ea** |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-area) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***Derived Variables***

:   **radDist** = distal-\>radius (exposed as **radDist**)

:   **dx** = distal-\>xLength

:   **dy** = distal-\>yLength

:   **dz** = distal-\>zLength

:   **px** = proximal-\>xLength

:   **py** = proximal-\>yLength

:   **pz** = proximal-\>zLength

:   **length** = sqrt(((dx - px) \* (dx - px) + (dy - py) \* (dy - py) +
    (dz - pz) \* (dz - pz))/(LEN \* LEN)) \* LEN (exposed as **length**)

***Conditional Derived Variables***

:   IF length = 0 \* LEN THEN

:     **surfaceArea** = 4 \* radDist \* radDist \* 3.14159265  (exposed
    as **surfaceArea**)

:   IF length \> 0 \* LEN THEN

:     **surfaceArea** = 2 \* radDist \* 3.14159265 \* length  (exposed
    as **surfaceArea**)
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="Segment">
      <xs:complexContent>
        <xs:extension base="BaseNonNegativeIntegerId">
          <xs:sequence>
            <xs:element name="parent" type="SegmentParent" minOccurs="0"/>
            <xs:element name="proximal" type="Point3DWithDiam" minOccurs="0"/>
            <xs:element name="distal" type="Point3DWithDiam" minOccurs="1"/>
          </xs:sequence>
          <xs:attribute name="name" type="xs:string" use="optional"/>
          <xs:attribute name="neuroLexId" type="NeuroLexId" use="optional"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=Segment){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import Segment
    from neuroml.utils import component_factory

    variable = component_factory(
        Segment,
        id: 'a NonNegativeInteger (required)' = None,
        name: 'a string (optional)' = None,
        neuro_lex_id: 'a NeuroLexId (optional)' = None,
        parent: 'a SegmentParent (optional)' = None,
        proximal: 'a Point3DWithDiam (optional)' = None,
        distal: 'a Point3DWithDiam (required)' = None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <segment id="3" name="Spine1">
        <parent segment="2" fractionAlong="0.5"/>
        <proximal x="25" y="0" z="0" diameter="0.1"/>
        <distal x="25" y="0.2" z="0" diameter="0.1"/>
    </segment>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <segment id="0" name="Soma">
        <proximal x="0" y="0" z="0" diameter="10"/>
        <distal x="10" y="0" z="0" diameter="10"/>
    </segment>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <segment id="1" name="Dendrite1">
        <parent segment="0"/>
        <distal x="20" y="0" z="0" diameter="3"/>
    </segment>
:::
:::
:::
:::
:::

::: {#segmentgroup .section}
[]{#schema-segmentgroup}

##### segmentGroup[\#](#segmentgroup "Link to this heading"){.headerlink}

*A method to describe a group of [[segment]{.std
.std-ref}](#schema-segment){.reference .internal}s in a
[[morphology]{.std .std-ref}](#schema-morphology){.reference .internal},
e.g. soma_group, dendrite_group, axon_group. While a name is useful to
describe the group, the **neuroLexId** attribute can be used to
explicitly specify the meaning of the group, e.g. sao1044911821 for
'Neuronal Cell Body', sao1211023249 for 'Dendrite'. The [[segment]{.std
.std-ref}](#schema-segment){.reference .internal}s in this group can be
specified as: a list of individual [[member]{.std
.std-ref}](#schema-member){.reference .internal} segments; a
[[path]{.std .std-ref}](#schema-path){.reference .internal}, all of the
segments along which should be included; a [[subTree]{.std
.std-ref}](#schema-subtree){.reference .internal} of the [[cell]{.std
.std-ref}](#schema-cell){.reference .internal} to include; other
segmentGroups to [[include]{.std .std-ref}](#schema-include){.reference
.internal} ( so all segments from those get included here ). An
[[inhomogeneousParameter]{.std
.std-ref}](#schema-inhomogeneousparameter){.reference .internal} can be
defined on the region of the cell specified by this group ( see
[[variableParameter]{.std
.std-ref}](#schema-variableparameter){.reference .internal} for usage
).*

::: {.sd-tab-set .docutils}
Text fields

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| **     | An id string for pointing to an entry in the NeuroLex        |
| neuroL | ontology. Use of this attribute is a shorthand for a full    |
| exId** | RDF based reference to the MIRIAM Resource                   |
|        | urn:miriam:neurolex, with an bqbiol:is qualifier.            |
+--------+--------------------------------------------------------------+
:::
:::

Child list

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    |                                                 | [[          |
| *not |                                                 | notes]{.std |
| es** |                                                 | .std        |
|      |                                                 | -ref}](Land |
|      |                                                 | ing.html#sc |
|      |                                                 | hema-notes) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   |                                                 | [[annot     |
| anno |                                                 | ation]{.std |
| tati |                                                 | .std-ref}   |
| on** |                                                 | ](Landing.h |
|      |                                                 | tml#schema- |
|      |                                                 | annotation) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Children list

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **pr |                                                 | [[pro       |
| oper |                                                 | perty]{.std |
| ty** |                                                 | .std-re     |
|      |                                                 | f}](Landing |
|      |                                                 | .html#schem |
|      |                                                 | a-property) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **m  |                                                 | [[m         |
| embe |                                                 | ember]{.std |
| rs** |                                                 | .std        |
|      |                                                 | -ref}](#sch |
|      |                                                 | ema-member) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    |                                                 | [           |
| *pat |                                                 | [path]{.std |
| hs** |                                                 | .s          |
|      |                                                 | td-ref}](#s |
|      |                                                 | chema-path) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **su |                                                 | [[su        |
| bTre |                                                 | bTree]{.std |
| es** |                                                 | .std-       |
|      |                                                 | ref}](#sche |
|      |                                                 | ma-subtree) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **in |                                                 | [[in        |
| clud |                                                 | clude]{.std |
| es** |                                                 | .std-       |
|      |                                                 | ref}](#sche |
|      |                                                 | ma-include) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   |                                                 | [[inhomo    |
| inho |                                                 | geneousPara |
| moge |                                                 | meter]{.std |
| neou |                                                 | .std-ref}   |
| sPar |                                                 | ](#schema-i |
| amet |                                                 | nhomogeneou |
| er** |                                                 | sparameter) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="SegmentGroup">
      <xs:complexContent>
        <xs:extension base="Base">
          <xs:sequence>
            <xs:element name="notes" type="Notes" minOccurs="0"/>
            <xs:element name="property" type="Property" minOccurs="0" maxOccurs="unbounded"/>
            <xs:element name="annotation" type="Annotation" minOccurs="0"/>
            <xs:element name="member" type="Member" minOccurs="0" maxOccurs="unbounded"/>
            <xs:element name="include" type="Include" minOccurs="0" maxOccurs="unbounded"/>
            <xs:element name="path" type="Path" minOccurs="0" maxOccurs="unbounded"/>
            <xs:element name="subTree" type="SubTree" minOccurs="0" maxOccurs="unbounded"/>
            <xs:element name="inhomogeneousParameter" type="InhomogeneousParameter" minOccurs="0" maxOccurs="unbounded"/>
          </xs:sequence>
          <xs:attribute name="neuroLexId" type="NeuroLexId" use="optional"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=SegmentGroup){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import SegmentGroup
    from neuroml.utils import component_factory

    variable = component_factory(
        SegmentGroup,
        id: 'a NonNegativeInteger (required)' = None,
        neuro_lex_id: 'a NeuroLexId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        members: 'list of Member(s) (optional)' = None,
        includes: 'list of Include(s) (optional)' = None,
        paths: 'list of Path(s) (optional)' = None,
        sub_trees: 'list of SubTree(s) (optional)' = None,
        inhomogeneous_parameters: 'list of InhomogeneousParameter(s) (optional)' = None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <segmentGroup id="dendrite_group" neuroLexId="sao1211023249">
        <member segment="1"/>
        <member segment="2"/>
        <member segment="3"/>
    </segmentGroup>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <segmentGroup id="soma_group" neuroLexId="sao1044911821">
        <member segment="0"/>
    </segmentGroup>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <segmentGroup id="spines" neuroLexId="sao1145756102">
        <member segment="3"/>
    </segmentGroup>
:::
:::
:::
:::
:::

::: {#member .section}
[]{#schema-member}

##### member[\#](#member "Link to this heading"){.headerlink}

*A single identified **segment** which is part of the
[[segmentGroup]{.std .std-ref}](#schema-segmentgroup){.reference
.internal}.*

::: {.sd-tab-set .docutils}
Text fields

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| **seg  |                                                              |
| ment** |                                                              |
+--------+--------------------------------------------------------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="Member">
      <xs:complexContent>
        <xs:extension base="BaseWithoutId">
          <xs:attribute name="segment" type="NonNegativeInteger" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=Member){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import Member
    from neuroml.utils import component_factory

    variable = component_factory(
        Member,
        segments: 'a NonNegativeInteger (required)' = None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <member segment="0"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <member segment="1"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <member segment="2"/>
:::
:::
:::
:::
:::

::: {#from .section}
[]{#schema-from}

##### from[\#](#from "Link to this heading"){.headerlink}

*In a [[path]{.std .std-ref}](#schema-path){.reference .internal} or
[[subTree]{.std .std-ref}](#schema-subtree){.reference .internal},
specifies which **segment** ( inclusive ) from which to calculate the
[[segmentGroup]{.std .std-ref}](#schema-segmentgroup){.reference
.internal}.*

::: {.sd-tab-set .docutils}
Text fields

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| **seg  |                                                              |
| ment** |                                                              |
+--------+--------------------------------------------------------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="SegmentEndPoint">
      <xs:complexContent>
        <xs:extension base="BaseWithoutId">
          <xs:attribute name="segment" type="NonNegativeInteger" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <from segment="1"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <from segment="1"/>
:::
:::
:::
:::
:::

::: {#to .section}
[]{#schema-to}

##### to[\#](#to "Link to this heading"){.headerlink}

*In a [[path]{.std .std-ref}](#schema-path){.reference .internal},
specifies which **segment** ( inclusive ) up to which to calculate the
[[segmentGroup]{.std .std-ref}](#schema-segmentgroup){.reference
.internal}.*

::: {.sd-tab-set .docutils}
Text fields

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| **seg  |                                                              |
| ment** |                                                              |
+--------+--------------------------------------------------------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="SegmentEndPoint">
      <xs:complexContent>
        <xs:extension base="BaseWithoutId">
          <xs:attribute name="segment" type="NonNegativeInteger" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <to segment="2"/>
:::
:::
:::
:::
:::

::: {#include .section}
[]{#schema-include}

##### include[\#](#include "Link to this heading"){.headerlink}

*Include all members of another [[segmentGroup]{.std
.std-ref}](#schema-segmentgroup){.reference .internal} in this group.*

::: {.sd-tab-set .docutils}
Text fields

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| **     |                                                              |
| href** |                                                              |
+--------+--------------------------------------------------------------+
| **se   |                                                              |
| gmentG |                                                              |
| roup** |                                                              |
+--------+--------------------------------------------------------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="Include">
      <xs:complexContent>
        <xs:extension base="BaseWithoutId">
          <xs:attribute name="segmentGroup" type="NmlId" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=Include){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import Include
    from neuroml.utils import component_factory

    variable = component_factory(
        Include,
        segment_groups: 'a NmlId (required)' = None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <include href="NML2_SingleCompHHCell.nml"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <include href="NML2_SimpleIonChannel.nml"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <include href="NML2_SimpleIonChannel.nml"/>
:::
:::
:::
:::
:::

::: {#path .section}
[]{#schema-path}

##### path[\#](#path "Link to this heading"){.headerlink}

*Include all the [[segment]{.std .std-ref}](#schema-segment){.reference
.internal}s between those specified by [[from]{.std
.std-ref}](#schema-from){.reference .internal} and [[to]{.std
.std-ref}](#schema-to){.reference .internal}, inclusive.*

::: {.sd-tab-set .docutils}
Child list

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **fr |                                                 | [           |
| om** |                                                 | [from]{.std |
|      |                                                 | .s          |
|      |                                                 | td-ref}](#s |
|      |                                                 | chema-from) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   |                                                 | [[to]{.std  |
| to** |                                                 | .std-ref}]( |
|      |                                                 | #schema-to) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="Path">
      <xs:complexContent>
        <xs:extension base="BaseWithoutId">
          <xs:sequence>
            <xs:element name="from" type="SegmentEndPoint" minOccurs="0"/>
            <xs:element name="to" type="SegmentEndPoint" minOccurs="0"/>
          </xs:sequence>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=Path){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import Path
    from neuroml.utils import component_factory

    variable = component_factory(
        Path,
        from_: 'a SegmentEndPoint (optional)' = None,
        to: 'a SegmentEndPoint (optional)' = None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <path>
        <from segment="1"/>
        <to segment="2"/>
    </path>
:::
:::
:::
:::
:::

::: {#subtree .section}
[]{#schema-subtree}

##### subTree[\#](#subtree "Link to this heading"){.headerlink}

*Include all the [[segment]{.std .std-ref}](#schema-segment){.reference
.internal}s distal to that specified by [[from]{.std
.std-ref}](#schema-from){.reference .internal} in the
[[segmentGroup]{.std .std-ref}](#schema-segmentgroup){.reference
.internal}.*

::: {.sd-tab-set .docutils}
Child list

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **fr |                                                 | [           |
| om** |                                                 | [from]{.std |
|      |                                                 | .s          |
|      |                                                 | td-ref}](#s |
|      |                                                 | chema-from) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="SubTree">
      <xs:complexContent>
        <xs:extension base="BaseWithoutId">
          <xs:choice>
            <xs:element name="from" type="SegmentEndPoint" minOccurs="0"/>
            <xs:element name="to" type="SegmentEndPoint" minOccurs="0"/>
          </xs:choice>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=SubTree){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import SubTree
    from neuroml.utils import component_factory

    variable = component_factory(
        SubTree,
        from_: 'a SegmentEndPoint (optional)' = None,
        to: 'a SegmentEndPoint (optional)' = None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <subTree>
        <from segment="1"/>
    </subTree>
:::
:::
:::
:::
:::

::: {#inhomogeneousparameter .section}
[]{#schema-inhomogeneousparameter}

##### inhomogeneousParameter[\#](#inhomogeneousparameter "Link to this heading"){.headerlink}

*An inhomogeneous parameter specified across the [[segmentGroup]{.std
.std-ref}](#schema-segmentgroup){.reference .internal} ( see
[[variableParameter]{.std
.std-ref}](#schema-variableparameter){.reference .internal} for usage
).*

::: {.sd-tab-set .docutils}
Text fields

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| **vari |                                                              |
| able** |                                                              |
+--------+--------------------------------------------------------------+
| **me   |                                                              |
| tric** |                                                              |
+--------+--------------------------------------------------------------+
:::
:::

Child list

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **pr |                                                 | [           |
| oxim |                                                 | [proximalDe |
| al** |                                                 | tails]{.std |
|      |                                                 | .s          |
|      |                                                 | td-ref}](#s |
|      |                                                 | chema-proxi |
|      |                                                 | maldetails) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   |                                                 | [[distalDe  |
| dist |                                                 | tails]{.std |
| al** |                                                 | .std-ref}]( |
|      |                                                 | #schema-dis |
|      |                                                 | taldetails) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="InhomogeneousParameter">
      <xs:complexContent>
        <xs:extension base="Base">
          <xs:sequence>
            <xs:element name="proximal" type="ProximalDetails" minOccurs="0"/>
            <xs:element name="distal" type="DistalDetails" minOccurs="0"/>
          </xs:sequence>
          <xs:attribute name="variable" type="xs:string" use="required"/>
          <xs:attribute name="metric" type="Metric" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=InhomogeneousParameter){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import InhomogeneousParameter
    from neuroml.utils import component_factory

    variable = component_factory(
        InhomogeneousParameter,
        id: 'a NmlId (required)' = None,
        variable: 'a string (required)' = None,
        metric: 'a Metric (required)' = None,
        proximal: 'a ProximalDetails (optional)' = None,
        distal: 'a DistalDetails (optional)' = None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <inhomogeneousParameter id="dendrite_group_x2" variable="r" metric="Path Length from root">
        <proximal translationStart="0"/>
        <distal normalizationEnd="1"/>
    </inhomogeneousParameter>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <inhomogeneousParameter id="dendrite_group_x1" variable="p" metric="Path Length from root"/>
:::
:::
:::
:::
:::

::: {#proximaldetails .section}
[]{#schema-proximaldetails}

##### proximalDetails[\#](#proximaldetails "Link to this heading"){.headerlink}

*What to do at the proximal point when creating an inhomogeneous
parameter.*

::: {.sd-tab-set .docutils}
Text fields

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| **     |                                                              |
| transl |                                                              |
| ationS |                                                              |
| tart** |                                                              |
+--------+--------------------------------------------------------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="ProximalDetails">
      <xs:complexContent>
        <xs:extension base="BaseWithoutId">
          <xs:attribute name="translationStart" type="xs:double" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=ProximalDetails){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import ProximalDetails
    from neuroml.utils import component_factory

    variable = component_factory(
        ProximalDetails,
        translation_start: 'a double (required)' = None,
    )
:::
:::
:::
:::
:::

::: {#distaldetails .section}
[]{#schema-distaldetails}

##### distalDetails[\#](#distaldetails "Link to this heading"){.headerlink}

*What to do at the distal point when creating an inhomogeneous
parameter.*

::: {.sd-tab-set .docutils}
Text fields

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| **     |                                                              |
| normal |                                                              |
| izatio |                                                              |
| nEnd** |                                                              |
+--------+--------------------------------------------------------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="DistalDetails">
      <xs:complexContent>
        <xs:extension base="BaseWithoutId">
          <xs:attribute name="normalizationEnd" type="xs:double" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=DistalDetails){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import DistalDetails
    from neuroml.utils import component_factory

    variable = component_factory(
        DistalDetails,
        normalization_end: 'a double (required)' = None,
    )
:::
:::
:::
:::
:::

::: {#morphology .section}
[]{#schema-morphology}

##### morphology[\#](#morphology "Link to this heading"){.headerlink}

*The collection of [[segment]{.std
.std-ref}](#schema-segment){.reference .internal}s which specify the 3D
structure of the cell, along with a number of [[segmentGroup]{.std
.std-ref}](#schema-segmentgroup){.reference .internal}s.*

::: {.sd-tab-set .docutils}
Children list

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **se |                                                 | [[se        |
| gmen |                                                 | gment]{.std |
| ts** |                                                 | .std-       |
|      |                                                 | ref}](#sche |
|      |                                                 | ma-segment) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    |                                                 | [[segment   |
| *seg |                                                 | Group]{.std |
| ment |                                                 | .std-ref}]  |
| Grou |                                                 | (#schema-se |
| ps** |                                                 | gmentgroup) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="Morphology">
      <xs:complexContent>
        <xs:extension base="Standalone">
          <xs:sequence>
            <xs:element name="segment" type="Segment" maxOccurs="unbounded"/>
            <xs:element name="segmentGroup" type="SegmentGroup" minOccurs="0" maxOccurs="unbounded"/>
          </xs:sequence>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=Morphology){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import Morphology
    from neuroml.utils import component_factory

    variable = component_factory(
        Morphology,
        id: 'a NmlId (required)' = None,
        metaid: 'a MetaId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        segments: 'list of Segment(s) (required)' = None,
        segment_groups: 'list of SegmentGroup(s) (optional)' = None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <morphology id="SpikingCell_morphology">
        <segment id="0" name="Soma">
            <proximal x="0" y="0" z="0" diameter="10"/>
            <distal x="10" y="0" z="0" diameter="10"/>
        </segment>
        <segment id="1" name="Dendrite1">
            <parent segment="0"/>
            <distal x="20" y="0" z="0" diameter="3"/>
        </segment>
        <segment id="2" name="Dendrite2">
            <parent segment="1"/>
            <distal x="30" y="0" z="0" diameter="1"/>
        </segment>
        <segment id="3" name="Spine1">
            <parent segment="2" fractionAlong="0.5"/>
            <proximal x="25" y="0" z="0" diameter="0.1"/>
            <distal x="25" y="0.2" z="0" diameter="0.1"/>
        </segment>
        <segmentGroup id="soma_group" neuroLexId="sao1044911821">
            <member segment="0"/>
        </segmentGroup>
        <segmentGroup id="dendrite_group" neuroLexId="sao1211023249">
            <member segment="1"/>
            <member segment="2"/>
            <member segment="3"/>
        </segmentGroup>
        <segmentGroup id="spines" neuroLexId="sao1145756102">
            <member segment="3"/>
        </segmentGroup>
    </morphology>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <morphology id="NeuroMorpho_PyrCell123">
        <segment id="0" name="Soma">
            <proximal x="0" y="0" z="0" diameter="10"/>
            <distal x="10" y="0" z="0" diameter="10"/>
        </segment>
    </morphology>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <morphology id="SimpleCell_Morphology">
        <segment id="0" name="Soma">
            <proximal x="0" y="0" z="0" diameter="10"/>
            <distal x="10" y="0" z="0" diameter="10"/>
        </segment>
        <segment id="1" name="MainDendrite1">
            <parent segment="0"/>
            <proximal x="10" y="0" z="0" diameter="3"/>
            <distal x="20" y="0" z="0" diameter="3"/>
        </segment>
        <segment id="2" name="MainDendrite2">
            <parent segment="1"/>
            <distal x="30" y="0" z="0" diameter="1"/>
        </segment>
        <segmentGroup id="soma_group" neuroLexId="sao1044911821">
            <member segment="0"/>
        </segmentGroup>
        <segmentGroup id="dendrite_group" neuroLexId="sao1211023249">
            <member segment="1"/>
            <member segment="2"/>
            <inhomogeneousParameter id="dendrite_group_x1" variable="p" metric="Path Length from root"/>
            <inhomogeneousParameter id="dendrite_group_x2" variable="r" metric="Path Length from root">
                <proximal translationStart="0"/>
                <distal normalizationEnd="1"/>
            </inhomogeneousParameter>
        </segmentGroup>
    </morphology>
:::
:::
:::
:::
:::

::: {#specificcapacitance .section}
[]{#schema-specificcapacitance}

##### specificCapacitance[\#](#specificcapacitance "Link to this heading"){.headerlink}

*Capacitance per unit area.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    |                                                 | [[spe       |
| *val |                                                 | cificCapaci |
| ue** |                                                 | tance]{.std |
|      |                                                 | .std-re     |
|      |                                                 | f}](Landing |
|      |                                                 | .html#schem |
|      |                                                 | a-dimension |
|      |                                                 | s-specificc |
|      |                                                 | apacitance) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Text fields

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| **se   |                                                              |
| gmentG |                                                              |
| roup** |                                                              |
+--------+--------------------------------------------------------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **s  |                                                 | [[spe       |
| pecC |                                                 | cificCapaci |
| ap** |                                                 | tance]{.std |
|      |                                                 | .std-re     |
|      |                                                 | f}](Landing |
|      |                                                 | .html#schem |
|      |                                                 | a-dimension |
|      |                                                 | s-specificc |
|      |                                                 | apacitance) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***Derived Variables***

:   **specCap** = value (exposed as **specCap**)
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="SpecificCapacitance">
      <xs:complexContent>
        <xs:extension base="BaseWithoutId">
          <xs:attribute name="value" type="Nml2Quantity_specificCapacitance" use="required"/>
          <xs:attribute name="segmentGroup" type="NmlId" use="optional" default="all"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=SpecificCapacitance){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import SpecificCapacitance
    from neuroml.utils import component_factory

    variable = component_factory(
        SpecificCapacitance,
        value: 'a Nml2Quantity_specificCapacitance (required)' = None,
        segment_groups: 'a NmlId (optional)' = 'all',
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <specificCapacitance segmentGroup="soma_group" value="1.0 uF_per_cm2"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <specificCapacitance segmentGroup="dendrite_group" value="2.0 uF_per_cm2"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <specificCapacitance segmentGroup="soma_group" value="1.0 uF_per_cm2"/>
:::
:::
:::
:::
:::

::: {#initmembpotential .section}
[]{#schema-initmembpotential}

##### initMembPotential[\#](#initmembpotential "Link to this heading"){.headerlink}

*Explicitly set initial membrane potential for the cell.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    |                                                 | [[vo        |
| *val |                                                 | ltage]{.std |
| ue** |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="InitMembPotential">
      <xs:complexContent>
        <xs:extension base="BaseWithoutId">
          <xs:attribute name="value" type="Nml2Quantity_voltage" use="required"/>
          <xs:attribute name="segmentGroup" type="NmlId" use="optional" default="all"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=InitMembPotential){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import InitMembPotential
    from neuroml.utils import component_factory

    variable = component_factory(
        InitMembPotential,
        value: 'a Nml2Quantity_voltage (required)' = None,
        segment_groups: 'a NmlId (optional)' = 'all',
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <initMembPotential value="-65mV"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <initMembPotential value="-65mV"/>
:::
:::
:::
:::
:::

::: {#spikethresh .section}
[]{#schema-spikethresh}

##### spikeThresh[\#](#spikethresh "Link to this heading"){.headerlink}

*Membrane potential at which to emit a spiking event. Note, usually the
spiking event will not be emitted again until the membrane potential has
fallen below this value and rises again to cross it in a positive
direction.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    |                                                 | [[vo        |
| *val |                                                 | ltage]{.std |
| ue** |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="SpikeThresh">
      <xs:complexContent>
        <xs:extension base="BaseWithoutId">
          <xs:attribute name="value" type="Nml2Quantity_voltage" use="required"/>
          <xs:attribute name="segmentGroup" type="NmlId" use="optional" default="all"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=SpikeThresh){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import SpikeThresh
    from neuroml.utils import component_factory

    variable = component_factory(
        SpikeThresh,
        value: 'a Nml2Quantity_voltage (required)' = None,
        segment_groups: 'a NmlId (optional)' = 'all',
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <spikeThresh value="-20mV"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <spikeThresh value="-20mV"/>
:::
:::
:::
:::
:::

::: {#membraneproperties .section}
[]{#schema-membraneproperties}

##### membraneProperties[\#](#membraneproperties "Link to this heading"){.headerlink}

*Properties specific to the membrane, such as the **populations** of
channels, **channelDensities,** **specificCapacitance,** etc.*

::: {.sd-tab-set .docutils}
Child list

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    |                                                 | [[i         |
| *ini |                                                 | nitMembPote |
| tMem |                                                 | ntial]{.std |
| bPot |                                                 | .std        |
| enti |                                                 | -ref}](#sch |
| al** |                                                 | ema-initmem |
|      |                                                 | bpotential) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **s  |                                                 | [[spikeT    |
| pike |                                                 | hresh]{.std |
| Thre |                                                 | .std-ref}   |
| sh** |                                                 | ](#schema-s |
|      |                                                 | pikethresh) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Children list

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **sp |                                                 | [[spe       |
| ecif |                                                 | cificCapaci |
| icCa |                                                 | tance]{.std |
| paci |                                                 | .std-r      |
| tanc |                                                 | ef}](#schem |
| es** |                                                 | a-specificc |
|      |                                                 | apacitance) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **p  |                                                 | [[baseC     |
| opul |                                                 | hannelPopul |
| atio |                                                 | ation]{.std |
| ns** |                                                 | .std-ref    |
|      |                                                 | }](#schema- |
|      |                                                 | basechannel |
|      |                                                 | population) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **ch |                                                 | [[ba        |
| anne |                                                 | seChannelDe |
| lDen |                                                 | nsity]{.std |
| siti |                                                 | .std-       |
| es** |                                                 | ref}](#sche |
|      |                                                 | ma-basechan |
|      |                                                 | neldensity) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **i  |                                                 | [[cu        |
| Ca** |                                                 | rrent]{.std |
|      |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   |                                                 | [[cu        |
| totC |                                                 | rrent]{.std |
| hanC |                                                 | .std-r      |
| urre |                                                 | ef}](Landin |
| nt** |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   |                                                 | [[spe       |
| totS |                                                 | cificCapaci |
| pecC |                                                 | tance]{.std |
| ap** |                                                 | .std-re     |
|      |                                                 | f}](Landing |
|      |                                                 | .html#schem |
|      |                                                 | a-dimension |
|      |                                                 | s-specificc |
|      |                                                 | apacitance) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **s  |                                                 | [           |
| urfa |                                                 | [area]{.std |
| ceAr |                                                 | .st         |
| ea** |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-area) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***Derived Variables***

:   **totSpecCap** = specificCapacitances\[\*\]-\>specCap(reduce method:
    add) (exposed as **totSpecCap**)

:   **totChanPopCurrent** = populations\[\*\]-\>i(reduce method: add)

:   **totChanDensCurrentDensity**
    = channelDensities\[\*\]-\>iDensity(reduce method: add)

:   **totChanCurrent** = totChanPopCurrent + (totChanDensCurrentDensity
    \* surfaceArea) (exposed as **totChanCurrent**)

:   **totChanPopCurrentCa** = populations\[ion='ca'\]-\>i(reduce method:
    add)

:   **totChanDensCurrentDensityCa**
    = channelDensities\[ion='ca'\]-\>iDensity(reduce method: add)

:   **iCa** = totChanPopCurrentCa + (totChanDensCurrentDensityCa \*
    surfaceArea) (exposed as **iCa**)
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="MembraneProperties">
      <xs:complexContent>
        <xs:extension base="BaseWithoutId">
          <xs:sequence>
            <xs:element name="channelPopulation" type="ChannelPopulation" minOccurs="0" maxOccurs="unbounded"/>
            <xs:element name="channelDensity" type="ChannelDensity" minOccurs="0" maxOccurs="unbounded"/>
            <xs:element name="channelDensityVShift" type="ChannelDensityVShift" minOccurs="0" maxOccurs="unbounded"/>
            <xs:element name="channelDensityNernst" type="ChannelDensityNernst" minOccurs="0" maxOccurs="unbounded"/>
            <xs:element name="channelDensityGHK" type="ChannelDensityGHK" minOccurs="0" maxOccurs="unbounded"/>
            <xs:element name="channelDensityGHK2" type="ChannelDensityGHK2" minOccurs="0" maxOccurs="unbounded"/>
            <xs:element name="channelDensityNonUniform" type="ChannelDensityNonUniform" minOccurs="0" maxOccurs="unbounded"/>
            <xs:element name="channelDensityNonUniformNernst" type="ChannelDensityNonUniformNernst" minOccurs="0" maxOccurs="unbounded"/>
            <xs:element name="channelDensityNonUniformGHK" type="ChannelDensityNonUniformGHK" minOccurs="0" maxOccurs="unbounded"/>
            <xs:element name="spikeThresh" type="SpikeThresh" minOccurs="0" maxOccurs="unbounded"/>
            <xs:element name="specificCapacitance" type="SpecificCapacitance" minOccurs="0" maxOccurs="unbounded"/>
            <xs:element name="initMembPotential" type="InitMembPotential" minOccurs="0" maxOccurs="unbounded"/>
          </xs:sequence>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=MembraneProperties){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import MembraneProperties
    from neuroml.utils import component_factory

    variable = component_factory(
        MembraneProperties,
        channel_populations: 'list of ChannelPopulation(s) (optional)' = None,
        channel_densities: 'list of ChannelDensity(s) (optional)' = None,
        channel_density_v_shifts: 'list of ChannelDensityVShift(s) (optional)' = None,
        channel_density_nernsts: 'list of ChannelDensityNernst(s) (optional)' = None,
        channel_density_ghks: 'list of ChannelDensityGHK(s) (optional)' = None,
        channel_density_ghk2s: 'list of ChannelDensityGHK2(s) (optional)' = None,
        channel_density_non_uniforms: 'list of ChannelDensityNonUniform(s) (optional)' = None,
        channel_density_non_uniform_nernsts: 'list of ChannelDensityNonUniformNernst(s) (optional)' = None,
        channel_density_non_uniform_ghks: 'list of ChannelDensityNonUniformGHK(s) (optional)' = None,
        spike_threshes: 'list of SpikeThresh(s) (required)' = None,
        specific_capacitances: 'list of SpecificCapacitance(s) (required)' = None,
        init_memb_potentials: 'list of InitMembPotential(s) (required)' = None,
        extensiontype_=None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <membraneProperties>
        <channelPopulation id="naChansDend" ionChannel="NaConductance" segment="2" number="120000" erev="50mV" ion="na"/>
        <channelDensity id="pasChans" ionChannel="pas" condDensity="3.0 S_per_m2" erev="-70mV" ion="non_specific"/>
        <channelDensity id="naChansSoma" ionChannel="NaConductance" segmentGroup="soma_group" condDensity="120.0 mS_per_cm2" erev="50mV" ion="na"/>
        <specificCapacitance segmentGroup="soma_group" value="1.0 uF_per_cm2"/>
        <specificCapacitance segmentGroup="dendrite_group" value="2.0 uF_per_cm2"/>
    </membraneProperties>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <membraneProperties>
        <channelDensity id="naChans" ionChannel="HH_Na" segmentGroup="soma_group" condDensity="120.0 mS_per_cm2" ion="na" erev="50mV"/>
        <!-- Ions present inside the cell. Note: a fixed reversal potential is specified here  
                <reversalPotential species="na" value="50mV"/>
                <reversalPotential species="k" value="-77mV"/>-->
    </membraneProperties>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <membraneProperties>
        <channelDensityNonUniform id="nonuniform_na_chans" ionChannel="NaConductance" erev="50mV" ion="na">
            <variableParameter parameter="condDensity" segmentGroup="dendrite_group">
                <inhomogeneousValue inhomogeneousParameter="dendrite_group_x1" value="5e-7 * exp(-p/200)"/>
            </variableParameter>
        </channelDensityNonUniform>
        <specificCapacitance segmentGroup="soma_group" value="1.0 uF_per_cm2"/>
    </membraneProperties>
:::
:::
:::
:::
:::

::: {#membraneproperties2capools .section}
[]{#schema-membraneproperties2capools}

##### membraneProperties2CaPools[\#](#membraneproperties2capools "Link to this heading"){.headerlink}

extends [[membraneProperties]{.std
.std-ref}](#schema-membraneproperties){.reference .internal}

*Variant of membraneProperties with 2 independent Ca pools.*

::: {.sd-tab-set .docutils}
Child list

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    |                                                 | [[i         |
| *ini |                                                 | nitMembPote |
| tMem |                                                 | ntial]{.std |
| bPot |                                                 | .std        |
| enti |                                                 | -ref}](#sch |
| al** |                                                 | ema-initmem |
|      |                                                 | bpotential) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **s  |                                                 | [[spikeT    |
| pike |                                                 | hresh]{.std |
| Thre |                                                 | .std-ref}   |
| sh** |                                                 | ](#schema-s |
|      |                                                 | pikethresh) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Children list

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **sp |                                                 | [[spe       |
| ecif |                                                 | cificCapaci |
| icCa |                                                 | tance]{.std |
| paci |                                                 | .std-r      |
| tanc |                                                 | ef}](#schem |
| es** |                                                 | a-specificc |
|      |                                                 | apacitance) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **p  |                                                 | [[baseC     |
| opul |                                                 | hannelPopul |
| atio |                                                 | ation]{.std |
| ns** |                                                 | .std-ref    |
|      |                                                 | }](#schema- |
|      |                                                 | basechannel |
|      |                                                 | population) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **ch |                                                 | [[ba        |
| anne |                                                 | seChannelDe |
| lDen |                                                 | nsity]{.std |
| siti |                                                 | .std-       |
| es** |                                                 | ref}](#sche |
|      |                                                 | ma-basechan |
|      |                                                 | neldensity) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **i  | *(from [[membraneProperties]{.std               | [[cu        |
| Ca** | .s                                              | rrent]{.std |
|      | td-ref}](#schema-membraneproperties){.reference | .std-r      |
|      | .internal})*                                    | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **iC |                                                 | [[cu        |
| a2** |                                                 | rrent]{.std |
|      |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   | *(from [[membraneProperties]{.std               | [[cu        |
| totC | .s                                              | rrent]{.std |
| hanC | td-ref}](#schema-membraneproperties){.reference | .std-r      |
| urre | .internal})*                                    | ef}](Landin |
| nt** |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   | *(from [[membraneProperties]{.std               | [[spe       |
| totS | .s                                              | cificCapaci |
| pecC | td-ref}](#schema-membraneproperties){.reference | tance]{.std |
| ap** | .internal})*                                    | .std-re     |
|      |                                                 | f}](Landing |
|      |                                                 | .html#schem |
|      |                                                 | a-dimension |
|      |                                                 | s-specificc |
|      |                                                 | apacitance) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **s  |                                                 | [           |
| urfa |                                                 | [area]{.std |
| ceAr |                                                 | .st         |
| ea** |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-area) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **s  | *(from [[membraneProperties]{.std               | [           |
| urfa | .s                                              | [area]{.std |
| ceAr | td-ref}](#schema-membraneproperties){.reference | .st         |
| ea** | .internal})*                                    | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-area) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***Derived Variables***

:   **totSpecCap** = specificCapacitances\[\*\]-\>specCap(reduce method:
    add) (exposed as **totSpecCap**)

:   **totChanPopCurrent** = populations\[\*\]-\>i(reduce method: add)

:   **totChanDensCurrentDensity**
    = channelDensities\[\*\]-\>iDensity(reduce method: add)

:   **totChanCurrent** = totChanPopCurrent + (totChanDensCurrentDensity
    \* surfaceArea) (exposed as **totChanCurrent**)

:   **totChanPopCurrentCa** = populations\[ion='ca'\]-\>i(reduce method:
    add)

:   **totChanDensCurrentDensityCa**
    = channelDensities\[ion='ca'\]-\>iDensity(reduce method: add)

:   **iCa** = totChanPopCurrentCa + (totChanDensCurrentDensityCa \*
    surfaceArea) (exposed as **iCa**)

:   **totChanPopCurrentCa2** = populations\[ion='ca2'\]-\>i(reduce
    method: add)

:   **totChanDensCurrentDensityCa2**
    = channelDensities\[ion='ca2'\]-\>iDensity(reduce method: add)

:   **iCa2** = totChanPopCurrentCa2 + (totChanDensCurrentDensityCa2 \*
    surfaceArea) (exposed as **iCa2**)
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="MembraneProperties2CaPools">
      <xs:complexContent>
        <xs:extension base="MembraneProperties">
          <xs:sequence>
            <xs:element name="channelDensityNernstCa2" type="ChannelDensityNernstCa2" minOccurs="0" maxOccurs="unbounded"/>
          </xs:sequence>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=MembraneProperties2CaPools){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import MembraneProperties2CaPools
    from neuroml.utils import component_factory

    variable = component_factory(
        MembraneProperties2CaPools,
        channel_populations: 'list of ChannelPopulation(s) (optional)' = None,
        channel_densities: 'list of ChannelDensity(s) (optional)' = None,
        channel_density_v_shifts: 'list of ChannelDensityVShift(s) (optional)' = None,
        channel_density_nernsts: 'list of ChannelDensityNernst(s) (optional)' = None,
        channel_density_ghks: 'list of ChannelDensityGHK(s) (optional)' = None,
        channel_density_ghk2s: 'list of ChannelDensityGHK2(s) (optional)' = None,
        channel_density_non_uniforms: 'list of ChannelDensityNonUniform(s) (optional)' = None,
        channel_density_non_uniform_nernsts: 'list of ChannelDensityNonUniformNernst(s) (optional)' = None,
        channel_density_non_uniform_ghks: 'list of ChannelDensityNonUniformGHK(s) (optional)' = None,
        spike_threshes: 'list of SpikeThresh(s) (required)' = None,
        specific_capacitances: 'list of SpecificCapacitance(s) (required)' = None,
        init_memb_potentials: 'list of InitMembPotential(s) (required)' = None,
        channel_density_nernst_ca2s: 'list of ChannelDensityNernstCa2(s) (optional)' = None,
    )
:::
:::
:::
:::
:::

::: {#biophysicalproperties .section}
[]{#schema-biophysicalproperties}

##### biophysicalProperties[\#](#biophysicalproperties "Link to this heading"){.headerlink}

*The biophysical properties of the [[cell]{.std
.std-ref}](#schema-cell){.reference .internal}, including the
[[membraneProperties]{.std
.std-ref}](#schema-membraneproperties){.reference .internal} and the
[[intracellularProperties]{.std
.std-ref}](#schema-intracellularproperties){.reference .internal}.*

::: {.sd-tab-set .docutils}
Child list

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   |                                                 | [[me        |
| memb |                                                 | mbranePrope |
| rane |                                                 | rties]{.std |
| Prop |                                                 | .std-       |
| erti |                                                 | ref}](#sche |
| es** |                                                 | ma-membrane |
|      |                                                 | properties) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **i  |                                                 | [[intrace   |
| ntra |                                                 | llularPrope |
| cell |                                                 | rties]{.std |
| ular |                                                 | .std-ref}]  |
| Prop |                                                 | (#schema-in |
| erti |                                                 | tracellular |
| es** |                                                 | properties) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   |                                                 | [[spe       |
| totS |                                                 | cificCapaci |
| pecC |                                                 | tance]{.std |
| ap** |                                                 | .std-re     |
|      |                                                 | f}](Landing |
|      |                                                 | .html#schem |
|      |                                                 | a-dimension |
|      |                                                 | s-specificc |
|      |                                                 | apacitance) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***Derived Variables***

:   **totSpecCap** = membraneProperties-\>totSpecCap (exposed as
    **totSpecCap**)
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="BiophysicalProperties">
      <xs:complexContent>
        <xs:extension base="Standalone">
          <xs:sequence>
            <xs:element name="membraneProperties" type="MembraneProperties"/>
            <xs:element name="intracellularProperties" type="IntracellularProperties" minOccurs="0"/>
            <xs:element name="extracellularProperties" type="ExtracellularProperties" minOccurs="0"/>
          </xs:sequence>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=BiophysicalProperties){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import BiophysicalProperties
    from neuroml.utils import component_factory

    variable = component_factory(
        BiophysicalProperties,
        id: 'a NmlId (required)' = None,
        metaid: 'a MetaId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        membrane_properties: 'a MembraneProperties (required)' = None,
        intracellular_properties: 'a IntracellularProperties (optional)' = None,
        extracellular_properties: 'a ExtracellularProperties (optional)' = None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <biophysicalProperties id="bio_cell">
        <membraneProperties>
            <channelPopulation id="naChansDend" ionChannel="NaConductance" segment="2" number="120000" erev="50mV" ion="na"/>
            <channelDensity id="pasChans" ionChannel="pas" condDensity="3.0 S_per_m2" erev="-70mV" ion="non_specific"/>
            <channelDensity id="naChansSoma" ionChannel="NaConductance" segmentGroup="soma_group" condDensity="120.0 mS_per_cm2" erev="50mV" ion="na"/>
            <specificCapacitance segmentGroup="soma_group" value="1.0 uF_per_cm2"/>
            <specificCapacitance segmentGroup="dendrite_group" value="2.0 uF_per_cm2"/>
        </membraneProperties>
        <intracellularProperties>
            <resistivity value="0.1 kohm_cm"/>
        </intracellularProperties>
    </biophysicalProperties>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <biophysicalProperties id="PyrCellChanDist">
        <membraneProperties>
            <channelDensity id="naChans" ionChannel="HH_Na" segmentGroup="soma_group" condDensity="120.0 mS_per_cm2" ion="na" erev="50mV"/>
            <!-- Ions present inside the cell. Note: a fixed reversal potential is specified here  
                <reversalPotential species="na" value="50mV"/>
                <reversalPotential species="k" value="-77mV"/>-->
        </membraneProperties>
        <intracellularProperties>
            <resistivity value="0.1 kohm_cm"/>
            <!-- REMOVED UNTIL WE CHECK HOW THE USAGE OF LEMS IMPACTS THIS...
                <biochemistry reactionScheme="InternalCaDynamics"/>  Ref to earlier pathway -->
        </intracellularProperties>
    </biophysicalProperties>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <biophysicalProperties id="biophys">
        <membraneProperties>
            <channelDensityNonUniform id="nonuniform_na_chans" ionChannel="NaConductance" erev="50mV" ion="na">
                <variableParameter parameter="condDensity" segmentGroup="dendrite_group">
                    <inhomogeneousValue inhomogeneousParameter="dendrite_group_x1" value="5e-7 * exp(-p/200)"/>
                </variableParameter>
            </channelDensityNonUniform>
            <specificCapacitance segmentGroup="soma_group" value="1.0 uF_per_cm2"/>
        </membraneProperties>
        <intracellularProperties>
            <resistivity value="0.1 kohm_cm"/>
        </intracellularProperties>
    </biophysicalProperties>
:::
:::
:::
:::
:::

::: {#biophysicalproperties2capools .section}
[]{#schema-biophysicalproperties2capools}

##### biophysicalProperties2CaPools[\#](#biophysicalproperties2capools "Link to this heading"){.headerlink}

*The biophysical properties of the [[cell]{.std
.std-ref}](#schema-cell){.reference .internal}, including the
[[membraneProperties2CaPools]{.std
.std-ref}](#schema-membraneproperties2capools){.reference .internal} and
the [[intracellularProperties2CaPools]{.std
.std-ref}](#schema-intracellularproperties2capools){.reference
.internal} for a cell with two Ca pools.*

::: {.sd-tab-set .docutils}
Child list

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   |                                                 | [           |
| memb |                                                 | [membranePr |
| rane |                                                 | operties2Ca |
| Prop |                                                 | Pools]{.std |
| erti |                                                 | .s          |
| es2C |                                                 | td-ref}](#s |
| aPoo |                                                 | chema-membr |
| ls** |                                                 | aneproperti |
|      |                                                 | es2capools) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **i  |                                                 | [[intr      |
| ntra |                                                 | acellularPr |
| cell |                                                 | operties2Ca |
| ular |                                                 | Pools]{.std |
| Prop |                                                 | .std-re     |
| erti |                                                 | f}](#schema |
| es2C |                                                 | -intracellu |
| aPoo |                                                 | larproperti |
| ls** |                                                 | es2capools) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   |                                                 | [[spe       |
| totS |                                                 | cificCapaci |
| pecC |                                                 | tance]{.std |
| ap** |                                                 | .std-re     |
|      |                                                 | f}](Landing |
|      |                                                 | .html#schem |
|      |                                                 | a-dimension |
|      |                                                 | s-specificc |
|      |                                                 | apacitance) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***Derived Variables***

:   **totSpecCap** = membraneProperties2CaPools-\>totSpecCap (exposed as
    **totSpecCap**)
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="BiophysicalProperties2CaPools">
      <xs:complexContent>
        <xs:extension base="Standalone">
          <xs:sequence>
            <xs:element name="membraneProperties2CaPools" type="MembraneProperties2CaPools"/>
            <xs:element name="intracellularProperties2CaPools" type="IntracellularProperties2CaPools" minOccurs="0"/>
            <xs:element name="extracellularProperties" type="ExtracellularProperties" minOccurs="0"/>
          </xs:sequence>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=BiophysicalProperties2CaPools){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import BiophysicalProperties2CaPools
    from neuroml.utils import component_factory

    variable = component_factory(
        BiophysicalProperties2CaPools,
        id: 'a NmlId (required)' = None,
        metaid: 'a MetaId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        membrane_properties2_ca_pools: 'a MembraneProperties2CaPools (required)' = None,
        intracellular_properties2_ca_pools: 'a IntracellularProperties2CaPools (optional)' = None,
        extracellular_properties: 'a ExtracellularProperties (optional)' = None,
    )
:::
:::
:::
:::
:::

::: {#intracellularproperties .section}
[]{#schema-intracellularproperties}

##### intracellularProperties[\#](#intracellularproperties "Link to this heading"){.headerlink}

*Biophysical properties related to the intracellular space within the
[[cell]{.std .std-ref}](#schema-cell){.reference .internal}, such as the
[[resistivity]{.std .std-ref}](#schema-resistivity){.reference
.internal} and the list of ionic [[species]{.std
.std-ref}](#schema-species){.reference .internal} present. **caConc**
and **caConcExt** are explicitly exposed here to facilitate accessing
these values from other Components, even though **caConcExt** is clearly
not an intracellular property.*

::: {.sd-tab-set .docutils}
Children list

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **r  |                                                 | [[resist    |
| esis |                                                 | ivity]{.std |
| tivi |                                                 | .std-ref}   |
| ty** |                                                 | ](#schema-r |
|      |                                                 | esistivity) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **s  |                                                 | [[sp        |
| peci |                                                 | ecies]{.std |
| esLi |                                                 | .std-       |
| st** |                                                 | ref}](#sche |
|      |                                                 | ma-species) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   |                                                 | [[concentr  |
| caCo |                                                 | ation]{.std |
| nc** |                                                 | .           |
|      |                                                 | std-ref}](L |
|      |                                                 | anding.html |
|      |                                                 | #schema-dim |
|      |                                                 | ensions-con |
|      |                                                 | centration) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    |                                                 | [[concentr  |
| *caC |                                                 | ation]{.std |
| oncE |                                                 | .           |
| xt** |                                                 | std-ref}](L |
|      |                                                 | anding.html |
|      |                                                 | #schema-dim |
|      |                                                 | ensions-con |
|      |                                                 | centration) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***Derived Variables***

:   **caConc** = speciesList\[ion='ca'\]-\>concentration(reduce method:
    add) (exposed as **caConc**)

:   **caConcExt** = speciesList\[ion='ca'\]-\>extConcentration(reduce
    method: add) (exposed as **caConcExt**)
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="IntracellularProperties">
      <xs:complexContent>
        <xs:extension base="BaseWithoutId">
          <xs:sequence>
            <xs:element name="species" type="Species" minOccurs="0" maxOccurs="unbounded"/>
            <xs:element name="resistivity" type="Resistivity" minOccurs="0" maxOccurs="unbounded"/>
          </xs:sequence>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=IntracellularProperties){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import IntracellularProperties
    from neuroml.utils import component_factory

    variable = component_factory(
        IntracellularProperties,
        species: 'list of Species(s) (optional)' = None,
        resistivities: 'list of Resistivity(s) (optional)' = None,
        extensiontype_=None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <intracellularProperties>
        <resistivity value="0.1 kohm_cm"/>
    </intracellularProperties>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <intracellularProperties>
        <resistivity value="0.1 kohm_cm"/>
        <!-- REMOVED UNTIL WE CHECK HOW THE USAGE OF LEMS IMPACTS THIS...
                <biochemistry reactionScheme="InternalCaDynamics"/>  Ref to earlier pathway -->
    </intracellularProperties>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <intracellularProperties>
        <resistivity value="0.1 kohm_cm"/>
    </intracellularProperties>
:::
:::
:::
:::
:::

::: {#intracellularproperties2capools .section}
[]{#schema-intracellularproperties2capools}

##### intracellularProperties2CaPools[\#](#intracellularproperties2capools "Link to this heading"){.headerlink}

extends [[intracellularProperties]{.std
.std-ref}](#schema-intracellularproperties){.reference .internal}

*Variant of intracellularProperties with 2 independent Ca pools.*

::: {.sd-tab-set .docutils}
Children list

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **s  |                                                 | [[sp        |
| peci |                                                 | ecies]{.std |
| esLi |                                                 | .std-       |
| st** |                                                 | ref}](#sche |
|      |                                                 | ma-species) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **r  |                                                 | [[resist    |
| esis |                                                 | ivity]{.std |
| tivi |                                                 | .std-ref}   |
| ty** |                                                 | ](#schema-r |
|      |                                                 | esistivity) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   | *(from [[intracellularProperties]{.std          | [[concentr  |
| caCo | .std-re                                         | ation]{.std |
| nc** | f}](#schema-intracellularproperties){.reference | .           |
|      | .internal})*                                    | std-ref}](L |
|      |                                                 | anding.html |
|      |                                                 | #schema-dim |
|      |                                                 | ensions-con |
|      |                                                 | centration) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **c  |                                                 | [[concentr  |
| aCon |                                                 | ation]{.std |
| c2** |                                                 | .           |
|      |                                                 | std-ref}](L |
|      |                                                 | anding.html |
|      |                                                 | #schema-dim |
|      |                                                 | ensions-con |
|      |                                                 | centration) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | *(from [[intracellularProperties]{.std          | [[concentr  |
| *caC | .std-re                                         | ation]{.std |
| oncE | f}](#schema-intracellularproperties){.reference | .           |
| xt** | .internal})*                                    | std-ref}](L |
|      |                                                 | anding.html |
|      |                                                 | #schema-dim |
|      |                                                 | ensions-con |
|      |                                                 | centration) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   |                                                 | [[concentr  |
| caCo |                                                 | ation]{.std |
| ncEx |                                                 | .           |
| t2** |                                                 | std-ref}](L |
|      |                                                 | anding.html |
|      |                                                 | #schema-dim |
|      |                                                 | ensions-con |
|      |                                                 | centration) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***Derived Variables***

:   **caConc2** = speciesList\[ion='ca2'\]-\>concentration(reduce
    method: add) (exposed as **caConc2**)

:   **caConcExt2** = speciesList\[ion='ca2'\]-\>extConcentration(reduce
    method: add) (exposed as **caConcExt2**)

:   **caConc** = speciesList\[ion='ca'\]-\>concentration(reduce method:
    add) (exposed as **caConc**)

:   **caConcExt** = speciesList\[ion='ca'\]-\>extConcentration(reduce
    method: add) (exposed as **caConcExt**)
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="IntracellularProperties2CaPools">
      <xs:complexContent>
        <xs:extension base="IntracellularProperties">
          </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=IntracellularProperties2CaPools){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import IntracellularProperties2CaPools
    from neuroml.utils import component_factory

    variable = component_factory(
        IntracellularProperties2CaPools,
        species: 'list of Species(s) (optional)' = None,
        resistivities: 'list of Resistivity(s) (optional)' = None,
    )
:::
:::
:::
:::
:::

::: {#resistivity .section}
[]{#schema-resistivity}

##### resistivity[\#](#resistivity "Link to this heading"){.headerlink}

*The resistivity, or specific axial resistance, of the cytoplasm.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    |                                                 | [[resist    |
| *val |                                                 | ivity]{.std |
| ue** |                                                 | .std-ref}]  |
|      |                                                 | (Landing.ht |
|      |                                                 | ml#schema-d |
|      |                                                 | imensions-r |
|      |                                                 | esistivity) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Text fields

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| **se   |                                                              |
| gmentG |                                                              |
| roup** |                                                              |
+--------+--------------------------------------------------------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="Resistivity">
      <xs:complexContent>
        <xs:extension base="BaseWithoutId">
          <xs:attribute name="value" type="Nml2Quantity_resistivity" use="required"/>
          <xs:attribute name="segmentGroup" type="NmlId" use="optional" default="all"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=Resistivity){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import Resistivity
    from neuroml.utils import component_factory

    variable = component_factory(
        Resistivity,
        value: 'a Nml2Quantity_resistivity (required)' = None,
        segment_groups: 'a NmlId (optional)' = 'all',
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <resistivity value="0.1 kohm_cm"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <resistivity value="0.1 kohm_cm"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <resistivity value="0.1 kohm_cm"/>
:::
:::
:::
:::
:::

::: {#concentrationmodel .section}
[]{#schema-concentrationmodel}

##### concentrationModel[\#](#concentrationmodel "Link to this heading"){.headerlink}

*Base for any model of an **ion** concentration which changes with time.
Internal ( **concentration** ) and external ( **extConcentration** )
values for the concentration of the ion are given.*

::: {.sd-tab-set .docutils}
Text fields

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| *      |                                                              |
| *ion** |                                                              |
+--------+--------------------------------------------------------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    |                                                 | [[concentr  |
| *con |                                                 | ation]{.std |
| cent |                                                 | .           |
| rati |                                                 | std-ref}](L |
| on** |                                                 | anding.html |
|      |                                                 | #schema-dim |
|      |                                                 | ensions-con |
|      |                                                 | centration) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **ex |                                                 | [[concentr  |
| tCon |                                                 | ation]{.std |
| cent |                                                 | .           |
| rati |                                                 | std-ref}](L |
| on** |                                                 | anding.html |
|      |                                                 | #schema-dim |
|      |                                                 | ensions-con |
|      |                                                 | centration) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **in |                                                 | [[concentr  |
| itia |                                                 | ation]{.std |
| lCon |                                                 | .           |
| cent |                                                 | std-ref}](L |
| rati |                                                 | anding.html |
| on** |                                                 | #schema-dim |
|      |                                                 | ensions-con |
|      |                                                 | centration) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **i  |                                                 | [[concentr  |
| niti |                                                 | ation]{.std |
| alEx |                                                 | .           |
| tCon |                                                 | std-ref}](L |
| cent |                                                 | anding.html |
| rati |                                                 | #schema-dim |
| on** |                                                 | ensions-con |
|      |                                                 | centration) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **s  |                                                 | [           |
| urfa |                                                 | [area]{.std |
| ceAr |                                                 | .st         |
| ea** |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-area) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***State Variables***

:   **concentration**: [[concentration]{.std
    .std-ref}](Landing.html#schema-dimensions-concentration){.reference
    .internal}  (exposed as **concentration**)

:   **extConcentration**: [[concentration]{.std
    .std-ref}](Landing.html#schema-dimensions-concentration){.reference
    .internal}  (exposed as **extConcentration**)

***On Start***

:   **concentration** = initialConcentration

:   **extConcentration** = initialExtConcentration
:::
:::
:::

::: {#decayingpoolconcentrationmodel .section}
[]{#schema-decayingpoolconcentrationmodel}

##### decayingPoolConcentrationModel[\#](#decayingpoolconcentrationmodel "Link to this heading"){.headerlink}

extends [[concentrationModel]{.std
.std-ref}](#schema-concentrationmodel){.reference .internal}

*Model of an intracellular buffering mechanism for **ion** ( currently
hard Coded to be calcium, due to requirement for **iCa** ) which has a
baseline level **restingConc** and tends to this value with time course
**decayConstant.** The ion is assumed to occupy a shell inside the
membrane of thickness **shellThickness.**.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    |                                                 | [           |
| *dec |                                                 | [time]{.std |
| ayCo |                                                 | .st         |
| nsta |                                                 | d-ref}](Lan |
| nt** |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **r  |                                                 | [[concentr  |
| esti |                                                 | ation]{.std |
| ngCo |                                                 | .           |
| nc** |                                                 | std-ref}](L |
|      |                                                 | anding.html |
|      |                                                 | #schema-dim |
|      |                                                 | ensions-con |
|      |                                                 | centration) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   |                                                 | [[l         |
| shel |                                                 | ength]{.std |
| lThi |                                                 | .std-       |
| ckne |                                                 | ref}](Landi |
| ss** |                                                 | ng.html#sch |
|      |                                                 | ema-dimensi |
|      |                                                 | ons-length) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Constants

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------------------+-----------------------------------+-------------+
| **Faraday** =      |                                   | [           |
| 96485.3C_per_mol   |                                   | [charge_per |
|                    |                                   | _mole]{.std |
|                    |                                   | .st         |
|                    |                                   | d-ref}](Lan |
|                    |                                   | ding.html#s |
|                    |                                   | chema-dimen |
|                    |                                   | sions-charg |
|                    |                                   | e-per-mole) |
|                    |                                   | {.reference |
|                    |                                   | .internal}  |
+--------------------+-----------------------------------+-------------+
| **AREA_SCALE** =   |                                   | [           |
| 1m2                |                                   | [area]{.std |
|                    |                                   | .st         |
|                    |                                   | d-ref}](Lan |
|                    |                                   | ding.html#s |
|                    |                                   | chema-dimen |
|                    |                                   | sions-area) |
|                    |                                   | {.reference |
|                    |                                   | .internal}  |
+--------------------+-----------------------------------+-------------+
| **LENGTH_SCALE** = |                                   | [[l         |
| 1m                 |                                   | ength]{.std |
|                    |                                   | .std-       |
|                    |                                   | ref}](Landi |
|                    |                                   | ng.html#sch |
|                    |                                   | ema-dimensi |
|                    |                                   | ons-length) |
|                    |                                   | {.reference |
|                    |                                   | .internal}  |
+--------------------+-----------------------------------+-------------+
:::
:::

Text fields

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| *      |                                                              |
| *ion** |                                                              |
+--------+--------------------------------------------------------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | *(from [[concentrationModel]{.std               | [[concentr  |
| *con | .s                                              | ation]{.std |
| cent | td-ref}](#schema-concentrationmodel){.reference | .           |
| rati | .internal})*                                    | std-ref}](L |
| on** |                                                 | anding.html |
|      |                                                 | #schema-dim |
|      |                                                 | ensions-con |
|      |                                                 | centration) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **ex | *(from [[concentrationModel]{.std               | [[concentr  |
| tCon | .s                                              | ation]{.std |
| cent | td-ref}](#schema-concentrationmodel){.reference | .           |
| rati | .internal})*                                    | std-ref}](L |
| on** |                                                 | anding.html |
|      |                                                 | #schema-dim |
|      |                                                 | ensions-con |
|      |                                                 | centration) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **i  |                                                 | [[cu        |
| Ca** |                                                 | rrent]{.std |
|      |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **in | *(from [[concentrationModel]{.std               | [[concentr  |
| itia | .s                                              | ation]{.std |
| lCon | td-ref}](#schema-concentrationmodel){.reference | .           |
| cent | .internal})*                                    | std-ref}](L |
| rati |                                                 | anding.html |
| on** |                                                 | #schema-dim |
|      |                                                 | ensions-con |
|      |                                                 | centration) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **i  | *(from [[concentrationModel]{.std               | [[concentr  |
| niti | .s                                              | ation]{.std |
| alEx | td-ref}](#schema-concentrationmodel){.reference | .           |
| tCon | .internal})*                                    | std-ref}](L |
| cent |                                                 | anding.html |
| rati |                                                 | #schema-dim |
| on** |                                                 | ensions-con |
|      |                                                 | centration) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **s  | *(from [[concentrationModel]{.std               | [           |
| urfa | .s                                              | [area]{.std |
| ceAr | td-ref}](#schema-concentrationmodel){.reference | .st         |
| ea** | .internal})*                                    | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-area) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***State Variables***

:   **concentration**: [[concentration]{.std
    .std-ref}](Landing.html#schema-dimensions-concentration){.reference
    .internal}  (exposed as **concentration**)

:   **extConcentration**: [[concentration]{.std
    .std-ref}](Landing.html#schema-dimensions-concentration){.reference
    .internal}  (exposed as **extConcentration**)

***On Start***

:   **concentration** = initialConcentration

:   **extConcentration** = initialExtConcentration

***On Conditions***

:   IF concentration \< 0 THEN

:      **concentration** = 0

***Derived Variables***

:   **effectiveRadius** = LENGTH_SCALE \* sqrt(surfaceArea/(AREA_SCALE
    \* (4 \* 3.14159)))

:   **innerRadius** = effectiveRadius - shellThickness

:   **shellVolume** = (4 \* (effectiveRadius \* effectiveRadius \*
    effectiveRadius) \* 3.14159 / 3) - (4 \* (innerRadius \* innerRadius
    \* innerRadius) \* 3.14159 / 3)

***Time Derivatives***

:   d **concentration** /dt = iCa / (2 \* Faraday \* shellVolume) -
    ((concentration - restingConc) / decayConstant)
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="DecayingPoolConcentrationModel">
      <xs:complexContent>
        <xs:extension base="Standalone">
          <xs:attribute name="ion" type="NmlId" use="required">
            <xs:annotation>
                </xs:annotation>
          </xs:attribute>
          <xs:attribute name="restingConc" type="Nml2Quantity_concentration" use="required"/>
          <xs:attribute name="decayConstant" type="Nml2Quantity_time" use="required"/>
          <xs:attribute name="shellThickness" type="Nml2Quantity_length" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=DecayingPoolConcentrationModel){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import DecayingPoolConcentrationModel
    from neuroml.utils import component_factory

    variable = component_factory(
        DecayingPoolConcentrationModel,
        id: 'a NmlId (required)' = None,
        metaid: 'a MetaId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        ion: 'a NmlId (required)' = None,
        resting_conc: 'a Nml2Quantity_concentration (required)' = None,
        decay_constant: 'a Nml2Quantity_time (required)' = None,
        shell_thickness: 'a Nml2Quantity_length (required)' = None,
        extensiontype_=None,
    )
:::
:::
:::
:::
:::

::: {#fixedfactorconcentrationmodel .section}
[]{#schema-fixedfactorconcentrationmodel}

##### fixedFactorConcentrationModel[\#](#fixedfactorconcentrationmodel "Link to this heading"){.headerlink}

extends [[concentrationModel]{.std
.std-ref}](#schema-concentrationmodel){.reference .internal}

*Model of buffering of concentration of an ion ( currently hard coded to
be calcium, due to requirement for **iCa** ) which has a baseline level
**restingConc** and tends to this value with time course
**decayConstant.** A fixed factor **rho** is used to scale the incoming
current *independently of the size of the compartment* to produce a
concentration change.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    |                                                 | [           |
| *dec |                                                 | [time]{.std |
| ayCo |                                                 | .st         |
| nsta |                                                 | d-ref}](Lan |
| nt** |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **r  |                                                 | [[concentr  |
| esti |                                                 | ation]{.std |
| ngCo |                                                 | .           |
| nc** |                                                 | std-ref}](L |
|      |                                                 | anding.html |
|      |                                                 | #schema-dim |
|      |                                                 | ensions-con |
|      |                                                 | centration) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **r  |                                                 | [[rho_f     |
| ho** |                                                 | actor]{.std |
|      |                                                 | .std-ref}   |
|      |                                                 | ](Landing.h |
|      |                                                 | tml#schema- |
|      |                                                 | dimensions- |
|      |                                                 | rho-factor) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Text fields

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| *      |                                                              |
| *ion** |                                                              |
+--------+--------------------------------------------------------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | *(from [[concentrationModel]{.std               | [[concentr  |
| *con | .s                                              | ation]{.std |
| cent | td-ref}](#schema-concentrationmodel){.reference | .           |
| rati | .internal})*                                    | std-ref}](L |
| on** |                                                 | anding.html |
|      |                                                 | #schema-dim |
|      |                                                 | ensions-con |
|      |                                                 | centration) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **ex | *(from [[concentrationModel]{.std               | [[concentr  |
| tCon | .s                                              | ation]{.std |
| cent | td-ref}](#schema-concentrationmodel){.reference | .           |
| rati | .internal})*                                    | std-ref}](L |
| on** |                                                 | anding.html |
|      |                                                 | #schema-dim |
|      |                                                 | ensions-con |
|      |                                                 | centration) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **i  |                                                 | [[cu        |
| Ca** |                                                 | rrent]{.std |
|      |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **in | *(from [[concentrationModel]{.std               | [[concentr  |
| itia | .s                                              | ation]{.std |
| lCon | td-ref}](#schema-concentrationmodel){.reference | .           |
| cent | .internal})*                                    | std-ref}](L |
| rati |                                                 | anding.html |
| on** |                                                 | #schema-dim |
|      |                                                 | ensions-con |
|      |                                                 | centration) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **i  | *(from [[concentrationModel]{.std               | [[concentr  |
| niti | .s                                              | ation]{.std |
| alEx | td-ref}](#schema-concentrationmodel){.reference | .           |
| tCon | .internal})*                                    | std-ref}](L |
| cent |                                                 | anding.html |
| rati |                                                 | #schema-dim |
| on** |                                                 | ensions-con |
|      |                                                 | centration) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **s  |                                                 | [           |
| urfa |                                                 | [area]{.std |
| ceAr |                                                 | .st         |
| ea** |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-area) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **s  | *(from [[concentrationModel]{.std               | [           |
| urfa | .s                                              | [area]{.std |
| ceAr | td-ref}](#schema-concentrationmodel){.reference | .st         |
| ea** | .internal})*                                    | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-area) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***State Variables***

:   **concentration**: [[concentration]{.std
    .std-ref}](Landing.html#schema-dimensions-concentration){.reference
    .internal}  (exposed as **concentration**)

:   **extConcentration**: [[concentration]{.std
    .std-ref}](Landing.html#schema-dimensions-concentration){.reference
    .internal}  (exposed as **extConcentration**)

***On Start***

:   **concentration** = initialConcentration

:   **extConcentration** = initialExtConcentration

***On Conditions***

:   IF concentration \< 0 THEN

:      **concentration** = 0

***Time Derivatives***

:   d **concentration** /dt = (iCa/surfaceArea) \* rho -
    ((concentration - restingConc) / decayConstant)
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="FixedFactorConcentrationModel">
      <xs:complexContent>
        <xs:extension base="Standalone">
          <xs:attribute name="ion" type="NmlId" use="required">
            <xs:annotation>
                </xs:annotation>
          </xs:attribute>
          <xs:attribute name="restingConc" type="Nml2Quantity_concentration" use="required"/>
          <xs:attribute name="decayConstant" type="Nml2Quantity_time" use="required"/>
          <xs:attribute name="rho" type="Nml2Quantity_rhoFactor" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=FixedFactorConcentrationModel){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import FixedFactorConcentrationModel
    from neuroml.utils import component_factory

    variable = component_factory(
        FixedFactorConcentrationModel,
        id: 'a NmlId (required)' = None,
        metaid: 'a MetaId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        ion: 'a NmlId (required)' = None,
        resting_conc: 'a Nml2Quantity_concentration (required)' = None,
        decay_constant: 'a Nml2Quantity_time (required)' = None,
        rho: 'a Nml2Quantity_rhoFactor (required)' = None,
    )
:::
:::
:::
:::
:::

::: {#fixedfactorconcentrationmodeltraub .section}
[]{#schema-fixedfactorconcentrationmodeltraub}

##### fixedFactorConcentrationModelTraub[\#](#fixedfactorconcentrationmodeltraub "Link to this heading"){.headerlink}

extends [[concentrationModel]{.std
.std-ref}](#schema-concentrationmodel){.reference .internal}

*Model of buffering of concentration of an ion ( currently hard coded to
be calcium, due to requirement for **iCa** ) which has a baseline level
**restingConc** and tends to this value with time course 1 / **beta.** A
fixed factor **phi** is used to scale the incoming current
*independently of the size of the compartment* to produce a
concentration change. Not recommended for use in models other than Traub
et al. 2005!*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **be |                                                 | [[per       |
| ta** |                                                 | _time]{.std |
|      |                                                 | .std-re     |
|      |                                                 | f}](Landing |
|      |                                                 | .html#schem |
|      |                                                 | a-dimension |
|      |                                                 | s-per-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **p  |                                                 | [[rho_f     |
| hi** |                                                 | actor]{.std |
|      |                                                 | .std-ref}   |
|      |                                                 | ](Landing.h |
|      |                                                 | tml#schema- |
|      |                                                 | dimensions- |
|      |                                                 | rho-factor) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **r  |                                                 | [[concentr  |
| esti |                                                 | ation]{.std |
| ngCo |                                                 | .           |
| nc** |                                                 | std-ref}](L |
|      |                                                 | anding.html |
|      |                                                 | #schema-dim |
|      |                                                 | ensions-con |
|      |                                                 | centration) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Text fields

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| **spe  |                                                              |
| cies** |                                                              |
+--------+--------------------------------------------------------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | *(from [[concentrationModel]{.std               | [[concentr  |
| *con | .s                                              | ation]{.std |
| cent | td-ref}](#schema-concentrationmodel){.reference | .           |
| rati | .internal})*                                    | std-ref}](L |
| on** |                                                 | anding.html |
|      |                                                 | #schema-dim |
|      |                                                 | ensions-con |
|      |                                                 | centration) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **ex | *(from [[concentrationModel]{.std               | [[concentr  |
| tCon | .s                                              | ation]{.std |
| cent | td-ref}](#schema-concentrationmodel){.reference | .           |
| rati | .internal})*                                    | std-ref}](L |
| on** |                                                 | anding.html |
|      |                                                 | #schema-dim |
|      |                                                 | ensions-con |
|      |                                                 | centration) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **i  |                                                 | [[cu        |
| Ca** |                                                 | rrent]{.std |
|      |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **in | *(from [[concentrationModel]{.std               | [[concentr  |
| itia | .s                                              | ation]{.std |
| lCon | td-ref}](#schema-concentrationmodel){.reference | .           |
| cent | .internal})*                                    | std-ref}](L |
| rati |                                                 | anding.html |
| on** |                                                 | #schema-dim |
|      |                                                 | ensions-con |
|      |                                                 | centration) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **i  | *(from [[concentrationModel]{.std               | [[concentr  |
| niti | .s                                              | ation]{.std |
| alEx | td-ref}](#schema-concentrationmodel){.reference | .           |
| tCon | .internal})*                                    | std-ref}](L |
| cent |                                                 | anding.html |
| rati |                                                 | #schema-dim |
| on** |                                                 | ensions-con |
|      |                                                 | centration) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **s  |                                                 | [           |
| urfa |                                                 | [area]{.std |
| ceAr |                                                 | .st         |
| ea** |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-area) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **s  | *(from [[concentrationModel]{.std               | [           |
| urfa | .s                                              | [area]{.std |
| ceAr | td-ref}](#schema-concentrationmodel){.reference | .st         |
| ea** | .internal})*                                    | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-area) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***State Variables***

:   **concentration**: [[concentration]{.std
    .std-ref}](Landing.html#schema-dimensions-concentration){.reference
    .internal}  (exposed as **concentration**)

:   **extConcentration**: [[concentration]{.std
    .std-ref}](Landing.html#schema-dimensions-concentration){.reference
    .internal}  (exposed as **extConcentration**)

***On Start***

:   **concentration** = initialConcentration

:   **extConcentration** = initialExtConcentration

***On Conditions***

:   IF concentration \< 0 THEN

:      **concentration** = 0

***Time Derivatives***

:   d **concentration** /dt = (iCa/surfaceArea) \* 1e-9 \* phi -
    ((concentration - restingConc) \* beta)
:::
:::
:::

::: {#species .section}
[]{#schema-species}

##### species[\#](#species "Link to this heading"){.headerlink}

*Description of a chemical species identified by **ion,** which has
internal, **concentration,** and external, **extConcentration** values
for its concentration.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **in |                                                 | [[concentr  |
| itia |                                                 | ation]{.std |
| lCon |                                                 | .           |
| cent |                                                 | std-ref}](L |
| rati |                                                 | anding.html |
| on** |                                                 | #schema-dim |
|      |                                                 | ensions-con |
|      |                                                 | centration) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **i  |                                                 | [[concentr  |
| niti |                                                 | ation]{.std |
| alEx |                                                 | .           |
| tCon |                                                 | std-ref}](L |
| cent |                                                 | anding.html |
| rati |                                                 | #schema-dim |
| on** |                                                 | ensions-con |
|      |                                                 | centration) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Text fields

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| *      |                                                              |
| *ion** |                                                              |
+--------+--------------------------------------------------------------+
| **se   |                                                              |
| gmentG |                                                              |
| roup** |                                                              |
+--------+--------------------------------------------------------------+
:::
:::

Component References

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   |                                                 | [[co        |
| conc |                                                 | ncentration |
| entr |                                                 | Model]{.std |
| atio |                                                 | .std-       |
| nMod |                                                 | ref}](#sche |
| el** |                                                 | ma-concentr |
|      |                                                 | ationmodel) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    |                                                 | [[concentr  |
| *con |                                                 | ation]{.std |
| cent |                                                 | .           |
| rati |                                                 | std-ref}](L |
| on** |                                                 | anding.html |
|      |                                                 | #schema-dim |
|      |                                                 | ensions-con |
|      |                                                 | centration) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **ex |                                                 | [[concentr  |
| tCon |                                                 | ation]{.std |
| cent |                                                 | .           |
| rati |                                                 | std-ref}](L |
| on** |                                                 | anding.html |
|      |                                                 | #schema-dim |
|      |                                                 | ensions-con |
|      |                                                 | centration) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***Structure***

:   CHILD INSTANCE: **concentrationModel**

***Derived Variables***

:   **concentration** = concentrationModel-\>concentration (exposed as
    **concentration**)

:   **extConcentration**
    = concentrationModel-\>extConcentration (exposed as
    **extConcentration**)
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="Species">
      <xs:complexContent>
        <xs:extension base="Base">
          <xs:attribute name="concentrationModel" type="NmlId" use="required"/>
          <xs:attribute name="ion" type="NmlId" use="optional">
            <xs:annotation>
            </xs:annotation>
          </xs:attribute>
          <xs:attribute name="initialConcentration" type="Nml2Quantity_concentration" use="required"/>
          <xs:attribute name="initialExtConcentration" type="Nml2Quantity_concentration" use="required"/>
          <xs:attribute name="segmentGroup" type="NmlId" use="optional" default="all"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=Species){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import Species
    from neuroml.utils import component_factory

    variable = component_factory(
        Species,
        id: 'a NmlId (required)' = None,
        concentration_model: 'a NmlId (required)' = None,
        ion: 'a NmlId (optional)' = None,
        initial_concentration: 'a Nml2Quantity_concentration (required)' = None,
        initial_ext_concentration: 'a Nml2Quantity_concentration (required)' = None,
        segment_groups: 'a NmlId (optional)' = 'all',
    )
:::
:::
:::
:::
:::

::: {#cell .section}
[]{#schema-cell}

##### cell[\#](#cell "Link to this heading"){.headerlink}

extends *[[baseCellMembPot]{.std
.std-ref}](#schema-basecellmembpot){.reference .internal}*

*Cell with [[segment]{.std .std-ref}](#schema-segment){.reference
.internal}s specified in a [[morphology]{.std
.std-ref}](#schema-morphology){.reference .internal} element along with
details on its [[biophysicalProperties]{.std
.std-ref}](#schema-biophysicalproperties){.reference .internal}. NOTE:
this can only be correctly simulated using jLEMS when there is a single
segment in the cell, and **v** of this cell represents the membrane
potential in that isopotential segment.*

::: {.sd-tab-set .docutils}
Text fields

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| **     |                                                              |
| neuroL |                                                              |
| exId** |                                                              |
+--------+--------------------------------------------------------------+
:::
:::

Child list

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   | Should only be used if morphology element is    | [[morph     |
| morp | outside the cell. This points to the id of the  | ology]{.std |
| holo | morphology.                                     | .std-ref    |
| gy** |                                                 | }](#schema- |
|      |                                                 | morphology) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | Should only be used if biophysicalProperties    | [[bioph     |
| *bio | element is outside the cell. This points to the | ysicalPrope |
| phys | id of the biophysicalProperties                 | rties]{.std |
| ical |                                                 | .std-ref    |
| Prop |                                                 | }](#schema- |
| erti |                                                 | biophysical |
| es** |                                                 | properties) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   |                                                 | [[concentr  |
| caCo |                                                 | ation]{.std |
| nc** |                                                 | .           |
|      |                                                 | std-ref}](L |
|      |                                                 | anding.html |
|      |                                                 | #schema-dim |
|      |                                                 | ensions-con |
|      |                                                 | centration) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    |                                                 | [[concentr  |
| *caC |                                                 | ation]{.std |
| oncE |                                                 | .           |
| xt** |                                                 | std-ref}](L |
|      |                                                 | anding.html |
|      |                                                 | #schema-dim |
|      |                                                 | ensions-con |
|      |                                                 | centration) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **i  |                                                 | [[cu        |
| Ca** |                                                 | rrent]{.std |
|      |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    |                                                 | [[cu        |
| *iCh |                                                 | rrent]{.std |
| anne |                                                 | .std-r      |
| ls** |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **iS |                                                 | [[cu        |
| yn** |                                                 | rrent]{.std |
|      |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **s  |                                                 | Di          |
| piki |                                                 | mensionless |
| ng** |                                                 |             |
+------+-------------------------------------------------+-------------+
| **s  |                                                 | [           |
| urfa |                                                 | [area]{.std |
| ceAr |                                                 | .st         |
| ea** |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-area) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   |                                                 | [[spe       |
| totS |                                                 | cificCapaci |
| pecC |                                                 | tance]{.std |
| ap** |                                                 | .std-re     |
|      |                                                 | f}](Landing |
|      |                                                 | .html#schem |
|      |                                                 | a-dimension |
|      |                                                 | s-specificc |
|      |                                                 | apacitance) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | Membrane potential *(from                       | [[vo        |
| *v** | [[baseCellMembPot]{.std                         | ltage]{.std |
|      | .std-ref}](#schema-basecellmembpot){.reference  | .std-r      |
|      | .internal})*                                    | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Spike event *(from [[baseSpikingCell]{.std      | Direction:  |
| *spi | .std-ref}](#schema-basespikingcell){.reference  | out         |
| ke** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Attachments

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **sy |                                                 | [[          |
| naps |                                                 | basePointCu |
| es** |                                                 | rrent]{.std |
|      |                                                 | .std        |
|      |                                                 | -ref}](Land |
|      |                                                 | ing.html#sc |
|      |                                                 | hema-basepo |
|      |                                                 | intcurrent) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***State Variables***

:   **v**: [[voltage]{.std
    .std-ref}](Landing.html#schema-dimensions-voltage){.reference
    .internal}  (exposed as **v**)

:   **spiking**: Dimensionless  (exposed as **spiking**)

***On Start***

:   **spiking** = 0

:   **v** = initMembPot

***On Conditions***

:   IF v \> thresh AND spiking \< 0.5 THEN

:      **spiking** = 1

:      EVENT OUT on port: **spike**

:   IF v \< thresh THEN

:      **spiking** = 0

***Derived Variables***

:   **initMembPot**
    = biophysicalProperties-\>membraneProperties-\>initMembPotential-\>value

:   **thresh**
    = biophysicalProperties-\>membraneProperties-\>spikeThresh-\>value

:   **surfaceArea** = morphology-\>segments\[\*\]-\>surfaceArea(reduce
    method: add) (exposed as **surfaceArea**)

:   **totSpecCap** = biophysicalProperties-\>totSpecCap (exposed as
    **totSpecCap**)

:   **totCap** = totSpecCap \* surfaceArea

:   **iChannels**
    = biophysicalProperties-\>membraneProperties-\>totChanCurrent (exposed
    as **iChannels**)

:   **iSyn** = synapses\[\*\]-\>i(reduce method: add) (exposed as
    **iSyn**)

:   **iCa** = biophysicalProperties-\>membraneProperties-\>iCa (exposed
    as **iCa**)

:   **caConc**
    = biophysicalProperties-\>intracellularProperties-\>caConc (exposed
    as **caConc**)

:   **caConcExt**
    = biophysicalProperties-\>intracellularProperties-\>caConcExt (exposed
    as **caConcExt**)

***Time Derivatives***

:   d **v** /dt = (iChannels + iSyn) / totCap
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="Cell">
      <xs:complexContent>
        <xs:extension base="BaseCell">
          <xs:sequence>
            <xs:element name="morphology" type="Morphology" minOccurs="0"/>
            <xs:element name="biophysicalProperties" type="BiophysicalProperties" minOccurs="0"/>
          </xs:sequence>
          <xs:attribute name="morphology" type="NmlId" use="optional">
            <xs:annotation>
                </xs:annotation>
          </xs:attribute>
          <xs:attribute name="biophysicalProperties" type="NmlId" use="optional">
            <xs:annotation>
                </xs:annotation>
          </xs:attribute>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=Cell){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import Cell
    from neuroml.utils import component_factory

    variable = component_factory(
        Cell,
        id: 'a NmlId (required)' = None,
        metaid: 'a MetaId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        neuro_lex_id: 'a NeuroLexId (optional)' = None,
        morphology_attr: 'a NmlId (optional)' = None,
        biophysical_properties_attr: 'a NmlId (optional)' = None,
        morphology: 'a Morphology (optional)' = None,
        biophysical_properties: 'a BiophysicalProperties (optional)' = None,
        extensiontype_=None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <cell id="SpikingCell" metaid="HippoCA1Cell">
        <notes>A Simple Spiking cell for testing purposes</notes>
        <annotation>
            <rdf:RDF xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#" xmlns:bqbiol="http://biomodels.net/biology-qualifiers/">
                <rdf:Description rdf:about="HippoCA1Cell">
                    <bqbiol:is>
                        <rdf:Bag>
                            <rdf:li rdf:resource="urn:miriam:neurondb:258"/>
                        </rdf:Bag>
                    </bqbiol:is>
                </rdf:Description>
            </rdf:RDF>
        </annotation>
        <morphology id="SpikingCell_morphology">
            <segment id="0" name="Soma">
                <proximal x="0" y="0" z="0" diameter="10"/>
                <distal x="10" y="0" z="0" diameter="10"/>
            </segment>
            <segment id="1" name="Dendrite1">
                <parent segment="0"/>
                <distal x="20" y="0" z="0" diameter="3"/>
            </segment>
            <segment id="2" name="Dendrite2">
                <parent segment="1"/>
                <distal x="30" y="0" z="0" diameter="1"/>
            </segment>
            <segment id="3" name="Spine1">
                <parent segment="2" fractionAlong="0.5"/>
                <proximal x="25" y="0" z="0" diameter="0.1"/>
                <distal x="25" y="0.2" z="0" diameter="0.1"/>
            </segment>
            <segmentGroup id="soma_group" neuroLexId="sao1044911821">
                <member segment="0"/>
            </segmentGroup>
            <segmentGroup id="dendrite_group" neuroLexId="sao1211023249">
                <member segment="1"/>
                <member segment="2"/>
                <member segment="3"/>
            </segmentGroup>
            <segmentGroup id="spines" neuroLexId="sao1145756102">
                <member segment="3"/>
            </segmentGroup>
        </morphology>
        <biophysicalProperties id="bio_cell">
            <membraneProperties>
                <channelPopulation id="naChansDend" ionChannel="NaConductance" segment="2" number="120000" erev="50mV" ion="na"/>
                <channelDensity id="pasChans" ionChannel="pas" condDensity="3.0 S_per_m2" erev="-70mV" ion="non_specific"/>
                <channelDensity id="naChansSoma" ionChannel="NaConductance" segmentGroup="soma_group" condDensity="120.0 mS_per_cm2" erev="50mV" ion="na"/>
                <specificCapacitance segmentGroup="soma_group" value="1.0 uF_per_cm2"/>
                <specificCapacitance segmentGroup="dendrite_group" value="2.0 uF_per_cm2"/>
            </membraneProperties>
            <intracellularProperties>
                <resistivity value="0.1 kohm_cm"/>
            </intracellularProperties>
        </biophysicalProperties>
    </cell>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <cell id="PyrCell" morphology="NeuroMorpho_PyrCell123" biophysicalProperties="PyrCellChanDist"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <cell id="SimpleCell">
        <morphology id="SimpleCell_Morphology">
            <segment id="0" name="Soma">
                <proximal x="0" y="0" z="0" diameter="10"/>
                <distal x="10" y="0" z="0" diameter="10"/>
            </segment>
            <segment id="1" name="MainDendrite1">
                <parent segment="0"/>
                <proximal x="10" y="0" z="0" diameter="3"/>
                <distal x="20" y="0" z="0" diameter="3"/>
            </segment>
            <segment id="2" name="MainDendrite2">
                <parent segment="1"/>
                <distal x="30" y="0" z="0" diameter="1"/>
            </segment>
            <segmentGroup id="soma_group" neuroLexId="sao1044911821">
                <member segment="0"/>
            </segmentGroup>
            <segmentGroup id="dendrite_group" neuroLexId="sao1211023249">
                <member segment="1"/>
                <member segment="2"/>
                <inhomogeneousParameter id="dendrite_group_x1" variable="p" metric="Path Length from root"/>
                <inhomogeneousParameter id="dendrite_group_x2" variable="r" metric="Path Length from root">
                    <proximal translationStart="0"/>
                    <distal normalizationEnd="1"/>
                </inhomogeneousParameter>
            </segmentGroup>
        </morphology>
        <biophysicalProperties id="biophys">
            <membraneProperties>
                <channelDensityNonUniform id="nonuniform_na_chans" ionChannel="NaConductance" erev="50mV" ion="na">
                    <variableParameter parameter="condDensity" segmentGroup="dendrite_group">
                        <inhomogeneousValue inhomogeneousParameter="dendrite_group_x1" value="5e-7 * exp(-p/200)"/>
                    </variableParameter>
                </channelDensityNonUniform>
                <specificCapacitance segmentGroup="soma_group" value="1.0 uF_per_cm2"/>
            </membraneProperties>
            <intracellularProperties>
                <resistivity value="0.1 kohm_cm"/>
            </intracellularProperties>
        </biophysicalProperties>
    </cell>
:::
:::
:::
:::
:::

::: {#cell2capools .section}
[]{#schema-cell2capools}

##### cell2CaPools[\#](#cell2capools "Link to this heading"){.headerlink}

extends [[cell]{.std .std-ref}](#schema-cell){.reference .internal}

*Variant of cell with two independent Ca2+ pools. Cell with
[[segment]{.std .std-ref}](#schema-segment){.reference .internal}s
specified in a [[morphology]{.std
.std-ref}](#schema-morphology){.reference .internal} element along with
details on its [[biophysicalProperties]{.std
.std-ref}](#schema-biophysicalproperties){.reference .internal}. NOTE:
this can only be correctly simulated using jLEMS when there is a single
segment in the cell, and **v** of this cell represents the membrane
potential in that isopotential segment.*

::: {.sd-tab-set .docutils}
Text fields

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| **     |                                                              |
| neuroL |                                                              |
| exId** |                                                              |
+--------+--------------------------------------------------------------+
:::
:::

Child list

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    |                                                 | [[bi        |
| *bio |                                                 | ophysicalPr |
| phys |                                                 | operties2Ca |
| ical |                                                 | Pools]{.std |
| Prop |                                                 | .std-       |
| erti |                                                 | ref}](#sche |
| es2C |                                                 | ma-biophysi |
| aPoo |                                                 | calproperti |
| ls** |                                                 | es2capools) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   | *(from [[cell]{.std                             | [[concentr  |
| caCo | .std-ref}](#schema-cell){.reference             | ation]{.std |
| nc** | .internal})*                                    | .           |
|      |                                                 | std-ref}](L |
|      |                                                 | anding.html |
|      |                                                 | #schema-dim |
|      |                                                 | ensions-con |
|      |                                                 | centration) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **c  |                                                 | [[concentr  |
| aCon |                                                 | ation]{.std |
| c2** |                                                 | .           |
|      |                                                 | std-ref}](L |
|      |                                                 | anding.html |
|      |                                                 | #schema-dim |
|      |                                                 | ensions-con |
|      |                                                 | centration) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | *(from [[cell]{.std                             | [[concentr  |
| *caC | .std-ref}](#schema-cell){.reference             | ation]{.std |
| oncE | .internal})*                                    | .           |
| xt** |                                                 | std-ref}](L |
|      |                                                 | anding.html |
|      |                                                 | #schema-dim |
|      |                                                 | ensions-con |
|      |                                                 | centration) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   |                                                 | [[concentr  |
| caCo |                                                 | ation]{.std |
| ncEx |                                                 | .           |
| t2** |                                                 | std-ref}](L |
|      |                                                 | anding.html |
|      |                                                 | #schema-dim |
|      |                                                 | ensions-con |
|      |                                                 | centration) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **i  | *(from [[cell]{.std                             | [[cu        |
| Ca** | .std-ref}](#schema-cell){.reference             | rrent]{.std |
|      | .internal})*                                    | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **iC |                                                 | [[cu        |
| a2** |                                                 | rrent]{.std |
|      |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | *(from [[cell]{.std                             | [[cu        |
| *iCh | .std-ref}](#schema-cell){.reference             | rrent]{.std |
| anne | .internal})*                                    | .std-r      |
| ls** |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **iS | *(from [[cell]{.std                             | [[cu        |
| yn** | .std-ref}](#schema-cell){.reference             | rrent]{.std |
|      | .internal})*                                    | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **s  | *(from [[cell]{.std                             | Di          |
| piki | .std-ref}](#schema-cell){.reference             | mensionless |
| ng** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| **s  | *(from [[cell]{.std                             | [           |
| urfa | .std-ref}](#schema-cell){.reference             | [area]{.std |
| ceAr | .internal})*                                    | .st         |
| ea** |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-area) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   | *(from [[cell]{.std                             | [[spe       |
| totS | .std-ref}](#schema-cell){.reference             | cificCapaci |
| pecC | .internal})*                                    | tance]{.std |
| ap** |                                                 | .std-re     |
|      |                                                 | f}](Landing |
|      |                                                 | .html#schem |
|      |                                                 | a-dimension |
|      |                                                 | s-specificc |
|      |                                                 | apacitance) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | Membrane potential *(from                       | [[vo        |
| *v** | [[baseCellMembPot]{.std                         | ltage]{.std |
|      | .std-ref}](#schema-basecellmembpot){.reference  | .std-r      |
|      | .internal})*                                    | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Spike event *(from [[baseSpikingCell]{.std      | Direction:  |
| *spi | .std-ref}](#schema-basespikingcell){.reference  | out         |
| ke** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Attachments

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **sy |                                                 | [[          |
| naps |                                                 | basePointCu |
| es** |                                                 | rrent]{.std |
|      |                                                 | .std        |
|      |                                                 | -ref}](Land |
|      |                                                 | ing.html#sc |
|      |                                                 | hema-basepo |
|      |                                                 | intcurrent) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***State Variables***

:   **v**: [[voltage]{.std
    .std-ref}](Landing.html#schema-dimensions-voltage){.reference
    .internal}  (exposed as **v**)

:   **spiking**: Dimensionless  (exposed as **spiking**)

***On Start***

:   **spiking** = 0

:   **v** = initMembPot

***On Conditions***

:   IF v \> thresh AND spiking \< 0.5 THEN

:      **spiking** = 1

:      EVENT OUT on port: **spike**

:   IF v \< thresh THEN

:      **spiking** = 0

***Derived Variables***

:   **initMembPot**
    = biophysicalProperties2CaPools-\>membraneProperties2CaPools-\>initMembPotential-\>value

:   **thresh**
    = biophysicalProperties2CaPools-\>membraneProperties2CaPools-\>spikeThresh-\>value

:   **surfaceArea** = morphology-\>segments\[\*\]-\>surfaceArea(reduce
    method: add) (exposed as **surfaceArea**)

:   **totSpecCap** = biophysicalProperties2CaPools-\>totSpecCap (exposed
    as **totSpecCap**)

:   **totCap** = totSpecCap \* surfaceArea

:   **iChannels**
    = biophysicalProperties2CaPools-\>membraneProperties2CaPools-\>totChanCurrent (exposed
    as **iChannels**)

:   **iSyn** = synapses\[\*\]-\>i(reduce method: add) (exposed as
    **iSyn**)

:   **iCa**
    = biophysicalProperties2CaPools-\>membraneProperties2CaPools-\>iCa (exposed
    as **iCa**)

:   **caConc**
    = biophysicalProperties2CaPools-\>intracellularProperties2CaPools-\>caConc (exposed
    as **caConc**)

:   **caConcExt**
    = biophysicalProperties2CaPools-\>intracellularProperties2CaPools-\>caConcExt (exposed
    as **caConcExt**)

:   **iCa2**
    = biophysicalProperties2CaPools-\>membraneProperties2CaPools-\>iCa2 (exposed
    as **iCa2**)

:   **caConc2**
    = biophysicalProperties2CaPools-\>intracellularProperties2CaPools-\>caConc2 (exposed
    as **caConc2**)

:   **caConcExt2**
    = biophysicalProperties2CaPools-\>intracellularProperties2CaPools-\>caConcExt2 (exposed
    as **caConcExt2**)

***Time Derivatives***

:   d **v** /dt = (iChannels + iSyn) / totCap
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="Cell2CaPools">
      <xs:complexContent>
        <xs:extension base="Cell">
          <xs:sequence>
            <xs:element name="biophysicalProperties2CaPools" type="BiophysicalProperties2CaPools" minOccurs="0"/>
          </xs:sequence>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=Cell2CaPools){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import Cell2CaPools
    from neuroml.utils import component_factory

    variable = component_factory(
        Cell2CaPools,
        id: 'a NmlId (required)' = None,
        metaid: 'a MetaId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        neuro_lex_id: 'a NeuroLexId (optional)' = None,
        morphology_attr: 'a NmlId (optional)' = None,
        biophysical_properties_attr: 'a NmlId (optional)' = None,
        morphology: 'a Morphology (optional)' = None,
        biophysical_properties: 'a BiophysicalProperties (optional)' = None,
        biophysical_properties2_ca_pools: 'a BiophysicalProperties2CaPools (optional)' = None,
    )
:::
:::
:::
:::
:::

::: {#basecellmembpotcap .section}
[]{#schema-basecellmembpotcap}

##### *baseCellMembPotCap*[\#](#basecellmembpotcap "Link to this heading"){.headerlink}

extends *[[baseCellMembPot]{.std
.std-ref}](#schema-basecellmembpot){.reference .internal}*

*Any cell with a membrane potential **v** with voltage units and a
membrane capacitance **C.** Also defines exposed value **iSyn** for
current due to external synapses and **iMemb** for total transmembrane
current ( usually channel currents plus **iSyn** ).*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Total capacitance of the cell membrane          | [[capaci    |
| *C** |                                                 | tance]{.std |
|      |                                                 | .std-ref}]  |
|      |                                                 | (Landing.ht |
|      |                                                 | ml#schema-d |
|      |                                                 | imensions-c |
|      |                                                 | apacitance) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Total current crossing the cell membrane        | [[cu        |
| *iMe |                                                 | rrent]{.std |
| mb** |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **iS | Total current due to synaptic inputs            | [[cu        |
| yn** |                                                 | rrent]{.std |
|      |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | Membrane potential *(from                       | [[vo        |
| *v** | [[baseCellMembPot]{.std                         | ltage]{.std |
|      | .std-ref}](#schema-basecellmembpot){.reference  | .std-r      |
|      | .internal})*                                    | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Spike event *(from [[baseSpikingCell]{.std      | Direction:  |
| *spi | .std-ref}](#schema-basespikingcell){.reference  | out         |
| ke** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="BaseCellMembPotCap">
      <xs:complexContent>
        <xs:extension base="BaseCell">
          <xs:attribute name="C" type="Nml2Quantity_capacitance" use="required">
            <xs:annotation>
              <xs:appinfo>
                <jxb:property name="Cap"/>
              </xs:appinfo>
            </xs:annotation>
          </xs:attribute>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=BaseCellMembPotCap){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import BaseCellMembPotCap
    from neuroml.utils import component_factory

    variable = component_factory(
        BaseCellMembPotCap,
        id: 'a NmlId (required)' = None,
        metaid: 'a MetaId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        neuro_lex_id: 'a NeuroLexId (optional)' = None,
        C: 'a Nml2Quantity_capacitance (required)' = None,
        extensiontype_=None,
    )
:::
:::
:::
:::
:::

::: {#baseiaf .section}
[]{#schema-baseiaf}

##### *baseIaf*[\#](#baseiaf "Link to this heading"){.headerlink}

extends *[[baseCellMembPot]{.std
.std-ref}](#schema-basecellmembpot){.reference .internal}*

*Base ComponentType for an integrate and fire cell which emits a spiking
event at membrane potential **thresh** and and resets to **reset**.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | The value the membrane potential is reset to on | [[vo        |
| *res | spiking                                         | ltage]{.std |
| et** |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   | The membrane potential at which to emit a       | [[vo        |
| thre | spiking event and reset voltage                 | ltage]{.std |
| sh** |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Membrane potential *(from                       | [[vo        |
| *v** | [[baseCellMembPot]{.std                         | ltage]{.std |
|      | .std-ref}](#schema-basecellmembpot){.reference  | .std-r      |
|      | .internal})*                                    | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Spike event *(from [[baseSpikingCell]{.std      | Direction:  |
| *spi | .std-ref}](#schema-basespikingcell){.reference  | out         |
| ke** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::
:::
:::

::: {#iaftaucell .section}
[]{#schema-iaftaucell}

##### iafTauCell[\#](#iaftaucell "Link to this heading"){.headerlink}

extends *[[baseIaf]{.std .std-ref}](#schema-baseiaf){.reference
.internal}*

*Integrate and fire cell which returns to its leak reversal potential of
**leakReversal** with a time constant **tau**.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **le |                                                 | [[vo        |
| akRe |                                                 | ltage]{.std |
| vers |                                                 | .std-r      |
| al** |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | The value the membrane potential is reset to on | [[vo        |
| *res | spiking *(from [[baseIaf]{.std                  | ltage]{.std |
| et** | .std-ref}](#schema-baseiaf){.reference          | .std-r      |
|      | .internal})*                                    | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **t  |                                                 | [           |
| au** |                                                 | [time]{.std |
|      |                                                 | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   | The membrane potential at which to emit a       | [[vo        |
| thre | spiking event and reset voltage *(from          | ltage]{.std |
| sh** | [[baseIaf]{.std                                 | .std-r      |
|      | .std-ref}](#schema-baseiaf){.reference          | ef}](Landin |
|      | .internal})*                                    | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Membrane potential *(from                       | [[vo        |
| *v** | [[baseCellMembPot]{.std                         | ltage]{.std |
|      | .std-ref}](#schema-basecellmembpot){.reference  | .std-r      |
|      | .internal})*                                    | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Spike event *(from [[baseSpikingCell]{.std      | Direction:  |
| *spi | .std-ref}](#schema-basespikingcell){.reference  | out         |
| ke** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***State Variables***

:   **v**: [[voltage]{.std
    .std-ref}](Landing.html#schema-dimensions-voltage){.reference
    .internal}  (exposed as **v**)

***On Start***

:   **v** = leakReversal

***On Conditions***

:   IF v \> thresh THEN

:      **v** = reset

:      EVENT OUT on port: **spike**

***Time Derivatives***

:   d **v** /dt = (leakReversal - v) / tau
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="IafTauCell">
      <xs:complexContent>
        <xs:extension base="BaseCell">
          <xs:attribute name="leakReversal" type="Nml2Quantity_voltage" use="required"/>
          <xs:attribute name="thresh" type="Nml2Quantity_voltage" use="required"/>
          <xs:attribute name="reset" type="Nml2Quantity_voltage" use="required"/>
          <xs:attribute name="tau" type="Nml2Quantity_time" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=IafTauCell){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import IafTauCell
    from neuroml.utils import component_factory

    variable = component_factory(
        IafTauCell,
        id: 'a NmlId (required)' = None,
        metaid: 'a MetaId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        neuro_lex_id: 'a NeuroLexId (optional)' = None,
        leak_reversal: 'a Nml2Quantity_voltage (required)' = None,
        thresh: 'a Nml2Quantity_voltage (required)' = None,
        reset: 'a Nml2Quantity_voltage (required)' = None,
        tau: 'a Nml2Quantity_time (required)' = None,
        extensiontype_=None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <iafTauCell id="iafTau" leakReversal="-50mV" thresh="-55mV" reset="-70mV" tau="30ms"/>
:::
:::
:::
:::
:::

::: {#iaftaurefcell .section}
[]{#schema-iaftaurefcell}

##### iafTauRefCell[\#](#iaftaurefcell "Link to this heading"){.headerlink}

extends [[iafTauCell]{.std .std-ref}](#schema-iaftaucell){.reference
.internal}

*Integrate and fire cell which returns to its leak reversal potential of
**leakReversal** with a time course **tau.** It has a refractory period
of **refract** after spiking.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **le | *(from [[iafTauCell]{.std                       | [[vo        |
| akRe | .std-ref}](#schema-iaftaucell){.reference       | ltage]{.std |
| vers | .internal})*                                    | .std-r      |
| al** |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **r  |                                                 | [           |
| efra |                                                 | [time]{.std |
| ct** |                                                 | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | The value the membrane potential is reset to on | [[vo        |
| *res | spiking *(from [[baseIaf]{.std                  | ltage]{.std |
| et** | .std-ref}](#schema-baseiaf){.reference          | .std-r      |
|      | .internal})*                                    | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **t  | *(from [[iafTauCell]{.std                       | [           |
| au** | .std-ref}](#schema-iaftaucell){.reference       | [time]{.std |
|      | .internal})*                                    | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   | The membrane potential at which to emit a       | [[vo        |
| thre | spiking event and reset voltage *(from          | ltage]{.std |
| sh** | [[baseIaf]{.std                                 | .std-r      |
|      | .std-ref}](#schema-baseiaf){.reference          | ef}](Landin |
|      | .internal})*                                    | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Membrane potential *(from                       | [[vo        |
| *v** | [[baseCellMembPot]{.std                         | ltage]{.std |
|      | .std-ref}](#schema-basecellmembpot){.reference  | .std-r      |
|      | .internal})*                                    | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Spike event *(from [[baseSpikingCell]{.std      | Direction:  |
| *spi | .std-ref}](#schema-basespikingcell){.reference  | out         |
| ke** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***State Variables***

:   **v**: [[voltage]{.std
    .std-ref}](Landing.html#schema-dimensions-voltage){.reference
    .internal}  (exposed as **v**)

:   **lastSpikeTime**: [[time]{.std
    .std-ref}](Landing.html#schema-dimensions-time){.reference
    .internal}

***On Start***

:   **v** = leakReversal

***Regime: refractory (initial)***

:   ***On Entry***

:      **lastSpikeTime** = t

:      **v** = reset

:   ***On Conditions***

:      IF t \> lastSpikeTime + refract THEN

:       TRANSITION to REGIME **integrating**

***Regime: integrating (initial)***

:   ***On Conditions***

:      IF v \> thresh THEN

:       EVENT OUT on port: **spike**

:       TRANSITION to REGIME **refractory**

:   ***Time Derivatives***

:      d **v** /dt = (leakReversal - v) / tau
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="IafTauRefCell">
      <xs:complexContent>
        <xs:extension base="IafTauCell">
          <xs:attribute name="refract" type="Nml2Quantity_time" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=IafTauRefCell){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import IafTauRefCell
    from neuroml.utils import component_factory

    variable = component_factory(
        IafTauRefCell,
        id: 'a NmlId (required)' = None,
        metaid: 'a MetaId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        neuro_lex_id: 'a NeuroLexId (optional)' = None,
        leak_reversal: 'a Nml2Quantity_voltage (required)' = None,
        thresh: 'a Nml2Quantity_voltage (required)' = None,
        reset: 'a Nml2Quantity_voltage (required)' = None,
        tau: 'a Nml2Quantity_time (required)' = None,
        refract: 'a Nml2Quantity_time (required)' = None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <iafTauRefCell id="iafTauRef" leakReversal="-50mV" thresh="-55mV" reset="-70mV" tau="30ms" refract="5ms"/>
:::
:::
:::
:::
:::

::: {#baseiafcapcell .section}
[]{#schema-baseiafcapcell}

##### *baseIafCapCell*[\#](#baseiafcapcell "Link to this heading"){.headerlink}

extends *[[baseCellMembPotCap]{.std
.std-ref}](#schema-basecellmembpotcap){.reference .internal}*

*Base Type for all Integrate and Fire cells with a capacitance **C,**
threshold **thresh** and reset membrane potential **reset**.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Total capacitance of the cell membrane *(from   | [[capaci    |
| *C** | [[baseCellMembPotCap]{.std                      | tance]{.std |
|      | .s                                              | .std-ref}]  |
|      | td-ref}](#schema-basecellmembpotcap){.reference | (Landing.ht |
|      | .internal})*                                    | ml#schema-d |
|      |                                                 | imensions-c |
|      |                                                 | apacitance) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    |                                                 | [[vo        |
| *res |                                                 | ltage]{.std |
| et** |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   |                                                 | [[vo        |
| thre |                                                 | ltage]{.std |
| sh** |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Total current crossing the cell membrane *(from | [[cu        |
| *iMe | [[baseCellMembPotCap]{.std                      | rrent]{.std |
| mb** | .s                                              | .std-r      |
|      | td-ref}](#schema-basecellmembpotcap){.reference | ef}](Landin |
|      | .internal})*                                    | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **iS | Total current due to synaptic inputs *(from     | [[cu        |
| yn** | [[baseCellMembPotCap]{.std                      | rrent]{.std |
|      | .s                                              | .std-r      |
|      | td-ref}](#schema-basecellmembpotcap){.reference | ef}](Landin |
|      | .internal})*                                    | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | Membrane potential *(from                       | [[vo        |
| *v** | [[baseCellMembPot]{.std                         | ltage]{.std |
|      | .std-ref}](#schema-basecellmembpot){.reference  | .std-r      |
|      | .internal})*                                    | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Spike event *(from [[baseSpikingCell]{.std      | Direction:  |
| *spi | .std-ref}](#schema-basespikingcell){.reference  | out         |
| ke** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::
:::
:::

::: {#iafcell .section}
[]{#schema-iafcell}

##### iafCell[\#](#iafcell "Link to this heading"){.headerlink}

extends *[[baseIafCapCell]{.std
.std-ref}](#schema-baseiafcapcell){.reference .internal}*

*Integrate and fire cell with capacitance **C,** **leakConductance** and
**leakReversal**.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Total capacitance of the cell membrane *(from   | [[capaci    |
| *C** | [[baseCellMembPotCap]{.std                      | tance]{.std |
|      | .s                                              | .std-ref}]  |
|      | td-ref}](#schema-basecellmembpotcap){.reference | (Landing.ht |
|      | .internal})*                                    | ml#schema-d |
|      |                                                 | imensions-c |
|      |                                                 | apacitance) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **l  |                                                 | [[conduc    |
| eakC |                                                 | tance]{.std |
| ondu |                                                 | .std-ref}]  |
| ctan |                                                 | (Landing.ht |
| ce** |                                                 | ml#schema-d |
|      |                                                 | imensions-c |
|      |                                                 | onductance) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **le |                                                 | [[vo        |
| akRe |                                                 | ltage]{.std |
| vers |                                                 | .std-r      |
| al** |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseIafCapCell]{.std                   | [[vo        |
| *res | .std-ref}](#schema-baseiafcapcell){.reference   | ltage]{.std |
| et** | .internal})*                                    | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   | *(from [[baseIafCapCell]{.std                   | [[vo        |
| thre | .std-ref}](#schema-baseiafcapcell){.reference   | ltage]{.std |
| sh** | .internal})*                                    | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Total current crossing the cell membrane *(from | [[cu        |
| *iMe | [[baseCellMembPotCap]{.std                      | rrent]{.std |
| mb** | .s                                              | .std-r      |
|      | td-ref}](#schema-basecellmembpotcap){.reference | ef}](Landin |
|      | .internal})*                                    | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **iS | Total current due to synaptic inputs *(from     | [[cu        |
| yn** | [[baseCellMembPotCap]{.std                      | rrent]{.std |
|      | .s                                              | .std-r      |
|      | td-ref}](#schema-basecellmembpotcap){.reference | ef}](Landin |
|      | .internal})*                                    | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | Membrane potential *(from                       | [[vo        |
| *v** | [[baseCellMembPot]{.std                         | ltage]{.std |
|      | .std-ref}](#schema-basecellmembpot){.reference  | .std-r      |
|      | .internal})*                                    | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Spike event *(from [[baseSpikingCell]{.std      | Direction:  |
| *spi | .std-ref}](#schema-basespikingcell){.reference  | out         |
| ke** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Attachments

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **sy |                                                 | [[          |
| naps |                                                 | basePointCu |
| es** |                                                 | rrent]{.std |
|      |                                                 | .std        |
|      |                                                 | -ref}](Land |
|      |                                                 | ing.html#sc |
|      |                                                 | hema-basepo |
|      |                                                 | intcurrent) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***State Variables***

:   **v**: [[voltage]{.std
    .std-ref}](Landing.html#schema-dimensions-voltage){.reference
    .internal}  (exposed as **v**)

***On Start***

:   **v** = leakReversal

***On Conditions***

:   IF v \> thresh THEN

:      **v** = reset

:      EVENT OUT on port: **spike**

***Derived Variables***

:   **iSyn** = synapses\[\*\]-\>i(reduce method: add) (exposed as
    **iSyn**)

:   **iMemb** = leakConductance \* (leakReversal - v) + iSyn (exposed as
    **iMemb**)

***Time Derivatives***

:   d **v** /dt = iMemb / C
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="IafCell">
      <xs:complexContent>
        <xs:extension base="BaseCell">
          <xs:attribute name="leakReversal" type="Nml2Quantity_voltage" use="required"/>
          <xs:attribute name="thresh" type="Nml2Quantity_voltage" use="required"/>
          <xs:attribute name="reset" type="Nml2Quantity_voltage" use="required"/>
          <xs:attribute name="C" type="Nml2Quantity_capacitance" use="required"/>
          <xs:attribute name="leakConductance" type="Nml2Quantity_conductance" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=IafCell){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import IafCell
    from neuroml.utils import component_factory

    variable = component_factory(
        IafCell,
        id: 'a NmlId (required)' = None,
        metaid: 'a MetaId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        neuro_lex_id: 'a NeuroLexId (optional)' = None,
        leak_reversal: 'a Nml2Quantity_voltage (required)' = None,
        thresh: 'a Nml2Quantity_voltage (required)' = None,
        reset: 'a Nml2Quantity_voltage (required)' = None,
        C: 'a Nml2Quantity_capacitance (required)' = None,
        leak_conductance: 'a Nml2Quantity_conductance (required)' = None,
        extensiontype_=None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <iafCell id="iaf" leakReversal="-50mV" thresh="-55mV" reset="-70mV" C="0.2nF" leakConductance="0.01uS"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <iafCell id="iaf" leakConductance="0.2nS" leakReversal="-70mV" thresh="-55mV" reset="-70mV" C="3.2pF"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <iafCell id="iaf" leakConductance="0.2nS" leakReversal="-70mV" thresh="-55mV" reset="-70mV" C="3.2pF"/>
:::
:::
:::
:::
:::

::: {#iafrefcell .section}
[]{#schema-iafrefcell}

##### iafRefCell[\#](#iafrefcell "Link to this heading"){.headerlink}

extends [[iafCell]{.std .std-ref}](#schema-iafcell){.reference
.internal}

*Integrate and fire cell with capacitance **C,** **leakConductance,**
**leakReversal** and refractory period **refract**.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Total capacitance of the cell membrane *(from   | [[capaci    |
| *C** | [[baseCellMembPotCap]{.std                      | tance]{.std |
|      | .s                                              | .std-ref}]  |
|      | td-ref}](#schema-basecellmembpotcap){.reference | (Landing.ht |
|      | .internal})*                                    | ml#schema-d |
|      |                                                 | imensions-c |
|      |                                                 | apacitance) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **l  | *(from [[iafCell]{.std                          | [[conduc    |
| eakC | .std-ref}](#schema-iafcell){.reference          | tance]{.std |
| ondu | .internal})*                                    | .std-ref}]  |
| ctan |                                                 | (Landing.ht |
| ce** |                                                 | ml#schema-d |
|      |                                                 | imensions-c |
|      |                                                 | onductance) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **le | *(from [[iafCell]{.std                          | [[vo        |
| akRe | .std-ref}](#schema-iafcell){.reference          | ltage]{.std |
| vers | .internal})*                                    | .std-r      |
| al** |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **r  |                                                 | [           |
| efra |                                                 | [time]{.std |
| ct** |                                                 | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseIafCapCell]{.std                   | [[vo        |
| *res | .std-ref}](#schema-baseiafcapcell){.reference   | ltage]{.std |
| et** | .internal})*                                    | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   | *(from [[baseIafCapCell]{.std                   | [[vo        |
| thre | .std-ref}](#schema-baseiafcapcell){.reference   | ltage]{.std |
| sh** | .internal})*                                    | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Total current crossing the cell membrane *(from | [[cu        |
| *iMe | [[baseCellMembPotCap]{.std                      | rrent]{.std |
| mb** | .s                                              | .std-r      |
|      | td-ref}](#schema-basecellmembpotcap){.reference | ef}](Landin |
|      | .internal})*                                    | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **iS | Total current due to synaptic inputs *(from     | [[cu        |
| yn** | [[baseCellMembPotCap]{.std                      | rrent]{.std |
|      | .s                                              | .std-r      |
|      | td-ref}](#schema-basecellmembpotcap){.reference | ef}](Landin |
|      | .internal})*                                    | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | Membrane potential *(from                       | [[vo        |
| *v** | [[baseCellMembPot]{.std                         | ltage]{.std |
|      | .std-ref}](#schema-basecellmembpot){.reference  | .std-r      |
|      | .internal})*                                    | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Spike event *(from [[baseSpikingCell]{.std      | Direction:  |
| *spi | .std-ref}](#schema-basespikingcell){.reference  | out         |
| ke** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Attachments

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **sy |                                                 | [[          |
| naps |                                                 | basePointCu |
| es** |                                                 | rrent]{.std |
|      |                                                 | .std        |
|      |                                                 | -ref}](Land |
|      |                                                 | ing.html#sc |
|      |                                                 | hema-basepo |
|      |                                                 | intcurrent) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***State Variables***

:   **v**: [[voltage]{.std
    .std-ref}](Landing.html#schema-dimensions-voltage){.reference
    .internal}  (exposed as **v**)

:   **lastSpikeTime**: [[time]{.std
    .std-ref}](Landing.html#schema-dimensions-time){.reference
    .internal}

***On Start***

:   **v** = leakReversal

***Derived Variables***

:   **iSyn** = synapses\[\*\]-\>i(reduce method: add) (exposed as
    **iSyn**)

:   **iMemb** = leakConductance \* (leakReversal - v) + iSyn (exposed as
    **iMemb**)

***Regime: refractory (initial)***

:   ***On Entry***

:      **lastSpikeTime** = t

:      **v** = reset

:   ***On Conditions***

:      IF t \> lastSpikeTime + refract THEN

:       TRANSITION to REGIME **integrating**

***Regime: integrating (initial)***

:   ***On Conditions***

:      IF v \> thresh THEN

:       EVENT OUT on port: **spike**

:       TRANSITION to REGIME **refractory**

:   ***Time Derivatives***

:      d **v** /dt = iMemb / C
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="IafRefCell">
      <xs:complexContent>
        <xs:extension base="IafCell">
          <xs:attribute name="refract" type="Nml2Quantity_time" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=IafRefCell){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import IafRefCell
    from neuroml.utils import component_factory

    variable = component_factory(
        IafRefCell,
        id: 'a NmlId (required)' = None,
        metaid: 'a MetaId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        neuro_lex_id: 'a NeuroLexId (optional)' = None,
        leak_reversal: 'a Nml2Quantity_voltage (required)' = None,
        thresh: 'a Nml2Quantity_voltage (required)' = None,
        reset: 'a Nml2Quantity_voltage (required)' = None,
        C: 'a Nml2Quantity_capacitance (required)' = None,
        leak_conductance: 'a Nml2Quantity_conductance (required)' = None,
        refract: 'a Nml2Quantity_time (required)' = None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <iafRefCell id="iafRef" leakReversal="-50mV" thresh="-55mV" reset="-70mV" C="0.2nF" leakConductance="0.01uS" refract="5ms"/>
:::
:::
:::
:::
:::

::: {#izhikevichcell .section}
[]{#schema-izhikevichcell}

##### izhikevichCell[\#](#izhikevichcell "Link to this heading"){.headerlink}

extends *[[baseCellMembPot]{.std
.std-ref}](#schema-basecellmembpot){.reference .internal}*

*Cell based on the 2003 model of Izhikevich, see
[http://izhikevich.org/publications/spikes.htm](http://izhikevich.org/publications/spikes.htm){.reference
.external}.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Time scale of the recovery variable U           | Di          |
| *a** |                                                 | mensionless |
+------+-------------------------------------------------+-------------+
| *    | Sensitivity of U to the subthreshold            | Di          |
| *b** | fluctuations of the membrane potential V        | mensionless |
+------+-------------------------------------------------+-------------+
| *    | After-spike reset value of V                    | Di          |
| *c** |                                                 | mensionless |
+------+-------------------------------------------------+-------------+
| *    | After-spike increase to U                       | Di          |
| *d** |                                                 | mensionless |
+------+-------------------------------------------------+-------------+
| **   | Spike threshold                                 | [[vo        |
| thre |                                                 | ltage]{.std |
| sh** |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   | Initial membrane potential                      | [[vo        |
| v0** |                                                 | ltage]{.std |
|      |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Constants

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------------------+-----------------------------------+-------------+
| **MSEC** = 1ms     |                                   | [           |
|                    |                                   | [time]{.std |
|                    |                                   | .st         |
|                    |                                   | d-ref}](Lan |
|                    |                                   | ding.html#s |
|                    |                                   | chema-dimen |
|                    |                                   | sions-time) |
|                    |                                   | {.reference |
|                    |                                   | .internal}  |
+--------------------+-----------------------------------+-------------+
| **MVOLT** = 1mV    |                                   | [[vo        |
|                    |                                   | ltage]{.std |
|                    |                                   | .std-r      |
|                    |                                   | ef}](Landin |
|                    |                                   | g.html#sche |
|                    |                                   | ma-dimensio |
|                    |                                   | ns-voltage) |
|                    |                                   | {.reference |
|                    |                                   | .internal}  |
+--------------------+-----------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Membrane recovery variable                      | Di          |
| *U** |                                                 | mensionless |
+------+-------------------------------------------------+-------------+
| *    | Membrane potential *(from                       | [[vo        |
| *v** | [[baseCellMembPot]{.std                         | ltage]{.std |
|      | .std-ref}](#schema-basecellmembpot){.reference  | .std-r      |
|      | .internal})*                                    | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Spike event *(from [[baseSpikingCell]{.std      | Direction:  |
| *spi | .std-ref}](#schema-basespikingcell){.reference  | out         |
| ke** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Attachments

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **sy |                                                 | [[ba        |
| naps |                                                 | sePointCurr |
| es** |                                                 | entDL]{.std |
|      |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-basepoin |
|      |                                                 | tcurrentdl) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***State Variables***

:   **v**: [[voltage]{.std
    .std-ref}](Landing.html#schema-dimensions-voltage){.reference
    .internal}  (exposed as **v**)

:   **U**: Dimensionless  (exposed as **U**)

***On Start***

:   **v** = v0

:   **U** = v0 \* b / MVOLT

***On Conditions***

:   IF v \> thresh THEN

:      **v** = c \* MVOLT

:      **U** = U + d

:      EVENT OUT on port: **spike**

***Derived Variables***

:   **ISyn** = synapses\[\*\]-\>I(reduce method: add)

***Time Derivatives***

:   d **v** /dt = (0.04 \* v\^2 / MVOLT + 5 \* v + (140.0 - U + ISyn) \*
    MVOLT)/MSEC

:   d **U** /dt = a \* (b \* v / MVOLT - U) / MSEC
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="IzhikevichCell">
      <xs:complexContent>
        <xs:extension base="BaseCell">
          <xs:attribute name="v0" type="Nml2Quantity_voltage" use="required"/>
          <xs:attribute name="thresh" type="Nml2Quantity_voltage" use="required"/>
          <xs:attribute name="a" type="Nml2Quantity_none" use="required"/>
          <xs:attribute name="b" type="Nml2Quantity_none" use="required"/>
          <xs:attribute name="c" type="Nml2Quantity_none" use="required"/>
          <xs:attribute name="d" type="Nml2Quantity_none" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=IzhikevichCell){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import IzhikevichCell
    from neuroml.utils import component_factory

    variable = component_factory(
        IzhikevichCell,
        id: 'a NmlId (required)' = None,
        metaid: 'a MetaId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        neuro_lex_id: 'a NeuroLexId (optional)' = None,
        v0: 'a Nml2Quantity_voltage (required)' = None,
        thresh: 'a Nml2Quantity_voltage (required)' = None,
        a: 'a Nml2Quantity_none (required)' = None,
        b: 'a Nml2Quantity_none (required)' = None,
        c: 'a Nml2Quantity_none (required)' = None,
        d: 'a Nml2Quantity_none (required)' = None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <izhikevichCell id="izBurst" v0="-70mV" thresh="30mV" a="0.02" b="0.2" c="-50.0" d="2"/>
:::
:::
:::
:::
:::

::: {#izhikevich2007cell .section}
[]{#schema-izhikevich2007cell}

##### izhikevich2007Cell[\#](#izhikevich2007cell "Link to this heading"){.headerlink}

extends *[[baseCellMembPotCap]{.std
.std-ref}](#schema-basecellmembpotcap){.reference .internal}*

*Cell based on the modified Izhikevich model in Izhikevich 2007,
Dynamical systems in neuroscience, MIT Press.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Total capacitance of the cell membrane *(from   | [[capaci    |
| *C** | [[baseCellMembPotCap]{.std                      | tance]{.std |
|      | .s                                              | .std-ref}]  |
|      | td-ref}](#schema-basecellmembpotcap){.reference | (Landing.ht |
|      | .internal})*                                    | ml#schema-d |
|      |                                                 | imensions-c |
|      |                                                 | apacitance) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | Time scale of recovery variable u               | [[per       |
| *a** |                                                 | _time]{.std |
|      |                                                 | .std-re     |
|      |                                                 | f}](Landing |
|      |                                                 | .html#schem |
|      |                                                 | a-dimension |
|      |                                                 | s-per-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | Sensitivity of recovery variable u to           | [[conduc    |
| *b** | subthreshold fluctuations of membrane potential | tance]{.std |
|      | v                                               | .std-ref}]  |
|      |                                                 | (Landing.ht |
|      |                                                 | ml#schema-d |
|      |                                                 | imensions-c |
|      |                                                 | onductance) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | After-spike reset value of v                    | [[vo        |
| *c** |                                                 | ltage]{.std |
|      |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | After-spike increase to u                       | [[cu        |
| *d** |                                                 | rrent]{.std |
|      |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    |                                                 | [[conduct   |
| *k** |                                                 | ance_per_vo |
|      |                                                 | ltage]{.std |
|      |                                                 | .std-ref}]( |
|      |                                                 | Landing.htm |
|      |                                                 | l#schema-di |
|      |                                                 | mensions-co |
|      |                                                 | nductance-p |
|      |                                                 | er-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   | Initial membrane potential                      | [[vo        |
| v0** |                                                 | ltage]{.std |
|      |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | Peak action potential value                     | [[vo        |
| *vpe |                                                 | ltage]{.std |
| ak** |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   | Resting membrane potential                      | [[vo        |
| vr** |                                                 | ltage]{.std |
|      |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   | Spike threshold                                 | [[vo        |
| vt** |                                                 | ltage]{.std |
|      |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Total current crossing the cell membrane *(from | [[cu        |
| *iMe | [[baseCellMembPotCap]{.std                      | rrent]{.std |
| mb** | .s                                              | .std-r      |
|      | td-ref}](#schema-basecellmembpotcap){.reference | ef}](Landin |
|      | .internal})*                                    | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **iS | Total current due to synaptic inputs *(from     | [[cu        |
| yn** | [[baseCellMembPotCap]{.std                      | rrent]{.std |
|      | .s                                              | .std-r      |
|      | td-ref}](#schema-basecellmembpotcap){.reference | ef}](Landin |
|      | .internal})*                                    | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | Membrane recovery variable                      | [[cu        |
| *u** |                                                 | rrent]{.std |
|      |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | Membrane potential *(from                       | [[vo        |
| *v** | [[baseCellMembPot]{.std                         | ltage]{.std |
|      | .std-ref}](#schema-basecellmembpot){.reference  | .std-r      |
|      | .internal})*                                    | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Spike event *(from [[baseSpikingCell]{.std      | Direction:  |
| *spi | .std-ref}](#schema-basespikingcell){.reference  | out         |
| ke** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Attachments

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **sy |                                                 | [[          |
| naps |                                                 | basePointCu |
| es** |                                                 | rrent]{.std |
|      |                                                 | .std        |
|      |                                                 | -ref}](Land |
|      |                                                 | ing.html#sc |
|      |                                                 | hema-basepo |
|      |                                                 | intcurrent) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***State Variables***

:   **v**: [[voltage]{.std
    .std-ref}](Landing.html#schema-dimensions-voltage){.reference
    .internal}  (exposed as **v**)

:   **u**: [[current]{.std
    .std-ref}](Landing.html#schema-dimensions-current){.reference
    .internal}  (exposed as **u**)

***On Start***

:   **v** = v0

:   **u** = 0

***On Conditions***

:   IF v \> vpeak THEN

:      **v** = c

:      **u** = u + d

:      EVENT OUT on port: **spike**

***Derived Variables***

:   **iSyn** = synapses\[\*\]-\>i(reduce method: add) (exposed as
    **iSyn**)

:   **iMemb** = k \* (v-vr) \* (v-vt) + iSyn - u (exposed as **iMemb**)

***Time Derivatives***

:   d **v** /dt = iMemb / C

:   d **u** /dt = a \* (b \* (v-vr) - u)
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="Izhikevich2007Cell">
      <xs:complexContent>
        <xs:extension base="BaseCellMembPotCap">
          <xs:attribute name="v0" type="Nml2Quantity_voltage" use="required"/>
          <xs:attribute name="k" type="Nml2Quantity_conductancePerVoltage" use="required"/>
          <xs:attribute name="vr" type="Nml2Quantity_voltage" use="required"/>
          <xs:attribute name="vt" type="Nml2Quantity_voltage" use="required"/>
          <xs:attribute name="vpeak" type="Nml2Quantity_voltage" use="required"/>
          <xs:attribute name="a" type="Nml2Quantity_pertime" use="required"/>
          <xs:attribute name="b" type="Nml2Quantity_conductance" use="required"/>
          <xs:attribute name="c" type="Nml2Quantity_voltage" use="required"/>
          <xs:attribute name="d" type="Nml2Quantity_current" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=Izhikevich2007Cell){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import Izhikevich2007Cell
    from neuroml.utils import component_factory

    variable = component_factory(
        Izhikevich2007Cell,
        id: 'a NmlId (required)' = None,
        metaid: 'a MetaId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        neuro_lex_id: 'a NeuroLexId (optional)' = None,
        C: 'a Nml2Quantity_capacitance (required)' = None,
        v0: 'a Nml2Quantity_voltage (required)' = None,
        k: 'a Nml2Quantity_conductancePerVoltage (required)' = None,
        vr: 'a Nml2Quantity_voltage (required)' = None,
        vt: 'a Nml2Quantity_voltage (required)' = None,
        vpeak: 'a Nml2Quantity_voltage (required)' = None,
        a: 'a Nml2Quantity_pertime (required)' = None,
        b: 'a Nml2Quantity_conductance (required)' = None,
        c: 'a Nml2Quantity_voltage (required)' = None,
        d: 'a Nml2Quantity_current (required)' = None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <izhikevich2007Cell id="iz2007RS" v0="-60mV" C="100 pF" k="0.7 nS_per_mV" vr="-60 mV" vt="-40 mV" vpeak="35 mV" a="0.03 per_ms" b="-2 nS" c="-50 mV" d="100 pA"/>
:::
:::
:::
:::
:::

::: {#adexiafcell .section}
[]{#schema-adexiafcell}

##### adExIaFCell[\#](#adexiafcell "Link to this heading"){.headerlink}

extends *[[baseCellMembPotCap]{.std
.std-ref}](#schema-basecellmembpotcap){.reference .internal}*

*Model based on Brette R and Gerstner W ( 2005 ) Adaptive Exponential
Integrate-and-Fire Model as an Effective Description of Neuronal
Activity. J Neurophysiol 94:3637-3642.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Total capacitance of the cell membrane *(from   | [[capaci    |
| *C** | [[baseCellMembPotCap]{.std                      | tance]{.std |
|      | .s                                              | .std-ref}]  |
|      | td-ref}](#schema-basecellmembpotcap){.reference | (Landing.ht |
|      | .internal})*                                    | ml#schema-d |
|      |                                                 | imensions-c |
|      |                                                 | apacitance) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   | Leak reversal potential                         | [[vo        |
| EL** |                                                 | ltage]{.std |
|      |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   | Spike threshold                                 | [[vo        |
| VT** |                                                 | ltage]{.std |
|      |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | Sub-threshold adaptation variable               | [[conduc    |
| *a** |                                                 | tance]{.std |
|      |                                                 | .std-ref}]  |
|      |                                                 | (Landing.ht |
|      |                                                 | ml#schema-d |
|      |                                                 | imensions-c |
|      |                                                 | onductance) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | Spike-triggered adaptation variable             | [[cu        |
| *b** |                                                 | rrent]{.std |
|      |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **de | Slope factor                                    | [[vo        |
| lT** |                                                 | ltage]{.std |
|      |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   | Leak conductance                                | [[conduc    |
| gL** |                                                 | tance]{.std |
|      |                                                 | .std-ref}]  |
|      |                                                 | (Landing.ht |
|      |                                                 | ml#schema-d |
|      |                                                 | imensions-c |
|      |                                                 | onductance) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **r  | Refractory period                               | [           |
| efra |                                                 | [time]{.std |
| ct** |                                                 | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | Reset potential                                 | [[vo        |
| *res |                                                 | ltage]{.std |
| et** |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **ta | Adaptation time constant                        | [           |
| uw** |                                                 | [time]{.std |
|      |                                                 | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   | Spike detection threshold                       | [[vo        |
| thre |                                                 | ltage]{.std |
| sh** |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Total current crossing the cell membrane *(from | [[cu        |
| *iMe | [[baseCellMembPotCap]{.std                      | rrent]{.std |
| mb** | .s                                              | .std-r      |
|      | td-ref}](#schema-basecellmembpotcap){.reference | ef}](Landin |
|      | .internal})*                                    | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **iS | Total current due to synaptic inputs *(from     | [[cu        |
| yn** | [[baseCellMembPotCap]{.std                      | rrent]{.std |
|      | .s                                              | .std-r      |
|      | td-ref}](#schema-basecellmembpotcap){.reference | ef}](Landin |
|      | .internal})*                                    | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | Membrane potential *(from                       | [[vo        |
| *v** | [[baseCellMembPot]{.std                         | ltage]{.std |
|      | .std-ref}](#schema-basecellmembpot){.reference  | .std-r      |
|      | .internal})*                                    | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | Adaptation current                              | [[cu        |
| *w** |                                                 | rrent]{.std |
|      |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Spike event *(from [[baseSpikingCell]{.std      | Direction:  |
| *spi | .std-ref}](#schema-basespikingcell){.reference  | out         |
| ke** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Attachments

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **sy |                                                 | [[          |
| naps |                                                 | basePointCu |
| es** |                                                 | rrent]{.std |
|      |                                                 | .std        |
|      |                                                 | -ref}](Land |
|      |                                                 | ing.html#sc |
|      |                                                 | hema-basepo |
|      |                                                 | intcurrent) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***State Variables***

:   **v**: [[voltage]{.std
    .std-ref}](Landing.html#schema-dimensions-voltage){.reference
    .internal}  (exposed as **v**)

:   **w**: [[current]{.std
    .std-ref}](Landing.html#schema-dimensions-current){.reference
    .internal}  (exposed as **w**)

:   **lastSpikeTime**: [[time]{.std
    .std-ref}](Landing.html#schema-dimensions-time){.reference
    .internal}

***On Start***

:   **v** = EL

:   **w** = 0

***Derived Variables***

:   **iSyn** = synapses\[\*\]-\>i(reduce method: add) (exposed as
    **iSyn**)

:   **iMemb** = -1 \* gL \* (v - EL) + gL \* delT \* exp((v - VT) /
    delT) - w + iSyn (exposed as **iMemb**)

***Regime: refractory (initial)***

:   ***On Entry***

:      **lastSpikeTime** = t

:      **v** = reset

:      **w** = w + b

:   ***On Conditions***

:      IF t \> lastSpikeTime + refract THEN

:       TRANSITION to REGIME **integrating**

:   ***Time Derivatives***

:      d **w** /dt = (a \* (v - EL) - w) / tauw

***Regime: integrating (initial)***

:   ***On Conditions***

:      IF v \> thresh THEN

:       EVENT OUT on port: **spike**

:       TRANSITION to REGIME **refractory**

:   ***Time Derivatives***

:      d **v** /dt = iMemb / C

:      d **w** /dt = (a \* (v - EL) - w) / tauw
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="AdExIaFCell">
      <xs:complexContent>
        <xs:extension base="BaseCellMembPotCap">
          <xs:attribute name="gL" type="Nml2Quantity_conductance" use="required"/>
          <xs:attribute name="EL" type="Nml2Quantity_voltage" use="required"/>
          <xs:attribute name="reset" type="Nml2Quantity_voltage" use="required"/>
          <xs:attribute name="VT" type="Nml2Quantity_voltage" use="required"/>
          <xs:attribute name="thresh" type="Nml2Quantity_voltage" use="required"/>
          <xs:attribute name="delT" type="Nml2Quantity_voltage" use="required"/>
          <xs:attribute name="tauw" type="Nml2Quantity_time" use="required"/>
          <xs:attribute name="refract" type="Nml2Quantity_time" use="required"/>
          <xs:attribute name="a" type="Nml2Quantity_conductance" use="required"/>
          <xs:attribute name="b" type="Nml2Quantity_current" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=AdExIaFCell){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import AdExIaFCell
    from neuroml.utils import component_factory

    variable = component_factory(
        AdExIaFCell,
        id: 'a NmlId (required)' = None,
        metaid: 'a MetaId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        neuro_lex_id: 'a NeuroLexId (optional)' = None,
        C: 'a Nml2Quantity_capacitance (required)' = None,
        g_l: 'a Nml2Quantity_conductance (required)' = None,
        EL: 'a Nml2Quantity_voltage (required)' = None,
        reset: 'a Nml2Quantity_voltage (required)' = None,
        VT: 'a Nml2Quantity_voltage (required)' = None,
        thresh: 'a Nml2Quantity_voltage (required)' = None,
        del_t: 'a Nml2Quantity_voltage (required)' = None,
        tauw: 'a Nml2Quantity_time (required)' = None,
        refract: 'a Nml2Quantity_time (required)' = None,
        a: 'a Nml2Quantity_conductance (required)' = None,
        b: 'a Nml2Quantity_current (required)' = None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <adExIaFCell id="adExBurst" C="281pF" gL="30nS" EL="-70.6mV" reset="-48.5mV" VT="-50.4mV" thresh="-40.4mV" refract="0ms" delT="2mV" tauw="40ms" a="4nS" b="0.08nA"/>
:::
:::
:::
:::
:::

::: {#fitzhughnagumocell .section}
[]{#schema-fitzhughnagumocell}

##### fitzHughNagumoCell[\#](#fitzhughnagumocell "Link to this heading"){.headerlink}

extends *[[baseCellMembPotDL]{.std
.std-ref}](#schema-basecellmembpotdl){.reference .internal}*

*Simple dimensionless model of spiking cell from FitzHugh and Nagumo.
Superseded by **fitzHughNagumo1969Cell** ( See
[NeuroML/NeuroML2#42](https://github.com/NeuroML/NeuroML2/issues/42){.github
.reference .external} ).*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    |                                                 | Di          |
| *I** |                                                 | mensionless |
+------+-------------------------------------------------+-------------+
:::
:::

Constants

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------------------+-----------------------------------+-------------+
| **SEC** = 1s       |                                   | [           |
|                    |                                   | [time]{.std |
|                    |                                   | .st         |
|                    |                                   | d-ref}](Lan |
|                    |                                   | ding.html#s |
|                    |                                   | chema-dimen |
|                    |                                   | sions-time) |
|                    |                                   | {.reference |
|                    |                                   | .internal}  |
+--------------------+-----------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Membrane potential *(from                       | Di          |
| *V** | [[baseCellMembPotDL]{.std                       | mensionless |
|      | .                                               |             |
|      | std-ref}](#schema-basecellmembpotdl){.reference |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| *    |                                                 | Di          |
| *W** |                                                 | mensionless |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Spike event *(from [[baseSpikingCell]{.std      | Direction:  |
| *spi | .std-ref}](#schema-basespikingcell){.reference  | out         |
| ke** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***State Variables***

:   **V**: Dimensionless  (exposed as **V**)

:   **W**: Dimensionless  (exposed as **W**)

***Time Derivatives***

:   d **V** /dt = ( (V - ((V\^3) / 3)) - W + I) / SEC

:   d **W** /dt = (0.08 \* (V + 0.7 - 0.8 \* W)) / SEC
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="FitzHughNagumoCell">
      <xs:complexContent>
        <xs:extension base="BaseCell">
          <xs:attribute name="I" type="Nml2Quantity_none" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=FitzHughNagumoCell){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import FitzHughNagumoCell
    from neuroml.utils import component_factory

    variable = component_factory(
        FitzHughNagumoCell,
        id: 'a NmlId (required)' = None,
        metaid: 'a MetaId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        neuro_lex_id: 'a NeuroLexId (optional)' = None,
        I: 'a Nml2Quantity_none (required)' = None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <fitzHughNagumoCell id="fn1" I="0.8"/>
:::
:::
:::
:::
:::

::: {#pinskyrinzelca3cell .section}
[]{#schema-pinskyrinzelca3cell}

##### pinskyRinzelCA3Cell[\#](#pinskyrinzelca3cell "Link to this heading"){.headerlink}

extends *[[baseCellMembPot]{.std
.std-ref}](#schema-basecellmembpot){.reference .internal}*

*Reduced CA3 cell model from Pinsky, P.F., Rinzel, J. Intrinsic and
network rhythmogenesis in a reduced traub model for CA3 neurons. J
Comput Neurosci 1, 39-60 ( 1994 ). See
[OpenSourceBrain/PinskyRinzelModel](https://github.com/OpenSourceBrain/PinskyRinzelModel){.github
.reference .external}.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   |                                                 | Di          |
| alph |                                                 | mensionless |
| ac** |                                                 |             |
+------+-------------------------------------------------+-------------+
| *    |                                                 | Di          |
| *bet |                                                 | mensionless |
| ac** |                                                 |             |
+------+-------------------------------------------------+-------------+
| **   |                                                 | [[spe       |
| cm** |                                                 | cificCapaci |
|      |                                                 | tance]{.std |
|      |                                                 | .std-re     |
|      |                                                 | f}](Landing |
|      |                                                 | .html#schem |
|      |                                                 | a-dimension |
|      |                                                 | s-specificc |
|      |                                                 | apacitance) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **e  |                                                 | [[vo        |
| Ca** |                                                 | ltage]{.std |
|      |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   |                                                 | [[vo        |
| eK** |                                                 | ltage]{.std |
|      |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   |                                                 | [[vo        |
| eL** |                                                 | ltage]{.std |
|      |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **e  |                                                 | [[vo        |
| Na** |                                                 | ltage]{.std |
|      |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    |                                                 | [[co        |
| *gAm |                                                 | nductanceDe |
| pa** |                                                 | nsity]{.std |
|      |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-conducta |
|      |                                                 | ncedensity) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **g  |                                                 | [[co        |
| Ca** |                                                 | nductanceDe |
|      |                                                 | nsity]{.std |
|      |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-conducta |
|      |                                                 | ncedensity) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **g  |                                                 | [[co        |
| KC** |                                                 | nductanceDe |
|      |                                                 | nsity]{.std |
|      |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-conducta |
|      |                                                 | ncedensity) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    |                                                 | [[co        |
| *gKa |                                                 | nductanceDe |
| hp** |                                                 | nsity]{.std |
|      |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-conducta |
|      |                                                 | ncedensity) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **gK |                                                 | [[co        |
| dr** |                                                 | nductanceDe |
|      |                                                 | nsity]{.std |
|      |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-conducta |
|      |                                                 | ncedensity) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **g  |                                                 | [[co        |
| Ld** |                                                 | nductanceDe |
|      |                                                 | nsity]{.std |
|      |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-conducta |
|      |                                                 | ncedensity) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **g  |                                                 | [[co        |
| Ls** |                                                 | nductanceDe |
|      |                                                 | nsity]{.std |
|      |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-conducta |
|      |                                                 | ncedensity) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **g  |                                                 | [[co        |
| Na** |                                                 | nductanceDe |
|      |                                                 | nsity]{.std |
|      |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-conducta |
|      |                                                 | ncedensity) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    |                                                 | [[co        |
| *gNm |                                                 | nductanceDe |
| da** |                                                 | nsity]{.std |
|      |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-conducta |
|      |                                                 | ncedensity) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   |                                                 | [[co        |
| gc** |                                                 | nductanceDe |
|      |                                                 | nsity]{.std |
|      |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-conducta |
|      |                                                 | ncedensity) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    |                                                 | [[currentDe |
| *iDe |                                                 | nsity]{.std |
| nd** |                                                 | .s          |
|      |                                                 | td-ref}](La |
|      |                                                 | nding.html# |
|      |                                                 | schema-dime |
|      |                                                 | nsions-curr |
|      |                                                 | entdensity) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    |                                                 | [[currentDe |
| *iSo |                                                 | nsity]{.std |
| ma** |                                                 | .s          |
|      |                                                 | td-ref}](La |
|      |                                                 | nding.html# |
|      |                                                 | schema-dime |
|      |                                                 | nsions-curr |
|      |                                                 | entdensity) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   |                                                 | Di          |
| pp** |                                                 | mensionless |
+------+-------------------------------------------------+-------------+
| **q  |                                                 | Di          |
| d0** |                                                 | mensionless |
+------+-------------------------------------------------+-------------+
:::
:::

Constants

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------------------+-----------------------------------+-------------+
| **MSEC** = 1 ms    |                                   | [           |
|                    |                                   | [time]{.std |
|                    |                                   | .st         |
|                    |                                   | d-ref}](Lan |
|                    |                                   | ding.html#s |
|                    |                                   | chema-dimen |
|                    |                                   | sions-time) |
|                    |                                   | {.reference |
|                    |                                   | .internal}  |
+--------------------+-----------------------------------+-------------+
| **MVOLT** = 1 mV   |                                   | [[vo        |
|                    |                                   | ltage]{.std |
|                    |                                   | .std-r      |
|                    |                                   | ef}](Landin |
|                    |                                   | g.html#sche |
|                    |                                   | ma-dimensio |
|                    |                                   | ns-voltage) |
|                    |                                   | {.reference |
|                    |                                   | .internal}  |
+--------------------+-----------------------------------+-------------+
| **UAMP_PER_CM2** = |                                   | [[currentDe |
| 1 uA_per_cm2       |                                   | nsity]{.std |
|                    |                                   | .s          |
|                    |                                   | td-ref}](La |
|                    |                                   | nding.html# |
|                    |                                   | schema-dime |
|                    |                                   | nsions-curr |
|                    |                                   | entdensity) |
|                    |                                   | {.reference |
|                    |                                   | .internal}  |
+--------------------+-----------------------------------+-------------+
| **Smax** = 125.0   |                                   | Di          |
|                    |                                   | mensionless |
+--------------------+-----------------------------------+-------------+
| **Vsyn** = 60.0 mV |                                   | [[vo        |
|                    |                                   | ltage]{.std |
|                    |                                   | .std-r      |
|                    |                                   | ef}](Landin |
|                    |                                   | g.html#sche |
|                    |                                   | ma-dimensio |
|                    |                                   | ns-voltage) |
|                    |                                   | {.reference |
|                    |                                   | .internal}  |
+--------------------+-----------------------------------+-------------+
| **betaqd** = 0.001 |                                   | Di          |
|                    |                                   | mensionless |
+--------------------+-----------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **C  |                                                 | Di          |
| ad** |                                                 | mensionless |
+------+-------------------------------------------------+-------------+
| **IC |                                                 | [[currentDe |
| ad** |                                                 | nsity]{.std |
|      |                                                 | .s          |
|      |                                                 | td-ref}](La |
|      |                                                 | nding.html# |
|      |                                                 | schema-dime |
|      |                                                 | nsions-curr |
|      |                                                 | entdensity) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   |                                                 | Di          |
| Si** |                                                 | mensionless |
+------+-------------------------------------------------+-------------+
| **   | Dendritic membrane potential                    | [[vo        |
| Vd** |                                                 | ltage]{.std |
|      |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   | Somatic membrane potential                      | [[vo        |
| Vs** |                                                 | ltage]{.std |
|      |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   |                                                 | Di          |
| Wi** |                                                 | mensionless |
+------+-------------------------------------------------+-------------+
| **   |                                                 | Di          |
| cd** |                                                 | mensionless |
+------+-------------------------------------------------+-------------+
| **   |                                                 | Di          |
| hs** |                                                 | mensionless |
+------+-------------------------------------------------+-------------+
| **   |                                                 | Di          |
| ns** |                                                 | mensionless |
+------+-------------------------------------------------+-------------+
| **   |                                                 | Di          |
| qd** |                                                 | mensionless |
+------+-------------------------------------------------+-------------+
| **   |                                                 | Di          |
| sd** |                                                 | mensionless |
+------+-------------------------------------------------+-------------+
| *    | Membrane potential *(from                       | [[vo        |
| *v** | [[baseCellMembPot]{.std                         | ltage]{.std |
|      | .std-ref}](#schema-basecellmembpot){.reference  | .std-r      |
|      | .internal})*                                    | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Spike event *(from [[baseSpikingCell]{.std      | Direction:  |
| *spi | .std-ref}](#schema-basespikingcell){.reference  | out         |
| ke** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***State Variables***

:   **Vs**: [[voltage]{.std
    .std-ref}](Landing.html#schema-dimensions-voltage){.reference
    .internal}  (exposed as **Vs**)

:   **Vd**: [[voltage]{.std
    .std-ref}](Landing.html#schema-dimensions-voltage){.reference
    .internal}  (exposed as **Vd**)

:   **Cad**: Dimensionless  (exposed as **Cad**)

:   **hs**: Dimensionless  (exposed as **hs**)

:   **ns**: Dimensionless  (exposed as **ns**)

:   **sd**: Dimensionless  (exposed as **sd**)

:   **cd**: Dimensionless  (exposed as **cd**)

:   **qd**: Dimensionless  (exposed as **qd**)

:   **Si**: Dimensionless  (exposed as **Si**)

:   **Wi**: Dimensionless  (exposed as **Wi**)

:   **Sisat**: Dimensionless

***On Start***

:   **Vs** = eL

:   **Vd** = eL

:   **qd** = qd0

***Derived Variables***

:   **v** = Vs (exposed as **v**)

:   **ICad** = gCa*sd*sd\*(Vd-eCa) (exposed as **ICad**)

:   **alphams_Vs**
    = 0.32\*(-46.9-Vs/MVOLT)/(exp((-46.9-Vs/MVOLT)/4.0)-1.0)

:   **betams_Vs** = 0.28\*(Vs/MVOLT+19.9)/(exp((Vs/MVOLT+19.9)/5.0)-1.0)

:   **Minfs_Vs** = alphams_Vs/(alphams_Vs+betams_Vs)

:   **alphans_Vs**
    = 0.016\*(-24.9-Vs/MVOLT)/(exp((-24.9-Vs/MVOLT)/5.0)-1.0)

:   **betans_Vs** = 0.25*exp(-1.0-0.025*Vs/MVOLT)

:   **alphahs_Vs** = 0.128\*exp((-43.0-Vs/MVOLT)/18.0)

:   **betahs_Vs** = 4.0/(1.0+exp((-20.0-Vs/MVOLT)/5.0))

:   **alphasd_Vd** = 1.6/(1.0+exp(-0.072\*(Vd/MVOLT-5.0)))

:   **betasd_Vd** = 0.02\*(Vd/MVOLT+8.9)/(exp((Vd/MVOLT+8.9)/5.0)-1.0)

:   **Iampa** = gAmpa*Wi*(Vd-Vsyn)

:   **Inmda**
    = gNmda*Sisat*(Vd-Vsyn)/(1.0+0.28*exp(-0.062*(Vd/MVOLT-60.0)))

:   **Isyn** = Iampa+Inmda

***Conditional Derived Variables***

:   IF 0.00002\*Cad \> 0.01 THEN

:     **alphaqd** = 0.01

:   OTHERWISE

:     **alphaqd** = 0.00002\*Cad

:   IF Cad/250 \> 1 THEN

:     **chid** = 1

:   OTHERWISE

:     **chid** = Cad/250

:   IF Vd \< -10\*MVOLT THEN

:     **alphacd_Vd** = exp((Vd/MVOLT+50.0)/11-(Vd/MVOLT+53.5)/27)/18.975

:   OTHERWISE

:     **alphacd_Vd** = 2.0\*exp((-53.5-Vd/MVOLT)/27.0)

:   IF Vd \< -10\*MVOLT THEN

:     **betacd_Vd** = (2.0\*exp((-53.5-Vd/MVOLT)/27.0)-alphacd_Vd)

:   OTHERWISE

:     **betacd_Vd** = 0

:   IF Si \> Smax THEN

:     **Sisat** = Smax

:   OTHERWISE

:     **Sisat** = Si

***Time Derivatives***

:   d **Vs** /dt =
    (-gLs\*(Vs-eL)-gNa\*(Minfs_Vs\^2)*hs*(Vs-eNa)-gKdr*ns*(Vs-eK)+(gc/pp)\*(Vd-Vs)+iSoma/pp)
    / cm

:   d **Vd** /dt =
    (iDend/(1.0-pp)-Isyn/(1.0-pp)-gLd\*(Vd-eL)-ICad-gKahp*qd*(Vd-eK)-gKC*cd*chid\*(Vd-eK)+(gc\*(Vs-Vd))/(1.0-pp))
    / cm

:   d **Cad** /dt = (-0.13*ICad/UAMP_PER_CM2-0.075*Cad) / MSEC

:   d **hs** /dt = (alphahs_Vs-(alphahs_Vs+betahs_Vs)\*hs) / MSEC

:   d **ns** /dt = (alphans_Vs-(alphans_Vs+betans_Vs)\*ns) / MSEC

:   d **sd** /dt = (alphasd_Vd-(alphasd_Vd+betasd_Vd)\*sd) / MSEC

:   d **cd** /dt = (alphacd_Vd-(alphacd_Vd+betacd_Vd)\*cd) / MSEC

:   d **qd** /dt = (alphaqd-(alphaqd+betaqd)\*qd) / MSEC

:   d **Si** /dt = -Si/150.0

:   d **Wi** /dt = -Wi/2.0
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="PinskyRinzelCA3Cell">
      <xs:complexContent>
        <xs:extension base="BaseCell">
          <xs:attribute name="iSoma" type="Nml2Quantity_currentDensity" use="required"/>
          <xs:attribute name="iDend" type="Nml2Quantity_currentDensity" use="required"/>
          <xs:attribute name="gc" type="Nml2Quantity_conductanceDensity" use="required"/>
          <xs:attribute name="gLs" type="Nml2Quantity_conductanceDensity" use="required"/>
          <xs:attribute name="gLd" type="Nml2Quantity_conductanceDensity" use="required"/>
          <xs:attribute name="gNa" type="Nml2Quantity_conductanceDensity" use="required"/>
          <xs:attribute name="gKdr" type="Nml2Quantity_conductanceDensity" use="required"/>
          <xs:attribute name="gCa" type="Nml2Quantity_conductanceDensity" use="required"/>
          <xs:attribute name="gKahp" type="Nml2Quantity_conductanceDensity" use="required"/>
          <xs:attribute name="gKC" type="Nml2Quantity_conductanceDensity" use="required"/>
          <xs:attribute name="gNmda" type="Nml2Quantity_conductanceDensity" use="required"/>
          <xs:attribute name="gAmpa" type="Nml2Quantity_conductanceDensity" use="required"/>
          <xs:attribute name="eNa" type="Nml2Quantity_voltage" use="required"/>
          <xs:attribute name="eCa" type="Nml2Quantity_voltage" use="required"/>
          <xs:attribute name="eK" type="Nml2Quantity_voltage" use="required"/>
          <xs:attribute name="eL" type="Nml2Quantity_voltage" use="required"/>
          <xs:attribute name="qd0" type="Nml2Quantity_none" use="required"/>
          <xs:attribute name="pp" type="Nml2Quantity_none" use="required"/>
          <xs:attribute name="alphac" type="Nml2Quantity_none" use="required"/>
          <xs:attribute name="betac" type="Nml2Quantity_none" use="required"/>
          <xs:attribute name="cm" type="Nml2Quantity_specificCapacitance" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=PinskyRinzelCA3Cell){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import PinskyRinzelCA3Cell
    from neuroml.utils import component_factory

    variable = component_factory(
        PinskyRinzelCA3Cell,
        id: 'a NmlId (required)' = None,
        metaid: 'a MetaId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        neuro_lex_id: 'a NeuroLexId (optional)' = None,
        i_soma: 'a Nml2Quantity_currentDensity (required)' = None,
        i_dend: 'a Nml2Quantity_currentDensity (required)' = None,
        gc: 'a Nml2Quantity_conductanceDensity (required)' = None,
        g_ls: 'a Nml2Quantity_conductanceDensity (required)' = None,
        g_ld: 'a Nml2Quantity_conductanceDensity (required)' = None,
        g_na: 'a Nml2Quantity_conductanceDensity (required)' = None,
        g_kdr: 'a Nml2Quantity_conductanceDensity (required)' = None,
        g_ca: 'a Nml2Quantity_conductanceDensity (required)' = None,
        g_kahp: 'a Nml2Quantity_conductanceDensity (required)' = None,
        g_kc: 'a Nml2Quantity_conductanceDensity (required)' = None,
        g_nmda: 'a Nml2Quantity_conductanceDensity (required)' = None,
        g_ampa: 'a Nml2Quantity_conductanceDensity (required)' = None,
        e_na: 'a Nml2Quantity_voltage (required)' = None,
        e_ca: 'a Nml2Quantity_voltage (required)' = None,
        e_k: 'a Nml2Quantity_voltage (required)' = None,
        e_l: 'a Nml2Quantity_voltage (required)' = None,
        qd0: 'a Nml2Quantity_none (required)' = None,
        pp: 'a Nml2Quantity_none (required)' = None,
        alphac: 'a Nml2Quantity_none (required)' = None,
        betac: 'a Nml2Quantity_none (required)' = None,
        cm: 'a Nml2Quantity_specificCapacitance (required)' = None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <pinskyRinzelCA3Cell id="pr2A" iSoma="0.75 uA_per_cm2" iDend="0 uA_per_cm2" gc="2.1 mS_per_cm2" qd0="0" gLs="0.1 mS_per_cm2" gLd="0.1 mS_per_cm2" gNa="30 mS_per_cm2" gKdr="15 mS_per_cm2" gCa="10 mS_per_cm2" gKahp="0.8 mS_per_cm2" gKC="15 mS_per_cm2" eNa="60 mV" eCa="80 mV" eK="-75 mV" eL="-60 mV" pp="0.5" cm="3 uF_per_cm2" alphac="2" betac="0.1" gNmda="0 mS_per_cm2" gAmpa="0 mS_per_cm2"/>
:::
:::
:::
:::
:::

::: {#hindmarshrose1984cell .section}
[]{#schema-hindmarshrose1984cell}

##### hindmarshRose1984Cell[\#](#hindmarshrose1984cell "Link to this heading"){.headerlink}

extends *[[baseCellMembPotCap]{.std
.std-ref}](#schema-basecellmembpotcap){.reference .internal}*

*The Hindmarsh Rose model is a simplified point cell model which
captures complex firing patterns of single neurons, such as periodic and
chaotic bursting. It has a fast spiking subsystem, which is a
generalization of the FitzHugh-Nagumo system, coupled to a slower
subsystem which allows the model to fire bursts. The dynamical variables
x, y, z correspond to the membrane potential, a recovery variable, and a
slower adaptation current, respectively. See Hindmarsh J. L., and Rose
R. M. ( 1984 ) A model of neuronal bursting using three coupled first
order differential equations. Proc. R. Soc. London, Ser. B 221:87--102.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Total capacitance of the cell membrane *(from   | [[capaci    |
| *C** | [[baseCellMembPotCap]{.std                      | tance]{.std |
|      | .s                                              | .std-ref}]  |
|      | td-ref}](#schema-basecellmembpotcap){.reference | (Landing.ht |
|      | .internal})*                                    | ml#schema-d |
|      |                                                 | imensions-c |
|      |                                                 | apacitance) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | cubic term in x nullcline                       | Di          |
| *a** |                                                 | mensionless |
+------+-------------------------------------------------+-------------+
| *    | quadratic term in x nullcline                   | Di          |
| *b** |                                                 | mensionless |
+------+-------------------------------------------------+-------------+
| *    | constant term in y nullcline                    | Di          |
| *c** |                                                 | mensionless |
+------+-------------------------------------------------+-------------+
| *    | quadratic term in y nullcline                   | Di          |
| *d** |                                                 | mensionless |
+------+-------------------------------------------------+-------------+
| *    | timescale separation between slow and fast      | Di          |
| *r** | subsystem (r greater than 0; r much less than   | mensionless |
|      | 1)                                              |             |
+------+-------------------------------------------------+-------------+
| *    | related to adaptation                           | Di          |
| *s** |                                                 | mensionless |
+------+-------------------------------------------------+-------------+
| *    | scaling of x for physiological membrane         | [[vo        |
| *v_s | potential                                       | ltage]{.std |
| cali |                                                 | .std-r      |
| ng** |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   |                                                 | Di          |
| x0** |                                                 | mensionless |
+------+-------------------------------------------------+-------------+
| **   | related to the system's resting potential       | Di          |
| x1** |                                                 | mensionless |
+------+-------------------------------------------------+-------------+
| **   |                                                 | Di          |
| y0** |                                                 | mensionless |
+------+-------------------------------------------------+-------------+
| **   |                                                 | Di          |
| z0** |                                                 | mensionless |
+------+-------------------------------------------------+-------------+
:::
:::

Constants

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------------------+-----------------------------------+-------------+
| **MSEC** = 1ms     |                                   | [           |
|                    |                                   | [time]{.std |
|                    |                                   | .st         |
|                    |                                   | d-ref}](Lan |
|                    |                                   | ding.html#s |
|                    |                                   | chema-dimen |
|                    |                                   | sions-time) |
|                    |                                   | {.reference |
|                    |                                   | .internal}  |
+--------------------+-----------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **c  |                                                 | Di          |
| hi** |                                                 | mensionless |
+------+-------------------------------------------------+-------------+
| *    | Total current crossing the cell membrane *(from | [[cu        |
| *iMe | [[baseCellMembPotCap]{.std                      | rrent]{.std |
| mb** | .s                                              | .std-r      |
|      | td-ref}](#schema-basecellmembpotcap){.reference | ef}](Landin |
|      | .internal})*                                    | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **iS | Total current due to synaptic inputs *(from     | [[cu        |
| yn** | [[baseCellMembPotCap]{.std                      | rrent]{.std |
|      | .s                                              | .std-r      |
|      | td-ref}](#schema-basecellmembpotcap){.reference | ef}](Landin |
|      | .internal})*                                    | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **p  |                                                 | Di          |
| hi** |                                                 | mensionless |
+------+-------------------------------------------------+-------------+
| **r  |                                                 | Di          |
| ho** |                                                 | mensionless |
+------+-------------------------------------------------+-------------+
| **s  |                                                 | Di          |
| piki |                                                 | mensionless |
| ng** |                                                 |             |
+------+-------------------------------------------------+-------------+
| *    | Membrane potential *(from                       | [[vo        |
| *v** | [[baseCellMembPot]{.std                         | ltage]{.std |
|      | .std-ref}](#schema-basecellmembpot){.reference  | .std-r      |
|      | .internal})*                                    | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    |                                                 | Di          |
| *x** |                                                 | mensionless |
+------+-------------------------------------------------+-------------+
| *    |                                                 | Di          |
| *y** |                                                 | mensionless |
+------+-------------------------------------------------+-------------+
| *    |                                                 | Di          |
| *z** |                                                 | mensionless |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Spike event *(from [[baseSpikingCell]{.std      | Direction:  |
| *spi | .std-ref}](#schema-basespikingcell){.reference  | out         |
| ke** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Attachments

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **sy |                                                 | [[          |
| naps |                                                 | basePointCu |
| es** |                                                 | rrent]{.std |
|      |                                                 | .std        |
|      |                                                 | -ref}](Land |
|      |                                                 | ing.html#sc |
|      |                                                 | hema-basepo |
|      |                                                 | intcurrent) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***State Variables***

:   **v**: [[voltage]{.std
    .std-ref}](Landing.html#schema-dimensions-voltage){.reference
    .internal}  (exposed as **v**)

:   **y**: Dimensionless  (exposed as **y**)

:   **z**: Dimensionless  (exposed as **z**)

:   **spiking**: Dimensionless  (exposed as **spiking**)

***On Start***

:   **v** = x0 \* v_scaling

:   **y** = y0

:   **z** = z0

***On Conditions***

:   IF v \> 0 AND spiking \< 0.5 THEN

:      **spiking** = 1

:      EVENT OUT on port: **spike**

:   IF v \< 0 THEN

:      **spiking** = 0

***Derived Variables***

:   **iSyn** = synapses\[\*\]-\>i(reduce method: add) (exposed as
    **iSyn**)

:   **x** = v / v_scaling (exposed as **x**)

:   **phi** = y - a \* x\^3 + b \* x\^2 (exposed as **phi**)

:   **chi** = c - d \* x\^2 - y (exposed as **chi**)

:   **rho** = s \* ( x - x1 ) - z (exposed as **rho**)

:   **iMemb** = (C \* (v_scaling \* (phi - z) / MSEC)) + iSyn (exposed
    as **iMemb**)

***Time Derivatives***

:   d **v** /dt = iMemb/C

:   d **y** /dt = chi / MSEC

:   d **z** /dt = r \* rho / MSEC
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="HindmarshRose1984Cell">
      <xs:complexContent>
        <xs:extension base="BaseCellMembPotCap">
          <xs:attribute name="a" type="Nml2Quantity_none" use="required"/>
          <xs:attribute name="b" type="Nml2Quantity_none" use="required"/>
          <xs:attribute name="c" type="Nml2Quantity_none" use="required"/>
          <xs:attribute name="d" type="Nml2Quantity_none" use="required"/>
          <xs:attribute name="s" type="Nml2Quantity_none" use="required"/>
          <xs:attribute name="x1" type="Nml2Quantity_none" use="required"/>
          <xs:attribute name="r" type="Nml2Quantity_none" use="required"/>
          <xs:attribute name="x0" type="Nml2Quantity_none" use="required"/>
          <xs:attribute name="y0" type="Nml2Quantity_none" use="required"/>
          <xs:attribute name="z0" type="Nml2Quantity_none" use="required"/>
          <xs:attribute name="v_scaling" type="Nml2Quantity_voltage" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=HindmarshRose1984Cell){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import HindmarshRose1984Cell
    from neuroml.utils import component_factory

    variable = component_factory(
        HindmarshRose1984Cell,
        id: 'a NmlId (required)' = None,
        metaid: 'a MetaId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        neuro_lex_id: 'a NeuroLexId (optional)' = None,
        C: 'a Nml2Quantity_capacitance (required)' = None,
        a: 'a Nml2Quantity_none (required)' = None,
        b: 'a Nml2Quantity_none (required)' = None,
        c: 'a Nml2Quantity_none (required)' = None,
        d: 'a Nml2Quantity_none (required)' = None,
        s: 'a Nml2Quantity_none (required)' = None,
        x1: 'a Nml2Quantity_none (required)' = None,
        r: 'a Nml2Quantity_none (required)' = None,
        x0: 'a Nml2Quantity_none (required)' = None,
        y0: 'a Nml2Quantity_none (required)' = None,
        z0: 'a Nml2Quantity_none (required)' = None,
        v_scaling: 'a Nml2Quantity_voltage (required)' = None,
    )
:::
:::
:::
:::
:::
:::

[]{#document-Userdocs/Schemas/Channels}

::: {#channels .section .tex2jax_ignore .mathjax_ignore}
[]{#schema-channels}

#### Channels[\#](#channels "Link to this heading"){.headerlink}

**Defines voltage ( and concentration ) gated ion channel models. Ion
channels will generally extend [[baseIonChannel]{.std
.std-ref}](#schema-baseionchannel){.reference .internal}. The most
commonly used voltage dependent gate will extend [[baseGate]{.std
.std-ref}](#schema-basegate){.reference .internal}.**

------------------------------------------------------------------------

Original ComponentType definitions:
[Channels.xml](https://github.com/NeuroML/NeuroML2/blob/master/NeuroML2CoreTypes//Channels.xml){.reference
.external}. Schema against which NeuroML based on these should be valid:
[NeuroML_v2.3.xsd](https://github.com/NeuroML/NeuroML2/tree/master/Schemas/NeuroML2/NeuroML_v2.3.xsd){.reference
.external}. Generated on 14/08/24 from
[this](https://github.com/NeuroML/NeuroML2/commit/352244cff605cb1ba24fa7c11757dc818fe90fd2){.reference
.external} commit. Please file any issues or questions at the [issue
tracker here](https://github.com/NeuroML/NeuroML2/issues){.reference
.external}.

------------------------------------------------------------------------

::: {#basevoltagedeprate .section}
[]{#schema-basevoltagedeprate}

##### *baseVoltageDepRate*[\#](#basevoltagedeprate "Link to this heading"){.headerlink}

*Base ComponentType for voltage dependent rate. Produces a time varying
rate **r** which depends on **v.**.*

::: {.sd-tab-set .docutils}
Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    |                                                 | [[per       |
| *r** |                                                 | _time]{.std |
|      |                                                 | .std-re     |
|      |                                                 | f}](Landing |
|      |                                                 | .html#schem |
|      |                                                 | a-dimension |
|      |                                                 | s-per-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    |                                                 | [[vo        |
| *v** |                                                 | ltage]{.std |
|      |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::
:::
:::

::: {#basevoltageconcdeprate .section}
[]{#schema-basevoltageconcdeprate}

##### *baseVoltageConcDepRate*[\#](#basevoltageconcdeprate "Link to this heading"){.headerlink}

extends *[[baseVoltageDepRate]{.std
.std-ref}](#schema-basevoltagedeprate){.reference .internal}*

*Base ComponentType for voltage and concentration dependent rate.
Produces a time varying rate **r** which depends on **v** and
**caConc.**.*

::: {.sd-tab-set .docutils}
Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseVoltageDepRate]{.std               | [[per       |
| *r** | .s                                              | _time]{.std |
|      | td-ref}](#schema-basevoltagedeprate){.reference | .std-re     |
|      | .internal})*                                    | f}](Landing |
|      |                                                 | .html#schem |
|      |                                                 | a-dimension |
|      |                                                 | s-per-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   |                                                 | [[concentr  |
| caCo |                                                 | ation]{.std |
| nc** |                                                 | .           |
|      |                                                 | std-ref}](L |
|      |                                                 | anding.html |
|      |                                                 | #schema-dim |
|      |                                                 | ensions-con |
|      |                                                 | centration) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseVoltageDepRate]{.std               | [[vo        |
| *v** | .s                                              | ltage]{.std |
|      | td-ref}](#schema-basevoltagedeprate){.reference | .std-r      |
|      | .internal})*                                    | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::
:::
:::

::: {#basehhrate .section}
[]{#schema-basehhrate}

##### *baseHHRate*[\#](#basehhrate "Link to this heading"){.headerlink}

extends *[[baseVoltageDepRate]{.std
.std-ref}](#schema-basevoltagedeprate){.reference .internal}*

*Base ComponentType for rate which follow one of the typical forms for
rate equations in the standard HH formalism, using the parameters
**rate,** **midpoint** and **scale**.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **mi |                                                 | [[vo        |
| dpoi |                                                 | ltage]{.std |
| nt** |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **ra |                                                 | [[per       |
| te** |                                                 | _time]{.std |
|      |                                                 | .std-re     |
|      |                                                 | f}](Landing |
|      |                                                 | .html#schem |
|      |                                                 | a-dimension |
|      |                                                 | s-per-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    |                                                 | [[vo        |
| *sca |                                                 | ltage]{.std |
| le** |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseVoltageDepRate]{.std               | [[per       |
| *r** | .s                                              | _time]{.std |
|      | td-ref}](#schema-basevoltagedeprate){.reference | .std-re     |
|      | .internal})*                                    | f}](Landing |
|      |                                                 | .html#schem |
|      |                                                 | a-dimension |
|      |                                                 | s-per-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseVoltageDepRate]{.std               | [[vo        |
| *v** | .s                                              | ltage]{.std |
|      | td-ref}](#schema-basevoltagedeprate){.reference | .std-r      |
|      | .internal})*                                    | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="HHRate">
      <xs:complexContent>
        <xs:extension base="BaseWithoutId">
          <xs:attribute name="type" type="NmlId" use="required"/>
          <xs:attribute name="rate" type="Nml2Quantity_pertime" use="optional"/>
          <xs:attribute name="midpoint" type="Nml2Quantity_voltage" use="optional"/>
          <xs:attribute name="scale" type="Nml2Quantity_voltage" use="optional"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::
:::
:::

::: {#hhexprate .section}
[]{#schema-hhexprate}

##### HHExpRate[\#](#hhexprate "Link to this heading"){.headerlink}

extends *[[baseHHRate]{.std .std-ref}](#schema-basehhrate){.reference
.internal}*

*Exponential form for rate equation ( Q: Should these be renamed
hhExpRate, etc? ).*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **mi | *(from [[baseHHRate]{.std                       | [[vo        |
| dpoi | .std-ref}](#schema-basehhrate){.reference       | ltage]{.std |
| nt** | .internal})*                                    | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **ra | *(from [[baseHHRate]{.std                       | [[per       |
| te** | .std-ref}](#schema-basehhrate){.reference       | _time]{.std |
|      | .internal})*                                    | .std-re     |
|      |                                                 | f}](Landing |
|      |                                                 | .html#schem |
|      |                                                 | a-dimension |
|      |                                                 | s-per-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseHHRate]{.std                       | [[vo        |
| *sca | .std-ref}](#schema-basehhrate){.reference       | ltage]{.std |
| le** | .internal})*                                    | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseVoltageDepRate]{.std               | [[per       |
| *r** | .s                                              | _time]{.std |
|      | td-ref}](#schema-basevoltagedeprate){.reference | .std-re     |
|      | .internal})*                                    | f}](Landing |
|      |                                                 | .html#schem |
|      |                                                 | a-dimension |
|      |                                                 | s-per-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseVoltageDepRate]{.std               | [[vo        |
| *v** | .s                                              | ltage]{.std |
|      | td-ref}](#schema-basevoltagedeprate){.reference | .std-r      |
|      | .internal})*                                    | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***Derived Variables***

:   **r** = rate \* exp((v - midpoint)/scale) (exposed as **r**)
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="HHRate">
      <xs:complexContent>
        <xs:extension base="BaseWithoutId">
          <xs:attribute name="type" type="NmlId" use="required"/>
          <xs:attribute name="rate" type="Nml2Quantity_pertime" use="optional"/>
          <xs:attribute name="midpoint" type="Nml2Quantity_voltage" use="optional"/>
          <xs:attribute name="scale" type="Nml2Quantity_voltage" use="optional"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::
:::
:::

::: {#hhsigmoidrate .section}
[]{#schema-hhsigmoidrate}

##### HHSigmoidRate[\#](#hhsigmoidrate "Link to this heading"){.headerlink}

extends *[[baseHHRate]{.std .std-ref}](#schema-basehhrate){.reference
.internal}*

*Sigmoidal form for rate equation.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **mi | *(from [[baseHHRate]{.std                       | [[vo        |
| dpoi | .std-ref}](#schema-basehhrate){.reference       | ltage]{.std |
| nt** | .internal})*                                    | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **ra | *(from [[baseHHRate]{.std                       | [[per       |
| te** | .std-ref}](#schema-basehhrate){.reference       | _time]{.std |
|      | .internal})*                                    | .std-re     |
|      |                                                 | f}](Landing |
|      |                                                 | .html#schem |
|      |                                                 | a-dimension |
|      |                                                 | s-per-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseHHRate]{.std                       | [[vo        |
| *sca | .std-ref}](#schema-basehhrate){.reference       | ltage]{.std |
| le** | .internal})*                                    | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseVoltageDepRate]{.std               | [[per       |
| *r** | .s                                              | _time]{.std |
|      | td-ref}](#schema-basevoltagedeprate){.reference | .std-re     |
|      | .internal})*                                    | f}](Landing |
|      |                                                 | .html#schem |
|      |                                                 | a-dimension |
|      |                                                 | s-per-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseVoltageDepRate]{.std               | [[vo        |
| *v** | .s                                              | ltage]{.std |
|      | td-ref}](#schema-basevoltagedeprate){.reference | .std-r      |
|      | .internal})*                                    | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***Derived Variables***

:   **r** = rate / (1 + exp(0 - (v - midpoint)/scale)) (exposed as
    **r**)
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="HHRate">
      <xs:complexContent>
        <xs:extension base="BaseWithoutId">
          <xs:attribute name="type" type="NmlId" use="required"/>
          <xs:attribute name="rate" type="Nml2Quantity_pertime" use="optional"/>
          <xs:attribute name="midpoint" type="Nml2Quantity_voltage" use="optional"/>
          <xs:attribute name="scale" type="Nml2Quantity_voltage" use="optional"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::
:::
:::

::: {#hhexplinearrate .section}
[]{#schema-hhexplinearrate}

##### HHExpLinearRate[\#](#hhexplinearrate "Link to this heading"){.headerlink}

extends *[[baseHHRate]{.std .std-ref}](#schema-basehhrate){.reference
.internal}*

*Exponential linear form for rate equation. Linear for large positive
**v,** exponentially decays for large negative **v.**.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **mi | *(from [[baseHHRate]{.std                       | [[vo        |
| dpoi | .std-ref}](#schema-basehhrate){.reference       | ltage]{.std |
| nt** | .internal})*                                    | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **ra | *(from [[baseHHRate]{.std                       | [[per       |
| te** | .std-ref}](#schema-basehhrate){.reference       | _time]{.std |
|      | .internal})*                                    | .std-re     |
|      |                                                 | f}](Landing |
|      |                                                 | .html#schem |
|      |                                                 | a-dimension |
|      |                                                 | s-per-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseHHRate]{.std                       | [[vo        |
| *sca | .std-ref}](#schema-basehhrate){.reference       | ltage]{.std |
| le** | .internal})*                                    | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseVoltageDepRate]{.std               | [[per       |
| *r** | .s                                              | _time]{.std |
|      | td-ref}](#schema-basevoltagedeprate){.reference | .std-re     |
|      | .internal})*                                    | f}](Landing |
|      |                                                 | .html#schem |
|      |                                                 | a-dimension |
|      |                                                 | s-per-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseVoltageDepRate]{.std               | [[vo        |
| *v** | .s                                              | ltage]{.std |
|      | td-ref}](#schema-basevoltagedeprate){.reference | .std-r      |
|      | .internal})*                                    | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***Derived Variables***

:   **x** = (v - midpoint) / scale

***Conditional Derived Variables***

:   IF x != 0 THEN

:     **r** = rate \* x / (1 - exp(0 - x))  (exposed as **r**)

:   IF x = 0 THEN

:     **r** = rate  (exposed as **r**)
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="HHRate">
      <xs:complexContent>
        <xs:extension base="BaseWithoutId">
          <xs:attribute name="type" type="NmlId" use="required"/>
          <xs:attribute name="rate" type="Nml2Quantity_pertime" use="optional"/>
          <xs:attribute name="midpoint" type="Nml2Quantity_voltage" use="optional"/>
          <xs:attribute name="scale" type="Nml2Quantity_voltage" use="optional"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::
:::
:::

::: {#basevoltagedepvariable .section}
[]{#schema-basevoltagedepvariable}

##### *baseVoltageDepVariable*[\#](#basevoltagedepvariable "Link to this heading"){.headerlink}

*Base ComponentType for voltage dependent variable **x,** which depends
on **v.** Can be used for inf/steady state of rate variable.*

::: {.sd-tab-set .docutils}
Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    |                                                 | Di          |
| *x** |                                                 | mensionless |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    |                                                 | [[vo        |
| *v** |                                                 | ltage]{.std |
|      |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::
:::
:::

::: {#basevoltageconcdepvariable .section}
[]{#schema-basevoltageconcdepvariable}

##### *baseVoltageConcDepVariable*[\#](#basevoltageconcdepvariable "Link to this heading"){.headerlink}

extends *[[baseVoltageDepVariable]{.std
.std-ref}](#schema-basevoltagedepvariable){.reference .internal}*

*Base ComponentType for voltage and calcium concentration dependent
variable **x,** which depends on **v** and **caConc.**.*

::: {.sd-tab-set .docutils}
Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseVoltageDepVariable]{.std           | Di          |
| *x** | .std-r                                          | mensionless |
|      | ef}](#schema-basevoltagedepvariable){.reference |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   |                                                 | [[concentr  |
| caCo |                                                 | ation]{.std |
| nc** |                                                 | .           |
|      |                                                 | std-ref}](L |
|      |                                                 | anding.html |
|      |                                                 | #schema-dim |
|      |                                                 | ensions-con |
|      |                                                 | centration) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseVoltageDepVariable]{.std           | [[vo        |
| *v** | .std-r                                          | ltage]{.std |
|      | ef}](#schema-basevoltagedepvariable){.reference | .std-r      |
|      | .internal})*                                    | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::
:::
:::

::: {#basehhvariable .section}
[]{#schema-basehhvariable}

##### *baseHHVariable*[\#](#basehhvariable "Link to this heading"){.headerlink}

extends *[[baseVoltageDepVariable]{.std
.std-ref}](#schema-basevoltagedepvariable){.reference .internal}*

*Base ComponentType for voltage dependent dimensionless variable which
follow one of the typical forms for variable equations in the standard
HH formalism, using the parameters **rate,** **midpoint,** **scale**.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **mi |                                                 | [[vo        |
| dpoi |                                                 | ltage]{.std |
| nt** |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **ra |                                                 | Di          |
| te** |                                                 | mensionless |
+------+-------------------------------------------------+-------------+
| *    |                                                 | [[vo        |
| *sca |                                                 | ltage]{.std |
| le** |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseVoltageDepVariable]{.std           | Di          |
| *x** | .std-r                                          | mensionless |
|      | ef}](#schema-basevoltagedepvariable){.reference |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseVoltageDepVariable]{.std           | [[vo        |
| *v** | .std-r                                          | ltage]{.std |
|      | ef}](#schema-basevoltagedepvariable){.reference | .std-r      |
|      | .internal})*                                    | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="HHVariable">
      <xs:complexContent>
        <xs:extension base="BaseWithoutId">
          <xs:attribute name="type" type="NmlId" use="required"/>
          <xs:attribute name="rate" type="xs:float" use="optional"/>
          <xs:attribute name="midpoint" type="Nml2Quantity_voltage" use="optional"/>
          <xs:attribute name="scale" type="Nml2Quantity_voltage" use="optional"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::
:::
:::

::: {#hhexpvariable .section}
[]{#schema-hhexpvariable}

##### HHExpVariable[\#](#hhexpvariable "Link to this heading"){.headerlink}

extends *[[baseHHVariable]{.std
.std-ref}](#schema-basehhvariable){.reference .internal}*

*Exponential form for variable equation.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **mi | *(from [[baseHHVariable]{.std                   | [[vo        |
| dpoi | .std-ref}](#schema-basehhvariable){.reference   | ltage]{.std |
| nt** | .internal})*                                    | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **ra | *(from [[baseHHVariable]{.std                   | Di          |
| te** | .std-ref}](#schema-basehhvariable){.reference   | mensionless |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseHHVariable]{.std                   | [[vo        |
| *sca | .std-ref}](#schema-basehhvariable){.reference   | ltage]{.std |
| le** | .internal})*                                    | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseVoltageDepVariable]{.std           | Di          |
| *x** | .std-r                                          | mensionless |
|      | ef}](#schema-basevoltagedepvariable){.reference |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseVoltageDepVariable]{.std           | [[vo        |
| *v** | .std-r                                          | ltage]{.std |
|      | ef}](#schema-basevoltagedepvariable){.reference | .std-r      |
|      | .internal})*                                    | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***Derived Variables***

:   **x** = rate \* exp((v - midpoint)/scale) (exposed as **x**)
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="HHVariable">
      <xs:complexContent>
        <xs:extension base="BaseWithoutId">
          <xs:attribute name="type" type="NmlId" use="required"/>
          <xs:attribute name="rate" type="xs:float" use="optional"/>
          <xs:attribute name="midpoint" type="Nml2Quantity_voltage" use="optional"/>
          <xs:attribute name="scale" type="Nml2Quantity_voltage" use="optional"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::
:::
:::

::: {#hhsigmoidvariable .section}
[]{#schema-hhsigmoidvariable}

##### HHSigmoidVariable[\#](#hhsigmoidvariable "Link to this heading"){.headerlink}

extends *[[baseHHVariable]{.std
.std-ref}](#schema-basehhvariable){.reference .internal}*

*Sigmoidal form for variable equation.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **mi | *(from [[baseHHVariable]{.std                   | [[vo        |
| dpoi | .std-ref}](#schema-basehhvariable){.reference   | ltage]{.std |
| nt** | .internal})*                                    | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **ra | *(from [[baseHHVariable]{.std                   | Di          |
| te** | .std-ref}](#schema-basehhvariable){.reference   | mensionless |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseHHVariable]{.std                   | [[vo        |
| *sca | .std-ref}](#schema-basehhvariable){.reference   | ltage]{.std |
| le** | .internal})*                                    | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseVoltageDepVariable]{.std           | Di          |
| *x** | .std-r                                          | mensionless |
|      | ef}](#schema-basevoltagedepvariable){.reference |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseVoltageDepVariable]{.std           | [[vo        |
| *v** | .std-r                                          | ltage]{.std |
|      | ef}](#schema-basevoltagedepvariable){.reference | .std-r      |
|      | .internal})*                                    | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***Derived Variables***

:   **x** = rate / (1 + exp(0 - (v - midpoint)/scale)) (exposed as
    **x**)
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="HHVariable">
      <xs:complexContent>
        <xs:extension base="BaseWithoutId">
          <xs:attribute name="type" type="NmlId" use="required"/>
          <xs:attribute name="rate" type="xs:float" use="optional"/>
          <xs:attribute name="midpoint" type="Nml2Quantity_voltage" use="optional"/>
          <xs:attribute name="scale" type="Nml2Quantity_voltage" use="optional"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::
:::
:::

::: {#hhexplinearvariable .section}
[]{#schema-hhexplinearvariable}

##### HHExpLinearVariable[\#](#hhexplinearvariable "Link to this heading"){.headerlink}

extends *[[baseHHVariable]{.std
.std-ref}](#schema-basehhvariable){.reference .internal}*

*Exponential linear form for variable equation. Linear for large
positive **v,** exponentially decays for large negative **v.**.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **mi | *(from [[baseHHVariable]{.std                   | [[vo        |
| dpoi | .std-ref}](#schema-basehhvariable){.reference   | ltage]{.std |
| nt** | .internal})*                                    | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **ra | *(from [[baseHHVariable]{.std                   | Di          |
| te** | .std-ref}](#schema-basehhvariable){.reference   | mensionless |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseHHVariable]{.std                   | [[vo        |
| *sca | .std-ref}](#schema-basehhvariable){.reference   | ltage]{.std |
| le** | .internal})*                                    | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseVoltageDepVariable]{.std           | Di          |
| *x** | .std-r                                          | mensionless |
|      | ef}](#schema-basevoltagedepvariable){.reference |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseVoltageDepVariable]{.std           | [[vo        |
| *v** | .std-r                                          | ltage]{.std |
|      | ef}](#schema-basevoltagedepvariable){.reference | .std-r      |
|      | .internal})*                                    | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***Derived Variables***

:   **a** = (v - midpoint) / scale

:   **x** = rate \* a / (1 - exp(0 - a)) (exposed as **x**)
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="HHVariable">
      <xs:complexContent>
        <xs:extension base="BaseWithoutId">
          <xs:attribute name="type" type="NmlId" use="required"/>
          <xs:attribute name="rate" type="xs:float" use="optional"/>
          <xs:attribute name="midpoint" type="Nml2Quantity_voltage" use="optional"/>
          <xs:attribute name="scale" type="Nml2Quantity_voltage" use="optional"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::
:::
:::

::: {#basevoltagedeptime .section}
[]{#schema-basevoltagedeptime}

##### *baseVoltageDepTime*[\#](#basevoltagedeptime "Link to this heading"){.headerlink}

*Base ComponentType for voltage dependent ComponentType producing value
**t** with dimension time ( e.g. for time course of rate variable ).
Note: time course would not normally be fit to exp/sigmoid etc.*

::: {.sd-tab-set .docutils}
Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    |                                                 | [           |
| *t** |                                                 | [time]{.std |
|      |                                                 | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    |                                                 | [[vo        |
| *v** |                                                 | ltage]{.std |
|      |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::
:::
:::

::: {#basevoltageconcdeptime .section}
[]{#schema-basevoltageconcdeptime}

##### *baseVoltageConcDepTime*[\#](#basevoltageconcdeptime "Link to this heading"){.headerlink}

extends *[[baseVoltageDepTime]{.std
.std-ref}](#schema-basevoltagedeptime){.reference .internal}*

*Base type for voltage and calcium concentration dependent ComponentType
producing value **t** with dimension time ( e.g. for time course of rate
variable ).*

::: {.sd-tab-set .docutils}
Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseVoltageDepTime]{.std               | [           |
| *t** | .s                                              | [time]{.std |
|      | td-ref}](#schema-basevoltagedeptime){.reference | .st         |
|      | .internal})*                                    | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   |                                                 | [[concentr  |
| caCo |                                                 | ation]{.std |
| nc** |                                                 | .           |
|      |                                                 | std-ref}](L |
|      |                                                 | anding.html |
|      |                                                 | #schema-dim |
|      |                                                 | ensions-con |
|      |                                                 | centration) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseVoltageDepTime]{.std               | [[vo        |
| *v** | .s                                              | ltage]{.std |
|      | td-ref}](#schema-basevoltagedeptime){.reference | .std-r      |
|      | .internal})*                                    | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::
:::
:::

::: {#fixedtimecourse .section}
[]{#schema-fixedtimecourse}

##### fixedTimeCourse[\#](#fixedtimecourse "Link to this heading"){.headerlink}

extends *[[baseVoltageDepTime]{.std
.std-ref}](#schema-basevoltagedeptime){.reference .internal}*

*Time course of a fixed magnitude **tau** which can be used for the time
course in [[gateHHtauInf]{.std
.std-ref}](#schema-gatehhtauinf){.reference .internal},
[[gateHHratesTau]{.std .std-ref}](#schema-gatehhratestau){.reference
.internal} or [[gateHHratesTauInf]{.std
.std-ref}](#schema-gatehhratestauinf){.reference .internal}.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **t  |                                                 | [           |
| au** |                                                 | [time]{.std |
|      |                                                 | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseVoltageDepTime]{.std               | [           |
| *t** | .s                                              | [time]{.std |
|      | td-ref}](#schema-basevoltagedeptime){.reference | .st         |
|      | .internal})*                                    | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseVoltageDepTime]{.std               | [[vo        |
| *v** | .s                                              | ltage]{.std |
|      | td-ref}](#schema-basevoltagedeptime){.reference | .std-r      |
|      | .internal})*                                    | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***Derived Variables***

:   **t** = tau (exposed as **t**)
:::
:::
:::

::: {#baseq10settings .section}
[]{#schema-baseq10settings}

##### *baseQ10Settings*[\#](#baseq10settings "Link to this heading"){.headerlink}

*Base ComponentType for a scaling to apply to gating variable time
course, usually temperature dependent.*

::: {.sd-tab-set .docutils}
Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **q  |                                                 | Di          |
| 10** |                                                 | mensionless |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **t  |                                                 | [[temper    |
| empe |                                                 | ature]{.std |
| ratu |                                                 | .std-ref}]  |
| re** |                                                 | (Landing.ht |
|      |                                                 | ml#schema-d |
|      |                                                 | imensions-t |
|      |                                                 | emperature) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="Q10Settings">
      <xs:attribute name="type" type="NmlId" use="required"/>
      <xs:attribute name="fixedQ10" type="Nml2Quantity_none" use="optional"/>
      <xs:attribute name="q10Factor" type="Nml2Quantity_none" use="optional"/>
      <xs:attribute name="experimentalTemp" type="Nml2Quantity_temperature" use="optional"/>
    </xs:complexType>
:::
:::
:::
:::
:::

::: {#q10fixed .section}
[]{#schema-q10fixed}

##### q10Fixed[\#](#q10fixed "Link to this heading"){.headerlink}

extends *[[baseQ10Settings]{.std
.std-ref}](#schema-baseq10settings){.reference .internal}*

*A fixed value, **fixedQ10,** for the scaling of the time course of the
gating variable.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **fi |                                                 | Di          |
| xedQ |                                                 | mensionless |
| 10** |                                                 |             |
+------+-------------------------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **q  | *(from [[baseQ10Settings]{.std                  | Di          |
| 10** | .std-ref}](#schema-baseq10settings){.reference  | mensionless |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **t  | *(from [[baseQ10Settings]{.std                  | [[temper    |
| empe | .std-ref}](#schema-baseq10settings){.reference  | ature]{.std |
| ratu | .internal})*                                    | .std-ref}]  |
| re** |                                                 | (Landing.ht |
|      |                                                 | ml#schema-d |
|      |                                                 | imensions-t |
|      |                                                 | emperature) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***Derived Variables***

:   **q10** = fixedQ10 (exposed as **q10**)
:::
:::
:::

::: {#q10exptemp .section}
[]{#schema-q10exptemp}

##### q10ExpTemp[\#](#q10exptemp "Link to this heading"){.headerlink}

extends *[[baseQ10Settings]{.std
.std-ref}](#schema-baseq10settings){.reference .internal}*

*A value for the Q10 scaling which varies as a standard function of the
difference between the current temperature, **temperature,** and the
temperature at which the gating variable equations were determined,
**experimentalTemp**.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **ex |                                                 | [[temper    |
| peri |                                                 | ature]{.std |
| ment |                                                 | .std-ref}]  |
| alTe |                                                 | (Landing.ht |
| mp** |                                                 | ml#schema-d |
|      |                                                 | imensions-t |
|      |                                                 | emperature) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    |                                                 | Di          |
| *q10 |                                                 | mensionless |
| Fact |                                                 |             |
| or** |                                                 |             |
+------+-------------------------------------------------+-------------+
:::
:::

Constants

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------------------+-----------------------------------+-------------+
| **TENDEGREES** =   |                                   | [[temper    |
| 10K                |                                   | ature]{.std |
|                    |                                   | .std-ref}]  |
|                    |                                   | (Landing.ht |
|                    |                                   | ml#schema-d |
|                    |                                   | imensions-t |
|                    |                                   | emperature) |
|                    |                                   | {.reference |
|                    |                                   | .internal}  |
+--------------------+-----------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **q  | *(from [[baseQ10Settings]{.std                  | Di          |
| 10** | .std-ref}](#schema-baseq10settings){.reference  | mensionless |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **t  | *(from [[baseQ10Settings]{.std                  | [[temper    |
| empe | .std-ref}](#schema-baseq10settings){.reference  | ature]{.std |
| ratu | .internal})*                                    | .std-ref}]  |
| re** |                                                 | (Landing.ht |
|      |                                                 | ml#schema-d |
|      |                                                 | imensions-t |
|      |                                                 | emperature) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***Derived Variables***

:   **q10** = q10Factor\^((temperature -
    experimentalTemp)/TENDEGREES) (exposed as **q10**)
:::
:::
:::

::: {#baseconductancescaling .section}
[]{#schema-baseconductancescaling}

##### *baseConductanceScaling*[\#](#baseconductancescaling "Link to this heading"){.headerlink}

*Base ComponentType for a scaling to apply to a gate's conductance, e.g.
temperature dependent scaling.*

::: {.sd-tab-set .docutils}
Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   |                                                 | Di          |
| fact |                                                 | mensionless |
| or** |                                                 |             |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **t  |                                                 | [[temper    |
| empe |                                                 | ature]{.std |
| ratu |                                                 | .std-ref}]  |
| re** |                                                 | (Landing.ht |
|      |                                                 | ml#schema-d |
|      |                                                 | imensions-t |
|      |                                                 | emperature) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::
:::
:::

::: {#q10conductancescaling .section}
[]{#schema-q10conductancescaling}

##### q10ConductanceScaling[\#](#q10conductancescaling "Link to this heading"){.headerlink}

extends *[[baseConductanceScaling]{.std
.std-ref}](#schema-baseconductancescaling){.reference .internal}*

*A value for the conductance scaling which varies as a standard function
of the difference between the current temperature, **temperature,** and
the temperature at which the conductance was originally determined,
**experimentalTemp**.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **ex |                                                 | [[temper    |
| peri |                                                 | ature]{.std |
| ment |                                                 | .std-ref}]  |
| alTe |                                                 | (Landing.ht |
| mp** |                                                 | ml#schema-d |
|      |                                                 | imensions-t |
|      |                                                 | emperature) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    |                                                 | Di          |
| *q10 |                                                 | mensionless |
| Fact |                                                 |             |
| or** |                                                 |             |
+------+-------------------------------------------------+-------------+
:::
:::

Constants

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------------------+-----------------------------------+-------------+
| **TENDEGREES** =   |                                   | [[temper    |
| 10K                |                                   | ature]{.std |
|                    |                                   | .std-ref}]  |
|                    |                                   | (Landing.ht |
|                    |                                   | ml#schema-d |
|                    |                                   | imensions-t |
|                    |                                   | emperature) |
|                    |                                   | {.reference |
|                    |                                   | .internal}  |
+--------------------+-----------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   | *(from [[baseConductanceScaling]{.std           | Di          |
| fact | .std-r                                          | mensionless |
| or** | ef}](#schema-baseconductancescaling){.reference |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **t  | *(from [[baseConductanceScaling]{.std           | [[temper    |
| empe | .std-r                                          | ature]{.std |
| ratu | ef}](#schema-baseconductancescaling){.reference | .std-ref}]  |
| re** | .internal})*                                    | (Landing.ht |
|      |                                                 | ml#schema-d |
|      |                                                 | imensions-t |
|      |                                                 | emperature) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***Derived Variables***

:   **factor** = q10Factor\^((temperature -
    experimentalTemp)/TENDEGREES) (exposed as **factor**)
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="Q10ConductanceScaling">
      <xs:complexContent>
        <xs:extension base="BaseWithoutId">
          <xs:attribute name="q10Factor" type="Nml2Quantity_none" use="required"/>
          <xs:attribute name="experimentalTemp" type="Nml2Quantity_temperature" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=Q10ConductanceScaling){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import Q10ConductanceScaling
    from neuroml.utils import component_factory

    variable = component_factory(
        Q10ConductanceScaling,
        q10_factor: 'a Nml2Quantity_none (required)' = None,
        experimental_temp: 'a Nml2Quantity_temperature (required)' = None,
    )
:::
:::
:::
:::
:::

::: {#baseconductancescalingcadependent .section}
[]{#schema-baseconductancescalingcadependent}

##### *baseConductanceScalingCaDependent*[\#](#baseconductancescalingcadependent "Link to this heading"){.headerlink}

extends *[[baseConductanceScaling]{.std
.std-ref}](#schema-baseconductancescaling){.reference .internal}*

*Base ComponentType for a scaling to apply to a gate's conductance which
depends on Ca concentration. Usually a generic expression of **caConc**
( so no standard, non-base form here ).*

::: {.sd-tab-set .docutils}
Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   | *(from [[baseConductanceScaling]{.std           | Di          |
| fact | .std-r                                          | mensionless |
| or** | ef}](#schema-baseconductancescaling){.reference |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   |                                                 | [[concentr  |
| caCo |                                                 | ation]{.std |
| nc** |                                                 | .           |
|      |                                                 | std-ref}](L |
|      |                                                 | anding.html |
|      |                                                 | #schema-dim |
|      |                                                 | ensions-con |
|      |                                                 | centration) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **t  | *(from [[baseConductanceScaling]{.std           | [[temper    |
| empe | .std-r                                          | ature]{.std |
| ratu | ef}](#schema-baseconductancescaling){.reference | .std-ref}]  |
| re** | .internal})*                                    | (Landing.ht |
|      |                                                 | ml#schema-d |
|      |                                                 | imensions-t |
|      |                                                 | emperature) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::
:::
:::

::: {#basegate .section}
[]{#schema-basegate}

##### *baseGate*[\#](#basegate "Link to this heading"){.headerlink}

*Base ComponentType for a voltage and/or concentration dependent gate.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    |                                                 | Di          |
| *ins |                                                 | mensionless |
| tanc |                                                 |             |
| es** |                                                 |             |
+------+-------------------------------------------------+-------------+
:::
:::

Child list

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    |                                                 | [[          |
| *not |                                                 | notes]{.std |
| es** |                                                 | .std        |
|      |                                                 | -ref}](Land |
|      |                                                 | ing.html#sc |
|      |                                                 | hema-notes) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    |                                                 | Di          |
| *fco |                                                 | mensionless |
| nd** |                                                 |             |
+------+-------------------------------------------------+-------------+
| *    |                                                 | Di          |
| *q** |                                                 | mensionless |
+------+-------------------------------------------------+-------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="GateHHUndetermined">
      <xs:complexContent>
        <xs:extension base="Base">
          <xs:sequence>
            <xs:element name="notes" type="Notes" minOccurs="0"/>
            <xs:element name="q10Settings" type="Q10Settings" minOccurs="0"/>
            <xs:element name="forwardRate" type="HHRate" minOccurs="0"/>
            <xs:element name="reverseRate" type="HHRate" minOccurs="0"/>
            <xs:element name="timeCourse" type="HHTime" minOccurs="0"/>
            <xs:element name="steadyState" type="HHVariable" minOccurs="0"/>
            <xs:element name="subGate" type="GateFractionalSubgate" minOccurs="0" maxOccurs="unbounded"/>
          </xs:sequence>
          <xs:attribute name="instances" type="PositiveInteger" use="required"/>
          <xs:attribute name="type" type="gateTypes" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::
:::
:::

::: {#gate .section}
[]{#schema-gate}

##### gate[\#](#gate "Link to this heading"){.headerlink}

extends *[[baseGate]{.std .std-ref}](#schema-basegate){.reference
.internal}*

*Conveniently named baseGate.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseGate]{.std                         | Di          |
| *ins | .std-ref}](#schema-basegate){.reference         | mensionless |
| tanc | .internal})*                                    |             |
| es** |                                                 |             |
+------+-------------------------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseGate]{.std                         | Di          |
| *fco | .std-ref}](#schema-basegate){.reference         | mensionless |
| nd** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseGate]{.std                         | Di          |
| *q** | .std-ref}](#schema-basegate){.reference         | mensionless |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="GateHHUndetermined">
      <xs:complexContent>
        <xs:extension base="Base">
          <xs:sequence>
            <xs:element name="notes" type="Notes" minOccurs="0"/>
            <xs:element name="q10Settings" type="Q10Settings" minOccurs="0"/>
            <xs:element name="forwardRate" type="HHRate" minOccurs="0"/>
            <xs:element name="reverseRate" type="HHRate" minOccurs="0"/>
            <xs:element name="timeCourse" type="HHTime" minOccurs="0"/>
            <xs:element name="steadyState" type="HHVariable" minOccurs="0"/>
            <xs:element name="subGate" type="GateFractionalSubgate" minOccurs="0" maxOccurs="unbounded"/>
          </xs:sequence>
          <xs:attribute name="instances" type="PositiveInteger" use="required"/>
          <xs:attribute name="type" type="gateTypes" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=GateHHUndetermined){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import GateHHUndetermined
    from neuroml.utils import component_factory

    variable = component_factory(
        GateHHUndetermined,
        id: 'a NmlId (required)' = None,
        instances: 'a PositiveInteger (required)' = None,
        type: 'a gateTypes (required)' = None,
        notes: 'a string (optional)' = None,
        q10_settings: 'a Q10Settings (optional)' = None,
        forward_rate: 'a HHRate (optional)' = None,
        reverse_rate: 'a HHRate (optional)' = None,
        time_course: 'a HHTime (optional)' = None,
        steady_state: 'a HHVariable (optional)' = None,
        sub_gates: 'list of GateFractionalSubgate(s) (optional)' = None,
    )
:::
:::
:::
:::
:::

::: {#gatehhrates .section}
[]{#schema-gatehhrates}

##### gateHHrates[\#](#gatehhrates "Link to this heading"){.headerlink}

extends [[gate]{.std .std-ref}](#schema-gate){.reference .internal}

*Gate which follows the general Hodgkin Huxley formalism.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseGate]{.std                         | Di          |
| *ins | .std-ref}](#schema-basegate){.reference         | mensionless |
| tanc | .internal})*                                    |             |
| es** |                                                 |             |
+------+-------------------------------------------------+-------------+
:::
:::

Child list

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **f  |                                                 | [[ba        |
| orwa |                                                 | seVoltageDe |
| rdRa |                                                 | pRate]{.std |
| te** |                                                 | .std-       |
|      |                                                 | ref}](#sche |
|      |                                                 | ma-basevolt |
|      |                                                 | agedeprate) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **r  |                                                 | [[ba        |
| ever |                                                 | seVoltageDe |
| seRa |                                                 | pRate]{.std |
| te** |                                                 | .std-       |
|      |                                                 | ref}](#sche |
|      |                                                 | ma-basevolt |
|      |                                                 | agedeprate) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Children list

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **q  |                                                 | [           |
| 10Se |                                                 | [baseQ10Set |
| ttin |                                                 | tings]{.std |
| gs** |                                                 | .s          |
|      |                                                 | td-ref}](#s |
|      |                                                 | chema-baseq |
|      |                                                 | 10settings) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    |                                                 | [[per       |
| *alp |                                                 | _time]{.std |
| ha** |                                                 | .std-re     |
|      |                                                 | f}](Landing |
|      |                                                 | .html#schem |
|      |                                                 | a-dimension |
|      |                                                 | s-per-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **be |                                                 | [[per       |
| ta** |                                                 | _time]{.std |
|      |                                                 | .std-re     |
|      |                                                 | f}](Landing |
|      |                                                 | .html#schem |
|      |                                                 | a-dimension |
|      |                                                 | s-per-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseGate]{.std                         | Di          |
| *fco | .std-ref}](#schema-basegate){.reference         | mensionless |
| nd** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| **i  |                                                 | Di          |
| nf** |                                                 | mensionless |
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseGate]{.std                         | Di          |
| *q** | .std-ref}](#schema-basegate){.reference         | mensionless |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| *    |                                                 | Di          |
| *rat |                                                 | mensionless |
| eSca |                                                 |             |
| le** |                                                 |             |
+------+-------------------------------------------------+-------------+
| **t  |                                                 | [           |
| au** |                                                 | [time]{.std |
|      |                                                 | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***State Variables***

:   **q**: Dimensionless  (exposed as **q**)

***On Start***

:   **q** = inf

***Derived Variables***

:   **rateScale** = q10Settings\[\*\]-\>q10(reduce method:
    multiply) (exposed as **rateScale**)

:   **alpha** = forwardRate-\>r (exposed as **alpha**)

:   **beta** = reverseRate-\>r (exposed as **beta**)

:   **fcond** = q\^instances (exposed as **fcond**)

:   **inf** = alpha/(alpha+beta) (exposed as **inf**)

:   **tau** = 1/((alpha+beta) \* rateScale) (exposed as **tau**)

***Time Derivatives***

:   d **q** /dt = (inf - q) / tau
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="GateHHRates">
      <xs:complexContent>
        <xs:extension base="Base">
          <xs:all>
            <xs:element name="notes" type="Notes" minOccurs="0"/>
            <xs:element name="q10Settings" type="Q10Settings" minOccurs="0"/>
            <xs:element name="forwardRate" type="HHRate" minOccurs="1"/>
            <xs:element name="reverseRate" type="HHRate" minOccurs="1"/>
          </xs:all>
          <xs:attribute name="instances" type="PositiveInteger" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=GateHHRates){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import GateHHRates
    from neuroml.utils import component_factory

    variable = component_factory(
        GateHHRates,
        id: 'a NmlId (required)' = None,
        instances: 'a PositiveInteger (required)' = None,
        notes: 'a string (optional)' = None,
        q10_settings: 'a Q10Settings (optional)' = None,
        forward_rate: 'a HHRate (required)' = None,
        reverse_rate: 'a HHRate (required)' = None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <gateHHrates id="m" instances="3">
        <forwardRate type="HHExpLinearRate" rate="1per_ms" midpoint="-40mV" scale="10mV"/>
        <reverseRate type="HHExpRate" rate="4per_ms" midpoint="-65mV" scale="-18mV"/>
    </gateHHrates>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <gateHHrates id="h" instances="1">
        <forwardRate type="HHExpRate" rate="0.07per_ms" midpoint="-65mV" scale="-20mV"/>
        <reverseRate type="HHSigmoidRate" rate="1per_ms" midpoint="-35mV" scale="10mV"/>
    </gateHHrates>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <gateHHrates id="m" instances="3">
        <forwardRate type="HHExpLinearRate" rate="1per_ms" midpoint="-40mV" scale="10mV"/>
        <reverseRate type="HHExpRate" rate="4per_ms" midpoint="-65mV" scale="-18mV"/>
    </gateHHrates>
:::
:::
:::
:::
:::

::: {#gatehhtauinf .section}
[]{#schema-gatehhtauinf}

##### gateHHtauInf[\#](#gatehhtauinf "Link to this heading"){.headerlink}

extends [[gate]{.std .std-ref}](#schema-gate){.reference .internal}

*Gate which follows the general Hodgkin Huxley formalism.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseGate]{.std                         | Di          |
| *ins | .std-ref}](#schema-basegate){.reference         | mensionless |
| tanc | .internal})*                                    |             |
| es** |                                                 |             |
+------+-------------------------------------------------+-------------+
:::
:::

Child list

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   |                                                 | [[ba        |
| time |                                                 | seVoltageDe |
| Cour |                                                 | pTime]{.std |
| se** |                                                 | .std-       |
|      |                                                 | ref}](#sche |
|      |                                                 | ma-basevolt |
|      |                                                 | agedeptime) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **s  |                                                 | [[baseVo    |
| tead |                                                 | ltageDepVar |
| ySta |                                                 | iable]{.std |
| te** |                                                 | .std-ref}   |
|      |                                                 | ](#schema-b |
|      |                                                 | asevoltaged |
|      |                                                 | epvariable) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Children list

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **q  |                                                 | [           |
| 10Se |                                                 | [baseQ10Set |
| ttin |                                                 | tings]{.std |
| gs** |                                                 | .s          |
|      |                                                 | td-ref}](#s |
|      |                                                 | chema-baseq |
|      |                                                 | 10settings) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseGate]{.std                         | Di          |
| *fco | .std-ref}](#schema-basegate){.reference         | mensionless |
| nd** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| **i  |                                                 | Di          |
| nf** |                                                 | mensionless |
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseGate]{.std                         | Di          |
| *q** | .std-ref}](#schema-basegate){.reference         | mensionless |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| *    |                                                 | Di          |
| *rat |                                                 | mensionless |
| eSca |                                                 |             |
| le** |                                                 |             |
+------+-------------------------------------------------+-------------+
| **t  |                                                 | [           |
| au** |                                                 | [time]{.std |
|      |                                                 | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***State Variables***

:   **q**: Dimensionless  (exposed as **q**)

***On Start***

:   **q** = inf

***Derived Variables***

:   **rateScale** = q10Settings\[\*\]-\>q10(reduce method:
    multiply) (exposed as **rateScale**)

:   **fcond** = q\^instances (exposed as **fcond**)

:   **inf** = steadyState-\>x (exposed as **inf**)

:   **tauUnscaled** = timeCourse-\>t

:   **tau** = tauUnscaled / rateScale (exposed as **tau**)

***Time Derivatives***

:   d **q** /dt = (inf - q) / tau
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="GateHHTauInf">
      <xs:complexContent>
        <xs:extension base="Base">
          <xs:all>
            <xs:element name="notes" type="Notes" minOccurs="0"/>
            <xs:element name="q10Settings" type="Q10Settings" minOccurs="0"/>
            <xs:element name="timeCourse" type="HHTime" minOccurs="1"/>
            <xs:element name="steadyState" type="HHVariable" minOccurs="1"/>
          </xs:all>
          <xs:attribute name="instances" type="PositiveInteger" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=GateHHTauInf){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import GateHHTauInf
    from neuroml.utils import component_factory

    variable = component_factory(
        GateHHTauInf,
        id: 'a NmlId (required)' = None,
        instances: 'a PositiveInteger (required)' = None,
        notes: 'a string (optional)' = None,
        q10_settings: 'a Q10Settings (optional)' = None,
        time_course: 'a HHTime (required)' = None,
        steady_state: 'a HHVariable (required)' = None,
    )
:::
:::
:::
:::
:::

::: {#gatehhinstantaneous .section}
[]{#schema-gatehhinstantaneous}

##### gateHHInstantaneous[\#](#gatehhinstantaneous "Link to this heading"){.headerlink}

extends [[gate]{.std .std-ref}](#schema-gate){.reference .internal}

*Gate which follows the general Hodgkin Huxley formalism but is
instantaneous, so tau = 0 and gate follows exactly inf value.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseGate]{.std                         | Di          |
| *ins | .std-ref}](#schema-basegate){.reference         | mensionless |
| tanc | .internal})*                                    |             |
| es** |                                                 |             |
+------+-------------------------------------------------+-------------+
:::
:::

Constants

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------------------+-----------------------------------+-------------+
| **SEC** = 1 s      |                                   | [           |
|                    |                                   | [time]{.std |
|                    |                                   | .st         |
|                    |                                   | d-ref}](Lan |
|                    |                                   | ding.html#s |
|                    |                                   | chema-dimen |
|                    |                                   | sions-time) |
|                    |                                   | {.reference |
|                    |                                   | .internal}  |
+--------------------+-----------------------------------+-------------+
:::
:::

Child list

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **s  |                                                 | [[baseVo    |
| tead |                                                 | ltageDepVar |
| ySta |                                                 | iable]{.std |
| te** |                                                 | .std-ref}   |
|      |                                                 | ](#schema-b |
|      |                                                 | asevoltaged |
|      |                                                 | epvariable) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseGate]{.std                         | Di          |
| *fco | .std-ref}](#schema-basegate){.reference         | mensionless |
| nd** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| **i  |                                                 | Di          |
| nf** |                                                 | mensionless |
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseGate]{.std                         | Di          |
| *q** | .std-ref}](#schema-basegate){.reference         | mensionless |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| **t  |                                                 | [           |
| au** |                                                 | [time]{.std |
|      |                                                 | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***Derived Variables***

:   **inf** = steadyState-\>x (exposed as **inf**)

:   **tau** = 0 \* SEC (exposed as **tau**)

:   **q** = inf (exposed as **q**)

:   **fcond** = q\^instances (exposed as **fcond**)
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="GateHHInstantaneous">
      <xs:complexContent>
        <xs:extension base="Base">
          <xs:all>
            <xs:element name="notes" type="Notes" minOccurs="0"/>
            <xs:element name="steadyState" type="HHVariable" minOccurs="1"/>
          </xs:all>
          <xs:attribute name="instances" type="PositiveInteger" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=GateHHInstantaneous){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import GateHHInstantaneous
    from neuroml.utils import component_factory

    variable = component_factory(
        GateHHInstantaneous,
        id: 'a NmlId (required)' = None,
        instances: 'a PositiveInteger (required)' = None,
        notes: 'a string (optional)' = None,
        steady_state: 'a HHVariable (required)' = None,
    )
:::
:::
:::
:::
:::

::: {#gatehhratestau .section}
[]{#schema-gatehhratestau}

##### gateHHratesTau[\#](#gatehhratestau "Link to this heading"){.headerlink}

extends [[gate]{.std .std-ref}](#schema-gate){.reference .internal}

*Gate which follows the general Hodgkin Huxley formalism.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseGate]{.std                         | Di          |
| *ins | .std-ref}](#schema-basegate){.reference         | mensionless |
| tanc | .internal})*                                    |             |
| es** |                                                 |             |
+------+-------------------------------------------------+-------------+
:::
:::

Child list

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **f  |                                                 | [[ba        |
| orwa |                                                 | seVoltageDe |
| rdRa |                                                 | pRate]{.std |
| te** |                                                 | .std-       |
|      |                                                 | ref}](#sche |
|      |                                                 | ma-basevolt |
|      |                                                 | agedeprate) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **r  |                                                 | [[ba        |
| ever |                                                 | seVoltageDe |
| seRa |                                                 | pRate]{.std |
| te** |                                                 | .std-       |
|      |                                                 | ref}](#sche |
|      |                                                 | ma-basevolt |
|      |                                                 | agedeprate) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   |                                                 | [[ba        |
| time |                                                 | seVoltageDe |
| Cour |                                                 | pTime]{.std |
| se** |                                                 | .std-       |
|      |                                                 | ref}](#sche |
|      |                                                 | ma-basevolt |
|      |                                                 | agedeptime) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Children list

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **q  |                                                 | [           |
| 10Se |                                                 | [baseQ10Set |
| ttin |                                                 | tings]{.std |
| gs** |                                                 | .s          |
|      |                                                 | td-ref}](#s |
|      |                                                 | chema-baseq |
|      |                                                 | 10settings) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    |                                                 | [[per       |
| *alp |                                                 | _time]{.std |
| ha** |                                                 | .std-re     |
|      |                                                 | f}](Landing |
|      |                                                 | .html#schem |
|      |                                                 | a-dimension |
|      |                                                 | s-per-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **be |                                                 | [[per       |
| ta** |                                                 | _time]{.std |
|      |                                                 | .std-re     |
|      |                                                 | f}](Landing |
|      |                                                 | .html#schem |
|      |                                                 | a-dimension |
|      |                                                 | s-per-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseGate]{.std                         | Di          |
| *fco | .std-ref}](#schema-basegate){.reference         | mensionless |
| nd** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| **i  |                                                 | Di          |
| nf** |                                                 | mensionless |
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseGate]{.std                         | Di          |
| *q** | .std-ref}](#schema-basegate){.reference         | mensionless |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| *    |                                                 | Di          |
| *rat |                                                 | mensionless |
| eSca |                                                 |             |
| le** |                                                 |             |
+------+-------------------------------------------------+-------------+
| **t  |                                                 | [           |
| au** |                                                 | [time]{.std |
|      |                                                 | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***State Variables***

:   **q**: Dimensionless  (exposed as **q**)

***On Start***

:   **q** = inf

***Derived Variables***

:   **rateScale** = q10Settings\[\*\]-\>q10(reduce method:
    multiply) (exposed as **rateScale**)

:   **alpha** = forwardRate-\>r (exposed as **alpha**)

:   **beta** = reverseRate-\>r (exposed as **beta**)

:   **fcond** = q\^instances (exposed as **fcond**)

:   **inf** = alpha/(alpha+beta) (exposed as **inf**)

:   **tauUnscaled** = timeCourse-\>t

:   **tau** = tauUnscaled / rateScale (exposed as **tau**)

***Time Derivatives***

:   d **q** /dt = (inf - q) / tau
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="GateHHRatesTau">
      <xs:complexContent>
        <xs:extension base="Base">
          <xs:all>
            <xs:element name="notes" type="Notes" minOccurs="0"/>
            <xs:element name="q10Settings" type="Q10Settings" minOccurs="0"/>
            <xs:element name="forwardRate" type="HHRate" minOccurs="1"/>
            <xs:element name="reverseRate" type="HHRate" minOccurs="1"/>
            <xs:element name="timeCourse" type="HHTime" minOccurs="1"/>
          </xs:all>
          <xs:attribute name="instances" type="PositiveInteger" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=GateHHRatesTau){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import GateHHRatesTau
    from neuroml.utils import component_factory

    variable = component_factory(
        GateHHRatesTau,
        id: 'a NmlId (required)' = None,
        instances: 'a PositiveInteger (required)' = None,
        notes: 'a string (optional)' = None,
        q10_settings: 'a Q10Settings (optional)' = None,
        forward_rate: 'a HHRate (required)' = None,
        reverse_rate: 'a HHRate (required)' = None,
        time_course: 'a HHTime (required)' = None,
    )
:::
:::
:::
:::
:::

::: {#gatehhratesinf .section}
[]{#schema-gatehhratesinf}

##### gateHHratesInf[\#](#gatehhratesinf "Link to this heading"){.headerlink}

extends [[gate]{.std .std-ref}](#schema-gate){.reference .internal}

*Gate which follows the general Hodgkin Huxley formalism.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseGate]{.std                         | Di          |
| *ins | .std-ref}](#schema-basegate){.reference         | mensionless |
| tanc | .internal})*                                    |             |
| es** |                                                 |             |
+------+-------------------------------------------------+-------------+
:::
:::

Child list

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **f  |                                                 | [[ba        |
| orwa |                                                 | seVoltageDe |
| rdRa |                                                 | pRate]{.std |
| te** |                                                 | .std-       |
|      |                                                 | ref}](#sche |
|      |                                                 | ma-basevolt |
|      |                                                 | agedeprate) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **r  |                                                 | [[ba        |
| ever |                                                 | seVoltageDe |
| seRa |                                                 | pRate]{.std |
| te** |                                                 | .std-       |
|      |                                                 | ref}](#sche |
|      |                                                 | ma-basevolt |
|      |                                                 | agedeprate) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **s  |                                                 | [[baseVo    |
| tead |                                                 | ltageDepVar |
| ySta |                                                 | iable]{.std |
| te** |                                                 | .std-ref}   |
|      |                                                 | ](#schema-b |
|      |                                                 | asevoltaged |
|      |                                                 | epvariable) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Children list

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **q  |                                                 | [           |
| 10Se |                                                 | [baseQ10Set |
| ttin |                                                 | tings]{.std |
| gs** |                                                 | .s          |
|      |                                                 | td-ref}](#s |
|      |                                                 | chema-baseq |
|      |                                                 | 10settings) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    |                                                 | [[per       |
| *alp |                                                 | _time]{.std |
| ha** |                                                 | .std-re     |
|      |                                                 | f}](Landing |
|      |                                                 | .html#schem |
|      |                                                 | a-dimension |
|      |                                                 | s-per-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **be |                                                 | [[per       |
| ta** |                                                 | _time]{.std |
|      |                                                 | .std-re     |
|      |                                                 | f}](Landing |
|      |                                                 | .html#schem |
|      |                                                 | a-dimension |
|      |                                                 | s-per-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseGate]{.std                         | Di          |
| *fco | .std-ref}](#schema-basegate){.reference         | mensionless |
| nd** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| **i  |                                                 | Di          |
| nf** |                                                 | mensionless |
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseGate]{.std                         | Di          |
| *q** | .std-ref}](#schema-basegate){.reference         | mensionless |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| *    |                                                 | Di          |
| *rat |                                                 | mensionless |
| eSca |                                                 |             |
| le** |                                                 |             |
+------+-------------------------------------------------+-------------+
| **t  |                                                 | [           |
| au** |                                                 | [time]{.std |
|      |                                                 | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***State Variables***

:   **q**: Dimensionless  (exposed as **q**)

***On Start***

:   **q** = inf

***Derived Variables***

:   **rateScale** = q10Settings\[\*\]-\>q10(reduce method:
    multiply) (exposed as **rateScale**)

:   **alpha** = forwardRate-\>r (exposed as **alpha**)

:   **beta** = reverseRate-\>r (exposed as **beta**)

:   **fcond** = q\^instances (exposed as **fcond**)

:   **inf** = steadyState-\>x (exposed as **inf**)

:   **tau** = 1/((alpha+beta) \* rateScale) (exposed as **tau**)

***Time Derivatives***

:   d **q** /dt = (inf - q) / tau
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="GateHHRatesInf">
      <xs:complexContent>
        <xs:extension base="Base">
          <xs:all>
            <xs:element name="notes" type="Notes" minOccurs="0"/>
            <xs:element name="q10Settings" type="Q10Settings" minOccurs="0"/>
            <xs:element name="forwardRate" type="HHRate" minOccurs="1"/>
            <xs:element name="reverseRate" type="HHRate" minOccurs="1"/>
            <xs:element name="steadyState" type="HHVariable" minOccurs="1"/>
          </xs:all>
          <xs:attribute name="instances" type="PositiveInteger" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=GateHHRatesInf){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import GateHHRatesInf
    from neuroml.utils import component_factory

    variable = component_factory(
        GateHHRatesInf,
        id: 'a NmlId (required)' = None,
        instances: 'a PositiveInteger (required)' = None,
        notes: 'a string (optional)' = None,
        q10_settings: 'a Q10Settings (optional)' = None,
        forward_rate: 'a HHRate (required)' = None,
        reverse_rate: 'a HHRate (required)' = None,
        steady_state: 'a HHVariable (required)' = None,
    )
:::
:::
:::
:::
:::

::: {#gatehhratestauinf .section}
[]{#schema-gatehhratestauinf}

##### gateHHratesTauInf[\#](#gatehhratestauinf "Link to this heading"){.headerlink}

extends [[gate]{.std .std-ref}](#schema-gate){.reference .internal}

*Gate which follows the general Hodgkin Huxley formalism.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseGate]{.std                         | Di          |
| *ins | .std-ref}](#schema-basegate){.reference         | mensionless |
| tanc | .internal})*                                    |             |
| es** |                                                 |             |
+------+-------------------------------------------------+-------------+
:::
:::

Child list

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **f  |                                                 | [[ba        |
| orwa |                                                 | seVoltageDe |
| rdRa |                                                 | pRate]{.std |
| te** |                                                 | .std-       |
|      |                                                 | ref}](#sche |
|      |                                                 | ma-basevolt |
|      |                                                 | agedeprate) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **r  |                                                 | [[ba        |
| ever |                                                 | seVoltageDe |
| seRa |                                                 | pRate]{.std |
| te** |                                                 | .std-       |
|      |                                                 | ref}](#sche |
|      |                                                 | ma-basevolt |
|      |                                                 | agedeprate) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   |                                                 | [[ba        |
| time |                                                 | seVoltageDe |
| Cour |                                                 | pTime]{.std |
| se** |                                                 | .std-       |
|      |                                                 | ref}](#sche |
|      |                                                 | ma-basevolt |
|      |                                                 | agedeptime) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **s  |                                                 | [[baseVo    |
| tead |                                                 | ltageDepVar |
| ySta |                                                 | iable]{.std |
| te** |                                                 | .std-ref}   |
|      |                                                 | ](#schema-b |
|      |                                                 | asevoltaged |
|      |                                                 | epvariable) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Children list

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **q  |                                                 | [           |
| 10Se |                                                 | [baseQ10Set |
| ttin |                                                 | tings]{.std |
| gs** |                                                 | .s          |
|      |                                                 | td-ref}](#s |
|      |                                                 | chema-baseq |
|      |                                                 | 10settings) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    |                                                 | [[per       |
| *alp |                                                 | _time]{.std |
| ha** |                                                 | .std-re     |
|      |                                                 | f}](Landing |
|      |                                                 | .html#schem |
|      |                                                 | a-dimension |
|      |                                                 | s-per-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **be |                                                 | [[per       |
| ta** |                                                 | _time]{.std |
|      |                                                 | .std-re     |
|      |                                                 | f}](Landing |
|      |                                                 | .html#schem |
|      |                                                 | a-dimension |
|      |                                                 | s-per-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseGate]{.std                         | Di          |
| *fco | .std-ref}](#schema-basegate){.reference         | mensionless |
| nd** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| **i  |                                                 | Di          |
| nf** |                                                 | mensionless |
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseGate]{.std                         | Di          |
| *q** | .std-ref}](#schema-basegate){.reference         | mensionless |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| *    |                                                 | Di          |
| *rat |                                                 | mensionless |
| eSca |                                                 |             |
| le** |                                                 |             |
+------+-------------------------------------------------+-------------+
| **t  |                                                 | [           |
| au** |                                                 | [time]{.std |
|      |                                                 | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***State Variables***

:   **q**: Dimensionless  (exposed as **q**)

***On Start***

:   **q** = inf

***Derived Variables***

:   **rateScale** = q10Settings\[\*\]-\>q10(reduce method:
    multiply) (exposed as **rateScale**)

:   **alpha** = forwardRate-\>r (exposed as **alpha**)

:   **beta** = reverseRate-\>r (exposed as **beta**)

:   **inf** = steadyState-\>x (exposed as **inf**)

:   **tauUnscaled** = timeCourse-\>t

:   **tau** = tauUnscaled / rateScale (exposed as **tau**)

:   **fcond** = q\^instances (exposed as **fcond**)

***Time Derivatives***

:   d **q** /dt = (inf - q) / tau
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="GateHHRatesTauInf">
      <xs:complexContent>
        <xs:extension base="Base">
          <xs:all>
            <xs:element name="notes" type="Notes" minOccurs="0"/>
            <xs:element name="q10Settings" type="Q10Settings" minOccurs="0"/>
            <xs:element name="forwardRate" type="HHRate" minOccurs="1"/>
            <xs:element name="reverseRate" type="HHRate" minOccurs="1"/>
            <xs:element name="timeCourse" type="HHTime" minOccurs="1"/>
            <xs:element name="steadyState" type="HHVariable" minOccurs="1"/>
          </xs:all>
          <xs:attribute name="instances" type="PositiveInteger" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=GateHHRatesTauInf){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import GateHHRatesTauInf
    from neuroml.utils import component_factory

    variable = component_factory(
        GateHHRatesTauInf,
        id: 'a NmlId (required)' = None,
        instances: 'a PositiveInteger (required)' = None,
        notes: 'a string (optional)' = None,
        q10_settings: 'a Q10Settings (optional)' = None,
        forward_rate: 'a HHRate (required)' = None,
        reverse_rate: 'a HHRate (required)' = None,
        time_course: 'a HHTime (required)' = None,
        steady_state: 'a HHVariable (required)' = None,
    )
:::
:::
:::
:::
:::

::: {#gatefractional .section}
[]{#schema-gatefractional}

##### gateFractional[\#](#gatefractional "Link to this heading"){.headerlink}

extends [[gate]{.std .std-ref}](#schema-gate){.reference .internal}

*Gate composed of subgates contributing with fractional conductance.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseGate]{.std                         | Di          |
| *ins | .std-ref}](#schema-basegate){.reference         | mensionless |
| tanc | .internal})*                                    |             |
| es** |                                                 |             |
+------+-------------------------------------------------+-------------+
:::
:::

Children list

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **q  |                                                 | [           |
| 10Se |                                                 | [baseQ10Set |
| ttin |                                                 | tings]{.std |
| gs** |                                                 | .s          |
|      |                                                 | td-ref}](#s |
|      |                                                 | chema-baseq |
|      |                                                 | 10settings) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **s  |                                                 | [[su        |
| ubGa |                                                 | bGate]{.std |
| te** |                                                 | .std-       |
|      |                                                 | ref}](#sche |
|      |                                                 | ma-subgate) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseGate]{.std                         | Di          |
| *fco | .std-ref}](#schema-basegate){.reference         | mensionless |
| nd** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseGate]{.std                         | Di          |
| *q** | .std-ref}](#schema-basegate){.reference         | mensionless |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| *    |                                                 | Di          |
| *rat |                                                 | mensionless |
| eSca |                                                 |             |
| le** |                                                 |             |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***Derived Variables***

:   **q** = subGate\[\*\]-\>qfrac(reduce method: add) (exposed as **q**)

:   **fcond** = q\^instances (exposed as **fcond**)

:   **rateScale** = q10Settings\[\*\]-\>q10(reduce method:
    multiply) (exposed as **rateScale**)
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="GateFractional">
      <xs:complexContent>
        <xs:extension base="Base">
          <xs:sequence>
            <xs:element name="notes" type="Notes" minOccurs="0"/>
            <xs:element name="q10Settings" type="Q10Settings" minOccurs="0"/>
            <xs:element name="subGate" type="GateFractionalSubgate" minOccurs="1" maxOccurs="unbounded"/>
          </xs:sequence>
          <xs:attribute name="instances" type="PositiveInteger" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=GateFractional){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import GateFractional
    from neuroml.utils import component_factory

    variable = component_factory(
        GateFractional,
        id: 'a NmlId (required)' = None,
        instances: 'a PositiveInteger (required)' = None,
        notes: 'a string (optional)' = None,
        q10_settings: 'a Q10Settings (optional)' = None,
        sub_gates: 'list of GateFractionalSubgate(s) (required)' = None,
    )
:::
:::
:::
:::
:::

::: {#subgate .section}
[]{#schema-subgate}

##### subGate[\#](#subgate "Link to this heading"){.headerlink}

*Gate composed of subgates contributing with fractional conductance.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    |                                                 | Di          |
| *fra |                                                 | mensionless |
| ctio |                                                 |             |
| nalC |                                                 |             |
| ondu |                                                 |             |
| ctan |                                                 |             |
| ce** |                                                 |             |
+------+-------------------------------------------------+-------------+
:::
:::

Child list

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    |                                                 | [[          |
| *not |                                                 | notes]{.std |
| es** |                                                 | .std        |
|      |                                                 | -ref}](Land |
|      |                                                 | ing.html#sc |
|      |                                                 | hema-notes) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   |                                                 | [[ba        |
| time |                                                 | seVoltageDe |
| Cour |                                                 | pTime]{.std |
| se** |                                                 | .std-       |
|      |                                                 | ref}](#sche |
|      |                                                 | ma-basevolt |
|      |                                                 | agedeptime) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **s  |                                                 | [[baseVo    |
| tead |                                                 | ltageDepVar |
| ySta |                                                 | iable]{.std |
| te** |                                                 | .std-ref}   |
|      |                                                 | ](#schema-b |
|      |                                                 | asevoltaged |
|      |                                                 | epvariable) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **i  |                                                 | Di          |
| nf** |                                                 | mensionless |
+------+-------------------------------------------------+-------------+
| *    |                                                 | Di          |
| *q** |                                                 | mensionless |
+------+-------------------------------------------------+-------------+
| *    |                                                 | Di          |
| *qfr |                                                 | mensionless |
| ac** |                                                 |             |
+------+-------------------------------------------------+-------------+
| **t  |                                                 | [           |
| au** |                                                 | [time]{.std |
|      |                                                 | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    |                                                 | Di          |
| *rat |                                                 | mensionless |
| eSca |                                                 |             |
| le** |                                                 |             |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***State Variables***

:   **q**: Dimensionless  (exposed as **q**)

***On Start***

:   **q** = inf

***Derived Variables***

:   **inf** = steadyState-\>x (exposed as **inf**)

:   **tauUnscaled** = timeCourse-\>t

:   **tau** = tauUnscaled / rateScale (exposed as **tau**)

:   **qfrac** = q \* fractionalConductance (exposed as **qfrac**)

***Time Derivatives***

:   d **q** /dt = (inf - q) / tau
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="GateFractionalSubgate">
      <xs:complexContent>
        <xs:extension base="Base">
          <xs:all>
            <xs:element name="notes" type="Notes" minOccurs="0"/>
            <xs:element name="q10Settings" type="Q10Settings" minOccurs="0"/>
            <xs:element name="steadyState" type="HHVariable" minOccurs="1"/>
            <xs:element name="timeCourse" type="HHTime" minOccurs="1"/>
          </xs:all>
          <xs:attribute name="fractionalConductance" type="Nml2Quantity_none" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=GateFractionalSubgate){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import GateFractionalSubgate
    from neuroml.utils import component_factory

    variable = component_factory(
        GateFractionalSubgate,
        id: 'a NmlId (required)' = None,
        fractional_conductance: 'a Nml2Quantity_none (required)' = None,
        notes: 'a string (optional)' = None,
        q10_settings: 'a Q10Settings (optional)' = None,
        steady_state: 'a HHVariable (required)' = None,
        time_course: 'a HHTime (required)' = None,
    )
:::
:::
:::
:::
:::

::: {#baseionchannel .section}
[]{#schema-baseionchannel}

##### *baseIonChannel*[\#](#baseionchannel "Link to this heading"){.headerlink}

*Base for all ion channel ComponentTypes.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **c  |                                                 | [[conduc    |
| ondu |                                                 | tance]{.std |
| ctan |                                                 | .std-ref}]  |
| ce** |                                                 | (Landing.ht |
|      |                                                 | ml#schema-d |
|      |                                                 | imensions-c |
|      |                                                 | onductance) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Text fields

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| **     |                                                              |
| neuroL |                                                              |
| exId** |                                                              |
+--------+--------------------------------------------------------------+
:::
:::

Child list

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    |                                                 | [[          |
| *not |                                                 | notes]{.std |
| es** |                                                 | .std        |
|      |                                                 | -ref}](Land |
|      |                                                 | ing.html#sc |
|      |                                                 | hema-notes) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   |                                                 | [[annot     |
| anno |                                                 | ation]{.std |
| tati |                                                 | .std-ref}   |
| on** |                                                 | ](Landing.h |
|      |                                                 | tml#schema- |
|      |                                                 | annotation) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    |                                                 | Di          |
| *fop |                                                 | mensionless |
| en** |                                                 |             |
+------+-------------------------------------------------+-------------+
| *    |                                                 | [[conduc    |
| *g** |                                                 | tance]{.std |
|      |                                                 | .std-ref}]  |
|      |                                                 | (Landing.ht |
|      |                                                 | ml#schema-d |
|      |                                                 | imensions-c |
|      |                                                 | onductance) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    |                                                 | [[vo        |
| *v** |                                                 | ltage]{.std |
|      |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::
:::
:::

::: {#ionchannelpassive .section}
[]{#schema-ionchannelpassive}

##### ionChannelPassive[\#](#ionchannelpassive "Link to this heading"){.headerlink}

extends [[ionChannel]{.std .std-ref}](#schema-ionchannel){.reference
.internal}

*Simple passive ion channel where the constant conductance through the
channel is equal to **conductance**.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **c  | *(from [[baseIonChannel]{.std                   | [[conduc    |
| ondu | .std-ref}](#schema-baseionchannel){.reference   | tance]{.std |
| ctan | .internal})*                                    | .std-ref}]  |
| ce** |                                                 | (Landing.ht |
|      |                                                 | ml#schema-d |
|      |                                                 | imensions-c |
|      |                                                 | onductance) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Text fields

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| **spe  |                                                              |
| cies** |                                                              |
+--------+--------------------------------------------------------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseIonChannel]{.std                   | Di          |
| *fop | .std-ref}](#schema-baseionchannel){.reference   | mensionless |
| en** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseIonChannel]{.std                   | [[conduc    |
| *g** | .std-ref}](#schema-baseionchannel){.reference   | tance]{.std |
|      | .internal})*                                    | .std-ref}]  |
|      |                                                 | (Landing.ht |
|      |                                                 | ml#schema-d |
|      |                                                 | imensions-c |
|      |                                                 | onductance) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseIonChannel]{.std                   | [[vo        |
| *v** | .std-ref}](#schema-baseionchannel){.reference   | ltage]{.std |
|      | .internal})*                                    | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***Derived Variables***

:   **fopen** = 1 (exposed as **fopen**)

:   **g** = conductance (exposed as **g**)
:::
:::
:::

::: {#ionchannelhh .section}
[]{#schema-ionchannelhh}

##### ionChannelHH[\#](#ionchannelhh "Link to this heading"){.headerlink}

extends *[[baseIonChannel]{.std
.std-ref}](#schema-baseionchannel){.reference .internal}*

*Note [[ionChannel]{.std .std-ref}](#schema-ionchannel){.reference
.internal} and [[ionChannelHH]{.std
.std-ref}](#schema-ionchannelhh){.reference .internal} are currently
functionally identical. This is needed since many existing examples use
ionChannel, some use ionChannelHH. NeuroML v2beta4 should remove one of
these, probably ionChannelHH.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **c  | *(from [[baseIonChannel]{.std                   | [[conduc    |
| ondu | .std-ref}](#schema-baseionchannel){.reference   | tance]{.std |
| ctan | .internal})*                                    | .std-ref}]  |
| ce** |                                                 | (Landing.ht |
|      |                                                 | ml#schema-d |
|      |                                                 | imensions-c |
|      |                                                 | onductance) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Text fields

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| **spe  |                                                              |
| cies** |                                                              |
+--------+--------------------------------------------------------------+
:::
:::

Children list

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   |                                                 | [[baseCo    |
| cond |                                                 | nductanceSc |
| ucta |                                                 | aling]{.std |
| nceS |                                                 | .std-ref}   |
| cali |                                                 | ](#schema-b |
| ng** |                                                 | aseconducta |
|      |                                                 | ncescaling) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    |                                                 | [           |
| *gat |                                                 | [gate]{.std |
| es** |                                                 | .s          |
|      |                                                 | td-ref}](#s |
|      |                                                 | chema-gate) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseIonChannel]{.std                   | Di          |
| *fop | .std-ref}](#schema-baseionchannel){.reference   | mensionless |
| en** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseIonChannel]{.std                   | [[conduc    |
| *g** | .std-ref}](#schema-baseionchannel){.reference   | tance]{.std |
|      | .internal})*                                    | .std-ref}]  |
|      |                                                 | (Landing.ht |
|      |                                                 | ml#schema-d |
|      |                                                 | imensions-c |
|      |                                                 | onductance) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseIonChannel]{.std                   | [[vo        |
| *v** | .std-ref}](#schema-baseionchannel){.reference   | ltage]{.std |
|      | .internal})*                                    | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***Derived Variables***

:   **conductanceScale** = conductanceScaling\[\*\]-\>factor(reduce
    method: multiply)

:   **fopen0** = gates\[\*\]-\>fcond(reduce method: multiply)

:   **fopen** = conductanceScale \* fopen0 (exposed as **fopen**)

:   **g** = conductance \* fopen (exposed as **g**)
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="IonChannelHH">
      <xs:complexContent>
        <xs:extension base="IonChannel"/>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=IonChannelHH){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import IonChannelHH
    from neuroml.utils import component_factory

    variable = component_factory(
        IonChannelHH,
        id: 'a NmlId (required)' = None,
        metaid: 'a MetaId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        neuro_lex_id: 'a NeuroLexId (optional)' = None,
        q10_conductance_scalings: 'list of Q10ConductanceScaling(s) (optional)' = None,
        species: 'a NmlId (optional)' = None,
        type: 'a channelTypes (optional)' = None,
        conductance: 'a Nml2Quantity_conductance (optional)' = None,
        gates: 'list of GateHHUndetermined(s) (optional)' = None,
        gate_hh_rates: 'list of GateHHRates(s) (optional)' = None,
        gate_h_hrates_taus: 'list of GateHHRatesTau(s) (optional)' = None,
        gate_hh_tau_infs: 'list of GateHHTauInf(s) (optional)' = None,
        gate_h_hrates_infs: 'list of GateHHRatesInf(s) (optional)' = None,
        gate_h_hrates_tau_infs: 'list of GateHHRatesTauInf(s) (optional)' = None,
        gate_hh_instantaneouses: 'list of GateHHInstantaneous(s) (optional)' = None,
        gate_fractionals: 'list of GateFractional(s) (optional)' = None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <ionChannelHH id="pas" conductance="10pS"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <ionChannelHH id="HH_Na" conductance="10pS" species="na">  
            
        </ionChannelHH>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <ionChannelHH id="NaConductance" conductance="10pS" species="na">
        <gateHHrates id="m" instances="3">
            <forwardRate type="HHExpLinearRate" rate="1per_ms" midpoint="-40mV" scale="10mV"/>
            <reverseRate type="HHExpRate" rate="4per_ms" midpoint="-65mV" scale="-18mV"/>
        </gateHHrates>
        <gateHHrates id="h" instances="1">
            <forwardRate type="HHExpRate" rate="0.07per_ms" midpoint="-65mV" scale="-20mV"/>
            <reverseRate type="HHSigmoidRate" rate="1per_ms" midpoint="-35mV" scale="10mV"/>
        </gateHHrates>
    </ionChannelHH>
:::
:::
:::
:::
:::

::: {#ionchannel .section}
[]{#schema-ionchannel}

##### ionChannel[\#](#ionchannel "Link to this heading"){.headerlink}

extends [[ionChannelHH]{.std .std-ref}](#schema-ionchannelhh){.reference
.internal}

*Note [[ionChannel]{.std .std-ref}](#schema-ionchannel){.reference
.internal} and [[ionChannelHH]{.std
.std-ref}](#schema-ionchannelhh){.reference .internal} are currently
functionally identical. This is needed since many existing examples use
ionChannel, some use ionChannelHH. NeuroML v2beta4 should remove one of
these, probably ionChannelHH.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **c  | *(from [[baseIonChannel]{.std                   | [[conduc    |
| ondu | .std-ref}](#schema-baseionchannel){.reference   | tance]{.std |
| ctan | .internal})*                                    | .std-ref}]  |
| ce** |                                                 | (Landing.ht |
|      |                                                 | ml#schema-d |
|      |                                                 | imensions-c |
|      |                                                 | onductance) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseIonChannel]{.std                   | Di          |
| *fop | .std-ref}](#schema-baseionchannel){.reference   | mensionless |
| en** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseIonChannel]{.std                   | [[conduc    |
| *g** | .std-ref}](#schema-baseionchannel){.reference   | tance]{.std |
|      | .internal})*                                    | .std-ref}]  |
|      |                                                 | (Landing.ht |
|      |                                                 | ml#schema-d |
|      |                                                 | imensions-c |
|      |                                                 | onductance) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseIonChannel]{.std                   | [[vo        |
| *v** | .std-ref}](#schema-baseionchannel){.reference   | ltage]{.std |
|      | .internal})*                                    | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***Derived Variables***

:   **conductanceScale** = conductanceScaling\[\*\]-\>factor(reduce
    method: multiply)

:   **fopen0** = gates\[\*\]-\>fcond(reduce method: multiply)

:   **fopen** = conductanceScale \* fopen0 (exposed as **fopen**)

:   **g** = conductance \* fopen (exposed as **g**)
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="IonChannel">
      <xs:complexContent>
        <xs:extension base="IonChannelScalable">
          <xs:choice>
            <xs:element name="gate" type="GateHHUndetermined" minOccurs="0" maxOccurs="unbounded"/>
            <xs:element name="gateHHrates" type="GateHHRates" minOccurs="0" maxOccurs="unbounded"/>
            <xs:element name="gateHHratesTau" type="GateHHRatesTau" minOccurs="0" maxOccurs="unbounded"/>
            <xs:element name="gateHHtauInf" type="GateHHTauInf" minOccurs="0" maxOccurs="unbounded"/>
            <xs:element name="gateHHratesInf" type="GateHHRatesInf" minOccurs="0" maxOccurs="unbounded"/>
            <xs:element name="gateHHratesTauInf" type="GateHHRatesTauInf" minOccurs="0" maxOccurs="unbounded"/>
            <xs:element name="gateHHInstantaneous" type="GateHHInstantaneous" minOccurs="0" maxOccurs="unbounded"/>
            <xs:element name="gateFractional" type="GateFractional" minOccurs="0" maxOccurs="unbounded"/>
          </xs:choice>
          <xs:attribute name="species" type="NmlId" use="optional"/>
          <xs:attribute name="type" type="channelTypes" use="optional"/>
          <xs:attribute name="conductance" type="Nml2Quantity_conductance" use="optional"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=IonChannel){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import IonChannel
    from neuroml.utils import component_factory

    variable = component_factory(
        IonChannel,
        id: 'a NmlId (required)' = None,
        metaid: 'a MetaId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        neuro_lex_id: 'a NeuroLexId (optional)' = None,
        q10_conductance_scalings: 'list of Q10ConductanceScaling(s) (optional)' = None,
        species: 'a NmlId (optional)' = None,
        type: 'a channelTypes (optional)' = None,
        conductance: 'a Nml2Quantity_conductance (optional)' = None,
        gates: 'list of GateHHUndetermined(s) (optional)' = None,
        gate_hh_rates: 'list of GateHHRates(s) (optional)' = None,
        gate_h_hrates_taus: 'list of GateHHRatesTau(s) (optional)' = None,
        gate_hh_tau_infs: 'list of GateHHTauInf(s) (optional)' = None,
        gate_h_hrates_infs: 'list of GateHHRatesInf(s) (optional)' = None,
        gate_h_hrates_tau_infs: 'list of GateHHRatesTauInf(s) (optional)' = None,
        gate_hh_instantaneouses: 'list of GateHHInstantaneous(s) (optional)' = None,
        gate_fractionals: 'list of GateFractional(s) (optional)' = None,
        extensiontype_=None,
    )
:::
:::
:::
:::
:::

::: {#ionchannelvshift .section}
[]{#schema-ionchannelvshift}

##### ionChannelVShift[\#](#ionchannelvshift "Link to this heading"){.headerlink}

extends [[ionChannel]{.std .std-ref}](#schema-ionchannel){.reference
.internal}

*Same as [[ionChannel]{.std .std-ref}](#schema-ionchannel){.reference
.internal}, but with a **vShift** parameter to change voltage activation
of gates. The exact usage of **vShift** in expressions for rates is
determined by the individual gates.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **c  | *(from [[baseIonChannel]{.std                   | [[conduc    |
| ondu | .std-ref}](#schema-baseionchannel){.reference   | tance]{.std |
| ctan | .internal})*                                    | .std-ref}]  |
| ce** |                                                 | (Landing.ht |
|      |                                                 | ml#schema-d |
|      |                                                 | imensions-c |
|      |                                                 | onductance) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   |                                                 | [[vo        |
| vShi |                                                 | ltage]{.std |
| ft** |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Text fields

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| **spe  |                                                              |
| cies** |                                                              |
+--------+--------------------------------------------------------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseIonChannel]{.std                   | Di          |
| *fop | .std-ref}](#schema-baseionchannel){.reference   | mensionless |
| en** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseIonChannel]{.std                   | [[conduc    |
| *g** | .std-ref}](#schema-baseionchannel){.reference   | tance]{.std |
|      | .internal})*                                    | .std-ref}]  |
|      |                                                 | (Landing.ht |
|      |                                                 | ml#schema-d |
|      |                                                 | imensions-c |
|      |                                                 | onductance) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseIonChannel]{.std                   | [[vo        |
| *v** | .std-ref}](#schema-baseionchannel){.reference   | ltage]{.std |
|      | .internal})*                                    | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="IonChannelVShift">
      <xs:complexContent>
        <xs:extension base="IonChannel">
          <xs:attribute name="vShift" type="Nml2Quantity_voltage" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=IonChannelVShift){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import IonChannelVShift
    from neuroml.utils import component_factory

    variable = component_factory(
        IonChannelVShift,
        id: 'a NmlId (required)' = None,
        metaid: 'a MetaId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        neuro_lex_id: 'a NeuroLexId (optional)' = None,
        q10_conductance_scalings: 'list of Q10ConductanceScaling(s) (optional)' = None,
        species: 'a NmlId (optional)' = None,
        type: 'a channelTypes (optional)' = None,
        conductance: 'a Nml2Quantity_conductance (optional)' = None,
        gates: 'list of GateHHUndetermined(s) (optional)' = None,
        gate_hh_rates: 'list of GateHHRates(s) (optional)' = None,
        gate_h_hrates_taus: 'list of GateHHRatesTau(s) (optional)' = None,
        gate_hh_tau_infs: 'list of GateHHTauInf(s) (optional)' = None,
        gate_h_hrates_infs: 'list of GateHHRatesInf(s) (optional)' = None,
        gate_h_hrates_tau_infs: 'list of GateHHRatesTauInf(s) (optional)' = None,
        gate_hh_instantaneouses: 'list of GateHHInstantaneous(s) (optional)' = None,
        gate_fractionals: 'list of GateFractional(s) (optional)' = None,
        v_shift: 'a Nml2Quantity_voltage (required)' = None,
    )
:::
:::
:::
:::
:::

::: {#ksstate .section}
[]{#schema-ksstate}

##### KSState[\#](#ksstate "Link to this heading"){.headerlink}

*One of the states in which a [[gateKS]{.std
.std-ref}](#schema-gateks){.reference .internal} can be. The rates of
transitions between these states are given by [[KSTransition]{.std
.std-ref}](#schema-kstransition){.reference .internal}s.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **r  |                                                 | Di          |
| elat |                                                 | mensionless |
| iveC |                                                 |             |
| ondu |                                                 |             |
| ctan |                                                 |             |
| ce** |                                                 |             |
+------+-------------------------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    |                                                 | Di          |
| *occ |                                                 | mensionless |
| upan |                                                 |             |
| cy** |                                                 |             |
+------+-------------------------------------------------+-------------+
| *    |                                                 | Di          |
| *q** |                                                 | mensionless |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***State Variables***

:   **occupancy**: Dimensionless  (exposed as **occupancy**)

***Derived Variables***

:   **q** = relativeConductance \* occupancy (exposed as **q**)
:::
:::
:::

::: {#closedstate .section}
[]{#schema-closedstate}

##### closedState[\#](#closedstate "Link to this heading"){.headerlink}

extends [[KSState]{.std .std-ref}](#schema-ksstate){.reference
.internal}

*A [[KSState]{.std .std-ref}](#schema-ksstate){.reference .internal}
with **relativeConductance** of 0.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **r  | *(from [[KSState]{.std                          | Di          |
| elat | .std-ref}](#schema-ksstate){.reference          | mensionless |
| iveC | .internal})*                                    |             |
| ondu |                                                 |             |
| ctan |                                                 |             |
| ce** |                                                 |             |
+------+-------------------------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | *(from [[KSState]{.std                          | Di          |
| *occ | .std-ref}](#schema-ksstate){.reference          | mensionless |
| upan | .internal})*                                    |             |
| cy** |                                                 |             |
+------+-------------------------------------------------+-------------+
| *    | *(from [[KSState]{.std                          | Di          |
| *q** | .std-ref}](#schema-ksstate){.reference          | mensionless |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="ClosedState">
      <xs:complexContent>
        <xs:extension base="Base">
          </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=ClosedState){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import ClosedState
    from neuroml.utils import component_factory

    variable = component_factory(
        ClosedState,
        id: 'a NmlId (required)' = None,
    )
:::
:::
:::
:::
:::

::: {#openstate .section}
[]{#schema-openstate}

##### openState[\#](#openstate "Link to this heading"){.headerlink}

extends [[KSState]{.std .std-ref}](#schema-ksstate){.reference
.internal}

*A [[KSState]{.std .std-ref}](#schema-ksstate){.reference .internal}
with **relativeConductance** of 1.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **r  | *(from [[KSState]{.std                          | Di          |
| elat | .std-ref}](#schema-ksstate){.reference          | mensionless |
| iveC | .internal})*                                    |             |
| ondu |                                                 |             |
| ctan |                                                 |             |
| ce** |                                                 |             |
+------+-------------------------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | *(from [[KSState]{.std                          | Di          |
| *occ | .std-ref}](#schema-ksstate){.reference          | mensionless |
| upan | .internal})*                                    |             |
| cy** |                                                 |             |
+------+-------------------------------------------------+-------------+
| *    | *(from [[KSState]{.std                          | Di          |
| *q** | .std-ref}](#schema-ksstate){.reference          | mensionless |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="OpenState">
      <xs:complexContent>
        <xs:extension base="Base">
          </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=OpenState){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import OpenState
    from neuroml.utils import component_factory

    variable = component_factory(
        OpenState,
        id: 'a NmlId (required)' = None,
    )
:::
:::
:::
:::
:::

::: {#ionchannelks .section}
[]{#schema-ionchannelks}

##### ionChannelKS[\#](#ionchannelks "Link to this heading"){.headerlink}

extends *[[baseIonChannel]{.std
.std-ref}](#schema-baseionchannel){.reference .internal}*

*A kinetic scheme based ion channel with multiple [[gateKS]{.std
.std-ref}](#schema-gateks){.reference .internal}s, each of which
consists of multiple [[KSState]{.std
.std-ref}](#schema-ksstate){.reference .internal}s and
[[KSTransition]{.std .std-ref}](#schema-kstransition){.reference
.internal}s giving the rates of transition between them.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **c  | *(from [[baseIonChannel]{.std                   | [[conduc    |
| ondu | .std-ref}](#schema-baseionchannel){.reference   | tance]{.std |
| ctan | .internal})*                                    | .std-ref}]  |
| ce** |                                                 | (Landing.ht |
|      |                                                 | ml#schema-d |
|      |                                                 | imensions-c |
|      |                                                 | onductance) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Text fields

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| **spe  |                                                              |
| cies** |                                                              |
+--------+--------------------------------------------------------------+
:::
:::

Children list

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   |                                                 | [[baseCo    |
| cond |                                                 | nductanceSc |
| ucta |                                                 | aling]{.std |
| nceS |                                                 | .std-ref}   |
| cali |                                                 | ](#schema-b |
| ng** |                                                 | aseconducta |
|      |                                                 | ncescaling) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    |                                                 | [[g         |
| *gat |                                                 | ateKS]{.std |
| es** |                                                 | .std        |
|      |                                                 | -ref}](#sch |
|      |                                                 | ema-gateks) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseIonChannel]{.std                   | Di          |
| *fop | .std-ref}](#schema-baseionchannel){.reference   | mensionless |
| en** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseIonChannel]{.std                   | [[conduc    |
| *g** | .std-ref}](#schema-baseionchannel){.reference   | tance]{.std |
|      | .internal})*                                    | .std-ref}]  |
|      |                                                 | (Landing.ht |
|      |                                                 | ml#schema-d |
|      |                                                 | imensions-c |
|      |                                                 | onductance) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseIonChannel]{.std                   | [[vo        |
| *v** | .std-ref}](#schema-baseionchannel){.reference   | ltage]{.std |
|      | .internal})*                                    | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***Derived Variables***

:   **fopen** = gates\[\*\]-\>fcond(reduce method: multiply) (exposed as
    **fopen**)

:   **g** = fopen \* conductance (exposed as **g**)
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="IonChannelKS">
      <xs:complexContent>
        <xs:extension base="Standalone">
          <xs:sequence>
            <xs:element name="gateKS" type="GateKS" minOccurs="0" maxOccurs="unbounded"/>
          </xs:sequence>
          <xs:attribute name="species" type="NmlId" use="optional"/>
          <xs:attribute name="conductance" type="Nml2Quantity_conductance" use="optional"/>
          <xs:attribute name="neuroLexId" type="NeuroLexId" use="optional"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=IonChannelKS){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import IonChannelKS
    from neuroml.utils import component_factory

    variable = component_factory(
        IonChannelKS,
        id: 'a NmlId (required)' = None,
        metaid: 'a MetaId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        species: 'a NmlId (optional)' = None,
        conductance: 'a Nml2Quantity_conductance (optional)' = None,
        neuro_lex_id: 'a NeuroLexId (optional)' = None,
        gate_kses: 'list of GateKS(s) (optional)' = None,
    )
:::
:::
:::
:::
:::

::: {#kstransition .section}
[]{#schema-kstransition}

##### KSTransition[\#](#kstransition "Link to this heading"){.headerlink}

*Specified the forward and reverse rates of transition between two
[[KSState]{.std .std-ref}](#schema-ksstate){.reference .internal}s in a
[[gateKS]{.std .std-ref}](#schema-gateks){.reference .internal}.*

::: {.sd-tab-set .docutils}
Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   |                                                 | [[per       |
| rf** |                                                 | _time]{.std |
|      |                                                 | .std-re     |
|      |                                                 | f}](Landing |
|      |                                                 | .html#schem |
|      |                                                 | a-dimension |
|      |                                                 | s-per-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   |                                                 | [[per       |
| rr** |                                                 | _time]{.std |
|      |                                                 | .std-re     |
|      |                                                 | f}](Landing |
|      |                                                 | .html#schem |
|      |                                                 | a-dimension |
|      |                                                 | s-per-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::
:::
:::

::: {#forwardtransition .section}
[]{#schema-forwardtransition}

##### forwardTransition[\#](#forwardtransition "Link to this heading"){.headerlink}

extends [[KSTransition]{.std .std-ref}](#schema-kstransition){.reference
.internal}

*A forward only [[KSTransition]{.std
.std-ref}](#schema-kstransition){.reference .internal} for a
[[gateKS]{.std .std-ref}](#schema-gateks){.reference .internal} which
specifies a **rate** ( type [[baseHHRate]{.std
.std-ref}](#schema-basehhrate){.reference .internal} ) which follows one
of the standard Hodgkin Huxley forms ( e.g. [[HHExpRate]{.std
.std-ref}](#schema-hhexprate){.reference .internal},
[[HHSigmoidRate]{.std .std-ref}](#schema-hhsigmoidrate){.reference
.internal}, [[HHExpLinearRate]{.std
.std-ref}](#schema-hhexplinearrate){.reference .internal}.*

::: {.sd-tab-set .docutils}
Constants

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------------------+-----------------------------------+-------------+
| **SEC** = 1s       |                                   | [           |
|                    |                                   | [time]{.std |
|                    |                                   | .st         |
|                    |                                   | d-ref}](Lan |
|                    |                                   | ding.html#s |
|                    |                                   | chema-dimen |
|                    |                                   | sions-time) |
|                    |                                   | {.reference |
|                    |                                   | .internal}  |
+--------------------+-----------------------------------+-------------+
:::
:::

Child list

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **ra |                                                 | [[baseH     |
| te** |                                                 | HRate]{.std |
|      |                                                 | .std-ref    |
|      |                                                 | }](#schema- |
|      |                                                 | basehhrate) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   | *(from [[KSTransition]{.std                     | [[per       |
| rf** | .std-ref}](#schema-kstransition){.reference     | _time]{.std |
|      | .internal})*                                    | .std-re     |
|      |                                                 | f}](Landing |
|      |                                                 | .html#schem |
|      |                                                 | a-dimension |
|      |                                                 | s-per-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   | *(from [[KSTransition]{.std                     | [[per       |
| rr** | .std-ref}](#schema-kstransition){.reference     | _time]{.std |
|      | .internal})*                                    | .std-re     |
|      |                                                 | f}](Landing |
|      |                                                 | .html#schem |
|      |                                                 | a-dimension |
|      |                                                 | s-per-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***Derived Variables***

:   **rf0** = rate-\>r

:   **rf** = rf0 (exposed as **rf**)

:   **rr** = 0/SEC (exposed as **rr**)
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="ForwardTransition">
      <xs:complexContent>
        <xs:extension base="Base">
          <xs:sequence>
            <xs:any processContents="skip" minOccurs="0" maxOccurs="unbounded"/>
          </xs:sequence>
          <xs:attribute name="from" type="NmlId" use="required"/>
          <xs:attribute name="to" type="NmlId" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=ForwardTransition){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import ForwardTransition
    from neuroml.utils import component_factory

    variable = component_factory(
        ForwardTransition,
        id: 'a NmlId (required)' = None,
        from_: 'a NmlId (required)' = None,
        to: 'a NmlId (required)' = None,
        anytypeobjs_=None,
    )
:::
:::
:::
:::
:::

::: {#reversetransition .section}
[]{#schema-reversetransition}

##### reverseTransition[\#](#reversetransition "Link to this heading"){.headerlink}

extends [[KSTransition]{.std .std-ref}](#schema-kstransition){.reference
.internal}

*A reverse only [[KSTransition]{.std
.std-ref}](#schema-kstransition){.reference .internal} for a
[[gateKS]{.std .std-ref}](#schema-gateks){.reference .internal} which
specifies a **rate** ( type [[baseHHRate]{.std
.std-ref}](#schema-basehhrate){.reference .internal} ) which follows one
of the standard Hodgkin Huxley forms ( e.g. [[HHExpRate]{.std
.std-ref}](#schema-hhexprate){.reference .internal},
[[HHSigmoidRate]{.std .std-ref}](#schema-hhsigmoidrate){.reference
.internal}, [[HHExpLinearRate]{.std
.std-ref}](#schema-hhexplinearrate){.reference .internal}.*

::: {.sd-tab-set .docutils}
Constants

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------------------+-----------------------------------+-------------+
| **SEC** = 1s       |                                   | [           |
|                    |                                   | [time]{.std |
|                    |                                   | .st         |
|                    |                                   | d-ref}](Lan |
|                    |                                   | ding.html#s |
|                    |                                   | chema-dimen |
|                    |                                   | sions-time) |
|                    |                                   | {.reference |
|                    |                                   | .internal}  |
+--------------------+-----------------------------------+-------------+
:::
:::

Child list

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **ra |                                                 | [[baseH     |
| te** |                                                 | HRate]{.std |
|      |                                                 | .std-ref    |
|      |                                                 | }](#schema- |
|      |                                                 | basehhrate) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   | *(from [[KSTransition]{.std                     | [[per       |
| rf** | .std-ref}](#schema-kstransition){.reference     | _time]{.std |
|      | .internal})*                                    | .std-re     |
|      |                                                 | f}](Landing |
|      |                                                 | .html#schem |
|      |                                                 | a-dimension |
|      |                                                 | s-per-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   | *(from [[KSTransition]{.std                     | [[per       |
| rr** | .std-ref}](#schema-kstransition){.reference     | _time]{.std |
|      | .internal})*                                    | .std-re     |
|      |                                                 | f}](Landing |
|      |                                                 | .html#schem |
|      |                                                 | a-dimension |
|      |                                                 | s-per-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***Derived Variables***

:   **rr0** = rate-\>r

:   **rr** = rr0 (exposed as **rr**)

:   **rf** = 0/SEC (exposed as **rf**)
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="ReverseTransition">
      <xs:complexContent>
        <xs:extension base="Base">
          <xs:sequence>
            <xs:any processContents="skip" minOccurs="0" maxOccurs="unbounded"/>
          </xs:sequence>
          <xs:attribute name="from" type="NmlId" use="required"/>
          <xs:attribute name="to" type="NmlId" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=ReverseTransition){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import ReverseTransition
    from neuroml.utils import component_factory

    variable = component_factory(
        ReverseTransition,
        id: 'a NmlId (required)' = None,
        from_: 'a NmlId (required)' = None,
        to: 'a NmlId (required)' = None,
        anytypeobjs_=None,
    )
:::
:::
:::
:::
:::

::: {#vhalftransition .section}
[]{#schema-vhalftransition}

##### vHalfTransition[\#](#vhalftransition "Link to this heading"){.headerlink}

extends [[KSTransition]{.std .std-ref}](#schema-kstransition){.reference
.internal}

*Transition which specifies both the forward and reverse rates of
transition.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    |                                                 | Di          |
| *gam |                                                 | mensionless |
| ma** |                                                 |             |
+------+-------------------------------------------------+-------------+
| **t  |                                                 | [           |
| au** |                                                 | [time]{.std |
|      |                                                 | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   |                                                 | [           |
| tauM |                                                 | [time]{.std |
| in** |                                                 | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    |                                                 | [[vo        |
| *vHa |                                                 | ltage]{.std |
| lf** |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    |                                                 | Di          |
| *z** |                                                 | mensionless |
+------+-------------------------------------------------+-------------+
:::
:::

Constants

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------------------+-----------------------------------+-------------+
| **kte** = 25.3mV   |                                   | [[vo        |
|                    |                                   | ltage]{.std |
|                    |                                   | .std-r      |
|                    |                                   | ef}](Landin |
|                    |                                   | g.html#sche |
|                    |                                   | ma-dimensio |
|                    |                                   | ns-voltage) |
|                    |                                   | {.reference |
|                    |                                   | .internal}  |
+--------------------+-----------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   | *(from [[KSTransition]{.std                     | [[per       |
| rf** | .std-ref}](#schema-kstransition){.reference     | _time]{.std |
|      | .internal})*                                    | .std-re     |
|      |                                                 | f}](Landing |
|      |                                                 | .html#schem |
|      |                                                 | a-dimension |
|      |                                                 | s-per-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   | *(from [[KSTransition]{.std                     | [[per       |
| rr** | .std-ref}](#schema-kstransition){.reference     | _time]{.std |
|      | .internal})*                                    | .std-re     |
|      |                                                 | f}](Landing |
|      |                                                 | .html#schem |
|      |                                                 | a-dimension |
|      |                                                 | s-per-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    |                                                 | [[vo        |
| *v** |                                                 | ltage]{.std |
|      |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***Derived Variables***

:   **rf0** = exp(z \* gamma \* (v - vHalf) / kte) / tau

:   **rr0** = exp(-z \* (1 - gamma) \* (v - vHalf) / kte) / tau

:   **rf** = 1 / (1/rf0 + tauMin) (exposed as **rf**)

:   **rr** = 1 / (1/rr0 + tauMin) (exposed as **rr**)
:::
:::
:::

::: {#tauinftransition .section}
[]{#schema-tauinftransition}

##### tauInfTransition[\#](#tauinftransition "Link to this heading"){.headerlink}

extends [[KSTransition]{.std .std-ref}](#schema-kstransition){.reference
.internal}

*KS Transition specified in terms of time constant [schema:tau]{.xref
.std .std-ref} and steady state [schema:inf]{.xref .std .std-ref}.*

::: {.sd-tab-set .docutils}
Child list

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   |                                                 | [[ba        |
| time |                                                 | seVoltageDe |
| Cour |                                                 | pTime]{.std |
| se** |                                                 | .std-       |
|      |                                                 | ref}](#sche |
|      |                                                 | ma-basevolt |
|      |                                                 | agedeptime) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **s  |                                                 | [[baseVo    |
| tead |                                                 | ltageDepVar |
| ySta |                                                 | iable]{.std |
| te** |                                                 | .std-ref}   |
|      |                                                 | ](#schema-b |
|      |                                                 | asevoltaged |
|      |                                                 | epvariable) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   | *(from [[KSTransition]{.std                     | [[per       |
| rf** | .std-ref}](#schema-kstransition){.reference     | _time]{.std |
|      | .internal})*                                    | .std-re     |
|      |                                                 | f}](Landing |
|      |                                                 | .html#schem |
|      |                                                 | a-dimension |
|      |                                                 | s-per-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   | *(from [[KSTransition]{.std                     | [[per       |
| rr** | .std-ref}](#schema-kstransition){.reference     | _time]{.std |
|      | .internal})*                                    | .std-re     |
|      |                                                 | f}](Landing |
|      |                                                 | .html#schem |
|      |                                                 | a-dimension |
|      |                                                 | s-per-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***Derived Variables***

:   **tau** = timeCourse-\>t

:   **inf** = steadyState-\>x

:   **rf** = inf/tau (exposed as **rf**)

:   **rr** = (1-inf)/tau (exposed as **rr**)
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="TauInfTransition">
      <xs:complexContent>
        <xs:extension base="Base">
          <xs:all>
            <xs:element name="steadyState" type="HHVariable"/>
            <xs:element name="timeCourse" type="HHTime"/>
          </xs:all>
          <xs:attribute name="from" type="NmlId" use="required"/>
          <xs:attribute name="to" type="NmlId" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=TauInfTransition){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import TauInfTransition
    from neuroml.utils import component_factory

    variable = component_factory(
        TauInfTransition,
        id: 'a NmlId (required)' = None,
        from_: 'a NmlId (required)' = None,
        to: 'a NmlId (required)' = None,
        steady_state: 'a HHVariable (required)' = None,
        time_course: 'a HHTime (required)' = None,
    )
:::
:::
:::
:::
:::

::: {#gateks .section}
[]{#schema-gateks}

##### gateKS[\#](#gateks "Link to this heading"){.headerlink}

extends *[[baseGate]{.std .std-ref}](#schema-basegate){.reference
.internal}*

*A gate which consists of multiple [[KSState]{.std
.std-ref}](#schema-ksstate){.reference .internal}s and
[[KSTransition]{.std .std-ref}](#schema-kstransition){.reference
.internal}s giving the rates of transition between them.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseGate]{.std                         | Di          |
| *ins | .std-ref}](#schema-basegate){.reference         | mensionless |
| tanc | .internal})*                                    |             |
| es** |                                                 |             |
+------+-------------------------------------------------+-------------+
:::
:::

Children list

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   |                                                 | [[KS        |
| stat |                                                 | State]{.std |
| es** |                                                 | .std-       |
|      |                                                 | ref}](#sche |
|      |                                                 | ma-ksstate) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **t  |                                                 | [[KSTrans   |
| rans |                                                 | ition]{.std |
| itio |                                                 | .std-ref}]  |
| ns** |                                                 | (#schema-ks |
|      |                                                 | transition) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **q  |                                                 | [           |
| 10Se |                                                 | [baseQ10Set |
| ttin |                                                 | tings]{.std |
| gs** |                                                 | .s          |
|      |                                                 | td-ref}](#s |
|      |                                                 | chema-baseq |
|      |                                                 | 10settings) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseGate]{.std                         | Di          |
| *fco | .std-ref}](#schema-basegate){.reference         | mensionless |
| nd** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| *    | *(from [[baseGate]{.std                         | Di          |
| *q** | .std-ref}](#schema-basegate){.reference         | mensionless |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| *    |                                                 | Di          |
| *rat |                                                 | mensionless |
| eSca |                                                 |             |
| le** |                                                 |             |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***Derived Variables***

:   **rateScale** = q10Settings\[\*\]-\>q10(reduce method:
    multiply) (exposed as **rateScale**)

:   **q** = states\[\*\]-\>q(reduce method: add) (exposed as **q**)

:   **fcond** = q\^instances (exposed as **fcond**)
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="GateKS">
      <xs:complexContent>
        <xs:extension base="Base">
          <xs:sequence>
            <xs:element name="notes" type="Notes" minOccurs="0"/>
            <xs:element name="q10Settings" type="Q10Settings" minOccurs="0"/>
            <xs:element name="closedState" type="ClosedState" minOccurs="1" maxOccurs="unbounded"/>
            <xs:element name="openState" type="OpenState" minOccurs="1" maxOccurs="unbounded"/>
            <xs:choice minOccurs="1" maxOccurs="unbounded">
              <xs:group ref="ForwardReverseTransition"/>
              <xs:element name="tauInfTransition" type="TauInfTransition"/>
            </xs:choice>
          </xs:sequence>
          <xs:attribute name="instances" type="PositiveInteger" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=GateKS){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import GateKS
    from neuroml.utils import component_factory

    variable = component_factory(
        GateKS,
        id: 'a NmlId (required)' = None,
        instances: 'a PositiveInteger (required)' = None,
        notes: 'a string (optional)' = None,
        q10_settings: 'a Q10Settings (optional)' = None,
        closed_states: 'list of ClosedState(s) (required)' = None,
        open_states: 'list of OpenState(s) (required)' = None,
        forward_transition: 'list of ForwardTransition(s) (required)' = None,
        reverse_transition: 'list of ReverseTransition(s) (required)' = None,
        tau_inf_transition: 'list of TauInfTransition(s) (required)' = None,
    )
:::
:::
:::
:::
:::
:::

[]{#document-Userdocs/Schemas/Synapses}

::: {#synapses .section .tex2jax_ignore .mathjax_ignore}
[]{#schema-synapses}

#### Synapses[\#](#synapses "Link to this heading"){.headerlink}

**A number of synaptic ComponentTypes for use in NeuroML 2 documents,
e.g. [[expOneSynapse]{.std .std-ref}](#schema-exponesynapse){.reference
.internal}, [[expTwoSynapse]{.std
.std-ref}](#schema-exptwosynapse){.reference .internal},
[[blockingPlasticSynapse]{.std
.std-ref}](#schema-blockingplasticsynapse){.reference .internal}. These
extend the [[baseSynapse]{.std
.std-ref}](#schema-basesynapse){.reference .internal} ComponentType.
Also defined continuously transmitting synapses, e.g.
[[gapJunction]{.std .std-ref}](#schema-gapjunction){.reference
.internal} and [[gradedSynapse]{.std
.std-ref}](#schema-gradedsynapse){.reference .internal}.**

------------------------------------------------------------------------

Original ComponentType definitions:
[Synapses.xml](https://github.com/NeuroML/NeuroML2/blob/master/NeuroML2CoreTypes//Synapses.xml){.reference
.external}. Schema against which NeuroML based on these should be valid:
[NeuroML_v2.3.xsd](https://github.com/NeuroML/NeuroML2/tree/master/Schemas/NeuroML2/NeuroML_v2.3.xsd){.reference
.external}. Generated on 14/08/24 from
[this](https://github.com/NeuroML/NeuroML2/commit/352244cff605cb1ba24fa7c11757dc818fe90fd2){.reference
.external} commit. Please file any issues or questions at the [issue
tracker here](https://github.com/NeuroML/NeuroML2/issues){.reference
.external}.

------------------------------------------------------------------------

::: {#basesynapse .section}
[]{#schema-basesynapse}

##### *baseSynapse*[\#](#basesynapse "Link to this heading"){.headerlink}

extends *[[basePointCurrent]{.std
.std-ref}](Landing.html#schema-basepointcurrent){.reference .internal}*

*Base type for all synapses, i.e. ComponentTypes which produce a current
( dimension current ) and change Dynamics in response to an incoming
event.*

[Bioportal entry for Computational Neuroscience Ontology related to
baseSynapse.](https://bioportal.bioontology.org/ontologies/CNO/?p=classes&amp;conceptid=cno_0000009){.reference
.external}

::: {.sd-tab-set .docutils}
Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | The total (usually time varying) current        | [[cu        |
| *i** | produced by this ComponentType *(from           | rrent]{.std |
|      | [[basePointCurrent]{.std                        | .std-r      |
|      | .std-ref}](L                                    | ef}](Landin |
|      | anding.html#schema-basepointcurrent){.reference | g.html#sche |
|      | .internal})*                                    | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   |                                                 | Direction:  |
| in** |                                                 | in          |
+------+-------------------------------------------------+-------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="BaseSynapse">
      <xs:complexContent>
        <xs:extension base="Standalone">
          <xs:attribute name="neuroLexId" type="NeuroLexId" use="optional"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=BaseSynapse){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import BaseSynapse
    from neuroml.utils import component_factory

    variable = component_factory(
        BaseSynapse,
        id: 'a NmlId (required)' = None,
        metaid: 'a MetaId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        neuro_lex_id: 'a NeuroLexId (optional)' = None,
        extensiontype_=None,
    )
:::
:::
:::
:::
:::

::: {#basevoltagedepsynapse .section}
[]{#schema-basevoltagedepsynapse}

##### *baseVoltageDepSynapse*[\#](#basevoltagedepsynapse "Link to this heading"){.headerlink}

extends *[[baseSynapse]{.std .std-ref}](#schema-basesynapse){.reference
.internal}*

*Base type for synapses with a dependence on membrane potential.*

::: {.sd-tab-set .docutils}
Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | The total (usually time varying) current        | [[cu        |
| *i** | produced by this ComponentType *(from           | rrent]{.std |
|      | [[basePointCurrent]{.std                        | .std-r      |
|      | .std-ref}](L                                    | ef}](Landin |
|      | anding.html#schema-basepointcurrent){.reference | g.html#sche |
|      | .internal})*                                    | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | The current may vary with the voltage exposed   | [[vo        |
| *v** | by the ComponentType on which this is placed    | ltage]{.std |
|      |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   | *(from [[baseSynapse]{.std                      | Direction:  |
| in** | .std-ref}](#schema-basesynapse){.reference      | in          |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="BaseVoltageDepSynapse">
      <xs:complexContent>
        <xs:extension base="BaseSynapse">

                </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=BaseVoltageDepSynapse){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import BaseVoltageDepSynapse
    from neuroml.utils import component_factory

    variable = component_factory(
        BaseVoltageDepSynapse,
        id: 'a NmlId (required)' = None,
        metaid: 'a MetaId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        neuro_lex_id: 'a NeuroLexId (optional)' = None,
        extensiontype_=None,
    )
:::
:::
:::
:::
:::

::: {#basesynapsedl .section}
[]{#schema-basesynapsedl}

##### *baseSynapseDL*[\#](#basesynapsedl "Link to this heading"){.headerlink}

extends *[[baseVoltageDepPointCurrentDL]{.std
.std-ref}](Landing.html#schema-basevoltagedeppointcurrentdl){.reference
.internal}*

*Base type for all synapses, i.e. ComponentTypes which produce a
dimensionless current and change Dynamics in response to an incoming
event.*

[Bioportal entry for Computational Neuroscience Ontology related to
baseSynapseDL.](https://bioportal.bioontology.org/ontologies/CNO/?p=classes&amp;conceptid=cno_0000009){.reference
.external}

::: {.sd-tab-set .docutils}
Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | The total (time varying) current produced by    | Di          |
| *I** | this ComponentType *(from                       | mensionless |
|      | [[basePointCurrentDL]{.std                      |             |
|      | .std-ref}](Lan                                  |             |
|      | ding.html#schema-basepointcurrentdl){.reference |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | The current may vary with the dimensionless     | Di          |
| *V** | voltage exposed by the ComponentType on which   | mensionless |
|      | this is placed *(from                           |             |
|      | [[baseVoltageDepPointCurrentDL]{.std            |             |
|      | .std-ref}](Landing.html#                        |             |
|      | schema-basevoltagedeppointcurrentdl){.reference |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::
:::
:::

::: {#basecurrentbasedsynapse .section}
[]{#schema-basecurrentbasedsynapse}

##### *baseCurrentBasedSynapse*[\#](#basecurrentbasedsynapse "Link to this heading"){.headerlink}

extends *[[baseSynapse]{.std .std-ref}](#schema-basesynapse){.reference
.internal}*

*Synapse model which produces a synaptic current.*

::: {.sd-tab-set .docutils}
Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | The total (usually time varying) current        | [[cu        |
| *i** | produced by this ComponentType *(from           | rrent]{.std |
|      | [[basePointCurrent]{.std                        | .std-r      |
|      | .std-ref}](L                                    | ef}](Landin |
|      | anding.html#schema-basepointcurrent){.reference | g.html#sche |
|      | .internal})*                                    | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   | *(from [[baseSynapse]{.std                      | Direction:  |
| in** | .std-ref}](#schema-basesynapse){.reference      | in          |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="BaseCurrentBasedSynapse">
      <xs:complexContent>
        <xs:extension base="BaseSynapse">

                </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=BaseCurrentBasedSynapse){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import BaseCurrentBasedSynapse
    from neuroml.utils import component_factory

    variable = component_factory(
        BaseCurrentBasedSynapse,
        id: 'a NmlId (required)' = None,
        metaid: 'a MetaId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        neuro_lex_id: 'a NeuroLexId (optional)' = None,
        extensiontype_=None,
    )
:::
:::
:::
:::
:::

::: {#alphacurrentsynapse .section}
[]{#schema-alphacurrentsynapse}

##### alphaCurrentSynapse[\#](#alphacurrentsynapse "Link to this heading"){.headerlink}

extends *[[baseCurrentBasedSynapse]{.std
.std-ref}](#schema-basecurrentbasedsynapse){.reference .internal}*

*Alpha current synapse: rise time and decay time are both **tau.**.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Baseline current increase after receiving a     | [[cu        |
| *iba | spike                                           | rrent]{.std |
| se** |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **t  | Time course for rise and decay                  | [           |
| au** |                                                 | [time]{.std |
|      |                                                 | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Properties

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------------------+-----------------------------------+-------------+
| **weight**         |                                   | Di          |
| (default: 1)       |                                   | mensionless |
+--------------------+-----------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | The total (usually time varying) current        | [[cu        |
| *i** | produced by this ComponentType *(from           | rrent]{.std |
|      | [[basePointCurrent]{.std                        | .std-r      |
|      | .std-ref}](L                                    | ef}](Landin |
|      | anding.html#schema-basepointcurrent){.reference | g.html#sche |
|      | .internal})*                                    | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   | *(from [[baseSynapse]{.std                      | Direction:  |
| in** | .std-ref}](#schema-basesynapse){.reference      | in          |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***State Variables***

:   **I**: [[current]{.std
    .std-ref}](Landing.html#schema-dimensions-current){.reference
    .internal}

:   **J**: [[current]{.std
    .std-ref}](Landing.html#schema-dimensions-current){.reference
    .internal}

***On Start***

:   **I** = 0

:   **J** = 0

***On Events***

:   EVENT IN on port: **in**

:      **J** = J + weight \* ibase

***Derived Variables***

:   **i** = I (exposed as **i**)

***Time Derivatives***

:   d **I** /dt = (2.7182818284590451\*J - I)/tau

:   d **J** /dt = -J/tau
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="AlphaCurrentSynapse">
      <xs:complexContent>
        <xs:extension base="BaseCurrentBasedSynapse">
          <xs:attribute name="tau" type="Nml2Quantity_time" use="required"/>
          <xs:attribute name="ibase" type="Nml2Quantity_current" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=AlphaCurrentSynapse){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import AlphaCurrentSynapse
    from neuroml.utils import component_factory

    variable = component_factory(
        AlphaCurrentSynapse,
        id: 'a NmlId (required)' = None,
        metaid: 'a MetaId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        neuro_lex_id: 'a NeuroLexId (optional)' = None,
        tau: 'a Nml2Quantity_time (required)' = None,
        ibase: 'a Nml2Quantity_current (required)' = None,
    )
:::
:::
:::
:::
:::

::: {#baseconductancebasedsynapse .section}
[]{#schema-baseconductancebasedsynapse}

##### *baseConductanceBasedSynapse*[\#](#baseconductancebasedsynapse "Link to this heading"){.headerlink}

extends *[[baseVoltageDepSynapse]{.std
.std-ref}](#schema-basevoltagedepsynapse){.reference .internal}*

*Synapse model which exposes a conductance **g** in addition to
producing a current. Not necessarily ohmic!!*

[Bioportal entry for Computational Neuroscience Ontology related to
baseConductanceBasedSynapse.](https://bioportal.bioontology.org/ontologies/CNO/?p=classes&amp;conceptid=cno_0000027){.reference
.external}

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **er | Reversal potential of the synapse               | [[vo        |
| ev** |                                                 | ltage]{.std |
|      |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | Baseline conductance, generally the maximum     | [[conduc    |
| *gba | conductance following a single spike            | tance]{.std |
| se** |                                                 | .std-ref}]  |
|      |                                                 | (Landing.ht |
|      |                                                 | ml#schema-d |
|      |                                                 | imensions-c |
|      |                                                 | onductance) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Time varying conductance through the synapse    | [[conduc    |
| *g** |                                                 | tance]{.std |
|      |                                                 | .std-ref}]  |
|      |                                                 | (Landing.ht |
|      |                                                 | ml#schema-d |
|      |                                                 | imensions-c |
|      |                                                 | onductance) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | The total (usually time varying) current        | [[cu        |
| *i** | produced by this ComponentType *(from           | rrent]{.std |
|      | [[basePointCurrent]{.std                        | .std-r      |
|      | .std-ref}](L                                    | ef}](Landin |
|      | anding.html#schema-basepointcurrent){.reference | g.html#sche |
|      | .internal})*                                    | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | The current may vary with the voltage exposed   | [[vo        |
| *v** | by the ComponentType on which this is placed    | ltage]{.std |
|      | *(from [[baseVoltageDepSynapse]{.std            | .std-r      |
|      | .std-                                           | ef}](Landin |
|      | ref}](#schema-basevoltagedepsynapse){.reference | g.html#sche |
|      | .internal})*                                    | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   | *(from [[baseSynapse]{.std                      | Direction:  |
| in** | .std-ref}](#schema-basesynapse){.reference      | in          |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="BaseConductanceBasedSynapse">
      <xs:complexContent>
        <xs:extension base="BaseVoltageDepSynapse">
          <xs:attribute name="gbase" type="Nml2Quantity_conductance" use="required"/>
          <xs:attribute name="erev" type="Nml2Quantity_voltage" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=BaseConductanceBasedSynapse){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import BaseConductanceBasedSynapse
    from neuroml.utils import component_factory

    variable = component_factory(
        BaseConductanceBasedSynapse,
        id: 'a NmlId (required)' = None,
        metaid: 'a MetaId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        neuro_lex_id: 'a NeuroLexId (optional)' = None,
        gbase: 'a Nml2Quantity_conductance (required)' = None,
        erev: 'a Nml2Quantity_voltage (required)' = None,
        extensiontype_=None,
    )
:::
:::
:::
:::
:::

::: {#baseconductancebasedsynapsetwo .section}
[]{#schema-baseconductancebasedsynapsetwo}

##### *baseConductanceBasedSynapseTwo*[\#](#baseconductancebasedsynapsetwo "Link to this heading"){.headerlink}

extends *[[baseVoltageDepSynapse]{.std
.std-ref}](#schema-basevoltagedepsynapse){.reference .internal}*

*Synapse model suited for a sum of two expTwoSynapses which exposes a
conductance **g** in addition to producing a current. Not necessarily
ohmic!!*

[Bioportal entry for Computational Neuroscience Ontology related to
baseConductanceBasedSynapseTwo.](https://bioportal.bioontology.org/ontologies/CNO/?p=classes&amp;conceptid=cno_0000027){.reference
.external}

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **er | Reversal potential of the synapse               | [[vo        |
| ev** |                                                 | ltage]{.std |
|      |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   | Baseline conductance 1                          | [[conduc    |
| gbas |                                                 | tance]{.std |
| e1** |                                                 | .std-ref}]  |
|      |                                                 | (Landing.ht |
|      |                                                 | ml#schema-d |
|      |                                                 | imensions-c |
|      |                                                 | onductance) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   | Baseline conductance 2                          | [[conduc    |
| gbas |                                                 | tance]{.std |
| e2** |                                                 | .std-ref}]  |
|      |                                                 | (Landing.ht |
|      |                                                 | ml#schema-d |
|      |                                                 | imensions-c |
|      |                                                 | onductance) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Time varying conductance through the synapse    | [[conduc    |
| *g** |                                                 | tance]{.std |
|      |                                                 | .std-ref}]  |
|      |                                                 | (Landing.ht |
|      |                                                 | ml#schema-d |
|      |                                                 | imensions-c |
|      |                                                 | onductance) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | The total (usually time varying) current        | [[cu        |
| *i** | produced by this ComponentType *(from           | rrent]{.std |
|      | [[basePointCurrent]{.std                        | .std-r      |
|      | .std-ref}](L                                    | ef}](Landin |
|      | anding.html#schema-basepointcurrent){.reference | g.html#sche |
|      | .internal})*                                    | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | The current may vary with the voltage exposed   | [[vo        |
| *v** | by the ComponentType on which this is placed    | ltage]{.std |
|      | *(from [[baseVoltageDepSynapse]{.std            | .std-r      |
|      | .std-                                           | ef}](Landin |
|      | ref}](#schema-basevoltagedepsynapse){.reference | g.html#sche |
|      | .internal})*                                    | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   | *(from [[baseSynapse]{.std                      | Direction:  |
| in** | .std-ref}](#schema-basesynapse){.reference      | in          |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="BaseConductanceBasedSynapseTwo">
      <xs:complexContent>
        <xs:extension base="BaseVoltageDepSynapse">
          <xs:attribute name="gbase1" type="Nml2Quantity_conductance" use="required"/>
          <xs:attribute name="gbase2" type="Nml2Quantity_conductance" use="required"/>
          <xs:attribute name="erev" type="Nml2Quantity_voltage" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=BaseConductanceBasedSynapseTwo){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import BaseConductanceBasedSynapseTwo
    from neuroml.utils import component_factory

    variable = component_factory(
        BaseConductanceBasedSynapseTwo,
        id: 'a NmlId (required)' = None,
        metaid: 'a MetaId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        neuro_lex_id: 'a NeuroLexId (optional)' = None,
        gbase1: 'a Nml2Quantity_conductance (required)' = None,
        gbase2: 'a Nml2Quantity_conductance (required)' = None,
        erev: 'a Nml2Quantity_voltage (required)' = None,
        extensiontype_=None,
    )
:::
:::
:::
:::
:::

::: {#exponesynapse .section}
[]{#schema-exponesynapse}

##### expOneSynapse[\#](#exponesynapse "Link to this heading"){.headerlink}

extends *[[baseConductanceBasedSynapse]{.std
.std-ref}](#schema-baseconductancebasedsynapse){.reference .internal}*

*Ohmic synapse model whose conductance rises instantaneously by (
**gbase** \* **weight** ) on receiving an event, and which decays
exponentially to zero with time course **tauDecay**.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **er | Reversal potential of the synapse *(from        | [[vo        |
| ev** | [[baseConductanceBasedSynapse]{.std             | ltage]{.std |
|      | .std-ref}](                                     | .std-r      |
|      | #schema-baseconductancebasedsynapse){.reference | ef}](Landin |
|      | .internal})*                                    | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | Baseline conductance, generally the maximum     | [[conduc    |
| *gba | conductance following a single spike *(from     | tance]{.std |
| se** | [[baseConductanceBasedSynapse]{.std             | .std-ref}]  |
|      | .std-ref}](                                     | (Landing.ht |
|      | #schema-baseconductancebasedsynapse){.reference | ml#schema-d |
|      | .internal})*                                    | imensions-c |
|      |                                                 | onductance) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **ta | Time course of decay                            | [           |
| uDec |                                                 | [time]{.std |
| ay** |                                                 | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Properties

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------------------+-----------------------------------+-------------+
| **weight**         |                                   | Di          |
| (default: 1)       |                                   | mensionless |
+--------------------+-----------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Time varying conductance through the synapse    | [[conduc    |
| *g** | *(from [[baseConductanceBasedSynapse]{.std      | tance]{.std |
|      | .std-ref}](                                     | .std-ref}]  |
|      | #schema-baseconductancebasedsynapse){.reference | (Landing.ht |
|      | .internal})*                                    | ml#schema-d |
|      |                                                 | imensions-c |
|      |                                                 | onductance) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | The total (usually time varying) current        | [[cu        |
| *i** | produced by this ComponentType *(from           | rrent]{.std |
|      | [[basePointCurrent]{.std                        | .std-r      |
|      | .std-ref}](L                                    | ef}](Landin |
|      | anding.html#schema-basepointcurrent){.reference | g.html#sche |
|      | .internal})*                                    | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | The current may vary with the voltage exposed   | [[vo        |
| *v** | by the ComponentType on which this is placed    | ltage]{.std |
|      | *(from [[baseVoltageDepSynapse]{.std            | .std-r      |
|      | .std-                                           | ef}](Landin |
|      | ref}](#schema-basevoltagedepsynapse){.reference | g.html#sche |
|      | .internal})*                                    | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   | *(from [[baseSynapse]{.std                      | Direction:  |
| in** | .std-ref}](#schema-basesynapse){.reference      | in          |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***State Variables***

:   **g**: [[conductance]{.std
    .std-ref}](Landing.html#schema-dimensions-conductance){.reference
    .internal}  (exposed as **g**)

***On Start***

:   **g** = 0

***On Events***

:   EVENT IN on port: **in**

:      **g** = g + (weight \* gbase)

***Derived Variables***

:   **i** = g \* (erev - v) (exposed as **i**)

***Time Derivatives***

:   d **g** /dt = -g / tauDecay
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="ExpOneSynapse">
      <xs:complexContent>
        <xs:extension base="BaseConductanceBasedSynapse">
          <xs:attribute name="tauDecay" type="Nml2Quantity_time" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=ExpOneSynapse){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import ExpOneSynapse
    from neuroml.utils import component_factory

    variable = component_factory(
        ExpOneSynapse,
        id: 'a NmlId (required)' = None,
        metaid: 'a MetaId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        neuro_lex_id: 'a NeuroLexId (optional)' = None,
        gbase: 'a Nml2Quantity_conductance (required)' = None,
        erev: 'a Nml2Quantity_voltage (required)' = None,
        tau_decay: 'a Nml2Quantity_time (required)' = None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <expOneSynapse id="syn1" gbase="5nS" erev="0mV" tauDecay="3ms"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <expOneSynapse id="syn2" gbase="10nS" erev="0mV" tauDecay="2ms"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <expOneSynapse id="syn1" gbase="5nS" erev="0mV" tauDecay="3ms"/>
:::
:::
:::
:::
:::

::: {#alphasynapse .section}
[]{#schema-alphasynapse}

##### alphaSynapse[\#](#alphasynapse "Link to this heading"){.headerlink}

extends *[[baseConductanceBasedSynapse]{.std
.std-ref}](#schema-baseconductancebasedsynapse){.reference .internal}*

*Ohmic synapse model where rise time and decay time are both **tau.**
Max conductance reached during this time ( assuming zero conductance
before ) is **gbase** \* **weight.**.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **er | Reversal potential of the synapse *(from        | [[vo        |
| ev** | [[baseConductanceBasedSynapse]{.std             | ltage]{.std |
|      | .std-ref}](                                     | .std-r      |
|      | #schema-baseconductancebasedsynapse){.reference | ef}](Landin |
|      | .internal})*                                    | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | Baseline conductance, generally the maximum     | [[conduc    |
| *gba | conductance following a single spike *(from     | tance]{.std |
| se** | [[baseConductanceBasedSynapse]{.std             | .std-ref}]  |
|      | .std-ref}](                                     | (Landing.ht |
|      | #schema-baseconductancebasedsynapse){.reference | ml#schema-d |
|      | .internal})*                                    | imensions-c |
|      |                                                 | onductance) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **t  | Time course of rise/decay                       | [           |
| au** |                                                 | [time]{.std |
|      |                                                 | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Properties

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------------------+-----------------------------------+-------------+
| **weight**         |                                   | Di          |
| (default: 1)       |                                   | mensionless |
+--------------------+-----------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Time varying conductance through the synapse    | [[conduc    |
| *g** | *(from [[baseConductanceBasedSynapse]{.std      | tance]{.std |
|      | .std-ref}](                                     | .std-ref}]  |
|      | #schema-baseconductancebasedsynapse){.reference | (Landing.ht |
|      | .internal})*                                    | ml#schema-d |
|      |                                                 | imensions-c |
|      |                                                 | onductance) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | The total (usually time varying) current        | [[cu        |
| *i** | produced by this ComponentType *(from           | rrent]{.std |
|      | [[basePointCurrent]{.std                        | .std-r      |
|      | .std-ref}](L                                    | ef}](Landin |
|      | anding.html#schema-basepointcurrent){.reference | g.html#sche |
|      | .internal})*                                    | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | The current may vary with the voltage exposed   | [[vo        |
| *v** | by the ComponentType on which this is placed    | ltage]{.std |
|      | *(from [[baseVoltageDepSynapse]{.std            | .std-r      |
|      | .std-                                           | ef}](Landin |
|      | ref}](#schema-basevoltagedepsynapse){.reference | g.html#sche |
|      | .internal})*                                    | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   | *(from [[baseSynapse]{.std                      | Direction:  |
| in** | .std-ref}](#schema-basesynapse){.reference      | in          |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***State Variables***

:   **g**: [[conductance]{.std
    .std-ref}](Landing.html#schema-dimensions-conductance){.reference
    .internal}  (exposed as **g**)

:   **A**: [[conductance]{.std
    .std-ref}](Landing.html#schema-dimensions-conductance){.reference
    .internal}

***On Start***

:   **g** = 0

:   **A** = 0

***On Events***

:   EVENT IN on port: **in**

:      **A** = A + (gbase\*weight)

***Derived Variables***

:   **i** = g \* (erev - v) (exposed as **i**)

***Time Derivatives***

:   d **g** /dt = (2.7182818284590451 \* A - g)/tau

:   d **A** /dt = -A / tau
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="AlphaSynapse">
      <xs:complexContent>
        <xs:extension base="BaseConductanceBasedSynapse">
          <xs:attribute name="tau" type="Nml2Quantity_time" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=AlphaSynapse){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import AlphaSynapse
    from neuroml.utils import component_factory

    variable = component_factory(
        AlphaSynapse,
        id: 'a NmlId (required)' = None,
        metaid: 'a MetaId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        neuro_lex_id: 'a NeuroLexId (optional)' = None,
        gbase: 'a Nml2Quantity_conductance (required)' = None,
        erev: 'a Nml2Quantity_voltage (required)' = None,
        tau: 'a Nml2Quantity_time (required)' = None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <alphaSynapse id="synalpha" gbase="0.5nS" erev="0mV" tau="2ms">
        <notes>An alpha synapse with time for rise equal to decay.</notes>
    </alphaSynapse>
:::
:::
:::
:::
:::

::: {#exptwosynapse .section}
[]{#schema-exptwosynapse}

##### expTwoSynapse[\#](#exptwosynapse "Link to this heading"){.headerlink}

extends *[[baseConductanceBasedSynapse]{.std
.std-ref}](#schema-baseconductancebasedsynapse){.reference .internal}*

*Ohmic synapse model whose conductance waveform on receiving an event
has a rise time of **tauRise** and a decay time of **tauDecay.** Max
conductance reached during this time ( assuming zero conductance before
) is **gbase** \* **weight.**.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **er | Reversal potential of the synapse *(from        | [[vo        |
| ev** | [[baseConductanceBasedSynapse]{.std             | ltage]{.std |
|      | .std-ref}](                                     | .std-r      |
|      | #schema-baseconductancebasedsynapse){.reference | ef}](Landin |
|      | .internal})*                                    | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | Baseline conductance, generally the maximum     | [[conduc    |
| *gba | conductance following a single spike *(from     | tance]{.std |
| se** | [[baseConductanceBasedSynapse]{.std             | .std-ref}]  |
|      | .std-ref}](                                     | (Landing.ht |
|      | #schema-baseconductancebasedsynapse){.reference | ml#schema-d |
|      | .internal})*                                    | imensions-c |
|      |                                                 | onductance) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **ta |                                                 | [           |
| uDec |                                                 | [time]{.std |
| ay** |                                                 | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **t  |                                                 | [           |
| auRi |                                                 | [time]{.std |
| se** |                                                 | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Derived parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **pe |                                                 | [           |
| akTi |                                                 | [time]{.std |
| me** |                                                 | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::

   **peakTime** = log(tauDecay / tauRise) \* (tauRise \*
tauDecay)/(tauDecay - tauRise)

::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   |                                                 | Di          |
| wave |                                                 | mensionless |
| form |                                                 |             |
| Fact |                                                 |             |
| or** |                                                 |             |
+------+-------------------------------------------------+-------------+
:::

   **waveformFactor** = 1 / (-exp(-peakTime / tauRise) + exp(-peakTime /
tauDecay))
:::

Properties

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------------------+-----------------------------------+-------------+
| **weight**         |                                   | Di          |
| (default: 1)       |                                   | mensionless |
+--------------------+-----------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Time varying conductance through the synapse    | [[conduc    |
| *g** | *(from [[baseConductanceBasedSynapse]{.std      | tance]{.std |
|      | .std-ref}](                                     | .std-ref}]  |
|      | #schema-baseconductancebasedsynapse){.reference | (Landing.ht |
|      | .internal})*                                    | ml#schema-d |
|      |                                                 | imensions-c |
|      |                                                 | onductance) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | The total (usually time varying) current        | [[cu        |
| *i** | produced by this ComponentType *(from           | rrent]{.std |
|      | [[basePointCurrent]{.std                        | .std-r      |
|      | .std-ref}](L                                    | ef}](Landin |
|      | anding.html#schema-basepointcurrent){.reference | g.html#sche |
|      | .internal})*                                    | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | The current may vary with the voltage exposed   | [[vo        |
| *v** | by the ComponentType on which this is placed    | ltage]{.std |
|      | *(from [[baseVoltageDepSynapse]{.std            | .std-r      |
|      | .std-                                           | ef}](Landin |
|      | ref}](#schema-basevoltagedepsynapse){.reference | g.html#sche |
|      | .internal})*                                    | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   | *(from [[baseSynapse]{.std                      | Direction:  |
| in** | .std-ref}](#schema-basesynapse){.reference      | in          |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***State Variables***

:   **A**: Dimensionless

:   **B**: Dimensionless

***On Start***

:   **A** = 0

:   **B** = 0

***On Events***

:   EVENT IN on port: **in**

:      **A** = A + (weight \* waveformFactor)

:      **B** = B + (weight \* waveformFactor)

***Derived Variables***

:   **g** = gbase \* (B - A) (exposed as **g**)

:   **i** = g \* (erev - v) (exposed as **i**)

***Time Derivatives***

:   d **A** /dt = -A / tauRise

:   d **B** /dt = -B / tauDecay
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="ExpTwoSynapse">
      <xs:complexContent>
        <xs:extension base="BaseConductanceBasedSynapse">
          <xs:attribute name="tauDecay" type="Nml2Quantity_time" use="required"/>
          <xs:attribute name="tauRise" type="Nml2Quantity_time" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=ExpTwoSynapse){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import ExpTwoSynapse
    from neuroml.utils import component_factory

    variable = component_factory(
        ExpTwoSynapse,
        id: 'a NmlId (required)' = None,
        metaid: 'a MetaId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        neuro_lex_id: 'a NeuroLexId (optional)' = None,
        gbase: 'a Nml2Quantity_conductance (required)' = None,
        erev: 'a Nml2Quantity_voltage (required)' = None,
        tau_decay: 'a Nml2Quantity_time (required)' = None,
        tau_rise: 'a Nml2Quantity_time (required)' = None,
        extensiontype_=None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <expTwoSynapse id="AMPA" gbase="0.5nS" erev="0mV" tauRise="1ms" tauDecay="2ms"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <expTwoSynapse id="synInput" gbase="8nS" erev="20mV" tauRise="1ms" tauDecay="5ms"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <expTwoSynapse id="synInputFast" gbase="1nS" erev="20mV" tauRise="0.2ms" tauDecay="1ms"/>
:::
:::
:::
:::
:::

::: {#expthreesynapse .section}
[]{#schema-expthreesynapse}

##### expThreeSynapse[\#](#expthreesynapse "Link to this heading"){.headerlink}

extends *[[baseConductanceBasedSynapseTwo]{.std
.std-ref}](#schema-baseconductancebasedsynapsetwo){.reference
.internal}*

*Ohmic synapse similar to expTwoSynapse but consisting of two components
that can differ in decay times and max conductances but share the same
rise time.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **er | Reversal potential of the synapse *(from        | [[vo        |
| ev** | [[baseConductanceBasedSynapseTwo]{.std          | ltage]{.std |
|      | .std-ref}](#sc                                  | .std-r      |
|      | hema-baseconductancebasedsynapsetwo){.reference | ef}](Landin |
|      | .internal})*                                    | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   | Baseline conductance 1 *(from                   | [[conduc    |
| gbas | [[baseConductanceBasedSynapseTwo]{.std          | tance]{.std |
| e1** | .std-ref}](#sc                                  | .std-ref}]  |
|      | hema-baseconductancebasedsynapsetwo){.reference | (Landing.ht |
|      | .internal})*                                    | ml#schema-d |
|      |                                                 | imensions-c |
|      |                                                 | onductance) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   | Baseline conductance 2 *(from                   | [[conduc    |
| gbas | [[baseConductanceBasedSynapseTwo]{.std          | tance]{.std |
| e2** | .std-ref}](#sc                                  | .std-ref}]  |
|      | hema-baseconductancebasedsynapsetwo){.reference | (Landing.ht |
|      | .internal})*                                    | ml#schema-d |
|      |                                                 | imensions-c |
|      |                                                 | onductance) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    |                                                 | [           |
| *tau |                                                 | [time]{.std |
| Deca |                                                 | .st         |
| y1** |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    |                                                 | [           |
| *tau |                                                 | [time]{.std |
| Deca |                                                 | .st         |
| y2** |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **t  |                                                 | [           |
| auRi |                                                 | [time]{.std |
| se** |                                                 | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Derived parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    |                                                 | [           |
| *pea |                                                 | [time]{.std |
| kTim |                                                 | .st         |
| e1** |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::

   **peakTime1** = log(tauDecay1 / tauRise) \* (tauRise \*
tauDecay1)/(tauDecay1 - tauRise)

::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    |                                                 | [           |
| *pea |                                                 | [time]{.std |
| kTim |                                                 | .st         |
| e2** |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::

   **peakTime2** = log(tauDecay2 / tauRise) \* (tauRise \*
tauDecay2)/(tauDecay2 - tauRise)

::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **w  |                                                 | Di          |
| avef |                                                 | mensionless |
| ormF |                                                 |             |
| acto |                                                 |             |
| r1** |                                                 |             |
+------+-------------------------------------------------+-------------+
:::

   **waveformFactor1** = 1 / (-exp(-peakTime1 / tauRise) +
exp(-peakTime1 / tauDecay1))

::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **w  |                                                 | Di          |
| avef |                                                 | mensionless |
| ormF |                                                 |             |
| acto |                                                 |             |
| r2** |                                                 |             |
+------+-------------------------------------------------+-------------+
:::

   **waveformFactor2** = 1 / (-exp(-peakTime2 / tauRise) +
exp(-peakTime2 / tauDecay2))
:::

Properties

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------------------+-----------------------------------+-------------+
| **weight**         |                                   | Di          |
| (default: 1)       |                                   | mensionless |
+--------------------+-----------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Time varying conductance through the synapse    | [[conduc    |
| *g** | *(from [[baseConductanceBasedSynapseTwo]{.std   | tance]{.std |
|      | .std-ref}](#sc                                  | .std-ref}]  |
|      | hema-baseconductancebasedsynapsetwo){.reference | (Landing.ht |
|      | .internal})*                                    | ml#schema-d |
|      |                                                 | imensions-c |
|      |                                                 | onductance) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | The total (usually time varying) current        | [[cu        |
| *i** | produced by this ComponentType *(from           | rrent]{.std |
|      | [[basePointCurrent]{.std                        | .std-r      |
|      | .std-ref}](L                                    | ef}](Landin |
|      | anding.html#schema-basepointcurrent){.reference | g.html#sche |
|      | .internal})*                                    | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | The current may vary with the voltage exposed   | [[vo        |
| *v** | by the ComponentType on which this is placed    | ltage]{.std |
|      | *(from [[baseVoltageDepSynapse]{.std            | .std-r      |
|      | .std-                                           | ef}](Landin |
|      | ref}](#schema-basevoltagedepsynapse){.reference | g.html#sche |
|      | .internal})*                                    | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   | *(from [[baseSynapse]{.std                      | Direction:  |
| in** | .std-ref}](#schema-basesynapse){.reference      | in          |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***State Variables***

:   **A**: Dimensionless

:   **B**: Dimensionless

:   **C**: Dimensionless

***On Start***

:   **A** = 0

:   **B** = 0

:   **C** = 0

***On Events***

:   EVENT IN on port: **in**

:      **A** = A + (gbase1*weight \* waveformFactor1 +
    gbase2*weight\*waveformFactor2 )/(gbase1+gbase2)

:      **B** = B + (weight \* waveformFactor1)

:      **C** = C + (weight \* waveformFactor2)

***Derived Variables***

:   **g** = gbase1\*(B - A) + gbase2\*(C-A) (exposed as **g**)

:   **i** = g \* (erev - v) (exposed as **i**)

***Time Derivatives***

:   d **A** /dt = -A / tauRise

:   d **B** /dt = -B / tauDecay1

:   d **C** /dt = -C / tauDecay2
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="ExpThreeSynapse">
      <xs:complexContent>
        <xs:extension base="BaseConductanceBasedSynapseTwo">
          <xs:attribute name="tauDecay1" type="Nml2Quantity_time" use="required"/>
          <xs:attribute name="tauDecay2" type="Nml2Quantity_time" use="required"/>
          <xs:attribute name="tauRise" type="Nml2Quantity_time" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=ExpThreeSynapse){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import ExpThreeSynapse
    from neuroml.utils import component_factory

    variable = component_factory(
        ExpThreeSynapse,
        id: 'a NmlId (required)' = None,
        metaid: 'a MetaId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        neuro_lex_id: 'a NeuroLexId (optional)' = None,
        gbase1: 'a Nml2Quantity_conductance (required)' = None,
        gbase2: 'a Nml2Quantity_conductance (required)' = None,
        erev: 'a Nml2Quantity_voltage (required)' = None,
        tau_decay1: 'a Nml2Quantity_time (required)' = None,
        tau_decay2: 'a Nml2Quantity_time (required)' = None,
        tau_rise: 'a Nml2Quantity_time (required)' = None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <expThreeSynapse id="synInputFastTwo" gbase1="1.5nS" tauRise="0.1ms" tauDecay1="0.7ms" gbase2="0.5nS" tauDecay2="2.5ms" erev="0mV"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <expThreeSynapse id="AMPA" gbase1="1.5nS" tauRise="0.1ms" tauDecay1="0.7ms" gbase2="0.5nS" tauDecay2="2.5ms" erev="0mV">
        <notes>A synapse consisting of one rise and two decay time courses.</notes>
    </expThreeSynapse>
:::
:::
:::
:::
:::

::: {#baseblockmechanism .section}
[]{#schema-baseblockmechanism}

##### *baseBlockMechanism*[\#](#baseblockmechanism "Link to this heading"){.headerlink}

*Base of any ComponentType which produces a varying scaling ( or
blockage ) of synaptic strength of magnitude **scaling**.*

::: {.sd-tab-set .docutils}
Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **b  |                                                 | Di          |
| lock |                                                 | mensionless |
| Fact |                                                 |             |
| or** |                                                 |             |
+------+-------------------------------------------------+-------------+
:::
:::
:::
:::

::: {#voltageconcdepblockmechanism .section}
[]{#schema-voltageconcdepblockmechanism}

##### voltageConcDepBlockMechanism[\#](#voltageconcdepblockmechanism "Link to this heading"){.headerlink}

extends *[[baseBlockMechanism]{.std
.std-ref}](#schema-baseblockmechanism){.reference .internal}*

*Synaptic blocking mechanism which varys with membrane potential across
the synapse, e.g. in NMDA receptor mediated synapses.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   |                                                 | [[concentr  |
| bloc |                                                 | ation]{.std |
| kCon |                                                 | .           |
| cent |                                                 | std-ref}](L |
| rati |                                                 | anding.html |
| on** |                                                 | #schema-dim |
|      |                                                 | ensions-con |
|      |                                                 | centration) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **s  |                                                 | [[concentr  |
| cali |                                                 | ation]{.std |
| ngCo |                                                 | .           |
| nc** |                                                 | std-ref}](L |
|      |                                                 | anding.html |
|      |                                                 | #schema-dim |
|      |                                                 | ensions-con |
|      |                                                 | centration) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **s  |                                                 | [[vo        |
| cali |                                                 | ltage]{.std |
| ngVo |                                                 | .std-r      |
| lt** |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Text fields

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| **spe  |                                                              |
| cies** |                                                              |
+--------+--------------------------------------------------------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **b  | *(from [[baseBlockMechanism]{.std               | Di          |
| lock | .s                                              | mensionless |
| Fact | td-ref}](#schema-baseblockmechanism){.reference |             |
| or** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    |                                                 | [[vo        |
| *v** |                                                 | ltage]{.std |
|      |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***Derived Variables***

:   **blockFactor** = 1/(1 + (blockConcentration / scalingConc)\* exp(-1
    \* (v / scalingVolt))) (exposed as **blockFactor**)
:::
:::
:::

::: {#baseplasticitymechanism .section}
[]{#schema-baseplasticitymechanism}

##### *basePlasticityMechanism*[\#](#baseplasticitymechanism "Link to this heading"){.headerlink}

*Base plasticity mechanism.*

::: {.sd-tab-set .docutils}
Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **pl |                                                 | Di          |
| asti |                                                 | mensionless |
| city |                                                 |             |
| Fact |                                                 |             |
| or** |                                                 |             |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   | This is where the plasticity mechanism receives | Direction:  |
| in** | spike events from the parent synapse.           | in          |
+------+-------------------------------------------------+-------------+
:::
:::
:::
:::

::: {#tsodyksmarkramdepmechanism .section}
[]{#schema-tsodyksmarkramdepmechanism}

##### tsodyksMarkramDepMechanism[\#](#tsodyksmarkramdepmechanism "Link to this heading"){.headerlink}

extends *[[basePlasticityMechanism]{.std
.std-ref}](#schema-baseplasticitymechanism){.reference .internal}*

*Depression-only Tsodyks-Markram model, as in Tsodyks and Markram 1997.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **i  |                                                 | Di          |
| nitR |                                                 | mensionless |
| elea |                                                 |             |
| sePr |                                                 |             |
| ob** |                                                 |             |
+------+-------------------------------------------------+-------------+
| **   |                                                 | [           |
| tauR |                                                 | [time]{.std |
| ec** |                                                 | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **pl | *(from [[basePlasticityMechanism]{.std          | Di          |
| asti | .std-re                                         | mensionless |
| city | f}](#schema-baseplasticitymechanism){.reference |             |
| Fact | .internal})*                                    |             |
| or** |                                                 |             |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   | This is where the plasticity mechanism receives | Direction:  |
| in** | spike events from the parent synapse. *(from    | in          |
|      | [[basePlasticityMechanism]{.std                 |             |
|      | .std-re                                         |             |
|      | f}](#schema-baseplasticitymechanism){.reference |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***Structure***

:   WITH **parent** AS **a**

:   WITH **this** AS **b**

:   EVENT CONNECTION from **a** TO **b**

***State Variables***

:   **R**: Dimensionless

***On Start***

:   **R** = 1

***On Events***

:   EVENT IN on port: **in**

:      **R** = R \* (1 - U)

***Derived Variables***

:   **U** = initReleaseProb

:   **plasticityFactor** = R \* U (exposed as **plasticityFactor**)

***Time Derivatives***

:   d **R** /dt = (1 - R) / tauRec
:::
:::
:::

::: {#tsodyksmarkramdepfacmechanism .section}
[]{#schema-tsodyksmarkramdepfacmechanism}

##### tsodyksMarkramDepFacMechanism[\#](#tsodyksmarkramdepfacmechanism "Link to this heading"){.headerlink}

extends *[[basePlasticityMechanism]{.std
.std-ref}](#schema-baseplasticitymechanism){.reference .internal}*

*Full Tsodyks-Markram STP model with both depression and facilitation,
as in Tsodyks, Pawelzik and Markram 1998.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **i  |                                                 | Di          |
| nitR |                                                 | mensionless |
| elea |                                                 |             |
| sePr |                                                 |             |
| ob** |                                                 |             |
+------+-------------------------------------------------+-------------+
| **   |                                                 | [           |
| tauF |                                                 | [time]{.std |
| ac** |                                                 | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   |                                                 | [           |
| tauR |                                                 | [time]{.std |
| ec** |                                                 | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **pl | *(from [[basePlasticityMechanism]{.std          | Di          |
| asti | .std-re                                         | mensionless |
| city | f}](#schema-baseplasticitymechanism){.reference |             |
| Fact | .internal})*                                    |             |
| or** |                                                 |             |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   | This is where the plasticity mechanism receives | Direction:  |
| in** | spike events from the parent synapse. *(from    | in          |
|      | [[basePlasticityMechanism]{.std                 |             |
|      | .std-re                                         |             |
|      | f}](#schema-baseplasticitymechanism){.reference |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***Structure***

:   WITH **parent** AS **a**

:   WITH **this** AS **b**

:   EVENT CONNECTION from **a** TO **b**

***State Variables***

:   **R**: Dimensionless

:   **U**: Dimensionless

***On Start***

:   **R** = 1

:   **U** = initReleaseProb

***On Events***

:   EVENT IN on port: **in**

:      **R** = R \* (1 - U)

:      **U** = U + initReleaseProb \* (1 - U)

***Derived Variables***

:   **plasticityFactor** = R \* U (exposed as **plasticityFactor**)

***Time Derivatives***

:   d **R** /dt = (1 - R) / tauRec

:   d **U** /dt = (initReleaseProb - U) / tauFac
:::
:::
:::

::: {#blockingplasticsynapse .section}
[]{#schema-blockingplasticsynapse}

##### blockingPlasticSynapse[\#](#blockingplasticsynapse "Link to this heading"){.headerlink}

extends [[expTwoSynapse]{.std
.std-ref}](#schema-exptwosynapse){.reference .internal}

*Biexponential synapse that allows for optional block and plasticity
mechanisms, which can be expressed as child elements.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **er | Reversal potential of the synapse *(from        | [[vo        |
| ev** | [[baseConductanceBasedSynapse]{.std             | ltage]{.std |
|      | .std-ref}](                                     | .std-r      |
|      | #schema-baseconductancebasedsynapse){.reference | ef}](Landin |
|      | .internal})*                                    | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | Baseline conductance, generally the maximum     | [[conduc    |
| *gba | conductance following a single spike *(from     | tance]{.std |
| se** | [[baseConductanceBasedSynapse]{.std             | .std-ref}]  |
|      | .std-ref}](                                     | (Landing.ht |
|      | #schema-baseconductancebasedsynapse){.reference | ml#schema-d |
|      | .internal})*                                    | imensions-c |
|      |                                                 | onductance) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **ta | *(from [[expTwoSynapse]{.std                    | [           |
| uDec | .std-ref}](#schema-exptwosynapse){.reference    | [time]{.std |
| ay** | .internal})*                                    | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **t  | *(from [[expTwoSynapse]{.std                    | [           |
| auRi | .std-ref}](#schema-exptwosynapse){.reference    | [time]{.std |
| se** | .internal})*                                    | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Derived parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **pe | *(from [[expTwoSynapse]{.std                    | [           |
| akTi | .std-ref}](#schema-exptwosynapse){.reference    | [time]{.std |
| me** | .internal})*                                    | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::

   **peakTime** = log(tauDecay / tauRise) \* (tauRise \*
tauDecay)/(tauDecay - tauRise)

::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   | *(from [[expTwoSynapse]{.std                    | Di          |
| wave | .std-ref}](#schema-exptwosynapse){.reference    | mensionless |
| form | .internal})*                                    |             |
| Fact |                                                 |             |
| or** |                                                 |             |
+------+-------------------------------------------------+-------------+
:::

   **waveformFactor** = 1 / (-exp(-peakTime / tauRise) + exp(-peakTime /
tauDecay))
:::

Children list

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **pl |                                                 | [[basePla   |
| asti |                                                 | sticityMech |
| city |                                                 | anism]{.std |
| Mech |                                                 | .std-ref}]  |
| anis |                                                 | (#schema-ba |
| ms** |                                                 | seplasticit |
|      |                                                 | ymechanism) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **b  |                                                 | [[ba        |
| lock |                                                 | seBlockMech |
| Mech |                                                 | anism]{.std |
| anis |                                                 | .std-       |
| ms** |                                                 | ref}](#sche |
|      |                                                 | ma-basebloc |
|      |                                                 | kmechanism) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Properties

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------------------+-----------------------------------+-------------+
| **weight**         |                                   | Di          |
| (default: 1)       |                                   | mensionless |
+--------------------+-----------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Time varying conductance through the synapse    | [[conduc    |
| *g** | *(from [[baseConductanceBasedSynapse]{.std      | tance]{.std |
|      | .std-ref}](                                     | .std-ref}]  |
|      | #schema-baseconductancebasedsynapse){.reference | (Landing.ht |
|      | .internal})*                                    | ml#schema-d |
|      |                                                 | imensions-c |
|      |                                                 | onductance) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | The total (usually time varying) current        | [[cu        |
| *i** | produced by this ComponentType *(from           | rrent]{.std |
|      | [[basePointCurrent]{.std                        | .std-r      |
|      | .std-ref}](L                                    | ef}](Landin |
|      | anding.html#schema-basepointcurrent){.reference | g.html#sche |
|      | .internal})*                                    | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | The current may vary with the voltage exposed   | [[vo        |
| *v** | by the ComponentType on which this is placed    | ltage]{.std |
|      | *(from [[baseVoltageDepSynapse]{.std            | .std-r      |
|      | .std-                                           | ef}](Landin |
|      | ref}](#schema-basevoltagedepsynapse){.reference | g.html#sche |
|      | .internal})*                                    | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   | *(from [[baseSynapse]{.std                      | Direction:  |
| in** | .std-ref}](#schema-basesynapse){.reference      | in          |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| *    | Used to relay incoming spikes to child          | Direction:  |
| *rel | plasticity mechanism                            | out         |
| ay** |                                                 |             |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***State Variables***

:   **A**: Dimensionless

:   **B**: Dimensionless

***On Start***

:   **A** = 0

:   **B** = 0

***On Events***

:   EVENT IN on port: **in**

:      **A** = A + (weight \* plasticityFactor \* waveformFactor)

:      **B** = B + (weight \* plasticityFactor \* waveformFactor)

:      EVENT OUT on port: **relay**

***Derived Variables***

:   **plasticityFactor**
    = plasticityMechanisms\[\*\]-\>plasticityFactor(reduce method:
    multiply)

:   **blockFactor** = blockMechanisms\[\*\]-\>blockFactor(reduce method:
    multiply)

:   **g** = blockFactor \* gbase \* (B - A) (exposed as **g**)

:   **i** = g \* (erev - v) (exposed as **i**)

***Time Derivatives***

:   d **A** /dt = -A / tauRise

:   d **B** /dt = -B / tauDecay
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="BlockingPlasticSynapse">
      <xs:complexContent>
        <xs:extension base="ExpTwoSynapse">
          <xs:sequence>
            <xs:element name="plasticityMechanism" type="PlasticityMechanism" minOccurs="0"/>
            <xs:element name="blockMechanism" type="BlockMechanism" minOccurs="0"/>
          </xs:sequence>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=BlockingPlasticSynapse){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import BlockingPlasticSynapse
    from neuroml.utils import component_factory

    variable = component_factory(
        BlockingPlasticSynapse,
        id: 'a NmlId (required)' = None,
        metaid: 'a MetaId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        neuro_lex_id: 'a NeuroLexId (optional)' = None,
        gbase: 'a Nml2Quantity_conductance (required)' = None,
        erev: 'a Nml2Quantity_voltage (required)' = None,
        tau_decay: 'a Nml2Quantity_time (required)' = None,
        tau_rise: 'a Nml2Quantity_time (required)' = None,
        plasticity_mechanism: 'a PlasticityMechanism (optional)' = None,
        block_mechanism: 'a BlockMechanism (optional)' = None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <blockingPlasticSynapse id="NMDA" gbase=".8nS" tauRise="1e-3s" tauDecay="13.3333e-3s" erev="0V">
        <blockMechanism type="voltageConcDepBlockMechanism" species="mg" blockConcentration="1.2mM" scalingConc="1.9205441817997078mM" scalingVolt="0.016129032258064516V"/>
    </blockingPlasticSynapse>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <blockingPlasticSynapse id="blockStpSynDep" gbase="1nS" erev="0mV" tauRise="0.1ms" tauDecay="2ms">
        <notes>A biexponential blocking synapse, with STD.</notes>
        <plasticityMechanism type="tsodyksMarkramDepMechanism" initReleaseProb="0.5" tauRec="120 ms"/>
        <blockMechanism type="voltageConcDepBlockMechanism" species="mg" blockConcentration="1.2 mM" scalingConc="1.920544 mM" scalingVolt="16.129 mV"/>
    </blockingPlasticSynapse>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <blockingPlasticSynapse id="blockStpSynDepFac" gbase="1nS" erev="0mV" tauRise="0.1ms" tauDecay="2ms">
        <notes>A biexponential blocking synapse with short term
                depression and facilitation.</notes>
        <plasticityMechanism type="tsodyksMarkramDepFacMechanism" initReleaseProb="0.5" tauRec="120 ms" tauFac="10 ms"/>
        <blockMechanism type="voltageConcDepBlockMechanism" species="mg" blockConcentration="1.2 mM" scalingConc="1.920544 mM" scalingVolt="16.129 mV"/>
    </blockingPlasticSynapse>
:::
:::
:::
:::
:::

::: {#doublesynapse .section}
[]{#schema-doublesynapse}

##### doubleSynapse[\#](#doublesynapse "Link to this heading"){.headerlink}

extends *[[baseVoltageDepSynapse]{.std
.std-ref}](#schema-basevoltagedepsynapse){.reference .internal}*

*Synapse consisting of two independent synaptic mechanisms ( e.g. AMPA-R
and NMDA-R ), which can be easily colocated in connections.*

::: {.sd-tab-set .docutils}
Paths

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| **sy   |                                                              |
| napse1 |                                                              |
| Path** |                                                              |
+--------+--------------------------------------------------------------+
| **sy   |                                                              |
| napse2 |                                                              |
| Path** |                                                              |
+--------+--------------------------------------------------------------+
:::
:::

Component References

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **sy |                                                 | [[baseSy    |
| naps |                                                 | napse]{.std |
| e1** |                                                 | .std-ref}   |
|      |                                                 | ](#schema-b |
|      |                                                 | asesynapse) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **sy |                                                 | [[baseSy    |
| naps |                                                 | napse]{.std |
| e2** |                                                 | .std-ref}   |
|      |                                                 | ](#schema-b |
|      |                                                 | asesynapse) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Properties

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------------------+-----------------------------------+-------------+
| **weight**         |                                   | Di          |
| (default: 1)       |                                   | mensionless |
+--------------------+-----------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | The total (usually time varying) current        | [[cu        |
| *i** | produced by this ComponentType *(from           | rrent]{.std |
|      | [[basePointCurrent]{.std                        | .std-r      |
|      | .std-ref}](L                                    | ef}](Landin |
|      | anding.html#schema-basepointcurrent){.reference | g.html#sche |
|      | .internal})*                                    | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | The current may vary with the voltage exposed   | [[vo        |
| *v** | by the ComponentType on which this is placed    | ltage]{.std |
|      | *(from [[baseVoltageDepSynapse]{.std            | .std-r      |
|      | .std-                                           | ef}](Landin |
|      | ref}](#schema-basevoltagedepsynapse){.reference | g.html#sche |
|      | .internal})*                                    | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   | *(from [[baseSynapse]{.std                      | Direction:  |
| in** | .std-ref}](#schema-basesynapse){.reference      | in          |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| *    | Used to relay incoming spikes to child          | Direction:  |
| *rel | mechanisms                                      | out         |
| ay** |                                                 |             |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***Structure***

:   WITH **this** AS **a**

:   WITH **synapse1Path** AS **b**

:   WITH **synapse2Path** AS **c**

:   CHILD INSTANCE: **synapse1**

:   CHILD INSTANCE: **synapse2**

:   EVENT CONNECTION from **a** TO **c**

:   EVENT CONNECTION from **a** TO **b**

***State Variables***

:   **weightFactor**: Dimensionless

***On Events***

:   EVENT IN on port: **in**

:      **weightFactor** = weight

:      EVENT OUT on port: **relay**

***Derived Variables***

:   **i1** = synapse1-\>i

:   **i2** = synapse2-\>i

:   **i** = weightFactor \* (i1 + i2) (exposed as **i**)
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="DoubleSynapse">
      <xs:complexContent>
        <xs:extension base="BaseVoltageDepSynapse">
          <xs:attribute name="synapse1" type="NmlId" use="required"/>
          <xs:attribute name="synapse2" type="NmlId" use="required"/>
          <xs:attribute name="synapse1Path" type="xs:string" use="required"/>
          <xs:attribute name="synapse2Path" type="xs:string" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=DoubleSynapse){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import DoubleSynapse
    from neuroml.utils import component_factory

    variable = component_factory(
        DoubleSynapse,
        id: 'a NmlId (required)' = None,
        metaid: 'a MetaId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        neuro_lex_id: 'a NeuroLexId (optional)' = None,
        synapse1: 'a NmlId (required)' = None,
        synapse2: 'a NmlId (required)' = None,
        synapse1_path: 'a string (required)' = None,
        synapse2_path: 'a string (required)' = None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <doubleSynapse id="AMPA_NMDA" synapse1="AMPA" synapse1Path="./AMPA" synapse2="NMDA" synapse2Path="./NMDA">
        <notes>A single "synapse" which contains both AMPA and NMDA. It is planned that the need for extra synapse1Path/synapse2Path attributes can be removed in later versions.</notes>
    </doubleSynapse>
:::
:::
:::
:::
:::

::: {#stdpsynapse .section}
[]{#schema-stdpsynapse}

##### stdpSynapse[\#](#stdpsynapse "Link to this heading"){.headerlink}

extends [[expTwoSynapse]{.std
.std-ref}](#schema-exptwosynapse){.reference .internal}

*Spike timing dependent plasticity mechanism, NOTE: EXAMPLE NOT YET
WORKING!!!*

[Bioportal entry for Computational Neuroscience Ontology related to
stdpSynapse.](https://bioportal.bioontology.org/ontologies/CNO/?p=classes&amp;conceptid=cno_0000034){.reference
.external}

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **er | Reversal potential of the synapse *(from        | [[vo        |
| ev** | [[baseConductanceBasedSynapse]{.std             | ltage]{.std |
|      | .std-ref}](                                     | .std-r      |
|      | #schema-baseconductancebasedsynapse){.reference | ef}](Landin |
|      | .internal})*                                    | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | Baseline conductance, generally the maximum     | [[conduc    |
| *gba | conductance following a single spike *(from     | tance]{.std |
| se** | [[baseConductanceBasedSynapse]{.std             | .std-ref}]  |
|      | .std-ref}](                                     | (Landing.ht |
|      | #schema-baseconductancebasedsynapse){.reference | ml#schema-d |
|      | .internal})*                                    | imensions-c |
|      |                                                 | onductance) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **ta | *(from [[expTwoSynapse]{.std                    | [           |
| uDec | .std-ref}](#schema-exptwosynapse){.reference    | [time]{.std |
| ay** | .internal})*                                    | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **t  | *(from [[expTwoSynapse]{.std                    | [           |
| auRi | .std-ref}](#schema-exptwosynapse){.reference    | [time]{.std |
| se** | .internal})*                                    | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Constants

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------------------+-----------------------------------+-------------+
| **tsinceRate** = 1 |                                   | Di          |
|                    |                                   | mensionless |
+--------------------+-----------------------------------+-------------+
| **longTime** =     |                                   | [           |
| 1000s              |                                   | [time]{.std |
|                    |                                   | .st         |
|                    |                                   | d-ref}](Lan |
|                    |                                   | ding.html#s |
|                    |                                   | chema-dimen |
|                    |                                   | sions-time) |
|                    |                                   | {.reference |
|                    |                                   | .internal}  |
+--------------------+-----------------------------------+-------------+
:::
:::

Derived parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **pe | *(from [[expTwoSynapse]{.std                    | [           |
| akTi | .std-ref}](#schema-exptwosynapse){.reference    | [time]{.std |
| me** | .internal})*                                    | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::

   **peakTime** = log(tauDecay / tauRise) \* (tauRise \*
tauDecay)/(tauDecay - tauRise)

::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   | *(from [[expTwoSynapse]{.std                    | Di          |
| wave | .std-ref}](#schema-exptwosynapse){.reference    | mensionless |
| form | .internal})*                                    |             |
| Fact |                                                 |             |
| or** |                                                 |             |
+------+-------------------------------------------------+-------------+
:::

   **waveformFactor** = 1 / (-exp(-peakTime / tauRise) + exp(-peakTime /
tauDecay))
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    |                                                 | Di          |
| *M** |                                                 | mensionless |
+------+-------------------------------------------------+-------------+
| *    |                                                 | Di          |
| *P** |                                                 | mensionless |
+------+-------------------------------------------------+-------------+
| *    | Time varying conductance through the synapse    | [[conduc    |
| *g** | *(from [[baseConductanceBasedSynapse]{.std      | tance]{.std |
|      | .std-ref}](                                     | .std-ref}]  |
|      | #schema-baseconductancebasedsynapse){.reference | (Landing.ht |
|      | .internal})*                                    | ml#schema-d |
|      |                                                 | imensions-c |
|      |                                                 | onductance) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | The total (usually time varying) current        | [[cu        |
| *i** | produced by this ComponentType *(from           | rrent]{.std |
|      | [[basePointCurrent]{.std                        | .std-r      |
|      | .std-ref}](L                                    | ef}](Landin |
|      | anding.html#schema-basepointcurrent){.reference | g.html#sche |
|      | .internal})*                                    | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   |                                                 | [           |
| tsin |                                                 | [time]{.std |
| ce** |                                                 | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | The current may vary with the voltage exposed   | [[vo        |
| *v** | by the ComponentType on which this is placed    | ltage]{.std |
|      | *(from [[baseVoltageDepSynapse]{.std            | .std-r      |
|      | .std-                                           | ef}](Landin |
|      | ref}](#schema-basevoltagedepsynapse){.reference | g.html#sche |
|      | .internal})*                                    | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   | *(from [[baseSynapse]{.std                      | Direction:  |
| in** | .std-ref}](#schema-basesynapse){.reference      | in          |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***State Variables***

:   **A**: Dimensionless

:   **B**: Dimensionless

:   **M**: Dimensionless  (exposed as **M**)

:   **P**: Dimensionless  (exposed as **P**)

:   **tsince**: [[time]{.std
    .std-ref}](Landing.html#schema-dimensions-time){.reference
    .internal}  (exposed as **tsince**)

***On Start***

:   **A** = 0

:   **B** = 0

:   **M** = 1

:   **P** = 1

:   **tsince** = longTime

***On Events***

:   EVENT IN on port: **in**

:      **A** = A + waveformFactor

:      **B** = B + waveformFactor

:      **tsince** = 0

***Derived Variables***

:   **g** = gbase \* (B - A) (exposed as **g**)

:   **i** = g \* (erev - v) (exposed as **i**)

***Time Derivatives***

:   d **A** /dt = -A / tauRise

:   d **B** /dt = -B / tauDecay

:   d **tsince** /dt = tsinceRate
:::
:::
:::

::: {#gapjunction .section}
[]{#schema-gapjunction}

##### gapJunction[\#](#gapjunction "Link to this heading"){.headerlink}

extends *[[baseSynapse]{.std .std-ref}](#schema-basesynapse){.reference
.internal}*

*Gap junction/single electrical connection.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **c  |                                                 | [[conduc    |
| ondu |                                                 | tance]{.std |
| ctan |                                                 | .std-ref}]  |
| ce** |                                                 | (Landing.ht |
|      |                                                 | ml#schema-d |
|      |                                                 | imensions-c |
|      |                                                 | onductance) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Properties

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------------------+-----------------------------------+-------------+
| **weight**         |                                   | Di          |
| (default: 1)       |                                   | mensionless |
+--------------------+-----------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | The total (usually time varying) current        | [[cu        |
| *i** | produced by this ComponentType *(from           | rrent]{.std |
|      | [[basePointCurrent]{.std                        | .std-r      |
|      | .std-ref}](L                                    | ef}](Landin |
|      | anding.html#schema-basepointcurrent){.reference | g.html#sche |
|      | .internal})*                                    | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    |                                                 | [[vo        |
| *v** |                                                 | ltage]{.std |
|      |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   | *(from [[baseSynapse]{.std                      | Direction:  |
| in** | .std-ref}](#schema-basesynapse){.reference      | in          |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***Derived Variables***

:   **vpeer** = peer-\>v

:   **i** = weight \* conductance \* (vpeer - v) (exposed as **i**)
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="GapJunction">
      <xs:complexContent>
        <xs:extension base="BaseSynapse">
          <xs:attribute name="conductance" type="Nml2Quantity_conductance" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=GapJunction){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import GapJunction
    from neuroml.utils import component_factory

    variable = component_factory(
        GapJunction,
        id: 'a NmlId (required)' = None,
        metaid: 'a MetaId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        neuro_lex_id: 'a NeuroLexId (optional)' = None,
        conductance: 'a Nml2Quantity_conductance (required)' = None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <gapJunction id="gj1" conductance="10pS"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <gapJunction id="gj1" conductance="10pS"/>
:::
:::
:::
:::
:::

::: {#basegradedsynapse .section}
[]{#schema-basegradedsynapse}

##### *baseGradedSynapse*[\#](#basegradedsynapse "Link to this heading"){.headerlink}

extends *[[baseSynapse]{.std .std-ref}](#schema-basesynapse){.reference
.internal}*

*Base type for graded synapses.*

::: {.sd-tab-set .docutils}
Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | The total (usually time varying) current        | [[cu        |
| *i** | produced by this ComponentType *(from           | rrent]{.std |
|      | [[basePointCurrent]{.std                        | .std-r      |
|      | .std-ref}](L                                    | ef}](Landin |
|      | anding.html#schema-basepointcurrent){.reference | g.html#sche |
|      | .internal})*                                    | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   | *(from [[baseSynapse]{.std                      | Direction:  |
| in** | .std-ref}](#schema-basesynapse){.reference      | in          |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::
:::
:::

::: {#silentsynapse .section}
[]{#schema-silentsynapse}

##### silentSynapse[\#](#silentsynapse "Link to this heading"){.headerlink}

extends *[[baseGradedSynapse]{.std
.std-ref}](#schema-basegradedsynapse){.reference .internal}*

*Dummy synapse which emits no current. Used as presynaptic endpoint for
analog synaptic connection.*

::: {.sd-tab-set .docutils}
Constants

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------------------+-----------------------------------+-------------+
| **AMP** = 1A       |                                   | [[cu        |
|                    |                                   | rrent]{.std |
|                    |                                   | .std-r      |
|                    |                                   | ef}](Landin |
|                    |                                   | g.html#sche |
|                    |                                   | ma-dimensio |
|                    |                                   | ns-current) |
|                    |                                   | {.reference |
|                    |                                   | .internal}  |
+--------------------+-----------------------------------+-------------+
:::
:::

Properties

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------------------+-----------------------------------+-------------+
| **weight**         |                                   | Di          |
| (default: 1)       |                                   | mensionless |
+--------------------+-----------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | The total (usually time varying) current        | [[cu        |
| *i** | produced by this ComponentType *(from           | rrent]{.std |
|      | [[basePointCurrent]{.std                        | .std-r      |
|      | .std-ref}](L                                    | ef}](Landin |
|      | anding.html#schema-basepointcurrent){.reference | g.html#sche |
|      | .internal})*                                    | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    |                                                 | [[vo        |
| *v** |                                                 | ltage]{.std |
|      |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   | *(from [[baseSynapse]{.std                      | Direction:  |
| in** | .std-ref}](#schema-basesynapse){.reference      | in          |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***Derived Variables***

:   **vpeer** = peer-\>v

:   **i** = 0 \* AMP (exposed as **i**)
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="SilentSynapse">
      <xs:complexContent>
        <xs:extension base="BaseSynapse">

                </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=SilentSynapse){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import SilentSynapse
    from neuroml.utils import component_factory

    variable = component_factory(
        SilentSynapse,
        id: 'a NmlId (required)' = None,
        metaid: 'a MetaId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        neuro_lex_id: 'a NeuroLexId (optional)' = None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <silentSynapse id="silent1"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <silentSynapse id="silent2"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <silentSynapse id="silent1"/>
:::
:::
:::
:::
:::

::: {#lineargradedsynapse .section}
[]{#schema-lineargradedsynapse}

##### linearGradedSynapse[\#](#lineargradedsynapse "Link to this heading"){.headerlink}

extends *[[baseGradedSynapse]{.std
.std-ref}](#schema-basegradedsynapse){.reference .internal}*

*Behaves just like a one way gap junction.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **c  |                                                 | [[conduc    |
| ondu |                                                 | tance]{.std |
| ctan |                                                 | .std-ref}]  |
| ce** |                                                 | (Landing.ht |
|      |                                                 | ml#schema-d |
|      |                                                 | imensions-c |
|      |                                                 | onductance) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Properties

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------------------+-----------------------------------+-------------+
| **weight**         |                                   | Di          |
| (default: 1)       |                                   | mensionless |
+--------------------+-----------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | The total (usually time varying) current        | [[cu        |
| *i** | produced by this ComponentType *(from           | rrent]{.std |
|      | [[basePointCurrent]{.std                        | .std-r      |
|      | .std-ref}](L                                    | ef}](Landin |
|      | anding.html#schema-basepointcurrent){.reference | g.html#sche |
|      | .internal})*                                    | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    |                                                 | [[vo        |
| *v** |                                                 | ltage]{.std |
|      |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   | *(from [[baseSynapse]{.std                      | Direction:  |
| in** | .std-ref}](#schema-basesynapse){.reference      | in          |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***Derived Variables***

:   **vpeer** = peer-\>v

:   **i** = weight \* conductance \* (vpeer - v) (exposed as **i**)
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="LinearGradedSynapse">
      <xs:complexContent>
        <xs:extension base="BaseSynapse">
          <xs:attribute name="conductance" type="Nml2Quantity_conductance" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=LinearGradedSynapse){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import LinearGradedSynapse
    from neuroml.utils import component_factory

    variable = component_factory(
        LinearGradedSynapse,
        id: 'a NmlId (required)' = None,
        metaid: 'a MetaId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        neuro_lex_id: 'a NeuroLexId (optional)' = None,
        conductance: 'a Nml2Quantity_conductance (required)' = None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <linearGradedSynapse id="gs1" conductance="5pS"/>
:::
:::
:::
:::
:::

::: {#gradedsynapse .section}
[]{#schema-gradedsynapse}

##### gradedSynapse[\#](#gradedsynapse "Link to this heading"){.headerlink}

extends *[[baseGradedSynapse]{.std
.std-ref}](#schema-basegradedsynapse){.reference .internal}*

*Graded/analog synapse. Based on synapse in Methods of
[http://www.nature.com/neuro/journal/v7/n12/abs/nn1352.html](http://www.nature.com/neuro/journal/v7/n12/abs/nn1352.html){.reference
.external}.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **V  | The half-activation voltage of the synapse      | [[vo        |
| th** |                                                 | ltage]{.std |
|      |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **c  |                                                 | [[conduc    |
| ondu |                                                 | tance]{.std |
| ctan |                                                 | .std-ref}]  |
| ce** |                                                 | (Landing.ht |
|      |                                                 | ml#schema-d |
|      |                                                 | imensions-c |
|      |                                                 | onductance) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | Slope of the activation curve                   | [[vo        |
| *del |                                                 | ltage]{.std |
| ta** |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **er | The reversal potential of the synapse           | [[vo        |
| ev** |                                                 | ltage]{.std |
|      |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | Rate constant for transmitter-receptor          | [[per       |
| *k** | dissociation rate                               | _time]{.std |
|      |                                                 | .std-re     |
|      |                                                 | f}](Landing |
|      |                                                 | .html#schem |
|      |                                                 | a-dimension |
|      |                                                 | s-per-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Properties

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------------------+-----------------------------------+-------------+
| **weight**         |                                   | Di          |
| (default: 1)       |                                   | mensionless |
+--------------------+-----------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | The total (usually time varying) current        | [[cu        |
| *i** | produced by this ComponentType *(from           | rrent]{.std |
|      | [[basePointCurrent]{.std                        | .std-r      |
|      | .std-ref}](L                                    | ef}](Landin |
|      | anding.html#schema-basepointcurrent){.reference | g.html#sche |
|      | .internal})*                                    | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **i  |                                                 | Di          |
| nf** |                                                 | mensionless |
+------+-------------------------------------------------+-------------+
| **t  |                                                 | [           |
| au** |                                                 | [time]{.std |
|      |                                                 | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    |                                                 | [[vo        |
| *v** |                                                 | ltage]{.std |
|      |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   | *(from [[baseSynapse]{.std                      | Direction:  |
| in** | .std-ref}](#schema-basesynapse){.reference      | in          |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***State Variables***

:   **s**: Dimensionless

***On Conditions***

:   IF (1-inf) \< 1e-4 THEN

:      **s** = inf

***Derived Variables***

:   **vpeer** = peer-\>v

:   **inf** = 1/(1 + exp((Vth - vpeer)/delta)) (exposed as **inf**)

:   **tau** = (1-inf)/k (exposed as **tau**)

:   **i** = weight \* conductance \* s \* (erev-v) (exposed as **i**)

***Conditional Derived Variables***

:   IF (1-inf) \> 1e-4 THEN

:     **s_rate** = (inf - s)/tau

:   OTHERWISE

:     **s_rate** = 0

***Time Derivatives***

:   d **s** /dt = s_rate
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="GradedSynapse">
      <xs:complexContent>
        <xs:extension base="BaseSynapse">
          <xs:attribute name="conductance" type="Nml2Quantity_conductance" use="required"/>
          <xs:attribute name="delta" type="Nml2Quantity_voltage" use="required"/>
          <xs:attribute name="Vth" type="Nml2Quantity_voltage" use="required"/>
          <xs:attribute name="k" type="Nml2Quantity_pertime" use="required"/>
          <xs:attribute name="erev" type="Nml2Quantity_voltage" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=GradedSynapse){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import GradedSynapse
    from neuroml.utils import component_factory

    variable = component_factory(
        GradedSynapse,
        id: 'a NmlId (required)' = None,
        metaid: 'a MetaId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        neuro_lex_id: 'a NeuroLexId (optional)' = None,
        conductance: 'a Nml2Quantity_conductance (required)' = None,
        delta: 'a Nml2Quantity_voltage (required)' = None,
        Vth: 'a Nml2Quantity_voltage (required)' = None,
        k: 'a Nml2Quantity_pertime (required)' = None,
        erev: 'a Nml2Quantity_voltage (required)' = None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <gradedSynapse id="gs2" conductance="5pS" delta="5mV" Vth="-55mV" k="0.025per_ms" erev="0mV"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <gradedSynapse id="gs1" conductance="0.1nS" delta="5mV" Vth="-35mV" k="0.025per_ms" erev="0mV"/>
:::
:::
:::
:::
:::
:::

[]{#document-Userdocs/Schemas/Inputs}

::: {#inputs .section .tex2jax_ignore .mathjax_ignore}
[]{#schema-inputs}

#### Inputs[\#](#inputs "Link to this heading"){.headerlink}

**A number of ComponentTypes for providing spiking ( e.g.
[[spikeGeneratorPoisson]{.std
.std-ref}](#schema-spikegeneratorpoisson){.reference .internal},
[[spikeArray]{.std .std-ref}](#schema-spikearray){.reference .internal}
) and current inputs ( e.g. [[pulseGenerator]{.std
.std-ref}](#schema-pulsegenerator){.reference .internal},
[[voltageClamp]{.std .std-ref}](#schema-voltageclamp){.reference
.internal}, [[timedSynapticInput]{.std
.std-ref}](#schema-timedsynapticinput){.reference .internal},
[[poissonFiringSynapse]{.std
.std-ref}](#schema-poissonfiringsynapse){.reference .internal} ) to
other ComponentTypes**

------------------------------------------------------------------------

Original ComponentType definitions:
[Inputs.xml](https://github.com/NeuroML/NeuroML2/blob/master/NeuroML2CoreTypes//Inputs.xml){.reference
.external}. Schema against which NeuroML based on these should be valid:
[NeuroML_v2.3.xsd](https://github.com/NeuroML/NeuroML2/tree/master/Schemas/NeuroML2/NeuroML_v2.3.xsd){.reference
.external}. Generated on 14/08/24 from
[this](https://github.com/NeuroML/NeuroML2/commit/352244cff605cb1ba24fa7c11757dc818fe90fd2){.reference
.external} commit. Please file any issues or questions at the [issue
tracker here](https://github.com/NeuroML/NeuroML2/issues){.reference
.external}.

------------------------------------------------------------------------

::: {#basepointcurrent .section}
[]{#schema-basepointcurrent}

##### *basePointCurrent*[\#](#basepointcurrent "Link to this heading"){.headerlink}

extends *[[baseStandalone]{.std
.std-ref}](Landing.html#schema-basestandalone){.reference .internal}*

*Base type for all ComponentTypes which produce a current **i** ( with
dimension current ).*

::: {.sd-tab-set .docutils}
Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | The total (usually time varying) current        | [[cu        |
| *i** | produced by this ComponentType                  | rrent]{.std |
|      |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::
:::
:::

::: {#basevoltagedeppointcurrent .section}
[]{#schema-basevoltagedeppointcurrent}

##### *baseVoltageDepPointCurrent*[\#](#basevoltagedeppointcurrent "Link to this heading"){.headerlink}

extends *[[basePointCurrent]{.std
.std-ref}](#schema-basepointcurrent){.reference .internal}*

*Base type for all ComponentTypes which produce a current **i** ( with
dimension current ) and require a voltage **v** exposed on the parent
Component, which would often be the membrane potential of a Component
extending [[baseCellMembPot]{.std
.std-ref}](Landing.html#schema-basecellmembpot){.reference .internal}.*

::: {.sd-tab-set .docutils}
Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | The total (usually time varying) current        | [[cu        |
| *i** | produced by this ComponentType *(from           | rrent]{.std |
|      | [[basePointCurrent]{.std                        | .std-r      |
|      | .std-ref}](#schema-basepointcurrent){.reference | ef}](Landin |
|      | .internal})*                                    | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | The current may vary with the voltage exposed   | [[vo        |
| *v** | by the ComponentType on which this is placed    | ltage]{.std |
|      |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::
:::
:::

::: {#basevoltagedeppointcurrentspiking .section}
[]{#schema-basevoltagedeppointcurrentspiking}

##### *baseVoltageDepPointCurrentSpiking*[\#](#basevoltagedeppointcurrentspiking "Link to this heading"){.headerlink}

extends *[[baseVoltageDepPointCurrent]{.std
.std-ref}](#schema-basevoltagedeppointcurrent){.reference .internal}*

*Base type for all ComponentTypes which produce a current **i,** require
a membrane potential **v** exposed on the parent and emit spikes ( on a
port **spike** ). The exposed variable **tsince** can be used for
plotting the time since the Component has spiked last.*

::: {.sd-tab-set .docutils}
Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | The total (usually time varying) current        | [[cu        |
| *i** | produced by this ComponentType *(from           | rrent]{.std |
|      | [[basePointCurrent]{.std                        | .std-r      |
|      | .std-ref}](#schema-basepointcurrent){.reference | ef}](Landin |
|      | .internal})*                                    | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   | Time since the last spike was emitted           | [           |
| tsin |                                                 | [time]{.std |
| ce** |                                                 | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | The current may vary with the voltage exposed   | [[vo        |
| *v** | by the ComponentType on which this is placed    | ltage]{.std |
|      | *(from [[baseVoltageDepPointCurrent]{.std       | .std-r      |
|      | .std-ref}]                                      | ef}](Landin |
|      | (#schema-basevoltagedeppointcurrent){.reference | g.html#sche |
|      | .internal})*                                    | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Port on which spikes are emitted                | Direction:  |
| *spi |                                                 | out         |
| ke** |                                                 |             |
+------+-------------------------------------------------+-------------+
:::
:::
:::
:::

::: {#basepointcurrentdl .section}
[]{#schema-basepointcurrentdl}

##### *basePointCurrentDL*[\#](#basepointcurrentdl "Link to this heading"){.headerlink}

*Base type for all ComponentTypes which produce a dimensionless current
**I.** There are many dimensionless equivalents of all the core current
producing ComponentTypes such as [[pulseGenerator]{.std
.std-ref}](#schema-pulsegenerator){.reference .internal} /
[[pulseGeneratorDL]{.std .std-ref}](#schema-pulsegeneratordl){.reference
.internal}, [[sineGenerator]{.std
.std-ref}](#schema-sinegenerator){.reference .internal} /
[[sineGeneratorDL]{.std .std-ref}](#schema-sinegeneratordl){.reference
.internal} and [[rampGenerator]{.std
.std-ref}](#schema-rampgenerator){.reference .internal} /
[[rampGeneratorDL]{.std .std-ref}](#schema-rampgeneratordl){.reference
.internal}.*

::: {.sd-tab-set .docutils}
Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | The total (time varying) current produced by    | Di          |
| *I** | this ComponentType                              | mensionless |
+------+-------------------------------------------------+-------------+
:::
:::
:::
:::

::: {#basevoltagedeppointcurrentdl .section}
[]{#schema-basevoltagedeppointcurrentdl}

##### *baseVoltageDepPointCurrentDL*[\#](#basevoltagedeppointcurrentdl "Link to this heading"){.headerlink}

extends *[[basePointCurrentDL]{.std
.std-ref}](#schema-basepointcurrentdl){.reference .internal}*

*Base type for all ComponentTypes which produce a dimensionless current
**I** and require a dimensionless membrane potential **V** exposed on
the parent Component.*

::: {.sd-tab-set .docutils}
Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | The total (time varying) current produced by    | Di          |
| *I** | this ComponentType *(from                       | mensionless |
|      | [[basePointCurrentDL]{.std                      |             |
|      | .s                                              |             |
|      | td-ref}](#schema-basepointcurrentdl){.reference |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | The current may vary with the dimensionless     | Di          |
| *V** | voltage exposed by the ComponentType on which   | mensionless |
|      | this is placed                                  |             |
+------+-------------------------------------------------+-------------+
:::
:::
:::
:::

::: {#basespikesource .section}
[]{#schema-basespikesource}

##### *baseSpikeSource*[\#](#basespikesource "Link to this heading"){.headerlink}

*Base for any ComponentType whose main purpose is to emit spikes ( on a
port **spike** ). The exposed variable **tsince** can be used for
plotting the time since the Component has spiked last.*

::: {.sd-tab-set .docutils}
Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   | Time since the last spike was emitted           | [           |
| tsin |                                                 | [time]{.std |
| ce** |                                                 | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Port on which spikes are emitted                | Direction:  |
| *spi |                                                 | out         |
| ke** |                                                 |             |
+------+-------------------------------------------------+-------------+
:::
:::
:::
:::

::: {#spikegenerator .section}
[]{#schema-spikegenerator}

##### spikeGenerator[\#](#spikegenerator "Link to this heading"){.headerlink}

extends *[[baseSpikeSource]{.std
.std-ref}](#schema-basespikesource){.reference .internal}*

*Simple generator of spikes at a regular interval set by **period**.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   | Time between spikes. The first spike will be    | [           |
| peri | emitted after this time.                        | [time]{.std |
| od** |                                                 | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Constants

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------------------+-----------------------------------+-------------+
| **SMALL_TIME** =   | A useful constant for use as a    | [           |
| 1e-9ms             | non zero time increment           | [time]{.std |
|                    |                                   | .st         |
|                    |                                   | d-ref}](Lan |
|                    |                                   | ding.html#s |
|                    |                                   | chema-dimen |
|                    |                                   | sions-time) |
|                    |                                   | {.reference |
|                    |                                   | .internal}  |
+--------------------+-----------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | When the next spike should ideally be emitted   | [           |
| *tne | (dt permitting)                                 | [time]{.std |
| xt** |                                                 | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   | Time since the last spike was emitted *(from    | [           |
| tsin | [[baseSpikeSource]{.std                         | [time]{.std |
| ce** | .std-ref}](#schema-basespikesource){.reference  | .st         |
|      | .internal})*                                    | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Port on which spikes are emitted *(from         | Direction:  |
| *spi | [[baseSpikeSource]{.std                         | out         |
| ke** | .std-ref}](#schema-basespikesource){.reference  |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***State Variables***

:   **tsince**: [[time]{.std
    .std-ref}](Landing.html#schema-dimensions-time){.reference
    .internal}  (exposed as **tsince**)

:   **tnext**: [[time]{.std
    .std-ref}](Landing.html#schema-dimensions-time){.reference
    .internal}  (exposed as **tnext**)

***On Start***

:   **tsince** = 0

:   **tnext** = period

***On Conditions***

:   IF tnext - t \< SMALL_TIME THEN

:      **tsince** = 0

:      **tnext** = tnext+period

:      EVENT OUT on port: **spike**

***Time Derivatives***

:   d **tsince** /dt = 1

:   d **tnext** /dt = 0
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="SpikeGenerator">
      <xs:complexContent>
        <xs:extension base="Standalone">
          <xs:attribute name="period" type="Nml2Quantity_time" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=SpikeGenerator){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import SpikeGenerator
    from neuroml.utils import component_factory

    variable = component_factory(
        SpikeGenerator,
        id: 'a NonNegativeInteger (required)' = None,
        metaid: 'a MetaId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        period: 'a Nml2Quantity_time (required)' = None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <spikeGenerator id="spikeGenRegular" period="20 ms"/>
:::
:::
:::
:::
:::

::: {#spikegeneratorrandom .section}
[]{#schema-spikegeneratorrandom}

##### spikeGeneratorRandom[\#](#spikegeneratorrandom "Link to this heading"){.headerlink}

extends *[[baseSpikeSource]{.std
.std-ref}](#schema-basespikesource){.reference .internal}*

*Generator of spikes with a random interspike interval of at least
**minISI** and at most **maxISI**.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   | Maximum interspike interval                     | [           |
| maxI |                                                 | [time]{.std |
| SI** |                                                 | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   | Minimum interspike interval                     | [           |
| minI |                                                 | [time]{.std |
| SI** |                                                 | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Constants

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------------------+-----------------------------------+-------------+
| **MSEC** = 1ms     | Required for converting time      | [           |
|                    | values to/from dimensionless      | [time]{.std |
|                    | quantities                        | .st         |
|                    |                                   | d-ref}](Lan |
|                    |                                   | ding.html#s |
|                    |                                   | chema-dimen |
|                    |                                   | sions-time) |
|                    |                                   | {.reference |
|                    |                                   | .internal}  |
+--------------------+-----------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **i  | The interval until the next spike               | [           |
| si** |                                                 | [time]{.std |
|      |                                                 | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | When the next spike should ideally be emitted   | [           |
| *tne | (dt permitting)                                 | [time]{.std |
| xt** |                                                 | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   | Time since the last spike was emitted *(from    | [           |
| tsin | [[baseSpikeSource]{.std                         | [time]{.std |
| ce** | .std-ref}](#schema-basespikesource){.reference  | .st         |
|      | .internal})*                                    | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Port on which spikes are emitted *(from         | Direction:  |
| *spi | [[baseSpikeSource]{.std                         | out         |
| ke** | .std-ref}](#schema-basespikesource){.reference  |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***State Variables***

:   **tsince**: [[time]{.std
    .std-ref}](Landing.html#schema-dimensions-time){.reference
    .internal}  (exposed as **tsince**)

:   **tnext**: [[time]{.std
    .std-ref}](Landing.html#schema-dimensions-time){.reference
    .internal}  (exposed as **tnext**)

:   **isi**: [[time]{.std
    .std-ref}](Landing.html#schema-dimensions-time){.reference
    .internal}  (exposed as **isi**)

***On Start***

:   **tsince** = 0

:   **isi** = minISI + MSEC \* random((maxISI - minISI) / MSEC)

:   **tnext** = isi

***On Conditions***

:   IF t \> tnext THEN

:      **isi** = minISI + MSEC \* random((maxISI - minISI) / MSEC)

:      **tsince** = 0

:      **tnext** = tnext+isi

:      EVENT OUT on port: **spike**

***Time Derivatives***

:   d **tsince** /dt = 1

:   d **tnext** /dt = 0
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="SpikeGeneratorRandom">
      <xs:complexContent>
        <xs:extension base="Standalone">
          <xs:attribute name="maxISI" type="Nml2Quantity_time" use="required"/>
          <xs:attribute name="minISI" type="Nml2Quantity_time" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=SpikeGeneratorRandom){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import SpikeGeneratorRandom
    from neuroml.utils import component_factory

    variable = component_factory(
        SpikeGeneratorRandom,
        id: 'a NonNegativeInteger (required)' = None,
        metaid: 'a MetaId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        max_isi: 'a Nml2Quantity_time (required)' = None,
        min_isi: 'a Nml2Quantity_time (required)' = None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <spikeGeneratorRandom id="spikeGenRandom" minISI="10 ms" maxISI="30 ms"/>
:::
:::
:::
:::
:::

::: {#spikegeneratorpoisson .section}
[]{#schema-spikegeneratorpoisson}

##### spikeGeneratorPoisson[\#](#spikegeneratorpoisson "Link to this heading"){.headerlink}

extends *[[baseSpikeSource]{.std
.std-ref}](#schema-basespikesource){.reference .internal}*

*Generator of spikes whose ISI is distributed according to an
exponential PDF with scale: 1 / **averageRate**.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **a  | The average rate at which spikes are emitted    | [[per       |
| vera |                                                 | _time]{.std |
| geRa |                                                 | .std-re     |
| te** |                                                 | f}](Landing |
|      |                                                 | .html#schem |
|      |                                                 | a-dimension |
|      |                                                 | s-per-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Constants

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------------------+-----------------------------------+-------------+
| **SMALL_TIME** =   |                                   | [           |
| 1e-9ms             |                                   | [time]{.std |
|                    |                                   | .st         |
|                    |                                   | d-ref}](Lan |
|                    |                                   | ding.html#s |
|                    |                                   | chema-dimen |
|                    |                                   | sions-time) |
|                    |                                   | {.reference |
|                    |                                   | .internal}  |
+--------------------+-----------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **i  | The interval until the next spike               | [           |
| si** |                                                 | [time]{.std |
|      |                                                 | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   | This is the ideal/perfect next spike time,      | [           |
| tnex | based on a newly generated isi, but dt          | [time]{.std |
| tIde | precision will mean that it's usually slightly  | .st         |
| al** | later than this                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | This is the next spike time for practical       | [           |
| *tne | purposes, ensuring that it's later than the     | [time]{.std |
| xtUs | current time                                    | .st         |
| ed** |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   | Time since the last spike was emitted *(from    | [           |
| tsin | [[baseSpikeSource]{.std                         | [time]{.std |
| ce** | .std-ref}](#schema-basespikesource){.reference  | .st         |
|      | .internal})*                                    | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Port on which spikes are emitted *(from         | Direction:  |
| *spi | [[baseSpikeSource]{.std                         | out         |
| ke** | .std-ref}](#schema-basespikesource){.reference  |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***State Variables***

:   **tsince**: [[time]{.std
    .std-ref}](Landing.html#schema-dimensions-time){.reference
    .internal}  (exposed as **tsince**)

:   **tnextIdeal**: [[time]{.std
    .std-ref}](Landing.html#schema-dimensions-time){.reference
    .internal}  (exposed as **tnextIdeal**)

:   **tnextUsed**: [[time]{.std
    .std-ref}](Landing.html#schema-dimensions-time){.reference
    .internal}  (exposed as **tnextUsed**)

:   **isi**: [[time]{.std
    .std-ref}](Landing.html#schema-dimensions-time){.reference
    .internal}  (exposed as **isi**)

***On Start***

:   **tsince** = 0

:   **isi** = -1 \* log(random(1)) / averageRate

:   **tnextIdeal** = isi

:   **tnextUsed** = isi

***On Conditions***

:   IF t \> tnextUsed THEN

:      **tsince** = 0

:      **isi** = -1 \* log(random(1)) / averageRate

:      **tnextIdeal** = (tnextIdeal+isi)

:      **tnextUsed** = tnextIdeal\*H( (tnextIdeal-t)/t ) +
    (t+SMALL_TIME)\*H( (t-tnextIdeal)/t )

:      EVENT OUT on port: **spike**

***Time Derivatives***

:   d **tsince** /dt = 1

:   d **tnextUsed** /dt = 0

:   d **tnextIdeal** /dt = 0
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="SpikeGeneratorPoisson">
      <xs:complexContent>
        <xs:extension base="Standalone">
          <xs:attribute name="averageRate" type="Nml2Quantity_pertime" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=SpikeGeneratorPoisson){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import SpikeGeneratorPoisson
    from neuroml.utils import component_factory

    variable = component_factory(
        SpikeGeneratorPoisson,
        id: 'a NonNegativeInteger (required)' = None,
        metaid: 'a MetaId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        average_rate: 'a Nml2Quantity_pertime (required)' = None,
        extensiontype_=None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <spikeGeneratorPoisson id="spikeGenPoisson" averageRate="50 Hz"/>
:::
:::
:::
:::
:::

::: {#spikegeneratorrefpoisson .section}
[]{#schema-spikegeneratorrefpoisson}

##### spikeGeneratorRefPoisson[\#](#spikegeneratorrefpoisson "Link to this heading"){.headerlink}

extends [[spikeGeneratorPoisson]{.std
.std-ref}](#schema-spikegeneratorpoisson){.reference .internal}

*Generator of spikes whose ISI distribution is the maximum entropy
distribution over \[ **minimumISI,** +infinity ) with mean: 1 /
**averageRate**.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **a  | The average rate at which spikes are emitted    | [[per       |
| vera | *(from [[spikeGeneratorPoisson]{.std            | _time]{.std |
| geRa | .std-                                           | .std-re     |
| te** | ref}](#schema-spikegeneratorpoisson){.reference | f}](Landing |
|      | .internal})*                                    | .html#schem |
|      |                                                 | a-dimension |
|      |                                                 | s-per-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   | The minimum interspike interval                 | [           |
| mini |                                                 | [time]{.std |
| mumI |                                                 | .st         |
| SI** |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Derived parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   | The average interspike interval                 | [           |
| aver |                                                 | [time]{.std |
| ageI |                                                 | .st         |
| si** |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::

   **averageIsi** = 1 / averageRate
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **i  | The interval until the next spike *(from        | [           |
| si** | [[spikeGeneratorPoisson]{.std                   | [time]{.std |
|      | .std-                                           | .st         |
|      | ref}](#schema-spikegeneratorpoisson){.reference | d-ref}](Lan |
|      | .internal})*                                    | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   | This is the ideal/perfect next spike time,      | [           |
| tnex | based on a newly generated isi, but dt          | [time]{.std |
| tIde | precision will mean that it's usually slightly  | .st         |
| al** | later than this *(from                          | d-ref}](Lan |
|      | [[spikeGeneratorPoisson]{.std                   | ding.html#s |
|      | .std-                                           | chema-dimen |
|      | ref}](#schema-spikegeneratorpoisson){.reference | sions-time) |
|      | .internal})*                                    | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | This is the next spike time for practical       | [           |
| *tne | purposes, ensuring that it's later than the     | [time]{.std |
| xtUs | current time *(from                             | .st         |
| ed** | [[spikeGeneratorPoisson]{.std                   | d-ref}](Lan |
|      | .std-                                           | ding.html#s |
|      | ref}](#schema-spikegeneratorpoisson){.reference | chema-dimen |
|      | .internal})*                                    | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   | Time since the last spike was emitted *(from    | [           |
| tsin | [[baseSpikeSource]{.std                         | [time]{.std |
| ce** | .std-ref}](#schema-basespikesource){.reference  | .st         |
|      | .internal})*                                    | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Port on which spikes are emitted *(from         | Direction:  |
| *spi | [[baseSpikeSource]{.std                         | out         |
| ke** | .std-ref}](#schema-basespikesource){.reference  |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***State Variables***

:   **tsince**: [[time]{.std
    .std-ref}](Landing.html#schema-dimensions-time){.reference
    .internal}  (exposed as **tsince**)

:   **tnextIdeal**: [[time]{.std
    .std-ref}](Landing.html#schema-dimensions-time){.reference
    .internal}  (exposed as **tnextIdeal**)

:   **tnextUsed**: [[time]{.std
    .std-ref}](Landing.html#schema-dimensions-time){.reference
    .internal}  (exposed as **tnextUsed**)

:   **isi**: [[time]{.std
    .std-ref}](Landing.html#schema-dimensions-time){.reference
    .internal}  (exposed as **isi**)

***On Start***

:   **tsince** = 0

:   **isi** = minimumISI - (averageIsi-minimumISI) \* log(random(1))

:   **tnextIdeal** = isi

:   **tnextUsed** = isi

***On Conditions***

:   IF t \> tnextUsed THEN

:      **tsince** = 0

:      **isi** = minimumISI - (averageIsi-minimumISI) \* log(random(1))

:      **tnextIdeal** = (tnextIdeal+isi)

:      **tnextUsed** = tnextIdeal\*H( (tnextIdeal-t)/t ) +
    (t+SMALL_TIME)\*H( (t-tnextIdeal)/t )

:      EVENT OUT on port: **spike**

***Time Derivatives***

:   d **tsince** /dt = 1

:   d **tnextUsed** /dt = 0

:   d **tnextIdeal** /dt = 0
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="SpikeGeneratorRefPoisson">
      <xs:complexContent>
        <xs:extension base="SpikeGeneratorPoisson">
          <xs:attribute name="minimumISI" type="Nml2Quantity_time" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=SpikeGeneratorRefPoisson){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import SpikeGeneratorRefPoisson
    from neuroml.utils import component_factory

    variable = component_factory(
        SpikeGeneratorRefPoisson,
        id: 'a NonNegativeInteger (required)' = None,
        metaid: 'a MetaId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        average_rate: 'a Nml2Quantity_pertime (required)' = None,
        minimum_isi: 'a Nml2Quantity_time (required)' = None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <spikeGeneratorRefPoisson id="spikeGenRefPoisson" averageRate="50 Hz" minimumISI="10 ms"/>
:::
:::
:::
:::
:::

::: {#poissonfiringsynapse .section}
[]{#schema-poissonfiringsynapse}

##### poissonFiringSynapse[\#](#poissonfiringsynapse "Link to this heading"){.headerlink}

extends *[[baseVoltageDepPointCurrentSpiking]{.std
.std-ref}](#schema-basevoltagedeppointcurrentspiking){.reference
.internal}*

*Poisson spike generator firing at **averageRate,** which is connected
to single **synapse** that is triggered every time a spike is generated,
producing an input current. See also
[[transientPoissonFiringSynapse]{.std
.std-ref}](#schema-transientpoissonfiringsynapse){.reference
.internal}.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **a  | The average rate at which spikes are emitted    | [[per       |
| vera |                                                 | _time]{.std |
| geRa |                                                 | .std-re     |
| te** |                                                 | f}](Landing |
|      |                                                 | .html#schem |
|      |                                                 | a-dimension |
|      |                                                 | s-per-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Constants

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------------------+-----------------------------------+-------------+
| **SMALL_TIME** =   |                                   | [           |
| 1e-9ms             |                                   | [time]{.std |
|                    |                                   | .st         |
|                    |                                   | d-ref}](Lan |
|                    |                                   | ding.html#s |
|                    |                                   | chema-dimen |
|                    |                                   | sions-time) |
|                    |                                   | {.reference |
|                    |                                   | .internal}  |
+--------------------+-----------------------------------+-------------+
:::
:::

Derived parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   | The average interspike interval                 | [           |
| aver |                                                 | [time]{.std |
| ageI |                                                 | .st         |
| si** |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::

   **averageIsi** = 1 / averageRate
:::

Paths

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| **s    | The target of the spikes, i.e. the synapse                   |
| pikeTa |                                                              |
| rget** |                                                              |
+--------+--------------------------------------------------------------+
:::
:::

Component References

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **s  |                                                 | [[baseSy    |
| ynap |                                                 | napse]{.std |
| se** |                                                 | .std-ref}]  |
|      |                                                 | (Landing.ht |
|      |                                                 | ml#schema-b |
|      |                                                 | asesynapse) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Properties

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------------------+-----------------------------------+-------------+
| **weight**         |                                   | Di          |
| (default: 1)       |                                   | mensionless |
+--------------------+-----------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | The total (usually time varying) current        | [[cu        |
| *i** | produced by this ComponentType *(from           | rrent]{.std |
|      | [[basePointCurrent]{.std                        | .std-r      |
|      | .std-ref}](#schema-basepointcurrent){.reference | ef}](Landin |
|      | .internal})*                                    | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **i  | The interval until the next spike               | [           |
| si** |                                                 | [time]{.std |
|      |                                                 | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   |                                                 | [           |
| tnex |                                                 | [time]{.std |
| tIde |                                                 | .st         |
| al** |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    |                                                 | [           |
| *tne |                                                 | [time]{.std |
| xtUs |                                                 | .st         |
| ed** |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   | Time since the last spike was emitted *(from    | [           |
| tsin | [[baseVoltageDepPointCurrentSpiking]{.std       | [time]{.std |
| ce** | .std-ref}](#schem                               | .st         |
|      | a-basevoltagedeppointcurrentspiking){.reference | d-ref}](Lan |
|      | .internal})*                                    | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | The current may vary with the voltage exposed   | [[vo        |
| *v** | by the ComponentType on which this is placed    | ltage]{.std |
|      | *(from [[baseVoltageDepPointCurrent]{.std       | .std-r      |
|      | .std-ref}]                                      | ef}](Landin |
|      | (#schema-basevoltagedeppointcurrent){.reference | g.html#sche |
|      | .internal})*                                    | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   | Note this is not used here. Will be removed in  | Direction:  |
| in** | future                                          | in          |
+------+-------------------------------------------------+-------------+
| *    | Port on which spikes are emitted                | Direction:  |
| *spi |                                                 | out         |
| ke** |                                                 |             |
+------+-------------------------------------------------+-------------+
| *    | Port on which spikes are emitted *(from         | Direction:  |
| *spi | [[baseVoltageDepPointCurrentSpiking]{.std       | out         |
| ke** | .std-ref}](#schem                               |             |
|      | a-basevoltagedeppointcurrentspiking){.reference |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***Structure***

:   WITH **this** AS **a**

:   WITH **spikeTarget** AS **b**

:   CHILD INSTANCE: **synapse**

:   EVENT CONNECTION from **a** TO **b**

***State Variables***

:   **tsince**: [[time]{.std
    .std-ref}](Landing.html#schema-dimensions-time){.reference
    .internal}  (exposed as **tsince**)

:   **tnextIdeal**: [[time]{.std
    .std-ref}](Landing.html#schema-dimensions-time){.reference
    .internal}  (exposed as **tnextIdeal**)

:   **tnextUsed**: [[time]{.std
    .std-ref}](Landing.html#schema-dimensions-time){.reference
    .internal}  (exposed as **tnextUsed**)

:   **isi**: [[time]{.std
    .std-ref}](Landing.html#schema-dimensions-time){.reference
    .internal}  (exposed as **isi**)

***On Start***

:   **tsince** = 0

:   **isi** = - averageIsi \* log(random(1))

:   **tnextIdeal** = isi

:   **tnextUsed** = isi

***On Conditions***

:   IF t \> tnextUsed THEN

:      **tsince** = 0

:      **isi** = - averageIsi \* log(1 - random(1))

:      **tnextIdeal** = (tnextIdeal+isi)

:      **tnextUsed** = tnextIdeal\*H( (tnextIdeal-t)/t ) +
    (t+SMALL_TIME)\*H( (t-tnextIdeal)/t )

:      EVENT OUT on port: **spike**

***Derived Variables***

:   **iSyn** = synapse-\>i

:   **i** = weight \* iSyn (exposed as **i**)

***Time Derivatives***

:   d **tsince** /dt = 1

:   d **tnextUsed** /dt = 0

:   d **tnextIdeal** /dt = 0
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="PoissonFiringSynapse">
      <xs:complexContent>
        <xs:extension base="Standalone">
          <xs:attribute name="averageRate" type="Nml2Quantity_pertime" use="required"/>
          <xs:attribute name="synapse" type="xs:string" use="required"/>
          <xs:attribute name="spikeTarget" type="xs:string" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=PoissonFiringSynapse){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import PoissonFiringSynapse
    from neuroml.utils import component_factory

    variable = component_factory(
        PoissonFiringSynapse,
        id: 'a NmlId (required)' = None,
        metaid: 'a MetaId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        average_rate: 'a Nml2Quantity_pertime (required)' = None,
        synapse: 'a string (required)' = None,
        spike_target: 'a string (required)' = None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <poissonFiringSynapse id="poissonFiringSyn" averageRate="10 Hz" synapse="synInput" spikeTarget="./synInput"/>
:::
:::
:::
:::
:::

::: {#transientpoissonfiringsynapse .section}
[]{#schema-transientpoissonfiringsynapse}

##### transientPoissonFiringSynapse[\#](#transientpoissonfiringsynapse "Link to this heading"){.headerlink}

extends *[[baseVoltageDepPointCurrentSpiking]{.std
.std-ref}](#schema-basevoltagedeppointcurrentspiking){.reference
.internal}*

*Poisson spike generator firing at **averageRate** after a **delay** and
for a **duration,** connected to single **synapse** that is triggered
every time a spike is generated, providing an input current. Similar to
ComponentType [[poissonFiringSynapse]{.std
.std-ref}](#schema-poissonfiringsynapse){.reference .internal}.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **a  |                                                 | [[per       |
| vera |                                                 | _time]{.std |
| geRa |                                                 | .std-re     |
| te** |                                                 | f}](Landing |
|      |                                                 | .html#schem |
|      |                                                 | a-dimension |
|      |                                                 | s-per-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    |                                                 | [           |
| *del |                                                 | [time]{.std |
| ay** |                                                 | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **du |                                                 | [           |
| rati |                                                 | [time]{.std |
| on** |                                                 | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Constants

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------------------+-----------------------------------+-------------+
| **SMALL_TIME** =   |                                   | [           |
| 1e-9ms             |                                   | [time]{.std |
|                    |                                   | .st         |
|                    |                                   | d-ref}](Lan |
|                    |                                   | ding.html#s |
|                    |                                   | chema-dimen |
|                    |                                   | sions-time) |
|                    |                                   | {.reference |
|                    |                                   | .internal}  |
+--------------------+-----------------------------------+-------------+
| **LONG_TIME** =    |                                   | [           |
| 1e9hour            |                                   | [time]{.std |
|                    |                                   | .st         |
|                    |                                   | d-ref}](Lan |
|                    |                                   | ding.html#s |
|                    |                                   | chema-dimen |
|                    |                                   | sions-time) |
|                    |                                   | {.reference |
|                    |                                   | .internal}  |
+--------------------+-----------------------------------+-------------+
:::
:::

Derived parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   |                                                 | [           |
| aver |                                                 | [time]{.std |
| ageI |                                                 | .st         |
| si** |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::

   **averageIsi** = 1 / averageRate
:::

Paths

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| **s    |                                                              |
| pikeTa |                                                              |
| rget** |                                                              |
+--------+--------------------------------------------------------------+
:::
:::

Component References

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **s  |                                                 | [[baseSy    |
| ynap |                                                 | napse]{.std |
| se** |                                                 | .std-ref}]  |
|      |                                                 | (Landing.ht |
|      |                                                 | ml#schema-b |
|      |                                                 | asesynapse) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Properties

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------------------+-----------------------------------+-------------+
| **weight**         |                                   | Di          |
| (default: 1)       |                                   | mensionless |
+--------------------+-----------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | The total (usually time varying) current        | [[cu        |
| *i** | produced by this ComponentType *(from           | rrent]{.std |
|      | [[basePointCurrent]{.std                        | .std-r      |
|      | .std-ref}](#schema-basepointcurrent){.reference | ef}](Landin |
|      | .internal})*                                    | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **i  |                                                 | [           |
| si** |                                                 | [time]{.std |
|      |                                                 | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   |                                                 | [           |
| tnex |                                                 | [time]{.std |
| tIde |                                                 | .st         |
| al** |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    |                                                 | [           |
| *tne |                                                 | [time]{.std |
| xtUs |                                                 | .st         |
| ed** |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   | Time since the last spike was emitted *(from    | [           |
| tsin | [[baseVoltageDepPointCurrentSpiking]{.std       | [time]{.std |
| ce** | .std-ref}](#schem                               | .st         |
|      | a-basevoltagedeppointcurrentspiking){.reference | d-ref}](Lan |
|      | .internal})*                                    | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | The current may vary with the voltage exposed   | [[vo        |
| *v** | by the ComponentType on which this is placed    | ltage]{.std |
|      | *(from [[baseVoltageDepPointCurrent]{.std       | .std-r      |
|      | .std-ref}]                                      | ef}](Landin |
|      | (#schema-basevoltagedeppointcurrent){.reference | g.html#sche |
|      | .internal})*                                    | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   | Note this is not used here. Will be removed in  | Direction:  |
| in** | future                                          | in          |
+------+-------------------------------------------------+-------------+
| *    | Port on which spikes are emitted                | Direction:  |
| *spi |                                                 | out         |
| ke** |                                                 |             |
+------+-------------------------------------------------+-------------+
| *    | Port on which spikes are emitted *(from         | Direction:  |
| *spi | [[baseVoltageDepPointCurrentSpiking]{.std       | out         |
| ke** | .std-ref}](#schem                               |             |
|      | a-basevoltagedeppointcurrentspiking){.reference |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***Structure***

:   WITH **this** AS **a**

:   WITH **spikeTarget** AS **b**

:   CHILD INSTANCE: **synapse**

:   EVENT CONNECTION from **a** TO **b**

***State Variables***

:   **tsince**: [[time]{.std
    .std-ref}](Landing.html#schema-dimensions-time){.reference
    .internal}  (exposed as **tsince**)

:   **tnextIdeal**: [[time]{.std
    .std-ref}](Landing.html#schema-dimensions-time){.reference
    .internal}  (exposed as **tnextIdeal**)

:   **tnextUsed**: [[time]{.std
    .std-ref}](Landing.html#schema-dimensions-time){.reference
    .internal}  (exposed as **tnextUsed**)

:   **isi**: [[time]{.std
    .std-ref}](Landing.html#schema-dimensions-time){.reference
    .internal}  (exposed as **isi**)

***On Start***

:   **tsince** = 0

:   **isi** = - averageIsi \* log(1 - random(1)) +delay

:   **tnextIdeal** = isi

:   **tnextUsed** = isi

***On Conditions***

:   IF t \> tnextUsed THEN

:      **tsince** = 0

:      **isi** = - averageIsi \* log(1 - random(1))

:      **tnextIdeal** = (tnextIdeal+isi) + H(((t+isi) -
    (delay+duration))/duration)\*LONG_TIME

:      **tnextUsed** = tnextIdeal\*H( (tnextIdeal-t)/t ) +
    (t+SMALL_TIME)\*H( (t-tnextIdeal)/t )

:      EVENT OUT on port: **spike**

***Derived Variables***

:   **iSyn** = synapse-\>i

:   **i** = weight \* iSyn (exposed as **i**)

***Time Derivatives***

:   d **tsince** /dt = 1

:   d **tnextUsed** /dt = 0

:   d **tnextIdeal** /dt = 0
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="TransientPoissonFiringSynapse">
      <xs:complexContent>
        <xs:extension base="Standalone">
          <xs:attribute name="averageRate" type="Nml2Quantity_pertime" use="required"/>
          <xs:attribute name="delay" type="Nml2Quantity_time" use="required"/>
          <xs:attribute name="duration" type="Nml2Quantity_time" use="required"/>
          <xs:attribute name="synapse" type="xs:string" use="required"/>
          <xs:attribute name="spikeTarget" type="xs:string" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=TransientPoissonFiringSynapse){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import TransientPoissonFiringSynapse
    from neuroml.utils import component_factory

    variable = component_factory(
        TransientPoissonFiringSynapse,
        id: 'a NmlId (required)' = None,
        metaid: 'a MetaId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        average_rate: 'a Nml2Quantity_pertime (required)' = None,
        delay: 'a Nml2Quantity_time (required)' = None,
        duration: 'a Nml2Quantity_time (required)' = None,
        synapse: 'a string (required)' = None,
        spike_target: 'a string (required)' = None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <transientPoissonFiringSynapse id="transPoissonFiringSyn" delay="50ms" duration="50ms" averageRate="300 Hz" synapse="synInputFast" spikeTarget="./synInputFast"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <transientPoissonFiringSynapse id="transPoissonFiringSyn2" delay="50ms" duration="500ms" averageRate="10 Hz" synapse="synInputFastTwo" spikeTarget="./synInputFastTwo"/>
:::
:::
:::
:::
:::

::: {#timedsynapticinput .section}
[]{#schema-timedsynapticinput}

##### timedSynapticInput[\#](#timedsynapticinput "Link to this heading"){.headerlink}

extends *[[baseVoltageDepPointCurrentSpiking]{.std
.std-ref}](#schema-basevoltagedeppointcurrentspiking){.reference
.internal}*

*Spike array connected to a single **synapse,** producing a current
triggered by each [[spike]{.std .std-ref}](#schema-spike){.reference
.internal} in the array.*

::: {.sd-tab-set .docutils}
Paths

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| **s    |                                                              |
| pikeTa |                                                              |
| rget** |                                                              |
+--------+--------------------------------------------------------------+
:::
:::

Component References

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **s  |                                                 | [[baseSy    |
| ynap |                                                 | napse]{.std |
| se** |                                                 | .std-ref}]  |
|      |                                                 | (Landing.ht |
|      |                                                 | ml#schema-b |
|      |                                                 | asesynapse) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Children list

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   |                                                 | [[          |
| spik |                                                 | spike]{.std |
| es** |                                                 | .st         |
|      |                                                 | d-ref}](#sc |
|      |                                                 | hema-spike) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Properties

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------------------+-----------------------------------+-------------+
| **weight**         |                                   | Di          |
| (default: 1)       |                                   | mensionless |
+--------------------+-----------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | The total (usually time varying) current        | [[cu        |
| *i** | produced by this ComponentType *(from           | rrent]{.std |
|      | [[basePointCurrent]{.std                        | .std-r      |
|      | .std-ref}](#schema-basepointcurrent){.reference | ef}](Landin |
|      | .internal})*                                    | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   | Time since the last spike was emitted *(from    | [           |
| tsin | [[baseVoltageDepPointCurrentSpiking]{.std       | [time]{.std |
| ce** | .std-ref}](#schem                               | .st         |
|      | a-basevoltagedeppointcurrentspiking){.reference | d-ref}](Lan |
|      | .internal})*                                    | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | The current may vary with the voltage exposed   | [[vo        |
| *v** | by the ComponentType on which this is placed    | ltage]{.std |
|      | *(from [[baseVoltageDepPointCurrent]{.std       | .std-r      |
|      | .std-ref}]                                      | ef}](Landin |
|      | (#schema-basevoltagedeppointcurrent){.reference | g.html#sche |
|      | .internal})*                                    | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   | This will receive events from the children      | Direction:  |
| in** |                                                 | in          |
+------+-------------------------------------------------+-------------+
| *    | Port on which spikes are emitted *(from         | Direction:  |
| *spi | [[baseVoltageDepPointCurrentSpiking]{.std       | out         |
| ke** | .std-ref}](#schem                               |             |
|      | a-basevoltagedeppointcurrentspiking){.reference |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***Structure***

:   WITH **this** AS **a**

:   WITH **spikeTarget** AS **b**

:   CHILD INSTANCE: **synapse**

:   EVENT CONNECTION from **a** TO **b**

***State Variables***

:   **tsince**: [[time]{.std
    .std-ref}](Landing.html#schema-dimensions-time){.reference
    .internal}  (exposed as **tsince**)

***On Events***

:   EVENT IN on port: **in**

:      **tsince** = 0

:      EVENT OUT on port: **spike**

***Derived Variables***

:   **iSyn** = synapse-\>i

:   **i** = weight \* iSyn (exposed as **i**)

***Time Derivatives***

:   d **tsince** /dt = 1
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="TimedSynapticInput">
      <xs:complexContent>
        <xs:extension base="Standalone">
          <xs:sequence>
            <xs:element name="spike" type="Spike" minOccurs="0" maxOccurs="unbounded"/>
          </xs:sequence>
          <xs:attribute name="synapse" type="NmlId" use="required"/>
          <xs:attribute name="spikeTarget" type="xs:string" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=TimedSynapticInput){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import TimedSynapticInput
    from neuroml.utils import component_factory

    variable = component_factory(
        TimedSynapticInput,
        id: 'a NmlId (required)' = None,
        metaid: 'a MetaId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        synapse: 'a NmlId (required)' = None,
        spike_target: 'a string (required)' = None,
        spikes: 'list of Spike(s) (optional)' = None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <timedSynapticInput id="synTrain" synapse="synInputFastTwo" spikeTarget="./synInputFastTwo">
        <spike id="0" time="2 ms"/>
        <spike id="1" time="15 ms"/>
        <spike id="2" time="27 ms"/>
        <spike id="3" time="40 ms"/>
        <spike id="4" time="45 ms"/>
        <spike id="5" time="50 ms"/>
        <spike id="6" time="52 ms"/>
        <spike id="7" time="54 ms"/>
        <spike id="8" time="54.5 ms"/>
        <spike id="9" time="54.6 ms"/>
        <spike id="10" time="54.7 ms"/>
        <spike id="11" time="54.8 ms"/>
        <spike id="12" time="54.9 ms"/>
        <spike id="13" time="55 ms"/>
        <spike id="14" time="55.1 ms"/>
        <spike id="15" time="55.2 ms"/>
    </timedSynapticInput>
:::
:::
:::
:::
:::

::: {#spikearray .section}
[]{#schema-spikearray}

##### spikeArray[\#](#spikearray "Link to this heading"){.headerlink}

extends *[[baseSpikeSource]{.std
.std-ref}](#schema-basespikesource){.reference .internal}*

*Set of spike ComponentTypes, each emitting one spike at a certain time.
Can be used to feed a predetermined spike train into a cell.*

::: {.sd-tab-set .docutils}
Children list

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   |                                                 | [[          |
| spik |                                                 | spike]{.std |
| es** |                                                 | .st         |
|      |                                                 | d-ref}](#sc |
|      |                                                 | hema-spike) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   | Time since the last spike was emitted *(from    | [           |
| tsin | [[baseSpikeSource]{.std                         | [time]{.std |
| ce** | .std-ref}](#schema-basespikesource){.reference  | .st         |
|      | .internal})*                                    | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   | This will receive events from the children      | Direction:  |
| in** |                                                 | in          |
+------+-------------------------------------------------+-------------+
| *    | Port on which spikes are emitted *(from         | Direction:  |
| *spi | [[baseSpikeSource]{.std                         | out         |
| ke** | .std-ref}](#schema-basespikesource){.reference  |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***State Variables***

:   **tsince**: [[time]{.std
    .std-ref}](Landing.html#schema-dimensions-time){.reference
    .internal}  (exposed as **tsince**)

***On Start***

:   **tsince** = 0

***On Events***

:   EVENT IN on port: **in**

:      **tsince** = 0

:      EVENT OUT on port: **spike**

***Time Derivatives***

:   d **tsince** /dt = 1
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="SpikeArray">
      <xs:complexContent>
        <xs:extension base="Standalone">
          <xs:sequence>
            <xs:element name="spike" type="Spike" minOccurs="0" maxOccurs="unbounded"/>
          </xs:sequence>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=SpikeArray){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import SpikeArray
    from neuroml.utils import component_factory

    variable = component_factory(
        SpikeArray,
        id: 'a NonNegativeInteger (required)' = None,
        metaid: 'a MetaId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        spikes: 'list of Spike(s) (optional)' = None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <spikeArray id="spkArr">
        <spike id="0" time="50 ms"/>
        <spike id="1" time="100 ms"/>
        <spike id="2" time="150 ms"/>
        <spike id="3" time="155 ms"/>
        <spike id="4" time="250 ms"/>
    </spikeArray>
:::
:::
:::
:::
:::

::: {#spike .section}
[]{#schema-spike}

##### spike[\#](#spike "Link to this heading"){.headerlink}

extends *[[baseSpikeSource]{.std
.std-ref}](#schema-basespikesource){.reference .internal}*

*Emits a single spike at the specified **time**.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **ti | Time at which to emit one spike event           | [           |
| me** |                                                 | [time]{.std |
|      |                                                 | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   | 0 signals not yet spiked, 1 signals has spiked  | Di          |
| spik |                                                 | mensionless |
| ed** |                                                 |             |
+------+-------------------------------------------------+-------------+
| **   | Time since the last spike was emitted *(from    | [           |
| tsin | [[baseSpikeSource]{.std                         | [time]{.std |
| ce** | .std-ref}](#schema-basespikesource){.reference  | .st         |
|      | .internal})*                                    | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Port on which spikes are emitted *(from         | Direction:  |
| *spi | [[baseSpikeSource]{.std                         | out         |
| ke** | .std-ref}](#schema-basespikesource){.reference  |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***Structure***

:   WITH **this** AS **a**

:   WITH **parent** AS **b**

:   EVENT CONNECTION from **a** TO **b**

***State Variables***

:   **tsince**: [[time]{.std
    .std-ref}](Landing.html#schema-dimensions-time){.reference
    .internal}  (exposed as **tsince**)

:   **spiked**: Dimensionless  (exposed as **spiked**)

***On Start***

:   **tsince** = 0

***On Conditions***

:   IF (t \>= time) AND (spiked = 0) THEN

:      **spiked** = 1

:      **tsince** = 0

:      EVENT OUT on port: **spike**

***Time Derivatives***

:   d **tsince** /dt = 1
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="Spike">
      <xs:complexContent>
        <xs:extension base="BaseNonNegativeIntegerId">
          <xs:attribute name="time" type="Nml2Quantity_time" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=Spike){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import Spike
    from neuroml.utils import component_factory

    variable = component_factory(
        Spike,
        id: 'a NonNegativeInteger (required)' = None,
        time: 'a Nml2Quantity_time (required)' = None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <spike id="0" time="50 ms"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <spike id="1" time="100 ms"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <spike id="2" time="150 ms"/>
:::
:::
:::
:::
:::

::: {#pulsegenerator .section}
[]{#schema-pulsegenerator}

##### pulseGenerator[\#](#pulsegenerator "Link to this heading"){.headerlink}

extends *[[basePointCurrent]{.std
.std-ref}](#schema-basepointcurrent){.reference .internal}*

*Generates a constant current pulse of a certain **amplitude** for a
specified **duration** after a **delay.** Scaled by **weight,** if set.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Amplitude of current pulse                      | [[cu        |
| *amp |                                                 | rrent]{.std |
| litu |                                                 | .std-r      |
| de** |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | Delay before change in current. Current is zero | [           |
| *del | prior to this.                                  | [time]{.std |
| ay** |                                                 | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **du | Duration for holding current at amplitude.      | [           |
| rati | Current is zero after delay + duration.         | [time]{.std |
| on** |                                                 | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Properties

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------------------+-----------------------------------+-------------+
| **weight**         |                                   | Di          |
| (default: 1)       |                                   | mensionless |
+--------------------+-----------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | The total (usually time varying) current        | [[cu        |
| *i** | produced by this ComponentType *(from           | rrent]{.std |
|      | [[basePointCurrent]{.std                        | .std-r      |
|      | .std-ref}](#schema-basepointcurrent){.reference | ef}](Landin |
|      | .internal})*                                    | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   | Note: this is not used here. Will be removed in | Direction:  |
| in** | future                                          | in          |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***State Variables***

:   **i**: [[current]{.std
    .std-ref}](Landing.html#schema-dimensions-current){.reference
    .internal}  (exposed as **i**)

***On Events***

:   EVENT IN on port: **in**

***On Conditions***

:   IF t \< delay THEN

:      **i** = 0

:   IF t \>= delay AND t \< duration + delay THEN

:      **i** = weight \* amplitude

:   IF t \>= duration + delay THEN

:      **i** = 0
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="PulseGenerator">
      <xs:complexContent>
        <xs:extension base="Standalone">
          <xs:attribute name="delay" type="Nml2Quantity_time" use="required"/>
          <xs:attribute name="duration" type="Nml2Quantity_time" use="required"/>
          <xs:attribute name="amplitude" type="Nml2Quantity_current" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=PulseGenerator){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import PulseGenerator
    from neuroml.utils import component_factory

    variable = component_factory(
        PulseGenerator,
        id: 'a NmlId (required)' = None,
        metaid: 'a MetaId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        delay: 'a Nml2Quantity_time (required)' = None,
        duration: 'a Nml2Quantity_time (required)' = None,
        amplitude: 'a Nml2Quantity_current (required)' = None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <pulseGenerator id="pulseGen1" delay="50ms" duration="200ms" amplitude="0.0032nA"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <pulseGenerator id="pulseGen2" delay="400ms" duration="200ms" amplitude="0.0020nA"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <pulseGenerator id="pulseGen3" delay="700ms" duration="200ms" amplitude="0.0010nA"/>
:::
:::
:::
:::
:::

::: {#compoundinput .section}
[]{#schema-compoundinput}

##### compoundInput[\#](#compoundinput "Link to this heading"){.headerlink}

extends *[[basePointCurrent]{.std
.std-ref}](#schema-basepointcurrent){.reference .internal}*

*Generates a current which is the sum of all its child
[[basePointCurrent]{.std .std-ref}](#schema-basepointcurrent){.reference
.internal} element, e.g. can be a combination of [[pulseGenerator]{.std
.std-ref}](#schema-pulsegenerator){.reference .internal},
[[sineGenerator]{.std .std-ref}](#schema-sinegenerator){.reference
.internal} elements producing a single **i.** Scaled by **weight,** if
set.*

::: {.sd-tab-set .docutils}
Children list

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **cu |                                                 | [[          |
| rren |                                                 | basePointCu |
| ts** |                                                 | rrent]{.std |
|      |                                                 | .st         |
|      |                                                 | d-ref}](#sc |
|      |                                                 | hema-basepo |
|      |                                                 | intcurrent) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Properties

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------------------+-----------------------------------+-------------+
| **weight**         |                                   | Di          |
| (default: 1)       |                                   | mensionless |
+--------------------+-----------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | The total (usually time varying) current        | [[cu        |
| *i** | produced by this ComponentType *(from           | rrent]{.std |
|      | [[basePointCurrent]{.std                        | .std-r      |
|      | .std-ref}](#schema-basepointcurrent){.reference | ef}](Landin |
|      | .internal})*                                    | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   | Note this is not used here. Will be removed in  | Direction:  |
| in** | future                                          | in          |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***On Events***

:   EVENT IN on port: **in**

***Derived Variables***

:   **i_total** = currents\[\*\]-\>i(reduce method: add)

:   **i** = weight \* i_total (exposed as **i**)
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="CompoundInput">
      <xs:complexContent>
        <xs:extension base="Standalone">
          <xs:sequence>
            <xs:element name="pulseGenerator" type="PulseGenerator" minOccurs="0" maxOccurs="unbounded"/>
            <xs:element name="sineGenerator" type="SineGenerator" minOccurs="0" maxOccurs="unbounded"/>
            <xs:element name="rampGenerator" type="RampGenerator" minOccurs="0" maxOccurs="unbounded"/>
          </xs:sequence>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=CompoundInput){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import CompoundInput
    from neuroml.utils import component_factory

    variable = component_factory(
        CompoundInput,
        id: 'a NmlId (required)' = None,
        metaid: 'a MetaId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        pulse_generators: 'list of PulseGenerator(s) (optional)' = None,
        sine_generators: 'list of SineGenerator(s) (optional)' = None,
        ramp_generators: 'list of RampGenerator(s) (optional)' = None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <compoundInput id="ci0">
        <pulseGenerator id="pg1" delay="50ms" duration="200ms" amplitude=".8 nA"/>
        <pulseGenerator id="pg2" delay="100ms" duration="100ms" amplitude=".4 nA"/>
        <sineGenerator id="sg0" phase="0" delay="125ms" duration="50ms" amplitude=".4nA" period="25ms"/>
    </compoundInput>
:::
:::
:::
:::
:::

::: {#compoundinputdl .section}
[]{#schema-compoundinputdl}

##### compoundInputDL[\#](#compoundinputdl "Link to this heading"){.headerlink}

extends *[[basePointCurrentDL]{.std
.std-ref}](#schema-basepointcurrentdl){.reference .internal}*

*Generates a current which is the sum of all its child
[[basePointCurrentDL]{.std
.std-ref}](#schema-basepointcurrentdl){.reference .internal} elements,
e.g. can be a combination of [[pulseGeneratorDL]{.std
.std-ref}](#schema-pulsegeneratordl){.reference .internal},
[[sineGeneratorDL]{.std .std-ref}](#schema-sinegeneratordl){.reference
.internal} elements producing a single **i.** Scaled by **weight,** if
set.*

::: {.sd-tab-set .docutils}
Children list

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **cu |                                                 | [[ba        |
| rren |                                                 | sePointCurr |
| ts** |                                                 | entDL]{.std |
|      |                                                 | .std-       |
|      |                                                 | ref}](#sche |
|      |                                                 | ma-basepoin |
|      |                                                 | tcurrentdl) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Properties

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------------------+-----------------------------------+-------------+
| **weight**         |                                   | Di          |
| (default: 1)       |                                   | mensionless |
+--------------------+-----------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | The total (time varying) current produced by    | Di          |
| *I** | this ComponentType *(from                       | mensionless |
|      | [[basePointCurrentDL]{.std                      |             |
|      | .s                                              |             |
|      | td-ref}](#schema-basepointcurrentdl){.reference |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   | Note this is not used here. Will be removed in  | Direction:  |
| in** | future                                          | in          |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***On Events***

:   EVENT IN on port: **in**

***Derived Variables***

:   **I_total** = currents\[\*\]-\>I(reduce method: add)

:   **I** = weight \* I_total (exposed as **I**)
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="CompoundInputDL">
      <xs:complexContent>
        <xs:extension base="Standalone">
          <xs:sequence>
            <xs:element name="pulseGeneratorDL" type="PulseGeneratorDL" minOccurs="0" maxOccurs="unbounded"/>
            <xs:element name="sineGeneratorDL" type="SineGeneratorDL" minOccurs="0" maxOccurs="unbounded"/>
            <xs:element name="rampGeneratorDL" type="RampGeneratorDL" minOccurs="0" maxOccurs="unbounded"/>
          </xs:sequence>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=CompoundInputDL){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import CompoundInputDL
    from neuroml.utils import component_factory

    variable = component_factory(
        CompoundInputDL,
        id: 'a NmlId (required)' = None,
        metaid: 'a MetaId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        pulse_generator_dls: 'list of PulseGeneratorDL(s) (optional)' = None,
        sine_generator_dls: 'list of SineGeneratorDL(s) (optional)' = None,
        ramp_generator_dls: 'list of RampGeneratorDL(s) (optional)' = None,
    )
:::
:::
:::
:::
:::

::: {#pulsegeneratordl .section}
[]{#schema-pulsegeneratordl}

##### pulseGeneratorDL[\#](#pulsegeneratordl "Link to this heading"){.headerlink}

extends *[[basePointCurrentDL]{.std
.std-ref}](#schema-basepointcurrentdl){.reference .internal}*

*Dimensionless equivalent of [[pulseGenerator]{.std
.std-ref}](#schema-pulsegenerator){.reference .internal}. Generates a
constant current pulse of a certain **amplitude** for a specified
**duration** after a **delay.** Scaled by **weight,** if set.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Amplitude of current pulse                      | Di          |
| *amp |                                                 | mensionless |
| litu |                                                 |             |
| de** |                                                 |             |
+------+-------------------------------------------------+-------------+
| *    | Delay before change in current. Current is zero | [           |
| *del | prior to this.                                  | [time]{.std |
| ay** |                                                 | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **du | Duration for holding current at amplitude.      | [           |
| rati | Current is zero after delay + duration.         | [time]{.std |
| on** |                                                 | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Properties

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------------------+-----------------------------------+-------------+
| **weight**         |                                   | Di          |
| (default: 1)       |                                   | mensionless |
+--------------------+-----------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | The total (time varying) current produced by    | Di          |
| *I** | this ComponentType *(from                       | mensionless |
|      | [[basePointCurrentDL]{.std                      |             |
|      | .s                                              |             |
|      | td-ref}](#schema-basepointcurrentdl){.reference |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   | Note this is not used here. Will be removed in  | Direction:  |
| in** | future                                          | in          |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***State Variables***

:   **I**: Dimensionless  (exposed as **I**)

***On Events***

:   EVENT IN on port: **in**

***On Conditions***

:   IF t \< delay THEN

:      **I** = 0

:   IF t \>= delay AND t \< duration + delay THEN

:      **I** = weight \* amplitude

:   IF t \>= duration + delay THEN

:      **I** = 0
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="PulseGeneratorDL">
      <xs:complexContent>
        <xs:extension base="Standalone">
          <xs:attribute name="delay" type="Nml2Quantity_time" use="required"/>
          <xs:attribute name="duration" type="Nml2Quantity_time" use="required"/>
          <xs:attribute name="amplitude" type="Nml2Quantity_none" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=PulseGeneratorDL){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import PulseGeneratorDL
    from neuroml.utils import component_factory

    variable = component_factory(
        PulseGeneratorDL,
        id: 'a NmlId (required)' = None,
        metaid: 'a MetaId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        delay: 'a Nml2Quantity_time (required)' = None,
        duration: 'a Nml2Quantity_time (required)' = None,
        amplitude: 'a Nml2Quantity_current (required)' = None,
    )
:::
:::
:::
:::
:::

::: {#sinegenerator .section}
[]{#schema-sinegenerator}

##### sineGenerator[\#](#sinegenerator "Link to this heading"){.headerlink}

extends *[[basePointCurrent]{.std
.std-ref}](#schema-basepointcurrent){.reference .internal}*

*Generates a sinusoidally varying current after a time **delay,** for a
fixed **duration.** The **period** and maximum **amplitude** of the
current can be set as well as the **phase** at which to start. Scaled by
**weight,** if set.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Maximum amplitude of current                    | [[cu        |
| *amp |                                                 | rrent]{.std |
| litu |                                                 | .std-r      |
| de** |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | Delay before change in current. Current is zero | [           |
| *del | prior to this.                                  | [time]{.std |
| ay** |                                                 | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **du | Duration for holding current at amplitude.      | [           |
| rati | Current is zero after delay + duration.         | [time]{.std |
| on** |                                                 | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   | Time period of oscillation                      | [           |
| peri |                                                 | [time]{.std |
| od** |                                                 | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | Phase (between 0 and 2\*pi) at which to start   | Di          |
| *pha | the varying current (i.e. at time given by      | mensionless |
| se** | delay)                                          |             |
+------+-------------------------------------------------+-------------+
:::
:::

Properties

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------------------+-----------------------------------+-------------+
| **weight**         |                                   | Di          |
| (default: 1)       |                                   | mensionless |
+--------------------+-----------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | The total (usually time varying) current        | [[cu        |
| *i** | produced by this ComponentType *(from           | rrent]{.std |
|      | [[basePointCurrent]{.std                        | .std-r      |
|      | .std-ref}](#schema-basepointcurrent){.reference | ef}](Landin |
|      | .internal})*                                    | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   |                                                 | Direction:  |
| in** |                                                 | in          |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***State Variables***

:   **i**: [[current]{.std
    .std-ref}](Landing.html#schema-dimensions-current){.reference
    .internal}  (exposed as **i**)

***On Events***

:   EVENT IN on port: **in**

***On Conditions***

:   IF t \< delay THEN

:      **i** = 0

:   IF t \>= delay AND t \< duration+delay THEN

:      **i** = weight \* amplitude \* sin(phase + (2 \* 3.14159265 \*
    (t-delay)/period) )

:   IF t \>= duration+delay THEN

:      **i** = 0
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="SineGenerator">
      <xs:complexContent>
        <xs:extension base="Standalone">
          <xs:attribute name="delay" type="Nml2Quantity_time" use="required"/>
          <xs:attribute name="phase" type="Nml2Quantity_none" use="required"/>
          <xs:attribute name="duration" type="Nml2Quantity_time" use="required"/>
          <xs:attribute name="amplitude" type="Nml2Quantity_current" use="required"/>
          <xs:attribute name="period" type="Nml2Quantity_time" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=SineGenerator){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import SineGenerator
    from neuroml.utils import component_factory

    variable = component_factory(
        SineGenerator,
        id: 'a NmlId (required)' = None,
        metaid: 'a MetaId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        delay: 'a Nml2Quantity_time (required)' = None,
        phase: 'a Nml2Quantity_none (required)' = None,
        duration: 'a Nml2Quantity_time (required)' = None,
        amplitude: 'a Nml2Quantity_current (required)' = None,
        period: 'a Nml2Quantity_time (required)' = None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <sineGenerator id="sg0" phase="0" delay="50ms" duration="200ms" amplitude="1.4nA" period="50ms"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <sineGenerator id="sg0" phase="0" delay="125ms" duration="50ms" amplitude=".4nA" period="25ms"/>
:::
:::
:::
:::
:::

::: {#sinegeneratordl .section}
[]{#schema-sinegeneratordl}

##### sineGeneratorDL[\#](#sinegeneratordl "Link to this heading"){.headerlink}

extends *[[basePointCurrentDL]{.std
.std-ref}](#schema-basepointcurrentdl){.reference .internal}*

*Dimensionless equivalent of [[sineGenerator]{.std
.std-ref}](#schema-sinegenerator){.reference .internal}. Generates a
sinusoidally varying current after a time **delay,** for a fixed
**duration.** The **period** and maximum **amplitude** of the current
can be set as well as the **phase** at which to start. Scaled by
**weight,** if set.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Maximum amplitude of current                    | Di          |
| *amp |                                                 | mensionless |
| litu |                                                 |             |
| de** |                                                 |             |
+------+-------------------------------------------------+-------------+
| *    | Delay before change in current. Current is zero | [           |
| *del | prior to this.                                  | [time]{.std |
| ay** |                                                 | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **du | Duration for holding current at amplitude.      | [           |
| rati | Current is zero after delay + duration.         | [time]{.std |
| on** |                                                 | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   | Time period of oscillation                      | [           |
| peri |                                                 | [time]{.std |
| od** |                                                 | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | Phase (between 0 and 2\*pi) at which to start   | Di          |
| *pha | the varying current (i.e. at time given by      | mensionless |
| se** | delay)                                          |             |
+------+-------------------------------------------------+-------------+
:::
:::

Properties

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------------------+-----------------------------------+-------------+
| **weight**         |                                   | Di          |
| (default: 1)       |                                   | mensionless |
+--------------------+-----------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | The total (time varying) current produced by    | Di          |
| *I** | this ComponentType *(from                       | mensionless |
|      | [[basePointCurrentDL]{.std                      |             |
|      | .s                                              |             |
|      | td-ref}](#schema-basepointcurrentdl){.reference |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   |                                                 | Direction:  |
| in** |                                                 | in          |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***State Variables***

:   **I**: Dimensionless  (exposed as **I**)

***On Events***

:   EVENT IN on port: **in**

***On Conditions***

:   IF t \< delay THEN

:      **I** = 0

:   IF t \>= delay AND t \< duration+delay THEN

:      **I** = weight \* amplitude \* sin(phase + (2 \* 3.14159265 \*
    (t-delay)/period) )

:   IF t \>= duration+delay THEN

:      **I** = 0
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="SineGeneratorDL">
      <xs:complexContent>
        <xs:extension base="Standalone">
          <xs:attribute name="delay" type="Nml2Quantity_time" use="required"/>
          <xs:attribute name="phase" type="Nml2Quantity_none" use="required"/>
          <xs:attribute name="duration" type="Nml2Quantity_time" use="required"/>
          <xs:attribute name="amplitude" type="Nml2Quantity_none" use="required"/>
          <xs:attribute name="period" type="Nml2Quantity_time" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=SineGeneratorDL){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import SineGeneratorDL
    from neuroml.utils import component_factory

    variable = component_factory(
        SineGeneratorDL,
        id: 'a NmlId (required)' = None,
        metaid: 'a MetaId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        delay: 'a Nml2Quantity_time (required)' = None,
        phase: 'a Nml2Quantity_none (required)' = None,
        duration: 'a Nml2Quantity_time (required)' = None,
        amplitude: 'a Nml2Quantity_current (required)' = None,
        period: 'a Nml2Quantity_time (required)' = None,
    )
:::
:::
:::
:::
:::

::: {#rampgenerator .section}
[]{#schema-rampgenerator}

##### rampGenerator[\#](#rampgenerator "Link to this heading"){.headerlink}

extends *[[basePointCurrent]{.std
.std-ref}](#schema-basepointcurrent){.reference .internal}*

*Generates a ramping current after a time **delay,** for a fixed
**duration.** During this time the current steadily changes from
**startAmplitude** to **finishAmplitude.** Scaled by **weight,** if
set.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Amplitude of current before time delay, and     | [[cu        |
| *bas | after time delay + duration                     | rrent]{.std |
| elin |                                                 | .std-r      |
| eAmp |                                                 | ef}](Landin |
| litu |                                                 | g.html#sche |
| de** |                                                 | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | Delay before change in current. Current is      | [           |
| *del | baselineAmplitude prior to this.                | [time]{.std |
| ay** |                                                 | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **du | Duration for holding current at amplitude.      | [           |
| rati | Current is baselineAmplitude after delay +      | [time]{.std |
| on** | duration.                                       | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **f  | Amplitude of linearly varying current at time   | [[cu        |
| inis | delay + duration                                | rrent]{.std |
| hAmp |                                                 | .std-r      |
| litu |                                                 | ef}](Landin |
| de** |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   | Amplitude of linearly varying current at time   | [[cu        |
| star | delay                                           | rrent]{.std |
| tAmp |                                                 | .std-r      |
| litu |                                                 | ef}](Landin |
| de** |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Properties

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------------------+-----------------------------------+-------------+
| **weight**         |                                   | Di          |
| (default: 1)       |                                   | mensionless |
+--------------------+-----------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | The total (usually time varying) current        | [[cu        |
| *i** | produced by this ComponentType *(from           | rrent]{.std |
|      | [[basePointCurrent]{.std                        | .std-r      |
|      | .std-ref}](#schema-basepointcurrent){.reference | ef}](Landin |
|      | .internal})*                                    | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   |                                                 | Direction:  |
| in** |                                                 | in          |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***State Variables***

:   **i**: [[current]{.std
    .std-ref}](Landing.html#schema-dimensions-current){.reference
    .internal}  (exposed as **i**)

***On Start***

:   **i** = baselineAmplitude

***On Events***

:   EVENT IN on port: **in**

***On Conditions***

:   IF t \< delay THEN

:      **i** = weight \* baselineAmplitude

:   IF t \>= delay AND t \< duration+delay THEN

:      **i** = weight \* (startAmplitude + (finishAmplitude -
    startAmplitude) \* (t - delay) / (duration))

:   IF t \>= duration+delay THEN

:      **i** = weight \* baselineAmplitude
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="RampGenerator">
      <xs:complexContent>
        <xs:extension base="Standalone">
          <xs:attribute name="delay" type="Nml2Quantity_time" use="required"/>
          <xs:attribute name="duration" type="Nml2Quantity_time" use="required"/>
          <xs:attribute name="startAmplitude" type="Nml2Quantity_current" use="required"/>
          <xs:attribute name="finishAmplitude" type="Nml2Quantity_current" use="required"/>
          <xs:attribute name="baselineAmplitude" type="Nml2Quantity_current" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=RampGenerator){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import RampGenerator
    from neuroml.utils import component_factory

    variable = component_factory(
        RampGenerator,
        id: 'a NmlId (required)' = None,
        metaid: 'a MetaId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        delay: 'a Nml2Quantity_time (required)' = None,
        duration: 'a Nml2Quantity_time (required)' = None,
        start_amplitude: 'a Nml2Quantity_current (required)' = None,
        finish_amplitude: 'a Nml2Quantity_current (required)' = None,
        baseline_amplitude: 'a Nml2Quantity_current (required)' = None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <rampGenerator id="rg0" delay="50ms" duration="200ms" startAmplitude="0.5nA" finishAmplitude="4nA" baselineAmplitude="0nA"/>
:::
:::
:::
:::
:::

::: {#rampgeneratordl .section}
[]{#schema-rampgeneratordl}

##### rampGeneratorDL[\#](#rampgeneratordl "Link to this heading"){.headerlink}

extends *[[basePointCurrentDL]{.std
.std-ref}](#schema-basepointcurrentdl){.reference .internal}*

*Dimensionless equivalent of [[rampGenerator]{.std
.std-ref}](#schema-rampgenerator){.reference .internal}. Generates a
ramping current after a time **delay,** for a fixed **duration.** During
this time the dimensionless current steadily changes from
**startAmplitude** to **finishAmplitude.** Scaled by **weight,** if
set.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Amplitude of current before time delay, and     | Di          |
| *bas | after time delay + duration                     | mensionless |
| elin |                                                 |             |
| eAmp |                                                 |             |
| litu |                                                 |             |
| de** |                                                 |             |
+------+-------------------------------------------------+-------------+
| *    | Delay before change in current. Current is      | [           |
| *del | baselineAmplitude prior to this.                | [time]{.std |
| ay** |                                                 | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **du | Duration for holding current at amplitude.      | [           |
| rati | Current is baselineAmplitude after delay +      | [time]{.std |
| on** | duration.                                       | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **f  | Amplitude of linearly varying current at time   | Di          |
| inis | delay + duration                                | mensionless |
| hAmp |                                                 |             |
| litu |                                                 |             |
| de** |                                                 |             |
+------+-------------------------------------------------+-------------+
| **   | Amplitude of linearly varying current at time   | Di          |
| star | delay                                           | mensionless |
| tAmp |                                                 |             |
| litu |                                                 |             |
| de** |                                                 |             |
+------+-------------------------------------------------+-------------+
:::
:::

Properties

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------------------+-----------------------------------+-------------+
| **weight**         |                                   | Di          |
| (default: 1)       |                                   | mensionless |
+--------------------+-----------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | The total (time varying) current produced by    | Di          |
| *I** | this ComponentType *(from                       | mensionless |
|      | [[basePointCurrentDL]{.std                      |             |
|      | .s                                              |             |
|      | td-ref}](#schema-basepointcurrentdl){.reference |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   |                                                 | Direction:  |
| in** |                                                 | in          |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***State Variables***

:   **I**: Dimensionless  (exposed as **I**)

***On Start***

:   **I** = baselineAmplitude

***On Events***

:   EVENT IN on port: **in**

***On Conditions***

:   IF t \< delay THEN

:      **I** = weight \* baselineAmplitude

:   IF t \>= delay AND t \< duration+delay THEN

:      **I** = weight \* (startAmplitude + (finishAmplitude -
    startAmplitude) \* (t - delay) / (duration))

:   IF t \>= duration+delay THEN

:      **I** = weight \* baselineAmplitude
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="RampGeneratorDL">
      <xs:complexContent>
        <xs:extension base="Standalone">
          <xs:attribute name="delay" type="Nml2Quantity_time" use="required"/>
          <xs:attribute name="duration" type="Nml2Quantity_time" use="required"/>
          <xs:attribute name="startAmplitude" type="Nml2Quantity_none" use="required"/>
          <xs:attribute name="finishAmplitude" type="Nml2Quantity_none" use="required"/>
          <xs:attribute name="baselineAmplitude" type="Nml2Quantity_none" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=RampGeneratorDL){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import RampGeneratorDL
    from neuroml.utils import component_factory

    variable = component_factory(
        RampGeneratorDL,
        id: 'a NmlId (required)' = None,
        metaid: 'a MetaId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        delay: 'a Nml2Quantity_time (required)' = None,
        duration: 'a Nml2Quantity_time (required)' = None,
        start_amplitude: 'a Nml2Quantity_current (required)' = None,
        finish_amplitude: 'a Nml2Quantity_current (required)' = None,
        baseline_amplitude: 'a Nml2Quantity_current (required)' = None,
    )
:::
:::
:::
:::
:::

::: {#voltageclamp .section}
[]{#schema-voltageclamp}

##### voltageClamp[\#](#voltageclamp "Link to this heading"){.headerlink}

extends *[[baseVoltageDepPointCurrent]{.std
.std-ref}](#schema-basevoltagedeppointcurrent){.reference .internal}*

*Voltage clamp. Applies a variable current **i** to try to keep parent
at **targetVoltage.** Not yet fully tested!!! Consider using
voltageClampTriple!!*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Delay before change in current. Current is zero | [           |
| *del | prior to this.                                  | [time]{.std |
| ay** |                                                 | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **du | Duration for attempting to keep parent at       | [           |
| rati | targetVoltage. Current is zero after delay +    | [time]{.std |
| on** | duration.                                       | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   | Current will be calculated by the difference in | [[resis     |
| simp | voltage between the target and parent, divided  | tance]{.std |
| leSe | by this value                                   | .std-ref}   |
| ries |                                                 | ](Landing.h |
| Resi |                                                 | tml#schema- |
| stan |                                                 | dimensions- |
| ce** |                                                 | resistance) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | Current will be applied to try to get parent to | [[vo        |
| *tar | this target voltage                             | ltage]{.std |
| getV |                                                 | .std-r      |
| olta |                                                 | ef}](Landin |
| ge** |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Properties

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------------------+-----------------------------------+-------------+
| **weight**         |                                   | Di          |
| (default: 1)       |                                   | mensionless |
+--------------------+-----------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | The total (usually time varying) current        | [[cu        |
| *i** | produced by this ComponentType *(from           | rrent]{.std |
|      | [[basePointCurrent]{.std                        | .std-r      |
|      | .std-ref}](#schema-basepointcurrent){.reference | ef}](Landin |
|      | .internal})*                                    | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | The current may vary with the voltage exposed   | [[vo        |
| *v** | by the ComponentType on which this is placed    | ltage]{.std |
|      | *(from [[baseVoltageDepPointCurrent]{.std       | .std-r      |
|      | .std-ref}]                                      | ef}](Landin |
|      | (#schema-basevoltagedeppointcurrent){.reference | g.html#sche |
|      | .internal})*                                    | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   | Note this is not used here. Will be removed in  | Direction:  |
| in** | future                                          | in          |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***State Variables***

:   **i**: [[current]{.std
    .std-ref}](Landing.html#schema-dimensions-current){.reference
    .internal}  (exposed as **i**)

***On Events***

:   EVENT IN on port: **in**

***On Conditions***

:   IF t \< delay THEN

:      **i** = 0

:   IF t \>= delay THEN

:      **i** = weight \* (targetVoltage - v) / simpleSeriesResistance

:   IF t \> duration + delay THEN

:      **i** = 0
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="VoltageClamp">
      <xs:complexContent>
        <xs:extension base="Standalone">
          <xs:attribute name="delay" type="Nml2Quantity_time" use="required"/>
          <xs:attribute name="duration" type="Nml2Quantity_time" use="required"/>
          <xs:attribute name="targetVoltage" type="Nml2Quantity_voltage" use="required"/>
          <xs:attribute name="simpleSeriesResistance" type="Nml2Quantity_resistance" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=VoltageClamp){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import VoltageClamp
    from neuroml.utils import component_factory

    variable = component_factory(
        VoltageClamp,
        id: 'a NmlId (required)' = None,
        metaid: 'a MetaId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        delay: 'a Nml2Quantity_time (required)' = None,
        duration: 'a Nml2Quantity_time (required)' = None,
        target_voltage: 'a Nml2Quantity_voltage (required)' = None,
        simple_series_resistance: 'a Nml2Quantity_resistance (required)' = None,
    )
:::
:::
:::
:::
:::

::: {#voltageclamptriple .section}
[]{#schema-voltageclamptriple}

##### voltageClampTriple[\#](#voltageclamptriple "Link to this heading"){.headerlink}

extends *[[baseVoltageDepPointCurrent]{.std
.std-ref}](#schema-basevoltagedeppointcurrent){.reference .internal}*

*Voltage clamp with 3 clamp levels. Applies a variable current **i** (
through **simpleSeriesResistance** ) to try to keep parent cell at
**conditioningVoltage** until time **delay,** **testingVoltage** until
**delay** + **duration,** and **returnVoltage** afterwards. Only enabled
if **active** = 1.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   | Whether the voltage clamp is active (1) or      | Di          |
| acti | inactive (0).                                   | mensionless |
| ve** |                                                 |             |
+------+-------------------------------------------------+-------------+
| **c  | Target voltage before time delay                | [[vo        |
| ondi |                                                 | ltage]{.std |
| tion |                                                 | .std-r      |
| ingV |                                                 | ef}](Landin |
| olta |                                                 | g.html#sche |
| ge** |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | Delay before switching from conditioningVoltage | [           |
| *del | to testingVoltage.                              | [time]{.std |
| ay** |                                                 | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **du | Duration to hold at testingVoltage.             | [           |
| rati |                                                 | [time]{.std |
| on** |                                                 | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | Target voltage after time duration              | [[vo        |
| *ret |                                                 | ltage]{.std |
| urnV |                                                 | .std-r      |
| olta |                                                 | ef}](Landin |
| ge** |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   | Current will be calculated by the difference in | [[resis     |
| simp | voltage between the target and parent, divided  | tance]{.std |
| leSe | by this value                                   | .std-ref}   |
| ries |                                                 | ](Landing.h |
| Resi |                                                 | tml#schema- |
| stan |                                                 | dimensions- |
| ce** |                                                 | resistance) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   | Target voltage between times delay and delay +  | [[vo        |
| test | duration                                        | ltage]{.std |
| ingV |                                                 | .std-r      |
| olta |                                                 | ef}](Landin |
| ge** |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Properties

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------------------+-----------------------------------+-------------+
| **weight**         |                                   | Di          |
| (default: 1)       |                                   | mensionless |
+--------------------+-----------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | The total (usually time varying) current        | [[cu        |
| *i** | produced by this ComponentType *(from           | rrent]{.std |
|      | [[basePointCurrent]{.std                        | .std-r      |
|      | .std-ref}](#schema-basepointcurrent){.reference | ef}](Landin |
|      | .internal})*                                    | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | The current may vary with the voltage exposed   | [[vo        |
| *v** | by the ComponentType on which this is placed    | ltage]{.std |
|      | *(from [[baseVoltageDepPointCurrent]{.std       | .std-r      |
|      | .std-ref}]                                      | ef}](Landin |
|      | (#schema-basevoltagedeppointcurrent){.reference | g.html#sche |
|      | .internal})*                                    | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   | Note this is not used here. Will be removed in  | Direction:  |
| in** | future                                          | in          |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***State Variables***

:   **i**: [[current]{.std
    .std-ref}](Landing.html#schema-dimensions-current){.reference
    .internal}  (exposed as **i**)

***On Events***

:   EVENT IN on port: **in**

***On Conditions***

:   IF active = 1 AND t \< delay THEN

:      **i** = weight \* (conditioningVoltage - v) /
    simpleSeriesResistance

:   IF active = 1 AND t \>= delay THEN

:      **i** = weight \* (testingVoltage - v) / simpleSeriesResistance

:   IF active = 1 AND t \> duration + delay THEN

:      **i** = weight \* (returnVoltage - v) / simpleSeriesResistance
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="VoltageClampTriple">
      <xs:complexContent>
        <xs:extension base="Standalone">
          <xs:attribute name="active" type="ZeroOrOne" use="required"/>
          <xs:attribute name="delay" type="Nml2Quantity_time" use="required"/>
          <xs:attribute name="duration" type="Nml2Quantity_time" use="required"/>
          <xs:attribute name="conditioningVoltage" type="Nml2Quantity_voltage" use="required"/>
          <xs:attribute name="testingVoltage" type="Nml2Quantity_voltage" use="required"/>
          <xs:attribute name="returnVoltage" type="Nml2Quantity_voltage" use="required"/>
          <xs:attribute name="simpleSeriesResistance" type="Nml2Quantity_resistance" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=VoltageClampTriple){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import VoltageClampTriple
    from neuroml.utils import component_factory

    variable = component_factory(
        VoltageClampTriple,
        id: 'a NmlId (required)' = None,
        metaid: 'a MetaId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        active: 'a ZeroOrOne (required)' = None,
        delay: 'a Nml2Quantity_time (required)' = None,
        duration: 'a Nml2Quantity_time (required)' = None,
        conditioning_voltage: 'a Nml2Quantity_voltage (required)' = None,
        testing_voltage: 'a Nml2Quantity_voltage (required)' = None,
        return_voltage: 'a Nml2Quantity_voltage (required)' = None,
        simple_series_resistance: 'a Nml2Quantity_resistance (required)' = None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <voltageClampTriple id="vClamp0" active="1" delay="50ms" duration="200ms" conditioningVoltage="-70mV" testingVoltage="-50mV" returnVoltage="-70mV" simpleSeriesResistance="1e6ohm"/>
:::
:::
:::
:::
:::
:::

[]{#document-Userdocs/Schemas/Networks}

::: {#networks .section .tex2jax_ignore .mathjax_ignore}
[]{#schema-networks}

#### Networks[\#](#networks "Link to this heading"){.headerlink}

**Network descriptions for NeuroML 2. Describes [[network]{.std
.std-ref}](#schema-network){.reference .internal} elements containing
[[population]{.std .std-ref}](#schema-population){.reference .internal}s
( potentially of type [[populationList]{.std
.std-ref}](#schema-populationlist){.reference .internal}, and so
specifying a list of cell [[location]{.std
.std-ref}](#schema-location){.reference .internal}s ),
[[projection]{.std .std-ref}](#schema-projection){.reference .internal}s
( i.e. lists of [[connection]{.std
.std-ref}](#schema-connection){.reference .internal}s ) and
[[input]{.std .std-ref}](#schema-input){.reference .internal}s.**

------------------------------------------------------------------------

Original ComponentType definitions:
[Networks.xml](https://github.com/NeuroML/NeuroML2/blob/master/NeuroML2CoreTypes//Networks.xml){.reference
.external}. Schema against which NeuroML based on these should be valid:
[NeuroML_v2.3.xsd](https://github.com/NeuroML/NeuroML2/tree/master/Schemas/NeuroML2/NeuroML_v2.3.xsd){.reference
.external}. Generated on 14/08/24 from
[this](https://github.com/NeuroML/NeuroML2/commit/352244cff605cb1ba24fa7c11757dc818fe90fd2){.reference
.external} commit. Please file any issues or questions at the [issue
tracker here](https://github.com/NeuroML/NeuroML2/issues){.reference
.external}.

------------------------------------------------------------------------

::: {#network .section}
[]{#schema-network}

##### network[\#](#network "Link to this heading"){.headerlink}

extends *[[baseStandalone]{.std
.std-ref}](Landing.html#schema-basestandalone){.reference .internal}*

*Network containing: [[population]{.std
.std-ref}](#schema-population){.reference .internal}s ( potentially of
type [[populationList]{.std
.std-ref}](#schema-populationlist){.reference .internal}, and so
specifying a list of cell [[location]{.std
.std-ref}](#schema-location){.reference .internal}s );
[[projection]{.std .std-ref}](#schema-projection){.reference .internal}s
( with lists of [[connection]{.std
.std-ref}](#schema-connection){.reference .internal}s ) and/or
[[explicitConnection]{.std
.std-ref}](#schema-explicitconnection){.reference .internal}s; and
[[inputList]{.std .std-ref}](#schema-inputlist){.reference .internal}s (
with lists of [[input]{.std .std-ref}](#schema-input){.reference
.internal}s ) and/or [[explicitInput]{.std
.std-ref}](#schema-explicitinput){.reference .internal}s. Note: often in
NeuroML this will be of type [[networkWithTemperature]{.std
.std-ref}](#schema-networkwithtemperature){.reference .internal} if
there are temperature dependent elements ( e.g. ion channels ).*

::: {.sd-tab-set .docutils}
Children list

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **r  |                                                 | [[r         |
| egio |                                                 | egion]{.std |
| ns** |                                                 | .std        |
|      |                                                 | -ref}](#sch |
|      |                                                 | ema-region) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **p  |                                                 | [[basePopul |
| opul |                                                 | ation]{.std |
| atio |                                                 | .           |
| ns** |                                                 | std-ref}](# |
|      |                                                 | schema-base |
|      |                                                 | population) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **p  |                                                 | [[proje     |
| roje |                                                 | ction]{.std |
| ctio |                                                 | .std-ref    |
| ns** |                                                 | }](#schema- |
|      |                                                 | projection) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **s  |                                                 | [[ex        |
| ynap |                                                 | plicitConne |
| ticC |                                                 | ction]{.std |
| onne |                                                 | .std-       |
| ctio |                                                 | ref}](#sche |
| ns** |                                                 | ma-explicit |
|      |                                                 | connection) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **el |                                                 | [[elec      |
| ectr |                                                 | tricalProje |
| ical |                                                 | ction]{.std |
| Proj |                                                 | .std-re     |
| ecti |                                                 | f}](#schema |
| on** |                                                 | -electrical |
|      |                                                 | projection) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **co |                                                 | [[cont      |
| ntin |                                                 | inuousProje |
| uous |                                                 | ction]{.std |
| Proj |                                                 | .std-re     |
| ecti |                                                 | f}](#schema |
| on** |                                                 | -continuous |
|      |                                                 | projection) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   |                                                 | [[explicit  |
| expl |                                                 | Input]{.std |
| icit |                                                 | .std-ref}]( |
| Inpu |                                                 | #schema-exp |
| ts** |                                                 | licitinput) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   |                                                 | [[inpu      |
| inpu |                                                 | tList]{.std |
| ts** |                                                 | .std-re     |
|      |                                                 | f}](#schema |
|      |                                                 | -inputlist) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="Network">
      <xs:complexContent>
        <xs:extension base="Standalone">
          <xs:sequence>
            <xs:element name="space" type="Space" minOccurs="0" maxOccurs="unbounded"/>
            <xs:element name="region" type="Region" minOccurs="0" maxOccurs="unbounded"/>
            <xs:element name="extracellularProperties" type="ExtracellularPropertiesLocal" minOccurs="0" maxOccurs="unbounded"/>
            <xs:element name="population" type="Population" maxOccurs="unbounded"/>
            <xs:element name="cellSet" type="CellSet" minOccurs="0" maxOccurs="unbounded"/>
            <xs:element name="synapticConnection" type="SynapticConnection" minOccurs="0" maxOccurs="unbounded"/>
            <xs:element name="projection" type="Projection" minOccurs="0" maxOccurs="unbounded"/>
            <xs:element name="electricalProjection" type="ElectricalProjection" minOccurs="0" maxOccurs="unbounded"/>
            <xs:element name="continuousProjection" type="ContinuousProjection" minOccurs="0" maxOccurs="unbounded"/>
            <xs:element name="explicitInput" type="ExplicitInput" minOccurs="0" maxOccurs="unbounded"/>
            <xs:element name="inputList" type="InputList" minOccurs="0" maxOccurs="unbounded"/>
          </xs:sequence>
          <xs:attribute name="type" type="networkTypes" use="optional"/>
          <xs:attribute name="temperature" type="Nml2Quantity_temperature" use="optional"/>
          <xs:attribute name="neuroLexId" type="NeuroLexId" use="optional"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=Network){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import Network
    from neuroml.utils import component_factory

    variable = component_factory(
        Network,
        id: 'a NmlId (required)' = None,
        metaid: 'a MetaId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        type: 'a networkTypes (optional)' = None,
        temperature: 'a Nml2Quantity_temperature (optional)' = None,
        neuro_lex_id: 'a NeuroLexId (optional)' = None,
        spaces: 'list of Space(s) (optional)' = None,
        regions: 'list of Region(s) (optional)' = None,
        extracellular_properties: 'list of ExtracellularPropertiesLocal(s) (optional)' = None,
        populations: 'list of Population(s) (required)' = None,
        cell_sets: 'list of CellSet(s) (optional)' = None,
        synaptic_connections: 'list of SynapticConnection(s) (optional)' = None,
        projections: 'list of Projection(s) (optional)' = None,
        electrical_projections: 'list of ElectricalProjection(s) (optional)' = None,
        continuous_projections: 'list of ContinuousProjection(s) (optional)' = None,
        explicit_inputs: 'list of ExplicitInput(s) (optional)' = None,
        input_lists: 'list of InputList(s) (optional)' = None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <network id="net1">
        <population id="iafPop1" component="iaf" size="1"/>
        <population id="iafPop2" component="iaf" size="1"/>
        <population id="iafPop3" component="iaf" size="1"/>
        <continuousProjection id="testLinearGradedConn" presynapticPopulation="iafPop1" postsynapticPopulation="iafPop2">
            <continuousConnection id="0" preCell="0" postCell="0" preComponent="silent1" postComponent="gs1"/>
        </continuousProjection>
        <continuousProjection id="testGradedConn" presynapticPopulation="iafPop1" postsynapticPopulation="iafPop3">
            <continuousConnection id="0" preCell="0" postCell="0" preComponent="silent2" postComponent="gs2"/>
        </continuousProjection>
        <explicitInput target="iafPop1[0]" input="pulseGen1" destination="synapses"/>
        <explicitInput target="iafPop1[0]" input="pulseGen2" destination="synapses"/>
        <explicitInput target="iafPop1[0]" input="pulseGen3" destination="synapses"/>
    </network>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <network id="net2">
        <population id="hhPop1" component="hhcell" size="1" type="populationList">
            <instance id="0">
                <location x="0" y="0" z="0"/>
            </instance>
        </population>
        <population id="hhPop2" component="hhcell" size="1" type="populationList">
            <instance id="0">
                <location x="100" y="0" z="0"/>
            </instance>
        </population>
        <continuousProjection id="testGradedConn" presynapticPopulation="hhPop1" postsynapticPopulation="hhPop2">
            <continuousConnectionInstanceW id="0" preCell="../hhPop1/0/hhcell" postCell="../hhPop2/0/hhcell" preComponent="silent1" postComponent="gs1" weight="1"/>
        </continuousProjection>
        <inputList id="i1" component="pulseGen1" population="hhPop1">
            <input id="0" target="../hhPop1/0/hhcell" destination="synapses"/>
        </inputList>
    </network>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <network id="PyrCellNet">
        <population id="Population1" component="PyrCell" extracellularProperties="extracellular" size="9"> 
            </population>
        <projection id="Proj1" presynapticPopulation="Population1" postsynapticPopulation="Population1" synapse="AMPA">
               
            </projection>
    </network>
:::
:::
:::
:::
:::

::: {#networkwithtemperature .section}
[]{#schema-networkwithtemperature}

##### networkWithTemperature[\#](#networkwithtemperature "Link to this heading"){.headerlink}

extends [[network]{.std .std-ref}](#schema-network){.reference
.internal}

*Same as [[network]{.std .std-ref}](#schema-network){.reference
.internal}, but with an explicit **temperature** for temperature
dependent elements ( e.g. ion channels ).*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **t  |                                                 | [[temper    |
| empe |                                                 | ature]{.std |
| ratu |                                                 | .std-ref}]  |
| re** |                                                 | (Landing.ht |
|      |                                                 | ml#schema-d |
|      |                                                 | imensions-t |
|      |                                                 | emperature) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::
:::
:::

::: {#basepopulation .section}
[]{#schema-basepopulation}

##### *basePopulation*[\#](#basepopulation "Link to this heading"){.headerlink}

extends *[[baseStandalone]{.std
.std-ref}](Landing.html#schema-basestandalone){.reference .internal}*

*A population of multiple instances of a specific **component,** which
anything which extends [[baseCell]{.std
.std-ref}](Landing.html#schema-basecell){.reference .internal}.*

::: {.sd-tab-set .docutils}
Component References

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    |                                                 | [[bas       |
| *com |                                                 | eCell]{.std |
| pone |                                                 | .std-re     |
| nt** |                                                 | f}](Landing |
|      |                                                 | .html#schem |
|      |                                                 | a-basecell) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Child list

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    |                                                 | [[          |
| *not |                                                 | notes]{.std |
| es** |                                                 | .std        |
|      |                                                 | -ref}](Land |
|      |                                                 | ing.html#sc |
|      |                                                 | hema-notes) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   |                                                 | [[annot     |
| anno |                                                 | ation]{.std |
| tati |                                                 | .std-ref}   |
| on** |                                                 | ](Landing.h |
|      |                                                 | tml#schema- |
|      |                                                 | annotation) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Children list

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **pr |                                                 | [[pro       |
| oper |                                                 | perty]{.std |
| ty** |                                                 | .std-re     |
|      |                                                 | f}](Landing |
|      |                                                 | .html#schem |
|      |                                                 | a-property) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::
:::
:::

::: {#population .section}
[]{#schema-population}

##### population[\#](#population "Link to this heading"){.headerlink}

extends *[[basePopulation]{.std
.std-ref}](#schema-basepopulation){.reference .internal}*

*A population of components, with just one parameter for the **size,**
i.e. number of components to create. Note: quite often this is used with
type= [[populationList]{.std
.std-ref}](#schema-populationlist){.reference .internal} which means the
size is determined by the number of [[instance]{.std
.std-ref}](#schema-instance){.reference .internal}s ( with
[[location]{.std .std-ref}](#schema-location){.reference .internal}s )
in the list. The **size** attribute is still set, and there will be a
validation error if this does not match the number in the list.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **si | Number of instances of this Component to create | Di          |
| ze** | when the population is instantiated             | mensionless |
+------+-------------------------------------------------+-------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="Population">
      <xs:complexContent>
        <xs:extension base="Standalone">
          <xs:choice>
            <xs:element name="layout" type="Layout" minOccurs="0"/>
            <xs:element name="instance" type="Instance" maxOccurs="unbounded"/>
          </xs:choice>
          <xs:attribute name="component" type="NmlId" use="required"/>
          <xs:attribute name="size" type="NonNegativeInteger" use="optional"/>
          <xs:attribute name="type" type="populationTypes" use="optional"/>
          <xs:attribute name="extracellularProperties" type="NmlId" use="optional"/>
          <xs:attribute name="neuroLexId" type="NeuroLexId" use="optional"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=Population){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import Population
    from neuroml.utils import component_factory

    variable = component_factory(
        Population,
        id: 'a NmlId (required)' = None,
        metaid: 'a MetaId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        component: 'a NmlId (required)' = None,
        size: 'a NonNegativeInteger (optional)' = None,
        type: 'a populationTypes (optional)' = None,
        extracellular_properties: 'a NmlId (optional)' = None,
        neuro_lex_id: 'a NeuroLexId (optional)' = None,
        layout: 'a Layout (optional)' = None,
        instances: 'list of Instance(s) (required)' = None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <population id="iafPop1" component="iaf" size="1"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <population id="iafPop2" component="iaf" size="1"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <population id="iafPop3" component="iaf" size="1"/>
:::
:::
:::
:::
:::

::: {#populationlist .section}
[]{#schema-populationlist}

##### populationList[\#](#populationlist "Link to this heading"){.headerlink}

extends *[[basePopulation]{.std
.std-ref}](#schema-basepopulation){.reference .internal}*

*An explicit list of [[instance]{.std
.std-ref}](#schema-instance){.reference .internal}s ( with
[[location]{.std .std-ref}](#schema-location){.reference .internal}s )
of components in the population.*

::: {.sd-tab-set .docutils}
Text fields

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| **     | Note: the size of the populationList to create is set by the |
| size** | number of explicitly defined instances. The size attribute   |
|        | is still set, and there will be a validation error if this   |
|        | does not match the number in the list.                       |
+--------+--------------------------------------------------------------+
:::
:::

Children list

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    |                                                 | [[ins       |
| *ins |                                                 | tance]{.std |
| tanc |                                                 | .std-r      |
| es** |                                                 | ef}](#schem |
|      |                                                 | a-instance) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::
:::
:::

::: {#instance .section}
[]{#schema-instance}

##### instance[\#](#instance "Link to this heading"){.headerlink}

*Specifies a single instance of a component in a [[population]{.std
.std-ref}](#schema-population){.reference .internal} ( placed at
[[location]{.std .std-ref}](#schema-location){.reference .internal} ).*

::: {.sd-tab-set .docutils}
Child list

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **lo |                                                 | [[loc       |
| cati |                                                 | ation]{.std |
| on** |                                                 | .std-r      |
|      |                                                 | ef}](#schem |
|      |                                                 | a-location) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="Instance">
      <xs:complexContent>
        <xs:extension base="BaseWithoutId">
          <xs:sequence>
            <xs:element name="location" type="Location"/>
          </xs:sequence>
          <xs:attribute name="id" type="xs:nonNegativeInteger"/>
          <xs:attribute name="i" type="xs:nonNegativeInteger"/>
          <xs:attribute name="j" type="xs:nonNegativeInteger"/>
          <xs:attribute name="k" type="xs:nonNegativeInteger"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=Instance){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import Instance
    from neuroml.utils import component_factory

    variable = component_factory(
        Instance,
        id: 'a nonNegativeInteger (optional)' = None,
        i: 'a nonNegativeInteger (optional)' = None,
        j: 'a nonNegativeInteger (optional)' = None,
        k: 'a nonNegativeInteger (optional)' = None,
        location: 'a Location (required)' = None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <instance id="0">
        <location x="0" y="0" z="0"/>
    </instance>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <instance id="0">
        <location x="100" y="0" z="0"/>
    </instance>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <instance id="0">
        <location x="0" y="0" z="0"/>
    </instance>
:::
:::
:::
:::
:::

::: {#location .section}
[]{#schema-location}

##### location[\#](#location "Link to this heading"){.headerlink}

*Specifies the ( x, y, z ) location of a single [[instance]{.std
.std-ref}](#schema-instance){.reference .internal} of a component in a
[[population]{.std .std-ref}](#schema-population){.reference
.internal}.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    |                                                 | Di          |
| *x** |                                                 | mensionless |
+------+-------------------------------------------------+-------------+
| *    |                                                 | Di          |
| *y** |                                                 | mensionless |
+------+-------------------------------------------------+-------------+
| *    |                                                 | Di          |
| *z** |                                                 | mensionless |
+------+-------------------------------------------------+-------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="Location">
      <xs:complexContent>
        <xs:extension base="BaseWithoutId">
          <xs:attribute name="x" type="xs:float" use="required"/>
          <xs:attribute name="y" type="xs:float" use="required"/>
          <xs:attribute name="z" type="xs:float" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=Location){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import Location
    from neuroml.utils import component_factory

    variable = component_factory(
        Location,
        x: 'a float (required)' = None,
        y: 'a float (required)' = None,
        z: 'a float (required)' = None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <location x="0" y="0" z="0"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <location x="100" y="0" z="0"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <location x="0" y="0" z="0"/>
:::
:::
:::
:::
:::

::: {#region .section}
[]{#schema-region}

##### region[\#](#region "Link to this heading"){.headerlink}

*Initial attempt to specify 3D region for placing cells. Work in
progress...*

::: {.sd-tab-set .docutils}
Child list

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    |                                                 | [[r         |
| *rec |                                                 | ectangularE |
| tang |                                                 | xtent]{.std |
| ular |                                                 | .std        |
| Exte |                                                 | -ref}](#sch |
| nt** |                                                 | ema-rectang |
|      |                                                 | ularextent) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="Region">
      <xs:complexContent>
        <xs:extension base="Base">
          <xs:sequence>
            <xs:any processContents="skip" minOccurs="0" maxOccurs="unbounded"/>
          </xs:sequence>
          <xs:attribute name="space" type="NmlId" use="optional"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=Region){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import Region
    from neuroml.utils import component_factory

    variable = component_factory(
        Region,
        id: 'a NmlId (required)' = None,
        spaces: 'a NmlId (optional)' = None,
        anytypeobjs_=None,
    )
:::
:::
:::
:::
:::

::: {#rectangularextent .section}
[]{#schema-rectangularextent}

##### rectangularExtent[\#](#rectangularextent "Link to this heading"){.headerlink}

*For defining a 3D rectangular box.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **x  |                                                 | Di          |
| Leng |                                                 | mensionless |
| th** |                                                 |             |
+------+-------------------------------------------------+-------------+
| **   |                                                 | Di          |
| xSta |                                                 | mensionless |
| rt** |                                                 |             |
+------+-------------------------------------------------+-------------+
| **y  |                                                 | Di          |
| Leng |                                                 | mensionless |
| th** |                                                 |             |
+------+-------------------------------------------------+-------------+
| **   |                                                 | Di          |
| ySta |                                                 | mensionless |
| rt** |                                                 |             |
+------+-------------------------------------------------+-------------+
| **z  |                                                 | Di          |
| Leng |                                                 | mensionless |
| th** |                                                 |             |
+------+-------------------------------------------------+-------------+
| **   |                                                 | Di          |
| zSta |                                                 | mensionless |
| rt** |                                                 |             |
+------+-------------------------------------------------+-------------+
:::
:::
:::
:::

::: {#projection .section}
[]{#schema-projection}

##### projection[\#](#projection "Link to this heading"){.headerlink}

*Projection from one population, **presynapticPopulation** to another,
**postsynapticPopulation,** through **synapse.** Contains lists of
[[connection]{.std .std-ref}](#schema-connection){.reference .internal}
or [[connectionWD]{.std .std-ref}](#schema-connectionwd){.reference
.internal} elements.*

::: {.sd-tab-set .docutils}
Paths

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| *      |                                                              |
| *presy |                                                              |
| naptic |                                                              |
| Popula |                                                              |
| tion** |                                                              |
+--------+--------------------------------------------------------------+
| **     |                                                              |
| postsy |                                                              |
| naptic |                                                              |
| Popula |                                                              |
| tion** |                                                              |
+--------+--------------------------------------------------------------+
:::
:::

Component References

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **s  |                                                 | [[baseSy    |
| ynap |                                                 | napse]{.std |
| se** |                                                 | .std-ref}]  |
|      |                                                 | (Landing.ht |
|      |                                                 | ml#schema-b |
|      |                                                 | asesynapse) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Children list

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **c  |                                                 | [[conne     |
| onne |                                                 | ction]{.std |
| ctio |                                                 | .std-ref    |
| ns** |                                                 | }](#schema- |
|      |                                                 | connection) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    |                                                 | [[connect   |
| *con |                                                 | ionWD]{.std |
| nect |                                                 | .std-ref}]  |
| ions |                                                 | (#schema-co |
| WD** |                                                 | nnectionwd) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="Projection">
      <xs:complexContent>
        <xs:extension base="BaseProjection">
          <xs:sequence>
            <xs:element name="connection" type="Connection" minOccurs="0" maxOccurs="unbounded"/>
            <xs:element name="connectionWD" type="ConnectionWD" minOccurs="0" maxOccurs="unbounded"/>
          </xs:sequence>
          <xs:attribute name="synapse" type="NmlId" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=Projection){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import Projection
    from neuroml.utils import component_factory

    variable = component_factory(
        Projection,
        id: 'a NmlId (required)' = None,
        presynaptic_population: 'a NmlId (required)' = None,
        postsynaptic_population: 'a NmlId (required)' = None,
        synapse: 'a NmlId (required)' = None,
        connections: 'list of Connection(s) (optional)' = None,
        connection_wds: 'list of ConnectionWD(s) (optional)' = None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <projection id="Proj1" presynapticPopulation="Population1" postsynapticPopulation="Population1" synapse="AMPA">
               
            </projection>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <projection id="internal1" presynapticPopulation="iafCells" postsynapticPopulation="iafCells" synapse="syn1">
                <synapseComponent component="syn1"/>-->
        <connection id="0" preCellId="../iafCells/0/iaf" postCellId="../iafCells/1/iaf"/>
    </projection>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <projection id="internal2" presynapticPopulation="iafCells" postsynapticPopulation="iafCells" synapse="syn2">
        <connection id="0" preCellId="../iafCells/0/iaf" postCellId="../iafCells/2/iaf"/>
    </projection>
:::
:::
:::
:::
:::

::: {#explicitconnection .section}
[]{#schema-explicitconnection}

##### explicitConnection[\#](#explicitconnection "Link to this heading"){.headerlink}

*Explicit event connection between components.*

::: {.sd-tab-set .docutils}
Text fields

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| **     |                                                              |
| target |                                                              |
| Port** |                                                              |
+--------+--------------------------------------------------------------+
:::
:::

Paths

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| **     |                                                              |
| from** |                                                              |
+--------+--------------------------------------------------------------+
| **to** |                                                              |
+--------+--------------------------------------------------------------+
:::
:::
:::
:::

::: {#connection .section}
[]{#schema-connection}

##### connection[\#](#connection "Link to this heading"){.headerlink}

*Event connection directly between named components, which gets
processed via a new instance of a **synapse** component which is created
on the target component. Normally contained inside a [[projection]{.std
.std-ref}](#schema-projection){.reference .internal} element.*

::: {.sd-tab-set .docutils}
Text fields

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| **d    |                                                              |
| estina |                                                              |
| tion** |                                                              |
+--------+--------------------------------------------------------------+
| **     |                                                              |
| preFra |                                                              |
| ctionA |                                                              |
| long** |                                                              |
+--------+--------------------------------------------------------------+
| **p    |                                                              |
| ostFra |                                                              |
| ctionA |                                                              |
| long** |                                                              |
+--------+--------------------------------------------------------------+
| **pr   |                                                              |
| eSegme |                                                              |
| ntId** |                                                              |
+--------+--------------------------------------------------------------+
| **pos  |                                                              |
| tSegme |                                                              |
| ntId** |                                                              |
+--------+--------------------------------------------------------------+
:::
:::

Paths

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| *      |                                                              |
| *preCe |                                                              |
| llId** |                                                              |
+--------+--------------------------------------------------------------+
| **     |                                                              |
| postCe |                                                              |
| llId** |                                                              |
+--------+--------------------------------------------------------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="Connection">
      <xs:complexContent>
        <xs:extension base="BaseConnectionOldFormat">
                </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=Connection){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import Connection
    from neuroml.utils import component_factory

    variable = component_factory(
        Connection,
        id: 'a NonNegativeInteger (required)' = None,
        neuro_lex_id: 'a NeuroLexId (optional)' = None,
        pre_cell_id: 'a Nml2PopulationReferencePath (required)' = None,
        pre_segment_id: 'a NonNegativeInteger (optional)' = '0',
        pre_fraction_along: 'a ZeroToOne (optional)' = '0.5',
        post_cell_id: 'a Nml2PopulationReferencePath (required)' = None,
        post_segment_id: 'a NonNegativeInteger (optional)' = '0',
        post_fraction_along: 'a ZeroToOne (optional)' = '0.5',
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <connection id="0" preCellId="../iafCells/0/iaf" postCellId="../iafCells/1/iaf"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <connection id="0" preCellId="../iafCells/0/iaf" postCellId="../iafCells/2/iaf"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <connection id="0" preCellId="../pop0/0/MultiCompCell" postCellId="../pop0/1/MultiCompCell" preSegmentId="0" preFractionAlong="0.5" postSegmentId="0" postFractionAlong="0.5"/>
:::
:::
:::
:::
:::

::: {#synapticconnection .section}
[]{#schema-synapticconnection}

##### synapticConnection[\#](#synapticconnection "Link to this heading"){.headerlink}

extends [[explicitConnection]{.std
.std-ref}](#schema-explicitconnection){.reference .internal}

*Explicit event connection between named components, which gets
processed via a new instance of a **synapse** component which is created
on the target component.*

::: {.sd-tab-set .docutils}
Text fields

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| **d    |                                                              |
| estina |                                                              |
| tion** |                                                              |
+--------+--------------------------------------------------------------+
:::
:::

Paths

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| **     |                                                              |
| from** |                                                              |
+--------+--------------------------------------------------------------+
| **to** |                                                              |
+--------+--------------------------------------------------------------+
:::
:::

Component References

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **s  |                                                 | [[baseSy    |
| ynap |                                                 | napse]{.std |
| se** |                                                 | .std-ref}]  |
|      |                                                 | (Landing.ht |
|      |                                                 | ml#schema-b |
|      |                                                 | asesynapse) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="SynapticConnection">
      <xs:complexContent>
        <xs:extension base="BaseWithoutId">
          <xs:attribute name="neuroLexId" type="NeuroLexId" use="optional"/>
          <xs:attribute name="from" type="Nml2PopulationReferencePath" use="required"/>
          <xs:attribute name="to" type="Nml2PopulationReferencePath" use="required"/>
          <xs:attribute name="synapse" type="NmlId" use="required"/>
          <xs:attribute name="destination" type="NmlId" use="optional"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=SynapticConnection){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import SynapticConnection
    from neuroml.utils import component_factory

    variable = component_factory(
        SynapticConnection,
        neuro_lex_id: 'a NeuroLexId (optional)' = None,
        from_: 'a Nml2PopulationReferencePath (required)' = None,
        to: 'a Nml2PopulationReferencePath (required)' = None,
        synapse: 'a NmlId (required)' = None,
        destination: 'a NmlId (optional)' = None,
    )
:::
:::
:::
:::
:::

::: {#synapticconnectionwd .section}
[]{#schema-synapticconnectionwd}

##### synapticConnectionWD[\#](#synapticconnectionwd "Link to this heading"){.headerlink}

extends [[synapticConnection]{.std
.std-ref}](#schema-synapticconnection){.reference .internal}

*Explicit event connection between named components, which gets
processed via a new instance of a **synapse** component which is created
on the target component, includes setting of **weight** and **delay**
for the synaptic connection.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    |                                                 | [           |
| *del |                                                 | [time]{.std |
| ay** |                                                 | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   |                                                 | Di          |
| weig |                                                 | mensionless |
| ht** |                                                 |             |
+------+-------------------------------------------------+-------------+
:::
:::

Paths

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| **     |                                                              |
| from** |                                                              |
+--------+--------------------------------------------------------------+
| **to** |                                                              |
+--------+--------------------------------------------------------------+
:::
:::
:::
:::

::: {#connectionwd .section}
[]{#schema-connectionwd}

##### connectionWD[\#](#connectionwd "Link to this heading"){.headerlink}

extends [[connection]{.std .std-ref}](#schema-connection){.reference
.internal}

*Event connection between named components, which gets processed via a
new instance of a synapse component which is created on the target
component, includes setting of **weight** and **delay** for the synaptic
connection.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    |                                                 | [           |
| *del |                                                 | [time]{.std |
| ay** |                                                 | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   |                                                 | Di          |
| weig |                                                 | mensionless |
| ht** |                                                 |             |
+------+-------------------------------------------------+-------------+
:::
:::

Text fields

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| **d    |                                                              |
| estina |                                                              |
| tion** |                                                              |
+--------+--------------------------------------------------------------+
| **     |                                                              |
| preFra |                                                              |
| ctionA |                                                              |
| long** |                                                              |
+--------+--------------------------------------------------------------+
| **p    |                                                              |
| ostFra |                                                              |
| ctionA |                                                              |
| long** |                                                              |
+--------+--------------------------------------------------------------+
| **pr   |                                                              |
| eSegme |                                                              |
| ntId** |                                                              |
+--------+--------------------------------------------------------------+
| **pos  |                                                              |
| tSegme |                                                              |
| ntId** |                                                              |
+--------+--------------------------------------------------------------+
:::
:::

Paths

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| *      |                                                              |
| *preCe |                                                              |
| llId** |                                                              |
+--------+--------------------------------------------------------------+
| **     |                                                              |
| postCe |                                                              |
| llId** |                                                              |
+--------+--------------------------------------------------------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="ConnectionWD">
      <xs:complexContent>
        <xs:extension base="BaseConnectionOldFormat">
          <xs:attribute name="weight" type="xs:float" use="required"/>
          <xs:attribute name="delay" type="Nml2Quantity_time" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=ConnectionWD){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import ConnectionWD
    from neuroml.utils import component_factory

    variable = component_factory(
        ConnectionWD,
        id: 'a NonNegativeInteger (required)' = None,
        neuro_lex_id: 'a NeuroLexId (optional)' = None,
        pre_cell_id: 'a Nml2PopulationReferencePath (required)' = None,
        pre_segment_id: 'a NonNegativeInteger (optional)' = '0',
        pre_fraction_along: 'a ZeroToOne (optional)' = '0.5',
        post_cell_id: 'a Nml2PopulationReferencePath (required)' = None,
        post_segment_id: 'a NonNegativeInteger (optional)' = '0',
        post_fraction_along: 'a ZeroToOne (optional)' = '0.5',
        weight: 'a float (required)' = None,
        delay: 'a Nml2Quantity_time (required)' = None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <connectionWD id="0" preCellId="../pop_EIF_cond_exp_isfa_ista[0]" postCellId="../pop_target[0]" weight="0.01" delay="10ms"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <connectionWD id="0" preCellId="../pop_EIF_cond_alpha_isfa_ista[0]" postCellId="../pop_target[1]" weight="0.005" delay="20ms"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <connectionWD id="0" preCellId="../pop_IF_curr_alpha[0]" postCellId="../pop_target[2]" weight="1" delay="30ms"/>
:::
:::
:::
:::
:::

::: {#electricalconnection .section}
[]{#schema-electricalconnection}

##### electricalConnection[\#](#electricalconnection "Link to this heading"){.headerlink}

*To enable connections between populations through gap junctions.*

::: {.sd-tab-set .docutils}
Component References

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **s  |                                                 | [[gapJun    |
| ynap |                                                 | ction]{.std |
| se** |                                                 | .std-ref}]  |
|      |                                                 | (Landing.ht |
|      |                                                 | ml#schema-g |
|      |                                                 | apjunction) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="ElectricalConnection">
      <xs:complexContent>
        <xs:extension base="BaseConnectionNewFormat">
          <xs:attribute name="synapse" type="NmlId" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=ElectricalConnection){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import ElectricalConnection
    from neuroml.utils import component_factory

    variable = component_factory(
        ElectricalConnection,
        id: 'a NonNegativeInteger (required)' = None,
        neuro_lex_id: 'a NeuroLexId (optional)' = None,
        pre_cell: 'a string (required)' = None,
        pre_segment: 'a NonNegativeInteger (optional)' = '0',
        pre_fraction_along: 'a ZeroToOne (optional)' = '0.5',
        post_cell: 'a string (required)' = None,
        post_segment: 'a NonNegativeInteger (optional)' = '0',
        post_fraction_along: 'a ZeroToOne (optional)' = '0.5',
        synapse: 'a NmlId (required)' = None,
        extensiontype_=None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <electricalConnection id="0" preCell="0" postCell="0" synapse="gj1"/>
:::
:::
:::
:::
:::

::: {#electricalconnectioninstance .section}
[]{#schema-electricalconnectioninstance}

##### electricalConnectionInstance[\#](#electricalconnectioninstance "Link to this heading"){.headerlink}

*To enable connections between populations through gap junctions.
Populations need to be of type [[populationList]{.std
.std-ref}](#schema-populationlist){.reference .internal} and contain
[[instance]{.std .std-ref}](#schema-instance){.reference .internal} and
[[location]{.std .std-ref}](#schema-location){.reference .internal}
elements.*

::: {.sd-tab-set .docutils}
Text fields

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| **     |                                                              |
| preFra |                                                              |
| ctionA |                                                              |
| long** |                                                              |
+--------+--------------------------------------------------------------+
| **p    |                                                              |
| ostFra |                                                              |
| ctionA |                                                              |
| long** |                                                              |
+--------+--------------------------------------------------------------+
| **     |                                                              |
| preSeg |                                                              |
| ment** |                                                              |
+--------+--------------------------------------------------------------+
| **p    |                                                              |
| ostSeg |                                                              |
| ment** |                                                              |
+--------+--------------------------------------------------------------+
:::
:::

Paths

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| **pre  |                                                              |
| Cell** |                                                              |
+--------+--------------------------------------------------------------+
| **post |                                                              |
| Cell** |                                                              |
+--------+--------------------------------------------------------------+
:::
:::

Component References

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **s  |                                                 | [[gapJun    |
| ynap |                                                 | ction]{.std |
| se** |                                                 | .std-ref}]  |
|      |                                                 | (Landing.ht |
|      |                                                 | ml#schema-g |
|      |                                                 | apjunction) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="ElectricalConnectionInstance">
      <xs:complexContent>
        <xs:extension base="ElectricalConnection"/>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=ElectricalConnectionInstance){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import ElectricalConnectionInstance
    from neuroml.utils import component_factory

    variable = component_factory(
        ElectricalConnectionInstance,
        id: 'a NonNegativeInteger (required)' = None,
        neuro_lex_id: 'a NeuroLexId (optional)' = None,
        pre_cell: 'a string (required)' = None,
        pre_segment: 'a NonNegativeInteger (optional)' = '0',
        pre_fraction_along: 'a ZeroToOne (optional)' = '0.5',
        post_cell: 'a string (required)' = None,
        post_segment: 'a NonNegativeInteger (optional)' = '0',
        post_fraction_along: 'a ZeroToOne (optional)' = '0.5',
        synapse: 'a NmlId (required)' = None,
        extensiontype_=None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <electricalConnectionInstance id="0" preCell="../iafPop1/0/iaf" postCell="../iafPop2/0/iaf" preSegment="0" preFractionAlong="0.5" postSegment="0" postFractionAlong="0.5" synapse="gj1"/>
:::
:::
:::
:::
:::

::: {#electricalconnectioninstancew .section}
[]{#schema-electricalconnectioninstancew}

##### electricalConnectionInstanceW[\#](#electricalconnectioninstancew "Link to this heading"){.headerlink}

extends [[electricalConnectionInstance]{.std
.std-ref}](#schema-electricalconnectioninstance){.reference .internal}

*To enable connections between populations through gap junctions.
Populations need to be of type [[populationList]{.std
.std-ref}](#schema-populationlist){.reference .internal} and contain
[[instance]{.std .std-ref}](#schema-instance){.reference .internal} and
[[location]{.std .std-ref}](#schema-location){.reference .internal}
elements. Includes setting of **weight** for the connection.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   |                                                 | Di          |
| weig |                                                 | mensionless |
| ht** |                                                 |             |
+------+-------------------------------------------------+-------------+
:::
:::

Text fields

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| **     |                                                              |
| preFra |                                                              |
| ctionA |                                                              |
| long** |                                                              |
+--------+--------------------------------------------------------------+
| **p    |                                                              |
| ostFra |                                                              |
| ctionA |                                                              |
| long** |                                                              |
+--------+--------------------------------------------------------------+
| **     |                                                              |
| preSeg |                                                              |
| ment** |                                                              |
+--------+--------------------------------------------------------------+
| **p    |                                                              |
| ostSeg |                                                              |
| ment** |                                                              |
+--------+--------------------------------------------------------------+
:::
:::

Paths

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| **pre  |                                                              |
| Cell** |                                                              |
+--------+--------------------------------------------------------------+
| **post |                                                              |
| Cell** |                                                              |
+--------+--------------------------------------------------------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="ElectricalConnectionInstanceW">
      <xs:complexContent>
        <xs:extension base="ElectricalConnectionInstance">
          <xs:attribute name="weight" type="xs:float" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=ElectricalConnectionInstanceW){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import ElectricalConnectionInstanceW
    from neuroml.utils import component_factory

    variable = component_factory(
        ElectricalConnectionInstanceW,
        id: 'a NonNegativeInteger (required)' = None,
        neuro_lex_id: 'a NeuroLexId (optional)' = None,
        pre_cell: 'a string (required)' = None,
        pre_segment: 'a NonNegativeInteger (optional)' = '0',
        pre_fraction_along: 'a ZeroToOne (optional)' = '0.5',
        post_cell: 'a string (required)' = None,
        post_segment: 'a NonNegativeInteger (optional)' = '0',
        post_fraction_along: 'a ZeroToOne (optional)' = '0.5',
        synapse: 'a NmlId (required)' = None,
        weight: 'a float (required)' = None,
    )
:::
:::
:::
:::
:::

::: {#electricalprojection .section}
[]{#schema-electricalprojection}

##### electricalProjection[\#](#electricalprojection "Link to this heading"){.headerlink}

*A projection between **presynapticPopulation** to another
**postsynapticPopulation** through gap junctions.*

::: {.sd-tab-set .docutils}
Component References

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    |                                                 | [[popul     |
| *pre |                                                 | ation]{.std |
| syna |                                                 | .std-ref    |
| ptic |                                                 | }](#schema- |
| Popu |                                                 | population) |
| lati |                                                 | {.reference |
| on** |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   |                                                 | [[popul     |
| post |                                                 | ation]{.std |
| syna |                                                 | .std-ref    |
| ptic |                                                 | }](#schema- |
| Popu |                                                 | population) |
| lati |                                                 | {.reference |
| on** |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Children list

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **c  |                                                 | [[elec      |
| onne |                                                 | tricalConne |
| ctio |                                                 | ction]{.std |
| ns** |                                                 | .std-re     |
|      |                                                 | f}](#schema |
|      |                                                 | -electrical |
|      |                                                 | connection) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **c  |                                                 | [[e         |
| onne |                                                 | lectricalCo |
| ctio |                                                 | nnectionIns |
| nIns |                                                 | tance]{.std |
| tanc |                                                 | .std        |
| es** |                                                 | -ref}](#sch |
|      |                                                 | ema-electri |
|      |                                                 | calconnecti |
|      |                                                 | oninstance) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="ElectricalProjection">
      <xs:complexContent>
        <xs:extension base="BaseProjection">
          <xs:sequence>
            <xs:element name="electricalConnection" type="ElectricalConnection" minOccurs="0" maxOccurs="unbounded"/>
            <xs:element name="electricalConnectionInstance" type="ElectricalConnectionInstance" minOccurs="0" maxOccurs="unbounded"/>
            <xs:element name="electricalConnectionInstanceW" type="ElectricalConnectionInstanceW" minOccurs="0" maxOccurs="unbounded"/>
          </xs:sequence>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=ElectricalProjection){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import ElectricalProjection
    from neuroml.utils import component_factory

    variable = component_factory(
        ElectricalProjection,
        id: 'a NmlId (required)' = None,
        presynaptic_population: 'a NmlId (required)' = None,
        postsynaptic_population: 'a NmlId (required)' = None,
        electrical_connections: 'list of ElectricalConnection(s) (optional)' = None,
        electrical_connection_instances: 'list of ElectricalConnectionInstance(s) (optional)' = None,
        electrical_connection_instance_ws: 'list of ElectricalConnectionInstanceW(s) (optional)' = None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <electricalProjection id="testGJconn" presynapticPopulation="iafPop1" postsynapticPopulation="iafPop2">
        <electricalConnectionInstance id="0" preCell="../iafPop1/0/iaf" postCell="../iafPop2/0/iaf" preSegment="0" preFractionAlong="0.5" postSegment="0" postFractionAlong="0.5" synapse="gj1"/>
    </electricalProjection>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <electricalProjection id="testGJconn" presynapticPopulation="iafPop1" postsynapticPopulation="iafPop2">
        <electricalConnection id="0" preCell="0" postCell="0" synapse="gj1"/>
    </electricalProjection>
:::
:::
:::
:::
:::

::: {#continuousconnection .section}
[]{#schema-continuousconnection}

##### continuousConnection[\#](#continuousconnection "Link to this heading"){.headerlink}

*An instance of a connection in a [[continuousProjection]{.std
.std-ref}](#schema-continuousprojection){.reference .internal} between
**presynapticPopulation** to another **postsynapticPopulation** through
a **preComponent** at the start and **postComponent** at the end. Can be
used for analog synapses.*

::: {.sd-tab-set .docutils}
Component References

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **pr |                                                 | [[b         |
| eCom |                                                 | aseGradedSy |
| pone |                                                 | napse]{.std |
| nt** |                                                 | .std-       |
|      |                                                 | ref}](Landi |
|      |                                                 | ng.html#sch |
|      |                                                 | ema-basegra |
|      |                                                 | dedsynapse) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    |                                                 | [[b         |
| *pos |                                                 | aseGradedSy |
| tCom |                                                 | napse]{.std |
| pone |                                                 | .std-       |
| nt** |                                                 | ref}](Landi |
|      |                                                 | ng.html#sch |
|      |                                                 | ema-basegra |
|      |                                                 | dedsynapse) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="ContinuousConnection">
      <xs:complexContent>
        <xs:extension base="BaseConnectionNewFormat">
          <xs:attribute name="preComponent" type="NmlId" use="required"/>
          <xs:attribute name="postComponent" type="NmlId" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=ContinuousConnection){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import ContinuousConnection
    from neuroml.utils import component_factory

    variable = component_factory(
        ContinuousConnection,
        id: 'a NonNegativeInteger (required)' = None,
        neuro_lex_id: 'a NeuroLexId (optional)' = None,
        pre_cell: 'a string (required)' = None,
        pre_segment: 'a NonNegativeInteger (optional)' = '0',
        pre_fraction_along: 'a ZeroToOne (optional)' = '0.5',
        post_cell: 'a string (required)' = None,
        post_segment: 'a NonNegativeInteger (optional)' = '0',
        post_fraction_along: 'a ZeroToOne (optional)' = '0.5',
        pre_component: 'a NmlId (required)' = None,
        post_component: 'a NmlId (required)' = None,
        extensiontype_=None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <continuousConnection id="0" preCell="0" postCell="0" preComponent="silent1" postComponent="gs1"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <continuousConnection id="0" preCell="0" postCell="0" preComponent="silent2" postComponent="gs2"/>
:::
:::
:::
:::
:::

::: {#continuousconnectioninstance .section}
[]{#schema-continuousconnectioninstance}

##### continuousConnectionInstance[\#](#continuousconnectioninstance "Link to this heading"){.headerlink}

*An instance of a connection in a [[continuousProjection]{.std
.std-ref}](#schema-continuousprojection){.reference .internal} between
**presynapticPopulation** to another **postsynapticPopulation** through
a **preComponent** at the start and **postComponent** at the end.
Populations need to be of type [[populationList]{.std
.std-ref}](#schema-populationlist){.reference .internal} and contain
[[instance]{.std .std-ref}](#schema-instance){.reference .internal} and
[[location]{.std .std-ref}](#schema-location){.reference .internal}
elements. Can be used for analog synapses.*

::: {.sd-tab-set .docutils}
Text fields

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| **     |                                                              |
| preFra |                                                              |
| ctionA |                                                              |
| long** |                                                              |
+--------+--------------------------------------------------------------+
| **p    |                                                              |
| ostFra |                                                              |
| ctionA |                                                              |
| long** |                                                              |
+--------+--------------------------------------------------------------+
| **     |                                                              |
| preSeg |                                                              |
| ment** |                                                              |
+--------+--------------------------------------------------------------+
| **p    |                                                              |
| ostSeg |                                                              |
| ment** |                                                              |
+--------+--------------------------------------------------------------+
:::
:::

Paths

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| **pre  |                                                              |
| Cell** |                                                              |
+--------+--------------------------------------------------------------+
| **post |                                                              |
| Cell** |                                                              |
+--------+--------------------------------------------------------------+
:::
:::

Component References

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **pr |                                                 | [[b         |
| eCom |                                                 | aseGradedSy |
| pone |                                                 | napse]{.std |
| nt** |                                                 | .std-       |
|      |                                                 | ref}](Landi |
|      |                                                 | ng.html#sch |
|      |                                                 | ema-basegra |
|      |                                                 | dedsynapse) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    |                                                 | [[b         |
| *pos |                                                 | aseGradedSy |
| tCom |                                                 | napse]{.std |
| pone |                                                 | .std-       |
| nt** |                                                 | ref}](Landi |
|      |                                                 | ng.html#sch |
|      |                                                 | ema-basegra |
|      |                                                 | dedsynapse) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="ContinuousConnectionInstance">
      <xs:complexContent>
        <xs:extension base="ContinuousConnection"/>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=ContinuousConnectionInstance){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import ContinuousConnectionInstance
    from neuroml.utils import component_factory

    variable = component_factory(
        ContinuousConnectionInstance,
        id: 'a NonNegativeInteger (required)' = None,
        neuro_lex_id: 'a NeuroLexId (optional)' = None,
        pre_cell: 'a string (required)' = None,
        pre_segment: 'a NonNegativeInteger (optional)' = '0',
        pre_fraction_along: 'a ZeroToOne (optional)' = '0.5',
        post_cell: 'a string (required)' = None,
        post_segment: 'a NonNegativeInteger (optional)' = '0',
        post_fraction_along: 'a ZeroToOne (optional)' = '0.5',
        pre_component: 'a NmlId (required)' = None,
        post_component: 'a NmlId (required)' = None,
        extensiontype_=None,
    )
:::
:::
:::
:::
:::

::: {#continuousconnectioninstancew .section}
[]{#schema-continuousconnectioninstancew}

##### continuousConnectionInstanceW[\#](#continuousconnectioninstancew "Link to this heading"){.headerlink}

extends [[continuousConnectionInstance]{.std
.std-ref}](#schema-continuousconnectioninstance){.reference .internal}

*An instance of a connection in a [[continuousProjection]{.std
.std-ref}](#schema-continuousprojection){.reference .internal} between
**presynapticPopulation** to another **postsynapticPopulation** through
a **preComponent** at the start and **postComponent** at the end.
Populations need to be of type [[populationList]{.std
.std-ref}](#schema-populationlist){.reference .internal} and contain
[[instance]{.std .std-ref}](#schema-instance){.reference .internal} and
[[location]{.std .std-ref}](#schema-location){.reference .internal}
elements. Can be used for analog synapses. Includes setting of
**weight** for the connection.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   |                                                 | Di          |
| weig |                                                 | mensionless |
| ht** |                                                 |             |
+------+-------------------------------------------------+-------------+
:::
:::

Text fields

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| **     |                                                              |
| preFra |                                                              |
| ctionA |                                                              |
| long** |                                                              |
+--------+--------------------------------------------------------------+
| **p    |                                                              |
| ostFra |                                                              |
| ctionA |                                                              |
| long** |                                                              |
+--------+--------------------------------------------------------------+
| **     |                                                              |
| preSeg |                                                              |
| ment** |                                                              |
+--------+--------------------------------------------------------------+
| **p    |                                                              |
| ostSeg |                                                              |
| ment** |                                                              |
+--------+--------------------------------------------------------------+
:::
:::

Paths

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| **pre  |                                                              |
| Cell** |                                                              |
+--------+--------------------------------------------------------------+
| **post |                                                              |
| Cell** |                                                              |
+--------+--------------------------------------------------------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="ContinuousConnectionInstanceW">
      <xs:complexContent>
        <xs:extension base="ContinuousConnectionInstance">
          <xs:attribute name="weight" type="xs:float" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=ContinuousConnectionInstanceW){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import ContinuousConnectionInstanceW
    from neuroml.utils import component_factory

    variable = component_factory(
        ContinuousConnectionInstanceW,
        id: 'a NonNegativeInteger (required)' = None,
        neuro_lex_id: 'a NeuroLexId (optional)' = None,
        pre_cell: 'a string (required)' = None,
        pre_segment: 'a NonNegativeInteger (optional)' = '0',
        pre_fraction_along: 'a ZeroToOne (optional)' = '0.5',
        post_cell: 'a string (required)' = None,
        post_segment: 'a NonNegativeInteger (optional)' = '0',
        post_fraction_along: 'a ZeroToOne (optional)' = '0.5',
        pre_component: 'a NmlId (required)' = None,
        post_component: 'a NmlId (required)' = None,
        weight: 'a float (required)' = None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <continuousConnectionInstanceW id="0" preCell="../hhPop1/0/hhcell" postCell="../hhPop2/0/hhcell" preComponent="silent1" postComponent="gs1" weight="1"/>
:::
:::
:::
:::
:::

::: {#continuousprojection .section}
[]{#schema-continuousprojection}

##### continuousProjection[\#](#continuousprojection "Link to this heading"){.headerlink}

*A projection between **presynapticPopulation** and
**postsynapticPopulation** through components **preComponent** at the
start and **postComponent** at the end of a [[continuousConnection]{.std
.std-ref}](#schema-continuousconnection){.reference .internal} or
[[continuousConnectionInstance]{.std
.std-ref}](#schema-continuousconnectioninstance){.reference .internal}.
Can be used for analog synapses.*

::: {.sd-tab-set .docutils}
Component References

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    |                                                 | [[popul     |
| *pre |                                                 | ation]{.std |
| syna |                                                 | .std-ref    |
| ptic |                                                 | }](#schema- |
| Popu |                                                 | population) |
| lati |                                                 | {.reference |
| on** |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   |                                                 | [[popul     |
| post |                                                 | ation]{.std |
| syna |                                                 | .std-ref    |
| ptic |                                                 | }](#schema- |
| Popu |                                                 | population) |
| lati |                                                 | {.reference |
| on** |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Children list

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **c  |                                                 | [[cont      |
| onne |                                                 | inuousConne |
| ctio |                                                 | ction]{.std |
| ns** |                                                 | .std-re     |
|      |                                                 | f}](#schema |
|      |                                                 | -continuous |
|      |                                                 | connection) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **c  |                                                 | [[c         |
| onne |                                                 | ontinuousCo |
| ctio |                                                 | nnectionIns |
| nIns |                                                 | tance]{.std |
| tanc |                                                 | .std        |
| es** |                                                 | -ref}](#sch |
|      |                                                 | ema-continu |
|      |                                                 | ousconnecti |
|      |                                                 | oninstance) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="ContinuousProjection">
      <xs:complexContent>
        <xs:extension base="BaseProjection">
          <xs:sequence>
            <xs:element name="continuousConnection" type="ContinuousConnection" minOccurs="0" maxOccurs="unbounded"/>
            <xs:element name="continuousConnectionInstance" type="ContinuousConnectionInstance" minOccurs="0" maxOccurs="unbounded"/>
            <xs:element name="continuousConnectionInstanceW" type="ContinuousConnectionInstanceW" minOccurs="0" maxOccurs="unbounded"/>
          </xs:sequence>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=ContinuousProjection){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import ContinuousProjection
    from neuroml.utils import component_factory

    variable = component_factory(
        ContinuousProjection,
        id: 'a NmlId (required)' = None,
        presynaptic_population: 'a NmlId (required)' = None,
        postsynaptic_population: 'a NmlId (required)' = None,
        continuous_connections: 'list of ContinuousConnection(s) (optional)' = None,
        continuous_connection_instances: 'list of ContinuousConnectionInstance(s) (optional)' = None,
        continuous_connection_instance_ws: 'list of ContinuousConnectionInstanceW(s) (optional)' = None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <continuousProjection id="testLinearGradedConn" presynapticPopulation="iafPop1" postsynapticPopulation="iafPop2">
        <continuousConnection id="0" preCell="0" postCell="0" preComponent="silent1" postComponent="gs1"/>
    </continuousProjection>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <continuousProjection id="testGradedConn" presynapticPopulation="iafPop1" postsynapticPopulation="iafPop3">
        <continuousConnection id="0" preCell="0" postCell="0" preComponent="silent2" postComponent="gs2"/>
    </continuousProjection>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <continuousProjection id="testGradedConn" presynapticPopulation="hhPop1" postsynapticPopulation="hhPop2">
        <continuousConnectionInstanceW id="0" preCell="../hhPop1/0/hhcell" postCell="../hhPop2/0/hhcell" preComponent="silent1" postComponent="gs1" weight="1"/>
    </continuousProjection>
:::
:::
:::
:::
:::

::: {#explicitinput .section}
[]{#schema-explicitinput}

##### explicitInput[\#](#explicitinput "Link to this heading"){.headerlink}

*An explicit input ( anything which extends [[basePointCurrent]{.std
.std-ref}](Landing.html#schema-basepointcurrent){.reference .internal} )
to a target cell in a population.*

::: {.sd-tab-set .docutils}
Text fields

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| **d    |                                                              |
| estina |                                                              |
| tion** |                                                              |
+--------+--------------------------------------------------------------+
| **     |                                                              |
| source |                                                              |
| Port** |                                                              |
+--------+--------------------------------------------------------------+
| **     |                                                              |
| target |                                                              |
| Port** |                                                              |
+--------+--------------------------------------------------------------+
:::
:::

Paths

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| **ta   |                                                              |
| rget** |                                                              |
+--------+--------------------------------------------------------------+
:::
:::

Component References

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    |                                                 | [[          |
| *inp |                                                 | basePointCu |
| ut** |                                                 | rrent]{.std |
|      |                                                 | .std        |
|      |                                                 | -ref}](Land |
|      |                                                 | ing.html#sc |
|      |                                                 | hema-basepo |
|      |                                                 | intcurrent) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="ExplicitInput">
      <xs:complexContent>
        <xs:extension base="BaseWithoutId">
          <xs:attribute name="target" type="Nml2PopulationReferencePath" use="required"/>
          <xs:attribute name="input" type="NmlId" use="required"/>
          <xs:attribute name="destination" type="NmlId"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=ExplicitInput){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import ExplicitInput
    from neuroml.utils import component_factory

    variable = component_factory(
        ExplicitInput,
        target: 'a Nml2PopulationReferencePath (required)' = None,
        input: 'a NmlId (required)' = None,
        destination: 'a NmlId (optional)' = None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <explicitInput target="iafPop1[0]" input="pulseGen1" destination="synapses"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <explicitInput target="iafPop1[0]" input="pulseGen2" destination="synapses"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <explicitInput target="iafPop1[0]" input="pulseGen3" destination="synapses"/>
:::
:::
:::
:::
:::

::: {#inputlist .section}
[]{#schema-inputlist}

##### inputList[\#](#inputlist "Link to this heading"){.headerlink}

*An explicit list of [[input]{.std .std-ref}](#schema-input){.reference
.internal}s to a **population.**.*

::: {.sd-tab-set .docutils}
Text fields

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| **     |                                                              |
| popula |                                                              |
| tion** |                                                              |
+--------+--------------------------------------------------------------+
:::
:::

Component References

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    |                                                 | [[          |
| *com |                                                 | basePointCu |
| pone |                                                 | rrent]{.std |
| nt** |                                                 | .std        |
|      |                                                 | -ref}](Land |
|      |                                                 | ing.html#sc |
|      |                                                 | hema-basepo |
|      |                                                 | intcurrent) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Children list

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   |                                                 | [[          |
| inpu |                                                 | input]{.std |
| ts** |                                                 | .st         |
|      |                                                 | d-ref}](#sc |
|      |                                                 | hema-input) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="InputList">
      <xs:complexContent>
        <xs:extension base="Base">
          <xs:sequence>
            <xs:element name="input" type="Input" minOccurs="0" maxOccurs="unbounded"/>
            <xs:element name="inputW" type="InputW" minOccurs="0" maxOccurs="unbounded"/>
          </xs:sequence>
          <xs:attribute name="population" type="NmlId" use="required"/>
          <xs:attribute name="component" type="NmlId" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=InputList){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import InputList
    from neuroml.utils import component_factory

    variable = component_factory(
        InputList,
        id: 'a NonNegativeInteger (required)' = None,
        populations: 'a NmlId (required)' = None,
        component: 'a NmlId (required)' = None,
        input: 'list of Input(s) (optional)' = None,
        input_ws: 'list of InputW(s) (optional)' = None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <inputList id="i1" component="pulseGen1" population="hhPop1">
        <input id="0" target="../hhPop1/0/hhcell" destination="synapses"/>
    </inputList>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <inputList id="i1" component="pulseGen1" population="iafPop1">
        <input id="0" target="../iafPop1/0/iaf" destination="synapses"/>
    </inputList>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <inputList id="i2" component="pulseGen2" population="iafPop2">
        <input id="0" target="../iafPop2/0/iaf" destination="synapses"/>
    </inputList>
:::
:::
:::
:::
:::

::: {#input .section}
[]{#schema-input}

##### input[\#](#input "Link to this heading"){.headerlink}

*Specifies a single input to a **target,** optionally giving the
**segmentId** ( default 0 ) and **fractionAlong** the segment ( default
0.5 ).*

::: {.sd-tab-set .docutils}
Text fields

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| *      | Optional specification of the segment to target, default 0   |
| *segme |                                                              |
| ntId** |                                                              |
+--------+--------------------------------------------------------------+
| **fra  | Optional specification of the fraction along the specified   |
| ctionA | segment, default 0.5                                         |
| long** |                                                              |
+--------+--------------------------------------------------------------+
| **d    |                                                              |
| estina |                                                              |
| tion** |                                                              |
+--------+--------------------------------------------------------------+
:::
:::

Paths

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| **ta   |                                                              |
| rget** |                                                              |
+--------+--------------------------------------------------------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="Input">
      <xs:complexContent>
        <xs:extension base="BaseNonNegativeIntegerId">
          <xs:attribute name="target" type="Nml2PopulationReferencePath" use="required"/>
          <xs:attribute name="destination" type="NmlId" use="required"/>
          <xs:attribute name="segmentId" type="NonNegativeInteger"/>
          <xs:attribute name="fractionAlong" type="ZeroToOne"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=Input){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import Input
    from neuroml.utils import component_factory

    variable = component_factory(
        Input,
        id: 'a NonNegativeInteger (required)' = None,
        target: 'a Nml2PopulationReferencePath (required)' = None,
        destination: 'a NmlId (required)' = None,
        segment_id: 'a NonNegativeInteger (optional)' = None,
        fraction_along: 'a ZeroToOne (optional)' = None,
        extensiontype_=None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <input id="0" target="../hhPop1/0/hhcell" destination="synapses"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <input id="0" target="../iafPop1/0/iaf" destination="synapses"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <input id="0" target="../iafPop2/0/iaf" destination="synapses"/>
:::
:::
:::
:::
:::

::: {#inputw .section}
[]{#schema-inputw}

##### inputW[\#](#inputw "Link to this heading"){.headerlink}

extends [[input]{.std .std-ref}](#schema-input){.reference .internal}

*Specifies input lists. Can set **weight** to scale individual inputs.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   |                                                 | Di          |
| weig |                                                 | mensionless |
| ht** |                                                 |             |
+------+-------------------------------------------------+-------------+
:::
:::

Text fields

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| **d    |                                                              |
| estina |                                                              |
| tion** |                                                              |
+--------+--------------------------------------------------------------+
:::
:::

Paths

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| **ta   |                                                              |
| rget** |                                                              |
+--------+--------------------------------------------------------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="InputW">
      <xs:complexContent>
        <xs:extension base="Input">
          <xs:attribute name="weight" type="xs:float" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=InputW){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import InputW
    from neuroml.utils import component_factory

    variable = component_factory(
        InputW,
        id: 'a NonNegativeInteger (required)' = None,
        target: 'a Nml2PopulationReferencePath (required)' = None,
        destination: 'a NmlId (required)' = None,
        segment_id: 'a NonNegativeInteger (optional)' = None,
        fraction_along: 'a ZeroToOne (optional)' = None,
        weight: 'a float (required)' = None,
    )
:::
:::
:::
:::
:::
:::

[]{#document-Userdocs/Schemas/PyNN}

::: {#pynn .section .tex2jax_ignore .mathjax_ignore}
[]{#schema-pynn}

#### PyNN[\#](#pynn "Link to this heading"){.headerlink}

**A number of ComponentType description of PyNN standard cells. All of
the cells extend [[basePyNNCell]{.std
.std-ref}](#schema-basepynncell){.reference .internal}, and the synapses
extend [[basePynnSynapse]{.std
.std-ref}](#schema-basepynnsynapse){.reference .internal}.**

------------------------------------------------------------------------

Original ComponentType definitions:
[PyNN.xml](https://github.com/NeuroML/NeuroML2/blob/master/NeuroML2CoreTypes//PyNN.xml){.reference
.external}. Schema against which NeuroML based on these should be valid:
[NeuroML_v2.3.xsd](https://github.com/NeuroML/NeuroML2/tree/master/Schemas/NeuroML2/NeuroML_v2.3.xsd){.reference
.external}. Generated on 14/08/24 from
[this](https://github.com/NeuroML/NeuroML2/commit/352244cff605cb1ba24fa7c11757dc818fe90fd2){.reference
.external} commit. Please file any issues or questions at the [issue
tracker here](https://github.com/NeuroML/NeuroML2/issues){.reference
.external}.

------------------------------------------------------------------------

::: {#basepynncell .section}
[]{#schema-basepynncell}

##### *basePyNNCell*[\#](#basepynncell "Link to this heading"){.headerlink}

extends *[[baseCellMembPot]{.std
.std-ref}](Landing.html#schema-basecellmembpot){.reference .internal}*

*Base type of any PyNN standard cell model. Note: membrane potential
**v** has dimensions voltage, but all other parameters are
dimensionless. This is to facilitate translation to and from PyNN
scripts in Python, where these parameters have implicit units, see
[http://neuralensemble.org/trac/PyNN/wiki/StandardModels](http://neuralensemble.org/trac/PyNN/wiki/StandardModels){.reference
.external}.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   |                                                 | Di          |
| cm** |                                                 | mensionless |
+------+-------------------------------------------------+-------------+
| **i_ |                                                 | Di          |
| offs |                                                 | mensionless |
| et** |                                                 |             |
+------+-------------------------------------------------+-------------+
| *    | This parameter is never used in the NeuroML2    | Di          |
| *tau | description of this cell! Any synapse producing | mensionless |
| _syn | a current can be placed on this cell            |             |
| _E** |                                                 |             |
+------+-------------------------------------------------+-------------+
| *    | This parameter is never used in the NeuroML2    | Di          |
| *tau | description of this cell! Any synapse producing | mensionless |
| _syn | a current can be placed on this cell            |             |
| _I** |                                                 |             |
+------+-------------------------------------------------+-------------+
| **   |                                                 | Di          |
| v_in |                                                 | mensionless |
| it** |                                                 |             |
+------+-------------------------------------------------+-------------+
:::
:::

Constants

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------------------+-----------------------------------+-------------+
| **MSEC** = 1ms     |                                   | [           |
|                    |                                   | [time]{.std |
|                    |                                   | .st         |
|                    |                                   | d-ref}](Lan |
|                    |                                   | ding.html#s |
|                    |                                   | chema-dimen |
|                    |                                   | sions-time) |
|                    |                                   | {.reference |
|                    |                                   | .internal}  |
+--------------------+-----------------------------------+-------------+
| **MVOLT** = 1mV    |                                   | [[vo        |
|                    |                                   | ltage]{.std |
|                    |                                   | .std-r      |
|                    |                                   | ef}](Landin |
|                    |                                   | g.html#sche |
|                    |                                   | ma-dimensio |
|                    |                                   | ns-voltage) |
|                    |                                   | {.reference |
|                    |                                   | .internal}  |
+--------------------+-----------------------------------+-------------+
| **NFARAD** = 1nF   |                                   | [[capaci    |
|                    |                                   | tance]{.std |
|                    |                                   | .std-ref}]  |
|                    |                                   | (Landing.ht |
|                    |                                   | ml#schema-d |
|                    |                                   | imensions-c |
|                    |                                   | apacitance) |
|                    |                                   | {.reference |
|                    |                                   | .internal}  |
+--------------------+-----------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **iS |                                                 | [[cu        |
| yn** |                                                 | rrent]{.std |
|      |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | Membrane potential *(from                       | [[vo        |
| *v** | [[baseCellMembPot]{.std                         | ltage]{.std |
|      | .std-ref}](                                     | .std-r      |
|      | Landing.html#schema-basecellmembpot){.reference | ef}](Landin |
|      | .internal})*                                    | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Spike event *(from [[baseSpikingCell]{.std      | Direction:  |
| *spi | .std-ref}](                                     | out         |
| ke** | Landing.html#schema-basespikingcell){.reference |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| **   |                                                 | Direction:  |
| spik |                                                 | in          |
| e_in |                                                 |             |
| _E** |                                                 |             |
+------+-------------------------------------------------+-------------+
| **   |                                                 | Direction:  |
| spik |                                                 | in          |
| e_in |                                                 |             |
| _I** |                                                 |             |
+------+-------------------------------------------------+-------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="basePyNNCell">
      <xs:complexContent>
        <xs:extension base="BaseCell">
          <xs:attribute name="cm" type="xs:float" use="required"/>
          <xs:attribute name="i_offset" type="xs:float" use="required"/>
          <xs:attribute name="tau_syn_E" type="xs:float" use="required"/>
          <xs:attribute name="tau_syn_I" type="xs:float" use="required"/>
          <xs:attribute name="v_init" type="xs:float" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::
:::
:::

::: {#basepynniafcell .section}
[]{#schema-basepynniafcell}

##### *basePyNNIaFCell*[\#](#basepynniafcell "Link to this heading"){.headerlink}

extends *[[basePyNNCell]{.std
.std-ref}](#schema-basepynncell){.reference .internal}*

*Base type of any PyNN standard integrate and fire model.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   | *(from [[basePyNNCell]{.std                     | Di          |
| cm** | .std-ref}](#schema-basepynncell){.reference     | mensionless |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| **i_ | *(from [[basePyNNCell]{.std                     | Di          |
| offs | .std-ref}](#schema-basepynncell){.reference     | mensionless |
| et** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| *    |                                                 | Di          |
| *tau |                                                 | mensionless |
| _m** |                                                 |             |
+------+-------------------------------------------------+-------------+
| **   |                                                 | Di          |
| tau_ |                                                 | mensionless |
| refr |                                                 |             |
| ac** |                                                 |             |
+------+-------------------------------------------------+-------------+
| *    | This parameter is never used in the NeuroML2    | Di          |
| *tau | description of this cell! Any synapse producing | mensionless |
| _syn | a current can be placed on this cell *(from     |             |
| _E** | [[basePyNNCell]{.std                            |             |
|      | .std-ref}](#schema-basepynncell){.reference     |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| *    | This parameter is never used in the NeuroML2    | Di          |
| *tau | description of this cell! Any synapse producing | mensionless |
| _syn | a current can be placed on this cell *(from     |             |
| _I** | [[basePyNNCell]{.std                            |             |
|      | .std-ref}](#schema-basepynncell){.reference     |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| **   | *(from [[basePyNNCell]{.std                     | Di          |
| v_in | .std-ref}](#schema-basepynncell){.reference     | mensionless |
| it** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| **v  |                                                 | Di          |
| _res |                                                 | mensionless |
| et** |                                                 |             |
+------+-------------------------------------------------+-------------+
| **   |                                                 | Di          |
| v_re |                                                 | mensionless |
| st** |                                                 |             |
+------+-------------------------------------------------+-------------+
| **v_ |                                                 | Di          |
| thre |                                                 | mensionless |
| sh** |                                                 |             |
+------+-------------------------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **iS | *(from [[basePyNNCell]{.std                     | [[cu        |
| yn** | .std-ref}](#schema-basepynncell){.reference     | rrent]{.std |
|      | .internal})*                                    | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | Membrane potential *(from                       | [[vo        |
| *v** | [[baseCellMembPot]{.std                         | ltage]{.std |
|      | .std-ref}](                                     | .std-r      |
|      | Landing.html#schema-basecellmembpot){.reference | ef}](Landin |
|      | .internal})*                                    | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Spike event *(from [[baseSpikingCell]{.std      | Direction:  |
| *spi | .std-ref}](                                     | out         |
| ke** | Landing.html#schema-basespikingcell){.reference |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| **   | *(from [[basePyNNCell]{.std                     | Direction:  |
| spik | .std-ref}](#schema-basepynncell){.reference     | in          |
| e_in | .internal})*                                    |             |
| _E** |                                                 |             |
+------+-------------------------------------------------+-------------+
| **   | *(from [[basePyNNCell]{.std                     | Direction:  |
| spik | .std-ref}](#schema-basepynncell){.reference     | in          |
| e_in | .internal})*                                    |             |
| _I** |                                                 |             |
+------+-------------------------------------------------+-------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="basePyNNIaFCell">
      <xs:complexContent>
        <xs:extension base="basePyNNCell">
          <xs:attribute name="tau_m" type="xs:float" use="required"/>
          <xs:attribute name="tau_refrac" type="xs:float" use="required"/>
          <xs:attribute name="v_reset" type="xs:float" use="required"/>
          <xs:attribute name="v_rest" type="xs:float" use="required"/>
          <xs:attribute name="v_thresh" type="xs:float" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::
:::
:::

::: {#basepynniafcondcell .section}
[]{#schema-basepynniafcondcell}

##### *basePyNNIaFCondCell*[\#](#basepynniafcondcell "Link to this heading"){.headerlink}

extends *[[basePyNNIaFCell]{.std
.std-ref}](#schema-basepynniafcell){.reference .internal}*

*Base type of conductance based PyNN IaF cell models.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   | *(from [[basePyNNCell]{.std                     | Di          |
| cm** | .std-ref}](#schema-basepynncell){.reference     | mensionless |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| **e  | This parameter is never used in the NeuroML2    | Di          |
| _rev | description of this cell! Any synapse producing | mensionless |
| _E** | a current can be placed on this cell            |             |
+------+-------------------------------------------------+-------------+
| **e  | This parameter is never used in the NeuroML2    | Di          |
| _rev | description of this cell! Any synapse producing | mensionless |
| _I** | a current can be placed on this cell            |             |
+------+-------------------------------------------------+-------------+
| **i_ | *(from [[basePyNNCell]{.std                     | Di          |
| offs | .std-ref}](#schema-basepynncell){.reference     | mensionless |
| et** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| *    | *(from [[basePyNNIaFCell]{.std                  | Di          |
| *tau | .std-ref}](#schema-basepynniafcell){.reference  | mensionless |
| _m** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| **   | *(from [[basePyNNIaFCell]{.std                  | Di          |
| tau_ | .std-ref}](#schema-basepynniafcell){.reference  | mensionless |
| refr | .internal})*                                    |             |
| ac** |                                                 |             |
+------+-------------------------------------------------+-------------+
| *    | This parameter is never used in the NeuroML2    | Di          |
| *tau | description of this cell! Any synapse producing | mensionless |
| _syn | a current can be placed on this cell *(from     |             |
| _E** | [[basePyNNCell]{.std                            |             |
|      | .std-ref}](#schema-basepynncell){.reference     |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| *    | This parameter is never used in the NeuroML2    | Di          |
| *tau | description of this cell! Any synapse producing | mensionless |
| _syn | a current can be placed on this cell *(from     |             |
| _I** | [[basePyNNCell]{.std                            |             |
|      | .std-ref}](#schema-basepynncell){.reference     |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| **   | *(from [[basePyNNCell]{.std                     | Di          |
| v_in | .std-ref}](#schema-basepynncell){.reference     | mensionless |
| it** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| **v  | *(from [[basePyNNIaFCell]{.std                  | Di          |
| _res | .std-ref}](#schema-basepynniafcell){.reference  | mensionless |
| et** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| **   | *(from [[basePyNNIaFCell]{.std                  | Di          |
| v_re | .std-ref}](#schema-basepynniafcell){.reference  | mensionless |
| st** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| **v_ | *(from [[basePyNNIaFCell]{.std                  | Di          |
| thre | .std-ref}](#schema-basepynniafcell){.reference  | mensionless |
| sh** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **iS | *(from [[basePyNNCell]{.std                     | [[cu        |
| yn** | .std-ref}](#schema-basepynncell){.reference     | rrent]{.std |
|      | .internal})*                                    | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | Membrane potential *(from                       | [[vo        |
| *v** | [[baseCellMembPot]{.std                         | ltage]{.std |
|      | .std-ref}](                                     | .std-r      |
|      | Landing.html#schema-basecellmembpot){.reference | ef}](Landin |
|      | .internal})*                                    | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Spike event *(from [[baseSpikingCell]{.std      | Direction:  |
| *spi | .std-ref}](                                     | out         |
| ke** | Landing.html#schema-basespikingcell){.reference |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| **   | *(from [[basePyNNCell]{.std                     | Direction:  |
| spik | .std-ref}](#schema-basepynncell){.reference     | in          |
| e_in | .internal})*                                    |             |
| _E** |                                                 |             |
+------+-------------------------------------------------+-------------+
| **   | *(from [[basePyNNCell]{.std                     | Direction:  |
| spik | .std-ref}](#schema-basepynncell){.reference     | in          |
| e_in | .internal})*                                    |             |
| _I** |                                                 |             |
+------+-------------------------------------------------+-------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="basePyNNIaFCondCell">
      <xs:complexContent>
        <xs:extension base="basePyNNIaFCell">
          <xs:attribute name="e_rev_E" type="xs:float" use="required"/>
          <xs:attribute name="e_rev_I" type="xs:float" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::
:::
:::

::: {#if-curr-alpha .section}
[]{#schema-if-curr-alpha}

##### IF_curr_alpha[\#](#if-curr-alpha "Link to this heading"){.headerlink}

extends *[[basePyNNIaFCell]{.std
.std-ref}](#schema-basepynniafcell){.reference .internal}*

*Leaky integrate and fire model with fixed threshold and
alpha-function-shaped post-synaptic current.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   | *(from [[basePyNNCell]{.std                     | Di          |
| cm** | .std-ref}](#schema-basepynncell){.reference     | mensionless |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| **i_ | *(from [[basePyNNCell]{.std                     | Di          |
| offs | .std-ref}](#schema-basepynncell){.reference     | mensionless |
| et** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| *    | *(from [[basePyNNIaFCell]{.std                  | Di          |
| *tau | .std-ref}](#schema-basepynniafcell){.reference  | mensionless |
| _m** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| **   | *(from [[basePyNNIaFCell]{.std                  | Di          |
| tau_ | .std-ref}](#schema-basepynniafcell){.reference  | mensionless |
| refr | .internal})*                                    |             |
| ac** |                                                 |             |
+------+-------------------------------------------------+-------------+
| *    | This parameter is never used in the NeuroML2    | Di          |
| *tau | description of this cell! Any synapse producing | mensionless |
| _syn | a current can be placed on this cell *(from     |             |
| _E** | [[basePyNNCell]{.std                            |             |
|      | .std-ref}](#schema-basepynncell){.reference     |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| *    | This parameter is never used in the NeuroML2    | Di          |
| *tau | description of this cell! Any synapse producing | mensionless |
| _syn | a current can be placed on this cell *(from     |             |
| _I** | [[basePyNNCell]{.std                            |             |
|      | .std-ref}](#schema-basepynncell){.reference     |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| **   | *(from [[basePyNNCell]{.std                     | Di          |
| v_in | .std-ref}](#schema-basepynncell){.reference     | mensionless |
| it** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| **v  | *(from [[basePyNNIaFCell]{.std                  | Di          |
| _res | .std-ref}](#schema-basepynniafcell){.reference  | mensionless |
| et** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| **   | *(from [[basePyNNIaFCell]{.std                  | Di          |
| v_re | .std-ref}](#schema-basepynniafcell){.reference  | mensionless |
| st** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| **v_ | *(from [[basePyNNIaFCell]{.std                  | Di          |
| thre | .std-ref}](#schema-basepynniafcell){.reference  | mensionless |
| sh** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **iS | *(from [[basePyNNCell]{.std                     | [[cu        |
| yn** | .std-ref}](#schema-basepynncell){.reference     | rrent]{.std |
|      | .internal})*                                    | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | Membrane potential *(from                       | [[vo        |
| *v** | [[baseCellMembPot]{.std                         | ltage]{.std |
|      | .std-ref}](                                     | .std-r      |
|      | Landing.html#schema-basecellmembpot){.reference | ef}](Landin |
|      | .internal})*                                    | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Spike event *(from [[baseSpikingCell]{.std      | Direction:  |
| *spi | .std-ref}](                                     | out         |
| ke** | Landing.html#schema-basespikingcell){.reference |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| **   | *(from [[basePyNNCell]{.std                     | Direction:  |
| spik | .std-ref}](#schema-basepynncell){.reference     | in          |
| e_in | .internal})*                                    |             |
| _E** |                                                 |             |
+------+-------------------------------------------------+-------------+
| **   | *(from [[basePyNNCell]{.std                     | Direction:  |
| spik | .std-ref}](#schema-basepynncell){.reference     | in          |
| e_in | .internal})*                                    |             |
| _I** |                                                 |             |
+------+-------------------------------------------------+-------------+
:::
:::

Attachments

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **sy |                                                 | [[baseSy    |
| naps |                                                 | napse]{.std |
| es** |                                                 | .std-ref}]  |
|      |                                                 | (Landing.ht |
|      |                                                 | ml#schema-b |
|      |                                                 | asesynapse) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***State Variables***

:   **v**: [[voltage]{.std
    .std-ref}](Landing.html#schema-dimensions-voltage){.reference
    .internal}  (exposed as **v**)

:   **lastSpikeTime**: [[time]{.std
    .std-ref}](Landing.html#schema-dimensions-time){.reference
    .internal}

***On Start***

:   **v** = v_init \* MVOLT

***Derived Variables***

:   **iSyn** = synapses\[\*\]-\>i(reduce method: add) (exposed as
    **iSyn**)

***Regime: refractory (initial)***

:   ***On Entry***

:      **lastSpikeTime** = t

:      **v** = v_reset \* MVOLT

:   ***On Conditions***

:      IF t \> lastSpikeTime + (tau_refrac\*MSEC) THEN

:       TRANSITION to REGIME **integrating**

***Regime: integrating (initial)***

:   ***On Conditions***

:      IF v \> v_thresh \* MVOLT THEN

:       EVENT OUT on port: **spike**

:       TRANSITION to REGIME **refractory**

:   ***Time Derivatives***

:      d **v** /dt = (MVOLT \* ((i_offset/cm) + ((v_rest - (v/MVOLT)) /
    tau_m))/MSEC) + (iSyn / (cm \* NFARAD))
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="IF_curr_alpha">
      <xs:complexContent>
        <xs:extension base="basePyNNIaFCell">
                </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=IF_curr_alpha){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import IF_curr_alpha
    from neuroml.utils import component_factory

    variable = component_factory(
        IF_curr_alpha,
        id: 'a NmlId (required)' = None,
        metaid: 'a MetaId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        neuro_lex_id: 'a NeuroLexId (optional)' = None,
        cm: 'a float (required)' = None,
        i_offset: 'a float (required)' = None,
        tau_syn_E: 'a float (required)' = None,
        tau_syn_I: 'a float (required)' = None,
        v_init: 'a float (required)' = None,
        tau_m: 'a float (required)' = None,
        tau_refrac: 'a float (required)' = None,
        v_reset: 'a float (required)' = None,
        v_rest: 'a float (required)' = None,
        v_thresh: 'a float (required)' = None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <IF_curr_alpha id="IF_curr_alpha" cm="1.0" i_offset="0.9" tau_m="20.0" tau_refrac="10.0" tau_syn_E="0.5" tau_syn_I="0.5" v_init="-65" v_reset="-62.0" v_rest="-65.0" v_thresh="-52.0"/>
:::
:::
:::
:::
:::

::: {#if-curr-exp .section}
[]{#schema-if-curr-exp}

##### IF_curr_exp[\#](#if-curr-exp "Link to this heading"){.headerlink}

extends *[[basePyNNIaFCell]{.std
.std-ref}](#schema-basepynniafcell){.reference .internal}*

*Leaky integrate and fire model with fixed threshold and
decaying-exponential post-synaptic current.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   | *(from [[basePyNNCell]{.std                     | Di          |
| cm** | .std-ref}](#schema-basepynncell){.reference     | mensionless |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| **i_ | *(from [[basePyNNCell]{.std                     | Di          |
| offs | .std-ref}](#schema-basepynncell){.reference     | mensionless |
| et** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| *    | *(from [[basePyNNIaFCell]{.std                  | Di          |
| *tau | .std-ref}](#schema-basepynniafcell){.reference  | mensionless |
| _m** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| **   | *(from [[basePyNNIaFCell]{.std                  | Di          |
| tau_ | .std-ref}](#schema-basepynniafcell){.reference  | mensionless |
| refr | .internal})*                                    |             |
| ac** |                                                 |             |
+------+-------------------------------------------------+-------------+
| *    | This parameter is never used in the NeuroML2    | Di          |
| *tau | description of this cell! Any synapse producing | mensionless |
| _syn | a current can be placed on this cell *(from     |             |
| _E** | [[basePyNNCell]{.std                            |             |
|      | .std-ref}](#schema-basepynncell){.reference     |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| *    | This parameter is never used in the NeuroML2    | Di          |
| *tau | description of this cell! Any synapse producing | mensionless |
| _syn | a current can be placed on this cell *(from     |             |
| _I** | [[basePyNNCell]{.std                            |             |
|      | .std-ref}](#schema-basepynncell){.reference     |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| **   | *(from [[basePyNNCell]{.std                     | Di          |
| v_in | .std-ref}](#schema-basepynncell){.reference     | mensionless |
| it** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| **v  | *(from [[basePyNNIaFCell]{.std                  | Di          |
| _res | .std-ref}](#schema-basepynniafcell){.reference  | mensionless |
| et** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| **   | *(from [[basePyNNIaFCell]{.std                  | Di          |
| v_re | .std-ref}](#schema-basepynniafcell){.reference  | mensionless |
| st** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| **v_ | *(from [[basePyNNIaFCell]{.std                  | Di          |
| thre | .std-ref}](#schema-basepynniafcell){.reference  | mensionless |
| sh** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **iS | *(from [[basePyNNCell]{.std                     | [[cu        |
| yn** | .std-ref}](#schema-basepynncell){.reference     | rrent]{.std |
|      | .internal})*                                    | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | Membrane potential *(from                       | [[vo        |
| *v** | [[baseCellMembPot]{.std                         | ltage]{.std |
|      | .std-ref}](                                     | .std-r      |
|      | Landing.html#schema-basecellmembpot){.reference | ef}](Landin |
|      | .internal})*                                    | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Spike event *(from [[baseSpikingCell]{.std      | Direction:  |
| *spi | .std-ref}](                                     | out         |
| ke** | Landing.html#schema-basespikingcell){.reference |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| **   | *(from [[basePyNNCell]{.std                     | Direction:  |
| spik | .std-ref}](#schema-basepynncell){.reference     | in          |
| e_in | .internal})*                                    |             |
| _E** |                                                 |             |
+------+-------------------------------------------------+-------------+
| **   | *(from [[basePyNNCell]{.std                     | Direction:  |
| spik | .std-ref}](#schema-basepynncell){.reference     | in          |
| e_in | .internal})*                                    |             |
| _I** |                                                 |             |
+------+-------------------------------------------------+-------------+
:::
:::

Attachments

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **sy |                                                 | [[baseSy    |
| naps |                                                 | napse]{.std |
| es** |                                                 | .std-ref}]  |
|      |                                                 | (Landing.ht |
|      |                                                 | ml#schema-b |
|      |                                                 | asesynapse) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***State Variables***

:   **v**: [[voltage]{.std
    .std-ref}](Landing.html#schema-dimensions-voltage){.reference
    .internal}  (exposed as **v**)

:   **lastSpikeTime**: [[time]{.std
    .std-ref}](Landing.html#schema-dimensions-time){.reference
    .internal}

***On Start***

:   **v** = v_init \* MVOLT

***Derived Variables***

:   **iSyn** = synapses\[\*\]-\>i(reduce method: add) (exposed as
    **iSyn**)

***Regime: refractory (initial)***

:   ***On Entry***

:      **lastSpikeTime** = t

:      **v** = v_reset \* MVOLT

:   ***On Conditions***

:      IF t \> lastSpikeTime + (tau_refrac\*MSEC) THEN

:       TRANSITION to REGIME **integrating**

***Regime: integrating (initial)***

:   ***On Conditions***

:      IF v \> v_thresh \* MVOLT THEN

:       EVENT OUT on port: **spike**

:       TRANSITION to REGIME **refractory**

:   ***Time Derivatives***

:      d **v** /dt = (MVOLT \* (((i_offset)/cm) + ((v_rest - (v/MVOLT))
    / tau_m))/MSEC) + (iSyn / (cm \* NFARAD))
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="IF_curr_exp">
      <xs:complexContent>
        <xs:extension base="basePyNNIaFCell">
                </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=IF_curr_exp){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import IF_curr_exp
    from neuroml.utils import component_factory

    variable = component_factory(
        IF_curr_exp,
        id: 'a NmlId (required)' = None,
        metaid: 'a MetaId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        neuro_lex_id: 'a NeuroLexId (optional)' = None,
        cm: 'a float (required)' = None,
        i_offset: 'a float (required)' = None,
        tau_syn_E: 'a float (required)' = None,
        tau_syn_I: 'a float (required)' = None,
        v_init: 'a float (required)' = None,
        tau_m: 'a float (required)' = None,
        tau_refrac: 'a float (required)' = None,
        v_reset: 'a float (required)' = None,
        v_rest: 'a float (required)' = None,
        v_thresh: 'a float (required)' = None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <IF_curr_exp id="IF_curr_exp" cm="1.0" i_offset="1.0" tau_m="20.0" tau_refrac="8.0" tau_syn_E="5.0" tau_syn_I="5.0" v_init="-65" v_reset="-70.0" v_rest="-65.0" v_thresh="-50.0"/>
:::
:::
:::
:::
:::

::: {#if-cond-alpha .section}
[]{#schema-if-cond-alpha}

##### IF_cond_alpha[\#](#if-cond-alpha "Link to this heading"){.headerlink}

extends *[[basePyNNIaFCondCell]{.std
.std-ref}](#schema-basepynniafcondcell){.reference .internal}*

*Leaky integrate and fire model with fixed threshold and
alpha-function-shaped post-synaptic conductance.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   | *(from [[basePyNNCell]{.std                     | Di          |
| cm** | .std-ref}](#schema-basepynncell){.reference     | mensionless |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| **e  | This parameter is never used in the NeuroML2    | Di          |
| _rev | description of this cell! Any synapse producing | mensionless |
| _E** | a current can be placed on this cell *(from     |             |
|      | [[basePyNNIaFCondCell]{.std                     |             |
|      | .st                                             |             |
|      | d-ref}](#schema-basepynniafcondcell){.reference |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| **e  | This parameter is never used in the NeuroML2    | Di          |
| _rev | description of this cell! Any synapse producing | mensionless |
| _I** | a current can be placed on this cell *(from     |             |
|      | [[basePyNNIaFCondCell]{.std                     |             |
|      | .st                                             |             |
|      | d-ref}](#schema-basepynniafcondcell){.reference |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| **i_ | *(from [[basePyNNCell]{.std                     | Di          |
| offs | .std-ref}](#schema-basepynncell){.reference     | mensionless |
| et** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| *    | *(from [[basePyNNIaFCell]{.std                  | Di          |
| *tau | .std-ref}](#schema-basepynniafcell){.reference  | mensionless |
| _m** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| **   | *(from [[basePyNNIaFCell]{.std                  | Di          |
| tau_ | .std-ref}](#schema-basepynniafcell){.reference  | mensionless |
| refr | .internal})*                                    |             |
| ac** |                                                 |             |
+------+-------------------------------------------------+-------------+
| *    | This parameter is never used in the NeuroML2    | Di          |
| *tau | description of this cell! Any synapse producing | mensionless |
| _syn | a current can be placed on this cell *(from     |             |
| _E** | [[basePyNNCell]{.std                            |             |
|      | .std-ref}](#schema-basepynncell){.reference     |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| *    | This parameter is never used in the NeuroML2    | Di          |
| *tau | description of this cell! Any synapse producing | mensionless |
| _syn | a current can be placed on this cell *(from     |             |
| _I** | [[basePyNNCell]{.std                            |             |
|      | .std-ref}](#schema-basepynncell){.reference     |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| **   | *(from [[basePyNNCell]{.std                     | Di          |
| v_in | .std-ref}](#schema-basepynncell){.reference     | mensionless |
| it** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| **v  | *(from [[basePyNNIaFCell]{.std                  | Di          |
| _res | .std-ref}](#schema-basepynniafcell){.reference  | mensionless |
| et** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| **   | *(from [[basePyNNIaFCell]{.std                  | Di          |
| v_re | .std-ref}](#schema-basepynniafcell){.reference  | mensionless |
| st** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| **v_ | *(from [[basePyNNIaFCell]{.std                  | Di          |
| thre | .std-ref}](#schema-basepynniafcell){.reference  | mensionless |
| sh** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **iS | *(from [[basePyNNCell]{.std                     | [[cu        |
| yn** | .std-ref}](#schema-basepynncell){.reference     | rrent]{.std |
|      | .internal})*                                    | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | Membrane potential *(from                       | [[vo        |
| *v** | [[baseCellMembPot]{.std                         | ltage]{.std |
|      | .std-ref}](                                     | .std-r      |
|      | Landing.html#schema-basecellmembpot){.reference | ef}](Landin |
|      | .internal})*                                    | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Spike event *(from [[baseSpikingCell]{.std      | Direction:  |
| *spi | .std-ref}](                                     | out         |
| ke** | Landing.html#schema-basespikingcell){.reference |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| **   | *(from [[basePyNNCell]{.std                     | Direction:  |
| spik | .std-ref}](#schema-basepynncell){.reference     | in          |
| e_in | .internal})*                                    |             |
| _E** |                                                 |             |
+------+-------------------------------------------------+-------------+
| **   | *(from [[basePyNNCell]{.std                     | Direction:  |
| spik | .std-ref}](#schema-basepynncell){.reference     | in          |
| e_in | .internal})*                                    |             |
| _I** |                                                 |             |
+------+-------------------------------------------------+-------------+
:::
:::

Attachments

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **sy |                                                 | [[baseSy    |
| naps |                                                 | napse]{.std |
| es** |                                                 | .std-ref}]  |
|      |                                                 | (Landing.ht |
|      |                                                 | ml#schema-b |
|      |                                                 | asesynapse) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***State Variables***

:   **v**: [[voltage]{.std
    .std-ref}](Landing.html#schema-dimensions-voltage){.reference
    .internal}  (exposed as **v**)

:   **lastSpikeTime**: [[time]{.std
    .std-ref}](Landing.html#schema-dimensions-time){.reference
    .internal}

***On Start***

:   **v** = v_init \* MVOLT

***Derived Variables***

:   **iSyn** = synapses\[\*\]-\>i(reduce method: add) (exposed as
    **iSyn**)

***Regime: refractory (initial)***

:   ***On Entry***

:      **lastSpikeTime** = t

:      **v** = v_reset \* MVOLT

:   ***On Conditions***

:      IF t \> lastSpikeTime + (tau_refrac\*MSEC) THEN

:       TRANSITION to REGIME **integrating**

***Regime: integrating (initial)***

:   ***On Conditions***

:      IF v \> v_thresh \* MVOLT THEN

:       EVENT OUT on port: **spike**

:       TRANSITION to REGIME **refractory**

:   ***Time Derivatives***

:      d **v** /dt = (MVOLT \* (((i_offset) / cm) + ((v_rest - (v /
    MVOLT)) / tau_m)) / MSEC) + (iSyn / (cm \* NFARAD))
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="IF_cond_alpha">
      <xs:complexContent>
        <xs:extension base="basePyNNIaFCondCell">
                </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=IF_cond_alpha){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import IF_cond_alpha
    from neuroml.utils import component_factory

    variable = component_factory(
        IF_cond_alpha,
        id: 'a NmlId (required)' = None,
        metaid: 'a MetaId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        neuro_lex_id: 'a NeuroLexId (optional)' = None,
        cm: 'a float (required)' = None,
        i_offset: 'a float (required)' = None,
        tau_syn_E: 'a float (required)' = None,
        tau_syn_I: 'a float (required)' = None,
        v_init: 'a float (required)' = None,
        tau_m: 'a float (required)' = None,
        tau_refrac: 'a float (required)' = None,
        v_reset: 'a float (required)' = None,
        v_rest: 'a float (required)' = None,
        v_thresh: 'a float (required)' = None,
        e_rev_E: 'a float (required)' = None,
        e_rev_I: 'a float (required)' = None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <IF_cond_alpha id="IF_cond_alpha" cm="1.0" e_rev_E="0.0" e_rev_I="-70.0" i_offset="0.9" tau_m="20.0" tau_refrac="5.0" tau_syn_E="0.3" tau_syn_I="0.5" v_init="-65" v_reset="-65.0" v_rest="-65.0" v_thresh="-50.0"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <IF_cond_alpha id="silent_cell" cm="1.0" e_rev_E="0.0" e_rev_I="-70.0" i_offset="0" tau_m="20.0" tau_refrac="5.0" tau_syn_E="5" tau_syn_I="10" v_init="-65" v_reset="-65.0" v_rest="-65.0" v_thresh="-50.0"/>
:::
:::
:::
:::
:::

::: {#if-cond-exp .section}
[]{#schema-if-cond-exp}

##### IF_cond_exp[\#](#if-cond-exp "Link to this heading"){.headerlink}

extends *[[basePyNNIaFCondCell]{.std
.std-ref}](#schema-basepynniafcondcell){.reference .internal}*

*Leaky integrate and fire model with fixed threshold and
exponentially-decaying post-synaptic conductance.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   | *(from [[basePyNNCell]{.std                     | Di          |
| cm** | .std-ref}](#schema-basepynncell){.reference     | mensionless |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| **e  | This parameter is never used in the NeuroML2    | Di          |
| _rev | description of this cell! Any synapse producing | mensionless |
| _E** | a current can be placed on this cell *(from     |             |
|      | [[basePyNNIaFCondCell]{.std                     |             |
|      | .st                                             |             |
|      | d-ref}](#schema-basepynniafcondcell){.reference |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| **e  | This parameter is never used in the NeuroML2    | Di          |
| _rev | description of this cell! Any synapse producing | mensionless |
| _I** | a current can be placed on this cell *(from     |             |
|      | [[basePyNNIaFCondCell]{.std                     |             |
|      | .st                                             |             |
|      | d-ref}](#schema-basepynniafcondcell){.reference |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| **i_ | *(from [[basePyNNCell]{.std                     | Di          |
| offs | .std-ref}](#schema-basepynncell){.reference     | mensionless |
| et** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| *    | *(from [[basePyNNIaFCell]{.std                  | Di          |
| *tau | .std-ref}](#schema-basepynniafcell){.reference  | mensionless |
| _m** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| **   | *(from [[basePyNNIaFCell]{.std                  | Di          |
| tau_ | .std-ref}](#schema-basepynniafcell){.reference  | mensionless |
| refr | .internal})*                                    |             |
| ac** |                                                 |             |
+------+-------------------------------------------------+-------------+
| *    | This parameter is never used in the NeuroML2    | Di          |
| *tau | description of this cell! Any synapse producing | mensionless |
| _syn | a current can be placed on this cell *(from     |             |
| _E** | [[basePyNNCell]{.std                            |             |
|      | .std-ref}](#schema-basepynncell){.reference     |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| *    | This parameter is never used in the NeuroML2    | Di          |
| *tau | description of this cell! Any synapse producing | mensionless |
| _syn | a current can be placed on this cell *(from     |             |
| _I** | [[basePyNNCell]{.std                            |             |
|      | .std-ref}](#schema-basepynncell){.reference     |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| **   | *(from [[basePyNNCell]{.std                     | Di          |
| v_in | .std-ref}](#schema-basepynncell){.reference     | mensionless |
| it** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| **v  | *(from [[basePyNNIaFCell]{.std                  | Di          |
| _res | .std-ref}](#schema-basepynniafcell){.reference  | mensionless |
| et** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| **   | *(from [[basePyNNIaFCell]{.std                  | Di          |
| v_re | .std-ref}](#schema-basepynniafcell){.reference  | mensionless |
| st** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| **v_ | *(from [[basePyNNIaFCell]{.std                  | Di          |
| thre | .std-ref}](#schema-basepynniafcell){.reference  | mensionless |
| sh** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **iS | *(from [[basePyNNCell]{.std                     | [[cu        |
| yn** | .std-ref}](#schema-basepynncell){.reference     | rrent]{.std |
|      | .internal})*                                    | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | Membrane potential *(from                       | [[vo        |
| *v** | [[baseCellMembPot]{.std                         | ltage]{.std |
|      | .std-ref}](                                     | .std-r      |
|      | Landing.html#schema-basecellmembpot){.reference | ef}](Landin |
|      | .internal})*                                    | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Spike event *(from [[baseSpikingCell]{.std      | Direction:  |
| *spi | .std-ref}](                                     | out         |
| ke** | Landing.html#schema-basespikingcell){.reference |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| **   | *(from [[basePyNNCell]{.std                     | Direction:  |
| spik | .std-ref}](#schema-basepynncell){.reference     | in          |
| e_in | .internal})*                                    |             |
| _E** |                                                 |             |
+------+-------------------------------------------------+-------------+
| **   | *(from [[basePyNNCell]{.std                     | Direction:  |
| spik | .std-ref}](#schema-basepynncell){.reference     | in          |
| e_in | .internal})*                                    |             |
| _I** |                                                 |             |
+------+-------------------------------------------------+-------------+
:::
:::

Attachments

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **sy |                                                 | [[baseSy    |
| naps |                                                 | napse]{.std |
| es** |                                                 | .std-ref}]  |
|      |                                                 | (Landing.ht |
|      |                                                 | ml#schema-b |
|      |                                                 | asesynapse) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***State Variables***

:   **v**: [[voltage]{.std
    .std-ref}](Landing.html#schema-dimensions-voltage){.reference
    .internal}  (exposed as **v**)

:   **lastSpikeTime**: [[time]{.std
    .std-ref}](Landing.html#schema-dimensions-time){.reference
    .internal}

***On Start***

:   **v** = v_init \* MVOLT

***Derived Variables***

:   **iSyn** = synapses\[\*\]-\>i(reduce method: add) (exposed as
    **iSyn**)

***Regime: refractory (initial)***

:   ***On Entry***

:      **lastSpikeTime** = t

:      **v** = v_reset \* MVOLT

:   ***On Conditions***

:      IF t \> lastSpikeTime + (tau_refrac\*MSEC) THEN

:       TRANSITION to REGIME **integrating**

***Regime: integrating (initial)***

:   ***On Conditions***

:      IF v \> v_thresh \* MVOLT THEN

:       EVENT OUT on port: **spike**

:       TRANSITION to REGIME **refractory**

:   ***Time Derivatives***

:      d **v** /dt = (MVOLT \* (((i_offset)/cm) + ((v_rest - (v /
    MVOLT)) / tau_m)) / MSEC) + (iSyn / (cm \* NFARAD))
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="IF_cond_exp">
      <xs:complexContent>
        <xs:extension base="basePyNNIaFCondCell">
                </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=IF_cond_exp){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import IF_cond_exp
    from neuroml.utils import component_factory

    variable = component_factory(
        IF_cond_exp,
        id: 'a NmlId (required)' = None,
        metaid: 'a MetaId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        neuro_lex_id: 'a NeuroLexId (optional)' = None,
        cm: 'a float (required)' = None,
        i_offset: 'a float (required)' = None,
        tau_syn_E: 'a float (required)' = None,
        tau_syn_I: 'a float (required)' = None,
        v_init: 'a float (required)' = None,
        tau_m: 'a float (required)' = None,
        tau_refrac: 'a float (required)' = None,
        v_reset: 'a float (required)' = None,
        v_rest: 'a float (required)' = None,
        v_thresh: 'a float (required)' = None,
        e_rev_E: 'a float (required)' = None,
        e_rev_I: 'a float (required)' = None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <IF_cond_exp id="IF_cond_exp" cm="1.0" e_rev_E="0.0" e_rev_I="-70.0" i_offset="1.0" tau_m="20.0" tau_refrac="5.0" tau_syn_E="5.0" tau_syn_I="5.0" v_init="-65" v_reset="-68.0" v_rest="-65.0" v_thresh="-52.0"/>
:::
:::
:::
:::
:::

::: {#eif-cond-exp-isfa-ista .section}
[]{#schema-eif-cond-exp-isfa-ista}

##### EIF_cond_exp_isfa_ista[\#](#eif-cond-exp-isfa-ista "Link to this heading"){.headerlink}

extends *[[basePyNNIaFCondCell]{.std
.std-ref}](#schema-basepynniafcondcell){.reference .internal}*

*Adaptive exponential integrate and fire neuron according to Brette R
and Gerstner W ( 2005 ) with exponentially-decaying post-synaptic
conductance.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    |                                                 | Di          |
| *a** |                                                 | mensionless |
+------+-------------------------------------------------+-------------+
| *    |                                                 | Di          |
| *b** |                                                 | mensionless |
+------+-------------------------------------------------+-------------+
| **   | *(from [[basePyNNCell]{.std                     | Di          |
| cm** | .std-ref}](#schema-basepynncell){.reference     | mensionless |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| **d  |                                                 | Di          |
| elta |                                                 | mensionless |
| _T** |                                                 |             |
+------+-------------------------------------------------+-------------+
| **e  | This parameter is never used in the NeuroML2    | Di          |
| _rev | description of this cell! Any synapse producing | mensionless |
| _E** | a current can be placed on this cell *(from     |             |
|      | [[basePyNNIaFCondCell]{.std                     |             |
|      | .st                                             |             |
|      | d-ref}](#schema-basepynniafcondcell){.reference |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| **e  | This parameter is never used in the NeuroML2    | Di          |
| _rev | description of this cell! Any synapse producing | mensionless |
| _I** | a current can be placed on this cell *(from     |             |
|      | [[basePyNNIaFCondCell]{.std                     |             |
|      | .st                                             |             |
|      | d-ref}](#schema-basepynniafcondcell){.reference |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| **i_ | *(from [[basePyNNCell]{.std                     | Di          |
| offs | .std-ref}](#schema-basepynncell){.reference     | mensionless |
| et** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| *    | *(from [[basePyNNIaFCell]{.std                  | Di          |
| *tau | .std-ref}](#schema-basepynniafcell){.reference  | mensionless |
| _m** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| **   | *(from [[basePyNNIaFCell]{.std                  | Di          |
| tau_ | .std-ref}](#schema-basepynniafcell){.reference  | mensionless |
| refr | .internal})*                                    |             |
| ac** |                                                 |             |
+------+-------------------------------------------------+-------------+
| *    | This parameter is never used in the NeuroML2    | Di          |
| *tau | description of this cell! Any synapse producing | mensionless |
| _syn | a current can be placed on this cell *(from     |             |
| _E** | [[basePyNNCell]{.std                            |             |
|      | .std-ref}](#schema-basepynncell){.reference     |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| *    | This parameter is never used in the NeuroML2    | Di          |
| *tau | description of this cell! Any synapse producing | mensionless |
| _syn | a current can be placed on this cell *(from     |             |
| _I** | [[basePyNNCell]{.std                            |             |
|      | .std-ref}](#schema-basepynncell){.reference     |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| *    |                                                 | Di          |
| *tau |                                                 | mensionless |
| _w** |                                                 |             |
+------+-------------------------------------------------+-------------+
| **   | *(from [[basePyNNCell]{.std                     | Di          |
| v_in | .std-ref}](#schema-basepynncell){.reference     | mensionless |
| it** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| **v  | *(from [[basePyNNIaFCell]{.std                  | Di          |
| _res | .std-ref}](#schema-basepynniafcell){.reference  | mensionless |
| et** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| **   | *(from [[basePyNNIaFCell]{.std                  | Di          |
| v_re | .std-ref}](#schema-basepynniafcell){.reference  | mensionless |
| st** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| **v  |                                                 | Di          |
| _spi |                                                 | mensionless |
| ke** |                                                 |             |
+------+-------------------------------------------------+-------------+
| **v_ | *(from [[basePyNNIaFCell]{.std                  | Di          |
| thre | .std-ref}](#schema-basepynniafcell){.reference  | mensionless |
| sh** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Derived parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    |                                                 | Di          |
| *eif |                                                 | mensionless |
| _thr |                                                 |             |
| esho |                                                 |             |
| ld** |                                                 |             |
+------+-------------------------------------------------+-------------+
:::

   **eif_threshold** = v_spike \* H(delta_T-1e-12) + v_thresh \*
H(-1\*delta_T+1e-9)
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **iS | *(from [[basePyNNCell]{.std                     | [[cu        |
| yn** | .std-ref}](#schema-basepynncell){.reference     | rrent]{.std |
|      | .internal})*                                    | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | Membrane potential *(from                       | [[vo        |
| *v** | [[baseCellMembPot]{.std                         | ltage]{.std |
|      | .std-ref}](                                     | .std-r      |
|      | Landing.html#schema-basecellmembpot){.reference | ef}](Landin |
|      | .internal})*                                    | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    |                                                 | Di          |
| *w** |                                                 | mensionless |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Spike event *(from [[baseSpikingCell]{.std      | Direction:  |
| *spi | .std-ref}](                                     | out         |
| ke** | Landing.html#schema-basespikingcell){.reference |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| **   | *(from [[basePyNNCell]{.std                     | Direction:  |
| spik | .std-ref}](#schema-basepynncell){.reference     | in          |
| e_in | .internal})*                                    |             |
| _E** |                                                 |             |
+------+-------------------------------------------------+-------------+
| **   | *(from [[basePyNNCell]{.std                     | Direction:  |
| spik | .std-ref}](#schema-basepynncell){.reference     | in          |
| e_in | .internal})*                                    |             |
| _I** |                                                 |             |
+------+-------------------------------------------------+-------------+
:::
:::

Attachments

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **sy |                                                 | [[baseSy    |
| naps |                                                 | napse]{.std |
| es** |                                                 | .std-ref}]  |
|      |                                                 | (Landing.ht |
|      |                                                 | ml#schema-b |
|      |                                                 | asesynapse) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***State Variables***

:   **v**: [[voltage]{.std
    .std-ref}](Landing.html#schema-dimensions-voltage){.reference
    .internal}  (exposed as **v**)

:   **w**: Dimensionless  (exposed as **w**)

:   **lastSpikeTime**: [[time]{.std
    .std-ref}](Landing.html#schema-dimensions-time){.reference
    .internal}

***On Start***

:   **v** = v_init \* MVOLT

:   **w** = 0

***Derived Variables***

:   **iSyn** = synapses\[\*\]-\>i(reduce method: add) (exposed as
    **iSyn**)

***Conditional Derived Variables***

:   IF delta_T \> 0 THEN

:     **delta_I** = delta_T \* exp(((v / MVOLT) - v_thresh) / delta_T)

:   IF delta_T = 0 THEN

:     **delta_I** = 0

***Regime: refractory (initial)***

:   ***On Entry***

:      **lastSpikeTime** = t

:      **v** = v_reset \* MVOLT

:      **w** = w+b

:   ***On Conditions***

:      IF t \> lastSpikeTime + (tau_refrac\*MSEC) THEN

:       TRANSITION to REGIME **integrating**

:   ***Time Derivatives***

:      d **w** /dt = (1 / tau_w) \* (a \* ((v / MVOLT) - v_rest) - w) /
    MSEC

***Regime: integrating (initial)***

:   ***On Conditions***

:      IF v \> eif_threshold \* MVOLT THEN

:       EVENT OUT on port: **spike**

:       TRANSITION to REGIME **refractory**

:   ***Time Derivatives***

:      d **v** /dt = (MVOLT \* ((-1 \* ((v / MVOLT) - v_rest) + delta_I)
    / tau_m + (i_offset - w) / cm) / MSEC) + (iSyn / (cm \* NFARAD))

:      d **w** /dt = (1 / tau_w) \* (a \* ((v / MVOLT) - v_rest) - w) /
    MSEC
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="EIF_cond_exp_isfa_ista">
      <xs:complexContent>
        <xs:extension base="basePyNNIaFCondCell">
          <xs:attribute name="a" type="xs:float" use="required"/>
          <xs:attribute name="b" type="xs:float" use="required"/>
          <xs:attribute name="delta_T" type="xs:float" use="required"/>
          <xs:attribute name="tau_w" type="xs:float" use="required"/>
          <xs:attribute name="v_spike" type="xs:float" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=EIF_cond_exp_isfa_ista){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import EIF_cond_exp_isfa_ista
    from neuroml.utils import component_factory

    variable = component_factory(
        EIF_cond_exp_isfa_ista,
        id: 'a NmlId (required)' = None,
        metaid: 'a MetaId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        neuro_lex_id: 'a NeuroLexId (optional)' = None,
        cm: 'a float (required)' = None,
        i_offset: 'a float (required)' = None,
        tau_syn_E: 'a float (required)' = None,
        tau_syn_I: 'a float (required)' = None,
        v_init: 'a float (required)' = None,
        tau_m: 'a float (required)' = None,
        tau_refrac: 'a float (required)' = None,
        v_reset: 'a float (required)' = None,
        v_rest: 'a float (required)' = None,
        v_thresh: 'a float (required)' = None,
        e_rev_E: 'a float (required)' = None,
        e_rev_I: 'a float (required)' = None,
        a: 'a float (required)' = None,
        b: 'a float (required)' = None,
        delta_T: 'a float (required)' = None,
        tau_w: 'a float (required)' = None,
        v_spike: 'a float (required)' = None,
        extensiontype_=None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <EIF_cond_exp_isfa_ista id="EIF_cond_exp_isfa_ista" a="0.0" b="0.0805" cm="0.281" delta_T="2.0" e_rev_E="0.0" e_rev_I="-80.0" i_offset="0.6" tau_m="9.3667" tau_refrac="5" tau_syn_E="5.0" tau_syn_I="5.0" tau_w="144.0" v_init="-65" v_reset="-68.0" v_rest="-70.6" v_spike="-40.0" v_thresh="-52.0"/>
:::
:::
:::
:::
:::

::: {#eif-cond-alpha-isfa-ista .section}
[]{#schema-eif-cond-alpha-isfa-ista}

##### EIF_cond_alpha_isfa_ista[\#](#eif-cond-alpha-isfa-ista "Link to this heading"){.headerlink}

extends *[[basePyNNIaFCondCell]{.std
.std-ref}](#schema-basepynniafcondcell){.reference .internal}*

*Adaptive exponential integrate and fire neuron according to Brette R
and Gerstner W ( 2005 ) with alpha-function-shaped post-synaptic
conductance.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    |                                                 | Di          |
| *a** |                                                 | mensionless |
+------+-------------------------------------------------+-------------+
| *    |                                                 | Di          |
| *b** |                                                 | mensionless |
+------+-------------------------------------------------+-------------+
| **   | *(from [[basePyNNCell]{.std                     | Di          |
| cm** | .std-ref}](#schema-basepynncell){.reference     | mensionless |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| **d  |                                                 | Di          |
| elta |                                                 | mensionless |
| _T** |                                                 |             |
+------+-------------------------------------------------+-------------+
| **e  | This parameter is never used in the NeuroML2    | Di          |
| _rev | description of this cell! Any synapse producing | mensionless |
| _E** | a current can be placed on this cell *(from     |             |
|      | [[basePyNNIaFCondCell]{.std                     |             |
|      | .st                                             |             |
|      | d-ref}](#schema-basepynniafcondcell){.reference |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| **e  | This parameter is never used in the NeuroML2    | Di          |
| _rev | description of this cell! Any synapse producing | mensionless |
| _I** | a current can be placed on this cell *(from     |             |
|      | [[basePyNNIaFCondCell]{.std                     |             |
|      | .st                                             |             |
|      | d-ref}](#schema-basepynniafcondcell){.reference |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| **i_ | *(from [[basePyNNCell]{.std                     | Di          |
| offs | .std-ref}](#schema-basepynncell){.reference     | mensionless |
| et** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| *    | *(from [[basePyNNIaFCell]{.std                  | Di          |
| *tau | .std-ref}](#schema-basepynniafcell){.reference  | mensionless |
| _m** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| **   | *(from [[basePyNNIaFCell]{.std                  | Di          |
| tau_ | .std-ref}](#schema-basepynniafcell){.reference  | mensionless |
| refr | .internal})*                                    |             |
| ac** |                                                 |             |
+------+-------------------------------------------------+-------------+
| *    | This parameter is never used in the NeuroML2    | Di          |
| *tau | description of this cell! Any synapse producing | mensionless |
| _syn | a current can be placed on this cell *(from     |             |
| _E** | [[basePyNNCell]{.std                            |             |
|      | .std-ref}](#schema-basepynncell){.reference     |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| *    | This parameter is never used in the NeuroML2    | Di          |
| *tau | description of this cell! Any synapse producing | mensionless |
| _syn | a current can be placed on this cell *(from     |             |
| _I** | [[basePyNNCell]{.std                            |             |
|      | .std-ref}](#schema-basepynncell){.reference     |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| *    |                                                 | Di          |
| *tau |                                                 | mensionless |
| _w** |                                                 |             |
+------+-------------------------------------------------+-------------+
| **   | *(from [[basePyNNCell]{.std                     | Di          |
| v_in | .std-ref}](#schema-basepynncell){.reference     | mensionless |
| it** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| **v  | *(from [[basePyNNIaFCell]{.std                  | Di          |
| _res | .std-ref}](#schema-basepynniafcell){.reference  | mensionless |
| et** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| **   | *(from [[basePyNNIaFCell]{.std                  | Di          |
| v_re | .std-ref}](#schema-basepynniafcell){.reference  | mensionless |
| st** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| **v  |                                                 | Di          |
| _spi |                                                 | mensionless |
| ke** |                                                 |             |
+------+-------------------------------------------------+-------------+
| **v_ | *(from [[basePyNNIaFCell]{.std                  | Di          |
| thre | .std-ref}](#schema-basepynniafcell){.reference  | mensionless |
| sh** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Derived parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    |                                                 | Di          |
| *eif |                                                 | mensionless |
| _thr |                                                 |             |
| esho |                                                 |             |
| ld** |                                                 |             |
+------+-------------------------------------------------+-------------+
:::

   **eif_threshold** = v_spike \* H(delta_T-1e-12) + v_thresh \*
H(-1\*delta_T+1e-9)
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **iS | *(from [[basePyNNCell]{.std                     | [[cu        |
| yn** | .std-ref}](#schema-basepynncell){.reference     | rrent]{.std |
|      | .internal})*                                    | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | Membrane potential *(from                       | [[vo        |
| *v** | [[baseCellMembPot]{.std                         | ltage]{.std |
|      | .std-ref}](                                     | .std-r      |
|      | Landing.html#schema-basecellmembpot){.reference | ef}](Landin |
|      | .internal})*                                    | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    |                                                 | Di          |
| *w** |                                                 | mensionless |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Spike event *(from [[baseSpikingCell]{.std      | Direction:  |
| *spi | .std-ref}](                                     | out         |
| ke** | Landing.html#schema-basespikingcell){.reference |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| **   | *(from [[basePyNNCell]{.std                     | Direction:  |
| spik | .std-ref}](#schema-basepynncell){.reference     | in          |
| e_in | .internal})*                                    |             |
| _E** |                                                 |             |
+------+-------------------------------------------------+-------------+
| **   | *(from [[basePyNNCell]{.std                     | Direction:  |
| spik | .std-ref}](#schema-basepynncell){.reference     | in          |
| e_in | .internal})*                                    |             |
| _I** |                                                 |             |
+------+-------------------------------------------------+-------------+
:::
:::

Attachments

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **sy |                                                 | [[baseSy    |
| naps |                                                 | napse]{.std |
| es** |                                                 | .std-ref}]  |
|      |                                                 | (Landing.ht |
|      |                                                 | ml#schema-b |
|      |                                                 | asesynapse) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***State Variables***

:   **v**: [[voltage]{.std
    .std-ref}](Landing.html#schema-dimensions-voltage){.reference
    .internal}  (exposed as **v**)

:   **w**: Dimensionless  (exposed as **w**)

:   **lastSpikeTime**: [[time]{.std
    .std-ref}](Landing.html#schema-dimensions-time){.reference
    .internal}

***On Start***

:   **v** = v_init \* MVOLT

:   **w** = 0

***Derived Variables***

:   **iSyn** = synapses\[\*\]-\>i(reduce method: add) (exposed as
    **iSyn**)

***Conditional Derived Variables***

:   IF delta_T \> 0 THEN

:     **delta_I** = delta_T \* exp(((v / MVOLT) - v_thresh) / delta_T)

:   IF delta_T = 0 THEN

:     **delta_I** = 0

***Regime: refractory (initial)***

:   ***On Entry***

:      **lastSpikeTime** = t

:      **v** = v_reset \* MVOLT

:      **w** = w + b

:   ***On Conditions***

:      IF t \> lastSpikeTime + (tau_refrac \* MSEC) THEN

:       TRANSITION to REGIME **integrating**

:   ***Time Derivatives***

:      d **w** /dt = (1 / tau_w) \* (a \* ((v / MVOLT) - v_rest) - w) /
    MSEC

***Regime: integrating (initial)***

:   ***On Conditions***

:      IF v \> eif_threshold \* MVOLT THEN

:       EVENT OUT on port: **spike**

:       TRANSITION to REGIME **refractory**

:   ***Time Derivatives***

:      d **v** /dt = (MVOLT \* ((-1 \* ( (v / MVOLT) - v_rest) +
    delta_I) / tau_m + (i_offset - w) / cm) / MSEC) + (iSyn / (cm \*
    NFARAD))

:      d **w** /dt = (1/ tau_w) \* (a\*((v/MVOLT)-v_rest) - w) /MSEC
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="EIF_cond_alpha_isfa_ista">
      <xs:complexContent>
        <xs:extension base="EIF_cond_exp_isfa_ista">
                </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=EIF_cond_alpha_isfa_ista){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import EIF_cond_alpha_isfa_ista
    from neuroml.utils import component_factory

    variable = component_factory(
        EIF_cond_alpha_isfa_ista,
        id: 'a NmlId (required)' = None,
        metaid: 'a MetaId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        neuro_lex_id: 'a NeuroLexId (optional)' = None,
        cm: 'a float (required)' = None,
        i_offset: 'a float (required)' = None,
        tau_syn_E: 'a float (required)' = None,
        tau_syn_I: 'a float (required)' = None,
        v_init: 'a float (required)' = None,
        tau_m: 'a float (required)' = None,
        tau_refrac: 'a float (required)' = None,
        v_reset: 'a float (required)' = None,
        v_rest: 'a float (required)' = None,
        v_thresh: 'a float (required)' = None,
        e_rev_E: 'a float (required)' = None,
        e_rev_I: 'a float (required)' = None,
        a: 'a float (required)' = None,
        b: 'a float (required)' = None,
        delta_T: 'a float (required)' = None,
        tau_w: 'a float (required)' = None,
        v_spike: 'a float (required)' = None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <EIF_cond_alpha_isfa_ista id="EIF_cond_alpha_isfa_ista" a="0.0" b="0.0805" cm="0.281" delta_T="0" e_rev_E="0.0" e_rev_I="-80.0" i_offset="0.6" tau_m="9.3667" tau_refrac="5" tau_syn_E="5.0" tau_syn_I="5.0" tau_w="144.0" v_init="-65" v_reset="-68.0" v_rest="-70.6" v_spike="-40.0" v_thresh="-52.0"/>
:::
:::
:::
:::
:::

::: {#hh-cond-exp .section}
[]{#schema-hh-cond-exp}

##### HH_cond_exp[\#](#hh-cond-exp "Link to this heading"){.headerlink}

extends *[[basePyNNCell]{.std
.std-ref}](#schema-basepynncell){.reference .internal}*

*Single-compartment Hodgkin-Huxley-type neuron with transient sodium and
delayed-rectifier potassium currents using the ion channel models from
Traub.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   | *(from [[basePyNNCell]{.std                     | Di          |
| cm** | .std-ref}](#schema-basepynncell){.reference     | mensionless |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| **e  |                                                 | Di          |
| _rev |                                                 | mensionless |
| _E** |                                                 |             |
+------+-------------------------------------------------+-------------+
| **e  |                                                 | Di          |
| _rev |                                                 | mensionless |
| _I** |                                                 |             |
+------+-------------------------------------------------+-------------+
| **e  |                                                 | Di          |
| _rev |                                                 | mensionless |
| _K** |                                                 |             |
+------+-------------------------------------------------+-------------+
| **e_ |                                                 | Di          |
| rev_ |                                                 | mensionless |
| Na** |                                                 |             |
+------+-------------------------------------------------+-------------+
| **   |                                                 | Di          |
| e_re |                                                 | mensionless |
| v_le |                                                 |             |
| ak** |                                                 |             |
+------+-------------------------------------------------+-------------+
| **   |                                                 | Di          |
| g_le |                                                 | mensionless |
| ak** |                                                 |             |
+------+-------------------------------------------------+-------------+
| **   |                                                 | Di          |
| gbar |                                                 | mensionless |
| _K** |                                                 |             |
+------+-------------------------------------------------+-------------+
| **g  |                                                 | Di          |
| bar_ |                                                 | mensionless |
| Na** |                                                 |             |
+------+-------------------------------------------------+-------------+
| **i_ | *(from [[basePyNNCell]{.std                     | Di          |
| offs | .std-ref}](#schema-basepynncell){.reference     | mensionless |
| et** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| *    | This parameter is never used in the NeuroML2    | Di          |
| *tau | description of this cell! Any synapse producing | mensionless |
| _syn | a current can be placed on this cell *(from     |             |
| _E** | [[basePyNNCell]{.std                            |             |
|      | .std-ref}](#schema-basepynncell){.reference     |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| *    | This parameter is never used in the NeuroML2    | Di          |
| *tau | description of this cell! Any synapse producing | mensionless |
| _syn | a current can be placed on this cell *(from     |             |
| _I** | [[basePyNNCell]{.std                            |             |
|      | .std-ref}](#schema-basepynncell){.reference     |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| **   | *(from [[basePyNNCell]{.std                     | Di          |
| v_in | .std-ref}](#schema-basepynncell){.reference     | mensionless |
| it** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| **v_ |                                                 | Di          |
| offs |                                                 | mensionless |
| et** |                                                 |             |
+------+-------------------------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    |                                                 | Di          |
| *h** |                                                 | mensionless |
+------+-------------------------------------------------+-------------+
| **iS | *(from [[basePyNNCell]{.std                     | [[cu        |
| yn** | .std-ref}](#schema-basepynncell){.reference     | rrent]{.std |
|      | .internal})*                                    | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    |                                                 | Di          |
| *m** |                                                 | mensionless |
+------+-------------------------------------------------+-------------+
| *    |                                                 | Di          |
| *n** |                                                 | mensionless |
+------+-------------------------------------------------+-------------+
| *    | Membrane potential *(from                       | [[vo        |
| *v** | [[baseCellMembPot]{.std                         | ltage]{.std |
|      | .std-ref}](                                     | .std-r      |
|      | Landing.html#schema-basecellmembpot){.reference | ef}](Landin |
|      | .internal})*                                    | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | Spike event *(from [[baseSpikingCell]{.std      | Direction:  |
| *spi | .std-ref}](                                     | out         |
| ke** | Landing.html#schema-basespikingcell){.reference |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
| **   | *(from [[basePyNNCell]{.std                     | Direction:  |
| spik | .std-ref}](#schema-basepynncell){.reference     | in          |
| e_in | .internal})*                                    |             |
| _E** |                                                 |             |
+------+-------------------------------------------------+-------------+
| **   | *(from [[basePyNNCell]{.std                     | Direction:  |
| spik | .std-ref}](#schema-basepynncell){.reference     | in          |
| e_in | .internal})*                                    |             |
| _I** |                                                 |             |
+------+-------------------------------------------------+-------------+
:::
:::

Attachments

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **sy |                                                 | [[baseSy    |
| naps |                                                 | napse]{.std |
| es** |                                                 | .std-ref}]  |
|      |                                                 | (Landing.ht |
|      |                                                 | ml#schema-b |
|      |                                                 | asesynapse) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***State Variables***

:   **v**: [[voltage]{.std
    .std-ref}](Landing.html#schema-dimensions-voltage){.reference
    .internal}  (exposed as **v**)

:   **m**: Dimensionless  (exposed as **m**)

:   **h**: Dimensionless  (exposed as **h**)

:   **n**: Dimensionless  (exposed as **n**)

***On Start***

:   **v** = v_init \* MVOLT

***Derived Variables***

:   **iSyn** = synapses\[\*\]-\>i(reduce method: add) (exposed as
    **iSyn**)

:   **iLeak** = g_leak \* (e_rev_leak - (v / MVOLT))

:   **iNa** = gbar_Na \* (m \* m \* m) \* h \* (e_rev_Na - (v / MVOLT))

:   **iK** = gbar_K \* (n \* n \* n \* n) \* (e_rev_K - (v / MVOLT))

:   **iMemb** = iLeak + iNa + iK + i_offset

:   **alpham** = 0.32 \* (13 - (v / MVOLT) + v_offset) / (exp((13 - (v /
    MVOLT) + v_offset) / 4.0) - 1)

:   **betam** = 0.28 \* ((v / MVOLT) - v_offset - 40) / (exp(((v /
    MVOLT) - v_offset - 40) / 5.0) - 1)

:   **alphah** = 0.128 \* exp((17 - (v / MVOLT) + v_offset) / 18.0)

:   **betah** = 4.0 / (1 + exp((40 - (v / MVOLT) + v_offset) / 5))

:   **alphan** = 0.032 \* (15 - (v / MVOLT) + v_offset) / (exp((15 - (v
    / MVOLT) + v_offset) / 5) - 1)

:   **betan** = 0.5 \* exp((10 - (v / MVOLT) + v_offset) / 40)

***Time Derivatives***

:   d **v** /dt = (MVOLT \* (iMemb / cm) / MSEC) + (iSyn / (cm \*
    NFARAD))

:   d **m** /dt = (alpham \* (1 - m) - betam \* m) / MSEC

:   d **h** /dt = (alphah \* (1 - h) - betah \* h) / MSEC

:   d **n** /dt = (alphan \* (1 - n) - betan \* n) / MSEC
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="HH_cond_exp">
      <xs:complexContent>
        <xs:extension base="basePyNNCell">
          <xs:attribute name="v_offset" type="xs:float" use="required"/>
          <xs:attribute name="e_rev_E" type="xs:float" use="required"/>
          <xs:attribute name="e_rev_I" type="xs:float" use="required"/>
          <xs:attribute name="e_rev_K" type="xs:float" use="required"/>
          <xs:attribute name="e_rev_Na" type="xs:float" use="required"/>
          <xs:attribute name="e_rev_leak" type="xs:float" use="required"/>
          <xs:attribute name="g_leak" type="xs:float" use="required"/>
          <xs:attribute name="gbar_K" type="xs:float" use="required"/>
          <xs:attribute name="gbar_Na" type="xs:float" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=HH_cond_exp){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import HH_cond_exp
    from neuroml.utils import component_factory

    variable = component_factory(
        HH_cond_exp,
        id: 'a NmlId (required)' = None,
        metaid: 'a MetaId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        neuro_lex_id: 'a NeuroLexId (optional)' = None,
        cm: 'a float (required)' = None,
        i_offset: 'a float (required)' = None,
        tau_syn_E: 'a float (required)' = None,
        tau_syn_I: 'a float (required)' = None,
        v_init: 'a float (required)' = None,
        v_offset: 'a float (required)' = None,
        e_rev_E: 'a float (required)' = None,
        e_rev_I: 'a float (required)' = None,
        e_rev_K: 'a float (required)' = None,
        e_rev_Na: 'a float (required)' = None,
        e_rev_leak: 'a float (required)' = None,
        g_leak: 'a float (required)' = None,
        gbar_K: 'a float (required)' = None,
        gbar_Na: 'a float (required)' = None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <HH_cond_exp id="HH_cond_exp" cm="0.2" e_rev_E="0.0" e_rev_I="-80.0" e_rev_K="-90.0" e_rev_Na="50.0" e_rev_leak="-65.0" g_leak="0.01" gbar_K="6.0" gbar_Na="20.0" i_offset="0.2" tau_syn_E="0.2" tau_syn_I="2.0" v_init="-65" v_offset="-63.0"/>
:::
:::
:::
:::
:::

::: {#basepynnsynapse .section}
[]{#schema-basepynnsynapse}

##### *basePynnSynapse*[\#](#basepynnsynapse "Link to this heading"){.headerlink}

extends *[[baseVoltageDepSynapse]{.std
.std-ref}](Landing.html#schema-basevoltagedepsynapse){.reference
.internal}*

*Base type for all PyNN synapses. Note, the current **I** produced is
dimensionless, but it requires a membrane potential **v** with dimension
voltage.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **t  |                                                 | Di          |
| au_s |                                                 | mensionless |
| yn** |                                                 |             |
+------+-------------------------------------------------+-------------+
:::
:::

Constants

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------------------+-----------------------------------+-------------+
| **MSEC** = 1ms     |                                   | [           |
|                    |                                   | [time]{.std |
|                    |                                   | .st         |
|                    |                                   | d-ref}](Lan |
|                    |                                   | ding.html#s |
|                    |                                   | chema-dimen |
|                    |                                   | sions-time) |
|                    |                                   | {.reference |
|                    |                                   | .internal}  |
+--------------------+-----------------------------------+-------------+
| **MVOLT** = 1mV    |                                   | [[vo        |
|                    |                                   | ltage]{.std |
|                    |                                   | .std-r      |
|                    |                                   | ef}](Landin |
|                    |                                   | g.html#sche |
|                    |                                   | ma-dimensio |
|                    |                                   | ns-voltage) |
|                    |                                   | {.reference |
|                    |                                   | .internal}  |
+--------------------+-----------------------------------+-------------+
| **NAMP** = 1nA     |                                   | [[cu        |
|                    |                                   | rrent]{.std |
|                    |                                   | .std-r      |
|                    |                                   | ef}](Landin |
|                    |                                   | g.html#sche |
|                    |                                   | ma-dimensio |
|                    |                                   | ns-current) |
|                    |                                   | {.reference |
|                    |                                   | .internal}  |
+--------------------+-----------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | The total (usually time varying) current        | [[cu        |
| *i** | produced by this ComponentType *(from           | rrent]{.std |
|      | [[basePointCurrent]{.std                        | .std-r      |
|      | .std-ref}](L                                    | ef}](Landin |
|      | anding.html#schema-basepointcurrent){.reference | g.html#sche |
|      | .internal})*                                    | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | The current may vary with the voltage exposed   | [[vo        |
| *v** | by the ComponentType on which this is placed    | ltage]{.std |
|      | *(from [[baseVoltageDepSynapse]{.std            | .std-r      |
|      | .std-ref}](Landin                               | ef}](Landin |
|      | g.html#schema-basevoltagedepsynapse){.reference | g.html#sche |
|      | .internal})*                                    | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   | *(from [[baseSynapse]{.std                      | Direction:  |
| in** | .std-re                                         | in          |
|      | f}](Landing.html#schema-basesynapse){.reference |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="BasePynnSynapse">
      <xs:complexContent>
        <xs:extension base="BaseSynapse">
          <xs:attribute name="tau_syn" type="xs:float" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=BasePynnSynapse){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import BasePynnSynapse
    from neuroml.utils import component_factory

    variable = component_factory(
        BasePynnSynapse,
        id: 'a NmlId (required)' = None,
        metaid: 'a MetaId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        neuro_lex_id: 'a NeuroLexId (optional)' = None,
        tau_syn: 'a float (required)' = None,
        extensiontype_=None,
    )
:::
:::
:::
:::
:::

::: {#expcondsynapse .section}
[]{#schema-expcondsynapse}

##### expCondSynapse[\#](#expcondsynapse "Link to this heading"){.headerlink}

extends *[[basePynnSynapse]{.std
.std-ref}](#schema-basepynnsynapse){.reference .internal}*

*Conductance based synapse with instantaneous rise and single
exponential decay ( with time constant tau_syn ).*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    |                                                 | Di          |
| *e_r |                                                 | mensionless |
| ev** |                                                 |             |
+------+-------------------------------------------------+-------------+
| **t  | *(from [[basePynnSynapse]{.std                  | Di          |
| au_s | .std-ref}](#schema-basepynnsynapse){.reference  | mensionless |
| yn** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Properties

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------------------+-----------------------------------+-------------+
| **weight**         |                                   | Di          |
| (default: 1)       |                                   | mensionless |
+--------------------+-----------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    |                                                 | Di          |
| *g** |                                                 | mensionless |
+------+-------------------------------------------------+-------------+
| *    | The total (usually time varying) current        | [[cu        |
| *i** | produced by this ComponentType *(from           | rrent]{.std |
|      | [[basePointCurrent]{.std                        | .std-r      |
|      | .std-ref}](L                                    | ef}](Landin |
|      | anding.html#schema-basepointcurrent){.reference | g.html#sche |
|      | .internal})*                                    | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | The current may vary with the voltage exposed   | [[vo        |
| *v** | by the ComponentType on which this is placed    | ltage]{.std |
|      | *(from [[baseVoltageDepSynapse]{.std            | .std-r      |
|      | .std-ref}](Landin                               | ef}](Landin |
|      | g.html#schema-basevoltagedepsynapse){.reference | g.html#sche |
|      | .internal})*                                    | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   | *(from [[baseSynapse]{.std                      | Direction:  |
| in** | .std-re                                         | in          |
|      | f}](Landing.html#schema-basesynapse){.reference |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***State Variables***

:   **g**: Dimensionless  (exposed as **g**)

***On Events***

:   EVENT IN on port: **in**

:      **g** = g+weight

***Derived Variables***

:   **i** = g \* (e_rev - (v/MVOLT)) \* NAMP (exposed as **i**)

***Time Derivatives***

:   d **g** /dt = -g / (tau_syn\*MSEC)
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="ExpCondSynapse">
      <xs:complexContent>
        <xs:extension base="BasePynnSynapse">
          <xs:attribute name="e_rev" type="xs:float" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=ExpCondSynapse){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import ExpCondSynapse
    from neuroml.utils import component_factory

    variable = component_factory(
        ExpCondSynapse,
        id: 'a NmlId (required)' = None,
        metaid: 'a MetaId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        neuro_lex_id: 'a NeuroLexId (optional)' = None,
        tau_syn: 'a float (required)' = None,
        e_rev: 'a float (required)' = None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <expCondSynapse id="syn1" tau_syn="5" e_rev="0"/>
:::
:::
:::
:::
:::

::: {#expcurrsynapse .section}
[]{#schema-expcurrsynapse}

##### expCurrSynapse[\#](#expcurrsynapse "Link to this heading"){.headerlink}

extends *[[basePynnSynapse]{.std
.std-ref}](#schema-basepynnsynapse){.reference .internal}*

*Current based synapse with instantaneous rise and single exponential
decay ( with time constant tau_syn ).*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **t  | *(from [[basePynnSynapse]{.std                  | Di          |
| au_s | .std-ref}](#schema-basepynnsynapse){.reference  | mensionless |
| yn** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Properties

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------------------+-----------------------------------+-------------+
| **weight**         |                                   | Di          |
| (default: 1)       |                                   | mensionless |
+--------------------+-----------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | The total (usually time varying) current        | [[cu        |
| *i** | produced by this ComponentType *(from           | rrent]{.std |
|      | [[basePointCurrent]{.std                        | .std-r      |
|      | .std-ref}](L                                    | ef}](Landin |
|      | anding.html#schema-basepointcurrent){.reference | g.html#sche |
|      | .internal})*                                    | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | The current may vary with the voltage exposed   | [[vo        |
| *v** | by the ComponentType on which this is placed    | ltage]{.std |
|      | *(from [[baseVoltageDepSynapse]{.std            | .std-r      |
|      | .std-ref}](Landin                               | ef}](Landin |
|      | g.html#schema-basevoltagedepsynapse){.reference | g.html#sche |
|      | .internal})*                                    | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   | *(from [[baseSynapse]{.std                      | Direction:  |
| in** | .std-re                                         | in          |
|      | f}](Landing.html#schema-basesynapse){.reference |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***State Variables***

:   **I**: Dimensionless

***On Events***

:   EVENT IN on port: **in**

:      **I** = I + weight

***Derived Variables***

:   **i** = I \* NAMP (exposed as **i**)

***Time Derivatives***

:   d **I** /dt = -I / (tau_syn\*MSEC)
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="ExpCurrSynapse">
      <xs:complexContent>
        <xs:extension base="BasePynnSynapse">

                </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=ExpCurrSynapse){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import ExpCurrSynapse
    from neuroml.utils import component_factory

    variable = component_factory(
        ExpCurrSynapse,
        id: 'a NmlId (required)' = None,
        metaid: 'a MetaId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        neuro_lex_id: 'a NeuroLexId (optional)' = None,
        tau_syn: 'a float (required)' = None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <expCurrSynapse id="syn3" tau_syn="5"/>
:::
:::
:::
:::
:::

::: {#alphacondsynapse .section}
[]{#schema-alphacondsynapse}

##### alphaCondSynapse[\#](#alphacondsynapse "Link to this heading"){.headerlink}

extends *[[basePynnSynapse]{.std
.std-ref}](#schema-basepynnsynapse){.reference .internal}*

*Alpha synapse: rise time and decay time are both tau_syn. Conductance
based synapse.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    |                                                 | Di          |
| *e_r |                                                 | mensionless |
| ev** |                                                 |             |
+------+-------------------------------------------------+-------------+
| **t  | *(from [[basePynnSynapse]{.std                  | Di          |
| au_s | .std-ref}](#schema-basepynnsynapse){.reference  | mensionless |
| yn** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Properties

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------------------+-----------------------------------+-------------+
| **weight**         |                                   | Di          |
| (default: 1)       |                                   | mensionless |
+--------------------+-----------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    |                                                 | Di          |
| *A** |                                                 | mensionless |
+------+-------------------------------------------------+-------------+
| *    |                                                 | Di          |
| *g** |                                                 | mensionless |
+------+-------------------------------------------------+-------------+
| *    | The total (usually time varying) current        | [[cu        |
| *i** | produced by this ComponentType *(from           | rrent]{.std |
|      | [[basePointCurrent]{.std                        | .std-r      |
|      | .std-ref}](L                                    | ef}](Landin |
|      | anding.html#schema-basepointcurrent){.reference | g.html#sche |
|      | .internal})*                                    | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | The current may vary with the voltage exposed   | [[vo        |
| *v** | by the ComponentType on which this is placed    | ltage]{.std |
|      | *(from [[baseVoltageDepSynapse]{.std            | .std-r      |
|      | .std-ref}](Landin                               | ef}](Landin |
|      | g.html#schema-basevoltagedepsynapse){.reference | g.html#sche |
|      | .internal})*                                    | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   | *(from [[baseSynapse]{.std                      | Direction:  |
| in** | .std-re                                         | in          |
|      | f}](Landing.html#schema-basesynapse){.reference |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***State Variables***

:   **g**: Dimensionless  (exposed as **g**)

:   **A**: Dimensionless  (exposed as **A**)

***On Events***

:   EVENT IN on port: **in**

:      **A** = A + weight

***Derived Variables***

:   **i** = g \* (e_rev - (v/MVOLT)) \* NAMP (exposed as **i**)

***Time Derivatives***

:   d **g** /dt = (2.7182818*A - g)/(tau_syn*MSEC)

:   d **A** /dt = -A /(tau_syn\*MSEC)
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="AlphaCondSynapse">
      <xs:complexContent>
        <xs:extension base="BasePynnSynapse">
          <xs:attribute name="e_rev" type="xs:float" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=AlphaCondSynapse){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import AlphaCondSynapse
    from neuroml.utils import component_factory

    variable = component_factory(
        AlphaCondSynapse,
        id: 'a NmlId (required)' = None,
        metaid: 'a MetaId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        neuro_lex_id: 'a NeuroLexId (optional)' = None,
        tau_syn: 'a float (required)' = None,
        e_rev: 'a float (required)' = None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <alphaCondSynapse id="syn2" tau_syn="5" e_rev="0"/>
:::
:::
:::
:::
:::

::: {#alphacurrsynapse .section}
[]{#schema-alphacurrsynapse}

##### alphaCurrSynapse[\#](#alphacurrsynapse "Link to this heading"){.headerlink}

extends *[[basePynnSynapse]{.std
.std-ref}](#schema-basepynnsynapse){.reference .internal}*

*Alpha synapse: rise time and decay time are both tau_syn. Current based
synapse.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **t  | *(from [[basePynnSynapse]{.std                  | Di          |
| au_s | .std-ref}](#schema-basepynnsynapse){.reference  | mensionless |
| yn** | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Properties

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------------------+-----------------------------------+-------------+
| **weight**         |                                   | Di          |
| (default: 1)       |                                   | mensionless |
+--------------------+-----------------------------------+-------------+
:::
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    |                                                 | [[cu        |
| *A** |                                                 | rrent]{.std |
|      |                                                 | .std-r      |
|      |                                                 | ef}](Landin |
|      |                                                 | g.html#sche |
|      |                                                 | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | The total (usually time varying) current        | [[cu        |
| *i** | produced by this ComponentType *(from           | rrent]{.std |
|      | [[basePointCurrent]{.std                        | .std-r      |
|      | .std-ref}](L                                    | ef}](Landin |
|      | anding.html#schema-basepointcurrent){.reference | g.html#sche |
|      | .internal})*                                    | ma-dimensio |
|      |                                                 | ns-current) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Requirements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | The current may vary with the voltage exposed   | [[vo        |
| *v** | by the ComponentType on which this is placed    | ltage]{.std |
|      | *(from [[baseVoltageDepSynapse]{.std            | .std-r      |
|      | .std-ref}](Landin                               | ef}](Landin |
|      | g.html#schema-basevoltagedepsynapse){.reference | g.html#sche |
|      | .internal})*                                    | ma-dimensio |
|      |                                                 | ns-voltage) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   | *(from [[baseSynapse]{.std                      | Direction:  |
| in** | .std-re                                         | in          |
|      | f}](Landing.html#schema-basesynapse){.reference |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***State Variables***

:   **I**: Dimensionless

:   **A**: Dimensionless  (exposed as **A**)

***On Events***

:   EVENT IN on port: **in**

:      **A** = A + weight

***Derived Variables***

:   **i** = I \* NAMP (exposed as **i**)

***Time Derivatives***

:   d **I** /dt = (2.7182818*A - I)/(tau_syn*MSEC)

:   d **A** /dt = -A /(tau_syn\*MSEC)
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="AlphaCurrSynapse">
      <xs:complexContent>
        <xs:extension base="BasePynnSynapse">

                </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=AlphaCurrSynapse){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import AlphaCurrSynapse
    from neuroml.utils import component_factory

    variable = component_factory(
        AlphaCurrSynapse,
        id: 'a NmlId (required)' = None,
        metaid: 'a MetaId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        neuro_lex_id: 'a NeuroLexId (optional)' = None,
        tau_syn: 'a float (required)' = None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <alphaCurrSynapse id="syn4" tau_syn="5"/>
:::
:::
:::
:::
:::

::: {#spikesourcepoisson .section}
[]{#schema-spikesourcepoisson}

##### SpikeSourcePoisson[\#](#spikesourcepoisson "Link to this heading"){.headerlink}

extends *[[baseSpikeSource]{.std
.std-ref}](Landing.html#schema-basespikesource){.reference .internal}*

*Spike source, generating spikes according to a Poisson process.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **du |                                                 | [           |
| rati |                                                 | [time]{.std |
| on** |                                                 | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **ra |                                                 | [[per       |
| te** |                                                 | _time]{.std |
|      |                                                 | .std-re     |
|      |                                                 | f}](Landing |
|      |                                                 | .html#schem |
|      |                                                 | a-dimension |
|      |                                                 | s-per-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    |                                                 | [           |
| *sta |                                                 | [time]{.std |
| rt** |                                                 | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Constants

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------------------+-----------------------------------+-------------+
| **LONG_TIME** =    |                                   | [           |
| 1e9hour            |                                   | [time]{.std |
|                    |                                   | .st         |
|                    |                                   | d-ref}](Lan |
|                    |                                   | ding.html#s |
|                    |                                   | chema-dimen |
|                    |                                   | sions-time) |
|                    |                                   | {.reference |
|                    |                                   | .internal}  |
+--------------------+-----------------------------------+-------------+
| **SMALL_TIME** =   |                                   | [           |
| 1e-9ms             |                                   | [time]{.std |
|                    |                                   | .st         |
|                    |                                   | d-ref}](Lan |
|                    |                                   | ding.html#s |
|                    |                                   | chema-dimen |
|                    |                                   | sions-time) |
|                    |                                   | {.reference |
|                    |                                   | .internal}  |
+--------------------+-----------------------------------+-------------+
:::
:::

Derived parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **e  |                                                 | [           |
| nd** |                                                 | [time]{.std |
|      |                                                 | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::

   **end** = start + duration
:::

Exposures

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **i  |                                                 | [           |
| si** |                                                 | [time]{.std |
|      |                                                 | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   |                                                 | [           |
| tnex |                                                 | [time]{.std |
| tIde |                                                 | .st         |
| al** |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    |                                                 | [           |
| *tne |                                                 | [time]{.std |
| xtUs |                                                 | .st         |
| ed** |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   | Time since the last spike was emitted *(from    | [           |
| tsin | [[baseSpikeSource]{.std                         | [time]{.std |
| ce** | .std-ref}](                                     | .st         |
|      | Landing.html#schema-basespikesource){.reference | d-ref}](Lan |
|      | .internal})*                                    | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Event Ports

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   |                                                 | Direction:  |
| in** |                                                 | in          |
+------+-------------------------------------------------+-------------+
| *    | Port on which spikes are emitted *(from         | Direction:  |
| *spi | [[baseSpikeSource]{.std                         | out         |
| ke** | .std-ref}](                                     |             |
|      | Landing.html#schema-basespikesource){.reference |             |
|      | .internal})*                                    |             |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***State Variables***

:   **tsince**: [[time]{.std
    .std-ref}](Landing.html#schema-dimensions-time){.reference
    .internal}  (exposed as **tsince**)

:   **tnextIdeal**: [[time]{.std
    .std-ref}](Landing.html#schema-dimensions-time){.reference
    .internal}  (exposed as **tnextIdeal**)

:   **tnextUsed**: [[time]{.std
    .std-ref}](Landing.html#schema-dimensions-time){.reference
    .internal}  (exposed as **tnextUsed**)

:   **isi**: [[time]{.std
    .std-ref}](Landing.html#schema-dimensions-time){.reference
    .internal}  (exposed as **isi**)

***On Start***

:   **isi** = start - log(random(1))/rate

:   **tsince** = 0

:   **tnextIdeal** = isi + H(((isi) -
    (start+duration))/duration)\*LONG_TIME

:   **tnextUsed** = tnextIdeal

***On Conditions***

:   IF t \> tnextUsed THEN

:      **isi** = -1 \* log(random(1))/rate

:      **tnextIdeal** = (tnextIdeal+isi) + H(((tnextIdeal+isi) -
    (start+duration))/duration)\*LONG_TIME

:      **tnextUsed** = tnextIdeal\*H( (tnextIdeal-t)/t ) +
    (t+SMALL_TIME)\*H( (t-tnextIdeal)/t )

:      **tsince** = 0

:      EVENT OUT on port: **spike**

***Time Derivatives***

:   d **tsince** /dt = 1

:   d **tnextUsed** /dt = 0

:   d **tnextIdeal** /dt = 0
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="SpikeSourcePoisson">
      <xs:complexContent>
        <xs:extension base="Standalone">
          <xs:attribute name="start" type="Nml2Quantity_time" use="required"/>
          <xs:attribute name="duration" type="Nml2Quantity_time" use="required"/>
          <xs:attribute name="rate" type="Nml2Quantity_pertime" use="required"/>
        </xs:extension>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: Python

::: {.sd-tab-content .docutils}
*[Go to the libNeuroML
documentation](https://libneuroml.readthedocs.io/en/latest/search.html?q=SpikeSourcePoisson){target="_blank"}*

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import SpikeSourcePoisson
    from neuroml.utils import component_factory

    variable = component_factory(
        SpikeSourcePoisson,
        id: 'a NonNegativeInteger (required)' = None,
        metaid: 'a MetaId (optional)' = None,
        notes: 'a string (optional)' = None,
        properties: 'list of Property(s) (optional)' = None,
        annotation: 'a Annotation (optional)' = None,
        start: 'a Nml2Quantity_time (required)' = None,
        duration: 'a Nml2Quantity_time (required)' = None,
        rate: 'a Nml2Quantity_pertime (required)' = None,
    )
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <SpikeSourcePoisson id="spikes1" start="50ms" duration="400ms" rate="50Hz"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <SpikeSourcePoisson id="spikes2" start="50ms" duration="300ms" rate="80Hz"/>
:::
:::
:::
:::
:::
:::

[]{#document-Userdocs/Schemas/Simulation}

::: {#simulation .section .tex2jax_ignore .mathjax_ignore}
[]{#schema-simulation}

#### Simulation[\#](#simulation "Link to this heading"){.headerlink}

**Specification of the LEMS Simulation element which is normally used to
define simulations of NeuroML2 files. Note: not actually part of NeuroML
v2, but this is required by much of the NeuroML toolchain for defining
Simulations ( which NeuroML model to use and how long to run for ), as
well as what to [[Display]{.std .std-ref}](#schema-display){.reference
.internal} and what to save in [[OutputFile]{.std
.std-ref}](#schema-outputfile){.reference .internal}s.**

------------------------------------------------------------------------

Original ComponentType definitions:
[Simulation.xml](https://github.com/NeuroML/NeuroML2/blob/master/NeuroML2CoreTypes//Simulation.xml){.reference
.external}. Schema against which NeuroML based on these should be valid:
[NeuroML_v2.3.xsd](https://github.com/NeuroML/NeuroML2/tree/master/Schemas/NeuroML2/NeuroML_v2.3.xsd){.reference
.external}. Generated on 14/08/24 from
[this](https://github.com/NeuroML/NeuroML2/commit/352244cff605cb1ba24fa7c11757dc818fe90fd2){.reference
.external} commit. Please file any issues or questions at the [issue
tracker here](https://github.com/NeuroML/NeuroML2/issues){.reference
.external}.

------------------------------------------------------------------------

::: {#id1 .section}
[]{#id2}

##### Simulation[\#](#id1 "Link to this heading"){.headerlink}

*The main element in a LEMS Simulation file. Defines the **length** of
simulation, the timestep ( dt ) **step** and an optional **seed** to use
for stochastic elements, as well as [[Display]{.std
.std-ref}](#schema-display){.reference .internal}s, [[OutputFile]{.std
.std-ref}](#schema-outputfile){.reference .internal}s and
[[EventOutputFile]{.std .std-ref}](#schema-eventoutputfile){.reference
.internal}s to record. Specifies a **target** component to run, usually
the id of a [[network]{.std
.std-ref}](Landing.html#schema-network){.reference .internal}.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   | Duration of the simulation run                  | [           |
| leng |                                                 | [time]{.std |
| th** |                                                 | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **st | Time step (dt) to use in the simulation         | [           |
| ep** |                                                 | [time]{.std |
|      |                                                 | .st         |
|      |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Text fields

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| **     | The seed to use in the random number generator for           |
| seed** | stochastic entities                                          |
+--------+--------------------------------------------------------------+
:::
:::

Component References

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   |                                                 | [s          |
| targ |                                                 | chema:compo |
| et** |                                                 | nent]{.xref |
|      |                                                 | .std        |
|      |                                                 | .std-ref}   |
+------+-------------------------------------------------+-------------+
:::
:::

Children list

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    |                                                 | [           |
| *met |                                                 | [Meta]{.std |
| as** |                                                 | .s          |
|      |                                                 | td-ref}](#s |
|      |                                                 | chema-meta) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **di |                                                 | [[Di        |
| spla |                                                 | splay]{.std |
| ys** |                                                 | .std-       |
|      |                                                 | ref}](#sche |
|      |                                                 | ma-display) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **o  |                                                 | [[Outpu     |
| utpu |                                                 | tFile]{.std |
| ts** |                                                 | .std-ref    |
|      |                                                 | }](#schema- |
|      |                                                 | outputfile) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **   |                                                 | [           |
| even |                                                 | [EventOutpu |
| ts** |                                                 | tFile]{.std |
|      |                                                 | .s          |
|      |                                                 | td-ref}](#s |
|      |                                                 | chema-event |
|      |                                                 | outputfile) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Dynamics

::: {.sd-tab-content .docutils}

***State Variables***

:   **t**: [[time]{.std
    .std-ref}](Landing.html#schema-dimensions-time){.reference
    .internal}
:::
:::
:::

::: {#display .section}
[]{#schema-display}

##### Display[\#](#display "Link to this heading"){.headerlink}

*Details of a display to generate ( usually a set of traces given by
[[Line]{.std .std-ref}](#schema-line){.reference .internal}s in a newly
opened window ) on completion of the simulation.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | A scaling of the time axis, e.g. 1ms means      | [           |
| *tim | display in milliseconds. Note: all quantities   | [time]{.std |
| eSca | are recorded in SI units                        | .st         |
| le** |                                                 | d-ref}](Lan |
|      |                                                 | ding.html#s |
|      |                                                 | chema-dimen |
|      |                                                 | sions-time) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| **xm | The maximum value on the x axis (i.e time       | Di          |
| ax** | variable) of the display                        | mensionless |
+------+-------------------------------------------------+-------------+
| **xm | The minimum value on the x axis (i.e time       | Di          |
| in** | variable) of the display                        | mensionless |
+------+-------------------------------------------------+-------------+
| **ym | The maximum value on the x axis of the display  | Di          |
| ax** |                                                 | mensionless |
+------+-------------------------------------------------+-------------+
| **ym | The minimum value on the y axis of the display  | Di          |
| in** |                                                 | mensionless |
+------+-------------------------------------------------+-------------+
:::
:::

Text fields

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| **t    | The title of the display, e.g. to use for the window         |
| itle** |                                                              |
+--------+--------------------------------------------------------------+
:::
:::

Children list

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    |                                                 | [           |
| *lin |                                                 | [Line]{.std |
| es** |                                                 | .s          |
|      |                                                 | td-ref}](#s |
|      |                                                 | chema-line) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::
:::
:::

::: {#line .section}
[]{#schema-line}

##### Line[\#](#line "Link to this heading"){.headerlink}

*Specification of a single time varying **quantity** to plot on the
[[Display]{.std .std-ref}](#schema-display){.reference .internal}. Note
that all quantities are handled internally in LEMS in SI units, and so a
**scale** should be used if it is to be displayed in other units.*

::: {.sd-tab-set .docutils}
Parameters

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| *    | A scaling factor to DIVIDE the quantity by. Can | [[Dimen     |
| *sca | be dimensional, so using scale=1mV means a      | sions]{.std |
| le** | value of -0.07V is displayed as -70.            | .std-ref}   |
|      | Alternatively, scale=0.001 would achieve the    | ](Landing.h |
|      | same thing.                                     | tml#schema- |
|      |                                                 | dimensions) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
| *    | An optional scaling of the time axis, e.g. 1ms  | [[Dimen     |
| *tim | means display in milliseconds. Note: if         | sions]{.std |
| eSca | present, this overrides timeScale from          | .std-ref}   |
| le** | *Display*                                       | ](Landing.h |
|      |                                                 | tml#schema- |
|      |                                                 | dimensions) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::

Text fields

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| **c    | A hex string for the color to display the trace for this     |
| olor** | quantity, e.g. #aa33ff                                       |
+--------+--------------------------------------------------------------+
:::
:::

Paths

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| **quan | Path to the quantity to display, see see                     |
| tity** | [https://docs.neuroml.org/Userdocs/Paths.ht                  |
|        | ml](https://docs.neuroml.org/Userdocs/Paths.html){.reference |
|        | .external}.                                                  |
+--------+--------------------------------------------------------------+
:::
:::
:::
:::

::: {#outputfile .section}
[]{#schema-outputfile}

##### OutputFile[\#](#outputfile "Link to this heading"){.headerlink}

*A file in which to save recorded values from the simulation.*

::: {.sd-tab-set .docutils}
Text fields

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| **     | Optional path to the directory in which to store the file    |
| path** |                                                              |
+--------+--------------------------------------------------------------+
| **file | Name of the file to generate. Can include a relative path    |
| Name** | (from the LEMS Simulation file location).                    |
+--------+--------------------------------------------------------------+
:::
:::

Children list

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **ou |                                                 | [[OutputC   |
| tput |                                                 | olumn]{.std |
| Colu |                                                 | .std-ref}]  |
| mn** |                                                 | (#schema-ou |
|      |                                                 | tputcolumn) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::
:::
:::

::: {#outputcolumn .section}
[]{#schema-outputcolumn}

##### OutputColumn[\#](#outputcolumn "Link to this heading"){.headerlink}

*Specification of a single time varying **quantity** to record during
the simulation. Note that all quantities are handled internally in LEMS
in SI units, and so the value for the quantity in the file ( as well as
time ) will be in SI units.*

::: {.sd-tab-set .docutils}
Paths

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| **quan | Path to the quantity to save, see see                        |
| tity** | [https://docs.neuroml.org/Userdocs/Paths.ht                  |
|        | ml](https://docs.neuroml.org/Userdocs/Paths.html){.reference |
|        | .external}. Note that all quantities are saved in SI units.  |
+--------+--------------------------------------------------------------+
:::
:::
:::
:::

::: {#eventoutputfile .section}
[]{#schema-eventoutputfile}

##### EventOutputFile[\#](#eventoutputfile "Link to this heading"){.headerlink}

*A file in which to save event information ( e.g. spikes from cells in a
population ) in a specified **format**.*

::: {.sd-tab-set .docutils}
Text fields

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| **     | Optional path to the directory in which to store the file    |
| path** |                                                              |
+--------+--------------------------------------------------------------+
| **file | Name of the file to generate. Can include a relative path    |
| Name** | (from the LEMS Simulation file location).                    |
+--------+--------------------------------------------------------------+
| **fo   | Takes values TIME_ID or ID_TIME, depending on the preferred  |
| rmat** | order of the time or event id (from *EventSelection*) in     |
|        | each row of the file                                         |
+--------+--------------------------------------------------------------+
:::
:::

Children list

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| **   |                                                 | [[EventSele |
| even |                                                 | ction]{.std |
| tSel |                                                 | .           |
| ecti |                                                 | std-ref}](# |
| on** |                                                 | schema-even |
|      |                                                 | tselection) |
|      |                                                 | {.reference |
|      |                                                 | .internal}  |
+------+-------------------------------------------------+-------------+
:::
:::
:::
:::

::: {#eventselection .section}
[]{#schema-eventselection}

##### EventSelection[\#](#eventselection "Link to this heading"){.headerlink}

*A specific source of events with an associated **id,** which will be
recorded inside the file specified in the parent [[EventOutputFile]{.std
.std-ref}](#schema-eventoutputfile){.reference .internal}. The attribute
**select** should point to a cell inside a [[population]{.std
.std-ref}](Landing.html#schema-population){.reference .internal} ( e.g.
hhpop\[0\], see
[https://docs.neuroml.org/Userdocs/Paths.html](https://docs.neuroml.org/Userdocs/Paths.html){.reference
.external} ), and the **eventPort** specifies the port for the emitted
events, which usually has id: spike. Note: the **id** used on this
element ( and appearing in the file alongside the event time ) can be
different from the id/index of the cell in the population.*

::: {.sd-tab-set .docutils}
Text fields

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| *      | The port on the cell which generates the events, usually:    |
| *event | spike                                                        |
| Port** |                                                              |
+--------+--------------------------------------------------------------+
:::
:::

Paths

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| **se   | The cell which will be emitting the events                   |
| lect** |                                                              |
+--------+--------------------------------------------------------------+
:::
:::
:::
:::

::: {#meta .section}
[]{#schema-meta}

##### Meta[\#](#meta "Link to this heading"){.headerlink}

*Metadata to add to simulation.*

::: {.sd-tab-set .docutils}
Text fields

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+--------+--------------------------------------------------------------+
| *      | Simulator name                                               |
| *for** |                                                              |
+--------+--------------------------------------------------------------+
| **me   | Integration method to use                                    |
| thod** |                                                              |
+--------+--------------------------------------------------------------+
| **abs  | Absolute tolerance for NEURON's cvode method                 |
| _toler |                                                              |
| ance** |                                                              |
+--------+--------------------------------------------------------------+
| **rel  | Relative tolerance for NEURON's cvode method                 |
| _toler |                                                              |
| ance** |                                                              |
+--------+--------------------------------------------------------------+
:::
:::
:::
:::
:::

[]{#document-Userdocs/Schemas/Index}

::: {#index .section .tex2jax_ignore .mathjax_ignore}
#### Index[\#](#index "Link to this heading"){.headerlink}

[]{#adexiafcell}

-   [[adExIaFCell]{.std
    .std-ref}](Landing.html#schema-adexiafcell){.reference .internal}
    []{#alphacondsynapse}

-   [[alphaCondSynapse]{.std
    .std-ref}](Landing.html#schema-alphacondsynapse){.reference
    .internal} []{#alphacurrentsynapse}

-   [[alphaCurrentSynapse]{.std
    .std-ref}](Landing.html#schema-alphacurrentsynapse){.reference
    .internal} []{#alphacurrsynapse}

-   [[alphaCurrSynapse]{.std
    .std-ref}](Landing.html#schema-alphacurrsynapse){.reference
    .internal} []{#alphasynapse}

-   [[alphaSynapse]{.std
    .std-ref}](Landing.html#schema-alphasynapse){.reference .internal}
    []{#annotation}

-   [[annotation]{.std
    .std-ref}](Landing.html#schema-annotation){.reference .internal}
    []{#area}

-   [[area]{.std
    .std-ref}](Landing.html#schema-dimensions-area){.reference
    .internal} []{#baseblockmechanism}

-   [[baseBlockMechanism]{.std
    .std-ref}](Landing.html#schema-baseblockmechanism){.reference
    .internal} []{#basebqbiol}

-   [[baseBqbiol]{.std
    .std-ref}](Landing.html#schema-basebqbiol){.reference .internal}
    []{#basecell}

-   [[baseCell]{.std .std-ref}](Landing.html#schema-basecell){.reference
    .internal} []{#basecellmembpot}

-   [[baseCellMembPot]{.std
    .std-ref}](Landing.html#schema-basecellmembpot){.reference
    .internal} []{#basecellmembpotcap}

-   [[baseCellMembPotCap]{.std
    .std-ref}](Landing.html#schema-basecellmembpotcap){.reference
    .internal} []{#basecellmembpotdl}

-   [[baseCellMembPotDL]{.std
    .std-ref}](Landing.html#schema-basecellmembpotdl){.reference
    .internal} []{#basechanneldensity}

-   [[baseChannelDensity]{.std
    .std-ref}](Landing.html#schema-basechanneldensity){.reference
    .internal} []{#basechanneldensitycond}

-   [[baseChannelDensityCond]{.std
    .std-ref}](Landing.html#schema-basechanneldensitycond){.reference
    .internal} []{#basechannelpopulation}

-   [[baseChannelPopulation]{.std
    .std-ref}](Landing.html#schema-basechannelpopulation){.reference
    .internal} []{#baseconductancebasedsynapse}

-   [[baseConductanceBasedSynapse]{.std
    .std-ref}](Landing.html#schema-baseconductancebasedsynapse){.reference
    .internal} []{#baseconductancebasedsynapsetwo}

-   [[baseConductanceBasedSynapseTwo]{.std
    .std-ref}](Landing.html#schema-baseconductancebasedsynapsetwo){.reference
    .internal} []{#baseconductancescaling}

-   [[baseConductanceScaling]{.std
    .std-ref}](Landing.html#schema-baseconductancescaling){.reference
    .internal} []{#baseconductancescalingcadependent}

-   [[baseConductanceScalingCaDependent]{.std
    .std-ref}](Landing.html#schema-baseconductancescalingcadependent){.reference
    .internal} []{#basecurrentbasedsynapse}

-   [[baseCurrentBasedSynapse]{.std
    .std-ref}](Landing.html#schema-basecurrentbasedsynapse){.reference
    .internal} []{#basegate}

-   [[baseGate]{.std .std-ref}](Landing.html#schema-basegate){.reference
    .internal} []{#basegradedsynapse}

-   [[baseGradedSynapse]{.std
    .std-ref}](Landing.html#schema-basegradedsynapse){.reference
    .internal} []{#basehhrate}

-   [[baseHHRate]{.std
    .std-ref}](Landing.html#schema-basehhrate){.reference .internal}
    []{#basehhvariable}

-   [[baseHHVariable]{.std
    .std-ref}](Landing.html#schema-basehhvariable){.reference .internal}
    []{#baseiaf}

-   [[baseIaf]{.std .std-ref}](Landing.html#schema-baseiaf){.reference
    .internal} []{#baseiafcapcell}

-   [[baseIafCapCell]{.std
    .std-ref}](Landing.html#schema-baseiafcapcell){.reference .internal}
    []{#baseionchannel}

-   [[baseIonChannel]{.std
    .std-ref}](Landing.html#schema-baseionchannel){.reference .internal}
    []{#baseplasticitymechanism}

-   [[basePlasticityMechanism]{.std
    .std-ref}](Landing.html#schema-baseplasticitymechanism){.reference
    .internal} []{#basepointcurrent}

-   [[basePointCurrent]{.std
    .std-ref}](Landing.html#schema-basepointcurrent){.reference
    .internal} []{#basepointcurrentdl}

-   [[basePointCurrentDL]{.std
    .std-ref}](Landing.html#schema-basepointcurrentdl){.reference
    .internal} []{#basepopulation}

-   [[basePopulation]{.std
    .std-ref}](Landing.html#schema-basepopulation){.reference .internal}
    []{#basepynncell}

-   [[basePyNNCell]{.std
    .std-ref}](Landing.html#schema-basepynncell){.reference .internal}
    []{#basepynniafcell}

-   [[basePyNNIaFCell]{.std
    .std-ref}](Landing.html#schema-basepynniafcell){.reference
    .internal} []{#basepynniafcondcell}

-   [[basePyNNIaFCondCell]{.std
    .std-ref}](Landing.html#schema-basepynniafcondcell){.reference
    .internal} []{#basepynnsynapse}

-   [[basePynnSynapse]{.std
    .std-ref}](Landing.html#schema-basepynnsynapse){.reference
    .internal} []{#baseq10settings}

-   [[baseQ10Settings]{.std
    .std-ref}](Landing.html#schema-baseq10settings){.reference
    .internal} []{#basespikesource}

-   [[baseSpikeSource]{.std
    .std-ref}](Landing.html#schema-basespikesource){.reference
    .internal} []{#basespikingcell}

-   [[baseSpikingCell]{.std
    .std-ref}](Landing.html#schema-basespikingcell){.reference
    .internal} []{#basestandalone}

-   [[baseStandalone]{.std
    .std-ref}](Landing.html#schema-basestandalone){.reference .internal}
    []{#basesynapse}

-   [[baseSynapse]{.std
    .std-ref}](Landing.html#schema-basesynapse){.reference .internal}
    []{#basesynapsedl}

-   [[baseSynapseDL]{.std
    .std-ref}](Landing.html#schema-basesynapsedl){.reference .internal}
    []{#basevoltageconcdeprate}

-   [[baseVoltageConcDepRate]{.std
    .std-ref}](Landing.html#schema-basevoltageconcdeprate){.reference
    .internal} []{#basevoltageconcdeptime}

-   [[baseVoltageConcDepTime]{.std
    .std-ref}](Landing.html#schema-basevoltageconcdeptime){.reference
    .internal} []{#basevoltageconcdepvariable}

-   [[baseVoltageConcDepVariable]{.std
    .std-ref}](Landing.html#schema-basevoltageconcdepvariable){.reference
    .internal} []{#basevoltagedeppointcurrent}

-   [[baseVoltageDepPointCurrent]{.std
    .std-ref}](Landing.html#schema-basevoltagedeppointcurrent){.reference
    .internal} []{#basevoltagedeppointcurrentdl}

-   [[baseVoltageDepPointCurrentDL]{.std
    .std-ref}](Landing.html#schema-basevoltagedeppointcurrentdl){.reference
    .internal} []{#basevoltagedeppointcurrentspiking}

-   [[baseVoltageDepPointCurrentSpiking]{.std
    .std-ref}](Landing.html#schema-basevoltagedeppointcurrentspiking){.reference
    .internal} []{#basevoltagedeprate}

-   [[baseVoltageDepRate]{.std
    .std-ref}](Landing.html#schema-basevoltagedeprate){.reference
    .internal} []{#basevoltagedepsynapse}

-   [[baseVoltageDepSynapse]{.std
    .std-ref}](Landing.html#schema-basevoltagedepsynapse){.reference
    .internal} []{#basevoltagedeptime}

-   [[baseVoltageDepTime]{.std
    .std-ref}](Landing.html#schema-basevoltagedeptime){.reference
    .internal} []{#basevoltagedepvariable}

-   [[baseVoltageDepVariable]{.std
    .std-ref}](Landing.html#schema-basevoltagedepvariable){.reference
    .internal} []{#biophysicalproperties}

-   [[biophysicalProperties]{.std
    .std-ref}](Landing.html#schema-biophysicalproperties){.reference
    .internal} []{#biophysicalproperties2capools}

-   [[biophysicalProperties2CaPools]{.std
    .std-ref}](Landing.html#schema-biophysicalproperties2capools){.reference
    .internal} []{#blockingplasticsynapse}

-   [[blockingPlasticSynapse]{.std
    .std-ref}](Landing.html#schema-blockingplasticsynapse){.reference
    .internal} []{#bqbiol_encodes}

-   [[bqbiol_encodes]{.std
    .std-ref}](Landing.html#schema-bqbiol-encodes){.reference .internal}
    []{#bqbiol_haspart}

-   [[bqbiol_hasPart]{.std
    .std-ref}](Landing.html#schema-bqbiol-haspart){.reference .internal}
    []{#bqbiol_hasproperty}

-   [[bqbiol_hasProperty]{.std
    .std-ref}](Landing.html#schema-bqbiol-hasproperty){.reference
    .internal} []{#bqbiol_hastaxon}

-   [[bqbiol_hasTaxon]{.std
    .std-ref}](Landing.html#schema-bqbiol-hastaxon){.reference
    .internal} []{#bqbiol_hasversion}

-   [[bqbiol_hasVersion]{.std
    .std-ref}](Landing.html#schema-bqbiol-hasversion){.reference
    .internal} []{#bqbiol_is}

-   [[bqbiol_is]{.std
    .std-ref}](Landing.html#schema-bqbiol-is){.reference .internal}
    []{#bqbiol_isdescribedby}

-   [[bqbiol_isDescribedBy]{.std
    .std-ref}](Landing.html#schema-bqbiol-isdescribedby){.reference
    .internal} []{#bqbiol_isencodedby}

-   [[bqbiol_isEncodedBy]{.std
    .std-ref}](Landing.html#schema-bqbiol-isencodedby){.reference
    .internal} []{#bqbiol_ishomologto}

-   [[bqbiol_isHomologTo]{.std
    .std-ref}](Landing.html#schema-bqbiol-ishomologto){.reference
    .internal} []{#bqbiol_ispartof}

-   [[bqbiol_isPartOf]{.std
    .std-ref}](Landing.html#schema-bqbiol-ispartof){.reference
    .internal} []{#bqbiol_ispropertyof}

-   [[bqbiol_isPropertyOf]{.std
    .std-ref}](Landing.html#schema-bqbiol-ispropertyof){.reference
    .internal} []{#bqbiol_isversionof}

-   [[bqbiol_isVersionOf]{.std
    .std-ref}](Landing.html#schema-bqbiol-isversionof){.reference
    .internal} []{#bqbiol_occursin}

-   [[bqbiol_occursIn]{.std
    .std-ref}](Landing.html#schema-bqbiol-occursin){.reference
    .internal} []{#bqmodel_is}

-   [[bqmodel_is]{.std
    .std-ref}](Landing.html#schema-bqmodel-is){.reference .internal}
    []{#bqmodel_isderivedfrom}

-   [[bqmodel_isDerivedFrom]{.std
    .std-ref}](Landing.html#schema-bqmodel-isderivedfrom){.reference
    .internal} []{#bqmodel_isdescribedby}

-   [[bqmodel_isDescribedBy]{.std
    .std-ref}](Landing.html#schema-bqmodel-isdescribedby){.reference
    .internal} []{#capacitance}

-   [[capacitance]{.std
    .std-ref}](Landing.html#schema-dimensions-capacitance){.reference
    .internal} []{#cell}

-   [[cell]{.std .std-ref}](Landing.html#schema-cell){.reference
    .internal} []{#cell2capools}

-   [[cell2CaPools]{.std
    .std-ref}](Landing.html#schema-cell2capools){.reference .internal}
    []{#channeldensity}

-   [[channelDensity]{.std
    .std-ref}](Landing.html#schema-channeldensity){.reference .internal}
    []{#channeldensityghk}

-   [[channelDensityGHK]{.std
    .std-ref}](Landing.html#schema-channeldensityghk){.reference
    .internal} []{#channeldensityghk2}

-   [[channelDensityGHK2]{.std
    .std-ref}](Landing.html#schema-channeldensityghk2){.reference
    .internal} []{#channeldensitynernst}

-   [[channelDensityNernst]{.std
    .std-ref}](Landing.html#schema-channeldensitynernst){.reference
    .internal} []{#channeldensitynernstca2}

-   [[channelDensityNernstCa2]{.std
    .std-ref}](Landing.html#schema-channeldensitynernstca2){.reference
    .internal} []{#channeldensitynonuniform}

-   [[channelDensityNonUniform]{.std
    .std-ref}](Landing.html#schema-channeldensitynonuniform){.reference
    .internal} []{#channeldensitynonuniformghk}

-   [[channelDensityNonUniformGHK]{.std
    .std-ref}](Landing.html#schema-channeldensitynonuniformghk){.reference
    .internal} []{#channeldensitynonuniformnernst}

-   [[channelDensityNonUniformNernst]{.std
    .std-ref}](Landing.html#schema-channeldensitynonuniformnernst){.reference
    .internal} []{#channeldensityvshift}

-   [[channelDensityVShift]{.std
    .std-ref}](Landing.html#schema-channeldensityvshift){.reference
    .internal} []{#channelpopulation}

-   [[channelPopulation]{.std
    .std-ref}](Landing.html#schema-channelpopulation){.reference
    .internal} []{#channelpopulationnernst}

-   [[channelPopulationNernst]{.std
    .std-ref}](Landing.html#schema-channelpopulationnernst){.reference
    .internal} []{#charge}

-   [[charge]{.std
    .std-ref}](Landing.html#schema-dimensions-charge){.reference
    .internal} []{#charge_per_mole}

-   [[charge_per_mole]{.std
    .std-ref}](Landing.html#schema-dimensions-charge-per-mole){.reference
    .internal} []{#closedstate}

-   [[closedState]{.std
    .std-ref}](Landing.html#schema-closedstate){.reference .internal}
    []{#compoundinput}

-   [[compoundInput]{.std
    .std-ref}](Landing.html#schema-compoundinput){.reference .internal}
    []{#compoundinputdl}

-   [[compoundInputDL]{.std
    .std-ref}](Landing.html#schema-compoundinputdl){.reference
    .internal} []{#concentration}

-   [[concentration]{.std
    .std-ref}](Landing.html#schema-dimensions-concentration){.reference
    .internal} []{#concentrationmodel}

-   [[concentrationModel]{.std
    .std-ref}](Landing.html#schema-concentrationmodel){.reference
    .internal} []{#conductance}

-   [[conductance]{.std
    .std-ref}](Landing.html#schema-dimensions-conductance){.reference
    .internal} []{#conductance_per_voltage}

-   [[conductance_per_voltage]{.std
    .std-ref}](Landing.html#schema-dimensions-conductance-per-voltage){.reference
    .internal} []{#conductanceDensity}

-   [[conductanceDensity]{.std
    .std-ref}](Landing.html#schema-dimensions-conductancedensity){.reference
    .internal} []{#connection}

-   [[connection]{.std
    .std-ref}](Landing.html#schema-connection){.reference .internal}
    []{#connectionwd}

-   [[connectionWD]{.std
    .std-ref}](Landing.html#schema-connectionwd){.reference .internal}
    []{#continuousconnection}

-   [[continuousConnection]{.std
    .std-ref}](Landing.html#schema-continuousconnection){.reference
    .internal} []{#continuousconnectioninstance}

-   [[continuousConnectionInstance]{.std
    .std-ref}](Landing.html#schema-continuousconnectioninstance){.reference
    .internal} []{#continuousconnectioninstancew}

-   [[continuousConnectionInstanceW]{.std
    .std-ref}](Landing.html#schema-continuousconnectioninstancew){.reference
    .internal} []{#continuousprojection}

-   [[continuousProjection]{.std
    .std-ref}](Landing.html#schema-continuousprojection){.reference
    .internal} []{#current}

-   [[current]{.std
    .std-ref}](Landing.html#schema-dimensions-current){.reference
    .internal} []{#currentDensity}

-   [[currentDensity]{.std
    .std-ref}](Landing.html#schema-dimensions-currentdensity){.reference
    .internal} []{#decayingpoolconcentrationmodel}

-   [[decayingPoolConcentrationModel]{.std
    .std-ref}](Landing.html#schema-decayingpoolconcentrationmodel){.reference
    .internal} []{#display}

-   [[Display]{.std .std-ref}](Landing.html#schema-display){.reference
    .internal} []{#distal}

-   [[distal]{.std .std-ref}](Landing.html#schema-distal){.reference
    .internal} []{#distaldetails}

-   [[distalDetails]{.std
    .std-ref}](Landing.html#schema-distaldetails){.reference .internal}
    []{#doublesynapse}

-   [[doubleSynapse]{.std
    .std-ref}](Landing.html#schema-doublesynapse){.reference .internal}
    []{#eif_cond_alpha_isfa_ista}

-   [[EIF_cond_alpha_isfa_ista]{.std
    .std-ref}](Landing.html#schema-eif-cond-alpha-isfa-ista){.reference
    .internal} []{#eif_cond_exp_isfa_ista}

-   [[EIF_cond_exp_isfa_ista]{.std
    .std-ref}](Landing.html#schema-eif-cond-exp-isfa-ista){.reference
    .internal} []{#electricalconnection}

-   [[electricalConnection]{.std
    .std-ref}](Landing.html#schema-electricalconnection){.reference
    .internal} []{#electricalconnectioninstance}

-   [[electricalConnectionInstance]{.std
    .std-ref}](Landing.html#schema-electricalconnectioninstance){.reference
    .internal} []{#electricalconnectioninstancew}

-   [[electricalConnectionInstanceW]{.std
    .std-ref}](Landing.html#schema-electricalconnectioninstancew){.reference
    .internal} []{#electricalprojection}

-   [[electricalProjection]{.std
    .std-ref}](Landing.html#schema-electricalprojection){.reference
    .internal} []{#eventoutputfile}

-   [[EventOutputFile]{.std
    .std-ref}](Landing.html#schema-eventoutputfile){.reference
    .internal} []{#eventselection}

-   [[EventSelection]{.std
    .std-ref}](Landing.html#schema-eventselection){.reference .internal}
    []{#expcondsynapse}

-   [[expCondSynapse]{.std
    .std-ref}](Landing.html#schema-expcondsynapse){.reference .internal}
    []{#expcurrsynapse}

-   [[expCurrSynapse]{.std
    .std-ref}](Landing.html#schema-expcurrsynapse){.reference .internal}
    []{#explicitconnection}

-   [[explicitConnection]{.std
    .std-ref}](Landing.html#schema-explicitconnection){.reference
    .internal} []{#explicitinput}

-   [[explicitInput]{.std
    .std-ref}](Landing.html#schema-explicitinput){.reference .internal}
    []{#exponesynapse}

-   [[expOneSynapse]{.std
    .std-ref}](Landing.html#schema-exponesynapse){.reference .internal}
    []{#expthreesynapse}

-   [[expThreeSynapse]{.std
    .std-ref}](Landing.html#schema-expthreesynapse){.reference
    .internal} []{#exptwosynapse}

-   [[expTwoSynapse]{.std
    .std-ref}](Landing.html#schema-exptwosynapse){.reference .internal}
    []{#fitzhughnagumocell}

-   [[fitzHughNagumoCell]{.std
    .std-ref}](Landing.html#schema-fitzhughnagumocell){.reference
    .internal} []{#fixedfactorconcentrationmodel}

-   [[fixedFactorConcentrationModel]{.std
    .std-ref}](Landing.html#schema-fixedfactorconcentrationmodel){.reference
    .internal} []{#fixedfactorconcentrationmodeltraub}

-   [[fixedFactorConcentrationModelTraub]{.std
    .std-ref}](Landing.html#schema-fixedfactorconcentrationmodeltraub){.reference
    .internal} []{#fixedtimecourse}

-   [[fixedTimeCourse]{.std
    .std-ref}](Landing.html#schema-fixedtimecourse){.reference
    .internal} []{#forwardtransition}

-   [[forwardTransition]{.std
    .std-ref}](Landing.html#schema-forwardtransition){.reference
    .internal} []{#from}

-   [[from]{.std .std-ref}](Landing.html#schema-from){.reference
    .internal} []{#gapjunction}

-   [[gapJunction]{.std
    .std-ref}](Landing.html#schema-gapjunction){.reference .internal}
    []{#gate}

-   [[gate]{.std .std-ref}](Landing.html#schema-gate){.reference
    .internal} []{#gatefractional}

-   [[gateFractional]{.std
    .std-ref}](Landing.html#schema-gatefractional){.reference .internal}
    []{#gatehhinstantaneous}

-   [[gateHHInstantaneous]{.std
    .std-ref}](Landing.html#schema-gatehhinstantaneous){.reference
    .internal} []{#gatehhrates}

-   [[gateHHrates]{.std
    .std-ref}](Landing.html#schema-gatehhrates){.reference .internal}
    []{#gatehhratesinf}

-   [[gateHHratesInf]{.std
    .std-ref}](Landing.html#schema-gatehhratesinf){.reference .internal}
    []{#gatehhratestau}

-   [[gateHHratesTau]{.std
    .std-ref}](Landing.html#schema-gatehhratestau){.reference .internal}
    []{#gatehhratestauinf}

-   [[gateHHratesTauInf]{.std
    .std-ref}](Landing.html#schema-gatehhratestauinf){.reference
    .internal} []{#gatehhtauinf}

-   [[gateHHtauInf]{.std
    .std-ref}](Landing.html#schema-gatehhtauinf){.reference .internal}
    []{#gateks}

-   [[gateKS]{.std .std-ref}](Landing.html#schema-gateks){.reference
    .internal} []{#gradedsynapse}

-   [[gradedSynapse]{.std
    .std-ref}](Landing.html#schema-gradedsynapse){.reference .internal}
    []{#hh_cond_exp}

-   [[HH_cond_exp]{.std
    .std-ref}](Landing.html#schema-hh-cond-exp){.reference .internal}
    []{#hhexplinearrate}

-   [[HHExpLinearRate]{.std
    .std-ref}](Landing.html#schema-hhexplinearrate){.reference
    .internal} []{#hhexplinearvariable}

-   [[HHExpLinearVariable]{.std
    .std-ref}](Landing.html#schema-hhexplinearvariable){.reference
    .internal} []{#hhexprate}

-   [[HHExpRate]{.std
    .std-ref}](Landing.html#schema-hhexprate){.reference .internal}
    []{#hhexpvariable}

-   [[HHExpVariable]{.std
    .std-ref}](Landing.html#schema-hhexpvariable){.reference .internal}
    []{#hhsigmoidrate}

-   [[HHSigmoidRate]{.std
    .std-ref}](Landing.html#schema-hhsigmoidrate){.reference .internal}
    []{#hhsigmoidvariable}

-   [[HHSigmoidVariable]{.std
    .std-ref}](Landing.html#schema-hhsigmoidvariable){.reference
    .internal} []{#hindmarshrose1984cell}

-   [[hindmarshRose1984Cell]{.std
    .std-ref}](Landing.html#schema-hindmarshrose1984cell){.reference
    .internal} []{#iafcell}

-   [[iafCell]{.std .std-ref}](Landing.html#schema-iafcell){.reference
    .internal} []{#iafrefcell}

-   [[iafRefCell]{.std
    .std-ref}](Landing.html#schema-iafrefcell){.reference .internal}
    []{#iaftaucell}

-   [[iafTauCell]{.std
    .std-ref}](Landing.html#schema-iaftaucell){.reference .internal}
    []{#iaftaurefcell}

-   [[iafTauRefCell]{.std
    .std-ref}](Landing.html#schema-iaftaurefcell){.reference .internal}
    []{#idealGasConstantDims}

-   [[idealGasConstantDims]{.std
    .std-ref}](Landing.html#schema-dimensions-idealgasconstantdims){.reference
    .internal} []{#if_cond_alpha}

-   [[IF_cond_alpha]{.std
    .std-ref}](Landing.html#schema-if-cond-alpha){.reference .internal}
    []{#if_cond_exp}

-   [[IF_cond_exp]{.std
    .std-ref}](Landing.html#schema-if-cond-exp){.reference .internal}
    []{#if_curr_alpha}

-   [[IF_curr_alpha]{.std
    .std-ref}](Landing.html#schema-if-curr-alpha){.reference .internal}
    []{#if_curr_exp}

-   [[IF_curr_exp]{.std
    .std-ref}](Landing.html#schema-if-curr-exp){.reference .internal}
    []{#include}

-   [[include]{.std .std-ref}](Landing.html#schema-include){.reference
    .internal} []{#includetype}

-   [[IncludeType]{.std
    .std-ref}](Landing.html#schema-includetype){.reference .internal}
    []{#inhomogeneousparameter}

-   [[inhomogeneousParameter]{.std
    .std-ref}](Landing.html#schema-inhomogeneousparameter){.reference
    .internal} []{#inhomogeneousvalue}

-   [[inhomogeneousValue]{.std
    .std-ref}](Landing.html#schema-inhomogeneousvalue){.reference
    .internal} []{#initmembpotential}

-   [[initMembPotential]{.std
    .std-ref}](Landing.html#schema-initmembpotential){.reference
    .internal} []{#input}

-   [[input]{.std .std-ref}](Landing.html#schema-input){.reference
    .internal} []{#inputlist}

-   [[inputList]{.std
    .std-ref}](Landing.html#schema-inputlist){.reference .internal}
    []{#inputw}

-   [[inputW]{.std .std-ref}](Landing.html#schema-inputw){.reference
    .internal} []{#instance}

-   [[instance]{.std .std-ref}](Landing.html#schema-instance){.reference
    .internal} []{#intracellularproperties}

-   [[intracellularProperties]{.std
    .std-ref}](Landing.html#schema-intracellularproperties){.reference
    .internal} []{#intracellularproperties2capools}

-   [[intracellularProperties2CaPools]{.std
    .std-ref}](Landing.html#schema-intracellularproperties2capools){.reference
    .internal} []{#ionchannel}

-   [[ionChannel]{.std
    .std-ref}](Landing.html#schema-ionchannel){.reference .internal}
    []{#ionchannelhh}

-   [[ionChannelHH]{.std
    .std-ref}](Landing.html#schema-ionchannelhh){.reference .internal}
    []{#ionchannelks}

-   [[ionChannelKS]{.std
    .std-ref}](Landing.html#schema-ionchannelks){.reference .internal}
    []{#ionchannelpassive}

-   [[ionChannelPassive]{.std
    .std-ref}](Landing.html#schema-ionchannelpassive){.reference
    .internal} []{#ionchannelvshift}

-   [[ionChannelVShift]{.std
    .std-ref}](Landing.html#schema-ionchannelvshift){.reference
    .internal} []{#izhikevich2007cell}

-   [[izhikevich2007Cell]{.std
    .std-ref}](Landing.html#schema-izhikevich2007cell){.reference
    .internal} []{#izhikevichcell}

-   [[izhikevichCell]{.std
    .std-ref}](Landing.html#schema-izhikevichcell){.reference .internal}
    []{#ksstate}

-   [[KSState]{.std .std-ref}](Landing.html#schema-ksstate){.reference
    .internal} []{#kstransition}

-   [[KSTransition]{.std
    .std-ref}](Landing.html#schema-kstransition){.reference .internal}
    []{#length}

-   [[length]{.std
    .std-ref}](Landing.html#schema-dimensions-length){.reference
    .internal} []{#line}

-   [[Line]{.std .std-ref}](Landing.html#schema-line){.reference
    .internal} []{#lineargradedsynapse}

-   [[linearGradedSynapse]{.std
    .std-ref}](Landing.html#schema-lineargradedsynapse){.reference
    .internal} []{#location}

-   [[location]{.std .std-ref}](Landing.html#schema-location){.reference
    .internal} []{#member}

-   [[member]{.std .std-ref}](Landing.html#schema-member){.reference
    .internal} []{#membraneproperties}

-   [[membraneProperties]{.std
    .std-ref}](Landing.html#schema-membraneproperties){.reference
    .internal} []{#membraneproperties2capools}

-   [[membraneProperties2CaPools]{.std
    .std-ref}](Landing.html#schema-membraneproperties2capools){.reference
    .internal} []{#meta}

-   [[Meta]{.std .std-ref}](Landing.html#schema-meta){.reference
    .internal} []{#morphology}

-   [[morphology]{.std
    .std-ref}](Landing.html#schema-morphology){.reference .internal}
    []{#network}

-   [[network]{.std .std-ref}](Landing.html#schema-network){.reference
    .internal} []{#networkwithtemperature}

-   [[networkWithTemperature]{.std
    .std-ref}](Landing.html#schema-networkwithtemperature){.reference
    .internal} []{#neuromldocument}

-   [[NeuroMLDocument]{.std
    .std-ref}](Landing.html#schema-neuromldocument){.reference
    .internal} []{#notes}

-   [[notes]{.std .std-ref}](Landing.html#schema-notes){.reference
    .internal} []{#openstate}

-   [[openState]{.std
    .std-ref}](Landing.html#schema-openstate){.reference .internal}
    []{#outputcolumn}

-   [[OutputColumn]{.std
    .std-ref}](Landing.html#schema-outputcolumn){.reference .internal}
    []{#outputfile}

-   [[OutputFile]{.std
    .std-ref}](Landing.html#schema-outputfile){.reference .internal}
    []{#parent}

-   [[parent]{.std .std-ref}](Landing.html#schema-parent){.reference
    .internal} []{#path}

-   [[path]{.std .std-ref}](Landing.html#schema-path){.reference
    .internal} []{#per_time}

-   [[per_time]{.std
    .std-ref}](Landing.html#schema-dimensions-per-time){.reference
    .internal} []{#per_voltage}

-   [[per_voltage]{.std
    .std-ref}](Landing.html#schema-dimensions-per-voltage){.reference
    .internal} []{#permeability}

-   [[permeability]{.std
    .std-ref}](Landing.html#schema-dimensions-permeability){.reference
    .internal} []{#pinskyrinzelca3cell}

-   [[pinskyRinzelCA3Cell]{.std
    .std-ref}](Landing.html#schema-pinskyrinzelca3cell){.reference
    .internal} []{#point3dwithdiam}

-   [[point3DWithDiam]{.std
    .std-ref}](Landing.html#schema-point3dwithdiam){.reference
    .internal} []{#pointcellcondbased}

-   [[pointCellCondBased]{.std
    .std-ref}](Landing.html#schema-pointcellcondbased){.reference
    .internal} []{#pointcellcondbasedca}

-   [[pointCellCondBasedCa]{.std
    .std-ref}](Landing.html#schema-pointcellcondbasedca){.reference
    .internal} []{#poissonfiringsynapse}

-   [[poissonFiringSynapse]{.std
    .std-ref}](Landing.html#schema-poissonfiringsynapse){.reference
    .internal} []{#population}

-   [[population]{.std
    .std-ref}](Landing.html#schema-population){.reference .internal}
    []{#populationlist}

-   [[populationList]{.std
    .std-ref}](Landing.html#schema-populationlist){.reference .internal}
    []{#projection}

-   [[projection]{.std
    .std-ref}](Landing.html#schema-projection){.reference .internal}
    []{#property}

-   [[property]{.std .std-ref}](Landing.html#schema-property){.reference
    .internal} []{#proximal}

-   [[proximal]{.std .std-ref}](Landing.html#schema-proximal){.reference
    .internal} []{#proximaldetails}

-   [[proximalDetails]{.std
    .std-ref}](Landing.html#schema-proximaldetails){.reference
    .internal} []{#pulsegenerator}

-   [[pulseGenerator]{.std
    .std-ref}](Landing.html#schema-pulsegenerator){.reference .internal}
    []{#pulsegeneratordl}

-   [[pulseGeneratorDL]{.std
    .std-ref}](Landing.html#schema-pulsegeneratordl){.reference
    .internal} []{#q10conductancescaling}

-   [[q10ConductanceScaling]{.std
    .std-ref}](Landing.html#schema-q10conductancescaling){.reference
    .internal} []{#q10exptemp}

-   [[q10ExpTemp]{.std
    .std-ref}](Landing.html#schema-q10exptemp){.reference .internal}
    []{#q10fixed}

-   [[q10Fixed]{.std .std-ref}](Landing.html#schema-q10fixed){.reference
    .internal} []{#rampgenerator}

-   [[rampGenerator]{.std
    .std-ref}](Landing.html#schema-rampgenerator){.reference .internal}
    []{#rampgeneratordl}

-   [[rampGeneratorDL]{.std
    .std-ref}](Landing.html#schema-rampgeneratordl){.reference
    .internal} []{#rdf_bag}

-   [[rdf_Bag]{.std .std-ref}](Landing.html#schema-rdf-bag){.reference
    .internal} []{#rdf_description}

-   [[rdf_Description]{.std
    .std-ref}](Landing.html#schema-rdf-description){.reference
    .internal} []{#rdf_li}

-   [[rdf_li]{.std .std-ref}](Landing.html#schema-rdf-li){.reference
    .internal} []{#rdf_rdf}

-   [[rdf_RDF]{.std .std-ref}](Landing.html#schema-rdf-rdf){.reference
    .internal} []{#rectangularextent}

-   [[rectangularExtent]{.std
    .std-ref}](Landing.html#schema-rectangularextent){.reference
    .internal} []{#region}

-   [[region]{.std .std-ref}](Landing.html#schema-region){.reference
    .internal} []{#resistance}

-   [[resistance]{.std
    .std-ref}](Landing.html#schema-dimensions-resistance){.reference
    .internal} []{#resistivity}

-   [[resistivity]{.std
    .std-ref}](Landing.html#schema-dimensions-resistivity){.reference
    .internal} []{#reversetransition}

-   [[reverseTransition]{.std
    .std-ref}](Landing.html#schema-reversetransition){.reference
    .internal} []{#rho_factor}

-   [[rho_factor]{.std
    .std-ref}](Landing.html#schema-dimensions-rho-factor){.reference
    .internal} []{#segment}

-   [[segment]{.std .std-ref}](Landing.html#schema-segment){.reference
    .internal} []{#segmentgroup}

-   [[segmentGroup]{.std
    .std-ref}](Landing.html#schema-segmentgroup){.reference .internal}
    []{#silentsynapse}

-   [[silentSynapse]{.std
    .std-ref}](Landing.html#schema-silentsynapse){.reference .internal}
    []{#simulation}

-   [[Simulation]{.std .std-ref}](Landing.html#id1){.reference
    .internal} []{#sinegenerator}

-   [[sineGenerator]{.std
    .std-ref}](Landing.html#schema-sinegenerator){.reference .internal}
    []{#sinegeneratordl}

-   [[sineGeneratorDL]{.std
    .std-ref}](Landing.html#schema-sinegeneratordl){.reference
    .internal} []{#species}

-   [[species]{.std .std-ref}](Landing.html#schema-species){.reference
    .internal} []{#specificCapacitance}

-   [[specificCapacitance]{.std
    .std-ref}](Landing.html#schema-dimensions-specificcapacitance){.reference
    .internal} []{#spike}

-   [[spike]{.std .std-ref}](Landing.html#schema-spike){.reference
    .internal} []{#spikearray}

-   [[spikeArray]{.std
    .std-ref}](Landing.html#schema-spikearray){.reference .internal}
    []{#spikegenerator}

-   [[spikeGenerator]{.std
    .std-ref}](Landing.html#schema-spikegenerator){.reference .internal}
    []{#spikegeneratorpoisson}

-   [[spikeGeneratorPoisson]{.std
    .std-ref}](Landing.html#schema-spikegeneratorpoisson){.reference
    .internal} []{#spikegeneratorrandom}

-   [[spikeGeneratorRandom]{.std
    .std-ref}](Landing.html#schema-spikegeneratorrandom){.reference
    .internal} []{#spikegeneratorrefpoisson}

-   [[spikeGeneratorRefPoisson]{.std
    .std-ref}](Landing.html#schema-spikegeneratorrefpoisson){.reference
    .internal} []{#spikesourcepoisson}

-   [[SpikeSourcePoisson]{.std
    .std-ref}](Landing.html#schema-spikesourcepoisson){.reference
    .internal} []{#spikethresh}

-   [[spikeThresh]{.std
    .std-ref}](Landing.html#schema-spikethresh){.reference .internal}
    []{#stdpsynapse}

-   [[stdpSynapse]{.std
    .std-ref}](Landing.html#schema-stdpsynapse){.reference .internal}
    []{#subgate}

-   [[subGate]{.std .std-ref}](Landing.html#schema-subgate){.reference
    .internal} []{#substance}

-   [[substance]{.std
    .std-ref}](Landing.html#schema-dimensions-substance){.reference
    .internal} []{#subtree}

-   [[subTree]{.std .std-ref}](Landing.html#schema-subtree){.reference
    .internal} []{#synapticconnection}

-   [[synapticConnection]{.std
    .std-ref}](Landing.html#schema-synapticconnection){.reference
    .internal} []{#synapticconnectionwd}

-   [[synapticConnectionWD]{.std
    .std-ref}](Landing.html#schema-synapticconnectionwd){.reference
    .internal} []{#tauinftransition}

-   [[tauInfTransition]{.std
    .std-ref}](Landing.html#schema-tauinftransition){.reference
    .internal} []{#temperature}

-   [[temperature]{.std
    .std-ref}](Landing.html#schema-dimensions-temperature){.reference
    .internal} []{#time}

-   [[time]{.std
    .std-ref}](Landing.html#schema-dimensions-time){.reference
    .internal} []{#timedsynapticinput}

-   [[timedSynapticInput]{.std
    .std-ref}](Landing.html#schema-timedsynapticinput){.reference
    .internal} []{#to}

-   [[to]{.std .std-ref}](Landing.html#schema-to){.reference .internal}
    []{#transientpoissonfiringsynapse}

-   [[transientPoissonFiringSynapse]{.std
    .std-ref}](Landing.html#schema-transientpoissonfiringsynapse){.reference
    .internal} []{#tsodyksmarkramdepfacmechanism}

-   [[tsodyksMarkramDepFacMechanism]{.std
    .std-ref}](Landing.html#schema-tsodyksmarkramdepfacmechanism){.reference
    .internal} []{#tsodyksmarkramdepmechanism}

-   [[tsodyksMarkramDepMechanism]{.std
    .std-ref}](Landing.html#schema-tsodyksmarkramdepmechanism){.reference
    .internal} []{#variableparameter}

-   [[variableParameter]{.std
    .std-ref}](Landing.html#schema-variableparameter){.reference
    .internal} []{#vhalftransition}

-   [[vHalfTransition]{.std
    .std-ref}](Landing.html#schema-vhalftransition){.reference
    .internal} []{#voltage}

-   [[voltage]{.std
    .std-ref}](Landing.html#schema-dimensions-voltage){.reference
    .internal} []{#voltageclamp}

-   [[voltageClamp]{.std
    .std-ref}](Landing.html#schema-voltageclamp){.reference .internal}
    []{#voltageclamptriple}

-   [[voltageClampTriple]{.std
    .std-ref}](Landing.html#schema-voltageclamptriple){.reference
    .internal} []{#voltageconcdepblockmechanism}

-   [[voltageConcDepBlockMechanism]{.std
    .std-ref}](Landing.html#schema-voltageconcdepblockmechanism){.reference
    .internal} []{#volume}

-   [[volume]{.std
    .std-ref}](Landing.html#schema-dimensions-volume){.reference
    .internal}
:::
:::
:::

[]{#document-Userdocs/NeuroMLv1}

::: {#neuroml-v1 .section .tex2jax_ignore .mathjax_ignore}
[]{#userdocs-neuromlv1}

### NeuroML v1[\#](#neuroml-v1 "Link to this heading"){.headerlink}

::: {.admonition .warning}
Warning

NeuroML v1.x is deprecated. This page is maintained for archival
purposes only.

Please use [[NeuroML v2]{.std
.std-ref}](Landing.html#userdocs-neuromlv2){.reference .internal}.

[[neuroConstruct]{.std
.std-ref}](Landing.html#userdocs-supporting-apps-neuroconstruct){.reference
.internal} can be used for converting NeuroML v1 models into NeuroML v2.
:::

There are three Levels of compliance to the NeuroML v1 specifications:

::: {#level-1 .section}
#### Level 1[\#](#level-1 "Link to this heading"){.headerlink}

-   [Metadata
    v1.8.1](https://github.com/NeuroML/org.neuroml1.model/blob/master/src/main/resources/NeuroML1Schemas/Level1/Metadata_v1.8.1.xsd){.reference
    .external}

-   [MorphML
    v1.8.1](https://github.com/NeuroML/org.neuroml1.model/blob/master/src/main/resources/NeuroML1Schemas/Level1/MorphML_v1.8.1.xsd){.reference
    .external}

Any Level 1 NeuroML v1 file will also be compliant to [this
schema](https://github.com/NeuroML/org.neuroml1.model/blob/master/src/main/resources/NeuroML1Schemas/Level1/NeuroML_Level1_v1.8.1.xsd){.reference
.external}.
:::

::: {#level-2 .section}
#### Level 2[\#](#level-2 "Link to this heading"){.headerlink}

-   [Biophysics
    v1.8.1](https://github.com/NeuroML/org.neuroml1.model/blob/master/src/main/resources/NeuroML1Schemas/Level2/Biophysics_v1.8.1.xsd){.reference
    .external}

-   [ChannelML
    v1.8.1](https://github.com/NeuroML/org.neuroml1.model/blob/master/src/main/resources/NeuroML1Schemas/Level2/ChannelML_v1.8.1.xsd){.reference
    .external}

Any Level 1 or Level 2 NeuroML v1 file will also be compliant to [this
schema](https://github.com/NeuroML/org.neuroml1.model/blob/master/src/main/resources/NeuroML1Schemas/Level2/NeuroML_Level2_v1.8.1.xsd){.reference
.external}.
:::

::: {#level-3 .section}
#### Level 3[\#](#level-3 "Link to this heading"){.headerlink}

-   [NetworkML
    v1.8.1](https://github.com/NeuroML/org.neuroml1.model/blob/master/src/main/resources/NeuroML1Schemas/Level3/NetworkML_v1.8.1.xsd){.reference
    .external}

Any Level 1 or Level 2 or Level 3 NeuroML v1 file will also be compliant
to [this
schema](https://github.com/NeuroML/org.neuroml1.model/blob/master/src/main/resources/NeuroML1Schemas/Level3/NeuroML_Level3_v1.8.1.xsd){.reference
.external}.

These files are archived in [this GitHub
repository](https://github.com/NeuroML/org.neuroml1.model/tree/master/src/main/resources/NeuroML1Schemas){.reference
.external}.
:::
:::

[]{#document-Userdocs/LEMSSchema}

::: {#lems .section .tex2jax_ignore .mathjax_ignore}
[]{#userdocs-lemsschema}

### LEMS[\#](#lems "Link to this heading"){.headerlink}

The current version of the LEMS specification is 0.7.6 and the schema
for this can be seen
[here](https://github.com/LEMS/LEMS/blob/master/Schemas/LEMS/LEMS_v0.7.6.xsd){.reference
.external}. The following figure, taken from Cannon et al. 2014
([\[[CGC+14](Landing.html#id10 "Robert C. Cannon, Padraig Gleeson, Sharon Crook, Gautham Ganapathy, Boris Marin, Eugenio Piasini, and R. Angus Silver. LEMS: a language for expressing complex biological models in concise and hierarchical form and its use in underpinning NeuroML 2. Frontiers in Neuroinformatics, 2014. doi:10.3389/fninf.2014.00079."){.reference
.internal}\]]{#id1}) shows the structure of LEMS models. The following
pages give details of all the elements that are included in LEMS. For
examples on LEMS, and using LEMS to extend NeuroML, please see the
relevant sections in the documentation.

<figure id="id2" class="align-center">
<a href="_images/lems-figure2.png"
class="reference internal image-reference"><img
src="_images/lems-figure2.png" style="width: 80%;"
alt="Structure of LEMS models" /></a>
<figcaption><p><span class="caption-number">Fig. 58 </span><span
class="caption-text"><em>(A)</em> Models in LEMS are specified using
ComponentType definitions with nested Dynamics elements. Any Parameter
or StateVariable declaration must refer to a Dimension element defined
at the top level. A Component element sets parameter values for a
particular instance of a ComponentType. Each Parameter value must refer
to one of the Unit elements defined at the top level. The Dynamics
element supports continuous time systems defined in terms of first order
differential equations, and event driven processing as specified by the
various “On…” elements. Multiple Regimes, each with independent
TimeDerivative expressions can be defined, along with the rules to
transition between them. <em>(B)</em> Example of a ComponentType, the
passive channel model from Figure 1. <em>©</em> The XML equivalent of
the ComponentType (top) and Component (bottom) for this model.
<em>(D)</em> Defining containment in LEMS, using Child (exactly one sub
element of the given type) or Children (zero or multiple copies).
<strong>(E)</strong> Extension in LEMS. Extending ComponentTypes inherit
the structure of the base type. Example Components in XML are shown in
<em>(D,E)</em>.</span><a href="#id2" class="headerlink"
title="Link to this image">#</a></p></figcaption>
</figure>

::: {.toctree-wrapper .compound}
[]{#document-Userdocs/LEMS_elements/Modelstructure}

::: {#model-structure .section .tex2jax_ignore .mathjax_ignore}
[]{#lemsschema-page-model-structure}

#### Model structure[\#](#model-structure "Link to this heading"){.headerlink}

**Models can be spread over multiple files. The root element in each
file is Lems.**

------------------------------------------------------------------------

Schema against which LEMS based on these should be valid:
[LEMS_v0.7.6.xsd](https://github.com/LEMS/LEMS/tree/master/Schemas/LEMS/LEMS_v0.7.6.xsd){.reference
.external}. Generated on 18/06/24 from
[this](https://github.com/LEMS/LEMS/commit/fd7b30eceb6735ac343745c8f6992bdde72b248b){.reference
.external} commit. Please file any issues or questions at the [issue
tracker here](https://github.com/LEMS/LEMS/issues){.reference
.external}.

------------------------------------------------------------------------

::: {#lems .section}
[]{#lemsschema-lems}

##### Lems[\#](#lems "Link to this heading"){.headerlink}

*Root element for any lems content*

::: {.sd-tab-set .docutils}
can contain these elements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+-------------+--------------------------------------------------------+
| **d         | [[Dimension]{.std                                      |
| imensions** | .s                                                     |
|             | td-ref}](Landing.html#lemsschema-dimension){.reference |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
| **          | [[Constant]{.std                                       |
| constants** | .std-ref}](#lemsschema-constant){.reference .internal} |
+-------------+--------------------------------------------------------+
| **units**   | [[Unit]{.std                                           |
|             | .std-ref}](Landing.html#lemsschema-unit){.reference    |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
| **a         | [[Assertion]{.std                                      |
| ssertions** | .s                                                     |
|             | td-ref}](Landing.html#lemsschema-assertion){.reference |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
| **compo     | [[ComponentType]{.std                                  |
| nentTypes** | .std-r                                                 |
|             | ef}](Landing.html#lemsschema-componenttype){.reference |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
| **c         | [[Component]{.std                                      |
| omponents** | .s                                                     |
|             | td-ref}](Landing.html#lemsschema-component){.reference |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
| **targets** | [[Target]{.std                                         |
|             | .std-ref}](#lemsschema-target){.reference .internal}   |
+-------------+--------------------------------------------------------+
:::
:::
:::
:::

::: {#target .section}
[]{#lemsschema-target}

##### Target[\#](#target "Link to this heading"){.headerlink}

*A lems file can contain many component definitions. A Target elements
specifies that a components should be treated as the entry point for
simulation or other processing*

::: {.sd-tab-set .docutils}
Properties

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------+-------------------------------------------------+
| *    | String      | Reference to the entry point component          |
| *com |             |                                                 |
| pone |             |                                                 |
| nt** |             |                                                 |
+------+-------------+-------------------------------------------------+
| **   | String      | Optional attribute specifying file in which to  |
| repo |             | save short report of simulation                 |
| rtFi |             |                                                 |
| le** |             |                                                 |
+------+-------------+-------------------------------------------------+
| *    | String      | Optional attribute specifying file in which to  |
| *tim |             | save times used in simulation                   |
| esFi |             |                                                 |
| le** |             |                                                 |
+------+-------------+-------------------------------------------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="Target">
      <xs:attribute name="component" type="xs:string" use="required"/>
      <xs:attribute name="reportFile" type="xs:string" use="optional">
        <xs:annotation>
          <xs:documentation>jLEMS only optional attribute to also write a short report with simulation duration, version, etc.</xs:documentation>
        </xs:annotation>
      </xs:attribute>
      <xs:attribute name="timesFile" type="xs:string" use="optional">
        <xs:annotation>
          <xs:documentation>jLEMS only optional attribute to also write a file containing actual times used in the simulation.</xs:documentation>
        </xs:annotation>
      </xs:attribute>
    </xs:complexType>
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <Target component="sim1"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Target component="sim1"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Target component="sim1"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Target component="sim1"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Target component="sim1"/>
:::
:::
:::
:::
:::

::: {#constant .section}
[]{#lemsschema-constant}

##### Constant[\#](#constant "Link to this heading"){.headerlink}

*A constant quantity: like a parameter for which the value is supplied
in the class definition itself rather than when a component is defined.*

::: {.sd-tab-set .docutils}
Properties

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------+-------------------------------------------------+
| **na | String      | A readable name for the constant.               |
| me** |             |                                                 |
+------+-------------+-------------------------------------------------+
| **   | String      | The symbol used in expressions to refer to this |
| symb |             | constant.                                       |
| ol** |             |                                                 |
+------+-------------+-------------------------------------------------+
| *    | String      | The value of a constant must be a plain number  |
| *val |             | (no units) giving the SI magnitude of the       |
| ue** |             | quantity or an expression involving only plain  |
|      |             | numbers or other constants.                     |
+------+-------------+-------------------------------------------------+
| *    | String      |                                                 |
| *dim |             |                                                 |
| ensi |             |                                                 |
| on** |             |                                                 |
+------+-------------+-------------------------------------------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="Constant">
      <xs:attribute name="name" type="xs:string" use="required"/>
      <xs:attribute name="dimension" type="xs:string" use="optional" default="none"/>
      <xs:attribute name="value" type="PhysicalQuantity" use="required"/>
      <xs:attribute name="description" type="xs:string" use="optional"/>
    </xs:complexType>
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <Constant name="kte" dimension="voltage" value="25.3mV"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Constant name="kte" dimension="voltage" value="25.3mV"/>
:::
:::
:::
:::
:::

::: {#include .section}
[]{#lemsschema-include}

##### Include[\#](#include "Link to this heading"){.headerlink}

*Include LEMS files in other LEMS files. Files are included where the
Include declaration occurs. The enclosing Lems block is stripped off and
the rest of the content included as is*

::: {.sd-tab-set .docutils}
Properties

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------+-------------------------------------------------+
| **fi | String      | the name or relative path of a file to be       |
| le** |             | included                                        |
+------+-------------+-------------------------------------------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="Include">
      <xs:attribute name="file" type="xs:string" use="required"/>
    </xs:complexType>
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <Include file="SimpleNetwork.xml"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Include file="SingleSimulation.xml"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Include file="ex2dims.xml"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Include file="hhchannel.xml"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Include file="hhcell.xml"/>
:::
:::
:::
:::
:::
:::

[]{#document-Userdocs/LEMS_elements/Unitsanddimensions}

::: {#units-and-dimensions .section .tex2jax_ignore .mathjax_ignore}
[]{#lemsschema-page-units-and-dimensions}

#### Units and dimensions[\#](#units-and-dimensions "Link to this heading"){.headerlink}

Schema against which LEMS based on these should be valid:
[LEMS_v0.7.6.xsd](https://github.com/LEMS/LEMS/tree/master/Schemas/LEMS/LEMS_v0.7.6.xsd){.reference
.external}. Generated on 18/06/24 from
[this](https://github.com/LEMS/LEMS/commit/fd7b30eceb6735ac343745c8f6992bdde72b248b){.reference
.external} commit. Please file any issues or questions at the [issue
tracker here](https://github.com/LEMS/LEMS/issues){.reference
.external}.

------------------------------------------------------------------------

::: {#dimension .section}
[]{#lemsschema-dimension}

##### Dimension[\#](#dimension "Link to this heading"){.headerlink}

*A Dimenson element associated a name with a particular combination of
the standards SI base dimensions, mass, lenght, time, current,
temperature and amount if substance (moles). Fractional dimensions are
not currently supported.*

::: {.sd-tab-set .docutils}
Properties

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------+-------------------------------------------------+
| **na | String      | The name to be used when referring to this      |
| me** |             | dimension from variable declaration or units    |
+------+-------------+-------------------------------------------------+
| *    | int         | Mass                                            |
| *m** |             |                                                 |
+------+-------------+-------------------------------------------------+
| *    | int         | Length                                          |
| *l** |             |                                                 |
+------+-------------+-------------------------------------------------+
| *    | int         | Time                                            |
| *t** |             |                                                 |
+------+-------------+-------------------------------------------------+
| *    | int         | Current                                         |
| *i** |             |                                                 |
+------+-------------+-------------------------------------------------+
| *    | int         | Temperature                                     |
| *k** |             |                                                 |
+------+-------------+-------------------------------------------------+
| *    | int         | Amount of substance                             |
| *n** |             |                                                 |
+------+-------------+-------------------------------------------------+
| *    | int         | Luminous intensity                              |
| *j** |             |                                                 |
+------+-------------+-------------------------------------------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="Dimension">
      <xs:attribute name="name" type="xs:string" use="required"/>
      <xs:attribute name="m" type="xs:integer" use="optional" default="0"/>
      <xs:attribute name="l" type="xs:integer" use="optional" default="0"/>
      <xs:attribute name="t" type="xs:integer" use="optional" default="0"/>
      <xs:attribute name="i" type="xs:integer" use="optional" default="0"/>
      <xs:attribute name="k" type="xs:integer" use="optional" default="0"/>
      <xs:attribute name="n" type="xs:integer" use="optional" default="0"/>
    </xs:complexType>
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <Dimension name="voltage" m="1" l="2" t="-3" i="-1"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Dimension name="time" t="1"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Dimension name="conductance" m="-1" l="-2" t="3" i="2"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Dimension name="capacitance" m="-1" l="-2" t="4" i="2"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Dimension name="current" i="1"/>
:::
:::
:::
:::
:::

::: {#unit .section}
[]{#lemsschema-unit}

##### Unit[\#](#unit "Link to this heading"){.headerlink}

*A Unit asociates a symbol with a dimension and a power of ten. For
non-metric units a scale can be provided, as in '1 inch = 0.0254 m'. In
this case there is a degeneracy between the power and the scale which is
best resolved by not using the two together. The offset parameter is
available for units which are not zero-offset, such as farenheit.*

::: {.sd-tab-set .docutils}
Properties

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------+-------------------------------------------------+
| **na | String      | As with constants, units are only referred to   |
| me** |             | within expressions using their symbols, so the  |
|      |             | name is just for readability.                   |
+------+-------------+-------------------------------------------------+
| **   | String      | The symbol is used to refer to this unit inside |
| symb |             | compound expressions coutaining a number and a  |
| ol** |             | unit symbol. Such expressions can only occur on |
|      |             | the right hand side of assignments statements.  |
+------+-------------+-------------------------------------------------+
| *    | String      | Reference to the dimension for this unit        |
| *dim |             |                                                 |
| ensi |             |                                                 |
| on** |             |                                                 |
+------+-------------+-------------------------------------------------+
| *    | int         | Power of ten                                    |
| *pow |             |                                                 |
| er** |             |                                                 |
+------+-------------+-------------------------------------------------+
| *    | double      | Scale, only to be used for scales which are not |
| *sca |             | powers of ten                                   |
| le** |             |                                                 |
+------+-------------+-------------------------------------------------+
| **   | double      | Offset for non zero-offset units                |
| offs |             |                                                 |
| et** |             |                                                 |
+------+-------------+-------------------------------------------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="Unit">
      <xs:attribute name="symbol" type="xs:string" use="required"/>
      <xs:attribute name="dimension" type="xs:string" use="required"/>
      <xs:attribute name="power" type="xs:integer" use="optional" default="0">
        <xs:annotation>
          <xs:documentation>Some have asked whether fractional dimensions should be allowed. Disallowing it until needed...</xs:documentation>
        </xs:annotation>
      </xs:attribute>
      <xs:attribute name="scale" type="xs:float" use="optional" default="1"/>
      <xs:attribute name="offset" type="xs:float" use="optional" default="0"/>
    </xs:complexType>
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <Unit symbol="mV" dimension="voltage" power="-3"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Unit symbol="ms" dimension="time" power="-3"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Unit symbol="pS" dimension="conductance" power="-12"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Unit symbol="nS" dimension="conductance" power="-9"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Unit symbol="uF" dimension="capacitance" power="-6"/>
:::
:::
:::
:::
:::

::: {#assertion .section}
[]{#lemsschema-assertion}

##### Assertion[\#](#assertion "Link to this heading"){.headerlink}

*Assertions are not strictly part of the model, but can be included in a
file as a consistency check.*

::: {.sd-tab-set .docutils}
Properties

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------+-------------------------------------------------+
| *    | String      | The name of a dimension                         |
| *dim |             |                                                 |
| ensi |             |                                                 |
| on** |             |                                                 |
+------+-------------+-------------------------------------------------+
| **m  | String      | An expression involving dimensions. The         |
| atch |             | dimensionality of the expression should match   |
| es** |             | the dimensionality of the dimension reference.  |
+------+-------------+-------------------------------------------------+
:::
:::
:::
:::
:::

[]{#document-Userdocs/LEMS_elements/Definingcomponenttypes}

::: {#defining-component-types .section .tex2jax_ignore .mathjax_ignore}
[]{#lemsschema-page-defining-component-types}

#### Defining component types[\#](#defining-component-types "Link to this heading"){.headerlink}

Schema against which LEMS based on these should be valid:
[LEMS_v0.7.6.xsd](https://github.com/LEMS/LEMS/tree/master/Schemas/LEMS/LEMS_v0.7.6.xsd){.reference
.external}. Generated on 18/06/24 from
[this](https://github.com/LEMS/LEMS/commit/fd7b30eceb6735ac343745c8f6992bdde72b248b){.reference
.external} commit. Please file any issues or questions at the [issue
tracker here](https://github.com/LEMS/LEMS/issues){.reference
.external}.

------------------------------------------------------------------------

::: {#componenttype .section}
[]{#lemsschema-componenttype}

##### ComponentType[\#](#componenttype "Link to this heading"){.headerlink}

*Root element for defining LEMS Component Types.*

::: {.sd-tab-set .docutils}
Properties

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------+-------------------------------------------------+
| **na | String      | The name of the component type. This can be     |
| me** |             | uses as an XML element name in the shorthand    |
|      |             | form whendefining components.                   |
+------+-------------+-------------------------------------------------+
| **e  | String      | The component type that this type inherits      |
| Xten |             | field definitions for, if any                   |
| ds** |             |                                                 |
+------+-------------+-------------------------------------------------+
:::
:::

can contain these elements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+-------------+--------------------------------------------------------+
| **p         | [[Parameter]{.std                                      |
| arameters** | .std-ref}](#lemsschema-parameter){.reference           |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
| **indexP    | [[IndexParameter]{.std                                 |
| arameters** | .std-ref}](#lemsschema-indexparameter){.reference      |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
| **derivedP  | [[DerivedParameter]{.std                               |
| arameters** | .std-ref}](#lemsschema-derivedparameter){.reference    |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
| **pathP     | [[PathParameter]{.std                                  |
| arameters** | .std-ref}](#lemsschema-pathparameter){.reference       |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
| **req       | [[Requirement]{.std                                    |
| uirements** | .std-ref}](#lemsschema-requirement){.reference         |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
| **c         | [[ComponentRequirement]{.std                           |
| omponentReq | .                                                      |
| uirements** | std-ref}](#lemsschema-componentrequirement){.reference |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
| **          | [[InstanceRequirement]{.std                            |
| instanceReq | .std-ref}](#lemsschema-instancerequirement){.reference |
| uirements** | .internal}                                             |
+-------------+--------------------------------------------------------+
| **          | [[Exposure]{.std                                       |
| exposures** | .std-ref}](#lemsschema-exposure){.reference .internal} |
+-------------+--------------------------------------------------------+
| **childs**  | [[Child]{.std .std-ref}](#lemsschema-child){.reference |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
| **          | [[Children]{.std                                       |
| childrens** | .std-ref}](#lemsschema-children){.reference .internal} |
+-------------+--------------------------------------------------------+
| **links**   | [[Link]{.std .std-ref}](#lemsschema-link){.reference   |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
| *           | [[ComponentReference]{.std                             |
| *componentR | .std-ref}](#lemsschema-componentreference){.reference  |
| eferences** | .internal}                                             |
+-------------+--------------------------------------------------------+
| **com       | [[ComponentTypeReference]{.std                         |
| ponentTypeR | .st                                                    |
| eferences** | d-ref}](#lemsschema-componenttypereference){.reference |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
| **          | [[Location]{.std                                       |
| locations** | .                                                      |
|             | std-ref}](Landing.html#lemsschema-location){.reference |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
| **          | [[Property]{.std                                       |
| propertys** | .std-ref}](#lemsschema-property){.reference .internal} |
+-------------+--------------------------------------------------------+
| **d         | [[Dynamics]{.std                                       |
| ynamicses** | .                                                      |
|             | std-ref}](Landing.html#lemsschema-dynamics){.reference |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
| **s         | [[Structure]{.std                                      |
| tructures** | .s                                                     |
|             | td-ref}](Landing.html#lemsschema-structure){.reference |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
| **si        | [[Simulation]{.std                                     |
| mulations** | .st                                                    |
|             | d-ref}](Landing.html#lemsschema-simulation){.reference |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
| **equ       | [[Equilibrium]{.std                                    |
| ilibriums** | .std                                                   |
|             | -ref}](Landing.html#lemsschema-equilibrium){.reference |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
| **p         | [[Procedure]{.std                                      |
| rocedures** | .s                                                     |
|             | td-ref}](Landing.html#lemsschema-procedure){.reference |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
| **          | [[Geometry]{.std                                       |
| geometrys** | .                                                      |
|             | std-ref}](Landing.html#lemsschema-geometry){.reference |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
| **fixeds**  | [[Fixed]{.std .std-ref}](#lemsschema-fixed){.reference |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
| **          | [[Constant]{.std                                       |
| constants** | .                                                      |
|             | std-ref}](Landing.html#lemsschema-constant){.reference |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
| **atta      | [[Attachments]{.std                                    |
| chmentses** | .std-ref}](#lemsschema-attachments){.reference         |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
| **e         | [[EventPort]{.std                                      |
| ventPorts** | .std-ref}](#lemsschema-eventport){.reference           |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
| **paths**   | [[Path]{.std .std-ref}](#lemsschema-path){.reference   |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
| **texts**   | [[Text]{.std .std-ref}](#lemsschema-text){.reference   |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
| **co        | [[Collection]{.std                                     |
| llections** | .std-ref}](#lemsschema-collection){.reference          |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
| **pairCo    | [[PairCollection]{.std                                 |
| llections** | .std-ref}](#lemsschema-paircollection){.reference      |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
| **abouts**  | [[About]{.std .std-ref}](#lemsschema-about){.reference |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
| **metas**   | [[Meta]{.std .std-ref}](#lemsschema-meta){.reference   |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="ComponentType">
      <xs:sequence>
        <xs:element name="Property" type="Property" minOccurs="0" maxOccurs="unbounded"/>
        <xs:element name="Parameter" type="Parameter" minOccurs="0" maxOccurs="unbounded"/>
        <xs:element name="DerivedParameter" type="DerivedParameter" minOccurs="0" maxOccurs="unbounded"/>
        <xs:element name="IndexParameter" type="IndexParameter" minOccurs="0" maxOccurs="unbounded"/>
        <xs:element name="Constant" type="Constant" minOccurs="0" maxOccurs="unbounded"/>
        <xs:element name="Child" type="Child" minOccurs="0" maxOccurs="unbounded"/>
        <xs:element name="Children" type="Children" minOccurs="0" maxOccurs="unbounded"/>
        <xs:element name="Fixed" type="Fixed" minOccurs="0" maxOccurs="unbounded"/>
        <xs:element name="Link" type="Link" minOccurs="0" maxOccurs="unbounded"/>
        <xs:element name="ComponentReference" type="ComponentReference" minOccurs="0" maxOccurs="unbounded"/>
        <xs:element name="Attachments" type="Attachments" minOccurs="0" maxOccurs="unbounded"/>
        <xs:element name="EventPort" type="EventPort" minOccurs="0" maxOccurs="unbounded"/>
        <xs:element name="Exposure" type="Exposure" minOccurs="0" maxOccurs="unbounded"/>
        <xs:element name="Requirement" type="Requirement" minOccurs="0" maxOccurs="unbounded"/>
        <xs:element name="ComponentRequirement" type="ComponentRequirement" minOccurs="0" maxOccurs="unbounded"/>
        <xs:element name="InstanceRequirement" type="InstanceRequirement" minOccurs="0" maxOccurs="unbounded"/>
        <xs:element name="Path" type="Path" minOccurs="0" maxOccurs="unbounded"/>
        <xs:element name="Text" type="Text" minOccurs="0" maxOccurs="unbounded"/>
        <xs:element name="Dynamics" type="Dynamics" minOccurs="0" maxOccurs="unbounded"/>
        <xs:element name="Structure" type="Structure" minOccurs="0" maxOccurs="1"/>
        <xs:element name="Simulation" type="Simulation" minOccurs="0" maxOccurs="1"/>
      </xs:sequence>
      <xs:attribute name="name" type="xs:string" use="required"/>
      <xs:attribute name="extends" type="xs:string" use="optional"/>
      <xs:attribute name="description" type="xs:string" use="optional"/>
    </xs:complexType>
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <ComponentType name="Population">
        <ComponentReference name="component" type="Component"/>
        <Parameter name="size" dimension="none"/>
        <Structure>
            <MultiInstantiate number="size" component="component"/>
        </Structure>
    </ComponentType>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <ComponentType name="EventConnectivity">
        <Link name="source" type="Population"/>
        <Link name="target" type="Population"/>
        <Child name="Connections" type="ConnectionPattern"/>
    </ComponentType>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <ComponentType name="Network">
        <Children name="populations" type="Population"/>
        <Children name="connectivities" type="EventConnectivity"/>
    </ComponentType>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <ComponentType name="AllAll" extends="ConnectionPattern">
        <Structure>
            <ForEach instances="../source" as="a">
                <ForEach instances="../target" as="b">
                    <EventConnection from="a" to="b"/>
                </ForEach>
            </ForEach>
        </Structure>
    </ComponentType>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <ComponentType name="ConnectionPattern"/>
:::
:::
:::
:::
:::

::: {#parameter .section}
[]{#lemsschema-parameter}

##### Parameter[\#](#parameter "Link to this heading"){.headerlink}

*A quantity, defined by name and dimension, that must be supplied when a
Component of the enclosing ComponentType is defined*

::: {.sd-tab-set .docutils}
Properties

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------+-------------------------------------------------+
| **na | String      | The name of the parameter. This is the name of  |
| me** |             | the attribute to be used when the parameter is  |
|      |             | supplied in a component definition              |
+------+-------------+-------------------------------------------------+
| *    | String      | The dimension, or 'none'. This should be the    |
| *dim |             | name of an already defined dimension element    |
| ensi |             |                                                 |
| on** |             |                                                 |
+------+-------------+-------------------------------------------------+
| **d  | String      | An optional description of the parameter        |
| escr |             |                                                 |
| ipti |             |                                                 |
| on** |             |                                                 |
+------+-------------+-------------------------------------------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="Parameter">
      <xs:complexContent>
        <xs:extension base="NamedDimensionalType"/>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <Parameter name="size" dimension="none"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Parameter name="xmin" dimension="none"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Parameter name="xmax" dimension="none"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Parameter name="ymin" dimension="none"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Parameter name="ymax" dimension="none"/>
:::
:::
:::
:::
:::

::: {#pathparameter .section}
[]{#lemsschema-pathparameter}

##### PathParameter[\#](#pathparameter "Link to this heading"){.headerlink}

*A parameter of which the value is a path expression. When a
ComponentType declares a PathParameter, a corresponding Component
definition should have an attibute with that name whose value is a path
expression that evaluates within the instance tree of the built model.
This is used, for example, in the definition of a group component class,
where the coresponding component specifies a path over the instance tree
which selectesthe items that should go in the group.*

::: {.sd-tab-set .docutils}
Properties

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------+-------------------------------------------------+
| **na | String      | Name of the parameter                           |
| me** |             |                                                 |
+------+-------------+-------------------------------------------------+
:::
:::
:::
:::

::: {#property .section}
[]{#lemsschema-property}

##### Property[\#](#property "Link to this heading"){.headerlink}

*An property on an instance of a component. Unlike a Parameter, a
Property can very from instance to instance. It should be set with an
Assign element within the build specification.*

::: {.sd-tab-set .docutils}
Properties

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------+-------------------------------------------------+
| **na | String      |                                                 |
| me** |             |                                                 |
+------+-------------+-------------------------------------------------+
| *    | String      |                                                 |
| *dim |             |                                                 |
| ensi |             |                                                 |
| on** |             |                                                 |
+------+-------------+-------------------------------------------------+
| **de | String      | The defaultValue for the property must be a     |
| faul |             | plain number (no units) giving the SI magnitude |
| tVal |             | of the quantity.                                |
| ue** |             |                                                 |
+------+-------------+-------------------------------------------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="Property">
      <xs:attribute name="name" type="xs:string" use="required"/>
      <xs:attribute name="dimension" type="xs:string" use="optional" default="none"/>
      <xs:attribute name="description" type="xs:string" use="optional"/>
      <xs:attribute name="defaultValue" type="xs:double" use="optional"/>
    </xs:complexType>
:::
:::
:::
:::
:::

::: {#derivedparameter .section}
[]{#lemsschema-derivedparameter}

##### DerivedParameter[\#](#derivedparameter "Link to this heading"){.headerlink}

*A parameter that is a function of the Component's Parameters, which
does not change with time. Its value can be supplied either with the
'value' attribute that evaluates within the scope of the definition, or
with the 'select' attribute which gives a path to 'primary' version of
the parameter. For example, setting
select='//MembranePotential\[species=channel/species\]/reversal' within
the appropriate context allows a channel's reversal potential to taken
from a single global setting according to its permeant ion, rather than
explicitly supplied locally.*

::: {.sd-tab-set .docutils}
Properties

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------+-------------------------------------------------+
| **na | String      | The name of the derived parameter               |
| me** |             |                                                 |
+------+-------------+-------------------------------------------------+
| *    | String      | The dimension, or 'none'. This should be the    |
| *dim |             | name of an already defined dimension element    |
| ensi |             |                                                 |
| on** |             |                                                 |
+------+-------------+-------------------------------------------------+
| **d  | String      | An optional description of the derived          |
| escr |             | parameter                                       |
| ipti |             |                                                 |
| on** |             |                                                 |
+------+-------------+-------------------------------------------------+
| **   | String      | Path to the parameter that supplies the value.  |
| sele |             | Exactly one of 'select' and 'value' is          |
| ct** |             | required.                                       |
+------+-------------+-------------------------------------------------+
| *    | String      | A string defining the value of the element      |
| *val |             |                                                 |
| ue** |             |                                                 |
+------+-------------+-------------------------------------------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="DerivedParameter">
      <xs:attribute name="name" type="xs:string" use="required"/>
      <xs:attribute name="dimension" type="xs:string" use="optional" default="none"/>
      <xs:attribute name="value" type="xs:string" use="required"/>
      <xs:attribute name="description" type="xs:string" use="optional"/>
    </xs:complexType>
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <DerivedParameter name="erev" dimension="voltage" select="//MembranePotential[species=channel/species]/reversal"/>
:::
:::
:::
:::
:::

::: {#fixed .section}
[]{#lemsschema-fixed}

##### Fixed[\#](#fixed "Link to this heading"){.headerlink}

*Fixes the value of a parameter in the parent class, so that it does not
have to be supplied separately in component definitions.*

::: {.sd-tab-set .docutils}
Properties

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------+-------------------------------------------------+
| *    | String      |                                                 |
| *par |             |                                                 |
| amet |             |                                                 |
| er** |             |                                                 |
+------+-------------+-------------------------------------------------+
| *    | String      |                                                 |
| *val |             |                                                 |
| ue** |             |                                                 |
+------+-------------+-------------------------------------------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="Fixed">
      <xs:attribute name="parameter" type="xs:string" use="required"/>
      <xs:attribute name="value" type="PhysicalQuantity" use="required"/>
      <xs:attribute name="description" type="xs:string" use="optional"/>
    </xs:complexType>
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <Fixed parameter="threshold" value="-45mV"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Fixed parameter="relativeConductance" value="0"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Fixed parameter="relativeConductance" value="1"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Fixed parameter="relativeConductance" value="0"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Fixed parameter="relativeConductance" value="1"/>
:::
:::
:::
:::
:::

::: {#requirement .section}
[]{#lemsschema-requirement}

##### Requirement[\#](#requirement "Link to this heading"){.headerlink}

*A Requirement gives the name and dimension of a quantity (parameter or
variable) that should be accessible within the scope of a model
component. This is only applicable for elements that can be included as
children of other elements, where the scope comprises its own parameters
and those in the scope of its enclosing element. Once a requirement has
been declared, then the quantity can be used within the Dynamics
definition of the component. It is the responsibility of an
implementation to check that the component is only used in a context in
which the requirement is met. A typical example is in defining membrand
bound components which require access to the membrane potential but
where the variable that holds the potential itself is defined in the top
level component.*

::: {.sd-tab-set .docutils}
Properties

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------+-------------------------------------------------+
| **na | String      | name                                            |
| me** |             |                                                 |
+------+-------------+-------------------------------------------------+
| *    | String      | The dimension, or 'none'. This should be the    |
| *dim |             | name of an already defined dimension element    |
| ensi |             |                                                 |
| on** |             |                                                 |
+------+-------------+-------------------------------------------------+
| **d  | String      | An optional description of the requirement      |
| escr |             |                                                 |
| ipti |             |                                                 |
| on** |             |                                                 |
+------+-------------+-------------------------------------------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="Requirement">
      <xs:complexContent>
        <xs:extension base="NamedDimensionalType"/>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <Requirement name="v" dimension="voltage"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Requirement name="v" dimension="voltage"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Requirement name="v" dimension="voltage"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Requirement name="v" dimension="voltage"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Requirement name="v" dimension="voltage"/>
:::
:::
:::
:::
:::

::: {#componentrequirement .section}
[]{#lemsschema-componentrequirement}

##### ComponentRequirement[\#](#componentrequirement "Link to this heading"){.headerlink}

*The name of a component or component reference that must exist in the
component hierarchy*

::: {.sd-tab-set .docutils}
Properties

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------+-------------------------------------------------+
| **na | String      | name                                            |
| me** |             |                                                 |
+------+-------------+-------------------------------------------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="ComponentRequirement">
      <xs:attribute name="name" type="xs:string" use="required"/>
    </xs:complexType>
:::
:::
:::
:::
:::

::: {#instancerequirement .section}
[]{#lemsschema-instancerequirement}

##### InstanceRequirement[\#](#instancerequirement "Link to this heading"){.headerlink}

*An instance that must be supplied at build time. Expressions can
contain references to quantities in the instance*

::: {.sd-tab-set .docutils}
Properties

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------+-------------------------------------------------+
| **na | String      | name                                            |
| me** |             |                                                 |
+------+-------------+-------------------------------------------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="InstanceRequirement">
      <xs:attribute name="name" type="xs:string" use="required"/>
      <xs:attribute name="type" type="xs:string" use="required"/>
    </xs:complexType>
:::
:::
:::
:::
:::

::: {#exposure .section}
[]{#lemsschema-exposure}

##### Exposure[\#](#exposure "Link to this heading"){.headerlink}

*A quantity that is made available to other components in the
simulation. Note that all variables in a Dynamics definition are
private. If other components need access to them, then the definition
should explicitly link them to an exposure defined in the component
class*

::: {.sd-tab-set .docutils}
Properties

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------+-------------------------------------------------+
| **na | String      | Name of the exposure element                    |
| me** |             |                                                 |
+------+-------------+-------------------------------------------------+
| *    | String      | The dimension, or 'none'. This should be the    |
| *dim |             | name of an already defined dimension element    |
| ensi |             |                                                 |
| on** |             |                                                 |
+------+-------------+-------------------------------------------------+
| **d  | String      | An optional description of the element          |
| escr |             |                                                 |
| ipti |             |                                                 |
| on** |             |                                                 |
+------+-------------+-------------------------------------------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="Exposure">
      <xs:complexContent>
        <xs:extension base="NamedDimensionalType"/>
      </xs:complexContent>
    </xs:complexType>
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <Exposure name="v" dimension="voltage"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Exposure name="tsince" dimension="time"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Exposure name="r" dimension="per_time"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Exposure name="fcond" dimension="none"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Exposure name="fcond" dimension="none"/>
:::
:::
:::
:::
:::

::: {#child .section}
[]{#lemsschema-child}

##### Child[\#](#child "Link to this heading"){.headerlink}

*Specifies that a component can have a child of a particular type. The
name supplied here can be used in path expressions to access the
component. This is useful, for example, where a component can have
multiple children of the same type but with different roles, such as the
forward and reverse transition rates in a channel.*

::: {.sd-tab-set .docutils}
Properties

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------+-------------------------------------------------+
| **na | String      | Name of the child                               |
| me** |             |                                                 |
+------+-------------+-------------------------------------------------+
| **ty | String      | Reference to a component class, the value       |
| pe** |             | should be the name of the target class.         |
+------+-------------+-------------------------------------------------+
| **d  | String      | An optional description of the child            |
| escr |             |                                                 |
| ipti |             |                                                 |
| on** |             |                                                 |
+------+-------------+-------------------------------------------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="Child">
      <xs:attribute name="name" type="xs:string" use="required"/>
      <xs:attribute name="type" type="xs:string" use="required"/>
      <xs:attribute name="description" type="xs:string" use="optional"/>
    </xs:complexType>
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <Child name="Connections" type="ConnectionPattern"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Child name="Forward" type="HHRate"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Child name="Reverse" type="HHRate"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Child name="Forward" type="HHRate"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Child name="Reverse" type="HHRate"/>
:::
:::
:::
:::
:::

::: {#children .section}
[]{#lemsschema-children}

##### Children[\#](#children "Link to this heading"){.headerlink}

*Specifies that a component can have children of a particular class. The
class may refer to an extendedtype, in which case components of any
class that extends the specified target class should be valid as child
components*

::: {.sd-tab-set .docutils}
Properties

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------+-------------------------------------------------+
| **na | String      | Name of the children                            |
| me** |             |                                                 |
+------+-------------+-------------------------------------------------+
| **ty | String      | The class of component allowed as children.     |
| pe** |             |                                                 |
+------+-------------+-------------------------------------------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="Children">
      <xs:attribute name="name" type="xs:string" use="required"/>
      <xs:attribute name="type" type="xs:string" use="optional"/>
      <xs:attribute name="min" type="xs:integer" use="optional"/>
      <xs:attribute name="max" type="xs:integer" use="optional"/>
      <xs:attribute name="description" type="xs:string" use="optional"/>
    </xs:complexType>
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <Children name="populations" type="Population"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Children name="connectivities" type="EventConnectivity"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Children name="lines" type="Line"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Children name="outputColumn" type="OutputColumn"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Children name="displays" type="Display"/>
:::
:::
:::
:::
:::

::: {#link .section}
[]{#lemsschema-link}

##### Link[\#](#link "Link to this heading"){.headerlink}

*Like a ComponentRef, but resolved relative to the enclosing object so
the target must already be in the model. One or the other should be
deprecated. The Link element has the same properties as ComponentRef.
The Link element just establishes a connection with the target
component, but leaves it in its existing place in the hierarchy.
Variables in the target component can be accessed via the name of the
link element.*

::: {.sd-tab-set .docutils}
Properties

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------+-------------------------------------------------+
| **na | String      | A name for the ComponentReference               |
| me** |             |                                                 |
+------+-------------+-------------------------------------------------+
| **ty | String      | The type of the target Component                |
| pe** |             |                                                 |
+------+-------------+-------------------------------------------------+
| **d  | String      | An optional description of the                  |
| escr |             | ComponentReference                              |
| ipti |             |                                                 |
| on** |             |                                                 |
+------+-------------+-------------------------------------------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="Link">
      <xs:attribute name="name" type="xs:string" use="required"/>
      <xs:attribute name="type" type="xs:string" use="required"/>
      <xs:attribute name="description" type="xs:string" use="optional"/>
    </xs:complexType>
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <Link name="source" type="Population"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Link name="target" type="Population"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Link name="from" type="KSState"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Link name="to" type="KSState"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Link name="from" type="KSState"/>
:::
:::
:::
:::
:::

::: {#componentreference .section}
[]{#lemsschema-componentreference}

##### ComponentReference[\#](#componentreference "Link to this heading"){.headerlink}

*A reference to another component. The target component can be accessed
with path expressions in the same way as a child component, but can be
defined independently*

::: {.sd-tab-set .docutils}
Properties

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------+-------------------------------------------------+
| **na | String      | A name for the ComponentReference               |
| me** |             |                                                 |
+------+-------------+-------------------------------------------------+
| **ty | String      | The type of the target Component                |
| pe** |             |                                                 |
+------+-------------+-------------------------------------------------+
| **d  | String      | An optional description of the                  |
| escr |             | ComponentReference                              |
| ipti |             |                                                 |
| on** |             |                                                 |
+------+-------------+-------------------------------------------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="ComponentReference">
      <xs:attribute name="name" type="xs:string" use="required"/>
      <xs:attribute name="type" type="xs:string" use="required"/>
      <xs:attribute name="local" type="xs:string" use="optional"/>
      <xs:attribute name="description" type="xs:string" use="optional"/>
    </xs:complexType>
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <ComponentReference name="component" type="Component"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <ComponentReference name="target" type="Component"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <ComponentReference name="channel" type="HHChannel"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <ComponentReference name="component" type="Component"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <ComponentReference name="synapse" type="Synapse"/>
:::
:::
:::
:::
:::

::: {#componenttypereference .section}
[]{#lemsschema-componenttypereference}

##### ComponentTypeReference[\#](#componenttypereference "Link to this heading"){.headerlink}

*This is used in conjunction with PathParameter elements to specify the
target class of selections defined within components operating over the
instance tree.*

::: {.sd-tab-set .docutils}
Properties

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------+-------------------------------------------------+
| **na | String      |                                                 |
| me** |             |                                                 |
+------+-------------+-------------------------------------------------+
:::
:::
:::
:::

::: {#collection .section}
[]{#lemsschema-collection}

##### Collection[\#](#collection "Link to this heading"){.headerlink}

*Specifies that instances of components based on this class can containe
a named collection of other instances. This provides for containers for
oprating on groups of instances with path and filter expressions defined
in components to operate over the instance tree.*

::: {.sd-tab-set .docutils}
Properties

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------+-------------------------------------------------+
| **na | String      |                                                 |
| me** |             |                                                 |
+------+-------------+-------------------------------------------------+
:::
:::
:::
:::

::: {#paircollection .section}
[]{#lemsschema-paircollection}

##### PairCollection[\#](#paircollection "Link to this heading"){.headerlink}

*Defines a named collection of paris of instances, similar to the
Collection element.*

::: {.sd-tab-set .docutils}
Properties

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------+-------------------------------------------------+
| **na | String      |                                                 |
| me** |             |                                                 |
+------+-------------+-------------------------------------------------+
:::
:::
:::
:::

::: {#eventport .section}
[]{#lemsschema-eventport}

##### EventPort[\#](#eventport "Link to this heading"){.headerlink}

*A port on a component that can send or receive events, depending on the
direction specified*

::: {.sd-tab-set .docutils}
Properties

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------+-------------------------------------------------+
| **na | String      | Name of the EventPort                           |
| me** |             |                                                 |
+------+-------------+-------------------------------------------------+
| *    | String      | 'IN' or 'OUT'                                   |
| *dir |             |                                                 |
| ecti |             |                                                 |
| on** |             |                                                 |
+------+-------------+-------------------------------------------------+
| **d  | String      | An optional description of the EventPort        |
| escr |             |                                                 |
| ipti |             |                                                 |
| on** |             |                                                 |
+------+-------------+-------------------------------------------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="EventPort">
      <xs:attribute name="name" type="xs:string" use="required"/>
      <xs:attribute name="direction" type="xs:string" use="required"/>
      <xs:attribute name="description" type="xs:string" use="optional"/>
    </xs:complexType>
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <EventPort name="spikes-in" direction="in"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <EventPort name="a" direction="out"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <EventPort name="in" direction="in"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <EventPort name="out" direction="out"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <EventPort name="in" direction="in"/>
:::
:::
:::
:::
:::

::: {#text .section}
[]{#lemsschema-text}

##### Text[\#](#text "Link to this heading"){.headerlink}

*Holds textual information that does not change the model but is needed
for other purposes such as labelling graphs.*

::: {.sd-tab-set .docutils}
Properties

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------+-------------------------------------------------+
| **na | String      | The textual content                             |
| me** |             |                                                 |
+------+-------------+-------------------------------------------------+
| **d  | String      | An optional description of the element          |
| escr |             |                                                 |
| ipti |             |                                                 |
| on** |             |                                                 |
+------+-------------+-------------------------------------------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="Text">
      <xs:attribute name="name" type="xs:string" use="required"/>
      <xs:attribute name="description" type="xs:string" use="optional"/>
    </xs:complexType>
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <Text name="title"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Text name="color"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Text name="path"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Text name="fileName"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Text name="destination"/>
:::
:::
:::
:::
:::

::: {#path .section}
[]{#lemsschema-path}

##### Path[\#](#path "Link to this heading"){.headerlink}

*Duplicates some functionality of PathParameter - the two should be
merged.*

::: {.sd-tab-set .docutils}
Properties

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------+-------------------------------------------------+
| **na | String      |                                                 |
| me** |             |                                                 |
+------+-------------+-------------------------------------------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="Path">
      <xs:attribute name="name" type="xs:string" use="required"/>
      <xs:attribute name="description" type="xs:string" use="optional"/>
    </xs:complexType>
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <Path name="quantity"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Path name="quantity"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Path name="from"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Path name="to"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Path name="quantity"/>
:::
:::
:::
:::
:::

::: {#attachments .section}
[]{#lemsschema-attachments}

##### Attachments[\#](#attachments "Link to this heading"){.headerlink}

*Specifies that a component can accept attached components of a
particular class. Attached components can be added at build time
dependent on other events. For scoping and access purposes they are like
child components. The cannonical use of attachments is in adding
synapses to a cell when a network connection is made.*

::: {.sd-tab-set .docutils}
Properties

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------+-------------------------------------------------+
| **na | String      | A name for the Attachments                      |
| me** |             |                                                 |
+------+-------------+-------------------------------------------------+
| **ty | String      | The type of the Attachments                     |
| pe** |             |                                                 |
+------+-------------+-------------------------------------------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="Attachments">
      <xs:attribute name="name" type="xs:string" use="required"/>
      <xs:attribute name="type" type="xs:string" use="required"/>
      <xs:attribute name="description" type="xs:string" use="optional"/>
    </xs:complexType>
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <Attachments name="synapses" type="Synapse"/>
:::
:::
:::
:::
:::

::: {#insertion .section}
[]{#lemsschema-insertion}

##### Insertion[\#](#insertion "Link to this heading"){.headerlink}
:::

::: {#integerparameter .section}
[]{#lemsschema-integerparameter}

##### IntegerParameter[\#](#integerparameter "Link to this heading"){.headerlink}

::: {.sd-tab-set .docutils}
Properties

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------+-------------------------------------------------+
| **na | String      | The name of the parameter. This is the name of  |
| me** |             | the attribute to be used when the parameter is  |
|      |             | supplied in a component definition              |
+------+-------------+-------------------------------------------------+
| *    | String      | The dimension, or 'none'. This should be the    |
| *dim |             | name of an already defined dimension element    |
| ensi |             |                                                 |
| on** |             |                                                 |
+------+-------------+-------------------------------------------------+
| **d  | String      | An optional description of the parameter        |
| escr |             |                                                 |
| ipti |             |                                                 |
| on** |             |                                                 |
+------+-------------+-------------------------------------------------+
:::
:::
:::
:::

::: {#indexparameter .section}
[]{#lemsschema-indexparameter}

##### IndexParameter[\#](#indexparameter "Link to this heading"){.headerlink}

::: {.sd-tab-set .docutils}
Properties

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------+-------------------------------------------------+
| **na | String      | The name of the parameter. This is the name of  |
| me** |             | the attribute to be used when the parameter is  |
|      |             | supplied in a component definition              |
+------+-------------+-------------------------------------------------+
| *    | String      | The dimension, or 'none'. This should be the    |
| *dim |             | name of an already defined dimension element    |
| ensi |             |                                                 |
| on** |             |                                                 |
+------+-------------+-------------------------------------------------+
| **d  | String      | An optional description of the parameter        |
| escr |             |                                                 |
| ipti |             |                                                 |
| on** |             |                                                 |
+------+-------------+-------------------------------------------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="IndexParameter">
      <xs:attribute name="name" type="xs:string" use="required"/>
    </xs:complexType>
:::
:::
:::
:::
:::

::: {#about .section}
[]{#lemsschema-about}

##### About[\#](#about "Link to this heading"){.headerlink}
:::

::: {#meta .section}
[]{#lemsschema-meta}

##### Meta[\#](#meta "Link to this heading"){.headerlink}

*Meta element to provide arbitrary metadata to LEMS simulations. Note
that this is not processed by the LEMS interpreter.*

::: {.sd-tab-set .docutils}
Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="Meta">
      <xs:sequence>
        <xs:any minOccurs="0" maxOccurs="unbounded" processContents="lax"/>
      </xs:sequence>
      <xs:anyAttribute processContents="skip"/>
    </xs:complexType>
:::
:::
:::
:::
:::
:::

[]{#document-Userdocs/LEMS_elements/Dynamics}

::: {#dynamics .section .tex2jax_ignore .mathjax_ignore}
[]{#lemsschema-page-dynamics}

#### Dynamics[\#](#dynamics "Link to this heading"){.headerlink}

Schema against which LEMS based on these should be valid:
[LEMS_v0.7.6.xsd](https://github.com/LEMS/LEMS/tree/master/Schemas/LEMS/LEMS_v0.7.6.xsd){.reference
.external}. Generated on 18/06/24 from
[this](https://github.com/LEMS/LEMS/commit/fd7b30eceb6735ac343745c8f6992bdde72b248b){.reference
.external} commit. Please file any issues or questions at the [issue
tracker here](https://github.com/LEMS/LEMS/issues){.reference
.external}.

------------------------------------------------------------------------

::: {#lemsschema-dynamics .section}
[]{#id1}

##### Dynamics[\#](#lemsschema-dynamics "Link to this heading"){.headerlink}

*Specifies the dynamical behavior of components build from this
ComponentType. Note that all variables in a Dynamics definition are
private. If other components need access to them, then the definition
should explicitly link them to an Exposure defined in the component
class*

::: {.sd-tab-set .docutils}
can contain these elements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+-------------+--------------------------------------------------------+
| **supers**  | [[Super]{.std .std-ref}](#lemsschema-super){.reference |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
| **derived   | [[DerivedVariable]{.std                                |
| Variables** | .std-ref}](#lemsschema-derivedvariable){.reference     |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
| **conditi   | [[ConditionalDerivedVariable]{.std                     |
| onalDerived | .std-re                                                |
| Variables** | f}](#lemsschema-conditionalderivedvariable){.reference |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
| **state     | [[StateVariable]{.std                                  |
| Variables** | .std-ref}](#lemsschema-statevariable){.reference       |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
| **timeDe    | [[TimeDerivative]{.std                                 |
| rivatives** | .std-ref}](#lemsschema-timederivative){.reference      |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
| **kinet     | [[KineticScheme]{.std                                  |
| icSchemes** | .std-ref}](#lemsschema-kineticscheme){.reference       |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
| *           | [[OnStart]{.std                                        |
| *onStarts** | .std-ref}](#lemsschema-onstart){.reference .internal}  |
+-------------+--------------------------------------------------------+
| *           | [[OnEvent]{.std                                        |
| *onEvents** | .std-ref}](#lemsschema-onevent){.reference .internal}  |
+-------------+--------------------------------------------------------+
| **onC       | [[OnCondition]{.std                                    |
| onditions** | .std-ref}](#lemsschema-oncondition){.reference         |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
| **stateSca  | [[StateScalarField]{.std                               |
| larFields** | .std-ref}](#lemsschema-statescalarfield){.reference    |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
| *           | [[DerivedScalarField]{.std                             |
| *derivedSca | .std-ref}](#lemsschema-derivedscalarfield){.reference  |
| larFields** | .internal}                                             |
+-------------+--------------------------------------------------------+
| **d         | [[DerivedPunctateField]{.std                           |
| erivedPunct | .                                                      |
| ateFields** | std-ref}](#lemsschema-derivedpunctatefield){.reference |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
| **regimes** | [[Regime]{.std                                         |
|             | .std-ref}](#lemsschema-regime){.reference .internal}   |
+-------------+--------------------------------------------------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="Dynamics">
      <xs:sequence>
        <xs:element name="StateVariable" type="StateVariable" minOccurs="0" maxOccurs="unbounded"/>
        <xs:element name="DerivedVariable" type="DerivedVariable" minOccurs="0" maxOccurs="unbounded"/>
        <xs:element name="ConditionalDerivedVariable" type="ConditionalDerivedVariable" minOccurs="0" maxOccurs="unbounded"/>
        <xs:element name="TimeDerivative" type="TimeDerivative" minOccurs="0" maxOccurs="unbounded"/>
        <xs:element name="OnStart" type="OnStart" minOccurs="0" maxOccurs="1"/>
        <xs:element name="OnEvent" type="OnEvent" minOccurs="0" maxOccurs="unbounded"/>
        <xs:element name="OnCondition" type="OnCondition" minOccurs="0" maxOccurs="unbounded"/>
        <xs:element name="Regime" type="Regime" minOccurs="0" maxOccurs="unbounded"/>
        <xs:element name="KineticScheme" type="KineticScheme" minOccurs="0" maxOccurs="1"/>
      </xs:sequence>
    </xs:complexType>
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <Dynamics>
        <StateVariable name="t" dimension="time"/>
    </Dynamics>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Dynamics>
        <StateVariable name="x" dimension="none"/>
        <DerivedVariable name="ex" dimension="none" value="exp(x)"/>
        <DerivedVariable name="q" dimension="none" value="ex / (1 + ex)"/>
        <DerivedVariable name="rf" dimension="per_time" select="Forward/r"/>
        <DerivedVariable name="rr" dimension="per_time" select="Reverse/r"/>
        <TimeDerivative variable="x" value="(1 + ex)^2 / ex * (rf * (1 - q) - rr * q)"/>
        <DerivedVariable name="fcond" dimension="none" exposure="fcond" value="q^power"/>
    </Dynamics>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Dynamics>
        <StateVariable name="q" dimension="none"/>
        <DerivedVariable dimension="per_time" name="rf" select="Forward/r"/>
        <DerivedVariable dimension="per_time" name="rr" select="Reverse/r"/>
        <TimeDerivative variable="q" value="rf * (1 - q) - rr * q"/>
        <DerivedVariable name="fcond" dimension="none" exposure="fcond" value="q^power"/>
    </Dynamics>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Dynamics>
        <OnStart>
            <StateAssignment variable="v" value="v0"/>
        </OnStart>
        <DerivedVariable name="totcurrent" dimension="current" select="populations[*]/current" reduce="add"/>
        <StateVariable name="v" exposure="v" dimension="voltage"/>
        <TimeDerivative variable="v" value="(totcurrent + injection) / capacitance"/>
    </Dynamics>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Dynamics>
        <StateVariable name="v" exposure="v" dimension="voltage"/>
        <TimeDerivative variable="v" value="leakConductance * (leakReversal - v) / capacitance"/>
        <OnEvent port="spikes-in">
            <StateAssignment variable="v" value="v + deltaV"/>
        </OnEvent>
    </Dynamics>
:::
:::
:::
:::
:::

::: {#statevariable .section}
[]{#lemsschema-statevariable}

##### StateVariable[\#](#statevariable "Link to this heading"){.headerlink}

*Specifies a state variable that stores values that define the state of
the system at a point in time. The dynamics of state variables may be
defined as Ordinary Differential Equations (ODEs) using the
*TimeDerivative* entity. They can also be discontinuously modified using
elements in *Dynamics* such as *OnStart*s, *OnEvent*s, and
*OnCondition*s. StateVariables may be linked to exposures.*

::: {.sd-tab-set .docutils}
Properties

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------+-------------------------------------------------+
| **na | String      | Name of the state variable                      |
| me** |             |                                                 |
+------+-------------+-------------------------------------------------+
| *    | String      | The dimension, or 'none'. This should be the    |
| *dim |             | name of an already defined dimension element    |
| ensi |             |                                                 |
| on** |             |                                                 |
+------+-------------+-------------------------------------------------+
| **ex | String      | If this variable is to be accessed from         |
| posu |             | outside, it should be linked to an Exposure     |
| re** |             | that is defined in the ComponentType.           |
+------+-------------+-------------------------------------------------+
| **d  | String      | An optional description of the state variable   |
| escr |             |                                                 |
| ipti |             |                                                 |
| on** |             |                                                 |
+------+-------------+-------------------------------------------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="StateVariable">
      <xs:attribute name="name" type="xs:string" use="required"/>
      <xs:attribute name="dimension" type="xs:string" use="optional" default="none"/>
      <xs:attribute name="exposure" type="xs:string" use="optional"/>
      <xs:attribute name="description" type="xs:string" use="optional"/>
    </xs:complexType>
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <StateVariable name="t" dimension="time"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <StateVariable name="v" exposure="v" dimension="voltage"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <StateVariable name="tsince" exposure="tsince" dimension="time"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <StateVariable name="tlast" dimension="time"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <StateVariable name="q" dimension="none"/>
:::
:::
:::
:::
:::

::: {#stateassignment .section}
[]{#lemsschema-stateassignment}

##### StateAssignment[\#](#stateassignment "Link to this heading"){.headerlink}

*Has 'variable' and 'value' fields*

::: {.sd-tab-set .docutils}
Properties

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------+-------------------------------------------------+
| **va | String      | The name of the variable                        |
| riab |             |                                                 |
| le** |             |                                                 |
+------+-------------+-------------------------------------------------+
| *    | String      | A string defining the value of the element      |
| *val |             |                                                 |
| ue** |             |                                                 |
+------+-------------+-------------------------------------------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="StateAssignment">
      <xs:attribute name="variable" type="xs:string" use="required"/>
      <xs:attribute name="value" type="xs:string" use="required"/>
    </xs:complexType>
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <StateAssignment variable="v" value="v + deltaV"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <StateAssignment variable="tsince" value="0"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <StateAssignment variable="tlast" value="t"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <StateAssignment variable="v" value="v0"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <StateAssignment variable="geff" value="0"/>
:::
:::
:::
:::
:::

::: {#timederivative .section}
[]{#lemsschema-timederivative}

##### TimeDerivative[\#](#timederivative "Link to this heading"){.headerlink}

*First order differential equations, functions of StateVariables and
Parameters, for how StateVariables change with time. Has a variable and
a value. The value is the rate of change of the variable.*

::: {.sd-tab-set .docutils}
Properties

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------+-------------------------------------------------+
| **va | String      | The name of the variable                        |
| riab |             |                                                 |
| le** |             |                                                 |
+------+-------------+-------------------------------------------------+
| *    | String      | A string defining the value of the element      |
| *val |             |                                                 |
| ue** |             |                                                 |
+------+-------------+-------------------------------------------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="TimeDerivative">
      <xs:attribute name="variable" type="xs:string" use="required"/>
      <xs:attribute name="value" type="xs:string" use="required"/>
    </xs:complexType>
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <TimeDerivative variable="v" value="leakConductance * (leakReversal - v) / capacitance"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <TimeDerivative variable="tsince" value="1"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <TimeDerivative variable="q" value="rf * (1 - q) - rr * q"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <TimeDerivative variable="x" value="(1 + ex)^2 / ex * (rf * (1 - q) - rr * q)"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <TimeDerivative variable="v" value="(totcurrent + injection) / capacitance"/>
:::
:::
:::
:::
:::

::: {#derivedvariable .section}
[]{#lemsschema-derivedvariable}

##### DerivedVariable[\#](#derivedvariable "Link to this heading"){.headerlink}

*A quantity that depends algebraically on other quantities in the model.
The 'value' field can be set to a mathematical expression, or the
'select' field to a path expression. If the path expression produces
multiple matches, then the 'reduce' field says how these are reduced to
a single value by taking the sum or product.*

::: {.sd-tab-set .docutils}
Properties

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------+-------------------------------------------------+
| **na | String      | Name of the derived variable                    |
| me** |             |                                                 |
+------+-------------+-------------------------------------------------+
| **   | String      | A path to the variable that supplies the value. |
| sele |             | Note that to select a variable from another     |
| ct** |             | component, the variable must be marked as an    |
|      |             | Exposure. Exactly one of 'select' and 'value'   |
|      |             | is required                                     |
+------+-------------+-------------------------------------------------+
| *    | String      | The dimension, or 'none'. This should be the    |
| *dim |             | name of an already defined dimension element    |
| ensi |             |                                                 |
| on** |             |                                                 |
+------+-------------+-------------------------------------------------+
| **d  | String      | An optional description of the derived variable |
| escr |             |                                                 |
| ipti |             |                                                 |
| on** |             |                                                 |
+------+-------------+-------------------------------------------------+
| **   | String      | Either 'add' or 'multiply'. This applies if     |
| redu |             | ther are multiple matches to the path or if     |
| ce** |             | 'required' is false. In the latter case, for    |
|      |             | multiply mode, multiplicative expressions in    |
|      |             | this variable behave as if the term was absent. |
|      |             | Additive expressions generate an error.         |
|      |             | Conversely, if set to 'add' then additive       |
|      |             | expressions behave as if it was not there and   |
|      |             | multiplicative ones generateand error.          |
+------+-------------+-------------------------------------------------+
| **ex | String      |                                                 |
| posu |             |                                                 |
| re** |             |                                                 |
+------+-------------+-------------------------------------------------+
| **re | boolean     | Set to true if it OK for this variable to be    |
| quir |             | absent. See 'reduce' for what happens in this   |
| ed** |             | case                                            |
+------+-------------+-------------------------------------------------+
| *    | String      | A string defining the value of the element      |
| *val |             |                                                 |
| ue** |             |                                                 |
+------+-------------+-------------------------------------------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="DerivedVariable">
      <xs:attribute name="name" type="xs:string" use="required"/>
      <xs:attribute name="dimension" type="xs:string" use="optional" default="none"/>
      <xs:attribute name="exposure" type="xs:string" use="optional"/>
      <xs:attribute name="description" type="xs:string" use="optional"/>
      <xs:attribute name="select" type="xs:string" use="optional"/>
      <xs:attribute name="value" type="xs:string" use="optional"/>
      <xs:attribute name="reduce" type="xs:string" use="optional"/>
      <xs:attribute name="required" type="xs:string" use="optional"/>
    </xs:complexType>
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <DerivedVariable name="tsince" dimension="time" exposure="tsince" value="t - tlast"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <DerivedVariable name="r" dimension="per_time" exposure="r" value="rate * exp((v - midpoint)/scale)"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <DerivedVariable name="r" dimension="per_time" exposure="r" value="rate / (1 + exp( -(v - midpoint)/scale))"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <DerivedVariable name="x" dimension="none" value="(v - midpoint) / scale"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <DerivedVariable name="r" dimension="per_time" exposure="r" value="rate * x / (1 - exp(-x))"/>
:::
:::
:::
:::
:::

::: {#onstart .section}
[]{#lemsschema-onstart}

##### OnStart[\#](#onstart "Link to this heading"){.headerlink}

::: {.sd-tab-set .docutils}
can contain these elements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+-------------+--------------------------------------------------------+
| **stateAs   | [[StateAssignment]{.std                                |
| signments** | .std-ref}](#lemsschema-stateassignment){.reference     |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
| **          | [[EventOut]{.std                                       |
| eventOuts** | .std-ref}](#lemsschema-eventout){.reference .internal} |
+-------------+--------------------------------------------------------+
| **tr        | [[Transition]{.std                                     |
| ansitions** | .std-ref}](#lemsschema-transition){.reference          |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="OnStart">
      <xs:sequence>
        <xs:element name="StateAssignment" type="StateAssignment" minOccurs="1" maxOccurs="unbounded"/>
      </xs:sequence>
    </xs:complexType>
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <OnStart>
        <StateAssignment variable="v" value="v0"/>
    </OnStart>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <OnStart>
        <StateAssignment variable="geff" value="0"/>
    </OnStart>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <OnStart>
        <StateAssignment variable="v" value="v0"/>
    </OnStart>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <OnStart>
        <StateAssignment variable="v" value="v0"/>
    </OnStart>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <OnStart>
        <StateAssignment variable="v" value="v0"/>
    </OnStart>
:::
:::
:::
:::
:::

::: {#oncondition .section}
[]{#lemsschema-oncondition}

##### OnCondition[\#](#oncondition "Link to this heading"){.headerlink}

::: {.sd-tab-set .docutils}
can contain these elements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+-------------+--------------------------------------------------------+
| **stateAs   | [[StateAssignment]{.std                                |
| signments** | .std-ref}](#lemsschema-stateassignment){.reference     |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
| **          | [[EventOut]{.std                                       |
| eventOuts** | .std-ref}](#lemsschema-eventout){.reference .internal} |
+-------------+--------------------------------------------------------+
| **tr        | [[Transition]{.std                                     |
| ansitions** | .std-ref}](#lemsschema-transition){.reference          |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="OnCondition">
      <xs:sequence>
        <xs:element name="StateAssignment" type="StateAssignment" minOccurs="0" maxOccurs="unbounded"/>
        <xs:element name="EventOut" type="EventOut" minOccurs="0" maxOccurs="unbounded"/>
        <xs:element name="Transition" type="Transition" minOccurs="0" maxOccurs="1"/>
      </xs:sequence>
      <xs:attribute name="test" type="xs:string" use="required"/>
    </xs:complexType>
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <OnCondition test="tsince .gt. period">
        <StateAssignment variable="tsince" value="0"/>
        <EventOut port="a"/>
    </OnCondition>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <OnCondition test="t - tlast .gt. period">
        <StateAssignment variable="tlast" value="t"/>
        <EventOut port="a"/>
    </OnCondition>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <OnCondition test="v .gt. threshold">
        <EventOut port="out"/>
        <Transition regime="refr"/>
    </OnCondition>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <OnCondition test="t .gt. tin + refractoryPeriod">
        <Transition regime="int"/>
    </OnCondition>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <OnCondition test="tsince .gt. period">
        <StateAssignment variable="tsince" value="0"/>
        <EventOut port="a"/>
    </OnCondition>
:::
:::
:::
:::
:::

::: {#onevent .section}
[]{#lemsschema-onevent}

##### OnEvent[\#](#onevent "Link to this heading"){.headerlink}

*Event handler block*

::: {.sd-tab-set .docutils}
Properties

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------+-------------------------------------------------+
| **po | String      | the port to listen on                           |
| rt** |             |                                                 |
+------+-------------+-------------------------------------------------+
:::
:::

can contain these elements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+-------------+--------------------------------------------------------+
| **stateAs   | [[StateAssignment]{.std                                |
| signments** | .std-ref}](#lemsschema-stateassignment){.reference     |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
| **          | [[EventOut]{.std                                       |
| eventOuts** | .std-ref}](#lemsschema-eventout){.reference .internal} |
+-------------+--------------------------------------------------------+
| **tr        | [[Transition]{.std                                     |
| ansitions** | .std-ref}](#lemsschema-transition){.reference          |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="OnEvent">
      <xs:sequence>
        <xs:element name="StateAssignment" type="StateAssignment" minOccurs="0" maxOccurs="unbounded"/>
        <xs:element name="EventOut" type="EventOut" minOccurs="0" maxOccurs="unbounded"/>
      </xs:sequence>
      <xs:attribute name="port" type="xs:string" use="required"/>
    </xs:complexType>
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <OnEvent port="spikes-in">
        <StateAssignment variable="v" value="v + deltaV"/>
    </OnEvent>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <OnEvent port="in">
        <StateAssignment variable="geff" value="geff + deltaG"/>
    </OnEvent>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <OnEvent port="in">
        <StateAssignment variable="v" value="v + deltaV"/>
    </OnEvent>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <OnEvent port="spikes-in">
        <StateAssignment variable="v" value="v + deltaV"/>
    </OnEvent>
:::
:::
:::
:::
:::

::: {#eventout .section}
[]{#lemsschema-eventout}

##### EventOut[\#](#eventout "Link to this heading"){.headerlink}

::: {.sd-tab-set .docutils}
Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="EventOut">
      <xs:attribute name="port" type="xs:string" use="required"/>
    </xs:complexType>
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <EventOut port="a"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <EventOut port="a"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <EventOut port="out"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <EventOut port="a"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <EventOut port="a"/>
:::
:::
:::
:::
:::

::: {#kineticscheme .section}
[]{#lemsschema-kineticscheme}

##### KineticScheme[\#](#kineticscheme "Link to this heading"){.headerlink}

*Allows the specification of systems that can be in one of a small
number of states at any time with probabilistic transitions between
states. This includes continuous time Markov processes as are used for
stochastic models of ion channels. A kinetic scheme does not itself
introduce any new elements or state variables. It is rather a way of
connecting quantities in existing components by saying that quantities
in the edge elements should be interpreted as transition rates among
quantities in the node elements.*

::: {.sd-tab-set .docutils}
Properties

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------+-------------------------------------------------+
| **na | String      | Name of kinetic scheme                          |
| me** |             |                                                 |
+------+-------------+-------------------------------------------------+
| *    | String      | Source of notes for scheme                      |
| *nod |             |                                                 |
| es** |             |                                                 |
+------+-------------+-------------------------------------------------+
| *    | String      | The element that provides the transitions for   |
| *edg |             | the scheme                                      |
| es** |             |                                                 |
+------+-------------+-------------------------------------------------+
| *    | String      | Name of state variable in state elements        |
| *sta |             |                                                 |
| teVa |             |                                                 |
| riab |             |                                                 |
| le** |             |                                                 |
+------+-------------+-------------------------------------------------+
| **   | String      | The name of the attribute in the rate element   |
| edge |             | that defines the source of the transition       |
| Sour |             |                                                 |
| ce** |             |                                                 |
+------+-------------+-------------------------------------------------+
| **   | String      | Attribute tha defines the target                |
| edge |             |                                                 |
| Targ |             |                                                 |
| et** |             |                                                 |
+------+-------------+-------------------------------------------------+
| **f  | String      | Name of forward rate exposure                   |
| orwa |             |                                                 |
| rdRa |             |                                                 |
| te** |             |                                                 |
+------+-------------+-------------------------------------------------+
| **r  | String      | Name of reverse rate exposure                   |
| ever |             |                                                 |
| seRa |             |                                                 |
| te** |             |                                                 |
+------+-------------+-------------------------------------------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="KineticScheme">
      <xs:attribute name="name" type="xs:string" use="required"/>
      <xs:attribute name="nodes" type="xs:string" use="required"/>
      <xs:attribute name="stateVariable" type="xs:string" use="required"/>
      <xs:attribute name="edges" type="xs:string" use="required"/>
      <xs:attribute name="edgeSource" type="xs:string" use="required"/>
      <xs:attribute name="edgeTarget" type="xs:string" use="required"/>
      <xs:attribute name="forwardRate" type="xs:string" use="required"/>
      <xs:attribute name="reverseRate" type="xs:string" use="required"/>
    </xs:complexType>
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <KineticScheme name="ks" nodes="states" stateVariable="occupancy" edges="transitions" edgeSource="from" edgeTarget="to" forwardRate="rf" reverseRate="rr"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <KineticScheme name="ks" nodes="states" stateVariable="occupancy" edges="transitions" edgeSource="from" edgeTarget="to" forwardRate="rf" reverseRate="rr" dependency="v" step="deltaV"/>
:::
:::
:::
:::
:::

::: {#regime .section}
[]{#lemsschema-regime}

##### Regime[\#](#regime "Link to this heading"){.headerlink}

*Allows the dynamics of a ComponentType to be expressed via a finite
state machine. Each regime has its internal dynamics, and conditions on
which transitions between regimes occur are specified using the
OnCondition element*

::: {.sd-tab-set .docutils}
Properties

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------+-------------------------------------------------+
| **na | String      | The name of the regime                          |
| me** |             |                                                 |
+------+-------------+-------------------------------------------------+
| **i  | String      | 'True' if this is the initial regime of the     |
| niti |             | system                                          |
| al** |             |                                                 |
+------+-------------+-------------------------------------------------+
:::
:::

can contain these elements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+-------------+--------------------------------------------------------+
| **derived   | [[DerivedVariable]{.std                                |
| Variables** | .std-ref}](#lemsschema-derivedvariable){.reference     |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
| **state     | [[StateVariable]{.std                                  |
| Variables** | .std-ref}](#lemsschema-statevariable){.reference       |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
| **timeDe    | [[TimeDerivative]{.std                                 |
| rivatives** | .std-ref}](#lemsschema-timederivative){.reference      |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
| *           | [[OnStart]{.std                                        |
| *onStarts** | .std-ref}](#lemsschema-onstart){.reference .internal}  |
+-------------+--------------------------------------------------------+
| *           | [[OnEntry]{.std                                        |
| *onEntrys** | .std-ref}](#lemsschema-onentry){.reference .internal}  |
+-------------+--------------------------------------------------------+
| *           | [[OnEvent]{.std                                        |
| *onEvents** | .std-ref}](#lemsschema-onevent){.reference .internal}  |
+-------------+--------------------------------------------------------+
| **onC       | [[OnCondition]{.std                                    |
| onditions** | .std-ref}](#lemsschema-oncondition){.reference         |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
| **req       | [lemsschema:requiredvar\_]{.xref .std .std-ref}        |
| uiredVars** |                                                        |
+-------------+--------------------------------------------------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="Regime">
      <xs:sequence>
        <xs:element name="TimeDerivative" type="TimeDerivative" minOccurs="0" maxOccurs="unbounded"/>
        <xs:element name="OnEntry" type="OnEntry" minOccurs="0" maxOccurs="1"/>
        <xs:element name="OnCondition" type="OnCondition" minOccurs="0" maxOccurs="unbounded"/>
      </xs:sequence>
      <xs:attribute name="name" type="xs:string" use="required"/>
      <xs:attribute name="initial" type="TrueOrFalse" use="optional"/>
    </xs:complexType>
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <Regime name="int" initial="true">
        <TimeDerivative variable="v" value="(current + gleak * (vleak - v)) / capacitance"/>
        <OnCondition test="v .gt. threshold">
            <EventOut port="out"/>
            <Transition regime="refr"/>
        </OnCondition>
        <OnEvent port="in">
            <StateAssignment variable="v" value="v + deltaV"/>
        </OnEvent>
    </Regime>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Regime name="refr">
        <OnEntry>
            <StateAssignment variable="tin" value="t"/>
            <StateAssignment variable="v" value="vreset"/>
        </OnEntry>
        <OnCondition test="t .gt. tin + refractoryPeriod">
            <Transition regime="int"/>
        </OnCondition>
    </Regime>
:::
:::
:::
:::
:::

::: {#onentry .section}
[]{#lemsschema-onentry}

##### OnEntry[\#](#onentry "Link to this heading"){.headerlink}

::: {.sd-tab-set .docutils}
can contain these elements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+-------------+--------------------------------------------------------+
| **stateAs   | [[StateAssignment]{.std                                |
| signments** | .std-ref}](#lemsschema-stateassignment){.reference     |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
| **          | [[EventOut]{.std                                       |
| eventOuts** | .std-ref}](#lemsschema-eventout){.reference .internal} |
+-------------+--------------------------------------------------------+
| **tr        | [[Transition]{.std                                     |
| ansitions** | .std-ref}](#lemsschema-transition){.reference          |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="OnEntry">
      <xs:sequence>
        <xs:element name="StateAssignment" type="StateAssignment" minOccurs="1" maxOccurs="unbounded"/>
      </xs:sequence>
    </xs:complexType>
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <OnEntry>
        <StateAssignment variable="tin" value="t"/>
        <StateAssignment variable="v" value="vreset"/>
    </OnEntry>
:::
:::
:::
:::
:::

::: {#transition .section}
[]{#lemsschema-transition}

##### Transition[\#](#transition "Link to this heading"){.headerlink}

::: {.sd-tab-set .docutils}
Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="Transition">
      <xs:attribute name="regime" type="xs:string" use="required"/>
    </xs:complexType>
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <Transition regime="int"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Transition regime="refr"/>
:::
:::
:::
:::
:::

::: {#super .section}
[]{#lemsschema-super}

##### Super[\#](#super "Link to this heading"){.headerlink}
:::

::: {#conditionalderivedvariable .section}
[]{#lemsschema-conditionalderivedvariable}

##### ConditionalDerivedVariable[\#](#conditionalderivedvariable "Link to this heading"){.headerlink}

::: {.sd-tab-set .docutils}
Properties

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------+-------------------------------------------------+
| **na | String      |                                                 |
| me** |             |                                                 |
+------+-------------+-------------------------------------------------+
| *    | String      |                                                 |
| *dim |             |                                                 |
| ensi |             |                                                 |
| on** |             |                                                 |
+------+-------------+-------------------------------------------------+
| **ex | String      |                                                 |
| posu |             |                                                 |
| re** |             |                                                 |
+------+-------------+-------------------------------------------------+
:::
:::

can contain these elements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+-------------+--------------------------------------------------------+
| **cases**   | [[Case]{.std .std-ref}](#lemsschema-case){.reference   |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="ConditionalDerivedVariable">
      <xs:sequence>
        <xs:element name="Case" type="Case" minOccurs="1" maxOccurs="unbounded"/>
      </xs:sequence>
      <xs:attribute name="name" type="xs:string" use="required"/>
      <xs:attribute name="dimension" type="xs:string" use="optional" default="none"/>
      <xs:attribute name="exposure" type="xs:string" use="optional"/>
    </xs:complexType>
:::
:::
:::
:::
:::

::: {#case .section}
[]{#lemsschema-case}

##### Case[\#](#case "Link to this heading"){.headerlink}

::: {.sd-tab-set .docutils}
Properties

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------+-------------------------------------------------+
| *    | String      | A string defining the value of the element      |
| *val |             |                                                 |
| ue** |             |                                                 |
+------+-------------+-------------------------------------------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="Case">
      <xs:attribute name="condition" type="xs:string" use="optional"/>
      <xs:attribute name="value" type="xs:string" use="required"/>
    </xs:complexType>
:::
:::
:::
:::
:::

::: {#equilibrium .section}
[]{#lemsschema-equilibrium}

##### Equilibrium[\#](#equilibrium "Link to this heading"){.headerlink}

::: {.sd-tab-set .docutils}
can contain these elements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+-------------+--------------------------------------------------------+
| **derived   | [[DerivedVariable]{.std                                |
| Variables** | .std-ref}](#lemsschema-derivedvariable){.reference     |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
:::
:::
:::
:::

::: {#statescalarfield .section}
[]{#lemsschema-statescalarfield}

##### StateScalarField[\#](#statescalarfield "Link to this heading"){.headerlink}
:::

::: {#derivedscalarfield .section}
[]{#lemsschema-derivedscalarfield}

##### DerivedScalarField[\#](#derivedscalarfield "Link to this heading"){.headerlink}
:::

::: {#derivedpunctatefield .section}
[]{#lemsschema-derivedpunctatefield}

##### DerivedPunctateField[\#](#derivedpunctatefield "Link to this heading"){.headerlink}
:::
:::

[]{#document-Userdocs/LEMS_elements/Structure}

::: {#structure .section .tex2jax_ignore .mathjax_ignore}
[]{#lemsschema-page-structure}

#### Structure[\#](#structure "Link to this heading"){.headerlink}

Schema against which LEMS based on these should be valid:
[LEMS_v0.7.6.xsd](https://github.com/LEMS/LEMS/tree/master/Schemas/LEMS/LEMS_v0.7.6.xsd){.reference
.external}. Generated on 18/06/24 from
[this](https://github.com/LEMS/LEMS/commit/fd7b30eceb6735ac343745c8f6992bdde72b248b){.reference
.external} commit. Please file any issues or questions at the [issue
tracker here](https://github.com/LEMS/LEMS/issues){.reference
.external}.

------------------------------------------------------------------------

::: {#lemsschema-structure .section}
[]{#id1}

##### Structure[\#](#lemsschema-structure "Link to this heading"){.headerlink}

*By default, each Component in a model gives rise to a single instance
of its state variables when the model is executed. The state variables
are then governed by the dynamics definition in the associated
ComponentType. Elements in the Structure declaration can be used to
change this behavior, for example to make multiple instances of the
state variables, or to instantiate a different component. A typical
application for the latter would be a Component that defines a
population of cells. The population Component might define the number of
cells it contains but would refer to a Component defined elsewhere for
the actual cell model to use.*

::: {.sd-tab-set .docutils}
can contain these elements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+-------------+--------------------------------------------------------+
| **buil      | [[BuildElement]{.std                                   |
| dElements** | .std-ref}](#lemsschema-buildelement){.reference        |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="Structure">
      <xs:sequence>
        <xs:element name="ChildInstance" type="ChildInstance" minOccurs="0" maxOccurs="unbounded"/>
        <xs:element name="MultiInstantiate" type="MultiInstantiate" minOccurs="0" maxOccurs="1"/>
        <xs:element name="ForEach" type="ForEach" minOccurs="0" maxOccurs="1"/>
        <xs:element name="With" type="With" minOccurs="0" maxOccurs="unbounded"/>
        <xs:element name="Tunnel" type="Tunnel" minOccurs="0" maxOccurs="1"/>
        <xs:element name="EventConnection" type="EventConnection" minOccurs="0" maxOccurs="unbounded"/>
      </xs:sequence>
    </xs:complexType>
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <Structure>
        <MultiInstantiate number="size" component="component"/>
    </Structure>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Structure>
        <ForEach instances="../source" as="a">
            <ForEach instances="../target" as="b">
                <EventConnection from="a" to="b"/>
            </ForEach>
        </ForEach>
    </Structure>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Structure>
        <ChildInstance component="channel"/>
    </Structure>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Structure>
        <With instance="from" as="a"/>
        <With instance="to" as="b"/>
        <EventConnection from="a" to="b" receiver="synapse" receiverContainer="destination" sourcePort="sourcePort" targetPort="targetPort"/>
    </Structure>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Structure>
        <MultiInstantiate number="size" component="component"/>
    </Structure>
:::
:::
:::
:::
:::

::: {#buildelement .section}
[]{#lemsschema-buildelement}

##### BuildElement[\#](#buildelement "Link to this heading"){.headerlink}

*Base class for elements that can be used in Structures*

::: {.sd-tab-set .docutils}
can contain these elements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+-------------+--------------------------------------------------------+
| **buil      | [[BuildElement]{.std                                   |
| dElements** | .std-ref}](#lemsschema-buildelement){.reference        |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
:::
:::
:::
:::

::: {#multiinstantiate .section}
[]{#lemsschema-multiinstantiate}

##### MultiInstantiate[\#](#multiinstantiate "Link to this heading"){.headerlink}

::: {.sd-tab-set .docutils}
can contain these elements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+-------------+--------------------------------------------------------+
| **assigns** | [[Assign]{.std                                         |
|             | .std-ref}](#lemsschema-assign){.reference .internal}   |
+-------------+--------------------------------------------------------+
| **buil      | [[BuildElement]{.std                                   |
| dElements** | .std-ref}](#lemsschema-buildelement){.reference        |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="MultiInstantiate">
      <xs:attribute name="component" type="xs:string" use="required"/>
      <xs:attribute name="number" type="xs:string" use="required"/>
    </xs:complexType>
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <MultiInstantiate number="size" component="component"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <MultiInstantiate number="size" component="component"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <MultiInstantiate number="size" component="component"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <MultiInstantiate number="size" component="component"/>
:::
:::
:::
:::
:::

::: {#coinstantiate .section}
[]{#lemsschema-coinstantiate}

##### CoInstantiate[\#](#coinstantiate "Link to this heading"){.headerlink}

::: {.sd-tab-set .docutils}
can contain these elements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+-------------+--------------------------------------------------------+
| **assigns** | [[Assign]{.std                                         |
|             | .std-ref}](#lemsschema-assign){.reference .internal}   |
+-------------+--------------------------------------------------------+
| **buil      | [[BuildElement]{.std                                   |
| dElements** | .std-ref}](#lemsschema-buildelement){.reference        |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
:::
:::
:::
:::

::: {#assign .section}
[]{#lemsschema-assign}

##### Assign[\#](#assign "Link to this heading"){.headerlink}

::: {.sd-tab-set .docutils}
Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="Assign">
      <xs:attribute name="property" type="xs:string" use="required"/>
      <xs:attribute name="value" type="xs:string" use="required"/>
    </xs:complexType>
:::
:::
:::
:::
:::

::: {#choose .section}
[]{#lemsschema-choose}

##### Choose[\#](#choose "Link to this heading"){.headerlink}

::: {.sd-tab-set .docutils}
can contain these elements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+-------------+--------------------------------------------------------+
| **buil      | [[BuildElement]{.std                                   |
| dElements** | .std-ref}](#lemsschema-buildelement){.reference        |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
:::
:::
:::
:::

::: {#childinstance .section}
[]{#lemsschema-childinstance}

##### ChildInstance[\#](#childinstance "Link to this heading"){.headerlink}

::: {.sd-tab-set .docutils}
can contain these elements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+-------------+--------------------------------------------------------+
| **assigns** | [[Assign]{.std                                         |
|             | .std-ref}](#lemsschema-assign){.reference .internal}   |
+-------------+--------------------------------------------------------+
| **buil      | [[BuildElement]{.std                                   |
| dElements** | .std-ref}](#lemsschema-buildelement){.reference        |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="ChildInstance">
      <xs:attribute name="component" type="xs:string" use="required"/>
    </xs:complexType>
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <ChildInstance component="channel"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <ChildInstance component="channel"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <ChildInstance component="channel"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <ChildInstance component="channel"/>
:::
:::
:::
:::
:::

::: {#foreach .section}
[]{#lemsschema-foreach}

##### ForEach[\#](#foreach "Link to this heading"){.headerlink}

::: {.sd-tab-set .docutils}
can contain these elements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+-------------+--------------------------------------------------------+
| **buil      | [[BuildElement]{.std                                   |
| dElements** | .std-ref}](#lemsschema-buildelement){.reference        |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="ForEach">
      <xs:sequence>
        <xs:element name="MultiInstantiate" type="MultiInstantiate" minOccurs="0" maxOccurs="1"/>
      </xs:sequence>
      <xs:attribute name="instances" type="xs:string" use="required"/>
      <xs:attribute name="as" type="xs:string" use="required"/>
    </xs:complexType>
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <ForEach instances="../source" as="a">
        <ForEach instances="../target" as="b">
            <EventConnection from="a" to="b"/>
        </ForEach>
    </ForEach>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <ForEach instances="../target" as="b">
        <EventConnection from="a" to="b"/>
    </ForEach>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <ForEach instances="../source" as="a">
        <ForEach instances="../target" as="b">
            <EventConnection from="a" to="b"/>
        </ForEach>
    </ForEach>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <ForEach instances="../target" as="b">
        <EventConnection from="a" to="b"/>
    </ForEach>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <ForEach instances="../source" as="a">
        <ForEach instances="../target" as="b">
            <EventConnection from="a" to="b"/>
        </ForEach>
    </ForEach>
:::
:::
:::
:::
:::

::: {#eventconnection .section}
[]{#lemsschema-eventconnection}

##### EventConnection[\#](#eventconnection "Link to this heading"){.headerlink}

::: {.sd-tab-set .docutils}
can contain these elements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+-------------+--------------------------------------------------------+
| **assigns** | [[Assign]{.std                                         |
|             | .std-ref}](#lemsschema-assign){.reference .internal}   |
+-------------+--------------------------------------------------------+
| **buil      | [[BuildElement]{.std                                   |
| dElements** | .std-ref}](#lemsschema-buildelement){.reference        |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="EventConnection">
      <xs:sequence>
        <xs:element name="Assign" type="Assign" minOccurs="0" maxOccurs="1"/>
      </xs:sequence>
      <xs:attribute name="from" type="xs:string" use="required"/>
      <xs:attribute name="to" type="xs:string" use="required"/>
      <xs:attribute name="sourcePort" type="xs:string" use="optional"/>
      <xs:attribute name="targetPort" type="xs:string" use="optional"/>
      <xs:attribute name="receiver" type="xs:string" use="optional"/>
      <xs:attribute name="receiverContainer" type="xs:string" use="optional"/>
      <xs:attribute name="delay" type="xs:string" use="optional"/>
    </xs:complexType>
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <EventConnection from="a" to="b"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <EventConnection from="a" to="b" receiver="synapse" receiverContainer="destination" sourcePort="sourcePort" targetPort="targetPort"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <EventConnection from="a" to="b"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <EventConnection from="a" to="b"/>
:::
:::
:::
:::
:::

::: {#tunnel .section}
[]{#lemsschema-tunnel}

##### Tunnel[\#](#tunnel "Link to this heading"){.headerlink}

::: {.sd-tab-set .docutils}
can contain these elements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+-------------+--------------------------------------------------------+
| **assigns** | [[Assign]{.std                                         |
|             | .std-ref}](#lemsschema-assign){.reference .internal}   |
+-------------+--------------------------------------------------------+
| **buil      | [[BuildElement]{.std                                   |
| dElements** | .std-ref}](#lemsschema-buildelement){.reference        |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="Tunnel">
      <xs:sequence>
        <xs:element name="Assign" type="Assign" minOccurs="0" maxOccurs="1"/>
      </xs:sequence>
      <xs:attribute name="name" type="xs:string" use="required"/>
      <xs:attribute name="endA" type="xs:string" use="required"/>
      <xs:attribute name="endB" type="xs:string" use="required"/>
      <xs:attribute name="componentA" type="xs:string" use="required"/>
      <xs:attribute name="componentB" type="xs:string" use="required"/>
    </xs:complexType>
:::
:::
:::
:::
:::

::: {#pairseventconnection .section}
[]{#lemsschema-pairseventconnection}

##### PairsEventConnection[\#](#pairseventconnection "Link to this heading"){.headerlink}

::: {.sd-tab-set .docutils}
can contain these elements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+-------------+--------------------------------------------------------+
| **buil      | [[BuildElement]{.std                                   |
| dElements** | .std-ref}](#lemsschema-buildelement){.reference        |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
:::
:::
:::
:::

::: {#pairfilter .section}
[]{#lemsschema-pairfilter}

##### PairFilter[\#](#pairfilter "Link to this heading"){.headerlink}

::: {.sd-tab-set .docutils}
can contain these elements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+-------------+--------------------------------------------------------+
| **buil      | [[BuildElement]{.std                                   |
| dElements** | .std-ref}](#lemsschema-buildelement){.reference        |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
:::
:::
:::
:::

::: {#includepair .section}
[]{#lemsschema-includepair}

##### IncludePair[\#](#includepair "Link to this heading"){.headerlink}

::: {.sd-tab-set .docutils}
can contain these elements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+-------------+--------------------------------------------------------+
| **buil      | [[BuildElement]{.std                                   |
| dElements** | .std-ref}](#lemsschema-buildelement){.reference        |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
:::
:::
:::
:::

::: {#with .section}
[]{#lemsschema-with}

##### With[\#](#with "Link to this heading"){.headerlink}

::: {.sd-tab-set .docutils}
can contain these elements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+-------------+--------------------------------------------------------+
| **buil      | [[BuildElement]{.std                                   |
| dElements** | .std-ref}](#lemsschema-buildelement){.reference        |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="With">
      <xs:attribute name="instance" type="xs:string" use="optional"/>
      <xs:attribute name="list" type="xs:string" use="optional"/>
      <xs:attribute name="index" type="xs:string" use="optional"/>
      <xs:attribute name="as" type="xs:string" use="required"/>
    </xs:complexType>
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <With instance="from" as="a"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <With instance="to" as="b"/>
:::
:::
:::
:::
:::

::: {#if .section}
[]{#lemsschema-if}

##### If[\#](#if "Link to this heading"){.headerlink}

::: {.sd-tab-set .docutils}
can contain these elements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+-------------+--------------------------------------------------------+
| **buil      | [[BuildElement]{.std                                   |
| dElements** | .std-ref}](#lemsschema-buildelement){.reference        |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
:::
:::
:::
:::

::: {#apply .section}
[]{#lemsschema-apply}

##### Apply[\#](#apply "Link to this heading"){.headerlink}

::: {.sd-tab-set .docutils}
can contain these elements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+-------------+--------------------------------------------------------+
| **buil      | [[BuildElement]{.std                                   |
| dElements** | .std-ref}](#lemsschema-buildelement){.reference        |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
:::
:::
:::
:::

::: {#gather .section}
[]{#lemsschema-gather}

##### Gather[\#](#gather "Link to this heading"){.headerlink}

::: {.sd-tab-set .docutils}
can contain these elements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+-------------+--------------------------------------------------------+
| **buil      | [[BuildElement]{.std                                   |
| dElements** | .std-ref}](#lemsschema-buildelement){.reference        |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
:::
:::
:::
:::

::: {#gatherpairs .section}
[]{#lemsschema-gatherpairs}

##### GatherPairs[\#](#gatherpairs "Link to this heading"){.headerlink}

::: {.sd-tab-set .docutils}
can contain these elements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+-------------+--------------------------------------------------------+
| **buil      | [[BuildElement]{.std                                   |
| dElements** | .std-ref}](#lemsschema-buildelement){.reference        |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
:::
:::
:::
:::
:::

[]{#document-Userdocs/LEMS_elements/Simulation}

::: {#simulation .section .tex2jax_ignore .mathjax_ignore}
[]{#lemsschema-page-simulation}

#### Simulation[\#](#simulation "Link to this heading"){.headerlink}

Schema against which LEMS based on these should be valid:
[LEMS_v0.7.6.xsd](https://github.com/LEMS/LEMS/tree/master/Schemas/LEMS/LEMS_v0.7.6.xsd){.reference
.external}. Generated on 18/06/24 from
[this](https://github.com/LEMS/LEMS/commit/fd7b30eceb6735ac343745c8f6992bdde72b248b){.reference
.external} commit. Please file any issues or questions at the [issue
tracker here](https://github.com/LEMS/LEMS/issues){.reference
.external}.

------------------------------------------------------------------------

::: {#lemsschema-simulation .section}
[]{#id1}

##### Simulation[\#](#lemsschema-simulation "Link to this heading"){.headerlink}

::: {.sd-tab-set .docutils}
can contain these elements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+-------------+--------------------------------------------------------+
| **records** | [[Record]{.std                                         |
|             | .std-ref}](#lemsschema-record){.reference .internal}   |
+-------------+--------------------------------------------------------+
| **eve       | [[EventRecord]{.std                                    |
| ntRecords** | .std-ref}](#lemsschema-eventrecord){.reference         |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
| **runs**    | [[Run]{.std .std-ref}](#lemsschema-run){.reference     |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
| **dat       | [[DataDisplay]{.std                                    |
| aDisplays** | .std-ref}](#lemsschema-datadisplay){.reference         |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
| **da        | [[DataWriter]{.std                                     |
| taWriters** | .std-ref}](#lemsschema-datawriter){.reference          |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
| **eve       | [[EventWriter]{.std                                    |
| ntWriters** | .std-ref}](#lemsschema-eventwriter){.reference         |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="Simulation">
      <xs:sequence>
        <xs:element name="DataDisplay" type="DataDisplay" minOccurs="0" maxOccurs="unbounded"/>
        <xs:element name="Record" type="Record" minOccurs="0" maxOccurs="unbounded"/>
        <xs:element name="EventRecord" type="EventRecord" minOccurs="0" maxOccurs="unbounded"/>
        <xs:element name="Run" type="Run" minOccurs="0" maxOccurs="unbounded"/>
        <xs:element name="DataWriter" type="DataWriter" minOccurs="0" maxOccurs="unbounded"/>
        <xs:element name="EventWriter" type="EventWriter" minOccurs="0" maxOccurs="unbounded"/>
        <xs:element name="Meta" type="Meta" minOccurs="0" maxOccurs="unbounded"/>
      </xs:sequence>
    </xs:complexType>
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <Simulation>
        <DataDisplay title="title" dataRegion="xmin,xmax,ymin,ymax"/>
    </Simulation>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Simulation>
        <Record quantity="quantity" timeScale="timeScale" scale="scale" color="color"/>
    </Simulation>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Simulation>
        <DataWriter path="path" fileName="fileName"/>
    </Simulation>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Simulation>
        <Record quantity="quantity"/>
    </Simulation>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Simulation>
        <Run component="target" variable="t" increment="step" total="length"/>
    </Simulation>
:::
:::
:::
:::
:::

::: {#record .section}
[]{#lemsschema-record}

##### Record[\#](#record "Link to this heading"){.headerlink}

::: {.sd-tab-set .docutils}
Properties

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------+-------------------------------------------------+
| **qu | String      | path to the parameter that will contain the     |
| anti |             | path to the quantity to be recorded             |
| ty** |             |                                                 |
+------+-------------+-------------------------------------------------+
| *    | String      | path to the element that defines the scale for  |
| *sca |             | rendering the quantity dimensionless            |
| le** |             |                                                 |
+------+-------------+-------------------------------------------------+
| *    | String      | hex format color suggestion for how the data    |
| *col |             | should be displayed                             |
| or** |             |                                                 |
+------+-------------+-------------------------------------------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="Record">
      <xs:attribute name="quantity" type="xs:string" use="required"/>
      <xs:attribute name="timeScale" type="xs:string" use="optional"/>
      <xs:attribute name="scale" type="xs:string" use="optional"/>
      <xs:attribute name="color" type="xs:string" use="optional"/>
    </xs:complexType>
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <Record quantity="quantity" timeScale="timeScale" scale="scale" color="color"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Record quantity="quantity"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Record quantity="quantity" timeScale="timeScale" scale="scale" color="color"/>
:::
:::
:::
:::
:::

::: {#eventrecord .section}
[]{#lemsschema-eventrecord}

##### EventRecord[\#](#eventrecord "Link to this heading"){.headerlink}

::: {.sd-tab-set .docutils}
Properties

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------+-------------------------------------------------+
| **qu | String      | path for the component which will emit spikes   |
| anti |             | to be recorded                                  |
| ty** |             |                                                 |
+------+-------------+-------------------------------------------------+
| *    | String      | event port for the component which will emit    |
| *eve |             | spikes                                          |
| ntPo |             |                                                 |
| rt** |             |                                                 |
+------+-------------+-------------------------------------------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="EventRecord">
      <xs:attribute name="quantity" type="xs:string" use="required"/>
      <xs:attribute name="eventPort" type="xs:string" use="required"/>
    </xs:complexType>
:::
:::
:::
:::
:::

::: {#datadisplay .section}
[]{#lemsschema-datadisplay}

##### DataDisplay[\#](#datadisplay "Link to this heading"){.headerlink}

::: {.sd-tab-set .docutils}
Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="DataDisplay">
      <xs:attribute name="title" type="xs:string" use="required"/>
      <xs:attribute name="dataRegion" type="xs:string" use="required"/>
    </xs:complexType>
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <DataDisplay title="title" dataRegion="xmin,xmax,ymin,ymax"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <DataDisplay title="title" dataRegion="xmin,xmax,ymin,ymax"/>
:::
:::
:::
:::
:::

::: {#datawriter .section}
[]{#lemsschema-datawriter}

##### DataWriter[\#](#datawriter "Link to this heading"){.headerlink}

::: {.sd-tab-set .docutils}
Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="DataWriter">
      <xs:attribute name="path" type="xs:string" use="required"/>
      <xs:attribute name="fileName" type="xs:string" use="required"/>
    </xs:complexType>
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <DataWriter path="path" fileName="fileName"/>
:::
:::
:::
:::
:::

::: {#eventwriter .section}
[]{#lemsschema-eventwriter}

##### EventWriter[\#](#eventwriter "Link to this heading"){.headerlink}

::: {.sd-tab-set .docutils}
Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="EventWriter">
      <xs:attribute name="path" type="xs:string" use="required"/>
      <xs:attribute name="fileName" type="xs:string" use="required"/>
      <xs:attribute name="format" type="xs:string" use="required"/>
    </xs:complexType>
:::
:::
:::
:::
:::

::: {#run .section}
[]{#lemsschema-run}

##### Run[\#](#run "Link to this heading"){.headerlink}

*The run element provides a way to make a model runnable. It should
point to the parameters that set the step size etc. The target
parameters have to be dimensionally consistent.*

::: {.sd-tab-set .docutils}
Properties

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------+-------------------------------------------------+
| *    | String      | name of the component reference that will set   |
| *com |             | the component to be run                         |
| pone |             |                                                 |
| nt** |             |                                                 |
+------+-------------+-------------------------------------------------+
| **va | String      |                                                 |
| riab |             |                                                 |
| le** |             |                                                 |
+------+-------------+-------------------------------------------------+
| *    | String      | path to the parameter that sets the step size   |
| *inc |             |                                                 |
| reme |             |                                                 |
| nt** |             |                                                 |
+------+-------------+-------------------------------------------------+
| *    | String      | path to the parameter that sets the total span  |
| *tot |             | of the independent variable to be run           |
| al** |             |                                                 |
+------+-------------+-------------------------------------------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="Run">
      <xs:attribute name="component" type="xs:string" use="required"/>
      <xs:attribute name="variable" type="xs:string" use="required"/>
      <xs:attribute name="increment" type="xs:string" use="required"/>
      <xs:attribute name="total" type="xs:string" use="required"/>
    </xs:complexType>
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <Run component="target" variable="t" increment="step" total="length"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Run component="target" variable="t" increment="step" total="length"/>
:::
:::
:::
:::
:::
:::

[]{#document-Userdocs/LEMS_elements/Procedure}

::: {#procedure .section .tex2jax_ignore .mathjax_ignore}
[]{#lemsschema-page-procedure}

#### Procedure[\#](#procedure "Link to this heading"){.headerlink}

Schema against which LEMS based on these should be valid:
[LEMS_v0.7.6.xsd](https://github.com/LEMS/LEMS/tree/master/Schemas/LEMS/LEMS_v0.7.6.xsd){.reference
.external}. Generated on 18/06/24 from
[this](https://github.com/LEMS/LEMS/commit/fd7b30eceb6735ac343745c8f6992bdde72b248b){.reference
.external} commit. Please file any issues or questions at the [issue
tracker here](https://github.com/LEMS/LEMS/issues){.reference
.external}.

------------------------------------------------------------------------

::: {#lemsschema-procedure .section}
[]{#id1}

##### Procedure[\#](#lemsschema-procedure "Link to this heading"){.headerlink}

::: {.sd-tab-set .docutils}
can contain these elements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+-------------+--------------------------------------------------------+
| **s         | [lemsschema:statement\_]{.xref .std .std-ref}          |
| tatements** |                                                        |
+-------------+--------------------------------------------------------+
:::
:::
:::
:::

::: {#equilibrate .section}
[]{#lemsschema-equilibrate}

##### Equilibrate[\#](#equilibrate "Link to this heading"){.headerlink}
:::

::: {#foreachcomponent .section}
[]{#lemsschema-foreachcomponent}

##### ForEachComponent[\#](#foreachcomponent "Link to this heading"){.headerlink}

::: {.sd-tab-set .docutils}
can contain these elements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+-------------+--------------------------------------------------------+
| **s         | [lemsschema:statement\_]{.xref .std .std-ref}          |
| tatements** |                                                        |
+-------------+--------------------------------------------------------+
:::
:::
:::
:::

::: {#print .section}
[]{#lemsschema-print}

##### Print[\#](#print "Link to this heading"){.headerlink}
:::
:::

[]{#document-Userdocs/LEMS_elements/DefiningComponents}

::: {#defining-components .section .tex2jax_ignore .mathjax_ignore}
[]{#lemsschema-page-defining-components}

#### Defining Components[\#](#defining-components "Link to this heading"){.headerlink}

Schema against which LEMS based on these should be valid:
[LEMS_v0.7.6.xsd](https://github.com/LEMS/LEMS/tree/master/Schemas/LEMS/LEMS_v0.7.6.xsd){.reference
.external}. Generated on 18/06/24 from
[this](https://github.com/LEMS/LEMS/commit/fd7b30eceb6735ac343745c8f6992bdde72b248b){.reference
.external} commit. Please file any issues or questions at the [issue
tracker here](https://github.com/LEMS/LEMS/issues){.reference
.external}.

------------------------------------------------------------------------

::: {#component .section}
[]{#lemsschema-component}

##### Component[\#](#component "Link to this heading"){.headerlink}

::: {.sd-tab-set .docutils}
Properties

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+------+-------------+-------------------------------------------------+
| **   | String      |                                                 |
| id** |             |                                                 |
+------+-------------+-------------------------------------------------+
| **na | String      | Name by which the component was declared - this |
| me** |             | shouldn't be accessible.                        |
+------+-------------+-------------------------------------------------+
| **de | String      | Name by which the component was declared - this |
| clar |             | shouldn't be accessible.                        |
| edTy |             |                                                 |
| pe** |             |                                                 |
+------+-------------+-------------------------------------------------+
| **ty | String      |                                                 |
| pe** |             |                                                 |
+------+-------------+-------------------------------------------------+
| **e  | String      |                                                 |
| Xten |             |                                                 |
| ds** |             |                                                 |
+------+-------------+-------------------------------------------------+
:::
:::

can contain these elements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+-------------+--------------------------------------------------------+
| **i         | [[Insertion]{.std                                      |
| nsertions** | .s                                                     |
|             | td-ref}](Landing.html#lemsschema-insertion){.reference |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
| **c         | [[Component]{.std                                      |
| omponents** | .std-ref}](#lemsschema-component){.reference           |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
| **abouts**  | [[About]{.std                                          |
|             | .std-ref}](Landing.html#lemsschema-about){.reference   |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
| **metas**   | [[Meta]{.std                                           |
|             | .std-ref}](Landing.html#lemsschema-meta){.reference    |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
:::
:::

Schema

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="Component">
      <xs:sequence>
        <xs:any minOccurs="0" maxOccurs="unbounded" processContents="lax"/>
      </xs:sequence>
      <xs:anyAttribute processContents="skip"/>
    </xs:complexType>
:::
:::
:::

Usage: XML

::: {.sd-tab-content .docutils}
::: {.highlight-xml .notranslate}
::: highlight
    <Component id="ctb" type="iaf1" threshold="-30 mV" refractoryPeriod="2 ms" capacitance="1uF"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Component id="celltype_c" type="iaf3" leakConductance="3 pS" refractoryPeriod="3 ms" threshold="45 mV" leakReversal="-50 mV" deltaV="5mV" capacitance="1uF"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Component id="gen1" type="spikeGenerator" period="30ms"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Component id="gen2" type="spikeGenerator2" period="32ms"/>
:::
:::

::: {.highlight-xml .notranslate}
::: highlight
    <Component id="iaf3cpt" type="iaf3" leakReversal="-50mV" deltaV="50mV" threshold="-30mV" leakConductance="50pS" refractoryPeriod="4ms" capacitance="1pF"/>
:::
:::
:::
:::
:::
:::

[]{#document-Userdocs/LEMS_elements/Geometry}

::: {#geometry .section .tex2jax_ignore .mathjax_ignore}
[]{#lemsschema-page-geometry}

#### Geometry[\#](#geometry "Link to this heading"){.headerlink}

Schema against which LEMS based on these should be valid:
[LEMS_v0.7.6.xsd](https://github.com/LEMS/LEMS/tree/master/Schemas/LEMS/LEMS_v0.7.6.xsd){.reference
.external}. Generated on 18/06/24 from
[this](https://github.com/LEMS/LEMS/commit/fd7b30eceb6735ac343745c8f6992bdde72b248b){.reference
.external} commit. Please file any issues or questions at the [issue
tracker here](https://github.com/LEMS/LEMS/issues){.reference
.external}.

------------------------------------------------------------------------

::: {#lemsschema-geometry .section}
[]{#id1}

##### Geometry[\#](#lemsschema-geometry "Link to this heading"){.headerlink}

*Specifies the geometrical interpretation of the properties of
components realizing this ComponentType.*

::: {.sd-tab-set .docutils}
can contain these elements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+-------------+--------------------------------------------------------+
| *           | [[Frustum]{.std                                        |
| *frustums** | .std-ref}](#lemsschema-frustum){.reference .internal}  |
+-------------+--------------------------------------------------------+
| **solids**  | [[Solid]{.std .std-ref}](#lemsschema-solid){.reference |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
| **          | [[Skeleton]{.std                                       |
| skeletons** | .std-ref}](#lemsschema-skeleton){.reference .internal} |
+-------------+--------------------------------------------------------+
:::
:::
:::
:::

::: {#frustum .section}
[]{#lemsschema-frustum}

##### Frustum[\#](#frustum "Link to this heading"){.headerlink}
:::

::: {#solid .section}
[]{#lemsschema-solid}

##### Solid[\#](#solid "Link to this heading"){.headerlink}
:::

::: {#location .section}
[]{#lemsschema-location}

##### Location[\#](#location "Link to this heading"){.headerlink}
:::

::: {#skeleton .section}
[]{#lemsschema-skeleton}

##### Skeleton[\#](#skeleton "Link to this heading"){.headerlink}

::: {.sd-tab-set .docutils}
can contain these elements

::: {.sd-tab-content .docutils}
::: pst-scrollable-table-container
+-------------+--------------------------------------------------------+
| **sca       | [[ScalarField]{.std                                    |
| larFields** | .std-ref}](#lemsschema-scalarfield){.reference         |
|             | .internal}                                             |
+-------------+--------------------------------------------------------+
:::
:::
:::
:::

::: {#scalarfield .section}
[]{#lemsschema-scalarfield}

##### ScalarField[\#](#scalarfield "Link to this heading"){.headerlink}
:::
:::
:::
:::
:::
:::

[]{#document-Userdocs/NeuroMLv2AndLEMS}

::: {#neuroml-2-and-lems .section .tex2jax_ignore .mathjax_ignore}
[]{#userdocs-nml2lems}

## NeuroML 2 and LEMS[\#](#neuroml-2-and-lems "Link to this heading"){.headerlink}

LEMS is an XML based language originally developed by Robert Cannon for
specifying generic models of hybrid dynamical systems. Models defined in
LEMS can also be simulated directly through a native interpreter.

-   **ComponentType** elements define the behaviour of a specific type
    of model and specify **Parameters**, **StateVariables**, and their
    **Dynamics** and **Structure** can be defined as templates for model
    elements (e.g. HH ion channels, abstract cells, etc.). The notion of
    a **ComponentType** is thus similar to that of a **class** in object
    oriented programming.

-   **Components** are instances of these types, with specific values of
    **Parameters** (e.g. HH squid axon Na+ channel, I&F cell with
    threshold -60mV, etc.). **Components** play the same role as
    **objects** in object oriented programming.

<figure id="id2" class="align-default">
<img src="_images/NeuroML2_LEMS_Overview_web.svg"
alt="Figure showing relationship between LEMS and NeuroML2" />
<figcaption><p><span class="caption-number">Fig. 59 </span><span
class="caption-text">This image (from Blundell et al. 2018 (<span
id="id1">[<a href="Landing.html#id13" class="reference internal"
title="Inga Blundell, Romain Brette, Thomas A. Cleland, Thomas G. Close, Daniel Coca, Andrew P. Davison, Sandra Diaz-Pier, Carlos Fernandez Musoles, Padraig Gleeson, Dan F. M. Goodman, Michael Hines, Michael W. Hopkins, Pramod Kumbhar, David R. Lester, Bóris Marin, Abigail Morrison, Eric Müller, Thomas Nowotny, Alexander Peyser, Dimitri Plotnikov, Paul Richmond, Andrew Rowley, Bernhard Rumpe, Marcel Stimberg, Alan B. Stokes, Adam Tomkins, Guido Trensch, Marmaduke Woodman, and Jochen Martin Eppler. Code generation in computational neuroscience: a review of tools and techniques. Frontiers in Neuroinformatics, 12:68, 2018. doi:10.3389/fninf.2018.00068.">BBC+18</a>]</span>))
shows the usage of LEMS <strong>ComponentTypes</strong> and
<strong>Components</strong> in NeuroML. Elements in NeuroML v2 are
<strong>Components</strong> which have a corresponding structural and
mathematical definition in LEMS
<strong>ComponentTypes</strong>.</span><a href="#id2" class="headerlink"
title="Link to this image">#</a></p></figcaption>
</figure>

On the left side of the figure, examples are shown of the (truncated)
XML representations of:

-   (blue) a [[network]{.std
    .std-ref}](Landing.html#schema-network){.reference .internal}
    containing two [[populations]{.std
    .std-ref}](Landing.html#schema-population){.reference .internal} of
    [[integrate-and-fire cells]{.std
    .std-ref}](Landing.html#schema-iafcell){.reference .internal}
    connected by a single [[projection]{.std
    .std-ref}](Landing.html#schema-projection){.reference .internal}
    between them;

-   (green) a [[spiking neuron model]{.std
    .std-ref}](Landing.html#schema-izhikevichcell){.reference .internal}
    as described by Izhikevich (2003);

-   (yellow) a [[conductance based synapse]{.std
    .std-ref}](Landing.html#schema-exponesynapse){.reference .internal}
    with a single exponential decay waveform.

On the right, the definition of the structure and dynamics of these
elements in the LEMS language is shown. Each element has a corresponding
**ComponentType** definition, describing the parameters (as well as
their dimensions, not shown) and the dynamics in terms of state
variables and their derivatives, any derived variables, and the
behaviour when certain conditions are met or events are received (for
example, the emission of a spike after a given threshold is crossed).

::: {#neuroml-2-component-type-definitions-in-lems .section}
[]{#userdocs-neuromlv2inlems}

### NeuroML 2 Component Type definitions in LEMS[\#](#neuroml-2-component-type-definitions-in-lems "Link to this heading"){.headerlink}

The standard set of **ComponentType** definitions for the core NeuroML2
elements are contained in a curated set of files (below) though users
are [[free to define their own ComponentTypes to extend the scope of the
language]{.std .std-ref}](Landing.html#userdocs-extending){.reference
.internal}.

-   [[Dimensions/units allowed]{.std
    .std-ref}](Landing.html#schema-neuromlcoredimensions){.reference
    .internal} ([source in
    LEMS](https://github.com/NeuroML/NeuroML2/blob/master/NeuroML2CoreTypes/NeuroMLCoreDimensions.xml?view=markup){.reference
    .external})

-   [[Cell models]{.std .std-ref}](Landing.html#schema-cells){.reference
    .internal} ([source in
    LEMS](https://github.com/NeuroML/NeuroML2/blob/master/NeuroML2CoreTypes/Cells.xml?view=markup){.reference
    .external})

-   [[Network elements]{.std
    .std-ref}](Landing.html#schema-networks){.reference .internal}
    ([source in
    LEMS](https://github.com/NeuroML/NeuroML2/blob/master/NeuroML2CoreTypes/Networks.xml?view=markup){.reference
    .external})

-   [[Ion channels]{.std
    .std-ref}](Landing.html#schema-channels){.reference .internal}
    ([source in
    LEMS](https://github.com/NeuroML/NeuroML2/blob/master/NeuroML2CoreTypes/Channels.xml?view=markup){.reference
    .external})

-   [[Synapse models]{.std
    .std-ref}](Landing.html#schema-synapses){.reference .internal}
    ([source in
    LEMS](https://github.com/NeuroML/NeuroML2/blob/master/NeuroML2CoreTypes/Synapses.xml?view=markup){.reference
    .external})

-   [[Mapping of PyNN cells & synapses]{.std
    .std-ref}](Landing.html#schema-pynn){.reference .internal} ([source
    in
    LEMS](https://github.com/NeuroML/NeuroML2/blob/master/NeuroML2CoreTypes/PyNN.xml?view=markup){.reference
    .external})

Here, for example, the [[izhikevich2007Cell]{.std
.std-ref}](Landing.html#schema-izhikevich2007cell){.reference .internal}
is defined in the [NeuroML
schema](https://github.com/NeuroML/NeuroML2/blob/master/Schemas/NeuroML2/NeuroML_v2.2.xsd){.reference
.external} as having the following internal attributes:

::: {.highlight-xml .notranslate}
::: highlight
    <xs:complexType name="Izhikevich2007Cell">
        <xs:complexContent>
            <xs:extension base="BaseCellMembPotCap">
                <xs:attribute name="v0" type="Nml2Quantity_voltage" use="required"/>
                <xs:attribute name="k" type="Nml2Quantity_conductancePerVoltage" use="required"/>
                <xs:attribute name="vr" type="Nml2Quantity_voltage" use="required"/>
                <xs:attribute name="vt" type="Nml2Quantity_voltage" use="required"/>
                <xs:attribute name="vpeak" type="Nml2Quantity_voltage" use="required"/>
                <xs:attribute name="a" type="Nml2Quantity_pertime" use="required"/>
                <xs:attribute name="b" type="Nml2Quantity_conductance" use="required"/>
                <xs:attribute name="c" type="Nml2Quantity_voltage" use="required"/>
                <xs:attribute name="d" type="Nml2Quantity_current" use="required"/>
            </xs:extension>
        </xs:complexContent>
    </xs:complexType>
:::
:::

Correspondingly, its **ComponentType** dynamics are defined in the LEMS
file,
[Cells.xml](https://github.com/NeuroML/NeuroML2/blob/master/NeuroML2CoreTypes/Cells.xml){.reference
.external}. (Note: you do not need to read the XML LEMS definitions, you
can see this information in a well formatted form [[here in the
documentation itself]{.std
.std-ref}](Landing.html#schema-izhikevich2007cell){.reference
.internal})

::: {.highlight-xml .notranslate}
::: highlight
    <ComponentType name="izhikevich2007Cell"
        extends="baseCellMembPotCap"
        description="Cell based on the modified Izhikevich model in Izhikevich 2007, Dynamical systems in neuroscience, MIT Press">

        <Parameter name="v0" dimension="voltage"/>

        <!--
        Defined in baseCellMembPotCap:
        <Parameter name="C" dimension="capacitance"/>
        -->
        <Parameter name="k" dimension="conductance_per_voltage"/>

        <Parameter name="vr" dimension="voltage"/>
        <Parameter name="vt" dimension="voltage"/>
        <Parameter name="vpeak" dimension="voltage"/>

        <Parameter name="a" dimension="per_time"/>
        <Parameter name="b" dimension="conductance"/>
        <Parameter name="c" dimension="voltage"/>
        <Parameter name="d" dimension="current"/>

        <Attachments name="synapses" type="basePointCurrent"/>

        <Exposure name="u" dimension="current"/>

        <Dynamics>

            <StateVariable name="v" dimension="voltage" exposure="v"/>
            <StateVariable name="u" dimension="current" exposure="u"/>

            <DerivedVariable name="iSyn"  dimension="current" exposure="iSyn" select="synapses[*]/i" reduce="add" />

            <DerivedVariable name="iMemb" dimension="current" exposure="iMemb" value="k * (v-vr) * (v-vt) + iSyn - u"/>

            <TimeDerivative variable="v" value="iMemb / C"/>
            <TimeDerivative variable="u" value="a * (b * (v-vr) - u)"/>

            <OnStart>
                <StateAssignment variable="v" value="v0"/>
                <StateAssignment variable="u" value="0"/>
            </OnStart>

            <OnCondition test="v .gt. vpeak">
                <StateAssignment variable="v" value="c"/>
                <StateAssignment variable="u" value="u + d"/>
                <EventOut port="spike"/>
            </OnCondition>

        </Dynamics>

    </ComponentType>
:::
:::

We can define **Component**s of the [[izhikevich2007Cell]{.std
.std-ref}](Landing.html#schema-izhikevich2007cell){.reference .internal}
**ComponentType** with the parameters we need. For example, the
[[izhikevich2007Cell]{.std
.std-ref}](Landing.html#schema-izhikevich2007cell){.reference .internal}
neuron model can exhibit different spiking behaviours, so we can define
a regular spiking **Component**, or another **Component** that exhibits
bursting.

::: {.highlight-xml .notranslate}
::: highlight
    <izhikevich2007Cell id="iz2007RS" v0 = "-60mV" C="100 pF" k = "0.7 nS_per_mV"
                        vr = "-60 mV" vt = "-40 mV" vpeak = "35 mV"
                        a = "0.03 per_ms" b = "-2 nS" c = "-50 mV" d = "100 pA"/>
:::
:::

Once these **Component**s are defined in the NeuroML document, we can
use **Instance**s of them to create populations and networks, and so on.

::: {.admonition-you-don-t-have-to-write-in-xml .admonition}
You don't have to write in XML...

A quick reminder that while XML files can be edited in a standard text
editor, you generally don't have to create/update them by hand. [[This
guide]{.std
.std-ref}](Landing.html#userdocs-getting-started-single-example){.reference
.internal} goes through the steps of creating an example using the
[[izhikevich2007Cell]{.std
.std-ref}](Landing.html#schema-izhikevich2007cell){.reference .internal}
model in Python using [[libNeuroML]{.std
.std-ref}](Landing.html#libneuroml){.reference .internal} and
[[pyNeuroML]{.std .std-ref}](Landing.html#pyneuroml){.reference
.internal}
:::

Using LEMS to specify the core of NeuroML version 2 has the following
significant advantages:

-   NeuroML 2 XML files can be used standalone by applications
    (exported/imported) in the same way as NeuroML v1.x, without
    reference to the LEMS definitions, easing the transition for v1.x
    compliant applications

-   Any NeuroML 2 **ComponentType** can be extended and will be
    usable/translatable by any application (e.g. jLEMS) which
    understands LEMS

The first point above means that a parsing application does not
necessarily have to natively read the LEMS type definition for, e.g. an
[[izhikevich2007Cell]{.std
.std-ref}](Landing.html#schema-izhikevich2007cell){.reference .internal}
element: it just has to map the NeuroML element parameters onto its own
object model implementing that entity. Ideally, the behaviour should be
the same − which could be ascertained by testing against the reference
LEMS interpreter implementation
([jLEMS](http://github.com/LEMS/jLEMS/){.reference .external}).

The second point above means that if an application does support LEMS,
it can automatically parse (and generate code for) a wide range of
NeuroML 2 cells, channels and synapses, including any new
**ComponentType** derived from these, without having to natively know
anything about channels, cell models, etc.

::: {.tip .admonition}
jnml and pynml handle both LEMS and NeuroML 2.

[[jNeuroML]{.std .std-ref}](Landing.html#jneuroml){.reference .internal}
and [[pynml]{.std .std-ref}](Landing.html#pyneuroml){.reference
.internal} handle both LEMS and NeuroML 2. They bundle jLEMS together
with the LEMS definitions for NeuroML 2 ComponentTypes, and can simulate
any LEMS model as well as many NeuroML 2 models.
:::
:::

::: {#more-information .section}
[]{#userdocs-neuromlv2inlems-more}

### More information[\#](#more-information "Link to this heading"){.headerlink}

While the [[tutorials]{.std
.std-ref}](Landing.html#userdocs-getting-started-neuroml){.reference
.internal} cover many of the key points of using LEMS with NeuroML,
there are some points which require further explanation:

-   [[What are the conventions/best practices to follow in naming
    NeuroML/LEMS files/elements?]{.std
    .std-ref}](Landing.html#userdocs-conventions){.reference .internal}

-   [[How are units and dimensions handled in NeuroML and LEMS?]{.std
    .std-ref}](Landing.html#userdocs-unitsanddimensions){.reference
    .internal}

-   [[How do I use a LEMS Simulation file to specify how to execute a
    NeuroML model?]{.std
    .std-ref}](Landing.html#userdocs-lemssimulation){.reference
    .internal}

-   [[How can I extend NeuroML model to include new types using
    LEMS?]{.std .std-ref}](Landing.html#userdocs-extending){.reference
    .internal}

-   [[What is the correct format/usage of paths and quantities in
    NeuroML and LEMS?]{.std
    .std-ref}](Landing.html#userdocs-paths){.reference .internal}
:::

::: {.toctree-wrapper .compound}
[]{#document-Userdocs/Conventions}

::: {#conventions .section .tex2jax_ignore .mathjax_ignore}
[]{#userdocs-conventions}

### Conventions[\#](#conventions "Link to this heading"){.headerlink}

This page documents various conventions in use in NeuroML.

::: {#prefer-underscores-instead-of-spaces .section}
[]{#userdocs-conventions-underscores}

#### Prefer underscores instead of spaces[\#](#prefer-underscores-instead-of-spaces "Link to this heading"){.headerlink}

In general, please prefer underscores [`_`{.docutils .literal
.notranslate}]{.pre} instead of spaces wherever possible, in filenames
and ids.
:::

::: {#component-ids-nmlid .section}
[]{#userdocs-conventions-nmlid}

#### Component IDs: NmlId[\#](#component-ids-nmlid "Link to this heading"){.headerlink}

Some Components take an [`id`{.docutils .literal .notranslate}]{.pre}
parameter of type [`NmlId`{.docutils .literal .notranslate}]{.pre} to
set an ID for them. They can then be referred to using their IDs when
constructing paths and so on.

IDs of type [`NmlId`{.docutils .literal .notranslate}]{.pre} in NeuroML
are strings and have certain constraints:

-   they **must** start with an alphabet (either small or capital) or an
    underscore

-   they may include alphabets, both small and capital letters, numbers
    and underscores

IDs are also checked during validation, so if an ID does not follow
these constraints, the validation will throw an error.
:::

::: {#file-naming .section}
[]{#userdocs-conventions-files}

#### File naming[\#](#file-naming "Link to this heading"){.headerlink}

When naming different NeuroML files, we suggest the following suffixes:

-   [`channel.nml`{.docutils .literal .notranslate}]{.pre} for NeuroML
    files describing ion channels, for example:
    [`Na.channel.nml`{.docutils .literal .notranslate}]{.pre}

-   [`cell.nml`{.docutils .literal .notranslate}]{.pre} for NeuroML
    files describing cells, for example: [`hh.cell.nml`{.docutils
    .literal .notranslate}]{.pre}

-   [`synapse.nml`{.docutils .literal .notranslate}]{.pre} for NeuroML
    files describing synapses, for example:
    [`AMPA.synapse.nml`{.docutils .literal .notranslate}]{.pre}

-   [`net.nml`{.docutils .literal .notranslate}]{.pre} for NeuroML files
    describing networks of cells, for example:
    [`excitatory.net.nml`{.docutils .literal .notranslate}]{.pre}

For LEMS files that describe simulations of NeuroML models ([["LEMS
Simulation files"]{.std
.std-ref}](Landing.html#userdocs-lemssimulation){.reference .internal}),
we suggest that:

-   file names start with the [`LEMS_`{.docutils .literal
    .notranslate}]{.pre} prefix,

-   file names end in [`xml`{.docutils .literal .notranslate}]{.pre}

For example [`LEMS_HH_Simulation.xml`{.docutils .literal
.notranslate}]{.pre}.

<figure id="id1" class="align-center">
<a href="_images/lems_nml_files.png"
class="reference internal image-reference"><img
src="_images/lems_nml_files.png"
style="width: 647.52px; height: 409.44px;"
alt="LEMS Simulation file and NeuroML file" /></a>
<figcaption><p><span class="caption-number">Fig. 60 </span><span
class="caption-text">Typical organisation for a NeuroML simulation. The
main NeuroML model is specified in a file with the network (<span
class="pre"><code
class="docutils literal notranslate">*.net.nml</code></span>), which can
include/point to files containing individual synapses (<span
class="pre"><code
class="docutils literal notranslate">*.synapse.nml</code></span>) or
cell files (<span class="pre"><code
class="docutils literal notranslate">*.cell.nml</code></span>). If the
latter are conductance based, they may include external channel files
(<span class="pre"><code
class="docutils literal notranslate">*.channel.nml</code></span>). The
main LEMS Simulation file only needs to include the network file, and
tools for running simulations of the model refer to just this LEMS file.
Exceptions to these conventions are frequent and simulations will run
perfectly well with all the elements inside the main LEMS file, but
using this scheme will maximise reusability of model elements.</span><a
href="#id1" class="headerlink"
title="Link to this image">#</a></p></figcaption>
</figure>
:::

::: {#neuron-segments .section}
[]{#userdocs-conventions-segments}

#### Neuron segments[\#](#neuron-segments "Link to this heading"){.headerlink}

When naming segments in multi-compartmental neuron models, we suggest
the following prefixes:

-   [`axon_`{.docutils .literal .notranslate}]{.pre} for axonal segments

-   [`dend_`{.docutils .literal .notranslate}]{.pre} for dendritic
    segments

-   [`soma_`{.docutils .literal .notranslate}]{.pre} for somatic
    segments

There are 3 specific recommended names for segment groups which contain
**ALL** of the somatic, dendritic or axonal segments

-   [`axon_group`{.docutils .literal .notranslate}]{.pre} for the group
    of all axonal segments

-   [`dendrite_group`{.docutils .literal .notranslate}]{.pre} for the
    group of all dendritic segments

-   [`soma_group`{.docutils .literal .notranslate}]{.pre} for the group
    of all somatic segments

Ideally every segment should be a member of one and only one of these
groups.
:::
:::

[]{#document-Userdocs/UnitsAndDimensions}

::: {#units-and-dimensions .section .tex2jax_ignore .mathjax_ignore}
[]{#userdocs-unitsanddimensions}

### Units and dimensions[\#](#units-and-dimensions "Link to this heading"){.headerlink}

Support for dimensional quantities is a fundamental (and essential)
feature of NeuroML, backed up by support for units and dimensions in
LEMS.

The basic rules are:

-   specify the **dimensions** of quantities in LEMS

-   use compatible **units** defined in the NeuroML schema in NeuroML
    models.

The main motivation for this is that fundamental expressions for
defining a model are independent of any particular units. For example,
Ohm's law, **V = I \* R** relates to quantities with dimensions voltage,
current and resistance, not millivolts, picoamps, ohms, etc.

Users can therefore use a wide range of commonly used units for each
dimension defined in the [[standard unit and dimension definitions]{.std
.std-ref}](Landing.html#schema-neuromlcoredimensions){.reference
.internal} of NeuroML 2 without worrying about conversion factors.

Additionally, please keep in mind that:

-   all quantities are saved and [[recorded]{.std
    .std-ref}](Landing.html#userdocs-quantitiesandrecording){.reference
    .internal} in SI Units

-   when plotting data using NeuroML/LEMS using the [[Line]{.std
    .std-ref}](Landing.html#schema-line){.reference .internal}
    component, users can use the [`scale`{.docutils .literal
    .notranslate}]{.pre} parameter to convert quantities to other units.
:::

[]{#document-Userdocs/Paths}

::: {#paths .section .tex2jax_ignore .mathjax_ignore}
[]{#userdocs-paths}

### Paths[\#](#paths "Link to this heading"){.headerlink}

Since NeuroMLv2 and LEMS are both XML based, entities in models and
simulations must be referred to using *paths*
([XPath](https://en.wikipedia.org/wiki/XPath){.reference .external}
like). This page documents how paths can be constructed, and how they
can be used to refer to entities in NeuroML/LEMS based models and
simulations (e.g. in a [[LEMS Simulation file]{.std
.std-ref}](Landing.html#userdocs-lemssimulation){.reference .internal}).

::: pst-scrollable-table-container
  operator                                               description       function                                                                                                                                                                                                                                                                                            example
  ------------------------------------------------------ ----------------- --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- ---------------------------------------------------------------------------------------------------------------------------
  [`/`{.code .docutils .literal .notranslate}]{.pre}     forward slash     used to split the levels in a path string                                                                                                                                                                                                                                                           see [[below]{.std .std-ref}](#userdocs-paths-example){.reference .internal}
  [`.`{.code .docutils .literal .notranslate}]{.pre}     single period     refers to the level of the current node (usually omitted)                                                                                                                                                                                                                                           see [[below]{.std .std-ref}](#userdocs-paths-example){.reference .internal}
  [`..`{.code .docutils .literal .notranslate}]{.pre}    two periods       refers to the level of the current node's parent node                                                                                                                                                                                                                                               see [[below]{.std .std-ref}](#userdocs-paths-example){.reference .internal}
  [`[x]`{.code .docutils .literal .notranslate}]{.pre}   square brackets   used to refer to a particular instance (in this case, [`x`{.code .docutils .literal .notranslate}]{.pre}) in Components/Elements that have a [`size`{.docutils .literal .notranslate}]{.pre} attribute (like [[population]{.std .std-ref}](Landing.html#schema-population){.reference .internal})   see [[below]{.std .std-ref}](#userdocs-paths-example){.reference .internal}
  [`:`{.code .docutils .literal .notranslate}]{.pre}     colon             used to refer to a particular Component instance for [`attachments`{.code .docutils .literal .notranslate}]{.pre}                                                                                                                                                                                   [ex](https://github.com/NeuroML/NeuroML2/blob/master/LEMSexamples/LEMS_NML2_Ex25_MultiComp.xml#L45){.reference .external}
:::

Paths start from any element and ascend/descend to refer to the entity
that is to be referenced.

::: {#example .section}
[]{#userdocs-paths-example}

#### Example[\#](#example "Link to this heading"){.headerlink}

For example, in the following block of code, based on the [[Izhikevich
network example]{.std
.std-ref}](Landing.html#userdocs-gettingstarted-izhikevichnetwork){.reference
.internal}, a network is defined in NeuroML with 2 populations:

::: {.highlight-xml .notranslate}
::: highlight
        <network id="IzNet">
            <population id="IzPop0" component="iz2007RS0" size="5">
                <property tag="color" value="0 0 .8"/>
            </population>
            <populationList id="IzPop1" component="iz2007RS0">
                <property tag="color" value=".8 0 0"/>
                <instance id=0>
                    <location x="0" y="0" z="0" />
                </instance>
                <instance id=1>
                    <location x="1" y="0" z="0" />
                </instance>
                <instance id=2>
                    <location x="2" y="0" z="0" />
                </instance>
                <instance id=3>
                    <location x="3" y="0" z="0" />
                </instance>
                <instance id=4>
                    <location x="4" y="0" z="0" />
                </instance>
            </populationList>
            <projection id="proj" presynapticPopulation="IzPop0" postsynapticPopulation="IzPop1" synapse="syn0">
                <connection id="0" preCellId="../IzPop0[0]" postCellId="../IzPop1/0"/>
                <connection id="1" preCellId="../IzPop0[0]" postCellId="../IzPop1/1"/>
                <connection id="2" preCellId="../IzPop0[0]" postCellId="../IzPop1/2"/>
                ...
            </projection>
            <explicitInput target="IzPop0[0]" input="pg_0"/>
            <explicitInput target="IzPop0[1]" input="pg_1"/>
            <explicitInput target="IzPop0[2]" input="pg_2"/>
            <explicitInput target="IzPop0[3]" input="pg_3"/>
            <explicitInput target="IzPop0[4]" input="pg_4"/>
        </network>
    </neuroml>
:::
:::

Here, in the [`explicitInput`{.docutils .literal .notranslate}]{.pre}
node, we need to refer to neurons of the [`IzPop0`{.docutils .literal
.notranslate}]{.pre} [`population`{.docutils .literal
.notranslate}]{.pre} node. Since [`explicitInput`{.docutils .literal
.notranslate}]{.pre} and [`population`{.docutils .literal
.notranslate}]{.pre} are *siblings* (both have the [`IzNet`{.docutils
.literal .notranslate}]{.pre} [`network`{.docutils .literal
.notranslate}]{.pre} as *parent*), they are at the same *level*.
Therefore, in [`explicitInput`{.docutils .literal .notranslate}]{.pre},
one can refer directly to [`IzPop0`{.docutils .literal
.notranslate}]{.pre}.

The [`projection`{.docutils .literal .notranslate}]{.pre} and
[`population`{.docutils .literal .notranslate}]{.pre} nodes are also
*siblings* and therefore are at the same level. So, in the
[`projection`{.docutils .literal .notranslate}]{.pre} tag also, we can
refer to the [`population`{.docutils .literal .notranslate}]{.pre} nodes
directly. The [`connection`{.docutils .literal .notranslate}]{.pre}
nodes, however, are *children* of the [`projection`{.docutils .literal
.notranslate}]{.pre} node. Therefore, for the [`connection`{.docutils
.literal .notranslate}]{.pre} nodes, the [`population`{.docutils
.literal .notranslate}]{.pre} nodes are at the *parent* level, and we
must use [`../IzPop0`{.docutils .literal .notranslate}]{.pre} to refer
to them.

[`../IzPop0`{.docutils .literal .notranslate}]{.pre} means "go up one
level to the parent level (to [`projection`{.docutils .literal
.notranslate}]{.pre}) and then refer to [`IzPop0`{.docutils .literal
.notranslate}]{.pre}". [`../`{.docutils .literal .notranslate}]{.pre}
can be used as many times as required and wherever required in the path.
For example, [`../../../`{.docutils .literal .notranslate}]{.pre} would
mean "go up three levels".
:::

::: {#helper-functions-in-pyneuroml .section}
[]{#userdocs-pyneuroml}

#### Helper functions in pyNeuroML[\#](#helper-functions-in-pyneuroml "Link to this heading"){.headerlink}

::: {.admonition .note}
Note

These functions require [[pyNeuroML]{.std
.std-ref}](Landing.html#pyneuroml){.reference .internal} version
0.5.18+, and [[pylems]{.std .std-ref}](Landing.html#pylems){.reference
.internal} version 0.5.8+.
:::

From version 0.5.18, [[pyNeuroML]{.std
.std-ref}](Landing.html#pyneuroml){.reference .internal} includes the
[list_recording_paths_for_exposures](https://pyneuroml.readthedocs.io/en/development/pyneuroml.html#pyneuroml.pynml.list_recording_paths_for_exposures){.reference
.external} helper function that can list the exposures and their
recordable paths from a NeuroML 2 model:

::: {.highlight-pycon .notranslate}
::: highlight
    >>> import pyneuroml.pynml
    >>> help(pynml.list_recording_paths_for_exposures)

    Help on function list_recording_paths_for_exposures in module pyneuroml.pynml:

    list_recording_paths_for_exposures(nml_doc_fn, substring='', target='')
        List the recording path strings for exposures.

        This wraps around `lems.model.list_recording_paths` to list the recording
        paths in the given NeuroML2 model. The only difference between the two is
        that the `lems.model.list_recording_paths` function is not aware of the
        NeuroML2 component types (since it's for any LEMS models in general), but
        this one is.
:::
:::

It can be run on the example [[Izhikevich network example]{.std
.std-ref}](Landing.html#userdocs-gettingstarted-izhikevichnetwork){.reference
.internal}:

::: {.highlight-pycon .notranslate}
::: highlight
    >>> pynml.list_recording_paths_for_exposures("izhikevich2007_network.nml", substring="", target="IzNet")
    ['IzNet/IzPop0[0]/iMemb',
     'IzNet/IzPop0[0]/iSyn',
     'IzNet/IzPop0[0]/u',
     'IzNet/IzPop0[0]/v',
     'IzNet/IzPop0[1]/iMemb',
     'IzNet/IzPop0[1]/iSyn',
     'IzNet/IzPop0[1]/u',
     'IzNet/IzPop0[1]/v',
     'IzNet/IzPop0[2]/iMemb',
     'IzNet/IzPop0[2]/iSyn',
     'IzNet/IzPop0[2]/u',
     'IzNet/IzPop0[2]/v',
     'IzNet/IzPop0[3]/iMemb',
     'IzNet/IzPop0[3]/iSyn',
     'IzNet/IzPop0[3]/u',
     'IzNet/IzPop0[3]/v',
     'IzNet/IzPop0[4]/iMemb',
     'IzNet/IzPop0[4]/iSyn',
     'IzNet/IzPop0[4]/u',
     'IzNet/IzPop0[4]/v',
     'IzNet/IzPop1[0]/iMemb',
    ..
    ]
:::
:::

Note that this function parsers the model description only, not the
built simulation description. Therefore, it will not necessarily list
the complete list of paths. Also worth noting is that since it parses
and iterates over the expanded representation of the model, it can be
slow and return long lists of results on larger models. It is therefore,
best to use this with the [`substring`{.docutils .literal
.notranslate}]{.pre} option to narrow its scope.

An associated helper function
[list_exposures](https://pyneuroml.readthedocs.io/en/development/pyneuroml.html?highlight=list_exposures#pyneuroml.pynml.list_exposures){.reference
.external} is also available:

::: {.highlight-pycon .notranslate}
::: highlight
    >>> import pyneuroml.pynml
    >>> help(pynml.list_exposures)

    list_exposures(nml_doc_fn, substring='')
        List exposures in a NeuroML model document file.

        This wraps around `lems.model.list_exposures` to list the exposures in a
        NeuroML2 model. The only difference between the two is that the
        `lems.model.list_exposures` function is not aware of the NeuroML2 component
        types (since it's for any LEMS models in general), but this one is.

        The returned dictionary is of the form:

        ..
            {
                "component": ["exp1", "exp2"]
            }
:::
:::

When run on the example [[Izhikevich network example]{.std
.std-ref}](Landing.html#userdocs-gettingstarted-izhikevichnetwork){.reference
.internal}, it will return:

::: {.highlight-pycon .notranslate}
::: highlight
    >>> pynml.list_exposures("izhikevich2007_network.nml")

    {<lems.model.component.FatComponent at 0x7f25b62caca0>: {'g': <lems.model.component.Exposure at 0x7f25dd1d2be0>,
      'i': <lems.model.component.Exposure at 0x7f25dc921e80>},
     <lems.model.component.FatComponent at 0x7f25b62cad00>: {'u': <lems.model.component.Exposure at 0x7f25b5f57400>,
      'iSyn': <lems.model.component.Exposure at 0x7f25b607a670>,
      'iMemb': <lems.model.component.Exposure at 0x7f25b607aa00>,
      'v': <lems.model.component.Exposure at 0x7f25b6500220>},
     <lems.model.component.FatComponent at 0x7f25b62cadf0>: {'i': <lems.model.component.Exposure at 0x7f25dc921e80>},
     <lems.model.component.FatComponent at 0x7f25b62caf70>: {'i': <lems.model.component.Exposure at 0x7f25dc921e80>},
     <lems.model.component.FatComponent at 0x7f25b5fc2ac0>: {'i': <lems.model.component.Exposure at 0x7f25dc921e80>},
     <lems.model.component.FatComponent at 0x7f25b65be9d0>: {'i': <lems.model.component.Exposure at 0x7f25dc921e80>},
     <lems.model.component.FatComponent at 0x7f25b65bed00>: {'i': <lems.model.component.Exposure at 0x7f25dc921e80>},
    ..
    }
:::
:::

This second function is primarily for use by the
[`list_recording_paths_for_exposures`{.docutils .literal
.notranslate}]{.pre} function.

As noted in the helper documentation, these are both based on a function
of the same name implemented in [[PyLEMS]{.std
.std-ref}](Landing.html#pylems){.reference .internal}, version 0.5.8+.
:::
:::

[]{#document-Userdocs/QuantitiesAndRecording}

::: {#quantities-and-recording .section .tex2jax_ignore .mathjax_ignore}
[]{#userdocs-quantitiesandrecording}

### Quantities and recording[\#](#quantities-and-recording "Link to this heading"){.headerlink}

In LEMS and NeuroML, [`quantities`{.docutils .literal
.notranslate}]{.pre} from all [`exposures`{.docutils .literal
.notranslate}]{.pre} and all [`events`{.docutils .literal
.notranslate}]{.pre} can be recorded by referring to them using
[[paths]{.std .std-ref}](Landing.html#userdocs-paths){.reference
.internal}. For examples, please see the [[Getting Started with
NeuroML]{.std
.std-ref}](Landing.html#userdocs-getting-started-neuroml){.reference
.internal} section.

::: {#recording-events .section}
[]{#userdocs-quantitiesandrecording-events}

#### Recording events[\#](#recording-events "Link to this heading"){.headerlink}

In NeuroML, all [`event`{.docutils .literal .notranslate}]{.pre}s can be
recorded to files declared using the [[EventOutputFile]{.std
.std-ref}](Landing.html#schema-eventoutputfile){.reference .internal}
component. Once an [`EventOutputFile`{.docutils .literal
.notranslate}]{.pre} has been declared, events to record can be selected
using the [[EventSelection]{.std
.std-ref}](Landing.html#schema-eventselection){.reference .internal}
component.

[[pyNeuroML]{.std .std-ref}](Landing.html#pyneuroml){.reference
.internal} provides the
[create_event_output_file](https://pyneuroml.readthedocs.io/en/development/pyneuroml.lems.html?highlight=add_selection_to_event_output_file#pyneuroml.lems.LEMSSimulation.LEMSSimulation.create_event_output_file){.reference
.external} function to create a [`EventOutputFile`{.docutils .literal
.notranslate}]{.pre} to record [`events`{.docutils .literal
.notranslate}]{.pre} to, and the
[add_selection_to_event_output_file](https://pyneuroml.readthedocs.io/en/development/pyneuroml.lems.html?highlight=add_selection_to_event_output_file#pyneuroml.lems.LEMSSimulation.LEMSSimulation.add_selection_to_event_output_file){.reference
.external} function to record [`events`{.docutils .literal
.notranslate}]{.pre} to the declared data file(s).
:::

::: {#recording-quantities-from-exposures .section}
[]{#userdocs-quantitiesandrecording-exposures}

#### Recording quantities from exposures[\#](#recording-quantities-from-exposures "Link to this heading"){.headerlink}

In NeuroML, all [`quantities`{.docutils .literal .notranslate}]{.pre}
can be recorded to files declared using the [[OutputFile]{.std
.std-ref}](Landing.html#schema-outputfile){.reference .internal}
component. Once the [`OutputFile`{.docutils .literal
.notranslate}]{.pre} has been declared, [`quantities`{.docutils .literal
.notranslate}]{.pre} to record can be selected using the
[[OutputColumn]{.std
.std-ref}](Landing.html#schema-outputcolumn){.reference .internal}
component.

[[pyNeuroML]{.std .std-ref}](Landing.html#pyneuroml){.reference
.internal} provides the
[create_output_file](https://pyneuroml.readthedocs.io/en/development/pyneuroml.lems.html?highlight=add_selection_to_event_output_file#pyneuroml.lems.LEMSSimulation.LEMSSimulation.create_event_output_file){.reference
.external} function to create a [`OutputFile`{.docutils .literal
.notranslate}]{.pre} to record [`quantities`{.docutils .literal
.notranslate}]{.pre} to, and the
[add_column_to_output_file](https://pyneuroml.readthedocs.io/en/development/pyneuroml.lems.html?highlight=add_selection_to_event_output_file#pyneuroml.lems.LEMSSimulation.LEMSSimulation.create_event_output_file){.reference
.external} function to select [`quantities`{.docutils .literal
.notranslate}]{.pre} to record to the declared data file(s).
:::
:::

[]{#document-Userdocs/LEMSSimulation}

::: {#lems-simulation-files .section .tex2jax_ignore .mathjax_ignore}
[]{#userdocs-lemssimulation}

### LEMS Simulation files[\#](#lems-simulation-files "Link to this heading"){.headerlink}

For many users, the most obvious place that LEMS is used is in the LEMS
Simulation file (usually [[LEMS\_\*.xml]{.std
.std-ref}](Landing.html#userdocs-conventions-files){.reference
.internal}).

In short, what a file like this does is:

-   point at the NeuroML file containing the model to simulate

-   include any other LEMS file it needs, including the [[NeuroML core
    type definitions]{.std
    .std-ref}](Landing.html#userdocs-neuromlv2inlems){.reference
    .internal}

-   specify how long to run the simulation for and the simulation
    timestep (dt)

-   say what to display when the simulation has finished (e.g. membrane
    potentials of selected cells)

-   say what to save to file, e.g. voltage traces, spike times

These files are crucial in many of the workflows for [[simulating
NeuroML models]{.std
.std-ref}](Landing.html#userdocs-simulators){.reference .internal}, and
are reused across different simulator targets, e.g. [`jnml`{.docutils
.literal .notranslate}]{.pre}` `{.docutils .literal
.notranslate}[`LEMS_MyNetwork.xml`{.docutils .literal
.notranslate}]{.pre} (run in jNeuroML), [`jnml`{.docutils .literal
.notranslate}]{.pre}` `{.docutils .literal
.notranslate}[`LEMS_MyNetwork.xml`{.docutils .literal
.notranslate}]{.pre}` `{.docutils .literal
.notranslate}[`-neuron`{.docutils .literal .notranslate}]{.pre} (convert
to NEURON), [`jnml`{.docutils .literal .notranslate}]{.pre}` `{.docutils
.literal .notranslate}[`LEMS_MyNetwork.xml`{.docutils .literal
.notranslate}]{.pre}` `{.docutils .literal
.notranslate}[`-brian2`{.docutils .literal .notranslate}]{.pre} (convert
to Brian2). See [[here]{.std
.std-ref}](Landing.html#userdocs-simulating-models-jnml){.reference
.internal} for more information.

<figure id="id1" class="align-center">
<a href="_images/lems_nml_files.png"
class="reference internal image-reference"><img
src="_images/lems_nml_files.png"
style="width: 647.52px; height: 409.44px;"
alt="LEMS Simulation file and NeuroML file" /></a>
<figcaption><p><span class="caption-number">Fig. 61 </span><span
class="caption-text">Typical organisation for a NeuroML simulation. The
main NeuroML model is specified in a file with the network (<span
class="pre"><code
class="docutils literal notranslate">*.net.nml</code></span>), which can
include/point to files containing individual synapses (<span
class="pre"><code
class="docutils literal notranslate">*.synapse.nml</code></span>) or
cell files (<span class="pre"><code
class="docutils literal notranslate">*.cell.nml</code></span>). If the
latter are conductance based, they may include external channel files
(<span class="pre"><code
class="docutils literal notranslate">*.channel.nml</code></span>). The
main LEMS Simulation file only needs to include the network file, and
tools for running simulations of the model refer to just this LEMS file.
Exceptions to these conventions are frequent and simulations will run
perfectly well with all the elements inside the main LEMS file, but
using this scheme will maximise reusability of model elements.</span><a
href="#id1" class="headerlink"
title="Link to this image">#</a></p></figcaption>
</figure>

::: {#specification-of-format .section}
#### Specification of format[\#](#specification-of-format "Link to this heading"){.headerlink}

See [[here]{.std .std-ref}](Landing.html#schema-simulation){.reference
.internal} for definition of the main elements used in the file,
including [[Display]{.std
.std-ref}](Landing.html#schema-display){.reference .internal},
[[OutputFile]{.std .std-ref}](Landing.html#schema-outputfile){.reference
.internal}, etc.
:::

::: {#quantities-and-paths .section}
#### Quantities and paths[\#](#quantities-and-paths "Link to this heading"){.headerlink}

Specifying the quantities to save/display in a LEMS Simulation file is
an important and sometimes confusing process. There is a [[dedicated
page]{.std
.std-ref}](Landing.html#userdocs-quantitiesandrecording){.reference
.internal} on quantities and paths in LEMS and NeuroML2.
:::

::: {#creating-lems-simulation-files .section}
#### Creating LEMS Simulation files[\#](#creating-lems-simulation-files "Link to this heading"){.headerlink}

Perhaps the easiest way to create a LEMS Simulation file is to base it
off of an existing example.

::: {.highlight-xml .notranslate}
::: highlight
    <Lems>

        <!-- Specify the Simulation element below as what LEMS should load. Save a
             report of the simulation (e.g. simulator version, run time) in a file-->
        <Target component="sim1" reportFile="report.txt"/>

        <Include file="Cells.xml"/>
        <Include file="Networks.xml"/>
        <Include file="Simulation.xml"/>

        <!-- Including file with a <neuroml> root, a "real" NeuroML 2 file -->
        <Include file="NML2_SingleCompHHCell.nml"/>

        <!-- What to run (from the above NeuroML file) and what duration/timestep -->
        <Simulation id="sim1" length="300ms" step="0.01ms" target="net1">

            <!-- Display a trace in a new window -->
            <Display id="d1" title="HH cell with simple morphology: voltage" timeScale="1ms" xmin="0" xmax="300" ymin="-90" ymax="50">
                    <Line id="v" quantity="hhpop[0]/v" color="#cccccc" scale="0.001" timeScale="1ms"/>
            </Display>

            <!-- Save a variable to file  -->
            <OutputFile id="of0" fileName="ex_v.dat">
                <OutputColumn id="v" quantity="hhpop[0]/v"/>
            </OutputFile>

            <!-- Save spike times from a cell to file  -->
            <EventOutputFile id="spikes" fileName="ex.spikes" format="TIME_ID">
                <EventSelection id="0" select="hhpop[0]" eventPort="spike"/>
            </EventOutputFile>

        </Simulation>

    </Lems>
:::
:::

Alternatively, it is possible to create a LEMS Simulation file in Python
file using pyNeuroML:

::: {.highlight-python .notranslate}
::: highlight
    from pyneuroml.lems import LEMSSimulation

    ls = LEMSSimulation('sim1', 500, 0.05, 'net1')
    ls.include_neuroml2_file('NML2_SingleCompHHCell.nml')

    ls.create_display('display0', "Voltages", "-90", "50")
    ls.add_line_to_display('display0', "v", "hhpop[0]/v", "1mV", "#ffffff")

    ls.create_output_file('Volts_file', "v.dat")
    ls.add_column_to_output_file('Volts_file', 'v', "hhpop[0]/v")

    ls.save_to_file()
:::
:::

See [this
example](https://github.com/NeuroML/pyNeuroML/blob/master/examples/create_new_lems_file.py){.reference
.external} for more details.
:::

::: {#what-about-sed-ml .section}
#### What about SED-ML?[\#](#what-about-sed-ml "Link to this heading"){.headerlink}

The Simulation Experiment Description Markup Language
([SED-ML](https://sed-ml.org/){.reference .external}) is used by a
number of other initiatives such as SBML for specifying simulation
setup, execution and basic analysis.

We chose to have a LEMS specific format for specifying simulations in
NeuroML2 as opposed to natively supporting SED-ML, mainly because of the
tight link to the LEMS language and [[jLEMS]{.std
.std-ref}](Landing.html#jlems){.reference .internal} package, i.e. all
of the NeuroML2 elements and elements in a LEMS simulation file have
underlying definitions in the LEMS language. However it is possible to
convert the LEMS simulation to the equivalent in SED-ML.

::: {#exporting-lems-simulation-descriptions-to-sed-ml .section}
##### Exporting LEMS simulation descriptions to SED-ML[\#](#exporting-lems-simulation-descriptions-to-sed-ml "Link to this heading"){.headerlink}

::: {.highlight-console .notranslate}
::: highlight
    # Using jnml
    jnml <LEMS simulation file> -sedml

    # Using pynml
    pynml <LEMS simulation file> -sedml
:::
:::
:::
:::
:::
:::
:::

[]{#document-Userdocs/ExtendingNeuroMLv2}

::: {#extending-neuroml .section .tex2jax_ignore .mathjax_ignore}
[]{#userdocs-extending}

## Extending NeuroML[\#](#extending-neuroml "Link to this heading"){.headerlink}

As a language, LEMS defines a set of built-in [`types`{.docutils
.literal .notranslate}]{.pre} which can be used together to build more
user-defined types. For example, Python defines [`int`{.docutils
.literal .notranslate}]{.pre}, [`float`{.docutils .literal
.notranslate}]{.pre}, [`str`{.docutils .literal .notranslate}]{.pre} and
so on as built-in types, and these can then be combined to define user
defined types, classes. An object of a particular class/type can be
instantiated by supplying values for the members defined in the
class/type.

**ComponentTypes** in LEMS are similar to classes in Python. They define
the membership structure of the type, but they do not specify values for
their members. Once a ComponentType has been defined, an instance of it
can be created by setting values for its members. This object is
referred to as a **Component** in LEMS.

Having definitions in LEMS allows their re-use, and all new
ComponentTypes can be submitted for inclusion to the NeuroMLv2
specification to be made accessible to other users.

-   Like NeuroML, LEMS also has a [well defined
    schema](https://github.com/LEMS/LEMS/tree/master/Schemas/LEMS){.reference
    .external} (XSD) that is used to validate LEMS XML files.

-   Also similar to NeuroML, you can use the [[LEMS Python tools]{.std
    .std-ref}](Landing.html#pylems){.reference .internal} to work with
    LEMS and do not need to work directly with the XML files.

The NeuroML2 standard is a list of [[curated LEMS ComponentTypes]{.std
.std-ref}](Landing.html#userdocs-neuromlv2){.reference .internal}. In
cases where the set of ComponentTypes defined in the NeuroML standard is
not sufficient for a particular modelling project, new ComponentTypes
can be defined to extend the NeuroMLv2 standard.

::: {#creating-new-componenttypes-with-existing-neuroml-componenttypes .section}
[]{#userdocs-extending-withneuroml}

### Creating new ComponentTypes with existing NeuroML ComponentTypes[\#](#creating-new-componenttypes-with-existing-neuroml-componenttypes "Link to this heading"){.headerlink}

Existing ComponentTypes defined in the NeuroMLv2 standard, when
sufficient, should be used to create new ComponentTypes. These new
ComponentTypes, since they consist of NeuroMLv2 ComponentTypes, *will be
valid against the NeuroMLv2 schema* (must use a [`<neuroml`{.code
.docutils .literal .notranslate}]{.pre}` `{.code .docutils .literal
.notranslate}[`..>`{.code .docutils .literal .notranslate}]{.pre} root
element). For convenience, the NeuroMLv2 schema includes a subset of the
[[LEMS elements]{.std
.std-ref}](Landing.html#userdocs-lemsschema){.reference .internal}.

An example of this type of extension of NeuroML can be see
[here](https://github.com/OpenSourceBrain/BlueBrainProjectShowcase/blob/master/NMC/NeuroML2/Ca_LVAst.channel.nml#L42){.reference
.external} where a new Calcium dependent ion channel Component requires
a new ComponentType [`Ca_LVAst_m_tau_tau`{.code .docutils .literal
.notranslate}]{.pre} that implements the time course of the gate.

However, please note that while the ComponentType will be valid NeuroML,
the new Components (instances) one creates of this ComponentType (and
models where Components are referenced/used) *will not*---since the
NeuroML schema *does not know of the new ComponentType*. The new
Components (and the models) will be valid LEMS. For this reason, while
the ComponentType file will use a [`<neuroml`{.code .docutils .literal
.notranslate}]{.pre}` `{.code .docutils .literal
.notranslate}[`..>`{.code .docutils .literal .notranslate}]{.pre} root
tag, the file containing its instantiated Components will use the
[`<Lems`{.code .docutils .literal .notranslate}]{.pre}` `{.code
.docutils .literal .notranslate}[`..>`{.code .docutils .literal
.notranslate}]{.pre} root tag.
:::

::: {#creating-new-componenttypes-with-lems-elements .section}
[]{#userdocs-extending-withlems}

### Creating new ComponentTypes with LEMS elements[\#](#creating-new-componenttypes-with-lems-elements "Link to this heading"){.headerlink}

When ComponentTypes from the NeuroMLv2 standard are not sufficient for
the creation of new ComponentTypes, one must use LEMS elements to do so.
The definitions of the [[NeuroMLv2 standard core ComponentTypes]{.std
.std-ref}](Landing.html#userdocs-neuromlv2inlems){.reference .internal}
are examples of this.

::: {#lems-elements .section}
[]{#userdocs-extending-withlems-types}

#### LEMS elements[\#](#lems-elements "Link to this heading"){.headerlink}

The list of built-in types provided by LEMS can be seen [[in the LEMS
documentation]{.std
.std-ref}](Landing.html#userdocs-lemsschema){.reference .internal}. As
the documentation notes, a ComponentType is the "Root element for
defining component types". It must contain a [`name`{.docutils .literal
.notranslate}]{.pre}, and can [`extend`{.docutils .literal
.notranslate}]{.pre} another ComponentType, thus inheriting its
members/attributes. Each ComponentType can contain members of other LEMS
types: [`Parameter`{.docutils .literal .notranslate}]{.pre},
[`DerivedParameter`{.docutils .literal .notranslate}]{.pre},
[`Dynamics`{.docutils .literal .notranslate}]{.pre},
[`Exposure`{.docutils .literal .notranslate}]{.pre} and so on.
:::

::: {#example-lorenz-model-for-cellular-convection .section}
[]{#userdocs-extending-withlems-examplexml}

#### Example: Lorenz model for cellular convection[\#](#example-lorenz-model-for-cellular-convection "Link to this heading"){.headerlink}

To see how to create new ComponentTypes using LEMS, let us create one
that is not neuroscience specific. We will first create it using the
plain XML and then see how it can be done using the Python pyLEMS API.

For this example, we will use the Lorenz model for cellular convection
[\[[Lor63](Landing.html#id16 "Edward N. Lorenz. Deterministic nonperiodic flow. Journal of Atmospheric Sciences, 20(2):130–141, 1963. doi:10.1175/1520-0469(1963)020<0130:dnf>2.0.co;2."){.reference
.internal}\]]{#id1}. The [Wikipedia
article](https://en.wikipedia.org/wiki/Lorenz_system#Overview){.reference
.external} provides a short summary of the model, and the equations that
govern it:

::: {#equation-39588480-0900-457c-872a-a79b84d928da .amsmath .math .notranslate .nohighlight}
[(1)[\#](#equation-39588480-0900-457c-872a-a79b84d928da "Permalink to this equation"){.headerlink}]{.eqno}\\\[\\begin{align}
\\frac{dx}{dt} &= \\sigma (y - x) \\\\ \\frac{dy}{dt} &= x (\\rho - z) -
y \\\\ \\frac{dz}{dt} &= xy - \\beta z \\end{align}\\\]
:::

So we can see here that we have three parameters:

-   [\\(\\sigma\\)]{.math .notranslate .nohighlight}

-   [\\(\\rho\\)]{.math .notranslate .nohighlight}

-   [\\(\\beta\\)]{.math .notranslate .nohighlight}

Next, [`x`{.docutils .literal .notranslate}]{.pre}, [`y`{.docutils
.literal .notranslate}]{.pre}, and [`z`{.docutils .literal
.notranslate}]{.pre} are the *state variables* for this model, with
initial values [`x0`{.docutils .literal .notranslate}]{.pre},
[`y0`{.docutils .literal .notranslate}]{.pre}, and [`z0`{.docutils
.literal .notranslate}]{.pre} respectively. We also want to be able to
observe the values of [`x`{.docutils .literal .notranslate}]{.pre},
[`y`{.docutils .literal .notranslate}]{.pre}, and [`z`{.docutils
.literal .notranslate}]{.pre}, so they must be *exposed* in the LEMS
definition.

Let us start with the XML definition of a ComponentType that will
describe this model. Each XML file must start with a [`<Lems>`{.docutils
.literal .notranslate}]{.pre} "root node". This includes information
about the version of the LEMS schema that this document is valid
against. In this case, we document that this LEMS file should be valid
against version 0.7.6 of the LEMS schema.

::: {.highlight-xml .notranslate}
::: highlight
    <Lems xmlns="http://www.neuroml.org/lems/0.7.6"
          xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
          xsi:schemaLocation="http://www.neuroml.org/lems/0.7.6 https://raw.github.com/LEMS/LEMS/master/Schemas/LEMS/LEMS_v0.7.6.xsd">

          <ComponentType name="lorenz1963" description="The Lorenz system is a simplified model for atomspheric convection, derived from the Navier Stokes equations.">

            <!-- Parameters: free parameters to be used in the model -->
            <Parameter name="sigma" dimension="none" description="Prandtl Number"/>
            <Parameter name="beta" dimension="none" description="Also named b elsewhere"/>
            <Parameter name="rho" dimension="none" description="Related to the Rayleigh number, also named r elsewhere"/>

            <!-- Initial Conditions: also free parameters to be set when creating a Component from the ComponentType -->
            <Parameter name="x0" dimension="none"/>
            <Parameter name="y0" dimension="none"/>
            <Parameter name="z0" dimension="none"/>


            <!-- Exposure: what we want to be able to record from the LEMS simulation -->
            <Exposure name="x" dimension="none"/>
            <Exposure name="y" dimension="none"/>
            <Exposure name="z" dimension="none"/>
          </ComponentType>
    </Lems>
:::
:::

Note that each parameter has a *dimension*, not a *unit*. This is
because LEMS allows us to use any valid units for each dimension, and
takes care of the conversion factors and so on. NeuroML also takes
advantage of this LEMS feature, as noted [[here]{.std
.std-ref}](Landing.html#userdocs-unitsanddimensions){.reference
.internal}.

Now, we can define the *dynamics* of the model, summarised in the
equations above:

::: {.highlight-xml .notranslate}
::: highlight
            <Dynamics>
                <!-- State variables: linked to Exposures so that they can be accessed -->
                <StateVariable name="x" dimension="none" exposure="x"/>
                <StateVariable name="y" dimension="none" exposure="y"/>
                <StateVariable name="z" dimension="none" exposure="z"/>

                <!-- Equations defining the dynamics of each state variable -->
                <TimeDerivative variable="x" value="( sigma * (y - x) ) / sec"/>
                <TimeDerivative variable="y" value="( rho * x - y - x * z ) / sec"/>
                <TimeDerivative variable="z" value="( x * y - beta * z) / sec"/>

                <!-- Actions to take on the start of a LEMS simulation -->
                <OnStart>
                    <StateAssignment variable="x" value="x0"/>
                    <StateAssignment variable="y" value="y0"/>
                    <StateAssignment variable="z" value="z0"/>
                </OnStart>
            </Dynamics>
:::
:::

Our LEMS file is almost complete. However, notice that we have used
[`sec`{.docutils .literal .notranslate}]{.pre} in the dynamics to denote
time but have not yet declared it. We define [`sec`{.docutils .literal
.notranslate}]{.pre} as a *constant* whose value is defined in the
ComponentType itself (and will not be set by us when instantiating a
Component of this ComponentType):

::: {.highlight-xml .notranslate}
::: highlight
            <Constant name="sec" dimension="time" value="1s"/>
:::
:::

Also note that while we have defined this constant, we have not yet
defined the [`time`{.docutils .literal .notranslate}]{.pre} dimension or
its units. We can do that outside the ComponentType:

::: {.highlight-xml .notranslate}
::: highlight
      <Dimension name="time" t="1"/>
      <Unit name="second" symbol="s" dimension="time" power="1"/>
      <Unit name="milli second" symbol="ms" dimension="time" power="-3"/>
:::
:::

We have defined two units for the time dimension, with their conversion
factors. LEMS will use this information to correctly convert all
dimensions as required. The NeuroMLv2 standard defines various
dimensions and their units [[in the schema]{.std
.std-ref}](Landing.html#schema-neuromlcoredimensions){.reference
.internal} for us to use.

The complete LEMS file will be this:

::: {.highlight-xml .notranslate}
::: highlight
    <Lems xmlns="http://www.neuroml.org/lems/0.7.6"
          xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
          xsi:schemaLocation="http://www.neuroml.org/lems/0.7.6 https://raw.github.com/LEMS/LEMS/master/Schemas/LEMS/LEMS_v0.7.6.xsd">

      <Dimension name="time" t="1"/>
      <Unit name="second" symbol="s" dimension="time" power="1"/>
      <Unit name="milli second" symbol="ms" dimension="time" power="-3"/>

      <ComponentType name="lorenz1963" description="The Lorenz system is a simplified model for atomspheric convection, derived from the Navier Stokes equations.">

        <!-- Parameters: free parameters to be used in the model -->
        <Parameter name="sigma" dimension="none" description="Prandtl Number"/>
        <Parameter name="beta" dimension="none" description="Also named b elsewhere"/>
        <Parameter name="rho" dimension="none" description="Related to the Rayleigh number, also named r elsewhere"/>

        <!-- Initial Conditions: also free parameters to be set when creating a Component from the ComponentType -->
        <Parameter name="x0" dimension="none"/>
        <Parameter name="y0" dimension="none"/>
        <Parameter name="z0" dimension="none"/>


        <!-- Exposure: what we want to be able to record from the LEMS simulation -->
        <Exposure name="x" dimension="none"/>
        <Exposure name="y" dimension="none"/>
        <Exposure name="z" dimension="none"/>
        <Constant name="sec" dimension="time" value="1s"/>

        <Dynamics>
            <!-- State variables: linked to Exposures so that they can be accessed -->
            <StateVariable name="x" dimension="none" exposure="x"/>
            <StateVariable name="y" dimension="none" exposure="y"/>
            <StateVariable name="z" dimension="none" exposure="z"/>

            <!-- Equations defining the dynamics of each state variable -->
            <TimeDerivative variable="x" value="( sigma * (y - x) ) / sec"/>
            <TimeDerivative variable="y" value="( rho * x - y - x * z ) / sec"/>
            <TimeDerivative variable="z" value="( x * y - beta * z) / sec"/>

            <!-- Actions to take on the start of a LEMS simulation -->
            <OnStart>
                <StateAssignment variable="x" value="x0"/>
                <StateAssignment variable="y" value="y0"/>
                <StateAssignment variable="z" value="z0"/>
            </OnStart>
        </Dynamics>
      </ComponentType>
    </Lems>
:::
:::

We now have a complete LEMS model declaration. To use this model, we
need to create an instance of the ComponentType, a Component. This
requires us to set the values of various parameters of the defined
model:

::: {.highlight-xml .notranslate}
::: highlight
    <lorenz1963 id="lorenzCell" sigma="10" beta="2.67" rho="28"
        x0="1.0" y0="1.0" z0="1.0"/>
:::
:::

Here, we've set parameters that result in the chaotic attractor regime.
We could also use different values for the parameters---like a class can
have many many objects with different parameters, a ComponentType can
have also have different Components.

Note that one can also define a Component using the standard constructor
form:

::: {.highlight-xml .notranslate}
::: highlight
      <Component id="lorenzCell" type="lorenz1963" sigma="10" beta="2.67" rho="28" x0="1.0" y0="1.0" z0="1.0"/>
:::
:::

The two forms are equivalent. As with other conventions, either form can
be used as long as it is used consistently.

The [`Include`{.docutils .literal .notranslate}]{.pre} element type
allows us to modularise our models. In NeuroML based models, we use it
to break our model down into small independent reusable files.
:::

::: {#writing-the-model-in-python-using-pylems .section}
[]{#userdocs-extending-lorenz-python}

#### Writing the model in Python using PyLEMS[\#](#writing-the-model-in-python-using-pylems "Link to this heading"){.headerlink}

While the underlying format for NeuroML and LEMS is XML, Python is the
suggested programming language for end users. In this section we will
see how the Lorenz model can be written using the [[PyLEMS]{.std
.std-ref}](Landing.html#pylems){.reference .internal} Python LEMS API.
The complete script is below:

::: {.highlight-python .notranslate}
::: highlight
    #!/usr/bin/env python3

    import lems.api as lems
    from lems.base.util import validate_lems

    model = lems.Model()

    model.add(lems.Dimension(name="time", t=1))
    model.add(lems.Unit(name="second", symbol="s", dimension="time", power=1))
    model.add(lems.Unit(name="milli second", symbol="ms", dimension="time", power=-3))

    lorenz = lems.ComponentType(name="lorenz1963", description="The Lorenz system is a simplified model for atomspheric convection, derived from the Navier Stokes equations")
    model.add(lorenz)

    lorenz.add(lems.Parameter(name="sigma", dimension="none", description="Prandtl Number"))
    lorenz.add(lems.Parameter(name="beta", dimension="none", description="Also named b elsewhere"))
    lorenz.add(lems.Parameter(name="rho", dimension="none", description="Related to the Rayleigh number, also named r elsewhere"))


    lorenz.add(lems.Parameter(name="x0", dimension="none"))
    lorenz.add(lems.Parameter(name="y0", dimension="none"))
    lorenz.add(lems.Parameter(name="z0", dimension="none"))

    lorenz.add(lems.Exposure(name="x", dimension="none"))
    lorenz.add(lems.Exposure(name="y", dimension="none"))
    lorenz.add(lems.Exposure(name="z", dimension="none"))

    lorenz.add(lems.Constant(name="sec", value="1s", dimension="time"))

    lorenz.dynamics.add(lems.StateVariable(name="x", dimension="none", exposure="x"))
    lorenz.dynamics.add(lems.StateVariable(name="y", dimension="none", exposure="y"))
    lorenz.dynamics.add(lems.StateVariable(name="z", dimension="none", exposure="z"))

    lorenz.dynamics.add(lems.TimeDerivative(variable="x", value="( sigma * (y - x)) / sec"))
    lorenz.dynamics.add(lems.TimeDerivative(variable="y", value="( rho * x - y - x * z ) / sec"))
    lorenz.dynamics.add(lems.TimeDerivative(variable="z", value="( x * y - beta * z) / sec"))

    onstart = lems.OnStart()
    onstart.add(lems.StateAssignment(variable="x", value="x0"))
    onstart.add(lems.StateAssignment(variable="y", value="y0"))
    onstart.add(lems.StateAssignment(variable="z", value="z0"))
    lorenz.dynamics.add(onstart)


    model.add(lems.Component(id_="lorenzCell", type_=lorenz.name, sigma="10",
                             beta="2.67", rho="28", x0="1.0", y0="1.0", z0="1.0"))

    file_name = "LEMS_lorenz.xml"
    model.export_to_file(file_name)


    validate_lems(file_name)
:::
:::

As you will see, the PyLEMS API exactly follows the XML constructs that
we used before. Running this script, let's call it
[`LorenzLems.py`{.docutils .literal .notranslate}]{.pre} gives us:

::: {.highlight-console .notranslate}
::: highlight
    $ python LorenzLems.py
    Validating LEMS_lorenz.xml against https://raw.githubusercontent.com/LEMS/LEMS/development/Schemas/LEMS/LEMS_v0.7.6.xsd
    It's valid!
:::
:::

The generated XML file is below. As you can see, it is identical to the
XML file that we wrote by hand in the previous section. You will also
see that the Python API also provides convenience functions, such as the
[`export_to_file`{.docutils .literal .notranslate}]{.pre} and
[`validate_lems`{.docutils .literal .notranslate}]{.pre} functions to
quickly save your model to an XML file, and validate it.

::: {.highlight-xml .notranslate}
::: highlight
    <?xml version="1.0" ?>
    <Lems xmlns="http://www.neuroml.org/lems/0.7.6" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.neuroml.org/lems/0.7.6 https://raw.githubusercontent.com/LEMS/LEMS/development/Schemas/LEMS/LEMS_v0.7.6.xsd">
      <Dimension name="time" t="1"/>
      <Unit symbol="s" dimension="time" power="1" scale="1.0"/>
      <Unit symbol="ms" dimension="time" power="-3" scale="1.0"/>
      <ComponentType name="lorenz1963" description="The Lorenz system is a simplified model for atomspheric convection, derived from the Navier Stokes equations">
        <Parameter name="sigma" dimension="none" description="Prandtl Number"/>
        <Parameter name="beta" dimension="none" description="Also named b elsewhere"/>
        <Parameter name="rho" dimension="none" description="Related to the Rayleigh number, also named r elsewhere"/>
        <Parameter name="x0" dimension="none"/>
        <Parameter name="y0" dimension="none"/>
        <Parameter name="z0" dimension="none"/>
        <Constant name="sec" value="1s" dimension="time"/>
        <Exposure name="x" dimension="none"/>
        <Exposure name="y" dimension="none"/>
        <Exposure name="z" dimension="none"/>
        <Dynamics>
          <StateVariable name="x" dimension="none" exposure="x"/>
          <StateVariable name="y" dimension="none" exposure="y"/>
          <StateVariable name="z" dimension="none" exposure="z"/>
          <TimeDerivative variable="x" value="( sigma * (y - x)) / sec"/>
          <TimeDerivative variable="y" value="( rho * x - y - x * z ) / sec"/>
          <TimeDerivative variable="z" value="( x * y - beta * z) / sec"/>
          <OnStart>
            <StateAssignment variable="x" value="x0"/>
            <StateAssignment variable="y" value="y0"/>
            <StateAssignment variable="z" value="z0"/>
          </OnStart>
        </Dynamics>
      </ComponentType>
      <Component id="lorenzCell" type="lorenz1963" sigma="10" beta="2.67" rho="28" x0="1.0" y0="1.0" z0="1.0"/>
    </Lems>
:::
:::

We strongly suggest that users use the Python tools when working with
both NeuroML and LEMS. Not only is Python easier to read and write than
XML, it also provides powerful programming constructs and has a rich
ecosystem of scientific software.
:::
:::

::: {#examples .section}
[]{#userdocs-extending-examples}

### Examples[\#](#examples "Link to this heading"){.headerlink}

Here are some examples of Components written using LEMS to extend
NeuroML that can be used as references.

-   [The Lorenz example XML source
    code](https://github.com/NeuroML/NeuroMLlite/blob/master/examples/test_files/Lorenz1963.xml){.reference
    .external}

-   [An example script for building a LEMS model using
    Python](https://github.com/LEMS/pylems/blob/master/examples/apitest2.py){.reference
    .external}

-   [Defining a new synapse in
    LEMS](https://github.com/OpenSourceBrain/BonoClopath2017/blob/master/NeuroML2/AMPA_NMDA.synapse.nml){.reference
    .external}

-   [Defining an ion channel in
    LEMS](https://github.com/OpenSourceBrain/SmithEtAl2013-L23DendriticSpikes/blob/master/NeuroML2/na.channel.nml){.reference
    .external}

-   [Defining a new Calcium pool in
    LEMS](https://github.com/OpenSourceBrain/PospischilEtAl2008/blob/master/NeuroML2/channels/Ca/Ca.nml){.reference
    .external}
:::
:::

[]{#document-Userdocs/Software/Software}

::: {#software-and-tools .section .tex2jax_ignore .mathjax_ignore}
[]{#userdocs-software}

## Software and Tools[\#](#software-and-tools "Link to this heading"){.headerlink}

::: {#core-neuroml-tools .section}
### Core NeuroML Tools[\#](#core-neuroml-tools "Link to this heading"){.headerlink}

The NeuroML initiative supports **a core set of libraries** (mainly in
Python and Java) to enable the creation/validation/analysis/simulation
of NeuroML models as well as to facilitate adding support for the
language to other applications.

<figure id="id1" class="align-center">
<a href="_images/pynml_jnml.svg"
class="reference internal image-reference"><img
src="_images/pynml_jnml.svg" style="width: 500px;"
alt="jNeuroML and pyNeuroML" /></a>
<figcaption><p><span class="caption-number">Fig. 62 </span><span
class="caption-text">Relationship between <a href="Landing.html#jlems"
class="reference internal"><span class="std std-ref">jLEMS</span></a>,
<a href="Landing.html#jneuroml" class="reference internal"><span
class="std std-ref">jNeuroML</span></a>, the <a
href="Landing.html#userdocs-neuromlv2" class="reference internal"><span
class="std std-ref">NeuroML 2 LEMS definitions</span></a>, <a
href="Landing.html#libneuroml" class="reference internal"><span
class="std std-ref">libNeuroML</span></a>, <a href="Landing.html#pylems"
class="reference internal"><span class="std std-ref">pyLEMS</span></a>
and <a href="Landing.html#pyneuroml" class="reference internal"><span
class="std std-ref">pyNeuroML</span></a>.</span><a href="#id1"
class="headerlink" title="Link to this image">#</a></p></figcaption>
</figure>

::: {#python-based-applications .section}
#### Python based applications[\#](#python-based-applications "Link to this heading"){.headerlink}

For most users, [[pyNeuroML]{.std
.std-ref}](Landing.html#pyneuroml){.reference .internal} will provide
all of the key functionality for building, validating, simulating,
visualising, and converting NeuroML 2 and LEMS models. It builds on
[[libNeuroML]{.std .std-ref}](Landing.html#libneuroml){.reference
.internal} and [[pyLEMS]{.std .std-ref}](Landing.html#pylems){.reference
.internal} and bundles all of the functionality of [[jNeuroML]{.std
.std-ref}](Landing.html#jneuroml){.reference .internal} to provide
access to this through a Python interface.
:::

::: {#java-based-applications .section}
#### Java based applications[\#](#java-based-applications "Link to this heading"){.headerlink}

[[jNeuroML]{.std .std-ref}](Landing.html#jneuroml){.reference .internal}
(for validating, simulating and converting NeuroML 2 models) and
[[jLEMS]{.std .std-ref}](Landing.html#jlems){.reference .internal} (for
simulating LEMS models) are the key applications created in Java for
supporting NeuroML 2/LEMS.
:::

::: {#neuroml-support-in-other-languages .section}
#### NeuroML support in other languages[\#](#neuroml-support-in-other-languages "Link to this heading"){.headerlink}

There are preliminary APIs for using NeuroML in [[C++]{.std
.std-ref}](Landing.html#neuromlc){.reference .internal} and
[[MATLAB]{.std .std-ref}](Landing.html#matlab){.reference .internal}.
:::
:::

::: {#other-neuroml-supporting-applications .section}
### Other NeuroML supporting applications[\#](#other-neuroml-supporting-applications "Link to this heading"){.headerlink}

Many other simulators, applications and libraries support NeuroML. See
[[here]{.std .std-ref}](Landing.html#userdocs-supporting){.reference
.internal} for more details.

A number of databases and neuroinformatics initiatives support NeuroML
as a core interchange format. See [[here]{.std
.std-ref}](Landing.html#userdocs-finding-models){.reference .internal}
for more details.
:::

::: {.toctree-wrapper .compound}
[]{#document-Userdocs/Software/pyNeuroML}

::: {#pyneuroml .section .tex2jax_ignore .mathjax_ignore}
[]{#id1}

### pyNeuroML[\#](#pyneuroml "Link to this heading"){.headerlink}

::: {.dropdown .admonition}
Suggested NeuroML tool

pyNeuroML is the suggested software tool for working with NeuroML. It
builds on [[jNeuroML]{.std .std-ref}](Landing.html#jneuroml){.reference
.internal}, [[libNeuroML]{.std
.std-ref}](Landing.html#libneuroml){.reference .internal}, and
[[pyLEMS]{.std .std-ref}](Landing.html#pylems){.reference .internal}.
:::

::: {.dropdown .admonition}
Citation

Please cite Vella et al.
([\[[VCC+14](Landing.html#id11 "Michael Vella, Robert C. Cannon, Sharon Crook, Andrew P. Davison, Gautham Ganapathy, Hugh P. C. Robinson, R. Angus Silver, and Padraig Gleeson. Libneuroml and pylems: using python to combine procedural and declarative modeling approaches in computational neuroscience. Frontiers in neuroinformatics, 8:38, 2014. doi:10.3389/fninf.2014.00038."){.reference
.internal}\]]{#id2}) if you use pyNeuroML.
:::

pyNeuroML is a Python package that allows you to work with NeuroML
models using the Python programming language. It includes all the API
functions provided by [[libNeuroML]{.std
.std-ref}](Landing.html#libneuroml){.reference .internal} and
[[pyLEMS]{.std .std-ref}](Landing.html#pylems){.reference .internal},
and also wraps all the functions that [[jNeuroML]{.std
.std-ref}](Landing.html#jneuroml){.reference .internal} provides, which
can therefore be used from within Python itself.

With pyNeuroML you can:

-   **Create** NeuroML models and simulations

-   **Validate** NeuroML v1.8.1 and v2.x files

-   **Simulate** NeuroML 2 models

-   **Export** NeuroML 2 and LEMS files to many formats such as Neuron,
    Brian, Matlab, etc.

-   **Import** other languages into LEMS (e.g. SBML)

-   **Visualise** NeuroML models and simulations

<figure id="id3" class="align-center">
<a href="_images/pynml_jnml.svg"
class="reference internal image-reference"><img
src="_images/pynml_jnml.svg" style="width: 500px;"
alt="jNeuroML and pyNeuroML" /></a>
<figcaption><p><span class="caption-number">Fig. 63 </span><span
class="caption-text">Relationship between <a href="Landing.html#jlems"
class="reference internal"><span class="std std-ref">jLEMS</span></a>,
<a href="Landing.html#jneuroml" class="reference internal"><span
class="std std-ref">jNeuroML</span></a>, the <a
href="Landing.html#userdocs-neuromlv2" class="reference internal"><span
class="std std-ref">NeuroML 2 LEMS definitions</span></a>, <a
href="Landing.html#libneuroml" class="reference internal"><span
class="std std-ref">libNeuroML</span></a>, <a href="Landing.html#pylems"
class="reference internal"><span class="std std-ref">pyLEMS</span></a>
and <a href="#pyneuroml" class="reference internal"><span
class="std std-ref">pyNeuroML</span></a>.</span><a href="#id3"
class="headerlink" title="Link to this image">#</a></p></figcaption>
</figure>

::: {#quick-start .section}
[]{#pyneuroml-quickstart}

#### Quick start[\#](#quick-start "Link to this heading"){.headerlink}

::: {#install-python-and-the-java-runtime-environment .section}
[]{#pyneuroml-install-python-and-java}

##### Install Python and the Java Runtime Environment[\#](#install-python-and-the-java-runtime-environment "Link to this heading"){.headerlink}

[Python](https://www.python.org/){.reference .external} is generally
pre-installed on all computers nowadays. However, if you do not have
Python installed on your system, please follow the official
[installation
instructions](https://www.python.org/downloads/){.reference .external}
to install Python on your computer. A number of Free/Open source
Integrated Development Environments (IDEs) are also available that make
working with Python (even) easier. An example list is
[here](https://opensource.com/resources/python/ides){.reference
.external}.

Since pyNeuroML wraps around jNeuroML which is written in Java, you will
need a Java Runtime Environment (JRE) installed on your system. On most
Linux systems [Free/Open source OpenJDK runtime
environment](https://openjdk.java.net/){.reference .external} is already
pre-installed. You can also install Oracle's proprietary Java platform
from their [download
page](https://www.oracle.com/java/technologies/javase-downloads.html){.reference
.external} if you prefer. Please refer to your operating system's
documentation to install a JRE.
:::

::: {#install-pyneuroml-with-pip .section}
[]{#pyneuroml-install-with-pip}

##### Install pyNeuroML with pip[\#](#install-pyneuroml-with-pip "Link to this heading"){.headerlink}

::: {.tip .dropdown .admonition}
Tip: Use a virtual environment

While using Python packages, it is suggested to use a virtual
environment to isolate the software you install from each other. Learn
more about using virtual environments in Python
[here](https://docs.python.org/3/tutorial/venv.html){.reference
.external}.
:::

The easiest way to install the latest version of pyNeuroML is using the
default Python package manager, [`pip`{.docutils .literal
.notranslate}]{.pre}:

::: {.highlight-console .notranslate}
::: highlight
    pip install pyneuroml
:::
:::

By default, this will only install the minimal set of packages required
to use pyNeuroML. To use pyNeuroML with specific [[supporting
tools]{.std .std-ref}](Landing.html#userdocs-supporting-apps){.reference
.internal}, please install them as required:

-   simulation backends

    -   [`pip`{.code .docutils .literal .notranslate}]{.pre}` `{.code
        .docutils .literal .notranslate}[`install`{.code .docutils
        .literal .notranslate}]{.pre}` `{.code .docutils .literal
        .notranslate}[`pyneuroml[neuron]`{.code .docutils .literal
        .notranslate}]{.pre}: NEURON

    -   [`pip`{.code .docutils .literal .notranslate}]{.pre}` `{.code
        .docutils .literal .notranslate}[`install`{.code .docutils
        .literal .notranslate}]{.pre}` `{.code .docutils .literal
        .notranslate}[`pyneuroml[brian]`{.code .docutils .literal
        .notranslate}]{.pre}: Brian2

    -   [`pip`{.code .docutils .literal .notranslate}]{.pre}` `{.code
        .docutils .literal .notranslate}[`install`{.code .docutils
        .literal .notranslate}]{.pre}` `{.code .docutils .literal
        .notranslate}[`pyneuroml[netpyne]`{.code .docutils .literal
        .notranslate}]{.pre}: NetPyNE

    -   [`pip`{.code .docutils .literal .notranslate}]{.pre}` `{.code
        .docutils .literal .notranslate}[`install`{.code .docutils
        .literal .notranslate}]{.pre}` `{.code .docutils .literal
        .notranslate}[`pyneuroml[tellurium]`{.code .docutils .literal
        .notranslate}]{.pre}: Tellurium

-   COMBINE formats (SEDML/SBML):

    -   [`pip`{.code .docutils .literal .notranslate}]{.pre}` `{.code
        .docutils .literal .notranslate}[`install`{.code .docutils
        .literal .notranslate}]{.pre}` `{.code .docutils .literal
        .notranslate}[`pyneuroml[combine]`{.code .docutils .literal
        .notranslate}]{.pre}

-   analysis and visualization

    -   [`pip`{.code .docutils .literal .notranslate}]{.pre}` `{.code
        .docutils .literal .notranslate}[`install`{.code .docutils
        .literal .notranslate}]{.pre}` `{.code .docutils .literal
        .notranslate}[`pyneuroml[analysis]`{.code .docutils .literal
        .notranslate}]{.pre}

-   visualization

    -   VisPy:

        -   [`pip`{.code .docutils .literal
            .notranslate}]{.pre}` `{.code .docutils .literal
            .notranslate}[`install`{.code .docutils .literal
            .notranslate}]{.pre}` `{.code .docutils .literal
            .notranslate}[`pyneuroml[vispy]`{.code .docutils .literal
            .notranslate}]{.pre}: Qt6 (default):

        -   [`pip`{.code .docutils .literal
            .notranslate}]{.pre}` `{.code .docutils .literal
            .notranslate}[`install`{.code .docutils .literal
            .notranslate}]{.pre}` `{.code .docutils .literal
            .notranslate}[`pyneuroml[vispy-qt5]`{.code .docutils
            .literal .notranslate}]{.pre} : Qt5

        -   [`pip`{.code .docutils .literal
            .notranslate}]{.pre}` `{.code .docutils .literal
            .notranslate}[`install`{.code .docutils .literal
            .notranslate}]{.pre}` `{.code .docutils .literal
            .notranslate}[`pyneuroml[vispy-common]`{.code .docutils
            .literal .notranslate}]{.pre}: to manually use another
            [supported
            backend](https://vispy.org/installation){.reference
            .external}:

    -   [`pip`{.code .docutils .literal .notranslate}]{.pre}` `{.code
        .docutils .literal .notranslate}[`install`{.code .docutils
        .literal .notranslate}]{.pre}` `{.code .docutils .literal
        .notranslate}[`pyneuroml[povray]`{.code .docutils .literal
        .notranslate}]{.pre}: Povray

    -   [`pip`{.code .docutils .literal .notranslate}]{.pre}` `{.code
        .docutils .literal .notranslate}[`install`{.code .docutils
        .literal .notranslate}]{.pre}` `{.code .docutils .literal
        .notranslate}[`pyneuroml[plotly]`{.code .docutils .literal
        .notranslate}]{.pre}: PlotLy

-   model fitting

    -   [`pip`{.code .docutils .literal .notranslate}]{.pre}` `{.code
        .docutils .literal .notranslate}[`install`{.code .docutils
        .literal .notranslate}]{.pre}` `{.code .docutils .literal
        .notranslate}[`pyneuroml[tune]`{.code .docutils .literal
        .notranslate}]{.pre}

-   HDF5 support

    -   [`pip`{.code .docutils .literal .notranslate}]{.pre}` `{.code
        .docutils .literal .notranslate}[`install`{.code .docutils
        .literal .notranslate}]{.pre}` `{.code .docutils .literal
        .notranslate}[`pyneuroml[hdf5]`{.code .docutils .literal
        .notranslate}]{.pre}

-   Neuroscience Gateway (NSG) support

    -   [`pip`{.code .docutils .literal .notranslate}]{.pre}` `{.code
        .docutils .literal .notranslate}[`install`{.code .docutils
        .literal .notranslate}]{.pre}` `{.code .docutils .literal
        .notranslate}[`pyneuroml[nsg]`{.code .docutils .literal
        .notranslate}]{.pre}

-   RDF annotations

    -   [`pip`{.code .docutils .literal .notranslate}]{.pre}` `{.code
        .docutils .literal .notranslate}[`install`{.code .docutils
        .literal .notranslate}]{.pre}` `{.code .docutils .literal
        .notranslate}[`pyneuroml[annotations]`{.code .docutils .literal
        .notranslate}]{.pre}

A number of "meta" packages are also available:

-   [`pip`{.code .docutils .literal .notranslate}]{.pre}` `{.code
    .docutils .literal .notranslate}[`install`{.code .docutils .literal
    .notranslate}]{.pre}` `{.code .docutils .literal
    .notranslate}[`pyneuroml[all]`{.code .docutils .literal
    .notranslate}]{.pre}: install everything

-   [`pip`{.code .docutils .literal .notranslate}]{.pre}` `{.code
    .docutils .literal .notranslate}[`install`{.code .docutils .literal
    .notranslate}]{.pre}` `{.code .docutils .literal
    .notranslate}[`pyneuroml[dev]`{.code .docutils .literal
    .notranslate}]{.pre}: install for development

-   [`pip`{.code .docutils .literal .notranslate}]{.pre}` `{.code
    .docutils .literal .notranslate}[`install`{.code .docutils .literal
    .notranslate}]{.pre}` `{.code .docutils .literal
    .notranslate}[`pyneuroml[doc]`{.code .docutils .literal
    .notranslate}]{.pre}: install for building documentation
:::

::: {#installing-optional-dependencies .section}
[]{#pyneuroml-optional}

##### Installing optional dependencies[\#](#installing-optional-dependencies "Link to this heading"){.headerlink}

The optional "extras" provided by pyNeuroML may require some additional
software to be installed that is not Python based, and so cannot be
automatically installed using [`pip`{.code .docutils .literal
.notranslate}]{.pre}:

::: {#neuron .section}
###### NEURON[\#](#neuron "Link to this heading"){.headerlink}

For compiling NEURON mod files, you also need a C compiler and the
[`make`{.docutils .literal .notranslate}]{.pre} utility installed on
your computer. Additionally, to run parallel simulations the
[MPI](https://en.wikipedia.org/wiki/Message_Passing_Interface){.reference
.external} libraries are also needed. Please see the [NEURON
installation
documentation](https://www.neuron.yale.edu/neuron/download){.reference
.external} for more information on installing NEURON on your computer.
:::

::: {#povray .section}
###### Povray[\#](#povray "Link to this heading"){.headerlink}

Requires the installation of the
[Povray](http://www.povray.org/){.reference .external} tool.
:::

::: {#vispy .section}
###### Vispy[\#](#vispy "Link to this heading"){.headerlink}

Requires installation of a [back
end](https://vispy.org/installation){.reference .external}. The default
is [Qt6](https://www.qt.io/){.reference .external}, but one can also use
Qt5, or a different back end. Vispy also makes use of GPUs via
[OpenGL](https://en.wikipedia.org/wiki/OpenGL){.reference .external}. So
a recent GPU is recommended for larger scale models.

For more information on individual simulation backends and extras,
please refer to their respective documentations.
:::
:::

::: {#installation-on-fedora-linux .section}
[]{#pyneuroml-install-fedora}

##### Installation on Fedora Linux[\#](#installation-on-fedora-linux "Link to this heading"){.headerlink}

PyNeuroML is tested for use on Fedora Linux by the [NeuroFedora
team](https://neuro.fedoraproject.org){.reference .external}. You can
install it on Fedora Linux directly using pip.

Please see the [project
documentation](https://docs.fedoraproject.org/en-US/neurofedora/){.reference
.external} for more information on Neuroscience software packages.
:::
:::

::: {#documentation .section}
[]{#pyneuroml-docs}

#### Documentation[\#](#documentation "Link to this heading"){.headerlink}

pyNeuroML provides a set of command line utilities along with an API to
use from within Python scripts:

::: {.dropdown .admonition}
TODO!

Check that all of these have usage documentation that is viewable using
the [`-h`{.docutils .literal .notranslate}]{.pre} flag. Issue filed:
[NeuroML/pyNeuroML#87](https://github.com/NeuroML/pyNeuroML/issues/87){.github
.reference .external}
:::

-   pynml

-   pynml-channelanalysis

-   pynml-modchananalysis

-   pynml-plotspikes

-   pynml-povray

-   pynml-sonata

-   pynml-summary

-   pynml-tune

These utilities are self-documented. So, to learn how these utilities
are to be used, run them with the [`-h`{.docutils .literal
.notranslate}]{.pre} flag. For example:

::: {.highlight-console .notranslate}
::: highlight
    pynml -h
    usage: pynml [-h|--help] [<shared options>] <one of the mutually-exclusive options>

    pyNeuroML v0.5.9: Python utilities for NeuroML2
        libNeuroML v0.2.54
        jNeuroML v0.10.2

    optional arguments:
      -h, --help            show this help message and exit

    Shared options:
      These options can be added to any of the mutually-exclusive options

      -verbose              Verbose output
      -java_max_memory MAX  Java memory for jNeuroML, e.g. 400M, 2G (used in
                            -Xmx argument to java)
      -nogui                Suppress GUI,
                            i.e. show no plots, just save results
      <LEMS/NeuroML 2 file>
                            LEMS/NeuroML 2 file to process

    ...
:::
:::

::: {#api-documentation .section}
[]{#pyneuroml-api-docs}

##### API documentation[\#](#api-documentation "Link to this heading"){.headerlink}

Detailed API documentation for pyNeuroML can be found
[here](https://pyneuroml.readthedocs.io/en/development/){.reference
.external}.

The pyNeuroML API is also self documented. You can use Python's in-built
documentation viewer [`pydoc`{.docutils .literal .notranslate}]{.pre} to
view the documentation for any of the package's modules and their
functions:

::: {.highlight-console .notranslate}
::: highlight
    pydoc pyneuroml
    Help on package pyneuroml:

    NAME
        pyneuroml

    PACKAGE CONTENTS
        analysis (package)
        lems (package)
        neuron (package)
        plot (package)
        povray (package)
        pynml
        swc (package)
        tune (package)

    DATA
        JNEUROML_VERSION = '0.10.2'

    VERSION
        0.5.9

    FILE
        /usr/lib/python3.9/site-packages/pyneuroml/__init__.py
:::
:::

::: {.highlight-console .notranslate}
::: highlight
    pydoc pyneuroml.analysis

    Help on package pyneuroml.analysis in pyneuroml:

    NAME
        pyneuroml.analysis

    PACKAGE CONTENTS
        ChannelDensityPlot
        ChannelHelper
        NML2ChannelAnalysis

    FUNCTIONS
        analyse_spiketime_vs_dt(nml2_file, target, duration, simulator, cell_v_path, dts, verbose=False, spike_threshold_mV=0, show_plot_already=True, save_figure_to=None, num_of_last_spikes=None)

        generate_current_vs_frequency_curve(nml2_file, cell_id, start_amp_nA=-0.1, end_amp_nA=0.1, step_nA=0.01, custom_amps_nA=[], analysis_duration=1000, analysis_delay=0, pre_zero_pulse=0, post_zero_pulse=0, dt=0.05, temperature='32degC', spike_threshold_mV=0.0, plot_voltage_traces=False, plot_if=True, plot_iv=False, xlim_if=None, ylim_if=None, xlim_iv=None, ylim_iv=None, label_xaxis=True, label_yaxis=True, show_volts_label=True, grid=True, font_size=12, if_iv_color='k', linewidth=1, bottom_left_spines_only=False, show_plot_already=True, save_voltage_traces_to=None, save_if_figure_to=None, save_iv_figure_to=None, save_if_data_to=None, save_iv_data_to=None, simulator='jNeuroML', num_processors=1, include_included=True, title_above_plot=False, return_axes=False, verbose=False)

    FILE
        /usr/lib/python3.9/site-packages/pyneuroml/analysis/__init__.py
:::
:::

Most IDEs are able to show you this information as you use them in your
Python scripts.
:::
:::

::: {#getting-help .section}
[]{#pyneuroml-gethelp}

#### Getting help[\#](#getting-help "Link to this heading"){.headerlink}

For any questions regarding pyNeuroML, please open an issue on the
GitHub issue tracker
[here](https://github.com/NeuroML/pyNeuroML/issues){.reference
.external}. Any bugs and feature requests can also be filed there.

You can also use any of the [[communication channels of the NeuroML
community]{.std .std-ref}](Landing.html#contact){.reference .internal}.
:::

::: {#development .section}
[]{#pyneuroml-development}

#### Development[\#](#development "Link to this heading"){.headerlink}

pyNeuroML is developed on GitHub at
[NeuroML/pyNeuroML](https://github.com/NeuroML/pyNeuroML){.github
.reference .external} under the [LPGL-3.0
license](https://github.com/NeuroML/pyNeuroML/blob/master/LICENSE.lesser){.reference
.external}. The repository contains the complete source code along with
instructions on building/installing pyNeuroML. Please follow the
instructions there to build pyNeuroML from source.
:::
:::

[]{#document-Userdocs/Software/libNeuroML}

::: {#libneuroml .section .tex2jax_ignore .mathjax_ignore}
[]{#id1}

### libNeuroML[\#](#libneuroml "Link to this heading"){.headerlink}

libNeuroML is a Python package for working with models specified in
NeuroML version 2. It provides a native Python object model
corresponding to the NeuroML schema. This allows users to build their
NeuroML models natively in Python without having to work directly with
the underlying XML representation. Additionally, libNeuroML includes
functions for the conversion of the Python representation of the NeuroML
model to and from the XML representation.

::: {.dropdown .admonition}
Use pyNeuroML

[[pyNeuroML]{.std .std-ref}](Landing.html#pyneuroml){.reference
.internal} builds on libNeuroML and includes additional utility
functions.
:::

::: {.dropdown .admonition}
Citation

Please cite Vella et al.
([\[[VCC+14](Landing.html#id11 "Michael Vella, Robert C. Cannon, Sharon Crook, Andrew P. Davison, Gautham Ganapathy, Hugh P. C. Robinson, R. Angus Silver, and Padraig Gleeson. Libneuroml and pylems: using python to combine procedural and declarative modeling approaches in computational neuroscience. Frontiers in neuroinformatics, 8:38, 2014. doi:10.3389/fninf.2014.00038."){.reference
.internal}\]]{#id2}) if you use libNeuroML.
:::

::: {#quick-start .section}
[]{#libneuroml-quickstart}

#### Quick start[\#](#quick-start "Link to this heading"){.headerlink}

::: {#install-python .section}
[]{#libneuroml-install-python}

##### Install Python[\#](#install-python "Link to this heading"){.headerlink}

[Python](https://www.python.org/){.reference .external} is generally
pre-installed on all computers nowadays. However, if you do not have
Python installed on your system, please follow the official
[installation
instructions](https://www.python.org/downloads/){.reference .external}
to install Python on your computer. A number of Free/Open source
Integrated Development Environments (IDEs) are also available that make
working with Python (even) easier. An example list is
[here](https://opensource.com/resources/python/ides){.reference
.external}.
:::

::: {#install-libneuroml-with-pip .section}
[]{#libneuroml-install-with-pip}

##### Install libNeuroML with pip[\#](#install-libneuroml-with-pip "Link to this heading"){.headerlink}

::: {.tip .dropdown .admonition}
Tip: Use a virtual environment

While using Python packages, it is suggested to use a virtual
environment to isolate the software you install from each other. Learn
more about using virtual environments in Python
[here](https://docs.python.org/3/tutorial/venv.html){.reference
.external}.
:::

The easiest way to install the latest version of libNeuroML is using the
default Python package manager, [`pip`{.docutils .literal
.notranslate}]{.pre}:

::: {.highlight-console .notranslate}
::: highlight
    pip install libNeuroML
:::
:::
:::

::: {#installation-on-fedora-linux .section}
[]{#libneuroml-install-fedora}

##### Installation on Fedora Linux[\#](#installation-on-fedora-linux "Link to this heading"){.headerlink}

On [Fedora](https://getfedora.org){.reference .external} Linux systems,
the [NeuroFedora](https://neuro.fedoraproject.org){.reference .external}
community provides libNeuroML in the [standard Fedora
repos](https://src.fedoraproject.org/rpms/python-libNeuroML){.reference
.external} and can be installed using the following commands:

::: {.highlight-console .notranslate}
::: highlight
    sudo dnf install python3-libNeuroML
:::
:::
:::
:::

::: {#documentation .section}
[]{#libneuroml-docs}

#### Documentation[\#](#documentation "Link to this heading"){.headerlink}

Detailed API documentation for libNeuroML can be found
[here](https://libneuroml.readthedocs.io/en/latest/){.reference
.external}. For more information on libNeuroML, please see Vella et al.
([\[[VCC+14](Landing.html#id11 "Michael Vella, Robert C. Cannon, Sharon Crook, Andrew P. Davison, Gautham Ganapathy, Hugh P. C. Robinson, R. Angus Silver, and Padraig Gleeson. Libneuroml and pylems: using python to combine procedural and declarative modeling approaches in computational neuroscience. Frontiers in neuroinformatics, 8:38, 2014. doi:10.3389/fninf.2014.00038."){.reference
.internal}\]]{#id3}) and Cannon et al.
([\[[CGC+14](Landing.html#id10 "Robert C. Cannon, Padraig Gleeson, Sharon Crook, Gautham Ganapathy, Boris Marin, Eugenio Piasini, and R. Angus Silver. LEMS: a language for expressing complex biological models in concise and hierarchical form and its use in underpinning NeuroML 2. Frontiers in Neuroinformatics, 2014. doi:10.3389/fninf.2014.00079."){.reference
.internal}\]]{#id4}).

The core classes in NeuroML are Python representations of the Component
Types defined in the [[NeuroML standard]{.std
.std-ref}](Landing.html#userdocs-neuromlv2){.reference .internal}. These
can be used to build NeuroML models in Python, and these models can then
be exported to the standard XML NeuroML representation. These core
classes also contain some utility functions to make it easier for users
to carry out common tasks.

<figure id="id5" class="align-center">
<img src="_images/libneuroml.png"
alt="Examples of mapping between Component names in the NeuroML schema and their corresponding libNeuroML Python classes." />
<figcaption><p><span class="caption-number">Fig. 64 </span><span
class="caption-text">Examples of mapping between Component names in the
NeuroML schema and their corresponding libNeuroML Python
classes.</span><a href="#id5" class="headerlink"
title="Link to this image">#</a></p></figcaption>
</figure>

Each NeuroML Component Type is represented here as a Python class. Due
to implementation limitations, whereas NeuroML Component Types use
[lower camel case
naming](https://en.wikipedia.org/wiki/Camel_case){.reference .external},
the Python classes here use [upper camel case
naming](https://en.wikipedia.org/wiki/Camel_case){.reference .external}.
So, for example, the [`adExIaFCell`{.docutils .literal
.notranslate}]{.pre} Component Type in the NeuroML schema becomes the
[`AdExIaFCell`{.docutils .literal .notranslate}]{.pre} class here, and
[`expTwoSynapse`{.docutils .literal .notranslate}]{.pre} becomes the
[`ExpTwoSynapse`{.docutils .literal .notranslate}]{.pre} class.

The [`child`{.docutils .literal .notranslate}]{.pre} and
[`children`{.docutils .literal .notranslate}]{.pre} elements that
NeuroML Component Types can have are represented in the Python classes
as variables. The variable names, to distinguish them from class names,
use [snake case](https://en.wikipedia.org/wiki/Snake_case){.reference
.external}. So for example, the [`cell`{.docutils .literal
.notranslate}]{.pre} NeuroML Component Type has a corresponding
[`Cell`{.docutils .literal .notranslate}]{.pre} Python class here. The
[`biophysicalProperties`{.docutils .literal .notranslate}]{.pre} child
Component Type in [`cell`{.docutils .literal .notranslate}]{.pre} is
represented as the [`biophysical_properties`{.docutils .literal
.notranslate}]{.pre} list variable in the [`Cell`{.docutils .literal
.notranslate}]{.pre} Python class. The class signatures list all the
child/children elements and text fields that the corresponding Component
Type possesses. To again use the [`Cell`{.docutils .literal
.notranslate}]{.pre} class as an example, the construction signature is
this:

::: {.highlight-python .notranslate}
::: highlight
    class neuroml.nml.nml.Cell(neuro_lex_id=None, id=None, metaid=None, notes=None, properties=None, annotation=None, morphology_attr=None, biophysical_properties_attr=None, morphology=None, biophysical_properties=None, extensiontype_=None, **kwargs_)
:::
:::

As can be seen here, it includes both the
[`biophysical_properties`{.docutils .literal .notranslate}]{.pre} and
[`morphology`{.docutils .literal .notranslate}]{.pre} child elements as
variables.

Please see the examples in the [[NeuroML documentation]{.std
.std-ref}](Landing.html#userdocs-getting-started-neuroml){.reference
.internal} to see usage examples of libNeuroML. Please also note that
this module is also included in the top level of the
[`neuroml`{.docutils .literal .notranslate}]{.pre} package, so you can
use these classes by importing neuroml:

::: {.highlight-python .notranslate}
::: highlight
    from neuroml import AdExIaFCell
:::
:::
:::

::: {#getting-help .section}
[]{#libneuroml-gethelp}

#### Getting help[\#](#getting-help "Link to this heading"){.headerlink}

For any questions regarding libNeuroML, please open an issue on the
GitHub issue tracker
[here](https://github.com/NeuralEnsemble/libNeuroML/issues){.reference
.external}. Any bugs and feature requests can also be filed there.

You can also use any of the [[communication channels of the NeuroML
community]{.std .std-ref}](Landing.html#contact){.reference .internal}.
:::

::: {#development .section}
[]{#libneuroml-development}

#### Development[\#](#development "Link to this heading"){.headerlink}

libNeuroML is developed on GitHub at
[NeuralEnsemble/libNeuroML](https://github.com/NeuralEnsemble/libNeuroML){.github
.reference .external} under the [BSD 3 clause
license](https://github.com/NeuralEnsemble/libNeuroML/blob/master/LICENSE){.reference
.external}. The repository contains the complete source code along with
instructions on building/installing libNeuroML. Please follow the
instructions there to build libNeuroML from source.
:::
:::

[]{#document-Userdocs/Software/pyLEMS}

::: {#pylems .section .tex2jax_ignore .mathjax_ignore}
[]{#id1}

### pyLEMS[\#](#pylems "Link to this heading"){.headerlink}

pyLEMS is a Python package which provides an API, as well as a simulator
for the [LEMS](http://lems.github.io/LEMS){.reference .external}
language. It can also be used to run NeuroML2 models.

::: {.dropdown .admonition}
Use pyNeuroML

[[pyNeuroML]{.std .std-ref}](Landing.html#pyneuroml){.reference
.internal} builds on pyLEMS and includes additional functions.
:::

::: {.dropdown .admonition}
Citation

Please cite Vella et al.
([\[[VCC+14](Landing.html#id11 "Michael Vella, Robert C. Cannon, Sharon Crook, Andrew P. Davison, Gautham Ganapathy, Hugh P. C. Robinson, R. Angus Silver, and Padraig Gleeson. Libneuroml and pylems: using python to combine procedural and declarative modeling approaches in computational neuroscience. Frontiers in neuroinformatics, 8:38, 2014. doi:10.3389/fninf.2014.00038."){.reference
.internal}\]]{#id2}) if you use pyLEMS.
:::

::: {#quick-start .section}
[]{#pylems-quickstart}

#### Quick start[\#](#quick-start "Link to this heading"){.headerlink}

::: {#install-python .section}
[]{#pylems-install-python}

##### Install Python[\#](#install-python "Link to this heading"){.headerlink}

[Python](https://www.python.org/){.reference .external} is generally
pre-installed on all computers nowadays. However, if you do not have
Python installed on your system, please follow the official
[installation
instructions](https://www.python.org/downloads/){.reference .external}
to install Python on your computer. A number of Free/Open source
Integrated Development Environments (IDEs) are also available that make
working with Python (even) easier. An example list is
[here](https://opensource.com/resources/python/ides){.reference
.external}.
:::

::: {#install-pylems-with-pip .section}
[]{#pylems-install-with-pip}

##### Install pyLEMS with pip[\#](#install-pylems-with-pip "Link to this heading"){.headerlink}

::: {.tip .dropdown .admonition}
Tip: Use a virtual environment

While using Python packages, it is suggested to use a virtual
environment to isolate the software you install from each other. Learn
more about using virtual environments in Python
[here](https://docs.python.org/3/tutorial/venv.html){.reference
.external}.
:::

The easiest way to install the latest version of pyLEMS is using the
default Python package manager, [`pip`{.docutils .literal
.notranslate}]{.pre}:

::: {.highlight-console .notranslate}
::: highlight
    pip install pyLEMS
:::
:::
:::

::: {#installation-on-fedora-linux .section}
[]{#pylems-install-fedora}

##### Installation on Fedora Linux[\#](#installation-on-fedora-linux "Link to this heading"){.headerlink}

On [Fedora](https://getfedora.org){.reference .external} Linux systems,
the [NeuroFedora](https://neuro.fedoraproject.org){.reference .external}
community provides pyLEMS in the [standard Fedora
repos](https://src.fedoraproject.org/rpms/python-pyLEMS){.reference
.external} and can be installed using the following commands:

::: {.highlight-console .notranslate}
::: highlight
    sudo dnf install python3-pyLEMS
:::
:::
:::
:::

::: {#documentation .section}
[]{#pylems-docs}

#### Documentation[\#](#documentation "Link to this heading"){.headerlink}

Detailed API documentation for PyLEMS can be found
[here](https://pylems.readthedocs.io/en/development/){.reference
.external}. pyLEMS provides the [`pylems`{.docutils .literal
.notranslate}]{.pre} command line utility that can be used to simulate
LEMS files. [`pylems`{.docutils .literal .notranslate}]{.pre} is self
documented, and you can learn about its usage using the [`-h`{.docutils
.literal .notranslate}]{.pre} flag:

::: {.highlight-console .notranslate}
::: highlight
    pylems -h
    usage: pylems [-h] [-I <Include directory>] [-nogui] [-dlems] <LEMS file>

    positional arguments:
      <LEMS file>           LEMS file to be simulated

    optional arguments:
      -h, --help            show this help message and exit
      -I <Include directory>
                            Directory to be searched for included files
      -nogui                If this is specified, just parse & simulate the model, but don't show any plots
      -dlems                If this is specified, export the LEMS file as dLEMS (distilled LEMS in JSON format, see https://github.com/borismarin/som-codegen)
:::
:::

To simulate a LEMS file:

::: {.highlight-console .notranslate}
::: highlight
    pylems lemsexample.xml
:::
:::

Please note that if you are simulating a NeuroML file you will have to
also specify the location of the [[NeuroML 2 LEMS definitions]{.std
.std-ref}](Landing.html#schema-neuromlcorecomptypes){.reference
.internal} with the [`-I`{.docutils .literal .notranslate}]{.pre}
option. We suggest that you use [[pyNeuroML]{.std
.std-ref}](Landing.html#pyneuroml){.reference .internal} where this is
not required:

::: {.highlight-console .notranslate}
::: highlight
    pylems -I <dir of NeuroML2 install>/NeuroML2CoreTypes/  LEMS_NeuroML2_Model.xml
:::
:::

For more information on pyLEMS, please see Vella et al.
([\[[VCC+14](Landing.html#id11 "Michael Vella, Robert C. Cannon, Sharon Crook, Andrew P. Davison, Gautham Ganapathy, Hugh P. C. Robinson, R. Angus Silver, and Padraig Gleeson. Libneuroml and pylems: using python to combine procedural and declarative modeling approaches in computational neuroscience. Frontiers in neuroinformatics, 8:38, 2014. doi:10.3389/fninf.2014.00038."){.reference
.internal}\]]{#id3}) and Cannon et al.
([\[[CGC+14](Landing.html#id10 "Robert C. Cannon, Padraig Gleeson, Sharon Crook, Gautham Ganapathy, Boris Marin, Eugenio Piasini, and R. Angus Silver. LEMS: a language for expressing complex biological models in concise and hierarchical form and its use in underpinning NeuroML 2. Frontiers in Neuroinformatics, 2014. doi:10.3389/fninf.2014.00079."){.reference
.internal}\]]{#id4}).

::: {#api-documentation .section}
[]{#pylems-api-docs}

##### API documentation[\#](#api-documentation "Link to this heading"){.headerlink}

Detailed API documentation for pyNeuroML can be found
[here](https://pylems.readthedocs.io/en/development/index.html){.reference
.external}.

The pyLEMS API is also self documented. You can use Python's in-built
documentation viewer [`pydoc`{.docutils .literal .notranslate}]{.pre} to
view the documentation for any of the package's modules and their
functions:

::: {.highlight-console .notranslate}
::: highlight
    Help on package lems:

    NAME
        lems

    DESCRIPTION
        @author: Gautham Ganapathy
        @organization: LEMS (http://neuroml.org/lems/, https://github.com/organizations/LEMS)
        @contact: gautham@lisphacker.org

    PACKAGE CONTENTS
        api
        base (package)
        dlems (package)
        model (package)
        parser (package)
        run
        sim (package)

    DATA
        logger = <Logger LEMS (WARNING)>

    VERSION
        0.5.2

    FILE
        /usr/lib/python3.9/site-packages/lems/__init__.py
:::
:::

Most IDEs are able to show you this information as you use them in your
Python scripts.
:::
:::

::: {#getting-help .section}
[]{#pylems-gethelp}

#### Getting help[\#](#getting-help "Link to this heading"){.headerlink}

For any questions regarding pyLEMS, please open an issue on the GitHub
issue tracker [here](https://github.com/LEMS/pylems/issues){.reference
.external}. Any bugs and feature requests can also be filed there.

You can also use any of the [[communication channels of the NeuroML
community]{.std .std-ref}](Landing.html#contact){.reference .internal}.
:::

::: {#development .section}
[]{#pylems-development}

#### Development[\#](#development "Link to this heading"){.headerlink}

pyLEMS is developed on GitHub at
[LEMS/pylems](https://github.com/LEMS/pylems){.github .reference
.external} under the [LGPL-3.0
license](https://github.com/LEMS/pylems/blob/master/LICENSE.lesser){.reference
.external}. The repository contains the complete source code along with
instructions on building/installing pyLEMS. Please follow the
instructions there to build pyLEMS from source.
:::
:::

[]{#document-Userdocs/Software/NeuroMLlite}

::: {#neuromllite .section .tex2jax_ignore .mathjax_ignore}
[]{#id1}

### NeuroMLlite[\#](#neuromllite "Link to this heading"){.headerlink}

NeuroMLlite is a common framework for reading/writing/generating network
specifications which builds on NeuroML 2. It is intended to provide a
high level specification which can be used to generate networks in
NeuroML and many other formats---including graphical and in neuronal
simulator formats.

::: {.note .dropdown .admonition}
Note: NeuroMLlite is under active development

Please [watch the GitHub
repository](https://github.com/NeuroML/NeuroMLlite){.reference
.external} to receive regular updates on its progress.
:::

::: {#quick-start .section}
[]{#neuromllite-quickstart}

#### Quick start[\#](#quick-start "Link to this heading"){.headerlink}

::: {#install-python .section}
[]{#neuromllite-install-python}

##### Install Python[\#](#install-python "Link to this heading"){.headerlink}

[Python](https://www.python.org/){.reference .external} is generally
pre-installed on all computers nowadays. However, if you do not have
Python installed on your system, please follow the official
[installation
instructions](https://www.python.org/downloads/){.reference .external}
to install Python on your computer. A number of Free/Open source
Integrated Development Environments (IDEs) are also available that make
working with Python (even) easier. An example list is
[here](https://opensource.com/resources/python/ides){.reference
.external}.
:::

::: {#install-neuromllite-with-pip .section}
[]{#neuromllite-install-with-pip}

##### Install NeuroMLlite with pip[\#](#install-neuromllite-with-pip "Link to this heading"){.headerlink}

::: {.tip .dropdown .admonition}
Tip: Use a virtual environment

While using Python packages, it is suggested to use a virtual
environment to isolate the software you install from each other. Learn
more about using virtual environments in Python
[here](https://docs.python.org/3/tutorial/venv.html){.reference
.external}.
:::

The easiest way to install the latest version of libNeuroML is using the
default Python package manager, [`pip`{.docutils .literal
.notranslate}]{.pre}:

::: {.highlight-console .notranslate}
::: highlight
    pip install neuromllite
:::
:::
:::

::: {#installation-on-fedora-linux .section}
[]{#neuromllite-install-fedora}

##### Installation on Fedora Linux[\#](#installation-on-fedora-linux "Link to this heading"){.headerlink}

On [Fedora](https://getfedora.org){.reference .external} Linux systems,
the [NeuroFedora](https://neuro.fedoraproject.org){.reference .external}
community provides pyNeuroML as a package in their [extras
repository](https://docs.fedoraproject.org/en-US/neurofedora/copr/){.reference
.external} and can be installed using the following commands:

::: {.highlight-console .notranslate}
::: highlight
    sudo dnf copr enable @neurofedora/neurofedora-extra
    sudo dnf install python3-neuromllite
:::
:::
:::
:::

::: {#documentation .section}
[]{#neuromllite-docs}

#### Documentation[\#](#documentation "Link to this heading"){.headerlink}

Along with a Python API, NeuroMLlite also provides a graphical user
interface [`nmllite-ui`{.docutils .literal .notranslate}]{.pre} that can
be used to create network models and export or simulate them using
different simulators supported by NeuroML.

::: {.highlight-console .notranslate}
::: highlight
    nmllite-ui

    NMLlite-UI v0.2.4: A GUI for loading NeuroMLlite files

    Usage:
        nmllite-ui Sim_xxx.json
             Load a NeuroMLlite file containing a Simulation, which refers to the Network to run
:::
:::

<figure id="id2" class="align-center">
<img src="_images/nmllite-example.png"
alt="Screenshot of NeuroMLlite UI showing an example simulation." />
<figcaption><p><span class="caption-number">Fig. 65 </span><span
class="caption-text">Screenshot of NeuroMLlite UI showing an example
simulation</span><a href="#id2" class="headerlink"
title="Link to this image">#</a></p></figcaption>
</figure>

::: {#api-documentation .section}
[]{#neuromllite-api-docs}

##### API documentation[\#](#api-documentation "Link to this heading"){.headerlink}

::: {.dropdown .admonition}
TODO!

Generate and publish API documentation for NeuroMLlite. Issue filed:
[NeuroML/NeuroMLlite#10](https://github.com/NeuroML/NeuroMLlite/issues/10){.github
.reference .external}
:::

The NeuroMLlite API is self documented. You can use Python's in-built
documentation viewer [`pydoc`{.docutils .literal .notranslate}]{.pre} to
view the documentation for any of the package's modules and their
functions:

::: {.highlight-console .notranslate}
::: highlight
    Help on package neuromllite:

    NAME
        neuromllite

    PACKAGE CONTENTS
        ArborHandler
        BBPConnectomeReader
        BaseTypes
        ConnectivityHandler
        DefaultNetworkHandler
        GraphVizHandler
        MatrixHandler
        NetworkGenerator
        NeuronHandler
        PsyNeuLinkHandler
        PsyNeuLinkReader
        PyNNHandler
        SonataHandler
        SonataReader
        gui (package)
        sweep (package)
        utils

    ...
:::
:::

Most IDEs are able to show you this information as you use them in your
Python scripts.

A number of examples showing how the NeuroMLlite Python API is to be
used are also included in the [GitHub
repository](https://github.com/NeuroML/NeuroMLlite/tree/master/examples){.reference
.external}. For instance,
[Example4.py](https://github.com/NeuroML/NeuroMLlite/blob/master/examples/Example4.py){.reference
.external} can be run in the following ways to generate different
representations of the created network model. Please see the [Readme
file](https://github.com/NeuroML/NeuroMLlite/blob/master/README.md){.reference
.external} included in the repository for more example usage.

::: {.highlight-console .notranslate}
::: highlight
    python Example4.py                  # Generate the network in JSON
    python Example4.py -nml             # Generate the network in NeuroML2
    python Example4.py -jnml            # Generate the network in NeuroML2 & run using jNeuroML
    python Example4.py -jnmlnetpyne     # Generate the network in NeuroML2 & run using NetPyNE
    python Example4.py -jnmlnrn         # Generate the network in NeuroML2 & run using NEURON
    python Example4.py -netpyne         # Generate & run the network directly in NetPyNE
    python Example4.py -pynnnest        # Generate & run the network in NEST using PyNN
    python Example4.py -pynnnrn         # Generate & run the network in NEURON using PyNN
    python Example4.py -pynnbrian       # Generate & run the network in Brian using PyNN
    ...
:::
:::
:::
:::

::: {#getting-help .section}
[]{#neuromllite-gethelp}

#### Getting help[\#](#getting-help "Link to this heading"){.headerlink}

For any questions regarding NeuroMLlite, please open an issue on the
GitHub issue tracker
[here](https://github.com/NeuroML/NeuroMLlite/issues){.reference
.external}. Any bugs and feature requests can also be filed there.

You can also use any of the [[communication channels of the NeuroML
community]{.std .std-ref}](Landing.html#contact){.reference .internal}.
:::

::: {#development .section}
[]{#neuromllite-development}

#### Development[\#](#development "Link to this heading"){.headerlink}

pyNeuroML is developed on GitHub at
[NeuroML/NeuroMLlite](https://github.com/NeuroML/NeuroMLlite){.github
.reference .external} under the [LPGL-3.0
license](https://github.com/NeuroML/NeuroMLlite/blob/master/LICENSE.lesser){.reference
.external}. The repository contains the complete source code along with
instructions on building/installing pyNeuroML. Please follow the
instructions there to build pyNeuroML from source.
:::
:::

[]{#document-Userdocs/Software/jNeuroML}

::: {#jneuroml .section .tex2jax_ignore .mathjax_ignore}
[]{#id1}

### jNeuroML[\#](#jneuroml "Link to this heading"){.headerlink}

jNeuroML is a Free/Open Source Java tool for working with LEMS and
NeuroML 2. It includes the [`jnml`{.docutils .literal
.notranslate}]{.pre} command line application, and can also be used as a
Java library.

With jNeuroML you can:

-   **Validate** NeuroML v1.8.1 and v2.x files

-   **Simulate** NeuroML 2 models

-   **Export** NeuroML 2 and LEMS files to many formats such as Neuron,
    Brian, Matlab, etc.

-   **Import** other languages into LEMS (e.g. SBML)

-   **Visualise** NeuroML models and simulations

::: {.dropdown .admonition}
Use pyNeuroML

[[pyNeuroML]{.std .std-ref}](Landing.html#pyneuroml){.reference
.internal} builds on jNeuroML and includes additional functions.
:::

::: {#quick-start .section}
[]{#jneuroml-quickstart}

#### Quick start[\#](#quick-start "Link to this heading"){.headerlink}

::: {#install-the-java-runtime-environment .section}
[]{#jneuroml-install-java}

##### Install the Java Runtime Environment[\#](#install-the-java-runtime-environment "Link to this heading"){.headerlink}

Since jNeuroML is written in Java, you will need a Java Runtime
Environment (JRE) installed on your system. On most Linux systems
[Free/Open source OpenJDK runtime
environment](https://openjdk.java.net/){.reference .external} is already
pre-installed. You can also install Oracle's proprietary Java platform
from their [download
page](https://www.oracle.com/java/technologies/javase-downloads.html){.reference
.external} if you prefer. Please refer to your operating system's
documentation to install a JRE.
:::

::: {#installation-using-pre-compiled-jar .section}
[]{#jneuroml-install-jar}

##### Installation using pre-compiled JAR[\#](#installation-using-pre-compiled-jar "Link to this heading"){.headerlink}

jNeuroML is provided as a pre-compiled ready-to-use Java JAR file that
can be used on any computer that has Java installed. Please download it
from the [GitHub release
page](https://github.com/NeuroML/jNeuroML/releases){.reference
.external} and unzip (extract) it in a preferred folder on your
computer:

::: {.highlight-console .notranslate}
::: highlight
    cd <folder where you downloaded the jNeuroML zip file>
    unzip jNeuroML.zip
:::
:::

This will extract the zip file to a new folder which will contain the
pre-compiled JAR file and runner scripts:

::: {.highlight-console .notranslate}
::: highlight
    ls jNeuroMLJar/
    jNeuroML-0.10.2-jar-with-dependencies.jar  jnml  jnml.bat  README
:::
:::

::: {.dropdown .admonition}
TODO

Add instructions on using the installer script.
[NeuroML/jNeuroML#76](https://github.com/NeuroML/jNeuroML/pull/76){.github
.reference .external}
:::
:::

::: {#installation-on-fedora-linux .section}
[]{#jneuroml-install-fedora}

##### Installation on Fedora Linux[\#](#installation-on-fedora-linux "Link to this heading"){.headerlink}

On [Fedora](https://getfedora.org){.reference .external} Linux systems,
the [NeuroFedora](https://neuro.fedoraproject.org){.reference .external}
community provides jNeuroML as a package in their [extras
repository](https://docs.fedoraproject.org/en-US/neurofedora/copr/){.reference
.external} and can be installed using the following commands:

::: {.highlight-console .notranslate}
::: highlight
    sudo dnf copr enable @neurofedora/neurofedora-extra
    sudo dnf install jneuroml
:::
:::
:::
:::

::: {#documentation .section}
[]{#jneuroml-docs}

#### Documentation[\#](#documentation "Link to this heading"){.headerlink}

Information on usage of the [`jnml`{.docutils .literal
.notranslate}]{.pre} command line application can be found with the -h
option:

::: {.highlight-console .notranslate}
::: highlight
    jnml -h

     jNeuroML v0.10.1
    Usage:

        jnml LEMSFile.xml
               Load LEMSFile.xml using jLEMS, parse it and validate it as LEMS, and execute the model it contains

        jnml LEMSFile.xml -nogui
               As above, parse and execute the model and save results, but don't show GUI

        ...
:::
:::

::: {#api-documentation .section}
[]{#jneuroml-api-docs}

##### API documentation[\#](#api-documentation "Link to this heading"){.headerlink}

The jNeuroML API is self documented. Please refer to the various
packages to learn their usage:

-   [NeuroML/jNeuroML](https://github.com/NeuroML/jNeuroML){.reference
    .external} (API Documentation
    [here](http://neuroml.github.io/jNeuroML){.reference .external})

-   [NeuroML/org.neuroml.model](https://github.com/NeuroML/org.neuroml.model){.reference
    .external} (API Documentation
    [here](http://neuroml.github.io/org.neuroml.model/index.html){.reference
    .external})

-   [NeuroML/org.neuroml.model.injectingplugin](https://github.com/NeuroML/org.neuroml.model.injectingplugin){.reference
    .external} (API Documentation
    [here](http://neuroml.github.io/org.neuroml.model.injectingplugin/index.html){.reference
    .external})

-   [NeuroML/org.neuroml.import: Import other formats into LEMS &
    combine with NeuroML
    models](https://github.com/NeuroML/org.neuroml.import){.reference
    .external} (API documentation
    [here](http://neuroml.github.io/org.neuroml.import/){.reference
    .external})

-   [NeuroML/org.neuroml.export: Export from NeuroML &
    LEMS](https://github.com/NeuroML/org.neuroml.export){.reference
    .external} (API Documentation
    [here](http://neuroml.github.io/org.neuroml.export/index.html){.reference
    .external})
:::
:::

::: {#getting-help .section}
[]{#jneuroml-gethelp}

#### Getting help[\#](#getting-help "Link to this heading"){.headerlink}

For any questions regarding jNeuroML, please open an issue on the GitHub
issue tracker
[here](https://github.com/NeuroML/jNeuroML/issues){.reference
.external}. Any bugs and feature requests can also be filed there.

You can also use any of the [[communication channels of the NeuroML
community]{.std .std-ref}](Landing.html#contact){.reference .internal}.
:::

::: {#development .section}
[]{#jneuroml-development}

#### Development[\#](#development "Link to this heading"){.headerlink}

jNeuroML is developed on GitHub at
[NeuroML/jNeuroML](https://github.com/NeuroML/jNeuroML){.github
.reference .external} under the [LPGL-3.0
license](https://github.com/NeuroML/jNeuroML/blob/master/LICENSE.lesser){.reference
.external}. The repository contains the complete source code along with
instructions on building/installing jNeuroML. Please follow the
instructions there to build jNeuroML from source.

::: {#nightly-pre-release-jar-builds .section}
[]{#jneuroml-nightlies}

##### Nightly (pre-release) jar builds:[\#](#nightly-pre-release-jar-builds "Link to this heading"){.headerlink}

::: {.admonition .warning}
Warning

Please note that these JARs are considered experimental and should only
be used for testing purposes.
:::

In case you want to use a development (un-released) version of jNeuroML,
you can download a development build following the steps below. You will
need to have the [Subversion](https://subversion.apache.org/){.reference
.external} tool installed on your system.

::: {.highlight-console .notranslate}
::: highlight
    svn checkout svn://svn.code.sf.net/p/neuroml/code/jNeuroMLJar
    cd jNeuroMLJar
:::
:::
:::
:::
:::

[]{#document-Userdocs/Software/jLEMS}

::: {#jlems .section .tex2jax_ignore .mathjax_ignore}
[]{#id1}

### jLEMS[\#](#jlems "Link to this heading"){.headerlink}

jLEMS is an interpreter for the Low Entropy Model Specification language
written in Java.

::: {.dropdown .admonition}
jLEMS is the reference implementation of LEMS

jLEMS was developed by Robert Cannon when the LEMS language was being
devised and serves at the key reference for how to implement/interpret
the language.
:::

::: {#quick-start .section}
[]{#jlems-quickstart}

#### Quick start[\#](#quick-start "Link to this heading"){.headerlink}

Since jLEMS is included in [[jNeuroML]{.std
.std-ref}](Landing.html#jneuroml){.reference .internal}, it does need
not to be installed it separately. Please follow the instructions on
installing jNeuroML provided [[here]{.std
.std-ref}](Landing.html#jneuroml-quickstart){.reference .internal}.

Please see the [[development section below]{.std
.std-ref}](#jlems-development){.reference .internal} for information on
building the jLEMS interpreter from source.
:::

::: {#documentation .section}
[]{#jlems-docs}

#### Documentation[\#](#documentation "Link to this heading"){.headerlink}

Detailed documentation on LEMS is maintained
[here](http://lems.github.io/LEMS/){.reference .external}. For more
information on LEMS, please also see Cannon et al.
([\[[CGC+14](Landing.html#id10 "Robert C. Cannon, Padraig Gleeson, Sharon Crook, Gautham Ganapathy, Boris Marin, Eugenio Piasini, and R. Angus Silver. LEMS: a language for expressing complex biological models in concise and hierarchical form and its use in underpinning NeuroML 2. Frontiers in Neuroinformatics, 2014. doi:10.3389/fninf.2014.00079."){.reference
.internal}\]]{#id2})
:::

::: {#getting-help .section}
[]{#jlems-gethelp}

#### Getting help[\#](#getting-help "Link to this heading"){.headerlink}

For any questions regarding jLEMS, please open an issue on the GitHub
issue tracker [here](https://github.com/LEMS/jLEMS/issues){.reference
.external}. Any bugs and feature requests can also be filed there.

You can also use any of the [[communication channels of the NeuroML
community]{.std .std-ref}](Landing.html#contact){.reference .internal}.
:::

::: {#development .section}
[]{#jlems-development}

#### Development[\#](#development "Link to this heading"){.headerlink}

jLEMS is developed on GitHub at
[LEMS/jLEMS](https://github.com/LEMS/jLEMS){.github .reference
.external} under the [MIT
license](https://github.com/LEMS/jLEMS/blob/master/LICENSE){.reference
.external}. The repository contains the complete source code along with
instructions on building/installing jLEMS.
:::
:::

[]{#document-Userdocs/Software/NeuroML_API}

::: {#neuroml-c-api .section .tex2jax_ignore .mathjax_ignore}
[]{#neuromlc}

### NeuroML C++ API[\#](#neuroml-c-api "Link to this heading"){.headerlink}

A C++ API for NeuroML.

::: {#quick-start .section}
[]{#neuromlc-quickstart}

#### Quick start[\#](#quick-start "Link to this heading"){.headerlink}

The C++ API is generated from the [[NeuroML specification]{.std
.std-ref}](Landing.html#userdocs-specification){.reference .internal}
using the [CodeSynthesis XSD XML Schema to C++ data binding
compiler](https://www.codesynthesis.com/products/xsd/){.reference
.external}. The C++ API needs to be compiled from source. Please refer
to the instructions in the [Readme
document](https://github.com/NeuroML/NeuroML_API/blob/master/README.md){.reference
.external} for instructions on building and installing the API.
:::

::: {#documentation .section}
[]{#neuromlc-docs}

#### Documentation[\#](#documentation "Link to this heading"){.headerlink}

For information on the generated C++ structure, please see the [XSD user
manual](http://www.codesynthesis.com/projects/xsd/documentation/cxx/tree/manual/){.reference
.external}.

::: {#api-documentation .section}
[]{#neuromlc-api-docs}

##### API documentation[\#](#api-documentation "Link to this heading"){.headerlink}

API documentation for the C++ API can be found
[here](https://neuroml.github.io/NeuroML_API/){.reference .external}. It
can also be generated while building the API from source, as documented
in the
[Readme](https://github.com/NeuroML/NeuroML_API/blob/master/README.md){.reference
.external}.
:::
:::

::: {#getting-help .section}
[]{#neuromlc-gethelp}

#### Getting help[\#](#getting-help "Link to this heading"){.headerlink}

For any questions regarding the C++ NeuroML API, please open an issue on
the GitHub issue tracker
[here](https://github.com/NeuroML/NeuroML_API/issues){.reference
.external}. Any bugs and feature requests can also be filed there.

You can also use any of the [[communication channels of the NeuroML
community]{.std .std-ref}](Landing.html#contact){.reference .internal}.
:::

::: {#development .section}
[]{#neuromlc-development}

#### Development[\#](#development "Link to this heading"){.headerlink}

The C++ NeuroML API is developed on GitHub at
[NeuroML/NeuroML_API](https://github.com/NeuroML/NeuroML_API){.github
.reference .external} under the [MIT
license](https://github.com/NeuroML/NeuroML_API/blob/master/License.txt){.reference
.external}.
:::
:::

[]{#document-Userdocs/Software/MatLab}

::: {#matlab-neuroml-toolbox .section .tex2jax_ignore .mathjax_ignore}
[]{#matlab}

### MatLab NeuroML Toolbox[\#](#matlab-neuroml-toolbox "Link to this heading"){.headerlink}

The NeuroML 2 Toolbox for MATLAB facilitates access to the Java NeuroML
2 API functionality ([[jNeuroML]{.std
.std-ref}](Landing.html#jneuroml){.reference .internal}) directly within
Matlab.

::: {#quick-start .section}
[]{#neuromlmatlab-quickstart}

#### Quick start[\#](#quick-start "Link to this heading"){.headerlink}

Please install jNeuroML following the instructions provided [[here]{.std
.std-ref}](Landing.html#jneuroml-quickstart){.reference .internal}. Run
Matlab and run the [`prefdir`{.docutils .literal .notranslate}]{.pre}
command to find the location of your preferences folder. Create a file
[`javaclasspath.txt`{.docutils .literal .notranslate}]{.pre} within that
folder containing, on a single line, the full path to the
[`jNeuroML-<version>-jar-with-dependencies.jar`{.docutils .literal
.notranslate}]{.pre} from jNeuroML.

Restart Matlab, and you will be able to access jNeuroML classes. You can
test your setup by validating an example file:

::: {.highlight-default .notranslate}
::: highlight
    import org.neuroml.model.util.NeuroML2Validator
    file = java.io.File('/full/path/to/model.nml');
    validator = NeuroML2Validator();
    validator.validateWithTests(file);
    disp(validator.getValidity())
:::
:::
:::

::: {#documentation .section}
[]{#neuromlmatlab-docs}

#### Documentation[\#](#documentation "Link to this heading"){.headerlink}

Please refer to the [[jNeuroML documentation]{.std
.std-ref}](Landing.html#jneuroml-docs){.reference .internal} for
information on the Java NeuroML API. Examples on using the Matlab
toolbox are available
[here](https://github.com/NeuroML/NeuroMLToolbox/blob/master/examples/run_examples.m){.reference
.external}.
:::

::: {#getting-help .section}
[]{#neuromlmatlab-gethelp}

#### Getting help[\#](#getting-help "Link to this heading"){.headerlink}

For any questions regarding the NeuroML Matlab toolbix, please open an
issue on the GitHub issue tracker
[here](https://github.com/NeuroML/NeuroMLToolbox/issues){.reference
.external}. Any bugs and feature requests can also be filed there.

You can also use any of the [[communication channels of the NeuroML
community]{.std .std-ref}](Landing.html#contact){.reference .internal}.
:::

::: {#development .section}
[]{#neuromlmatlab-development}

#### Development[\#](#development "Link to this heading"){.headerlink}

The NeuroML Matlab toolbox is developed on GitHub at
[NeuroML/NeuroMLToolbox](https://github.com/NeuroML/NeuroMLToolbox){.github
.reference .external}.
:::
:::

[]{#document-Userdocs/Software/SupportingTools}

::: {#tools-and-resources-with-neuroml-support .section .tex2jax_ignore .mathjax_ignore}
[]{#userdocs-supporting}

### Tools and resources with NeuroML support[\#](#tools-and-resources-with-neuroml-support "Link to this heading"){.headerlink}

Apart from the **[[core NeuroML tools]{.std
.std-ref}](Landing.html#userdocs-software){.reference .internal}** (e.g.
[[pyNeuroML]{.std .std-ref}](Landing.html#pyneuroml){.reference
.internal}, [[jNeuroML]{.std
.std-ref}](Landing.html#jneuroml){.reference .internal}) there are many
other applications, libraries and databases which support NeuroML 2 and
LEMS.

[![h](_images/neuron.png)](#neuron)
[![h](_images/neuroconstruct.png)](#neuroconstruct)
[![h](_images/netpyne.png)](#netpyne)
[![h](_images/genesis.png)](#genesis) [![h](_images/moose.jpg)](#moose)
[![h](_images/brian2.png)](#brian) [![h](_images/arbor.png)](#arbor)
[![h](_images/pynn.png)](#pynn)
[![h](_images/mdf.png)](#model-description-format-mdf)
[![h](_images/openworm.png)](#openworm)
[![h](_images/tvb.png)](#the-virtual-brain)
[![h](_images/lfpy.png)](#lfpy)
[![h](_images/biosimulators.png)](#biosimulators)
[![h](_images/neuronland.png)](#neuronland)
[![h](_images/cx3d.png)](#cx3d) [![h](_images/trees.png)](#trees)
[![h](_images/trakem2.png)](#trakem2)
[![h](_images/neuronvisio.png)](#neuronvisio)
[![h](_images/myokit.png)](#myokit)
[![h](_images/geppetto.png)](#geppetto)
[![h](_images/catmaid.png)](#catmaid)

These tools take a [[number of different approaches]{.std
.std-ref}](Landing.html#userdocs-neuroml-support-approaches){.reference
.internal} to adding NeuroML support, from dealing with the format
natively to allowing import/export of (subsets of) the language, to an
external application generating scripts/code for use in the simulator.

::: {.note .admonition}
Please help us keep this page up to date.

Tools listed here may have moved to new locations, or may no longer be
maintained, and others may be missing. Please [file
issues](https://github.com/NeuroML/Documentation/issues/new/choose){.reference
.external} if you can help update this information.
:::

::: {#applications-with-neuroml-support .section}
[]{#userdocs-supporting-apps}

#### Applications with NeuroML support[\#](#applications-with-neuroml-support "Link to this heading"){.headerlink}

::: {#neuron .section}
[]{#userdocs-supporting-apps-neuron}

##### NEURON[\#](#neuron "Link to this heading"){.headerlink}

![NEURON logo](_images/neuron.png)

The [NEURON](http://www.neuron.yale.edu/neuron/){.reference .external}
simulation environment is one of the main target platforms for a
standard facilitating exchange of neuronal models. [[jNeuroML]{.std
.std-ref}](Landing.html#jneuroml){.reference .internal} can be used to
convert NeuroML2/LEMS models to NEURON. NEURON simulations can also be
generated from NeuroML model components by
[neuroConstruct](http://www.neuroconstruct.org){.reference .external}.

See also [[NetPyNE]{.std
.std-ref}](Landing.html#userdocs-netpyne){.reference .internal}, which
builds on NEURON.

There is a **dedicated page on NEURON/NeuroML interactions**
[[here]{.std .std-ref}](Landing.html#userdocs-neuron){.reference
.internal}.
:::

::: {#netpyne .section}
[]{#userdocs-supporting-apps-netpyne}

##### NetPyNE[\#](#netpyne "Link to this heading"){.headerlink}

![NetPyNE logo](_images/netpyne.png)

[NetPyNE](http://netpyne.org/){.reference .external} is a Python package
to facilitate the development, simulation, parallelization, analysis,
and optimization of biological neuronal networks using the NEURON
simulator. NetPyNE can import from and export to NeuroML. NetPyNE also
provides a web based [Graphical User
Interface](https://github.com/MetaCell/NetPyNE-UI/wiki){.reference
.external}.

There is a **dedicated page on NetPyNE/NeuroML interactions**
[[here]{.std .std-ref}](Landing.html#userdocs-netpyne){.reference
.internal}.
:::

::: {#neuroconstruct .section}
[]{#userdocs-supporting-apps-neuroconstruct}

##### neuroConstruct[\#](#neuroconstruct "Link to this heading"){.headerlink}

![Neuroconstruct logo](_images/neuroconstruct.png)

[neuroConstruct](http://www.neuroconstruct.org){.reference .external} is
a Java based application for constructing 3D networks of biologically
realistic neurons. The current version can generate code for the
[NEURON](http://www.neuron.yale.edu/neuron/){.reference .external},
[GENESIS](http://genesis-sim.org/){.reference .external},
[PSICS](http://www.psics.org/index.html){.reference .external} and
[PyNN](http://neuralensemble.org/PyNN/){.reference .external} platforms
and also provides import/export support for MorphML, ChannelML and
NetworkML (from NeuroML v1) and for NeuroMLv2 cells and networks.

More info on the support for NeuroML in neuroConstruct is available
[here](http://www.neuroconstruct.org/docs/neuroml.html){.reference
.external}.
:::

::: {#genesis .section}
[]{#userdocs-supporting-apps-genesis}

##### GENESIS[\#](#genesis "Link to this heading"){.headerlink}

![GENESIS logo](_images/genesis.png)

[GENESIS](http://genesis-sim.org/){.reference .external} is a commonly
used neuronal simulation environment and was a main target platform for
the NeuroMLv1 specifications. Full GENESIS simulations can be generated
from NeuroMLv1 model components by
[neuroConstruct](http://www.neuroconstruct.org){.reference .external}.

Due to the lack of active development of GENESIS, support for mapping to
GENESIS in NeuroMLv2 has been deprecated in favour of [[MOOSE]{.std
.std-ref}](#userdocs-supporting-apps-moose){.reference .internal}.
:::

::: {#moose .section}
[]{#userdocs-supporting-apps-moose}

##### MOOSE[\#](#moose "Link to this heading"){.headerlink}

![MOOSE logo](_images/moose.jpg)

[MOOSE](https://moose.ncbs.res.in/){.reference .external} is the
Multiscale Object-Oriented Simulation Environment. It is the base and
numerical core for large, detailed multi-scale simulations that span
computational neuroscience and systems biology. It is based on a
complete reimplementation of the GENESIS 2 core.

More information on running NeuroML models in MOOSE can be found
[[here]{.std
.std-ref}](Landing.html#userdocs-simulating-models-moose){.reference
.internal}.

There is a **dedicated page on MOOSE/NeuroML interactions** [[here]{.std
.std-ref}](Landing.html#userdocs-moose){.reference .internal}.
:::

::: {#brian .section}
[]{#userdocs-supporting-apps-brian}

##### BRIAN[\#](#brian "Link to this heading"){.headerlink}

![Brian logo](_images/brian2.png)

[Brian](https://briansimulator.org/){.reference .external} is an easy to
use, Python based simulator of spiking networks.

There is a **dedicated page on Brian/NeuroML interactions** [[here]{.std
.std-ref}](Landing.html#userdocs-brian){.reference .internal}.
:::

::: {#eden .section}
[]{#userdocs-supporting-apps-eden}

##### EDEN[\#](#eden "Link to this heading"){.headerlink}

[EDEN](https://gitlab.com/neurocomputing-lab/Inferior_OliveEMC/eden){.reference
.external} is a recently developed simulation engine which incorporates
native NeuroML 2 support from the start.

Initial tests of using EDEN with NeuroML models and example code can be
found [here](https://github.com/OpenSourceBrain/EDENShowcase){.reference
.external}.

There is a **dedicated page on EDEN/NeuroML interactions** [[here]{.std
.std-ref}](Landing.html#userdocs-eden){.reference .internal}.
:::

::: {#arbor .section}
[]{#userdocs-supporting-apps-arbor}

##### Arbor[\#](#arbor "Link to this heading"){.headerlink}

![Arbor logo](_images/arbor.png)

[Arbor](https://arbor-sim.org/){.reference .external} is a high
performance multicompartmental neural simulation library. Addition of
support for NeuroML2 and LEMS is under active development. See
[here](https://docs.arbor-sim.org/en/stable/fileformat/neuroml.html){.reference
.external}.

Example code for interactions between NeuroML models and Arbor can be
found
[here](https://github.com/OpenSourceBrain/ArborShowcase){.reference
.external}.

There is a **dedicated page on Arbor/NeuroML interactions** [[here]{.std
.std-ref}](Landing.html#userdocs-arbor){.reference .internal}.
:::

::: {#pynn .section}
[]{#userdocs-supporting-apps-pynn}

##### PyNN[\#](#pynn "Link to this heading"){.headerlink}

![PyNN logo](_images/pynn.png)

[PyNN](http://neuralensemble.org/PyNN/){.reference .external} is a
Python package for simulator independent specification of neuronal
network models. Model code can be developed using the PyNN API and then
run using [NEURON](http://www.neuron.yale.edu/neuron/){.reference
.external}, [NEST](https://nest-simulator.org/){.reference .external} or
[Brian](https://briansimulator.org/){.reference .external}. The
developed model also can be stored as a NeuroML document. The latest
version of [neuroConstruct](http://www.neuroconstruct.org){.reference
.external} can be used to generate executable scripts for PyNN based
simulators based on NeuroML components, although the majority of
multicompartmental conductance based models which are available in
neuroConstruct are outside the current scope of the PyNN API.

More info on the latest support for running NeuroML models in PyNN and
vice versa can be found
[here](https://github.com/NeuroML/NeuroML2/issues/73){.reference
.external}.
:::

::: {#nest .section}
[]{#userdocs-supporting-apps-nest}

##### NEST[\#](#nest "Link to this heading"){.headerlink}

![NEST logo](_images/nest-logo.png)

NEST is a simulator for spiking neural network models that focuses on
the dynamics, size and structure of neural systems rather than on the
exact morphology of individual neurons.

There is a **dedicated page on NEST/NeuroML interactions** [[here]{.std
.std-ref}](Landing.html#userdocs-nest){.reference .internal}.
:::

::: {#openworm .section}
[]{#userdocs-supporting-apps-openworm}

##### OpenWorm[\#](#openworm "Link to this heading"){.headerlink}

![OpenWorm logo](_images/openworm.png)

The [OpenWorm](http://openworm.org/){.reference .external} project aims
to create a simulation platform to build digital *in-silico* living
systems, starting with a *C. elegans* virtual organism simulation. The
simulations and associated tools are being developed in a fully open
source manner. NeuroML is being used for the description of the 302
neurons in the worm's nervous system, both for morphological description
of the cells and their electrical properties.

The [c302 subproject](https://github.com/openworm/c302){.reference
.external} in OpenWorm has the latest developments in the NeuroML
version of the worm nervous system.

Members of the OpenWorm project are also creating a general purpose
neuronal simulator (for both electrical and physical simulations) which
will have parallelism and native support for NeuroML built in from the
start (see [[Geppetto]{.std
.std-ref}](#userdocs-supporting-apps-geppetto){.reference .internal}.
:::

::: {#model-description-format-mdf .section}
[]{#userdocs-supporting-apps-mdf}

##### Model Description Format (MDF)[\#](#model-description-format-mdf "Link to this heading"){.headerlink}

![MDF logo](_images/mdf.png)

[ModECI Model Description Format
(MDF)](https://github.com/ModECI/MDF){.reference .external} is an open
source, community-supported standard and associated library of tools for
expressing computational models in a form that allows them to be
exchanged between diverse programming languages and execution
environments, with a particular focus on machine learning, artificial
intelligence and computational neuroscience.

There will be full compatibility between NeuroML and MDF for specifying
neuronal models. See
[here](https://github.com/ModECI/MDF/blob/main/examples/NeuroML/README.md){.reference
.external} for ongoing work in this direction.
:::

::: {#the-virtual-brain .section}
[]{#userdocs-supporting-apps-tvb}

##### The Virtual Brain[\#](#the-virtual-brain "Link to this heading"){.headerlink}

![TVB logo](_images/tvb.png)

[The Virtual Brain (TVB)](https://www.thevirtualbrain.org){.reference
.external} offers a simulation environment for large-scale brain
networks. It allows network properties, in particular the brain's
structural connectivity, to be incorporated into models, and so TVB can
simulate whole brain behaviour as is commonly observed in clinical
scanners (e.g. EEG, MEG, fMRI).

Initial work mapping networks in TVB to/from NeuroML 2 and LEMS can be
found
[here](https://github.com/OpenSourceBrain/TheVirtualBrainShowcase){.reference
.external}. See also the work of the [INCF Network Specification Working
Group in this
area](https://github.com/NeuralEnsemble/Networks_SIG/issues?q=is%3Aissue+is%3Aopen+label%3ATVB){.reference
.external}.
:::

::: {#lfpy .section}
[]{#userdocs-supporting-apps-lfpy}

##### LFPy[\#](#lfpy "Link to this heading"){.headerlink}

![LFPy logo](_images/lfpy.png)

[LFPy](https://lfpy.readthedocs.io/en/latest/){.reference .external} is
a Python package for calculation of extracellular potentials from
multicompartment neuron models. It relies on the NEURON simulator and
uses the Python interface it provides. LFPy provides a set of easy to
use Python classes for setting up the model, running simulations and
calculating the extracellular potentials arising from activity in the
model neuron. Initial support for loading of NeuroML morphologies has
been added.
:::

::: {#biosimulators .section}
[]{#userdocs-supporting-apps-biosimulators}

##### BioSimulators[\#](#biosimulators "Link to this heading"){.headerlink}

![BioSimulators logo](_images/biosimulators.png)

[BioSimulators](https://biosimulators.org/){.reference .external}
provides a registry and platform supporting a broad range of modeling
frameworks, model formats, simulation algorithms, and simulation tools.

See for example
[https://biosimulators.org/simulators/pyneuroml/latest](https://biosimulators.org/simulators/pyneuroml/latest){.reference
.external}.
:::

::: {#n2a .section}
[]{#userdocs-supporting-apps-n2a}

##### N2A[\#](#n2a "Link to this heading"){.headerlink}

"Neurons to Algorithms" (N2A) is a language for modeling neural systems,
along with a software tool for editing models and simulating them.

There is a **dedicated page on N2A/NeuroML interactions** [[here]{.std
.std-ref}](Landing.html#userdocs-n2a){.reference .internal}.
:::

::: {#neuronland .section}
[]{#userdocs-supporting-apps-neuronland}

##### NeuronLand[\#](#neuronland "Link to this heading"){.headerlink}

![NeuronLand logo](_images/neuronland.png)

[NeuronLand](http://neuronland.org/){.reference .external} provides
NLMorphologyConverter, which is a command line program for converting
between over 20 different 3D neuron morphology formats, and
NLMorphologyViewer, which provides a simple interface for viewing these
data. Both of these tools provide import and export of MorphML.
:::

::: {#cx3d .section}
[]{#userdocs-supporting-apps-cx3d}

##### CX3D[\#](#cx3d "Link to this heading"){.headerlink}

![CX3D logo](_images/cx3d.png)

[CX3D](http://www.ini.uzh.ch/~amw/seco/cx3d/){.reference .external} is a
tool for simulating the growth of cortex in 3D. There was a preliminary
implementation of export of generated networks to NeuroML in CX3D.
:::

::: {#trees-toolbox .section}
[]{#userdocs-supporting-apps-trees}

##### TREES toolbox[\#](#trees-toolbox "Link to this heading"){.headerlink}

![Trees logo](_images/trees.png)

The [TREES toolbox](https://github.com/cuntzlab/treestoolbox){.reference
.external} is an application in MATLAB which allows: automatic
reconstruction of neuronal branching from microscopy image stacks and
generation of synthetic axonal and dendritic trees; visualisation,
editing and analysis of neuronal trees; comparison of branching patterns
between neurons; and investigation of how dendritic and axonal branching
depends on local optimization of total wiring and conduction distance.

The latest version of the TREES toolbox includes basic functionality for
exporting cells in NeuroML v1.x Level 1 (MorphML) or as a NeuroML
v2alpha morphology file.
:::

::: {#trakem2 .section}
[]{#userdocs-supporting-apps-trackem}

##### TrakEM2[\#](#trakem2 "Link to this heading"){.headerlink}

![TrakEM2 logo](_images/trakem2.png)

[TrakEM2](http://www.ini.uzh.ch/~acardona/trakem2.html){.reference
.external} is an ImageJ plugin for morphological data mining,
three-dimensional modelling and image stitching, registration, editing
and annotation. As of v0.8n, a menu item "Export - NeuroML..." gives an
option to export to MorphML (the anatomy of the arbors only) or NeuroML
(the whole network with anatomy and synapses), for the selected trees or
all trees.
:::

::: {#neuronvisio .section}
[]{#userdocs-supporting-apps-neurovisio}

##### Neuronvisio[\#](#neuronvisio "Link to this heading"){.headerlink}

![Neuronvisio logo](_images/neuronvisio.png)

[Neuronvisio](http://neuronvisio.org/){.reference .external} is a
Graphical User Interface for NEURON simulator environment with 3D
capabilities. Neuronvisio makes easy to select and investigate sections'
properties, it offers easy integration with matplotlib for the plotting
the results. It can save the geometry using NeuroML and the simulation
results in a customised and extensible HDF5 format; the results can then
be reload in the software and analysed at a later stage, without
re-running the simulation.
:::

::: {#catmaid .section}
[]{#userdocs-supporting-apps-catmaid}

##### CATMAID[\#](#catmaid "Link to this heading"){.headerlink}

![CATMAID logo](_images/catmaid.png)

[CATMAID](https://catmaid.readthedocs.io/en/stable/){.reference
.external} is the Collaborative Annotation Toolkit for Massive Amounts
of Image Data, and is a widely used tool for online reconstruction and
annotation of connectomics data. Initial support for export of
reconstructed neurons in NeuroML format has been added.
:::

::: {#myokit .section}
[]{#userdocs-supporting-apps-myokit}

##### Myokit[\#](#myokit "Link to this heading"){.headerlink}

![Myokit logo](_images/myokit.png)

[Myokit](http://myokit.org/){.reference .external} (the Maastricht
Myocyte Toolkit) is a Python-based software package created by Michael
Clerx to simplify the use of numerical models in the analysis of cardiac
myocytes. Initial support for importing ChannelML [has been
added](https://myokit.readthedocs.io/en/stable/api_formats/index.html#api-formats){.reference
.external}.
:::

::: {#geppetto .section}
[]{#userdocs-supporting-apps-geppetto}

##### Geppetto[\#](#geppetto "Link to this heading"){.headerlink}

![Geppetto logo](_images/geppetto.png)

[Geppetto](http://www.geppetto.org/){.reference .external} is a
web-based multi-algorithm, multi-scale simulation platform designed to
support the simulation of complex biological systems and their
surrounding environment. It is open source and is being developed as
part of the [[OpenWorm project]{.std
.std-ref}](#userdocs-supporting-apps-openworm){.reference .internal} to
create an *in-silico* model of the nematode *C. elegans*. It has had
inbuilt support for NeuroML 2/LEMS from the start, and is suitable for
many other types of neuronal models.
:::
:::

::: {#other-legacy-tools .section}
[]{#userdocs-supporting-other}

#### Other/legacy tools[\#](#other-legacy-tools "Link to this heading"){.headerlink}

::: {.admonition-older-applications .admonition}
Older applications

Note: many of the applications listed below are no longer in active
development or links no longer work.
:::

::: {#psics .section}
[]{#userdocs-supporting-other-psics}

##### PSICS[\#](#psics "Link to this heading"){.headerlink}

The latest version of
[neuroConstruct](http://www.neuroconstruct.org){.reference .external}
can be used to generate executable scripts for
[PSICS](http://www.psics.org/index.html){.reference .external} based on
NeuroML components.
:::

::: {#whole-brain-catalogue .section}
[]{#userdocs-supporting-other-wbc}

##### Whole Brain Catalogue[\#](#whole-brain-catalogue "Link to this heading"){.headerlink}

The [Whole Brain Catalog](https://twitter.com/braincatalog){.reference
.external} was a graphical interface that allowed multiscale
neuroscience data to be visualised relative to a 3D brain atlas.
:::

::: {#pcsim .section}
[]{#userdocs-supporting-other-pcsim}

##### PCSIM[\#](#pcsim "Link to this heading"){.headerlink}

[PCSIM](http://www.lsm.tugraz.at/pcsim/){.reference .external} is a tool
in C++ for simulating large scale networks of cells and synapses.
:::

::: {#neuromantic .section}
[]{#userdocs-supporting-other-neuromantic}

##### Neuromantic[\#](#neuromantic "Link to this heading"){.headerlink}

[Neuromantic](https://sourceforge.net/projects/neuromantic/){.reference
.external} is a freeware tool for neuronal reconstruction (similar in
some ways to part of Neurolucida's functionality). Neuromantic mainly
uses SWC/Cvapp format, but the latest version can import and export
MorphML.
:::

::: {#neurospaces-genesis-3 .section}
[]{#userdocs-supporting-other-neurospaces}

##### Neurospaces/ GENESIS 3[\#](#neurospaces-genesis-3 "Link to this heading"){.headerlink}

The [Neurospaces/ GENESIS
3](http://neurospaces.sourceforge.net/){.reference .external} project is
developing a modular reimplementation of the core of GENESIS 2 along
with a number of other components for computational neuroscience as part
of the GENESIS 3 initiative. Neurospaces/GENESIS 3 currently supports
reading of passive models in NeuroML format (morphology + passive
parameters).
:::

::: {#splitneuron .section}
[]{#userdocs-supporting-other-splitneuron}

##### SplitNeuron[\#](#splitneuron "Link to this heading"){.headerlink}

[SplitNeuron](https://sourceforge.net/projects/splitneuron/){.reference
.external} is a library written in C for data structures and functions
extending SQLite to simulate large-scale networks of Izhikevich Simple
Model compartments. SplitNeuron answers a fundamental issue in
large-scale simulation, data transfer between storage and functional
software: it uses database not only for data storage but also as
simulation engine, moving computation to data rather than using storage
systems only for data holding. This choice offers more features with
less code to write and a unique way of accessing data for further
analysis. Features under development include direct import and
cell/network creation from NeuroML.
:::

::: {#neuranim .section}
[]{#userdocs-supporting-other-neuranim}

##### NeurAnim[\#](#neuranim "Link to this heading"){.headerlink}

[NeurAnim](https://sourceforge.net/projects/neuranim/){.reference
.external} is a research aid for computational neuroscience. It is used
to visualise and animate neural network simulations in 3D, and to render
movies of these animations for use in presentations. Networks stored in
the instance based representation of NetworkML can be loaded and
visualised.
:::

::: {#cnrun .section}
[]{#userdocs-supporting-other-cnrun}

##### CNrun[\#](#cnrun "Link to this heading"){.headerlink}

[CNrun](http://johnhommer.com/academic/code/cnrun/){.reference
.external} is a neuronal network model simulator, similar in purpose to
NEURON except that individual neurons are not compartmentalised. It was
built from refactored code written by Thomas Nowotny. It reads in
network topology description from a NeuroML file, where the cell_type
attribute determines the unit class, one of the in-built neuron types of
CNrun (e.g. Hodgkin Huxley cell by Traub and Miles (1991), Poisson
oscillator, van der Pol oscillator).
:::

::: {#neugen .section}
[]{#userdocs-supporting-other-neugen}

##### NeuGen[\#](#neugen "Link to this heading"){.headerlink}

[NeuGen](https://durus.gcsc.uni-frankfurt.de/~neugen/){.reference
.external} is an application in Java which is able to generate networks
of synaptically connected morphologically detailed neurons, as in a
cortical column. NeuGen generates sets of neurons of the different
morphological classes of the cortex, e.g. pyramidal cells and stellate
neurons, and connects these networks in 3D. The latest version of NeuGen
can export the generated networks to NeuroML. Some manual editing of the
generated files is required to make them valid. The developers have been
informed of the required updates which will be incorporated soon.
:::

::: {#morphforge .section}
[]{#userdocs-supporting-other-morphforge}

##### morphforge[\#](#morphforge "Link to this heading"){.headerlink}

[morphforge](https://github.com/mikehulluk/morphforge){.reference
.external} is a high level, simulator independent, Python library for
building simulations of small populations of multi-compartmental
neurons. It was built as part of the PhD thesis of Mike Hull (Uni.
Edinburgh): Investigating the role of electrical coupling in small
populations of interneurons in Xenopus laevis tadpoles. Loading of
morphologies in MorphML format is supported, and loading of channel
descriptions from ChannelML is in progress. Future development of
morphforge will be closely aligned with the development of the
multicompartmental modelling API in Python (libNeuroML).
:::

::: {#neurotranslate .section}
[]{#userdocs-supporting-other-neurotranslate}

##### NeuroTranslate[\#](#neurotranslate "Link to this heading"){.headerlink}

[NeuroTranslate](https://github.com/nathanjordan/NeuroTranslate){.reference
.external} is a tool that translates input files between two different
languages, the NCS (Neo-Cortical Simulator) input language and NeuroML
format. It provides a user-friendly interface, which can be used to both
create and edit simulations.
:::

::: {#moogli .section}
[]{#userdocs-supporting-other-moogli}

##### Moogli[\#](#moogli "Link to this heading"){.headerlink}

[Moogli](https://moose.ncbs.res.in/readthedocs/user/py/graphics/index_graphics.html){.reference
.external} (a sister project of [[MOOSE]{.std
.std-ref}](#userdocs-supporting-apps-moose){.reference .internal} is a
simulator independent OpenGL based visualization tool for neural
simulations. Moogli can visualize morphology of single/multiple neurons
or network of neurons, and can also visualize activity in these cells.
Loading of morphologies in MorphML and NeuroML formats is supported.
:::
:::

::: {.toctree-wrapper .compound}
[]{#document-Userdocs/Software/Tools/Approaches}

::: {#approaches-to-adding-neuroml-support .section .tex2jax_ignore .mathjax_ignore}
[]{#userdocs-neuroml-support-approaches}

#### Approaches to adding NeuroML support[\#](#approaches-to-adding-neuroml-support "Link to this heading"){.headerlink}

There are a number of ways that a neuronal simulator can add "support
for NeuroML", depending on how deeply it embeds/supports the elements of
the language.

::: {#commonly-used-approaches .section}
##### Commonly used approaches[\#](#commonly-used-approaches "Link to this heading"){.headerlink}

::: {#native-support-for-neuroml-elements .section}
[]{#userdocs-neuroml-support-approaches-native}

###### 1) Native support for NeuroML elements[\#](#native-support-for-neuroml-elements "Link to this heading"){.headerlink}

A simulator may have an equivalent internal representation of the core
concepts from NeuroML2/LEMS, and so be able to natively read/write these
formats.

This is the approach taken in [[jNeuroML]{.std
.std-ref}](Landing.html#jneuroml){.reference .internal} and [[EDEN]{.std
.std-ref}](Landing.html#userdocs-eden){.reference .internal}.
:::

::: {#native-ability-to-import-neuroml-elements .section}
[]{#userdocs-neuroml-support-approaches-native-import}

###### 2) Native ability to import NeuroML elements[\#](#native-ability-to-import-neuroml-elements "Link to this heading"){.headerlink}

Another approach is for simulators to natively support importing (a
subset of) NeuroML models, whereby the NeuroML components are converted
to the equivalent entities in the simulator's internal representation of
the model.

This is the approach taken in [[MOOSE]{.std
.std-ref}](Landing.html#userdocs-supporting-apps-moose){.reference
.internal}, [[Arbor]{.std
.std-ref}](Landing.html#userdocs-arbor){.reference .internal} and
[[NetPyNE]{.std .std-ref}](Landing.html#userdocs-netpyne){.reference
.internal}.
:::

::: {#native-ability-to-export-neuroml-elements .section}
[]{#userdocs-neuroml-support-approaches-native-export}

###### 3) Native ability to export NeuroML elements[\#](#native-ability-to-export-neuroml-elements "Link to this heading"){.headerlink}

Some simulators allow models to be created with their preferred native
model description format, and then exported in valid NeuroML.

This is the approach taken in [[NEURON]{.std
.std-ref}](Landing.html#userdocs-neuron){.reference .internal} and
[[NetPyNE]{.std .std-ref}](Landing.html#userdocs-netpyne){.reference
.internal}. It is also possible to export [[PyNN]{.std
.std-ref}](Landing.html#userdocs-pynn){.reference .internal} models to
NeuroML equivalents.
:::

::: {#rd-party-mapping-to-simulators-own-format .section}
[]{#userdocs-neuroml-support-approaches-mapping}

###### 4) 3rd party mapping to simulator's own format[\#](#rd-party-mapping-to-simulators-own-format "Link to this heading"){.headerlink}

This is the approach taken in [[NEURON]{.std
.std-ref}](Landing.html#userdocs-neuron){.reference .internal} via
[[jNeuroML]{.std .std-ref}](Landing.html#jneuroml){.reference
.internal}.
:::
:::
:::

[]{#document-Userdocs/Software/Tools/NEURON}

::: {#neuron-and-neuroml .section .tex2jax_ignore .mathjax_ignore}
[]{#userdocs-neuron}

#### NEURON and NeuroML[\#](#neuron-and-neuroml "Link to this heading"){.headerlink}

![NEURON logo](_images/neuron.png)

[NEURON](http://www.neuron.yale.edu/neuron){.reference .external} is a
widely used simulation environment and is one of the main target
platforms for a standard facilitating exchange of neuronal models.

::: {#simulating-neuroml-models-in-neuron .section}
[]{#userdocs-neuron-simulating}

##### Simulating NeuroML models in NEURON[\#](#simulating-neuroml-models-in-neuron "Link to this heading"){.headerlink}

[[jNeuroML]{.std .std-ref}](Landing.html#jneuroml){.reference .internal}
or [[pyNeuroML]{.std .std-ref}](Landing.html#pyneuroml){.reference
.internal} can be used to convert NeuroML2/LEMS models to NEURON. This
involves pointing at a [[LEMS Simulation file]{.std
.std-ref}](Landing.html#userdocs-lemssimulation){.reference .internal}
describing what to simulate, and using the [`-neuron`{.docutils .literal
.notranslate}]{.pre} option:

::: {.highlight-console .notranslate}
::: highlight
    # Simulate the model using NEURON with python/hoc/mod files generated by jNeuroML
    jnml <LEMS simulation file> -neuron -run

    # Simulate the model using NEURON with python/hoc/mod files generated by pyNeuroML
    pynml <LEMS simulation file> -neuron -run
:::
:::

These commands generate a PyNeuron script and run it (a file ending in
[`_nrn.py`{.docutils .literal .notranslate}]{.pre}). So you must have
NEURON installed on your system, with its Python bindings (PyNeuron).
Skipping the [`-run`{.docutils .literal .notranslate}]{.pre} flag will
generate the Python script but will not run it: you can run it manually
later. Adding [`-nogui`{.docutils .literal .notranslate}]{.pre} will
suppress the NEURON graphical elements/menu opening and just run the
model in NEURON in the background

You can also run LEMS simulations using the NEURON simulator using the
[[pyNeuroML]{.std .std-ref}](Landing.html#pyneuroml){.reference
.internal} API:

::: {.highlight-python .notranslate}
::: highlight
    from pyneuroml.pynml import run_lems_with_jneuroml_neuron

    ...

    run_lems_with_jneuroml_neuron(lems_file_name)
:::
:::
:::

::: {#setting-the-neuron-home-environment-variable .section}
[]{#userdocs-neuron-envvar}

##### Setting the NEURON_HOME environment variable[\#](#setting-the-neuron-home-environment-variable "Link to this heading"){.headerlink}

Since it is possible to install multiple versions of NEURON in different
places, the NeuroML tools need to be told where the NEURON tools are. To
do this, they look at the [`NEURON_HOME`{.docutils .literal
.notranslate}]{.pre} environment variable. This needs to hold the path
to where the binary ([`bin`{.docutils .literal .notranslate}]{.pre})
folder holding the NEURON tools such as [`nrniv`{.docutils .literal
.notranslate}]{.pre} are located. On Linux like systems, one can use
[`which`{.docutils .literal .notranslate}]{.pre} to find these tools and
set the variable:

::: {.highlight-bash .notranslate}
::: highlight
    $ which nrniv
    ~/.local/share/virtualenvs/neuroml-311-dev/bin/nrniv

    $ export NEURON_HOME="~/.local/share/virtualenvs/neuroml-311-dev/"
:::
:::

One can combine these commands together also:

::: {.highlight-bash .notranslate}
::: highlight
    $ export NEURON_HOME="$(dirname $(dirname $(which nrniv)))"
:::
:::
:::

::: {#using-neuroconstruct .section}
[]{#userdocs-neuron-neuroconstruct}

##### Using neuroConstruct[\#](#using-neuroconstruct "Link to this heading"){.headerlink}

NEURON simulations can also be generated from NeuroML model components
by [[neuroConstruct]{.std
.std-ref}](Landing.html#userdocs-supporting-apps-neuroconstruct){.reference
.internal}, but most of this functionality is related to [[NeuroML
v1]{.std .std-ref}](Landing.html#userdocs-neuromlv1){.reference
.internal}.
:::
:::

[]{#document-Userdocs/Software/Tools/NetPyNE}

::: {#netpyne-and-neuroml .section .tex2jax_ignore .mathjax_ignore}
[]{#userdocs-netpyne}

#### NetPyNE and NeuroML[\#](#netpyne-and-neuroml "Link to this heading"){.headerlink}

![NetPyNE logo](_images/netpyne.png)

[NetPyNE](http://netpyne.org){.reference .external} is a Python package
to facilitate the development, simulation, parallelization, analysis,
and optimization of biological neuronal networks using the NEURON
simulator. NetPyNE can import from and export to NeuroML. NetPyNE also
provides a web based [Graphical User
Interface](https://github.com/MetaCell/NetPyNE-UI/wiki){.reference
.external}.

::: {#importing-neuroml-into-netpyne .section}
##### Importing NeuroML into NetPyNE[\#](#importing-neuroml-into-netpyne "Link to this heading"){.headerlink}

An example of how to import a network in NeuroML into NetPyNE can be
found
[here](https://github.com/Neurosim-lab/netpyne/blob/development/examples/NeuroMLImport/SimpleNet_import.py){.reference
.external}.
:::

::: {#exporting-neuroml-from-netpyne .section}
##### Exporting NeuroML from NetPyNE[\#](#exporting-neuroml-from-netpyne "Link to this heading"){.headerlink}

An example of how to export a network built using NetPyNE to NeuroML can
be found
[here](https://github.com/OpenSourceBrain/NetPyNEShowcase/blob/master/NetPyNE/HHSmall/HH_export.py){.reference
.external}.
:::

::: {#running-netpyne-on-osbv2 .section}
##### Running NetPyNE on OSBv2[\#](#running-netpyne-on-osbv2 "Link to this heading"){.headerlink}

Building and running NetPyNE models will be a core feature of Open
Source Brain v2.0. See
[here](https://docs.opensourcebrain.org/OSBv2/NetPyNE.html){.reference
.external} for more details.
:::

::: {#neuromllite .section}
##### NeuroMLlite[\#](#neuromllite "Link to this heading"){.headerlink}

NetPyNE is also a key target for cross simulator network creation using
[[NeuroMLlite]{.std .std-ref}](Landing.html#neuromllite){.reference
.internal}. There are ongoing plans for greater alignment between
formats used for network specification in NetPyNE and NeuroMLlite.
:::
:::

[]{#document-Userdocs/Software/Tools/PyNN}

::: {#pynn-and-neuroml .section .tex2jax_ignore .mathjax_ignore}
[]{#userdocs-pynn}

#### PyNN and NeuroML[\#](#pynn-and-neuroml "Link to this heading"){.headerlink}

![PyNN logo](_images/pynn.png)

[PyNN](http://neuralensemble.org/PyNN/){.reference .external} is a
Python package for simulator independent specification of neuronal
network models. Model code can be developed using the PyNN API and then
run using [NEURON](http://www.neuron.yale.edu/neuron/){.reference
.external}, [NEST](https://nest-simulator.org/){.reference .external} or
[Brian](https://briansimulator.org/){.reference .external}. The
developed model also can be stored as a NeuroML document.

The latest version of [[neuroConstruct]{.std
.std-ref}](Landing.html#userdocs-supporting-apps-neuroconstruct){.reference
.internal} can be used to generate executable scripts for PyNN based
simulators based on NeuroML components, although the majority of
multicompartmental conductance based models which are available in
neuroConstruct are outside the current scope of the PyNN API.

See
[OpenSourceBrain/PyNNShowcase](https://github.com/OpenSourceBrain/PyNNShowcase){.github
.reference .external} for examples of usage of NeuroML and PyNN.

More info on the latest support for running NeuroML models in PyNN and
vice versa can be found
[here](https://github.com/NeuroML/NeuroML2/issues/73){.reference
.external}.

PyNN is also a key target for cross simulator network creation using
[[NeuroMLlite]{.std .std-ref}](Landing.html#neuromllite){.reference
.internal}.
:::

[]{#document-Userdocs/Software/Tools/Brian}

::: {#brian-and-neuroml .section .tex2jax_ignore .mathjax_ignore}
[]{#userdocs-brian}

#### Brian and NeuroML[\#](#brian-and-neuroml "Link to this heading"){.headerlink}

![Brian logo](_images/brian2.png)

[Brian](https://briansimulator.org/){.reference .external} is an easy to
use, Python based simulator of spiking networks.

::: {#converting-neuroml-model-to-brian .section}
##### Converting NeuroML model to Brian[\#](#converting-neuroml-model-to-brian "Link to this heading"){.headerlink}

[[jNeuroML]{.std .std-ref}](Landing.html#jneuroml){.reference .internal}
or [[pyNeuroML]{.std .std-ref}](Landing.html#pyneuroml){.reference
.internal} can be used to convert NeuroML2/LEMS models to [Brian version
2](https://github.com/brian-team/brian2){.reference .external}. This
involves pointing at a [[LEMS Simulation file]{.std
.std-ref}](Landing.html#userdocs-lemssimulation){.reference .internal}
describing what to simulate, and using the [`-brian2`{.docutils .literal
.notranslate}]{.pre} option:

::: {.highlight-console .notranslate}
::: highlight
    # Using jnml
    jnml <LEMS simulation file> -brian2

    # Using pynml
    pynml <LEMS simulation file> -brian2
:::
:::

This command generates a Python script (a file ending in
[`_brian2.py`{.docutils .literal .notranslate}]{.pre}) which can be run
in Python and will simulate the model and plot/save the results, as
outlined in the [[LEMS Simulation file]{.std
.std-ref}](Landing.html#userdocs-lemssimulation){.reference .internal}.

Notes:

-   Only single compartment cells can be converted to Brian format so
    far. While there is support in Brian for multicompartmental cell
    simulation, this is not yet covered in the jNeuroML based export.

-   There has been support for converting NeuroML models to Brian v1
    (using [`-brian`{.docutils .literal .notranslate}]{.pre}), but since
    this version of Brian is deprecated, and only supports Python 2,
    this export is no longer actively developed.

-   There is limited support for executing networks of cells in Brian,
    and the most likely route for adding this functionality is via
    [[NeuroMLlite]{.std .std-ref}](Landing.html#neuromllite){.reference
    .internal}.
:::

::: {#examples .section}
##### Examples[\#](#examples "Link to this heading"){.headerlink}

Example code for interactions between NeuroML models and Brian can be
found
[here](https://github.com/OpenSourceBrain/BrianShowcase){.reference
.external}.
:::
:::

[]{#document-Userdocs/Software/Tools/MOOSE}

::: {#moose-and-neuroml .section .tex2jax_ignore .mathjax_ignore}
[]{#userdocs-moose}

#### MOOSE and NeuroML[\#](#moose-and-neuroml "Link to this heading"){.headerlink}

[MOOSE](https://moose.ncbs.res.in/){.reference .external} is the
Multiscale Object-Oriented Simulation Environment. It is the base and
numerical core for large, detailed multi-scale simulations that span
computational neuroscience and systems biology. It is based on a
complete reimplementation of the GENESIS 2 core.

Some tests of using MOOSE with NeuroML models and example code can be
found in the [MOOSE
Showcase](https://github.com/OpenSourceBrain/MOOSEShowcase){.reference
.external} repository.

::: {#simulating-neuroml-models-in-moose .section}
##### Simulating NeuroML models in MOOSE[\#](#simulating-neuroml-models-in-moose "Link to this heading"){.headerlink}

You can export NeuroML models to the MOOSE simulator format using
[[jNeuroML]{.std .std-ref}](Landing.html#jneuroml){.reference .internal}
or [[pyNeuroML]{.std .std-ref}](Landing.html#pyneuroml){.reference
.internal}, pointing at a [[LEMS Simulation file]{.std
.std-ref}](Landing.html#userdocs-lemssimulation){.reference .internal}
describing what to simulate, and using the [`-moose`{.docutils .literal
.notranslate}]{.pre} option:

::: {.highlight-console .notranslate}
::: highlight
    # Using jnml
    jnml <LEMS simulation file> -moose

    # Using pynml
    pynml <LEMS simulation file> -moose
:::
:::
:::
:::

[]{#document-Userdocs/Software/Tools/EDEN}

::: {#eden-and-neuroml .section .tex2jax_ignore .mathjax_ignore}
[]{#userdocs-eden}

#### EDEN and NeuroML[\#](#eden-and-neuroml "Link to this heading"){.headerlink}

[EDEN](https://gitlab.com/neurocomputing-lab/Inferior_OliveEMC/eden){.reference
.external} is a recently developed simulation engine which incorporates
native NeuroML 2 support from the start.

Initial tests of using EDEN with NeuroML models and example code can be
found [here](https://github.com/OpenSourceBrain/EDENShowcase){.reference
.external}.
:::

[]{#document-Userdocs/Software/Tools/Arbor}

::: {#arbor-and-neuroml .section .tex2jax_ignore .mathjax_ignore}
[]{#userdocs-arbor}

#### Arbor and NeuroML[\#](#arbor-and-neuroml "Link to this heading"){.headerlink}

![Arbor logo](_images/arbor.png)

[Arbor](https://arbor-sim.org/){.reference .external} is a high
performance multicompartmental neural simulation library. Addition of
support for NeuroML2 and LEMS is under active development.

::: {#importing-neuroml-into-arbor .section}
##### Importing NeuroML into Arbor[\#](#importing-neuroml-into-arbor "Link to this heading"){.headerlink}

The current approach to supporting NeuroML in Arbor involves [[importing
NeuroML to Arbor's internal format]{.std
.std-ref}](Landing.html#userdocs-neuroml-support-approaches-native-import){.reference
.internal}.

See
[here](https://docs.arbor-sim.org/en/stable/fileformat/neuroml.html){.reference
.external} for Arbor's own documentation on this. It involves calling
the
[neuroml()](https://docs.arbor-sim.org/en/stable/python/morphology.html#arbor.neuroml){.reference
.external} method in arbor pointing at the NeuroML file containing the
cell you wish to load:

::: {.highlight-python .notranslate}
::: highlight
    nml = arbor.neuroml('mymorphology.cell.nml')
:::
:::

See
[here](https://github.com/OpenSourceBrain/ArborShowcase/blob/main/NeuroML2/test_arbor.py){.reference
.external} for a worked example of this, importing a multicompartmental
cell with only a passive membrane conductance.

::: {#support-for-channels-synapses-in-lems .section}
###### Support for channels/synapses in LEMS[\#](#support-for-channels-synapses-in-lems "Link to this heading"){.headerlink}

There is work under way to allow reading of the dynamics of ion channels
and synapses which are specified in LEMS into Arbor.

See
[thorstenhater/nmlcc](https://github.com/thorstenhater/nmlcc){.github
.reference .external} for more details.
:::
:::

::: {#network-models-in-arbor-with-neuromllite .section}
##### Network models in Arbor with NeuroMLlite[\#](#network-models-in-arbor-with-neuromllite "Link to this heading"){.headerlink}

There is preliminary support for building network specified in
[[NeuroMLlite]{.std .std-ref}](Landing.html#neuromllite){.reference
.internal} format directly in Arbor. See
[here](https://github.com/NeuroML/NeuroMLlite/tree/master/examples/arbor){.reference
.external} for an example.
:::

::: {#examples .section}
##### Examples[\#](#examples "Link to this heading"){.headerlink}

Example code for interactions between NeuroML models and Arbor can be
found in the [Arbor
Showcase](https://github.com/OpenSourceBrain/ArborShowcase){.reference
.external} repository.
:::
:::

[]{#document-Userdocs/Software/Tools/N2A}

::: {#n2a-and-neuroml .section .tex2jax_ignore .mathjax_ignore}
[]{#userdocs-n2a}

#### N2A and NeuroML[\#](#n2a-and-neuroml "Link to this heading"){.headerlink}

"Neurons to Algorithms" (N2A) is a language for modeling neural systems,
along with a software tool for editing models and simulating them

See
[sandialabs/n2a](https://github.com/sandialabs/n2a/wiki/Backend%20LEMS){.github
.reference .external} for information on the interactions between
NeuroML/LEMS and N2A.
:::

[]{#document-Userdocs/Software/Tools/NEST}

::: {#nest-and-neuroml .section .tex2jax_ignore .mathjax_ignore}
[]{#userdocs-nest}

#### NEST and NeuroML[\#](#nest-and-neuroml "Link to this heading"){.headerlink}

![NEST logo](_images/nest-logo.png)

NEST is a simulator for spiking neural network models that focuses on
the dynamics, size and structure of neural systems rather than on the
exact morphology of individual neurons. The development of NEST is
coordinated by the NEST Initiative.

NEST is ideal for networks of spiking neurons of any size, for example:

-   Models of information processing e.g. in the visual or auditory
    cortex of mammals,

-   Models of network activity dynamics, e.g. laminar cortical networks
    or balanced random networks,

-   Models of learning and plasticity.

See
[OpenSourceBrain/NESTShowcase](https://github.com/OpenSourceBrain/NESTShowcase){.github
.reference .external} for examples of usage of NeuroML and NEST.
:::

[]{#document-Userdocs/Software/Tools/SWC}

::: {#swc-and-neuroml .section .tex2jax_ignore .mathjax_ignore}
[]{#userdocs-swc}

#### SWC and NeuroML[\#](#swc-and-neuroml "Link to this heading"){.headerlink}

The SWC format was developed to cover most of the information common
between Neurolucida, NEURON, and GENESIS formats. It is used by
resources such as [NeuroMorpho.org](http://NeuroMorpho.org){.reference
.external}.

Information on the SWC format can be found in the [NeuroMorpho
FAQ](http://neuromorpho.org/myfaq.jsp){.reference .external} under the
"What is SWC format" entry.

Recommended applications for converting SWC into NeuroML are CVApp and
neuroConstruct (see below).

::: {#tools .section}
##### Tools[\#](#tools "Link to this heading"){.headerlink}

A number of tools support conversion of SWC to NeuroML.

::: {#cvapp .section}
[]{#userdocs-cvapp-tools-cvapp}

###### CVApp[\#](#cvapp "Link to this heading"){.headerlink}

[CVApp](https://github.com/NeuroML/Cvapp-NeuroMorpho.org){.reference
.external} is a standalone Java tool that can visualize SWC files (for
example from [NeuroMorpho.org](https://neuromorpho.org){.reference
.external}) and export them into NeuroML2.

<figure id="id1" class="align-center">
<a href="_images/cvapp.png"
class="reference internal image-reference"><img src="_images/cvapp.png"
style="width: 80%;" alt="Screenshot of CVApp" /></a>
<figcaption><p><span class="caption-number">Fig. 66 </span><span
class="caption-text">Screenshot of CVApp</span><a href="#id1"
class="headerlink" title="Link to this image">#</a></p></figcaption>
</figure>

One can select "NeuroMLv2" from the "Save As" drop down box to export
the loaded reconstruction to NeuroML.
:::

::: {#neuroconstruct .section}
[]{#userdocs-cvapp-tools-neuroconstruct}

###### neuroConstruct[\#](#neuroconstruct "Link to this heading"){.headerlink}

[[neuroConstruct]{.std
.std-ref}](Landing.html#userdocs-supporting-apps-neuroconstruct){.reference
.internal} includes functionality to interactively convert CVapp (SWC)
files to NeuroML2. Please see the [neuroConstruct
documentation](http://www.neuroconstruct.org/docs/import.html){.reference
.external} for more information.
:::
:::
:::
:::
:::
:::
:::

[]{#document-Userdocs/Publications}

::: {#citing-neuroml-and-related-publications .section .tex2jax_ignore .mathjax_ignore}
[]{#citing}

## Citing NeuroML and related publications[\#](#citing-neuroml-and-related-publications "Link to this heading"){.headerlink}

This page documents how one can cite NeuroML in their work, and lists
publications associated with the NeuroML initiative.

::: {#citing-neuroml .section}
### Citing NeuroML[\#](#citing-neuroml "Link to this heading"){.headerlink}

Please cite NeuroML in your work whenever you have used it. Generally,
you should cite the particular paper while discussing NeuroML in the
text, and also note and cite the specific version of the NeuroML tool
that has been used in the work.

::: {#papers .section}
#### Papers[\#](#papers "Link to this heading"){.headerlink}

Please cite the following papers as required:

::: {#neuroml-2-and-lems .section}
[]{#papers-neuroml2}

##### NeuroML 2 and LEMS[\#](#neuroml-2-and-lems "Link to this heading"){.headerlink}

::: {.tip .admonition}
The main citation for NeuroML 2

Please cite the following paper when discussing NeuroML v2.0 or LEMS.
:::

**Cannon RC, Gleeson P, Crook S, Ganapathy G, Marin B, Piasini E and
Silver RA (2014)** [LEMS: A language for expressing complex biological
models in concise and hierarchical form and its use in underpinning
NeuroML
2](http://journal.frontiersin.org/Journal/10.3389/fninf.2014.00079/abstract),
*Frontiers in Neuroinformatics* 8: 79.

::: {.highlight-bibtex .notranslate}
::: highlight
    @Article{Cannon2014,
      author    = {Robert C. Cannon and Padraig Gleeson and Sharon Crook and Gautham Ganapathy and Boris Marin and Eugenio Piasini and R. Angus Silver},
      title     = {{LEMS}: a language for expressing complex biological models in concise and hierarchical form and its use in underpinning {NeuroML} 2},
      doi       = {10.3389/fninf.2014.00079},
      volume    = {8},
      journal   = {Frontiers in Neuroinformatics},
      publisher = {Frontiers Media {SA}},
      year      = {2014},
    }
:::
:::
:::

::: {#libneuroml-and-pylems .section}
[]{#papers-libneuroml}

##### libNeuroML and PyLEMS[\#](#libneuroml-and-pylems "Link to this heading"){.headerlink}

::: {.tip .admonition}
Citation for Python & NeuroML

Please cite the following paper when using the Python NeuroML libraries
:::

**Vella M, Cannon RC, Crook S, Davison AP, Ganapathy G, Robinson HP,
Silver RA and Gleeson P (2014)** [libNeuroML and PyLEMS: using Python to
combine procedural and declarative modeling approaches in computational
neuroscience.](http://journal.frontiersin.org/Journal/10.3389/fninf.2014.00038/abstract)
*Frontiers in Neuroinformatics* 8: 38.

::: {.highlight-bibtex .notranslate}
::: highlight
    @Article{Vella2014,
      author       = {Vella, Michael and Cannon, Robert C. and Crook, Sharon and Davison, Andrew P. and Ganapathy, Gautham and Robinson, Hugh P. C. and Silver, R. Angus and Gleeson, Padraig},
      title        = {libNeuroML and PyLEMS: using Python to combine procedural and declarative modeling approaches in computational neuroscience.},
      doi          = {10.3389/fninf.2014.00038},
      pages        = {38},
      volume       = {8},
      journal      = {Frontiers in neuroinformatics},
      year         = {2014},
    }
:::
:::
:::

::: {#neuroml-v1 .section}
[]{#papers-neuroml1}

##### NeuroML v1[\#](#neuroml-v1 "Link to this heading"){.headerlink}

::: {.tip .admonition}
Citation for NeuroML v1

Please cite the following paper when discussing NeuroML version 1.
(deprecated)
:::

**Gleeson, P., S. Crook, R. C. Cannon, M. L. Hines, G. O. Billings, et
al. (2010)** [NeuroML: A Language for Describing Data Driven Models of
Neurons and Networks with a High Degree of Biological
Detail.](http://www.ploscompbiol.org/article/info%3Adoi%2F10.1371%2Fjournal.pcbi.1000815)
*PLoS Computational Biology* 6(6): e1000815.

::: {.highlight-bibtex .notranslate}
::: highlight
    @Article{Gleeson2010,
      author    = {Padraig Gleeson and Sharon Crook and Robert C. Cannon and Michael L. Hines and Guy O. Billings and Matteo Farinella and Thomas M. Morse and Andrew P. Davison and Subhasis Ray and Upinder S. Bhalla and Simon R. Barnes and Yoana D. Dimitrova and R. Angus Silver},
      title     = {{NeuroML}: A Language for Describing Data Driven Models of Neurons and Networks with a High Degree of Biological Detail},
      doi       = {10.1371/journal.pcbi.1000815},
      editor    = {Karl J. Friston},
      number    = {6},
      pages     = {e1000815},
      volume    = {6},
      journal   = {{PLoS} Computational Biology},
      publisher = {Public Library of Science ({PLoS})},
      year      = {2010},
    }
:::
:::
:::

::: {#open-source-brain .section}
[]{#papers-osb}

##### Open Source Brain[\#](#open-source-brain "Link to this heading"){.headerlink}

This paper describes version 1 of the [Open Source
Brain](https://www.opensourcebrain.org){.reference .external} platform.
Please cite this paper if you have made use of OSB in your work:

**Gleeson P, Cantarelli M, Marin B, Quintana A, Earnshaw M, et al.
(2019)** [Open Source Brain: a collaborative resource for visualizing,
analyzing, simulating and developing standardized models of neurons and
circuits.](https://www.cell.com/neuron/fulltext/S0896-6273(19)30444-1)
*Neuron* 103 (3):395--411

::: {.highlight-bibtex .notranslate}
::: highlight
    @Article{Gleeson2019,
      author    = {Padraig Gleeson and Matteo Cantarelli and Boris Marin and Adrian Quintana and Matt Earnshaw and Sadra Sadeh and Eugenio Piasini and Justas Birgiolas and Robert C. Cannon and N. Alex Cayco-Gajic and Sharon Crook and Andrew P. Davison and Salvador Dura-Bernal and Andr{\'{a}}s Ecker and Michael L. Hines and Giovanni Idili and Frederic Lanore and Stephen D. Larson and William W. Lytton and Amitava Majumdar and Robert A. McDougal and Subhashini Sivagnanam and Sergio Solinas and Rokas Stanislovas and Sacha J. van Albada and Werner van Geit and R. Angus Silver},
      title     = {Open Source Brain: A Collaborative Resource for Visualizing, Analyzing, Simulating, and Developing Standardized Models of Neurons and Circuits},
      doi       = {10.1016/j.neuron.2019.05.019},
      number    = {3},
      pages     = {395--411},
      volume    = {103},
      journal   = {Neuron},
      publisher = {Elsevier {BV}},
      year      = {2019},
    }
:::
:::
:::
:::

::: {#software .section}
#### Software[\#](#software "Link to this heading"){.headerlink}

It is important to cite software used in scientific work to:

-   aid reproducibility of work

-   to ensure that the developers of tools receive credit for their
    work.

You can learn more about Software Citation Principles as set out by the
[F1000 Software Citation working
group](https://www.force11.org/group/software-citation-working-group){.reference
.external} in [this work](https://peerj.com/articles/cs-86/){.reference
.external}
[\[[SKNG16](Landing.html#id20 "Arfon M. Smith, Daniel S. Katz, Kyle E. Niemeyer, and FORCE11 Software Citation Working Group. Software citation principles. PeerJ Computer Science, 2:e86, 2016. URL: https://doi.org/10.7717/peerj-cs.86, doi:10.7717/peerj-cs.86."){.reference
.internal}\]]{#id1}.

You can obtain the version of [[pyNeuroML]{.std
.std-ref}](Landing.html#pyneuroml){.reference .internal} and associated
tools using the following command (with example output):

::: {.highlight-console .notranslate}
::: highlight
    $ pynml -version
    pyNeuroML v0.5.20 (libNeuroML v0.3.1, jNeuroML v0.11.1)
:::
:::

Each NeuroML software tool has a unique DOI and reference associated
with each release at the [Zenodo archival
facility](https://zenodo.org/search?page=1&amp;size=20&amp;q=neuroml){.reference
.external}. On each entry, you will be able to find the DOI and citation
of the particular version you are using, and you will also be able to
download the citation in different formats at the bottom of the right
hand side bar.
:::
:::

::: {#other-publications .section}
### Other publications[\#](#other-publications "Link to this heading"){.headerlink}

This section lists other publications related to NeuroML.

**Günay, C. et al. (2008)** [Computational intelligence in
electrophysiology: Trends and open
problems](http://springerlink.com/content/2177143636k14816/). In
Smolinski, Milanova and Hassanien, eds. *Applications of Computational
Intelligence in Biology*. Springer, Berlin/Heidelberg.

**Gleeson, P., V. Steuber, and R. A. Silver (2007)** [neuroConstruct: A
Tool for Modeling Networks of Neurons in 3D
Space](http://www.ncbi.nlm.nih.gov/pubmed/17442244). *Neuron.*
54(2):219-235.

**Cannon R. C., M. O. Gewaltig, P. Gleeson, U. S. Bhalla, H. Cornelis,
M. L. Hines, F. W. Howell, E. Muller, J. R. Stiles, S. Wils, E. De
Schutter (2007)** [Interoperability of Neuroscience Modeling Software:
Current Status and Future
Directions.](http://www.ncbi.nlm.nih.gov/pubmed/17873374)
*Neuroinformatics* Volume 5, 127-138.

**Crook, S., P. Gleeson, F. Howell, J. Svitak and R.A. Silver (2007)**
[MorphML: Level 1 of the NeuroML standards for neuronal morphology data
and model specification.](http://www.ncbi.nlm.nih.gov/pubmed/17873371)
*Neuroinformatics.* 5(2):96-104.

**Crook, S. and F. Howell (2007)** [XML for data representation and
model specification in
neuroscience.](http://www.amazon.com/Neuroinformatics-Methods-Molecular-Biology-Koslow/dp/1588297209)
In *Methods in Molecular Biology Book Series: Neuroinformatics*. ed. C.
Crasto, Humana Press.

**Crook, S., D. Beeman, P. Gleeson and F. Howell (2005)** [XML for model
specification in neuroscience: An introduction and workshop
summary.](http://www.brains-minds-media.org/archive/228) *Brains, Minds,
and Media.* 1:bmm228 (urn:nbn:de:0009-3-2282).

**Qi, W. and S. Crook (2004)** [Tools for neuroinformatic data exchange:
An XML application for neuronal morphology
data.](https://www.sciencedirect.com/science/article/pii/S0925231204001766)
*Neurocomputing.* 58-60C:1091-1095.

**Goddard, N., M. Hucka, F. Howell, H. Cornelis, K. Shankar and D.
Beeman (2001)** [Towards NeuroML: Model description methods for
collaborative modeling in
neuroscience](http://www.pubmedcentral.nih.gov/articlerender.fcgi?artid=1088511).
*Philosophical Transactions of the Royal Society B*. 356:1209-1228.

::: {#book-chapters .section}
#### Book Chapters[\#](#book-chapters "Link to this heading"){.headerlink}

**Crook, SM, HE Plesser, AP Davison (2013)** Lessons from the past:
approaches for reproducibility in computational neuroscience. In *JM
Bower*, ed. 20 Years of Computational Neuroscience, Springer

**Gleeson, P, V Steuber, RA Silver and S Crook (2012)** NeuroML. In *Le
Novere*, ed. Computational Systems Biology, Springer.
:::

::: {#abstracts .section}
#### Abstracts[\#](#abstracts "Link to this heading"){.headerlink}

**Cannon, R, P Gleeson, S Crook, RA Silver (2012)** [A declarative model
specification system allowing NeuroML to be extended with user-defined
component types](http://www.biomedcentral.com/1471-2202/13/S1/P42). *BMC
Neuroscience*. 13(Suppl 1): P42.

**Gleeson P, S Crook, A Silver, R Cannon (2011)** [Development of
NeuroML version 2.0: Greater extensibility, support for abstract
neuronal models and interaction with Systems Biology
languages](http://www.biomedcentral.com/1471-2202/12/S1/P29). *BMC
Neuroscience*. 12:P29.

**Gleeson, P., S. Crook, S. Barnes and R.A. Silver (2008)**
[Interoperable model components for biologically realistic single neuron
and network models implemented in
NeuroML.](http://frontiersin.org/conferences/individual_abstract_listing.php?conferid=2&pap=491&ind_abs=1&pg=5)
*Frontiers in Neuroinformatics. Conference Abstract: Neuroinformatics
2008.* doi: 10.3389/conf.neuro.11.2008.01.135.

**Larson, S. and M. Martone (2008)** [A spatial framework for
multi-scale computational
neuroanatomy.](http://frontiersin.org/conferences/individual_abstract_listing.php?conferid=2&pap=490&ind_abs=1)
*Frontiers in Neuroinformatics. Conference Abstract: Neuroinformatics
2008.* doi: 10.3389/conf.neuro.11.2008.01.134.

**Crook, S., P. Gleeson and R.A. Silver (2007)** NetworkML: Level 3 of
the NeuroML standards for multiscale model specification and exchange.
*Society for Neuroscience Abstracts.* 102.28.

**Gleeson, P., S. Crook, V. Steuber and R.A. Silver (2007)** [Using
NeuroML and neuroConstruct to build neuronal network models for multiple
simulators](http://www.biomedcentral.com/1471-2202/8/S2/P1). *BMC
Neuroscience.* 8(2):P101.
:::
:::
:::

[]{#document-Userdocs/FAQ}

::: {#frequently-asked-questions-faq .section .tex2jax_ignore .mathjax_ignore}
[]{#userdocs-faq}

## Frequently asked questions (FAQ)[\#](#frequently-asked-questions-faq "Link to this heading"){.headerlink}

::: {.note .admonition}
Please help improve the FAQ.

This page lists some commonly asked questions related to NeuroML. Please
[open
issues](https://github.com/NeuroML/Documentation/issues){.reference
.external} to add more entries to this FAQ.
:::

::: {#are-length-0-segments-allowed-in-neuroml .section}
[]{#userdocs-faq-zero-length-segments}

### 1. Are length 0 segments allowed in NeuroML?[\#](#are-length-0-segments-allowed-in-neuroml "Link to this heading"){.headerlink}

Discussion link:
[NeuroML/NeuroML2#115](https://github.com/NeuroML/NeuroML2/issues/115){.github
.reference .external}

There are a lot of SWC reconstructions which have adjacent points, which
would get converted to zero length segments. This shouldn't be an issue
for most visualisation applications, so no need for them to say that
they can't visualise the cell if they see it's invalid.

The [`jnml`{.docutils .literal .notranslate}]{.pre}` `{.docutils
.literal .notranslate}[`-validate`{.docutils .literal
.notranslate}]{.pre} option could throw a warning when it sees these
segments, but currently doesn't (it could be added
[here](https://github.com/NeuroML/org.neuroml.model/blob/development/src/main/java/org/neuroml/model/util/NeuroML2Validator.java#L199){.reference
.external}).

For individual simulators, they could have an issue with this, if they
map each segment to a compartment (as Moose might), but for Neuron using
cables/sections with multiple segments, it shouldn't matter as long as
the section doesn't just have one segment.

So ideally it should be the application which loads the NeuroML in (or
the conversion/export code) which decides whether this is an issue.
:::

::: {#what-is-the-difference-between-reader-writer-methods-in-pyneuroml-and-libneuroml .section}
[]{#userdocs-faq-pyneuroml-or-neuroml}

### 2. What is the difference between reader/writer methods in pyNeuroML and libNeuroML?[\#](#what-is-the-difference-between-reader-writer-methods-in-pyneuroml-and-libneuroml "Link to this heading"){.headerlink}

Both [[libNeuroML]{.std .std-ref}](Landing.html#libneuroml){.reference
.internal} and [[pyNeuroML]{.std
.std-ref}](Landing.html#pyneuroml){.reference .internal} include methods
that can read and write NeuroML files. However, they are not the same.

libNeuroML is the low level Python API for working with NeuroML. The
loaders/writers included here can therefore read/write NeuroML files.
However, these are "low level" functions and do not include additional
features.

The readers/writers in pyNeuroML use these low level functions from
libNeuroML but also run other checks and include other features.

So,

-   [pyneuroml.io.read_neuroml2_file](https://pyneuroml.readthedocs.io/en/latest/pyneuroml.io.html#pyneuroml.io.read_neuroml2_file){.reference
    .external} should be preferred over
    [neuroml.loaders.read_neuroml2_file](https://libneuroml.readthedocs.io/en/latest/userdocs/loaders.html#neuroml.loaders.read_neuroml2_file){.reference
    .external}: it also allows pre-loading validation checks, and it
    also handles morphologies referenced in other files.

-   [pyneuroml.io.write_neuroml2_file](https://pyneuroml.readthedocs.io/en/latest/pyneuroml.io.html#pyneuroml.io.write_neuroml2_file){.reference
    .external} should be preferred over
    [neuroml.writers.NeuroMLWriter.write](https://libneuroml.readthedocs.io/en/latest/userdocs/writers.html#neuroml.writers.NeuroMLWriter.write){.reference
    .external}: it also validates the file after writing it.
:::
:::

[]{#document-Userdocs/Walkthroughs/Walkthroughs}

::: {#walk-throughs .section .tex2jax_ignore .mathjax_ignore}
[]{#userdocs-walkthroughs}

## Walk throughs[\#](#walk-throughs "Link to this heading"){.headerlink}

This chapter documents a number of real-world tasks for users to refer
to.

::: {.toctree-wrapper .compound}
[]{#document-Userdocs/Walkthroughs/RayEtAl2020/RayEtAl2020}

::: {#converting-ray-et-al-2020-to-neuroml .section .tex2jax_ignore .mathjax_ignore}
[]{#userdocs-walkthroughs-rayetal2020}

### Converting Ray et al 2020 to NeuroML[\#](#converting-ray-et-al-2020-to-neuroml "Link to this heading"){.headerlink}

This section documents the conversion of Ray et al 2020
[\[[RAS20](Landing.html#id21 "Subhasis Ray, Zane N. Aldworth, and Mark A. Stopfer. Feedback inhibition and its control in an insect olfactory circuit. eLife, 9:e53281, 2020. URL: https://doi.org/10.7554/eLife.53281, doi:10.7554/eLife.53281."){.reference
.internal}\]]{#id1}, which was originally implemented in NEURON, to
NeuroML. It broadly follows the steps outlined in the [[converting
models]{.std
.std-ref}](Landing.html#userdocs-creating-models-converting-conductance){.reference
.internal} section.

For any queries, please contact Ankur Sinha on any of the NeuroML
channels.

::: {.toctree-wrapper .compound}
[]{#document-Userdocs/Walkthroughs/RayEtAl2020/Setup}

::: {#setting-up .section .tex2jax_ignore .mathjax_ignore}
#### Setting up[\#](#setting-up "Link to this heading"){.headerlink}

::: {#step-1-find-the-original-model-code .section}
##### Step 1) Find the original model code[\#](#step-1-find-the-original-model-code "Link to this heading"){.headerlink}

The original code is published on
[ModelDB](https://modeldb.science/262670){.reference .external}.
:::

::: {#step-2-create-github-and-open-source-brain-accounts-for-sharing-the-code .section}
##### Step 2) Create GitHub and Open Source Brain accounts for sharing the code[\#](#step-2-create-github-and-open-source-brain-accounts-for-sharing-the-code "Link to this heading"){.headerlink}

::: {#a-sign-up-to-github-and-open-source-brain .section}
###### 2a) Sign up to GitHub and Open Source Brain[\#](#a-sign-up-to-github-and-open-source-brain "Link to this heading"){.headerlink}

We signed in to GitHub and OSBv1
:::

::: {#b-create-github-repository .section}
###### 2b) Create GitHub repository[\#](#b-create-github-repository "Link to this heading"){.headerlink}

ModelDB provides GitHub repositories for all its models now. This model
is available on GitHub here:
[ModelDBRepository/262670](https://github.com/ModelDBRepository/262670){.github
.reference .external}. The Open Source Brain (OSB) organization on
GitHub also keeps a "fork" of these repositories to allow users to
easily add them to both Open Source Brain v1 and v2. This fork is here,
and is the one that we will work with:
[OpenSourceBrain/262670](https://github.com/OpenSourceBrain/262670){.github
.reference .external}.

For the conversion, I (Ankur) created a fork of this repository with a
new branch to work in:
[sanjayankur31/262670](https://github.com/sanjayankur31/262670){.github
.reference .external}. A pull request work flow was used to submit
converted bits back to the repository.

The first step was to re-organise the code to prepare it for conversion.
All the existing code was moved to a new NEURON folder, and a new
NeuroML2 folder set up to store the NeuroML version.
:::

::: {#c-create-open-source-brain-project .section}
###### 2c) Create Open Source Brain project[\#](#c-create-open-source-brain-project "Link to this heading"){.headerlink}

A new project was created on OSBv1 and linked to the OSB repository:
[https://v1.opensourcebrain.org/projects/locust-mushroom-body](https://v1.opensourcebrain.org/projects/locust-mushroom-body){.reference
.external}.
:::
:::
:::

[]{#document-Userdocs/Walkthroughs/RayEtAl2020/Conversion}

::: {#converting-to-neuroml .section .tex2jax_ignore .mathjax_ignore}
#### Converting to NeuroML[\#](#converting-to-neuroml "Link to this heading"){.headerlink}

On inspection of the model, we see that it has two biophysically
detailed cell models:

-   the GGN (Giant GABAergic Neuron)

-   the KC (Kenyon cell)

::: {#converting-the-giant-gabaergic-neuron .section}
##### Converting the Giant GABAergic Neuron[\#](#converting-the-giant-gabaergic-neuron "Link to this heading"){.headerlink}

::: {#step-1-exporting-morphology-of-the-ggn .section}
###### Step 1) Exporting morphology of the GGN[\#](#step-1-exporting-morphology-of-the-ggn "Link to this heading"){.headerlink}

Let's start with the GGN first. It's morphology is defined as an SWC
file in [this
file](https://github.com/OpenSourceBrain/262670/blob/master/NEURON/mb/cell_templates/GGN_20170309_sc.swc){.reference
.external}. One can download this file and view the morphology in a
tool, like the [HBP morphology
viewer](https://neuroinformatics.nl/HBP/morphology-viewer/#){.reference
.external}.

<figure id="id1" class="align-center">
<a href="_images/GGN.png"
class="reference internal image-reference"><img src="_images/GGN.png"
style="width: 384.0px; height: 199.60000000000002px;"
alt="Screen shot showing visualisation of the GGN in the HBP morphology viewer" /></a>
<figcaption><p><span class="caption-number">Fig. 67 </span><span
class="caption-text">Visualisation of the GGN in the HBP morphology
viewer.</span><a href="#id1" class="headerlink"
title="Link to this image">#</a></p></figcaption>
</figure>

A [NEURON HOC
script](https://github.com/OpenSourceBrain/262670/blob/master/NEURON/mb/cell_templates/GGN_20170309_sc.hoc){.reference
.external} that includes the full morphology and the biophysics is also
included.

Let us export the morphology first. pyNeuroML includes the
[`export_to_neuroml2`{.docutils .literal .notranslate}]{.pre} helper
function that exports a cell model in NEURON to NeuroML. We can write a
short script to use this function to export the morphology from the
provided HOC script.

::: {.highlight-python .notranslate}
::: highlight
    #!/usr/bin/env python3
    """
    Convert cell morphology to NeuroML.

    We only export morphologies here. We add the biophysics manually.

    File: NeuroML2/scripts/cell2nml.py
    """

    import os
    import sys

    import pyneuroml
    from pyneuroml.neuron import export_to_neuroml2
    from neuron import h


    def main(acell):
        """Main runner method.

        :param acell: name of cell
        :returns: None

        """
        loader_hoc_file = f"{acell}_loader.hoc"
        loader_hoc_file_txt = """
        /*load_file("nrngui.hoc")*/
        load_file("stdrun.hoc")
        xopen("../../NEURON/mb/cell_templates/GGN_20170309_sc.hoc")
        objref cell
        cell = new GGN_20170309_sc()
        """

        with open(loader_hoc_file, 'w') as f:
            print(loader_hoc_file_txt, file=f)

        export_to_neuroml2(loader_hoc_file, f"{acell}.morph.cell.nml",
                           includeBiophysicalProperties=False, validate=False)

        os.remove(loader_hoc_file)
        # Note--a couple of diameters are 0.0, modified to 0.001 to validate the
        # model


    if __name__ == "__main__":
        if len(sys.argv) != 2:
            print("This script only accepts one argument.")
            sys.exit(1)
        main(sys.argv[1])
:::
:::

What we're doing here is using the HOC script to build the cell model in
NEURON, and then exporting it to NeuroML. Calling it as
[`python`{.docutils .literal .notranslate}]{.pre}` `{.docutils .literal
.notranslate}[`cellmorph2nml.py`{.docutils .literal
.notranslate}]{.pre}` `{.docutils .literal .notranslate}[`GGN`{.docutils
.literal .notranslate}]{.pre} will create a new file:
[`GGN.morph.cell.nml`{.docutils .literal .notranslate}]{.pre} which
contains the morphology of the cell in NeuroML format. Note that while
[`export_to_neuroml2`{.docutils .literal .notranslate}]{.pre} does allow
exporting the biophysics of the cell, it is better to add these manually
later once one has gone through and converted the required ion channels
and so on.

We can visualise the morphology using the pyNeuroML tools:

::: {.highlight-bash .notranslate}
::: highlight
    pynml-plotmorph -i GGN.morph.cell.nml
:::
:::

<figure id="id2" class="align-center">
<a href="_images/GGN-vispy.png"
class="reference internal image-reference"><img
src="_images/GGN-vispy.png" style="width: 384.0px; height: 216.0px;"
alt="Screen shot showing visualisation of the GGN using `pynml-plotmorph`" /></a>
<figcaption><p><span class="caption-number">Fig. 68 </span><span
class="caption-text">Visualisation of the GGN using <span
class="pre"><code
class="docutils literal notranslate">pynml-plotmorph</code></span></span><a
href="#id2" class="headerlink"
title="Link to this image">#</a></p></figcaption>
</figure>
:::

::: {#step-2-adding-biophysics-to-the-ggn .section}
###### Step 2) Adding biophysics to the GGN[\#](#step-2-adding-biophysics-to-the-ggn "Link to this heading"){.headerlink}

Now that we have the morphology of the GGN exported, we can add the
biophysics. We need to inspect the original model code to learn about
the biophysics. In this model, for the GGN cell, the biophysics are
included in the HOC script:

::: {.highlight-default .notranslate}
::: highlight
    proc biophys() {
      forsec all {
        Ra = 100.0
        cm = 1
        insert pas
          g_pas = 0.03e-3   // S/cm2 - as per Laurent et al 1990 RM = 33kohm-cm2
          e_pas = -51
      }
    }
:::
:::

As we see here, this is a passive cell without any ion channels. To add
the biophysics, we write a simple Python script that will make use of
the pyNeuroML API. The complete script is present in the
[repository](https://github.com/OpenSourceBrain/262670/blob/master/NeuroML2/postprocess_cells.py){.reference
.external}:

::: {.highlight-python .notranslate}
::: highlight
    def load_and_setup_cell(cellname: str):
        """Load a cell, and clean it to prepare it for further modifications.

        These operations are common for all cells.

        :param cellname: name of cell.
            the file containing the cell should then be <cell>.morph.cell.nml
        :returns: document with cell
        :rtype: neuroml.NeuroMLDocument

        """
        celldoc = read_neuroml2_file(
            f"{cellname}.morph.cell.nml"
        )  # type: neuroml.NeuroMLDocument
        cell = celldoc.cells[0]  # type: neuroml.Cell
        celldoc.networks = []
        cell.id = cellname
        cell.notes = cell.notes.replace("GGN_20170309_sc_0_0", cellname)
        cell.notes += ". Reference: Subhasis Ray, Zane N Aldworth, Mark A Stopfer (2020) Feedback inhibition and its control in an insect olfactory circuit eLife 9:e53281."

        [
            default_all_group,
            default_soma_group,
            default_dendrite_group,
            default_axon_group,
        ] = cell.setup_default_segment_groups(
            use_convention=True,
            default_groups=["all", "soma_group", "dendrite_group", "axon_group"],
        )

        # populate default groups
        for sg in cell.morphology.segment_groups:
            if "soma" in sg.id and sg.id != "soma_group":
                default_soma_group.add(neuroml.Include(segment_groups=sg.id))
            if "axon" in sg.id and sg.id != "axon_group":
                default_axon_group.add(neuroml.Include(segment_groups=sg.id))
            if "dend" in sg.id and sg.id != "dendrite_group":
                default_dendrite_group.add(neuroml.Include(segment_groups=sg.id))

        cell.optimise_segment_groups()

        return celldoc


    def postprocess_GGN():
        """Post process GGN and add biophysics."""
        cellname = "GGN"
        celldoc = load_and_setup_cell(cellname)
        cell = celldoc.cells[0]  # type: neuroml.Cell

        # biophysics
        # all
        cell.add_channel_density(
            nml_cell_doc=celldoc,
            cd_id="pas",
            ion_channel="pas",
            cond_density="0.00003 S_per_cm2",
            erev="-51 mV",
            group_id="all",
            ion="non_specific",
            ion_chan_def_file="channels/pas.channel.nml",
        )
        cell.set_resistivity("0.1 kohm_cm", group_id="all")
        cell.set_specific_capacitance("1 uF_per_cm2", group_id="all")
        cell.set_init_memb_potential("-80mV")

        # L1 validation
        # cell.validate(recursive=True)
        cell.summary(morph=False, biophys=True)
        # use pynml writer to also run L2 validation
        write_neuroml2_file(celldoc, f"{cellname}.cell.nml")
:::
:::

The [`load_and_setup_cell`{.docutils .literal .notranslate}]{.pre}
function does some basic clean up and set up of the cell. It ensures
that the various segments that were exported from NEURON are placed into
the conventional segment groups. The [`postprocess_GGN`{.docutils
.literal .notranslate}]{.pre} function then adds the passive biophysics
to the cell.

The [`pas`{.docutils .literal .notranslate}]{.pre} channel is a standard
implementation of a passive ion channel. The rest are membrane
properties--resistivity, specific capacitance and so on. Once this is
set up, we write the cell to a new file.

The GGN cell has now been converted. Since the GGN is a simple passive
cell, we won't test its biophysics just yet.
:::
:::

::: {#converting-the-kenyon-cell .section}
##### Converting the Kenyon Cell[\#](#converting-the-kenyon-cell "Link to this heading"){.headerlink}

The KC cell is defined in the [`kc_1_comp.hoc`{.docutils .literal
.notranslate}]{.pre} file. Whereas the GGN cell had a complex morphology
but passive biophysics, the KC cell has very simple morphology---a
single compartment---but does contain active channels:

::: {.highlight-default .notranslate}
::: highlight
    create soma


    objref all
    proc subsets() { local i
      objref all
      all = new SectionList()
      soma all.append()
    }
    proc geom() {
         soma {  // Total Cm = 4 pF
            L = 6.366
            diam = 20
        }
    }

    proc biophys() {
      forsec all {
        Ra = 35.4
        cm = 1
        insert pas
          g_pas = 9.75e-5         // S/cm2
          e_pas = -70             // mV
        insert kv
          gbar_kv = 1.5e-3       // S/cm2
        insert ka
          gbar_ka = 1.4525e-2    // S/cm2
        insert kst
          gbar_kst = 2.0275e-3   // S/cm2
        insert naf
          gbar_naf = 3.5e-2      // S/cm2
        insert nas
          gbar_nas = 3e-3        // S/cm2
          ek = -81.0             // mV
          ena = 58.0               // mV
      }
    }
:::
:::

::: {#step-1-converting-ion-channels .section}
###### Step 1) Converting ion channels[\#](#step-1-converting-ion-channels "Link to this heading"){.headerlink}

For this cell, we will first convert the various ion channel models.
These are included in the [`mod`{.docutils .literal .notranslate}]{.pre}
folder. An inspection tells us that these are all Hodgkin-Huxley type
ion channels that use similar formalisms.

The first thing to do is to generate plots of time courses and steady
states of the various ion channels. These can be done easily using
[`pynml-modchananalysis`{.docutils .literal .notranslate}]{.pre} command
line tool included in pyNeuroML. We begin with the [`nas`{.docutils
.literal .notranslate}]{.pre} channel:

::: {.highlight-default .notranslate}
::: highlight
    pynml-modchananalysis -modFile nas_wustenberg.mod nas
:::
:::

This will generate two plots, one for the steady state dynamics
([`inf`{.docutils .literal .notranslate}]{.pre}), and one for the time
course ([`tau`{.docutils .literal .notranslate}]{.pre}) for activation
variables in the channels:

<figure id="id3" class="align-center">
<a
href="_images/Time_course(s)_of_activation_variables_in_nas_at_6.3_degC.png"
class="reference internal image-reference"><img
src="_images/Time_course%28s%29_of_activation_variables_in_nas_at_6.3_degC.png"
style="width: 768.0px; height: 388.8px;"
alt="Image showing time course of activation variables of nas channel, generated with pynml-modchananalysis" /></a>
<figcaption><p><span class="caption-number">Fig. 69 </span><span
class="caption-text">Time course of activation variables of nas channel,
generated with pynml-modchananalysis</span><a href="#id3"
class="headerlink" title="Link to this image">#</a></p></figcaption>
</figure>

<figure id="id4" class="align-center">
<a
href="_images/Steady_state(s)_of_activation_variables_in_nas_at_6.3_degC.png"
class="reference internal image-reference"><img
src="_images/Steady_state%28s%29_of_activation_variables_in_nas_at_6.3_degC.png"
style="width: 768.0px; height: 388.8px;"
alt="Image showing steady state dynamics of activation variables of nas channel, generated with pynml-modchananalysis" /></a>
<figcaption><p><span class="caption-number">Fig. 70 </span><span
class="caption-text">Steady state dynamics of activation variables of
nas channel, generated with pynml-modchananalysis</span><a href="#id4"
class="headerlink" title="Link to this image">#</a></p></figcaption>
</figure>

The mod file defining the nas channel is shown below:

::: {.highlight-default .notranslate}
::: highlight
    : nas_wustenberg.mod --- 
    : 
    : Filename: nas_wustenberg.mod
    : Description: 
    : Author: Subhasis Ray
    : Maintainer: 
    : Created: Wed Dec 13 19:06:03 EST 2017
    : Version: 
    : Last-Updated: Mon Jun 18 14:38:15 2018 (-0400)
    :           By: Subhasis Ray
    : URL: 
    : Doc URL: 
    : Keywords: 
    : Compatibility: 

    : Commentary: 
    : 
    : NEURON implementation of slow Na+ channel ( NAS ) from Wustenberg
    : DG, Boytcheva M, Grunewald B, Byrne JH, Menzel R, Baxter DA

    : This is slow Na+ channel in Apis mellifera Kenyon cells :(cultured).

    TITLE Slow NA+ current in honey bee KC from Wustenberg et al 2004

    COMMENT
      NEURON implementation by Subhasis Ray (ray dot subhasis at gmail dot com).

    ENDCOMMENT

    INDEPENDENT { t FROM 0 TO 1 WITH 1 (ms) }

    NEURON { 
            SUFFIX nas
            USEION na READ ena WRITE ina
            RANGE gbar, ina, g
    }

    UNITS {
            (S) = (siemens)
            (mV) = (millivolt) 
            (mA) = (milliamp) 
    }
     
    PARAMETER { 
            gbar = 0.0      (mho/cm2)
    }
     
    ASSIGNED { 
        ena  (mV)
            v    (mV)
            ina  (mA/cm2)
            g    (S/cm2)
            minf
        hinf
            mtau (ms)
            htau (ms)
    }
     
    STATE {
        m
        h
    }

    BREAKPOINT { 
            SOLVE states METHOD cnexp 
            g = gbar * m * m * m * h
            ina = g * ( v - ena )
    }
     
    INITIAL { 
            settables(v)
        m = minf
            h  = hinf
    } 

    DERIVATIVE states { 
            settables(v) 
            h' = (hinf - h) / htau
        m' = (minf - m ) / mtau
    }

    : Parameters from the article (Table 2):
    :
    :       E,mV    g,nS            taumax,ms       taumin,ms       Vh1,mV  s1      Vh2,mV  s2      N
    : INa                     minf                                  -30.1   6.65                    3
    :                         taum  0.83            0.093           -20.3   6.45
    : INaF  58      140       hinf                                  -51.4   5.9                     1
    :                         tauh  1.66            0.12            -8.03   8.69
    :
    : INaS  58      12        hinf                                  -51.4   5.9                     1
    :                         tauh  12.24           1.9             -32.6   8
    : The equations are:
    :       minf = 1 / ( 1 + exp((Vh - V) / s))
    :       hinf = 1 / ( 1 + exp((V - Vh) / s))
    :       taum = (taumax - taumin) / (1 + exp((V - Vh1) / s1)) + taumin

    PROCEDURE settables(v (mV)) { 
    UNITSOFF
            TABLE minf, hinf, mtau, htau FROM -120 TO 40 WITH 641
            minf  = 1.0 / (1 + exp((-30.1 - v) / 6.65))
            hinf  = 1.0 / (1 + exp((v + 51.4) / 5.9 ))
            mtau = (0.83 - 0.093) / (1 + exp((v + 20.3) / 6.45)) + 0.093
            htau = (12.24 - 1.9) / (1 + exp((v + 32.6) / 8.0)) + 1.9
    UNITSON
    }

    : nas_wustenberg.mod ends here
:::
:::

Here we have the [`m`{.docutils .literal .notranslate}]{.pre} and
[`h`{.docutils .literal .notranslate}]{.pre} activation variables.
Although they are described in the standard Hodgkin Huxley formalism, in
the [`settables`{.docutils .literal .notranslate}]{.pre} procedure, we
can see that their values are only calculated in the range of -120mV to
40mV. The value does not change beyond 40mV. This could be done for a
number of reasons. Perhaps the cell's membrane potential does not go
beyond 40mV.

We will attempt to remain faithful to the mod file in our conversion, so
we will also incorporate this feature.

Since we know this is a Hodgkin Huxley type channel, we can search the
schema to see if there are any elements that can describe it. A search
shows us that the
[ionChannelHH](https://docs.neuroml.org/Userdocs/Schemas/Channels.html#ionchannelhh){.reference
.external} component type exists in the schema/standard. In the schema,
this is identical to
[ionChannel](https://docs.neuroml.org/Userdocs/Schemas/Channels.html#ionchannel){.reference
.external}. The usage examples included in the documentation indicate
that we can use these elements from the standard to describe the ion
channel here. We need to:

-   include the [`m`{.docutils .literal .notranslate}]{.pre} gate, which
    has 3 sub units ([`m^3`{.docutils .literal .notranslate}]{.pre})

-   include the [`h`{.docutils .literal .notranslate}]{.pre} gate, which
    has 1 sub unit ([`h^1`{.docutils .literal .notranslate}]{.pre})

-   formalise the equations that are used to calculate the steady state
    ([`inf`{.docutils .literal .notranslate}]{.pre})and time course
    ([`tau`{.docutils .literal .notranslate}]{.pre}) for these
    gates/activation variables.

To begin with, let us ignore the restriction included in the mod file at
40mV. Our NeuroML description will look something like this:

::: {.highlight-xml .notranslate}
::: highlight
    <ionChannel id="nas" type="ionChannelHH" conductance="1pS" species="na">

    <!-- some more details to be added here -->

    </ionChannelHH>
:::
:::

Now, there are number of different ways of expressing the dynamics of
the activation variables. (See [this page for an introduction to HH
formalism](https://www.st-andrews.ac.uk/~wjh/hh_model_intro/){.reference
.external}). One can use the values of the forward and reverse rates
(called [`alpha`{.docutils .literal .notranslate}]{.pre} and
[`beta`{.docutils .literal .notranslate}]{.pre} in general) to calculate
the steady state and time course. Another possibility is that
[`alpha`{.docutils .literal .notranslate}]{.pre} and [`beta`{.docutils
.literal .notranslate}]{.pre} are not given, and instead the equations
for the steady state and time course are. The latter is the case here:

::: {.highlight-default .notranslate}
::: highlight
    inf = 1 / ( 1 + exp((Vh - V) / s) )
:::
:::

If the rates are given, one can use the
[gateHHrates](https://docs.neuroml.org/Userdocs/Schemas/Channels.html#gatehhrates){.reference
.external} component type from the standard. If the steady state and
time course are, we can use
[gateHHtauInf](https://docs.neuroml.org/Userdocs/Schemas/Channels.html#gatehhtauinf){.reference
.external}. There are also other components that can be used if a
combination of rates and/or steady state and time course are given.

The equation above is in the form of a [sigmoid
function](https://en.wikipedia.org/wiki/Sigmoid_function){.reference
.external}. Another search in the standard shows us that we have the
[HHSigmoidVariable](https://docs.neuroml.org/Userdocs/Schemas/Channels.html#hhsigmoidvariable){.reference
.external} component type that can be used to represent a rate that is a
sigmoid function. The [`dynamics`{.docutils .literal
.notranslate}]{.pre} tab tells us that the equation is represented as:

::: {.highlight-default .notranslate}
::: highlight
    x = rate / (1 + exp(0 - (v - midpoint)/scale))
:::
:::

Putting the equation for [`minf`{.docutils .literal .notranslate}]{.pre}
and [`hinf`{.docutils .literal .notranslate}]{.pre} together with this
form:

::: {.highlight-default .notranslate}
::: highlight
    x = rate / (1 + exp(0 - (v - midpoint)/scale))
    minf  = 1.0 / (1 + exp((-30.1 - v) / 6.65))
    hinf  = 1.0 / (1 + exp((v + 51.4) / 5.9 ))
:::
:::

We can see that for [`minf`{.docutils .literal .notranslate}]{.pre},
[`rate`{.docutils .literal .notranslate}]{.pre}` `{.docutils .literal
.notranslate}[`=`{.docutils .literal .notranslate}]{.pre}` `{.docutils
.literal .notranslate}[`1`{.docutils .literal
.notranslate}]{.pre}` `{.docutils .literal .notranslate}[`per`{.docutils
.literal .notranslate}]{.pre}` `{.docutils .literal
.notranslate}[`ms`{.docutils .literal .notranslate}]{.pre},
[`scale`{.docutils .literal .notranslate}]{.pre}` `{.docutils .literal
.notranslate}[`=`{.docutils .literal .notranslate}]{.pre}` `{.docutils
.literal .notranslate}[`6.65mV`{.docutils .literal .notranslate}]{.pre}
and [`midpoint`{.docutils .literal .notranslate}]{.pre}` `{.docutils
.literal .notranslate}[`=`{.docutils .literal
.notranslate}]{.pre}` `{.docutils .literal
.notranslate}[`-30.1mV`{.docutils .literal .notranslate}]{.pre}.
Similarly, for [`hinf`{.docutils .literal .notranslate}]{.pre},
[`rate`{.docutils .literal .notranslate}]{.pre}` `{.docutils .literal
.notranslate}[`=`{.docutils .literal .notranslate}]{.pre}` `{.docutils
.literal .notranslate}[`1`{.docutils .literal
.notranslate}]{.pre}` `{.docutils .literal .notranslate}[`per`{.docutils
.literal .notranslate}]{.pre}` `{.docutils .literal
.notranslate}[`ms`{.docutils .literal .notranslate}]{.pre},
[`scale`{.docutils .literal .notranslate}]{.pre}` `{.docutils .literal
.notranslate}[`=`{.docutils .literal .notranslate}]{.pre}` `{.docutils
.literal .notranslate}[`-5.9mV`{.docutils .literal .notranslate}]{.pre}
and [`midpoint`{.docutils .literal .notranslate}]{.pre}` `{.docutils
.literal .notranslate}[`=`{.docutils .literal
.notranslate}]{.pre}` `{.docutils .literal
.notranslate}[`-51.4mV`{.docutils .literal .notranslate}]{.pre}.

::: {.highlight-xml .notranslate}
::: highlight
    <ionChannel id="nas" type="ionChannelHH" conductance="1pS" species="na">

        <gate id="m" instances="3" type="gateHHtauInf">
            <steadyState type="HHSigmoidVariable" midpoint="-30.1mV" scale="6.65mV" rate="1" />
            <timeCourse />
        </gate>

        <gate id="h" instances="1" type="gateHHtauInf">
            <steadyState type="HHSigmoidVariable" midpoint="-51.4mV" scale="-5.9mV" rate="1" />
            <timeCourse />
        </gate>
    </ionChannel>
:::
:::

The time course is given by:

::: {.highlight-default .notranslate}
::: highlight
    taum = (taumax - taumin) / (1 + exp((V - Vh1) / s1)) + taumin
:::
:::

Even though this is also a sigmoid, it is not, unfortunately, a standard
form. A component type does not exist in the NeuroML standard that can
encapsulate this form (It has more parameters, and has an additional
[`+`{.docutils .literal .notranslate}]{.pre}` `{.docutils .literal
.notranslate}[`taumin`{.docutils .literal .notranslate}]{.pre}).

This is not a problem, though, because NeuroML can be easily extended
using [[LEMS]{.std .std-ref}](Landing.html#userdocs-lems){.reference
.internal}. We can write a new component type to encapsulate these
dynamics based on the [LEMS
definition](https://github.com/NeuroML/NeuroML2/blob/development/NeuroML2CoreTypes/Channels.xml#L99){.reference
.external} of the
[HHSigmoidVariable](https://docs.neuroml.org/Userdocs/Schemas/Channels.html#hhsigmoidvariable){.reference
.external} component type:

::: {.highlight-xml .notranslate}
::: highlight
    <ComponentType name="HHSigmoidVariable"
                   extends="baseHHVariable"
                   description="Sigmoidal form for variable equation">
        <Dynamics>
            <DerivedVariable name="x" dimension="none" exposure="x" value="rate / (1 + exp(0 - (v - midpoint)/scale))"/>
        </Dynamics>
    </ComponentType>
:::
:::

Our new component will be this, and we will save it in a different file
that we can then "include" in the nas channel definition file:

::: {.highlight-xml .notranslate}
::: highlight
    <!-- Saved as RayTau.nml -->
    <?xml version="1.0" encoding="ISO-8859-1"?>
    <neuroml xmlns="http://www.neuroml.org/schema/neuroml2" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.neuroml.org/schema/neuroml2 https://raw.github.com/NeuroML/NeuroML2/development/Schemas/NeuroML2/NeuroML_v2beta4.xsd" id="NeuroML_ionChannel">

        <ComponentType name="Ray_tau"
                      extends="baseVoltageDepTime"
                      description="Tau parameter">

            <Parameter name="max_tau" dimension="time"/>
            <Parameter name="min_tau" dimension="time"/>
            <Parameter name="midpoint" dimension="voltage"/>
            <Parameter name="scale" dimension="voltage"/>
            <Dynamics>
                <DerivedVariable name="t" dimension="time" exposure="t" value="(((max_tau - min_tau) / (1 + exp(0 - (v - midpoint) / scale))) + min_tau)"/>
            </Dynamics>
        </ComponentType>
    </neuroml>
:::
:::

Note that it is very similar to the HHSigmoidVariable definition. The
only difference is that we have had to define additional parameters to
use in our equation. Also note that [`HHSigmoidVariable`{.docutils
.literal .notranslate}]{.pre} extends [`baseHHVariable`{.docutils
.literal .notranslate}]{.pre} which extends
[`baseVoltageDepVariable`{.docutils .literal .notranslate}]{.pre}.
However, since we know that [`tau`{.docutils .literal
.notranslate}]{.pre} is a time value, we extend the
[`baseVoltageDepTime`{.docutils .literal .notranslate}]{.pre} component
type instead. This is very similar to
[`baseVoltageDepVariable`{.docutils .literal .notranslate}]{.pre}, but
is designed for component types producing time values, such as the time
course here.

We save this as a different component type (a class), and we will
provide it parameters to create our time courses for both [`m`{.docutils
.literal .notranslate}]{.pre} and [`h`{.docutils .literal
.notranslate}]{.pre} activation variables. Our completed file will look
like this:

::: {.highlight-xml .notranslate}
::: highlight
    <!-- saved as nas.channel.nml -->
    <?xml version="1.0" encoding="ISO-8859-1"?>
    <neuroml xmlns="http://www.neuroml.org/schema/neuroml2" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.neuroml.org/schema/neuroml2 https://raw.github.com/NeuroML/NeuroML2/development/Schemas/NeuroML2/NeuroML_v2beta4.xsd" id="NeuroML_ionChannel">

        <notes>NeuroML file containing a single ion channel</notes>
        <include href="RayTau.nml" />

        <ionChannel id="nas" type="ionChannelHH" conductance="1pS" species="na">

            <gate id="m" instances="3" type="gateHHtauInf">
                <steadyState type="HHSigmoidVariable" midpoint="-30.1mV" scale="6.65mV" rate="1" />
                <timeCourse type="Ray_tau" min_tau="0.83 ms" max_tau="0.093 ms" midpoint="-20.3 mV" scale="6.45mV"/>
            </gate>

            <gate id="h" instances="1" type="gateHHtauInf">
                <steadyState type="HHSigmoidVariable" midpoint="-51.4mV" scale="-5.9mV" rate="1" />
                <timeCourse type="Ray_tau" min_tau="1.9 ms" max_tau="12.24 ms" midpoint="-32.6 mV" scale="-8.0mV"/>
            </gate>
        </ionChannel>

    </neuroml>
:::
:::

Running [`pynml-channelanalysis`{.docutils .literal
.notranslate}]{.pre}` `{.docutils .literal
.notranslate}[`nas.nml`{.docutils .literal .notranslate}]{.pre} will
generate the graphs for the steady state and time course from our
channel definition, and you will see that these are the same as the
graphs we generated from the mod files.

We are almost there, but we have a little more work to do here. Remember
that the original mod file limited the value of steady state at 40mV? We
have not incorporated that into our channel file yet.

Since the [`HHSigmoidVariable`{.docutils .literal .notranslate}]{.pre}
we have used for the steady state does not allow multiple equations, we
will write a new component type for the steady state also:

::: {.highlight-xml .notranslate}
::: highlight
    <?xml version="1.0" encoding="ISO-8859-1"?>
    <neuroml xmlns="http://www.neuroml.org/schema/neuroml2" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.neuroml.org/schema/neuroml2 https://raw.github.com/NeuroML/NeuroML2/development/Schemas/NeuroML2/NeuroML_v2beta4.xsd" id="NeuroML_ionChannel">

        <ComponentType name="Ray_inf"
            extends="baseVoltageDepVariable"
            description="Inf parameter for Ray et al 2020" >

            <Constant name="table_max" dimension="voltage" value="40 mV"/>
            <Parameter name="rate" dimension="none"/>
            <Parameter name="midpoint" dimension="voltage"/>
            <Parameter name="scale" dimension="voltage"/>
            <Dynamics>
                <ConditionalDerivedVariable name="x" dimension="per_time" exposure="x">
                    <Case condition="v .gt. table_max" value="(rate / (1 + exp(0 - (table_max - midpoint) / scale)))"/>
                    <Case value="(rate / (1 + exp(0 - (v - midpoint) / scale)))"/>
                </ConditionalDerivedVariable>
            </Dynamics>
        </ComponentType>
    <neuroml/>
:::
:::

The only difference here is that instead of the
[`DerivedVariable`{.docutils .literal .notranslate}]{.pre}, we have used
a [`ConditionalDerivedVariable`{.docutils .literal .notranslate}]{.pre}
that allows conditional dynamics. We have two cases here:

-   if [`v`{.docutils .literal .notranslate}]{.pre} is greater than
    [`table_max`{.docutils .literal .notranslate}]{.pre} (which is
    40mV), the rate is the value at [`v=40mV`{.docutils .literal
    .notranslate}]{.pre}

-   otherwise, the rate is calculated from the value of [`v`{.docutils
    .literal .notranslate}]{.pre}

::: {#a-kv-naf-kst .section}
###### 1a) Kv, Naf, Kst[\#](#a-kv-naf-kst "Link to this heading"){.headerlink} {aria-level="7"}

The nas channel is now done. If we look at the other channels---naf, kv,
and kst---they follow similar formalisms. So, we can re-use our newly
created component types, [`Ray_inf`{.docutils .literal
.notranslate}]{.pre} and [`Ray_tau`{.docutils .literal
.notranslate}]{.pre}. In fact, we consolidate them in a single file,
[`RaySigmoid.nml`{.docutils .literal .notranslate}]{.pre}, and "include"
this in the channel definition files. For example, here is
[`kv.channel.nml`{.docutils .literal .notranslate}]{.pre}:

::: {.highlight-xml .notranslate}
::: highlight
    <?xml version="1.0" encoding="ISO-8859-1"?>
    <neuroml xmlns="http://www.neuroml.org/schema/neuroml2" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.neuroml.org/schema/neuroml2 https://raw.github.com/NeuroML/NeuroML2/development/Schemas/NeuroML2/NeuroML_v2beta4.xsd" id="NeuroML_ionChannel">

        <notes>NeuroML file containing a single ion channel</notes>
        <include href="RaySigmoid.nml" />

        <ionChannel id="kv" conductance="1pS" type="ionChannelHH" species="k">

            <notes>
                    Implementation of A type K+ channel ( KV ) from Wustenberg DG, Boytcheva M, Grunewald B, Byrne JH, Menzel R, Baxter DA.
                    This is a delayed rectifier type K+ channel in Apis mellifera Kenyon cells (cultured).
            </notes>

            <!-- custom component types because the tables in the mod files only go to 40 -->
            <gate id="m" type="gateHHtauInf" instances="4">
                <steadyState type="Ray_inf" rate="1.0" midpoint="-37.6mV" scale="27.24mV"/>
                <timeCourse type="Ray_tau" min_tau="1.85 ms" max_tau="3.53 ms" midpoint="45.0 mV" scale="-13.71mV"/>
            </gate>

        </ionChannel>
    </neuroml>
:::
:::

Plots for the steady state and time course are:

<figure id="id5" class="align-center">
<a
href="_images/Steady_state(s)_of_activation_variables_of_kv_from_kv.channel.nml_at_6.3_degC.png"
class="reference internal image-reference"><img
src="_images/Steady_state%28s%29_of_activation_variables_of_kv_from_kv.channel.nml_at_6.3_degC.png"
style="width: 768.0px; height: 388.8px;"
alt="Image showing steady state dynamics of activation variables of kv channel, generated with pynml-channelanalysis" /></a>
<figcaption><p><span class="caption-number">Fig. 71 </span><span
class="caption-text">Steady state dynamics of activation variables of kv
channel, generated with pynml-channelanalysis</span><a href="#id5"
class="headerlink" title="Link to this image">#</a></p></figcaption>
</figure>

<figure id="id6" class="align-center">
<a
href="_images/Time_Course(s)_of_activation_variables_of_kv_from_kv.channel.nml_at_6.3_degC.png"
class="reference internal image-reference"><img
src="_images/Time_Course%28s%29_of_activation_variables_of_kv_from_kv.channel.nml_at_6.3_degC.png"
style="width: 768.0px; height: 388.8px;"
alt="Image showing time course of activation variables of kv channel, generated with pynml-channelanalysis" /></a>
<figcaption><p><span class="caption-number">Fig. 72 </span><span
class="caption-text">Time course of activation variables of kv channel,
generated with pynml-channelanalysis</span><a href="#id6"
class="headerlink" title="Link to this image">#</a></p></figcaption>
</figure>

In these graphs, the effect of the conditional at 40mV becomes more
apparent.
:::

::: {#b-ka .section}
###### 1b) Ka[\#](#b-ka "Link to this heading"){.headerlink} {aria-level="7"}

The last remaining channel is the ka channel. It's dynamics are defined
in the mod file as:

::: {.highlight-default .notranslate}
::: highlight
    PROCEDURE settables(v (mV)) { 
    UNITSOFF
            TABLE minf, hinf, mtau, htau FROM -120 TO 40 WITH 641
            minf  = 1.0 / (1 + exp((-20.1 - v)/16.1))
            hinf  = 1.0 / ( 1 + exp( ( v + 74.7 ) / 7 ) )
            mtau = (1.65 - 0.35) / ((1 + exp(- (v + 70) / 4.0)) * (1 + exp((v + 20) / 12.0))) + 0.35
            htau = (90 - 2.5) / ((1 + exp(- (v + 60) / 25.0)) * (1 + exp((v + 62) / 16.0))) + 2.5
    UNITSON
    }
:::
:::

The steady state here follows the same formalism, but the time course
does not. So, we need to create new component type to encapsulate the
time courses here, similar to the [`Ray_tau`{.docutils .literal
.notranslate}]{.pre} component type that we did before:

::: {.highlight-xml .notranslate}
::: highlight
        <ComponentType name="Ray_ka_tau"
                       extends="baseVoltageDepTime"
                       description="Tau parameter to describe ka">

            <Parameter name="max_tau" dimension="time"/>
            <Parameter name="min_tau" dimension="time"/>
            <Parameter name="midpoint1" dimension="voltage"/>
            <Parameter name="scale1" dimension="voltage"/>
            <Parameter name="midpoint2" dimension="voltage"/>
            <Parameter name="scale2" dimension="voltage"/>
            <Constant name="table_max" dimension="voltage" value="40 mV"/>
            <Dynamics>
                <ConditionalDerivedVariable name="t" dimension="time" exposure="t" >
                    <Case condition="v .gt. table_max" value="(max_tau - min_tau) / ((1 + exp(-(table_max + midpoint1) / scale1)) * ( 1 + exp((table_max + midpoint2) / scale2))) + min_tau"/>
                    <Case value="(max_tau - min_tau) / ((1 + exp(-(v + midpoint1) / scale1)) * ( 1 + exp((v + midpoint2) / scale2))) + min_tau"/>

                </ConditionalDerivedVariable>
            </Dynamics>
        </ComponentType>
:::
:::

We also add this to our [`RaySigmoid.nml`{.docutils .literal
.notranslate}]{.pre} file. The ka.channel.nml file will, finally, look
like this:

::: {.highlight-xml .notranslate}
::: highlight
    <?xml version="1.0" encoding="ISO-8859-1"?>
    <neuroml xmlns="http://www.neuroml.org/schema/neuroml2" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.neuroml.org/schema/neuroml2 https://raw.github.com/NeuroML/NeuroML2/development/Schemas/NeuroML2/NeuroML_v2beta4.xsd" id="NeuroML_ionChannel">

        <notes>NeuroML file containing a single ion channel</notes>
        <include href="RaySigmoid.nml" />

        <ionChannel id="ka" conductance="1pS" type="ionChannelHH" species="k">

            <notes>
                    Implementation of A type K+ channel ( KA ) from Wustenberg DG, Boytcheva M, Grunewald B, Byrne JH, Menzel R, Baxter DA.
                    This is transient A type K+ channel in Apis mellifera Kenyon cells (cultured).
            </notes>

            <!-- custom component types because the tables in the mod files only go to 40 -->
            <gate id="m" type="gateHHtauInf" instances="3">
                <timeCourse type="Ray_ka_tau" midpoint1="70mV" midpoint2="2.0mV" scale1="4.0mV" scale2="12.0mV" min_tau="0.35ms" max_tau="1.65ms"/>
                <steadyState type="Ray_inf" rate="1.0" midpoint="-20.1mV" scale="16.1mV"/>
            </gate>

            <gate id="h" type="gateHHtauInf" instances="1">
                <timeCourse type="Ray_ka_tau" midpoint1="60mV" midpoint2="62.0mV" scale1="25.0mV" scale2="16.0mV" min_tau="2.5ms" max_tau="90.0ms"/>
                <steadyState type="Ray_inf" rate="1.0" midpoint="-74.7mV" scale="-7.0mV"/>
            </gate>

        </ionChannel>
    </neuroml>
:::
:::

That is all the ion channels converted.

The ion channels are usually the most involved to convert because one
must understand their initial descriptions in the mod files. Even though
the [NMODL
language](https://nrn.readthedocs.io/en/8.2.2/python/modelspec/programmatic/mechanisms/nmodl.html#nmodl){.reference
.external} used in mod files does have a well defined structure, like
general programming languages, it is free-flowing. This means that
different people can write the same dynamics in different ways. On the
other hand, NeuroML and LEMS are more formal with more strict
structures, and once channels are converted to these formats, they are
much easier to understand.
:::
:::

::: {#step-2-creating-the-morphology .section}
###### Step 2) Creating the morphology[\#](#step-2-creating-the-morphology "Link to this heading"){.headerlink}

Since the morphology of the KC is a single compartment, we don't need to
export it from NEURON. We can create it ourselves.

The morphology is given in the HOC script:

::: {.highlight-default .notranslate}
::: highlight
    proc geom() {
         soma {  // Total Cm = 4 pF
            L = 6.366
            diam = 20
        }
    }
:::
:::

We can create this using a Python script:

::: {.highlight-python .notranslate}
::: highlight
        celldoc = component_factory(
            "NeuroMLDocument", id="KC_doc"
        )  # type: neuroml.NeuroMLDocument
        cell = celldoc.add("Cell", id="KC", validate=False)  # type: neuroml.Cell
        cell.setup_nml_cell()
        cell.add_segment([0, 0, 0, 20], [0, 0, 6.366, 20], seg_type="soma")
:::
:::

The [`setup_nml_cell`{.docutils .literal .notranslate}]{.pre} and
[`add_segment`{.docutils .literal .notranslate}]{.pre} methods are part
of the [Cell class in the standard
API](https://libneuroml.readthedocs.io/en/latest/userdocs/coreclasses.html#cell){.reference
.external}.

Now that we have the morphology and the ion channels for the KC, we can
add the biophysics to the morphology to complete the cell:

::: {.highlight-python .notranslate}
::: highlight
        # biophysics
        # all
        cell.set_resistivity("35.4 ohm_cm", group_id="all")
        cell.set_specific_capacitance("1 uF_per_cm2", group_id="all")
        cell.set_init_memb_potential("-70mV")
        cell.set_spike_thresh("-10mV")

        cell.add_channel_density(
            nml_cell_doc=celldoc,
            cd_id="pas",
            ion_channel="pas",
            cond_density="9.75e-5 S_per_cm2",
            erev="-70 mV",
            group_id="all",
            ion="non_specific",
            ion_chan_def_file="channels/pas.channel.nml",
        )

        # K
        cell.add_channel_density(
            nml_cell_doc=celldoc,
            cd_id="kv",
            ion_channel="kv",
            cond_density="1.5e-3 S_per_cm2",
            erev="-81 mV",
            group_id="all",
            ion="k",
            ion_chan_def_file="channels/kv.channel.nml",
        )
        cell.add_channel_density(
            nml_cell_doc=celldoc,
            cd_id="ka",
            ion_channel="ka",
            cond_density="1.4525e-2 S_per_cm2",
            erev="-81 mV",
            group_id="all",
            ion="k",
            ion_chan_def_file="channels/ka.channel.nml",
        )
        cell.add_channel_density(
            nml_cell_doc=celldoc,
            cd_id="kst",
            ion_channel="kst",
            cond_density="2.0275e-3 S_per_cm2",
            erev="-81 mV",
            group_id="all",
            ion="k",
            ion_chan_def_file="channels/kst.channel.nml",
        )
        # Na
        cell.add_channel_density(
            nml_cell_doc=celldoc,
            cd_id="naf",
            ion_channel="naf",
            cond_density="3.5e-2 S_per_cm2",
            erev="58 mV",
            group_id="all",
            ion="na",
            ion_chan_def_file="channels/naf.channel.nml",
        )
        cell.add_channel_density(
            nml_cell_doc=celldoc,
            cd_id="nas",
            ion_channel="nas",
            cond_density="3e-3 S_per_cm2",
            erev="58 mV",
            group_id="all",
            ion="na",
            ion_chan_def_file="channels/nas.channel.nml",
        )
:::
:::

This completes the cell. We export it to a NeuroML file.
:::

::: {#step-3-testing-the-model .section}
###### Step 3) Testing the model[\#](#step-3-testing-the-model "Link to this heading"){.headerlink}

Finally, we want to test our NeuroML conversion against the original
cell to see that it exhibits the same dynamics. The
[`test_kc.py`{.docutils .literal .notranslate}]{.pre} script runs a
simple step current simulation with a single KC cell and shows its
membrane potentials:

<figure id="id7" class="align-center">
<a href="_images/KC-NEURON.png"
class="reference internal image-reference"><img
src="_images/KC-NEURON.png"
style="width: 768.0px; height: 401.20000000000005px;"
alt="Image showing the membrane potential of the NEURON implementation of the KC cell model with a step current." /></a>
<figcaption><p><span class="caption-number">Fig. 73 </span><span
class="caption-text">The membrane potential of the NEURON implementation
of the KC cell model with a step current.</span><a href="#id7"
class="headerlink" title="Link to this image">#</a></p></figcaption>
</figure>

We write a quick simulation to reproduce these using our NeuroML model:

::: {.highlight-python .notranslate}
::: highlight
    def step_current_omv_kc():
        """Create a step current simulation OMV LEMS file"""
        # read the cell file, modify it, write a new one
        netdoc = read_neuroml2_file("KC.cell.nml")
        kc_cell = netdoc.cells[0]
        net = netdoc.add(neuroml.Network, id="KC_net", validate=False)
        pop = net.add(neuroml.Population, id="KC_pop", component=kc_cell.id, size=1)

        # should be same as test_kc.py
        pg = netdoc.add(
            neuroml.PulseGenerator(
                id="pg", delay="100ms", duration="500ms",
                amplitude="16pA"
            )
        )

        # Add these to cells
        input_list = net.add(
            neuroml.InputList(id="input_list", component=pg.id, populations=pop.id)
        )
        aninput = input_list.add(
            neuroml.Input(
                id="0",
                target="../%s[0]" % (pop.id),
                destination="synapses",
                segment_id="0",
            )
        )
        write_neuroml2_file(netdoc, "KC.net.nml")

        generate_lems_file_for_neuroml(
            sim_id="KC_step_test",
            target=net.id,
            neuroml_file="KC.net.nml",
            duration="700ms",
            dt="0.01ms",
            lems_file_name="LEMS_KC_step_test.xml",
            nml_doc=netdoc,
            gen_spike_saves_for_all_somas=True,
            target_dir=".",
            gen_saves_for_quantities={
                "k.dat": ["KC_pop[0]/biophys/membraneProperties/kv/iDensity"]
            },
            copy_neuroml=False
        )

        data = run_lems_with_jneuroml_neuron(
            "LEMS_KC_step_test.xml", load_saved_data=True, compile_mods=True
        )

        generate_plot(
            xvalues=[data["t"]],
            yvalues=[data["KC_pop[0]/v"]],
            title="Membrane potential: KC",
        )
:::
:::

This will generate graphs of the KC's membrane potential.

<figure id="id8" class="align-center">
<a href="_images/KC-NeuroML.png"
class="reference internal image-reference"><img
src="_images/KC-NeuroML.png" style="width: 768.0px; height: 388.8px;"
alt="Image showing the membrane potential of the NeuroML implementation of the KC cell model with a step current." /></a>
<figcaption><p><span class="caption-number">Fig. 74 </span><span
class="caption-text">The membrane potential of the NeuroML
implementation of the KC cell model with a step current.</span><a
href="#id8" class="headerlink"
title="Link to this image">#</a></p></figcaption>
</figure>

As we can see, the membrane potentials look very similar. In the next
page, we will also set up some more validation tests to better verify
that the NEURON and NeuroML implementations produce the same dynamics.

Since this writes a LEMS simulation file also, we can also run the LEMS
file directly for later verification:

::: {.highlight-default .notranslate}
::: highlight
    pynml LEMS_KC_step_test.xml -neuron -nogui
:::
:::
:::
:::
:::

[]{#document-Userdocs/Walkthroughs/RayEtAl2020/OMV}

::: {#adding-omv-tests .section .tex2jax_ignore .mathjax_ignore}
#### Adding OMV tests[\#](#adding-omv-tests "Link to this heading"){.headerlink}

Now that we have converted the cell models to NeuroML, we want to ensure
that we get the same behaviour from the NeuroML converted cell model as
from the original NEURON code. For this, we have the [Open Source Model
Validation (OMV)
framework](https://github.com/OpenSourceBrain/osb-model-validation/){.reference
.external}. The idea here is that we write simple test files that
[`omv`{.docutils .literal .notranslate}]{.pre} will run, and we provide
data that [`omv`{.docutils .literal .notranslate}]{.pre} can check
against to see if the tests results are correct.

We will add OMV tests for the KC here.

::: {#step-1-getting-spike-data-from-the-neuron-model .section}
##### Step 1) Getting spike data from the NEURON model[\#](#step-1-getting-spike-data-from-the-neuron-model "Link to this heading"){.headerlink}

For [`omv`{.docutils .literal .notranslate}]{.pre} to test the model
against, we need to generate some spike data that it knows to be the
expected values. The [`test_kc.py`{.docutils .literal
.notranslate}]{.pre} script already provides us with the spike data. So,
we can run the script and note down the spike times in a Model Emergent
Properties (MEP) file:

::: {.highlight-yaml .notranslate}
::: highlight
    system: Testing a detailed cell

    experiments:
      stepKC:
        expected:
          spike times: [144.45000000000556, 177.64999999997536, 210.29999999994567, 242.74999999991616, 275.0499999998868, 307.19999999985754, 339.2249999998284, 371.07499999979945, 402.74999999977064, 434.1499999997421, 465.2249999997138, 495.87499999968594, 525.9749999997221, 555.3499999998289, 583.8499999999326]

    # generated from test-kc.py
:::
:::
:::

::: {#step-2-adding-tests .section}
##### Step 2) Adding tests[\#](#step-2-adding-tests "Link to this heading"){.headerlink}

Next, we can write a OSB Model Test (OMT) file to test the model using
the step current simulation we had written before:

::: {.highlight-yaml .notranslate}
::: highlight
    target: LEMS_KC_step_test.xml
    engine: jNeuroML_NEURON
    mep: .test.kc.mep
    experiments:
      stepKC:
        observables:
          spike times:
            file:
              path: KC_step_test.KC_pop.v.dat
              columns: [0,1]
              scaling: [1000,1000]
            spike detection:
              method: threshold
              threshold: -10.
            tolerance: 0.00
:::
:::

Note that we start with a tolerance of 0 here. Let us run the test and
see what we get:

::: {.highlight-console .notranslate}
::: highlight
    $ omv test .test.kc.jnmlneuron.omt

    [omv]
    [omv] Running the tests defined in .test.kc.jnmlnrn.omt
    [omv] =================================================
    [omv]   Found 1 experiment(s) to run on engine: jNeuroML_NEURON
    [omv] PATH: :/home/asinha/.local/share/virtualenvs/neuroml-311-dev/bin
    [omv] Env vars: {'PYTHONPATH': '/home/asinha/local/lib/python/site-packages', 'NEURON_HOME': '/home/asinha/.local/share/virtualenvs/neuroml-311-dev', 'JNML_HOME': PosixPath('/usr/bin')}
    [omv]   Running file ./LEMS_KC_step_test.xml with jNeuroML_NEURON, env: {'NEURON_HOME': '/home/asinha/.local/share/virtualenvs/neuroml-311-dev', 'JNML_HOME': PosixPath('/usr/bin')}
    [omv]     Running the commands: [/usr/bin/jnml /home/asinha/Documents/02_Code/00_mine/models/RayEtAl2020/NeuroML2/LEMS_KC_step_test.xml -neuron -nogui -run] in (/home/asinha/Documents/02_Code/00_mine/models/RayEtAl2020/NeuroML2; cwd=/home/asinha/Documents/02_Code/00_mine/models/RayEtAl2020/NeuroML2; shell=False; env={'NEURON_HOME': '/home/asinha/.local/share/virtualenvs/neuroml-311-dev', 'JNML_HOME': PosixPath('/usr/bin')})
    [omv]     Commands: ['/usr/bin/jnml', '/home/asinha/Documents/02_Code/00_mine/models/RayEtAl2020/NeuroML2/LEMS_KC_step_test.xml', '-neuron', '-nogui', '-run'] completed successfully
    [omv]   Success with running jNeuroML_NEURON
    [omv]   Running checks for experiment: stepKC
    [omv]
    [omv] Comparison of
                        (observed data): [143.95, 176.76, 209.23, 241.52, 273.66, 305.68, 337.57, 369.33, 400.94, 432.38, 463.62, 494.59999999999997, 525.28, 555.59, 585.45]
                        and
                        (expected data): [144.45000000000556, 177.64999999997536, 210.29999999994567, 242.74999999991616, 275.0499999998868, 307.19999999985754, 339.2249999998284, 371.07499999979945, 402.74999999977064, 434.1499999997421, 465.2249999997138, 495.87499999968594, 525.9749999997221, 555.3499999998289, 583.8499999999326]
                        failed against tolerance 0
    [omv]   A better tolerance to try is: 0.005087969567027844
    [omv]       Observable                        Test Passed
    [omv]       --------------------------------------------------
    [omv]       spike times                            ✘
    [omv]       +++++++++++++++++++++ Error info ++++++++++++++++++
    [omv]        Return code: 0
:::
:::

We see that there's a slight difference in the spike times we obtain
from our implementation. This is not unexpected. Small variations in how
the mod files are written, or how the morphology is set up can result in
small variations in the spike times. [`omv`{.docutils .literal
.notranslate}]{.pre} will suggest a tolerance value for us to use. The
smaller the tolerance, the better.

We update our file to use the suggested tolerance and re-run the test:

::: {.highlight-console .notranslate}
::: highlight
    $ omv test .test.kc.jnmlnrn.omt
    [omv]
    [omv] Running the tests defined in .test.kc.jnmlnrn.omt
    [omv] =================================================
    [omv]   Found 1 experiment(s) to run on engine: jNeuroML_NEURON
    [omv] PATH: :/home/asinha/.local/share/virtualenvs/neuroml-311-dev/bin
    [omv] Env vars: {'PYTHONPATH': '/home/asinha/local/lib/python/site-packages', 'NEURON_HOME': '/home/asinha/.local/share/virtualenvs/neuroml-311-dev', 'JNML_HOME': PosixPath('/usr/bin')}
    [omv]   Running file ./LEMS_KC_step_test.xml with jNeuroML_NEURON, env: {'NEURON_HOME': '/home/asinha/.local/share/virtualenvs/neuroml-311-dev', 'JNML_HOME': PosixPath('/usr/bin')}
    [omv]     Running the commands: [/usr/bin/jnml /home/asinha/Documents/02_Code/00_mine/models/RayEtAl2020/NeuroML2/LEMS_KC_step_test.xml -neuron -nogui -run] in (/home/asinha/Documents/02_Code/00_mine/models/RayEtAl2020/NeuroML2; cwd=/home/asinha/Documents/02_Code/00_mine/models/RayEtAl2020/NeuroML2; shell=False; env={'NEURON_HOME': '/home/asinha/.local/share/virtualenvs/neuroml-311-dev', 'JNML_HOME': PosixPath('/usr/bin')})
    [omv]     Commands: ['/usr/bin/jnml', '/home/asinha/Documents/02_Code/00_mine/models/RayEtAl2020/NeuroML2/LEMS_KC_step_test.xml', '-neuron', '-nogui', '-run'] completed successfully
    [omv]   Success with running jNeuroML_NEURON
    [omv]   Running checks for experiment: stepKC
    [omv]
    [omv]       Observable                        Test Passed
    [omv]       --------------------------------------------------
    [omv]       spike times                            ✔
    [omv]
    [omv]                   =================================
    [omv]                   Test passed: .test.kc.jnmlnrn.omt
:::
:::

We can also add a simple validation test that will allow
[`omv`{.docutils .literal .notranslate}]{.pre} to validate the various
NeuroML files in the repository:

::: {.highlight-yaml .notranslate}
::: highlight
    target: "*.c*.nml"
    engine: jNeuroML_validate
:::
:::

One can run all the OMV tests in a repository at once:

::: {.highlight-console .notranslate}
::: highlight
    $ omv all
    [omv] Python 3. Ignoring tests for non Py3 compatible engines: False
    [omv]
    [omv]
    [omv] Running the tests defined in ./.test.kc.jnmlnrn.omt
    [omv] ===================================================
    [omv]   Found 1 experiment(s) to run on engine: jNeuroML_NEURON
    [omv] PATH: :/home/asinha/.local/share/virtualenvs/neuroml-311-dev/bin
    [omv] Env vars: {'PYTHONPATH': '/home/asinha/local/lib/python/site-packages', 'NEURON_HOME': '/home/asinha/.local/share/virtualenvs/neuroml-311-dev', 'JNML_HOME': PosixPath('/usr/bin')}
    [omv]   Running file ./LEMS_KC_step_test.xml with jNeuroML_NEURON, env: {'NEURON_HOME': '/home/asinha/.local/share/virtualenvs/neuroml-311-dev', 'JNML_HOME': PosixPath('/usr/bin')}
    [omv]     Running the commands: [/usr/bin/jnml /home/asinha/Documents/02_Code/00_mine/models/RayEtAl2020/NeuroML2/LEMS_KC_step_test.xml -neuron -nogui -run] in (/home/asinha/Documents/02_Code/00_mine/models/RayEtAl2020/NeuroML2; cwd=/home/asinha/Documents/02_Code/00_mine/models/RayEtAl2020/NeuroML2; shell=False; env={'NEURON_HOME': '/home/asinha/.local/share/virtualenvs/neuroml-311-dev', 'JNML_HOME': PosixPath('/usr/bin')})
    [omv]     Commands: ['/usr/bin/jnml', '/home/asinha/Documents/02_Code/00_mine/models/RayEtAl2020/NeuroML2/LEMS_KC_step_test.xml', '-neuron', '-nogui', '-run'] completed successfully
    [omv]   Success with running jNeuroML_NEURON
    [omv]   Running checks for experiment: stepKC
    [omv]
    [omv]       Observable                        Test Passed
    [omv]       --------------------------------------------------
    [omv]       spike times                            ✔
    [omv]
    [omv]
    [omv]       [ Test 1 of 2 complete - failed so far: 0 ]
    [omv]
    [omv]
    [omv] Running the tests defined in ./.test.validate.omt
    [omv] =================================================
    [omv]   No mep file specified. Will only run simulation using: jNeuroML_validate
    [omv]   Found 1 experiment(s) to run on engine: jNeuroML_validate
    [omv]   Running with jNeuroML_validate, using: ['/usr/bin/jnml', '-validate', '/home/asinha/Documents/02_Code/00_mine/models/RayEtAl2020/NeuroML2/GGN.morph.cell.nml', '/home/asinha/Documents/02_Code/00_mine/models/RayEtAl2020/NeuroML2/GGN.cell.nml', '/home/asinha/Documents/02_Code/00_mine/models/RayEtAl2020/NeuroML2/KC.cell.nml']...
    [omv]     Running the commands: [/usr/bin/jnml -validate /home/asinha/Documents/02_Code/00_mine/models/RayEtAl2020/NeuroML2/GGN.morph.cell.nml /home/asinha/Documents/02_Code/00_mine/models/RayEtAl2020/NeuroML2/GGN.cell.nml /home/asinha/Documents/02_Code/00_mine/models/RayEtAl2020/NeuroML2/KC.cell.nml] in (/home/asinha/Documents/02_Code/00_mine/models/RayEtAl2020/NeuroML2; cwd=/home/asinha/Documents/02_Code/00_mine/models/RayEtAl2020/NeuroML2; shell=False; env={'JNML_HOME': PosixPath('/usr/bin')})
    [omv]     Commands: ['/usr/bin/jnml', '-validate', '/home/asinha/Documents/02_Code/00_mine/models/RayEtAl2020/NeuroML2/GGN.morph.cell.nml', '/home/asinha/Documents/02_Code/00_mine/models/RayEtAl2020/NeuroML2/GGN.cell.nml', '/home/asinha/Documents/02_Code/00_mine/models/RayEtAl2020/NeuroML2/KC.cell.nml'] completed successfully
    [omv]   Running checks for experiment: Dry run
    [omv]
    [omv]       Observable                        Test Passed
    [omv]       --------------------------------------------------
    [omv]       dry                                    ✔
    [omv]
    [omv]
    [omv]       [ Test 2 of 2 complete - failed so far: 0 ]
    [omv]
    [omv]                             -------------
    [omv]                             2 test(s) run
    [omv]                             -------------
    [omv]
    [omv]                           All tests passing!
    [omv]                           ==================
:::
:::
:::

::: {#step-3-setting-up-continuous-testing-on-github-actions .section}
##### Step 3) Setting up continuous testing on GitHub Actions[\#](#step-3-setting-up-continuous-testing-on-github-actions "Link to this heading"){.headerlink}

Now that we have set up OMV test, we want to set up ["continuous
testing"
(CT)](https://en.wikipedia.org/wiki/Continuous_testing){.reference
.external}. What this means is that we want these test to run
automatically whenever we make any changes. On GitHub, we can do these
using [GitHub Actions](https://docs.github.com/en/actions){.reference
.external}.

To set up a GitHub Action, we need to set up a "workflow file" in the
[`.github/workflows`{.docutils .literal .notranslate}]{.pre} directory.
We create one called [`omv-ci.yml`{.docutils .literal
.notranslate}]{.pre}:

::: {.highlight-yaml .notranslate}
::: highlight
    name: Continuous build using OMV

    on:
      schedule:
        - cron: "1 1 1 */2 *"
      push:
        branches: [ master, development, experimental ]
      pull_request:
        branches: [ master, development, experimental ]

    jobs:
      build:

        runs-on: ubuntu-latest
        strategy:
          fail-fast: false
          matrix:
            python-version: [ "3.8", "3.10"]
            engine: [ jNeuroML_validate, jNeuroML_NEURON ]

        steps:
        - uses: actions/checkout@v4

        - name: Set up Python  ${{ matrix.python-version }}
          uses: actions/setup-python@v5
          with:
            python-version:  ${{ matrix.python-version }}

        - name: Install OMV
          run: |
            pip install OSBModelValidation
            pip install scipy sympy matplotlib cython pandas tables

        - name: Run OMV tests on engine ${{ matrix.engine }}
          run: |
            omv all -V --engine=${{ matrix.engine }}

        - name: OMV final version info
          run: |
            omv list -V # list installed engines
            env
:::
:::

This installs OMV and runs all test using it in the repository.
Additionally, it tests this out on a couple of Python versions.

Now, whenever we make a change to the repository---either as a pull
request, or as a direct push of a commit, these tests will be run
immediately to tell us if the model is still working as it should. These
can be seen in the ["actions" tab in the GitHub
Repository](https://github.com/OpenSourceBrain/262670/actions){.reference
.external}.
:::
:::
:::
:::
:::
:::
:::

::: {.toctree-wrapper .compound}
[]{#document-Events/Events}

::: {#neuroml-outreach-and-events .section .tex2jax_ignore .mathjax_ignore}
[]{#neuromlevents-main}

## NeuroML outreach and events[\#](#neuroml-outreach-and-events "Link to this heading"){.headerlink}

The NeuroML community organises regular training and outreach events.
Recent meetings are listed below (please see the individual pages for
more details):

-   [[July 2024: NeuroML tutorial at CNS 2024]{.std
    .std-ref}](Landing.html#neuromlevents-202407cns){.reference
    .internal}

-   [[April 2024: NeuroML hackathon at HARMONY 2024]{.std
    .std-ref}](Landing.html#neuromlevents-202404harmony){.reference
    .internal}

-   [[June 2022: NeuroML tutorial at CNS\*2022 satellite tutorials]{.std
    .std-ref}](Landing.html#neuromlevents-202206cns){.reference
    .internal}

-   [[April 2022: NeuroML development workshop at HARMONY 2022]{.std
    .std-ref}](Landing.html#neuromlevents-202104harmony){.reference
    .internal}

-   [[October 2021: NeuroML development workshop at COMBINE
    meeting]{.std
    .std-ref}](Landing.html#neuromlevents-202109combine){.reference
    .internal}

-   [[August 2021: NeuroML tutorial at INCF Training Weeks]{.std
    .std-ref}](Landing.html#neuromlevents-202108incf){.reference
    .internal}

-   [[July 2021: NeuroML tutorial at CNS\*2021]{.std
    .std-ref}](Landing.html#neuromlevents-202107cns2021){.reference
    .internal}

-   [[March 2021: NeuroML hackathon at HARMONY 2021]{.std
    .std-ref}](Landing.html#neuromlevents-202103harmony){.reference
    .internal}
:::

[]{#document-Events/202407-CNS2024}

::: {#july-2024-neuroml-tutorial-at-cns-2024 .section .tex2jax_ignore .mathjax_ignore}
[]{#neuromlevents-202407cns}

## July 2024: NeuroML tutorial at CNS 2024[\#](#july-2024-neuroml-tutorial-at-cns-2024 "Link to this heading"){.headerlink}

::: {.tip .admonition}
Registration for the annual meeting of the Organisation of Computational
Neuroscience, CNS 2024, is open.

Please register for the CNS 2024
[here](https://www.cnsorg.org/cns-2024-quick){.reference .external}.
:::

We will be running a full day NeuroML tutorial during the [annual
meeting of the Organisation of Computational Neuroscience,
2024](https://www.cnsorg.org/cns-2024-quick){.reference .external}
meeting in Natal, Brazil from July 20--24.

**Standardised, data-driven computational modelling with NeuroML using
Open Source Brian**

Data-driven models of neurons and circuits are important for
understanding how the properties of membrane conductances, synapses,
dendrites and the anatomical connectivity between neurons generate the
complex dynamical behaviours of brain circuits in health and disease.
However, even though data and models are being made publicly available
in recent years and the use of standards such as [Neurodata Without
Borders (NWB)](https://nwb.org){.reference .external} and NeuroML to
promote FAIR (Findable, Accessible, Interoperable and Reusable)
neuroscience is on the rise, development of data driven models remains
hampered by the difficulty of finding appropriate data and the inherent
complexity involved in their construction.

The [Open Source Brain web platform
(OSB)](https://opensourcebrain.org){.reference .external} brings
together data, accompanying analysis tools, and computational models in
a single, scalable resource. It indexes repositories from established
sources such as the [DANDI data
archive](https://dandiarchive.org){.reference .external}, the [ModelDB
model sharing archive](https://modeldb.science){.reference .external},
and GitHub to provide easy access to a plethora of experimental data and
models, including a large number standardised in NWB and NeuroML formats

OSB also incorporates the NeuroML software ecosystem. NeuroML is an
established community standard and software ecosystem that enables the
development of biophysically detailed models using a declarative,
simulator independent description. The software ecosystem supports all
steps of the model lifecycle and allows users to automatically generate
code and run their NeuroML models using a number of well established
simulation engines (NEURON/NetPyNE).

Read the NeuroML pre-print here:
[https://www.biorxiv.org/content/10.1101/2023.12.07.570537v1](https://www.biorxiv.org/content/10.1101/2023.12.07.570537v1){.reference
.external}

::: {#agenda .section}
[]{#neuromlevents-202407cns-agenda}

### Agenda[\#](#agenda "Link to this heading"){.headerlink}

In this tutorial, attendees will learn about:

-   Finding data and models on OSB

-   NeuroML and its software ecosystem

-   Using NeuroML models on OSB

-   Building and simulating new NeuroML models constrained by the data
    on OSB

We will also provide assistance with advanced tasks, and discuss new
features to further aid researchers.
:::

::: {#times-and-dates .section}
[]{#neuromlevents-202407cns-when}

### Times and dates[\#](#times-and-dates "Link to this heading"){.headerlink}

More details to follow
:::

::: {#registration .section}
[]{#neuromlevents-202407cns-registration}

### Registration[\#](#registration "Link to this heading"){.headerlink}

To take part in the tutorial, please register
[here](https://www.cnsorg.org/cns-2024-quick){.reference .external} for
CNS 2024.

We look forward to working with the community to drive further uptake of
NeuroML compliant models and tools!
:::
:::

[]{#document-Events/202404-Harmony}

::: {#april-2024-neuroml-hackathon-at-harmony-2024 .section .tex2jax_ignore .mathjax_ignore}
[]{#neuromlevents-202404harmony}

## April 2024: NeuroML hackathon at HARMONY 2024[\#](#april-2024-neuroml-hackathon-at-harmony-2024 "Link to this heading"){.headerlink}

::: {.tip .admonition}
Registration for the COMBINE initiative's HARMONY 2024 meeting is free.

Please register for the COMBINE HARMONY 2024 meeting
[here](https://co.mbine.org/author/harmony-2024/){.reference .external}
if you are coming to our NeuroML workshop. Registration for HARMONY is
free.
:::

We will be running a NeuroML workshop during the upcoming [COMBINE
network's HARMONY
2024](https://co.mbine.org/author/harmony-2024/){.reference .external}
meeting on Tues 9th April 2024 in London, entitled:

**NeuroML hackathon: convert your neuron and network models to open,
standardised, reusable formats**

This will be an opportunity for developers of models in computational
neuroscience to get an introduction to the aims and structure of
NeuroML, a guide to the tools available for building/converting their
models to NeuroML, and to receive hands on help with expressing their
models (or other published models they are interested in) in NeuroML
format, making them more open, accessible and reusable.

::: {#agenda .section}
[]{#neuromlevents-202404harmony-agenda}

### Agenda[\#](#agenda "Link to this heading"){.headerlink}

More details to follow
:::

::: {#times-and-dates .section}
[]{#neuromlevents-202404harmony-when}

### Times and dates[\#](#times-and-dates "Link to this heading"){.headerlink}

More details to follow
:::

::: {#registration .section}
[]{#neuromlevents-202404harmony-registration}

### Registration[\#](#registration "Link to this heading"){.headerlink}

To take part in the workshop, please register
[here](https://co.mbine.org/author/harmony-2024/){.reference .external}
for the HARMONY meeting (registration is free).
:::

::: {#open-an-issue-beforehand .section}
[]{#neuromlevents-202404harmony-prepare}

### Open an issue beforehand\![\#](#open-an-issue-beforehand "Link to this heading"){.headerlink}

While it will be possible to raise and discuss new issues at the
workshop, it will be easier to manage and plan work/discussions if you
open an issue with a description of the problem you are trying to
address at:
[NeuroML/NeuroML2#issues](https://github.com/NeuroML/NeuroML2/issues){.github
.reference .external}.
:::

::: {#slack .section}
[]{#neuromlevents-202404harmony-slack}

### Slack[\#](#slack "Link to this heading"){.headerlink}

To aid communication with the community during (and after) the meeting,
we have a **Slack channel** for NeuroML related discussions. Please
contact [Padraig Gleeson](mailto:p_DOT_gleeson_AT_ucl.ac.uk){.reference
.external} for an invite.

We look forward to working with the community to drive further uptake of
NeuroML compliant models and tools!
:::
:::

[]{#document-Events/20220630-CNS2022}

::: {#june-2022-neuroml-tutorial-at-cns-2022-satellite-tutorials .section .tex2jax_ignore .mathjax_ignore}
[]{#neuromlevents-202206cns}

## June 2022: NeuroML tutorial at CNS\*2022 satellite tutorials[\#](#june-2022-neuroml-tutorial-at-cns-2022-satellite-tutorials "Link to this heading"){.headerlink}

An online NeuroML tutorial will be held at the [CNS\*2022 satellite
tutorials](https://ocns.github.io/SoftwareWG/pages/software-wg-satellite-tutorials-at-cns-2022.html){.reference
.external}. Registration for the satellite tutorials is free, but
required.

This tutorial is intended for members of the research community
interested in learning more about how NeuroML and its related
technologies facilitates the standardization, sharing, and collaborative
development of models.

::: {#times-and-dates .section}
[]{#neuromlevents-202206cns-when}

### Times and dates[\#](#times-and-dates "Link to this heading"){.headerlink}

-   Dates: June 30, 2022

-   Time: : [1400--1700
    UTC](https://www.timeanddate.com/worldclock/fixedtime.html?iso=20220630T14){.reference
    .external}
:::

::: {#target-audience .section}
[]{#neuromlevents-202206cns-who}

### Target audience[\#](#target-audience "Link to this heading"){.headerlink}

Anyone who is already familiar with computational modelling, but is keen
to standardise, share and collaboratively develop their models.
:::

::: {#where .section}
[]{#neuromlevents-202206cns-where}

### Where[\#](#where "Link to this heading"){.headerlink}

The tutorial be done online via Zoom and will make use of the [Open
Source Brain v2](https://v2.opensourcebrain.org){.reference .external}
integrated web research platform.

Please register for the [CNS\*2022 satellite
tutorials](https://ocns.github.io/SoftwareWG/pages/software-wg-satellite-tutorials-at-cns-2022.html){.reference
.external} to receive the Zoom links.
:::

::: {#agenda .section}
[]{#neuromlevents-202206cns-agenda}

### Agenda[\#](#agenda "Link to this heading"){.headerlink}

::: {#part-1-introduction-to-neuroml .section}
#### Part 1: Introduction to NeuroML[\#](#part-1-introduction-to-neuroml "Link to this heading"){.headerlink}

-   Overview of NeuroML

-   Introduce the NeuroML tool chain

-   Introduce main documentation

-   Related technologies and initiatives
:::

::: {#part-2-hands-on-demonstrations-of-building-and-using-neuroml-models .section}
#### Part 2: Hands on demonstrations of building and using NeuroML models[\#](#part-2-hands-on-demonstrations-of-building-and-using-neuroml-models "Link to this heading"){.headerlink}

-   Izhikevich neuron hands on tutorial

-   Spiking neuron network tutorial

-   Single compartment HH neuron tutorial

-   Multi compartmental HH neuron tutorial
:::
:::
:::

[]{#document-Events/202204-Harmony}

::: {#april-2022-neuroml-development-workshop-at-harmony-2022 .section .tex2jax_ignore .mathjax_ignore}
[]{#neuromlevents-202104harmony}

## April 2022: NeuroML development workshop at HARMONY 2022[\#](#april-2022-neuroml-development-workshop-at-harmony-2022 "Link to this heading"){.headerlink}

::: {.tip .admonition}
Registration for the COMBINE initiative's HARMONY 2022 meeting is free.

Please register for the COMBINE HARMONY 2022 meeting
[here](https://combine-org.github.io/events/){.reference .external} if
you are coming to our NeuroML workshop. Registration for HARMONY is
free.
:::

We will be running a NeuroML development workshop during the upcoming
[COMBINE network's HARMONY
2022](https://combine-org.github.io/events/){.reference .external}
meeting on Thus 28 April 2022. This will be an opportunity for anyone
interested in developing NeuroML or adding support for the format to
their application talk about their work and hear about other
developments.

::: {#agenda .section}
[]{#neuromlevents-202104harmony-agenda}

### Agenda[\#](#agenda "Link to this heading"){.headerlink}

The agenda for the meeting can be found
[here](https://docs.google.com/document/d/1GOkmylIlLH3dbxB2Pmis7VxFPzipstLBdXn5uBu4rv8/edit){.reference
.external}.
:::

::: {#times-and-dates .section}
[]{#neuromlevents-202104harmony-when}

### Times and dates[\#](#times-and-dates "Link to this heading"){.headerlink}

The workshop will take place on **Thus 28 April 2022** at 15:00-18:00
UTC
([converter](https://www.timeanddate.com/worldclock/converter.html?iso=20220427T150000&amp;p1=136&amp;p2=1440&amp;p3=137&amp;p4=179&amp;p5=37&amp;p6=176&amp;p7=248){.reference
.external}).
:::

::: {#registration .section}
[]{#neuromlevents-202104harmony-registration}

### Registration[\#](#registration "Link to this heading"){.headerlink}

To take part in the workshop, please register
[here](https://combine-org.github.io/events/){.reference .external} for
the HARMONY meeting (registration is free).

You will get sent details to access the HARMONY agenda, which will have
links to the **Zoom session for the NeuroML workshop**.
:::

::: {#open-an-issue-beforehand .section}
[]{#neuromlevents-202104harmony-prepare}

### Open an issue beforehand\![\#](#open-an-issue-beforehand "Link to this heading"){.headerlink}

While it will be possible to raise and discuss new issues at the
workshop, it will be easier to manage and plan work/discussions if you
open an issue with a description of the problem you are trying to
address at:
[NeuroML/NeuroML2#issues](https://github.com/NeuroML/NeuroML2/issues){.github
.reference .external}.
:::

::: {#slack .section}
[]{#neuromlevents-202104harmony-slack}

### Slack[\#](#slack "Link to this heading"){.headerlink}

To aid communication with the community during (and after) the meeting,
we have a **Slack channel** for NeuroML related discussions. Please
contact [Padraig Gleeson](mailto:p_DOT_gleeson_AT_ucl.ac.uk){.reference
.external} for an invite.

We look forward to working with the community to drive further uptake of
NeuroML compliant models and tools!
:::
:::

[]{#document-Events/202109-COMBINE}

::: {#october-2021-neuroml-development-workshop-at-combine-meeting .section .tex2jax_ignore .mathjax_ignore}
[]{#neuromlevents-202109combine}

## October 2021: NeuroML development workshop at COMBINE meeting[\#](#october-2021-neuroml-development-workshop-at-combine-meeting "Link to this heading"){.headerlink}

::: {.tip .admonition}
Registration for the COMBINE 2021 meeting is free.

Register for the COMBINE 2021 meeting
[here](https://combine-org.github.io/events/){.reference .external}.
Registration is free.
:::

A NeuroML development workshop will be held as part of the [annual
COMBINE meeting in October
2021](https://combine-org.github.io/events/){.reference .external}.

The general theme of the workshop is to discuss the current status of
the NeuroML standard and the complete software ecosystem, and future
development plans.

::: {#times-and-dates .section}
[]{#neuromlevents-202109combine-when}

### Times and dates[\#](#times-and-dates "Link to this heading"){.headerlink}

-   13 October 2021

-   8-11am PDT/11-2pm EST/4-7pm UK/5-8pm CET/8:30-11:30 IST
:::

::: {#target-audience .section}
[]{#neuromlevents-202109combine-who}

### Target audience[\#](#target-audience "Link to this heading"){.headerlink}

Everyone that is involved/interested in developing tools that
use/integrate with NeuroML is encouraged to join.

Please register for the COMBINE meeting (free of charge) to receive
access to the complete schedule of the meeting, including links to the
various virtual meetings/sessions.
:::

::: {#agenda-minutes .section}
### Agenda/minutes[\#](#agenda-minutes "Link to this heading"){.headerlink}

The agenda/minutes for the meeting can be found
[here](https://docs.google.com/document/d/1rZp6fggUe1vlo5fYK-CiUP__fdJV8xYg-wrkpIp0dHk/edit){.reference
.external}.
:::
:::

[]{#document-Events/202108-INCF-Training-Week}

::: {#august-2021-neuroml-tutorial-at-incf-training-weeks .section .tex2jax_ignore .mathjax_ignore}
[]{#neuromlevents-202108incf}

## August 2021: NeuroML tutorial at INCF Training Weeks[\#](#august-2021-neuroml-tutorial-at-incf-training-weeks "Link to this heading"){.headerlink}

A NeuroML tutorial will be held at the [Virtual INCF Neuroinformatics
Training Weeks
2021](https://www.incf.org/virtual-incf-neuroinformatics-training-week-2021){.reference
.external}.

This tutorial is intended for members of the research community
interested in learning more about how NeuroML and its related
technologies facilitates the standardization, sharing, and collaborative
development of models.

::: {#times-and-dates .section}
[]{#neuromlevents-202108incf-when}

### Times and dates[\#](#times-and-dates "Link to this heading"){.headerlink}

This tutorial will be offered twice during the Neuroinformatics Training
Week: session 1 is targeted to participants residing in Europe, Africa,
and the Americas while session 2 is targeted to participants residing in
Asia and Australia.

Session 1:

-   Dates: 23 Aug 2021

-   Time: : 11:00-15:00 EDT / 17:00-21:00 CEST

Session 2

-   Dates: 26 Aug 2021

-   Time: 09:00-13:00 CEST / 16:00-20:00 JST / 17:00-21:00 AEST
:::

::: {#target-audience .section}
[]{#neuromlevents-202108incf-who}

### Target audience[\#](#target-audience "Link to this heading"){.headerlink}

Anyone who is already familiar with computational modelling, but is keen
to standardise, share and collaboratively develop their models.
:::

::: {#agenda .section}
[]{#neuromlevents-202108incf-agenda}

### Agenda[\#](#agenda "Link to this heading"){.headerlink}

::: {#part-1-introduction-to-neuroml .section}
#### Part 1: Introduction to NeuroML[\#](#part-1-introduction-to-neuroml "Link to this heading"){.headerlink}

-   Overview of NeuroML

-   Introduce the NeuroML tool chain

-   Introduce main documentation

-   Related technologies and initiatives
:::

::: {#part-2-hands-on-demonstrations-of-building-and-using-neuroml-models .section}
#### Part 2: Hands on demonstrations of building and using NeuroML models[\#](#part-2-hands-on-demonstrations-of-building-and-using-neuroml-models "Link to this heading"){.headerlink}

-   Izhikevich neuron hands on tutorial

-   Spiking neuron network tutorial

-   Single compartment HH neuron tutorial

-   Multi compartmental HH neuron tutorial
:::
:::
:::

[]{#document-Events/202107-CNS2021}

::: {#july-2021-neuroml-tutorial-at-cns-2021 .section .tex2jax_ignore .mathjax_ignore}
[]{#neuromlevents-202107cns2021}

## July 2021: NeuroML tutorial at CNS\*2021[\#](#july-2021-neuroml-tutorial-at-cns-2021 "Link to this heading"){.headerlink}

::: {.tip .admonition}
Register for the 30th Annual meeting of the Organization for
Computational Neurosciences (OCNS).

Register for the CNS\*2021
[here](http://www.cnsorg.org/cns-2021){.reference .external}.
:::

We will be running a half day tutorial at the 30th annual meeting of the
Organization for Computational Neurosciences (OCNS):
[CNS\*2021](https://www.cnsorg.org/cns-2021){.reference .external}.

The goal of the tutorial is to teach users to: **build, visualise,
analyse and simulate models using NeuroML**.

::: {#why-take-part .section}
[]{#neuromlevents-202107cns2021-why}

### Why take part?[\#](#why-take-part "Link to this heading"){.headerlink}

This tutorial is aimed at new and current NeuroML users. We will start
with a quick introduction to the NeuroML standard and the associated
software ecosystem, after which we will proceed to conduct hands-on
sessions to show how one can build computational models with NeuroML.
:::

::: {#times-and-dates .section}
[]{#neuromlevents-202107cns2021-when}

### Times and dates[\#](#times-and-dates "Link to this heading"){.headerlink}

-   [Friday 2nd July
    1500UTC](https://www.timeanddate.com/worldclock/fixedtime.html?msg=NeuroML+tutorial+at+CNS%2A2021&amp;iso=20210702T11&amp;p1=179&amp;ah=3){.reference
    .external}.
:::

::: {#registration .section}
[]{#neuromlevents-202107cns2021-registration}

### Registration[\#](#registration "Link to this heading"){.headerlink}

To take part in the tutorial, please register
[here](https://www.cnsorg.org/cns-2021){.reference .external} for the
CNS\*2021 meeting.
:::

::: {#pre-requisites .section}
[]{#neuromlevents-202107cns2021-prepare}

### Pre-requisites[\#](#pre-requisites "Link to this heading"){.headerlink}

The sessions will make use of the NeuroML Python tools. Please follow
the documentation to install them on your system if you wish to use them
locally:

-   [[PyNeuroML]{.std .std-ref}](Landing.html#pyneuroml){.reference
    .internal}

-   [[libNeuroML]{.std .std-ref}](Landing.html#libneuroml){.reference
    .internal}

You can also use the interactive Jupyter notebooks from the
documentation if you prefer
([example](https://docs.neuroml.org/Userdocs/NML2_examples/SingleNeuron.html){.reference
.external}). These can be run on Binder and Google Collab in your web
browser and do not require you to install anything locally on your
computer.
:::

::: {#slack .section}
[]{#neuromlevents-202107cns2021-slack}

### Slack[\#](#slack "Link to this heading"){.headerlink}

To aid communication with the community during (and after) the meeting,
we have a **Slack channel** for NeuroML related discussions. Please
contact [Padraig Gleeson](mailto:p_DOT_gleeson_AT_ucl.ac.uk){.reference
.external} for an invite.

You can also contact the NeuroML community using one of our other
[[channels]{.std .std-ref}](Landing.html#contact-chat){.reference
.internal}.

We look forward to working with the community to drive further uptake of
NeuroML compliant models and tools!
:::
:::

[]{#document-Events/202103-Harmony}

::: {#march-2021-neuroml-hackathon-at-harmony-2021 .section .tex2jax_ignore .mathjax_ignore}
[]{#neuromlevents-202103harmony}

## March 2021: NeuroML hackathon at HARMONY 2021[\#](#march-2021-neuroml-hackathon-at-harmony-2021 "Link to this heading"){.headerlink}

::: {.tip .admonition}
Registration for the COMBINE initiative's HARMONY 2021 meeting is free.

Register for the COMBINE: HARMONY 2021 meeting
[here](http://co.mbine.org/events/HARMONY_2021){.reference .external}.
Registration is free.
:::

We will be running 3 online NeuroML hackathon sessions during the
upcoming [COMBINE: HARMONY
2021](http://co.mbine.org/events/HARMONY_2021){.reference .external}
meeting on 23-25th March. The general theme of the sessions will be:
**learn to build, visualise, analyse and simulate your models using
NeuroML**.

::: {#why-take-part .section}
[]{#neuromlevents-202103harmony-why}

### Why take part?[\#](#why-take-part "Link to this heading"){.headerlink}

These hackathons will give members of the neuroscience community the
chance to:

-   Get high level introductions to the NeuroML language and tool chain

-   Meet the NeuroML core development team and editors

-   Find out the latest information on which simulators/applications
    support NeuroML

-   Open, discuss and work on issues related to converting your model to
    NeuroML, or supporting NeuroML in your simulator

-   Learn how to share your models with the community
:::

::: {#times-and-dates .section}
[]{#neuromlevents-202103harmony-when}

### Times and dates[\#](#times-and-dates "Link to this heading"){.headerlink}

All sessions will be online and take place over 3 hours (9am-noon
Pacific; 12-3pm EST time; 4-7pm UK/UTC; 5-8pm CET, 9:30pm-12:30am IST;
note non-standard US/EU time differences that week). The broad focus of
each of the sessions (dependent on interests of attendees) is:

-   Tues 23rd March: Introduction to NeuroML, general questions about
    usage

-   Wed 24th March: Detailed cell/conductance based models (e.g.
    converting channels to NeuroML)

-   Thus 25th March: Abstract/point neuron networks including PyNN
    interactions
:::

::: {#registration .section}
[]{#neuromlevents-202103harmony-registration}

### Registration[\#](#registration "Link to this heading"){.headerlink}

To take part in the hackathon, please register
[here](http://co.mbine.org/events/HARMONY_2021){.reference .external}
for the HARMONY meeting (registration is free). You will get sent
details to access the [agenda](https://harmony2021.sched.com){.reference
.external}, which will have links to the Zoom sessions for each of the
days.
:::

::: {#open-an-issue-beforehand .section}
[]{#neuromlevents-202103harmony-prepare}

### Open an issue beforehand\![\#](#open-an-issue-beforehand "Link to this heading"){.headerlink}

While it will be possible to raise and discuss new issues at the
hackathons, it will be easier to manage and plan work/discussions if you
open an issue with a description of the problem you are trying to
address at:
[NeuroML/NeuroML2#issues](https://github.com/NeuroML/NeuroML2/issues){.github
.reference .external}.
:::

::: {#slack .section}
[]{#neuromlevents-202103harmony-slack}

### Slack[\#](#slack "Link to this heading"){.headerlink}

To aid communication with the community during (and after) the meeting,
we have a **Slack channel** for NeuroML related discussions. Please
contact [Padraig Gleeson](mailto:p_DOT_gleeson_AT_ucl.ac.uk){.reference
.external} for an invite.

We look forward to working with the community to drive further uptake of
NeuroML compliant models and tools!
:::
:::

[]{#document-Events/2012-Edinburgh}

::: {#march-2012-fourth-neuroml-development-workshop .section .tex2jax_ignore .mathjax_ignore}
[]{#neuromlevents-2012edinburgh}

## March 2012: Fourth NeuroML Development Workshop[\#](#march-2012-fourth-neuroml-development-workshop "Link to this heading"){.headerlink}

Convergence in Computational Neuroscience 2012 Joint BrainScaleS
CodeJam/NeuroML workshop, Edinburgh, 12-16th March

The NeuroML Development Workshops and the
[BrainScaleS](http://brainscales.kip.uni-heidelberg.de/){.reference
.external} (previously FACETS)
[CodeJams](http://neuralensemble.org/meetings.html){.reference
.external} have been two important initiatives in recent years for
developers of tools in computational and systems neuroscience to present
their latest work, exchange ideas and work at achieving interoperability
between software applications for investigating brain function. This
year these groups held a joint workshop (Convergence in Computational
Neuroscience) on March 12th-16th in the Informatics Forum in Edinburgh,
UK. The meeting was held at the [Informatics
Forum](https://www.ed.ac.uk/informatics/about/location/forum){.reference
.external} in Edinburgh, UK, from 12th to 16th March 2012.

::: {#meeting-report .section}
### Meeting report[\#](#meeting-report "Link to this heading"){.headerlink}

Note: details of the meeting activities from Wednesday 14th to Friday
16th are available on the [NeuralEnsemble.org
webpage](http://neuralensemble.org/meetings/CodeJam5/){.reference
.external}.

::: {#monday-12th-march-neuroml-development-workshop-day-1 .section}
#### Monday 12th March: NeuroML Development Workshop Day 1[\#](#monday-12th-march-neuroml-development-workshop-day-1 "Link to this heading"){.headerlink}

::: {#morning-session-current-state-of-neuroml-2-development-relationship-to-other-initiatives .section}
##### Morning session: Current state of NeuroML 2 development & relationship to other initiatives[\#](#morning-session-current-state-of-neuroml-2-development-relationship-to-other-initiatives "Link to this heading"){.headerlink}

Chair: Andrew Davison

::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| Time | Session                                         | Speaker     |
+======+=================================================+=============+
| 0    | **Welcome & goals of meeting**                  | Angus       |
| 9:00 |                                                 | Silver      |
|      | Angus welcomed attendees, thanking in           |             |
|      | particular out local organisers at University   |             |
|      | of Edinburgh, Mike Hull and Mika Pelko!         |             |
+------+-------------------------------------------------+-------------+
| 0    | **Update on latest developments in NeuroML      | Padraig     |
| 9:05 | 2/LEMS                                          | Gleeson     |
|      | ([PPT](https://docs.neuroml.org/_static/NeuroM  |             |
|      | L2012/PGleeson_NeuroMLIntro2012.ppt){.reference |             |
|      | .external})**                                   |             |
|      |                                                 |             |
|      | Padraig presented an introduction to NeuroML,   |             |
|      | starting with an overview of the modular nature |             |
|      | of NeuroML v1.x, advantages of the use of XML,  |             |
|      | examples of neuronal models in NeuroML, current |             |
|      | tools which support the language, (including    |             |
|      | the recently added                              |             |
|      | [Neur                                           |             |
|      | oMorpho.org](http://NeuroMorpho.org){.reference |             |
|      | .external} and Channelpedia).                   |             |
|      |                                                 |             |
|      | The requirements for v2.0 were presented.       |             |
|      | Explicit definitions of model component         |             |
|      | behaviour allows description of the dynamics of |             |
|      | model components in a simulator independent,    |             |
|      | machine readable way. The relationship between  |             |
|      | LEMS and NeuroML2 was discussed. A short        |             |
|      | overview of NeuroML 2.0 was given including     |             |
|      | dimensions/units. Example of adaptive           |             |
|      | exponential integrate and cell network was      |             |
|      | presented. An overview of libNeuroML was given. |             |
|      | Export to NEURON, neuroConstruct, interaction   |             |
|      | with SBML was shown.                            |             |
+------+-------------------------------------------------+-------------+
| 0    | **Introduction to NineML & libNineML            | Mike Hull   |
| 9:30 | ([PDF](https://docs.neuroml.org/_               |             |
|      | static/NeuroML2012/MHull_NineML.pdf){.reference |             |
|      | .external})**                                   |             |
|      |                                                 |             |
|      | Mike gave an introduction to the NineML object  |             |
|      | model and libNineML. The INCF Task Force in     |             |
|      | Multiscale Modelling created the language,      |             |
|      | consisting of an Abstraction Layer and User     |             |
|      | Layer. Mike's presentation focused on the       |             |
|      | abstraction layer which contains many terms for |             |
|      | the object model. Core object in the            |             |
|      | Abstraction layer were presented:               |             |
|      | ComponentClass, Interface with Parameters,      |             |
|      | Ports (AnalogPorts, EventPorts and Reduce       |             |
|      | Ports), Dynamics with StateVariables and a      |             |
|      | Regime Transition Graph (with Transitions,      |             |
|      | StateAssignments, Aliases).                     |             |
|      |                                                 |             |
|      | libNineML (in Python) loads and saves models    |             |
|      | from/to XML to/from Python that helps with code |             |
|      | generation, turns models into NEST, NEURON and  |             |
|      | PyNN.                                           |             |
+------+-------------------------------------------------+-------------+
| 1    | **The COMBINE Initiative                        | Nicolas le  |
| 0:00 | ([PDF](https://docs.neuroml.org/_static/NeuroM  | Novere      |
|      | L2012/NleNovere_NeuroML-COMBINE.pdf){.reference |             |
|      | .external})**                                   |             |
|      |                                                 |             |
|      | Nicolas presented the Combine Initiative:       |             |
|      | Standards for describing the whole life-cycle   |             |
|      | of modelling. Different communities favour      |             |
|      | different types of models that are more suited  |             |
|      | for their domain. Current standardisation       |             |
|      | efforts depend on the initial people,           |             |
|      | individual funding structure, IP issues.        |             |
|      | Specifications, API's, test-suites, etc. really |             |
|      | need industry-grade support which is not        |             |
|      | compatible with standard academic usages and    |             |
|      | possibilities.                                  |             |
|      |                                                 |             |
|      | The vision of COMBINE is to pave the space of   |             |
|      | model descriptions with coordination of         |             |
|      | standard development (without interference with |             |
|      | the development). There are criteria for        |             |
|      | inclusion in the core COMBINE standards: new    |             |
|      | standards must be different from those already  |             |
|      | included, described in technical specification  |             |
|      | documents, free, open, developed and used by    |             |
|      | more than one team, democratically elected      |             |
|      | members, mature software support including API, |             |
|      | and must be actively developed.                 |             |
|      |                                                 |             |
|      | COMBINE organises joint meetings replacing      |             |
|      | standards specific ones (e.g. SBML Hackathon).  |             |
|      | Next COMBINE meeting is in Toronto in August.   |             |
|      | HARMONY for hacking will be in Maastricht in    |             |
|      | May.                                            |             |
+------+-------------------------------------------------+-------------+
| 1    | **Coffee**                                      |             |
| 0:30 |                                                 |             |
+------+-------------------------------------------------+-------------+
| 1    | **The International Neuroinformatics            | Sean Hill   |
| 1:00 | Coordinating Facility**                         |             |
|      |                                                 |             |
|      | Sean presented the motivation for, current      |             |
|      | structure and the aims of the INCF. The goal of |             |
|      | neuroscience is to understand the brain. We're  |             |
|      | at a crisis point in understanding disorders.   |             |
|      | Big Pharma is pulling out of neuroscience due   |             |
|      | to the high cost and risk of understanding      |             |
|      | these disorders. Past centuries have focused on |             |
|      | obtaining observations. More recently, models   |             |
|      | were used to understand these observations.     |             |
|      | Today we have eScience as a new way of handling |             |
|      | large-scale data, modelling, simulations,       |             |
|      | linking data, etc. One of the INCF's goals is   |             |
|      | to transform neuroscience into an eScience from |             |
|      | level of molecules to clinic.                   |             |
|      |                                                 |             |
|      | Integration of databases is a goal, which       |             |
|      | requires standardized data formats. There are   |             |
|      | 16 member countries in INCF. He gave an         |             |
|      | introduction to the 4 programs from the past    |             |
|      | few years: Digital Brain Atlasing, Multi-Scale  |             |
|      | Modeling, Ontologies of Neural Structures and   |             |
|      | Standards for Datasharing.                      |             |
|      |                                                 |             |
|      | He discussed future plans for the INCF          |             |
|      | "Cyberinfrastructure", including a discussion   |             |
|      | of the planned INCF cloud "Dropbox" for data    |             |
|      | which could include metadata tags enabling      |             |
|      | global search.                                  |             |
+------+-------------------------------------------------+-------------+
| 1    | **Collaborative Modelling Repository update**   | Padraig     |
| 1:30 |                                                 | Gleeson     |
|      | Padraig presented the initial work towards an   |             |
|      | open source, collaborative repository for       |             |
|      | NeuroML models, the Open Source Brain project.  |             |
|      | A preliminary version of this is avaliable      |             |
|      | here. It will be based on a version control     |             |
|      | repository (initially Mercurial) storing the    |             |
|      | model files. It will have automatic generation  |             |
|      | of online documentation of the models from      |             |
|      | ChannelML, MorphML, etc. Connectivity matrices  |             |
|      | for network connections, etc. can be generated  |             |
|      | for models which are stored in neuroConstruct   |             |
|      | format. NeuroLex IDs can be used to identify    |             |
|      | cells and channels to other resources. Feedback |             |
|      | on the initial implementation was welcomed.     |             |
+------+-------------------------------------------------+-------------+
| 1    | **Open discussion on model specification        |             |
| 2:00 | initiatives**                                   |             |
|      |                                                 |             |
|      | Differences were pointed out between INCF and   |             |
|      | COMBINE approaches to standards development.    |             |
|      | IEEE provides an infrastructure for review,     |             |
|      | etc. Nicolas discussed the burdens of obtaining |             |
|      | "official" standards board recognition. Many of |             |
|      | the COMBINE procedures imitate W3C procedures.  |             |
|      | A good standard is one that works. COMBINE      |             |
|      | criteria don't say anything about the standard  |             |
|      | document itself.                                |             |
+------+-------------------------------------------------+-------------+
| 1    | **Lunch**                                       |             |
| 3:00 |                                                 |             |
+------+-------------------------------------------------+-------------+
:::
:::

::: {#afternoon-session-specification-of-detailed-biophysical-components-in-neuroml-2-lems .section}
##### Afternoon session: Specification of detailed biophysical components in NeuroML 2/LEMS[\#](#afternoon-session-specification-of-detailed-biophysical-components-in-neuroml-2-lems "Link to this heading"){.headerlink}

Chair: Sharon Crook

::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| Time | Session                                         | Speaker     |
+======+=================================================+=============+
| 1    | **Representing channels, synapses & conductance | Robert      |
| 4:00 | based models                                    | Cannon      |
|      | ([                                              |             |
|      | PDF](https://docs.neuroml.org/_static/NeuroML20 |             |
|      | 12/RCannon_ModellingIonChannels.pdf){.reference |             |
|      | .external})**                                   |             |
|      |                                                 |             |
|      | Robert gave a presentation on ways to represent |             |
|      | synapses and conductances. He defined the       |             |
|      | Nernst equation with XML based on Hille's       |             |
|      | description. It is still not clear how some     |             |
|      | things should be done, in particular how to     |             |
|      | handle dimensions and units. Currently,         |             |
|      | dimensions are defined and then assertions      |             |
|      | about relationships among dimensions can be     |             |
|      | made. Units are not defined until NeuroML is    |             |
|      | written (with numerical values, e.g. -70mV).    |             |
|      | There was some discussion of how dimensions     |             |
|      | should be defined. Physicists solved this       |             |
|      | problem by developing SI units. Both space and  |             |
|      | no space between values and units are allowed   |             |
|      | in LEMS. There was some discussion of how       |             |
|      | events can be handled in LEMS.                  |             |
+------+-------------------------------------------------+-------------+
| 1    | **Experiences with using NeuroML 2**            | Avrama      |
| 4:20 |                                                 | Blackwell   |
|      | Avrama gave a brief report of her hands on      |             |
|      | experience with using NeuroML 2. She has a      |             |
|      | medium spiny projection neuron model which      |             |
|      | she's translating to GENESIS. She didn't want   |             |
|      | to use a GUI unless absolutely necessary and    |             |
|      | has been manually editing the XML. Many of the  |             |
|      | (non calcium dependent) channels have already   |             |
|      | been converted to NeuroML 2. She can only run   |             |
|      | single compartment versions of her model since  |             |
|      | LEMS doesn't yet support multi compartmental    |             |
|      | models. She has produced some some multi        |             |
|      | segment morphologies in NML2, even though these |             |
|      | can't be used in LEMS based simulations yet.    |             |
|      | She requires a way to specify distance from the |             |
|      | soma. Another difficulty is not being able to   |             |
|      | define a template (dendritic) subbranch and add |             |
|      | it multiple times to the cell. She will add     |             |
|      | spines later. Calcium dependent channels are a  |             |
|      | work in progress in NeuroML v2.0, but some      |             |
|      | useful simulations have already been done with  |             |
|      | her developing model.                           |             |
+------+-------------------------------------------------+-------------+
| 1    | **Implementing cerebellar models in PyNEURON,   | Sergio      |
| 4:40 | neuroConstruct & NeuroML**                      | Solinas     |
|      |                                                 |             |
|      | Sergio is developing cerebellar models (Golgi   |             |
|      | cells and granular cells) in a network of       |             |
|      | granule layer. Solinas S, Nieus T and D\`Angelo |             |
|      | E (2010) A realistic large-scale model of the   |             |
|      | cerebellum granular layer predicts circuit      |             |
|      | spatio-temporal filtering properties. Frontiers |             |
|      | in Cellular Neuroscience                        |             |
|      | ([link](                                        |             |
|      | http://www.frontiersin.org/Cellular_Neuroscienc |             |
|      | e/10.3389/fncel.2010.00012/abstract){.reference |             |
|      | .external}) gives an overview of the network.   |             |
|      | There were improvements made to the model in    |             |
|      | 2011 and it was translated to Python Neuron for |             |
|      | parallel simulation on cluster. Added gap       |             |
|      | junctions, more realistic inputs. Python eased  |             |
|      | improvements to the model.                      |             |
+------+-------------------------------------------------+-------------+
| 1    | **Large scale cortical models for studying LFPs | Richard     |
| 5:00 | ([PPT]                                          | Tomsett     |
|      | (https://docs.neuroml.org/_static/NeuroML2012/R |             |
|      | Tomsett_LargeScaleCorticalModel.ppt){.reference |             |
|      | .external})**                                   |             |
|      |                                                 |             |
|      | Richard presented his work on developing large  |             |
|      | scale cortical models for studying Local Field  |             |
|      | Potentials. What network properties cause       |             |
|      | pathological dynamics? Much data comes from     |             |
|      | electrodes in vitro. Gaute Einevol's work looks |             |
|      | at how dendritic structure affects the field    |             |
|      | potentials in a network (Linden et al Neuron    |             |
|      | 2010). Richard focused on Bush and Sejnowski J  |             |
|      | Neurosci Methods 1993 method to reduce model    |             |
|      | and see what the LFP looks like (and compared   |             |
|      | to Linden data). Then since it looked pretty    |             |
|      | good he created a network of these reduced      |             |
|      | models for simulation and analysis. Then looked |             |
|      | at results from Utah array in Matlab. Next      |             |
|      | he'll add Gaussian connectivity and some        |             |
|      | patches and long range connections.             |             |
+------+-------------------------------------------------+-------------+
| 1    | **Coffee**                                      |             |
| 5:15 |                                                 |             |
+------+-------------------------------------------------+-------------+
| 1    | **Break out sessions**                          |             |
| 5:30 |                                                 |             |
|      | -   Channel and synapse specifications          |             |
|      |                                                 |             |
|      | -   Proposed structure for abstract neuron      |             |
|      |     model hierarchy                             |             |
+------+-------------------------------------------------+-------------+
| 1    | **Reconvene and presentated discussions**       |             |
| 7:30 |                                                 |             |
+------+-------------------------------------------------+-------------+
| 1    | **Close**                                       |             |
| 8:00 |                                                 |             |
+------+-------------------------------------------------+-------------+
:::
:::
:::

::: {#tuesday-13th-march-neuroml-development-workshop-day-2 .section}
#### Tuesday 13th March: NeuroML Development Workshop Day 2[\#](#tuesday-13th-march-neuroml-development-workshop-day-2 "Link to this heading"){.headerlink}

::: {#morning-session-representing-morphologies-support-for-detailed-neuronal-simulators-relationship-to-connectomics-initiatives .section}
##### Morning session: Representing morphologies/support for detailed neuronal simulators/relationship to connectomics initiatives[\#](#morning-session-representing-morphologies-support-for-detailed-neuronal-simulators-relationship-to-connectomics-initiatives "Link to this heading"){.headerlink}

Chair: Michael Hines

::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| Time | Session                                         | Speaker     |
+======+=================================================+=============+
| 0    | **The Neural Tissue                             | James       |
| 9:00 | Simulator([PDF                                  | Kozloski    |
|      | ](https://docs.neuroml.org/_static/NeuroML2012/ |             |
|      | JKozloski_NeuroML_workshop_2012.pdf){.reference |             |
|      | .external})**                                   |             |
|      |                                                 |             |
|      | James presented his work on the Neural Tissue   |             |
|      | Simulator, much of which was contained in the   |             |
|      | recent publication: Kozloski J and Wagner J     |             |
|      | (2011) An ultrascalable solution to large-scale |             |
|      | neural tissue simulation. Frontiers in          |             |
|      | Neuroinformatics. 5:15.                         |             |
|      | ([l                                             |             |
|      | ink](http://www.frontiersin.org/neuroinformatic |             |
|      | s/10.3389/fninf.2011.00015/abstract){.reference |             |
|      | .external}).                                    |             |
|      |                                                 |             |
|      | The key goals of this work are: to develop a    |             |
|      | simulator capable of testing mappings to        |             |
|      | various machine architectures, both parallel    |             |
|      | and multithreaded; to develop support for high  |             |
|      | level, abstract model definitions and           |             |
|      | simulation specifications; and to create an     |             |
|      | extensible simulator, able to map arbitrary,    |             |
|      | domain level models directly to a variety of    |             |
|      | data arrangements and computational             |             |
|      | implementations.                                |             |
|      |                                                 |             |
|      | James discussed the process of defining the     |             |
|      | model (using the Model Definition & Graph       |             |
|      | Specification Languages), how the model         |             |
|      | elements are partitioned on the computing       |             |
|      | resources, and how these elements communicate   |             |
|      | during simulation to solve the model equations. |             |
|      | He discussed the specific case of simulating    |             |
|      | cortical columns when synapses were determined  |             |
|      | through contact detection algorithms. He also   |             |
|      | presented some results for how the simulator    |             |
|      | scales for larger networks. The Neural Tissue   |             |
|      | Simulator is not currently publicly available,  |             |
|      | but James is keen to make it available, and to  |             |
|      | build a community of users. NeuroML support is  |             |
|      | also planned                                    |             |
+------+-------------------------------------------------+-------------+
| 0    | **The Blue Brain Project**                      | Eilif       |
| 9:30 |                                                 | Muller      |
|      | Eilif presented an overview of the Blue Brain   |             |
|      | Project's efforts to reverse engineer a P14 Rat |             |
|      | non-barrel somatosensory cortical column. Based |             |
|      | on a database of anatomical reconstructions,    |             |
|      | electrophysiology, etc. they will fill the      |             |
|      | cortical column with cells based on known       |             |
|      | location, probability distributions.            |             |
|      | Morphologies for those classes of cells are     |             |
|      | taken from library of cells called a collage    |             |
|      | with some rules about how they fit in based on  |             |
|      | constraints from reconstructions. Some of these |             |
|      | cells have been "repaired" due to axon cuts in  |             |
|      | reconstructions.                                |             |
|      |                                                 |             |
|      | Functional circuits are also based on           |             |
|      | biological data. Electrical behaviors are based |             |
|      | on classifications based on firing patterns     |             |
|      | observed in experiments. This is combinatorial  |             |
|      | since each morphological class has a number of  |             |
|      | possible firing patterns. They use genetic      |             |
|      | algorithm to adjust parameters which are set up |             |
|      | based on what is known (gene expression, etc).  |             |
|      | Channelome project uses cell culture and        |             |
|      | automated patch clamp by robot and then         |             |
|      | automated model fitting for data that are then  |             |
|      | posted to Channelpedia. Channels there are      |             |
|      | available in ChannelML. Synaptic                |             |
|      | parameterization and validation for functional  |             |
|      | synapses are also based on database of recorded |             |
|      | synaptic properties.                            |             |
|      |                                                 |             |
|      | In silico model is compared to in vitro using   |             |
|      | same protocols as experiments. There was a      |             |
|      | standards and interoperability discussion: they |             |
|      | are mostly using custom formats other than what |             |
|      | they use with NEURON. Eilif welcomed greater    |             |
|      | support for more widely used standards.         |             |
+------+-------------------------------------------------+-------------+
| 1    | **Tools for the dense reconstruction of         | Moritz      |
| 0:00 | neuronal circuits**                             | H           |
|      |                                                 | elmstaedter |
|      | Moritz gave an overview of his recent work with |             |
|      | Winfried Denk and Frank Briggman, which is      |             |
|      | continuing in his own lab. They have used       |             |
|      | Serial Block-Face Electron Microscopy (SBEM),   |             |
|      | to investigate the connectivity in blocks of    |             |
|      | neuronal tissue, which has been the subject of  |             |
|      | a number of recent publications, e.g. K.L.      |             |
|      | Briggman, Helmstaedter, M. and W. Denk, Wiring  |             |
|      | specificity in the direction-selectivity        |             |
|      | circuit of the mammalian retina. Nature 471,    |             |
|      | 183-188, 2011.                                  |             |
|      | ([link](http://www.nature.com/nature/journ      |             |
|      | al/v471/n7337/full/nature09818.html){.reference |             |
|      | .external}).                                    |             |
|      |                                                 |             |
|      | He also discussed the application               |             |
|      | [K                                              |             |
|      | NOSSOS](http://www.knossostool.org/){.reference |             |
|      | .external} which was developed to facilitate    |             |
|      | the reconstruction of neuronal morphologies     |             |
|      | from such data. While this tool uses a          |             |
|      | proprietary format for storing morphologies, it |             |
|      | is open source and Mortiz was keen to integrate |             |
|      | the application with other tools using NeuroML. |             |
+------+-------------------------------------------------+-------------+
| 1    | **Coffee**                                      |             |
| 0:30 |                                                 |             |
+------+-------------------------------------------------+-------------+
| 1    | **The OpenWorm project: Using NeuroML in a      | Stephen     |
| 1:00 | highly detailed model of C. elegans             | Larson      |
|      | ([PPT](https://docs.neuroml.org/_static         |             |
|      | /NeuroML2012/Open_Worm_03-13-12.ppt){.reference |             |
|      | .external})**                                   |             |
|      |                                                 |             |
|      | Stephen presented the OpenWorm project. This    |             |
|      | ambitious project aims to build an in silico    |             |
|      | model of C. elegans. This well studied system   |             |
|      | with \~1000 cells and 302 identified neurons is |             |
|      | an ideal system with which to attempt a full    |             |
|      | simulation of a living organism down to         |             |
|      | cellular scale. Many different approaches are   |             |
|      | being take in the project at the moment,        |             |
|      | including investigatin Smoothed Particle        |             |
|      | Hydrodynamics (SHP) to allow simulation of the  |             |
|      | interaction of the worm with its environment,   |             |
|      | and creating a new parallel (GPU based)         |             |
|      | simulator in Java which will support physical   |             |
|      | and electrical simulations. Full                |             |
|      | reconstructions of all of the cells are         |             |
|      | available in 3D and those of the neurons have   |             |
|      | been converted to NeuroML, which will form the  |             |
|      | basis of a simulation of the worm's nervous     |             |
|      | system.                                         |             |
|      |                                                 |             |
|      | One of the concrete outcomes of code sprinting  |             |
|      | work on the fringes of the NeuroML/CodeJam      |             |
|      | meeting was the updated OpenWorm Browser, where |             |
|      | the full cellular structure of the worm can be  |             |
|      | browsed in 3D.                                  |             |
+------+-------------------------------------------------+-------------+
| 1    | **Open discussion on tool support for large     |             |
| 1:30 | scale neural simulations**                      |             |
+------+-------------------------------------------------+-------------+
| 1    | **Lunch (provided)**                            |             |
| 3:00 |                                                 |             |
+------+-------------------------------------------------+-------------+
:::
:::

::: {#afternoon-session-best-practices-when-implementing-support-for-neuroml-in-simulators .section}
##### Afternoon session: Best practices when implementing support for NeuroML in simulators[\#](#afternoon-session-best-practices-when-implementing-support-for-neuroml-in-simulators "Link to this heading"){.headerlink}

Chair: Avrama Blackwell

::: pst-scrollable-table-container
+------+-------------------------------------------------+-------------+
| Time | Session                                         | Speaker     |
+======+=================================================+=============+
| 1    | **Introduction to SED-ML                        | Dagmar      |
| 4:00 | ([PDF]                                          | Waltemath   |
|      | (https://docs.neuroml.org/_static/NeuroML2012/D |             |
|      | Waltemath_sed-ml_edinburgh2012_.pdf){.reference |             |
|      | .external})**                                   |             |
|      |                                                 |             |
|      | Dagmar gave an overview of the motivation       |             |
|      | behind the development of SED-ML, the           |             |
|      | Simulation Experiment Description language ,    |             |
|      | the current status of the specification, and    |             |
|      | some of the uses it has been put to so far. It  |             |
|      | compliments a model description in SBML or      |             |
|      | NeuroML and allow specification of the          |             |
|      | simulation algorithm used to run the model, any |             |
|      | changes made to the parameters specified in the |             |
|      | model description, the simulation duration,     |             |
|      | what variables were saved during the            |             |
|      | simulation, and how that data was processed.    |             |
|      |                                                 |             |
|      | In SED-ML you can define a uniform time course  |             |
|      | with an initial time and start and end time.    |             |
|      | This needs to be expanded to other possible     |             |
|      | time courses. Multiple tasks (simulations) can  |             |
|      | be defined. For example, run the original       |             |
|      | database model and the changed model. Output    |             |
|      | can be set up as 2D or 3D plots or a datatable. |             |
|      | SED-ML has an elected board of editors.         |             |
|      | Contribution to SED-ML is encouraged.           |             |
|      | Sourceforge can be used for feature requests    |             |
|      | and this will move forward as people            |             |
|      | contribute.                                     |             |
+------+-------------------------------------------------+-------------+
| 1    | **Introduction to CNO: an ontology for          | Yann le     |
| 4:20 | annotating computational neuroscience models    | Franc       |
|      | ([PDF](https://docs.neuroml.org/_               |             |
|      | static/NeuroML2012/YleFranc_CNO.pdf){.reference |             |
|      | .external})**                                   |             |
|      |                                                 |             |
|      | Yann presented an introduction to CNO: an       |             |
|      | ontology for annotating computational           |             |
|      | neuroscience models. All classes must have a    |             |
|      | unique identifier, a label (name) and a         |             |
|      | human-readable definition. Relationships among  |             |
|      | classes are specified with relations. Examples  |             |
|      | are subsumption relations, associative          |             |
|      | relations, etc. We then can associate this      |             |
|      | semantic information with parts of XML files.   |             |
+------+-------------------------------------------------+-------------+
| 1    | **NeuroLex & NIF update                         | Stephen     |
| 4:40 | ([PPT](https://docs.neuroml.org/_static/NeuroM  | Larson      |
|      | L2012/NeuroLexNIFupdate_3-13-12.ppt){.reference |             |
|      | .external})**                                   |             |
|      |                                                 |             |
|      | Stephen gave us an update on NeuroLex and the   |             |
|      | Neuroscience Information Framework (NIF). We    |             |
|      | need for an online parts list for the brain.    |             |
|      | NeuroLex is built on Wiki technology with extra |             |
|      | functionality to create structured knowledge    |             |
|      | where anyone can create or edit. It currently   |             |
|      | as about 18,000 concepts. NIF funds curators    |             |
|      | from NIH money and also looks for volunteers.   |             |
|      | In the future they want to dominate Google      |             |
|      | searches with NeuroLex terms. Looking to Yelp   |             |
|      | for how they display info including images and  |             |
|      | related queries and such. Some place for        |             |
|      | community comments. Another goal is to expose   |             |
|      | high quality linked data with example of an     |             |
|      | open linked data graph.                         |             |
+------+-------------------------------------------------+-------------+
| 1    | **libSBML and SBML L3                           | Sarah       |
| 5:00 | ([PDF](https://docs.neuroml.org/_static/NeuroM  | Keating     |
|      | L2012/SKeating_libsbml-and-sbml.pdf){.reference |             |
|      | .external})**                                   |             |
|      |                                                 |             |
|      | Sarah gave a brief overview of libSBML and SBML |             |
|      | Level 3. libSBML, which provides an API for     |             |
|      | creating, editing and saving SBML in many       |             |
|      | languages (e.g. C++, Python, Java, Ruby, Perl)  |             |
|      | has been instrumental in the growth of the      |             |
|      | number of applications supporting SBML.         |             |
|      |                                                 |             |
|      | SBML Level 3 has a modular architecture,        |             |
|      | featuring a core specification (roughly in line |             |
|      | with previous SBML releases) and a number of    |             |
|      | specialist packages, which applications can     |             |
|      | choose to support or not. Examples of these     |             |
|      | packages incluse layout for storing the spatial |             |
|      | topology of a model's network diagram, comp for |             |
|      | defining how a model is composed from other     |             |
|      | models and spatial for describing models that   |             |
|      | involve a spatial component. libSBML already    |             |
|      | has a generic framework to support extensions   |             |
|      | for generic packages.                           |             |
+------+-------------------------------------------------+-------------+
| 1    | **Coffee**                                      |             |
| 5:15 |                                                 |             |
+------+-------------------------------------------------+-------------+
| 1    | **Break out sessions**                          |             |
| 5:30 |                                                 |             |
|      | -   How best to map generic model descriptions  |             |
|      |     to a given simulator                        |             |
|      |                                                 |             |
|      | -   Support for morphologies                    |             |
+------+-------------------------------------------------+-------------+
| 1    | **Reconvene and presentated discussions**       |             |
| 7:30 |                                                 |             |
+------+-------------------------------------------------+-------------+
| 1    | **Close of meeting**                            |             |
| 8:00 |                                                 |             |
+------+-------------------------------------------------+-------------+
:::
:::
:::

::: {#wednesday-14th-marchfriday-16th-march .section}
#### Wednesday 14th March--Friday 16th March[\#](#wednesday-14th-marchfriday-16th-march "Link to this heading"){.headerlink}

Full details of the meetings from Wed-Fri are available on the
[NeuralEnsemble.org
webpage](http://neuralensemble.org/meetings/CodeJam5/){.reference
.external} for the meeting.
:::
:::

::: {#funding .section}
### Funding[\#](#funding "Link to this heading"){.headerlink}

The NeuroML part of this workshop was made possible with funding from:

-   The National Institutes of Health

-   Wellcome

-   The UK Neuroinformatics Node

-   The International Neuroinformatics Co-ordinating Facility (INCF)
:::
:::

[]{#document-Events/PastEvents}

::: {#past-neuroml-events .section .tex2jax_ignore .mathjax_ignore}
[]{#neuromlevents-past}

## Past NeuroML Events[\#](#past-neuroml-events "Link to this heading"){.headerlink}

A number of developer workshops and editorial board meetings have been
held since 2008 to coordinate and promote the work of the NeuroML
community. These are listed [[here]{.std
.std-ref}](Landing.html#neuromlevents-meetings){.reference .internal}.

There has been significant NeuroML involvement also at the meetings
organised by the Open Source Brain initiative. See
[here](https://www.opensourcebrain.org/docs#Meetings){.reference
.external} for more details.
:::
:::

::: {.toctree-wrapper .compound}
[]{#document-NeuroMLOrg/CommunicationChannels}

::: {#getting-in-touch .section .tex2jax_ignore .mathjax_ignore}
[]{#contact}

## Getting in touch[\#](#getting-in-touch "Link to this heading"){.headerlink}

We're happy to talk with users, developers and modellers about using
NeuroML in their work.

::: {#mailing-list .section}
[]{#contact-ml}

### Mailing list[\#](#mailing-list "Link to this heading"){.headerlink}

For announcements, general discussion, queries, and troubleshooting
related to NeuroML please use the mailing list:
[https://lists.sourceforge.net/lists/listinfo/neuroml-technology](https://lists.sourceforge.net/lists/listinfo/neuroml-technology){.reference
.external}.
:::

::: {#chat-channels .section}
[]{#contact-chat}

### Chat channels[\#](#chat-channels "Link to this heading"){.headerlink}

A Gitter/Matrix chat channels for queries are also available. One can
access it either via Gitter or Matrix/Element.

-   [Gitter](https://gitter.im/NeuroML/community){.reference .external}

-   [Matrix/Element](https://matrix.to/#/!EQLdKYsJxEfGHAybdP:gitter.im?via=gitter.im&amp;via=matrix.org){.reference
    .external}

Please note that activity in these rooms depends on time zones and the
availability of community members. So, if you do not get a response
soon, please post to the mailing list listed above or file an issue on
GitHub as noted below.
:::

::: {#issues-related-to-the-libraries-or-specification .section}
[]{#contact-issues}

### Issues related to the libraries or specification[\#](#issues-related-to-the-libraries-or-specification "Link to this heading"){.headerlink}

-   Please file general issues related to NeuroML at the
    [NeuroML/NeuroML2 repository on
    GitHub](https://github.com/NeuroML/NeuroML2/issues){.reference
    .external}.

-   Please file issues related to LEMS and jLEMS at the [LEMS/jLEMS
    repository on
    GitHub](https://github.com/LEMS/jLEMS/issues){.reference .external}.

-   Additionally, please file issues related to the different NeuroML
    core tools at their individual [[GitHub repositories]{.std
    .std-ref}](Landing.html#userdocs-software){.reference .internal}.
:::

::: {#social-media .section}
[]{#contact-social}

### Social media[\#](#social-media "Link to this heading"){.headerlink}

You can follow NeuroML related updates on Twitter at
[\@NeuroML](https://twitter.com/NeuroML){.reference .external}.
:::
:::

[]{#document-NeuroMLOrg/Standards}

::: {#overview-of-standards-in-neuroscience .section .tex2jax_ignore .mathjax_ignore}
[]{#initiative-overview-standards}

## Overview of standards in neuroscience[\#](#overview-of-standards-in-neuroscience "Link to this heading"){.headerlink}

In biology, several community standards have been developed to describe
experimental data (e.g. Brain Imaging Data Structure
[BIDS](https://bids.neuroimaging.io/){.reference .external}, Neurodata
Without Borders [NWB](https://nwb.org){.reference .external} and
computational models (e.g. Systems Biology Markup Language
[SBML](https://sbml.org){.reference .external},
[CellML](https://cellml.org){.reference .external}, Scalable Open
Network Architecture TemplAte
[SONATA](https://github.com/AllenInstitute/sonata){.reference
.external}, [PyNN](https://neuralensemble.org/docs/PyNN/){.reference
.external}, and Neural Open Markup Language (NeuroML). These standards
have enabled open and interoperable ecosystems of software applications,
libraries, and databases to emerge, facilitating the sharing of research
outputs, an endeavour encouraged by a growing number of funding agencies
and scientific journals.

::: {#neuroml-as-a-standard .section}
[]{#initiative-overview-standards-neuroml}

### NeuroML as a standard[\#](#neuroml-as-a-standard "Link to this heading"){.headerlink}

NeuroML is an international, collaborative initiative to develop a
language for describing detailed models of neural systems, which will
serve as a standard data format for defining and exchanging descriptions
of neuronal cell and network models.

::: {#endorsed-incf-standard .section}
[]{#initiative-overview-standards-neuroml-incf}

#### Endorsed INCF standard[\#](#endorsed-incf-standard "Link to this heading"){.headerlink}

The mission of INCF is to promote the uptake of FAIR data management
practices in neuroscience through the development of standards and best
practices that support open, FAIR, and citable neuroscience. INCF also
provides training on how standards and best practices facilitate
reproducibility and enable the sharing of data and code.

NeuroML is an [INCF endorsed
standard](https://www.incf.org/sbp/neuroml){.reference .external}.
:::

::: {#combine-standard .section}
[]{#initiative-overview-standards-neuroml-combine}

#### COMBINE standard[\#](#combine-standard "Link to this heading"){.headerlink}

The "COmputational Modeling in BIology NEtwork" (COMBINE) is an
initiative to coordinate the development of the various community
standards and formats for computational models. By doing so, it is
expected that the federated projects will develop a set of interoperable
and non-overlapping standards covering all aspects of modelling in
biology. The global COMBINE effort is led by the COMBINE Coordination
Board.

Building on the experience of mature projects, which already have stable
specifications, software support, user-base and community governance,
COMBINE will help foster or support fledgling efforts aimed at filling
gaps or new needs. As those efforts mature, they may become part of the
core set of COMBINE standards.

One of the initial activities of COMBINE is to coordinate the
organization of scientific and technical events common to several
standards. Those events, as others related to our field of research are
gathered in a calendar.

NeuroML is a [COMBINE official
standard](http://co.mbine.org/standards/neuroml){.reference .external}.
:::
:::
:::

[]{#document-NeuroMLOrg/History}

::: {#a-brief-history-of-neuroml .section .tex2jax_ignore .mathjax_ignore}
[]{#history}

## A brief history of NeuroML[\#](#a-brief-history-of-neuroml "Link to this heading"){.headerlink}

::: {#the-early-days .section}
### The early days[\#](#the-early-days "Link to this heading"){.headerlink}

The concept of NeuroML was first introduced in an article by Goddard et
al. (2001)
[\[[GHH+01](Landing.html#id3 "Nigel H. Goddard, Michael Hucka, Fred Howell, Hugo Cornelis, Kavita Shankar, and David Beeman. Towards NeuroML: model description methods for collaborative modelling in neuroscience. Philosophical Transactions of the Royal Society of London. Series B: Biological Sciences, 356(1412):1209–1228, 8 2001. doi:10.1098/rstb.2001.0910."){.reference
.internal}\]]{#id1}, following meetings at the University of Edinburgh
where initial templates and an overall structure for a model description
language for computational modelling in neuroscience were discussed. The
proposal extended general purpose structures for neuroscience data
proposed by Gardner et al. (2001)
[\[[GKA+01](Landing.html#id4 "D. Gardner, K. H. Knuth, M. Abato, S. M. Erde, T. White, R. DeBellis, and E. P. Gardner. Common data model for neuroscience data and data model exchange. Journal of the American Medical Informatics Association, 8(1):17–33, 1 2001. doi:10.1136/jamia.2001.0080017."){.reference
.internal}\]]{#id2}.

At that time, the design principles for NeuroML were closely linked with
a specific software architecture in which a base application loads a
range of plug-ins to handle different aspects of a simulation
experiment. The simulation platform
[Neosim](http://www.neurogems.org/neosim2/){.reference .external}
provided an implementation of this approach (Howell et al. 2003
[\[[HCG+03](Landing.html#id5 "F. Howell, R. Cannon, N. Goddard, H. Bringmann, P. Rogister, and H. Cornelis. Linking computational neuroscience simulation tools—a pragmatic approach to component-based development. Neurocomputing, 52-54:289–294, 6 2003. doi:10.1016/s0925-2312(02)00781-6."){.reference
.internal}\]]{#id3}), and early NeuroML development was closely aligned
to this architecture. Fred Howell and Robert Cannon developed a software
library, the NeuroML Development Kit (NDK), to simplify the process of
working with XML serializations of models. This library implemented a
particular dialect of XML but did not define particular structures at
the model description level. Instead, Neosim plug-in developers were
free to develop their own structures and serialize them via the NDK, in
the hope that some consensus would emerge around the most useful ones.

In practice, few developers beyond the Edinburgh group developed or used
such structures and the resulting XML was too application specific to
gain wider adoption. The Neosim project was completed in 2005.
:::

::: {#neuroml-v1-x .section}
### NeuroML v1.x[\#](#neuroml-v1-x "Link to this heading"){.headerlink}

Based on discussions with Howell and Cannon about the need to develop a
consensus for describing widely used model components, Sharon Crook
worked with the neuroanatomy community on a language for describing
neuronal morphologies in XML, **MorphML** (Qi and Crook 2004
[\[[QC04](Landing.html#id6 "Weihong Qi and Sharon Crook. Tools for neuroinformatic data exchange: an XML application for neuronal morphology data. Neurocomputing, 58-60:1091–1095, 6 2004. doi:10.1016/j.neucom.2004.01.171."){.reference
.internal}\]]{#id4}). At the same time, Padraig Gleeson, working with
Angus Silver, was developing [[neuroConstruct]{.std
.std-ref}](Landing.html#userdocs-supporting-apps-neuroconstruct){.reference
.internal}, for generating neuronal simulations for the NEURON and
GENESIS simulators (Gleeson et al. 2007
[\[[GSS07](Landing.html#id7 "Padraig Gleeson, Volker Steuber, and R. Angus Silver. neuroConstruct: a tool for modeling networks of neurons in 3d space. Neuron, 54(2):219–235, 4 2007. doi:10.1016/j.neuron.2007.03.025."){.reference
.internal}\]]{#id5}), which had its own internal simulator independent
representation for morphologies, channel and networks.

It was agreed that these efforts should be merged under the banner of
NeuroML, and the [[v1.x structure of NeuroML]{.std
.std-ref}](Landing.html#userdocs-neuromlv1){.reference .internal} was
created. A modular approach containing **MorphML**, **ChannelML** and
**NetworkML** was adopted to allow application developers to support
only those parts of the language needed by their application (Crook et
al. 2007
[\[[CGH+07](Landing.html#id8 "Sharon Crook, Padraig Gleeson, Fred Howell, Joseph Svitak, and R. Angus Silver. MorphML: level 1 of the NeuroML standards for neuronal morphology data and model specification. Neuroinformatics, 5(2):96–104, 5 2007. doi:10.1007/s12021-007-0003-6."){.reference
.internal}\]]{#id6}, Gleeson et al. 2010
[\[[GCC+10](Landing.html#id9 "Padraig Gleeson, Sharon Crook, Robert C. Cannon, Michael L. Hines, Guy O. Billings, Matteo Farinella, Thomas M. Morse, Andrew P. Davison, Subhasis Ray, Upinder S. Bhalla, Simon R. Barnes, Yoana D. Dimitrova, and R. Angus Silver. NeuroML: a language for describing data driven models of neurons and networks with a high degree of biological detail. PLoS Computational Biology, 6(6):e1000815, 2010. doi:10.1371/journal.pcbi.1000815."){.reference
.internal}\]]{#id7}). XML schema files for this version of the standard
have been available since 2006. The motivation, structure and
functionality of this version is described in detail in Gleeson et al.
2010, while the specification of the language is outlined in the
[Supporting
Information](http://www.ploscompbiol.org/article/info:doi/10.1371/journal.pcbi.1000815#s5){.reference
.external} of that publication.

For converting NeuroML v1 models/files to NeuroML2, users can use
[[neuroConstruct]{.std
.std-ref}](Landing.html#userdocs-supporting-apps-neuroconstruct){.reference
.internal}.
:::

::: {#neuroml-v2-x-introducing-lems .section}
### NeuroML v2.x - introducing LEMS...[\#](#neuroml-v2-x-introducing-lems "Link to this heading"){.headerlink}

NeuroML2 development was started in 2011. The main motivation for
NeuroML2 was the lack of extensibility of NeuroML v1.x; every new model
type which was introduced into the language required an update to the
Schema, updates to the text documentation and an implementation in each
of the native formats of the target simulators. NeuroML2 is built on the
**LEMS (Low Entropy Model Specification) language**, which allows
machine readable definitions of the cell, channel and synapse models
which form the core of the language. This increases transparency of
model structure and dynamics and facilitates automatic mapping of the
models to multiple simulation formats. More details on the structure of
LEMS and how it is used in NeuroML2 can be found in Cannon et al. 2014
[\[[CGC+14](Landing.html#id10 "Robert C. Cannon, Padraig Gleeson, Sharon Crook, Gautham Ganapathy, Boris Marin, Eugenio Piasini, and R. Angus Silver. LEMS: a language for expressing complex biological models in concise and hierarchical form and its use in underpinning NeuroML 2. Frontiers in Neuroinformatics, 2014. doi:10.3389/fninf.2014.00079."){.reference
.internal}\]]{#id8} and [[here]{.std
.std-ref}](Landing.html#userdocs-specification){.reference .internal}.

In parallel with development of NeuroML2 and LEMS, software libraries
for reading, writing and running simulations using the languages are
under active development in Java ([[jNeuroML]{.std
.std-ref}](Landing.html#jneuroml){.reference .internal}) and Python
([[libNeuroML]{.std .std-ref}](Landing.html#libneuroml){.reference
.internal} and [[pyLEMS]{.std .std-ref}](Landing.html#pylems){.reference
.internal} (see Vella et al. 2014
[\[[VCC+14](Landing.html#id11 "Michael Vella, Robert C. Cannon, Sharon Crook, Andrew P. Davison, Gautham Ganapathy, Hugh P. C. Robinson, R. Angus Silver, and Padraig Gleeson. Libneuroml and pylems: using python to combine procedural and declarative modeling approaches in computational neuroscience. Frontiers in neuroinformatics, 8:38, 2014. doi:10.3389/fninf.2014.00038."){.reference
.internal}\]]{#id9}) and [[pyNeuroML]{.std
.std-ref}](Landing.html#pyneuroml){.reference .internal}).

The NeuroML specifications are developed by the [[NeuroML Editorial
Board]{.std .std-ref}](Landing.html#neuromlorg-board){.reference
.internal} and overseen by its [[Scientific Committee]{.std
.std-ref}](Landing.html#neuromlorg-scientificcommittee){.reference
.internal}. NeuroML specifications and the associated libraries are
developed on GitHub and an overview of current activities can be found
[here](https://github.com/NeuroML/NeuroML2/projects/1){.reference
.external}.

::: {.admonition-recent-releases-of-neuroml2 .admonition}
Recent releases of NeuroML2

For full details on the recent releases of NeuroML see:
[here](https://github.com/NeuroML/NeuroML2/blob/master/HISTORY.md){.reference
.external}.
:::
:::

::: {#the-future .section}
### The future[\#](#the-future "Link to this heading"){.headerlink}

[[NeuroMLlite]{.std .std-ref}](Landing.html#neuromllite){.reference
.internal} is under active development, which will significantly enhance
the range of network models which can be expressed (in a concise JSON
based format) and run in NeuroML supporting simulators. This work will
form the basis of NeuroML v3.0.
:::
:::

[]{#document-NeuroMLOrg/Board}

::: {#neuroml-editorial-board .section .tex2jax_ignore .mathjax_ignore}
[]{#neuromlorg-board}

## NeuroML Editorial Board[\#](#neuroml-editorial-board "Link to this heading"){.headerlink}

An elected board of editors has been formed to manage the NeuroML
specification development process. The editorial board consists of five
members, elected by the NeuroML community. The editors are responsible
for producing and maintaining the official documentation for the NeuroML
language, and work in collaboration with the [[Scientific
Committee]{.std
.std-ref}](Landing.html#neuromlorg-scientificcommittee){.reference
.internal} who provide oversight and guidance.

Due to the close link between the development of NeuroML 2 and LEMS,
this group is also responsible for producing a stable specification for
the subset of LEMS used by NeuroML 2.

::: {#current-editorial-board .section}
### Current Editorial Board[\#](#current-editorial-board "Link to this heading"){.headerlink}

The five current members of the NeuroML Editorial Board are:

-   Padraig Gleeson

-   Boris Marin

-   Sotirios Panagiotou

-   Subhasis Ray

-   Ankur Sinha

Ankur Sinha and Subhasis Ray were elected for three year terms in 2024
(2025--2027). Padraig Gleeson, Boris Marin and Sotirios Panagiotou were
elected for three year terms in 2022 (2023--2025).

::: container-fluid
::: {.row .my-2 .py-2}
::: {.col-sm-4 .px-2}
[![Sotirios
Panagiotou](_images/sotirios.jpg){style="height: 150px;"}](_images/sotirios.jpg){.reference
.internal .image-reference}

**Sotirios Panagiotou**\
PhD candidate\
Erasmus Medical Center\
Rotterdam, Netherlands\
[Website](https://scholar.google.com/citations?user=8XdZ5jUAAAAJ&amp;hl=en&amp;oi=ao){.reference
.external}
:::

::: {.col-sm-4 .px-2}
[![Padraig
Gleeson](_images/padraig2.jpeg){style="height: 150px;"}](_images/padraig2.jpeg){.reference
.internal .image-reference}

**Padraig Gleeson**\
University College London\
UK\
[Website](http://www.opensourcebrain.org/users/4){.reference .external}
:::

::: {.col-sm-4 .px-2}
[![Boris
Marin](_images/boris.jpg){style="height: 150px;"}](_images/boris.jpg){.reference
.internal .image-reference}

**Boris Marin**\
Universidade Federal do ABC\
Brazil\
[Website](https://www.opensourcebrain.org/users/67){.reference
.external}
:::

::: {.col-sm-4 .px-2}
[![Subhasis
Ray](_images/subhasis.jpg){style="height: 150px;"}](_images/subhasis.jpg){.reference
.internal .image-reference}

**Subhasis Ray**\
TCG CREST\
Kolkata, India\
[Website](https://rayslab.github.io/){.reference .external}
:::

::: {.col-sm-4 .px-2}
[![Ankur
Sinha](_images/ankur.png){style="height: 150px;"}](_images/ankur.png){.reference
.internal .image-reference}

**Ankur Sinha**\
University College London\
UK\
[Website](https://ankursinha.in){.reference .external}
:::
:::
:::

Information on past editors and elections can be found [[here]{.std
.std-ref}](Landing.html#neuromlorg-boardhistory){.reference .internal}.
:::

::: {#procedures .section}
### Procedures[\#](#procedures "Link to this heading"){.headerlink}

The procedures for election of the editorial board, its
responsibilities, size and activities were heavily inspired by other
initiatives like
[SBML](http://sbml.org/Documents/SBML_Development_Process#The_SBML_Editors){.reference
.external}, that have had successful editorial teams for many years. The
[COMBINE initiative](http://co.mbine.org/standards){.reference
.external} seeks to promote community developed standards in
computational biology, and the NeuroML editorial board will work with
this initiative to ensure best practices in specification preparation.
:::

::: {#responsibilities-of-neuroml-editors .section}
### Responsibilities of NeuroML Editors[\#](#responsibilities-of-neuroml-editors "Link to this heading"){.headerlink}

Their main responsibilities are:

-   Defining and documenting the procedure for specification production:
    scope, release frequency, update procedures, form of specification
    (web based or single pdf, etc.). This should be based on some or all
    of the recommendations for community based standards development
    from the COMBINE initiative. These procedures for specification
    production will have to be agreed with the [[Scientific
    Committee]{.std
    .std-ref}](Landing.html#neuromlorg-scientificcommittee){.reference
    .internal}.

-   Preparing the core specification for the NeuroML language.

-   Testing reference implementations of NeuroML compliant applications.

-   Preparing a specification for the LEMS language. This can be a
    subset of the language supported by the reference implementations in
    Java (jLEMS) and Python (pyLEMS), but will have to cover all of the
    LEMS elements required to specify the ComponentType definitions for
    NeuroML 2.

-   Responding to community queries about the specification.

-   Establishing a procedure for incorporating major changes into the
    specification (in cooperation with the [[Scientific Committee]{.std
    .std-ref}](Landing.html#neuromlorg-scientificcommittee){.reference
    .internal}).

Participation in the editorial board will be on a volunteer basis, there
is no central funding to support this work.
:::

::: {.toctree-wrapper .compound}
[]{#document-NeuroMLOrg/BoardHistory}

::: {#history-of-the-neuroml-editorial-board .section .tex2jax_ignore .mathjax_ignore}
[]{#neuromlorg-boardhistory}

### History of the NeuroML Editorial Board[\#](#history-of-the-neuroml-editorial-board "Link to this heading"){.headerlink}

This page documents the previous members of the NeuroML Board.

::: {#initial-election-of-editors-2013 .section}
#### Initial election of editors (2013)[\#](#initial-election-of-editors-2013 "Link to this heading"){.headerlink}

The first election of an editorial board for NeuroML took place in
May/June 2013.

-   The electorate consisted of the members of the NeuroML mailing lists
    on 3rd May 2013.

-   Anyone on these lists could nominate someone to be an editor. Self
    nominations were also allowed.

-   The three candidates who received the highest number of votes will
    serve three year terms and the two with the next highest number of
    votes will serve two year terms.

-   Nicolas Le Novère (lenov -at-
    [babraham.ac.uk](http://babraham.ac.uk){.reference .external}) was
    the returning officer for this initial election.
:::

::: {#election-of-editors-2015 .section}
#### Election of editors (2015)[\#](#election-of-editors-2015 "Link to this heading"){.headerlink}

The second election of an editorial board for NeuroML took place in
June/July 2015.

-   The electorate consisted of the members of the NeuroML mailing lists
    on 18 June 2015.

-   Anyone on these lists could nominate someone to be an editor. Self
    nominations were also allowed.

-   The two candidates who received the highest number of votes would
    serve three year terms.

-   Nicolas Le Novère (lenov -at-
    [babraham.ac.uk](http://babraham.ac.uk){.reference .external}) was
    the returning officer for this election.

-   Results were announced
    [here](https://sourceforge.net/p/neuroml/mailman/message/34331970/){.reference
    .external}.
:::

::: {#election-of-editors-2016 .section}
#### Election of editors (2016)[\#](#election-of-editors-2016 "Link to this heading"){.headerlink}

The third election of an editorial board for NeuroML took place in
July/August 2016.

-   The electorate consisted of the members of the NeuroML mailing lists
    on 18 July 2016.

-   Anyone on these lists could nominate someone to be an editor. Self
    nominations were also allowed.

-   The three candidates who received the highest number of votes would
    serve three year terms.

-   Nicolas Le Novère (lenov -at-
    [babraham.ac.uk](http://babraham.ac.uk){.reference .external}) was
    the returning officer for this election.
:::

::: {#election-of-editors-2018 .section}
#### Election of editors (2018)[\#](#election-of-editors-2018 "Link to this heading"){.headerlink}

The fourth election of an editorial board for NeuroML took place in
Nov/Dec 2018.

-   The electorate consisted of the members of the NeuroML mailing lists
    as well as anyone who had made significant contributions to any of
    the NeuroML GitHub repositories in the past 3 years.

-   Anyone on the electorate could nominate someone to be an editor.
    Self nominations were also allowed.

-   The two candidates who received the highest number of votes would
    serve three year terms.

-   Salvador Dura-Bernal was elected outright on the first round of
    voting and Andrew Davison was elected in a run off between the two
    next highest placed candidates who received the same number of
    votes.

-   Malin Sandstrom at the INCF was the returning officer for this
    election.
:::

::: {#election-of-editors-2019 .section}
#### Election of editors (2019)[\#](#election-of-editors-2019 "Link to this heading"){.headerlink}

The fifth election of an editorial board for NeuroML took place in
Nov/Dec 2019.

-   The electorate consisted of the members of the NeuroML mailing lists
    as well as anyone who had made significant contributions to any of
    the NeuroML GitHub repositories in the past 3 years.

-   Anyone on the electorate could nominate someone to be an editor.
    Self nominations were also allowed.

-   The three candidates who received the highest number of votes would
    serve three year terms.

-   Padraig Gleeson, Boris Marin and Justas Birgiolas were nominated,
    and eventually all elected to serve as editors.

-   Malin Sandstrom at the INCF was the returning officer for this
    election.
:::

::: {#election-of-editors-2022 .section}
#### Election of editors (2022)[\#](#election-of-editors-2022 "Link to this heading"){.headerlink}

The sixth election of an editorial board for NeuroML took place in
Nov/Dec 2021.

-   The electorate consisted of the members of the NeuroML mailing lists
    as well as anyone who had made significant contributions to any of
    the NeuroML GitHub repositories in the past 3 years.

-   Anyone on the electorate could nominate someone to be an editor.
    Self nominations were also allowed.

-   The three candidates who received the highest number of votes would
    serve three year terms.

-   Salvador Dura-Bernal and Ankur Sinha were nominated, and eventually
    all elected to serve as editors.

-   Sharon Crook was the returning officer for this election.
:::

::: {#election-of-editors-2024 .section}
#### Election of editors (2024)[\#](#election-of-editors-2024 "Link to this heading"){.headerlink}

The seventh election of an editorial board for NeuroML took place in
Nov/Dec 2023.

-   The electorate consisted of the members of the NeuroML mailing lists
    as well as anyone who had made significant contributions to any of
    the NeuroML GitHub repositories in the past 3 years.

-   Anyone on the electorate could nominate someone to be an editor.
    Self nominations were also allowed.

-   The three candidates who received the highest number of votes would
    serve three year terms.

-   Padraig Gleeson, Boris Marin, Sotirios Panagiotou and Subhasis Ray
    were nominated. Padraig Gleeson, Boris Marin, Sotirios Panagiotou
    received highest votes and were elected to serve as editors.

-   Cengiz Gunay was the returning officer for this election.
:::

::: {#election-of-editors-2025 .section}
#### Election of editors (2025)[\#](#election-of-editors-2025 "Link to this heading"){.headerlink}

The eight election of an editorial board for NeuroML took place in
Nov/Dec 2024.

-   The electorate consisted of the members of the NeuroML mailing lists
    as well as anyone who had made significant contributions to any of
    the NeuroML GitHub repositories in the past 3 years.

-   Anyone on the electorate could nominate someone to be an editor.
    Self nominations were also allowed.

-   The two candidates who received the highest number of votes would
    serve three year terms.

-   Ankur Sinha and Subhasis Ray were nominated. Ankur Sinha and
    Subhasis were both elected to serve as editors.

-   Cengiz Gunay was the returning officer for this election.
:::
:::

[]{#document-NeuroMLOrg/BoardMeetingReports}

::: {#workshop-and-meeting-reports .section .tex2jax_ignore .mathjax_ignore}
[]{#neuromlevents-meetings}

### Workshop and Meeting reports[\#](#workshop-and-meeting-reports "Link to this heading"){.headerlink}

Information and minutes of various NeuroML meetings can be found here.

::: pst-scrollable-table-container
+-------------+-------------+------------------------------------------+
| Meeting     | Location    | Summary                                  |
+=============+=============+==========================================+
| 2021        | Online at   | A NeuroML development workshop was held  |
| NeuroML     | COMBINE     | as part of the annual COMBINE meeting in |
| Development | 2021        | October 2021.\                           |
| workshop &  |             | [[More info]{.std                        |
| Editorial   |             | .std-ref}](Landing.html                  |
| Board       |             | #neuromlevents-202109combine){.reference |
| Meeting     |             | .internal}.                              |
+-------------+-------------+------------------------------------------+
| 2019        | CNS\*2019   | The fifth NeuroML Editorial Board        |
| NeuroML     | Meeting,    | Meeting took place at the CNS meeting in |
| Editorial   | Barcelona   | Barcelona, Monday 15th July, 2019.\      |
| Board       |             | [Minutes](https:/                        |
| Meeting     |             | /docs.neuroml.org/_static/files/NeuroMLE |
|             |             | ditorialBoardMeeting2019.pdf){.reference |
|             |             | .external}.                              |
+-------------+-------------+------------------------------------------+
| 2018        | Online      | The fourth NeuroML Editorial Board       |
| NeuroML     |             | Meeting took place via video conference  |
| Editorial   |             | on 6th July 2018 between the NeuroML     |
| Board       |             | Editorial Board and interested members   |
| Meeting     |             | of the community to get an update on all |
|             |             | current NeuroML related activities.\     |
|             |             | [Minutes](https:/                        |
|             |             | /docs.neuroml.org/_static/files/NeuroMLE |
|             |             | ditorialBoardMeeting2018.pdf){.reference |
|             |             | .external}.                              |
+-------------+-------------+------------------------------------------+
| 2016        | Janelia     | The third NeuroML Editorial Board        |
| NeuroML     | Research    | Meeting took place after the             |
| Editorial   | Campus,     | **Collaborative Development of           |
| Board       | Virginia,   | Data-Driven Models of Neural Systems**   |
| Meeting     | USA         | conference held at Janelia Research      |
|             |             | Campus in Sept 2016. More details on the |
|             |             | main conference can be found             |
|             |             | [here]                                   |
|             |             | (https://www.janelia.org/you-janelia/con |
|             |             | ferences/collaborative-development-data- |
|             |             | driven-models-neural-systems){.reference |
|             |             | .external}.\                             |
|             |             | [Minutes](https:/                        |
|             |             | /docs.neuroml.org/_static/files/NeuroMLE |
|             |             | ditorialBoardMeeting2016.pdf){.reference |
|             |             | .external}.                              |
+-------------+-------------+------------------------------------------+
| 2015        | Alghero,    | The second NeuroML Editorial Board       |
| NeuroML     | Sardinia,   | Meeting took place prior to the Open     |
| Editorial   | Italy       | Source Brain 2015 meeting held in        |
| Board       |             | Sardinia. More details on the main       |
| Meeting @   |             | meeting can be found                     |
| OSB 2015    |             | [here](http://opensourcebrain.org        |
|             |             | /docs/Help/Meetings#OSB_2015){.reference |
|             |             | .external}.\                             |
|             |             | [Minutes](https:/                        |
|             |             | /docs.neuroml.org/_static/files/NeuroMLE |
|             |             | ditorialBoardMeeting2015.pdf){.reference |
|             |             | .external}.                              |
+-------------+-------------+------------------------------------------+
| 2014        | Alghero,    | The first official NeuroML Editorial     |
| NeuroML     | Sardinia,   | Board Meeting took place prior to the    |
| Editorial   | Italy       | Open Source Brain 2014 meeting held in   |
| Board       |             | Sardinia. More details on the main       |
| Meeting @   |             | meeting can be found                     |
| OSB 2014    |             | [here](http://opensourcebrain.org        |
|             |             | /docs/Help/Meetings#OSB_2014){.reference |
|             |             | .external}.\                             |
|             |             | [Minutes](https:/                        |
|             |             | /docs.neuroml.org/_static/files/NeuroMLE |
|             |             | ditorialBoardMeeting2014.pdf){.reference |
|             |             | .external}.                              |
+-------------+-------------+------------------------------------------+
| 2013        | Alghero,    | The NeuroML Development Workshop was     |
| NeuroML     | Sardinia,   | merged into the Open Source Brain        |
| Meeting     | Italy       | kickoff meeting in Alghero, Sardinia.    |
| Development |             | More details on this meeting can be      |
| Workshop @  |             | found                                    |
| OSB 2013    |             | [here](http://www.opensourcebrain.or     |
|             |             | g/projects/osb/wiki/Meetings){.reference |
|             |             | .external}. Discussions on the state of  |
|             |             | NeuroML and future developments took     |
|             |             | place during the main meeting.           |
+-------------+-------------+------------------------------------------+
| 2012        | Informatics | The NeuroML workshop at was combined     |
| NeuroML     | Forum,      | with the                                 |
| Development | Edinburgh,  | [BrainScaleS](http://brains              |
| Workshop    | UK          | cales.kip.uni-heidelberg.de/){.reference |
|             |             | .external} (previously FACETS) CodeJam   |
|             |             | meeting.\                                |
|             |             | [[Minutes]{.std                          |
|             |             | .std-ref}](Landing.html                  |
|             |             | #neuromlevents-2012edinburgh){.reference |
|             |             | .internal}.                              |
+-------------+-------------+------------------------------------------+
| 2011        | University  | A key outcome of third NeuroML           |
| NeuroML     | College     | Development Workshop was the creation of |
| Development | London, UK  | a Scientific Committee for NeuroML.\     |
| Workshop    |             | [Min                                     |
|             |             | utes](https://docs.neuroml.org/_static/f |
|             |             | iles/NeuroMLWorkshop2011.pdf){.reference |
|             |             | .external}.                              |
+-------------+-------------+------------------------------------------+
| 2010        | Arizona     | The second NeuroML Development Workshop  |
| NeuroML     | State       | was held in Arizona State University to  |
| Development | University, | plan for version 2.0 of the NeuroML      |
| Workshop    | USA         | model description language. There was    |
|             |             | also a Symposium on Multiscale           |
|             |             | Approaches to Understanding Neural       |
|             |             | Plasticity held at ASU before the main   |
|             |             | meeting and a number of tutorials on     |
|             |             | software for multiscale modeling given   |
|             |             | by the meeting participants on the       |
|             |             | following day.\                          |
|             |             | [Min                                     |
|             |             | utes](https://docs.neuroml.org/_static/f |
|             |             | iles/NeuroMLWorkshop2010.pdf){.reference |
|             |             | .external}.                              |
+-------------+-------------+------------------------------------------+
| 2009        | University  | The focus of the workshop was to refine  |
| NeuroML     | College     | the specifications for describing models |
| Development | London, UK  | of channel kinetics and the biophysical  |
| Workshop    |             | properties of cells. Special thanks to   |
|             |             | the Wellcome Trust, the INCF, and the    |
|             |             | NSF for their generous support of this   |
|             |             | endeavour.\                              |
|             |             | [Min                                     |
|             |             | utes](https://docs.neuroml.org/_static/f |
|             |             | iles/NeuroMLWorkshop2009.pdf){.reference |
|             |             | .external}.                              |
+-------------+-------------+------------------------------------------+
| 2008 CNS    | Portland,   | Padraig Gleeson and Sharon Crook         |
| Workshop    | Oregon, USA | moderated a workshop on                  |
|             |             | "Interoperability of Software for        |
|             |             | Computational and Experimental           |
|             |             | Neuroscience" at the 2008 Computational  |
|             |             | Neuroscience Meeting.                    |
+-------------+-------------+------------------------------------------+
:::
:::
:::
:::

[]{#document-NeuroMLOrg/ScientificCommittee}

::: {#neuroml-scientific-committee .section .tex2jax_ignore .mathjax_ignore}
[]{#neuromlorg-scientificcommittee}

## NeuroML Scientific Committee[\#](#neuroml-scientific-committee "Link to this heading"){.headerlink}

The responsibilities of the NeuroML Scientific Committee are:

-   To advise on the scientific focus of the NeuroML initiative; to
    ensure that the structure of the language is based on the latest
    knowledge of neuronal anatomy and physiology.

-   To agree on the technical implementation for the core specifications
    (in collaboration with the [[NeuroML Editorial Board]{.std
    .std-ref}](Landing.html#neuromlorg-board){.reference .internal}) and
    to ensure that best practices are encouraged in model specification.

-   To promote NeuroML internationally, both the core specifications and
    the tools which support the language.

-   To define the governance structure of the NeuroML Initiative and
    outline a path towards a specification process with dedicated,
    elected editors.

-   To engage with other standardisation and databasing initiatives in
    the computational neuroscience and wider biology fields.

-   To review and agree on extensions to the core specifications and the
    scope of the initiative; to address issues the community raises
    regarding the direction of the initiative.

::: {#current-members .section}
### Current Members[\#](#current-members "Link to this heading"){.headerlink}

::: container-fluid
::: {.row .my-2 .py-2}
::: {.col-sm-4 .px-2}
[![Upi
Bhalla](_images/bhalla.png){style="height: 150px;"}](_images/bhalla.png){.reference
.internal .image-reference}

**Upi Bhalla**\
NCBS\
Bangalore, India\
[Website](http://www.ncbs.res.in/bhalla){.reference .external}
:::

::: {.col-sm-4 .px-2}
[![Avrama
Blackwell](_images/avrama.jpg){style="height: 150px;"}](_images/avrama.jpg){.reference
.internal .image-reference}

**Avrama Blackwell**\
Krasnow Institute of Advanced Studies\
George Mason University, USA\
[Website](http://krasnow1.gmu.edu/CENlab/avrama.html){.reference
.external}
:::

::: {.col-sm-4 .px-2}
[![Hugo
Cornells](_images/hugo.png){style="height: 150px;"}](_images/hugo.png){.reference
.internal .image-reference}

**Hugo Cornells**\
K.U. Leuven\
Belgium\
[Website](http://neurospaces.sourceforge.net/){.reference .external}
:::

::: {.col-sm-4 .px-2}
[![Sharon
Crook](_images/rsz_crook.jpg){style="height: 150px;"}](_images/rsz_crook.jpg){.reference
.internal .image-reference}

**Sharon Crook**\
Arizona State University\
USA\
[Website](https://iconlab.asu.edu/crook.html){.reference .external}
:::

::: {.col-sm-4 .px-2}
[![Andrew
Davison](_images/andrew.jpg){style="height: 150px;"}](_images/andrew.jpg){.reference
.internal .image-reference}

**Andrew Davison**\
CNRS, Gif-sur-Yvette\
France\
[Website](http://www.andrewdavison.info/){.reference .external}
:::

::: {.col-sm-4 .px-2}
[![Salvador
Dura-Bernal](_images/salva.png){style="height: 150px;"}](_images/salva.png){.reference
.internal .image-reference}

**Salvador Dura-Bernal**\
SUNY Downstate\
Brooklyn, USA\
[Website](http://www.neurosimlab.com/users/salvador-dura-bernal){.reference
.external}
:::

::: {.col-sm-4 .px-2}
[![Robert
McDougal](_images/robertmcdougal.png){style="height: 150px;"}](_images/robertmcdougal.png){.reference
.internal .image-reference}

**Robert McDougal**\
Yale University\
USA\
[Website](https://medicine.yale.edu/lab/shepherd/profile/robert_mcdougal/){.reference
.external}
:::

::: {.col-sm-4 .px-2}
[![Lyle
Graham](_images/lyle.png){style="height: 150px;"}](_images/lyle.png){.reference
.internal .image-reference}

**Lyle Graham**\
Université Paris Descartes\
Paris, France\
[Website](http://lyle.neurophysics.eu/){.reference .external}
:::

::: {.col-sm-4 .px-2}
[![Cengiz
Gunay](_images/cgunay.JPG){style="height: 150px;"}](_images/cgunay.JPG){.reference
.internal .image-reference}

**Cengiz Gunay**\
Georgia Gwinnett College\
USA\
[Website](https://www.ggc.edu/about-ggc/directory/cengiz-gunay){.reference
.external}
:::

::: {.col-sm-4 .px-2}
[![Michael
Hines](_images/michael.png){style="height: 150px;"}](_images/michael.png){.reference
.internal .image-reference}

**Michael Hines**\
Yale University\
USA\
[Website](http://www.neuron.yale.edu/neuron/credits){.reference
.external}
:::

::: {.col-sm-4 .px-2}
[![Angus
Silver](_images/angus.png){style="height: 150px;"}](_images/angus.png){.reference
.internal .image-reference}

**Angus Silver**\
University College London\
London, UK\
[Website](https://silverlab.org){.reference .external}
:::
:::
:::
:::

::: {#past-members .section}
### Past Members[\#](#past-members "Link to this heading"){.headerlink}

(Note: past members who are currently members of the [[NeuroML Editorial
Board]{.std .std-ref}](Landing.html#neuromlorg-board){.reference
.internal} are not listed.)

-   Robert Cannon
:::
:::

[]{#document-NeuroMLOrg/Funding}

::: {#funding-and-acknowledgements .section .tex2jax_ignore .mathjax_ignore}
[]{#neuromlorg-funding}

## Funding and Acknowledgements[\#](#funding-and-acknowledgements "Link to this heading"){.headerlink}

The NeuroML effort has been made possible by funding from research
councils in the UK, EU, and the USA.

  --------------------------------------------------------------------------------------------------------------------------------- ------------------------------------------------------------------------------------------------------------
                                   ![UK Medical Research Council](_images/mrc.jpg){.align-center}                                   [UK Medical Research Council](http://www.mrc.ac.uk/){.reference .external}
                   ![UK Biotechnology and Biological Sciences Research Council](_images/bbsrc.gif){.align-center}                   [UK Biotechnology and Biological Sciences Research Council](http://www.bbsrc.ac.uk/){.reference .external}
                                  ![National Institutes of Health](_images/nih.gif){.align-center}                                  [National Institutes of Health](http://www.nimh.nih.gov/){.reference .external}
                                     ![EU Synapse Project](_images/EUS_200px.gif){.align-center}                                    [EU Synapse Project](http://www.eusynapse.mpg.de/){.reference .external}
                                   ![National Science Foundation](_images/nsf.gif){.align-center}                                   [National Science Foundation](http://nsf.gov/){.reference .external}
                      ![International Neuroinformatics Coordinating Facility](_images/incf.png){.align-center}                      [International Neuroinformatics Coordinating Facility](http://incf.org/){.reference .external}
   [![Wellcome](_images/wtlogo.png){.align-center style="width: 30%;"}](_images/wtlogo.png){.reference .internal .image-reference}  [Wellcome](http://www.wellcome.ac.uk/){.reference .external}
  --------------------------------------------------------------------------------------------------------------------------------- ------------------------------------------------------------------------------------------------------------
:::

[]{#document-NeuroMLOrg/OutreachTraining}

::: {#outreach-and-training .section .tex2jax_ignore .mathjax_ignore}
[]{#neuromlorg-outreachtraining}

## Outreach and training[\#](#outreach-and-training "Link to this heading"){.headerlink}

The NeuroML community has a strong record of participating in training
and outreach activities. Information on tutorials and workshops can be
seen in the Events pages.

::: {#google-summer-of-code .section}
[]{#neuromlorg-outreachtraining-gsoc}

### Google summer of Code[\#](#google-summer-of-code "Link to this heading"){.headerlink}

The NeuroML community participates annually in the [Google Summer of
Code](https://summerofcode.withgoogle.com){.reference .external} under
the [INCF](https://incf.org){.reference .external} organisation.
Projects are centred around the standardisation of published models in
NeuroML to make these standardised versions available on the [Open
Source Brain](https://opensourcebrain.org){.reference .external}
platform and improving the NeuroML tools wherever possible.

::: {#neuromlorg-outreachtraining-gsoc-2025 .section}
[]{#id1}

#### 2025[\#](#neuromlorg-outreachtraining-gsoc-2025 "Link to this heading"){.headerlink}

-   Hengye Zhu: [Developing standardised biophysically detailed neuronal
    circuit models using NeuroML/PyNN/Open Source
    Brain](https://github.com/OpenSourceBrain/Macaque_auditory_thalamocortical_model_data/tree/feat-neuroml-gsoc/NeuroML2){.reference
    .external}.
:::

::: {#neuromlorg-outreachtraining-gsoc-2024 .section}
[]{#id2}

#### 2024[\#](#neuromlorg-outreachtraining-gsoc-2024 "Link to this heading"){.headerlink}

-   Aditya Pandey: [Implementation of an SWC to NeuroML converter in
    PyNeuroML](https://github.com/AdityaPandeyCN/GSOC2024_INCF){.reference
    .external}.

-   Ioannis Daras: [Incorporation of new features into an advanced cross
    platform 3D viewer for NeuroML cells and
    networks](https://github.com/lej0hn/GSOC_2024){.reference
    .external}.
:::

::: {#neuromlorg-outreachtraining-gsoc-2022 .section}
[]{#id3}

#### 2022[\#](#neuromlorg-outreachtraining-gsoc-2022 "Link to this heading"){.headerlink}

-   Anuja Negi: [Simulating multiscale models of the mouse visual cortex
    in
    NeuroML](https://github.com/OpenSourceBrain/AllenInstituteNeuroML/blob/master/GSoC22.md){.reference
    .external}

-   Shayan Shafquat: [Open source, cross simulator, large scale cortical
    models in
    NeuroML](https://github.com/OpenSourceBrain/BahlEtAl2012_ReducedL5PyrCell/tree/master/GSoC-2022#readme){.reference
    .external}
:::

::: {#other-past-gsoc-projects .section}
[]{#neuromlorg-outreachtraining-gsoc-past}

#### Other past GSoC projects[\#](#other-past-gsoc-projects "Link to this heading"){.headerlink}

-   2020: Ronaldo Nunes: [Conversion of large scale cortical models into
    PyNN/NeuroML](https://summerofcode.withgoogle.com/archive/2020/projects/4613840035119104){.reference
    .external}

-   2018: Jessica Dafflon: [Implementation of Two Neural Mass Models on
    the Open Source Brain
    Platform](https://summerofcode.withgoogle.com/archive/2018/projects/6446944821968896){.reference
    .external}

-   2017: Andras Ecker: [Conversion of a large scale hippocampal network
    model to
    NeuroML](https://summerofcode.withgoogle.com/archive/2017/projects/4673827971792896){.reference
    .external}

-   2016: Rokas Stanislavos: [Building cortical network models in
    NeuroML2 using procedural and declarative programming
    approaches](https://summerofcode.withgoogle.com/archive/2016/projects/5650639363244032){.reference
    .external}

-   2015: Justas Birgiolas: [Convert several large-scale thalamocortical
    models to NeuroML &
    PyNN](https://www.google-melange.com/archive/gsoc/2015/orgs/incf/projects/birgiolasj.html){.reference
    .external}

-   2014: Ramón Martínez Mayorquin: [Open source, cross simulator, large
    scale cortical
    models](https://www.google-melange.com/archive/gsoc/2014/orgs/incf/projects/h_mayorquin.html){.reference
    .external}

-   2013: Vitor Chaud: [Open source, cross simulator, models of cortical
    circuits](https://www.google-melange.com/archive/gsoc/2013/orgs/incf/projects/vitorchaud.html){.reference
    .external}

-   2012: Mike Vella: [Simulator-independent Python API for
    multi-compartmental
    modeling](https://www.google-melange.com/archive/gsoc/2012/orgs/incf/projects/vellamike.html){.reference
    .external}
:::
:::
:::

[]{#document-NeuroMLOrg/Contributors}

::: {#neuroml-contributors .section .tex2jax_ignore .mathjax_ignore}
[]{#neuromlorg-contributors}

## NeuroML contributors[\#](#neuroml-contributors "Link to this heading"){.headerlink}

This page lists contributors to the various NeuroML and related
repositories, listed in no particular order. It is generated
periodically, most recently on 29/07/24. See also the current [[NeuroML
Editorial Board]{.std
.std-ref}](Landing.html#neuromlorg-board){.reference .internal} and the
[[Scientific Committee]{.std
.std-ref}](Landing.html#neuromlorg-scientificcommittee){.reference
.internal}. The list of repositories can be seen on [[the repositories
page]{.std .std-ref}](Landing.html#neuromlorg-repositories){.reference
.internal}.

-   [\@pgleeson](https://github.com/pgleeson){.reference .external}

-   [\@JustasB](https://github.com/JustasB){.reference .external}

-   [\@sanjayankur31](https://github.com/sanjayankur31){.reference
    .external}

-   [\@tarelli](https://github.com/tarelli){.reference .external}

-   [\@borismarin](https://github.com/borismarin){.reference .external}

-   [\@mattearnshaw](https://github.com/mattearnshaw){.reference
    .external}

-   [\@adrianq](https://github.com/adrianq){.reference .external}

-   [\@epiasini](https://github.com/epiasini){.reference .external}

-   [\@RokasSt](https://github.com/RokasSt){.reference .external}

-   [\@dilawar](https://github.com/dilawar){.reference .external}

-   [\@russelljjarvis](https://github.com/russelljjarvis){.reference
    .external}

-   [\@FinnK](https://github.com/FinnK){.reference .external}

-   [\@kapilkd13](https://github.com/kapilkd13){.reference .external}

-   [\@gidili](https://github.com/gidili){.reference .external}

-   [\@wvangeit](https://github.com/wvangeit){.reference .external}

-   [\@hugh-osborne](https://github.com/hugh-osborne){.reference
    .external}

-   [\@lungd](https://github.com/lungd){.reference .external}

-   [\@orena1](https://github.com/orena1){.reference .external}

-   [\@ChihweiLHBird](https://github.com/ChihweiLHBird){.reference
    .external}

-   [\@rgerkin](https://github.com/rgerkin){.reference .external}

-   [\@allcontributors\[bot\]](https://github.com/apps/allcontributors){.reference
    .external}

-   [\@shayanshafquat](https://github.com/shayanshafquat){.reference
    .external}

-   [\@ccluri](https://github.com/ccluri){.reference .external}

-   [\@mwatts15](https://github.com/mwatts15){.reference .external}

-   [\@34383c](https://github.com/34383c){.reference .external}

-   [\@andrisecker](https://github.com/andrisecker){.reference
    .external}

-   [\@jrieke](https://github.com/jrieke){.reference .external}

-   [\@scrook](https://github.com/scrook){.reference .external}

-   [\@jonc125](https://github.com/jonc125){.reference .external}

-   [\@davidt0x](https://github.com/davidt0x){.reference .external}

-   [\@doorkn-b](https://github.com/doorkn-b){.reference .external}

-   [\@ramcdougal](https://github.com/ramcdougal){.reference .external}

-   [\@keszybz](https://github.com/keszybz){.reference .external}

-   [\@avrama](https://github.com/avrama){.reference .external}

-   [\@jsnowacki](https://github.com/jsnowacki){.reference .external}

-   [\@rocapp](https://github.com/rocapp){.reference .external}

-   [\@prex1030](https://github.com/prex1030){.reference .external}

-   [\@frothga](https://github.com/frothga){.reference .external}

-   [\@fywang](https://github.com/fywang){.reference .external}

-   [\@DavidPoliakoff](https://github.com/DavidPoliakoff){.reference
    .external}

-   [\@jonsalaz](https://github.com/jonsalaz){.reference .external}

-   [\@cewarr](https://github.com/cewarr){.reference .external}

-   [\@Ermentrout](https://github.com/Ermentrout){.reference .external}

-   [\@miscco](https://github.com/miscco){.reference .external}

-   [\@lisphacker](https://github.com/lisphacker){.reference .external}

-   [\@dokato](https://github.com/dokato){.reference .external}

-   [\@robertcannon](https://github.com/robertcannon){.reference
    .external}

-   [\@JLLeitschuh](https://github.com/JLLeitschuh){.reference
    .external}

-   [\@waffle-iron](https://github.com/waffle-iron){.reference
    .external}

-   [\@volker01](https://github.com/volker01){.reference .external}

-   [\@theiera](https://github.com/theiera){.reference .external}

-   [\@apdavison](https://github.com/apdavison){.reference .external}

-   [\@vitorchaud](https://github.com/vitorchaud){.reference .external}

-   [\@hglabska](https://github.com/hglabska){.reference .external}

-   [\@adityagilra](https://github.com/adityagilra){.reference
    .external}

-   [\@kkalou](https://github.com/kkalou){.reference .external}

-   [\@osb-admin](https://github.com/osb-admin){.reference .external}

-   [\@tcstewar](https://github.com/tcstewar){.reference .external}

-   [\@jplang](https://github.com/jplang){.reference .external}

-   [\@marutosi](https://github.com/marutosi){.reference .external}

-   [\@edavis10](https://github.com/edavis10){.reference .external}

-   [\@winterheart](https://github.com/winterheart){.reference
    .external}

-   [\@jbbarth](https://github.com/jbbarth){.reference .external}

-   [\@jgoerzen](https://github.com/jgoerzen){.reference .external}

-   [\@h-mayorquin](https://github.com/h-mayorquin){.reference
    .external}

-   [\@robertvi](https://github.com/robertvi){.reference .external}

-   [\@stellaprins](https://github.com/stellaprins){.reference
    .external}

-   [\@harveymannering](https://github.com/harveymannering){.reference
    .external}

-   [\@acardona](https://github.com/acardona){.reference .external}

-   [\@rizland](https://github.com/rizland){.reference .external}

-   [\@SteMasoli](https://github.com/SteMasoli){.reference .external}

-   [\@MFarinella](https://github.com/MFarinella){.reference .external}

-   [\@yates9](https://github.com/yates9){.reference .external}

-   [\@Neurophile](https://github.com/Neurophile){.reference .external}

-   [\@EmmanuelleChaigneau](https://github.com/EmmanuelleChaigneau){.reference
    .external}

-   [\@slarson](https://github.com/slarson){.reference .external}

-   [\@usama57](https://github.com/usama57){.reference .external}

-   [\@dasj-osb](https://github.com/dasj-osb){.reference .external}

-   [\@muratkrty](https://github.com/muratkrty){.reference .external}

-   [\@arnabiswas](https://github.com/arnabiswas){.reference .external}

-   [\@souravsingh](https://github.com/souravsingh){.reference
    .external}

-   [\@pshwetank](https://github.com/pshwetank){.reference .external}

-   [\@yuumi15](https://github.com/yuumi15){.reference .external}

-   [\@albada](https://github.com/albada){.reference .external}

-   [\@jhnnsnk](https://github.com/jhnnsnk){.reference .external}

-   [\@anujanegi](https://github.com/anujanegi){.reference .external}

-   [\@nikiita013](https://github.com/nikiita013){.reference .external}

-   [\@kedoxey](https://github.com/kedoxey){.reference .external}

-   [\@ChristophMetzner](https://github.com/ChristophMetzner){.reference
    .external}

-   [\@evakh](https://github.com/evakh){.reference .external}

-   [\@BiaDarkia](https://github.com/BiaDarkia){.reference .external}

-   [\@Mcdonoughd](https://github.com/Mcdonoughd){.reference .external}

-   [\@sdrsd](https://github.com/sdrsd){.reference .external}

-   [\@salvadord](https://github.com/salvadord){.reference .external}

-   [\@kperun](https://github.com/kperun){.reference .external}

-   [\@ddanny](https://github.com/ddanny){.reference .external}

-   [\@jrmartin](https://github.com/jrmartin){.reference .external}

-   [\@koppen](https://github.com/koppen){.reference .external}

-   [\@tofi86](https://github.com/tofi86){.reference .external}

-   [\@yuba](https://github.com/yuba){.reference .external}

-   [\@smith](https://github.com/smith){.reference .external}

-   [\@abronte](https://github.com/abronte){.reference .external}

-   [\@dpbus](https://github.com/dpbus){.reference .external}

-   [\@korun](https://github.com/korun){.reference .external}

-   [\@wronglink](https://github.com/wronglink){.reference .external}

-   [\@clairvy](https://github.com/clairvy){.reference .external}

-   [\@RobFerrer](https://github.com/RobFerrer){.reference .external}

-   [\@ryansch](https://github.com/ryansch){.reference .external}

-   [\@yujideveloper](https://github.com/yujideveloper){.reference
    .external}

-   [\@chantra](https://github.com/chantra){.reference .external}

-   [\@mattmueller](https://github.com/mattmueller){.reference
    .external}

-   [\@tfliss](https://github.com/tfliss){.reference .external}

-   [\@BKriener](https://github.com/BKriener){.reference .external}

-   [\@JessyD](https://github.com/JessyD){.reference .external}

-   [\@jasonxanthakis](https://github.com/jasonxanthakis){.reference
    .external}

-   [\@jimboH](https://github.com/jimboH){.reference .external}

-   [\@vrhaynes](https://github.com/vrhaynes){.reference .external}

-   [\@abhineet99](https://github.com/abhineet99){.reference .external}

-   [\@filippomc](https://github.com/filippomc){.reference .external}

-   [\@kmantel](https://github.com/kmantel){.reference .external}

-   [\@zsinnema](https://github.com/zsinnema){.reference .external}

-   [\@Muhaddatha](https://github.com/Muhaddatha){.reference .external}

-   [\@D-GopalKrishna](https://github.com/D-GopalKrishna){.reference
    .external}

-   [\@Aiga115](https://github.com/Aiga115){.reference .external}

-   [\@vidhya-metacell](https://github.com/vidhya-metacell){.reference
    .external}

-   [\@vidhya-longani](https://github.com/vidhya-longani){.reference
    .external}

-   [\@etowett](https://github.com/etowett){.reference .external}

-   [\@SimaoBolota-MetaCell](https://github.com/SimaoBolota-MetaCell){.reference
    .external}

-   [\@Salam-Dalloul](https://github.com/Salam-Dalloul){.reference
    .external}

-   [\@ronaldonunes](https://github.com/ronaldonunes){.reference
    .external}

-   [\@pfeffer90](https://github.com/pfeffer90){.reference .external}

-   [\@stephprince](https://github.com/stephprince){.reference
    .external}

-   [\@hrani](https://github.com/hrani){.reference .external}

-   [\@upibhalla](https://github.com/upibhalla){.reference .external}

-   [\@asiaszmek](https://github.com/asiaszmek){.reference .external}

-   [\@subhacom](https://github.com/subhacom){.reference .external}

-   [\@aviralg](https://github.com/aviralg){.reference .external}

-   [\@malav4994](https://github.com/malav4994){.reference .external}

-   [\@bhanu151](https://github.com/bhanu151){.reference .external}

-   [\@analkumar2](https://github.com/analkumar2){.reference .external}

-   [\@bitdeli-chef](https://github.com/bitdeli-chef){.reference
    .external}

-   [\@physicalist](https://github.com/physicalist){.reference
    .external}

-   [\@anhknguyen96](https://github.com/anhknguyen96){.reference
    .external}

-   [\@Daksh1603](https://github.com/Daksh1603){.reference .external}

-   [\@yarikoptic](https://github.com/yarikoptic){.reference .external}

-   [\@musicinmybrain](https://github.com/musicinmybrain){.reference
    .external}

-   [\@tommorse](https://github.com/tommorse){.reference .external}

-   [\@neuroman314](https://github.com/neuroman314){.reference
    .external}

-   [\@pramodk](https://github.com/pramodk){.reference .external}

-   [\@alexsavulescu](https://github.com/alexsavulescu){.reference
    .external}

-   [\@ferdonline](https://github.com/ferdonline){.reference .external}

-   [\@olupton](https://github.com/olupton){.reference .external}

-   [\@vellamike](https://github.com/vellamike){.reference .external}

-   [\@clbarnes](https://github.com/clbarnes){.reference .external}

-   [\@mattions](https://github.com/mattions){.reference .external}

-   [\@lebedov](https://github.com/lebedov){.reference .external}

-   [\@baladkb](https://github.com/baladkb){.reference .external}

-   [\@jefferis](https://github.com/jefferis){.reference .external}

-   [\@mstimberg](https://github.com/mstimberg){.reference .external}

-   [\@arosh](https://github.com/arosh){.reference .external}

-   [\@unidesigner](https://github.com/unidesigner){.reference
    .external}

-   [\@lucasklmn](https://github.com/lucasklmn){.reference .external}
:::

[]{#document-NeuroMLOrg/Repositories}

::: {#neuroml-repositories .section .tex2jax_ignore .mathjax_ignore}
[]{#neuromlorg-repositories}

## NeuroML repositories[\#](#neuroml-repositories "Link to this heading"){.headerlink}

This page lists repositories related to NeuroML, listed in no particular
order. It is generated periodically, most recently on 29/07/24. A
complete list of contributors can be seen [[here]{.std
.std-ref}](Landing.html#neuromlorg-contributors){.reference .internal}.

For the status of tests on standardized NeuroML models on Open Source
Brain, please see this page:
[OpenSourceBrain/.github](https://github.com/OpenSourceBrain/.github/blob/main/testsheet/README.md){.github
.reference .external}.

-   [NeuroML/Cvapp-NeuroMorpho.org](https://github.com/NeuroML/Cvapp-NeuroMorpho.org){.reference
    .external}

-   [NeuroML/org.neuroml.model](https://github.com/NeuroML/org.neuroml.model){.reference
    .external}

-   [NeuroML/org.neuroml.model.injectingplugin](https://github.com/NeuroML/org.neuroml.model.injectingplugin){.reference
    .external}

-   [NeuroML/org.neuroml1.model](https://github.com/NeuroML/org.neuroml1.model){.reference
    .external}

-   [NeuroML/org.neuroml.visualiser](https://github.com/NeuroML/org.neuroml.visualiser){.reference
    .external}

-   [NeuroML/NeuroML2](https://github.com/NeuroML/NeuroML2){.reference
    .external}

-   [NeuroML/jNeuroML](https://github.com/NeuroML/jNeuroML){.reference
    .external}

-   [NeuroML/org.neuroml.export](https://github.com/NeuroML/org.neuroml.export){.reference
    .external}

-   [NeuroML/org.neuroml.import](https://github.com/NeuroML/org.neuroml.import){.reference
    .external}

-   [NeuroML/NeuroMLToPOV-Ray](https://github.com/NeuroML/NeuroMLToPOV-Ray){.reference
    .external}

-   [NeuroML/NML2_LEMS_Examples](https://github.com/NeuroML/NML2_LEMS_Examples){.reference
    .external}

-   [NeuroML/NeuroMLWebsite](https://github.com/NeuroML/NeuroMLWebsite){.reference
    .external}

-   [NeuroML/pyNeuroML](https://github.com/NeuroML/pyNeuroML){.reference
    .external}

-   [NeuroML/neuroml2model](https://github.com/NeuroML/neuroml2model){.reference
    .external}

-   [NeuroML/Presentations](https://github.com/NeuroML/Presentations){.reference
    .external}

-   [NeuroML/NeuroMLToolbox](https://github.com/NeuroML/NeuroMLToolbox){.reference
    .external}

-   [NeuroML/NeuroML_API](https://github.com/NeuroML/NeuroML_API){.reference
    .external}

-   [NeuroML/NetworkShorthand](https://github.com/NeuroML/NetworkShorthand){.reference
    .external}

-   [NeuroML/mod2neuroml](https://github.com/NeuroML/mod2neuroml){.reference
    .external}

-   [NeuroML/neuroml2modelLite](https://github.com/NeuroML/neuroml2modelLite){.reference
    .external}

-   [NeuroML/NeuroMLlite](https://github.com/NeuroML/NeuroMLlite){.reference
    .external}

-   [NeuroML/Documentation](https://github.com/NeuroML/Documentation){.reference
    .external}

-   [NeuroML/.github](https://github.com/NeuroML/.github){.reference
    .external}

-   [NeuroML/stochdiff](https://github.com/NeuroML/stochdiff){.reference
    .external}

-   [NeuroML/xppy](https://github.com/NeuroML/xppy){.reference
    .external}

-   [NeuroML/n2a](https://github.com/NeuroML/n2a){.reference .external}

-   [NeuroML/xppaut](https://github.com/NeuroML/xppaut){.reference
    .external}

-   [LEMS/pylems](https://github.com/LEMS/pylems){.reference .external}

-   [LEMS/jLEMS](https://github.com/LEMS/jLEMS){.reference .external}

-   [LEMS/LEMS](https://github.com/LEMS/LEMS){.reference .external}

-   [LEMS/org.lemsml.model](https://github.com/LEMS/org.lemsml.model){.reference
    .external}

-   [LEMS/expr-parser](https://github.com/LEMS/expr-parser){.reference
    .external}

-   [LEMS/lems-domogen-maven-plugin](https://github.com/LEMS/lems-domogen-maven-plugin){.reference
    .external}

-   [OpenSourceBrain/CA1PyramidalCell](https://github.com/OpenSourceBrain/CA1PyramidalCell){.reference
    .external}

-   [OpenSourceBrain/OSB_API](https://github.com/OpenSourceBrain/OSB_API){.reference
    .external}

-   [OpenSourceBrain/CerebellarNucleusNeuron](https://github.com/OpenSourceBrain/CerebellarNucleusNeuron){.reference
    .external}

-   [OpenSourceBrain/GranCellLayer](https://github.com/OpenSourceBrain/GranCellLayer){.reference
    .external}

-   [OpenSourceBrain/GranCellRothmanIf](https://github.com/OpenSourceBrain/GranCellRothmanIf){.reference
    .external}

-   [OpenSourceBrain/GranCellSolinasEtAl10](https://github.com/OpenSourceBrain/GranCellSolinasEtAl10){.reference
    .external}

-   [OpenSourceBrain/GranuleCell](https://github.com/OpenSourceBrain/GranuleCell){.reference
    .external}

-   [OpenSourceBrain/GranuleCellVSCS](https://github.com/OpenSourceBrain/GranuleCellVSCS){.reference
    .external}

-   [OpenSourceBrain/IzhikevichModel](https://github.com/OpenSourceBrain/IzhikevichModel){.reference
    .external}

-   [OpenSourceBrain/MainenEtAl_PyramidalCell](https://github.com/OpenSourceBrain/MainenEtAl_PyramidalCell){.reference
    .external}

-   [OpenSourceBrain/PurkinjeCell](https://github.com/OpenSourceBrain/PurkinjeCell){.reference
    .external}

-   [OpenSourceBrain/RothmanEtAl_KoleEtAl_PyrCell](https://github.com/OpenSourceBrain/RothmanEtAl_KoleEtAl_PyrCell){.reference
    .external}

-   [OpenSourceBrain/SolinasEtAl-GolgiCell](https://github.com/OpenSourceBrain/SolinasEtAl-GolgiCell){.reference
    .external}

-   [OpenSourceBrain/VervaekeEtAl-GolgiCellNetwork](https://github.com/OpenSourceBrain/VervaekeEtAl-GolgiCellNetwork){.reference
    .external}

-   [OpenSourceBrain/Thalamocortical](https://github.com/OpenSourceBrain/Thalamocortical){.reference
    .external}

-   [OpenSourceBrain/PyloricNetwork](https://github.com/OpenSourceBrain/PyloricNetwork){.reference
    .external}

-   [OpenSourceBrain/MorrisLecarModel](https://github.com/OpenSourceBrain/MorrisLecarModel){.reference
    .external}

-   [OpenSourceBrain/StriatalSpinyProjectionNeuron](https://github.com/OpenSourceBrain/StriatalSpinyProjectionNeuron){.reference
    .external}

-   [OpenSourceBrain/CelegansNeuromechanicalGaitModulation](https://github.com/OpenSourceBrain/CelegansNeuromechanicalGaitModulation){.reference
    .external}

-   [OpenSourceBrain/PospischilEtAl2008](https://github.com/OpenSourceBrain/PospischilEtAl2008){.reference
    .external}

-   [OpenSourceBrain/Drosophila_Projection_Neuron](https://github.com/OpenSourceBrain/Drosophila_Projection_Neuron){.reference
    .external}

-   [OpenSourceBrain/NengoNeuroML](https://github.com/OpenSourceBrain/NengoNeuroML){.reference
    .external}

-   [OpenSourceBrain/NeuroElectroSciUnit](https://github.com/OpenSourceBrain/NeuroElectroSciUnit){.reference
    .external}

-   [OpenSourceBrain/NeuroMorpho](https://github.com/OpenSourceBrain/NeuroMorpho){.reference
    .external}

-   [OpenSourceBrain/redmine](https://github.com/OpenSourceBrain/redmine){.reference
    .external}

-   [OpenSourceBrain/CSAShowcase](https://github.com/OpenSourceBrain/CSAShowcase){.reference
    .external}

-   [OpenSourceBrain/L5bPyrCellHayEtAl2011](https://github.com/OpenSourceBrain/L5bPyrCellHayEtAl2011){.reference
    .external}

-   [OpenSourceBrain/neuroConstructShowcase](https://github.com/OpenSourceBrain/neuroConstructShowcase){.reference
    .external}

-   [OpenSourceBrain/PinskyRinzelModel](https://github.com/OpenSourceBrain/PinskyRinzelModel){.reference
    .external}

-   [OpenSourceBrain/Brunel2000](https://github.com/OpenSourceBrain/Brunel2000){.reference
    .external}

-   [OpenSourceBrain/NineMLShowcase](https://github.com/OpenSourceBrain/NineMLShowcase){.reference
    .external}

-   [OpenSourceBrain/BrianShowcase](https://github.com/OpenSourceBrain/BrianShowcase){.reference
    .external}

-   [OpenSourceBrain/MUSICShowcase](https://github.com/OpenSourceBrain/MUSICShowcase){.reference
    .external}

-   [OpenSourceBrain/SBMLShowcase](https://github.com/OpenSourceBrain/SBMLShowcase){.reference
    .external}

-   [OpenSourceBrain/HindmarshRose1984](https://github.com/OpenSourceBrain/HindmarshRose1984){.reference
    .external}

-   [OpenSourceBrain/FitzHugh-Nagumo](https://github.com/OpenSourceBrain/FitzHugh-Nagumo){.reference
    .external}

-   [OpenSourceBrain/BluehiveShowcase](https://github.com/OpenSourceBrain/BluehiveShowcase){.reference
    .external}

-   [OpenSourceBrain/NIFShowcase](https://github.com/OpenSourceBrain/NIFShowcase){.reference
    .external}

-   [OpenSourceBrain/CATMAIDShowcase](https://github.com/OpenSourceBrain/CATMAIDShowcase){.reference
    .external}

-   [OpenSourceBrain/VogelsSprekelerEtAl2011](https://github.com/OpenSourceBrain/VogelsSprekelerEtAl2011){.reference
    .external}

-   [OpenSourceBrain/cereb_grc_mc](https://github.com/OpenSourceBrain/cereb_grc_mc){.reference
    .external}

-   [OpenSourceBrain/NSGPortalShowcase](https://github.com/OpenSourceBrain/NSGPortalShowcase){.reference
    .external}

-   [OpenSourceBrain/ACnet2](https://github.com/OpenSourceBrain/ACnet2){.reference
    .external}

-   [OpenSourceBrain/CNOShowcase](https://github.com/OpenSourceBrain/CNOShowcase){.reference
    .external}

-   [OpenSourceBrain/DentateGyrus2005](https://github.com/OpenSourceBrain/DentateGyrus2005){.reference
    .external}

-   [OpenSourceBrain/ghk-nernst](https://github.com/OpenSourceBrain/ghk-nernst){.reference
    .external}

-   [OpenSourceBrain/korngreen-pyramidal](https://github.com/OpenSourceBrain/korngreen-pyramidal){.reference
    .external}

-   [OpenSourceBrain/neuroConstruct_to_be_deleted](https://github.com/OpenSourceBrain/neuroConstruct_to_be_deleted){.reference
    .external}

-   [OpenSourceBrain/NEURONShowcase](https://github.com/OpenSourceBrain/NEURONShowcase){.reference
    .external}

-   [OpenSourceBrain/LarkumEtAl2009](https://github.com/OpenSourceBrain/LarkumEtAl2009){.reference
    .external}

-   [OpenSourceBrain/GPUShowcase](https://github.com/OpenSourceBrain/GPUShowcase){.reference
    .external}

-   [OpenSourceBrain/FPGAShowcase](https://github.com/OpenSourceBrain/FPGAShowcase){.reference
    .external}

-   [OpenSourceBrain/FarinellaEtAl_NMDAspikes](https://github.com/OpenSourceBrain/FarinellaEtAl_NMDAspikes){.reference
    .external}

-   [OpenSourceBrain/BlueBrainProjectShowcase](https://github.com/OpenSourceBrain/BlueBrainProjectShowcase){.reference
    .external}

-   [OpenSourceBrain/OSB_Metadata](https://github.com/OpenSourceBrain/OSB_Metadata){.reference
    .external}

-   [OpenSourceBrain/OSB_Status](https://github.com/OpenSourceBrain/OSB_Status){.reference
    .external}

-   [OpenSourceBrain/osb-model-validation](https://github.com/OpenSourceBrain/osb-model-validation){.reference
    .external}

-   [OpenSourceBrain/ModelDBShowcase](https://github.com/OpenSourceBrain/ModelDBShowcase){.reference
    .external}

-   [OpenSourceBrain/OSB_Documentation](https://github.com/OpenSourceBrain/OSB_Documentation){.reference
    .external}

-   [OpenSourceBrain/V1NetworkModels](https://github.com/OpenSourceBrain/V1NetworkModels){.reference
    .external}

-   [OpenSourceBrain/PotjansDiesmann2014](https://github.com/OpenSourceBrain/PotjansDiesmann2014){.reference
    .external}

-   [OpenSourceBrain/AllenInstituteNeuroML](https://github.com/OpenSourceBrain/AllenInstituteNeuroML){.reference
    .external}

-   [OpenSourceBrain/StochasticityShowcase](https://github.com/OpenSourceBrain/StochasticityShowcase){.reference
    .external}

-   [OpenSourceBrain/VierlingClaassenEtAl2010](https://github.com/OpenSourceBrain/VierlingClaassenEtAl2010){.reference
    .external}

-   [OpenSourceBrain/OSB_Videos](https://github.com/OpenSourceBrain/OSB_Videos){.reference
    .external}

-   [OpenSourceBrain/Contribute](https://github.com/OpenSourceBrain/Contribute){.reference
    .external}

-   [OpenSourceBrain/SmithEtAl2013-L23DendriticSpikes](https://github.com/OpenSourceBrain/SmithEtAl2013-L23DendriticSpikes){.reference
    .external}

-   [OpenSourceBrain/MiglioreEtAl14_OlfactoryBulb3D](https://github.com/OpenSourceBrain/MiglioreEtAl14_OlfactoryBulb3D){.reference
    .external}

-   [OpenSourceBrain/WeilerEtAl08-LaminarCortex](https://github.com/OpenSourceBrain/WeilerEtAl08-LaminarCortex){.reference
    .external}

-   [OpenSourceBrain/Cerebellum3DDemo](https://github.com/OpenSourceBrain/Cerebellum3DDemo){.reference
    .external}

-   [OpenSourceBrain/dLGNinterneuronHalnesEtAl2011](https://github.com/OpenSourceBrain/dLGNinterneuronHalnesEtAl2011){.reference
    .external}

-   [OpenSourceBrain/GranularLayerSolinasNieusDAngelo2010](https://github.com/OpenSourceBrain/GranularLayerSolinasNieusDAngelo2010){.reference
    .external}

-   [OpenSourceBrain/CommunityModellingCA1](https://github.com/OpenSourceBrain/CommunityModellingCA1){.reference
    .external}

-   [OpenSourceBrain/FergusonEtAl2014-CA1PyrCell](https://github.com/OpenSourceBrain/FergusonEtAl2014-CA1PyrCell){.reference
    .external}

-   [OpenSourceBrain/VERTEXShowcase](https://github.com/OpenSourceBrain/VERTEXShowcase){.reference
    .external}

-   [OpenSourceBrain/MOOSEShowcase](https://github.com/OpenSourceBrain/MOOSEShowcase){.reference
    .external}

-   [OpenSourceBrain/recaptcha](https://github.com/OpenSourceBrain/recaptcha){.reference
    .external}

-   [OpenSourceBrain/OpenCortex](https://github.com/OpenSourceBrain/OpenCortex){.reference
    .external}

-   [OpenSourceBrain/OlfactoryTest](https://github.com/OpenSourceBrain/OlfactoryTest){.reference
    .external}

-   [OpenSourceBrain/FergusonEtAl2013-PVFastFiringCell](https://github.com/OpenSourceBrain/FergusonEtAl2013-PVFastFiringCell){.reference
    .external}

-   [OpenSourceBrain/M1NetworkModel](https://github.com/OpenSourceBrain/M1NetworkModel){.reference
    .external}

-   [OpenSourceBrain/NESTShowcase](https://github.com/OpenSourceBrain/NESTShowcase){.reference
    .external}

-   [OpenSourceBrain/Hippocampus3DDemo](https://github.com/OpenSourceBrain/Hippocampus3DDemo){.reference
    .external}

-   [OpenSourceBrain/OSB_Samples](https://github.com/OpenSourceBrain/OSB_Samples){.reference
    .external}

-   [OpenSourceBrain/org.geppetto.persistence](https://github.com/OpenSourceBrain/org.geppetto.persistence){.reference
    .external}

-   [OpenSourceBrain/redmine_github_hook](https://github.com/OpenSourceBrain/redmine_github_hook){.reference
    .external}

-   [OpenSourceBrain/PyNNShowcase](https://github.com/OpenSourceBrain/PyNNShowcase){.reference
    .external}

-   [OpenSourceBrain/neuralensemble-docker](https://github.com/OpenSourceBrain/neuralensemble-docker){.reference
    .external}

-   [OpenSourceBrain/WangBuzsaki1996](https://github.com/OpenSourceBrain/WangBuzsaki1996){.reference
    .external}

-   [OpenSourceBrain/GolgiCellDendGapJunctions](https://github.com/OpenSourceBrain/GolgiCellDendGapJunctions){.reference
    .external}

-   [OpenSourceBrain/NetPyNEShowcase](https://github.com/OpenSourceBrain/NetPyNEShowcase){.reference
    .external}

-   [OpenSourceBrain/Poirazi2003-CA1PyramidalCell](https://github.com/OpenSourceBrain/Poirazi2003-CA1PyramidalCell){.reference
    .external}

-   [OpenSourceBrain/Ferrante2009-DentateGyrusGranuleCell](https://github.com/OpenSourceBrain/Ferrante2009-DentateGyrusGranuleCell){.reference
    .external}

-   [OpenSourceBrain/Hemond2008-CA3PyramidalCell](https://github.com/OpenSourceBrain/Hemond2008-CA3PyramidalCell){.reference
    .external}

-   [OpenSourceBrain/geppetto-osb](https://github.com/OpenSourceBrain/geppetto-osb){.reference
    .external}

-   [OpenSourceBrain/SynapticIntegration](https://github.com/OpenSourceBrain/SynapticIntegration){.reference
    .external}

-   [OpenSourceBrain/NorenbergEtAl2010_DGBasketCell](https://github.com/OpenSourceBrain/NorenbergEtAl2010_DGBasketCell){.reference
    .external}

-   [OpenSourceBrain/TheVirtualBrainShowcase](https://github.com/OpenSourceBrain/TheVirtualBrainShowcase){.reference
    .external}

-   [OpenSourceBrain/tutorials](https://github.com/OpenSourceBrain/tutorials){.reference
    .external}

-   [OpenSourceBrain/BartosEtAl2002](https://github.com/OpenSourceBrain/BartosEtAl2002){.reference
    .external}

-   [OpenSourceBrain/destexhe_jcns_2009](https://github.com/OpenSourceBrain/destexhe_jcns_2009){.reference
    .external}

-   [OpenSourceBrain/MouseLightShowcase](https://github.com/OpenSourceBrain/MouseLightShowcase){.reference
    .external}

-   [OpenSourceBrain/BonoClopath2017](https://github.com/OpenSourceBrain/BonoClopath2017){.reference
    .external}

-   [OpenSourceBrain/SpineShowcase](https://github.com/OpenSourceBrain/SpineShowcase){.reference
    .external}

-   [OpenSourceBrain/SadehEtAl2017-InhibitionStabilizedNetworks](https://github.com/OpenSourceBrain/SadehEtAl2017-InhibitionStabilizedNetworks){.reference
    .external}

-   [OpenSourceBrain/WilsonCowan](https://github.com/OpenSourceBrain/WilsonCowan){.reference
    .external}

-   [OpenSourceBrain/MultiscaleISN](https://github.com/OpenSourceBrain/MultiscaleISN){.reference
    .external}

-   [OpenSourceBrain/PINGnets](https://github.com/OpenSourceBrain/PINGnets){.reference
    .external}

-   [OpenSourceBrain/DynaSimShowcase](https://github.com/OpenSourceBrain/DynaSimShowcase){.reference
    .external}

-   [OpenSourceBrain/Amacrine](https://github.com/OpenSourceBrain/Amacrine){.reference
    .external}

-   [OpenSourceBrain/del-Molino2017](https://github.com/OpenSourceBrain/del-Molino2017){.reference
    .external}

-   [OpenSourceBrain/MejiasEtAl2016](https://github.com/OpenSourceBrain/MejiasEtAl2016){.reference
    .external}

-   [OpenSourceBrain/NWBShowcase](https://github.com/OpenSourceBrain/NWBShowcase){.reference
    .external}

-   [OpenSourceBrain/L23PyramidalCellTutorial](https://github.com/OpenSourceBrain/L23PyramidalCellTutorial){.reference
    .external}

-   [OpenSourceBrain/ConnectivityShowcase](https://github.com/OpenSourceBrain/ConnectivityShowcase){.reference
    .external}

-   [OpenSourceBrain/IonChannelGenealogyShowcase](https://github.com/OpenSourceBrain/IonChannelGenealogyShowcase){.reference
    .external}

-   [OpenSourceBrain/TobinEtAl2017](https://github.com/OpenSourceBrain/TobinEtAl2017){.reference
    .external}

-   [OpenSourceBrain/CalciumImagingDriftingGrating](https://github.com/OpenSourceBrain/CalciumImagingDriftingGrating){.reference
    .external}

-   [OpenSourceBrain/PsyNeuLinkShowcase](https://github.com/OpenSourceBrain/PsyNeuLinkShowcase){.reference
    .external}

-   [OpenSourceBrain/EbnerEtAl2019](https://github.com/OpenSourceBrain/EbnerEtAl2019){.reference
    .external}

-   [OpenSourceBrain/OSBv2](https://github.com/OpenSourceBrain/OSBv2){.reference
    .external}

-   [OpenSourceBrain/JoglekarEtAl18](https://github.com/OpenSourceBrain/JoglekarEtAl18){.reference
    .external}

-   [OpenSourceBrain/BahlEtAl2012_ReducedL5PyrCell](https://github.com/OpenSourceBrain/BahlEtAl2012_ReducedL5PyrCell){.reference
    .external}

-   [OpenSourceBrain/SussilloAndAbbott2009](https://github.com/OpenSourceBrain/SussilloAndAbbott2009){.reference
    .external}

-   [OpenSourceBrain/DemirtasEtAl19](https://github.com/OpenSourceBrain/DemirtasEtAl19){.reference
    .external}

-   [OpenSourceBrain/SinusoidalVoltageProtocols](https://github.com/OpenSourceBrain/SinusoidalVoltageProtocols){.reference
    .external}

-   [OpenSourceBrain/ONNXShowcase](https://github.com/OpenSourceBrain/ONNXShowcase){.reference
    .external}

-   [OpenSourceBrain/ServiceStatus](https://github.com/OpenSourceBrain/ServiceStatus){.reference
    .external}

-   [OpenSourceBrain/ArborShowcase](https://github.com/OpenSourceBrain/ArborShowcase){.reference
    .external}

-   [OpenSourceBrain/hh-testing](https://github.com/OpenSourceBrain/hh-testing){.reference
    .external}

-   [OpenSourceBrain/HNN](https://github.com/OpenSourceBrain/HNN){.reference
    .external}

-   [OpenSourceBrain/Documentation](https://github.com/OpenSourceBrain/Documentation){.reference
    .external}

-   [OpenSourceBrain/GSoC_2021_OSB_NWB](https://github.com/OpenSourceBrain/GSoC_2021_OSB_NWB){.reference
    .external}

-   [OpenSourceBrain/EDENShowcase](https://github.com/OpenSourceBrain/EDENShowcase){.reference
    .external}

-   [OpenSourceBrain/BindsNETShowcase](https://github.com/OpenSourceBrain/BindsNETShowcase){.reference
    .external}

-   [OpenSourceBrain/moose-core](https://github.com/OpenSourceBrain/moose-core){.reference
    .external}

-   [OpenSourceBrain/.github](https://github.com/OpenSourceBrain/.github){.reference
    .external}

-   [OpenSourceBrain/OSBv2_Showcase](https://github.com/OpenSourceBrain/OSBv2_Showcase){.reference
    .external}

-   [OpenSourceBrain/DANDIArchiveShowcase](https://github.com/OpenSourceBrain/DANDIArchiveShowcase){.reference
    .external}

-   [OpenSourceBrain/NeuroDataShare](https://github.com/OpenSourceBrain/NeuroDataShare){.reference
    .external}

-   [OpenSourceBrain/OSB_homepage](https://github.com/OpenSourceBrain/OSB_homepage){.reference
    .external}

-   [OpenSourceBrain/pynsgr](https://github.com/OpenSourceBrain/pynsgr){.reference
    .external}

-   [OpenSourceBrain/BarrelCortexInSilicoShowcase](https://github.com/OpenSourceBrain/BarrelCortexInSilicoShowcase){.reference
    .external}

-   [OpenSourceBrain/Maki-MarttunenEtAl2020](https://github.com/OpenSourceBrain/Maki-MarttunenEtAl2020){.reference
    .external}

-   [NeuralEnsemble/libNeuroML](https://github.com/NeuralEnsemble/libNeuroML){.reference
    .external}

-   [scrook/neuroml-db](https://github.com/scrook/neuroml-db){.reference
    .external}

-   [NeuralEnsemble/neurotune](https://github.com/NeuralEnsemble/neurotune){.reference
    .external}

-   [NeuralEnsemble/pyelectro](https://github.com/NeuralEnsemble/pyelectro){.reference
    .external}
:::

[]{#document-NeuroMLOrg/CoC}

::: {#code-of-conduct .section .tex2jax_ignore .mathjax_ignore}
[]{#coc}

## Code of Conduct[\#](#code-of-conduct "Link to this heading"){.headerlink}

Everyone is welcome in the NeuroML community. We request everyone
interacting on the NeuroML channels in any capacity to treat each other
respectfully. Please:

-   act in good faith

-   be friendly, welcoming, respectful, and patient

-   be mindful and considerate

-   be open, use prefer and promote Open Science practices.

If you experience or become aware of behaviour that does not adhere to
the Code of Conduct, please contact the moderators of the channel/event
you are in.
:::
:::

::: {.toctree-wrapper .compound}
[]{#document-Devdocs/Devdocs}

::: {#overview .section .tex2jax_ignore .mathjax_ignore}
[]{#devdocs-overview}

## Overview[\#](#overview "Link to this heading"){.headerlink}

This section will contain information for those who wish to **contribute
to the development** of the NeuroML standard and associated tools.

An overview of the NeuroML **release process** can be found [[here]{.std
.std-ref}](Landing.html#devdocs-release){.reference .internal}.

The relationship of NeuroML to a number of other tools and standards in
computational neuroscience, and the practical steps taken thus far to
ensure interoperability, can be found [[here]{.std
.std-ref}](Landing.html#devdocs-interaction){.reference .internal}.

The following project Kanban boards are used to consolidate issues:

-   [NeuroML](https://github.com/orgs/NeuroML/projects/4/views/1){.reference
    .external}: for all repositories under the NeuroML GitHub
    organization

-   [LEMS](https://github.com/orgs/LEMS/projects/2/views/1){.reference
    .external}: for all repositories under the LEMS GitHub organization

-   [NeuralEnsemble](https://github.com/orgs/NeuralEnsemble/projects/1/views/1){.reference
    .external}: for all NeuroML related repositories in the Neural
    Ensemble GitHub organization

::: {.toctree-wrapper .compound}
[]{#document-Devdocs/DevSOP}

::: {#contribution-guidelines .section .tex2jax_ignore .mathjax_ignore}
[]{#devdocs-devsop}

### Contribution guidelines[\#](#contribution-guidelines "Link to this heading"){.headerlink}

Thank you for your interest in contributing to NeuroML. Welcome!

This page documents the contribution guidelines for all NeuroML related
repositories.

Please do remember that these are *guidelines* but not rules that must
be strictly followed. We think these are reasonable ideas to follow and
they help us maintain a high code quality while making it easier and
more efficient for all of us to work together. However, there may be
cases where they can not be followed, and that's fine too.

::: {#code-of-conduct .section}
[]{#devdocs-devsop-coc}

#### Code of conduct[\#](#code-of-conduct "Link to this heading"){.headerlink}

All NeuroML projects are governed by the [[Code of Conduct]{.std
.std-ref}](Landing.html#coc){.reference .internal}. By participating,
you are expected to uphold this code. Please report unacceptable
behaviour to the moderators of the communication channel you are in.
:::

::: {#structure-of-repositories .section}
[]{#devdocs-devsop-repos}

#### Structure of repositories[\#](#structure-of-repositories "Link to this heading"){.headerlink}

-   All NeuroML repositories use the
    [Git](https://git-scm.com/){.reference .external} version control
    system.

-   Contributions are made using [pull
    requests](https://docs.github.com/en/github/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-pull-requests){.reference
    .external}.

-   Each NeuroML software tool resides in its own GitHub repository
    under the [NeuroML GitHub
    Organization](https://github.com/NeuroML){.reference .external},
    apart from
    [libNeuroML](https://github.com/NeuralEnsemble/libNeuroML){.reference
    .external} which is developed in collaboration with the
    [NeuralEnsemble community](http://neuralensemble.org/){.reference
    .external} and so lives under their GitHub organization.

-   LEMS repositories are housed under the [LEMS GitHub
    Organization](https://github.com/LEMS){.reference .external}.

You can find links to these on the respective pages for each [[software
tool]{.std .std-ref}](Landing.html#userdocs-software){.reference
.internal}.

The NeuroML standard itself (schema and ComponentType definitions) is
housed in its own repository
[here](https://github.com/NeuroML/NeuroML2){.reference .external}.

(devdocs:devsop:repos:zenhub)

::: {#kanban-board-on-zenhub .section}
##### Kanban board on Zenhub[\#](#kanban-board-on-zenhub "Link to this heading"){.headerlink}

An overview of the various repositories, tasks, issues, and so on can be
seen on the [NeuroML Kanban board on
Zenhub](https://app.zenhub.com/workspaces/neuroml-development-605c92c7c670460016e497ab/board?filterLogic=any&amp;repos=7225220,6579766,7225426,299352189,78101103,129064858,8460738,6171449,6171626,27832592,78100679,6171646,3740176,4614078,7146844,4326891&amp;showPipelineDescriptions=false){.reference
.external}.
:::
:::

::: {#versioning .section}
[]{#devdocs-devsop-versioning}

#### Versioning[\#](#versioning "Link to this heading"){.headerlink}

All NeuroML repositories (including the standard) follow [Semantic
versioning](https://semver.org/){.reference .external}. This means that
the version string consists of three components:
[`MAJOR.MINOR.PATCH`{.docutils .literal .notranslate}]{.pre}:

-   the MAJOR version is incremented when incompatible API changes are
    made,

-   the MINOR version is incremented when functionality is added in a
    backwards compatible manner, and

-   the PATCH version is incremented when backwards compatible bug fixes
    are made.
:::

::: {#git-branches .section}
[]{#devdocs-devsop-branches}

#### Git branches[\#](#git-branches "Link to this heading"){.headerlink}

-   Please develop against the [`development`{.docutils .literal
    .notranslate}]{.pre} branch in all repositories. This branch is
    merged into [`master`{.docutils .literal .notranslate}]{.pre} via a
    pull request when a new release is made. This ensures that all tests
    are run at each step to verify correctness. As a result, the
    [`master`{.docutils .literal .notranslate}]{.pre} branch of all
    repositories holds the stable version of the standard and tools,
    while the [`development`{.docutils .literal .notranslate}]{.pre}
    branch holds the next, unstable version that is being worked upon.

-   For branch names, please consider using the [Git
    flow](https://nvie.com/posts/a-successful-git-branching-model/){.reference
    .external} naming convention (not mandatory but strongly suggested):

    -   prefix feature branches with [`feat/`{.docutils .literal
        .notranslate}]{.pre} or [`enh/`{.docutils .literal
        .notranslate}]{.pre} (for enhancement)

    -   prefix bugfix branches with [`bugfix/`{.docutils .literal
        .notranslate}]{.pre} or [`fix/`{.docutils .literal
        .notranslate}]{.pre}

    -   pull requests addressing specific tickets may also mention them
        in the branch name. E.g., [`bugfix/issue-22`{.docutils .literal
        .notranslate}]{.pre}.
:::

::: {#git-commits .section}
[]{#devdocs-devsop-commits}

#### Git commits[\#](#git-commits "Link to this heading"){.headerlink}

Git commit messages are extremely important because they allow us to
nicely track the complete development history of the project. Here are
some guidelines on writing good commit messages:

-   Each commit should ideally only address one issue. It should be
    self-contained (should not group together lots of changes). Tip: use
    [`git`{.docutils .literal .notranslate}]{.pre}` `{.docutils .literal
    .notranslate}[`add`{.docutils .literal
    .notranslate}]{.pre}` `{.docutils .literal
    .notranslate}[`-p`{.docutils .literal .notranslate}]{.pre} to break
    your work down into logical, small commits).

-   Write good commit messages. Read [this
    post](https://chris.beams.io/posts/git-commit/){.reference
    .external} to see how to write meaningful, useful commit messages
    and why they are important.

-   We strongly suggest using the [Conventional
    Commit](https://www.conventionalcommits.org/en/v1.0.0/#summary){.reference
    .external} specification. In short:

    -   Each commit is of the form [`<type>[optional`{.docutils .literal
        .notranslate}]{.pre}` `{.docutils .literal
        .notranslate}[`scope]:`{.docutils .literal
        .notranslate}]{.pre}` `{.docutils .literal
        .notranslate}[`description`{.docutils .literal
        .notranslate}]{.pre}, followed by the text body of the commit
        after a blank line, and then any optional references etc. as
        footer.

    -   The [`type`{.docutils .literal .notranslate}]{.pre} can be one
        of: [`fix`{.docutils .literal .notranslate}]{.pre},
        [`feat`{.docutils .literal .notranslate}]{.pre},
        [`build`{.docutils .literal .notranslate}]{.pre},
        [`chore`{.docutils .literal .notranslate}]{.pre},
        [`ci`{.docutils .literal .notranslate}]{.pre}, [`docs`{.docutils
        .literal .notranslate}]{.pre}, [`refactor`{.docutils .literal
        .notranslate}]{.pre}, [`perf`{.docutils .literal
        .notranslate}]{.pre}, [`test`{.docutils .literal
        .notranslate}]{.pre}, and so on depending on what the commit is
        doing.

    -   Any backwards incompatible, breaking change must be clearly
        noted in the commit using the [`BREAKING`{.docutils .literal
        .notranslate}]{.pre}` `{.docutils .literal
        .notranslate}[`CHANGE`{.docutils .literal .notranslate}]{.pre}
        phrase. This corresponds to a major version update (as noted
        above in the versioning section).
:::

::: {#code-style-java .section}
[]{#devdocs-devsop-style-java}

#### Code style: Java[\#](#code-style-java "Link to this heading"){.headerlink}

TODO
:::

::: {#code-style-python .section}
[]{#devdocs-devsop-style-python}

#### Code style: Python[\#](#code-style-python "Link to this heading"){.headerlink}

-   While Python 2 is still supported even though it is [no longer
    supported by the Python
    community](https://pythonclock.org){.reference .external}, given
    that most Python modules (numpy/scipy/matplotlib/sphinx) have
    dropped support for this deprecated Python version, NeuroML will
    also drop support in the near future. Therefore, we strongly suggest
    using Python 3.

-   For Python repositories, please use
    [Black](https://black.readthedocs.io/){.reference .external} to
    format your code before committing and submitting a pull request.

-   We also strongly suggest linting using
    [flake8](https://flake8.pycqa.org/){.reference .external}.

-   Please use [type
    hints](https://docs.python.org/3/library/typing.html?highlight=type%20hint){.reference
    .external} in your code and run
    [mypy](https://mypy.readthedocs.io/en/stable/){.reference .external}
    to test it for correctness. You can see the [mypy
    cheatsheet](https://mypy.readthedocs.io/en/stable/cheat_sheet.html){.reference
    .external} to quickly see how to do this. Since NeuroML is currently
    still supporting Python 2, we use the Python 2 style to maintain
    compatibility (this also works with Python 3).

-   Deprecations should be clearly noted in the code, and in the commit
    message. You may use the [Sphinx deprecated
    directive](https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-deprecated){.reference
    .external} along with the Python
    [DeprecationWarning](https://docs.python.org/3/library/exceptions.html#DeprecationWarning){.reference
    .external}, for example.
:::

::: {#documentation .section}
[]{#devdocs-devsop-docs}

#### Documentation[\#](#documentation "Link to this heading"){.headerlink}

All tools include their own documentation in their repositories. Please
feel free to improve this documentation and submit pull requests.

When contributing fixes and enhancements, please remember to document
your classes/functions and code in general. Not only does this allow
others to understand your code, it also allows us to auto-generate
documentation using various tools.

-   For the Java repositories, please use the standard
    [Javadoc](https://www.oracle.com/technical-resources/articles/java/javadoc-tool.html){.reference
    .external} syntax.

-   For the Python repositories, please document your code using the
    standard [Sphinx
    reStructuredText](https://www.sphinx-doc.org){.reference .external}
    system. For functions and so on, you can use the provided
    [fields](https://www.sphinx-doc.org/en/master/usage/restructuredtext/domains.html?#python-signatures){.reference
    .external}.

Where applicable, please add examples and so on to the software
documentation to ensure that users can find the information quickly.
Additionally, please remember to consider if this primary NeuroML
documentation here needs to be updated.

Please use [Semantic Line Breaks](https://sembr.org/){.reference
.external} wherever possible.
:::

::: {#testing .section}
[]{#devdocs-devsop-testing}

#### Testing[\#](#testing "Link to this heading"){.headerlink}

-   Before submitting a pull request, please run the various tests to
    confirm your changes. You can see how they are run in the various
    GitHub workflow files (in the [`.github/workflows/`{.docutils
    .literal .notranslate}]{.pre} folder in each repository). They will
    be run on all pull requests automatically so you can also verify
    your changes there.

-   For a new feature addition, please remember to include a unit test.

-   For a bug fix, please include a regression test.
:::
:::

[]{#document-Devdocs/ReleaseProcess}

::: {#release-process .section .tex2jax_ignore .mathjax_ignore}
[]{#devdocs-release}

### Release Process[\#](#release-process "Link to this heading"){.headerlink}

::: {#overview .section}
#### Overview[\#](#overview "Link to this heading"){.headerlink}

In general, work is carried out in the **development** branches of the
[main NeuroML
repositories](https://github.com/NeuroML/.github/blob/main/testsheet/README.md){.reference
.external} and these are merged to **master** branches on a new major
release, e.g. move from NeuroML v2.1 to v2.2.

A single page showing the **status of the automated test** as well as
any **open Pull Requests** on all of the core NeuroML repositories can
be found
[here](https://github.com/NeuroML/.github/blob/main/testsheet/README.md){.reference
.external}.
:::

::: {#steps-for-new-major-release .section}
#### Steps for new major release[\#](#steps-for-new-major-release "Link to this heading"){.headerlink}

These are the steps required for a new release of the NeuroML
development tools.

::: pst-scrollable-table-container
  Task                                                                                                                                                                                                                                                                                         Version this was last done
  -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- ----------------------------
  Commit final stable work in development branches                                                                                                                                                                                                                                             v2.3
  Make releases (not just tag - generates DOI) previous development versions of individual repos                                                                                                                                                                                               v2.3
  Increment all version numbers - to distinguish release from previous development version                                                                                                                                                                                                     v2.3
  Test all development branches - rerun GitHub Actions at least once                                                                                                                                                                                                                           v2.3
  Recheck all READMEs & docs                                                                                                                                                                                                                                                                   v2.3
  Run & check [test.py](https://github.com/NeuroML/NeuroML2/blob/master/test.py){.reference .external} in NeuroML2 repo                                                                                                                                                                        v2.3
  Check through issues for closed & easily closable ones                                                                                                                                                                                                                                       v2.3
  Update version in [[documentation pages]{.std .std-ref}](Landing.html#userdocs-neuromlv2){.reference .internal}                                                                                                                                                                              v2.3
  Update [HISTORY.md](https://github.com/NeuroML/NeuroML2/blob/master/HISTORY.md){.reference .external} in NeuroML2                                                                                                                                                                            v2.3
  pylems: Update README; Merge to master; Tag release; Release to pip                                                                                                                                                                                                                          v2.3
  libNeuroML: Update README; Retest; Merge to master; Tag release; Release to pip; Check [installation docs](https://libneuroml.readthedocs.org/en/latest/install.html){.reference .external}                                                                                                  v2.3
  pyNeuroML: Update Readme; Tag release; Release to pip                                                                                                                                                                                                                                        v2.3
  NeuroMLlite: Update Readme; Tag release; Release to pip                                                                                                                                                                                                                                      v2.3
  Java repositories ([[jNeuroML]{.std .std-ref}](Landing.html#jneuroml){.reference .internal}, org.neuroml.\* etc.): Merge development to master; Tag releases                                                                                                                                 v2.3
  Rebuild jNeuroML & commit to [jNeuroMLJar](https://sourceforge.net/p/neuroml/code/HEAD/tree/jNeuroMLJar/){.reference .external} and use latest for [jNeuroML for OMV](https://github.com/OpenSourceBrain/osb-model-validation/blob/master/omv/engines/getjnml.py#L8){.reference .external}   v2.3
  Add new binary release on [NeuroML/jNeuroML](https://github.com/NeuroML/jNeuroML/releases){.github .reference .external}                                                                                                                                                                     v2.3
  Update version used in [neuroConstruct](https://github.com/NeuralEnsemble/neuroConstruct){.reference .external}                                                                                                                                                                              v2.3
  Update docs on [http://docs.neuroml.org](https://docs.neuroml.org){.reference .external}                                                                                                                                                                                                     v2.3
  Update version on [COMBINE website](https://github.com/combine-org/combine-org.github.io/blob/master/content/authors/NeuroML/_index.md){.reference .external}                                                                                                                                v2.2
  ANNOUNCE (mailing list, Twitter)                                                                                                                                                                                                                                                             v2.2
  Increment version numbers in all development branches                                                                                                                                                                                                                                        v2.3
  DOI on [Zenodo](https://doi.org/10.5281/zenodo.593108){.reference .external}                                                                                                                                                                                                                 v2.3
  Update NeuroML [milestones](https://github.com/NeuroML/NeuroML2/milestones){.reference .external}                                                                                                                                                                                            v2.2
  New release of [neuroConstruct](https://github.com/NeuralEnsemble/neuroConstruct/releases){.reference .external}                                                                                                                                                                             v2.3
  Test toolchain on Windows...                                                                                                                                                                                                                                                                 v2.0
:::
:::
:::

[]{#document-Devdocs/UpdatingStandard}

::: {#making-changes-to-the-neuroml-standard .section .tex2jax_ignore .mathjax_ignore}
[]{#devdocs-updating-standard}

### Making changes to the NeuroML standard[\#](#making-changes-to-the-neuroml-standard "Link to this heading"){.headerlink}

The NeuroML standard is stored in two sets of files, each serving a
specific purpose:

-   the NeuroML [XML Schema
    Definition](https://en.wikipedia.org/wiki/XML_Schema_(W3C)){.reference
    .external} (XSD) file: this specifies the structure of a valid
    NeuroML XML file: what XML tags may be used and the how they are
    related

-   the NeuroML [LEMS](http://lems.github.io/LEMS/){.reference
    .external} ComponentType definition XML files: these include the
    definitions of the NeuroML standard ComponentTypes in LEMS
    constructs, which include the mathematical details underlying these
    ComponentTypes

These files are housed in the
[NeuroML](https://github.com/NeuroML/NeuroML2/){.reference .external}
repository.

The XSD schema file is used to validate NeuroML XML files, as shown in
the [[page on validating NeuroML files]{.std
.std-ref}](Landing.html#userdocs-validating-models){.reference
.internal}. Further, the NeuroML Python model in [[libNeuroML]{.std
.std-ref}](Landing.html#libneuroml){.reference .internal} is also
generated from the XSD file using the
[generateDS](http://www.davekuhlman.org/generateDS.html){.reference
.external} utility.

The LEMS ComponentType definition XML files are also used for a series
of additional validation tests, and since they include the details of
the underlying dynamics for all ComponentTypes, they are also used for
the simulation of NeuroML models either using the reference LEMS
interpreter, [[jLEMS]{.std .std-ref}](Landing.html#jlems){.reference
.internal}, or through automated code generation for supported
simulation platforms (via [[jNeuroML]{.std
.std-ref}](Landing.html#jneuroml){.reference .internal}). Additionally,
the LEMS definition files are also used the generate the [[human
readable schema documentation]{.std
.std-ref}](Landing.html#userdocs-neuromlv2){.reference .internal}
included in this documentation resource.

The two sets of files are therefore, tightly coupled. Any changes to the
XSD file must also be followed by corresponding changes to the LEMS
definition files.

::: {#procedure .section}
[]{#devdocs-updating-standard-proc}

#### Procedure[\#](#procedure "Link to this heading"){.headerlink}

::: {.admonition-pr-waiting .admonition}
PR waiting

TODO: A pull request to include the [`transfer_docs_to_xsd.py`{.docutils
.literal .notranslate}]{.pre} script in the repository is in review
here:
[NeuroML/NeuroML2#172](https://github.com/NeuroML/NeuroML2/pull/172){.github
.reference .external}
:::

The suggested way of making changes to these files is via pull requests
to the NeuroML repository which will undergo review by the NeuroML
editorial board and the development team. As noted in the [[general
contribution guidelines]{.std
.std-ref}](Landing.html#devdocs-devsop){.reference .internal}, the
[`development`{.docutils .literal .notranslate}]{.pre} branch tracks the
next release of the NeuroML standard. So, all pull request must be made
against the [`development`{.docutils .literal .notranslate}]{.pre}
branch.

-   New ComponentTypes, and their elements (parameters, variables etc.)
    that are added in the LEMS definition XML files should be properly
    documented.

-   After both sets of files have been modified, please run the
    [`transfer_docs_to_xsd.py`{.docutils .literal .notranslate}]{.pre}
    script in the [`scripts`{.docutils .literal .notranslate}]{.pre}
    folder to copy documentation over from the XML files to the XSD
    schema file. This script will also run basic sanity checks to ensure
    that all ComponentTypes in the LEMS XML definition files are
    represented in the XSD schema file and vice-versa.

-   Please run [`xmllint`{.docutils .literal .notranslate}]{.pre} on the
    files to ensure they are formatted correctly.

-   Please make individual commits for changes to the XSD file, and the
    XML files. This ensures that their change history is clearly
    maintained.
:::

::: {#regenerating-schema-documentation .section}
[]{#devdocs-updating-standard-schema-docs}

#### Regenerating schema documentation[\#](#regenerating-schema-documentation "Link to this heading"){.headerlink}

Once the pull request has been merged in the NeuroML repository, the
[[human readable schema documentation included in this documentation
resource]{.std .std-ref}](Landing.html#userdocs-neuromlv2){.reference
.internal} must be updated. This is done by running the
[generate-jupyter-ast.py](https://github.com/NeuroML/Documentation/blob/master/scripts/schemas/generate-jupyter-ast.py){.reference
.external} script included in the [documentation source
repository](https://github.com/NeuroML/Documentation){.reference
.external}. This will read the LEMS XML definition files and regenerate
the corresponding documentation pages. A pull request can then be opened
with the updated pages.
:::

::: {#updating-the-java-api-org-neuroml-model .section}
[]{#devdocs-updating-standard-org-neuroml-model}

#### Updating the Java API: org.neuroml.model[\#](#updating-the-java-api-org-neuroml-model "Link to this heading"){.headerlink}

TODO: Document what needs to be done for
[NeuroML/org.neuroml.model](https://github.com/NeuroML/org.neuroml.model){.github
.reference .external}
:::

::: {#updating-the-python-api-libneuroml .section}
[]{#devdocs-updating-standard-libneuroml}

#### Updating the Python API: libNeuroML[\#](#updating-the-python-api-libneuroml "Link to this heading"){.headerlink}

::: {.admonition-pr-waiting .admonition}
PR waiting

TODO: A pull request to include the [`regenerate-nml.sh`{.docutils
.literal .notranslate}]{.pre} script in the repository is in review
here:
[NeuralEnsemble/libNeuroML#110](https://github.com/NeuralEnsemble/libNeuroML/pull/110){.github
.reference .external}
:::

Any changes to the XSD schema file require regeneration of the [Python
object model in
libNeuroML](https://github.com/NeuralEnsemble/libNeuroML/blob/development/neuroml/nml/nml.py){.reference
.external}:

-   copy over the updated XSD schema file to the
    [`neuroml/nml/`{.docutils .literal .notranslate}]{.pre} directory in
    the [`development`{.docutils .literal .notranslate}]{.pre} branch

-   commit the new XSD file

-   run the [`regenerate-nml.sh`{.docutils .literal .notranslate}]{.pre}
    script to regenerate and reformat [`nml.py`{.docutils .literal
    .notranslate}]{.pre}

-   build and install libNeuroML into a new virtual environment

-   run all tests using [`pytest`{.docutils .literal
    .notranslate}]{.pre}

-   run all examples and ensure that they run correctly (please see the
    [GitHub actions
    workflow](https://github.com/NeuralEnsemble/libNeuroML/blob/master/.github/workflows/ci.yml#L44){.reference
    .external} for more information)

-   if all checks pass successfully, a pull request can be opened
:::

::: {#updating-the-c-api .section}
[]{#devdocs-updating-standard-c-api}

#### Updating the C++ API[\#](#updating-the-c-api "Link to this heading"){.headerlink}

TODO: Document what needs to be done for
[NeuroML/NeuroML_API](https://github.com/NeuroML/NeuroML_API/){.github
.reference .external}
:::
:::
:::
:::

[]{#document-Devdocs/InteractionOtherBits}

::: {#interaction-with-other-languages-and-standards .section .tex2jax_ignore .mathjax_ignore}
[]{#devdocs-interaction}

## Interaction with other languages and standards[\#](#interaction-with-other-languages-and-standards "Link to this heading"){.headerlink}

::: {.admonition-needs-work .admonition}
Needs work

TODO: Add more information to each of these
:::

::: {#pynn .section}
[]{#devdocs-interaction-pynn}

### PyNN[\#](#pynn "Link to this heading"){.headerlink}

[NeuroML/NeuroML2#73](https://github.com/NeuroML/NeuroML2/issues/73){.github
.reference .external}
:::

::: {#sbml .section}
[]{#devdocs-interaction-sbml}

### SBML[\#](#sbml "Link to this heading"){.headerlink}

[OpenSourceBrain/SBMLShowcase](https://github.com/OpenSourceBrain/SBMLShowcase){.github
.reference .external}
:::

::: {#sonata .section}
[]{#devdocs-interaction-sonata}

### Sonata[\#](#sonata "Link to this heading"){.headerlink}

[https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1007696](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1007696){.reference
.external}
:::

::: {#nineml-spineml .section}
[]{#devdocs-interaction-nineml}

### NineML & SpineML[\#](#nineml-spineml "Link to this heading"){.headerlink}

[OpenSourceBrain/NineMLShowcase](https://github.com/OpenSourceBrain/NineMLShowcase){.github
.reference .external}
:::

::: {#modeci-mdf .section}
[]{#devdocs-interaction-mdf}

### ModECI MDF[\#](#modeci-mdf "Link to this heading"){.headerlink}

[http://www.modeci.org/](http://www.modeci.org/){.reference .external}
:::

::: {#swc .section}
[]{#devdocs-interaction-swc}

### SWC[\#](#swc "Link to this heading"){.headerlink}

[http://www.neuronland.org/NLMorphologyConverter/MorphologyFormats/SWC/Spec.html](http://www.neuronland.org/NLMorphologyConverter/MorphologyFormats/SWC/Spec.html){.reference
.external}
[http://www.neuromorpho.org/myfaq.jsp](http://www.neuromorpho.org/myfaq.jsp){.reference
.external}
:::
:::
:::

::: {.toctree-wrapper .compound}
[]{#document-Reference/Glossary}

::: {#glossary .section .tex2jax_ignore .mathjax_ignore}
## Glossary[\#](#glossary "Link to this heading"){.headerlink}

-   XML: Extensible Markup Language (XML) is a markup language that
    defines a set of rules for encoding documents in a format that is
    both human-readable and machine-readable. (Read full entry on
    [Wikipedia](https://en.wikipedia.org/wiki/XML){.reference
    .external})
:::

[]{#document-Reference/zBibliography}

::: {#bibliography .section .tex2jax_ignore .mathjax_ignore}
## Bibliography[\#](#bibliography "Link to this heading"){.headerlink}

::: {#id1 .docutils .container}
::: {.citation-list role="list"}
::: {#id15 .citation role="doc-biblioentry"}
[[\[]{.fn-bracket}BRB+16[\]]{.fn-bracket}]{.label}

Marianne J Bezaire, Ivan Raikov, Kelly Burk, Dhrumil Vyas, and Ivan
Soltesz. Interneuronal mechanisms of hippocampal theta oscillations in a
full-scale model of the rodent ca1 circuit. *eLife*, 5:e18566, dec 2016.
URL:
[https://doi.org/10.7554/eLife.18566](https://doi.org/10.7554/eLife.18566){.reference
.external},
[doi:10.7554/eLife.18566](https://doi.org/10.7554/eLife.18566){.reference
.external}.
:::

::: {#id23 .citation role="doc-biblioentry"}
[[\[]{.fn-bracket}BHG+23[\]]{.fn-bracket}]{.label}

Justas Birgiolas, Vergil Haynes, Padraig Gleeson, Richard C. Gerkin,
Suzanne W. Dietrich, and Sharon Crook. Neuroml-db: sharing and
characterizing data-driven neuroscience models described in neuroml.
*PLOS Computational Biology*, 19(3):1--29, 2023. URL:
[https://doi.org/10.1371/journal.pcbi.1010941](https://doi.org/10.1371/journal.pcbi.1010941){.reference
.external},
[doi:10.1371/journal.pcbi.1010941](https://doi.org/10.1371/journal.pcbi.1010941){.reference
.external}.
:::

::: {#id13 .citation role="doc-biblioentry"}
[[\[]{.fn-bracket}BBC+18[\]]{.fn-bracket}]{.label}

Inga Blundell, Romain Brette, Thomas A. Cleland, Thomas G. Close, Daniel
Coca, Andrew P. Davison, Sandra Diaz-Pier, Carlos Fernandez Musoles,
Padraig Gleeson, Dan F. M. Goodman, Michael Hines, Michael W. Hopkins,
Pramod Kumbhar, David R. Lester, Bóris Marin, Abigail Morrison, Eric
Müller, Thomas Nowotny, Alexander Peyser, Dimitri Plotnikov, Paul
Richmond, Andrew Rowley, Bernhard Rumpe, Marcel Stimberg, Alan B.
Stokes, Adam Tomkins, Guido Trensch, Marmaduke Woodman, and
Jochen Martin Eppler. Code generation in computational neuroscience: a
review of tools and techniques. *Frontiers in Neuroinformatics*, 12:68,
2018.
[doi:10.3389/fninf.2018.00068](https://doi.org/10.3389/fninf.2018.00068){.reference
.external}.
:::

::: {#id10 .citation role="doc-biblioentry"}
[[\[]{.fn-bracket}CGC+14[\]]{.fn-bracket}]{.label}

Robert C. Cannon, Padraig Gleeson, Sharon Crook, Gautham Ganapathy,
Boris Marin, Eugenio Piasini, and R. Angus Silver. LEMS: a language for
expressing complex biological models in concise and hierarchical form
and its use in underpinning NeuroML 2. *Frontiers in Neuroinformatics*,
2014.
[doi:10.3389/fninf.2014.00079](https://doi.org/10.3389/fninf.2014.00079){.reference
.external}.
:::

::: {#id8 .citation role="doc-biblioentry"}
[[\[]{.fn-bracket}CGH+07[\]]{.fn-bracket}]{.label}

Sharon Crook, Padraig Gleeson, Fred Howell, Joseph Svitak, and R. Angus
Silver. MorphML: level 1 of the NeuroML standards for neuronal
morphology data and model specification. *Neuroinformatics*,
5(2):96--104, 5 2007.
[doi:10.1007/s12021-007-0003-6](https://doi.org/10.1007/s12021-007-0003-6){.reference
.external}.
:::

::: {#id17 .citation role="doc-biblioentry"}
[[\[]{.fn-bracket}FHA+15[\]]{.fn-bracket}]{.label}

KA Ferguson, CYL Huh, B Amilhon, S Williams, and FK Skinner. Data set of
CA1 pyramidal cell recordings using an intact whole hippocampus
preparation, including recordings of rebound firing (V2). May 2015. URL:
[https://doi.org/10.5281/zenodo.17794](https://doi.org/10.5281/zenodo.17794){.reference
.external},
[doi:10.5281/zenodo.17794](https://doi.org/10.5281/zenodo.17794){.reference
.external}.
:::

::: {#id4 .citation role="doc-biblioentry"}
[[\[]{.fn-bracket}GKA+01[\]]{.fn-bracket}]{.label}

D. Gardner, K. H. Knuth, M. Abato, S. M. Erde, T. White, R. DeBellis,
and E. P. Gardner. Common data model for neuroscience data and data
model exchange. *Journal of the American Medical Informatics
Association*, 8(1):17--33, 1 2001.
[doi:10.1136/jamia.2001.0080017](https://doi.org/10.1136/jamia.2001.0080017){.reference
.external}.
:::

::: {#id12 .citation role="doc-biblioentry"}
[[\[]{.fn-bracket}GCM+19[\]]{.fn-bracket}]{.label}

Padraig Gleeson, Matteo Cantarelli, Boris Marin, Adrian Quintana, Matt
Earnshaw, Sadra Sadeh, Eugenio Piasini, Justas Birgiolas, Robert C.
Cannon, N. Alex Cayco-Gajic, Sharon Crook, Andrew P. Davison, Salvador
Dura-Bernal, András Ecker, Michael L. Hines, Giovanni Idili, Frederic
Lanore, Stephen D. Larson, William W. Lytton, Amitava Majumdar,
Robert A. McDougal, Subhashini Sivagnanam, Sergio Solinas, Rokas
Stanislovas, Sacha J. van Albada, Werner van Geit, and R. Angus Silver.
Open source brain: a collaborative resource for visualizing, analyzing,
simulating, and developing standardized models of neurons and circuits.
*Neuron*, 103(3):395--411, 2019.
[doi:10.1016/j.neuron.2019.05.019](https://doi.org/10.1016/j.neuron.2019.05.019){.reference
.external}.
:::

::: {#id9 .citation role="doc-biblioentry"}
[[\[]{.fn-bracket}GCC+10[\]]{.fn-bracket}]{.label}

Padraig Gleeson, Sharon Crook, Robert C. Cannon, Michael L. Hines,
Guy O. Billings, Matteo Farinella, Thomas M. Morse, Andrew P. Davison,
Subhasis Ray, Upinder S. Bhalla, Simon R. Barnes, Yoana D. Dimitrova,
and R. Angus Silver. NeuroML: a language for describing data driven
models of neurons and networks with a high degree of biological detail.
*PLoS Computational Biology*, 6(6):e1000815, 2010.
[doi:10.1371/journal.pcbi.1000815](https://doi.org/10.1371/journal.pcbi.1000815){.reference
.external}.
:::

::: {#id7 .citation role="doc-biblioentry"}
[[\[]{.fn-bracket}GSS07[\]]{.fn-bracket}]{.label}

Padraig Gleeson, Volker Steuber, and R. Angus Silver. neuroConstruct: a
tool for modeling networks of neurons in 3d space. *Neuron*,
54(2):219--235, 4 2007.
[doi:10.1016/j.neuron.2007.03.025](https://doi.org/10.1016/j.neuron.2007.03.025){.reference
.external}.
:::

::: {#id3 .citation role="doc-biblioentry"}
[[\[]{.fn-bracket}GHH+01[\]]{.fn-bracket}]{.label}

Nigel H. Goddard, Michael Hucka, Fred Howell, Hugo Cornelis, Kavita
Shankar, and David Beeman. Towards NeuroML: model description methods
for collaborative modelling in neuroscience. *Philosophical Transactions
of the Royal Society of London. Series B: Biological Sciences*,
356(1412):1209--1228, 8 2001.
[doi:10.1098/rstb.2001.0910](https://doi.org/10.1098/rstb.2001.0910){.reference
.external}.
:::

::: {#id14 .citation role="doc-biblioentry"}
[[\[]{.fn-bracket}HH52[\]]{.fn-bracket}]{.label}

Alan L. Hodgkin and Andrew F. Huxley. A quantitative description of
membrane current and its application to conduction and excitation in
nerve. *The Journal of physiology*, 117(4):500, 1952.
:::

::: {#id5 .citation role="doc-biblioentry"}
[[\[]{.fn-bracket}HCG+03[\]]{.fn-bracket}]{.label}

F. Howell, R. Cannon, N. Goddard, H. Bringmann, P. Rogister, and
H. Cornelis. Linking computational neuroscience simulation tools---a
pragmatic approach to component-based development. *Neurocomputing*,
52-54:289--294, 6 2003.
[doi:10.1016/s0925-2312(02)00781-6](https://doi.org/10.1016/s0925-2312(02)00781-6){.reference
.external}.
:::

::: {#id2 .citation role="doc-biblioentry"}
[[\[]{.fn-bracket}Izh07[\]]{.fn-bracket}]{.label}

Eugene M. Izhikevich. *Dynamical systems in neuroscience*. MIT Press,
2007.
:::

::: {#id16 .citation role="doc-biblioentry"}
[[\[]{.fn-bracket}Lor63[\]]{.fn-bracket}]{.label}

Edward N. Lorenz. Deterministic nonperiodic flow. *Journal of
Atmospheric Sciences*, 20(2):130--141, 1963.
[doi:10.1175/1520-0469(1963)020\<0130:dnf\>2.0.co;2](https://doi.org/10.1175/1520-0469(1963)020%3C0130:dnf%3E2.0.co;2){.reference
.external}.
:::

::: {#id19 .citation role="doc-biblioentry"}
[[\[]{.fn-bracket}PBM04[\]]{.fn-bracket}]{.label}

Astrid A. Prinz, Dirk Bucher, and Eve Marder. Similar network activity
from disparate circuit parameters. *Nature Neuroscience*,
7(12):1345--1352, 2004.
[doi:10.1038/nn1352](https://doi.org/10.1038/nn1352){.reference
.external}.
:::

::: {#id6 .citation role="doc-biblioentry"}
[[\[]{.fn-bracket}QC04[\]]{.fn-bracket}]{.label}

Weihong Qi and Sharon Crook. Tools for neuroinformatic data exchange: an
XML application for neuronal morphology data. *Neurocomputing*,
58-60:1091--1095, 6 2004.
[doi:10.1016/j.neucom.2004.01.171](https://doi.org/10.1016/j.neucom.2004.01.171){.reference
.external}.
:::

::: {#id21 .citation role="doc-biblioentry"}
[[\[]{.fn-bracket}RAS20[\]]{.fn-bracket}]{.label}

Subhasis Ray, Zane N. Aldworth, and Mark A. Stopfer. Feedback inhibition
and its control in an insect olfactory circuit. *eLife*, 9:e53281, 2020.
URL:
[https://doi.org/10.7554/eLife.53281](https://doi.org/10.7554/eLife.53281){.reference
.external},
[doi:10.7554/eLife.53281](https://doi.org/10.7554/eLife.53281){.reference
.external}.
:::

::: {#id18 .citation role="doc-biblioentry"}
[[\[]{.fn-bracket}RGF+11[\]]{.fn-bracket}]{.label}

Cyrille Rossant, Dan F. Goodman, Bertrand Fontaine, Jonathan
Platkiewicz, Anna Magnusson, and Romain Brette. Fitting neuron models to
spike trains. *Frontiers in Neuroscience*, 5:9, 2011. URL:
[https://www.frontiersin.org/article/10.3389/fnins.2011.00009](https://www.frontiersin.org/article/10.3389/fnins.2011.00009){.reference
.external},
[doi:10.3389/fnins.2011.00009](https://doi.org/10.3389/fnins.2011.00009){.reference
.external}.
:::

::: {#id22 .citation role="doc-biblioentry"}
[[\[]{.fn-bracket}SGM+25[\]]{.fn-bracket}]{.label}

Ankur Sinha, Padraig Gleeson, Bóris Marin, Salvador Dura-Bernal,
Sotirios Panagiotou, Sharon Crook, Matteo Cantarelli, Robert C. Cannon,
Andrew P. Davison, Harsha Gurnani, and Robin Angus Silver. The neuroml
ecosystem for standardized multi-scale modeling in neuroscience.
*eLife*, 13:RP95135, 2025. URL:
[https://doi.org/10.7554/eLife.95135](https://doi.org/10.7554/eLife.95135){.reference
.external},
[doi:10.7554/eLife.95135](https://doi.org/10.7554/eLife.95135){.reference
.external}.
:::

::: {#id20 .citation role="doc-biblioentry"}
[[\[]{.fn-bracket}SKNG16[\]]{.fn-bracket}]{.label}

Arfon M. Smith, Daniel S. Katz, Kyle E. Niemeyer, and FORCE11 Software
Citation Working Group. Software citation principles. *PeerJ Computer
Science*, 2:e86, 2016. URL:
[https://doi.org/10.7717/peerj-cs.86](https://doi.org/10.7717/peerj-cs.86){.reference
.external},
[doi:10.7717/peerj-cs.86](https://doi.org/10.7717/peerj-cs.86){.reference
.external}.
:::

::: {#id11 .citation role="doc-biblioentry"}
[[\[]{.fn-bracket}VCC+14[\]]{.fn-bracket}]{.label}

Michael Vella, Robert C. Cannon, Sharon Crook, Andrew P. Davison,
Gautham Ganapathy, Hugh P. C. Robinson, R. Angus Silver, and Padraig
Gleeson. Libneuroml and pylems: using python to combine procedural and
declarative modeling approaches in computational neuroscience.
*Frontiers in neuroinformatics*, 8:38, 2014.
[doi:10.3389/fninf.2014.00038](https://doi.org/10.3389/fninf.2014.00038){.reference
.external}.
:::
:::
:::
:::
:::
:::

::: prev-next-area
:::
:::

::: {.bd-sidebar-secondary .bd-toc}
::: {.sidebar-secondary-items .sidebar-secondary__inner}
::: sidebar-secondary-item
::: {.page-toc .tocsection .onthispage}
Contents
:::

[User documentation]{.caption-text}

-   [Mission and Aims](#document-Userdocs/MissionAndAims){.reference
    .internal .nav-link}
-   [How to use this documentation](#document-Userdocs/Usage){.reference
    .internal .nav-link}
-   [Get NeuroML](#document-Userdocs/GetNeuroML){.reference .internal
    .nav-link}
-   [Getting started with
    NeuroML](#document-Userdocs/GettingStarted){.reference .internal
    .nav-link}
    -   [Simulating a regular spiking Izhikevich
        neuron](#document-Userdocs/SingleNeuronExample){.reference
        .internal .nav-link}
    -   [Interactive single Izhikevich neuron NeuroML
        example](#document-Userdocs/NML2_examples/SingleNeuron){.reference
        .internal .nav-link}
    -   [A two population network of regular spiking Izhikevich
        neurons](#document-Userdocs/IzhikevichNetworkExample){.reference
        .internal .nav-link}
    -   [Interactive two population network
        example](#document-Userdocs/NML2_examples/IzhikevichNetwork){.reference
        .internal .nav-link}
    -   [Simulating a single compartment Hodgkin-Huxley
        neuron](#document-Userdocs/SingleCompartmentHHExample){.reference
        .internal .nav-link}
    -   [Interactive single compartment HH
        example](#document-Userdocs/NML2_examples/HH_single_compartment){.reference
        .internal .nav-link}
    -   [Simulating a multi compartment OLM
        neuron](#document-Userdocs/MultiCompartmentOLMexample){.reference
        .internal .nav-link}
    -   [Interactive multi-compartment OLM cell
        example](#document-Userdocs/NML2_examples/OLM){.reference
        .internal .nav-link}
    -   [Create novel NeuroML models from components on
        NeuroML-DB](#document-Userdocs/NML2_examples/NeuroML-DB){.reference
        .internal .nav-link}
-   [Finding and sharing NeuroML
    models](#document-Userdocs/FindingNeuroMLModels){.reference
    .internal .nav-link}
-   [Creating NeuroML
    models](#document-Userdocs/CreatingNeuroMLModels){.reference
    .internal .nav-link}
    -   [Converting models to NeuroML and sharing them on Open Source
        Brain](#document-Userdocs/ConvertingModels){.reference .internal
        .nav-link}
    -   [Handling Morphology
        Files](#document-Userdocs/ImportingMorphologyFiles){.reference
        .internal .nav-link}
    -   [HDF5 support](#document-Userdocs/HDF5){.reference .internal
        .nav-link}
    -   [Maintaining provenance in NeuroML
        models](#document-Userdocs/Provenance){.reference .internal
        .nav-link}
-   [Validating NeuroML
    Models](#document-Userdocs/ValidatingNeuroMLModels){.reference
    .internal .nav-link}
-   [Visualising NeuroML
    Models](#document-Userdocs/VisualisingNeuroMLModels){.reference
    .internal .nav-link}
    -   [Visualising and analysing ion channel
        models](#document-Userdocs/VisualisingChannels){.reference
        .internal .nav-link}
    -   [Visualising and analysing cell
        models](#document-Userdocs/VisualisingCells){.reference
        .internal .nav-link}
-   [Simulating NeuroML
    Models](#document-Userdocs/SimulatingNeuroMLModels){.reference
    .internal .nav-link}
-   [Optimising/fitting NeuroML
    Models](#document-Userdocs/OptimisingNeuroMLModels){.reference
    .internal .nav-link}
-   [Testing/validating NeuroML
    Models](#document-Userdocs/TestingNeuroMLModels){.reference
    .internal .nav-link}
-   [LEMS: Low Entropy Model
    Specification](#document-Userdocs/LEMS){.reference .internal
    .nav-link}
    -   [Model structure
        overview](#document-Userdocs/LEMSOverview){.reference .internal
        .nav-link}
    -   [Example 1: Dimensions, Units, ComponentTypes and
        Components](#document-Userdocs/LEMSExample1){.reference
        .internal .nav-link}
    -   [Example 2: tidying up example
        1](#document-Userdocs/LEMSExample2){.reference .internal
        .nav-link}
    -   [Main model](Landing.html#main-model){.reference .internal
        .nav-link}
    -   [Included files](Landing.html#included-files){.reference
        .internal .nav-link}
    -   [Example 3: Connection dependent synaptic
        components](#document-Userdocs/LEMSExample3){.reference
        .internal .nav-link}
    -   [Example 4: Kinetic
        schemes](#document-Userdocs/LEMSExample4){.reference .internal
        .nav-link}
    -   [Example 5: References and
        paths](#document-Userdocs/LEMSExample5){.reference .internal
        .nav-link}
    -   [Example 6: User defined types for simulation and
        display](#document-Userdocs/LEMSExample6){.reference .internal
        .nav-link}
    -   [Example 7: User defined types for networks and
        populations](#document-Userdocs/LEMSExample7){.reference
        .internal .nav-link}
    -   [Example 8: Regimes in Dynamics
        definitions](#document-Userdocs/LEMSExample8){.reference
        .internal .nav-link}
-   [Schema/Specification](#document-Userdocs/Specification){.reference
    .internal .nav-link}
    -   [NeuroML v2](#document-Userdocs/NeuroMLv2){.reference .internal
        .nav-link}
        -   [NeuroMLDocument](#document-Userdocs/Schemas/NeuroMLDocument){.reference
            .internal .nav-link}
        -   [NeuroMLCoreDimensions](#document-Userdocs/Schemas/NeuroMLCoreDimensions){.reference
            .internal .nav-link}
        -   [NeuroMLCoreCompTypes](#document-Userdocs/Schemas/NeuroMLCoreCompTypes){.reference
            .internal .nav-link}
        -   [Cells](#document-Userdocs/Schemas/Cells){.reference
            .internal .nav-link}
        -   [Channels](#document-Userdocs/Schemas/Channels){.reference
            .internal .nav-link}
        -   [Synapses](#document-Userdocs/Schemas/Synapses){.reference
            .internal .nav-link}
        -   [Inputs](#document-Userdocs/Schemas/Inputs){.reference
            .internal .nav-link}
        -   [Networks](#document-Userdocs/Schemas/Networks){.reference
            .internal .nav-link}
        -   [PyNN](#document-Userdocs/Schemas/PyNN){.reference .internal
            .nav-link}
        -   [Simulation](#document-Userdocs/Schemas/Simulation){.reference
            .internal .nav-link}
        -   [Index](#document-Userdocs/Schemas/Index){.reference
            .internal .nav-link}
    -   [NeuroML v1](#document-Userdocs/NeuroMLv1){.reference .internal
        .nav-link}
    -   [LEMS](#document-Userdocs/LEMSSchema){.reference .internal
        .nav-link}
        -   [Model
            structure](#document-Userdocs/LEMS_elements/Modelstructure){.reference
            .internal .nav-link}
        -   [Units and
            dimensions](#document-Userdocs/LEMS_elements/Unitsanddimensions){.reference
            .internal .nav-link}
        -   [Defining component
            types](#document-Userdocs/LEMS_elements/Definingcomponenttypes){.reference
            .internal .nav-link}
        -   [Dynamics](#document-Userdocs/LEMS_elements/Dynamics){.reference
            .internal .nav-link}
        -   [Structure](#document-Userdocs/LEMS_elements/Structure){.reference
            .internal .nav-link}
        -   [Simulation](#document-Userdocs/LEMS_elements/Simulation){.reference
            .internal .nav-link}
        -   [Procedure](#document-Userdocs/LEMS_elements/Procedure){.reference
            .internal .nav-link}
        -   [Defining
            Components](#document-Userdocs/LEMS_elements/DefiningComponents){.reference
            .internal .nav-link}
        -   [Geometry](#document-Userdocs/LEMS_elements/Geometry){.reference
            .internal .nav-link}
-   [NeuroML 2 and LEMS](#document-Userdocs/NeuroMLv2AndLEMS){.reference
    .internal .nav-link}
    -   [Conventions](#document-Userdocs/Conventions){.reference
        .internal .nav-link}
    -   [Units and
        dimensions](#document-Userdocs/UnitsAndDimensions){.reference
        .internal .nav-link}
    -   [Paths](#document-Userdocs/Paths){.reference .internal
        .nav-link}
    -   [Quantities and
        recording](#document-Userdocs/QuantitiesAndRecording){.reference
        .internal .nav-link}
    -   [LEMS Simulation
        files](#document-Userdocs/LEMSSimulation){.reference .internal
        .nav-link}
-   [Extending
    NeuroML](#document-Userdocs/ExtendingNeuroMLv2){.reference .internal
    .nav-link}
-   [Software and
    Tools](#document-Userdocs/Software/Software){.reference .internal
    .nav-link}
    -   [pyNeuroML](#document-Userdocs/Software/pyNeuroML){.reference
        .internal .nav-link}
    -   [libNeuroML](#document-Userdocs/Software/libNeuroML){.reference
        .internal .nav-link}
    -   [pyLEMS](#document-Userdocs/Software/pyLEMS){.reference
        .internal .nav-link}
    -   [NeuroMLlite](#document-Userdocs/Software/NeuroMLlite){.reference
        .internal .nav-link}
    -   [jNeuroML](#document-Userdocs/Software/jNeuroML){.reference
        .internal .nav-link}
    -   [jLEMS](#document-Userdocs/Software/jLEMS){.reference .internal
        .nav-link}
    -   [NeuroML C++
        API](#document-Userdocs/Software/NeuroML_API){.reference
        .internal .nav-link}
    -   [MatLab NeuroML
        Toolbox](#document-Userdocs/Software/MatLab){.reference
        .internal .nav-link}
    -   [Tools and resources with NeuroML
        support](#document-Userdocs/Software/SupportingTools){.reference
        .internal .nav-link}
        -   [Approaches to adding NeuroML
            support](#document-Userdocs/Software/Tools/Approaches){.reference
            .internal .nav-link}
        -   [NEURON and
            NeuroML](#document-Userdocs/Software/Tools/NEURON){.reference
            .internal .nav-link}
        -   [NetPyNE and
            NeuroML](#document-Userdocs/Software/Tools/NetPyNE){.reference
            .internal .nav-link}
        -   [PyNN and
            NeuroML](#document-Userdocs/Software/Tools/PyNN){.reference
            .internal .nav-link}
        -   [Brian and
            NeuroML](#document-Userdocs/Software/Tools/Brian){.reference
            .internal .nav-link}
        -   [MOOSE and
            NeuroML](#document-Userdocs/Software/Tools/MOOSE){.reference
            .internal .nav-link}
        -   [EDEN and
            NeuroML](#document-Userdocs/Software/Tools/EDEN){.reference
            .internal .nav-link}
        -   [Arbor and
            NeuroML](#document-Userdocs/Software/Tools/Arbor){.reference
            .internal .nav-link}
        -   [N2A and
            NeuroML](#document-Userdocs/Software/Tools/N2A){.reference
            .internal .nav-link}
        -   [NEST and
            NeuroML](#document-Userdocs/Software/Tools/NEST){.reference
            .internal .nav-link}
        -   [SWC and
            NeuroML](#document-Userdocs/Software/Tools/SWC){.reference
            .internal .nav-link}
-   [Citing NeuroML and related
    publications](#document-Userdocs/Publications){.reference .internal
    .nav-link}
-   [Frequently asked questions
    (FAQ)](#document-Userdocs/FAQ){.reference .internal .nav-link}
-   [Walk
    throughs](#document-Userdocs/Walkthroughs/Walkthroughs){.reference
    .internal .nav-link}
    -   [Converting Ray et al 2020 to
        NeuroML](#document-Userdocs/Walkthroughs/RayEtAl2020/RayEtAl2020){.reference
        .internal .nav-link}
        -   [Setting
            up](#document-Userdocs/Walkthroughs/RayEtAl2020/Setup){.reference
            .internal .nav-link}
        -   [Converting to
            NeuroML](#document-Userdocs/Walkthroughs/RayEtAl2020/Conversion){.reference
            .internal .nav-link}
        -   [Adding OMV
            tests](#document-Userdocs/Walkthroughs/RayEtAl2020/OMV){.reference
            .internal .nav-link}

[NeuroML events]{.caption-text}

-   [NeuroML outreach and events](#document-Events/Events){.reference
    .internal .nav-link}
-   [July 2024: NeuroML tutorial at CNS
    2024](#document-Events/202407-CNS2024){.reference .internal
    .nav-link}
-   [April 2024: NeuroML hackathon at HARMONY
    2024](#document-Events/202404-Harmony){.reference .internal
    .nav-link}
-   [June 2022: NeuroML tutorial at CNS\*2022 satellite
    tutorials](#document-Events/20220630-CNS2022){.reference .internal
    .nav-link}
-   [April 2022: NeuroML development workshop at HARMONY
    2022](#document-Events/202204-Harmony){.reference .internal
    .nav-link}
-   [October 2021: NeuroML development workshop at COMBINE
    meeting](#document-Events/202109-COMBINE){.reference .internal
    .nav-link}
-   [August 2021: NeuroML tutorial at INCF Training
    Weeks](#document-Events/202108-INCF-Training-Week){.reference
    .internal .nav-link}
-   [July 2021: NeuroML tutorial at
    CNS\*2021](#document-Events/202107-CNS2021){.reference .internal
    .nav-link}
-   [March 2021: NeuroML hackathon at HARMONY
    2021](#document-Events/202103-Harmony){.reference .internal
    .nav-link}
-   [March 2012: Fourth NeuroML Development
    Workshop](#document-Events/2012-Edinburgh){.reference .internal
    .nav-link}
-   [Past NeuroML Events](#document-Events/PastEvents){.reference
    .internal .nav-link}

[The NeuroML Initiative]{.caption-text}

-   [Getting in
    touch](#document-NeuroMLOrg/CommunicationChannels){.reference
    .internal .nav-link}
-   [Overview of standards in
    neuroscience](#document-NeuroMLOrg/Standards){.reference .internal
    .nav-link}
-   [A brief history of
    NeuroML](#document-NeuroMLOrg/History){.reference .internal
    .nav-link}
-   [NeuroML Editorial Board](#document-NeuroMLOrg/Board){.reference
    .internal .nav-link}
    -   [History of the NeuroML Editorial
        Board](#document-NeuroMLOrg/BoardHistory){.reference .internal
        .nav-link}
    -   [Workshop and Meeting
        reports](#document-NeuroMLOrg/BoardMeetingReports){.reference
        .internal .nav-link}
-   [NeuroML Scientific
    Committee](#document-NeuroMLOrg/ScientificCommittee){.reference
    .internal .nav-link}
-   [Funding and
    Acknowledgements](#document-NeuroMLOrg/Funding){.reference .internal
    .nav-link}
-   [Outreach and
    training](#document-NeuroMLOrg/OutreachTraining){.reference
    .internal .nav-link}
-   [NeuroML contributors](#document-NeuroMLOrg/Contributors){.reference
    .internal .nav-link}
-   [NeuroML repositories](#document-NeuroMLOrg/Repositories){.reference
    .internal .nav-link}
-   [Code of Conduct](#document-NeuroMLOrg/CoC){.reference .internal
    .nav-link}

[Developer documentation]{.caption-text}

-   [Overview](#document-Devdocs/Devdocs){.reference .internal
    .nav-link}
    -   [Contribution guidelines](#document-Devdocs/DevSOP){.reference
        .internal .nav-link}
    -   [Release Process](#document-Devdocs/ReleaseProcess){.reference
        .internal .nav-link}
    -   [Making changes to the NeuroML
        standard](#document-Devdocs/UpdatingStandard){.reference
        .internal .nav-link}
-   [Interaction with other languages and
    standards](#document-Devdocs/InteractionOtherBits){.reference
    .internal .nav-link}

[Reference]{.caption-text}

-   [Glossary](#document-Reference/Glossary){.reference .internal
    .nav-link}
-   [Bibliography](#document-Reference/zBibliography){.reference
    .internal .nav-link}
:::
:::
:::
:::

::: {.bd-footer-content__inner .container}
::: footer-item
By NeuroML contributors
:::

::: footer-item
© Copyright 2025.\
:::

::: footer-item
:::

::: footer-item
:::
:::
:::
:::
:::
